commit f61f6a81d4c9b0d25d33a0312fb8eba3ddad8205 (refs/changes/02/4502/5)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2018-07-17T18:13:00-04:00 (1 year, 3 months ago)
    
    OS-7069 zone hang during shutdown after interface address conflict

diff --git a/usr/src/uts/common/inet/ip/ip_if.c b/usr/src/uts/common/inet/ip/ip_if.c
index c4b60daa68..f5ac4b30c6 100644
--- a/usr/src/uts/common/inet/ip/ip_if.c
+++ b/usr/src/uts/common/inet/ip/ip_if.c
@@ -1383,7 +1383,17 @@ ill_capability_probe(ill_t *ill)
 static void
 ill_capability_wait(ill_t *ill)
 {
-	while (ill->ill_capab_pending_cnt != 0) {
+	/*
+	 * I'm in this ill's squeue.  The ILL_CONDEMNED flag can only be set
+	 * by someone who is the writer.  If it's set, we won't be receiving
+	 * any more messages from below that can cv_broadcast() to us.  So
+	 * also check in the loop for ILL_CONDEMNED, since we drop-and-
+	 * reacquire the squeue.
+	 */
+	ASSERT(IAM_WRITER_ILL(ill));
+
+	while (ill->ill_capab_pending_cnt != 0 &&
+	    !(ill->ill_flags & ILL_CONDEMNED)) {
 		mutex_enter(&ill->ill_dlpi_capab_lock);
 		ipsq_exit(ill->ill_phyint->phyint_ipsq);
 		cv_wait(&ill->ill_dlpi_capab_cv, &ill->ill_dlpi_capab_lock);
