commit aceca587e089571c5ee2a6bbd2ad7c5ab4d71e27 (refs/changes/02/4502/3)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2018-07-12T13:25:03-04:00 (1 year, 3 months ago)
    
    OS-7069 zone hang during shutdown after interface address conflict

diff --git a/usr/src/uts/common/inet/ip/ip_if.c b/usr/src/uts/common/inet/ip/ip_if.c
index c4b60daa68..d59e025d74 100644
--- a/usr/src/uts/common/inet/ip/ip_if.c
+++ b/usr/src/uts/common/inet/ip/ip_if.c
@@ -22,8 +22,8 @@
  * Copyright (c) 1991, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 1990 Mentat Inc.
  * Copyright (c) 2013 by Delphix. All rights reserved.
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
  * Copyright (c) 2014, OmniTI Computer Consulting, Inc. All rights reserved.
+ * Copyright 2018, Joyent, Inc.
  */
 
 /*
@@ -12768,11 +12768,14 @@ ill_dl_down(ill_t *ill)
 	mutex_exit(&ill->ill_lock);
 
 	if (mp != NULL) {
+		boolean_t already_condemned;
+
 		ip1dbg(("ill_dl_down: %s (%u) for %s\n",
 		    dl_primstr(*(int *)mp->b_rptr), *(int *)mp->b_rptr,
 		    ill->ill_name));
 		mutex_enter(&ill->ill_lock);
 		ill->ill_state_flags |= ILL_DL_UNBIND_IN_PROGRESS;
+		already_condemned = (ill->ill_state_flags & ILL_CONDEMNED);
 		mutex_exit(&ill->ill_lock);
 		/*
 		 * ip_rput does not pass up normal (M_PROTO) DLPI messages
@@ -12785,13 +12788,26 @@ ill_dl_down(ill_t *ill)
 		 * by DL_UNBIND. Also the DLD capability disable needs a
 		 * cv_wait'able context.
 		 */
-		if (ill->ill_state_flags & ILL_CONDEMNED)
+		if (already_condemned)
 			ill_capability_dld_disable(ill);
-		ill_capability_reset(ill, B_FALSE);
+		else
+			ill_capability_reset(ill, B_FALSE);
 		ill_dlpi_send(ill, mp);
 
-		/* Wait for the capability reset to finish */
-		ill_capability_wait(ill);
+		if (!already_condemned) {
+			boolean_t condemned_now;
+			/*
+			 * Check one more time in case we got condemned
+			 * in-between...
+			 */
+			mutex_enter(&ill->ill_lock);
+			condemned_now = (ill->ill_state_flags & ILL_CONDEMNED);
+			mutex_exit(&ill->ill_lock);
+
+			/* Wait for the capability reset to finish */
+			if (!condemned_now)
+				ill_capability_wait(ill);
+		}
 	}
 }
 
