commit a8844e693827790a0c4d4710605ac2b629cdff1e (refs/changes/02/4502/1)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2018-07-09T23:43:47-04:00 (1 year, 3 months ago)
    
    OS-7069 zone hang during shutdown after interface address conflict

diff --git a/usr/src/uts/common/inet/ip/ip_if.c b/usr/src/uts/common/inet/ip/ip_if.c
index c4b60daa68..57fe8dd73f 100644
--- a/usr/src/uts/common/inet/ip/ip_if.c
+++ b/usr/src/uts/common/inet/ip/ip_if.c
@@ -22,8 +22,8 @@
  * Copyright (c) 1991, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright (c) 1990 Mentat Inc.
  * Copyright (c) 2013 by Delphix. All rights reserved.
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
  * Copyright (c) 2014, OmniTI Computer Consulting, Inc. All rights reserved.
+ * Copyright 2017, Joyent, Inc.
  */
 
 /*
@@ -12787,11 +12787,14 @@ ill_dl_down(ill_t *ill)
 		 */
 		if (ill->ill_state_flags & ILL_CONDEMNED)
 			ill_capability_dld_disable(ill);
-		ill_capability_reset(ill, B_FALSE);
+		else
+			ill_capability_reset(ill, B_FALSE);
 		ill_dlpi_send(ill, mp);
 
-		/* Wait for the capability reset to finish */
-		ill_capability_wait(ill);
+		if (!(ill->ill_state_flags & ILL_CONDEMNED)) {
+			/* Wait for the capability reset to finish */
+			ill_capability_wait(ill);
+		}
 	}
 }
 
