commit 6b71f9f441616d87fd515f3f81d589747e566588 (refs/changes/73/3873/2)
Author: Jason King <jason.king@joyent.com>
Date:   2018-04-25T15:56:06-05:00 (1 year, 5 months ago)
    
    OS-6909 mandoc(1) should be large file aware

diff --git a/usr/src/cmd/mandoc/Makefile.common b/usr/src/cmd/mandoc/Makefile.common
index ac5b6fe2ac..c96c088d97 100644
--- a/usr/src/cmd/mandoc/Makefile.common
+++ b/usr/src/cmd/mandoc/Makefile.common
@@ -12,6 +12,7 @@
 #
 # Copyright 2014 Garrett D'Amore <garrett@damore.org>
 # Copyright 2017 Nexenta Systems, Inc.
+# Copyright 2018, Joyent, Inc.
 #
 
 PROG=		mandoc
@@ -78,6 +79,7 @@ OBJS +=		compat_ohash.o
 CFLAGS +=	$(CC_VERBOSE)
 CSTD =	$(CSTD_GNU99)
 
-CPPFLAGS +=	-DOSNAME="\"illumos\""
+CPPFLAGS +=	-DOSNAME="\"illumos\"" \
+		-D_LARGEFILE_SOURCE=1 -D_FILE_OFFSET_BITS=64
 
 LDLIBS +=	-lz
