commit 5ff5ef1a21bd7df91d8efb79cc44dccc8cf76911 (refs/changes/73/3873/4)
Author: Jason King <jason.king@joyent.com>
Date:   2018-04-26T14:14:07-05:00 (1 year, 5 months ago)
    
    OS-6909 mandoc(1) should be large file aware

diff --git a/usr/src/cmd/mandoc/Makefile.common b/usr/src/cmd/mandoc/Makefile.common
index ac5b6fe2ac..e47ef79569 100644
--- a/usr/src/cmd/mandoc/Makefile.common
+++ b/usr/src/cmd/mandoc/Makefile.common
@@ -12,6 +12,7 @@
 #
 # Copyright 2014 Garrett D'Amore <garrett@damore.org>
 # Copyright 2017 Nexenta Systems, Inc.
+# Copyright 2018, Joyent, Inc.
 #
 
 PROG=		mandoc
@@ -78,6 +79,6 @@ OBJS +=		compat_ohash.o
 CFLAGS +=	$(CC_VERBOSE)
 CSTD =	$(CSTD_GNU99)
 
-CPPFLAGS +=	-DOSNAME="\"illumos\""
+CPPFLAGS +=	-DOSNAME="\"illumos\"" -D_FILE_OFFSET_BITS=64
 
 LDLIBS +=	-lz
diff --git a/usr/src/tools/mandoc/Makefile b/usr/src/tools/mandoc/Makefile
index e6b9571841..e3b1b39422 100644
--- a/usr/src/tools/mandoc/Makefile
+++ b/usr/src/tools/mandoc/Makefile
@@ -11,7 +11,7 @@
 
 #
 # Copyright 2017 Nexenta Systems, Inc.
-# Copyright 2017 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 #
 
 CMDDIR=		$(SRC)/cmd/mandoc
@@ -21,6 +21,7 @@ include		$(SRC)/tools/Makefile.tools
 include		$(CMDDIR)/Makefile.common
 
 CPPFLAGS += -_gcc=-nostdinc -I/usr/include -I$(NATIVE_ADJUNCT)/include
+CPPFLAGS += -D_FILE_OFFSET_BITS=64
 LDFLAGS += -L$(NATIVE_ADJUNCT)/lib -R$(NATIVE_ADJUNCT)/lib
 OBJS +=		fts.o \
 		reallocarray.o \
