commit def3ed0fe7ee1a5ce86d439e7b73d61575330e67 (refs/changes/73/3873/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-04-25T15:53:34-05:00 (1 year, 5 months ago)
    
    fix

diff --git a/usr/src/cmd/mandoc/Makefile.common b/usr/src/cmd/mandoc/Makefile.common
index 1d2847eb74..c96c088d97 100644
--- a/usr/src/cmd/mandoc/Makefile.common
+++ b/usr/src/cmd/mandoc/Makefile.common
@@ -80,6 +80,6 @@ CFLAGS +=	$(CC_VERBOSE)
 CSTD =	$(CSTD_GNU99)
 
 CPPFLAGS +=	-DOSNAME="\"illumos\"" \
-		-D_LARGEFILE_SOURCE=1 -D_FILE_OFFSET_BITS=1
+		-D_LARGEFILE_SOURCE=1 -D_FILE_OFFSET_BITS=64
 
 LDLIBS +=	-lz
