commit cee9597d202bcc099e218829dee8c328fe599a43 (refs/changes/18/2218/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-07-14T10:29:00-07:00 (2 years, 3 months ago)
    
    joyent/node-docker-registry-client#19 v2.gcrio.test.js failure

diff --git a/test/v2.gcrio.test.js b/test/v2.gcrio.test.js
index 0c8c0f5..895352b 100644
--- a/test/v2.gcrio.test.js
+++ b/test/v2.gcrio.test.js
@@ -231,11 +231,17 @@ test('v2 gcr.io', function (tt) {
             t.equal(first.headers['docker-distribution-api-version'],
                 'registry/2.0');
 
-            t.ok(stream);
-            t.equal(stream.statusCode, 200);
-            t.equal(stream.headers['content-type'],
-                'application/octet-stream');
-            t.ok(stream.headers['content-length']);
+            t.ok(stream, 'got a stream');
+            t.equal(stream.statusCode, 200,
+                'got a 200 HTTP status: ' + stream.statusCode);
+            // At some point (by July 2017 at least), gcr started responding
+            // with `Content-Type: text/html` for this blob response for this
+            // image. Hrm.
+            // t.equal(stream.headers['content-type'],
+            //     'application/octet-stream');
+            t.ok(stream.headers['content-length'],
+                'got a Content-Length header: '
+                    + stream.headers['content-length']);
 
             var numBytes = 0;
             var hash = crypto.createHash(digest.split(':')[0]);
