From bc5bd1dcdc17d6c07abdb3594c5da4e76736ce1e Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Tue, 7 Feb 2017 00:25:29 +0000
Subject: [PATCH] MANTA-3085 gc_create_links must limit concurrent operations

---
 bin/gc_create_links.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/bin/gc_create_links.js b/bin/gc_create_links.js
index d222de7..2d675d0 100755
--- a/bin/gc_create_links.js
+++ b/bin/gc_create_links.js
@@ -7,7 +7,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -134,7 +134,7 @@ function processLinkFile(objPath, cb) {
                 }
 
                 //Create dirs, then link, then delete.
-                vasync.forEachParallel({
+                vasync.forEachPipeline({
                         func: MANTA_CLIENT.mkdirp.bind(MANTA_CLIENT),
                         inputs: dirs
                 }, function (err2, results) {
@@ -143,7 +143,7 @@ function processLinkFile(objPath, cb) {
                                 return;
                         }
 
-                        vasync.forEachParallel({
+                        vasync.forEachPipeline({
                                 func: function link(linkObj, subcb) {
                                         LOG.info({ linkObj: linkObj },
                                                  'linking object');
@@ -232,7 +232,7 @@ function createGcLinks(opts, cb) {
                         return;
                 }
 
-                vasync.forEachParallel({
+                vasync.forEachPipeline({
                         func: findAndVerifyJob,
                         inputs: objs
                 }, function (err2, results) {
-- 
2.21.0

