commit 588f06b362068cdb9bbee6f7ba14d525231bc644 (refs/changes/96/2596/2)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-09-20T16:47:07+00:00 (2 years, 1 month ago)
    
    MANATEE-365 pg_dump backup does not work with PostgreSQL 9.6.3

diff --git a/pg_dump/pg_backup_common.sh b/pg_dump/pg_backup_common.sh
index 983babd..d66fbe3 100644
--- a/pg_dump/pg_backup_common.sh
+++ b/pg_dump/pg_backup_common.sh
@@ -108,9 +108,17 @@ function mount_data_set
     # get pg dir size
     PG_DIR_SIZE=$(du -s $PG_DIR | cut -f1)
 
+    # Versions of PG after 9.5 removed the checkpoint_segments parameter
+    # if we're running on 9.2, we'll tune it, otherwise leave it alone.
+    PG_STARTUP_OPTIONS="-c logging_collector=off -c fsync=off \
+        -c synchronous_commit=off -c checkpoint_timeout=1h"
+    PG_SERVER_VERSION=$(postgres --version | cut -d' ' -f3)
+    if [[ $PG_SERVER_VERSION == 9.2* ]]; then
+        PG_STARTUP_OPTIONS+=" -c checkpoint_segments=100"
+    fi
+
     ctrun -o noorphan sudo -u postgres postgres -D $PG_DIR -p 23456 \
-        -c logging_collector=off -c fsync=off -c synchronous_commit=off \
-        -c checkpoint_segments=100 -c checkpoint_timeout=1h &
+         $PG_STARTUP_OPTIONS &
     PG_PID=$!
     [[ $? -eq 0 ]] || fatal "unable to start postgres"
 
