commit a20d8db204f1948b7ecf72ae1e0e802282a3e42b
Author: Sean Chittenden <sean@chittenden.org>
Date:   2017-07-06T08:09:02-07:00 (2 years, 3 months ago)
    
    Integrate changes from CM-1084 back into production configs.

diff --git a/etc/postgresql.manta.coal.conf b/etc/postgresql.manta.coal.conf
index 9bfe123..79acd49 100644
--- a/etc/postgresql.manta.coal.conf
+++ b/etc/postgresql.manta.coal.conf
@@ -162,7 +162,7 @@ shared_buffers = 32MB     # min 128kB
 wal_level = hot_standby # minimal, archive, or hot_standby
           # (change requires restart)
 fsync = on       # turns forced synchronization on or off
-synchronous_commit = on # synchronization level;
+synchronous_commit = off # synchronization level;
           # off, local, remote_write, or on
 #wal_sync_method = fsync    # the default is the first option
           # supported by the operating system:
@@ -172,7 +172,7 @@ synchronous_commit = on # synchronization level;
           #   fsync_writethrough
           #   open_sync
 # Not neccessary on ZFS, see http://www.postgresql.org/docs/current/static/wal-reliability.html
-full_page_writes = off          # recover from partial page writes
+full_page_writes = on          # recover from partial page writes
 #wal_buffers = -1     # min 32kB, -1 sets based on shared_buffers
           # (change requires restart)
 #wal_writer_delay = 200ms   # 1-10000 milliseconds
diff --git a/etc/postgresql.manta.lab.conf b/etc/postgresql.manta.lab.conf
index 0f2362e..e4c63cc 100644
--- a/etc/postgresql.manta.lab.conf
+++ b/etc/postgresql.manta.lab.conf
@@ -182,10 +182,10 @@ full_page_writes = off      # recover from partial page writes
 
 # - Checkpoints -
 
-#checkpoint_segments = 3    # in logfile segments, min 1, 16MB each
-#checkpoint_timeout = 5min    # range 30s-1h
-#checkpoint_completion_target = 0.5 # checkpoint target duration, 0.0 - 1.0
-#checkpoint_warning = 30s   # 0 disables
+checkpoint_segments = 30    # in logfile segments, min 1, 16MB each
+checkpoint_timeout = 5min    # range 30s-1h
+checkpoint_completion_target = 0.8 # checkpoint target duration, 0.0 - 1.0
+checkpoint_warning = 30s   # 0 disables
 
 # - Archiving -
 
diff --git a/etc/postgresql.manta.production.conf b/etc/postgresql.manta.production.conf
index 93046de..c695859 100644
--- a/etc/postgresql.manta.production.conf
+++ b/etc/postgresql.manta.production.conf
@@ -182,10 +182,10 @@ full_page_writes = off      # recover from partial page writes
 
 # - Checkpoints -
 
-checkpoint_segments = 1    # in logfile segments, min 1, 16MB each
-checkpoint_timeout = 30s # range 30s-1h
-#checkpoint_completion_target = 0.5 # checkpoint target duration, 0.0 - 1.0
-#checkpoint_warning = 30s   # 0 disables
+checkpoint_segments = 30   # in logfile segments, min 1, 16MB each
+checkpoint_timeout = 5s # range 30s-1h
+checkpoint_completion_target = 0.8 # checkpoint target duration, 0.0 - 1.0
+checkpoint_warning = 30s   # 0 disables
 
 # - Archiving -
 
