commit 67e0d8f4fb1db3e2f2a8bb76be61539fe81431fb (refs/changes/78/3678/2)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-03-18T17:20:45+00:00 (1 year, 7 months ago)
    
    OS-6790 LTP rmdir03 failing

diff --git a/usr/src/lib/brand/lx/testing/ltp_skiplist b/usr/src/lib/brand/lx/testing/ltp_skiplist
index 09ff2cacc1..40da511040 100644
--- a/usr/src/lib/brand/lx/testing/ltp_skiplist
+++ b/usr/src/lib/brand/lx/testing/ltp_skiplist
@@ -9,7 +9,7 @@
 # http://www.illumos.org/license/CDDL.
 #
 
-# Copyright 2017 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 
 # Broken tests
 poll02		# OS-3997
@@ -76,6 +76,9 @@ switch01
 ustat01
 ustat02
 
+# tests that need further investigation when time allows
+rmdir03
+
 # needs a config with a dev or mounting
 # As of February 2017, an increasing number of tests require device support
 # (often via a zeroed file mounted via loopback) to function properly.
