From c8c6cd09b9e851d519d68a9f448670dd650338ef Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Wed, 3 May 2017 06:40:28 +0000
Subject: [PATCH] OS-6100 OS-6082 somewhat short on follow through

---
 Makefile      | 8 ++++++++
 Makefile.defs | 7 -------
 2 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index a132e6e..b7ed818 100644
--- a/Makefile
+++ b/Makefile
@@ -96,6 +96,14 @@ GITDESCRIBE = \
 
 TARBALL =	$(NAME)-$(BRANCH)-$(TIMESTAMP)-$(GITDESCRIBE).tgz
 
+#
+# Some software (e.g., OpenSSL 0.9.8) is very particular about the Perl
+# interpreter used during the build.  This is the full path to the version
+# built during the strap build, which is safe to use on the build machine:
+#
+NATIVE_PERL =	$(STRAPPROTO)/usr/perl5/5.12/bin/perl
+
+
 all: $(SUBDIRS)
 
 strap: $(STRAP_SUBDIRS)
diff --git a/Makefile.defs b/Makefile.defs
index eec470d..91b61ee 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -57,13 +57,6 @@ SEPARATE_BUILD =
 
 CPPFLAGS =	$(SYSINCDIRS:%=-isystem $(DESTDIR)/%)
 
-#
-# Some software (e.g., OpenSSL 0.9.8) is very particular about the Perl
-# interpreter used during the build.  This is the full path to the version
-# built during the strap build, which is safe to use on the build machine:
-#
-NATIVE_PERL =	$(STRAPPROTO)/usr/perl5/5.12/bin/perl
-
 #
 # We ship platform-private perl modules (e.g. from ntp) in this directory,
 # and using this perl interpeter:
-- 
2.21.0

