commit c12589eaf9406fccd800cac935d2cfe36c68e20d (refs/changes/00/1000/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton instance audit a567a843

diff --git a/lib/do_instance/do_audit.js b/lib/do_instance/do_audit.js
index 82b09f8..ce4060f 100644
--- a/lib/do_instance/do_audit.js
+++ b/lib/do_instance/do_audit.js
@@ -27,7 +27,6 @@ var sortDefault = 'id,time';
 
 
 function do_audit(subcmd, opts, args, cb) {
-    var self = this;
     if (opts.help) {
         this.do_help('help', {}, [subcmd], cb);
         return;
@@ -51,23 +50,25 @@ function do_audit(subcmd, opts, args, cb) {
 
     var arg = args[0];
     var uuid;
+    var tritonapi = this.top.tritonapi;
 
-    if (common.isUUID(arg)) {
-        uuid = arg;
-        go1();
-    } else {
-        self.top.tritonapi.getInstance(arg, function (err, inst) {
-            if (err) {
-                cb(err);
-                return;
-            }
-            uuid = inst.id;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (common.isUUID(arg)) {
+            uuid = arg;
             go1();
-        });
-    }
+        } else {
+            tritonapi.getInstance(arg, function (err, inst) {
+                if (err) {
+                    cb(err);
+                    return;
+                }
+                uuid = inst.id;
+                go1();
+            });
+        }});
 
     function go1() {
-        self.top.tritonapi.cloudapi.machineAudit(uuid, function (err, audit) {
+        tritonapi.cloudapi.machineAudit(uuid, function (err, audit) {
             if (err) {
                 cb(err);
                 return;
