commit ad8bf30863b80754a63c7dffe18d46eb38bf32ab (refs/changes/54/3954/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-05-15T11:12:52-05:00 (1 year, 5 months ago)
    
    OS-6935 Possible race in lx_start_nfs_lockd

diff --git a/usr/src/uts/common/brand/lx/os/lx_lockd.c b/usr/src/uts/common/brand/lx/os/lx_lockd.c
index 234e815d3f..9c76c57352 100644
--- a/usr/src/uts/common/brand/lx/os/lx_lockd.c
+++ b/usr/src/uts/common/brand/lx/os/lx_lockd.c
@@ -77,6 +77,11 @@ lx_lockd_alive(pid_t lockd_pid)
 	}
 
 	mutex_enter(&p->p_lock);
+	if (p->p_stat == SZOMB || (p->p_flag & SEXITING) != 0) {
+		mutex_exit(&p->p_lock);
+		mutex_exit(&pidlock);
+		return (B_FALSE);
+	}
 	vp = p->p_exec;
 	VN_HOLD(vp);
 	mutex_exit(&p->p_lock);
