commit 22d423702ea8e017b089c0734446a2c157cd165f (refs/changes/56/5056/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-11-14T13:06:00+00:00 (11 months ago)
    
    OS-7373 gcc7 strapfix is incomplete

diff --git a/Makefile b/Makefile
index cd18df8..167ca81 100644
--- a/Makefile
+++ b/Makefile
@@ -41,7 +41,7 @@ SUBDIRS = \
 COMMA=,
 EXTRA_COMPILERS = $(subst $(COMMA), , $(SHADOW_COMPILERS))
 SUBDIRS +=	$(EXTRA_COMPILERS)
-STRAPFIX +=	$(EXTRA_COMPILERS)
+STRAPFIX +=	$(PRIMARY_COMPILER) $(EXTRA_COMPILERS)
 STRAPFIX_SUBDIRS=$(STRAPFIX:%=%.strapfix)
 
 else
@@ -181,6 +181,7 @@ $(STRAPFIX_SUBDIRS): $(SUBDIRS)
 	(cd $$(basename $@ .strapfix) && \
 	    PKG_CONFIG_LIBDIR="" \
 	    STRAP=$(STRAP) \
+	    PRIMARY_COMPILER=$(PRIMARY_COMPILER) \
 	    $(MAKE) DESTDIR=$(DESTDIR) strapfix)
 
 else
diff --git a/gcc4/Makefile b/gcc4/Makefile
index 6fa7269..9a485ea 100644
--- a/gcc4/Makefile
+++ b/gcc4/Makefile
@@ -82,8 +82,10 @@ ifeq ($(STRAP),strap)
 		$(DESTDIR)/$(PREFIX)/lib/64/libgcc_s.so.1 \
 		$(DESTDIR)/$(PREFIX)/lib/64/libstdc++.so.6
 	STRAPFIX_FILES = $(STRAPFIX_FILES_32) $(STRAPFIX_FILES_64)
+ifeq ($(PRIMARY_COMPILER),gcc4)
 	STRAPFIX_BINS = gcc g++ cpp
 endif
+endif
 
 #
 # Unlike everything else, gcc is built to be a cross-compiler, really.  It
diff --git a/gcc7/Makefile b/gcc7/Makefile
index fffe5a5..0d4351c 100644
--- a/gcc7/Makefile
+++ b/gcc7/Makefile
@@ -80,6 +80,9 @@ ifeq ($(STRAP),strap)
 		$(DESTDIR)/$(PREFIX)/lib/64/libgcc_s.so.1 \
 		$(DESTDIR)/$(PREFIX)/lib/64/libstdc++.so.6
 	STRAPFIX_FILES = $(STRAPFIX_FILES_32) $(STRAPFIX_FILES_64)
+ifeq ($(PRIMARY_COMPILER),gcc7)
+	STRAPFIX_BINS = gcc g++ cpp
+endif
 endif
 
 #
@@ -174,6 +177,10 @@ strapfix: $(STRAPFIX_LINKS) $(STRAPFIX_FILES) | install
 		    'dyn:runpath $(DESTDIR)/$(PREFIX)/lib/64' \
 		    $$f; \
 	done
+	mkdir -p $(DESTDIR)/usr/bin
+	for f in $(STRAPFIX_BINS); do \
+		ln -sf $(DESTDIR)/$(PREFIX)/bin/$$f $(DESTDIR)/usr/bin/$$f; \
+	done
 
 $(STRAPFIX_LINKS):
 	mkdir -p `dirname $@`
