From d251237127ae2876542ae934ca6dbe012355e7be Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 26 Apr 2019 10:46:17 +0200
Subject: [PATCH] TRITON-1299 need logarchiver-agent service to run along side
 hermes-actor

---
 deps/hermes | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/hermes b/deps/hermes
index f39da64..1cf50fa 160000
--- a/deps/hermes
+++ b/deps/hermes
@@ -1 +1 @@
-Subproject commit f39da64fb3ce370e0f3b24d8d96a9c5e6ecd0a79
+Subproject commit 1cf50fa80a674a5a427df733ac58eb162b7aadba
-- 
2.21.0

