commit d62c8c6dabed519130d057a97fee0927d6a365f4 (refs/changes/18/4518/3)
Author: Kody A Kantor <kody@kkantor.com>
Date:   2018-07-18T17:15:08+00:00 (1 year, 3 months ago)
    
    MANTA-3808 add functionality to disable mackerel and mola jobs

diff --git a/deps/jsstyle b/deps/jsstyle
index da42b50..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit da42b50ceb12d431437b32efd4c411a8e2fac0c8
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
diff --git a/lib/main.js b/lib/main.js
index 63f926a..69ce9f0 100755
--- a/lib/main.js
+++ b/lib/main.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var mod_assert = require('assert-plus');
@@ -118,6 +118,39 @@ function main() {
                 process.exit(1);
         }
 
+        if (config.disableAllJobs === true) {
+                console.log('all jobs are disabled, exiting.');
+                process.exit(0);
+        }
+
+        /* check if this job is enabled or not. */
+        var isEnabled = true;
+        switch (opts.jobName) {
+                case 'storage':
+                        isEnabled = (config.meterStorageEnabled === true);
+                        break;
+                case 'compute':
+                        isEnabled = (config.meterComputeEnabled === true);
+                        break;
+                case 'request':
+                case 'accessLogs':
+                        isEnabled = (config.meterRequestEnabled === true);
+                        break;
+                case 'summarizeDaily':
+                        isEnabled = (config.meterDailyEnabled === true);
+                        break;
+                default:
+                        console.log('unknown job name');
+                        break;
+        }
+
+        /* exit if job is not enabled. */
+        if (isEnabled === false) {
+                console.log(opts.jobName, 'job is disabled, exiting.');
+                process.exit(0);
+        }
+
+
         if (opts.configOnly) {
                 mod_meter.configureJob({
                         jobConfig: jobConfig,
diff --git a/sapi_manifests/mackerel/template b/sapi_manifests/mackerel/template
index c21c167..f87c7e4 100755
--- a/sapi_manifests/mackerel/template
+++ b/sapi_manifests/mackerel/template
@@ -24,5 +24,35 @@
     },
     "jobsFile": "etc/jobs.json",
     "mantaBaseDirectory": "/poseidon/stor/usage",
-    "lookupFile": "assets/etc/lookup.json"
+    "lookupFile": "assets/etc/lookup.json",
+    {{#DISABLE_ALL_JOBS}}
+    "disableAllJobs": {{DISABLE_ALL_JOBS}},
+    {{/DISABLE_ALL_JOBS}}
+    {{^DISABLE_ALL_JOBS}}
+    "disableAllJobs": false,
+    {{/DISABLE_ALL_JOBS}}
+    {{#METER_STORAGE_ENABLED}}
+    "meterStorageEnabled": {{METER_STORAGE_ENABLED}},
+    {{/METER_STORAGE_ENABLED}}
+    {{^METER_STORAGE_ENABLED}}
+    "meterStorageEnabled": true,
+    {{/METER_STORAGE_ENABLED}}
+    {{#METER_COMPUTE_ENABLED}}
+    "meterComputeEnabled": {{METER_COMPUTE_ENABLED}},
+    {{/METER_COMPUTE_ENABLED}}
+    {{^METER_COMPUTE_ENABLED}}
+    "meterComputeEnabled": true,
+    {{/METER_COMPUTE_ENABLED}}
+    {{#METER_REQUEST_ENABLED}}
+    "meterRequestEnabled": {{METER_REQUEST_ENABLED}},
+    {{/METER_REQUEST_ENABLED}}
+    {{^METER_REQUEST_ENABLED}}
+    "meterRequestEnabled": true,
+    {{/METER_REQUEST_ENABLED}}
+    {{#METER_PREV_DAY_ENABLED}}
+    "meterPrevDayEnabled": {{METER_PREV_DAY_ENABLED}}
+    {{/METER_PREV_DAY_ENABLED}}
+    {{^METER_PREV_DAY_ENABLED}}
+    "meterPrevDayEnabled": true
+    {{/METER_PREV_DAY_ENABLED}}
 }
