commit 52f57468539f5bf86230b900cf35dc216cec1eb6 (refs/changes/77/3677/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-03-18T08:43:24-07:00 (1 year, 7 months ago)
    
    OS-6742 "vmadm reboot" is not implemented for bhyve

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 6c0f3463..b7c39f32 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -14431,19 +14431,20 @@ function reset(uuid, log, callback)
 /*
  * This handles rebooting docker=true zones.
  */
-function doDockerReboot(vmobj, options, callback)
+function doStopStartReboot(vmobj, options, callback)
 {
     var log = options.log;
     var tracers_obj;
 
     if (process.env.EXPERIMENTAL_VMJS_TRACING) {
-        tracers_obj = traceUntilCallback('docker-reboot', log, callback);
+        tracers_obj = traceUntilCallback('docker-stopstart-reboot', log,
+            callback);
         callback = tracers_obj.callback;
         log = tracers_obj.log;
     }
 
     log.debug({vmobj_pid: vmobj.pid, timeout: options.timeout},
-        'doDockerReboot');
+        'doStopStartReboot');
 
     if (!options.timeout) {
         options.timeout = 10;
@@ -14680,8 +14681,8 @@ exports.reboot = function (uuid, options, callback)
             // re-add log in case it was changed
             options.log = log;
 
-            if (vmobj.docker) {
-                doDockerReboot(vmobj, options, cb);
+            if (vmobj.docker || vmobj.brand === 'bhyve') {
+                doStopStartReboot(vmobj, options, cb);
             } else {
                 doReboot(vmobj, options, cb);
             }
