From e2621e73c02690f7b79733f7bd69ef7d20e967cd Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 18 Jun 2018 12:13:25 -0700
Subject: [PATCH] backout MANTA-3719 Need Firewall Rules for Postgres (needs
 more work)

---
 boot/setup.sh | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index e191aa6..a1b3ec1 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2018, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -34,7 +34,6 @@ function manta_manatee_setup {
 
     echo "Adding local manifest directories"
     manta_add_manifest_dir "/opt/smartdc/manatee"
-    manta_add_manifest_dir "/opt/smartdc/waferlock"
 
     # MANTA-1360 no args to manta_common_setup so we don't rotate the 'manatee'
     # entry which stomps over the other manatee logs
@@ -85,9 +84,6 @@ function common_enable_services {
     #
     echo "Starting prefaulter"
     svccfg import /opt/smartdc/manatee/smf/manifests/pg_prefaulter.xml
-
-    echo "Starting waferlock"
-    svccfg import /opt/smartdc/waferlock/smf/manifests/waferlock.xml
 }
 
 function common_manatee_setup {
@@ -197,7 +193,6 @@ function manta_setup_manatee_env {
     manta_add_logadm_entry "manatee-sitter"
     manta_add_logadm_entry "manatee-backupserver"
     manta_add_logadm_entry "manatee-snapshotter"
-    manta_add_logadm_entry "waferlock"
     manta_add_logadm_entry "postgresql" "/var/pg"
     manta_add_logadm_entry "pgdump" "/var/log/manatee"
 }
-- 
2.21.0

