From 63fe4d87140e0040077164fd7ea71e56445922ef Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike.zeller@joyent.com>
Date: Mon, 26 Feb 2018 09:44:46 -0800
Subject: [PATCH] TRITON-179 AddNic should check for existence of ipv4_ips
 properly

---
 lib/nics.js | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/nics.js b/lib/nics.js
index e740a12..0a0fbc3 100644
--- a/lib/nics.js
+++ b/lib/nics.js
@@ -167,7 +167,11 @@ function addNic(req, res, next) {
         // anyways. If a user passed in a specific ip we will use that.
         assert.uuid(networkArg.ipv4_uuid, 'ipv4_uuid');
         networkUuid = networkArg.ipv4_uuid;
-        networkIp = networkArg.ipv4_ips[0];
+
+        if (networkArg.ipv4_ips) {
+            networkIp = networkArg.ipv4_ips[0];
+        }
+
         break;
     default:
         return next(new InvalidArgumentError(
-- 
2.21.0

