From c09ca918647fde1f2076436eefd69d5cae2d9ba4 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Fri, 7 Sep 2018 04:37:53 +0000
Subject: [PATCH] OS-7224 ship the certutil(1) manual page

---
 manifest             | 1 +
 nss-nspr/install-nss | 6 ++++++
 2 files changed, 7 insertions(+)

diff --git a/manifest b/manifest
index c40d74b..5e4caa3 100644
--- a/manifest
+++ b/manifest
@@ -375,6 +375,7 @@ s usr/lib/mps/secv1/libsoftokn3.chk=../libsoftokn3.chk
 s usr/lib/mps/secv1/libsoftokn3.so=../libsoftokn3.so
 s usr/lib/mps/secv1/libsqlite3.so=../libsqlite3.so
 s usr/lib/mps/secv1/libssl3.so=../libssl3.so
+f usr/share/man/man1/certutil.1 0444 root bin
 # ntp
 f lib/svc/manifest/network/ntp.xml 0444 root sys
 f etc/security/auth_attr.d/ntp 0444 root sys
diff --git a/nss-nspr/install-nss b/nss-nspr/install-nss
index ab59637..9821623 100644
--- a/nss-nspr/install-nss
+++ b/nss-nspr/install-nss
@@ -23,14 +23,20 @@ ROOT=$(pwd)
 BIN_PREFIX=${DESTDIR}/usr/bin
 PREFIX=${DESTDIR}/usr/lib/mps
 INC_PREFIX=${DESTDIR}/usr/include/mps
+MAN1_PREFIX=${DESTDIR}/usr/share/man/man1
 
 echo PREFIX=${PREFIX}
 mkdir -p ${PREFIX}
 mkdir -p ${BIN_PREFIX}
 mkdir -p ${INC_PREFIX}
+mkdir -p ${MAN1_PREFIX}
 
 source ../install.subr
 
+cd ${ROOT}/${VERS}
+
+_install N nss/doc/nroff/certutil.1 $MAN1_PREFIX/certutil.1 555
+
 cd ${ROOT}/${VERS}/dist/SunOS5.11_i86pc_gcc_OPT.OBJ
 
 _install N lib/libfreebl3.chk ${PREFIX}/libfreebl3.chk 555
-- 
2.21.0

