commit da65d9f40675d35b3d1d1b658993986931812b8a (refs/changes/04/204/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-08-04T18:06:07+00:00 (3 years, 2 months ago)
    
    NAPI-371 Deleting IPs on older networks broken by NAPI-363

diff --git a/lib/models/ip/index.js b/lib/models/ip/index.js
index 4bb31ea..25faa1d 100644
--- a/lib/models/ip/index.js
+++ b/lib/models/ip/index.js
@@ -538,7 +538,7 @@ function deleteIP(app, log, params, callback) {
     if (use_strings) {
         val.ipaddr = ip.toString();
     } else {
-        val.ip = util_ip.aton(key);
+        val.ip = util_ip.aton(ip.toString());
     }
 
     log.info(params, 'deleteIP: deleting IP %s', ip.toString());
