From da65d9f40675d35b3d1d1b658993986931812b8a Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Thu, 4 Aug 2016 18:06:07 +0000
Subject: [PATCH] NAPI-371 Deleting IPs on older networks broken by NAPI-363

---
 lib/models/ip/index.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/models/ip/index.js b/lib/models/ip/index.js
index 4bb31ea..25faa1d 100644
--- a/lib/models/ip/index.js
+++ b/lib/models/ip/index.js
@@ -538,7 +538,7 @@ function deleteIP(app, log, params, callback) {
     if (use_strings) {
         val.ipaddr = ip.toString();
     } else {
-        val.ip = util_ip.aton(key);
+        val.ip = util_ip.aton(ip.toString());
     }
 
     log.info(params, 'deleteIP: deleting IP %s', ip.toString());
-- 
2.21.0

