commit a64bdf05f70ec54731028e8c5248472daa9f3663 (refs/changes/86/3086/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-12-11T12:11:26-08:00 (1 year, 10 months ago)
    
    TOOLS-1949 sdcnode updates for Dec 2017 node.js security releases

diff --git a/Makefile b/Makefile
index 80f4157..2af1711 100644
--- a/Makefile
+++ b/Makefile
@@ -50,7 +50,7 @@ CLEAN_FILES	+= node_modules cscope.files
 
 # The prebuilt sdcnode version we want. See
 # "tools/mk/Makefile.node_prebuilt.targ" for details.
-NODE_PREBUILT_VERSION=v4.8.5
+NODE_PREBUILT_VERSION=v4.8.7
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
 	NODE_PREBUILT_TAG=zone
