commit cb357e69c19ec675d58d60f492136c537e0bcce5 (refs/changes/52/1052/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-02T14:41:24-08:00 (2 years, 10 months ago)
    
    joyent/node-cueball#57 insomniac resolver leads to sadness after #54

diff --git a/lib/resolver.js b/lib/resolver.js
index 6e64563..7076224 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -484,10 +484,18 @@ CueBallDNSResolver.prototype.state_aaaa = function (S) {
 		});
 	});
 	if (haveV6) {
-		this.r_srvRem = this.r_srvs.slice();
 		this.r_nextV6 = undefined;
+		this.r_srvRem = this.r_srvs.slice();
 		S.gotoState('aaaa_next');
 	} else {
+		var d = new Date();
+		/*
+		 * The extra +100 is to make sure we come back here *after* the
+		 * NIC cache has definitely expired.
+		 */
+		d.setTime(
+		    CueBallDNSResolver._nicCacheUpdated + NIC_CACHE_TTL + 100);
+		this.r_nextV6 = d;
 		S.gotoState('a');
 	}
 };
