commit 16f5987f90d0c6d2733158d0f13155462bbdf276 (refs/changes/81/1881/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-05-01T16:03:55-07:00 (2 years, 5 months ago)
    
    CNS-147 want unified tag parser for triton.cns.services

diff --git a/test/machines/tags.js b/test/machines/tags.js
index e86c7b5..5b64f2f 100644
--- a/test/machines/tags.js
+++ b/test/machines/tags.js
@@ -173,9 +173,12 @@ module.exports = function (suite, client, other, machine, callback) {
         }
 
         var unrecognizedMsg = 'Unrecognized special triton tag "triton.foo"';
-        var stringMsg = '"triton.cns.services" must be a string';
-        var booleanMsg = '"triton.cns.disable" must be a boolean';
-        var dnsMsg = '"_foo.bar" is not DNS safe';
+        var stringMsg = 'Triton tag "triton.cns.services" value must be a ' +
+            'string: true (boolean)';
+        var booleanMsg = 'Triton tag "triton.cns.disable" value must be a ' +
+            'boolean: "true" (string)';
+        var dnsMsg = 'invalid "triton.cns.services" tag: Expected DNS name ' +
+            'but "_" found.';
         var dockerMsg = 'Special tag "docker:label:com.docker." not supported';
 
         function postBadTritonTag(_, next) {
