From 462660d096cfdbc5f3c8140ad4c7dee688d9d6de Mon Sep 17 00:00:00 2001
From: Ante Vojvodic <ante@atcorner.hr>
Date: Wed, 19 Jun 2019 09:32:36 +0200
Subject: [PATCH] #839: configure should import configure-build if exists

---
 configure | 14 +++-----------
 1 file changed, 3 insertions(+), 11 deletions(-)

diff --git a/configure b/configure
index cb46dc5e..02ac4dcd 100755
--- a/configure
+++ b/configure
@@ -31,8 +31,9 @@ conf_pkgsrcurl="https://download.joyent.com/pub/build/pkgsrc"
 # Build defaults which are unlikely to change, but may be overridden by
 # configure-build.
 #
-if [[ ! -f configure-build ]]; then
-	source default.configure-build
+source default.configure-build
+if [[ -f configure-build ]]; then
+	source configure-build || fatal "failed to source configure-build"
 fi
 
 #
@@ -58,13 +59,6 @@ function fatal
 	exit 1
 }
 
-function source_vars
-{
-	if [[ -f "configure-build" ]]; then
-		source "$file" || fatal "failed to source configure-build"
-	fi
-}
-
 function check_loc
 {
 	[[ "$(readlink -f "$PWD")" != "$(readlink -f "$conf_arg0_dir")" ]] && \
@@ -478,8 +472,6 @@ read -r -d '' usage <<EOF
 		shadow compilers, comma delimited (gcc7,gcc#) [default: none]
 EOF
 
-source_vars
-
 while getopts "cdhp:rs:" arg; do
 	case $arg in
 	c)
-- 
2.21.0

