commit c194ed75d57bc9433068846930b5201fac17b1c2 (refs/changes/57/857/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-07T19:27:19-05:00 (2 years, 11 months ago)
    
    joyent/node-manta#274 mkdirp should not overwrite existing dirs

diff --git a/CHANGES.md b/CHANGES.md
index 206b5bb..8c48256 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -7,6 +7,7 @@
 ## 3.1.3
 
 - joyent/node-manta#277 mjob fails with "mjob: AssertionError: body (object) is required"
+- joyent/node-manta#274 `mput -p` no longer overwrites metadata of parent directories
 
 ## 3.1.2
 
diff --git a/lib/client.js b/lib/client.js
index 4fe45e2..788eb8d 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -1738,9 +1738,14 @@ MantaClient.prototype.mkdirp = function mkdirp(dir, opts, cb) {
         var tmp = dirs.slice(0, i).join('/');
         var _dir =
             path.posix.normalize(sprintf('/%s/%s/%s', root, tmp, _d));
-
-        tasks.push(function _mkdir(_, _cb) {
-            self.mkdir(_dir, _opts, _cb);
+        tasks.push(function _checkDir(_, _cb) {
+            self.info(_dir, _opts, function (err, check) {
+                if (err && err.name === 'NotFoundError') {
+                    self.mkdir(_dir, _opts, _cb);
+                } else {
+                    _cb(null);
+                }
+            });
         });
     });
 
diff --git a/test/client.test.js b/test/client.test.js
index 7b27aab..0743bec 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -29,6 +29,7 @@ var CHILD1 = SUBDIR1 + '/child1-' + libuuid.v4().split('-')[0]; // object
 var CHILD2 = SUBDIR2 + '/child2-' + libuuid.v4().split('-')[0]; // link
 var NOENTSUB1 = SUBDIR1 + '/a/b/c';
 var NOENTSUB2 = SUBDIR1 + '/d/e/f';
+var SUBDIR_274 = ROOT + '/node-manta-test-client-' + libuuid.v4().split('-')[0];
 var SPECIALOBJ1 = SUBDIR1 + '/' + 'before-\r-after';
 
 var SUBDIR1_NOBJECTS = 1;
@@ -291,6 +292,45 @@ test('put with mkdirp', function (t) {
     });
 });
 
+test('#274 put preserve metadata', function (t) {
+    var opts = {
+        headers: {
+            'm-billy-batson': 'captain-marvel'
+        }
+    };
+    var text = 'Shazam!';
+    var size = Buffer.byteLength(text);
+    var stream = new MemoryStream();
+    var self = this;
+
+    this.client.mkdirp(SUBDIR_274, function onMkdir(err0) {
+        t.ifError(err0);
+        self.client.chattr(SUBDIR_274, opts, function onChattr(err1) {
+            t.ifError(err1);
+            self.client.put(SUBDIR_274 + '/a/b/c/obj', stream, {
+                size: size,
+                mkdirs: true
+            }, function onPut(err2) {
+                t.ifError(err2);
+                self.client.info(SUBDIR_274, function onInfo(err3, info) {
+                    t.ifError(err3);
+                    t.ok(info, 'got info: ' + info);
+                    t.ok(info.headers, 'got info.headers: ' + info.headers);
+                    var headers = info.headers || {};
+                    t.ok(headers['m-billy-batson'], 'metadata exists');
+                    t.equal(headers['m-billy-batson'], 'captain-marvel');
+                    t.done();
+                });
+            });
+        });
+    });
+
+    process.nextTick(function () {
+        stream.write(text);
+        stream.end();
+    });
+});
+
 test('streams', function (t) {
     var client = this.client;
     var stream = new MemoryStream();
