From 896ca99be001542546ace7d21a6f385e8327bbaf Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 13 Aug 2018 12:40:18 +0000
Subject: [PATCH] 9674 Let's scrap AVS/sdbc

---
 overlay/generic/etc/minor_perm | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/overlay/generic/etc/minor_perm b/overlay/generic/etc/minor_perm
index 639401d8..52aa933f 100644
--- a/overlay/generic/etc/minor_perm
+++ b/overlay/generic/etc/minor_perm
@@ -48,7 +48,6 @@ wusb_ca:* 0666 root sys
 wusb_df:* 0666 root sys
 ipw:* 0666 root sys
 clone:ipw 0666 root sys
-ii:* 0666 root sys
 fcip:* 0600 root sys
 ral:* 0666 root sys
 clone:ral 0666 root sys
@@ -96,10 +95,6 @@ qlc:* 0666 root sys
 qlge:* 0666 root sys
 igb:* 0666 root sys
 clone:igb 0666 root sys
-ncall:* 0666 root sys
-nsctl:* 0666 root sys
-nskern:* 0666 root sys
-sdbc:* 0666 root sys
 afe:* 0666 root sys
 clone:afe 0666 root sys
 scsa1394:* 0666 root sys
-- 
2.21.0

