From 83fd53f36e17eacb62c475a4ba9654f712a3e4e8 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 5 Jul 2019 20:02:08 +0200
Subject: [PATCH] TRITON-1784 Get rid of Nodeunit for VMAPI tests

---
 Makefile      | 8 ++++----
 package.json  | 2 +-
 test/runtests | 9 ++++++---
 3 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/Makefile b/Makefile
index 8ea2d11..35e6e5b 100644
--- a/Makefile
+++ b/Makefile
@@ -36,7 +36,7 @@ NAME = vmapi
 #
 # Tools
 #
-NODEUNIT  := ./node_modules/.bin/nodeunit
+TAP_EXEC = ./node_modules/.bin/tap
 
 #
 # Files
@@ -76,13 +76,13 @@ RELSTAGEDIR          := /tmp/$(NAME)-$(STAMP)
 # Repo-specific targets
 #
 .PHONY: all
-all: $(SMF_MANIFESTS) | $(NODEUNIT) sdc-scripts
+all: $(SMF_MANIFESTS) | $(TAP_EXEC) sdc-scripts
 	$(NPM) install
 
-$(NODEUNIT): | $(NPM_EXEC)
+$(TAP_EXEC): | $(NPM_EXEC)
 	$(NPM) install
 
-CLEAN_FILES += $(NODEUNIT) ./node_modules/nodeunit
+CLEAN_FILES += $(TAP_EXEC) ./node_modules/.bin/tap
 
 BASE_IMAGE_UUID = 04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f
 BUILDIMAGE_NAME = $(NAME)
diff --git a/package.json b/package.json
index 2bfc24b..0601f1d 100644
--- a/package.json
+++ b/package.json
@@ -37,7 +37,7 @@
   },
   "devDependencies": {
     "byline": "5.0.0",
-    "nodeunit": "0.11.3"
+    "tap": "12.6.1"
   },
   "scripts": {
     "start": "node ./server.js"
diff --git a/test/runtests b/test/runtests
index 555a2f1..cb917cc 100755
--- a/test/runtests
+++ b/test/runtests
@@ -85,7 +85,9 @@ start_time=$(date +%s)
 
 TOP=$(cd $(dirname $0)/../; pwd)
 NODE_INSTALL=$TOP/build/node
-NODEUNIT=./node_modules/.bin/nodeunit
+TAP_EXEC=./node_modules/.bin/tap
+TEST_JOBS=10
+TEST_TIMEOUT_S=1200
 
 
 # Options.
@@ -209,6 +211,7 @@ fi
 # Run the tests.
 if [[ -n "$test_files" ]]; then
     VMAPI_IP=$VMAPI_IP CNAPI_IP=$CNAPI_IP NAPI_IP=$NAPI_IP \
-    PATH=$NODE_INSTALL/bin:$PATH $NODEUNIT --reporter \
-    $opt_reporter $test_files | tee $OUTPUT_DIR/vmapi.tap
+    PATH=$NODE_INSTALL/bin:$PATH $TAP_EXEC --timeout $TEST_TIMEOUT_S \
+    -j $TEST_JOBS -o $OUTPUT_DIR/vmapi.tap $test_files \
+    -R $opt_reporter
 fi
-- 
2.21.0

