commit 6e67c9fe851334d2f91e3a77944987bcf40be2b1 (refs/changes/14/2814/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-10-18T17:41:34+00:00 (2 years ago)
    
    OS-5969 Want support for QLogic QL41000/45000 series devices

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index f32881c6..8e29dede 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -1013,6 +1013,16 @@ pmcs "pciex11f8,8001"
 pseudo "zconsnex"
 pseudo "zfdnex"
 pvscsi "pci15ad,7c0"
+qede "pciex1077,1634"
+qede "pciex1077,1629"
+qede "pciex1077,1630"
+qede "pciex1077,1656"
+qede "pciex1077,1654"
+qede "pciex1077,1666"
+qede "pciex1077,8070"
+qede "pciex1077,8071"
+qede "pciex1077,8072"
+qede "pciex1077,8073"
 qlc "pci1077,132"
 qlc "pci1077,2200"
 qlc "pci1077,2300"
diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index eeb69a15..1ecd0d7d 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -296,3 +296,4 @@ vmxnet3s 299
 smrt 300
 iwn 301
 xhci 302
+qede 303
