commit e7d872974f4b34028ef061f8881f7169be4df1a0 (refs/changes/09/2609/1)
Author: Jared Morrow <jm@joyent.com>
Date:   2017-09-19T13:27:11-06:00 (2 years, 1 month ago)
    
    MANTA-3418: Updating to latest manta-scripts (253ae31)

diff --git a/deps/manta-scripts b/deps/manta-scripts
index 4b86dd1..253ae31 160000
--- a/deps/manta-scripts
+++ b/deps/manta-scripts
@@ -1 +1 @@
-Subproject commit 4b86dd132d68cb177f066a83d90b8fdc9ac2cde9
+Subproject commit 253ae31df06a8b5a3d520e3874d0eb060dda4fa3
