commit 3a519a7c006842fee692c420851de6ec21cfe859 (refs/changes/61/3561/3)
Author: Jeff Emershaw <jeff.emershaw@joyent.com>
Date:   2018-06-22T17:01:27+02:00 (1 year, 4 months ago)
    
    PUBAPI-1378 Provide UpdateFabricNetwork endpoint

diff --git a/deps/jsstyle b/deps/jsstyle
index d75b7ca..ccb145b 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit d75b7ca8308be17c80e2b120f2a01d4a0c20d8a8
+Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
diff --git a/lib/cloudapi/networks.js b/lib/cloudapi/networks.js
index 4395e54..c350337 100644
--- a/lib/cloudapi/networks.js
+++ b/lib/cloudapi/networks.js
@@ -1,11 +1,10 @@
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var schema_util = require('../util.js');
 var s_prim = require('../primitive.js');
 var base = s_prim.base;
-var qual = s_prim.qual;
 
 var s_create_network = {
 	'type': 'object',
@@ -18,13 +17,30 @@ var s_create_network = {
 		'provision_end_ip': base.str.v4addr.req,
 		'gateway': base.str.v4addr.opt,
 		'resolvers':
-		    schema_util.array_of(base.str.v4addr.opt),
+			schema_util.array_of(base.str.v4addr.opt),
 		'routes': base.obj.opt,
 		'vlan_id': base.intg.vlanid.req
 	},
 	'additionalProperties': false
 };
 
+var s_update_network = {
+	'type': 'object',
+	'properties': {
+		'id': base.str.uuid.req,
+		'name': base.str.netname.opt,
+		'routes': base.obj.opt,
+		'description': base.str.opt,
+		'resolvers':
+			schema_util.array_of(base.str.v4addr.opt),
+		'provision_start_ip': base.str.v4addr.opt,
+		'provision_end_ip': base.str.v4addr.opt,
+		'vlan_id': base.intg.vlanid.req
+	},
+	'additionalProperties': false,
+	'required': [ 'vlan_id', 'id' ]
+};
+
 var s_get_network = {
 	'type': 'object',
 	'properties': {
@@ -47,6 +63,7 @@ var s_list_networks = {
 module.exports = {
     CreateFabricNetwork: s_create_network,
     DeleteFabricNetwork: s_get_network,
+    UpdateFabricNetwork: s_update_network,
     GetFabricNetwork: s_get_network,
     ListFabricNetworks: s_list_networks
 };
diff --git a/test/tst.cloudapi.js b/test/tst.cloudapi.js
index 1fbaec8..913ddd4 100644
--- a/test/tst.cloudapi.js
+++ b/test/tst.cloudapi.js
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var extend = require('xtend');
@@ -60,6 +60,33 @@ test('CreateFabricNetwork', function (t) {
 	t.end();
 });
 
+test('UpdateFabricNetwork', function (t) {
+	var schema = s_cloudapi.UpdateFabricNetwork;
+	var baseParams = {
+		'id': '451640f5-a385-4924-bc3f-a16cf7eacaf2',
+		'name': 'asdf',
+		'vlan_id': 4
+	};
+
+	h.expectValidationErrors(t, schema, {}, [
+		h.errMissing('id'),
+		h.errMissing('vlan_id')
+	]);
+
+	h.expectSingleValidationError(t, schema,
+		extend(baseParams, { 'name': {} }),
+		'name', h.msg.objStr);
+
+	h.expectSingleValidationError(t, schema,
+		extend(baseParams, { 'routes': 'a' }),
+		'routes', h.msg.strObj);
+
+	h.expectSingleValidationError(t, schema,
+		extend(baseParams, { 'resolvers': 'a' }),
+		'resolvers', h.msg.strArr);
+
+	t.end();
+});
 
 test('CreateFabricVLAN', function (t) {
 	var schema = s_cloudapi.CreateFabricVLAN;
@@ -94,11 +121,11 @@ test('UpdateConfig', function (t) {
 	var schema = s_cloudapi.UpdateConfig;
 
 	h.expectSingleValidationError(t, schema,
-	    { 'default_network': 'asdf' },
+		{ 'default_network': 'asdf' },
 		'default_network', h.msg.uuid);
 
 	h.expectSuccess(t, schema,
-	    { 'default_network': '47c7466a-1813-470c-8805-0384f399566e' },
+	{ 'default_network': '47c7466a-1813-470c-8805-0384f399566e' },
 		'default_network', h.msg.uuid);
 
 	t.end();
