commit c8a1fde5fce37c8022b54e51ca462422dff86738
Author: Dan McDonald <danmcd@joyent.com>
Date:   2019-07-10T17:25:44-04:00 (3 months ago)
    
    OS-7885 manifest is missing /etc/ipf/smartos_version

diff --git a/overlay/generic/manifest b/overlay/generic/manifest
index c273c0d9..2d517eac 100644
--- a/overlay/generic/manifest
+++ b/overlay/generic/manifest
@@ -25,6 +25,8 @@ d etc/inet 0755 root sys
 f etc/inet/ntp.conf 0644 root sys
 f etc/inet/services 0644 root sys
 f etc/inittab 0644 root sys
+d etc/ipf 0755 root sys
+f etc/ipf/smartos_version 0444 root sys
 f etc/issue 0644 root sys
 f etc/logadm.conf 0644 root sys
 f etc/minor_perm 0644 root sys
