From 1e02e3f790309e478e8ddd7ae0afeef0ec74e61b Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 7 Feb 2019 19:47:03 +0000
Subject: [PATCH] TRITON-1177 hermes build looking for node in all the wrong
 places (fix build)

---
 Makefile | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/Makefile b/Makefile
index 08000b3..bd361e9 100644
--- a/Makefile
+++ b/Makefile
@@ -166,8 +166,13 @@ $(DESTDIR)$(PREFIX)/sapi_manifests/%: $(PWD)/sapi_manifests/%
 	@mkdir -p `dirname $@`
 	cp $^ $@
 
+$(DESTDIR)$(PREFIX)/node_modules: $(STAMP_NODE_MODULES)
+	rm -rf $@
+	cp -r $(PWD)/node_modules $@
+
 clean::
 	rm -rf $(PWD)/node_modules
+	rm -rf $(PWD)/make_stamps/node-modules
 	rm -rf $(PWD)/proto
 
 include ./tools/mk/Makefile.deps
-- 
2.21.0

