From a244d170c40da672e781f0e2cd0ee5e300e4602e Mon Sep 17 00:00:00 2001
From: Jordan Hendricks <jordan.hendricks@joyent.com>
Date: Wed, 9 Jan 2019 21:22:08 +0000
Subject: [PATCH] MANTA-4060 mpicker could more easily show utilization of all
 storage nodes

---
 bin/mpicker | 41 +++++++++++++++++++++++++++++++++++------
 1 file changed, 35 insertions(+), 6 deletions(-)

diff --git a/bin/mpicker b/bin/mpicker
index b3653a5..22ec753 100755
--- a/bin/mpicker
+++ b/bin/mpicker
@@ -8,7 +8,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 var mod_assert = require('assert-plus');
@@ -174,18 +174,24 @@ MPicker.prototype.do_poll = function do_poll(subcmd, opts, args, cb) {
         return;
     }
 
+    if (opts.percentage && opts.percentage > 100) {
+        cb(new VError('percentage should not be greater than 100 percent'));
+        return;
+    }
+
     p_opts = {
         log: LOG,
         defaultMaxStreamingSizeMB: cfg.defaultMaxStreamingSizeMB ||
             DEF_MAX_STREAMING_SIZE_MB,
-        maxUtilizationPct: cfg.storage.maxUtilizationPct ||
+        maxUtilizationPct: opts.percentage || cfg.storage.maxUtilizationPct ||
             DEF_MAX_PERCENT_UTIL,
         maxOperatorUtilizationPct: cfg.storage.maxOperatorUtilizationPct ||
-            DEF_MAX_PERCENT_UTIL,
+            DEF_MAX_OPERATOR_PERCENT_UTIL,
         multiDC: cfg.storage.multiDC,
         moray: cfg.storage.moray,
         lag: cfg.storage.lag
     };
+
     if (interval) {
         p_opts.interval = interval * 1000;
     }
@@ -193,10 +199,15 @@ MPicker.prototype.do_poll = function do_poll(subcmd, opts, args, cb) {
 
     function printTabularSummary(db) {
         mod_assert.object(db, 'db');
+        var summary, dcs, columns, rows;
 
-        var columns, rows;
+        summary = [];
+        summary.push('polling options:');
+        summary.push(' max percentage utilization: ' + p_opts.maxUtilizationPct);
+        summary.push(' max operator percentage utilization: ' + p_opts.maxOperatorUtilizationPct);
+        summary.push('');
 
-        var dcs = Object.keys(db);
+        dcs = Object.keys(db);
 
         var dc_label = 'DATACENTER';
         var msi_label = 'MANTA_STORAGE_ID';
@@ -238,6 +249,10 @@ MPicker.prototype.do_poll = function do_poll(subcmd, opts, args, cb) {
             });
         });
 
+        if (!opts.header) {
+            console.log(summary.join('\n'));
+        }
+
 
         mod_tab.emitTable({
             columns: columns,
@@ -300,6 +315,13 @@ MPicker.prototype.do_poll.options = commonOptions.concat([
         helpArg: 'FILE',
         completionType: 'file'
     },
+    {
+        names: ['percentage', 'p'],
+        type: 'positiveInteger',
+        help: 'Poll with custom utilization percentage maximum (overrides ' +
+            'config/default values)',
+        helpArg: 'PERCENTAGE'
+    },
     {
         names: ['all', 'a'],
         type: 'bool',
@@ -315,7 +337,7 @@ MPicker.prototype.do_poll.options = commonOptions.concat([
     {
         names: [ 'omitHeader', 'H' ],
         type: 'bool',
-        help: 'Omit header row from tabular output',
+        help: 'Omit summary and header row from tabular output',
         default: false
     },
     {
@@ -350,6 +372,13 @@ MPicker.prototype.do_poll.help = [
     '       "moray": a Moray configuration object',
     '       "lag": allowable grace period on storage records in ms (optional)',
     '',
+    'You may override the maximum percent utilization field in the ',
+    'configuration file using the `-p` flag.  For example, if you wanted to ',
+    'see storage information for all storage nodes, regardless of how close ',
+    'to capacity they are, you could poll with a max percentage utilization ',
+    'of 100:',
+    '   mpicker poll -p 100',
+    '',
     'You may also print the `poll` output in JSON format, which is suitable',
     'as input to the `choose` subcommand.  For example, to simulate choosing',
     '3 replicas for an object of size 150 MB from a given Moray\'s storage',
-- 
2.21.0

