From 48da2844407b91f09040816d6572b75cb46a7201 Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Wed, 11 Oct 2017 15:20:04 -0700
Subject: [PATCH] joyent/node-cueball#132 Expose counters and simple metrics
 from CueBallConnectionPool

---
 lib/pool.js | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..3d55225 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,19 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.poolStats = function () {
+	var self = this;
+	var stats = {};
+
+	stats['counters'] = self.p_counters;
+	stats['totalConnections'] = Object.keys(self.p_connections).length;
+	stats['idleConnections'] = self.p_idleq.q_len;
+	stats['pendingConnections'] = self.p_initq.q_len;
+	stats['waiterCount'] = self.p_waiters.q_len;
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
-- 
2.21.0

