From f8599124b02f03805af2b61aee00c67e91fee1f7 Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Wed, 11 Oct 2017 15:20:04 -0700
Subject: [PATCH] joyent/node-cueball#132 Expose counters and simple metrics
 from CueBallConnectionPool

---
 lib/pool.js | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..ee37660 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,25 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.getStats = function () {
+	var self = this;
+
+	// Get a snapshot of current counter values.
+	var counters = {};
+	Object.keys(this.p_counters).forEach(function (k) {
+		counters[k] = self.p_counters[k];
+	});
+
+	var stats = {};
+	stats['counters'] = counters;
+	stats['totalConnections'] = Object.keys(this.p_connections).length;
+	stats['idleConnections'] = this.p_idleq.length;
+	stats['pendingConnections'] = this.p_initq.length;
+	stats['waiterCount'] = this.p_waiters.length;
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
-- 
2.21.0

