From c952d0abff31d50985bbc04f5271467b6723102f Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Wed, 11 Oct 2017 15:20:04 -0700
Subject: [PATCH] joyent/node-cueball#132 Expose counters and simple metrics
 from CueBallConnectionPool

---
 lib/pool.js | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..e72d0b9 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,26 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.getStats = function () {
+	var self = this;
+
+	// Get a snapshot of current counter values.
+	var counters = {};
+	Object.keys(this.p_counters).forEach(function (k) {
+		counters[k] = self.p_counters[k];
+	});
+
+	var stats = {
+		'counters': counters,
+		'totalConnections': Object.keys(self.p_connections).length,
+		'idleConnections': self.p_idleq.length,
+		'pendingConnections': self.p_initq.length,
+		'waiterCount': self.p_waiters.length
+	};
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
-- 
2.21.0

