From ac43af5be288c7c060d9edfae7f54b3e0788e3ca Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Wed, 11 Oct 2017 15:20:04 -0700
Subject: [PATCH] joyent/node-cueball#132 Expose counters and simple metrics
 from CueBallConnectionPool

---
 lib/pool.js | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/lib/pool.js b/lib/pool.js
index 84cab63..627634c 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -729,6 +729,18 @@ CueBallConnectionPool.prototype.printConnections = function () {
 	console.log('dead:', self.p_dead);
 };
 
+CueBallConnectionPool.prototype.getStats = function () {
+	var stats = {};
+
+	stats['counters'] = this.p_counters;
+	stats['totalConnections'] = Object.keys(this.p_connections).length;
+	stats['idleConnections'] = this.p_idleq.length;
+	stats['pendingConnections'] = this.p_initq.length;
+	stats['waiterCount'] = this.p_waiters.length;
+
+	return (stats);
+};
+
 CueBallConnectionPool.prototype.claim = function (options, cb) {
 	var self = this;
 	var done = false;
-- 
2.21.0

