From b0dbde4f0771a239c61d0510d23ed6c1b4cf1e4b Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Tue, 13 Mar 2018 21:12:05 +0000
Subject: [PATCH] NET-387 portolan command line asks for --vnet_id but really
 wants --vnet-id

---
 bin/portolan | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/portolan b/bin/portolan
index 4de1039..d248407 100755
--- a/bin/portolan
+++ b/bin/portolan
@@ -70,7 +70,7 @@ var OPTS = {
     },
 
     vnet_id: {
-        names: [ 'vnet-id' ],
+        names: [ 'vnet_id', 'vnet-id' ],
         type: 'positiveInteger',
         help: 'Virtual network ID'
     },
-- 
2.21.0

