From c70bc6619e20dfb1829d62be04867b30fed0147c Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Tue, 8 Oct 2019 19:03:53 +0000
Subject: [PATCH] OS-6146 Update copyright date in /etc/release (fix build)

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index e5ff5dca..1a7017cc 100755
--- a/configure
+++ b/configure
@@ -436,7 +436,7 @@ RPCGEN=/opt/local/bin/rpcgen;			export RPCGEN
 ASTBINDIR=/opt/local/ast/bin;			export ASTBINDIR
 LD_TOXIC_PATH="\$ROOT/lib:\$ROOT/usr/lib";	export LD_TOXIC_PATH
 # generate buildversion for illumos buildversion module
-BUILDVERSION_EXEC="${conf_root}/tools/build_etcrelease";
+BUILDVERSION_EXEC="${conf_root}/tools/build_etcrelease -g";
 						export BUILDVERSION_EXEC
 
 EOF
-- 
2.21.0

