From 42d225c27308544b9d64b7d9b509c68046728245 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 21 May 2019 15:40:55 -0400
Subject: [PATCH] TRITON-1698 update sdc-systems-tests to node v6

---
 Makefile | 6 +++---
 deps/eng | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 99bc96c..9004c06 100644
--- a/Makefile
+++ b/Makefile
@@ -22,10 +22,10 @@ JSL_FILES_NODE	 = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 
-# sdc-minimal-multiarch-lts 15.4.1
-NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
+# minimal-64-lts 18.4.0
+NODE_PREBUILT_IMAGE=c2c31b00-1d60-11e9-9a77-ff9f06554b0f
 ifeq ($(shell uname -s),SunOS)
-	NODE_PREBUILT_VERSION=v0.10.48
+	NODE_PREBUILT_VERSION=v6.17.0
 	NODE_PREBUILT_TAG=gz
 else
 	NPM	:= $(shell which npm)
diff --git a/deps/eng b/deps/eng
index 126c0f0..d25b8fc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 126c0f032b8bfddd45b14e8ee14e73e9798a013f
+Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
-- 
2.21.0

