From 6fe5755698e91248d3d3d1c09f4fcb2e9cd9175a Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Fri, 14 Jul 2017 14:16:39 +0200
Subject: [PATCH] OS-6234 want a simple refcount implementation in DDI

---
 usr/src/uts/common/sys/sunddi.h | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/usr/src/uts/common/sys/sunddi.h b/usr/src/uts/common/sys/sunddi.h
index 1d94c8fd2c..586ba2f897 100644
--- a/usr/src/uts/common/sys/sunddi.h
+++ b/usr/src/uts/common/sys/sunddi.h
@@ -53,6 +53,7 @@
 #include <sys/sysevent.h>
 #include <sys/thread.h>
 #include <sys/stream.h>
+#include <sys/atomic.h>
 #if defined(__GNUC__) && defined(_ASM_INLINES) && defined(_KERNEL)
 #include <asm/sunddi.h>
 #endif
@@ -2285,6 +2286,33 @@ void ddi_register_aliases(plat_alias_t *pali, uint64_t npali);
 dev_info_t *ddi_alias_redirect(char *alias);
 char *ddi_curr_redirect(char *curr);
 
+/*
+ * DDI refcount interface
+ */
+typedef struct ddi_refcount {
+	uint64_t rc_count;
+} ddi_refcount_t;
+
+#define	ddi_refcount_create(rc)			((rc)->rc_count = 0)
+#define	ddi_refcount_create_untracked(rc)	((rc)->rc_count = 0)
+#define	ddi_refcount_destroy(rc)		((rc)->rc_count = 0)
+#define	ddi_refcount_destroy_many(rc, number)	((rc)->rc_count = 0)
+#define	ddi_refcount_is_zero(rc)		((rc)->rc_count == 0)
+#define	ddi_refcount_count(rc)			((rc)->rc_count)
+#define	ddi_refcount_add(rc, holder) \
+	atomic_inc_64_nv(&(rc)->rc_count)
+#define	ddi_refcount_remove(rc, holder) \
+	atomic_dec_64_nv(&(rc)->rc_count)
+#define	ddi_refcount_add_many(rc, number, holder) \
+	atomic_add_64_nv(&(rc)->rc_count, number)
+#define	ddi_refcount_remove_many(rc, number, holder) \
+	atomic_add_64_nv(&(rc)->rc_count, -number)
+#define	ddi_refcount_transfer(dst, src) { \
+	uint64_t __tmp = (src)->rc_count; \
+	atomic_add_64(&(src)->rc_count, -__tmp); \
+	atomic_add_64(&(dst)->rc_count, __tmp); \
+}
+
 #endif	/* _KERNEL */
 
 #ifdef	__cplusplus
-- 
2.21.0

