commit 29a350b2110e11c143cd98a2fe0a175d141d7720
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-07-09T16:24:55-06:00 (3 months ago)
    
    MANTA-4397 SAPI configuration for boray

diff --git a/boot/setup.sh b/boot/setup.sh
index 8e00a5f..fd08d17 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -27,11 +27,14 @@ ZONE_UUID=`/usr/bin/zonename`
 export PATH=$SVC_ROOT/bin:$SVC_ROOT/build/node/bin:/opt/local/bin:/usr/sbin/:/usr/bin:$PATH
 
 function manta_setup_boray {
-    svccfg import /opt/smartdc/muppet/smf/manifests/boray.xml
+    svccfg import /opt/smartdc/boray/smf/manifests/boray.xml
     svcadm enable boray
-    [[ $? -eq 0 ]] || fatal "Unable to start boray"
+    [[ $? -eq 0 ]] || fatal "unable to start boray"
 }
 
+source ${DIR}/scripts/util.sh
+source ${DIR}/scripts/services.sh
+
 echo "Running common setup scripts"
 manta_common_presetup
 
@@ -43,7 +46,7 @@ manta_common_setup "boray" 0
 manta_ensure_zk
 
 echo "Setting up Boray"
-manta_setup_boray_config
+manta_setup_boray
 
 manta_common_setup_end
 
diff --git a/sapi_manifests/boray/manifest.json b/sapi_manifests/boray/manifest.json
index 45a182e..cc5f22b 100644
--- a/sapi_manifests/boray/manifest.json
+++ b/sapi_manifests/boray/manifest.json
@@ -1,5 +1,5 @@
 {
     "name": "boray",
-    "path": "/opt/smartdc/boray/etc/config.json",
+    "path": "/opt/smartdc/boray/etc/config.toml",
     "master": true
 }
diff --git a/sapi_manifests/boray/template b/sapi_manifests/boray/template
index 2c63c08..eb3b6b7 100644
--- a/sapi_manifests/boray/template
+++ b/sapi_manifests/boray/template
@@ -1,2 +1,121 @@
-{
-}
+[log]
+{{#BORAY_LOG_LEVEL}}
+level = "{{BORAY_LOG_LEVEL}}"
+{{/BORAY_LOG_LEVEL}}
+{{^BORAY_LOG_LEVEL}}
+level = "info"
+{{/BORAY_LOG_LEVEL}}
+
+[server]
+{{#BORAY_SERVER_HOST}}
+host = "{{BORAY_SERVER_HOST}}"
+{{/BORAY_SERVER_HOST}}
+{{^BORAY_SERVER_HOST}}
+host = "0.0.0.0"
+{{/BORAY_SERVER_HOST}}
+{{#BORAY_SERVER_PORT}}
+port = {{BORAY_SERVER_PORT}}
+{{/BORAY_SERVER_PORT}}
+{{^BORAY_SERVER_PORT}}
+port = 2030
+{{/BORAY_SERVER_PORT}}
+
+[metrics]
+{{#BORAY_METRICS_HOST}}
+host = "{{BORAY_METRICS_HOST}}"
+{{/BORAY_METRICS_HOST}}
+{{^BORAY_METRICS_HOST}}
+host = "0.0.0.0"
+{{/BORAY_METRICS_HOST}}
+{{#BORAY_METRICS_PORT}}
+port = {{BORAY_METRICS_PORT}}
+{{/BORAY_METRICS_PORT}}
+{{^BORAY_METRICS_PORT}}
+port = 3020
+{{/BORAY_METRICS_PORT}}
+
+[database]
+user = "boray"
+{{#BORAY_DATABASE_HOST}}
+host = "{{BORAY_DATABASE_HOST}}"
+{{/BORAY_DATABASE_HOST}}
+{{^BORAY_DATABASE_HOST}}
+host = "127.0.0.1"
+{{/BORAY_DATABASE_HOST}}
+{{#BORAY_DATABASE_PORT}}
+port = {{BORAY_DATABASE_PORT}}
+{{/BORAY_DATABASE_PORT}}
+{{^BORAY_DATABASE_PORT}}
+port = 5432
+{{/BORAY_DATABASE_PORT}}
+{{#BORAY_DATABASE_DATABASE}}
+database = "{{BORAY_DATABASE_DATABASE}}"
+{{/BORAY_DATABASE_DATABASE}}
+{{^BORAY_DATABASE_DATABASE}}
+database = "boray"
+{{/BORAY_DATABASE_DATABASE}}
+{{#BORAY_DATABASE_APPLICATION_NAME}}
+application_name = "{{BORAY_DATABASE_APPLICATION_NAME}}"
+{{/BORAY_DATABASE_APPLICATION_NAME}}
+{{^BORAY_DATABASE_APPLICATION_NAME}}
+application_name = "boray"
+{{/BORAY_DATABASE_APPLICATION_NAME}}
+{{#BORAY_DATABASE_TLS_MODE}}
+tls_mode = "{{BORAY_DATABASE_TLS_MODE}}"
+{{/BORAY_DATABASE_TLS_MODE}}
+{{^BORAY_DATABASE_TLS_MODE}}
+tls_mode = "disable"
+{{/BORAY_DATABASE_TLS_MODE}}
+
+[cueball]
+{{#BORAY_CUEBALL_MAX_CONNECTIONS}}
+max_connections = {{BORAY_CUEBALL_MAX_CONNECTIONS}}
+{{/BORAY_CUEBALL_MAX_CONNECTIONS}}
+{{^BORAY_CUEBALL_MAX_CONNECTIONS}}
+max_connections = 10
+{{/BORAY_CUEBALL_MAX_CONNECTIONS}}
+{{#BORAY_CUEBALL_CLAIM_TIMEOUT}}
+claim_timeout = {{BORAY_CUEBALL_CLAIM_TIMEOUT}}
+{{/BORAY_CUEBALL_CLAIM_TIMEOUT}}
+{{^BORAY_CUEBALL_CLAIM_TIMEOUT}}
+claim_timeout = 500 # milliseconds
+{{/BORAY_CUEBALL_CLAIM_TIMEOUT}}
+{{#BORAY_CUEBALL_REBALANCER_ACTION_DELAY}}
+rebalancer_action_delay = {{BORAY_CUEBALL_REBALANCER_ACTION_DELAY}}
+{{/BORAY_CUEBALL_REBALANCER_ACTION_DELAY}}
+{{^BORAY_CUEBALL_REBALANCER_ACTION_DELAY}}
+rebalancer_action_delay = 100 # milliseconds
+{{/BORAY_CUEBALL_REBALANCER_ACTION_DELAY}}
+
+
+[tokio]
+{{#BORAY_TOKIO_CORE_THREADS}}
+core_threads = {{BORAY_TOKIO_CORE_THREADS}}
+{{/BORAY_TOKIO_CORE_THREADS}}
+{{^BORAY_TOKIO_CORE_THREADS}}
+core_threads = 4
+{{/BORAY_TOKIO_CORE_THREADS}}
+{{#BORAY_TOKIO_BLOCKING_THREADS}}
+blocking_threads = {{BORAY_TOKIO_BLOCKING_THREADS}}
+{{/BORAY_TOKIO_BLOCKING_THREADS}}
+{{^BORAY_TOKIO_BLOCKING_THREADS}}
+blocking_threads = 200
+{{/BORAY_TOKIO_BLOCKING_THREADS}}
+{{#BORAY_TOKIO_THREAD_KEEP_ALIVE}}
+thread_keep_alive = {{BORAY_TOKIO_THREAD_KEEP_ALIVE}}
+{{/BORAY_TOKIO_THREAD_KEEP_ALIVE}}
+{{^BORAY_TOKIO_THREAD_KEEP_ALIVE}}
+thread_keep_alive = 60
+{{/BORAY_TOKIO_THREAD_KEEP_ALIVE}}
+{{#BORAY_TOKIO_THREAD_STACK_SIZE}}
+thread_stack_size = {{BORAY_TOKIO_THREAD_STACK_SIZE}}
+{{/BORAY_TOKIO_THREAD_STACK_SIZE}}
+{{^BORAY_TOKIO_THREAD_STACK_SIZE}}
+thread_stack_size = 2097152
+{{/BORAY_TOKIO_THREAD_STACK_SIZE}}
+{{#BORAY_TOKIO_THREAD_NAME_PREFIX}}
+thread_name_prefix = "{{BORAY_TOKIO_THREAD_NAME_PREFIX}}"
+{{/BORAY_TOKIO_THREAD_NAME_PREFIX}}
+{{^BORAY_TOKIO_THREAD_NAME_PREFIX}}
+thread_name_prefix = "boray-worker-"
+{{/BORAY_TOKIO_THREAD_NAME_PREFIX}}
diff --git a/smf/manifests/boray.xml.in b/smf/manifests/boray.xml.in
index beaefa3..f199618 100644
--- a/smf/manifests/boray.xml.in
+++ b/smf/manifests/boray.xml.in
@@ -43,8 +43,9 @@
         </dependency>
 
         <exec_method type="method" name="start" timeout_seconds="10"
-          exec="/opt/smartdc/boray/bin/boray &amp;">
-            <method_context>
+          exec="./bin/boray -c ./etc/config.toml &amp;">
+          <method_context working_directory="/opt/smartdc/boray">
+                <method_credential user="root" group="nobody" />
             </method_context>
         </exec_method>
 
