commit 80b6065f4188ede7ac93099b00c5d2b18524cb0e (refs/changes/59/459/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-09-13T11:51:49-07:00 (3 years, 1 month ago)
    
    CMON-9 add cmon builds to purge-mg-builds process

diff --git a/tools/purge-mg-builds b/tools/purge-mg-builds
index 964ec17..0271476 100755
--- a/tools/purge-mg-builds
+++ b/tools/purge-mg-builds
@@ -97,6 +97,8 @@ TTL_DAYS_FROM_NAME='{
     "binder": 365,
     "ca": 365,
     "cloudapi": 365,
+    "cmon": 365,
+    "cmon-agent": 365,
     "cn-agent": 365,
     "cnapi": 365,
     "cns": 365,
