From e1aab09f2e15872ea0458210a5dcbe6e4cf13f92 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Fri, 15 Mar 2019 15:07:37 -0400
Subject: [PATCH] TOOLS-2101 jsstyle submodules too old for old perls

---
 .gitmodules | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.gitmodules b/.gitmodules
index c967475..7ff1faa 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -6,7 +6,7 @@
 	url = git://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
 	url = https://github.com/joyent/eng.git
-- 
2.21.0

