commit ce4ba4a3737aaf20fec282e287832648a0525a24 (refs/changes/49/3349/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-02-09T23:44:34+00:00 (1 year, 8 months ago)
    
    PKCS#11 tests should not use CRYPTO_INVALID_SESSION

diff --git a/usr/src/test/crypto-tests/tests/common/cryptotest_pkcs.c b/usr/src/test/crypto-tests/tests/common/cryptotest_pkcs.c
index 4a98a2aafc..db15968b49 100644
--- a/usr/src/test/crypto-tests/tests/common/cryptotest_pkcs.c
+++ b/usr/src/test/crypto-tests/tests/common/cryptotest_pkcs.c
@@ -11,10 +11,12 @@
 
 /*
  * Copyright 2015 Nexenta Systems, Inc.  All rights reserved.
+ * Copyright 2018, Joyent, Inc.
  */
 
 #include <stdio.h>
 #include <cryptoutil.h>
+#include <security/cryptoki.h>
 
 #include "cryptotest.h"
 
@@ -64,7 +66,7 @@ cryptotest_init(cryptotest_t *arg, size_t fg)
 
 	op->mechname = arg->mechname;
 
-	op->hsession = CRYPTO_INVALID_SESSION;
+	op->hsession = CK_INVALID_HANDLE;
 	op->fg = fg;
 
 	if (op->out == NULL)
@@ -87,7 +89,7 @@ int
 cryptotest_close(crypto_op_t *op)
 {
 	(void) C_DestroyObject(op->hsession, op->keyt);
-	if (op->hsession != CRYPTO_INVALID_SESSION)
+	if (op->hsession != CK_INVALID_HANDLE)
 		(void) cryptotest_close_session(op->hsession);
 	free(op);
 	return (C_Finalize(NULL));
