From cce07e6d21fa870d7a307108fb67644b56a86229 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Tue, 11 Dec 2018 10:24:11 -0700
Subject: [PATCH] joyent/node-fast#22 Fix copy-paste bug in fast_server module

---
 lib/fast_server.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/fast_server.js b/lib/fast_server.js
index 1c633cb..4cc50c0 100644
--- a/lib/fast_server.js
+++ b/lib/fast_server.js
@@ -227,7 +227,7 @@ function FastServer(args)
 	this.fs_nrequests_failed = 0;		/* count of reqs failed */
 
 	if (this.fs_collector) {
-		if (this.fc_collector.FIXED_BUCKETS === true) {
+		if (this.fs_collector.FIXED_BUCKETS === true) {
 			fixed_buckets = true;
 		}
 		this.fs_request_counter = this.fs_collector.counter({
-- 
2.21.0

