commit 3a00a96957f9d9d003c0cb980890c42de1451337 (refs/changes/37/2437/4)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-23T20:39:19+00:00 (2 years, 2 months ago)
    
    MANTA-3340 muskie mpu could log a bit quieter

diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index ded74d8..a3976d2 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -312,7 +312,8 @@ function validateParts(req, res, next) {
                 var e = errors[i];
                 assert.number(e.statusCode);
                 if (e.statusCode >= 500) {
-                    log.error('internal error: ' + e);
+                    e = new verror.VError(e, 'internal error');
+                    log.warn(e);
                 } else if (e.statusCode >= 400) {
                     next(e);
                     return;
@@ -434,8 +435,10 @@ function finalizeUpload(req, res, next) {
 
             if (err) {
                 results.operations.forEach(function (r) {
-                    log.error('error with shark ' + r.result.shark +
-                        ': ' + r.err);
+                    log.warn({
+                        err: r.err,
+                        shark: r.result.shark
+                    }, 'error with shark');
                 });
                 next(new SharksExhaustedError());
             } else {
@@ -458,7 +461,7 @@ function finalizeUpload(req, res, next) {
                     var msg = 'mako nodes returned different md5 sums for ' +
                         'the same object';
 
-                    log.error({
+                    log.warn({
                         results: results.operations.map(function (r) {
                             return ({
                                 shark: r.result.shark,
diff --git a/lib/uploads/common.js b/lib/uploads/common.js
index e22ca84..cc43a1c 100644
--- a/lib/uploads/common.js
+++ b/lib/uploads/common.js
@@ -886,7 +886,9 @@ function commitUpload(partsMD5, size, md5, cb) {
 
                     self.req.moray.commitMPU(opts, function (err3, meta) {
                         if (err3) {
-                            log.error('error batching data: ' + err3);
+                            err3 = new verror.VError(err3,
+                                'error batching data');
+                            log.warn(err3);
                             cb(err3);
                         } else {
                             log.debug('batch successful');
