From 430e333151fb38a90ba526eb2c37a9d30fb1a220 Mon Sep 17 00:00:00 2001
From: Nick Zivkovic <nick.zivkovic@joyent.com>
Date: Fri, 20 Apr 2018 20:56:10 +0000
Subject: [PATCH] TRITON-327 sdcadm gets VMAPI-error on fabric-setup

---
 lib/post-setup/fabrics.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/post-setup/fabrics.js b/lib/post-setup/fabrics.js
index d08c3f7..24745c7 100644
--- a/lib/post-setup/fabrics.js
+++ b/lib/post-setup/fabrics.js
@@ -712,7 +712,7 @@ Fabrics.prototype.updateFabricsSvcs = function updateFabricsSvcs(cb) {
     if (!self.configChanged) {
         return cb(null);
     }
-    var svcs = ['napi', 'vmapi', 'dhcpd'];
+    var svcs = ['napi', 'dhcpd', 'vmapi'];
 
     self.progress('Restarting config of services using "fabric_cfg": %s',
     svcs.join(', '));
@@ -720,7 +720,7 @@ Fabrics.prototype.updateFabricsSvcs = function updateFabricsSvcs(cb) {
     if (self.dockerSetup) {
         svcs.push('docker');
     }
-    vasync.forEachParallel({
+    vasync.forEachPipeline({
         inputs: svcs,
         func: function updateSvc(svc, next) {
             common.spawnRun({
-- 
2.21.0

