commit 6253c437278a9e9ee009dee1fd6318ab75294db7 (refs/changes/31/4831/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2018-09-12T17:29:27+00:00 (1 year, 1 month ago)
    
    OS-7242 $(VARPD_MAPFILES) escaped out of the varpd makefiles

diff --git a/usr/src/lib/librename/Makefile b/usr/src/lib/librename/Makefile
index 222523d9a1..1278c6cc9d 100644
--- a/usr/src/lib/librename/Makefile
+++ b/usr/src/lib/librename/Makefile
@@ -30,7 +30,7 @@ lint :=		TARGET = lint
 
 all clean clobber lint: $(SUBDIRS)
 
-install: 	$(SUBDIRS) $(VARPD_MAPFILES) install_h
+install: 	$(SUBDIRS) install_h
 
 install_h: 	$(ROOTHDRS)
 
diff --git a/usr/src/lib/libsff/Makefile b/usr/src/lib/libsff/Makefile
index 39c13c30bf..134bf99c6d 100644
--- a/usr/src/lib/libsff/Makefile
+++ b/usr/src/lib/libsff/Makefile
@@ -30,7 +30,7 @@ lint :=		TARGET = lint
 
 all clean clobber lint: $(SUBDIRS)
 
-install: 	$(SUBDIRS) $(VARPD_MAPFILES) install_h
+install: 	$(SUBDIRS) install_h
 
 install_h: 	$(ROOTHDRS)
 
