From 75192a868cdced71327f4c49109cda967434d98b Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Fri, 15 Mar 2019 15:07:22 -0400
Subject: [PATCH] TOOLS-2101 jsstyle submodules too old for old perls

---
 .gitmodules  | 2 +-
 deps/jsstyle | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 88a739a..73bd0dc 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
 	url = git://github.com/davepacheco/javascriptlint.git
diff --git a/deps/jsstyle b/deps/jsstyle
index da42b50..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit da42b50ceb12d431437b32efd4c411a8e2fac0c8
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
-- 
2.21.0

