From fd33b842396ad4073fb0e253f045db04a98f35db Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Mon, 28 May 2018 14:27:51 -0700
Subject: [PATCH] TRITON-436 remove obsolete envvars from Amon test suite setup

---
 test/runtests | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/test/runtests b/test/runtests
index 77f1520..fcb0979 100755
--- a/test/runtests
+++ b/test/runtests
@@ -131,9 +131,6 @@ export UFDS_ROOTDN=$CONFIG_ufds_ldap_root_dn
 export UFDS_PASSWORD=$CONFIG_ufds_ldap_root_pw
 export VMAPI_URL="http://${CONFIG_vmapi_admin_ips%%,*}"
 export CNAPI_URL="http://${CONFIG_cnapi_admin_ips%%,*}"
-export REDIS_HOST=$(echo $CONFIG_redis_admin_ips | cut -d, -f1)
-export REDIS_PORT=6379
-export DATACENTER_NAME=$CONFIG_datacenter_name
 
 echo ""
 echo "# Datacenter config:"
@@ -143,9 +140,6 @@ echo "# UFDS_ROOTDN is $UFDS_ROOTDN"
 echo '# UFDS_PASSWORD is ***'
 echo "# VMAPI_URL is $VMAPI_URL"
 echo "# CNAPI_URL is $CNAPI_URL"
-echo "# REDIS_HOST is $REDIS_HOST"
-echo "# REDIS_PORT is $REDIS_PORT"
-echo "# DATACENTER_NAME is $DATACENTER_NAME"
 
 
 # Currently not sure if we need to run from $TOP. Let's just do so.
-- 
2.21.0

