From a42f276b5809307bbf0861a6cc882562de48df12 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 14 Dec 2018 16:22:30 -0800
Subject: [PATCH] MANTA-4035 retire propeller

---
 update/repos.sh | 1 -
 1 file changed, 1 deletion(-)

diff --git a/update/repos.sh b/update/repos.sh
index 8281754..5fa9174 100755
--- a/update/repos.sh
+++ b/update/repos.sh
@@ -33,7 +33,6 @@ REPOS=(
     manta-medusa
     manta-mola
     manta-muskie
-    manta-propeller
     manta-reshard
     manta-wrasse
     moray
-- 
2.21.0

