commit a42f276b5809307bbf0861a6cc882562de48df12 (refs/changes/48/5248/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-12-14T16:22:30-08:00 (10 months ago)
    
    MANTA-4035 retire propeller

diff --git a/update/repos.sh b/update/repos.sh
index 8281754..5fa9174 100755
--- a/update/repos.sh
+++ b/update/repos.sh
@@ -33,7 +33,6 @@ REPOS=(
     manta-medusa
     manta-mola
     manta-muskie
-    manta-propeller
     manta-reshard
     manta-wrasse
     moray
