From e21b8048bf39652bb32416f0cb3b8f77d80f3755 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Wed, 11 Oct 2017 14:34:22 -0400
Subject: [PATCH] HEAD-2369 cache=false with sdc-server setup still creates a
 cache device

---
 scripts/joysetup.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/joysetup.sh b/scripts/joysetup.sh
index 97b83d34..4572505f 100755
--- a/scripts/joysetup.sh
+++ b/scripts/joysetup.sh
@@ -625,7 +625,7 @@ if [[ "$(zpool list)" == "no pools available" ]] \
         dl3="${CONFIG_layout}"
     fi
 
-    if ! /usr/bin/disklayout ${dl0} "${dl1}" ${dl2} "${dl3}"> ${POOL_FILE}; then
+    if ! /usr/bin/disklayout ${dl0} ${dl1:+"$dl1"} ${dl2} "${dl3}"> ${POOL_FILE}; then
         fatal "disk layout failed"
     fi
 
-- 
2.21.0

