From b00a8418c893936cf571d0035ca8deaab66a3d4e Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 23 May 2017 21:06:53 +0000
Subject: [PATCH] Add yet another CR

---
 README.md | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/README.md b/README.md
index d51b47e..e686744 100644
--- a/README.md
+++ b/README.md
@@ -4,3 +4,5 @@
 Can we rewind things?
 
 This is an update
+
+This is the second update
-- 
2.21.0

