commit ddde3927c85ec3c418681c58f0b8e4fd5ad4d88d
Author: Mohamed Khalfella <mohamed.khalfella@joyent.com>
Date:   2019-09-04T04:46:16+00:00 (5 weeks ago)
    
    OS-7975 bhyve should require exactly one primary nic

diff --git a/usr/src/lib/brand/bhyve/zone/boot.c b/usr/src/lib/brand/bhyve/zone/boot.c
index 6e6f3dc640..ebbe7ef9dc 100644
--- a/usr/src/lib/brand/bhyve/zone/boot.c
+++ b/usr/src/lib/brand/bhyve/zone/boot.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -377,6 +377,7 @@ add_nets(int *argc, char **argv)
 {
 	char *nets;
 	char *net;
+	char *next_net = NULL;
 	char *lasts;
 	int nextpcifn = 1;		/* 0 reserved for primary */
 	char slotconf[MAXNAMELEN];
@@ -386,10 +387,11 @@ add_nets(int *argc, char **argv)
 		return (0);
 	}
 
-	for (net = strtok_r(nets, " ", &lasts); net != NULL;
-	    net = strtok_r(NULL, " ", &lasts)) {
+	for (net = strtok_r(nets, " ", &lasts); net != NULL; net = next_net) {
 		int pcifn;
 
+		next_net = strtok_r(NULL, " ", &lasts);
+
 		/* zoneadmd is not careful about a trailing delimiter. */
 		if (net[0] == '\0') {
 			continue;
@@ -421,6 +423,13 @@ add_nets(int *argc, char **argv)
 		    add_arg(argc, argv, slotconf) != 0) {
 			return (-1);
 		}
+
+		/* Make sure there is a "primary" net */
+		if (next_net == NULL && primary == NULL) {
+			(void) printf("Error: "
+			    "no primary net has been specified\n");
+			return (-1);
+		}
 	}
 
 	return (0);
