From 1382fa6e6c3dd6c13321778ab0c4ff759848ae22 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Wed, 14 Mar 2018 22:08:08 +0000
Subject: [PATCH] TOOLS-1302 sdcadm update failed when importing image

---
 lib/procedures/shared.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/procedures/shared.js b/lib/procedures/shared.js
index 82f6f5e..4c0751e 100644
--- a/lib/procedures/shared.js
+++ b/lib/procedures/shared.js
@@ -762,7 +762,8 @@ function imgadmInstallRemote(opts, callback) {
 
     var argv = [
         '/opt/smartdc/bin/sdc-oneachnode',
-        format('-n %s ', server),
+        '-n', server,
+        '-T', '300',
         format('/usr/sbin/imgadm import -q %s', img.uuid)
     ];
 
-- 
2.21.0

