From d0052bfa3ddf38a16633ab8deaec021d609fa8d0 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Fri, 10 Aug 2018 20:13:30 +0000
Subject: [PATCH] joyent/manta-garbage-collector#48 repo cleanup/improvements

---
 lib/mako_instruction_uploader.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/mako_instruction_uploader.js b/lib/mako_instruction_uploader.js
index f18ea3d..0f5eba7 100644
--- a/lib/mako_instruction_uploader.js
+++ b/lib/mako_instruction_uploader.js
@@ -142,7 +142,7 @@ _listen_for_instructions()
 
 		var keys = [];
 		var lines = [];
-		var marked_space = [];
+		var marked_space = 0;
 
 		instruction.lines.forEach(function (elem) {
 			keys.push(elem.key);
-- 
2.21.0

