commit d0052bfa3ddf38a16633ab8deaec021d609fa8d0 (refs/changes/60/4660/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-08-10T20:13:30+00:00 (1 year, 2 months ago)
    
    joyent/manta-garbage-collector#48 repo cleanup/improvements

diff --git a/lib/mako_instruction_uploader.js b/lib/mako_instruction_uploader.js
index f18ea3d..0f5eba7 100644
--- a/lib/mako_instruction_uploader.js
+++ b/lib/mako_instruction_uploader.js
@@ -142,7 +142,7 @@ _listen_for_instructions()
 
 		var keys = [];
 		var lines = [];
-		var marked_space = [];
+		var marked_space = 0;
 
 		instruction.lines.forEach(function (elem) {
 			keys.push(elem.key);
