From f6232d26e20018d0dcc5e38c5a39e9ccdecad228 Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Tue, 7 May 2019 01:54:11 +0000
Subject: [PATCH] MANTA-4142 Add metricPorts mdata variable to electric-moray

---
 boot/setup.sh | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index 44bb06d..da7ca0e 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -202,6 +202,19 @@ function manta_setup_electric_moray {
             fatal "unable to start $electric_moray_instance"
     done
 
+    #
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    #
+    # The metricPorts values are derived from the electric-moray service's
+    # "SIZE" SAPI metadata. We don't need to worry about keeping the metricPorts
+    # updated if this variable changes, because such a change does not affect
+    # already-provisioned zones. This is because electric-moray zones pull the
+    # "SIZE" variable from /var/tmp/metadata.json, which is only written once,
+    # when the zone is provisioned -- it is not managed by config-agent.
+    #
+    mdata-put metricPorts $(IFS=','; echo "${kangs[*]}")
+
     unset IFS
 }
 
-- 
2.21.0

