commit 4d1a20241a0c2022a1b0417772320ca0215550c3
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-03-01T22:56:31+00:00 (7 months ago)
    
    MANTA-4142 Add metricPorts mdata variable to electric-moray

diff --git a/boot/setup.sh b/boot/setup.sh
index 44bb06d..b802e87 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -202,6 +202,10 @@ function manta_setup_electric_moray {
             fatal "unable to start $electric_moray_instance"
     done
 
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    mdata-put metricPorts $(IFS=','; echo "${kangs[*]}")
+
     unset IFS
 }
 
