commit a063b4a0aa0290ccf4ae8e3841165ad60542d76f (refs/changes/03/4503/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-07-10T17:30:34+02:00 (1 year, 3 months ago)
    
    TRITON-556 Update coal-setup.sh

diff --git a/tools/coal-setup.sh b/tools/coal-setup.sh
index 2e12da5..fb13888 100755
--- a/tools/coal-setup.sh
+++ b/tools/coal-setup.sh
@@ -185,7 +185,12 @@ fi
 echo "# Enable volapi"
 sdcadm post-setup volapi
 sdcadm experimental nfs-volumes cloudapi
+sdcadm experimental nfs-volumes cloudapi-automount
 sdcadm experimental nfs-volumes docker
+sdcadm experimental nfs-volumes docker-automount
+
+echo "# Enable cns"
+sdcadm post-setup cns
 
 if ! $(nictagadm exists sdc_underlay 2>/dev/null); then
     echo "# Enable external and underlay nics"
