From a063b4a0aa0290ccf4ae8e3841165ad60542d76f Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Tue, 10 Jul 2018 17:30:34 +0200
Subject: [PATCH] TRITON-556 Update coal-setup.sh

---
 tools/coal-setup.sh | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tools/coal-setup.sh b/tools/coal-setup.sh
index 2e12da5..fb13888 100755
--- a/tools/coal-setup.sh
+++ b/tools/coal-setup.sh
@@ -185,7 +185,12 @@ fi
 echo "# Enable volapi"
 sdcadm post-setup volapi
 sdcadm experimental nfs-volumes cloudapi
+sdcadm experimental nfs-volumes cloudapi-automount
 sdcadm experimental nfs-volumes docker
+sdcadm experimental nfs-volumes docker-automount
+
+echo "# Enable cns"
+sdcadm post-setup cns
 
 if ! $(nictagadm exists sdc_underlay 2>/dev/null); then
     echo "# Enable external and underlay nics"
-- 
2.21.0

