commit 9e979707091997c40fcf01e7c949d1513dda1d7b (refs/changes/10/3210/1)
Author: Mike Zeller <mike.zeller@joyent.com>
Date:   2018-01-12T13:53:08-08:00 (1 year, 9 months ago)
    
    CNAPI-734 updating sysinfo is missing an argument

diff --git a/lib/ur.js b/lib/ur.js
index 7716b54..8891540 100644
--- a/lib/ur.js
+++ b/lib/ur.js
@@ -87,7 +87,8 @@ Ur.prototype.serverSysinfo = function (uuid, opts, callback) {
         'timeoutSeconds': opts.timeoutSeconds,
         'message': {
             'type': 'file',
-            'file': '/usr/bin/sysinfo'
+            'file': '/usr/bin/sysinfo',
+            'args': ['-f']
         }
     };
     opts.uuid = uuid;
