From 9e979707091997c40fcf01e7c949d1513dda1d7b Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike.zeller@joyent.com>
Date: Fri, 12 Jan 2018 13:53:08 -0800
Subject: [PATCH] CNAPI-734 updating sysinfo is missing an argument

---
 lib/ur.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/ur.js b/lib/ur.js
index 7716b54..8891540 100644
--- a/lib/ur.js
+++ b/lib/ur.js
@@ -87,7 +87,8 @@ Ur.prototype.serverSysinfo = function (uuid, opts, callback) {
         'timeoutSeconds': opts.timeoutSeconds,
         'message': {
             'type': 'file',
-            'file': '/usr/bin/sysinfo'
+            'file': '/usr/bin/sysinfo',
+            'args': ['-f']
         }
     };
     opts.uuid = uuid;
-- 
2.21.0

