commit 1ff5a2569c5b2a03e1e5e4e750f715ce6468ffc7
Author: John Levon <john.levon@joyent.com>
Date:   2019-06-28T12:18:59+00:00 (3 months ago)
    
    MANTA-2843 Manta zones lacking CPU cap thwart provisioning

diff --git a/deps/eng b/deps/eng
index d25b8fc..940d8fb 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
+Subproject commit 940d8fbbef6f4d0e6a9e45409fc4646eda8255e8
diff --git a/lib/post-setup/dev-headnode-prov.js b/lib/post-setup/dev-headnode-prov.js
index 22a855e..8a5e5ad 100644
--- a/lib/post-setup/dev-headnode-prov.js
+++ b/lib/post-setup/dev-headnode-prov.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -44,7 +44,8 @@ function makeHeadnodeProvisionable(opts, cb) {
             });
         },
         function tweakCnapiConfig(ctx, next) {
-            if (ctx.cnapi.metadata.ALLOC_FILTER_HEADNODE === false &&
+            if (ctx.cnapi.metadata.ALLOC_FILTER_CAPNESS === false &&
+                ctx.cnapi.metadata.ALLOC_FILTER_HEADNODE === false &&
                 ctx.cnapi.metadata.ALLOC_FILTER_MIN_RESOURCES === false) {
                 progress('CNAPI is already configured to allow headnode ' +
                     'provisioning and over-provisioning');
@@ -57,6 +58,7 @@ function makeHeadnodeProvisionable(opts, cb) {
                 ' and over-provisioning');
             var update = {
                 metadata: {
+                    ALLOC_FILTER_CAPNESS: false,
                     ALLOC_FILTER_HEADNODE: false,
                     ALLOC_FILTER_MIN_RESOURCES: false
                 }
@@ -133,8 +135,8 @@ function do_dev_headnode_prov(subcmd, opts, args, cb) {
 do_dev_headnode_prov.help = (
     'Make the headnode provisionable, for development and testing.\n' +
     '\n' +
-    'This is done via `ALLOC_FILTER_HEADNODE` and \n' +
-    '`ALLOC_FILTER_MIN_RESOURCES` SAPI configuration of the CNAPI service.\n' +
+    'This is done via `ALLOC_FILTER_CAPNESS`, `ALLOC_FILTER_HEADNODE`, and \n' +
+    '`ALLOC_FILTER_MIN_RESOURCES` SAPI parameters of the CNAPI service.\n' +
     'See https://github.com/joyent/sdc-cnapi/blob/master/docs/index.md#sapi-configuration\n' +
     '\n' +
     'Usage:\n' +
diff --git a/man/man1/sdcadm.1.ronn b/man/man1/sdcadm.1.ronn
index f353931..372e7ae 100644
--- a/man/man1/sdcadm.1.ronn
+++ b/man/man1/sdcadm.1.ronn
@@ -573,10 +573,11 @@ Create the "cns" service and a first instance.
 
 ### sdcadm post-setup dev-headnode-prov
 
-Make the headnode provisionable, for development and testing.
+Make the headnode provisionable, for development and testing. This allows
+a Manta deployment to co-exist with standard Triton provisioning.
 
-This is done via `ALLOC_FILTER_HEADNODE` and `ALLOC_FILTER_MIN_RESOURCES`
-SAPI configuration of the CNAPI service. See
+This is done via `ALLOC_FILTER_CAPNESS`, `ALLOC_FILTER_HEADNODE`,
+and `ALLOC_FILTER_MIN_RESOURCES` SAPI parameters of the CNAPI service. See
 [SAPI configuration](https://github.com/joyent/sdc-cnapi/blob/master/docs/index.md#sapi-configuration )
 
 ### sdcadm platform \[options\] COMMAND \[args...\]
@@ -1135,4 +1136,4 @@ Enables/disables support for various NFS volumes features.
 
 ## COPYRIGHT
 
-sdcadm Copyright (c) 2018 Joyent Inc., All rights reserved.
+sdcadm Copyright 2019 Joyent, Inc.
diff --git a/test/post-setup.test.js b/test/post-setup.test.js
index 08e9b45..2d94a00 100644
--- a/test/post-setup.test.js
+++ b/test/post-setup.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -185,7 +185,8 @@ test('sdcadm post-setup dev-headnode-prov', function (t) {
 
             var svc = JSON.parse(stdout)[0];
 
-            if (svc.metadata.ALLOC_FILTER_HEADNODE === false &&
+            if (svc.metadata.ALLOC_FILTER_CAPNESS === false &&
+                svc.metadata.ALLOC_FILTER_HEADNODE === false &&
                 svc.metadata.ALLOC_FILTER_MIN_RESOURCES === false) {
                 t.end();
                 return;
