commit 2020a42ce18b140ab7dde08b1e9eb4a71d18e9d5 (refs/changes/69/3969/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-04-30T13:20:07-07:00 (1 year, 5 months ago)
    
    MANTA-3672 want support for cross-account RBAC role membership

diff --git a/lib/server/server.js b/lib/server/server.js
index e1c8031..4cc7308 100644
--- a/lib/server/server.js
+++ b/lib/server/server.js
@@ -163,12 +163,12 @@ function Server(opts) {
     server.get({
         name: 'getAccountByUuid',
         path: '/accounts/:accountid'
-    }, [getAccount, sendAuth]);
+    }, [getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getAccount',
         path: '/accounts'
-    }, [getAccountUuid, getAccount, sendAuth]);
+    }, [getAccountUuid, getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getUserByUuid',
@@ -383,7 +383,11 @@ function getUser(req, res, next) {
 
 
 function getRoles(req, res, next) {
-    var roles = req.auth.user.roles || [];
+    var roles;
+    if (req.auth.user)
+        roles = req.auth.user.roles || [];
+    else
+        roles = req.auth.account.roles || [];
     req.log.debug({roles: roles}, 'getRoles handler: entered');
     lib.getRoles({
         roles: roles,
