From 2020a42ce18b140ab7dde08b1e9eb4a71d18e9d5 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 30 Apr 2018 13:20:07 -0700
Subject: [PATCH] MANTA-3672 want support for cross-account RBAC role
 membership

---
 lib/server/server.js | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/lib/server/server.js b/lib/server/server.js
index e1c8031..4cc7308 100644
--- a/lib/server/server.js
+++ b/lib/server/server.js
@@ -163,12 +163,12 @@ function Server(opts) {
     server.get({
         name: 'getAccountByUuid',
         path: '/accounts/:accountid'
-    }, [getAccount, sendAuth]);
+    }, [getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getAccount',
         path: '/accounts'
-    }, [getAccountUuid, getAccount, sendAuth]);
+    }, [getAccountUuid, getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getUserByUuid',
@@ -383,7 +383,11 @@ function getUser(req, res, next) {
 
 
 function getRoles(req, res, next) {
-    var roles = req.auth.user.roles || [];
+    var roles;
+    if (req.auth.user)
+        roles = req.auth.user.roles || [];
+    else
+        roles = req.auth.account.roles || [];
     req.log.debug({roles: roles}, 'getRoles handler: entered');
     lib.getRoles({
         roles: roles,
-- 
2.21.0

