From e98293c5de260d7aa89410fdaa054326a0af364c Mon Sep 17 00:00:00 2001
From: Nick Zivkovic <nick.zivkovic@joyent.com>
Date: Fri, 22 Jun 2018 19:52:20 +0000
Subject: [PATCH] TRITON-540 Need to update the ufds-watcher submodule again

---
 deps/sdc-napi-ufds-watcher | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/sdc-napi-ufds-watcher b/deps/sdc-napi-ufds-watcher
index 2d4e757..c41edaa 160000
--- a/deps/sdc-napi-ufds-watcher
+++ b/deps/sdc-napi-ufds-watcher
@@ -1 +1 @@
-Subproject commit 2d4e75711d2212d808bb167b0b3c8d9eecb7d7f1
+Subproject commit c41edaa852414a628e41ae6b133db08e9700ab4b
-- 
2.21.0

