From 481c249d866e94e4d3a5364fe15504b1105cd7f5 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 5 Sep 2018 19:00:53 +0200
Subject: [PATCH] TRITON-752 sdcadm tests broken after TRITON-723 fix

---
 test/check-health.test.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/test/check-health.test.js b/test/check-health.test.js
index ac0ce16..be72aed 100644
--- a/test/check-health.test.js
+++ b/test/check-health.test.js
@@ -8,7 +8,6 @@
  * Copyright 2018, Joyent, Inc.
  */
 
-
 var test = require('tape').test;
 var exec = require('child_process').exec;
 var vasync = require('vasync');
@@ -214,8 +213,8 @@ test('sdcadm check-health with disabled papi', function (t) {
 test('sdcadm check-health -q with disabled papi', function (t) {
     exec('sdcadm check-health -q', function (err, stdout, stderr) {
         t.equal(err && err.code, 1, 'errcode is 1');
-        t.equal('', stdout, 'empty stdout');
-        t.notEqual('', stderr, 'not empty stderr');
+        t.equal('', stderr, 'empty stderr');
+        t.notEqual('', stdout, 'not empty stdout');
         t.end();
     });
 });
@@ -251,8 +250,9 @@ test('check-health when binder is down', function (t) {
                     t.equal(err.killed, false, 'process not killed');
                     t.notEqual(stdout, '', 'empty stdout');
                     t.notEqual(
-                        stderr.indexOf('Binder service seems to be down'), -1,
+                        stdout.indexOf('Binder service seems to be down'), -1,
                         'binder off stderr');
+                    t.notEqual(stderr, '', 'empty stderr');
                     next();
                 });
             },
-- 
2.21.0

