From 1f6403f1229c3a78df47bf9c502b7e89e72e733c Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Thu, 10 Nov 2016 19:54:08 +0000
Subject: [PATCH] OS-5652 platform build needs to work with python 2.7 (fix
 pkgsrc)

---
 pkgsrc/onbld/packlist | 82 +++++++++++++++++++++----------------------
 1 file changed, 41 insertions(+), 41 deletions(-)

diff --git a/pkgsrc/onbld/packlist b/pkgsrc/onbld/packlist
index e42e6772..7697095d 100644
--- a/pkgsrc/onbld/packlist
+++ b/pkgsrc/onbld/packlist
@@ -84,47 +84,47 @@
 ./lib/perl/onbld_elfmod_vertype.pm
 ./lib/perl/onbld_elfmod.pm
 ./lib/python
-./lib/python2.6/onbld/__init__.py
-./lib/python2.6/onbld/__init__.pyc
-./lib/python2.6/onbld/Checks/__init__.py
-./lib/python2.6/onbld/Checks/__init__.pyc
-./lib/python2.6/onbld/Checks/Cddl.py
-./lib/python2.6/onbld/Checks/Cddl.pyc
-./lib/python2.6/onbld/Checks/CmtBlk.py
-./lib/python2.6/onbld/Checks/CmtBlk.pyc
-./lib/python2.6/onbld/Checks/Comments.py
-./lib/python2.6/onbld/Checks/Comments.pyc
-./lib/python2.6/onbld/Checks/Copyright.py
-./lib/python2.6/onbld/Checks/Copyright.pyc
-./lib/python2.6/onbld/Checks/CStyle.py
-./lib/python2.6/onbld/Checks/CStyle.pyc
-./lib/python2.6/onbld/Checks/DbLookups.py
-./lib/python2.6/onbld/Checks/DbLookups.pyc
-./lib/python2.6/onbld/Checks/HdrChk.py
-./lib/python2.6/onbld/Checks/HdrChk.pyc
-./lib/python2.6/onbld/Checks/JStyle.py
-./lib/python2.6/onbld/Checks/JStyle.pyc
-./lib/python2.6/onbld/Checks/Keywords.py
-./lib/python2.6/onbld/Checks/Keywords.pyc
-./lib/python2.6/onbld/Checks/ManLint.py
-./lib/python2.6/onbld/Checks/ManLint.pyc
-./lib/python2.6/onbld/Checks/Mapfile.py
-./lib/python2.6/onbld/Checks/Mapfile.pyc
-./lib/python2.6/onbld/Checks/ProcessCheck.py
-./lib/python2.6/onbld/Checks/ProcessCheck.pyc
-./lib/python2.6/onbld/hgext/__init__.py
-./lib/python2.6/onbld/hgext/__init__.pyc
-./lib/python2.6/onbld/hgext/cdm.py
-./lib/python2.6/onbld/Scm/__init__.py
-./lib/python2.6/onbld/Scm/__init__.pyc
-./lib/python2.6/onbld/Scm/Backup.py
-./lib/python2.6/onbld/Scm/Backup.pyc
-./lib/python2.6/onbld/Scm/Ignore.py
-./lib/python2.6/onbld/Scm/Ignore.pyc
-./lib/python2.6/onbld/Scm/Version.py
-./lib/python2.6/onbld/Scm/Version.pyc
-./lib/python2.6/onbld/Scm/WorkSpace.py
-./lib/python2.6/onbld/Scm/WorkSpace.pyc
+./lib/python2.7/onbld/__init__.py
+./lib/python2.7/onbld/__init__.pyc
+./lib/python2.7/onbld/Checks/__init__.py
+./lib/python2.7/onbld/Checks/__init__.pyc
+./lib/python2.7/onbld/Checks/Cddl.py
+./lib/python2.7/onbld/Checks/Cddl.pyc
+./lib/python2.7/onbld/Checks/CmtBlk.py
+./lib/python2.7/onbld/Checks/CmtBlk.pyc
+./lib/python2.7/onbld/Checks/Comments.py
+./lib/python2.7/onbld/Checks/Comments.pyc
+./lib/python2.7/onbld/Checks/Copyright.py
+./lib/python2.7/onbld/Checks/Copyright.pyc
+./lib/python2.7/onbld/Checks/CStyle.py
+./lib/python2.7/onbld/Checks/CStyle.pyc
+./lib/python2.7/onbld/Checks/DbLookups.py
+./lib/python2.7/onbld/Checks/DbLookups.pyc
+./lib/python2.7/onbld/Checks/HdrChk.py
+./lib/python2.7/onbld/Checks/HdrChk.pyc
+./lib/python2.7/onbld/Checks/JStyle.py
+./lib/python2.7/onbld/Checks/JStyle.pyc
+./lib/python2.7/onbld/Checks/Keywords.py
+./lib/python2.7/onbld/Checks/Keywords.pyc
+./lib/python2.7/onbld/Checks/ManLint.py
+./lib/python2.7/onbld/Checks/ManLint.pyc
+./lib/python2.7/onbld/Checks/Mapfile.py
+./lib/python2.7/onbld/Checks/Mapfile.pyc
+./lib/python2.7/onbld/Checks/ProcessCheck.py
+./lib/python2.7/onbld/Checks/ProcessCheck.pyc
+./lib/python2.7/onbld/hgext/__init__.py
+./lib/python2.7/onbld/hgext/__init__.pyc
+./lib/python2.7/onbld/hgext/cdm.py
+./lib/python2.7/onbld/Scm/__init__.py
+./lib/python2.7/onbld/Scm/__init__.pyc
+./lib/python2.7/onbld/Scm/Backup.py
+./lib/python2.7/onbld/Scm/Backup.pyc
+./lib/python2.7/onbld/Scm/Ignore.py
+./lib/python2.7/onbld/Scm/Ignore.pyc
+./lib/python2.7/onbld/Scm/Version.py
+./lib/python2.7/onbld/Scm/Version.pyc
+./lib/python2.7/onbld/Scm/WorkSpace.py
+./lib/python2.7/onbld/Scm/WorkSpace.pyc
 ./man/man1onbld/bldenv.1onbld
 ./man/man1onbld/bringovercheck.1onbld
 ./man/man1onbld/cddlchk.1onbld
-- 
2.21.0

