commit bee6957d1f29030075695bca560843bb3c43bae7 (refs/changes/08/2708/2)
Author: Sam Gwydir <sam.gwydir@joyent.com>
Date:   2017-10-06T21:30:04+00:00 (2 years ago)
    
    MANATEE-373 Want additional postgres modules for admin purposes

diff --git a/etc/9.6/postgresql.manta.coal.conf b/etc/9.6/postgresql.manta.coal.conf
index 5170bb8..2209619 100644
--- a/etc/9.6/postgresql.manta.coal.conf
+++ b/etc/9.6/postgresql.manta.coal.conf
@@ -142,7 +142,7 @@ shared_buffers = 32MB                   # min 128kB
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/etc/9.6/postgresql.manta.lab.conf b/etc/9.6/postgresql.manta.lab.conf
index 9bd60ee..9714901 100644
--- a/etc/9.6/postgresql.manta.lab.conf
+++ b/etc/9.6/postgresql.manta.lab.conf
@@ -142,7 +142,7 @@ maintenance_work_mem = 1GB              # min 1MB
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/etc/9.6/postgresql.manta.production.conf b/etc/9.6/postgresql.manta.production.conf
index 10f800d..dd6c0b1 100644
--- a/etc/9.6/postgresql.manta.production.conf
+++ b/etc/9.6/postgresql.manta.production.conf
@@ -142,7 +142,7 @@ maintenance_work_mem = 4GB              # min 1MB
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/package.json b/package.json
index 9bbd8a1..11dbc67 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     },
     "dependencies": {
         "manta": "1.2.6",
-        "manatee": "git+https://github.com/joyent/manatee.git#34238c25",
+        "manatee": "git+https://github.com/joyent/manatee.git#6a769d97",
         "carrier": "0.2.0"
     },
     "sdcDependencies": {
