commit 3ab70dfa20d8c1c6e799546afdc2433d522e80c0 (refs/changes/08/2708/3)
Author: Sam Gwydir <sam.gwydir@joyent.com>
Date:   2017-10-06T22:41:48+00:00 (2 years ago)
    
    MANATEE-373 Want additional postgres modules for admin purposes

diff --git a/etc/9.6/postgresql.manta.coal.conf b/etc/9.6/postgresql.manta.coal.conf
index 6d02d7b..d82ef9e 100644
--- a/etc/9.6/postgresql.manta.coal.conf
+++ b/etc/9.6/postgresql.manta.coal.conf
@@ -147,7 +147,7 @@ dynamic_shared_memory_type = sysv       # the default is the first option
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/etc/9.6/postgresql.manta.lab.conf b/etc/9.6/postgresql.manta.lab.conf
index 6341d4a..031ccb2 100644
--- a/etc/9.6/postgresql.manta.lab.conf
+++ b/etc/9.6/postgresql.manta.lab.conf
@@ -147,7 +147,7 @@ dynamic_shared_memory_type = sysv       # the default is the first option
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/etc/9.6/postgresql.manta.production.conf b/etc/9.6/postgresql.manta.production.conf
index 9a650e6..6a09688 100644
--- a/etc/9.6/postgresql.manta.production.conf
+++ b/etc/9.6/postgresql.manta.production.conf
@@ -147,7 +147,7 @@ dynamic_shared_memory_type = sysv       # the default is the first option
 
 #max_files_per_process = 1000           # min 25
                                         # (change requires restart)
-shared_preload_libraries = 'pg_stat_statements' # (change requires restart)
+shared_preload_libraries = 'auto_explain,pg_buffercache,pg_freespacemap,pg_stat_statements,pg_visibility,pgstattuple' # (change requires restart)
 
 # - Cost-Based Vacuum Delay -
 
diff --git a/package.json b/package.json
index 9bbd8a1..11dbc67 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     },
     "dependencies": {
         "manta": "1.2.6",
-        "manatee": "git+https://github.com/joyent/manatee.git#34238c25",
+        "manatee": "git+https://github.com/joyent/manatee.git#6a769d97",
         "carrier": "0.2.0"
     },
     "sdcDependencies": {
