commit e4b9942eb90056392cc52ca238b1eccf572f1c29 (refs/changes/75/2175/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-06-30T10:32:47-07:00 (2 years, 3 months ago)
    
    DOCKER-1076 leftover jsprim module loading breaks tests suite

diff --git a/test/lib/common.js b/test/lib/common.js
index a45b81c..02d9fa1 100644
--- a/test/lib/common.js
+++ b/test/lib/common.js
@@ -17,7 +17,6 @@ var deepEqual = require('deep-equal');
 var difflet = require('difflet');
 var exec = require('child_process').exec;
 var fmt = require('util').format;
-var jsprim = require('jsprim');
 var libuuid = require('libuuid');
 var VError = require('verror').VError;
 
