From 0f575a6380e776f642e8f95951eaf70b0b28816c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 21 Mar 2019 13:46:49 +0000
Subject: [PATCH] MANTA-4191 remove deps/node submodule remnants from mola and
 madtom

---
 .gitmodules | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 43a90e3..54b4439 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -7,9 +7,6 @@
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
 	url = git://github.com/davepacheco/javascriptlint.git
-[submodule "deps/node"]
-	path = deps/node
-	url = git://github.com/joyent/node.git
 [submodule "deps/manta-scripts"]
 	path = deps/manta-scripts
 	url = git://github.com/joyent/manta-scripts.git
-- 
2.21.0

