From d96d135ecc57efc3aa582a90e33862bbbb53d2fd Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Mon, 11 Feb 2019 15:04:24 +0000
Subject: [PATCH] OS-7576 ctfdump -c should include non-root types

---
 usr/src/cmd/ctfdump/ctfdump.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/usr/src/cmd/ctfdump/ctfdump.c b/usr/src/cmd/ctfdump/ctfdump.c
index e241c35b47..ec879fcc06 100644
--- a/usr/src/cmd/ctfdump/ctfdump.c
+++ b/usr/src/cmd/ctfdump/ctfdump.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -982,7 +982,7 @@ ctfdump_source(void)
 		    strerror(errno));
 	}
 
-	if (ctf_type_iter(g_fp, B_FALSE, ctfsrc_collect_types_cb,
+	if (ctf_type_iter(g_fp, B_TRUE, ctfsrc_collect_types_cb,
 	    idnames) == CTF_ERR) {
 		warnx("failed to collect types: %s",
 		    ctf_errmsg(ctf_errno(g_fp)));
-- 
2.21.0

