commit d96d135ecc57efc3aa582a90e33862bbbb53d2fd (refs/changes/61/5561/1)
Author: John Levon <john.levon@joyent.com>
Date:   2019-02-11T15:04:24+00:00 (8 months ago)
    
    OS-7576 ctfdump -c should include non-root types

diff --git a/usr/src/cmd/ctfdump/ctfdump.c b/usr/src/cmd/ctfdump/ctfdump.c
index e241c35b47..ec879fcc06 100644
--- a/usr/src/cmd/ctfdump/ctfdump.c
+++ b/usr/src/cmd/ctfdump/ctfdump.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -982,7 +982,7 @@ ctfdump_source(void)
 		    strerror(errno));
 	}
 
-	if (ctf_type_iter(g_fp, B_FALSE, ctfsrc_collect_types_cb,
+	if (ctf_type_iter(g_fp, B_TRUE, ctfsrc_collect_types_cb,
 	    idnames) == CTF_ERR) {
 		warnx("failed to collect types: %s",
 		    ctf_errmsg(ctf_errno(g_fp)));
