commit e0e711d5a740663af5480daff7b390aff0d1d3d8 (refs/changes/15/3315/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-02-08T20:51:41+00:00 (1 year, 8 months ago)
    
    OS-6594 zoneadmd forgot to setenv _ZONECFG_device_resources

diff --git a/usr/src/cmd/zoneadmd/zoneadmd.c b/usr/src/cmd/zoneadmd/zoneadmd.c
index 5d3adcebc6..2155ec0e5f 100644
--- a/usr/src/cmd/zoneadmd/zoneadmd.c
+++ b/usr/src/cmd/zoneadmd/zoneadmd.c
@@ -22,7 +22,7 @@
 /*
  * Copyright (c) 2003, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright 2014 Nexenta Systems, Inc. All rights reserved.
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  * Copyright (c) 2016 by Delphix. All rights reserved.
  */
 
@@ -866,6 +866,7 @@ setup_subproc_env(boolean_t debug)
 	if ((res = zonecfg_setdevent(snap_hndl)) != Z_OK)
 		goto done;
 
+	dev_resources[0] = '\0';
 	while (zonecfg_getdevent(snap_hndl, &dtab) == Z_OK) {
 		struct zone_res_attrtab *rap;
 		char *match;
@@ -874,6 +875,7 @@ setup_subproc_env(boolean_t debug)
 
 		(void) strlcat(dev_resources, match, sizeof (dev_resources));
 		(void) strlcat(dev_resources, " ", sizeof (dev_resources));
+		set_zonecfg_env(RSRC_DEV, match, "path", match);
 
 		for (rap = dtab.zone_dev_attrp; rap != NULL;
 		    rap = rap->zone_res_attr_next)
@@ -881,6 +883,14 @@ setup_subproc_env(boolean_t debug)
 			    rap->zone_res_attr_name, rap->zone_res_attr_value);
 	}
 
+	for (char *p = dev_resources; *p != '\0'; p++) {
+		if (*p == '-') {
+			*p = '_';
+		}
+	}
+
+	(void) setenv("_ZONECFG_device_resources", dev_resources, 1);
+
 	(void) zonecfg_enddevent(snap_hndl);
 
 	if ((res = zonecfg_setattrent(snap_hndl)) != Z_OK)
