commit 051df89df9bf815914a681b2026e572a2e22d076 (refs/changes/89/4889/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-09-27T15:22:34+00:00 (1 year ago)
    
    OS-7225 KVM tries to include wrong kvm_para.h

diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index ed1aaab..26cbb53 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -22,8 +22,12 @@
 #include <pthread.h>
 #include <sys/utsname.h>
 #ifdef CONFIG_KVM_PARA
+#ifdef __sun__
+#include <sys/kvm_para.h>
+#else
 #include <linux/kvm_para.h>
 #endif
+#endif
 #include <sys/ioctl.h>
 
 #include "kvm.h"
