commit 582f0fd2dc7e721361281c595415f5f6e24edc6f (refs/changes/89/4889/2)
Author: John Levon <john.levon@joyent.com>
Date:   2018-09-27T15:24:02+00:00 (1 year ago)
    
    OS-7225 KVM tries to include wrong kvm_para.h

diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index ed1aaab..a732d28 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -2,7 +2,7 @@
  * qemu/kvm integration, x86 specific code
  *
  * Copyright (C) 2006-2008 Qumranet Technologies
- * Portions Copyright 2011 Joyent, Inc.
+ * Portions Copyright 2018 Joyent, Inc.
  *
  * Licensed under the terms of the GNU GPL version 2 or higher.
  */
@@ -22,8 +22,12 @@
 #include <pthread.h>
 #include <sys/utsname.h>
 #ifdef CONFIG_KVM_PARA
+#ifdef __sun__
+#include <sys/kvm_para.h>
+#else
 #include <linux/kvm_para.h>
 #endif
+#endif
 #include <sys/ioctl.h>
 
 #include "kvm.h"
