From ff155387103dea553438cbc670160ddb3c757cda Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Fri, 28 Sep 2018 09:54:26 +0000
Subject: [PATCH] OS-7225 KVM tries to include wrong kvm_para.h

---
 qemu-kvm-x86.c    | 4 ++--
 target-i386/kvm.c | 6 +-----
 2 files changed, 3 insertions(+), 7 deletions(-)

diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index ed1aaab..cb3dad0 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -2,7 +2,7 @@
  * qemu/kvm integration, x86 specific code
  *
  * Copyright (C) 2006-2008 Qumranet Technologies
- * Portions Copyright 2011 Joyent, Inc.
+ * Portions Copyright 2018 Joyent, Inc.
  *
  * Licensed under the terms of the GNU GPL version 2 or higher.
  */
@@ -22,7 +22,7 @@
 #include <pthread.h>
 #include <sys/utsname.h>
 #ifdef CONFIG_KVM_PARA
-#include <linux/kvm_para.h>
+#include <sys/kvm_para.h>
 #endif
 #include <sys/ioctl.h>
 
diff --git a/target-i386/kvm.c b/target-i386/kvm.c
index bb3f684..8f28911 100644
--- a/target-i386/kvm.c
+++ b/target-i386/kvm.c
@@ -10,7 +10,7 @@
  * This work is licensed under the terms of the GNU GPL, version 2 or later.
  * See the COPYING file in the top-level directory.
  *
- * Portions Copyright 2011 Joyent, Inc.
+ * Portions Copyright 2018 Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -36,11 +36,7 @@
 #include "kvm_x86.h"
 
 #ifdef CONFIG_KVM_PARA
-#ifdef __sun__
 #include <sys/kvm_para.h>
-#else
-#include <linux/kvm_para.h>
-#endif
 #endif
 
 #ifdef __sun__
-- 
2.21.0

