From 582f0fd2dc7e721361281c595415f5f6e24edc6f Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Thu, 27 Sep 2018 15:24:02 +0000
Subject: [PATCH] OS-7225 KVM tries to include wrong kvm_para.h

---
 qemu-kvm-x86.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index ed1aaab..a732d28 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -2,7 +2,7 @@
  * qemu/kvm integration, x86 specific code
  *
  * Copyright (C) 2006-2008 Qumranet Technologies
- * Portions Copyright 2011 Joyent, Inc.
+ * Portions Copyright 2018 Joyent, Inc.
  *
  * Licensed under the terms of the GNU GPL version 2 or higher.
  */
@@ -22,8 +22,12 @@
 #include <pthread.h>
 #include <sys/utsname.h>
 #ifdef CONFIG_KVM_PARA
+#ifdef __sun__
+#include <sys/kvm_para.h>
+#else
 #include <linux/kvm_para.h>
 #endif
+#endif
 #include <sys/ioctl.h>
 
 #include "kvm.h"
-- 
2.21.0

