From 8f50cc802b39e2e09aced9751cfb3da0199e5ceb Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 29 Sep 2016 14:06:35 -0700
Subject: [PATCH] joyent/node-cueball#27 dynamic resolver mode broken

---
 lib/resolver.js | 3 ++-
 package.json    | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/lib/resolver.js b/lib/resolver.js
index 0b7bb66..85e208a 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -304,13 +304,14 @@ CueBallDNSResolver.prototype.state_check_ns = function (S) {
 		this.r_bootstrap =
 		    CueBallDNSResolver.bootstrapResolvers[notIp[0]];
 		if (this.r_bootstrap === undefined) {
-			this.r_bootstrap = new CueBallDNSResolver({
+			var res = new CueBallDNSResolver({
 				domain: notIp[0],
 				service: '_dns._udp',
 				defaultPort: 53,
 				log: this.r_log,
 				recovery: this.r_recovery
 			});
+			this.r_bootstrap = res.r_fsm;
 			CueBallDNSResolver.bootstrapResolvers[notIp[0]] =
 			    this.r_bootstrap;
 		}
diff --git a/package.json b/package.json
index 9448f88..ad16a87 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "0.5.0",
+  "version": "0.5.1",
   "description": "",
   "main": "lib/index.js",
   "dependencies": {
-- 
2.21.0

