From 1fd45ba80830c12375f31a0f625d38f97e226eaa Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 16 Nov 2016 01:48:54 +0000
Subject: [PATCH] MORAY-350 Moray throws ReferenceError: InternalError is not
 defined

---
 lib/pg.js | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/pg.js b/lib/pg.js
index eb9a693..a85d674 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -25,6 +25,7 @@ var dtrace = require('./dtrace');
 
 var mod_errors = require('./errors');
 var ConnectTimeoutError = mod_errors.ConnectTimeoutError;
+var InternalError = mod_errors.InternalError;
 var InvalidIndexDefinitionError = mod_errors.InvalidIndexDefinitionError;
 var QueryTimeoutError = mod_errors.QueryTimeoutError;
 var UniqueAttributeError = mod_errors.UniqueAttributeError;
-- 
2.21.0

