commit 1fd45ba80830c12375f31a0f625d38f97e226eaa (refs/changes/06/906/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-11-16T01:49:51+00:00 (2 years, 11 months ago)
    
    MORAY-350 Moray throws ReferenceError: InternalError is not defined

diff --git a/lib/pg.js b/lib/pg.js
index eb9a693..a85d674 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -25,6 +25,7 @@ var dtrace = require('./dtrace');
 
 var mod_errors = require('./errors');
 var ConnectTimeoutError = mod_errors.ConnectTimeoutError;
+var InternalError = mod_errors.InternalError;
 var InvalidIndexDefinitionError = mod_errors.InvalidIndexDefinitionError;
 var QueryTimeoutError = mod_errors.QueryTimeoutError;
 var UniqueAttributeError = mod_errors.UniqueAttributeError;
