commit aacead0cfc6ff1b732bad12e641406253859d278 (refs/changes/33/533/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2016-09-26T17:14:22+00:00 (3 years ago)
    
    backout PORTAL-2165: needs more work

diff --git a/lib/cloudapi.js b/lib/cloudapi.js
index ffd3581..3a415c1 100644
--- a/lib/cloudapi.js
+++ b/lib/cloudapi.js
@@ -170,8 +170,6 @@ function _addToQuery(req, obj) {
  * @param {Object} options object (required):
  *        - {String} url (required) CloudAPI location.
  *        - {String} account (optional) the login name to use (default my).
- *        - {Object} logFields (optional) the additional fields to include
- *          into the log records.
  *        - {Number} logLevel (optional) an enum value for the logging level.
  *        - {String} version (optional) api version (default ~7.2).
  *        - {Function} sign (required) callback function to use for signing
@@ -189,17 +187,10 @@ function CloudAPI(options) {
     assert.string(options.url, 'options.url');
     assert.func(options.sign, 'options.sign');
     assert.optionalString(options.account, 'options.account');
-    assert.optionalObject(options.logFields, 'options.logFields');
 
-    if (options.logFields) {
-        options.log = log.child(options.logFields);
-    } else {
-        options.log = log.child();
-    }
-
-    if (options.logLevel) {
-        options.log.level(options.logLevel);
-    }
+    if (options.logLevel)
+        log.level(options.logLevel);
+    options.log = log;
 
     if (options.api_version)
         options.version = options.api_version;
