From aacead0cfc6ff1b732bad12e641406253859d278 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Mon, 26 Sep 2016 17:14:22 +0000
Subject: [PATCH] backout PORTAL-2165: needs more work

---
 lib/cloudapi.js | 15 +++------------
 1 file changed, 3 insertions(+), 12 deletions(-)

diff --git a/lib/cloudapi.js b/lib/cloudapi.js
index ffd3581..3a415c1 100644
--- a/lib/cloudapi.js
+++ b/lib/cloudapi.js
@@ -170,8 +170,6 @@ function _addToQuery(req, obj) {
  * @param {Object} options object (required):
  *        - {String} url (required) CloudAPI location.
  *        - {String} account (optional) the login name to use (default my).
- *        - {Object} logFields (optional) the additional fields to include
- *          into the log records.
  *        - {Number} logLevel (optional) an enum value for the logging level.
  *        - {String} version (optional) api version (default ~7.2).
  *        - {Function} sign (required) callback function to use for signing
@@ -189,17 +187,10 @@ function CloudAPI(options) {
     assert.string(options.url, 'options.url');
     assert.func(options.sign, 'options.sign');
     assert.optionalString(options.account, 'options.account');
-    assert.optionalObject(options.logFields, 'options.logFields');
 
-    if (options.logFields) {
-        options.log = log.child(options.logFields);
-    } else {
-        options.log = log.child();
-    }
-
-    if (options.logLevel) {
-        options.log.level(options.logLevel);
-    }
+    if (options.logLevel)
+        log.level(options.logLevel);
+    options.log = log;
 
     if (options.api_version)
         options.version = options.api_version;
-- 
2.21.0

