From a7bb4b22886fad6ccdb0944cb24ff1085f3eec77 Mon Sep 17 00:00:00 2001
From: Brittany Wald <brittany.wald@joyent.com>
Date: Fri, 8 Dec 2017 02:09:14 +0000
Subject: [PATCH] MANTA-3514 electric-moray error message wording change test
 failure

---
 test/batch.test.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/test/batch.test.js b/test/batch.test.js
index da78fcb..7f9c163 100644
--- a/test/batch.test.js
+++ b/test/batch.test.js
@@ -443,7 +443,8 @@ test('same key transforms to different values', function (t) {
 
 test('bad requests array: empty array', function (t) {
     runBadBatch(this.client, t, [],
-        'must specify an array with at least one request');
+        'batch expects "requests" (args[0]) to be an array with at least one' +
+        ' request but received an empty array');
 });
 
 
-- 
2.21.0

