commit a7bb4b22886fad6ccdb0944cb24ff1085f3eec77 (refs/changes/64/3064/1)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2017-12-08T02:09:14+00:00 (1 year, 10 months ago)
    
    MANTA-3514 electric-moray error message wording change test failure

diff --git a/test/batch.test.js b/test/batch.test.js
index da78fcb..7f9c163 100644
--- a/test/batch.test.js
+++ b/test/batch.test.js
@@ -443,7 +443,8 @@ test('same key transforms to different values', function (t) {
 
 test('bad requests array: empty array', function (t) {
     runBadBatch(this.client, t, [],
-        'must specify an array with at least one request');
+        'batch expects "requests" (args[0]) to be an array with at least one' +
+        ' request but received an empty array');
 });
 
 
