commit 86364bcc063404b551cba1c7b1e17efa652f305d (refs/changes/63/3263/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-01-25T14:24:38-05:00 (1 year, 8 months ago)
    
    RELENG-612 prefer (git+)https urls

diff --git a/.gitmodules b/.gitmodules
index ef29a86..04abb3e 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,12 +1,12 @@
 [submodule "deps/restdown"]
 	path = deps/restdown
-	url = git://github.com/trentm/restdown.git
+	url = https://github.com/trentm/restdown.git
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
-	url = git://github.com/davepacheco/javascriptlint.git
+	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/davepacheco/jsstyle.git
 [submodule "deps/sdc-scripts"]
 	path = deps/sdc-scripts
-	url = git@github.com:joyent/sdc-scripts.git
+	url = https://github.com/joyent/sdc-scripts.git
