From aed36f67118a4449cd2fd31b17cece618844eda8 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 7 Dec 2016 23:24:15 +0000
Subject: [PATCH] joyent/node-lockfd#7 node v4 support (v2.x)

---
 CHANGES.md   | 4 ++++
 Makefile     | 9 ++++++---
 package.json | 4 ++--
 3 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 71f113a..ff9bf09 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,10 @@
 
 (nothing yet)
 
+## 2.0.1
+
+- joyent/node-lockfd#7 Node v4 support
+
 ## 2.0.0
 
 - joyent/node-lockfd#6 Re-branding the 1.1.0 release. The flock(3C)
diff --git a/Makefile b/Makefile
index c6a209e..de581d3 100644
--- a/Makefile
+++ b/Makefile
@@ -41,10 +41,13 @@ TOP             := $(shell pwd)
 #
 
 .PHONY: all
-all: $(LOCKFD_BINDING) ./node_modules
+all: $(LOCKFD_BINDING) | ./node_modules
 	$(NPM) install
 
-$(LOCKFD_BINDING):
+./node_modules:
+	$(NPM) install
+
+$(LOCKFD_BINDING): | ./node_modules
 	cd src && make
 
 $(ESLINT):
@@ -53,7 +56,7 @@ $(ESLINT):
 $(TAPE):
 	$(NPM) install
 
-CLEAN_FILES += $(TAPE) ./node_modules/tape $(LOCKFD_BINDING)
+CLEAN_FILES += $(LOCKFD_BINDING) src/lockfd.o src/v8plus_errno.h ./node_modules
 
 .PHONY: test
 test: $(TAPE)
diff --git a/package.json b/package.json
index 4934b96..6468bcb 100644
--- a/package.json
+++ b/package.json
@@ -1,12 +1,12 @@
 {
   "name": "lockfd",
   "description": "thin wrapper around fcntl(F_SETLK) and flock()",
-  "version": "2.0.0",
+  "version": "2.0.1",
   "author": "Joshua M. Clulow <jmc@joyent.com>",
   "repository": "git://github.com/joyent/node-lockfd.git",
   "main": "./lib/index.js",
   "dependencies": {
-    "v8plus": "~0.3.1"
+    "v8plus": "~1.0.2"
   },
   "scripts": {
     "postinstall": "cd src && gmake $(eval echo ${MAKE_OVERRIDES})",
-- 
2.21.0

