From 0c7389ff18a2b060754ea055547bf7c78e329144 Mon Sep 17 00:00:00 2001
From: Jared Morrow <jm@joyent.com>
Date: Thu, 16 Aug 2018 14:39:23 -0600
Subject: [PATCH] MANTA-3877 Add no_reindex option to manta bucket creation for
 GC trigger update

---
 lib/moray.js | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/lib/moray.js b/lib/moray.js
index 35f2981..f90476e 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -35,7 +35,15 @@ var utils = require('./utils');
 var sprintf = util.format;
 
 var BUCKET = process.env.MANTA_RING_BUCKET || 'manta';
+/*
+ * NOTE: Care must be taken when deploying a incremented version of the manta
+ * bucket for large databases. Currently `no_reindex = true` is being
+ * passed into createBucket below. If new columns are added to the manta
+ * bucket, this reindex option should be removed or at least revisited.
+ * Do not change BUCKET_VERSION without discussing a deployment strategy.
+ */
 var BUCKET_VERSION = 2;
+
 /* JSSTYLED */
 var ROOT_RE = /^\/[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\/stor$/;
 var SCHEMA = {
@@ -109,6 +117,7 @@ function setupMantaBuckets(log, client, cb) {
             opts: {
                 index: SCHEMA,
                 post: POST,
+                no_reindex: true, // See comment above BUCKET_VERSION definition
                 options: {
                     version: BUCKET_VERSION
                 }
-- 
2.21.0

