commit bba7c2acae1a6a6a9105a7fd9af34b953e3683b1
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T14:50:31-05:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/Makefile b/Makefile
index bd361e9..e826f15 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ TOP =			$(PWD)
 # hermes-actor runs, and it also happens to work with hermes and hermes-proxy
 # even though they run in the sdc zone.
 #
-NODE_PREBUILT_VERSION = v6.15.1
+NODE_PREBUILT_VERSION = v6.17.0
 NODE_PREBUILT_TAG = gz
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_IMAGE =   18b094b0-eb01-11e5-80c1-175dac7ddf02
