From bba7c2acae1a6a6a9105a7fd9af34b953e3683b1 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 5 Mar 2019 14:50:31 -0500
Subject: [PATCH] TOOLS-2186 sdcnode updates for Feburary 2019 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index bd361e9..e826f15 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,7 @@ TOP =			$(PWD)
 # hermes-actor runs, and it also happens to work with hermes and hermes-proxy
 # even though they run in the sdc zone.
 #
-NODE_PREBUILT_VERSION = v6.15.1
+NODE_PREBUILT_VERSION = v6.17.0
 NODE_PREBUILT_TAG = gz
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_IMAGE =   18b094b0-eb01-11e5-80c1-175dac7ddf02
-- 
2.21.0

