From d5424d0aa42aea0142f0e64da5ce0269a103884c Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Mon, 10 Apr 2017 22:57:29 +0000
Subject: [PATCH] OS-6045 usb_ep_ss_comp_descr_t(9S) points to the void

---
 manifest | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/manifest b/manifest
index 3b95d09fdf..4a07938dac 100644
--- a/manifest
+++ b/manifest
@@ -20359,7 +20359,7 @@ f usr/share/man/man9s/usb_dev_qlf_descr.9s 0444 root bin
 f usr/share/man/man9s/usb_ep_descr.9s 0444 root bin
 f usr/share/man/man9s/usb_ep_xdescr.9s 0444 root bin
 f usr/share/man/man9s/usb_ep_ss_comp_descr.9s 0444 root bin
-s usr/share/man/man9s/usb_ep_ss_comp_descr_t.9s=usb_ep_ss_comp_descr
+s usr/share/man/man9s/usb_ep_ss_comp_descr_t.9s=usb_ep_ss_comp_descr.9s
 f usr/share/man/man9s/usb_if_descr.9s 0444 root bin
 s usr/share/man/man9s/usb_intr_req_t.9s=usb_intr_req.9s
 f usr/share/man/man9s/usb_intr_req.9s 0444 root bin
-- 
2.21.0

