From ca737d0824ee9ccbcc836bbcb6779188876602b0 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Fri, 10 Feb 2017 23:20:10 +0000
Subject: [PATCH] MANTA-3161 sdc-manta.git should run catest with the right
 node

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 74b9d2f..316845a 100644
--- a/Makefile
+++ b/Makefile
@@ -92,7 +92,7 @@ check:: $(NODE_EXEC)
 
 .PHONY: test
 test: | $(CATEST)
-	$(CATEST) -a
+	PATH="$(TOP)/build/node/bin:$$PATH" $(CATEST) -a
 
 $(CATEST): deps/catest/.git
 
-- 
2.21.0

