From e0a1542b9fb6790bdf5bc4c630fef30bc29afca4 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 20 Dec 2018 15:41:14 +0100
Subject: [PATCH] OS-7408 want native libvmm (fix deps)

---
 usr/src/lib/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/usr/src/lib/Makefile b/usr/src/lib/Makefile
index e04e4d82cd..8e48dda5ed 100644
--- a/usr/src/lib/Makefile
+++ b/usr/src/lib/Makefile
@@ -690,6 +690,7 @@ libtsnet:	libtsol libsecdb
 libtsol:	libsecdb
 libuuid:	libdlpi
 libv12n:	libds libuuid
+libvmm:		libvmmapi
 libvolmgt:	libadm
 libvrrpadm:	libdladm libscf
 libvscan:	libscf libsecdb
-- 
2.21.0

