commit e0a1542b9fb6790bdf5bc4c630fef30bc29afca4 (refs/changes/73/5273/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-12-20T15:41:14+01:00 (10 months ago)
    
    OS-7408 want native libvmm (fix deps)

diff --git a/usr/src/lib/Makefile b/usr/src/lib/Makefile
index e04e4d82cd..8e48dda5ed 100644
--- a/usr/src/lib/Makefile
+++ b/usr/src/lib/Makefile
@@ -690,6 +690,7 @@ libtsnet:	libtsol libsecdb
 libtsol:	libsecdb
 libuuid:	libdlpi
 libv12n:	libds libuuid
+libvmm:		libvmmapi
 libvolmgt:	libadm
 libvrrpadm:	libdladm libscf
 libvscan:	libscf libsecdb
