commit 6c82e344fe6eb8b884db5e75c5f0afc454504962 (refs/changes/57/1757/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-04-06T17:40:23+02:00 (2 years, 6 months ago)
    
    TOOLS-1748 sdcadm reprovision timeout after TOOLS-1715

diff --git a/lib/procedures/shared.js b/lib/procedures/shared.js
index f0632a4..bc4f2a7 100644
--- a/lib/procedures/shared.js
+++ b/lib/procedures/shared.js
@@ -820,7 +820,7 @@ function reprovisionRemote(opts, callback) {
     ur.exec({
         sdcadm: sdcadm,
         server: opts.server,
-        execTimeout: 300,
+        execTimeout: 300 * 1000,
         cmd: format('/usr/sbin/vmadm reprovision %s <<< \'%s\'',
                 opts.zonename, reproJSON)
     }, function (execErr, result) {
diff --git a/lib/ur.js b/lib/ur.js
index 281985e..3165a12 100644
--- a/lib/ur.js
+++ b/lib/ur.js
@@ -66,7 +66,7 @@ function execOnAllNodes(opts, cb) {
  *      - cmd {String} The command/script to run.
  *      - server {UUID} The UUID of the server on which to run.
  *      - log {Bunyan Logger} Optional.
- *      - execTimeout {Number} Optional. Default 30s.
+ *      - execTimeout {Number} in milliseconds. Optional. Default 30s.
  *      - env {Object} Optional. Default '{}'.
  * @param cb {Function} `function (err, result)`
  */
