commit db78d8777fc45a304c6e92d84cf7f96c0f663d46 (refs/changes/66/2266/1)
Author: Jared Morrow <jm@joyent.com>
Date:   2017-07-24T12:30:11-06:00 (2 years, 2 months ago)
    
    MANTA-3109 manta log uploader retries multiple times in parallel

diff --git a/deps/manta-scripts b/deps/manta-scripts
index cd72e80..4b86dd1 160000
--- a/deps/manta-scripts
+++ b/deps/manta-scripts
@@ -1 +1 @@
-Subproject commit cd72e801576de5a4b1aec35a7f79409d3f1eebb5
+Subproject commit 4b86dd132d68cb177f066a83d90b8fdc9ac2cde9
