From e93c45f719fb8b91617e1031fa730eeffe7f9847 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <josh@wilsdon.ca>
Date: Fri, 16 Mar 2018 20:25:17 -0700
Subject: [PATCH] OS-6782 add 'Bhyve Max Vcpus' to sysinfo

---
 src/sysinfo | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/sysinfo b/src/sysinfo
index c648e182..67873b78 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -200,9 +200,15 @@ function get_smartos_cpu_info()
 function get_bhyve_capability()
 {
     Bhyve_Capable="false"
+    Bhyve_Max_Vcpus="0"
+
     if [[ -x /usr/lib/brand/bhyve/bhhwcompat ]]; then
         if /usr/lib/brand/bhyve/bhhwcompat; then
             Bhyve_Capable="true"
+            vcpus=$(/usr/lib/brand/bhyve/bhhwcompat -c | /usr/bin/grep "^[0-9]*$")
+            if [[ -n ${vcpus} ]]; then
+                Bhyve_Max_Vcpus=${vcpus}
+            fi
         fi
     fi
 }
@@ -666,6 +672,7 @@ CPU_Type='${CPU_Version}'
 CPU_Virtualization='${CPU_Virtualization}'
 CPU_Physical_Cores=${CPU_Count}
 Bhyve_Capable='${Bhyve_Capable}'
+Bhyve_Max_Vcpus=${Bhyve_Max_Vcpus}
 Nic_Tags=${NicTagList}
 Setup='${Setup_complete}'
 END
@@ -799,6 +806,7 @@ END
   "Setup": "${Setup_complete}",
   "VM Capable": ${VM_Capable},
   "Bhyve Capable": ${Bhyve_Capable},
+  "Bhyve Max Vcpus": ${Bhyve_Max_Vcpus},
   "CPU Type": "${CPU_Version}",
   "CPU Virtualization": "${CPU_Virtualization}",
   "CPU Physical Cores": ${CPU_Count},
-- 
2.21.0

