commit 7b4b7da8e9d558f4495ec78d553390f71f372017 (refs/changes/58/858/2)
Author: Cameron Watters <cameron.watters@joyent.com>
Date:   2016-11-08T09:41:21-07:00 (2 years, 11 months ago)
    
    DOCKER-967 properly order extraHosts entries in /etc/hosts

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 4f7bacab..22e78bd0 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -2177,7 +2177,7 @@ function createHostConfFileMounts(vmobj, opts, log, callback) {
     });
     if (extraHosts) {
         extraHosts.forEach(function (h) {
-            hostsContents += h[0] + '\t' + h[1] + '\n';
+            hostsContents += h[1] + '\t' + h[0] + '\n';
         });
     }
     hostsContents += hostLinkContents;
diff --git a/src/vm/tests/test-docker.js b/src/vm/tests/test-docker.js
index a1952d43..3aab4de4 100644
--- a/src/vm/tests/test-docker.js
+++ b/src/vm/tests/test-docker.js
@@ -762,7 +762,7 @@ test('test docker VM with "docker:extraHosts"', function (t) {
             var hostsPath = path.resolve(
                 vmobj.zonepath, 'root', 'etc', 'hosts');
             var hostsContent = fs.readFileSync(hostsPath, 'utf8');
-            var foo = /^foo\t1.2.3.4$/m;
+            var foo = /^1.2.3.4\tfoo$/m;
             t.ok(foo.test(hostsContent),
                 util.format('%s entry found in hosts path (%s): %j',
                     foo, hostsPath, hostsContent));
@@ -781,8 +781,8 @@ test('test docker VM with "docker:extraHosts"', function (t) {
             var hostsPath = path.resolve(
                 vmobj.zonepath, 'root', 'etc', 'hosts');
             var hostsContent = fs.readFileSync(hostsPath, 'utf8');
-            var foo = /^foo\t1.2.3.4$/m;
-            var bar = /^bar\t5.6.7.8$/m;
+            var foo = /^1.2.3.4\tfoo$/m;
+            var bar = /^5.6.7.8\tbar$/m;
             t.notOk(foo.test(hostsContent),
                 util.format('%s entry NOT found in hosts path (%s): %j',
                     foo, hostsPath, hostsContent));
