commit c3e26092c64593555fa48ec3f2ed19c9bbc1614b
Author: John Levon <john.levon@joyent.com>
Date:   2019-05-08T13:20:43+00:00 (5 months ago)
    
    OS-7790 enable smatch build for illumos-joyent

diff --git a/configure b/configure
index 75bff507..ff23cf70 100755
--- a/configure
+++ b/configure
@@ -270,6 +270,11 @@ function generate_env
 	done
 	unset IFS
 
+	if [[ "$ENABLE_SMATCH" = "yes" ]]; then
+		SMATCHBIN=\$CODEMGR_WS/usr/src/tools/proto/root_\$MACH-nd/opt/onbld/bin/\$MACH/smatch
+		SHADOW_CCS+=" smatch,$SMATCHBIN,smatch"
+	fi
+
 	lprefix=$(echo $conf_root | tr / _)
 	[[ $? -eq 0 ]] || fatal "failed to create lock prefix"
 
@@ -438,13 +443,15 @@ read -r -d '' usage <<EOF
 		primary compiler version [default: gcc4]
 	-r
 		full strap build (no cache) [default: no]
+	-S
+		do *not* run smatch [default is to run smatch]
 	-s gcc7
 		shadow compilers, comma delimited (gcc7,gcc#) [default: none]
 EOF
 
 source_vars
 
-while getopts "cdhp:rs:" arg; do
+while getopts "cdhp:rSs:" arg; do
 	case $arg in
 	c)
 		ILLUMOS_CLOBBER=yes ;;
@@ -460,11 +467,15 @@ while getopts "cdhp:rs:" arg; do
 	h)
 		echo "$usage"
 		exit 0 ;;
+
 	p)
 		PRIMARY_COMPILER=$OPTARG
 		FORCE_STRAP_REBUILD=yes ;;
 	r)
 		FORCE_STRAP_REBUILD=yes ;;
+
+	S)
+		ENABLE_SMATCH=no ;;
 	s)
 		SHADOW_COMPILERS=$OPTARG
 		FORCE_STRAP_REBUILD=yes ;;
@@ -477,6 +488,7 @@ done
 [[ -n "$ILLUMOS_ENABLE_DEBUG" ]] || ILLUMOS_ENABLE_DEBUG=no
 [[ -n "$ILLUMOS_CLOBBER" ]] || ILLUMOS_CLOBBER=no
 [[ -n "$PRIMARY_COMPILER" ]] || PRIMARY_COMPILER=gcc4
+[[ -n "$ENABLE_SMATCH" ]] || ENABLE_SMATCH=yes
 
 PRIMARY_COMPILER_VER=$(echo $PRIMARY_COMPILER | sed 's/^gcc//')
 
@@ -487,6 +499,7 @@ ILLUMOS_ENABLE_DEBUG=$ILLUMOS_ENABLE_DEBUG
 PRIMARY_COMPILER=$PRIMARY_COMPILER
 PRIMARY_COMPILER_VER=$PRIMARY_COMPILER_VER
 SHADOW_COMPILERS=$SHADOW_COMPILERS
+ENABLE_SMATCH=$ENABLE_SMATCH
 EOF
 
 echo "Doing pre-flight checks... \c "
