From 4a2f0f514c9d085abed6e5529d7c33c5d2ecd2ec Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Wed, 16 May 2018 14:01:27 -0400
Subject: [PATCH] TRITON-388 napi crashes when first starting

---
 lib/napi.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/napi.js b/lib/napi.js
index 3f6ae21..4eb9b8f 100644
--- a/lib/napi.js
+++ b/lib/napi.js
@@ -295,9 +295,9 @@ NAPI.prototype.state_init.buckets = function (S) {
     models.init(self, function (err) {
         if (err) {
             self.log.error(err, 'Error initializing models; retrying in 10s');
-            S.timeout(function () {
+            S.timeout(10000, function () {
                 S.gotoState('init.buckets');
-            }, 10000);
+            });
             return;
         }
 
-- 
2.21.0

