commit abcd21aaafa386c4fd468b328cb38bc061698c4b (refs/changes/84/2884/3)
Author: Jonathan Perkin <jperkin@joyent.com>
Date:   2017-12-05T12:38:50+00:00 (1 year, 10 months ago)
    
    OS-6428 bad free in ctf_dwarf_init_die

diff --git a/usr/src/lib/libctf/common/ctf_dwarf.c b/usr/src/lib/libctf/common/ctf_dwarf.c
index 811a55bc64..fb5aac2da3 100644
--- a/usr/src/lib/libctf/common/ctf_dwarf.c
+++ b/usr/src/lib/libctf/common/ctf_dwarf.c
@@ -2649,10 +2649,13 @@ static void
 ctf_dwarf_free_dies(ctf_die_t *cdies, int ndies)
 {
 	int i;
+	ctf_die_t *cdp;
 
 	ctf_dprintf("Beginning to free dies\n");
 	for (i = 0; i < ndies; i++) {
-		ctf_dwarf_free_die(&cdies[i]);
+		cdp = &cdies[i];
+		if (cdp->cd_elf != NULL)
+			ctf_dwarf_free_die(cdp);
 	}
 
 	ctf_free(cdies, sizeof (ctf_die_t) * ndies);
@@ -2731,48 +2734,38 @@ ctf_dwarf_init_die(int fd, Elf *elf, ctf_die_t *cdp, int ndie, char *errbuf,
 		cdp->cd_elf = elf;
 		cdp->cd_maxoff = nexthdr - 1;
 		cdp->cd_ctfp = ctf_fdcreate(fd, &ret);
-		if (cdp->cd_ctfp == NULL) {
-			ctf_free(cdp, sizeof (ctf_die_t));
+		if (cdp->cd_ctfp == NULL)
 			return (ret);
-		}
 		avl_create(&cdp->cd_map, ctf_dwmap_comp, sizeof (ctf_dwmap_t),
 		    offsetof(ctf_dwmap_t, cdm_avl));
 		cdp->cd_errbuf = errbuf;
 		cdp->cd_errlen = errlen;
-		bzero(&cdp->cd_vars, sizeof (ctf_list_t));
-		bzero(&cdp->cd_funcs, sizeof (ctf_list_t));
-		bzero(&cdp->cd_bitfields, sizeof (ctf_list_t));
 
 		if ((ret = ctf_dwarf_die_elfenc(elf, cdp, errbuf,
 		    errlen)) != 0) {
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ret);
 		}
 
 		if ((ret = ctf_dwarf_sib(cdp, NULL, &cu)) != 0) {
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ret);
 		}
 		if (cu == NULL) {
 			(void) snprintf(errbuf, errlen,
 			    "file does not contain DWARF data\n");
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ECTF_CONVBKERR);
 		}
 
 		if ((ret = ctf_dwarf_child(cdp, cu, &child)) != 0) {
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ret);
 		}
 		if (child == NULL) {
 			(void) snprintf(errbuf, errlen,
 			    "file does not contain DWARF data\n");
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ECTF_CONVBKERR);
 		}
 
@@ -2781,7 +2774,6 @@ ctf_dwarf_init_die(int fd, Elf *elf, ctf_die_t *cdp, int ndie, char *errbuf,
 
 		if ((cdp->cd_cmh = ctf_merge_init(fd, &ret)) == NULL) {
 			avl_destroy(&cdp->cd_map);
-			ctf_free(cdp, sizeof (ctf_die_t));
 			return (ret);
 		}
 
@@ -2847,6 +2839,7 @@ ctf_dwarf_convert(int fd, Elf *elf, uint_t nthrs, int *errp, ctf_file_t **fpp,
 		*errp = ENOMEM;
 		return (CTF_CONV_ERROR);
 	}
+	bzero(cdies, sizeof (ctf_die_t) * ndies);
 
 	for (i = 0; i < ndies; i++) {
 		cdp = &cdies[i];
