commit 80b5dde48ed62c3524683c1b99d1e9a786cdecbd (refs/changes/34/4834/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-09-13T19:18:44+02:00 (1 year, 1 month ago)
    
    TRITON-770 Fix refreservation tests for kvm broken by TRITON-591

diff --git a/src/vm/tests/test-create-kvm.js b/src/vm/tests/test-create-kvm.js
index a9724c1c..db71c130 100644
--- a/src/vm/tests/test-create-kvm.js
+++ b/src/vm/tests/test-create-kvm.js
@@ -295,8 +295,8 @@ test('test default refreservation', function(t) {
                 }
                 disks = obj.disks;
                 t.ok(disks.length === 2, 'VM has 2 disks');
-                t.ok(disks[0].refreservation === disks[0].size, 'disk 0 has correct refreservation: ' + disks[0].refreservation + '/' + disks[0].size);
-                t.ok(disks[1].refreservation === disks[1].size, 'disk 1 has correct refreservation: ' + disks[1].refreservation + '/' + disks[1].size);
+                t.ok(disks[0].refreservation > disks[0].size, 'disk 0 has correct refreservation: ' + disks[0].refreservation + '/' + disks[0].size);
+                t.ok(disks[1].refreservation > disks[1].size, 'disk 1 has correct refreservation: ' + disks[1].refreservation + '/' + disks[1].size);
                 state.vmobj = obj;
                 cb();
             });
