commit 64abde85c1d3fb961f673c6bb9eb961c74d81866 (refs/changes/34/4834/4)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-09-17T19:06:58+02:00 (1 year, 1 month ago)
    
    TRITON-770 Fix refreservation tests for kvm broken by TRITON-591

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index f7b97338..696c1d63 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -1336,21 +1336,10 @@ exports.validate = function (brand, action, payload, options, callback)
                 var d_idx = 0;
                 if (payload.hasOwnProperty(d)) {
                     payload[d].forEach(function (disk) {
-                        if (disk.hasOwnProperty('refreservation')) {
-                            if (disk.refreservation < 0) {
+                        if (disk.hasOwnProperty('refreservation') &&
+                            disk.refreservation < 0) {
                                 errors.bad_values.push(d + '.' + d_idx
                                     + '.refreservation');
-                            } else if (disk.size
-                                && disk.refreservation > disk.size) {
-
-                                errors.bad_values.push(d + '.' + d_idx
-                                    + '.refreservation');
-                            } else if (disk.image_size
-                                && disk.refreservation > disk.image_size) {
-
-                                errors.bad_values.push(d + '.' + d_idx
-                                    + '.refreservation');
-                            }
                         }
                         d_idx++;
                     });
diff --git a/src/vm/tests/test-create-kvm.js b/src/vm/tests/test-create-kvm.js
index a9724c1c..f84c9edb 100644
--- a/src/vm/tests/test-create-kvm.js
+++ b/src/vm/tests/test-create-kvm.js
@@ -1,4 +1,4 @@
-// Copyright 2015 Joyent, Inc.  All rights reserved.
+// Copyright 2018 Joyent, Inc.  All rights reserved.
 //
 // These tests ensure that create works with specific options set.
 //
@@ -295,8 +295,8 @@ test('test default refreservation', function(t) {
                 }
                 disks = obj.disks;
                 t.ok(disks.length === 2, 'VM has 2 disks');
-                t.ok(disks[0].refreservation === disks[0].size, 'disk 0 has correct refreservation: ' + disks[0].refreservation + '/' + disks[0].size);
-                t.ok(disks[1].refreservation === disks[1].size, 'disk 1 has correct refreservation: ' + disks[1].refreservation + '/' + disks[1].size);
+                t.ok(disks[0].refreservation > disks[0].size, 'disk 0 has correct refreservation: ' + disks[0].refreservation + '>' + disks[0].size);
+                t.ok(disks[1].refreservation > disks[1].size, 'disk 1 has correct refreservation: ' + disks[1].refreservation + '>' + disks[1].size);
                 state.vmobj = obj;
                 cb();
             });
