From a072744484cd6ce1b8efbf283921dad4ff6462ab Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Wed, 11 Oct 2017 00:54:22 +0000
Subject: [PATCH] MANTA-2422 moray connection limits should be tunable by SAPI

---
 sapi_manifests/moray/template | 40 +++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/sapi_manifests/moray/template b/sapi_manifests/moray/template
index 00b9609..5b6ebfa 100644
--- a/sapi_manifests/moray/template
+++ b/sapi_manifests/moray/template
@@ -12,26 +12,66 @@
       "zk": {
         "connStr": "{{#ZK_SERVERS}}{{host}}:{{port}}{{^last}},{{/last}}{{/ZK_SERVERS}}",
         "opts": {
+{{#MORAY_ZK_SESSION_TIMEOUT}}
+          "sessionTimeout": {{MORAY_ZK_SESSION_TIMEOUT}},
+{{/MORAY_ZK_SESSION_TIMEOUT}}
+{{^MORAY_ZK_SESSION_TIMEOUT}}
           "sessionTimeout": 60000,
+{{/MORAY_ZK_SESSION_TIMEOUT}}
+{{#MORAY_ZK_SPIN_DELAY}}
+          "spinDelay": {{MORAY_ZK_SPIN_DELAY}},
+{{/MORAY_ZK_SPIN_DELAY}}
+{{^MORAY_ZK_SPIN_DELAY}}
           "spinDelay": 1000,
+{{/MORAY_ZK_SPIN_DELAY}}
+{{#MORAY_ZK_RETRIES}}
+          "retries": {{MORAY_ZK_RETRIES}}
+{{/MORAY_ZK_RETRIES}}
+{{^MORAY_ZK_RETRIES}}
           "retries": 2
+{{/MORAY_ZK_RETRIES}}
         }
       }
     },
     "pg": {
+{{#MORAY_PG_CONN_TIMEOUT}}
+      "connectTimeout": {{MORAY_PG_CONN_TIMEOUT}},
+{{/MORAY_PG_CONN_TIMEOUT}}
+{{^MORAY_PG_CONN_TIMEOUT}}
       "connectTimeout": 4000,
+{{/MORAY_PG_CONN_TIMEOUT}}
+{{#MORAY_PG_CHECK_INTERVAL}}
+      "checkInterval": {{MORAY_PG_CHECK_INTERVAL}},
+{{/MORAY_PG_CHECK_INTERVAL}}
+{{^MORAY_PG_CHECK_INTERVAL}}
       "checkInterval": 90000,
+{{/MORAY_PG_CHECK_INTERVAL}}
       "maxConnections": {{MORAY_MAX_PG_CONNS}},
 {{#MORAY_MAX_QUEUE_LENGTH}}
       "maxQueueLength": {{MORAY_MAX_QUEUE_LENGTH}},
 {{/MORAY_MAX_QUEUE_LENGTH}}
+{{#MORAY_PG_MAX_IDLE_TIME}}
+      "maxIdleTime": {{MORAY_PG_MAX_IDLE_TIME}},
+{{/MORAY_PG_MAX_IDLE_TIME}}
+{{^MORAY_PG_MAX_IDLE_TIME}}
       "maxIdleTime": 270000,
+{{/MORAY_PG_MAX_IDLE_TIME}}
       "user": "moray"
     }
   },
   "cache": {
+{{#MORAY_CACHE_SIZE}}
+    "size": {{MORAY_CACHE_SIZE}},
+{{/MORAY_CACHE_SIZE}}
+{{^MORAY_CACHE_SIZE}}
     "size": 5000,
+{{/MORAY_CACHE_SIZE}}
+{{#MORAY_CACHE_EXPIRY}}
+    "expiry": {{MORAY_CACHE_EXPIRY}}
+{{/MORAY_CACHE_EXPIRY}}
+{{^MORAY_CACHE_EXPIRY}}
     "expiry": 60
+{{/MORAY_CACHE_EXPIRT}}
   },
   "server_uuid": "{{auto.SERVER_UUID}}",
   "datacenter": "{{DATACENTER}}"
-- 
2.21.0

