commit 5aade41cd4b1273639dd50a2cecf02bfaaceda61 (refs/changes/87/987/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    trailing whitespace

diff --git a/lib/do_package/do_get.js b/lib/do_package/do_get.js
index 3b4bb1c..4ba4b52 100644
--- a/lib/do_package/do_get.js
+++ b/lib/do_package/do_get.js
@@ -34,7 +34,7 @@ function do_get(subcmd, opts, args, callback) {
             if (err) {
                 return callback(err);
             }
-            
+
             if (opts.json) {
                 console.log(JSON.stringify(pkg));
             } else {
