commit 13d902018a776c711b37765f074a8d4aba4cae19 (refs/changes/67/5367/2)
Author: Angela Fong2 <angela.fong@joyent.com>
Date:   2019-01-23T10:27:30-08:00 (9 months ago)
    
    MANTA-3970 Nginx connection count should be tunable via SAPI

diff --git a/README.md b/README.md
index 40dcd0b..6a3acfd 100644
--- a/README.md
+++ b/README.md
@@ -92,6 +92,14 @@ server at `http://1.2.3.4:8080/`.
 
 ## SAPI Tunables
 
+    MAKO_WORKER_PROCESSES      Number of nginx worker processes
+
+    MAKO_WORKER_CONNECTIONS    Maximum number of connections per nginx worker
+                               process
+
+    MAKO_THREAD_POOL_SIZE      Number of threads to use for multi-part upload
+                               thread pool
+
     MAKO_PROCESS_MANIFEST      Boolean value which when set, enables the post
                                processing of a mako manifest to produce a
                                summary as well.  The summary provides
diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index 030f7a2..45183fe 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -16,7 +16,12 @@ thread_pool mpu threads=8;
 error_log /var/log/mako-error.log info;
 
 events {
+{{#MAKO_WORKER_CONNECTIONS}}
+        worker_connections {{MAKO_WORKER_CONNECTIONS}};
+{{/MAKO_WORKER_CONNECTIONS}}
+{{^MAKO_WORKER_CONNECTIONS}}
 	worker_connections  1024;
+{{/MAKO_WORKER_CONNECTIONS}}
 	use eventport;
 }
 
