From 8f53a54cccd187731948db4b8997951f5259286f Mon Sep 17 00:00:00 2001
From: Angela Fong2 <angela.fong@joyent.com>
Date: Mon, 14 Jan 2019 17:11:34 -0800
Subject: [PATCH] MANTA-3970 Nginx connection count should be tunable via SAPI

---
 sapi_manifests/mako/template | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index 030f7a2..45183fe 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -16,7 +16,12 @@ thread_pool mpu threads=8;
 error_log /var/log/mako-error.log info;
 
 events {
+{{#MAKO_WORKER_CONNECTIONS}}
+        worker_connections {{MAKO_WORKER_CONNECTIONS}};
+{{/MAKO_WORKER_CONNECTIONS}}
+{{^MAKO_WORKER_CONNECTIONS}}
 	worker_connections  1024;
+{{/MAKO_WORKER_CONNECTIONS}}
 	use eventport;
 }
 
-- 
2.21.0

