commit 8f53a54cccd187731948db4b8997951f5259286f (refs/changes/67/5367/1)
Author: Angela Fong2 <angela.fong@joyent.com>
Date:   2019-01-14T17:11:34-08:00 (9 months ago)
    
    MANTA-3970 Nginx connection count should be tunable via SAPI

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index 030f7a2..45183fe 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -16,7 +16,12 @@ thread_pool mpu threads=8;
 error_log /var/log/mako-error.log info;
 
 events {
+{{#MAKO_WORKER_CONNECTIONS}}
+        worker_connections {{MAKO_WORKER_CONNECTIONS}};
+{{/MAKO_WORKER_CONNECTIONS}}
+{{^MAKO_WORKER_CONNECTIONS}}
 	worker_connections  1024;
+{{/MAKO_WORKER_CONNECTIONS}}
 	use eventport;
 }
 
