commit 19710890b80df33b3e2a4a5a3e523d5bfdec3a4b (refs/changes/28/2428/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2017-08-18T12:04:10-04:00 (2 years, 2 months ago)
    
    PUBAPI-1426 sdc-cloudapi incorrectly handles machine create test failure

diff --git a/test/machines.70.test.js b/test/machines.70.test.js
index 238b6f9..3ca64cf 100644
--- a/test/machines.70.test.js
+++ b/test/machines.70.test.js
@@ -152,7 +152,7 @@ test('Wait For Running', function (t) {
 
 
 test('Get Machine,  with Firewall Enabled', function (t) {
-    if (MACHINE_UUID) {
+    if (!MACHINE_UUID) {
         return t.end();
     }
 
