commit 4413f7632052c21e8c008b56b00b276d38fc050a (refs/changes/19/1019/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:06-05:00 (2 years, 10 months ago)
    
    ./bin/triton services

diff --git a/lib/do_services.js b/lib/do_services.js
index b8e90be..8850451 100644
--- a/lib/do_services.js
+++ b/lib/do_services.js
@@ -31,38 +31,44 @@ function do_services(subcmd, opts, args, callback) {
 
     var columns = opts.o.split(',');
     var sort = opts.s.split(',');
+    var tritonapi = this.tritonapi;
 
-    this.tritonapi.cloudapi.listServices(function (err, services) {
-        if (err) {
-            callback(err);
-            return;
+    common.cliSetupTritonApi({cli: this}, function onSetup(setupErr) {
+        if (setupErr) {
+            callback(setupErr);
         }
+        tritonapi.cloudapi.listServices(function (err, services) {
+            if (err) {
+                callback(err);
+                return;
+            }
 
-        if (opts.json) {
-            console.log(JSON.stringify(services));
-        } else {
-            /*
-             * services are returned in the form of:
-             * {name: 'endpoint', name2: 'endpoint2', ...}
-             * we "normalize" them for use by tabula and JSON stream
-             * by making them an array
-             */
-            var svcs = [];
-            Object.keys(services).forEach(function (key) {
-                svcs.push({
-                    name: key,
-                    endpoint: services[key]
+            if (opts.json) {
+                console.log(JSON.stringify(services));
+            } else {
+                /*
+                 * services are returned in the form of:
+                 * {name: 'endpoint', name2: 'endpoint2', ...}
+                 * we "normalize" them for use by tabula and JSON stream
+                 * by making them an array
+                 */
+                var svcs = [];
+                Object.keys(services).forEach(function (key) {
+                    svcs.push({
+                        name: key,
+                        endpoint: services[key]
+                    });
                 });
-            });
-            tabula(svcs, {
-                skipHeader: opts.H,
-                columns: columns,
-                sort: sort,
-                dottedLookup: true
-            });
-        }
+                tabula(svcs, {
+                    skipHeader: opts.H,
+                    columns: columns,
+                    sort: sort,
+                    dottedLookup: true
+                });
+            }
 
-        callback();
+            callback();
+        });
     });
 }
 
