commit 9f1eb2addd0acf4b187c1b263395f7902cea12b9 (refs/changes/07/4507/2)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-07-12T17:06:48+02:00 (1 year, 3 months ago)
    
    OS-7074 deadlock with vmm_detach() and iommu_cleanup()

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index a8381a9c0a..68afa627ee 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -1551,7 +1551,7 @@ vmm_do_vm_destroy_locked(vmm_softc_t *sc, boolean_t clean_zsd)
 int
 vmm_do_vm_destroy(vmm_softc_t *sc, boolean_t clean_zsd)
 {
-	int 		err;
+	int		err;
 
 	mutex_enter(&vmmdev_mtx);
 	mutex_enter(&vmm_mtx);
@@ -1922,7 +1922,8 @@ vmm_detach(dev_info_t *dip, ddi_detach_cmd_t cmd)
 	}
 
 	/* Ensure that all resources have been cleaned up */
-	mutex_enter(&vmmdev_mtx);
+	if (mutex_tryenter(&vmmdev_mtx) == 0)
+		return (DDI_FAILURE);
 
 	if (vmmdev_inst_count != 0) {
 		mutex_exit(&vmmdev_mtx);
