commit 9bea0f21e95b206e1294b263ca5d5810489bed73 (refs/changes/07/4507/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-07-12T17:03:52+02:00 (1 year, 3 months ago)
    
    OS-7074 deadlock with vmm_detach() and iommu_cleanup()

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index a8381a9c0a..ff1224f9de 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -1922,7 +1922,8 @@ vmm_detach(dev_info_t *dip, ddi_detach_cmd_t cmd)
 	}
 
 	/* Ensure that all resources have been cleaned up */
-	mutex_enter(&vmmdev_mtx);
+	if (mutex_tryenter(&vmmdev_mtx) == 0)
+		return (DDI_FAILURE);
 
 	if (vmmdev_inst_count != 0) {
 		mutex_exit(&vmmdev_mtx);
