From 101fd1b505abadce7f28b9b0827f26d45a1fbf39 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Mon, 23 Apr 2018 12:26:58 -0700
Subject: [PATCH] TRITON-335 Better "latest" links for SmartDataCenter and
 platform

---
 tools/mantaput-bits | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/tools/mantaput-bits b/tools/mantaput-bits
index 23043c0..c8868dd 100755
--- a/tools/mantaput-bits
+++ b/tools/mantaput-bits
@@ -122,9 +122,13 @@ done
 # upload the md5sums and config.mk
 echo -e $md5sums | mput -v -H 'content-type: text/plain' /${MANTA_USER}${UPLOAD_BASE_DIR}/${upload_subdir}/md5sums.txt
 mput -v -f $BITS_DIR/config.mk /${MANTA_USER}${UPLOAD_BASE_DIR}/${upload_subdir}/config.mk
-# now update the latest link
 
+# now update the branch latest link
 echo "/${MANTA_USER}${UPLOAD_BASE_DIR}/${upload_subdir}" | mput -v -H 'content-type: text/plain' /${MANTA_USER}${UPLOAD_BASE_DIR}/${UPLOAD_BRANCH}-latest
+# If this is a release branch, also update latest link
+if [[ $UPLOAD_BRANCH =~ ^release ]]; then
+    echo "/${MANTA_USER}${UPLOAD_BASE_DIR}/${upload_subdir}" | mput -v -H 'content-type: text/plain' /${MANTA_USER}${UPLOAD_BASE_DIR}/latest
+fi
 
 echo "Uploaded to /${MANTA_USER}${UPLOAD_BASE_DIR}/${upload_subdir}"
 
-- 
2.21.0

