From 6ba4b0ae7788fa79d43a7735ddf769be6fbec738 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 29 Nov 2016 07:51:44 -0500
Subject: [PATCH] joyent/node-cueball#54 skip AAAA lookups on machines with no
 v6 address

---
 lib/resolver.js | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/lib/resolver.js b/lib/resolver.js
index aa0474f..8cf5b41 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -34,6 +34,7 @@ const mod_ipaddr = require('ipaddr.js');
 const mod_fs = require('fs');
 const mod_crypto = require('crypto');
 const mod_verror = require('verror');
+const mod_os = require('os');
 
 const FSM = mod_mooremachine.FSM;
 const EventEmitter = mod_events.EventEmitter;
@@ -465,9 +466,21 @@ CueBallDNSResolver.prototype.state_srv_error = function (S) {
 };
 
 CueBallDNSResolver.prototype.state_aaaa = function (S) {
-	this.r_srvRem = this.r_srvs.slice();
-	this.r_nextV6 = undefined;
-	S.gotoState('aaaa_next');
+	var nics = mod_os.networkInterfaces();
+	var haveV6 = false;
+	Object.keys(nics).forEach(function (k) {
+		nics[k].forEach(function (addr) {
+			if (addr.family === 'IPv6' && addr.address !== '::1')
+				haveV6 = true;
+		});
+	});
+	if (haveV6) {
+		this.r_srvRem = this.r_srvs.slice();
+		this.r_nextV6 = undefined;
+		S.gotoState('aaaa_next');
+	} else {
+		S.gotoState('a');
+	}
 };
 
 CueBallDNSResolver.prototype.state_aaaa_next = function (S) {
-- 
2.21.0

