commit 0e5c11a04dabb46c23bfc5ef6057e845aa54e789 (refs/changes/13/3813/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-04-12T09:21:22+00:00 (1 year, 6 months ago)
    
    OS-6879 Need management of ppt devices

diff --git a/overlay/generic/etc/ppt_matches b/overlay/generic/etc/ppt_matches
new file mode 100644
index 00000000..03ff7e03
--- /dev/null
+++ b/overlay/generic/etc/ppt_matches
@@ -0,0 +1 @@
+pci10de,15f0
diff --git a/overlay/generic/manifest b/overlay/generic/manifest
index 64d0f21b..d8b88125 100644
--- a/overlay/generic/manifest
+++ b/overlay/generic/manifest
@@ -36,6 +36,7 @@ f etc/opasswd 0644 root sys
 f etc/ouser_attr 0644 root sys
 f etc/passwd 0644 root sys
 f etc/path_to_inst 0444 root root
+f etc/ppt_matches 0444 root root
 f etc/profile 0644 root sys
 f etc/resolv.conf 0644 netadm netadm
 f etc/rtc_config 0644 root root
diff --git a/src/img/CHANGES.md b/src/img/CHANGES.md
index 3855d6b9..b3260374 100644
--- a/src/img/CHANGES.md
+++ b/src/img/CHANGES.md
@@ -5,6 +5,10 @@ Known issues:
 - Docker image imports are experimental. Docker image import also only supports
   Docker Registry v2.
 
+## 3.9.1
+
+- OS-5979 ensure imgadm uses the provided req_id
+
 ## 3.9.0
 
 - TRITON-178 add support for image creation for bhyve VMs. Also includes a
diff --git a/src/img/lib/imgadm.js b/src/img/lib/imgadm.js
index 62a750a1..aad2eebb 100644
--- a/src/img/lib/imgadm.js
+++ b/src/img/lib/imgadm.js
@@ -4099,9 +4099,16 @@ IMGADM.prototype.publishImage = function publishImage(opts, callback) {
         'options.manifestFile.files[0].compression');
     var self = this;
 
+    // Set x-request-id on all IMGAPI calls.
+    var headers = {};
+    if (self.log && self.log.fields && self.log.fields.req_id) {
+        headers['x-request-id'] = self.log.fields.req_id;
+    }
+
     var client = imgapi.createClient({
         agent: false,
         url: opts.url,
+        headers: headers,
         log: self.log.child({component: 'api', url: opts.url}, true),
         rejectUnauthorized: (process.env.IMGADM_INSECURE !== '1'),
         userAgent: self.userAgent
diff --git a/src/img/lib/sources/imgapi.js b/src/img/lib/sources/imgapi.js
index 270e93ff..f51e5916 100644
--- a/src/img/lib/sources/imgapi.js
+++ b/src/img/lib/sources/imgapi.js
@@ -44,9 +44,14 @@ function ImgapiSource(opts) {
 
     this.__defineGetter__('client', function () {
         if (this._client === undefined) {
+            var headers = {};
+            if (self.log && self.log.fields && self.log.fields.req_id) {
+                headers['x-request-id'] = self.log.fields.req_id;
+            }
             this._client = imgapi.createClient({
                 url: self.normUrl,
                 version: '~2',
+                headers: headers,
                 log: self.log,
                 rejectUnauthorized: (opts.insecure !== undefined ?
                     opts.insecure : process.env.IMGADM_INSECURE !== '1'),
diff --git a/src/img/package.json b/src/img/package.json
index 84e2d61f..2b968dfe 100644
--- a/src/img/package.json
+++ b/src/img/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgadm",
   "description": "Manage SmartOS virtual machine images.",
-  "version": "3.9.0",
+  "version": "3.9.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
diff --git a/src/sysinfo b/src/sysinfo
index 67873b78..c28215f2 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -103,14 +103,24 @@ function get_smbios_system_info()
 {
     # This puts the variables we're pulling out into the local environment
     eval $(smbios -t SMB_TYPE_SYSTEM | tr -d "\"" \
-        | egrep "Manufacturer: |Product: |Serial Number: |UUID: |SKU Number: |Version: |Family: " \
+        | egrep "Manufacturer: |Product: |Serial Number: |UUID \(Endian-corrected\): |UUID: |SKU Number: |Version: |Family: " \
         | sed -e 's/^ *//' \
         -e 's/: /="/' \
         -e 's/ *$/"/' \
         -e 's/Serial Number/Serial_Number/' \
         -e 's/SKU Number/SKU_Number/' \
         -e 's/Version/HW_Version/' \
-        -e 's/Family/HW_Family/')
+        -e 's/Family/HW_Family/' \
+        -e 's/UUID (Endian-corrected)/Fixed_UUID/')
+
+    vendor=$(smbios -t SMB_TYPE_BIOS | grep "Vendor: " \
+        | sed -e 's/^.*Vendor: //;s/ *$//')
+
+    # If we are a bhyve guest, we can use the correct UUID format; otherwise we
+    # use the network-endian one for compatibility reasons.
+    if [ "$vendor" = "BHYVE" -a -n "$Fixed_UUID" ]; then
+        UUID=$Fixed_UUID
+    fi
 
     # overwrite UUID if config dictates otherwise
     tmp_uuid=$(/usr/bin/bootparams | grep "^override_uuid=" | cut -f2 -d'=')
