commit 91bba6d61b71f872318bfbdfea1231ca68a53f8c (refs/changes/05/3805/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-04-11T11:55:49-07:00 (1 year, 6 months ago)
    
    joyent/node-sshpk-agent#14 Support sshpk 1.14.x and higher

diff --git a/lib/client.js b/lib/client.js
index 7795a5b..9aec5f7 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -5,7 +5,7 @@ var sshpk = require('sshpk');
 var sshpkUtils = require('sshpk/lib/utils');
 var util = require('util');
 var errors = require('./errors');
-var SSHBuffer = require('sshpk/lib/ssh-buffer');
+var SSHBuffer = require('./ssh-buffer');
 var AgentProtocolError = errors.AgentProtocolError;
 
 var ClientFSM = require('./client-fsm');
@@ -382,8 +382,9 @@ function certToBuffer(cert, k) {
 		 * For some reason the public key gets encoded again for
 		 * ed25519 certs. The mysteries will never cease.
 		 */
-		sshbuf.writePart(k.part.R);
-		sshbuf.writePart(k.part.r);
+		sshbuf.writeBuffer(k.part.A.data);
+		var edk = Buffer.concat([k.part.A.data, k.part.k.data]);
+		sshbuf.writeBuffer(edk);
 		break;
 	default:
 		throw (new Error('Key type ' + k.type + ' not supported'));
diff --git a/package.json b/package.json
index 73d0d7a..83c68ef 100644
--- a/package.json
+++ b/package.json
@@ -29,7 +29,7 @@
     "assert-plus": "^1.0.0",
     "mooremachine": "^2.0.1",
     "readable-stream": "^2.1.4",
-    "sshpk": ">=1.13.0 < 1.13.2",
+    "sshpk": ">=1.14.0 < 1.15.0",
     "verror": "^1.10.0"
   },
   "devDependencies": {
