commit 893fd7ad1a3b85b384b466e39a8e0b5d7dadf1e2 (refs/changes/06/3506/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-02-27T15:37:45-08:00 (1 year, 7 months ago)
    
    TRITON-189 config-agent change in TRITON-184

diff --git a/agent.js b/agent.js
index 9f94266..f07a1fe 100644
--- a/agent.js
+++ b/agent.js
@@ -135,7 +135,8 @@ async.waterfall([
 				// Allow the config file to specify the
 				// instance UUID(s). This is used for load
 				// testing.
-				if (config.instances.length === 0) {
+				if (!config.instances ||
+				    config.instances.length === 0) {
 					config.instances = [ zonename ];
 				}
 			}
