commit 46a48d3ece32e751f8fc0028c1e4b0bbd02ed99e (refs/changes/89/3989/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-05-22T22:07:24+00:00 (1 year, 5 months ago)
    
    joyent/node-fast#19 tst.server.js fails after #14

diff --git a/test/tst.server.js b/test/tst.server.js
index e0a22f7..984dbe5 100644
--- a/test/tst.server.js
+++ b/test/tst.server.js
@@ -25,7 +25,6 @@ var mod_fastdemo = require('../lib/demo_server');
 var mod_protocol = require('../lib/fast_protocol');
 var mod_testcommon = require('./common');
 
-var EventEmitter = require('events');
 var VError = require('verror');
 
 var testLog;
@@ -1122,7 +1121,6 @@ serverTestCases = [ {
 		var choice = 0;
 		var nrequests = 200;
 		var ncomplete = 0;
-		var e = new EventEmitter();
 
 		var queue = mod_vasync.queuev({
 			'concurrency': 100,
@@ -1135,7 +1133,7 @@ serverTestCases = [ {
 					'maxObjectsToBuffer': 1
 				}, function () {
 					if (++ncomplete == nrequests) {
-						e.emit('complete');
+						queue.close();
 					}
 				});
 				qcallback();
@@ -1150,7 +1148,7 @@ serverTestCases = [ {
 		 * that just finished sending requests. This should
 		 * trigger the 'onConnsDestroyed' callback.
 		 */
-		e.on('complete', function () {
+		queue.on('end', function () {
 			var WAIT = 3000;
 			finish_ts = process.hrtime();
 			tctx.ts_clients.forEach(function (c) {
@@ -1183,7 +1181,6 @@ serverTestCases = [ {
 
 			callback();
 		}
-		queue.close();
 	});
     }
 } ];
