From d7b4948209c3f0e4b979351cd15e84db478b20c8 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 16 Aug 2017 12:14:16 -0700
Subject: [PATCH] joyent/node-cueball#123 ConnectionSet leak during failure

---
 lib/set.js | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/set.js b/lib/set.js
index 5e80905..62c8332 100644
--- a/lib/set.js
+++ b/lib/set.js
@@ -618,7 +618,7 @@ CueBallConnectionSet.prototype._incrCounter = function (counter) {
  *        |
  *        |
  *        v
- *    +--------+
+ *    +--------+ fsm failed
  *    |        | .drain()
  *    |  init  +-----------------------------+
  *    |        |                             |
@@ -704,6 +704,8 @@ LogicalConnection.prototype.state_init = function (S) {
 			if (self.lc_fsm.isInState('idle')) {
 				self.lc_hdl.try(self.lc_fsm);
 			}
+		} else if (st === 'failed' || st === 'cancelled') {
+			S.gotoState('stopped');
 		}
 	});
 
@@ -712,6 +714,8 @@ LogicalConnection.prototype.state_init = function (S) {
 			if (self.lc_hdl.isInState('waiting')) {
 				self.lc_hdl.try(self.lc_fsm);
 			}
+		} else if (st === 'failed') {
+			S.gotoState('stopped');
 		}
 	});
 
-- 
2.21.0

