commit e3d8de190a253b19f550842057526a5056242b32 (refs/changes/62/3762/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-04-02T16:21:08-07:00 (1 year, 6 months ago)
    
    TRITON-288 sdcnode updates for March 2018 node.js security releases

diff --git a/Makefile b/Makefile
index cccfeae9..a3de7c11 100644
--- a/Makefile
+++ b/Makefile
@@ -36,7 +36,7 @@ JSSTYLE_FLAGS	 = -o "indent=2,doxygen,unparenthesized-return=0,line-length=120"
 REPO_MODULES	 =
 SMF_MANIFESTS_IN = smf/manifests/adminui.xml.in
 
-NODE_PREBUILT_VERSION=v4.8.7
+NODE_PREBUILT_VERSION=v4.9.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
 	NODE_PREBUILT_TAG=zone
