commit 60de181a6ce0c84cb423eb98882cd652babede17 (refs/changes/86/4186/2)
Author: Rui Loura <rui@joyent.com>
Date:   2018-06-14T20:35:39+00:00 (1 year, 4 months ago)
    
    MANTA-3753 muppet should leverage autometadata instead of checking nictags

diff --git a/muppet.js b/muppet.js
index f7f8290..b24df0d 100644
--- a/muppet.js
+++ b/muppet.js
@@ -58,29 +58,28 @@ function getUntrustedIPs(cfg, callback) {
 
         cfg.log.info({ nics: nics }, 'Looked up NICs');
 
-        function addIPsFromNics(nic) {
-            // Skip NICs on trusted networks.
-            if (nic.nic_tag === 'admin' || nic.nic_tag === 'manta') {
-                return;
+        function _pushIP(ip) {
+            // Skip the admin and trusted IPs.
+            if ((net.isIPv4(ip) || net.isIPv6(ip)) &&
+                ip !== cfg.adminIP && ip !== cfg.trustedIP) {
+
+                cfg.untrustedIPs.push(ip);
             }
+        }
 
+        function _addIPsFromNics(nic) {
             if (nic.hasOwnProperty('ips')) {
                 nic.ips.forEach(function parseIP(addr) {
-                    const ip = addr.split('/')[0];
-                    if (net.isIPv4(ip) || net.isIPv6(ip)) {
-                        cfg.untrustedIPs.push(ip);
-                    }
+                    _pushIP(addr.split('/')[0]);
                 });
             } else if (nic.hasOwnProperty('ip')) {
-                if (net.isIPv4(nic.ip)) {
-                    cfg.untrustedIPs.push(nic.ip);
-                }
+                _pushIP(nic.ip);
             } else {
                 cfg.log.warn({ nic: nic }, 'NIC has no IP addresses');
             }
         }
 
-        nics.forEach(addIPsFromNics);
+        nics.forEach(_addIPsFromNics);
         callback();
     });
 }
diff --git a/sapi_manifests/muppet/template b/sapi_manifests/muppet/template
index 17c3478..8e4dfc7 100644
--- a/sapi_manifests/muppet/template
+++ b/sapi_manifests/muppet/template
@@ -1,6 +1,8 @@
 {
   "name": "manta.{{REGION}}.{{DNS_DOMAIN}}",
   "trustedIP": "{{auto.MANTA_IP}}",
+  "adminIP": "{{auto.ADMIN_IP}}",
+  "adminIPS":  {{auto.ADMIN_IPS}},
   "zookeeper": {
     "servers": [
       {{#ZK_SERVERS}}
