commit 3b2ad69fd06226482d6276a7af5baef9eb835df6 (refs/changes/05/5705/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2019-03-04T21:16:11+00:00 (7 months ago)
    
    OS-7632 smartos configure should heed OS-7341

diff --git a/configure b/configure
index d70c53d2..8c0ae268 100755
--- a/configure
+++ b/configure
@@ -442,8 +442,8 @@ read -r -d '' usage <<EOF
 		primary compiler version [default: gcc4]
 	-r
 		full strap build (no cache) [default: no]
-	-s gcc6,gcc7
-		shadow compilers [default: none]
+	-s gcc7
+		shadow compilers, comma delimited (gcc7,gcc#) [default: none]
 EOF
 
 source_vars
