commit 68b3b7b926004a0b85f12a1223cec646ffab9743 (refs/changes/53/5653/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2019-02-27T21:55:38+00:00 (7 months ago)
    
    MANTA-4135 garbage-collector doesn't handle empty GC_ASSIGNED_SHARDS

diff --git a/sapi_manifests/manta-garbage-collector/template b/sapi_manifests/manta-garbage-collector/template
index e666713..c62c152 100644
--- a/sapi_manifests/manta-garbage-collector/template
+++ b/sapi_manifests/manta-garbage-collector/template
@@ -29,10 +29,10 @@
 		{
 			"host": "{{host}}"
 		}{{^last}},{{/last}}{{/GC_ASSIGNED_SHARDS}}
-		{{^GC_ASSIGNED_SHARDS}}
+		{{^GC_ASSIGNED_SHARDS}}{{#GC_SHARD_NUM_LO}}{{#GC_SHARD_NUM_HI}}
 		{{GC_SHARD_NUM_LO}},
 		{{GC_SHARD_NUM_HI}}
-		{{/GC_ASSIGNED_SHARDS}}
+		{{/GC_SHARD_NUM_HI}}{{/GC_SHARD_NUM_LO}}{{/GC_ASSIGNED_SHARDS}}
 	],
 	"buckets": [{{#GC_ASSIGNED_BUCKETS}}
 		{
