commit 5191bd90dc0c805d64c466d5980185d64bf9c98c (refs/changes/74/5174/2)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-13T10:28:53-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 305da56..9b821a9 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright 2016, Joyent, Inc.
+# Copyright 2018, Joyent, Inc.
 #
 
 #
@@ -32,7 +32,7 @@ JSSTYLE_FLAGS	= -f tools/jsstyle.conf
 ESLINT_FILES	= $(JS_FILES)
 
 ifeq ($(shell uname -s),SunOS)
-	NODE_PREBUILT_VERSION =	v6.14.3
+	NODE_PREBUILT_VERSION =	v6.15.1
 	NODE_PREBUILT_TAG =	zone
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
 endif
