commit 3e6fc939acaf0563d60443536de1145fbd6f7a94 (refs/changes/74/5174/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-03T16:11:18-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 305da56..c83185f 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright 2016, Joyent, Inc.
+# Copyright 2018, Joyent, Inc.
 #
 
 #
@@ -32,7 +32,7 @@ JSSTYLE_FLAGS	= -f tools/jsstyle.conf
 ESLINT_FILES	= $(JS_FILES)
 
 ifeq ($(shell uname -s),SunOS)
-	NODE_PREBUILT_VERSION =	v6.14.3
+	NODE_PREBUILT_VERSION =	v6.15.0
 	NODE_PREBUILT_TAG =	zone
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
 endif
