commit 6e4060b0f6c98ea3bf4a7b7aef06599f3d4ef30c (refs/changes/78/3178/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2018-01-08T12:04:19-08:00 (1 year, 9 months ago)
    
    DOCKER-1138 Errors With Docker-Compose and TritonFS

diff --git a/lib/validate.js b/lib/validate.js
index 268c9e2..25770aa 100644
--- a/lib/validate.js
+++ b/lib/validate.js
@@ -593,7 +593,11 @@ function validateCreateVolume(req, res, next) {
     var validVolumeSize = false;
     var volumeDriver = req.params.Driver;
     var volumeName = req.params.Name;
-    var volumeSize = req.params.DriverOpts.size;
+    var volumeSize;
+
+    if (req.params.DriverOpts) {
+        volumeSize = req.params.DriverOpts.size;
+    }
 
     if (typeof (volumeDriver) !== 'string') {
         next(new errors.ValidationError('volume driver must be a string'));
