commit 567606ff9ee7d1e9bdbe9ad3a582fa0b05f39c3b
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2019-05-14T23:04:48+00:00 (5 months ago)
    
    OS-7026 Add DDI support for Upgradable Firmware Modules

diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index a92bd1e2..83cf7447 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -304,3 +304,4 @@ viona 306
 ppt 307
 coretemp 308
 amdf17nbdf 309
+ufm 310
