commit f8025ecea2755de5a9baa7c64179e8eee8e3eaa4
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2019-04-25T21:06:14+00:00 (6 months ago)
    
    OS-7026 Add DDI support for Upgradable Firmware Modules

diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index 26eaeddb..cabaa7a6 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -302,3 +302,4 @@ buildstamp 304
 vmm 305
 viona 306
 ppt 307
+ufm 308
