From f8025ecea2755de5a9baa7c64179e8eee8e3eaa4 Mon Sep 17 00:00:00 2001
From: Rob Johnston <rob.johnston@joyent.com>
Date: Thu, 25 Apr 2019 21:06:14 +0000
Subject: [PATCH] OS-7026 Add DDI support for Upgradable Firmware Modules

---
 overlay/generic/etc/name_to_major | 1 +
 1 file changed, 1 insertion(+)

diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index 26eaeddb..cabaa7a6 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -302,3 +302,4 @@ buildstamp 304
 vmm 305
 viona 306
 ppt 307
+ufm 308
-- 
2.21.0

