From 567606ff9ee7d1e9bdbe9ad3a582fa0b05f39c3b Mon Sep 17 00:00:00 2001
From: Rob Johnston <rob.johnston@joyent.com>
Date: Thu, 25 Apr 2019 21:06:14 +0000
Subject: [PATCH] OS-7026 Add DDI support for Upgradable Firmware Modules

---
 overlay/generic/etc/name_to_major | 1 +
 1 file changed, 1 insertion(+)

diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index a92bd1e2..83cf7447 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -304,3 +304,4 @@ viona 306
 ppt 307
 coretemp 308
 amdf17nbdf 309
+ufm 310
-- 
2.21.0

