From 91f6eee4efa86ee8f2ad758bbb9ee28211bb7415 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 29 Sep 2016 15:43:26 -0700
Subject: [PATCH] joyent/node-cueball#29 duplicate 'end' handler in
 ConnectionFSM

---
 lib/connection-fsm.js | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/lib/connection-fsm.js b/lib/connection-fsm.js
index 130aaca..197341d 100644
--- a/lib/connection-fsm.js
+++ b/lib/connection-fsm.js
@@ -410,10 +410,6 @@ ConnectionFSM.prototype.state_idle = function (S) {
 		S.gotoState('error');
 		self.cf_pool._incrCounter('end-during-idle');
 	});
-	S.on(this.cf_conn, 'end', function () {
-		self.cf_lastError = new mod_errors.ConnectionClosedError(self);
-		S.gotoState('error');
-	});
 	S.on(this, 'closeAsserted', function () {
 		S.gotoState('closed');
 	});
@@ -494,10 +490,6 @@ ConnectionFSM.prototype.state_ping = function (S) {
 		S.gotoState('error');
 		self.cf_pool._incrCounter('end-during-ping');
 	});
-	S.on(this.cf_conn, 'end', function () {
-		self.cf_lastError = new mod_errors.ConnectionClosedError(self);
-		S.gotoState('error');
-	});
 	S.on(this, 'closeAsserted', function () {
 		self.cf_lastError = new mod_errors.ConnectionClosedError(self);
 		S.gotoState('error');
@@ -563,10 +555,6 @@ ConnectionFSM.prototype.state_busy = function (S) {
 		self.cf_closeAfter = true;
 		self.cf_pool._incrCounter('end-during-busy');
 	});
-	S.on(this.cf_conn, 'end', function () {
-		self.cf_lastError = new mod_errors.ConnectionClosedError(self);
-		S.gotoState('error');
-	});
 	S.on(this.cf_conn, 'close', function () {
 		self.cf_lastError = new mod_errors.ConnectionClosedError(self);
 		S.gotoState('error');
-- 
2.21.0

