From 950d9db9dcd57ebe00faf0c3a9e5dc6017bbdb35 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 5 Jul 2016 21:01:29 +0000
Subject: [PATCH] NAPI-399 Don't run 'make test' during MG build

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 01d9e5c..7010fb1 100644
--- a/Makefile
+++ b/Makefile
@@ -1459,7 +1459,7 @@ napi: $(NAPI_BITS) napi_image
 $(NAPI_BITS): build/sdc-napi
 	@echo "# Build napi: branch $(SDC_NAPI_BRANCH), sha $(SDC_NAPI_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
 	mkdir -p $(BITS_DIR)
-	(cd build/sdc-napi && NPM_CONFIG_CACHE=$(MG_CACHE_DIR)/npm TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) gmake pkg test release publish)
+	(cd build/sdc-napi && NPM_CONFIG_CACHE=$(MG_CACHE_DIR)/npm TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) gmake pkg release publish)
 	@echo "# Created napi bits (time `date -u +%Y%m%dT%H%M%SZ`):"
 	@ls -l $(NAPI_BITS)
 	@echo ""
-- 
2.21.0

