commit 1feabf0920228975b34008249902d8e3b6deb851 (refs/changes/25/1225/1)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2017-01-06T21:30:05-08:00 (2 years, 9 months ago)
    
    OS-5885 should mount /dev/shm w/ correct permissions

diff --git a/src/dockerinit/src/dockerinit.c b/src/dockerinit/src/dockerinit.c
index ff599f72..96492682 100644
--- a/src/dockerinit/src/dockerinit.c
+++ b/src/dockerinit/src/dockerinit.c
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -787,9 +787,22 @@ mountOSDevFD()
 void
 mountLXDevShm()
 {
+    int flags = MS_DATA | MS_OPTIONSTR;
+    char opts[MAX_MNTOPT_STR];
+    int optslen;
+
+    /* initialize */
+    opts[0] = '\0';
+
     dlog("MOUNT /dev/shm (shm)\n");
 
-    if (mount("shm", "/dev/shm", MS_DATA, "tmpfs", NULL, 0) != 0) {
+    /*
+     * Linux defaults to mode=1777 for tmpfs mounts. So should we.
+     */
+    (void) strlcat(opts, "mode=1777", sizeof (opts));
+
+    optslen = sizeof (opts);
+    if (mount("shm", "/dev/shm", flags, "tmpfs", NULL, 0, opts, optslen) != 0) {
         fatal(ERR_MOUNT_DEVSHM, "failed to mount /dev/shm: %s\n",
             strerror(errno));
     }
