From 0f325f906446e153dd19049e6147177625ab3cdd Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 28 Oct 2016 18:09:49 -0700
Subject: [PATCH] MANTA-3018 binder still serving 5s TTLs for SRV additionals

---
 lib/server.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/server.js b/lib/server.js
index 08158d7..c517713 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -257,7 +257,7 @@ function resolve(options, query, cb) {
                                             ports.length < 1)
                                                 ports = [s.port];
                                         var ar, sr, nm;
-                                        var rttl = host[host.type].ttl;
+                                        var rttl = host.ttl || 30;
                                         if (service !== undefined) {
                                                 nm = host.name + '.' + domain;
                                                 ports.forEach(function (p) {
-- 
2.21.0

