commit 0f325f906446e153dd19049e6147177625ab3cdd (refs/changes/16/816/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-10-28T18:10:35-07:00 (2 years, 11 months ago)
    
    MANTA-3018 binder still serving 5s TTLs for SRV additionals

diff --git a/lib/server.js b/lib/server.js
index 08158d7..c517713 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -257,7 +257,7 @@ function resolve(options, query, cb) {
                                             ports.length < 1)
                                                 ports = [s.port];
                                         var ar, sr, nm;
-                                        var rttl = host[host.type].ttl;
+                                        var rttl = host.ttl || 30;
                                         if (service !== undefined) {
                                                 nm = host.name + '.' + domain;
                                                 ports.forEach(function (p) {
