From 513e5f4b87c4e296511b293812fd041ecc8f2a0f Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Mon, 9 Jul 2018 20:39:41 +0000
Subject: [PATCH] TRITON-574 sdcadm platform assign is broken with rack aware
 networking

---
 lib/platform.js | 34 +++++++++++++++++++++++++++++++++-
 1 file changed, 33 insertions(+), 1 deletion(-)

diff --git a/lib/platform.js b/lib/platform.js
index 548b5a4..956b11b 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1066,6 +1066,7 @@ Platform.prototype.assign = function assign(opts, callback) {
     var serverRecs = [];
     var assignServers = [];
     var uuids = [];
+    var admin_tags = [];
     var headnode;
     var progress = self.progress;
 
@@ -1083,10 +1084,41 @@ Platform.prototype.assign = function assign(opts, callback) {
         progress('Updating booter cache for servers');
 
         vasync.pipeline({funcs: [
+            function getAdminNicTags(_, next) {
+                self.sdcadm.napi.listNetworkPools({name: 'admin'},
+                    function (err, pools) {
+
+                    if (err) {
+                        next(new errors.SDCClientError(err, 'napi'));
+                        return;
+                    }
+
+                    if (pools.length > 1) {
+                        progress('Multiple netowrk pools named admin, '
+                            + 'using %s', pools[0].name);
+                    }
+
+                    /*
+                     * The presence of an admin network pool is optional.
+                     */
+                    if (pools.length === 0) {
+                        next();
+                        return;
+                    }
+
+                    admin_tags = pools[0].nic_tags_present;
+                    next();
+                });
+
+            },
             function (_, next) {
+                if (admin_tags.indexOf('admin') === -1) {
+                    admin_tags.push('admin');
+                }
+
                 var listOpts = {
                     belongs_to_type: 'server',
-                    nic_tags_provided: 'admin'
+                    nic_tags_provided: admin_tags
                 };
                 if (!opts.all) {
                     listOpts.belongs_to_uuid = serveruuids;
-- 
2.21.0

