From dedcd7d676581645ec53cb97b2ff41d5bb7e771a Mon Sep 17 00:00:00 2001
From: Patrick Mooney <pmooney@pfmooney.com>
Date: Thu, 3 Nov 2016 23:25:59 +0000
Subject: [PATCH] OS-5769 localedef yacc can race

---
 usr/src/cmd/localedef/Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/usr/src/cmd/localedef/Makefile b/usr/src/cmd/localedef/Makefile
index e8c0c5f16c..95fc87312b 100644
--- a/usr/src/cmd/localedef/Makefile
+++ b/usr/src/cmd/localedef/Makefile
@@ -229,6 +229,8 @@ CATDIRS		= \
 DATA		= $(CATDIRS:%=%/LCL_DATA)
 
 .PARALLEL:
+# The output from yacc must be emitted serially
+.NO_PARALLEL: parser.tab.c parser.tab.h
 
 DCOLL		= LC_COLLATE/LCL_DATA
 DCTYPE		= LC_CTYPE/LCL_DATA
-- 
2.21.0

