From c71d373410ec9aceb3ee8d6ef7f901ecf7aa4430 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Thu, 4 Jan 2018 13:44:33 +0000
Subject: [PATCH] MANTA-3477 create "manta-reshard" zone image

---
 Makefile        | 44 +++++++++++++++++++++++++++++++++++++++++++-
 targets.json.in | 27 +++++++++++++++++++++++++++
 2 files changed, 70 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7010fb1..439893c 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright 2016 Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -1744,6 +1744,48 @@ clean_cns:
 	(cd build/triton-cns && gmake distclean)
 
 
+#---- Manta Resharding System
+
+_manta-reshard_stamp=$(MANTA_RESHARD_BRANCH)-$(TIMESTAMP)-g$(MANTA_RESHARD_SHA)
+MANTA_RESHARD_BITS=$(BITS_DIR)/manta-reshard/manta-reshard-pkg-$(_manta-reshard_stamp).tar.bz2
+MANTA_RESHARD_IMAGE_BIT=$(BITS_DIR)/manta-reshard/manta-reshard-zfs-$(_manta-reshard_stamp).zfs.gz
+MANTA_RESHARD_MANIFEST_BIT=$(BITS_DIR)/manta-reshard/manta-reshard-zfs-$(_manta-reshard_stamp).imgmanifest
+
+.PHONY: manta-reshard
+manta-reshard: $(MANTA_RESHARD_BITS) manta-reshard_image
+
+$(MANTA_RESHARD_BITS): build/manta-reshard
+	@echo "# Build manta-reshard: branch $(MANTA_RESHARD_BRANCH), sha $(MANTA_RESHARD_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
+	mkdir -p $(BITS_DIR)
+	cd build/manta-reshard && NPM_CONFIG_CACHE=$(MG_CACHE_DIR)/npm TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) gmake release publish
+	@echo "# Created manta-reshard bits (time `date -u +%Y%m%dT%H%M%SZ`):"
+	@ls -l $(MANTA_RESHARD_BITS)
+	@echo ""
+
+.PHONY: manta-reshard_image
+manta-reshard_image: $(MANTA_RESHARD_IMAGE_BIT)
+
+$(MANTA_RESHARD_IMAGE_BIT): $(MANTA_RESHARD_BITS)
+	@echo "# Build manta-reshard_image: branch $(MANTA_RESHARD_BRANCH), sha $(MANTA_RESHARD_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
+	./tools/prep_dataset_in_jpc.sh -i "$(MANTA_RESHARD_IMAGE_UUID)" -t $(MANTA_RESHARD_BITS) \
+		-b "manta-reshard" \
+		-o "$(MANTA_RESHARD_IMAGE_BIT)" -p $(MANTA_RESHARD_PKGSRC) -O "$(MG_OUT_PATH)" \
+		-t $(MANTA_RESHARD_EXTRA_TARBALLS) -n $(MANTA_RESHARD_IMAGE_NAME) \
+		-v $(_manta-reshard_stamp) -d $(MANTA_RESHARD_IMAGE_DESCRIPTION)
+	@echo "# Created manta-reshard image (time `date -u +%Y%m%dT%H%M%SZ`):"
+	@ls -l $$(dirname $(MANTA_RESHARD_IMAGE_BIT))
+	@echo ""
+
+manta-reshard_publish_image: $(MANTA_RESHARD_IMAGE_BIT)
+	@echo "# Publish manta-reshard image to SDC Updates repo."
+	$(UPDATES_IMGADM) import -ddd -m $(MANTA_RESHARD_MANIFEST_BIT) -f $(MANTA_RESHARD_IMAGE_BIT)
+
+
+clean_manta-reshard:
+	$(RM) -rf $(BITS_DIR)/manta-reshard
+	cd build/manta-reshard && gmake distclean
+
+
 #---- Mola
 
 _mola_stamp=$(MANTA_MOLA_BRANCH)-$(TIMESTAMP)-g$(MANTA_MOLA_SHA)
diff --git a/targets.json.in b/targets.json.in
index f2009ce..fb6973c 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1430,6 +1430,33 @@ cat <<EOF
     ]
   },
 
+  "manta-reshard": {
+    "appliance": "true",
+    "build_platform": "20141030T081701Z",
+    "image_name": "manta-reshard",
+    "image_description": "Manta Resharding System",
+    "image_version": "1.0.0",
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "coreutils-8.23nb2"
+    ],
+    "repos": [
+        {"url": "git@github.com:joyent/manta-reshard.git"}
+    ],
+    "public": true,
+    "deps": [
+      "amon",
+      "config-agent",
+      "registrar"
+    ],
+    "tarballs": [
+      {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
+      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"},
+      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"}
+    ]
+  },
+
   "manta-deployment": {
     "appliance": "true",
     "build_platform": "20141030T081701Z",
-- 
2.21.0

