commit a2c3eeb9004d06f65b52c9c2b7bb39b9291aa755 (refs/changes/22/1022/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:06-05:00 (2 years, 10 months ago)
    
    And the first warning worksmith

diff --git a/lib/do_profile/profilecommon.js b/lib/do_profile/profilecommon.js
index 168e021..6fd794e 100644
--- a/lib/do_profile/profilecommon.js
+++ b/lib/do_profile/profilecommon.js
@@ -224,12 +224,22 @@ function profileDockerSetup(opts, cb) {
 
         function mentionSettingUp(arg, next) {
             console.log('Setting up profile "%s" to use Docker.', profile.name);
-            //     console.log(wordwrap('\nWARNING: Docker does not
-            //         support ' + 'encrypted (password protected)
-            //         keys. If you continue, a ' + 'decrypted copy of
-            //         your SSH private key will be stored on ' +
-            //         'disk for with Docker.'));
-            next();
+            console.log(wordwrap(
+                '\nWARNING: Docker uses TLS based authentication with a ' +
+                 'different security model from SSH keys. As a result, the n' +
+                    'Docker client can not currently support encrypted ' +
+                    '(password protected) keys or SSH agents. If you ' +
+                    'continue, the Triton CLI will attempt to format your' +
+                    'SSH *private* key as an unencrypted TLS cert and place ' +
+                    'it in ~/.triton/docker for use by the Docker client.'));
+            common.promptYesNo({msg: 'Continue? [y/n] '}, function (answer) {
+                if (answer !== 'y') {
+                    console.error('Aborting');
+                    next(true);
+                } else {
+                    next();
+                }
+            });
         },
 
         function checkSshPrivKey(arg, next) {
