commit 395631d9abc3730b42a3072b12830e51c0395375 (refs/changes/93/1793/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2017-04-14T01:05:51+00:00 (2 years, 6 months ago)
    
    FWAPI-277 /test/integration/helpers.js should be removed

diff --git a/test/integration/helpers.js b/test/integration/helpers.js
deleted file mode 100644
index 778ae6d..0000000
--- a/test/integration/helpers.js
+++ /dev/null
@@ -1,28 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2017, Joyent, Inc.
- */
-
-/*
- * Test helpers for FWAPI integration tests
- */
-
-'use strict';
-
-var common = require('../lib/common');
-
-
-
-// --- Exports
-
-
-
-module.exports = {
-    ifErr: common.ifErr,
-    uuidSort: common.uuidSort
-};
