commit 06ea566d1af0119c76451bcaac3eb406d6eeb01d (refs/changes/67/2467/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-08-25T16:18:11-07:00 (2 years, 1 month ago)
    
    DOCKER-705 "docker build" containers are assigned public IP

diff --git a/lib/backends/sdc/build.js b/lib/backends/sdc/build.js
index 3f58a65..215067a 100644
--- a/lib/backends/sdc/build.js
+++ b/lib/backends/sdc/build.js
@@ -113,7 +113,6 @@ function buildImage(opts, callback) {
             account: req.account,
             app: req.app,
             clientApiVersion: req.clientApiVersion,
-            requireExternalNetwork: true,  // use external network
             image: scratchImage,
             log: log,
             name: 'build_' + req_id,
diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 516d192..0d8361e 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -549,9 +549,8 @@ function addRulesToFWAPI(opts, rules, payload, callback) {
  *  External:
  *
  *   An external network is added in these cases:
- *    1. opts.requireExternalNetwork is set, or
- *    2. fabrics are *not* enabled, or
- *    3. fabrics are enabled and the user wants to expose ports
+ *    1. fabrics are *not* enabled, or
+ *    2. fabrics are enabled and the user wants to expose ports
  *
  *   The user can specify which external network is used by setting the
  *   'triton.network.public' container label (tag), this specifies the external
@@ -569,8 +568,6 @@ function addNetworksToPayload(opts, container, payload, callback) {
         'opts.config.overlay.externalPool');
     assert.optionalBool(opts.config.overlay.enabled,
         'opts.config.overlay.enabled');
-    assert.optionalBool(opts.requireExternalNetwork,
-        'opts.requireExternalNetwork');
     assert.func(callback, 'callback');
 
     var log = opts.log;
@@ -601,8 +598,7 @@ function addNetworksToPayload(opts, container, payload, callback) {
         },
 
         function addExternalNetwork(_, next) {
-            if (!opts.requireExternalNetwork && opts.config.overlay.enabled
-                && !publishingPorts(container)) {
+            if (!opts.config.overlay.enabled && !publishingPorts(container)) {
                 // DOCKER-1045: for fabrics, it is an error if the
                 // triton.network.public label is used and no ports are being
                 // published.
@@ -711,10 +707,8 @@ function externalNetworkByName(opts, container, payload, callback) {
  * or if 'bridge' is supplied, we will use the user's default fabric
  * network, stored in UFDS.
  *
- * Additionally, if the user is publishing ports or we have the
- * `requireExternalNetwork` option (currently used by docker build,
- * see DOCKER-705), we will also attach to configured network pool
- * (typically the external/public pool).
+ * Additionally, if the user is publishing ports we will also attach to the
+ * configured network pool (typically the external/public pool).
  */
 function defaultFabricNetwork(opts, payload, callback) {
     assert.object(opts, 'opts');
@@ -2843,8 +2837,6 @@ function createContainer(opts, callback) {
     assert.string(opts.req_id, 'opts.req_id');
     assert.object(opts.account, 'opts.account');
     assert.number(opts.clientApiVersion, 'opts.clientApiVersion');
-    assert.optionalBool(opts.requireExternalNetwork,
-        'opts.requireExternalNetwork');
 
     var log = opts.log || this.log;
     var name = opts.name;
@@ -2867,7 +2859,6 @@ function createContainer(opts, callback) {
         buildVmPayload({
             app: opts.app,
             config: config,
-            requireExternalNetwork: opts.requireExternalNetwork,
             image: opts.image,
             log: log,
             req_id: opts.req_id,
