commit 9042ab02ef82773566b7c2675464fc22d46af243 (refs/changes/93/4593/2)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2018-07-30T11:16:58+01:00 (1 year, 2 months ago)
    
    TRITON-639 amon-agent postinstall broken by TRITON-478

diff --git a/Makefile b/Makefile
index 6de743a..872dc27 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -61,6 +61,16 @@ else
 	MAKE = make
 	TAR ?= tar
 endif
+
+#
+# Due to the unfortunate nature of npm, the Node Package Manager, there appears
+# to be no way to assemble our dependencies without running the lifecycle
+# scripts.  These lifecycle scripts should not be run except in the context of
+# an agent installation or uninstallation, so we provide a magic environment
+# varible to disable them here.
+#
+NPM_ENV = SDC_AGENT_SKIP_LIFECYCLE=yes MAKE=$(MAKE)
+
 NODE_DEV := ./node_modules/.bin/node-dev
 TAP := ./node_modules/.bin/tap
 JSSTYLE_FLAGS := -f tools/jsstyle.conf
@@ -78,35 +88,38 @@ all: common plugins agent testbuild relay master dev sdc-scripts
 
 .PHONY: common
 common: | $(NPM_EXEC)
-	(cd common && MAKE=$(MAKE) $(NPM) install && $(NPM) link)
+	(cd common && $(NPM_ENV) $(NPM) install && $(NPM) link)
 
 .PHONY: plugins
 plugins: | $(NPM_EXEC)
-	(cd plugins && $(NPM) install && $(NPM) link)
+	(cd plugins && $(NPM_ENV) $(NPM) install && $(NPM) link)
 
 .PHONY: agent
 agent: common plugins | $(NPM_EXEC)
-	(cd agent && $(NPM) link amon-common amon-plugins && MAKE=$(MAKE) $(NPM) install)
+	(cd agent && $(NPM) link amon-common amon-plugins && \
+	$(NPM_ENV) $(NPM) install)
 
 .PHONY: relay
 relay: common plugins testbuild | $(NPM_EXEC)
-	(cd relay && $(NPM) link amon-common amon-plugins && MAKE=$(MAKE) $(NPM) install)
+	(cd relay && $(NPM) link amon-common amon-plugins && \
+	$(NPM_ENV) $(NPM) install)
 
 .PHONY: master
 master: common plugins | $(NPM_EXEC)
-	(cd master && $(NPM) link amon-common amon-plugins && MAKE=$(MAKE) $(NPM) install)
+	(cd master && $(NPM) link amon-common amon-plugins && \
+	$(NPM_ENV) $(NPM) install)
 
 # 'testbuild' is the name for building in the 'test' dir. Want 'make test'
 # to actually *run* the tests.
 .PHONY: testbuild
 testbuild: | $(NPM_EXEC)
-	(cd test && MAKE=$(MAKE) $(NPM) install)
+	(cd test && $(NPM_ENV) $(NPM) install)
 
 # "dev" is the name for the top-level dev package
 .PHONY: dev
 dev: common | $(NPM_EXEC)
 	$(NPM) link amon-common
-	$(NPM) install
+	$(NPM_ENV) $(NPM) install
 
 
 #
diff --git a/agent/pkg/postinstall.sh b/agent/pkg/postinstall.sh
index a6b7f23..681d852 100755
--- a/agent/pkg/postinstall.sh
+++ b/agent/pkg/postinstall.sh
@@ -28,7 +28,7 @@ fi
 set -o errexit
 set -o pipefail
 
-if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
+if [[ "$SDC_AGENT_SKIP_LIFECYCLE" == "yes" ]]; then
     echo "Skipping amon-agent postinstall (assuming build-time install)"
     exit 0
 fi
diff --git a/relay/pkg/postinstall.sh b/relay/pkg/postinstall.sh
index b375032..067ede3 100755
--- a/relay/pkg/postinstall.sh
+++ b/relay/pkg/postinstall.sh
@@ -18,9 +18,9 @@ fi
 
 set -o xtrace
 
-if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
-  echo "Skipping amon-relay postinstall (assuming build-time install)"
-  exit 0
+if [[ "$SDC_AGENT_SKIP_LIFECYCLE" == "yes" ]]; then
+    echo "Skipping amon-relay postinstall (assuming build-time install)"
+    exit 0
 fi
 
 export SMFDIR=$npm_config_smfdir
