commit c0a1ad751387329210237f6fed3a3aa91dbe18e2 (refs/changes/93/4593/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2018-07-27T14:20:47+00:00 (1 year, 2 months ago)
    
    TRITON-639 amon-agent postinstall broken by TRITON-478

diff --git a/agent/pkg/postinstall.sh b/agent/pkg/postinstall.sh
index a6b7f23..5b586b0 100755
--- a/agent/pkg/postinstall.sh
+++ b/agent/pkg/postinstall.sh
@@ -28,7 +28,8 @@ fi
 set -o errexit
 set -o pipefail
 
-if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
+if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]] && \
+	[[ -z "$SMF_FMRI" ]]; then
     echo "Skipping amon-agent postinstall (assuming build-time install)"
     exit 0
 fi
diff --git a/relay/pkg/postinstall.sh b/relay/pkg/postinstall.sh
index b375032..7e0495e 100755
--- a/relay/pkg/postinstall.sh
+++ b/relay/pkg/postinstall.sh
@@ -18,7 +18,8 @@ fi
 
 set -o xtrace
 
-if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
+if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]] && \
+	[[ -z "$SMF_FMRI" ]]; then
   echo "Skipping amon-relay postinstall (assuming build-time install)"
   exit 0
 fi
