commit fe0b9094b6dc91bf3a5934d7373424e98be63156 (refs/changes/75/3675/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-03-17T08:22:02+00:00 (1 year, 7 months ago)
    
    OS-6178 Support "routes" for KVM and bhyve instances

diff --git a/src/vm/node_modules/proptable.js b/src/vm/node_modules/proptable.js
index 4092fcd0..7bbb6c5e 100644
--- a/src/vm/node_modules/proptable.js
+++ b/src/vm/node_modules/proptable.js
@@ -1619,8 +1619,10 @@ exports.properties = {
         json: 'routes',
         payload: {
             allowed: {
+                'bhyve': ['create', 'receive'],
                 'joyent': ['create', 'receive'],
                 'joyent-minimal': ['create', 'receive'],
+                'kvm': ['create', 'receive'],
                 'lx': ['create', 'receive']
             },
             type: 'flat-object'
@@ -1635,8 +1637,10 @@ exports.properties = {
     }, set_routes: {
         payload: {
             allowed: {
+                'bhyve': ['update'],
                 'joyent': ['update'],
                 'joyent-minimal': ['update'],
+                'kvm': ['update'],
                 'lx': ['update']
             },
             type: 'flat-object'
