commit cc388fd9a0696d47353e165429601f37a321d55a (refs/changes/31/4231/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-06-15T09:58:32-05:00 (1 year, 4 months ago)
    
    Interface additions for route shootdowns

diff --git a/usr/src/uts/common/sys/overlay_target.h b/usr/src/uts/common/sys/overlay_target.h
index 9dce4e1cae..28c0559acb 100644
--- a/usr/src/uts/common/sys/overlay_target.h
+++ b/usr/src/uts/common/sys/overlay_target.h
@@ -304,6 +304,7 @@ typedef struct overlay_targ_list {
 #define	OVERLAY_TARG_CACHE_REMOVE	(OVERLAY_TARG_IOCTL | 0x32)
 #define	OVERLAY_TARG_CACHE_FLUSH	(OVERLAY_TARG_IOCTL | 0x33)
 #define	OVERLAY_TARG_CACHE_ITER		(OVERLAY_TARG_IOCTL | 0x34)
+#define	OVERLAY_TARG_CACHE_REMOVE_NET	(OVERLAY_TARG_IOCTL | 0x35)
 
 /*
  * This is a pretty arbitrary number that we're constraining ourselves to
@@ -322,11 +323,24 @@ typedef struct overlay_targ_cache_entry {
 	overlay_target_point_t	otce_dest;
 } overlay_targ_cache_entry_t;
 
+typedef struct overlay_targ_cache_net_entry {
+	struct in6_addr		otcne_src;
+	struct in6_addr		otcne_dst;
+	uint16_t		otcne_vlan;		/* src vlan */
+	uint8_t			otcne_src_prefixlen;
+	uint8_t			otcne_dst_prefixlen;
+} overlay_targ_cache_net_entry_t;
+
 typedef struct overlay_targ_cache {
 	datalink_id_t			otc_linkid;
 	overlay_targ_cache_entry_t	otc_entry;
 } overlay_targ_cache_t;
 
+typedef struct overlay_targ_cache_net {
+	datalink_id_t			otcn_linkid;
+	overlay_targ_cache_net_entry_t	otcn_entry;
+} overlay_targ_cache_net_t;
+
 typedef struct overlay_targ_cache_iter {
 	datalink_id_t			otci_linkid;
 	uint32_t			otci_pad;
