commit fc2d3fa11cf30c6e3e8be2e3edcd3b5b9c8f1f9d (refs/changes/14/5714/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-03-05T16:31:42+00:00 (7 months ago)
    
    OS-7634 enable zfs-test cliroot zpool_get_002_pos

diff --git a/usr/src/test/zfs-tests/runfiles/smartos.run b/usr/src/test/zfs-tests/runfiles/smartos.run
index 2443f6fc6a..18a491c403 100644
--- a/usr/src/test/zfs-tests/runfiles/smartos.run
+++ b/usr/src/test/zfs-tests/runfiles/smartos.run
@@ -24,7 +24,7 @@ post = cleanup
 outputdir = /var/tmp/test_results
 
 [/opt/zfs-tests/tests/functional/cli_root/zpool_get]
-tests = ['zpool_get_001_pos', 'zpool_get_003_pos',
+tests = ['zpool_get_001_pos', 'zpool_get_002_pos', 'zpool_get_003_pos',
     'zpool_get_004_neg']
 
 [/opt/zfs-tests/tests/functional/cli_root/zpool_import]
