From 45c78d9ff2735be8fc52a8fca2459016a7fa70d6 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Thu, 22 Jun 2017 16:25:48 -0400
Subject: [PATCH] OS-6204 async.forEachSeries RangeError with many inputs

---
 src/node_modules/async.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/node_modules/async.js b/src/node_modules/async.js
index 20afb3bd..7b7f97f2 100644
--- a/src/node_modules/async.js
+++ b/src/node_modules/async.js
@@ -150,7 +150,7 @@
                         callback();
                     }
                     else {
-                        iterate();
+                        setImmediate(iterate);
                     }
                 }
             });
-- 
2.21.0

