commit 45c78d9ff2735be8fc52a8fca2459016a7fa70d6 (refs/changes/37/2137/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2017-06-22T16:25:48-04:00 (2 years, 4 months ago)
    
    OS-6204 async.forEachSeries RangeError with many inputs

diff --git a/src/node_modules/async.js b/src/node_modules/async.js
index 20afb3bd..7b7f97f2 100644
--- a/src/node_modules/async.js
+++ b/src/node_modules/async.js
@@ -150,7 +150,7 @@
                         callback();
                     }
                     else {
-                        iterate();
+                        setImmediate(iterate);
                     }
                 }
             });
