commit a2ad851a560a8429a4c4d59f7793ef523b2987ab (refs/changes/37/2137/3)
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2017-06-22T23:16:27+00:00 (2 years, 4 months ago)
    
    OS-6203 VM.lookup breaks if "options" argument is empty

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 3ce23319..14974ede 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -1476,7 +1476,6 @@ exports.lookup = function (search, options, callback)
     var key;
     var matches;
     var need_fields = [];
-    var results = [];
     var tracers_obj;
     var transform;
 
@@ -1559,8 +1558,7 @@ exports.lookup = function (search, options, callback)
 
     lookup_opts = {log: log, fields: need_fields};
     vmload.getVmobjs(lookupFilter, lookup_opts, function gotVMs(err, vmobjs) {
-        var r;
-        var short_results = [];
+        var short_results;
 
         if (err) {
             callback(err);
@@ -1589,9 +1587,9 @@ exports.lookup = function (search, options, callback)
                 }
             }));
         } else {
-            for (r in vmobjs) {
-                short_results.push(results[r].uuid);
-            }
+            short_results = vmobjs.map(function getUUID(vmobj) {
+                return vmobj.uuid;
+            });
             callback(null, short_results);
         }
     });
