commit 4bcdf77a46d38729edacc72ccc07d4f2c1350219
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2019-07-10T11:36:38+02:00 (3 months ago)
    
    OS-7882 nvmeadm attach/detach of ignored namespace causes panic

diff --git a/usr/src/uts/common/io/nvme/nvme.c b/usr/src/uts/common/io/nvme/nvme.c
index cfe314b03e..4b6bfb06ab 100644
--- a/usr/src/uts/common/io/nvme/nvme.c
+++ b/usr/src/uts/common/io/nvme/nvme.c
@@ -13,7 +13,7 @@
  * Copyright 2018 Nexenta Systems, Inc.
  * Copyright 2016 Tegile Systems, Inc. All rights reserved.
  * Copyright (c) 2016 The MathWorks, Inc.  All rights reserved.
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -3858,6 +3858,9 @@ nvme_ioctl_detach(nvme_t *nvme, int nsid, nvme_ioctl_t *nioc, int mode,
 	if (nsid == 0)
 		return (EINVAL);
 
+	if (nvme->n_ns[nsid - 1].ns_ignore)
+		return (0);
+
 	rv = bd_detach_handle(nvme->n_ns[nsid - 1].ns_bd_hdl);
 	if (rv != DDI_SUCCESS)
 		rv = EBUSY;
@@ -3888,6 +3891,14 @@ nvme_ioctl_attach(nvme_t *nvme, int nsid, nvme_ioctl_t *nioc, int mode,
 
 	kmem_free(idns, sizeof (nvme_identify_nsid_t));
 
+	if (nvme->n_ns[nsid - 1].ns_ignore)
+		return (ENOTSUP);
+
+	if (nvme->n_ns[nsid - 1].ns_bd_hdl == NULL)
+		nvme->n_ns[nsid - 1].ns_bd_hdl = bd_alloc_handle(
+		    &nvme->n_ns[nsid - 1], &nvme_bd_ops, &nvme->n_prp_dma_attr,
+		    KM_SLEEP);
+
 	rv = bd_attach_handle(nvme->n_dip, nvme->n_ns[nsid - 1].ns_bd_hdl);
 	if (rv != DDI_SUCCESS)
 		rv = EBUSY;
