commit 2bc25fb6a29438146bb9e4dfca62c5dd4f0bc636 (refs/changes/04/3304/2)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2018-02-06T20:52:04-05:00 (1 year, 8 months ago)
    
    NAPI-xxx Clean up CI noise in NAPI

diff --git a/lib/models/aggregation.js b/lib/models/aggregation.js
index 0b9a7b5..e391fde 100644
--- a/lib/models/aggregation.js
+++ b/lib/models/aggregation.js
@@ -219,8 +219,8 @@ function validateMACs(opts, name, list, callback) {
 
         for (var i = 0; i < nicObjs.length; i++) {
             // Make sure all NICs have the same belongs_to_uuid
-            if (nicObjs[i].params.belongs_to_uuid
-                !== nicObjs[0].params.belongs_to_uuid) {
+            if (nicObjs[i].params.belongs_to_uuid !==
+                nicObjs[0].params.belongs_to_uuid) {
                 callback(errors.invalidParam(name,
                     constants.msg.AGGR_MATCH));
                 return;
diff --git a/lib/models/network.js b/lib/models/network.js
index b2fa213..15c52cc 100644
--- a/lib/models/network.js
+++ b/lib/models/network.js
@@ -390,8 +390,8 @@ function provisionRangeUpdates(app, log, network, update, callback) {
     var toMove = [];
 
     if (update.hasOwnProperty('provision_start_ip') &&
-        update.provision_start_ip.compare(network.params.provision_start_ip)
-        !== 0) {
+        update.provision_start_ip.compare(
+            network.params.provision_start_ip) !== 0) {
         toMove.push({
             before: util_ip.ipAddrMinus(network.params.provision_start_ip, 1),
             after: util_ip.ipAddrMinus(update.provision_start_ip, 1)
@@ -399,8 +399,8 @@ function provisionRangeUpdates(app, log, network, update, callback) {
     }
 
     if (update.hasOwnProperty('provision_end_ip') &&
-        update.provision_end_ip.compare(network.params.provision_end_ip)
-        !== 0) {
+        update.provision_end_ip.compare(
+            network.params.provision_end_ip) !== 0) {
         toMove.push({
             before: util_ip.ipAddrPlus(network.params.provision_end_ip, 1),
             after: util_ip.ipAddrPlus(update.provision_end_ip, 1)
