From 6ebb5f4197b989a01c00680af9c024c698750a8d Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 17 Oct 2016 17:22:09 -0700
Subject: [PATCH] MANTA-3008 binder always serving TTLs as 5 seconds

---
 lib/server.js | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index 6821828..0b7f444 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -162,7 +162,15 @@ function resolve(options, query, cb) {
                         query.setError('eserver');
                 } else {
                         var addr;
-                        var ttl = record.ttl;
+
+                        var ttl = (record[record.type] || {}).ttl;
+                        /*
+                         * Default the TTL to 30 seconds (the default ZK
+                         * session timeout)
+                         */
+                        if (ttl === undefined)
+                                ttl = 30;
+
                         if (service !== undefined &&
                             record.type !== 'service') {
                                 query.setError('eserver');
@@ -229,6 +237,7 @@ function resolve(options, query, cb) {
                                             ports.length < 1)
                                                 ports = [s.port];
                                         var ar, sr, nm;
+                                        var rttl = host[host.type].ttl;
                                         if (service !== undefined) {
                                                 nm = host.name + '.' + domain;
                                                 ports.forEach(function (p) {
@@ -240,11 +249,11 @@ function resolve(options, query, cb) {
                                                 });
                                                 ar = new ARecord(a);
                                                 query.addAdditional(nm, ar,
-                                                    ttl);
+                                                    rttl);
                                         } else {
                                                 ar = new ARecord(a);
                                                 query.addAnswer(domain, ar,
-                                                    ttl);
+                                                    rttl);
                                         }
                                 }
                                 break;
-- 
2.21.0

