commit 6acbe8351238106e57396b08a1b78f37fe0f45bb (refs/changes/71/1671/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-03-16T14:11:15+01:00 (2 years, 7 months ago)
    
    TOOLS-1724 sdcadm trying to read image_uuid of undefined

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 2cca542..6ee8435 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -2785,7 +2785,7 @@ SdcAdm.prototype.getCandidateImages = function getCandidateImages(opts, cb) {
             currImgUuids = Object.keys(currImgUuids);
             if (currImgUuids.length === 0) {
                 // No insts -> use the image_uuid set on the service.
-                if (opts.service.params.image_uuid) {
+                if (opts.service.params && opts.service.params.image_uuid) {
                     currImgUuids.push(opts.service.params.image_uuid);
                 }
             }
