From 6acbe8351238106e57396b08a1b78f37fe0f45bb Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 16 Mar 2017 14:11:15 +0100
Subject: [PATCH] TOOLS-1724 sdcadm trying to read image_uuid of undefined

---
 lib/sdcadm.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 2cca542..6ee8435 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -2785,7 +2785,7 @@ SdcAdm.prototype.getCandidateImages = function getCandidateImages(opts, cb) {
             currImgUuids = Object.keys(currImgUuids);
             if (currImgUuids.length === 0) {
                 // No insts -> use the image_uuid set on the service.
-                if (opts.service.params.image_uuid) {
+                if (opts.service.params && opts.service.params.image_uuid) {
                     currImgUuids.push(opts.service.params.image_uuid);
                 }
             }
-- 
2.21.0

