From d938af94ee94987e5672f6c64112c096ae8d58e0 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 10 Oct 2016 16:46:32 -0400
Subject: [PATCH] RELENG-612 prefer (git+)https urls

---
 .gitmodules  | 8 ++++----
 package.json | 4 ++--
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 7625872..d11873b 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,15 +1,15 @@
 [submodule "deps/restdown"]
 	path = deps/restdown
-	url = git://github.com/trentm/restdown.git
+	url = https://github.com/trentm/restdown.git
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
-	url = git://github.com/davepacheco/javascriptlint.git
+	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/davepacheco/jsstyle.git
 [submodule "deps/sdc-scripts"]
 	path = deps/sdc-scripts
 	url = git@github.com:joyent/sdc-scripts.git
 [submodule "deps/restdown-brand-remora"]
 	path = deps/restdown-brand-remora
-	url = git@github.com:joyent/restdown-brand-remora.git
+	url = https://github.com/joyent/restdown-brand-remora.git
diff --git a/package.json b/package.json
index bd5541b..35dc8b5 100644
--- a/package.json
+++ b/package.json
@@ -15,8 +15,8 @@
     "bunyan": "0.22.0",
     "sigyan": "0.2.0",
     "backoff": "2.3.0",
-    "moray": "git+ssh://git@github.com:joyent/node-moray.git#b84ef0e",
-    "sdc-clients": "git+ssh://git@github.com:joyent/node-sdc-clients.git#a7366a02",
+    "moray": "git+https://github.com/joyent/node-moray.git#b84ef0e",
+    "sdc-clients": "git+https://git@github.com/joyent/node-sdc-clients.git#a7366a02",
     "ldapjs": "0.7.0",
     "vasync": "1.4.0",
     "clone": "0.1.6"
-- 
2.21.0

