commit 5912959789a06fab44eaacff2ed6575e603df630 (refs/changes/11/2711/2)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-10-09T20:58:00+00:00 (2 years ago)
    
    MANATEE-372 disable wal sender timeout

diff --git a/etc/9.2/postgresql.manta.coal.conf b/etc/9.2/postgresql.manta.coal.conf
index de4b415..56cb4a2 100644
--- a/etc/9.2/postgresql.manta.coal.conf
+++ b/etc/9.2/postgresql.manta.coal.conf
@@ -210,7 +210,7 @@ full_page_writes = off          # recover from partial page writes
 max_wal_senders = 15    # max number of walsender processes
         # (change requires restart)
 wal_keep_segments = 50    # in logfile segments, 16MB each; 0 disables
-#replication_timeout = 60s  # in milliseconds; 0 disables
+replication_timeout = 0   # in milliseconds; 0 disables
 
 # - Master Server -
 
diff --git a/etc/9.2/postgresql.manta.lab.conf b/etc/9.2/postgresql.manta.lab.conf
index 4b0e9f8..b1c0369 100644
--- a/etc/9.2/postgresql.manta.lab.conf
+++ b/etc/9.2/postgresql.manta.lab.conf
@@ -210,7 +210,7 @@ full_page_writes = off      # recover from partial page writes
 max_wal_senders = 15    # max number of walsender processes
         # (change requires restart)
 wal_keep_segments = 100    # in logfile segments, 16MB each; 0 disables
-#replication_timeout = 60s  # in milliseconds; 0 disables
+replication_timeout = 0    # in milliseconds; 0 disables
 
 # - Master Server -
 
diff --git a/etc/9.2/postgresql.manta.production.conf b/etc/9.2/postgresql.manta.production.conf
index 8c9ccbd..6094e18 100644
--- a/etc/9.2/postgresql.manta.production.conf
+++ b/etc/9.2/postgresql.manta.production.conf
@@ -210,7 +210,7 @@ checkpoint_timeout = 30s # range 30s-1h
 max_wal_senders = 15    # max number of walsender processes
         # (change requires restart)
 wal_keep_segments = 1000    # in logfile segments, 16MB each; 0 disables
-#replication_timeout = 60s  # in milliseconds; 0 disables
+replication_timeout = 0     # in milliseconds; 0 disables
 
 # - Master Server -
 
diff --git a/etc/9.6/postgresql.manta.coal.conf b/etc/9.6/postgresql.manta.coal.conf
index d82ef9e..dc6a3da 100644
--- a/etc/9.6/postgresql.manta.coal.conf
+++ b/etc/9.6/postgresql.manta.coal.conf
@@ -239,7 +239,7 @@ max_wal_size = 480MB
 max_wal_senders = 15            # max number of walsender processes
                                 # (change requires restart)
 wal_keep_segments = 50          # in logfile segments, 16MB each; 0 disables
-#wal_sender_timeout = 60s       # in milliseconds; 0 disables
+wal_sender_timeout = 0          # in milliseconds; 0 disables
 
 #max_replication_slots = 0      # max number of replication slots
                                 # (change requires restart)
diff --git a/etc/9.6/postgresql.manta.lab.conf b/etc/9.6/postgresql.manta.lab.conf
index 031ccb2..d745fe8 100644
--- a/etc/9.6/postgresql.manta.lab.conf
+++ b/etc/9.6/postgresql.manta.lab.conf
@@ -239,7 +239,7 @@ max_wal_size = 480MB
 max_wal_senders = 15            # max number of walsender processes
                                 # (change requires restart)
 wal_keep_segments = 100         # in logfile segments, 16MB each; 0 disables
-#wal_sender_timeout = 60s       # in milliseconds; 0 disables
+wal_sender_timeout = 0          # in milliseconds; 0 disables
 
 #max_replication_slots = 0      # max number of replication slots
                                 # (change requires restart)
diff --git a/etc/9.6/postgresql.manta.production.conf b/etc/9.6/postgresql.manta.production.conf
index 6a09688..72f5d84 100644
--- a/etc/9.6/postgresql.manta.production.conf
+++ b/etc/9.6/postgresql.manta.production.conf
@@ -239,7 +239,7 @@ max_wal_size = 480MB
 max_wal_senders = 15            # max number of walsender processes
                                 # (change requires restart)
 wal_keep_segments = 1000        # in logfile segments, 16MB each; 0 disables
-#wal_sender_timeout = 60s       # in milliseconds; 0 disables
+wal_sender_timeout = 0          # in milliseconds; 0 disables
 
 #max_replication_slots = 0      # max number of replication slots
                                 # (change requires restart)
