commit 7f5f7ef8f09cc5524b7541b51e6145b90bfc9682
Author: Jason King <jason.king@joyent.com>
Date:   2019-04-09T19:09:37+00:00 (6 months ago)
    
    OS-7723 libdemangle missing dependency on libcustr

diff --git a/usr/src/lib/Makefile b/usr/src/lib/Makefile
index 669cffa0d3..88fb25ddf2 100644
--- a/usr/src/lib/Makefile
+++ b/usr/src/lib/Makefile
@@ -615,6 +615,7 @@ libcmdutils:	libavl
 libcpc:		libpctx
 libcrypt:	libgen
 libctf:		libdwarf
+libdemangle: libcustr
 libdevid:	libdevinfo
 libdevinfo:	libsec libgen
 libdhcpagent:	libdhcputil libuuid libdlpi libcontract
