commit 7cc760631132f531a791438f2b6470840d635268 (refs/changes/76/1976/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-05-18T16:18:27-07:00 (2 years, 5 months ago)
    
    MANTA-3274 add a tools/rsync-to to sdc-manta

diff --git a/tools/rsync-to b/tools/rsync-to
new file mode 100755
index 0000000..a838081
--- /dev/null
+++ b/tools/rsync-to
@@ -0,0 +1,39 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright (c) 2017, Joyent, Inc.
+#
+
+#
+# Rsync the master in this working copy to the install on the given HN.
+#
+
+#set -o xtrace
+set -o errexit
+
+TOP=$(cd $(dirname $0)/../; pwd)
+NODE=$1
+
+if [[ -z "$MANTA_ZONE" ]]; then
+    MANTA_ZONE=$(ssh $NODE "vmadm lookup -1 alias=manta0" 2>/dev/null)
+fi
+echo "MANTA_ZONE: $MANTA_ZONE"
+
+extraOpts=
+if [[ $(uname -s) != "SunOS" ]]; then
+    extraOpts="--exclude *.node --exclude build"
+fi
+
+rsync -av ${TOP}/ \
+    $NODE:/zones/$MANTA_ZONE/root/opt/smartdc/manta-deployment/ \
+    $extraOpts \
+    --exclude .git/ \
+    --exclude /boot/ \
+    --exclude /deps/ \
+    --exclude /tools/ \
+    --exclude /tmp/
