commit a3e48b2b7e203fa3e6912b4918bc82242cddc22b (refs/changes/03/4203/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2018-06-13T20:38:42+00:00 (1 year, 4 months ago)
    
    OS-7019 missing break in hma_fpu_alloc

diff --git a/usr/src/uts/i86pc/os/hma_fpu.c b/usr/src/uts/i86pc/os/hma_fpu.c
index f207470e85..14cfa8baed 100644
--- a/usr/src/uts/i86pc/os/hma_fpu.c
+++ b/usr/src/uts/i86pc/os/hma_fpu.c
@@ -105,6 +105,7 @@ hma_fpu_alloc(int kmflag)
 	case FP_FXSAVE:
 		bzero(fpu->hf_guest_fpu.fpu_regs.kfpu_u.kfpu_generic,
 		    sizeof (struct fxsave_state));
+		break;
 	case FP_XSAVE:
 		bzero(fpu->hf_guest_fpu.fpu_regs.kfpu_u.kfpu_generic,
 		    cpuid_get_xsave_size());
