From 8c14019e8e380ecf55e119765f81930d3c864f28 Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Fri, 16 Mar 2018 11:28:28 -0500
Subject: [PATCH] Fix .WAIT in Makefile

---
 usr/src/lib/varpd/Makefile | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/usr/src/lib/varpd/Makefile b/usr/src/lib/varpd/Makefile
index daa849572e..a0e4537ab9 100644
--- a/usr/src/lib/varpd/Makefile
+++ b/usr/src/lib/varpd/Makefile
@@ -10,10 +10,10 @@
 #
 
 #
-# Copyright 2015 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 #
 
-SUBDIRS = libvarpd .WAIT direct files svp
+SUBDIRS = libvarpd direct files svp
 
 all :=		TARGET = all
 clean :=	TARGET = clean
@@ -26,6 +26,7 @@ lint :=		TARGET = lint
 .KEEP_STATE:
 
 all clean clobber install install_h check lint: $(SUBDIRS)
+direct files svp: libvarpd
 
 $(SUBDIRS):  FRC
 	@cd $@; pwd; $(MAKE) $(TARGET)
-- 
2.21.0

