From 66b7928da196f9135ab6493974cd048e33791bf7 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Mon, 29 Aug 2016 12:49:30 -0700
Subject: [PATCH] TOOLS-1535 gerrit could be better with names and faces

---
 images/appserver/etc/gerrit.config.base | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/images/appserver/etc/gerrit.config.base b/images/appserver/etc/gerrit.config.base
index d1e6159..ef22120 100644
--- a/images/appserver/etc/gerrit.config.base
+++ b/images/appserver/etc/gerrit.config.base
@@ -37,6 +37,18 @@
 [cache]
 	directory = cache
 
+#
+# The "web_sessions" cache tracks open user sessions.  These effectively map a
+# cookie in your browser back to an identity.  The maximum age of entries in
+# this cache determines how often a user must sign in again.
+#
+# Note that the Gerrit documentation suggests that a value of 0 will allow
+# entries to remain in the cache until it fills.  This appears to not to be the
+# case in practice.  Instead, we have selected a large positive value.
+#
+[cache "web_sessions"]
+	maxAge = 1 y
+
 [plugins]
 	allowRemoteAdmin = true
 
-- 
2.21.0

