From c9fdbf3286ca0e3c2de0edb1180dc7f06adfbc99 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Wed, 21 Dec 2016 15:18:48 -0500
Subject: [PATCH] better error messages

---
 CHANGES.md    | 4 +++-
 lib/config.js | 5 ++---
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index a3b6b12..0c6090e 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,8 +10,10 @@ Known issues:
 - PUBAPI-1205/PUBAPI-1171/PUBAPI-1351 The handling of legacy `SDC_*`
   environmental variables has been cleaned up.  These environment
   variables are used for compatibility with the node-smartdc toolset.
-   * `SDC_TESTING` is now evaluated the same way as noe-smartdc.  Any
+   * `SDC_TESTING` is now evaluated the same way as node-smartdc.  Any
      set value but the empty string is true.
+   * Errors on boolean environment variables will now identify
+     the variable at fault.
 
 - [joyent/node-triton#137] Improve the handling for the getting started case
   when a user may not have envvars or a profile setup.
diff --git a/lib/config.js b/lib/config.js
index df790b7..a8b03c5 100644
--- a/lib/config.js
+++ b/lib/config.js
@@ -288,13 +288,12 @@ function _loadEnvProfile() {
     {
         return null;
     }
-
     if (process.env.TRITON_TLS_INSECURE) {
         envProfile.insecure = common.boolFromString(
-            process.env.TRITON_TLS_INSECURE);
+            process.env.TRITON_TLS_INSECURE, undefined, 'TRITON_TLS_INSECURE');
     } else if (process.env.SDC_TLS_INSECURE) {
         envProfile.insecure = common.boolFromString(
-            process.env.SDC_TLS_INSECURE);
+            process.env.SDC_TLS_INSECURE, undefined, 'SDC_TLS_INSECURE');
     } else if (process.env.SDC_TESTING) { // deprecated
         envProfile.insecure = true;
     }
-- 
2.21.0

