commit 8ddd87e94fc3b33c4b99c444d25eeeebd2a54997 (refs/changes/30/3230/4)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-01-19T13:34:04+00:00 (1 year, 9 months ago)
    
    OS-6562 lx_mremap broken under DEBUG build

diff --git a/usr/src/uts/common/brand/lx/syscall/lx_mem.c b/usr/src/uts/common/brand/lx/syscall/lx_mem.c
index 836edbfa14..4940f5af5f 100644
--- a/usr/src/uts/common/brand/lx/syscall/lx_mem.c
+++ b/usr/src/uts/common/brand/lx/syscall/lx_mem.c
@@ -12,7 +12,7 @@
 /*
  * Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
  * Use is subject to license terms.
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -513,7 +513,8 @@ lx_get_mapping(uintptr_t find_addr, size_t find_size, lx_segmap_t *mp,
 	uint_t prot;
 	caddr_t saddr, eaddr, naddr;
 
-	AS_LOCK_ENTER(as, RW_READER);
+	/* pr_getprot asserts that the as is held as a writer */
+	AS_LOCK_ENTER(as, RW_WRITER);
 
 	seg = as_segat(as, (caddr_t)find_addr);
 	if (seg == NULL || (seg->s_flags & S_HOLE) != 0) {
@@ -1005,12 +1006,17 @@ lx_mremap(uintptr_t old_addr, size_t old_size, size_t new_size, int flags,
 	mflags |= MAP_SHARED;
 
 	rval = (long)smmap64((void *)new_addr, new_size, prot, mflags, fd, off);
+
+	/* "close" the temporary fd for the mapping */
+	mutex_enter(&fp->f_tlock);
+	setf(fd, NULL);
+	unfalloc(fp);
+
+	/* check the result from smmap64 */
 	if (ttolwp(curthread)->lwp_errno != 0) {
-		(void) close(fd);
 		rval = set_errno(ENOMEM);
 		goto out;
 	}
-	(void) close(fd);
 
 	/*
 	 * Our mapping succeeded; we're now going to rip down the old mapping.
@@ -1083,7 +1089,7 @@ lx_u2u_copy(void *src, void *dst, size_t len)
 			as_pageunlock(p_as, ppa_src, sp, mlen, S_READ);
 			return (EFAULT);
 		}
-		bcopy(sp, dp, mlen);
+		ucopy(sp, dp, mlen);
 		no_fault();		/* calls smap_enable */
 
 		as_pageunlock(p_as, ppa_dst, dp, mlen, S_WRITE);
