commit ce8e912c0b6039a2e86e7038f405aa3a4852c943
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-05-07T15:35:03+00:00 (5 months ago)
    
    OS-7775 platform builds on base-64 need gcc4.x to bootstrap

diff --git a/configure b/configure
index 75bff507..e99bfe64 100755
--- a/configure
+++ b/configure
@@ -91,6 +91,9 @@ function install_pkgin
 	pkglist="$pkglist p5-XML-Parser gettext python27 py27-expat"
 	pkglist="$pkglist coreutils gsed pkg_alternatives cdrtools"
 	pkglist="$pkglist py27-sqlite3 nasm pigz"
+	# If /opt/local/bin/gcc isn't 4.x, install gcc 4.9 also,
+	# needed in order to build our local copy of gcc 4.4.
+	pkglist="$pkglist $PKGSRC_GCC4X_PKG"
 
 	for pkg in $pkglist; do
 		if ! pkg_info -qe $pkg; then
@@ -480,6 +483,14 @@ done
 
 PRIMARY_COMPILER_VER=$(echo $PRIMARY_COMPILER | sed 's/^gcc//')
 
+HOST_GCC_IS_4X=$(/opt/local/bin/gcc --version | head -1 | fgrep '(GCC) 4')
+if [[ -n "$HOST_GCC_IS_4X" ]]; then
+	HOST_GCC4_PREFIX=/opt/local/bin
+else
+	HOST_GCC4_PREFIX=/opt/local/gcc49/bin
+	PKGSRC_GCC4X_PKG="gcc49"
+fi
+
 cat >build.env <<EOF
 FORCE_STRAP_REBUILD=$FORCE_STRAP_REBUILD
 ILLUMOS_CLOBBER=$ILLUMOS_CLOBBER
@@ -487,6 +498,7 @@ ILLUMOS_ENABLE_DEBUG=$ILLUMOS_ENABLE_DEBUG
 PRIMARY_COMPILER=$PRIMARY_COMPILER
 PRIMARY_COMPILER_VER=$PRIMARY_COMPILER_VER
 SHADOW_COMPILERS=$SHADOW_COMPILERS
+HOST_GCC4_PREFIX=$HOST_GCC4_PREFIX
 EOF
 
 echo "Doing pre-flight checks... \c "
