commit 29a8f38032d82bce3aa87cd4fc5e5e2742a54d7e (refs/changes/32/4832/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-09-12T23:56:05+00:00 (1 year, 1 month ago)
    
    MANTA-3945 garbage-collector loads index shard list incorrectly

diff --git a/cmd/server.js b/cmd/server.js
index a4599d0..d4b8eb6 100644
--- a/cmd/server.js
+++ b/cmd/server.js
@@ -182,13 +182,29 @@ load_index_shard_range(ctx, done)
 		return;
 	}
 
-	ctx.ctx_cfg.index_shard_lo = parseInt(
-			index_shards[0].host.charAt(0), 10);
+	/*
+	 * While shards in production Manta deployments are numeric by
+	 * convention, they don't have to be. This assumption needs to
+	 * be removed.
+	 */
+	function parse_shard_number(shard) {
+		return (parseInt(shard.substring(0, shard.indexOf('.')), 10));
+	}
+
+	var init = parse_shard_number(index_shards[0].host);
+
+	ctx.ctx_cfg.index_shard_lo = init;
+	ctx.ctx_cfg.index_shard_hi = init;
 
 	index_shards.forEach(function (shard) {
-		if (shard.last) {
-			ctx.ctx_cfg.index_shard_hi = parseInt(
-				shard.host.charAt(0), 10);
+		var shard_num = parse_shard_number(shard.host);
+
+		if (ctx.ctx_cfg.index_shard_lo > shard_num) {
+			ctx.ctx_cfg.index_shard_lo = shard_num;
+		}
+
+		if (ctx.ctx_cfg.index_shard_hi < shard_num) {
+			ctx.ctx_cfg.index_shard_hi = shard_num;
 		}
 	});
 
