From 647bc2e6e30e21abf80a12f4ea32e10614b30439 Mon Sep 17 00:00:00 2001
From: Jason King <jason.brian.king@gmail.com>
Date: Tue, 13 Jun 2017 22:38:53 +0000
Subject: [PATCH] OS-6186 illumos#8369 breaks smartos-live build

---
 src/dockerinit/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/dockerinit/Makefile b/src/dockerinit/Makefile
index 4b5d03d0..8d7639c8 100644
--- a/src/dockerinit/Makefile
+++ b/src/dockerinit/Makefile
@@ -28,7 +28,7 @@ GXX =		$(NATIVEDIR)/usr/bin/g++
 SYSINCDIRS =	/usr/include
 SYSLIBDIRS =	/usr/lib /lib
 
-CPPFLAGS =	$(SYSINCDIRS:%=-isystem $(DESTDIR)/%)
+CPPFLAGS =	$(SYSINCDIRS:%=-isystem $(DESTDIR)/%) -D_LARGEFILE64_SOURCE
 LDFLAGS =	$(SYSLIBDIRS:%=-L$(DESTDIR)/%) \
 		-Wl,-zassert-deflib -Wl,-zfatal-warnings
 
-- 
2.21.0

