commit b9b88b4cc0f7ea99aa4d5948832281d3ad26cfe7 (refs/changes/98/3298/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-02-05T16:40:04-08:00 (1 year, 8 months ago)
    
    OS-6585 Update mdb_v8 to 1.2.1

diff --git a/mdb_v8/Makefile b/mdb_v8/Makefile
index 6c57dfd..4f6828a 100644
--- a/mdb_v8/Makefile
+++ b/mdb_v8/Makefile
@@ -23,7 +23,7 @@
 # 3. From one directory up, run tar cvzf mdb_v8-<ver>.tar.gz mdb_v8-<ver>
 #
 
-VER =	mdb_v8-1.1.4
+VER =	mdb_v8-1.2.1
 
 include ../Makefile.defs
 include ../Makefile.targ
diff --git a/mdb_v8/mdb_v8-1.1.4.tar.gz b/mdb_v8/mdb_v8-1.1.4.tar.gz
deleted file mode 100644
index 5554cc9..0000000
Binary files a/mdb_v8/mdb_v8-1.1.4.tar.gz and /dev/null differ
diff --git a/mdb_v8/mdb_v8-1.2.1.tar.gz b/mdb_v8/mdb_v8-1.2.1.tar.gz
new file mode 100644
index 0000000..7737e5d
Binary files /dev/null and b/mdb_v8/mdb_v8-1.2.1.tar.gz differ
