commit a651bd57db8892570251d928b06325c03bce3dc5 (refs/changes/48/1548/2)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2017-02-21T11:24:46-08:00 (2 years, 8 months ago)
    
    TOOLS-1701 sdcnode should support `make -j`

diff --git a/tools/build-a-node b/tools/build-a-node
index 66b68a4..c8dfe90 100755
--- a/tools/build-a-node
+++ b/tools/build-a-node
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -125,7 +125,12 @@ buildEnv=$(echo "$buildOpts" | json build_env)
 if [[ -n "$buildEnv" ]]; then
     eval "export $buildEnv"
 fi
-make V=1 CC=$GCC CXX=$GXX
+
+JOBS_ARG=
+if [[ -n ${MAKE_JOBS} ]]; then
+    JOBS_ARG="-j${MAKE_JOBS}"
+fi
+make ${JOBS_ARG} V=1 CC=$GCC CXX=$GXX
 make V=1 CC=$GCC CXX=$GXX install
 
 # Optionally include GCC runtime libs. If a version of GCC was used to compile
