commit dddb38750360f402dcabdd58726616137f511fa3 (refs/changes/12/3112/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-12-14T11:53:16-08:00 (1 year, 10 months ago)
    
    SAPI-299 improve some instances.test.js debugability

diff --git a/test/instances.test.js b/test/instances.test.js
index 136cde4..46e0143 100644
--- a/test/instances.test.js
+++ b/test/instances.test.js
@@ -188,7 +188,7 @@ test('put/get/del vm instance', function (t) {
             }
         }
 
-        t.ok(found, 'found service' + inst.uuid);
+        t.ok(found, 'found instance ' + inst.uuid);
     };
 
     var uri_inst = '/instances/' + inst.uuid;
@@ -770,6 +770,7 @@ test('invalid zone parameters', function (t) {
      */
     inst.params.ram = 10 * 1024 * 1024 * 1024 * 1024;  // 10 TB
     inst.params.networks = [ { name: 'admin', ip: '192.168.1.1'} ];
+    inst.params.alias = common.getUniqueTestResourceName('sapitest-invalidram');
 
     var uri_inst = '/instances/' + inst.uuid;
 
