From dddb38750360f402dcabdd58726616137f511fa3 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 14 Dec 2017 11:53:16 -0800
Subject: [PATCH] SAPI-299 improve some instances.test.js debugability

---
 test/instances.test.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/test/instances.test.js b/test/instances.test.js
index 136cde4..46e0143 100644
--- a/test/instances.test.js
+++ b/test/instances.test.js
@@ -188,7 +188,7 @@ test('put/get/del vm instance', function (t) {
             }
         }
 
-        t.ok(found, 'found service' + inst.uuid);
+        t.ok(found, 'found instance ' + inst.uuid);
     };
 
     var uri_inst = '/instances/' + inst.uuid;
@@ -770,6 +770,7 @@ test('invalid zone parameters', function (t) {
      */
     inst.params.ram = 10 * 1024 * 1024 * 1024 * 1024;  // 10 TB
     inst.params.networks = [ { name: 'admin', ip: '192.168.1.1'} ];
+    inst.params.alias = common.getUniqueTestResourceName('sapitest-invalidram');
 
     var uri_inst = '/instances/' + inst.uuid;
 
-- 
2.21.0

