commit 6abc5d64d7868546ae9e01ba3044751f080e306a (refs/changes/20/3120/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2017-12-18T12:07:35-05:00 (1 year, 10 months ago)
    
    TRITON-24 node-triton ListNetworkIPs has unordered results

diff --git a/lib/common.js b/lib/common.js
index 2347242..a55ab4e 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -1301,7 +1301,26 @@ function argvFromLine(line) {
     return argv;
 }
 
+/*
+ * Convert an IPv4 address (as a string) to a number
+ */
+function ipv4ToLong(ip) {
+    var l = 0;
+
+    assert.string(ip, 'ip');
+    ip.split('.').forEach(function processIpOctet(octet) {
+        octet = parseInt(octet, 10);
+
+        assert.number(octet, 'octet');
+        assert(octet >= 0, 'octet >= 0');
+        assert(octet < 256, 'octet < 256');
 
+        l <<= 8;
+        l += octet;
+    });
+
+    return l;
+}
 
 //---- exports
 
@@ -1339,6 +1358,7 @@ module.exports = {
     objFromKeyValueArgs: objFromKeyValueArgs,
     argvFromLine: argvFromLine,
     jsonPredFromKv: jsonPredFromKv,
-    monotonicTimeDiffMs: monotonicTimeDiffMs
+    monotonicTimeDiffMs: monotonicTimeDiffMs,
+    ipv4ToLong: ipv4ToLong
 };
 // vim: set softtabstop=4 shiftwidth=4:
diff --git a/lib/do_network/do_ip/do_list.js b/lib/do_network/do_ip/do_list.js
index 854045d..9c010be 100644
--- a/lib/do_network/do_ip/do_list.js
+++ b/lib/do_network/do_ip/do_list.js
@@ -41,7 +41,21 @@ function do_list(subcmd, opts, args, callback) {
     }
     columns = columns.split(',');
 
-    var sort = opts.s.split(',');
+    var sort = opts.s.split(',').map(function mapSort(field) {
+        var obj = {
+            field: field
+        };
+
+        switch (field) {
+        case 'ip':
+            obj.keyFunc = common.ipv4ToLong;
+            break;
+        default:
+            break;
+        }
+
+        return obj;
+    });
 
     vasync.pipeline({arg: {cli: this.top}, funcs: [
         common.cliSetupTritonApi,
