commit 7faab0752344c5c3f7b7167c649b5c02410dabb6 (refs/changes/90/3590/4)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-04-11T12:55:55-07:00 (1 year, 6 months ago)
    
    TRITON-53 x-account image clone

diff --git a/CHANGES.md b/CHANGES.md
index 20db701..2723c89 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,10 @@
 # IMGAPI changelog
 
+## 4.4.0
+
+- TRITON-53 x-account image clone, adds a new CloneImage
+  /images/:uuid/clone endpoint.
+
 ## 4.3.0
 
 - TRITON-178 Add support for image creation with bhyve brand (requires platform
diff --git a/docs/index.md b/docs/index.md
index d2f2bd9..f7aa669 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -621,6 +621,7 @@ and relevant for images in an IMGAPI server that uses [channels](#channels).
 | [EnableImage](#EnableImage)                       | POST /images/:uuid?action=enable                           | Enable the image.                                                             |
 | [AddImageAcl](#AddImageAcl)                       | POST /images/:uuid/acl?action=add                          | Add account UUIDs to the image ACL.                                           |
 | [RemoveImageAcl](#RemoveImageAcl)                 | POST /images/:uuid/acl?action=remove                       | Remove account UUIDs from the image ACL.                                      |
+| [CloneImage](#CloneImage)                         | POST /images/:uuid/clone                                   | Clone this image.
 | [AddImageIcon](#AddImageIcon)                     | POST /images/:uuid/icon                                    | Add the image icon.                                                           |
 | [GetImageIcon](#GetImageIcon)                     | GET /images/:uuid/icon                                     | Get the image icon file.                                                      |
 | [DeleteImageIcon](#DeleteImageIcon)               | DELETE /images/:uuid/icon                                  | Remove the image icon.                                                        |
@@ -1900,6 +1901,66 @@ CLI tool:
 
 
 
+## CloneImage (POST /images/:uuid/clone?account=:account)
+
+Clone this image. The provided account must be on the image ACL - see
+[AddImageAcl](#AddImageAcl). This makes a copy of the given image (including
+origin images). The newly cloned image will have account as the image owner and
+the cloned image will have an empty ACL. A successful image clone operation also
+removes the given account from the original image ACL (as there is no need to
+have it shared with account anymore).
+
+### Inputs
+
+| Field                 | Type | Required? | Default | Notes                                                |
+| --------------------- | ---- | --------- | ------- | ---------------------------------------------------- |
+| account (query param) | UUID | Yes       | -       | This account the cloned image will be assigned to.   |
+
+### Returns
+
+The cloned image object.
+
+### Errors
+
+See [Errors](#errors) section above.
+
+### Example
+
+Raw API tool:
+
+    $ sdc-imgapi /images/e70502b0-705e-498e-a810-53a03980eabf/clone?account=ab0896af-bf8c-48d4-885c-6573a94b1895 -X POST
+    HTTP/1.1 200 OK
+    Content-Type: application/json
+    Content-Length: 356
+    Date: Tue, 08 Jan 2013 20:21:17 GMT
+    Server: IMGAPI/1.0.0
+    x-request-id: f5645880-59d0-11e2-b638-4b6ffa4ca56f
+    x-response-time: 110
+    x-server-name: 70f0978d-7efa-4c45-8ebf-8cb9e3a887f7
+    Connection: keep-alive
+
+    {
+      "uuid": "e70502b0-705e-498e-a810-53a03980eabf",
+      "owner": "ab0896af-bf8c-48d4-885c-6573a94b1895",
+      "name": "foo",
+      "version": "1.0.0",
+      "state": "active",
+      "disabled": false,
+      "public": false,
+      "published_at": "2013-01-08T20:21:17.932Z",
+      "type": "zone-dataset",
+      "os": "smartos",
+      "files": [
+        {
+          "sha1": "cd0e0510c4a0799551687901077d7c4c06a4ebd8",
+          "size": 42,
+          "compression": "bzip2"
+        }
+      ],
+      "acl": []
+    }
+
+
 ## AdminImportImage (POST /images/:uuid?action=import)
 
 Import an image (preserving its `uuid` and `published_at` fields).
diff --git a/lib/errors.js b/lib/errors.js
index f5b45f9..50fd0b1 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -378,6 +378,22 @@ NotImageOwnerError.statusCode = 422;
 NotImageOwnerError.description = 'The caller is not the owner of this image.';
 
 
+function ImageNotShared(cause, account, imageUuid) {
+    RestError.call(this, {
+        restCode: this.constructor.restCode,
+        statusCode: this.constructor.statusCode,
+        message: format('Image %s is not shared with account %s',
+            account, imageUuid),
+        cause: cause
+    });
+}
+util.inherits(ImageNotShared, RestError);
+ImageNotShared.prototype.name = 'ImageNotShared';
+ImageNotShared.restCode = 'ImageNotShared';
+ImageNotShared.statusCode = 422;
+ImageNotShared.description = 'The image is not shared with this account.';
+
+
 function NotMantaPathOwnerError(cause, account, mpath) {
     if (mpath === undefined) {
         mpath = account;
@@ -608,6 +624,7 @@ module.exports = {
     StorageUnsupportedError: StorageUnsupportedError,
     RemoteSourceError: RemoteSourceError,
     OwnerDoesNotExistError: OwnerDoesNotExistError,
+    ImageNotShared: ImageNotShared,
     AccountDoesNotExistError: AccountDoesNotExistError,
     NotImageOwnerError: NotImageOwnerError,
     NotMantaPathOwnerError: NotMantaPathOwnerError,
diff --git a/lib/images.js b/lib/images.js
index 4ded88d..85a84eb 100644
--- a/lib/images.js
+++ b/lib/images.js
@@ -274,12 +274,7 @@ Image.prototype.addFile = function addFile(app, file, log, callback) {
  * Move the image file from this, into the given Image instance.
  *
  * @param app {App} The IMGAPI app.
- * @param file {Object} Describes the uploaded file, with keys:
- *      - `sha1` {String}
- *      - `size` {Integer}
- *      - `stor` {String}
- *      - `contentMD5` {String}
- *      - `mtime` {String} ISO date string
+ * @param toImage {Object} The description Image instance.
  * @param log {Bunyan Logger}
  * @param callback {Function} `function (err)` where `err` is some internal
  *      detail (i.e. it should be wrapped for the user).
@@ -316,6 +311,35 @@ function moveFileToImage(app, toImage, log, callback) {
 };
 
 
+/**
+ * Copy the image files from this, into the given Image instance.
+ *
+ * @param app {App} The IMGAPI app.
+ * @param toImage {Object} The destination Image instance.
+ * @param log {Bunyan Logger}
+ * @param callback {Function} `function (err)` where `err` is some internal
+ *      detail (i.e. it should be wrapped for the user).
+ */
+Image.prototype.copyFilesToImage =
+function copyFilesToImage(app, toImage, log, callback) {
+    var files = this.files;
+    assert.equal(files.length, 1, 'Expect exactly one image file');
+    var file = files[0];
+    var self = this;
+
+    var stor = app.getStor(file.stor);
+    stor.copyFileBetweenImages(self, toImage, 'file0',
+            function _copyFileCb(err) {
+        if (err) {
+            callback(err);
+            return;
+        }
+        // Add the file image metadata.
+        toImage.addFile(app, file, log, callback);
+    });
+};
+
+
 /**
  * Add an uploaded icon to this Image instance. The file will have already
  * be written out (to disk or to manta, depending).
@@ -543,6 +567,80 @@ Image.prototype.channelRemove = function channelRemove(
 };
 
 
+/**
+ * Delete this image.
+ *
+ * @param app {App} The IMGAPI app.
+ * @param log {Bunyan Logger}
+ * @param callback {Function} `function (err)` where `err` is okay to
+ *      use for an API reponse (i.e. doesn't expose internal details).
+ */
+Image.prototype.delete = function imageDetete(app, log, callback) {
+    var self = this;
+
+    vasync.pipeline({funcs: [
+        deleteModel,
+        deleteFiles,
+        deleteIconFile
+    ]}, callback);
+
+    // Delete the model.
+    // Note: We delete the manifest entry first to make sure the entry
+    // goes away, if subsequent deletion of files from storage fails,
+    // then that is just internally logged for operators to cleanup.
+    function deleteModel(_, next) {
+        app.db.del(self.uuid, function (delErr) {
+            if (delErr) {
+                next(delErr);
+                return;
+            }
+            app.cacheInvalidateDelete('Image', self);
+            next();
+        });
+    }
+
+    function deleteFiles(_, next) {
+        // Delete all files.
+        async.forEach(
+            self.files,
+            function deleteOneFile(file, nextFile) {
+                var stor = app.getStor(file.stor);
+                stor.deleteImageFile(self, nextFile);
+            },
+            function doneDeletes(fileErr) {
+                if (fileErr) {
+                    log.error({err: fileErr, image: self},
+                        'error deleting image file(s), this image may ' +
+                        'have zombie files which must be remove ' +
+                        'manually by an operator');
+                    next(errors.parseErrorFromStorage(fileErr,
+                        'error deleting image file'));
+                    return;
+                }
+                next();
+            }
+        );
+    }
+
+    function deleteIconFile(_, next) {
+        if (!self.icon) {
+            next();
+            return;
+        }
+        var stor = app.getStor(self.icon.stor);
+        stor.deleteImageFile(self, 'icon', function (fileErr) {
+            if (fileErr) {
+                log.error({err: fileErr, image: self},
+                    'error deleting model icon, this image may ' +
+                    'have a zombie icon file which must be ' +
+                    'removed manually by an operator');
+            }
+            next();
+        });
+    }
+};
+
+
 /**
  * Get an image from the database.
  *
@@ -3818,6 +3916,214 @@ function apiListImageJobs(req, res, next) {
 }
 
 
+function apiCloneImage(req, res, next) {
+    assert.object(req._app, 'req._app');
+    assert.object(req._image, 'req._image');
+    assert.ok(req._image.activated, 'req._image.activated');
+
+    var account;
+    var app = req._app;
+    var clonedImages = [];
+    var image = req._image;
+    var log = req.log;
+    var originImage;
+
+    if (req.query && req.query.account) {
+        account = req.query.account;
+    }
+
+    // Validation.
+    if (!UUID_RE.test(account)) {
+        next(new errors.InvalidParameterError(
+            format('invalid "account": not a UUID: "%s"', account),
+            [ { field: 'account', code: 'Invalid' } ]));
+        return;
+    }
+
+    if (!req._image.acl || req._image.acl.indexOf(account) === -1) {
+        next(new errors.ImageNotShared(null, account, image.uuid));
+        return;
+    }
+
+    // XXX: Trent, are there other image types we don't want to clone?
+    if (image.type === 'docker') {
+        next(new errors.ValidationFailedError(
+            'docker images cannot be cloned',
+            [ { field: 'type', code: 'Invalid' } ]));
+        return;
+    }
+
+    log.debug({image: image}, 'cloneImage: start');
+
+    vasync.pipeline({arg: {}, funcs: [
+        // Find and validate each of the origin images:
+        // 1. must be activated or disabled (disabled means cannot provision)
+        // 2. must be owned by the given user (or admin)
+        function lookupOriginImages(ctx, pipeNext) {
+            // The images to clone, starting from the base.
+            ctx.imagesToClone = [image];
+
+            if (!image.origin) {
+                pipeNext();
+                return;
+            }
+
+            var maxImageCount = 100;
+
+            function validateOneOriginImage(currentOrigin, subnext) {
+                Image.get(app, currentOrigin, log, function (err, img) {
+                    if (err) {
+                        subnext(err);
+                        return;
+                    }
+                    // Note that we allow disabled origin images, as that just
+                    // means that the origin image is not provisionable.
+                    if (img.state === 'unactivated') {
+                        subnext(new errors.OriginIsNotActiveError(
+                            currentOrigin));
+                        return;
+                    }
+                    // Stop cloning when an admin image is found.
+                    if ((img.owner === constants.UNSET_OWNER_UUID) ||
+                            (img.owner === app.config.adminUuid)) {
+                        originImage = img;
+                        subnext();
+                        return;
+                    }
+                    ctx.imagesToClone.unshift(img);
+                    // Follow the origin chain.
+                    if (img.origin) {
+                        if (ctx.imagesToClone.length >= maxImageCount) {
+                            subnext(new errors.InternalError(format(
+                                'Origin chain too long, exceeds %d images',
+                                maxImageCount)));
+                            return;
+                        }
+                        validateOneOriginImage(img.origin, subnext);
+                        return;
+                    }
+                    subnext();
+                });
+            }
+
+            validateOneOriginImage(image.origin, pipeNext);
+        },
+
+        // Clone each of the images, but leave them unactivated.
+        function cloneImages(ctx, pipeNext) {
+            assert.arrayOfObject(ctx.imagesToClone, 'ctx.imagesToClone');
+
+            log.info({uuid: image.uuid},
+                'cloneImage: %d images to clone',
+                ctx.imagesToClone.length);
+
+            function cloneOneImage(img, nextImageCb) {
+                vasync.pipeline({ arg: {}, funcs: [
+                    cloneImgMetadata,
+                    cloneImgFiles,
+                    activateClonedImg
+                ]}, nextImageCb);
+
+                function cloneImgMetadata(subctx, subnext) {
+                    var imgData = img.serialize(app.mode, '*');
+                    var lastIdx = ctx.imagesToClone.length - 1;
+                    var isFinalImg = (img === ctx.imagesToClone[lastIdx]);
+                    // Remove these image fields.
+                    delete imgData.acl;  // No access given to others.
+                    delete imgData.uuid; // So we get a new uuid.
+                    delete imgData.published_at; // Will be updated.
+                    // Change the owner and update the origin.
+                    imgData.owner = account;
+                    if (originImage) {
+                        imgData.origin = originImage.uuid;
+                    }
+                    // Disable intermediate images (so they are not shown as
+                    // provisionable), otherwise it would be strange when you
+                    // clone image X, but then you see image X, Y and Z (where
+                    // Y and Z are parent images of X).
+                    if (!isFinalImg) {
+                        imgData.disabled = true;
+                    }
+                    // Create the clone.
+                    Image.create(app, imgData, false, false,
+                            function (cErr, newimg) {
+                        if (cErr) {
+                            subnext(cErr);
+                            return;
+                        }
+                        subctx.newimg = newimg;
+                        originImage = newimg;
+                        clonedImages.push(newimg);
+                        subnext();
+                    });
+                }
+
+                function cloneImgFiles(subctx, subnext) {
+                    img.copyFilesToImage(app, subctx.newimg, log, subnext);
+                }
+
+                function activateClonedImg(subctx, subnext) {
+                    subctx.newimg.activate(app, log, function _imgActCb(err) {
+                        if (!err) {
+                            log.info({uuid: image.uuid,
+                                    clone_uuid: subctx.newimg.uuid},
+                                'image cloned');
+                        }
+                        subnext(err);
+                    });
+                }
+            }
+
+            function cleanupDeleteOneImage(img, deleteNextImg) {
+                img.delete(app, log, function _delOneImgCb(delErr) {
+                    if (delErr) {
+                        log.error({uuid: img.uuid},
+                            'clone failure: unable to delete cloned image: %s',
+                            delErr);
+                    }
+                    deleteNextImg();
+                });
+            }
+
+            vasync.forEachPipeline({
+                func: cloneOneImage,
+                inputs: ctx.imagesToClone
+            }, function _cloneImgPipeCb(err) {
+                if (err) {
+                    // Failure - cleanup newly created/cloned images.
+                    vasync.forEachPipeline({
+                        func: cleanupDeleteOneImage,
+                        inputs: clonedImages
+                    }, function _deleteImgPipeCb() {
+                        pipeNext(err);
+                        return;
+                    });
+                }
+                pipeNext();
+            });
+        },
+
+        // Once clone is successful, the original image does not need to be
+        // shared with this account anymore.
+        function unshareImage(ctx, pipeNext) {
+            image.removeAcl(app, [account], log, pipeNext);
+        }
+
+    ]}, function (err) {
+        if (err) {
+            log.error({account: account, image: image.uuid},
+                'Unable to clone image: %s', err);
+            next(err);
+            return;
+        }
+
+        var finalImage = clonedImages[clonedImages.length - 1];
+        res.send(finalImage.serialize(app.mode, req.getVersion()));
+        next();
+    });
+}
+
+
 /**
  * Ensure the 'uuid' request param is valid, else this is a 404.
  */
@@ -4164,6 +4470,14 @@ function mountApi(server, reqAuth, reqPassiveAuth) {
         channels.reqChannel,
         reqGetImage,    // add `req._image`, ensure access
         apiListImageJobs);
+    server.post(
+        {path: '/images/:uuid/clone', name: 'CloneImage'},
+        reqAuth,
+        reqValidUuid,
+        channels.reqChannel,
+        restify.bodyParser({mapParams: false}),
+        reqGetImage,
+        apiCloneImage);
 }
 
 
diff --git a/lib/storage.js b/lib/storage.js
index db83ddf..e1c0c73 100644
--- a/lib/storage.js
+++ b/lib/storage.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 /*
@@ -147,6 +147,18 @@ Storage.prototype.moveFileBetweenImages =
     function (fromImage, toImage, filename, callback) {
 };
 
+/**
+ * Copies an image file from one image to another image.
+ *
+ * @param fromImage {Image}
+ * @param toImage {Image}
+ * @param filename {String} Image filename (i.e. 'file0')
+ * @param callback {Function} `function (err)`
+ */
+Storage.prototype.copyFileBetweenImages =
+    function (fromImage, toImage, filename, callback) {
+};
+
 
 /**
  * Returns the archive path for image manifests. Image manifests are archived
@@ -479,6 +491,64 @@ function (fromImage, toImage, filename, callback) {
 };
 
 
+LocalStorage.prototype.copyFileBetweenImages =
+function (fromImage, toImage, filename, callback) {
+    assert.object(fromImage, 'fromImage');
+    assert.uuid(fromImage.uuid, 'fromImage.uuid');
+    assert.object(toImage, 'toImage');
+    assert.uuid(toImage.uuid, 'toImage.uuid');
+    assert.string(filename, 'filename');
+    assert.func(callback, 'callback');
+
+    var fromPath = this.storPathFromImageUuid(fromImage.uuid, filename);
+    var toPath = this.storPathFromImageUuid(toImage.uuid, filename);
+
+    var toDir = path.dirname(toPath);
+    mkdirp(toDir, function (mkdirErr) {
+        var cbCalled = false;
+        var readStream;
+        var writeStream;
+
+        if (mkdirErr) {
+            callback(mkdirErr);
+            return;
+        }
+
+        // Create read stream.
+        try {
+            readStream = fs.createReadStream(fromPath);
+        } catch (ex) {
+            callback(new errors.ValidationFailedError(ex,
+                'source file not found'));
+            return;
+        }
+
+        // Create write stream.
+        try {
+            writeStream = fs.createWriteStream(toPath);
+        } catch (ex) {
+            callback(new errors.ValidationFailedError(ex,
+                'destination file not writable'));
+            return;
+        }
+
+        function done(err) {
+            if (!cbCalled) {
+                callback(err);
+                cbCalled = true;
+            }
+        }
+
+        readStream.on('error', done);
+        writeStream.on('error', done);
+        writeStream.on('close', function _writeStreamCloseCb(ex) {
+            done();
+        });
+        readStream.pipe(writeStream);
+    });
+};
+
+
 LocalStorage.prototype.archiveImageManifest = function (manifest, callback) {
     assert.object(manifest, 'manifest');
     assert.string(manifest.uuid, 'manifest.uuid');
@@ -792,6 +862,29 @@ function (fromImage, toImage, filename, callback) {
     });
 };
 
+MantaStorage.prototype.copyFileBetweenImages =
+function (fromImage, toImage, filename, callback) {
+    assert.object(fromImage, 'fromImage');
+    assert.uuid(fromImage.uuid, 'fromImage.uuid');
+    assert.object(toImage, 'toImage');
+    assert.uuid(toImage.uuid, 'toImage.uuid');
+    assert.string(filename, 'filename');
+    assert.func(callback, 'callback');
+
+    var fromPath = this._storPathFromImageUuid(fromImage.uuid, filename);
+    var toPath = this._storPathFromImageUuid(toImage.uuid, filename);
+    var toDir = path.dirname(toPath);
+
+    var self = this;
+
+    self.client.mkdirp(toDir, function (mkdirErr) {
+        if (mkdirErr) {
+            return callback(mkdirErr);
+        }
+        self.client.ln(fromPath, toPath, callback);
+    });
+};
+
 MantaStorage.prototype.archiveImageManifest =
 function (manifest, callback) {
     assert.object(manifest, 'manifest');
diff --git a/package.json b/package.json
index ebd267e..139cadd 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgapi",
   "description": "Image API to manage images for SDC 7",
-  "version": "4.3.0",
+  "version": "4.4.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
