commit 33d23c37d2eb2ca2b63fbf9f43ee197c5457fd72 (refs/changes/20/4020/1)
Author: sjorge <sjorge@blackdot.be>
Date:   2018-05-29T21:49:49+02:00 (1 year, 4 months ago)
    
    OS-6968 Increase sieze of root by 2M

diff --git a/tools/build_live b/tools/build_live
index 36f87177..340e0504 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -911,7 +911,7 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_root_size=276000
+bi_root_size=278000
 if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
     grep "DEBUG enabled" >/dev/null; then
 	#
