commit 7a40aecde72d655a4fd1d81befa4f4c4a6ed4f47 (refs/changes/44/544/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-09-28T17:25:33-07:00 (3 years ago)
    
    MANTA-2994 Update node version for OpenSSL 1.0.2j/1.0.1u vulnerability fixes

diff --git a/Makefile b/Makefile
index bf5d036..6fce20d 100644
--- a/Makefile
+++ b/Makefile
@@ -55,7 +55,7 @@ CLEAN_FILES += node_modules
 #
 NAME 			= muskie
 NODE_PREBUILT_TAG       = zone
-NODE_PREBUILT_VERSION	:= v0.10.46
+NODE_PREBUILT_VERSION	:= v0.10.47
 NODE_PREBUILT_IMAGE     = fd2cc906-8938-11e3-beab-4359c665ac99
 
 include ./tools/mk/Makefile.defs
