commit 32e562443a08730fab73a8fe17b70b9253f11a46 (refs/changes/04/1004/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton instance get 516a51c5-f9cc-4145-93a9-71d0a518ae8d

diff --git a/lib/do_instance/do_get.js b/lib/do_instance/do_get.js
index 18afe2d..3663a22 100644
--- a/lib/do_instance/do_get.js
+++ b/lib/do_instance/do_get.js
@@ -19,15 +19,21 @@ function do_get(subcmd, opts, args, cb) {
         return cb(new Error('invalid args: ' + args));
     }
 
-    this.top.tritonapi.getInstance(args[0], function (err, inst) {
-        if (inst) {
-            if (opts.json) {
-                console.log(JSON.stringify(inst));
-            } else {
-                console.log(JSON.stringify(inst, null, 4));
-            }
+    var tritonapi = this.top.tritonapi;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-        cb(err);
+        tritonapi.getInstance(args[0], function (err, inst) {
+            if (inst) {
+                if (opts.json) {
+                    console.log(JSON.stringify(inst));
+                } else {
+                    console.log(JSON.stringify(inst, null, 4));
+                }
+            }
+            cb(err);
+        });
     });
 }
 
