commit 6d898986a072af80249ea82ed75129b2bbe98f8d (refs/changes/65/3865/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-04-24T21:01:45+00:00 (1 year, 5 months ago)
    
    OS-6827 Update vmadm to capture new log after OS-6718

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 3400f3ce..d3cd0cf5 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -6550,6 +6550,16 @@ function archiveVM(uuid, options, callback)
                 dst: archive_dirname,
                 targ: path.join(archive_dirname, 'cores')
             });
+            patterns_to_archive.push({
+                src: path.join(vmobj.zonepath, 'logs/console.log*'),
+                dst: path.join(archive_dirname, 'console'),
+                create_dst_dir: true
+            });
+            patterns_to_archive.push({
+                src: path.join(vmobj.zonepath, 'logs/platform.log*'),
+                dst: path.join(archive_dirname, 'platform'),
+                create_dst_dir: true
+            });
 
             if (vmobj.brand === 'kvm') {
                 patterns_to_archive.push({
@@ -6562,17 +6572,6 @@ function archiveVM(uuid, options, callback)
                     dst: archive_dirname,
                     targ: path.join(archive_dirname, 'startvm')
                 });
-            } else if (vmobj.brand === 'bhyve') {
-                patterns_to_archive.push({
-                    src: path.join(vmobj.zonepath, 'root/tmp/zhyve.log'),
-                    dst: path.join(archive_dirname, 'vmlogs'),
-                    create_dst_dir: true
-                });
-                patterns_to_archive.push({
-                    src: path.join(vmobj.zonepath, 'root/var/run/*.log*'),
-                    dst: path.join(archive_dirname, 'vmlogs'),
-                    create_dst_dir: true
-                });
             } else if (vmobj.docker) {
 
                 function logTimestamp() {
