commit eb7fa9915f1daf39892fa2cf859240cd9d233462 (refs/changes/98/4698/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-08-16T16:40:20+02:00 (1 year, 2 months ago)
    
    TRITON-695 sdcadm tests broken with "Missing assets_* in SAPI"

diff --git a/lib/cli/do_check_config.js b/lib/cli/do_check_config.js
index fcf19b7..182c6c5 100644
--- a/lib/cli/do_check_config.js
+++ b/lib/cli/do_check_config.js
@@ -300,7 +300,7 @@ CheckConfig.prototype.execute = function cExecute(opts, args, cb) {
             // value for *_pw.
             services.forEach(function checkSvcCfgVars(s) {
                 if (s === 'cns' || s === 'portolan' || s === 'docker' ||
-                    s === 'cmon' || s === 'volapi') {
+                    s === 'cmon' || s === 'volapi' || s === 'assets') {
                     return;
                 }
                 if (!sdc[s + '_root_pw'] && s !== 'manta' && s !== 'sapi') {
