From 5fb389a4a5de6ceea6e201249248a9d050e07d50 Mon Sep 17 00:00:00 2001
From: Peter Marton <pmarton@netflix.com>
Date: Mon, 31 Dec 2018 10:24:57 -0800
Subject: [PATCH] joyent/node-verror#63 do not call sprintf with one argument

---
 lib/verror.js | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/lib/verror.js b/lib/verror.js
index 8663dde..9e1dfd6 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -114,8 +114,15 @@ function parseConstructorArguments(args)
 		});
 	}
 
+	/*
+	 * Do not call sprintf with one argument, performance optimization.
+	 * It also addresses the issue when someone passes user input and it
+	 * contains format string charahcters like %s without additional args.
+	 */
 	if (sprintf_args.length === 0) {
 		shortmessage = '';
+	} else if (sprintf_args.length === 1) {
+		shortmessage = sprintf_args[0];
 	} else {
 		shortmessage = sprintf.apply(null, sprintf_args);
 	}
-- 
2.21.0

