From a1c56780dfd1d30b44d2638711aee85fb2828cdf Mon Sep 17 00:00:00 2001
From: Trent Mick <trent.mick@joyent.com>
Date: Wed, 4 Jan 2017 15:20:47 -0800
Subject: [PATCH] MANTA-3071 mls reports no results, no error

---
 CHANGES.md |  3 +++
 bin/mls    | 21 ++++++++++++++++-----
 2 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 58b56f4..9b435c2 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,9 @@
 
 ## not yet released
 
+- MANTA-3071 Fix case where a transient error in one request to the Manta
+  server can make `mls` exit 0 without reporting results.
+
 ## 4.1.1
 
 - joyent/node-manta#293 '~~/' works, '~~' does now too.
diff --git a/bin/mls b/bin/mls
index b84568d..7e78ed8 100755
--- a/bin/mls
+++ b/bin/mls
@@ -204,13 +204,24 @@ function printEntry(opts, obj) {
     }
 
     options.paths.forEach(function (p) {
-        client.ls(p, options, function (err, res) {
-            if (err) {
-                if (err.name === 'InvalidDirectoryError') {
-                    printEntry(options, err.info);
+        client.ls(p, options, function (lsErr, res) {
+            if (lsErr) {
+                if (lsErr.name === 'InvalidDirectoryError') {
+                    printEntry(options, lsErr.info);
                     cb();
                 } else {
-                    client.get(p, cb);
+                    /*
+                     * The first request done by `client.ls` is a HEAD, for
+                     * which there is no response body with good error details.
+                     * Fallback to `client.get` to get hopefully helpful
+                     * error details.
+                     *
+                     * If `client.get` succeeds (e.g. if lsErr was transient),
+                     * then use the lsErr that we have.
+                     */
+                    client.get(p, function (getErr) {
+                        cb(getErr || lsErr);
+                    });
                 }
                 return;
             }
-- 
2.21.0

