commit 4e221969582ecd73944f8e4d96c497f85b64ae4e (refs/changes/36/5036/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-11-07T11:15:13-08:00 (11 months ago)
    
    TRITON-940 Add metricPorts metadata to binder for cmon-agent

diff --git a/boot/setup.sh b/boot/setup.sh
index 85ab7de..75cbbdc 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -156,6 +156,9 @@ else # FLAVOR == "sdc"
     sdc_log_rotation_add zookeeper /var/log/zookeeper/zookeeper.log 1g
     sdc_log_rotation_setup_end
 
+    # Add metadata for cmon-agent discovery
+    mdata-put metricPorts 1053
+
     # All done, run boilerplate end-of-setup
     sdc_setup_complete
 
diff --git a/package.json b/package.json
index 30dbcdf..8f844b0 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "binder",
   "description": "SmartDataService DNS Service",
-  "version": "1.2.0",
+  "version": "1.2.1",
   "author": "Joyent (joyent.com)",
   "license": "MPL-2.0",
   "private": true,
