commit 864a30f2703da75d7a34546240a6137c0435971d (refs/changes/14/3714/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-03-23T19:45:36+00:00 (1 year, 7 months ago)
    
    OS-6814 OS-6726 missed exception_lists/check_rtime

diff --git a/exception_lists/check_rtime b/exception_lists/check_rtime
index d98cdabbc6..9ab3b5da18 100644
--- a/exception_lists/check_rtime
+++ b/exception_lists/check_rtime
@@ -76,7 +76,7 @@ EXEC_STACK	^opt/os-tests/tests/secflags/stacky$
 
 # Objects for which we allow relocations to the text segment
 TEXTREL		^platform/.*/MACH(kernel)/unix$
-TEXTREL		^usr/sbin/amd64/bhyve$
+TEXTREL		^usr/sbin/bhyve$
 
 # Directories and files that are allowed to have no direct bound symbols
 NODIRECT	^platform/.*/MACH(kernel)/unix$
