commit c39f1cdffa52ffcf70dff9eea2b607ac47c522f6 (refs/changes/20/1120/2)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2016-12-14T20:24:46+00:00 (2 years, 10 months ago)
    
    VOLAPI-30 Make VOLAPI use node v4.x

diff --git a/targets.json.in b/targets.json.in
index ed0329c..cc58457 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -453,7 +453,8 @@ cat <<EOF
     "image_name": "volapi",
     "image_description": "SDC Volumes API",
     "image_version": "1.0.0",
-    "image_uuid": "de411e86-548d-11e4-a4b7-3bb60478632a",
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
     "repos": [
       {"url": "git@github.com:joyent/sdc-volapi.git"}
     ],
@@ -464,7 +465,13 @@ cat <<EOF
       "registrar",
       "https://download.joyent.com/pub/build/sdcnode"
     ],
-    "pkgsrc": [],
+    "pkgsrc": [
+        "coreutils-8.23nb2",
+        "curl-7.47.1",
+        "gsed-4.2.2nb4",
+        "patch-2.7.5",
+        "sudo-1.8.15"
+    ],
     "tarballs": [
       {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
