commit 0bad8333294beafce97678b0fdd1b7d553a6b73a
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T14:51:17-05:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/Makefile b/Makefile
index d00c68f..4e257b9 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ ESLINT_FILES := $(JS_FILES)
 
 # Should be the same version as the platform's /usr/node/bin/node.
 NODE_PREBUILT_TAG =	gz
-NODE_PREBUILT_VERSION =	v6.15.1
+NODE_PREBUILT_VERSION =	v6.17.0
 ifeq ($(shell uname -s),SunOS)
 	# sdc-smartos/1.6.3.
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
