commit 7bda8d7951ec631c92bd105ca55a54c97d229115 (refs/changes/62/3262/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-01-25T14:22:15-05:00 (1 year, 8 months ago)
    
    RELENG-612 prefer (git+)https urls

diff --git a/.gitmodules b/.gitmodules
index 7f63af3..b3d9976 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,9 +1,9 @@
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
-	url = git://github.com/davepacheco/javascriptlint.git
+	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/davepacheco/jsstyle.git
 [submodule "deps/sdc-scripts"]
 	path = deps/sdc-scripts
-	url = git@github.com:joyent/sdc-scripts.git
+	url = https://github.com/joyent/sdc-scripts.git
