From 39e68e7347acadc4e9647b614f900acc653abd73 Mon Sep 17 00:00:00 2001
From: Sean Chittenden <sean@chittenden.org>
Date: Fri, 23 Jun 2017 09:21:53 -0700
Subject: [PATCH] Improve readability: wrap lines along complete thoughts.

---
 lib/postgresMgr.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/postgresMgr.js b/lib/postgresMgr.js
index dc7ecc4..620fc9d 100644
--- a/lib/postgresMgr.js
+++ b/lib/postgresMgr.js
@@ -1497,8 +1497,8 @@ PostgresMgr.prototype._initDb = function (callback) {
 
         },
         function _initDb(_, cb) {
-            var cmd = 'sudo -u ' + self._dbUser + ' ' + self._pgInitDbPath +
-                ' -D ' + self._dataDir;
+            var cmd = 'sudo -u ' + self._dbUser + ' ' +
+                self._pgInitDbPath + ' -D ' + self._dataDir;
             log.info({cmd: cmd}, 'PostgresMgr.initDb: initializing db');
             exec(cmd, function (err, stdout, stderr) {
                 // ignore errors since the db could already be initialized
-- 
2.21.0

