commit 26e197eee1ac8c178826a3375667936b6268d0e2
Author: John Levon <john.levon@joyent.com>
Date:   2019-04-25T16:13:23+00:00 (6 months ago)
    
    TRITON-1353 Triton systems should ship with HT disabled

diff --git a/lib/cnapi/server.js b/lib/cnapi/server.js
index 21770f2..d9cd53d 100644
--- a/lib/cnapi/server.js
+++ b/lib/cnapi/server.js
@@ -203,13 +203,17 @@ var s_sysinfo = {
 		'Bhyve Max Vcpus': base.intg.nonneg.opt,
 		'CPU Type': base.str.nonempty.req,
 		'CPU Virtualization': base.str.req,
-		'CPU Physical Cores': base.intg.pos.req,
+		'CPU Physical Cores': base.intg.pos.opt,
 		'Admin NIC Tag': base.str.opt, // OS-7075 (20180719)
 		'Admin IP': base.str.v4addr.opt, // TRITON-841 (20181206)
 		'UUID': base.str.uuid.req,
 		'Hostname': base.str.hostname.opt,
 		'HVM API': base.bool.opt,
-		'CPU Total Cores': base.intg.pos.req,
+		'CPU Total Cores': base.intg.pos.opt,
+		'CPU Socket Count': base.intg.pos.opt,
+		'CPU Core Count': base.intg.pos.opt,
+		'CPU Online Count': base.intg.pos.opt,
+		'CPU Count': base.intg.pos.opt,
 		'MiB of Memory': base.intg.pos.req,
 		'Zpool': base.str.nonempty.req,
 		'Zpool Disks': s_si_disklist_r,
diff --git a/lib/dapi/server.js b/lib/dapi/server.js
index c9e24b2..c74ac04 100644
--- a/lib/dapi/server.js
+++ b/lib/dapi/server.js
@@ -50,6 +50,10 @@ var s_sysinfo = {
 		'Bhyve Capable': base.bool.opt,
 		'Bhyve Max Vcpus': base.intg.nonneg.opt,
 		'CPU Total Cores': base.intg.pos.req,
+		'CPU Socket Count': base.intg.pos.opt,
+		'CPU Core Count': base.intg.pos.opt,
+		'CPU Online Count': base.intg.pos.opt,
+		'CPU Count': base.intg.pos.opt,
 		'Admin NIC Tag': base.str.opt, // OS-7075 (20180719)
 		'Admin IP': base.str.v4addr.opt, // TRITON-841 (20181206)
 		'HVM API': base.bool.opt,
