commit 31dd4a5e5afd46082b242c123b0e8b9b821a4024
Author: John Levon <john.levon@joyent.com>
Date:   2019-05-01T12:06:38+00:00 (5 months ago)
    
    TRITON-1353 Triton systems should ship with HT disabled

diff --git a/lib/cnapi/server.js b/lib/cnapi/server.js
index 21770f2..d9cd53d 100644
--- a/lib/cnapi/server.js
+++ b/lib/cnapi/server.js
@@ -203,13 +203,17 @@ var s_sysinfo = {
 		'Bhyve Max Vcpus': base.intg.nonneg.opt,
 		'CPU Type': base.str.nonempty.req,
 		'CPU Virtualization': base.str.req,
-		'CPU Physical Cores': base.intg.pos.req,
+		'CPU Physical Cores': base.intg.pos.opt,
 		'Admin NIC Tag': base.str.opt, // OS-7075 (20180719)
 		'Admin IP': base.str.v4addr.opt, // TRITON-841 (20181206)
 		'UUID': base.str.uuid.req,
 		'Hostname': base.str.hostname.opt,
 		'HVM API': base.bool.opt,
-		'CPU Total Cores': base.intg.pos.req,
+		'CPU Total Cores': base.intg.pos.opt,
+		'CPU Socket Count': base.intg.pos.opt,
+		'CPU Core Count': base.intg.pos.opt,
+		'CPU Online Count': base.intg.pos.opt,
+		'CPU Count': base.intg.pos.opt,
 		'MiB of Memory': base.intg.pos.req,
 		'Zpool': base.str.nonempty.req,
 		'Zpool Disks': s_si_disklist_r,
diff --git a/lib/dapi/server.js b/lib/dapi/server.js
index c9e24b2..63ca6b4 100644
--- a/lib/dapi/server.js
+++ b/lib/dapi/server.js
@@ -49,7 +49,11 @@ var s_sysinfo = {
 	'properties': {
 		'Bhyve Capable': base.bool.opt,
 		'Bhyve Max Vcpus': base.intg.nonneg.opt,
-		'CPU Total Cores': base.intg.pos.req,
+		'CPU Total Cores': base.intg.pos.opt,
+		'CPU Socket Count': base.intg.pos.opt,
+		'CPU Core Count': base.intg.pos.opt,
+		'CPU Online Count': base.intg.pos.opt,
+		'CPU Count': base.intg.pos.opt,
 		'Admin NIC Tag': base.str.opt, // OS-7075 (20180719)
 		'Admin IP': base.str.v4addr.opt, // TRITON-841 (20181206)
 		'HVM API': base.bool.opt,
