commit 0c18cc6ac965a00d91bb7f69649ad6f3b0bb7d29 (refs/changes/11/2211/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-07-12T13:04:14-07:00 (2 years, 3 months ago)
    
    TOOLS-1807 node updates for july 2017 node.js security releases

diff --git a/node.js/Makefile b/node.js/Makefile
index 9511c36..e8edf57 100644
--- a/node.js/Makefile
+++ b/node.js/Makefile
@@ -52,6 +52,7 @@ endif
 
 AUTOCONF_OPTS += \
 	--with-dtrace \
+	--without-snapshot \
 	--openssl-use-sys \
 	--openssl-libpath=$(DESTDIR)/lib \
 	--openssl-includes=$(DESTDIR)/usr/include \
