commit dfdbd2cde09f2c229e4941f5d58b3823cd99e5f5 (refs/changes/94/3794/3)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-05-31T15:35:12-04:00 (1 year, 4 months ago)
    
    TRITON-315 unit tests for removeOldCNToolsTarballs (WIP/POC)

diff --git a/.gitignore b/.gitignore
index 0bbb6d6..73f5c48 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,5 @@
+/coverage
+/.nyc_output
 /node_modules
 /tmp
 /docs/*.json
diff --git a/Makefile b/Makefile
index aa396a4..5580d6c 100644
--- a/Makefile
+++ b/Makefile
@@ -26,6 +26,8 @@ ESLINT_FILES	= $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 CLEAN_FILES += ./node_modules ./build/sdcadm-*.sh ./build/sdcadm-*.imgmanifest ./build/shar-image ./man/man1/sdcadm.1 ./etc/sdcadm.completion
+TAPE := ./node_modules/.bin/tape
+NYC := ./node_modules/.bin/nyc
 
 
 NODE_PREBUILT_VERSION=v6.14.0
@@ -58,6 +60,10 @@ all: | $(NPM_EXEC)
 	$(NODE) ./node_modules/.bin/kthxbai
 	rm -rf ./node_modules/.bin/kthxbai ./node_modules/kthxbai
 
+$(TAPE):
+
+$(NYC):
+
 .PHONY: shar
 shar:
 	./tools/mk-shar -o $(TOP)/build -s $(STAMP)
@@ -66,6 +72,14 @@ shar:
 test:
 	./test/runtests
 
+.PHONY: unit
+unit: | $(TAPE)
+	$(TAPE) 'test/unit/**/*.test.js'
+
+.PHONY: coverage
+coverage: | $(NYC) $(TAPE)
+	$(NYC) $(TAPE) 'test/unit/**/*.test.js'
+
 .PHONY: release
 release: all man completion shar
 
diff --git a/lib/steps/index.js b/lib/steps/index.js
index 5b6e24c..bb81417 100644
--- a/lib/steps/index.js
+++ b/lib/steps/index.js
@@ -38,5 +38,6 @@ module.exports = {};
         module.exports[symbol] = mod[symbol];
     });
 });
+module.exports.usbkey = require('./usbkey');
 
 // vim: set softtabstop=4 shiftwidth=4:
diff --git a/lib/steps/usbkey.js b/lib/steps/usbkey.js
new file mode 100644
index 0000000..0e9d0e6
--- /dev/null
+++ b/lib/steps/usbkey.js
@@ -0,0 +1,45 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2018 Joyent, Inc.
+ */
+/* jsl:ignore */
+'use strict';
+/* jsl:end */
+
+const fs = require('fs');
+const path = require('path');
+
+var assert = require('assert-plus');
+
+// keep a finite number cn_tools backups on the usb key
+function removeOldCNToolsTarballs(arg, next) {
+    assert.optionalFunc(arg.progress, 'arg.progress');
+
+    const progress = arg.progress || console.log;
+    const backupPath = '/usbkey/extra/joysetup/';
+    var tarballs = fs.readdirSync(backupPath).filter(
+        function isCNTools(p) {
+            return (p.startsWith('cn_tools.') &&
+                    p.endsWith('tar.gz') &&
+                    p !== 'cn_tools.tar.gz');
+        });
+    tarballs.sort();
+    tarballs.reverse();
+    const toDelete = tarballs.slice(4);
+    if (toDelete.length) {
+        progress('Removing old cn backups: ' + toDelete.join(', '));
+        toDelete.forEach(function rmBall(fname) {
+            fs.unlinkSync(path.join(backupPath, fname));
+        });
+    }
+    next();
+}
+
+module.exports = {
+    removeOldCNToolsTarballs: removeOldCNToolsTarballs
+};
diff --git a/package.json b/package.json
index 55cd69c..f1297d6 100644
--- a/package.json
+++ b/package.json
@@ -37,6 +37,8 @@
     "marked-man": "0.1.4",
     "eslint": "4.13.1",
     "eslint-plugin-joyent": "~2.0.0"
+    "mock-fs": "4.4.x",
+    "tap": "^12.0.1"
   },
   "engines": {
     "node": ">=0.10"
diff --git a/test/unit/steps/usbkey.test.js b/test/unit/steps/usbkey.test.js
new file mode 100644
index 0000000..7ec299b
--- /dev/null
+++ b/test/unit/steps/usbkey.test.js
@@ -0,0 +1,108 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2018 Joyent, Inc.
+ */
+'use strict';
+
+const fs = require('fs');
+
+const mockfs = require('mock-fs');
+const test = require('tape');
+
+const usbkey = require('../../../lib/steps/usbkey'); // TODO: how to not be gross?
+
+const joysetupDir = '/usbkey/extra/joysetup/';
+
+test('removeOldCNToolsTarballs', function(outer) {
+    test('empty', function (t) {
+        mockfs({[joysetupDir]: {}});
+        usbkey.removeOldCNToolsTarballs({}, function() {
+            const stats = fs.statSync(joysetupDir);
+            t.ok(stats.isDirectory(joysetupDir));
+            const files = fs.readdirSync(joysetupDir)
+            t.equal(files.length, 0);
+            mockfs.restore();
+            t.end();
+        });
+    });
+
+    test('other-files-untouched', function (t) {
+        mockfs({[joysetupDir]:
+                {'cn_tools.tar.gz': '',
+                 'joysetup.sh': '',
+                 'agentsetup.sh': '',
+                 'cn_tools.20180208T185323.tar.gz': ''}});
+        usbkey.removeOldCNToolsTarballs({}, function() {
+            const files = fs.readdirSync(joysetupDir)
+            t.ok(files.length == 4);
+            mockfs.restore();
+            t.end();
+        });
+    });
+
+    test('prune', function (t) {
+        const isCNTools = function isCNTools(p) {
+            return (p.startsWith('cn_tools.') &&
+                    p.endsWith('tar.gz') &&
+                    p !== 'cn_tools.tar.gz');
+        };
+        const oldFormatTools = ['cn_tools.2017-12-14T15:57:59.179Z.tar.gz',
+                           'cn_tools.2017-08-31T14:36:03.728Z.tar.gz'];
+        const oldTools = ['cn_tools.20180201T000000.tar.gz',
+                          'cn_tools.20180201T204654.tar.gz'];
+        const cnTools = ['cn_tools.20180207T220522.tar.gz',
+                         'cn_tools.20180208T181727.tar.gz',
+                         'cn_tools.20180208T181853.tar.gz',
+                         'cn_tools.20180208T185323.tar.gz'];
+        const stdFiles = 
+              {'cn_tools.tar.gz': '',
+               'joysetup.sh': '',
+               'agentsetup.sh': ''};
+        const checkRemoved = function checkRemoved(tt) {
+                const files = fs.readdirSync(joysetupDir)
+                tt.equal(files.length, 7);
+                const remainingTools = files.filter(isCNTools);
+                const remainingOther = files.filter(function(p) {
+                    return (!isCNTools(p));
+                });
+                tt.deepEqual(remainingTools.slice().sort(),
+                             cnTools.slice().sort());
+                tt.deepEqual(remainingOther.slice().sort(),
+                             Object.keys(stdFiles).sort());
+                mockfs.restore();
+                tt.end()
+
+        };
+        
+        test('prune-old-tarballs', function (tt) {
+            const fakeFiles = Object.assign({}, stdFiles);
+            for (let fname of [].concat(cnTools, oldTools)) {
+                fakeFiles[fname] = '';
+            }
+            mockfs({[joysetupDir]: fakeFiles});
+            usbkey.removeOldCNToolsTarballs({}, function() {
+                checkRemoved(tt);
+            });
+        });
+
+        test('prune-old-format', function (tt) {
+            const fakeFiles = Object.assign({}, stdFiles);
+            for (let fname of [].concat(cnTools, oldTools, oldFormatTools)) {
+                fakeFiles[fname] = '';
+            }
+            mockfs({[joysetupDir]: fakeFiles});
+            usbkey.removeOldCNToolsTarballs({}, function() {
+                checkRemoved(tt);
+            });
+        });
+
+        t.end();
+    });
+
+    outer.end();
+});
