commit cc8c51056696fd10c5ced746abc8aece008374ff
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-07-04T16:31:08+02:00 (3 months ago)
    
    TRITON-1796 re-enable sdcadm manatee tests

diff --git a/test/runtests b/test/runtests
index e354ae3..8104423 100755
--- a/test/runtests
+++ b/test/runtests
@@ -135,7 +135,7 @@ set +o errexit
 for file in $test_files; do
     test_file=$(basename $file)
     echo "# $test_file"
-    NO_MANATEE_HA=1 PATH=$NODE_INSTALL/bin:$PATH $TAPE $file \
+    PATH=$NODE_INSTALL/bin:$PATH $TAPE $file \
         | tee $RESULTS/$test_file.tap
     retval=$?
     if [[ "${retval}" != "0" ]]; then
