From cc8c51056696fd10c5ced746abc8aece008374ff Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 4 Jul 2019 16:31:08 +0200
Subject: [PATCH] TRITON-1796 re-enable sdcadm manatee tests

---
 test/runtests | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/runtests b/test/runtests
index e354ae3..8104423 100755
--- a/test/runtests
+++ b/test/runtests
@@ -135,7 +135,7 @@ set +o errexit
 for file in $test_files; do
     test_file=$(basename $file)
     echo "# $test_file"
-    NO_MANATEE_HA=1 PATH=$NODE_INSTALL/bin:$PATH $TAPE $file \
+    PATH=$NODE_INSTALL/bin:$PATH $TAPE $file \
         | tee $RESULTS/$test_file.tap
     retval=$?
     if [[ "${retval}" != "0" ]]; then
-- 
2.21.0

