From 046129cc8e74c41ef366180689075962ea24d652 Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Sat, 6 Aug 2016 08:29:33 +0200
Subject: [PATCH] TOOLS-1503 sdcadm fails to print an error with "%j"

---
 lib/cli/index.js | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/lib/cli/index.js b/lib/cli/index.js
index b72ffa5..f621180 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -22,6 +22,7 @@ var Cmdln = cmdln.Cmdln;
 var strsplit = require('strsplit');
 var vasync = require('vasync');
 var uuid = require('node-uuid');
+var extsprintf = require('extsprintf');
 
 var common = require('../common');
 var errors = require('../errors');
@@ -133,8 +134,9 @@ CLI.prototype.init = function init(opts, args, callback) {
      */
     this.progress = function progress() {
         var args_ = Array.prototype.slice.call(arguments);
-        self.log.debug.apply(self.log, [ {progress: true} ].concat(args_));
-        console.log.apply(null, args_);
+        var msg = extsprintf.sprintf.apply(null, args_);
+        self.log.debug({progress: true}, msg);
+        console.log(msg);
     };
 
     // Cmdln class handles `opts.help`.
-- 
2.21.0

