From 2d628f5ca6088ad5b775ffa25470b510143653ce Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Sat, 6 Aug 2016 08:29:33 +0200
Subject: [PATCH] TOOLS-1503 sdcadm fails to print an error with "%j"

---
 lib/cli/index.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/cli/index.js b/lib/cli/index.js
index b72ffa5..e4d8b76 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -22,6 +22,7 @@ var Cmdln = cmdln.Cmdln;
 var strsplit = require('strsplit');
 var vasync = require('vasync');
 var uuid = require('node-uuid');
+var extsprintf = require('extsprintf');
 
 var common = require('../common');
 var errors = require('../errors');
@@ -134,7 +135,7 @@ CLI.prototype.init = function init(opts, args, callback) {
     this.progress = function progress() {
         var args_ = Array.prototype.slice.call(arguments);
         self.log.debug.apply(self.log, [ {progress: true} ].concat(args_));
-        console.log.apply(null, args_);
+        extsprintf.printf.apply(null, args_);
     };
 
     // Cmdln class handles `opts.help`.
-- 
2.21.0

