commit 2d628f5ca6088ad5b775ffa25470b510143653ce (refs/changes/48/248/1)
Author: Pedro P. Candel <pedro@joyent.com>
Date:   2016-08-06T08:30:01+02:00 (3 years, 2 months ago)
    
    TOOLS-1503 sdcadm fails to print an error with "%j"

diff --git a/lib/cli/index.js b/lib/cli/index.js
index b72ffa5..e4d8b76 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -22,6 +22,7 @@ var Cmdln = cmdln.Cmdln;
 var strsplit = require('strsplit');
 var vasync = require('vasync');
 var uuid = require('node-uuid');
+var extsprintf = require('extsprintf');
 
 var common = require('../common');
 var errors = require('../errors');
@@ -134,7 +135,7 @@ CLI.prototype.init = function init(opts, args, callback) {
     this.progress = function progress() {
         var args_ = Array.prototype.slice.call(arguments);
         self.log.debug.apply(self.log, [ {progress: true} ].concat(args_));
-        console.log.apply(null, args_);
+        extsprintf.printf.apply(null, args_);
     };
 
     // Cmdln class handles `opts.help`.
