commit 046129cc8e74c41ef366180689075962ea24d652 (refs/changes/48/248/3)
Author: Pedro P. Candel <pedro@joyent.com>
Date:   2016-08-16T13:25:04+02:00 (3 years, 2 months ago)
    
    TOOLS-1503 sdcadm fails to print an error with "%j"

diff --git a/lib/cli/index.js b/lib/cli/index.js
index b72ffa5..f621180 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -22,6 +22,7 @@ var Cmdln = cmdln.Cmdln;
 var strsplit = require('strsplit');
 var vasync = require('vasync');
 var uuid = require('node-uuid');
+var extsprintf = require('extsprintf');
 
 var common = require('../common');
 var errors = require('../errors');
@@ -133,8 +134,9 @@ CLI.prototype.init = function init(opts, args, callback) {
      */
     this.progress = function progress() {
         var args_ = Array.prototype.slice.call(arguments);
-        self.log.debug.apply(self.log, [ {progress: true} ].concat(args_));
-        console.log.apply(null, args_);
+        var msg = extsprintf.sprintf.apply(null, args_);
+        self.log.debug({progress: true}, msg);
+        console.log(msg);
     };
 
     // Cmdln class handles `opts.help`.
