commit 52fdb69ac92e8fc92a7ec207e133867a3f16bd68 (refs/changes/48/2448/8)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-29T22:24:22+00:00 (2 years, 1 month ago)
    
    joyent/node-manta#323 return response argument from client.commitUpload

diff --git a/CHANGES.md b/CHANGES.md
index ebd0725..a87eceb 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,13 @@
 
 ## not yet released
 
+## 4.5.0
+
+Minor bump due to augmentation of the client API.
+
+- joyent/node-manta#323 return response argument from client.commitUpload
+- joyent/node-manta#318 node-manta nodejs version support
+- joyent/node-manta#322 test7 make target should be test8 given node v8
 - joyent/node-manta#321 document mlogin's use of poseidon assets
 
 ## 4.4.3
diff --git a/lib/client.js b/lib/client.js
index 4a557c4..6b487d8 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -3599,7 +3599,7 @@ MantaClient.prototype.commitUpload = function commitUpload(id, p, opts, cb) {
                 log.debug(err2);
                 cb(err2);
             } else {
-                cb(null);
+                cb(null, res);
             }
         });
     });
diff --git a/package.json b/package.json
index 6b7df03..5125357 100644
--- a/package.json
+++ b/package.json
@@ -7,7 +7,7 @@
         "type": "git",
         "url": "git://github.com/joyent/node-manta.git"
     },
-    "version": "4.4.3",
+    "version": "4.5.0",
     "main": "./lib/index.js",
     "dependencies": {
         "assert-plus": "^1.0.0",
diff --git a/test/client.test.js b/test/client.test.js
index 1eb783e..ce5bb41 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -802,12 +802,14 @@ test('commit upload', function (t) {
     };
 
     var self = this;
-    self.client.commitUpload(UPLOAD1, ETAGS1, opts, function (err) {
+    self.client.commitUpload(UPLOAD1, ETAGS1, opts, function (err, res) {
         t.ifError(err);
         if (err) {
             t.done();
             return;
         }
+        t.ok(res);
+        t.equal(res.statusCode, 201);
         self.client.getUpload(UPLOAD1, opts, function (err2, upload) {
             t.ifError(err2);
             if (err2) {
@@ -849,6 +851,18 @@ test('commit upload', function (t) {
     });
 });
 
+test('errant commit upload returns undefined res', function (t) {
+    var opts = {
+        account: this.client.user
+    };
+    var self = this;
+    self.client.commitUpload(libuuid.v4(), ETAGS1, opts, function (err, res) {
+        t.ok(err);
+        t.ok(res === undefined);
+        t.done();
+    });
+});
+
 test('abort upload', function (t) {
     var opts = {
         account: this.client.user
