commit c17d702adfca07af0716e342289c4502566f9f93 (refs/changes/48/2448/3)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-24T22:58:25+00:00 (2 years, 1 month ago)
    
    joyent/node-manta#323 return response argument from client.commitUpload

diff --git a/lib/client.js b/lib/client.js
index 4a557c4..6b487d8 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -3599,7 +3599,7 @@ MantaClient.prototype.commitUpload = function commitUpload(id, p, opts, cb) {
                 log.debug(err2);
                 cb(err2);
             } else {
-                cb(null);
+                cb(null, res);
             }
         });
     });
diff --git a/test/client.test.js b/test/client.test.js
index 1eb783e..2079f2d 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -802,12 +802,14 @@ test('commit upload', function (t) {
     };
 
     var self = this;
-    self.client.commitUpload(UPLOAD1, ETAGS1, opts, function (err) {
+    self.client.commitUpload(UPLOAD1, ETAGS1, opts, function (err, res) {
         t.ifError(err);
         if (err) {
             t.done();
             return;
         }
+        t.ok(res);
+        t.equal(res.statusCode, 201);
         self.client.getUpload(UPLOAD1, opts, function (err2, upload) {
             t.ifError(err2);
             if (err2) {
@@ -849,6 +851,18 @@ test('commit upload', function (t) {
     });
 });
 
+test('errant commit returns undefined res', function (t) {
+    var opts = {
+        account: this.client.user
+    };
+    var self = this;
+    self.client.commitUpload(libuuid.v4(), ETAGS1, opts, function (err, res) {
+        t.ok(err);
+        t.ok(res === undefined);
+        t.done();
+    });
+});
+
 test('abort upload', function (t) {
     var opts = {
         account: this.client.user
