From e5c9c1ff4f764e45b9c847314b394502354a7137 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Wed, 23 Aug 2017 21:51:19 +0000
Subject: [PATCH] joyent/node-manta#323 return response argument from
 client.commitUpload

---
 lib/client.js | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/client.js b/lib/client.js
index 4a557c4..123dc8c 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -3589,7 +3589,7 @@ MantaClient.prototype.commitUpload = function commitUpload(id, p, opts, cb) {
         if (err) {
             err = new Verror(err, 'signing request');
             log.debug(err);
-            cb(err);
+            cb(err, null);
             return;
         }
 
@@ -3597,9 +3597,9 @@ MantaClient.prototype.commitUpload = function commitUpload(id, p, opts, cb) {
         self.jsonClient.post(options, body, function (err2, req, res) {
             if (err2) {
                 log.debug(err2);
-                cb(err2);
+                cb(err2, null);
             } else {
-                cb(null);
+                cb(null, res);
             }
         });
     });
-- 
2.21.0

