commit 18e241feccecbf3665e351dc0606f8d6530d2813 (refs/changes/80/1880/3)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-05-02T01:50:37+00:00 (2 years, 5 months ago)
    
    TOOLS-1761 node-moray-sandbox should use a lower maxConnections setting

diff --git a/Makefile b/Makefile
index 5671231..7a49724 100644
--- a/Makefile
+++ b/Makefile
@@ -45,7 +45,9 @@ all: $(TAPE)
 	$(NPM) rebuild
 
 $(ESLINT): | $(NPM_EXEC)
-	$(NPM) install
+	$(NPM) install \
+	    eslint@`json -f package.json devDependencies.eslint` \
+	    eslint-plugin-joyent@`json -f package.json devDependencies.eslint-plugin-joyent`
 
 $(ISTANBUL): | $(NPM_EXEC)
 	$(NPM) install
diff --git a/README.md b/README.md
index 7ca812e..8086730 100644
--- a/README.md
+++ b/README.md
@@ -70,6 +70,13 @@ stopping Postgres. If there is only a single client, for convenience, you can
 call `.stop()` on a client and it will handle closing the client for you before
 stopping Postgres.
 
+Each additional Moray connected to a Postgres instance maintains an additional
+set of connections to the database. With enough Moray instances running, you
+can hit Postgres' maximum connection limit, which will prevent spawning new
+instances. This is normally not an issue, but can be easy to hit with Postgres
+9.2 on Mac OS X, where the max is 20. If that happens to you, newer versions of
+Postgres can help alleviate the issue.
+
 If you want to test your application's ability to handle errors, you can
 inject a series of errors to be returned by future Moray client calls. For
 example, to make the first call to `.batch()` actually call out to Moray,
diff --git a/lib/sandbox.js b/lib/sandbox.js
index d5ad200..7210258 100644
--- a/lib/sandbox.js
+++ b/lib/sandbox.js
@@ -69,7 +69,7 @@ function mkMorayConfig(log, connstr) {
             audit: false,
             standalone: {
                 pg: {
-                    maxConnections: 15,
+                    maxConnections: 5,
                     queryTimeout: 30000
                 },
                 url: connstr
@@ -179,6 +179,11 @@ MoraySandbox.prototype._startMoray = function startMoray(connstr, callback) {
     var self = this;
     var config = mkMorayConfig(self.log, connstr);
     var server = mod_moray_server.createServer(config.server);
+
+    server.on('error', function onError(err) {
+        callback(new VError(err, 'Failed to start Moray server'));
+    });
+
     server.on('ready', function setupDB() {
         server.db_conn.pg(function (cErr, pg) {
             if (cErr) {
diff --git a/package.json b/package.json
index d1e0b69..3510e80 100644
--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "name": "moray-sandbox",
   "description": "Library for managing temporary, sandboxed Moray instances",
   "main": "./lib/index.js",
-  "version": "0.1.0",
+  "version": "0.1.1",
   "keywords": [ "moray" ],
   "repository": {
     "type": "git",
