commit 5e4b5a57dc382b3ab1787a60465a3d2c48b8dbd0
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-05-07T18:33:20+00:00 (5 months ago)
    
    OS-7769 Want Intel PCH temperature sensor

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 0ac01ecc..64b11760 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -1030,6 +1030,15 @@ oce "pciex19a2,700"
 oce "pciex19a2,710"
 ohci "pciclass,0c0310"
 pcata "pccard,disk"
+pchtemp "pci8086,8ca4"
+pchtemp "pci8086,8c24"
+pchtemp "pci8086,8d24"
+pchtemp "pci8086,9d31"
+pchtemp "pci8086,a1b1"
+pchtemp "pci8086,a231"
+pchtemp "pci8086,a2b1"
+pchtemp "pci8086,a131"
+pchtemp "pci8086,a379"
 pci_pci "pci1011,1"
 pci_pci "pci1011,21"
 pci_pci "pci1014,22"
diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index a92bd1e2..f63a4cea 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -304,3 +304,4 @@ viona 306
 ppt 307
 coretemp 308
 amdf17nbdf 309
+pchtemp 310
