commit 13d0d87b201223cf515e1865e3db14fc93ffc2fc (refs/changes/30/1330/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-01-24T21:53:34+00:00 (2 years, 8 months ago)
    
    MANTA-3112 nginx suffers from port_getn() ETIME bug

diff --git a/src/event/modules/ngx_eventport_module.c b/src/event/modules/ngx_eventport_module.c
index dafa27f6..157a8e47 100644
--- a/src/event/modules/ngx_eventport_module.c
+++ b/src/event/modules/ngx_eventport_module.c
@@ -468,6 +468,14 @@ ngx_eventport_process_events(ngx_cycle_t *cycle, ngx_msec_t timer,
         ngx_time_update();
     }
 
+    /*
+     * There's a long standing condition with event ports that port_getn() may
+     * return ETIME, but there may be events available, in particular, if we
+     * were waiting for more events than were available.
+     */
+    if (n == -1 && err == ETIME && events > 0)
+	    n = 0;
+
     if (n == -1) {
         if (err == ETIME) {
             if (timer != NGX_TIMER_INFINITE) {
