commit d827df7542c49865c39a43ce22e4adbe74139896 (refs/changes/43/3043/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-12-05T15:24:17-08:00 (1 year, 10 months ago)
    
    WORKFLOW-221 upgrade to node v4

diff --git a/targets.json.in b/targets.json.in
index f2009ce..202ac95 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -517,9 +517,14 @@ cat <<EOF
     "image_name": "workflow",
     "image_description": "SDC Workflow",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
     "pkgsrc": [
-      "apg-2.2.3"
+      "coreutils-8.23nb2",
+      "curl-7.47.1",
+      "gsed-4.2.2nb4",
+      "patch-2.7.5",
+      "sudo-1.8.15"
     ],
     "repos": [
       {"url": "git@github.com:joyent/sdc-workflow.git"}
