From ef075b2d06b34a65cdb5d822f28dbc0f36e147df Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Tue, 5 Dec 2017 16:23:23 -0800
Subject: [PATCH] WORKFLOW-221 upgrade to node v4

---
 targets.json.in | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index f2009ce..6c4a6f0 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -517,9 +517,15 @@ cat <<EOF
     "image_name": "workflow",
     "image_description": "SDC Workflow",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
     "pkgsrc": [
-      "apg-2.2.3"
+      "coreutils-8.23nb2",
+      "curl-7.47.1",
+      "gsed-4.2.2nb4",
+      "patch-2.7.5",
+      "apg-2.3.0bnb3", // Required by vmapi provision workflow
+      "sudo-1.8.15"
     ],
     "repos": [
       {"url": "git@github.com:joyent/sdc-workflow.git"}
-- 
2.21.0

