commit 0b3006561a6f194bc18d1b01a5d95013b158ca04
Author: John Levon <john.levon@joyent.com>
Date:   2019-07-29T09:51:18+00:00 (2 months ago)
    
    TOOLS-2299 bin/reflash needs to regenerate build.spec.merged

diff --git a/bin/reflash b/bin/reflash
index b05fe881..ece66249 100755
--- a/bin/reflash
+++ b/bin/reflash
@@ -179,6 +179,12 @@ check_prerequisites()
     else
         fatal "Unsupported platform ${uname}"
     fi
+
+    # make sure we're using an up to date build spec
+    (cd $ROOT && gmake --quiet build.spec.merged) || {
+        echo "failed to update $ROOT/build.spec.merged" 2>&1
+        exit 1
+    }
 }
 
 check_user_sanity()
