commit f70abb21a227e976de3064c349664c9bb4647cc9 (refs/changes/28/28/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-07-05T18:44:54-07:00 (3 years, 3 months ago)
    
    joyent/node-zkstream#3 travis-ci is broken

diff --git a/test/zkserver.js b/test/zkserver.js
index 280eeb8..7164be8 100644
--- a/test/zkserver.js
+++ b/test/zkserver.js
@@ -17,7 +17,10 @@ const mod_ps = require('ps-node');
 const mod_events = require('events');
 
 function ZKServer(opts) {
-	this.zk_cmds = ['zkServer.sh', 'zkServer'];
+	this.zk_cmds = ['zkServer.sh', 'zkServer',
+	    '/usr/share/zookeeper/bin/zkServer.sh',
+	    '/usr/local/bin/zkServer.sh',
+	    '/opt/local/sbin/zkServer.sh'];
 	this.zk_opts = opts;
 	this.zk_tokill = [];
 	if (opts && opts.command)
