From 736270952c78b2bdbef4de7bd7a560e449bfae49 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 5 Jul 2016 18:44:54 -0700
Subject: [PATCH] joyent/node-zkstream#3 travis-ci is broken

---
 test/zkserver.js | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/test/zkserver.js b/test/zkserver.js
index 280eeb8..7164be8 100644
--- a/test/zkserver.js
+++ b/test/zkserver.js
@@ -17,7 +17,10 @@ const mod_ps = require('ps-node');
 const mod_events = require('events');
 
 function ZKServer(opts) {
-	this.zk_cmds = ['zkServer.sh', 'zkServer'];
+	this.zk_cmds = ['zkServer.sh', 'zkServer',
+	    '/usr/share/zookeeper/bin/zkServer.sh',
+	    '/usr/local/bin/zkServer.sh',
+	    '/opt/local/sbin/zkServer.sh'];
 	this.zk_opts = opts;
 	this.zk_tokill = [];
 	if (opts && opts.command)
-- 
2.21.0

