commit 5fe69cf48caac47efab8953bb36f440334dac6bf (refs/changes/02/1902/1)
Author: Jason King <jason.king@joyent.com>
Date:   2017-05-04T13:33:34-05:00 (2 years, 5 months ago)
    
    NAPI-412 portolan-moray.logReq should sort results

diff --git a/lib/consumer.js b/lib/consumer.js
index f135a09..0bc6f21 100644
--- a/lib/consumer.js
+++ b/lib/consumer.js
@@ -304,7 +304,12 @@ function logReq(opts, callback) {
 
     var cnUuid = opts.cnUuid;
     var filter = fmt('cn_uuid=%s', cnUuid);
-    var searchOpts = { limit: opts.limit };
+    var searchOpts = {
+        limit: opts.limit,
+        sort: {
+            attribute: '_mtime'
+        }
+    };
     var logs = [];
 
     /*
