commit 0d2965ff3d2db1ae137f0259e1a2836bc6a50a5e (refs/changes/02/1902/2)
Author: Jason King <jason.king@joyent.com>
Date:   2017-05-04T13:44:45-05:00 (2 years, 5 months ago)
    
    NAPI-412 portolan-moray.logReq should sort results

diff --git a/lib/consumer.js b/lib/consumer.js
index f135a09..99e38cc 100644
--- a/lib/consumer.js
+++ b/lib/consumer.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016, Joyent, Inc.
+ * Copyright 2017, Joyent, Inc.
  */
 
 /*
@@ -304,7 +304,12 @@ function logReq(opts, callback) {
 
     var cnUuid = opts.cnUuid;
     var filter = fmt('cn_uuid=%s', cnUuid);
-    var searchOpts = { limit: opts.limit };
+    var searchOpts = {
+        limit: opts.limit,
+        sort: {
+            attribute: '_mtime'
+        }
+    };
     var logs = [];
 
     /*
