commit abb70de01ada7f0a5b5f0d744eea8790245dc443 (refs/changes/48/2048/4)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-06-08T16:10:32+02:00 (2 years, 4 months ago)
    
    ZAPI-754 use cueball for talking to other SDC services

diff --git a/lib/apis/cnapi.js b/lib/apis/cnapi.js
index 19afd42..c6fb1bf 100644
--- a/lib/apis/cnapi.js
+++ b/lib/apis/cnapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -26,7 +26,7 @@ function Cnapi(options) {
         url: options.url,
         version: '*',
         log: options.log,
-        retry: { retries: 2 }
+        agent: options.agent
     });
 }
 
diff --git a/lib/apis/imgapi.js b/lib/apis/imgapi.js
index 7854dbc..c9d7d6b 100644
--- a/lib/apis/imgapi.js
+++ b/lib/apis/imgapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -21,7 +21,11 @@ var sdc = require('sdc-clients');
  */
 function Imgapi(options) {
     this.log = options.log;
-    this.client = new sdc.IMGAPI({ url: options.url, log: options.log });
+    this.client = new sdc.IMGAPI({
+        url: options.url,
+        log: options.log,
+        agent: options.agent
+    });
 }
 
 
diff --git a/lib/apis/napi.js b/lib/apis/napi.js
index 1441b15..1906804 100644
--- a/lib/apis/napi.js
+++ b/lib/apis/napi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -27,7 +27,8 @@ function Napi(options) {
     this.client = restify.createJsonClient({
         url: options.url,
         version: '*',
-        log: options.log
+        log: options.log,
+        agent: options.agent
     });
 }
 
diff --git a/lib/apis/papi.js b/lib/apis/papi.js
index bd5006e..7dad134 100644
--- a/lib/apis/papi.js
+++ b/lib/apis/papi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -21,7 +21,11 @@ var sdc = require('sdc-clients');
  */
 function Papi(options) {
     this.log = options.log;
-    this.client = sdc.PAPI({ url: options.url, log: options.log });
+    this.client = sdc.PAPI({
+        url: options.url,
+        log: options.log,
+        agent: options.agent
+    });
 }
 
 
diff --git a/package.json b/package.json
index 39fe63b..8b9609c 100644
--- a/package.json
+++ b/package.json
@@ -10,6 +10,7 @@
     "backoff": "2.5.0",
     "bunyan": "1.8.1",
     "changefeed": "1.3.0",
+    "cueball": "2.2.7",
     "dashdash": "1.14.1",
     "deep-diff": "0.3.3",
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
diff --git a/sapi_manifests/vmapi/template b/sapi_manifests/vmapi/template
index ecad82c..ffac729 100644
--- a/sapi_manifests/vmapi/template
+++ b/sapi_manifests/vmapi/template
@@ -1,7 +1,4 @@
 {
-    "binder": {
-        "domain": "{{{BINDER_SERVICE}}}"
-    },
     "logLevel": "debug",
     "maxSockets": 100,
     "useVmAgent": true,
@@ -64,6 +61,27 @@
             "resolvers": ["{{{BINDER_SERVICE}}}"]
         }
     },
+    "cueballHttpAgent": {
+        "resolvers": ["{{{BINDER_SERVICE}}}"],
+        "initialDomains": [
+            "{{{CNAPI_SERVICE}}}",
+            "{{{NAPI_SERVICE}}}",
+            "{{{WORKFLOW_SERVICE}}}",
+            "{{{IMGAPI_SERVICE}}}",
+            "{{{PAPI_SERVICE}}}"
+        ],
+        "spares": 4,
+        "maximum": 100,
+        "recovery": {
+            "default": {
+                "timeout": 2000,
+                "maxTimeout": 8000,
+                "retries": 5,
+                "delay": 250,
+                "maxDelay": 1000
+            }
+        }
+    },
     "changefeed": {
         "resources": [
             {
diff --git a/server.js b/server.js
index 7645456..08d6e88 100644
--- a/server.js
+++ b/server.js
@@ -14,6 +14,7 @@
 
 var assert = require('assert-plus');
 var changefeed = require('changefeed');
+var cueball = require('cueball');
 var fs = require('fs');
 var http = require('http');
 var https = require('https');
@@ -67,29 +68,39 @@ function createApiClients(config, parentLog) {
     assert.object(config, 'config');
     assert.object(parentLog, 'parentLog');
 
+    var agent;
+    if (config.cueballHttpAgent) {
+        agent = new cueball.HttpAgent(config.cueballHttpAgent);
+    }
+
     assert.object(config.cnapi, 'config.cnapi');
     var cnapiClientOpts = jsprim.deepCopy(config.cnapi);
     cnapiClientOpts.log = parentLog.child({ component: 'cnapi' }, true);
+    cnapiClientOpts.agent = agent;
     var cnapiClient = new CNAPI(cnapiClientOpts);
 
     assert.object(config.imgapi, 'config.imgapi');
     var imgapiClientOpts = jsprim.deepCopy(config.imgapi);
     imgapiClientOpts.log = parentLog.child({ component: 'imgapi' }, true);
+    imgapiClientOpts.agent = agent;
     var imgapiClient = new IMGAPI(imgapiClientOpts);
 
     assert.object(config.napi, 'config.napi');
     var napiClientOpts = jsprim.deepCopy(config.napi);
     napiClientOpts.log = parentLog.child({ component: 'napi' }, true);
+    napiClientOpts.agent = agent;
     var napiClient = new NAPI(napiClientOpts);
 
     assert.object(config.papi, 'config.papi');
     var papiClientOpts = jsprim.deepCopy(config.papi);
     papiClientOpts.log = parentLog.child({ component: 'papi' }, true);
+    papiClientOpts.agent = agent;
     var papiClient = new PAPI(papiClientOpts);
 
     assert.object(config.wfapi, 'config.wfapi');
     var wfapiClientOpts = jsprim.deepCopy(config.wfapi);
     wfapiClientOpts.log = parentLog.child({ component: 'wfapi' }, true);
+    wfapiClientOpts.agent = agent;
     var wfapiClient = new WFAPI(wfapiClientOpts);
 
     return {
