commit d0da3e77d8ee33fc1c3cc927b6c110abc4049af3 (refs/changes/21/2521/2)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-09-11T18:33:31-04:00 (2 years, 1 month ago)
    
    more changes according to first CR

diff --git a/lib/volapi.js b/lib/volapi.js
index 2d23ab9..5de7c9e 100644
--- a/lib/volapi.js
+++ b/lib/volapi.js
@@ -120,7 +120,8 @@ function pollVolumeStateChange(client, volumeUuid, options, callback) {
             } else if (volumeInSuccessState) {
                 callback(null, volume);
             } else {
-                if (latestPollTimeFromStartInMs < options.timeout) {
+                if (options.timeout === undefined ||
+                    latestPollTimeFromStartInMs < options.timeout) {
                     setTimeout(_poll, POLL_VOLUME_STATE_CHANGE_PERIOD_IN_MS);
                 } else {
                     callback(new Error('Timeout when polling for state ' +
