From d0da3e77d8ee33fc1c3cc927b6c110abc4049af3 Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Mon, 11 Sep 2017 18:33:20 -0400
Subject: [PATCH] more changes according to first CR

---
 lib/volapi.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/volapi.js b/lib/volapi.js
index 2d23ab9..5de7c9e 100644
--- a/lib/volapi.js
+++ b/lib/volapi.js
@@ -120,7 +120,8 @@ function pollVolumeStateChange(client, volumeUuid, options, callback) {
             } else if (volumeInSuccessState) {
                 callback(null, volume);
             } else {
-                if (latestPollTimeFromStartInMs < options.timeout) {
+                if (options.timeout === undefined ||
+                    latestPollTimeFromStartInMs < options.timeout) {
                     setTimeout(_poll, POLL_VOLUME_STATE_CHANGE_PERIOD_IN_MS);
                 } else {
                     callback(new Error('Timeout when polling for state ' +
-- 
2.21.0

