commit 3e04e0dd1803641fcd25e3f8e9b19b18507aa2d4 (refs/changes/64/4764/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-08-29T17:20:19+00:00 (1 year, 1 month ago)
    
    OS-7145 missing bounds check

diff --git a/usr/src/uts/common/brand/lx/os/lx_audit.c b/usr/src/uts/common/brand/lx/os/lx_audit.c
index 65b8d35e41..6f82c079c2 100644
--- a/usr/src/uts/common/brand/lx/os/lx_audit.c
+++ b/usr/src/uts/common/brand/lx/os/lx_audit.c
@@ -862,6 +862,9 @@ lx_audit_syscall_exit(int sysnum, long ret)
 	if (lxzd->lxzd_audit_enabled == LXAE_DISABLED)
 		return;
 
+	if (sysnum >= LX_NSYSCALLS)
+		return;
+
 	asp = lxzd->lxzd_audit_state;
 	ASSERT(asp != NULL);
 
