commit a6846b42f9c7f954075af3437f637b46f650d39e (refs/changes/85/1785/1)
Author: Jason King <jason.king@joyent.com>
Date:   2017-04-12T15:55:59-05:00 (2 years, 6 months ago)
    
    TOOLS-1752 Create tool for creating Triton origin images

diff --git a/.gitignore b/.gitignore
index 7e7a84d..f0c6abe 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,5 @@
 /node_modules
 /tmp
 /npm-debug.log
+/npm-0-stamp
+/image-0-stamp
diff --git a/Makefile b/Makefile
index 9c2f1a9..906c587 100644
--- a/Makefile
+++ b/Makefile
@@ -4,10 +4,29 @@
 # Makefile for triton-origin-image
 #
 
+NPM = npm
+CREATE_IMAGE = ./create-image.sh
+CONFIG = ./data.json
+
+# Should this default to /Joyent_Dev/stor/builds/.... ?
+MANTA_PATH =
+
+# -------
 
 #
 # Targets
 #
-.PHONY: all
-all:
+.PHONY: all image
+
+all: npm-0-stamp image-0-stamp
+
+npm-0-stamp:
+	$(NPM) install
+	touch $@
+
+image-0-stamp: npm-0-stamp
+	$(CREATE_IMAGE) -f $(CONFIG) $(MANTA_PATH)
+	touch $@
 
+clean:
+	rm -f npm-0-stamp image-0-stamp
diff --git a/package.json b/package.json
index 93cc55a..1e3432c 100644
--- a/package.json
+++ b/package.json
@@ -6,7 +6,7 @@
   "private": true,
   "homepage": "https://github.com/joyent/triton-origin-image",
   "dependencies": {
-    "triton": "5.1.1"
+    "triton": "5.2.0"
   },
   "repository": {
     "type": "git",
