From a4ad2460400303b447fb265daa70ca699a47830c Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Mon, 16 Sep 2019 13:34:24 +0000
Subject: [PATCH] OS-7945 kvm needs to be built with retpolines

---
 Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index cee6818..71245cc 100644
--- a/Makefile
+++ b/Makefile
@@ -123,7 +123,9 @@ KERNEL_CFLAGS = \
 #
 ifneq ($(PRIMARY_COMPILER_VER),4)
 KERNEL_CFLAGS += \
-	-fno-shrink-wrap
+	-fno-shrink-wrap \
+	-mindirect-branch=thunk-extern \
+	-mindirect-branch-register
 endif
 
 USER_CFLAGS = \
-- 
2.21.0

