commit a4ad2460400303b447fb265daa70ca699a47830c
Author: John Levon <john.levon@joyent.com>
Date:   2019-09-16T13:34:24+00:00 (3 weeks ago)
    
    OS-7945 kvm needs to be built with retpolines

diff --git a/Makefile b/Makefile
index cee6818..71245cc 100644
--- a/Makefile
+++ b/Makefile
@@ -123,7 +123,9 @@ KERNEL_CFLAGS = \
 #
 ifneq ($(PRIMARY_COMPILER_VER),4)
 KERNEL_CFLAGS += \
-	-fno-shrink-wrap
+	-fno-shrink-wrap \
+	-mindirect-branch=thunk-extern \
+	-mindirect-branch-register
 endif
 
 USER_CFLAGS = \
