From 79055b1e6029dea8e79c60ded394933a5151b07d Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Tue, 8 Aug 2017 13:31:37 -0700
Subject: [PATCH] MANTA-3187 Muskie logs object IDs incorrectly for PUTs

---
 lib/audit.js | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/lib/audit.js b/lib/audit.js
index d352464..3841d4c 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -227,8 +227,19 @@ function auditLogger(options) {
             secure: req.secure
         };
 
+        /*
+         * For operations which save objects, the final objectId may no longer
+         * match the objectId in the request metadata. If our request has an
+         * updated objectId, it will be in req.objectId. We prefer to report
+         * the final objectId, and report the earlier objectId as prevObjectId.
+         */
         if (req.metadata && req.metadata.objectId) {
-            obj.objectId = req.metadata.objectId;
+            if (req.objectId && req.objectId !== req.metadata.objectId) {
+                obj.prevObjectId = req.metadata.objectId;
+                obj.objectId = req.objectId;
+            } else {
+                obj.objectId = req.metadata.objectId;
+            }
         }
         obj.sharksContacted = req.sharksContacted;
         obj.entryShard = req.entryShard;
-- 
2.21.0

