commit 9d3569df9a279f879db49d1b7e96e8612b801f6c (refs/changes/45/2345/3)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-08-10T00:06:55+00:00 (2 years, 2 months ago)
    
    MANTA-3187 Muskie logs object IDs incorrectly for PUTs

diff --git a/lib/audit.js b/lib/audit.js
index d352464..3841d4c 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -227,8 +227,19 @@ function auditLogger(options) {
             secure: req.secure
         };
 
+        /*
+         * For operations which save objects, the final objectId may no longer
+         * match the objectId in the request metadata. If our request has an
+         * updated objectId, it will be in req.objectId. We prefer to report
+         * the final objectId, and report the earlier objectId as prevObjectId.
+         */
         if (req.metadata && req.metadata.objectId) {
-            obj.objectId = req.metadata.objectId;
+            if (req.objectId && req.objectId !== req.metadata.objectId) {
+                obj.prevObjectId = req.metadata.objectId;
+                obj.objectId = req.objectId;
+            } else {
+                obj.objectId = req.metadata.objectId;
+            }
         }
         obj.sharksContacted = req.sharksContacted;
         obj.entryShard = req.entryShard;
