commit 5031f0398be607cc11416db2b9fbd51b13caa8e9 (refs/changes/45/2345/8)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-08-23T21:22:36+00:00 (2 years, 2 months ago)
    
    MANTA-3187 Muskie logs object IDs incorrectly for PUTs

diff --git a/lib/audit.js b/lib/audit.js
index 8876d61..f45fa40 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -246,8 +246,27 @@ function auditLogger(options) {
             secure: req.secure
         };
 
+        /*
+         * Muskie keeps a reference to the Moray metadata loaded for the
+         * request at req.metadata. In cases where a new object ID is
+         * generated (such as a PUT overwriting an existing object,) we
+         * want to differentiate between the new object ID for the
+         * object and the previous one (and supply both objectIds.)
+         */
         if (req.metadata && req.metadata.objectId) {
-            obj.objectId = req.metadata.objectId;
+            if (req.objectId && req.objectId !== req.metadata.objectId) {
+                obj.prevObjectId = req.metadata.objectId;
+                obj.objectId = req.objectId;
+            } else {
+                obj.objectId = req.metadata.objectId;
+            }
+        } else if (req.objectId) {
+            /*
+             * For a PUT which is creating a fresh object, the req.metadata
+             * will have no objectId, we still want to log its objectId, if
+             * available.
+             */
+            obj.objectId = req.objectId;
         }
         obj.sharksContacted = req.sharksContacted;
         obj.entryShard = req.entryShard;
