commit db978497a76af4ec5a3ca343f0c67beb4c6e2c78 (refs/changes/45/945/3)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2016-11-24T11:44:22-08:00 (2 years, 10 months ago)
    
    CNAPI-679 CNAPI should use node v4.x

diff --git a/Makefile b/Makefile
index d0955e0..0c13979 100644
--- a/Makefile
+++ b/Makefile
@@ -42,11 +42,11 @@ JSSTYLE_FLAGS	= -o indent=4,doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/cnapi.xml.in
 SMF_DTD		= $(REPO_ROOT)/tools/service_bundle.dtd.1
 
-NODE_PREBUILT_VERSION=v0.10.42
+NODE_PREBUILT_VERSION=v4.6.1
 NODE_PREBUILT_TAG=zone
 ifeq ($(shell uname -s),SunOS)
-	# Allow building on a SmartOS image other than sdc-smartos@1.6.3.
-	NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
+	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
+	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
 endif
 
 COAL ?= root@10.99.99.7
diff --git a/lib/apis/moray.js b/lib/apis/moray.js
index 7f556a0..5252688 100644
--- a/lib/apis/moray.js
+++ b/lib/apis/moray.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -16,6 +16,7 @@ var moray_client = require('moray');
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
 var async = require('async');
+var verror = require('verror');
 
 var BUCKETS = {
     'servers': {
@@ -120,13 +121,7 @@ Moray.prototype.connect = function () {
             serializers: bunyan.stdSerializers
         }),
         reconnect: true,
-        noCache: true,
-        connectTimeout: 10000,
-        retry: {
-            retries: Infinity,
-            minTimeout: 1000,
-            maxTimeout: 16000
-        }
+        noCache: true
     });
 
     function onConnect() {
@@ -166,18 +161,20 @@ Moray.prototype.ensureClientReady = function (callback) {
                         return;
                     }
 
-                    if (error.name === 'ConnectTimeoutError') {
+                    if (verror.hasCauseWithName(error, 'ConnectTimeoutError')) {
                         setTimeout(next, 1000);
                         return;
                     }
 
-                    if (error.name === 'BucketNotFoundError') {
+                    if (verror.hasCauseWithName(error, 'BucketNotFoundError')) {
                         run = false;
                         next();
                         return;
                     }
 
-                    if (error.name !== 'NoDatabasePeersError') {
+                    if (!verror.hasCauseWithName(error,
+                                                 'NoDatabasePeersError'))
+                    {
                         self.log.info(
                             'Received %s from moray', error.message);
                         self.log.info({ error: error });
@@ -221,7 +218,9 @@ Moray.prototype.initializeBuckets = function (callback) {
 
                     function onbucket(error) {
                         if (error) {
-                            if (error.name === 'BucketNotFoundError') {
+                            if (verror.hasCauseWithName(error,
+                                                        'BucketNotFoundError'))
+                            {
                                 self.log.info(
                                     'Moray bucket \'%s\','
                                     + ' does not yet exist, we will create it',
@@ -256,7 +255,9 @@ Moray.prototype.initializeBuckets = function (callback) {
                 function (error, obj) {
                     if (error) {
 
-                        if (error.name === 'ObjectNotFoundError') {
+                        if (verror.hasCauseWithName(error,
+                                                    'ObjectNotFoundError'))
+                        {
                             var ModelServer = require('../models/server');
                             self.log.info(
                                 'Default object does not yet exist, creating'
diff --git a/lib/common.js b/lib/common.js
index 18f6bd0..4d0a0dc 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -5,11 +5,11 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 var fs = require('fs');
-var uuid = require('node-uuid');
+var libuuid = require('libuuid');
 var qs = require('querystring');
 
 
@@ -72,7 +72,7 @@ function timestamp() {
 
 
 function genId() {
-    return uuid.v4();
+    return libuuid.create();
 }
 
 
diff --git a/lib/endpoints/tasks.js b/lib/endpoints/tasks.js
index 6ec11d4..7ae4527 100644
--- a/lib/endpoints/tasks.js
+++ b/lib/endpoints/tasks.js
@@ -5,11 +5,12 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 var restify = require('restify');
 var async = require('async');
+var verror = require('verror');
 var ModelServer = require('../models/server');
 var buckets = require('../apis/moray').BUCKETS;
 
@@ -33,7 +34,9 @@ Task.get = function (req, res, next) {
         buckets.tasks.name,
         req.params.taskid,
         function (error, obj) {
-            if (error && error.name === 'ObjectNotFoundError') {
+            if (error && verror.hasCauseWithName(error,
+                                                 'ObjectNotFoundError'))
+            {
                 next(new restify.ResourceNotFoundError(
                     'no such task found'));
                 return;
@@ -87,7 +90,9 @@ Task.wait = function (req, res, next) {
             req.params.taskid,
             { noCache: true },
             function (err, obj) {
-                if (err && err.name === 'ObjectNotFoundError') {
+                if (err && verror.hasCauseWithName(err,
+                                                   'ObjectNotFoundError'))
+                {
                     next(new restify.ResourceNotFoundError(
                         'no such task found'));
                     return;
diff --git a/lib/endpoints/zfs.js b/lib/endpoints/zfs.js
index 12bff3b..a18258d 100644
--- a/lib/endpoints/zfs.js
+++ b/lib/endpoints/zfs.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -34,8 +34,12 @@ function ZFS() {}
 
 ZFS.listDatasets = function (req, res, next) {
     var server = req.stash.server;
+    var opts = {
+        params: req.params,
+        req: req
+    };
 
-    server.zfsTask('zfs_list_datasets', req.params, function (err, results) {
+    server.zfsTask('zfs_list_datasets', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot list datasets:' + err.message)));
@@ -62,8 +66,12 @@ ZFS.listDatasets = function (req, res, next) {
 
 ZFS.createDataset = function (req, res, next) {
     var server = req.stash.server;
+    var opts = {
+        params: req.params,
+        req: req
+    };
 
-    server.zfsTask('zfs_create_dataset', req.params, function (err) {
+    server.zfsTask('zfs_create_dataset', opts, function (err) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot create dataset:' + err.message)));
@@ -92,13 +100,17 @@ ZFS.createDataset = function (req, res, next) {
 
 ZFS.createSnapshot = function (req, res, next) {
     var server = req.stash.server;
-
     var name = req.params.name;
     var snapshot = req.params.dataset + '@' + name;
 
+    var opts = {
+        params: { dataset: snapshot },
+        req: req
+    };
+
     server.zfsTask(
         'zfs_snapshot_dataset',
-        { dataset: snapshot },
+        opts,
         function (err) {
             if (err) {
                 next(new restify.InternalError(
@@ -132,9 +144,14 @@ ZFS.rollbackSnapshot = function (req, res, next) {
     var name = req.params.name;
     var snapshot = req.params.dataset + '@' + name;
 
+    var opts = {
+        params: { dataset: snapshot },
+        req: req
+    };
+
     server.zfsTask(
         'zfs_rollback_dataset',
-        { dataset: snapshot },
+        opts,
         function (err) {
             if (err) {
                 next(new restify.InternalError(
@@ -163,7 +180,12 @@ ZFS.rollbackSnapshot = function (req, res, next) {
 ZFS.listSnapshots = function (req, res, next) {
     var server = req.stash.server;
 
-    server.zfsTask('zfs_list_snapshots', req.params, function (err, results) {
+    var opts = {
+        params: req.params,
+        req: req
+    };
+
+    server.zfsTask('zfs_list_snapshots', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot list snapshots:' + err.message)));
@@ -208,12 +230,17 @@ function findProps(params) {
 ZFS.getAllProperties = function (req, res, next) {
     var server = req.stash.server;
 
-    var options = {};
+    var params = {};
     var properties = findProps(req.params);
     if (properties.length > 0)
-        options.properties = properties;
+        params.properties = properties;
+
+    var opts = {
+        params: params,
+        req: req
+    };
 
-    server.zfsTask('zfs_get_properties', options, function (err, results) {
+    server.zfsTask('zfs_get_properties', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot get properties:' + err.message)));
@@ -253,15 +280,18 @@ ZFS.getProperties = function (req, res, next) {
     var server = req.stash.server;
     var dataset = req.params.dataset;
 
-    var options = {};
-    options.dataset = dataset;
+    var params = {};
+    params.dataset = dataset;
     var properties = findProps(req.params);
     if (properties.length > 0)
-        options.properties = properties;
+        params.properties = properties;
 
-    // XXX Handle case where pool is specified or not
+    var opts = {
+        params: params,
+        req: req
+    };
 
-    server.zfsTask('zfs_get_properties', options, function (err, results) {
+    server.zfsTask('zfs_get_properties', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot get properties:' + err.message)));
@@ -299,7 +329,12 @@ ZFS.getProperties = function (req, res, next) {
 ZFS.setProperties = function (req, res, next) {
     var server = req.stash.server;
 
-    server.zfsTask('zfs_set_properties', req.params, function (err) {
+    var opts = {
+        params: req.params,
+        req: req
+    };
+
+    server.zfsTask('zfs_set_properties', opts, function (err) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot set properties:' + err.message)));
@@ -327,7 +362,13 @@ ZFS.setProperties = function (req, res, next) {
 ZFS.destroyDataset = function (req, res, next) {
     var server = req.stash.server;
 
-    server.zfsTask('zfs_destroy_dataset', req.params, function (err) {
+    var opts = {
+        params: req.params,
+        req: req
+    };
+
+
+    server.zfsTask('zfs_destroy_dataset', opts, function (err) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot destroy dataset:' + err.message)));
@@ -354,7 +395,12 @@ ZFS.destroyDataset = function (req, res, next) {
 ZFS.listZpools = function (req, res, next) {
     var server = req.stash.server;
 
-    server.zfsTask('zfs_list_pools', req.params, function (err, results) {
+    var opts = {
+        params: req.params,
+        req: req
+    };
+
+    server.zfsTask('zfs_list_pools', opts, function (err, results) {
         if (err) {
             return (next(new restify.InternalError(
                 'cannot list pools:' + err.message)));
diff --git a/lib/models/image.js b/lib/models/image.js
index 9d0f5cf..40fb1b7 100644
--- a/lib/models/image.js
+++ b/lib/models/image.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -16,7 +16,6 @@
 var async = require('async');
 var sprintf = require('sprintf').sprintf;
 var util = require('util');
-var nodeuuid = require('node-uuid');
 
 var ModelBase = require('./base');
 var ModelServer;
diff --git a/lib/models/server.js b/lib/models/server.js
index 334d75e..67b5e0e 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -849,7 +849,9 @@ ModelServer.prototype.getRaw = function (extras, callback) {
             buckets.servers.name,
             uuid,
             function (error, obj) {
-                if (error && error.name === 'ObjectNotFoundError') {
+                if (error && verror.hasCauseWithName(error,
+                                                     'ObjectNotFoundError'))
+                {
                     self.exists = false;
                     self.log.error('Server %s not found in moray', uuid);
 
@@ -872,7 +874,8 @@ ModelServer.prototype.getRaw = function (extras, callback) {
                         uuid,
                         function (getStatusError, statusObj) {
                             if (getStatusError &&
-                                getStatusError.name === 'ObjectNotFoundError')
+                                verror.hasCauseWithName(getStatusError,
+                                                        'ObjectNotFoundError'))
                             {
                                 self.log.error(
                                     'last_heartbeat for %s not found in moray',
@@ -1254,7 +1257,7 @@ function Server$getLastHeartbeat(callback) {
         onGetObject);
 
     function onGetObject(error, obj) {
-        if (error && error.name === 'ObjectNotFoundError') {
+        if (error && verror.hasCauseWithName(error, 'ObjectNotFoundError')) {
             self.log.error(
                 'last_heartbeat for %s not found in moray', uuid);
             callback();
@@ -2176,7 +2179,8 @@ function (opts) {
 
             client = restify.createJsonClient(cOpts);
 
-            log.info('posting task to %s%s', cOpts.url, rOpts.path);
+            log.info('posting task to %s%s (req_id=%s)',
+                     cOpts.url, rOpts.path, req_id);
 
             // write initial task to moray
             // post http request
@@ -2361,7 +2365,25 @@ ModelServer.prototype.sendRequest = function (opts, cb) {
     });
 };
 
-ModelServer.prototype.zfsTask = function (task, options, callback) {
+ModelServer.prototype.zfsTask = function (task, opts, callback) {
+    var self = this;
+
+    var request = {
+        task: task,
+        cb: function (error, taskstatus) {
+        },
+        evcb: function () {},
+        synccb: function (error, result) {
+            callback(error, result);
+        },
+        req_id: opts.req.getId(),
+        params: opts.params
+    };
+
+    self.sendTaskRequest(request);
+};
+
+ModelServer.prototype.zfsTaskOld = function (task, options, callback) {
     var self = this;
 
     var request = {
diff --git a/lib/models/vm.js b/lib/models/vm.js
index f914029..7c62c10 100644
--- a/lib/models/vm.js
+++ b/lib/models/vm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -17,7 +17,6 @@ var async = require('async');
 var assert = require('assert-plus');
 var sprintf = require('sprintf').sprintf;
 var util = require('util');
-var nodeuuid = require('node-uuid');
 var sdcClients = require('sdc-clients');
 
 var ModelBase = require('./base');
diff --git a/lib/models/waitlist.js b/lib/models/waitlist.js
index 5e3f332..2a3860c 100644
--- a/lib/models/waitlist.js
+++ b/lib/models/waitlist.js
@@ -15,7 +15,7 @@
 
 var assert = require('assert-plus');
 var async = require('async');
-var libuuid = require('node-uuid');
+var libuuid = require('libuuid');
 var once = require('once');
 var sprintf = require('sprintf').sprintf;
 var vasync = require('vasync');
@@ -546,7 +546,7 @@ ModelWaitlist.getTicket = function (uuid, callback) {
         buckets.waitlist_tickets.name, uuid, onGet);
 
     function onGet(error, obj) {
-        if (error && error.name === 'ObjectNotFoundError') {
+        if (error && verror.hasCauseWithName(error, 'ObjectNotFoundError')) {
             self.log.error('Ticket %s not found in moray', uuid);
             callback();
             return;
@@ -572,7 +572,7 @@ ModelWaitlist.prototype.getServerQueue = function (callback) {
         self.uuid,
         function (err, response) {
             var res = { etag: null };
-            if (err && err.name === 'ObjectNotFoundError') {
+            if (err && verror.hasCauseWithName(err, 'ObjectNotFoundError')) {
                 self.log.error(
                     'Ticket queue for %s not found in moray',
                     self.uuid);
@@ -638,14 +638,14 @@ ModelWaitlist.prototype.ensureServerQueue = function (callback) {
                 serverqueue,
                 { etag: etag },
                 function (err, response) {
-                    if (err && (err.name === 'EtagConflictError' ||
-                               (err.name === 'UniqueAttributeError')))
+                    if (err &&
+                       (verror.hasCauseWithName(err, 'EtagConflictError') ||
+                       (verror.hasCauseWithName(err, 'UniqueAttributeError'))))
                     {
                         attempts++;
-                        self.log.warn(
-                            'waitlist (%s) collision on queue ' +
+                        self.log.warn({ err: err },
+                            'waitlist collision on queue ' +
                             'initialization after %s attempts, retrying',
-                            err.name,
                             attempts);
                         process.nextTick(function () {
                             self.ensureServerQueue(wfcb);
@@ -743,8 +743,9 @@ function (ticket, serverqueue, etag, callback) {
         },
         function (wfcb) {
             moray.batch(data, function (err, meta) {
-                if (err && (err.name === 'EtagConflictError' ||
-                           (err.name === 'UniqueAttributeError')))
+                if (err &&
+                   (verror.hasCauseWithName(err, 'EtagConflictError') ||
+                   (verror.hasCauseWithName(err, 'UniqueAttributeError'))))
                 {
                     process.nextTick(function () {
                         self.ensureServerQueue(function (err2, resp) {
@@ -856,8 +857,9 @@ function (ticket, serverqueue, etag, callback) {
         },
         function (wfcb) {
             moray.batch(data, function (err, meta) {
-                if (err && (err.name === 'EtagConflictError' ||
-                           (err.name === 'UniqueAttributeError')))
+                if (err &&
+                   (verror.hasCauseWithName(err, 'EtagConflictError') ||
+                   (verror.hasCauseWithName(err, 'UniqueAttributeError'))))
                 {
                     process.nextTick(function () {
                         self.ensureServerQueue(function (err2, resp) {
@@ -971,8 +973,9 @@ function (ticket, serverqueue, etag, callback) {
             self.log.info({ batch: data },
                           'doing batch write after finishing %s', ticket.uuid);
             moray.batch(data, function (err, meta) {
-                if (err && (err.name === 'EtagConflictError' ||
-                           (err.name === 'UniqueAttributeError')))
+                if (err &&
+                   (verror.hasCauseWithName(err, 'EtagConflictError') ||
+                   (verror.hasCauseWithName(err, 'UniqueAttributeError'))))
                 {
                     process.nextTick(function () {
                         self.ensureServerQueue(function (err2, resp) {
@@ -1046,8 +1049,8 @@ function (ticket, serverqueue, etag, callback) {
 
     var moray = ModelWaitlist.getMoray();
     moray.batch(data, function (err, meta) {
-        if (err && (err.name === 'EtagConflictError' ||
-                   (err.name === 'UniqueAttributeError')))
+        if (err && (verror.hasCauseWithName(err, 'EtagConflictError') ||
+                   (verror.hasCauseWithName(err, 'UniqueAttributeError'))))
         {
             process.nextTick(function () {
                 self.ensureServerQueue(function (err2, resp) {
@@ -1079,7 +1082,7 @@ ModelWaitlist.prototype.createTicket = function (params, callback) {
     assert.string(params.id, 'params.id');
     assert.string(params.expires_at, 'params.expires_at');
 
-    var ticket_uuid = libuuid.v4();
+    var ticket_uuid = libuuid.create();
 
     var serverqueue = null;
     var etag = null;
diff --git a/package.json b/package.json
index 693c140..6a3930f 100644
--- a/package.json
+++ b/package.json
@@ -6,29 +6,26 @@
   "private": true,
   "dependencies": {
     "amqp": "git+https://github.com/postwait/node-amqp.git#7a421793796bf66f5741282624c3db3ae266b3b5",
-    "assert-plus": "0.1.5",
+    "assert-plus": "1.0.0",
     "async": "1.5.2",
-    "bunyan": "1.3.4",
+    "bunyan": "1.8.5",
     "dapi": "git+https://github.com/joyent/sdc-designation.git#db24809071ed2c2e5fe26fbd0b8d27d4b0f01d0a",
     "deep-equal": "git+https://github.com/substack/node-deep-equal.git#b2cfeb95da6982f8bc2e676231820a3a47385db3",
     "dox": "0.4.1",
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
-    "filed": "0.0.5",
     "jsprim": "1.3.0",
-    "memwatch": "git+https://github.com/lloyd/node-memwatch.git#475ce024a53c74ad5baff27003b81499896e81a5",
-    "moray": "git+https://github.com/joyent/node-moray.git#fd5781bc",
-    "node-uuid": "1.2.0",
+    "moray": "2.0.0",
+    "libuuid": "0.2.1",
     "nodeunit": "0.9.1",
     "once": "1.3.0",
-    "restify": "^3.0.1",
+    "restify": "4.2.0",
     "restify-validator": "0.3.0",
-    "sdc-clients": "git+https://github.com/joyent/node-sdc-clients.git#121bd661b8301a040677f65a289d59043171ce01",
-    "sprintf": "0.1.1",
+    "sdc-clients": "10.0.1",
+    "sprintf": "0.1.5",
     "trace-event": "1.3.0",
-    "vasync": "git+https://github.com/davepacheco/node-vasync.git#e8743bb07a09dd83ea2df7e6d241c8a78b67aec6",
-    "verror": "1.3.4",
-    "watershed": "0.3.0",
-    "wf-client": "git+https://github.com/joyent/sdc-wf-client.git#a8590b497e40aea87610fa1c370fcbed64f43367",
+    "vasync": "1.6.3",
+    "verror": "1.9.0",
+    "wf-client": "0.2.0",
     "wf-shared": "git+https://github.com/joyent/sdc-wf-shared.git#9aed2e8"
   },
   "devDependencies": {
diff --git a/scripts/cnapi-delete-all-servers.sh b/scripts/cnapi-delete-all-servers.sh
index b09820d..53bd340 100755
--- a/scripts/cnapi-delete-all-servers.sh
+++ b/scripts/cnapi-delete-all-servers.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2016, Joyent, Inc.
 #
 
 set -x xtrace
@@ -21,6 +21,8 @@ zlogin $UUID svcadm disable cnapi
 zlogin $UUID /opt/smartdc/cnapi/{build/node/bin/node,node_modules/.bin/delbucket} -h $CONFIG_moray_admin_ips cnapi_servers &
 zlogin $UUID /opt/smartdc/cnapi/{build/node/bin/node,node_modules/.bin/delbucket} -h $CONFIG_moray_admin_ips cnapi_waitlist_tickets &
 zlogin $UUID /opt/smartdc/cnapi/{build/node/bin/node,node_modules/.bin/delbucket} -h $CONFIG_moray_admin_ips cnapi_waitlist_queues &
+zlogin $UUID /opt/smartdc/cnapi/{build/node/bin/node,node_modules/.bin/delbucket} -h $CONFIG_moray_admin_ips cnapi_status &
+zlogin $UUID /opt/smartdc/cnapi/{build/node/bin/node,node_modules/.bin/delbucket} -h $CONFIG_moray_admin_ips cnapi_tasks &
 
 wait
 zlogin $UUID svcadm enable cnapi
