commit 7a26c3b6539fecf080204ad44f3775f8f6f932da (refs/changes/88/988/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton account update triton_cns_enabled=false

diff --git a/lib/do_account/do_update.js b/lib/do_account/do_update.js
index fa0f62f..876debc 100644
--- a/lib/do_account/do_update.js
+++ b/lib/do_account/do_update.js
@@ -30,7 +30,9 @@ function do_update(subcmd, opts, args, callback) {
     var log = this.log;
     var tritonapi = this.top.tritonapi;
 
-    vasync.pipeline({arg: {}, funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
+
         function gatherDataArgs(ctx, next) {
             if (opts.file) {
                 next();
