commit 10454305f84ff41d88d20398376b6861da20f443
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-18T15:37:13+00:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index ed0c5c1..4991c6c 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -6,7 +6,7 @@
 	url = git://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
 	url = https://github.com/joyent/eng.git
diff --git a/deps/jsstyle b/deps/jsstyle
index ab8f1fc..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit ab8f1fc8d90db8484bbed8a526ff40add5aa1c15
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
