commit 9fcbc2bb305f0d780e24e9285cc54ea6eab0f723 (refs/changes/27/2227/3)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-08-07T23:15:16+12:00 (2 years, 2 months ago)
    
    ADMINUI-2314: Upgrade Node.js version from 0.10 to 4.x

diff --git a/.gitignore b/.gitignore
index ca4a2fec..18042af7 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
+*.swp
 /node_modules
 /tmp
 build
diff --git a/Makefile b/Makefile
index cf21c053..438df634 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -21,16 +21,12 @@
 # other projects too, you should add these to the original versions of the
 # included Makefiles (in eng.git) so that other teams can use them too.
 #
-
-#
-# Tools
-#
-MOCHA_PHANTOMJS	:= ./node_modules/.bin/mocha-phantomjs
+NAME		:= adminui
 
 #
 # Files
 #
-JS_FILES	:= $(shell ls *.js) $(shell find lib test -name '*.js')
+JS_FILES	:= $(shell ls *.js) $(shell find lib -name '*.js')
 JSL_CONF_NODE	 = tools/jsl.node.conf
 JSL_FILES_NODE   = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
@@ -39,10 +35,11 @@ JSSTYLE_FLAGS    = -o "indent=2,doxygen,unparenthesized-return=0,line-length=120
 REPO_MODULES	 =
 SMF_MANIFESTS_IN = smf/manifests/adminui.xml.in
 
-NODE_PREBUILT_VERSION=v0.10.48
-NODE_PREBUILT_TAG=zone
-NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
-
+NODE_PREBUILT_VERSION=v4.8.4
+ifeq ($(shell uname -s),SunOS)
+	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
+	NODE_PREBUILT_TAG=zone
+endif
 
 
 include ./tools/mk/Makefile.defs
@@ -53,8 +50,15 @@ else
 endif
 include ./tools/mk/Makefile.smf.defs
 
+
+#
+# Variables
+#
+
+# Mountain Gorilla-spec'd versioning.
+
 ROOT            := $(shell pwd)
-RELEASE_TARBALL := adminui-pkg-$(STAMP).tar.bz2
+RELEASE_TARBALL := $(NAME)-pkg-$(STAMP).tar.bz2
 RELSTAGEDIR          := /tmp/$(STAMP)
 
 
@@ -73,7 +77,7 @@ all: $(SMF_MANIFESTS) node_modules js sdc-scripts
 .PHONY: dev
 dev: node_modules_dev sdc-scripts
 
-.PHONY: node_modules
+.PHONY: node_modules node_modules_dev
 node_modules: | $(NPM_EXEC)
 	$(NPM) install --production
 
@@ -103,13 +107,13 @@ test: | $(JS_BUNDLE)
 .PHONY: release
 release: all deps docs $(SMF_MANIFESTS)
 	@echo "Building $(RELEASE_TARBALL)"
-	@mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/adminui
+	@mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)
 	@mkdir -p $(RELSTAGEDIR)/site
 	@touch $(RELSTAGEDIR)/site/.do-not-delete-me
-	cp -r $(ROOT)/* $(RELSTAGEDIR)/root/opt/smartdc/adminui/
-	rm $(RELSTAGEDIR)/root/opt/smartdc/adminui/image-version.json
-	@echo "{\"version\": \"$(STAMP)\"}" >> $(RELSTAGEDIR)/root/opt/smartdc/adminui/image-version.json
-	rm -rf $(RELSTAGEDIR)/root/opt/smartdc/adminui/deps
+	cp -r $(ROOT)/* $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)/
+	rm $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)/image-version.json
+	@echo "{\"version\": \"$(STAMP)\"}" >> $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)/image-version.json
+	rm -rf $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)/deps
 	mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/boot
 	cp -R $(ROOT)/deps/sdc-scripts/* $(RELSTAGEDIR)/root/opt/smartdc/boot/
 	cp -R $(ROOT)/boot/* $(RELSTAGEDIR)/root/opt/smartdc/boot/
@@ -123,8 +127,8 @@ publish: release
 		echo "error: 'BITS_DIR' must be set for 'publish' target"; \
 		exit 1; \
 	fi
-	mkdir -p $(BITS_DIR)/adminui
-	cp $(ROOT)/$(RELEASE_TARBALL) $(BITS_DIR)/adminui/$(RELEASE_TARBALL)
+	mkdir -p $(BITS_DIR)/$(NAME)
+	cp $(ROOT)/$(RELEASE_TARBALL) $(BITS_DIR)/$(NAME)/$(RELEASE_TARBALL)
 
 
 
diff --git a/README.md b/README.md
index e10e031d..574c3bfe 100644
--- a/README.md
+++ b/README.md
@@ -46,7 +46,7 @@ services running in
     make check
 
     make devrun
-    # devrun automatically restarts node process and incrementally re-builds assets
+    # devrun rebuilds client-side assets and runs server
     ```
 
 - testing:
diff --git a/boot/configure.sh b/boot/configure.sh
index a9916488..3db8e218 100755
--- a/boot/configure.sh
+++ b/boot/configure.sh
@@ -7,14 +7,14 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
 set -o xtrace
 
 echo "Updating adminui service manifest"
-$(/opt/local/bin/gsed -i"" \
+$(/usr/bin/sed -i"" \
   -e "s/@@PREFIX@@/\/opt\/smartdc\/adminui/g" \
   /opt/smartdc/adminui/smf/manifests/adminui.xml)
 
diff --git a/lib/amon.js b/lib/amon.js
index aeee89e8..319a6c77 100644
--- a/lib/amon.js
+++ b/lib/amon.js
@@ -9,7 +9,7 @@
  */
 
 var restify = require('restify');
-var fmt = require('sys').format;
+var fmt = require('util').format;
 
 module.exports = {
     createProbe: createProbe,
diff --git a/lib/sdc-clients.js b/lib/sdc-clients.js
index a0a893e7..d53ee5b3 100644
--- a/lib/sdc-clients.js
+++ b/lib/sdc-clients.js
@@ -105,7 +105,7 @@ module.exports.createClients = function (config) {
         cfg.log = opts.log;
         cfg.reconnect = true;
         cfg.retry = {
-            retries: Infinity,
+            retries: 999999,  // sdc-clients won't take Infinity
             minTimeout: 1000,
             maxTimeout: 16000
         };
diff --git a/lib/servers.js b/lib/servers.js
index 735b3e65..3a40592a 100644
--- a/lib/servers.js
+++ b/lib/servers.js
@@ -9,7 +9,7 @@
  */
 
 var _ = require('underscore');
-var format = require('sys').format;
+var format = require('util').format;
 var getRequestHeaders = require('./get-request-headers');
 var utils = require('./utils');
 var async = require('async');
diff --git a/lib/sessions.js b/lib/sessions.js
index a3b16353..1d84cf64 100644
--- a/lib/sessions.js
+++ b/lib/sessions.js
@@ -5,11 +5,12 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var assert = require('assert');
 var crypto = require('crypto');
+var verror = require('verror');
 
 var TOKEN_TIMEOUT = 300 * 1000;
 var SESSIONS_BUCKET = 'sdc_adminui_sessions';
@@ -101,7 +102,7 @@ Sessions.prototype.get = function (id, cb) {
     this.client.getObject(SESSIONS_BUCKET, id, function (err, obj) {
         self.log.debug('session.get cb', err, obj);
         if (err) {
-            if (err.name === 'ObjectNotFoundError') {
+            if (verror.hasCauseWithName(err, 'ObjectNotFoundError')) {
                 self.log.debug('session doesn\'t exist %s', id);
                 return cb(null, null);
             } else {
diff --git a/package.json b/package.json
index 4d6fa6a4..85c9bc5c 100644
--- a/package.json
+++ b/package.json
@@ -1,67 +1,51 @@
 {
   "author": "Kevin Chan",
   "name": "adminui",
-  "description": "SmartDataCenter Operations Portal",
-  "version": "1.0.6",
+  "description": "Triton Operations Portal",
+  "version": "1.1.0",
   "private": true,
   "repository": {
     "git": "git@git.github.com:joyent/sdc-adminui.git"
   },
   "engines": {
-    "node": "0.8.x"
+    "node": ">=0.10"
   },
   "browser": "./www/adminui.js",
   "main": "./server.js",
   "dependencies": {
-    "assert-plus": "0.1.5",
+    "assert-plus": "1.0.0",
     "async": "0.2.10",
     "backbone": "1.1.2",
     "backbone.paginator": "2.0.2",
     "browserify": "3.46.1",
     "browserify-shim": "2.0.10",
-    "bunyan": "1.3.5",
+    "bunyan": "1.8.10",
     "classnames": "1.2.2",
     "clone": "0.1.19",
-    "envify": "1.2.0",
     "filed": "0.1.0",
-    "glob": "3.2.11",
     "handlebars": "1.2.1",
-    "kang": "0.0.8",
+    "kang": "1.1.0",
     "less": "1.7.5",
     "mime": "1.2.11",
-    "moray": "git+ssh://git@github.com:joyent/node-moray.git#b84ef0e",
-    "node-uuid": "1.4.3",
+    "moray": "3.2.0",
+    "node-uuid": "1.4.7",
     "promise": "6.1.0",
     "react": "0.13.0",
     "reactify": "1.0.0",
-    "restify": "2.8.5",
-    "sdc-clients": "10.1.0",
-    "ufds": "1.1.3",
-    "sdc-events": "git+ssh://git@github.com:joyent/sdc-events.git",
+    "restify": "4.3.1",
+    "sdc-clients": "10.2.1",
+    "ufds": "1.3.0",
     "superagent": "0.21.0",
     "through": "2.3.4",
     "trace-event": "1.3.0",
     "uglify-js": "2.4.16",
     "underscore": "1.8.3",
     "underscore.string": "3.0.3",
-    "watchify": "0.10.2"
+    "verror": "1.10.0"
   },
   "devDependencies": {
-    "node-dev": "^2.3.0",
-    "jest-cli": "^0.1.15",
     "react-tools": "0.13.0"
   },
-  "jest": {
-    "scriptPreprocessor": "<rootDir>/test/jsx-preprocessor.js",
-    "unmockedModulePathPatterns": [
-      "<rootDir>/node_modules/react",
-      "<rootDir>/node_modules/backbone",
-      "<rootDir>/node_modules/superagent"
-    ],
-    "testPathIgnorePatterns": [
-      "<rootDir>/node_modules"
-    ]
-  },
   "scripts": {
     "test": "jest",
     "start": "node server.js | bunyan"
diff --git a/test/jsx-preprocessor.js b/test/jsx-preprocessor.js
deleted file mode 100644
index ce9840c8..00000000
--- a/test/jsx-preprocessor.js
+++ /dev/null
@@ -1,18 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2014, Joyent, Inc.
- */
-
-// preprocessor.js
-var ReactTools = require('react-tools');
-
-module.exports = {
-    process: function (src) {
-        return ReactTools.transform(src);
-    }
-};
diff --git a/tools/build-js b/tools/build-js
index f192560b..13fb80e6 100755
--- a/tools/build-js
+++ b/tools/build-js
@@ -16,8 +16,7 @@ var LOG = require('bunyan').createLogger({name:'bundler'});
 var async = require('async');
 var fs = require('fs');
 var path = require('path');
-var browserify = require('watchify/node_modules/browserify');
-var watchify = require('watchify');
+var browserify = require('browserify');
 
 var shim = require('browserify-shim');
 var format = require('util').format;
@@ -111,8 +110,13 @@ Object.keys(shimConfig).forEach(function(k) {
 });
 
 var prepAppBundle = function() {
-    var bundler = process.env.WATCH ? watchify : browserify;
-    var bundle = shim(bundler({ extensions: ['.js', '.jsx'] }), shimConfig);
+    var opts = {
+        extensions: ['.js', '.jsx'],
+        debug: true
+    };
+
+    var bundle = shim(browserify(opts), shimConfig);
+
     bundle
         .transform(require.resolve(join(root, './transforms/tpl')))
         .transform(require('reactify'))
@@ -132,8 +136,11 @@ var prepAppBundle = function() {
 };
 
 function prepLibsBundle() {
-    var opts = {noParse: [ require.resolve(join(root, './lib/epoch.0.5.2.min')) ]};
-    var blibs = process.env.WATCH ? watchify(opts) : browserify(opts);
+    var opts = {
+        noParse: [ require.resolve(join(root, './lib/epoch.0.5.2.min')) ]
+    };
+
+    var blibs = browserify(opts);
     blibs.require('react');
     blibs.require('underscore');
     blibs.require('underscore.string');
@@ -203,7 +210,7 @@ function bundle(bundler, out, cb) {
 
     LOG.info('[%s] building', out);
 
-    var bs = bundler.bundle({debug:true});
+    var bs = bundler.bundle();
 
     var bytes, time;
     bundler.on('bytes', function(b) { bytes = b; });
diff --git a/tools/devrun.sh b/tools/devrun.sh
index 3b4ed1b7..cc6e1080 100755
--- a/tools/devrun.sh
+++ b/tools/devrun.sh
@@ -6,12 +6,12 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 port=$(cat `pwd`/etc/config.json | json sslport)
 
-echo "Will rebundle when needed";
-WATCH=true tools/build-js | ./node_modules/bunyan/bin/bunyan &
-echo "***** Starting adminui server on port $port"
-./node_modules/.bin/node-dev server.js NODE=$(which node) | ./node_modules/.bin/bunyan
+echo "***** Bundling client-side assets"
+tools/build-js | ./node_modules/bunyan/bin/bunyan
+echo "***** Starting server on port $port"
+node server.js | ./node_modules/.bin/bunyan
diff --git a/tools/mk/Makefile.smf.targ b/tools/mk/Makefile.smf.targ
index 0fa03ce4..f78de968 100644
--- a/tools/mk/Makefile.smf.targ
+++ b/tools/mk/Makefile.smf.targ
@@ -23,7 +23,7 @@ check-manifests: $(SMF_MANIFESTS:%=%.smfchk)
 %.smfchk: %
 	$(XMLLINT) --path $(dir $(SMF_DTD)) --dtdvalid $(SMF_DTD) $^
 
-check: check-manifests
+check:: check-manifests
 
 $(SMF_MANIFESTS): %: %.in
 	$(SED) -e 's#@@NODE@@#@@PREFIX@@/$(NODE_INSTALL)/bin/node#' $< > $@
diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index ea190008..3f6d54b3 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -209,7 +209,7 @@ check-jsstyle:  $(JSSTYLE_EXEC)
 	$(JSSTYLE) $(JSSTYLE_FLAGS) $(JSSTYLE_FILES)
 
 .PHONY: check
-check: check-jsl check-json $(JSSTYLE_TARGET) check-bash
+check:: check-jsl check-json $(JSSTYLE_TARGET) check-bash
 	@echo check ok
 
 .PHONY: clean
diff --git a/tools/npmfreeze.js b/tools/npmfreeze.js
index abc71288..671a8cbe 100755
--- a/tools/npmfreeze.js
+++ b/tools/npmfreeze.js
@@ -82,13 +82,15 @@ Object.keys(deps).forEach(function(k) {
             process.exit(code);
         }
 
-        var val;
-        eval('val = ' + json);
+        var val = JSON.parse(json);
 
         deps[k] = val['dist-tags'].latest;
         wait--;
     });
 });
 
-return (wait === 0 ? done() : waitForDone());
-
+if (wait === 0) {
+    done();
+} else {
+    waitForDone();
+}
diff --git a/www/js/components/__tests__/user-tile-test.js b/www/js/components/__tests__/user-tile-test.js
deleted file mode 100644
index e0e28a3a..00000000
--- a/www/js/components/__tests__/user-tile-test.js
+++ /dev/null
@@ -1,27 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2014, Joyent, Inc.
- */
-
-/** @jsx: React.DOM */
-
-jest.dontMock('../user-tile.jsx');
-
-describe('UserTile', function() {
-
-    it('renders user-tile', function() {
-        var React = require('react/addons');
-        var TestUtils = React.addons.TestUtils;
-        var UserTile = require('../user-tile.jsx');
-
-        var userTile = new Usertile({uuid: '12b2e590-e0ea-11e3-8b68-0800200c9a66'});
-        TestUtils.renderIntoDocument(userTile);
-        var tile = TestUtils.findRenderedDOMComponentWithTag(userTile, 'div');
-        expect(tile.getDOMNode().class).toEqual('user-tile');
-    });
-});
