commit e8dff3c1fbcf774755e28ac9312eb849f54e116d (refs/changes/55/3355/1)
Author: Mike Zeller <mike.zeller@joyent.com>
Date:   2018-02-12T18:08:01-08:00 (1 year, 8 months ago)
    
    TRITON-132 Fix support for preprovisioning instance NICs

diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 33c09ec..e74eb62 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -1165,7 +1165,7 @@ function preProvisionNics(req, cb) {
             });
 
             if (netNics.length > 0) {
-                nics = nics.concat(netNics);
+                nics.push.apply(nics, netNics);
                 done();
                 return;
             }
