From e8dff3c1fbcf774755e28ac9312eb849f54e116d Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike.zeller@joyent.com>
Date: Mon, 12 Feb 2018 18:08:01 -0800
Subject: [PATCH] TRITON-132 Fix support for preprovisioning instance NICs

---
 lib/endpoints/vms.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 33c09ec..e74eb62 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -1165,7 +1165,7 @@ function preProvisionNics(req, cb) {
             });
 
             if (netNics.length > 0) {
-                nics = nics.concat(netNics);
+                nics.push.apply(nics, netNics);
                 done();
                 return;
             }
-- 
2.21.0

