commit d2317dccc0c5a9c9bbc6d0f45d6204b05e570a44 (refs/changes/73/1573/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-02-24T15:03:54-08:00 (2 years, 7 months ago)
    
    DOCKER-1003 Update tests to account for FWAPI-260

diff --git a/test/integration/cli-run-ports.test.js b/test/integration/cli-run-ports.test.js
index b59fe0a..8b31691 100644
--- a/test/integration/cli-run-ports.test.js
+++ b/test/integration/cli-run-ports.test.js
@@ -192,8 +192,8 @@ test('no port args', function (tt) {
                 tag: 'sdc_docker'
             },
             expected: [
-                'FROM tag sdc_docker TO tag sdc_docker ALLOW tcp PORT all',
-                'FROM tag sdc_docker TO tag sdc_docker ALLOW udp PORT all'
+                'FROM tag "sdc_docker" TO tag "sdc_docker" ALLOW tcp PORT all',
+                'FROM tag "sdc_docker" TO tag "sdc_docker" ALLOW udp PORT all'
             ]
         });
     });
