From a3cd71de3bed3c7870048301968222d6c05108d4 Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Thu, 10 Nov 2016 09:54:45 -0800
Subject: [PATCH] ADMINUI-2354 allow ufds_master credentials to differ from
 local ufds

---
 sapi_manifests/adminui/template | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sapi_manifests/adminui/template b/sapi_manifests/adminui/template
index 991d7b45..b46d43a2 100644
--- a/sapi_manifests/adminui/template
+++ b/sapi_manifests/adminui/template
@@ -9,7 +9,12 @@
 	"ufdsMaster": {
 		"url": "ldaps://{{^ufds_remote_ip}}{{{ufds_domain}}}{{/ufds_remote_ip}}{{{ufds_remote_ip}}}",
 		"bindDN": "{{{ufds_ldap_root_dn}}}",
+		{{^ufds_remote_ldap_root_pw}}
 		"bindPassword": "{{{ufds_ldap_root_pw}}}",
+		{{/ufds_remote_ldap_root_pw}}
+		{{#ufds_remote_ldap_root_pw}}
+		"bindPassword": "{{{ufds_remote_ldap_root_pw}}}",
+		{{/ufds_remote_ldap_root_pw}}
 		"clientTimeout": 240000,
 		"cache": {
 			"size": 5000,
-- 
2.21.0

