commit 135a6284f18e7cdc9ff14db7229d927e0f4e7b2d (refs/changes/53/2553/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-09-12T18:27:20-07:00 (2 years, 1 month ago)
    
    MANTA-3421 binder crashes on EHOSTUNREACH

diff --git a/lib/server.js b/lib/server.js
index 38cac7c..97e9a51 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -351,10 +351,11 @@ function createServer(options) {
         assert.object(options.log, 'options.log');
         assert.optionalObject(options.recursion, 'options.recursion');
         assert.string(options.dnsDomain, 'options.dnsDomain');
+        var log = options.log;
 
         var server = mname.createServer({
                 name: options.name || 'binder',
-                log: options.log
+                log: log
         });
 
         server.on('query', function onQuery(query, cb) {
@@ -451,6 +452,15 @@ function createServer(options) {
                 }, 'DNS query');
         });
 
+        server.on('error', function (err) {
+                if (err.toString().indexOf('EHOSTUNREACH') != -1) {
+                        log.error(err, 'cannot reply to DNS traffic: ' +
+                            'is there asymmetric routing?');
+                        return;
+                }
+                throw (err);
+        });
+
         server.start = function start(callback) {
                 var done = 0;
                 server.listenUdp({
