commit 9ef247a7a440af8a7454b617a16e859b09d8919e (refs/changes/53/2553/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-09-12T18:33:07-07:00 (2 years, 1 month ago)
    
    MANTA-3421 binder crashes on EHOSTUNREACH

diff --git a/deps/manta-scripts b/deps/manta-scripts
index 4b86dd1..cd72e80 160000
--- a/deps/manta-scripts
+++ b/deps/manta-scripts
@@ -1 +1 @@
-Subproject commit 4b86dd132d68cb177f066a83d90b8fdc9ac2cde9
+Subproject commit cd72e801576de5a4b1aec35a7f79409d3f1eebb5
diff --git a/lib/server.js b/lib/server.js
index 38cac7c..eae6ee0 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -351,10 +351,11 @@ function createServer(options) {
         assert.object(options.log, 'options.log');
         assert.optionalObject(options.recursion, 'options.recursion');
         assert.string(options.dnsDomain, 'options.dnsDomain');
+        var log = options.log;
 
         var server = mname.createServer({
                 name: options.name || 'binder',
-                log: options.log
+                log: log
         });
 
         server.on('query', function onQuery(query, cb) {
@@ -451,6 +452,22 @@ function createServer(options) {
                 }, 'DNS query');
         });
 
+        server.on('error', function (err) {
+                /*
+                 * The EHOSTUNREACH error seems to have a numeric errno
+                 * in its .code property instead of a string. We'll check for
+                 * a string instead so we stand a chance of testing this on
+                 * other platforms (instead of depending on the exact errno
+                 * value).
+                 */
+                if (err.toString().indexOf('EHOSTUNREACH') != -1) {
+                        log.error(err, 'cannot reply to DNS traffic: ' +
+                            'is there asymmetric routing?');
+                        return;
+                }
+                throw (err);
+        });
+
         server.start = function start(callback) {
                 var done = 0;
                 server.listenUdp({
