commit 8a162b7a0522cbb6468d991ed416f9af3f879ae7
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2016-09-09T21:47:56+00:00 (3 years, 1 month ago)
    
    AGENT-1035 net-agent needs new rsync-to script

diff --git a/tools/rsync-to b/tools/rsync-to
index d975b95..989fa9e 100755
--- a/tools/rsync-to
+++ b/tools/rsync-to
@@ -28,7 +28,7 @@ if [[ $(uname -s) != "SunOS" ]]; then
 fi
 
 rsync -av ${TOP}/ \
-    $NODE:/var/tmp/net-agent/ \
+    $NODE:/opt/smartdc/agents/lib/node_modules/net-agent/ \
     $extraOpts \
     --exclude .git/ \
     --exclude /deps/ \
@@ -36,4 +36,4 @@ rsync -av ${TOP}/ \
     --exclude /tools/ \
     --exclude /tmp/
 
-ssh $NODE cp -R /opt/smartdc/agents/lib/node_modules/net-agent/node_modules/ /var/tmp/net-agent/
\ No newline at end of file
+ssh $NODE svcadm restart net-agent
