From 057a784dc344f347fe1e1554d36ac6a4699fe4b7 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 9 Aug 2017 14:26:25 -0700
Subject: [PATCH] joyent/node-triton#225 improve 'make test-in-parallel'

---
 Makefile     | 5 +++--
 package.json | 3 ++-
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index 8129a3e..b2fac48 100644
--- a/Makefile
+++ b/Makefile
@@ -37,8 +37,9 @@ test-integration:
 
 .PHONY: test-in-parallel
 test-in-parallel:
-	NODE_NDEBUG= prove -j15 -e ./node_modules/.bin/tape \
-		test/unit/*.test.js test/integration/*.test.js
+	ls test/unit/*.test.js test/integration/*.test.js \
+	    | parallel ./node_modules/.bin/tape \
+	    | ./node_modules/.bin/tap-summary --no-ansi --no-progress
 
 .PHONY: clean
 clean::
diff --git a/package.json b/package.json
index da34b75..3b0c1b5 100644
--- a/package.json
+++ b/package.json
@@ -32,7 +32,8 @@
     "wordwrap": "1.0.0"
   },
   "devDependencies": {
-    "tape": "4.2.0"
+    "tape": "4.2.0",
+    "tap-summary": "3.0.2"
   },
   "main": "./lib",
   "scripts": {
-- 
2.21.0

