commit a4b08ddfee1c35713b26a1ae0685616c4dfe3064 (refs/changes/68/3368/1)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2018-02-14T15:38:00-08:00 (1 year, 8 months ago)
    
    ADMINUI-2398 CN Setup always shows 'Setup Failed' during setup

diff --git a/sapi_manifests/adminui/template b/sapi_manifests/adminui/template
index b46d43a2..74a02725 100644
--- a/sapi_manifests/adminui/template
+++ b/sapi_manifests/adminui/template
@@ -77,5 +77,5 @@
 			}
 		}
 	},
-	"serverSetupJobName": "server-setup-1.0.2"
+	"serverSetupJobName": "server-setup-*"
 }
