From 7642e1d74dba582ecfc08fb456dc68ca6720b895 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 18 Apr 2018 21:47:06 +0000
Subject: [PATCH] NAPI-364 Unable to use VLAN 1 (fix unit tests)

---
 test/unit/networks.test.js | 1 -
 1 file changed, 1 deletion(-)

diff --git a/test/unit/networks.test.js b/test/unit/networks.test.js
index b21b6fd..3c7cb74 100644
--- a/test/unit/networks.test.js
+++ b/test/unit/networks.test.js
@@ -300,7 +300,6 @@ test('Create network - invalid parameters', function (t) {
 
         ['vlan_id', 'a', constants.VLAN_MSG],
         ['vlan_id', '-1', constants.VLAN_MSG],
-        ['vlan_id', '1', constants.VLAN_MSG],
         ['vlan_id', '4095', constants.VLAN_MSG],
 
         ['provision_start_ip', fmt('10.0.%d.254', num - 1), MSG.start_outside],
-- 
2.21.0

