commit 7642e1d74dba582ecfc08fb456dc68ca6720b895 (refs/changes/40/3840/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-04-18T21:47:06+00:00 (1 year, 6 months ago)
    
    NAPI-364 Unable to use VLAN 1 (fix unit tests)

diff --git a/test/unit/networks.test.js b/test/unit/networks.test.js
index b21b6fd..3c7cb74 100644
--- a/test/unit/networks.test.js
+++ b/test/unit/networks.test.js
@@ -300,7 +300,6 @@ test('Create network - invalid parameters', function (t) {
 
         ['vlan_id', 'a', constants.VLAN_MSG],
         ['vlan_id', '-1', constants.VLAN_MSG],
-        ['vlan_id', '1', constants.VLAN_MSG],
         ['vlan_id', '4095', constants.VLAN_MSG],
 
         ['provision_start_ip', fmt('10.0.%d.254', num - 1), MSG.start_outside],
