From d5fabc2c4cfa7f269547d18bf1fa3c66d352f828 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 22 Nov 2018 11:03:03 +0000
Subject: [PATCH] MANTA-4022 propeller's nodeunit dep cannot install with old
 node (fix build)

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 2e659ef..fe9a794 100644
--- a/Makefile
+++ b/Makefile
@@ -32,7 +32,7 @@ NPM             := npm
 # Files
 #
 DOC_FILES        = $(shell find docs -name '*.md' | cut -d '/' -f 2)
-JS_FILES        := $(shell ls *.js) $(shell find lib test bin -name '*.js')
+JS_FILES        := $(shell ls *.js) $(shell find lib bin -name '*.js')
 JSL_CONF_NODE    = tools/jsl.node.conf
 JSL_FILES_NODE   = $(JS_FILES)
 JSSTYLE_FILES    = $(JS_FILES)
-- 
2.21.0

