commit a0f8966ec87873cd0b9b60846cb087c07ff75565 (refs/changes/26/3226/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-01-16T21:08:15-05:00 (1 year, 9 months ago)
    
    TOOLS-1976 sdcadm crashes in ensureDelegateDataset

diff --git a/lib/procedures/shared.js b/lib/procedures/shared.js
index 3d72e81..000b9a6 100644
--- a/lib/procedures/shared.js
+++ b/lib/procedures/shared.js
@@ -623,7 +623,6 @@ function disableVMRegistrar(arg, next) {
 
 function waitUntilVMNotInDNS(arg, next) {
     assert.object(arg, 'arg');
-    assert.object(arg.sdcadm, 'arg.sdcadm');
     assert.func(arg.progress, 'arg.progress');
     assert.object(arg.log, 'arg.log');
     assert.object(arg.cnapi, 'arg.cnapi');
@@ -650,7 +649,6 @@ function waitUntilVMNotInDNS(arg, next) {
 
 function waitUntilVmInDNS(arg, next) {
     assert.object(arg, 'arg');
-    assert.object(arg.sdcadm, 'arg.sdcadm');
     assert.func(arg.progress, 'arg.progress');
     assert.object(arg.log, 'arg.log');
     assert.object(arg.cnapi, 'arg.cnapi');
@@ -962,7 +960,6 @@ function updateVmUserScriptRemote(arg, next) {
 
 function ensureDelegateDataset(arg, next) {
     assert.object(arg, 'arg');
-    assert.object(arg.sdcadm, 'arg.sdcadm');
     assert.func(arg.progress, 'arg.progress');
     assert.object(arg.log, 'arg.log');
     assert.object(arg.cnapi, 'arg.cnapi');
