commit 5bdfbdd52e9edd3856a30dc4e7635e43359bc23a (refs/changes/71/5371/1)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2019-01-16T11:42:56-05:00 (9 months ago)
    
    OS-7516 so_krecv_unblock() double-mutex-exits

diff --git a/usr/src/uts/common/fs/sockfs/sockcommon_subr.c b/usr/src/uts/common/fs/sockfs/sockcommon_subr.c
index 7bdd64393b..90b0b24d9a 100644
--- a/usr/src/uts/common/fs/sockfs/sockcommon_subr.c
+++ b/usr/src/uts/common/fs/sockfs/sockcommon_subr.c
@@ -24,7 +24,7 @@
  */
 /*
  * Copyright 2014, OmniTI Computer Consulting, Inc. All rights reserved.
- * Copyright 2015 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -2517,5 +2517,5 @@ so_krecv_unblock(sonode_t *so)
 
 	so->so_rcv_queued = 0;
 	(void) so_check_flow_control(so);
-	mutex_exit(&so->so_lock);
+	/* so_check_flow_control() always drops so->so_lock, so we're good. */
 }
