commit acb3902dcdb95bc39222e1e48f812fd1e1eb0894 (refs/changes/09/1009/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    instance snapshot {create,delete,get,list}

diff --git a/lib/do_instance/do_snapshot/do_create.js b/lib/do_instance/do_snapshot/do_create.js
index d8aa522..de7b5cf 100644
--- a/lib/do_instance/do_snapshot/do_create.js
+++ b/lib/do_instance/do_snapshot/do_create.js
@@ -47,7 +47,8 @@ function do_create(subcmd, opts, args, cb) {
         createOpts.name = opts.name;
     }
 
-    vasync.pipeline({arg: {}, funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
         function createSnapshot(ctx, next) {
             ctx.start = Date.now();
 
diff --git a/lib/do_instance/do_snapshot/do_delete.js b/lib/do_instance/do_snapshot/do_delete.js
index 99d6c9b..e85237c 100644
--- a/lib/do_instance/do_snapshot/do_delete.js
+++ b/lib/do_instance/do_snapshot/do_delete.js
@@ -61,7 +61,8 @@ function do_delete(subcmd, opts, args, cb) {
         });
     }
 
-    vasync.pipeline({funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
         function confirm(_, next) {
             if (opts.force) {
                 return next();
diff --git a/lib/do_instance/do_snapshot/do_get.js b/lib/do_instance/do_snapshot/do_get.js
index 437a724..7bf02f3 100644
--- a/lib/do_instance/do_snapshot/do_get.js
+++ b/lib/do_instance/do_snapshot/do_get.js
@@ -36,22 +36,27 @@ function do_get(subcmd, opts, args, cb) {
     var name = args[1];
     var cli = this.top;
 
-    cli.tritonapi.getInstanceSnapshot({
-        id: id,
-        name: name
-    }, function onSnapshot(err, snapshot) {
-        if (err) {
-            cb(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-
-        if (opts.json) {
-            console.log(JSON.stringify(snapshot));
-        } else {
-            console.log(JSON.stringify(snapshot, null, 4));
-        }
-
-        cb();
+        cli.tritonapi.getInstanceSnapshot({
+            id: id,
+            name: name
+        }, function onSnapshot(err, snapshot) {
+            if (err) {
+                cb(err);
+                return;
+            }
+
+            if (opts.json) {
+                console.log(JSON.stringify(snapshot));
+            } else {
+                console.log(JSON.stringify(snapshot, null, 4));
+            }
+
+            cb();
+        });
     });
 }
 
diff --git a/lib/do_instance/do_snapshot/do_list.js b/lib/do_instance/do_snapshot/do_list.js
index 622f576..162dfc3 100644
--- a/lib/do_instance/do_snapshot/do_list.js
+++ b/lib/do_instance/do_snapshot/do_list.js
@@ -40,35 +40,40 @@ function do_list(subcmd, opts, args, cb) {
     var cli = this.top;
     var machineId = args[0];
 
-    cli.tritonapi.listInstanceSnapshots({
-        id: machineId
-    }, function onSnapshots(err, snapshots) {
-        if (err) {
-            cb(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-
-        if (opts.json) {
-            common.jsonStream(snapshots);
-        } else {
-            var columns = COLUMNS_DEFAULT;
-
-            if (opts.o) {
-                columns = opts.o;
-            } else if (opts.long) {
-                columns = COLUMNS_DEFAULT;
+        cli.tritonapi.listInstanceSnapshots({
+            id: machineId
+        }, function onSnapshots(err, snapshots) {
+            if (err) {
+                cb(err);
+                return;
             }
 
-            columns = columns.split(',');
-            var sort = opts.s.split(',');
-
-            tabula(snapshots, {
-                skipHeader: opts.H,
-                columns: columns,
-                sort: sort
-            });
-        }
-        cb();
+            if (opts.json) {
+                common.jsonStream(snapshots);
+            } else {
+                var columns = COLUMNS_DEFAULT;
+
+                if (opts.o) {
+                    columns = opts.o;
+                } else if (opts.long) {
+                    columns = COLUMNS_DEFAULT;
+                }
+
+                columns = columns.split(',');
+                var sort = opts.s.split(',');
+
+                tabula(snapshots, {
+                    skipHeader: opts.H,
+                    columns: columns,
+                    sort: sort
+                });
+            }
+            cb();
+        });
     });
 }
 
