commit 310d0a4089ca1b9ce42db6daa33557cf320c1053 (refs/changes/42/1142/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-12-18T00:49:27+00:00 (2 years, 10 months ago)
    
    OS-5856 Want support for XXV710

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 35dfbf9d..2bc1c105 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -497,6 +497,8 @@ i40e "pciex8086,1584"
 i40e "pciex8086,1585"
 i40e "pciex8086,1586"
 i40e "pciex8086,1589"
+i40e "pciex8086,158a"
+i40e "pciex8086,158b"
 i915 "pci8086,2562"
 i915 "pci8086,2572"
 i915 "pci8086,2582"
