From 8e35295a5f7b9a3b1ab1aa11614e21754a20c517 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Tue, 9 Apr 2019 11:25:58 +0100
Subject: [PATCH] TOOLS-2238 bits-upload targets should allow an override

---
 tools/mk/Makefile.targ | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index 159e095..0c3fe0b 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -457,8 +457,11 @@ buildimage: stamp-buildimage-prep release $(AGENTS:%=%-prebuilt)
 
 #
 # Upload the build products from this component's ./bits directory
-# to either manta or a filesystem path.
+# to either manta or a filesystem path. We allow consumers to override
+# this, supplying their own custom bits-upload* targets by setting
+# ENGBLD_BITS_UPLOAD_OVERRIDE.
 #
+ifndef ENGBLD_BITS_UPLOAD_OVERRIDE
 .PHONY: bits-upload
 bits-upload: publish $(BUILDIMAGE_TARG)
 	$(TOP)/deps/eng/tools/bits-upload.sh \
@@ -486,6 +489,7 @@ bits-upload-latest:
 		-d $(ENGBLD_DEST_OUT_PATH)/$(NAME) \
 		-D $(ENGBLD_BITS_DIR) \
 		-n $(NAME)
+endif
 
 #
 # Do an npm install for the dependencies of buildimage itself.
-- 
2.21.0

