From 28178b88ccb18fc15df3403d411a03852e735bc5 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 20 Apr 2017 08:43:51 +0000
Subject: [PATCH] OS-6067 tools/build_illumos fails if run from tcsh

---
 tools/build_illumos | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tools/build_illumos b/tools/build_illumos
index c362cb80..e0ba6ff9 100755
--- a/tools/build_illumos
+++ b/tools/build_illumos
@@ -15,6 +15,7 @@
 #
 
 # Setup for building stuff!
+unset SHELL
 unset CC
 unset CXX
 ROOT=`pwd`
-- 
2.21.0

