commit 28178b88ccb18fc15df3403d411a03852e735bc5 (refs/changes/13/1813/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2017-04-20T08:43:51+00:00 (2 years, 6 months ago)
    
    OS-6067 tools/build_illumos fails if run from tcsh

diff --git a/tools/build_illumos b/tools/build_illumos
index c362cb80..e0ba6ff9 100755
--- a/tools/build_illumos
+++ b/tools/build_illumos
@@ -15,6 +15,7 @@
 #
 
 # Setup for building stuff!
+unset SHELL
 unset CC
 unset CXX
 ROOT=`pwd`
