From af63331ce94131f357028f9ad26fe09d2acee6fa Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Tue, 22 Jan 2019 23:51:29 +0000
Subject: [PATCH] TOOLS-2062 eng.git should be a git submodule of Triton/Manta
 repositories (fix Makefile)

---
 Makefile | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 9935014..9d0672f 100644
--- a/Makefile
+++ b/Makefile
@@ -94,12 +94,12 @@ release: all $(SMF_MANIFESTS)
 
 .PHONY: publish
 publish: release
-	@if [[ -z "$(BITS_DIR)" ]]; then \
-		@echo "error: 'BITS_DIR' must be set for 'publish' target"; \
+	@if [[ -z "$(ENGBLD_BITS_DIR)" ]]; then \
+		echo "error: 'ENGBLD_BITS_DIR' must be set for 'publish' target"; \
 		exit 1; \
 	fi
-	mkdir -p $(BITS_DIR)/registrar
-	cp $(ROOT)/$(RELEASE_TARBALL) $(BITS_DIR)/registrar/$(RELEASE_TARBALL)
+	mkdir -p $(ENGBLD_BITS_DIR)/registrar
+	cp $(ROOT)/$(RELEASE_TARBALL) $(ENGBLD_BITS_DIR)/registrar/$(RELEASE_TARBALL)
 
 
 include ./deps/eng/tools/mk/Makefile.deps
-- 
2.21.0

