commit d7e12c10e7596b0b4a8a9b79e044cb3a5f78d2d3 (refs/changes/58/5158/1)
Author: Rui Loura <rui@joyent.com>
Date:   2018-11-28T20:10:10+00:00 (10 months ago)
    
    TRITON-648 vmapi should report version number in response header

diff --git a/lib/vmapi.js b/lib/vmapi.js
index 3bea4e2..8ff57fc 100644
--- a/lib/vmapi.js
+++ b/lib/vmapi.js
@@ -16,6 +16,7 @@ var assert = require('assert-plus');
 var async = require('async');
 var bunyan = require('bunyan');
 var EffluentLogger = require('effluent-logger');
+var fs = require('fs');
 var once = require('once');
 var restify = require('restify');
 var trace_event = require('trace-event');
@@ -181,7 +182,7 @@ VmapiApp.prototype._initApis = function _initApis(options) {
 
     // Init VmapiApp server
     this.server = restify.createServer({
-        name: 'VMAPI',
+        name: 'VMAPI/' + getVersion(),
         log: log.child({ component: 'api' }, true),
         version: apiVersion,
         serverName: 'SmartDataCenter',
@@ -460,5 +461,12 @@ function formatJSON(req, res, body, callback) {
     callback(null, formattedJson);
 }
 
+function getVersion() {
+    var pkg = fs.readFileSync(__dirname + '/../package.json', 'utf8');
+    var ver = JSON.parse(pkg).version;
+    assert.string(ver, 'version');
+    return ver;
+}
+
 
 module.exports = VmapiApp;
