commit 5735010caaa5ab5400a11b0018e2953ec5e6682b (refs/changes/27/5327/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2019-01-08T16:53:20+00:00 (9 months ago)
    
    OS-7488 vm delete fails after OS-7436

diff --git a/src/vm/node_modules/vminfod/vminfod.js b/src/vm/node_modules/vminfod/vminfod.js
index 92a7fed0..b5131232 100644
--- a/src/vm/node_modules/vminfod/vminfod.js
+++ b/src/vm/node_modules/vminfod/vminfod.js
@@ -2314,17 +2314,19 @@ Vminfod.prototype.handleZpoolEvent = function handleZpoolEvent(obj) {
         func: function handleZfsEventTask(extras, cb) {
             var datasetObj = {};
             var uuid;
-            var zpools = self.zonesZpools();
 
             /*
-             * Check to see if the zpool for the event seen is in the list of
-             * zpools in use by VMs on the system.
+             * If this doesn't look like it could be a zonepath dataset, ensure
+             * that the dataset's zpool is one that is used by any VM.
              */
-            if (!zpools.hasOwnProperty(zpool)) {
-                self.log.debug({zpool: zpool, zpools: Object.keys(zpools)},
-                    'ignoring %s zfs event for zpool: %s', dsname, zpool);
-                cb();
-                return;
+            if (parts.length !== 2) {
+                var zpools = self.zonesZpools();
+                if (!zpools.hasOwnProperty(zpool)) {
+                    self.log.debug({zpool: zpool, zpools: Object.keys(zpools)},
+                        'ignoring %s zfs event for zpool: %s', dsname, zpool);
+                    cb();
+                    return;
+                }
             }
 
             if (parts[1]) {
