From 678d53d5ba3db8b0968a111c8a17381870f0c488 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Wed, 28 Aug 2019 09:50:49 -0700
Subject: [PATCH] joyent/manta-thoth#95 thoth should work inside a job

---
 bin/thoth | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/bin/thoth b/bin/thoth
index 8dea1a6..d31ac7d 100755
--- a/bin/thoth
+++ b/bin/thoth
@@ -1,7 +1,7 @@
 #!/usr/bin/env node
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 var mod_ctype = require('ctype');
@@ -3549,10 +3549,12 @@ var connect = function ()
 	var i, opts = {};
 	var mvars = [ 'MANTA_URL', 'MANTA_USER', 'MANTA_KEY_ID' ];
 
-	for (i = 0; i < mvars.length; i++) {
-		if (!process.env[mvars[i]]) {
-			fatal('expected ' + mvars[i] +
-			    ' environment variable to be set');
+	if (process.env.MANTA_NO_AUTH !== 'true') {
+		for (i = 0; i < mvars.length; i++) {
+			if (!process.env[mvars[i]]) {
+				fatal('expected ' + mvars[i] +
+				    ' environment variable to be set');
+			}
 		}
 	}
 
-- 
2.21.0

