commit 2402a86d10245acc51826b83d8cd6c4f6e1fe22b
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-05-31T09:34:51-06:00 (4 months ago)
    
    MANTA-4307 Make boray and electric-boray proper Manta services

diff --git a/config/services/boray/service.json b/config/services/boray/service.json
new file mode 100644
index 0000000..2160cbc
--- /dev/null
+++ b/config/services/boray/service.json
@@ -0,0 +1,9 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 256
+	},
+	"metadata": {
+		"MORAY_MAX_PG_CONNS": 16
+	}
+}
diff --git a/config/services/boray/service.json.lab b/config/services/boray/service.json.lab
new file mode 100644
index 0000000..86322f7
--- /dev/null
+++ b/config/services/boray/service.json.lab
@@ -0,0 +1,5 @@
+{
+	"params": {
+		"ram": 1024
+	}
+}
diff --git a/config/services/boray/service.json.production b/config/services/boray/service.json.production
new file mode 100644
index 0000000..4504983
--- /dev/null
+++ b/config/services/boray/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 32768,
+                "quota": 100
+	}
+}
diff --git a/config/services/buckets-postgres/service.json b/config/services/buckets-postgres/service.json
new file mode 100644
index 0000000..8a347c3
--- /dev/null
+++ b/config/services/buckets-postgres/service.json
@@ -0,0 +1,10 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 2048,
+		"max_lwps": 256000
+	},
+	"metadata": {
+		"PG_DIR": "/manatee/pg/data"
+	}
+}
diff --git a/config/services/buckets-postgres/service.json.lab b/config/services/buckets-postgres/service.json.lab
new file mode 100644
index 0000000..519cc8a
--- /dev/null
+++ b/config/services/buckets-postgres/service.json.lab
@@ -0,0 +1,6 @@
+{
+	"params": {
+                "quota": 30
+	}
+}
+
diff --git a/config/services/buckets-postgres/service.json.production b/config/services/buckets-postgres/service.json.production
new file mode 100644
index 0000000..d7e80c9
--- /dev/null
+++ b/config/services/buckets-postgres/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 32768,
+                "quota": 4000
+	}
+}
diff --git a/config/services/electric-boray/service.json b/config/services/electric-boray/service.json
new file mode 100644
index 0000000..ce40041
--- /dev/null
+++ b/config/services/electric-boray/service.json
@@ -0,0 +1,7 @@
+{
+	"params": {
+		"networks": [ "manta", "admin" ],
+		"ram": 2048,
+		"mdata_exec_timeout": 3600
+	}
+}
diff --git a/config/services/electric-boray/service.json.lab b/config/services/electric-boray/service.json.lab
new file mode 100644
index 0000000..49a2fab
--- /dev/null
+++ b/config/services/electric-boray/service.json.lab
@@ -0,0 +1,5 @@
+{
+	"params": {
+		"ram": 3072
+	}
+}
diff --git a/config/services/electric-boray/service.json.production b/config/services/electric-boray/service.json.production
new file mode 100644
index 0000000..4d57911
--- /dev/null
+++ b/config/services/electric-boray/service.json.production
@@ -0,0 +1,6 @@
+{
+	"params": {
+		"ram": 16384,
+                "quota": 100
+	}
+}
diff --git a/docs/man/man1/manta-adm.md b/docs/man/man1/manta-adm.md
index 65bced0..5cf829d 100644
--- a/docs/man/man1/manta-adm.md
+++ b/docs/man/man1/manta-adm.md
@@ -64,6 +64,15 @@ Services that are part of the Manta application include:
 **authcache**
   Stores user identity information
 
+**boray**
+  PostgreSQL interface for manta buckets system
+
+**buckets-postgres**
+  PostgreSQL databases used for storing object metadata for the buckets system
+
+**electric-boray**
+  Boray proxy for buckets that handles sharding using consistent hashing
+
 **electric-moray**
   Moray proxy that handles sharding using consistent hashing
 
@@ -695,7 +704,7 @@ See above for information on the `-l`, `-H`, and `-o` options.
 
 `-j, --json`
   Instead of the default text-based output, emit a JSON representation of the
-  summary information reported with the `-s/--summary` command.  This format is 
+  summary information reported with the `-s/--summary` command.  This format is
   suitable for use with `manta-adm update`.  This option cannot be combined with
   `-c/--bycn`, `-a/--all`, `-H/--omit-header`, or `-o/--columns`, and it _must_
   be combined with `-s/--summary`.  (Future versions of this command may support
diff --git a/lib/deploy.js b/lib/deploy.js
index 59bbc83..07c4ec8 100644
--- a/lib/deploy.js
+++ b/lib/deploy.js
@@ -277,7 +277,7 @@ Deployer.prototype.deploy = function (options, svcname, callback)
 				var e = new Error(message);
 				e.message = message;
 				log.error(message);
-				return (cb(e));
+			return (cb(e));
 			}
 
 			self.service = svcs[0];
@@ -973,7 +973,8 @@ function createInstance(self, app, svc, cb) {
 	/*
 	 * Prefix with the shard for things that are shardable...
 	 */
-	if (['postgres', 'moray'].indexOf(self.svcname) !== -1) {
+	if (['postgres', 'moray', 'buckets-postgres',
+	    'boray'].indexOf(self.svcname) !== -1) {
 		params.alias = shard + '.' + params.alias;
 	}
 
@@ -1016,6 +1017,18 @@ function createInstance(self, app, svc, cb) {
 		    shard, service_root);
 	}
 
+	if (self.svcname === 'buckets-postgres') {
+		metadata.SERVICE_NAME = sprintf('%s.boray.%s',
+		    shard, service_root);
+		metadata.MANATEE_SHARD_PATH = sprintf('/manatee/%s',
+		    metadata.SERVICE_NAME);
+	}
+
+	if (self.svcname === 'boray') {
+		metadata.SERVICE_NAME = sprintf('%s.boray.%s',
+		    shard, service_root);
+	}
+
 	if (self.svcname === 'storage') {
 		metadata.SERVICE_NAME = sprintf('stor.%s', service_root);
 	}
diff --git a/lib/services.js b/lib/services.js
index e965e92..36ff6dc 100644
--- a/lib/services.js
+++ b/lib/services.js
@@ -54,7 +54,10 @@ var mSvcNames = [
     'reshard',
     'pgstatsmon',
     'garbage-collector',
-    'prometheus'
+    'prometheus',
+    'buckets-postgres',
+    'boray',
+    'electric-boray'
 ];
 
 /*
@@ -80,7 +83,10 @@ var mSvcConfigs = {
     'pgstatsmon':	{ 'oneach': true,  'probes': true,  'sharded': false },
     'garbage-collector': { 'oneach': true, 'probes': true,  'sharded': false },
     'prometheus':	{ 'oneach': true,  'probes': true,  'sharded': false },
-    'marlin':		{ 'oneach': false, 'probes': false, 'sharded': false }
+    'marlin':		{ 'oneach': false, 'probes': false, 'sharded': false },
+    'buckets-postgres':	{ 'oneach': true,  'probes': true,  'sharded': true  },
+    'boray':		{ 'oneach': true,  'probes': true,  'sharded': true  },
+    'electric-boray':	{ 'oneach': true,  'probes': true,  'sharded': false }
 };
 
 exports.mSvcNames = mSvcNames;
diff --git a/man/man1/manta-adm.1 b/man/man1/manta-adm.1
index 43968fb..732dd6b 100644
--- a/man/man1/manta-adm.1
+++ b/man/man1/manta-adm.1
@@ -60,6 +60,15 @@ Services that are part of the Manta application include:
 \fBauthcache\fP
 Stores user identity information
 .TP
+\fBboray\fP
+PostgreSQL interface for manta buckets system
+.TP
+\fBbuckets\-postgres\fP
+PostgreSQL databases used for storing object metadata for the buckets system
+.TP
+\fBelectric\-boray\fP
+Boray proxy for buckets that handles sharding using consistent hashing
+.TP
 \fBelectric\-moray\fP
 Moray proxy that handles sharding using consistent hashing
 .TP
@@ -103,7 +112,7 @@ PostgreSQL databases used for storing object and job metadata
 Monitoring system for Postgres
 .TP
 \fBprometheus\fP
-Time-series database that aggregates Manta and Triton metrics
+Time\-series database that aggregates Manta and Triton metrics
 .TP
 \fBreshard\fP
 Automated resharding system for Manta
@@ -739,7 +748,7 @@ logically apply).  The count for each group is also reported.  With
 .TP
 \fB\fC\-j, \-\-json\fR
 Instead of the default text\-based output, emit a JSON representation of the
-summary information reported with the \fB\fC\-s/\-\-summary\fR command.  This format is 
+summary information reported with the \fB\fC\-s/\-\-summary\fR command.  This format is
 suitable for use with \fB\fCmanta\-adm update\fR\&.  This option cannot be combined with
 \fB\fC\-c/\-\-bycn\fR, \fB\fC\-a/\-\-all\fR, \fB\fC\-H/\-\-omit\-header\fR, or \fB\fC\-o/\-\-columns\fR, and it \fImust\fP
 be combined with \fB\fC\-s/\-\-summary\fR\&.  (Future versions of this command may support
diff --git a/test/tst.services.js b/test/tst.services.js
index 12da351..7ac8d8d 100644
--- a/test/tst.services.js
+++ b/test/tst.services.js
@@ -41,7 +41,10 @@ var knownServices = [
     'reshard',
     'pgstatsmon',
     'garbage-collector',
-    'prometheus'
+    'prometheus',
+    'buckets-postgres',
+    'boray',
+    'electric-boray'
 ];
 
 function main()
