From 751eea1ed4a1c3b1571e62b8822c63d87863f033 Mon Sep 17 00:00:00 2001
From: mguegan <mguegan@virtua.ch>
Date: Wed, 14 Dec 2016 10:09:30 +0100
Subject: [PATCH] joyent/mdata-client#8 Add initial support for OpenBSD 6.0

---
 Makefile   | 12 ++++++++++++
 plat/bsd.c |  4 +++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index b464961..af0c297 100644
--- a/Makefile
+++ b/Makefile
@@ -86,6 +86,18 @@ PLATFORM_OK = true
 MANSECT = 1
 endif
 
+ifeq ($(UNAME_S),OpenBSD)
+CC = cc
+
+CTFMERGE = /usr/bin/true
+CTFCONVERT = /usr/bin/true
+
+CFILES += plat/bsd.c plat/unix_common.c
+HDRS += plat/unix_common.h
+PLATFORM_OK = true
+MANSECT = 1
+endif
+
 ifeq ($(PLATFORM_OK),false)
 $(error Unknown platform: $(UNAME_S))
 endif
diff --git a/plat/bsd.c b/plat/bsd.c
index dfb3aa3..7d3f8a4 100644
--- a/plat/bsd.c
+++ b/plat/bsd.c
@@ -25,8 +25,10 @@
 #include "dynstr.h"
 #include "plat/unix_common.h"
 
-#ifdef __NetBSD__
+#if defined(__NetBSD__)
 #define	SERIAL_DEVICE	"/dev/tty01"
+#elif defined(__OpenBSD__)
+#define	SERIAL_DEVICE	"/dev/cua01"
 #else
 #define	SERIAL_DEVICE	"/dev/cuau1"
 #endif
-- 
2.21.0

