From c54c17edee4f1156488ebdbd6ea31a3ef2eef44b Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 12 Dec 2017 15:56:42 -0800
Subject: [PATCH] TRITON-27 vm-agent: update eslint to 2.x (fix "make check" to
 update eslint if necessary)

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 993edee..9eafbc2 100644
--- a/Makefile
+++ b/Makefile
@@ -142,7 +142,7 @@ dumpvar:
 
 # eslint ftw
 ESLINT = ./node_modules/.bin/eslint
-$(ESLINT): | $(NPM_EXEC)
+$(ESLINT): package.json | $(NPM_EXEC)
 	$(RUN_NPM_INSTALL)
 
 check:: check-eslint
-- 
2.21.0

