From 1d6270522282718da78c6c790b6c796fec95173e Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 5 Dec 2017 15:20:37 -0800
Subject: [PATCH] joyent/node-sshpk#39 Missing NULL in x509 RSA certs

---
 lib/formats/x509.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/formats/x509.js b/lib/formats/x509.js
index 2396574..cad74b5 100644
--- a/lib/formats/x509.js
+++ b/lib/formats/x509.js
@@ -523,6 +523,8 @@ function writeTBSCert(cert, der) {
 
 	der.startSequence();
 	der.writeOID(SIGN_ALGS[sig.algo]);
+	if (sig.algo.match(/^rsa-/))
+		der.writeNull();
 	der.endSequence();
 
 	cert.issuer.toAsn1(der);
-- 
2.21.0

