commit 1d6270522282718da78c6c790b6c796fec95173e (refs/changes/42/3042/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-12-05T15:20:37-08:00 (1 year, 10 months ago)
    
    joyent/node-sshpk#39 Missing NULL in x509 RSA certs

diff --git a/lib/formats/x509.js b/lib/formats/x509.js
index 2396574..cad74b5 100644
--- a/lib/formats/x509.js
+++ b/lib/formats/x509.js
@@ -523,6 +523,8 @@ function writeTBSCert(cert, der) {
 
 	der.startSequence();
 	der.writeOID(SIGN_ALGS[sig.algo]);
+	if (sig.algo.match(/^rsa-/))
+		der.writeNull();
 	der.endSequence();
 
 	cert.issuer.toAsn1(der);
