From a63425315e18a9b4b28cbf14e03d0c84c6cdf2c9 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 22 Jun 2018 12:20:12 -0700
Subject: [PATCH] MANTA-3783 waferlock should whitelist binder zones

---
 sapi_manifests/waferlock/template | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/sapi_manifests/waferlock/template b/sapi_manifests/waferlock/template
index bff129b..fda573d 100644
--- a/sapi_manifests/waferlock/template
+++ b/sapi_manifests/waferlock/template
@@ -24,6 +24,7 @@
 		"paths": ["/moray"],
 		"domains": [
 			"{{{moray_domain}}}",
+			"{{{binder_domain}}}",
 			{{! Don't use pgstatsmon_domain, might not be set. }}
 			"pgstatsmon.{{{datacenter_name}}}.{{{dns_domain}}}"
 		],
@@ -38,7 +39,8 @@
 			"ops.{{{DOMAIN_NAME}}}",
 			"madtom.{{{DOMAIN_NAME}}}",
 			"pgstatsmon.{{{DOMAIN_NAME}}}",
-			"reshard.{{{DOMAIN_NAME}}}"
+			"reshard.{{{DOMAIN_NAME}}}",
+			"nameservice.{{{DOMAIN_NAME}}}"
 		],
 		"sapi_url": "{{{SAPI_URL}}}",
 		"sapi_services": ["manta/moray"],
-- 
2.21.0

