commit e5329cf8e22b9f0f6aeebf0f750e8afa69875115 (refs/changes/23/2523/6)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-11-09T09:24:27-08:00 (1 year, 11 months ago)
    
    HEAD-2316 Add sdc-volapi script to headnode tools

diff --git a/Makefile b/Makefile
index c1483da8..887fecbf 100644
--- a/Makefile
+++ b/Makefile
@@ -71,7 +71,8 @@ SDC_ZONE_COMMANDS = \
 	sdc-vmapi \
 	sdc-waitforjob \
 	sdc-workflow \
-	updates-imgadm
+	updates-imgadm \
+	sdc-volapi
 
 SDC_ZONE_MAN_LINKS = \
 	$(SDC_ZONE_COMMANDS:%=$(PROTO)/opt/smartdc/man/man1/%.1)
