commit 4f12edbbe9e371716de468adedabfd88ea8838d7 (refs/changes/23/2523/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-09-08T10:45:58-04:00 (2 years, 1 month ago)
    
    HEAD-2316 Add sdc-volapi script to headnode tools

diff --git a/Makefile b/Makefile
index c1483da8..887fecbf 100644
--- a/Makefile
+++ b/Makefile
@@ -71,7 +71,8 @@ SDC_ZONE_COMMANDS = \
 	sdc-vmapi \
 	sdc-waitforjob \
 	sdc-workflow \
-	updates-imgadm
+	updates-imgadm \
+	sdc-volapi
 
 SDC_ZONE_MAN_LINKS = \
 	$(SDC_ZONE_COMMANDS:%=$(PROTO)/opt/smartdc/man/man1/%.1)
