From 4f12edbbe9e371716de468adedabfd88ea8838d7 Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Wed, 4 May 2016 12:21:45 +0200
Subject: [PATCH] HEAD-2316 Add sdc-volapi script to headnode tools

---
 Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index c1483da8..887fecbf 100644
--- a/Makefile
+++ b/Makefile
@@ -71,7 +71,8 @@ SDC_ZONE_COMMANDS = \
 	sdc-vmapi \
 	sdc-waitforjob \
 	sdc-workflow \
-	updates-imgadm
+	updates-imgadm \
+	sdc-volapi
 
 SDC_ZONE_MAN_LINKS = \
 	$(SDC_ZONE_COMMANDS:%=$(PROTO)/opt/smartdc/man/man1/%.1)
-- 
2.21.0

