commit 34a35617abb8f245685cbb529453c0bde1ba8366 (refs/changes/63/463/1)
Author: Jordan Paige Hendricks <jordan.hendricks@joyent.com>
Date:   2016-09-13T22:55:27+00:00 (3 years, 1 month ago)
    
    MANTA-2908 muskie logs misleading about remoteIp and remotePort

diff --git a/lib/audit.js b/lib/audit.js
index 50582d4..36bfa92 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -209,7 +209,13 @@ function auditLogger(options) {
             operation: name,
             billable_operation: op,
             bytesTransferred: req._size,
-            remoteAddress: req.connection._xff,
+            /*
+             * If the request has an X-Forwarded-For header from haproxy,
+             * logicalRemoteAddress will be the logical source IP of the client.
+             * Otherwise, it will be the same as remoteAddress.
+             */
+            logicalRemoteAddress: req.connection._xff,
+            remoteAddress: req.connection.remoteAddress,
             remotePort: req.connection.remotePort,
             req_id: req.id,
             reqHeaderLength: reqHeaderLength,
