From 9ec333e9b6660c115b644ff81efa0a191e689bf2 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 20 Sep 2016 16:47:48 -0700
Subject: [PATCH] joyent/node-cueball#23 Fix for #15 broke "additional"
 processing

---
 lib/resolver.js | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/lib/resolver.js b/lib/resolver.js
index 9dbb9fa..ac845c8 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -805,9 +805,10 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 			var cache = {};
 			var addns = msg.getAdditionals();
 			addns.forEach(function (rr) {
-				if (rr.type !== type) {
+				if (rr.type !== 'A' && rr.type !== 'AAAA') {
 					if (rr.type === 'CNAME' ||
-					    rr.type === 'DNAME')
+					    rr.type === 'DNAME' ||
+					    rr.type === 'OPT')
 						return;
 					self.r_log.warn('got unsupported ' +
 					    'additional rrtype: %s', rr.type);
-- 
2.21.0

