commit a4b14c362472f2e99b87b8c7d156b12fd8dd5799 (refs/changes/11/511/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-09-20T16:48:22-07:00 (3 years, 1 month ago)
    
    joyent/node-cueball#23 Fix for #15 broke "additional" processing

diff --git a/lib/resolver.js b/lib/resolver.js
index 9dbb9fa..68bf42d 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -805,7 +805,7 @@ CueBallDNSResolver.prototype.resolve = function (domain, type, timeout) {
 			var cache = {};
 			var addns = msg.getAdditionals();
 			addns.forEach(function (rr) {
-				if (rr.type !== type) {
+				if (rr.type !== 'A' && rr.type !== 'AAAA') {
 					if (rr.type === 'CNAME' ||
 					    rr.type === 'DNAME')
 						return;
