From 58d4acb4ed7381b220f11836e13ac5e883f41196 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 8 Mar 2017 12:56:28 +0100
Subject: [PATCH] TOOLS-1714 "sdcadm platform install" error handling could be
 improved

---
 lib/platform.js | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/lib/platform.js b/lib/platform.js
index f9eba36..dbbb9a4 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -731,10 +731,12 @@ Platform.prototype.install = function install(opts, callback) {
                         'in channel \'%s\'', opts.image,
                         self.sdcadm.updates.channel || defaultChannel);
             } else {
-                progress('In order not to have to re-download image, ' +
-                         '%s has been left behind.', filepath);
-                progress('After correcting above problem, rerun ' +
-                         '`sdcadm platform install %s`.', filepath);
+                if (filepath) {
+                    progress('In order not to have to re-download image, ' +
+                             '%s has been left behind.', filepath);
+                    progress('After correcting above problem, rerun ' +
+                             '`sdcadm platform install %s`.', filepath);
+                }
             }
             updateHistory(hist, function (_histErr) {
                 callback(err);
-- 
2.21.0

