commit 58d4acb4ed7381b220f11836e13ac5e883f41196 (refs/changes/25/1625/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-03-08T19:24:16+00:00 (2 years, 7 months ago)
    
    TOOLS-1714 "sdcadm platform install" error handling could be improved

diff --git a/lib/platform.js b/lib/platform.js
index f9eba36..dbbb9a4 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -731,10 +731,12 @@ Platform.prototype.install = function install(opts, callback) {
                         'in channel \'%s\'', opts.image,
                         self.sdcadm.updates.channel || defaultChannel);
             } else {
-                progress('In order not to have to re-download image, ' +
-                         '%s has been left behind.', filepath);
-                progress('After correcting above problem, rerun ' +
-                         '`sdcadm platform install %s`.', filepath);
+                if (filepath) {
+                    progress('In order not to have to re-download image, ' +
+                             '%s has been left behind.', filepath);
+                    progress('After correcting above problem, rerun ' +
+                             '`sdcadm platform install %s`.', filepath);
+                }
             }
             updateHistory(hist, function (_histErr) {
                 callback(err);
