commit 249fef658625e98e406d9f6d924078223b3eface (refs/changes/12/1012/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton instance wait 8e65bf9b

diff --git a/lib/do_instance/do_wait.js b/lib/do_instance/do_wait.js
index 529acb2..a472238 100644
--- a/lib/do_instance/do_wait.js
+++ b/lib/do_instance/do_wait.js
@@ -12,6 +12,7 @@
 
 var vasync = require('vasync');
 
+var common = require('../common');
 var distractions = require('../distractions');
 var errors = require('../errors');
 
@@ -34,7 +35,8 @@ function do_wait(subcmd, opts, args, cb) {
     var done = 0;
     var instFromId = {};
 
-    vasync.pipeline({funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
         function getInsts(_, next) {
             vasync.forEachParallel({
                 inputs: ids,
