From ec308873181f90f86aafc06d28296ac6c9bbbd7d Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike@mikezeller.net>
Date: Thu, 20 Jun 2019 19:50:08 +0000
Subject: [PATCH] TOOLS-2278 Exclude deps in rust-illumos-priv

---
 Cargo.toml | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Cargo.toml b/Cargo.toml
index e2a28b5..40a5917 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,5 +12,6 @@ categories = [
     "os",
     "os::unix-apis"
 ]
+exclude = ["deps/*"]
 
 [dependencies]
-- 
2.21.0

