commit 504e09bb99cdbbe99739ebc5241b40e07a8ae333
Author: John Levon <john.levon@joyent.com>
Date:   2019-03-30T13:52:47+00:00 (6 months ago)
    
    OS-7422 iPXE should be provided directly from joyent/ipxe

diff --git a/Makefile b/Makefile
index 6ccfb43..85e9ff5 100644
--- a/Makefile
+++ b/Makefile
@@ -2538,25 +2538,25 @@ clean_manta-deployment:
 
 
 
-#---- sdcboot (boot utilities for sdc-headnode)
+#---- ipxe (ipxe fork for sdc-headnode)
 
-_sdcboot_stamp=$(SDCBOOT_BRANCH)-$(TIMESTAMP)-g$(SDCBOOT_SHA)
-SDCBOOT_BITS=$(BITS_DIR)/sdcboot/sdcboot-$(_sdcboot_stamp).tgz
+_ipxe_stamp=$(IPXE_BRANCH)-$(TIMESTAMP)-g$(IPXE_SHA)
+IPXE_BITS=$(BITS_DIR)/ipxe/ipxe-$(_ipxe_stamp).tgz
 
-.PHONY: sdcboot
-sdcboot: $(SDCBOOT_BITS)
+.PHONY: ipxe
+ipxe: $(IPXE_BITS)
 
-$(SDCBOOT_BITS): build/sdcboot
+$(IPXE_BITS): build/ipxe
 	mkdir -p $(BITS_DIR)
-	(cd build/sdcboot && TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) \
+	(cd build/ipxe && TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) \
 	    gmake pkg release publish)
-	@echo "# Created sdcboot bits (time `date -u +%Y%m%dT%H%M%SZ`):"
-	@ls -l $(SDCBOOT_BITS)
+	@echo "# Created ipxe bits (time `date -u +%Y%m%dT%H%M%SZ`):"
+	@ls -l $(IPXE_BITS)
 	@echo ""
 
-clean_sdcboot:
-	$(RM) -rf $(BITS_DIR)/sdcboot
-	(cd build/sdcboot && gmake clean)
+clean_ipxe:
+	$(RM) -rf $(BITS_DIR)/ipxe
+	(cd build/ipxe && gmake clean)
 
 #---- firmware-tools (Legacy-mode FDUM facilities and firmware for Joyent HW)
 
diff --git a/targets.json.in b/targets.json.in
index 388e73b..e5c4c0f 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -44,6 +44,7 @@ function build_headnode_target()
       ${firmware}
       "fwapi",
       "imgapi",
+      "ipxe",
       "sdc-manatee",
       "mahi",
       "manta-deployment",
@@ -55,7 +56,6 @@ function build_headnode_target()
       "sapi",
       "sdc",
       "sdcadm",
-      "sdcboot",
       "ufds",
       "vmapi",
       "workflow"
@@ -606,6 +606,15 @@ cat <<EOF
     ]
   },
 
+  "ipxe": {
+    "build_platform": "20151126T062538Z",
+    "repos": [
+        {"url": "$MG_GIT_PREFIX/ipxe.git"}
+    ],
+    "public": true,
+    "deps": []
+  },
+
   "sdc": {
     "appliance": "true",
     "build_platform": "20151126T062538Z",
@@ -1451,15 +1460,6 @@ cat <<EOF
     ]
   },
 
-  "sdcboot": {
-    "build_platform": "20151126T062538Z",
-    "repos": [
-        {"url": "$MG_GIT_PREFIX/sdcboot.git"}
-    ],
-    "public": true,
-    "deps": []
-  },
-
   ${FIRMWARE_TOOLS_ENTRY}
   ${HEADNODE_TARGETS}
 
diff --git a/tools/purge-mg-builds b/tools/purge-mg-builds
index 8fcf5d5..9298bd6 100755
--- a/tools/purge-mg-builds
+++ b/tools/purge-mg-builds
@@ -112,6 +112,7 @@ TTL_DAYS_FROM_NAME='{
     "hagfish-watcher": 365,
     "heartbeater": 365,
     "imgapi": 365,
+    "ipxe": 365,
     "mackerel": 365,
     "madtom": 365,
     "mahi": 365,
@@ -145,7 +146,6 @@ TTL_DAYS_FROM_NAME='{
     "sdc-system-tests": 365,
     "sdc-zookeeper": 365,
     "sdcadm": 365,
-    "sdcboot": 365,
     "smartlogin": 365,
     "smartos": 365,
     "ufds": 365,
