commit 667ea1e7c971e38cfbc7d274fe3636c8c0b8824b (refs/changes/11/5311/2)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-01-04T18:17:10+00:00 (9 months ago)
    
    OS-7480 xhci port capabilities warning is confusing

diff --git a/usr/src/uts/common/io/usb/hcd/xhci/xhci.c b/usr/src/uts/common/io/usb/hcd/xhci/xhci.c
index eabdd4fd9e..d404ea1d18 100644
--- a/usr/src/uts/common/io/usb/hcd/xhci/xhci.c
+++ b/usr/src/uts/common/io/usb/hcd/xhci/xhci.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -1485,7 +1485,7 @@ xhci_port_count(xhci_t *xhcip)
 			fusb2 = first;
 		} else {
 			xhci_error(xhcip, "encountered port capabilities with "
-			    "unknown major USB version: %d\n", rvers);
+			    "unknown USB version: %x.%x\n", maj, min);
 		}
 	}
 
