From 890801c7c50fb82427a4166505f364d0e3314d2d Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Mon, 22 Oct 2018 15:20:14 -0400
Subject: [PATCH] TRITON-895 want more context around vmadm failures during
 smartdc/init

---
 scripts/headnode.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/scripts/headnode.sh b/scripts/headnode.sh
index 08c1140f..a5431b1d 100755
--- a/scripts/headnode.sh
+++ b/scripts/headnode.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2017 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 #
 
 #
@@ -530,7 +530,7 @@ function create_zone {
         > ${payload_file}
     fi
 
-    cat ${payload_file} | vmadm create
+    VMADM_DEBUG_LEVEL='debug' vmadm create -f "$payload_file"
 
     local loops=
     local zonepath=
-- 
2.21.0

