commit 4f21481a9e0829d8dfce5ef11cef80740f2b5c5d (refs/changes/28/1728/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-03-30T10:34:16-07:00 (2 years, 6 months ago)
    
    MON-354 test-amon is failing in nightly

diff --git a/test/maintenances.test.js b/test/maintenances.test.js
index a8b999f..3b4af85 100644
--- a/test/maintenances.test.js
+++ b/test/maintenances.test.js
@@ -435,9 +435,10 @@ test('maint 2: create maint window', function (t) {
     });
 });
 
-test('maint 2: stop amontestzone', {timeout: 60000}, function (t) {
+test('maint 2: stop amontestzone', function (t) {
     notifications = []; // reset
-    common.vmStop({uuid: prep.amontestzone.uuid, timeout: 60000},
+    // I've seen this VM stop on nightly-2 take >1 min.
+    common.vmStop({uuid: prep.amontestzone.uuid, timeout: 3 * 60 * 1000},
         function (err) {
             t.ifError(err, 'stopped amontestzone');
             t.end();
diff --git a/test/prep.js b/test/prep.js
index 5fce639..516f7f4 100644
--- a/test/prep.js
+++ b/test/prep.js
@@ -346,7 +346,7 @@ function createAmontestzone(next) {
                 common.waitForVmapiJob({
                         vmapiClient: vmapiClient,
                         jobInfo: jobInfo,
-                        timeout: 2 * 60 * 1000 /* 2 minutes */
+                        timeout: 5 * 60 * 1000 /* 5 minutes */
                     }, function (err3) {
                         if (err3)
                             return next(err3);
diff --git a/tools/rsync-test-to b/tools/rsync-test-to
new file mode 100755
index 0000000..c6c0139
--- /dev/null
+++ b/tools/rsync-test-to
@@ -0,0 +1,38 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright 2017, Joyent, Inc.
+#
+
+#
+# Rsync the *test suite* files in this working copy to the install on the QA
+# "nightly" staging setup (i.e. the small DC setup from last night's build of
+# SDC master).
+#
+
+#set -o xtrace
+set -o errexit
+
+TOP=$(cd $(dirname $0)/../; pwd)
+NODE=$1
+
+extraOpts=
+if [[ $(uname -s) != "SunOS" ]]; then
+    extraOpts="--exclude *.node --exclude build"
+fi
+
+rsync -av ${TOP}/test/ \
+    $NODE:/opt/smartdc/agents/lib/node_modules/amon-relay/test/ \
+    $extraOpts \
+    --exclude "*.log" \
+    --exclude "config.json" \
+    --exclude "prep.json" \
+    --exclude tmp/
+
+rsync -av ${TOP}/plugins/test/ \
+    $NODE:/opt/smartdc/agents/lib/node_modules/amon-relay/node_modules/amon-plugins/test/
