commit 4443d985e2d8bc5a119de37b0a22a47780ed62ba
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-08-15T06:40:20-06:00 (8 weeks ago)
    
    MANTA-4288 Update moray to use node-fast 2.7.0

diff --git a/lib/schema.js b/lib/schema.js
index 470bbbb..894d056 100644
--- a/lib/schema.js
+++ b/lib/schema.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -378,6 +378,7 @@ AJV_ENV.addSchema({
     properties: {
         'log': { allOf: [ { '$ref': 'object' } ] },
         'bunyan': { allOf: [ { '$ref': 'object' } ] },
+        'fast': { allOf: [ { '$ref': 'object' } ] },
         'name': { allOf: [ { '$ref': 'string' } ] },
         'file': { allOf: [ { '$ref': 'string' } ] },
         'bindip': { allOf: [ { '$ref': 'string' } ] },
diff --git a/lib/server.js b/lib/server.js
index 79bc5d2..69f6489 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 var util = require('util');
@@ -145,11 +145,20 @@ function MorayServer(options) {
         objectCache: this.ms_objectCache
     };
 
+    var crc_mode = options.fast.crc_mode ||
+        mod_fast.FAST_CHECKSUM_V1;
+
+    collector.gauge({
+        name: 'crc_mode',
+        help: 'The node-fast CRC compatibilty mode of the Fast server'
+    }).set(crc_mode);
+
     var socket = mod_net.createServer({ 'allowHalfOpen': true });
     var server = new mod_fast.FastServer({
         log: log.child({ component: 'fast' }),
         collector: collector,
-        server: socket
+        server: socket,
+        crc_mode: crc_mode
     });
 
     var methods = [
@@ -300,6 +309,10 @@ MorayServer.prototype.close = function close() {
 function createServer(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
+    assert.optionalObject(options.fast, 'options.fast');
+    options.fast = options.fast || {};
+    assert.optionalNumber(options.fast.crc_mode,
+        'options.fast.crc_mode');
 
     return new MorayServer(options);
 }
diff --git a/package.json b/package.json
index 4925370..7c96805 100644
--- a/package.json
+++ b/package.json
@@ -17,7 +17,7 @@
         "dtrace-provider": "~0.8",
         "deep-equal": "0.0.0",
         "exeunt": "^1.1.1",
-        "fast": "2.3.2",
+        "fast": "2.7.0",
         "forkexec": "^1.1.0",
         "ip6addr": "0.2.2",
         "jsprim": "1.4.0",
@@ -27,7 +27,7 @@
         "libuuid": "0.2.1",
         "lru-cache": "2.5.0",
         "node-manatee": "git+https://github.com/joyent/node-manatee.git#262828a8",
-        "moray": "3.1.1",
+        "moray": "git+https://github.com/joyent/node-moray.git#MANTA-4287",
         "microtime": "2.1.6",
         "once": "1.3.0",
         "pg": "6.2.4",
diff --git a/sapi_manifests/moray/template b/sapi_manifests/moray/template
index 961538f..fae4fc5 100644
--- a/sapi_manifests/moray/template
+++ b/sapi_manifests/moray/template
@@ -6,6 +6,9 @@
       "type": "udp"
     }
   },
+  "fast": {
+    "crc_mode": 2
+  },
   "manatee": {
     "manatee": {
       "path": "/manatee/{{SERVICE_NAME}}",
