commit 7f6c4f6146477f619634d032330fadfa2e848ad6 (refs/changes/76/3376/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-02-15T14:48:25-08:00 (1 year, 8 months ago)
    
    TRITON-138 config-agent should not use setInterval

diff --git a/CHANGES.md b/CHANGES.md
index 0cd4387..080328c 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,11 @@
 # sdc-config-agent changelog
 
+## 1.6.1
+
+- [TRITON-138] Avoid setInterval in polling in case the process (e.g. calling
+  SAPI GetConfig) takes longer than the poll interval. This can result in
+  uselessly piling on SAPI.
+
 ## 1.6.0
 
 - [AGENT-1086] Support network per rack Manta deployments by allowing 
diff --git a/agent.js b/agent.js
index 713d0f3..a5b76bc 100644
--- a/agent.js
+++ b/agent.js
@@ -5,11 +5,11 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
- * SmartDataCenter config agent
+ * Triton config agent
  *
  * This agent periodically gathers config information for this instance
  * from SAPI, renders file templates (in "sapi_templates/..." dirs) and, if
@@ -20,6 +20,7 @@ var assert = require('assert-plus');
 var async = require('async');
 var fs = require('fs');
 var optimist = require('optimist');
+var uuidv4 = require('uuid/v4');
 var util = require('./lib/common/util');
 
 var Agent = require('./lib/agent/agent');
@@ -95,6 +96,33 @@ var zonename;
 // TODO(refactor): pass autoMetadata as an opt to `new Agent`.
 var autoMetadata = config.autoMetadata = {};
 
+
+/*
+ * This is the normal operating mode of config-agent: periodically checking
+ * for config updates from SAPI and refreshing manifests/config files.
+ */
+function startPeriodicRefresh() {
+	var checkOnce = function () {
+		var startTime = Date.now();
+		var reqId = uuidv4();
+		log.trace({req_id: reqId}, 'start periodic check');
+
+		agent.checkAndRefresh(function doneCheck(err) {
+			log.debug(
+				{
+					err: err,
+					req_id: reqId,
+					elapsedMs: Date.now() - startTime
+				},
+				'done periodic check');
+
+			setTimeout(checkOnce, config.pollInterval);
+		});
+	};
+
+	setTimeout(checkOnce, config.pollInterval);
+}
+
 async.waterfall([
 	// TODO(refactor) move this to Agent.init
 	function gatherInsts(cb) {
@@ -237,8 +265,7 @@ async.waterfall([
 				cb(err);
 			});
 		} else {
-			setInterval(agent.checkAndRefresh.bind(agent),
-				config.pollInterval);
+			startPeriodicRefresh();
 			cb(null);
 		}
 	}
diff --git a/package.json b/package.json
index 4639ca8..c9856d2 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "config-agent",
     "description": "SmartDataCenter Config Agent",
-    "version": "1.5.0",
+    "version": "1.6.1",
     "author": "Joyent (joyent.com)",
     "dependencies": {
         "assert-plus": "0.1.5",
@@ -16,6 +16,7 @@
         "restify": "2.8.5",
         "sdc-clients": "9.0.3",
         "sprintf-js": "1.0.3",
+        "uuid": "3.2.1",
         "vasync": "1.6.3",
         "verror": "1.6.0"
     },
