commit 89ca6b21954f6bab61d711348862d196a68a1f22 (refs/changes/03/1203/1)
Author: Dillon Amburgey <dillona@dillona.com>
Date:   2017-01-03T20:17:52-05:00 (2 years, 9 months ago)
    
    joyent/triton#223 External network shouldn't have QEMU DHCP service

diff --git a/tools/coal-linux-kvm-setup b/tools/coal-linux-kvm-setup
index 350afc7..7aeac8b 100755
--- a/tools/coal-linux-kvm-setup
+++ b/tools/coal-linux-kvm-setup
@@ -118,8 +118,6 @@ ADMIN_HOST_IP=$(echo ${ADMIN_NETWORK} | sed 's/\.[0-9]*$/.254/')
 EXTERNAL_NAME=sdc-external
 EXTERNAL_NETWORK=10.88.88.0
 EXTERNAL_NETMASK=255.255.255.0
-EXTERNAL_DHCP_START=10.88.88.10
-EXTERNAL_DHCP_END=10.88.88.254
 EXTERNAL_HOST_IP=10.88.88.2
 
 # Checks if a network exists and is active. Destroys and undefines it as needed.
@@ -180,11 +178,7 @@ cat > ${VIRSH_NET_EXTERNAL_XML} << EOF
      <name>${EXTERNAL_NAME}</name>
      <bridge name="sdc-ext" />
      <forward mode="nat" />
-     <ip address="${EXTERNAL_HOST_IP}" netmask="${EXTERNAL_NETMASK}">
-          <dhcp>
-              <range start="${EXTERNAL_DHCP_START}" end="${EXTERNAL_DHCP_END}" />
-          </dhcp>
-     </ip>
+     <ip address="${EXTERNAL_HOST_IP}" netmask="${EXTERNAL_NETMASK}" />
 </network>
 EOF
 
