commit dd8c649210773c2bb117c0f310bf9c21230f35b9 (refs/changes/74/1974/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-05-18T14:53:56-07:00 (2 years, 5 months ago)
    
    TOOLS-1781 sdcadm: improve 'sdcadm help completion' output

diff --git a/lib/cli/do_completion.js b/lib/cli/do_completion.js
index 94e2674..42becd5 100644
--- a/lib/cli/do_completion.js
+++ b/lib/cli/do_completion.js
@@ -30,14 +30,11 @@ do_completion.options = [
 do_completion.help = [
     'Output bash completion code for the `sdcadm` CLI.',
     '',
-    'Installation:',
-    '    sdcadm completion >> ~/.bashrc',
+    'By default, sdcadm installation should setup for Bash completion.',
+    'However, you can update the completions as follows:',
     '',
-    'Or maybe:',
-    '    sdcadm completion > /usr/local/etc/bash_completion.d/sdcadm',
-    '',
-    'Note that sdcadm experimental, platform and post-setup sub-commands',
-    'are not yet included.'
+    '    sdcadm completion >/opt/smartdc/sdcadm/etc/sdcadm.completion \\',
+    '       && source /opt/smartdc/sdcadm/etc/sdcadm.completion'
 ].join('\n');
 do_completion.hidden = true;
 
diff --git a/tools/rsync-to b/tools/rsync-to
index c9aafcc..f89fb59 100755
--- a/tools/rsync-to
+++ b/tools/rsync-to
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -29,7 +29,11 @@ else
     ssh $NODE rm -rf $BASEDIR/node_modules
 fi
 
-for name in bin etc smf lib node_modules test package.json; do
+toSync="bin etc smf lib test package.json"
+if [[ -d ${TOP}/node_modules ]]; then
+    toSync="$toSync node_modules"
+fi
+for name in $toSync; do
     if [[ -d ${TOP}/$name ]]; then
         rsync -av ${TOP}/$name/ $NODE:$BASEDIR/$name/ $extraOpts
     else
