From fb28f83adf55094d3945b1e8e78a2471be15dab1 Mon Sep 17 00:00:00 2001
From: Matt Smillie <matt.smillie@joyent.com>
Date: Fri, 29 Jul 2016 11:05:26 -0700
Subject: [PATCH] DOCKER-899: Fix docker failures when fabrics are not enabled

---
 lib/backends/sdc/containers.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 1d205a2..596b776 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -617,7 +617,7 @@ function externalNetworkByName(opts, payload, callback) {
             return;
         }
 
-        payload.networks = [ { uuid: networks[0], primary: true} ];
+        payload.networks = [ { uuid: networks[0].uuid, primary: true} ];
         callback();
         return;
     });
-- 
2.21.0

