commit fb28f83adf55094d3945b1e8e78a2471be15dab1 (refs/changes/69/169/1)
Author: Matt Smillie <matt.smillie@joyent.com>
Date:   2016-07-29T11:05:26-07:00 (3 years, 2 months ago)
    
    DOCKER-899: Fix docker failures when fabrics are not enabled

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 1d205a2..596b776 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -617,7 +617,7 @@ function externalNetworkByName(opts, payload, callback) {
             return;
         }
 
-        payload.networks = [ { uuid: networks[0], primary: true} ];
+        payload.networks = [ { uuid: networks[0].uuid, primary: true} ];
         callback();
         return;
     });
