commit 60add02fff475a79ce582dad6d3cf2604af8db3f (refs/changes/02/4602/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-07-30T16:37:03+00:00 (1 year, 2 months ago)
    
    joyent/gerritbot#5 disable check for repository "joyent/postgres"

diff --git a/server.js b/server.js
index c3a8003..86f2bc9 100644
--- a/server.js
+++ b/server.js
@@ -36,6 +36,7 @@ repoHasMakeCheck['joyent/illumos-extra'] = false;
 repoHasMakeCheck['joyent/illumos-joyent'] = false;
 repoHasMakeCheck['joyent/zfs_snapshot_tar'] = false;
 repoHasMakeCheck['joyent/illumos-kvm'] = false;
+repoHasMakeCheck['joyent/postgres'] = false;
 
 var dockerKeyPem = mod_fs.readFileSync(config.docker.keyFile);
 var dockerKey = mod_sshpk.parsePrivateKey(dockerKeyPem);
