From 1f47054f5b8559fff85442dbf68153e3101a8644 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Fri, 4 Aug 2017 15:58:03 -0700
Subject: [PATCH] DOCKER-1086 Docker build copy and no trailing slash after a
 cached command

---
 lib/build.js | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/lib/build.js b/lib/build.js
index 2d39412..9cabd01 100644
--- a/lib/build.js
+++ b/lib/build.js
@@ -693,7 +693,27 @@ Builder.prototype.cmdAdd = function cmdAdd(cmd, callback) {
     // Note: copyInfos are populated in the cmdAddPreFn.
     assert.arrayOfObject(cmd.ctx.copyInfos, 'cmd.ctx.copyInfos');
 
-    this.performCopy(cmd, cmd.ctx.copyInfos, callback);
+    var builder = this;
+
+    // DOCKER-935
+    // When the previous cmd was cached, update the copyInfo, as the previous
+    // copyInfo was generated when there was no real file content (as the
+    // zone had not been reprovisioned yet), and as such the copyInfo.destPath
+    // could be incorrect.
+    var lastLayer = builder.layers[builder.layers.length-2];
+    if (lastLayer && lastLayer.cmd === null) {
+        builder.log.info('cmdAdd: update copyInfo since last layer was cached');
+        builder.cmdAddPreFn(cmd, function (err) {
+            if (err) {
+                callback(err);
+                return;
+            }
+            builder.performCopy(cmd, cmd.ctx.copyInfos, callback);
+        });
+        return;
+    }
+
+    builder.performCopy(cmd, cmd.ctx.copyInfos, callback);
 };
 
 Builder.prototype.addArgEntry = function addArgEntry(name, value) {
-- 
2.21.0

