From 41056707276fa310d4bff8ac166150863af25d0b Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 14 Jun 2018 17:22:02 -0700
Subject: [PATCH] MANTA-3719 Need Firewall Rules for Postgres

---
 boot/setup.sh | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index a1b3ec1..e191aa6 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -34,6 +34,7 @@ function manta_manatee_setup {
 
     echo "Adding local manifest directories"
     manta_add_manifest_dir "/opt/smartdc/manatee"
+    manta_add_manifest_dir "/opt/smartdc/waferlock"
 
     # MANTA-1360 no args to manta_common_setup so we don't rotate the 'manatee'
     # entry which stomps over the other manatee logs
@@ -84,6 +85,9 @@ function common_enable_services {
     #
     echo "Starting prefaulter"
     svccfg import /opt/smartdc/manatee/smf/manifests/pg_prefaulter.xml
+
+    echo "Starting waferlock"
+    svccfg import /opt/smartdc/waferlock/smf/manifests/waferlock.xml
 }
 
 function common_manatee_setup {
@@ -193,6 +197,7 @@ function manta_setup_manatee_env {
     manta_add_logadm_entry "manatee-sitter"
     manta_add_logadm_entry "manatee-backupserver"
     manta_add_logadm_entry "manatee-snapshotter"
+    manta_add_logadm_entry "waferlock"
     manta_add_logadm_entry "postgresql" "/var/pg"
     manta_add_logadm_entry "pgdump" "/var/log/manatee"
 }
-- 
2.21.0

