commit 83ea204acd4e1d5986d560c7b3bf2bbe9c795c45 (refs/changes/39/4439/2)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-07-12T19:38:00+00:00 (1 year, 3 months ago)
    
    OS-7054 bhyve should halt on triple fault

diff --git a/usr/src/uts/i86pc/io/vmm/vmm.c b/usr/src/uts/i86pc/io/vmm/vmm.c
index 0f722facb0..a4b669ab81 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm.c
@@ -2324,9 +2324,7 @@ nested_fault(struct vm *vm, int vcpuid, uint64_t info1, uint64_t info2,
 	if (type1 == VM_INTINFO_HWEXCEPTION && vector1 == IDT_DF) {
 		VCPU_CTR2(vm, vcpuid, "triple fault: info1(%#lx), info2(%#lx)",
 		    info1, info2);
-#ifdef	__FreeBSD__
 		vm_suspend(vm, VM_SUSPEND_TRIPLEFAULT);
-#endif
 		*retinfo = 0;
 		return (0);
 	}
