commit 764eb2b9881f462b43467380182f4250da1930fb (refs/changes/01/3201/3)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-01-29T17:18:07+13:00 (1 year, 8 months ago)
    
    TRITON-19 Triton equivalent to AWS' termination protection

diff --git a/lib/index.js b/lib/index.js
index 7be9f55..18ea188 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -42,6 +42,7 @@ var typeFromKey = {
     'triton.cns.disable': 'boolean',
     'triton.cns.reverse_ptr': 'string',
     'triton.network.public': 'string',
+    'triton.instance.undeletable': 'boolean',
 
     // Internal testing Triton tags
     'triton._test.string': 'string',
@@ -144,6 +145,12 @@ var validatorFromKey = {
         return null;
     },
 
+    'triton.instance.undeletable':
+    function validateTritonInstanceUndeletable(val) {
+        assert.bool(val, 'val');
+        return null;
+    },
+
     /*
      * Triton tags used for internal testing.
      */
diff --git a/test/unit.test.js b/test/unit.test.js
index c66aa65..2dc3fd0 100644
--- a/test/unit.test.js
+++ b/test/unit.test.js
@@ -181,9 +181,28 @@ var cases = [
         key: 'triton.cns.services',
         str: '5foobar',
         val: '5foobar'
-    }
+    },
 
     // TODO: triton.cns.reverse_ptr
+
+    {
+        key: 'triton.instance.undeletable',
+        str: 'true',
+        val: true
+    },
+    {
+        key: 'triton.instance.undeletable',
+        str: 'false',
+        val: false
+    },
+    {
+        key: 'triton.instance.undeletable',
+        str: 'booga',
+        /* JSSTYLED */
+        err: /Triton tag "triton.instance.undeletable" value must be "true" or "false": "booga"/,
+        /* JSSTYLED */
+        errmsg: /Triton tag "triton.instance.undeletable" value must be a boolean: "booga"/
+    }
 ];
 
 
