commit 45e436996c1bad20b5ddb9120a0457e47f8a798f (refs/changes/01/3201/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-12-06T15:33:41+13:00 (1 year, 10 months ago)
    
    TRITON-19: Triton equivalent to AWS' termination protection

diff --git a/lib/index.js b/lib/index.js
index 7be9f55..cda450c 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -42,6 +42,7 @@ var typeFromKey = {
     'triton.cns.disable': 'boolean',
     'triton.cns.reverse_ptr': 'string',
     'triton.network.public': 'string',
+    'triton.instance.protection': 'boolean',
 
     // Internal testing Triton tags
     'triton._test.string': 'string',
@@ -144,6 +145,12 @@ var validatorFromKey = {
         return null;
     },
 
+    'triton.instance.protection':
+    function validateTritonInstanceProtection(val) {
+        assert.bool(val, 'val');
+        return null;
+    },
+
     /*
      * Triton tags used for internal testing.
      */
diff --git a/test/unit.test.js b/test/unit.test.js
index c66aa65..49053f6 100644
--- a/test/unit.test.js
+++ b/test/unit.test.js
@@ -181,9 +181,28 @@ var cases = [
         key: 'triton.cns.services',
         str: '5foobar',
         val: '5foobar'
-    }
+    },
 
     // TODO: triton.cns.reverse_ptr
+
+    {
+        key: 'triton.instance.protection',
+        str: 'true',
+        val: true
+    },
+    {
+        key: 'triton.instance.protection',
+        str: 'false',
+        val: false
+    },
+    {
+        key: 'triton.instance.protection',
+        str: 'booga',
+        /* JSSTYLED */
+        err: /Triton tag "triton.instance.protection" value must be "true" or "false": "booga"/,
+        /* JSSTYLED */
+        errmsg: /Triton tag "triton.instance.protection" value must be a boolean: "booga"/
+    }
 ];
 
 
