commit 1b0348831ce3bf49c14e3007c978acf131a8e13e (refs/changes/81/3681/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-03-19T14:53:05+00:00 (1 year, 7 months ago)
    
    OS-6734 need LIST_FOREACH_SAFE

diff --git a/usr/src/uts/common/sys/queue.h b/usr/src/uts/common/sys/queue.h
index 39a476450f..87de5c27b5 100644
--- a/usr/src/uts/common/sys/queue.h
+++ b/usr/src/uts/common/sys/queue.h
@@ -179,6 +179,11 @@ struct {								\
 		(var);							\
 		(var) = ((var)->field.le_next))
 
+#define	LIST_FOREACH_SAFE(var, head, field, tvar)			\
+	for ((var) = LIST_FIRST((head));				\
+	    (var) && ((tvar) = LIST_NEXT((var), field), 1);		\
+	    (var) = (tvar))
+
 /*
  * List access methods.
  */
