From 1b0348831ce3bf49c14e3007c978acf131a8e13e Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Fri, 16 Mar 2018 21:49:23 +0000
Subject: [PATCH] OS-6734 need LIST_FOREACH_SAFE

---
 usr/src/uts/common/sys/queue.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/usr/src/uts/common/sys/queue.h b/usr/src/uts/common/sys/queue.h
index 39a476450f..87de5c27b5 100644
--- a/usr/src/uts/common/sys/queue.h
+++ b/usr/src/uts/common/sys/queue.h
@@ -179,6 +179,11 @@ struct {								\
 		(var);							\
 		(var) = ((var)->field.le_next))
 
+#define	LIST_FOREACH_SAFE(var, head, field, tvar)			\
+	for ((var) = LIST_FIRST((head));				\
+	    (var) && ((tvar) = LIST_NEXT((var), field), 1);		\
+	    (var) = (tvar))
+
 /*
  * List access methods.
  */
-- 
2.21.0

