From e43f5e6d7ae1d81ec4fa96596f5739c698b70ab2 Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Wed, 31 Aug 2016 08:04:03 +0200
Subject: [PATCH] TOOLS-1538 sdcadm should not throw ENOENT when checking for
 sdc-usbkey

---
 lib/common.js | 42 +++++++++++++++++++++++++++++++++---------
 1 file changed, 33 insertions(+), 9 deletions(-)

diff --git a/lib/common.js b/lib/common.js
index a565ceb..8c38fd7 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -1295,12 +1295,17 @@ function isUsbKeyMounted(log, cb) {
     function sdcUsbkeyExists(ctx, next) {
         fs.stat(ctx.utility, function (er, st) {
             if (er) {
-                return next(er);
+                if (er.code === 'ENOENT') {
+                    next();
+                } else {
+                    next(er);
+                }
+                return;
             }
             if (st.isFile()) {
                 ctx.utilityExists = true;
             }
-            return next();
+            next();
         });
     }
 
@@ -1330,11 +1335,17 @@ function isUsbKeyMounted(log, cb) {
 
     function checkMeyMounted(ctx, next) {
         if (ctx.keyIsMounted || ctx.utilityExists) {
-            return next();
+            next();
+            return;
         }
         fs.stat(ctx.key, function (er, st) {
             if (er) {
-                return next(er);
+                if (er.code === 'ENOENT') {
+                    next();
+                } else {
+                    next(er);
+                }
+                return;
             }
             if (st.isFile()) {
                 ctx.keyIsMounted = true;
@@ -1378,11 +1389,17 @@ function mountUsbKey(log, cb) {
 
     function sdcUsbkeyExists(ctx, next) {
         if (ctx.keyIsMounted) {
-            return next();
+            next();
+            return;
         }
         fs.stat(ctx.utility, function (er, st) {
             if (er) {
-                return next(er);
+                if (er.code === 'ENOENT') {
+                    next();
+                } else {
+                    next(er);
+                }
+                return;
             }
             if (st.isFile()) {
                 ctx.utilityExists = true;
@@ -1481,16 +1498,23 @@ function unmountUsbKey(log, cb) {
 
     function sdcUsbkeyExists(ctx, next) {
         if (!ctx.keyIsMounted) {
-            return next();
+            next();
+            return;
         }
         fs.stat(ctx.utility, function (er, st) {
             if (er) {
-                return next(er);
+                if (er.code === 'ENOENT') {
+                    next();
+                } else {
+                    next(er);
+                }
+                return;
             }
             if (st.isFile()) {
                 ctx.utilityExists = true;
             }
-            return next();
+            next();
+            return;
         });
     }
 
-- 
2.21.0

