commit 73f52d388539a8de1da5d8c08368b213b2348c1e (refs/changes/59/4859/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-09-21T12:22:45-04:00 (1 year, 1 month ago)
    
    OS-7256 smartos installer slow at drawing horizontal rule

diff --git a/overlay/generic/smartdc/lib/smartos_prompt_config.sh b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
index 6a0b37ac..352379b9 100644
--- a/overlay/generic/smartdc/lib/smartos_prompt_config.sh
+++ b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
@@ -578,35 +578,38 @@ updatenicstates()
 
 printheader()
 {
-	local newline=
+	local newline=false
 	local cols=`tput cols`
 	local subheader=$1
 
 	if [[ $(getanswer "simple_headers") == "true" ]]; then
-	echo "> ${subheader}"
-	return
+		echo "> ${subheader}"
+		return
 	fi
 
-	if [ $cols -gt 80 ] ;then
-	newline='\n'
+	if ((cols > 80)); then
+		newline=true
 	fi
 
 	clear
 	printf " %-40s\n" "SmartOS Setup"
 	printf " %-40s%38s\n" "$subheader" "https://wiki.smartos.org/install"
-	for i in {1..80} ; do printf "-" ; done && printf "$newline"
+
+	printf -- '-%.0s' {1..80}
+	if $newline; then
+		echo
+	fi
 }
 
 print_warning()
 {
 	clear
 	printf "WARNING\n"
-	for i in {1..80} ; do printf "-" ; done && printf "\n"
-	printf "\n$1\n"
+	printf -- '-%.0s' {1..80}
+	printf "\n\n%s\n\n" "$1"
 
-	prmpt_str="\nPress [enter] to continue "
-	printf "$prmpt_str"
-	read continue
+	prmpt_str="Press [enter] to continue "
+	read -p "$prmpt_str"
 }
 
 nicsup()
