commit 9898211ab69de3b70ce7986f8b0ccf4d466dd679 (refs/changes/92/292/8)
Author: Johanan Lai <lai.johanan@gmail.com>
Date:   2016-08-17T00:44:53-07:00 (3 years, 2 months ago)
    
    ADMINUI-2333 Incompletely setup CN shows NaN% disk utilization

diff --git a/www/js/components/pages/server/disk-overview.jsx b/www/js/components/pages/server/disk-overview.jsx
index daa884b3..8acd8ac5 100644
--- a/www/js/components/pages/server/disk-overview.jsx
+++ b/www/js/components/pages/server/disk-overview.jsx
@@ -20,9 +20,23 @@ var ServerDiskOverview = React.createClass({
     },
     render: function() {
         var server = this.props.server.toJSON();
+        if (!server.vms) {
+            return null;
+        }
 
-        var provisionable = utils.getReadableSize(server.unreserved_disk * 1048576);
-        var provisioned = utils.getReadableSize(server.disk_pool_size_bytes - (server.unreserved_disk * 1048576));
+        var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
+        var provisionable = server.unreserved_disk * 1048576;
+        var total = server.disk_pool_size_bytes;
+        
+        if (provisioned < 0 || !provisioned) { provisioned = 0; }
+        if (provisionable < 0 || !provisionable) {
+            if (!provisionable && server.disk_pool_size_bytes) { provisionable = server.disk_pool_size_bytes; }
+            else { provisionable = 0; }
+        }
+        if (total < 0 || !provisionable) { total = 0; }
+
+        var provisioned = utils.getReadableSize(provisioned);
+        var provisionable = utils.getReadableSize(provisionable);
         var total = utils.getReadableSize(server.disk_pool_size_bytes);
 
         return <div className="disk-overview">
diff --git a/www/js/components/pages/server/disk-utilization-circle.jsx b/www/js/components/pages/server/disk-utilization-circle.jsx
index 99faae0d..46358875 100644
--- a/www/js/components/pages/server/disk-utilization-circle.jsx
+++ b/www/js/components/pages/server/disk-utilization-circle.jsx
@@ -31,8 +31,11 @@ var ServerMemoryUtilizationCircle = React.createClass({
         var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
         var provisionable = server.unreserved_disk * 1048576;
 
-        if (provisioned < 0) { provisioned = 0; }
-        if (provisionable < 0) { provisionable = 0; }
+        if (provisioned < 0 || !provisioned) { provisioned = 0; }
+        if (provisionable < 0 || !provisionable) {
+            if (!provisionable && server.disk_pool_size_bytes) { provisionable = server.disk_pool_size_bytes; }
+            else { provisionable = 0; }
+        }
 
         var pieData = [
             {label: 'Provisioned', value: provisioned },
@@ -72,8 +75,10 @@ var ServerMemoryUtilizationCircle = React.createClass({
         var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
         var total = server.disk_pool_size_bytes;
 
-        var util_percent = Math.round(provisioned / total * 100);
-        if (util_percent < 0) { util_percent = 0; }
+        if (provisioned < 0 || !provisioned) { provisioned = 0; }
+        if (total < 0 || !server.unreserved_disk) { total = 0; } 
+
+        var util_percent = (provisioned && total > 0) ? Math.round(provisioned / total * 100) : 0;
 
         var pctSize, labelSize;
         if (diameter > 100) {
