commit a2d4574eb0418371a41ec43ca381d1650142d13c (refs/changes/92/292/1)
Author: Johanan Lai <lai.johanan@gmail.com>
Date:   2016-08-15T16:13:17-07:00 (3 years, 2 months ago)
    
    ADMINUI-2333 Incomplete CoaL setup shows NaN% disk utilization

diff --git a/www/js/components/pages/server/disk-utilization-circle.jsx b/www/js/components/pages/server/disk-utilization-circle.jsx
index 99faae0d..909eefe7 100644
--- a/www/js/components/pages/server/disk-utilization-circle.jsx
+++ b/www/js/components/pages/server/disk-utilization-circle.jsx
@@ -31,8 +31,12 @@ var ServerMemoryUtilizationCircle = React.createClass({
         var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
         var provisionable = server.unreserved_disk * 1048576;
 
-        if (provisioned < 0) { provisioned = 0; }
-        if (provisionable < 0) { provisionable = 0; }
+        if (provisioned < 0 || typeof provisioned !== "number") { provisioned = 0; }
+        if (provisionable < 0) {
+            provisionable = 0;
+        } else if (typeof provisioned !== "number") {
+            provisionable = server.disk_pool_size_bytes;
+        }
 
         var pieData = [
             {label: 'Provisioned', value: provisioned },
@@ -73,7 +77,7 @@ var ServerMemoryUtilizationCircle = React.createClass({
         var total = server.disk_pool_size_bytes;
 
         var util_percent = Math.round(provisioned / total * 100);
-        if (util_percent < 0) { util_percent = 0; }
+        if (util_percent < 0 || typeof util_percent !== "number") { util_percent = 0; }
 
         var pctSize, labelSize;
         if (diameter > 100) {
