commit be50ee5bdd1140768c9e4fbd7af1a63669f0a74f (refs/changes/93/993/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton fwrule get 172b1c32-2edd-4cfa-ac70-b2bcd0be0c5c

diff --git a/lib/do_fwrule/do_get.js b/lib/do_fwrule/do_get.js
index b919036..474c415 100644
--- a/lib/do_fwrule/do_get.js
+++ b/lib/do_fwrule/do_get.js
@@ -33,21 +33,26 @@ function do_get(subcmd, opts, args, cb) {
     }
 
     var id = args[0];
-    var cli = this.top;
+    var tritonapi = this.top.tritonapi;
 
-    cli.tritonapi.getFirewallRule(id, function onRule(err, fwrule) {
-        if (err) {
-            cb(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-
-        if (opts.json) {
-            console.log(JSON.stringify(fwrule));
-        } else {
-            console.log(JSON.stringify(fwrule, null, 4));
-        }
-
-        cb();
+        tritonapi.getFirewallRule(id, function onRule(err, fwrule) {
+            if (err) {
+                cb(err);
+                return;
+            }
+
+            if (opts.json) {
+                console.log(JSON.stringify(fwrule));
+            } else {
+                console.log(JSON.stringify(fwrule, null, 4));
+            }
+
+            cb();
+        });
     });
 }
 
