From bf06d0a35931f721c09839261d8dadde9ee2889f Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Mon, 17 Oct 2016 13:45:47 +0200
Subject: [PATCH] TOOLS-1573 sdcadm experimental avail ignores agent services

---
 lib/cli/index.js | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/lib/cli/index.js b/lib/cli/index.js
index f621180..6c5ad9a 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -295,12 +295,7 @@ CLI.prototype._specFromArgs = function _specFromArgs(opts, args, cb) {
             // All unresolved services from M3 go here.
             var unsupportedInsts = ['nat', 'hostvolume',
                 // removed agents:
-                'provisioner', 'heartbeater', 'zonetracker',
-                // TODO (AGENT-882: Not yet implemented individual updates):
-                'cabase', 'cainstsvc',
-                // TODO (rollback AGENT-909 during phase 2):
-                'config-agent', 'agents_core', 'smartlogin', 'amon-agent',
-                'amon-relay', 'hagfish-watcher', 'firewaller'
+                'provisioner', 'heartbeater', 'zonetracker'
             ];
             // Explicit and separate array to show which instances are locked
             // and therefore not available to udpate with --all
-- 
2.21.0

