commit 5dee57efc64bf4c0c7b54324baa9c39f57293ce9
Author: Jason King <jason.king@joyent.com>
Date:   2019-07-24T21:05:57+00:00 (3 months ago)
    
    OS-7902 Document mdb(1) ::ps flags

diff --git a/usr/src/cmd/mdb/common/modules/genunix/genunix.c b/usr/src/cmd/mdb/common/modules/genunix/genunix.c
index 95b564ed3a..348777dcf0 100644
--- a/usr/src/cmd/mdb/common/modules/genunix/genunix.c
+++ b/usr/src/cmd/mdb/common/modules/genunix/genunix.c
@@ -354,6 +354,19 @@ ps(uintptr_t addr, uint_t flags, int argc, const mdb_arg_t *argv)
 	return (DCMD_OK);
 }
 
+static void
+ps_help(void)
+{
+	mdb_printf("Display processes.\n\n",
+	    "Options:\n"
+	    "    -f\tDisplay command arguments\n",
+	    "    -l\tDisplay LWPs\n",
+	    "    -T\tDisplay tasks\n",
+	    "	 -P\tDisplay projects\n",
+	    "    -z\tDisplay zones\n",
+	    "    -t\tDisplay threads\n");
+}
+
 #define	PG_NEWEST	0x0001
 #define	PG_OLDEST	0x0002
 #define	PG_PIPE_OUT	0x0004
@@ -4036,7 +4049,8 @@ static const mdb_dcmd_t dcmds[] = {
 	{ "panicinfo", NULL, "print panic information", panicinfo },
 	{ "pid2proc", "?", "convert PID to proc_t address", pid2proc },
 	{ "project", NULL, "display kernel project(s)", project },
-	{ "ps", "[-fltzTP]", "list processes (and associated thr,lwp)", ps },
+	{ "ps", "[-fltzTP]", "list processes (and associated thr,lwp)", ps,
+	    ps_help },
 	{ "pflags", NULL, "display various proc_t flags", pflags },
 	{ "pgrep", "[-x] [-n | -o] pattern",
 		"pattern match against all processes", pgrep },
