From 779c88b260609582468f325d56fa68fe14ba99dd Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 29 Jul 2019 19:28:52 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index d25b8fc..01f669d 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
+Subproject commit 01f669da1f5cb1439d573fc42efd2dc43a7c6184
-- 
2.21.0

