From e764e3b94ca7678de6085820ccd574a9dae2ae9a Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Thu, 10 Oct 2019 10:01:32 -0400
Subject: [PATCH] TOOLS-2330 explicitly set SHELL in Makefile.defs

---
 tools/mk/Makefile.defs | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/tools/mk/Makefile.defs b/tools/mk/Makefile.defs
index 81caa42..a47c556 100644
--- a/tools/mk/Makefile.defs
+++ b/tools/mk/Makefile.defs
@@ -53,6 +53,10 @@
 # destinations, but sometimes to modify the behavior of the build.
 #
 
+
+# Be explicit about assuming bash for inline shell execution
+SHELL ?= /bin/bash
+
 TOP ?= $(shell pwd)
 
 #
-- 
2.17.2 (Apple Git-113)

