From 3993fcbf67283b74024af68a982f1cb10683aa86 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Mon, 12 Aug 2019 16:21:41 +0000
Subject: [PATCH] OS-7948 /usr/lib/fs/smbfs/mount shipped with wrong
 permissions

---
 manifest | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/manifest b/manifest
index efb0c5b6ac..20881e2b43 100644
--- a/manifest
+++ b/manifest
@@ -5898,9 +5898,9 @@ s usr/lib/fs/smbfs/amd64/libshare_smbfs.so=libshare_smbfs.so.1
 f usr/lib/fs/smbfs/dfshares 0555 root bin
 f usr/lib/fs/smbfs/libshare_smbfs.so.1 0755 root sys
 s usr/lib/fs/smbfs/libshare_smbfs.so=libshare_smbfs.so.1
-f usr/lib/fs/smbfs/mount 0555 root bin
+f usr/lib/fs/smbfs/mount 4555 root bin
 f usr/lib/fs/smbfs/share 0555 root bin
-f usr/lib/fs/smbfs/umount 0555 root bin
+f usr/lib/fs/smbfs/umount 4555 root bin
 f usr/lib/fs/smbfs/unshare 0555 root bin
 d usr/lib/fs/tmpfs 0755 root sys
 f usr/lib/fs/tmpfs/mount 0555 root bin
-- 
2.21.0

