commit 50bfea31b0e15a1fd4e7c97932b7da49b0e2a06c
Author: Rui Loura <rui@joyent.com>
Date:   2019-09-20T16:11:15-04:00 (2 weeks ago)
    
    MANTA-4569 bump rust-sharkspotter version to 0.3.0 in Cargo.toml

diff --git a/Cargo.toml b/Cargo.toml
index 8c0bea9..8c90d55 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "sharkspotter"
-version = "0.2.0"
+version = "0.3.0"
 authors = ["Rui Loura <rui@joyent.com>", "Jon Anderson <jon.anderson@joyent.com>"]
 edition = "2018"
 
