From 69a00a5ff84af315b269219203d35c3fd5f35a6c Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Tue, 9 May 2017 14:27:34 -0700
Subject: [PATCH] MANTA-3230 muppet should remove RC4, 3DES.

---
 etc/stud.conf | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/etc/stud.conf b/etc/stud.conf
index a2417ff..9c30c90 100644
--- a/etc/stud.conf
+++ b/etc/stud.conf
@@ -36,7 +36,7 @@ pem-file = "/opt/smartdc/muppet/etc/ssl.pem"
 #
 # Preferred cipher list taken from
 # https://github.com/cloudflare/sslconfig
-ciphers = "EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:HIGH:RC4-SHA:!MD5:!aNULL:!PSK"
+ciphers = "EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:HIGH:!MD5:!aNULL:!PSK"
 
 # Enforce server cipher list order
 #
@@ -121,3 +121,5 @@ write-proxy = on
 #
 # type: boolean
 proxy-proxy = off
+
+# EOF
-- 
2.21.0

