From 492a5f841b1b1249683bd38750577d9379b58340 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Wed, 10 May 2017 12:11:42 -0700
Subject: [PATCH] MANTA-3230 muppet should remove RC4, 3DES.

---
 etc/stud.conf | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/etc/stud.conf b/etc/stud.conf
index a2417ff..ed9a35e 100644
--- a/etc/stud.conf
+++ b/etc/stud.conf
@@ -34,9 +34,9 @@ pem-file = "/opt/smartdc/muppet/etc/ssl.pem"
 # Run openssl ciphers for list of available ciphers.
 # type: string
 #
-# Preferred cipher list taken from
-# https://github.com/cloudflare/sslconfig
-ciphers = "EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:EECDH+3DES:RSA+3DES:HIGH:RC4-SHA:!MD5:!aNULL:!PSK"
+# Preferred cipher list as per RFD 93
+# https://github.com/joyent/rfd/tree/master/rfd/0093
+ciphers = "EECDH+CHACHA20:EECDH+AES128:RSA+AES128:EECDH+AES256:RSA+AES256:HIGH:!MD5:!aNULL:!PSK"
 
 # Enforce server cipher list order
 #
@@ -121,3 +121,5 @@ write-proxy = on
 #
 # type: boolean
 proxy-proxy = off
+
+# EOF
-- 
2.21.0

