commit cdc31e72dc25205880f1eb23e4b0158ce8dcdff9 (refs/changes/19/3919/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-05-08T17:50:07+02:00 (1 year, 5 months ago)
    
    OS-6937 support large page mapping for bhyve IOMMU domains

diff --git a/usr/src/uts/i86pc/io/vmm/intel/vtd.c b/usr/src/uts/i86pc/io/vmm/intel/vtd.c
index 2626d9dbed..91a39645d2 100644
--- a/usr/src/uts/i86pc/io/vmm/intel/vtd.c
+++ b/usr/src/uts/i86pc/io/vmm/intel/vtd.c
@@ -705,6 +705,7 @@ vtd_create_domain(vm_paddr_t maxaddr)
 	if ((uintptr_t)dom->ptp & PAGE_MASK)
 		panic("vtd_create_domain: ptp (%p) not page aligned", dom->ptp);
 
+#ifdef __FreeBSD__
 #ifdef notyet
 	/*
 	 * XXX superpage mappings for the iommu do not work correctly.
@@ -722,6 +723,9 @@ vtd_create_domain(vm_paddr_t maxaddr)
 	 */
 	dom->spsmask = VTD_CAP_SPS(vtdmap->cap);
 #endif
+#else
+	dom->spsmask = VTD_CAP_SPS(vtdmap->cap);
+#endif
 
 	SLIST_INSERT_HEAD(&domhead, dom, next);
 
