commit 0fd87c660b6fa4f3787c87d7c65e54179b4098d4 (refs/changes/24/5624/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-22T14:38:51+00:00 (8 months ago)
    
    TRITON-1248 hermes should .gitignore more build artifacts

diff --git a/.gitignore b/.gitignore
index e2b4aef..dd0a1cc 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,4 +4,6 @@ bin/node
 node/
 proto/
 downloads/
+/build
+/make_stamps
 .*.sw?
