From 0fd87c660b6fa4f3787c87d7c65e54179b4098d4 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 22 Feb 2019 14:38:51 +0000
Subject: [PATCH] TRITON-1248 hermes should .gitignore more build artifacts

---
 .gitignore | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/.gitignore b/.gitignore
index e2b4aef..dd0a1cc 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,4 +4,6 @@ bin/node
 node/
 proto/
 downloads/
+/build
+/make_stamps
 .*.sw?
-- 
2.21.0

