From 4f8bddc8a92733b28c7c5e0f924edb5977182782 Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Mon, 27 Nov 2017 23:00:17 +0000
Subject: [PATCH] OS-6465 update ACPI to latest version and bring in iasl

---
 configure | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/configure b/configure
index 74a38f17..7afe1a2f 100755
--- a/configure
+++ b/configure
@@ -83,7 +83,7 @@ function install_pkgin
 	if pkgin av build-essential | grep ^build-essential >/dev/null; then
 		pkglist="build-essential"
 	else
-		pkglist="gmake binutils autoconf automake bison"
+		pkglist="gmake binutils autoconf automake bison m4"
 		pkglist="$pkglist libtool-base gcc-compiler"
 	fi
 
@@ -310,6 +310,8 @@ AR=/usr/bin/ar;					export AR
 MAKE="$conf_root/proto.strap/usr/bin/dmake";	export MAKE
 LEX=/opt/local/bin/lex;				export LEX
 YACC=/opt/local/bin/yacc;			export YACC
+BISON=/opt/local/bin/bison;			export BISON
+GM4=/opt/local/bin/gm4;				export GM4
 RPCGEN=/opt/local/bin/rpcgen;			export RPCGEN
 ASTBINDIR=/opt/local/ast/bin;			export ASTBINDIR
 LD_TOXIC_PATH="\$ROOT/lib:\$ROOT/usr/lib";	export LD_TOXIC_PATH
-- 
2.21.0

