From ec99de2b64475809b46b14836516f23e22590460 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Tue, 26 Feb 2019 11:14:26 +0000
Subject: [PATCH] TRITON-1141 convert sdc-agents-core to engbld framework (fix
 node version)

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fdd4711..1cfc5c7 100644
--- a/Makefile
+++ b/Makefile
@@ -27,7 +27,7 @@ JSSTYLE_FLAGS    = -o indent=4,doxygen,unparenthesized-return=0
 NODE_PREBUILT_VERSION=v0.12.14
 NODE_PREBUILT_TAG=gz
 ifeq ($(shell uname -s),SunOS)
-    NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
+    NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
 endif
 
 ENGBLD_REQUIRE := $(shell git submodule update --init deps/eng)
-- 
2.21.0

