commit 06292d5c4579586b37bdb53e0e81342ec88d6196 (refs/changes/47/3547/1)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2018-03-06T11:26:49-08:00 (1 year, 7 months ago)
    
    MORAY-461 moray VError 'object is not a function'

diff --git a/lib/pg.js b/lib/pg.js
index 695f8c6..f79b2a6 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -19,7 +19,8 @@ var once = require('once');
 var pg = require('pg');
 require('pg-parse-float')(pg);
 var pooling = require('pooling');
-var VError = require('verror');
+var verror = require('verror');
+var VError = verror.VError;
 
 var dtrace = require('./dtrace');
 
