From 06292d5c4579586b37bdb53e0e81342ec88d6196 Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Tue, 6 Mar 2018 11:26:49 -0800
Subject: [PATCH] MORAY-461 moray VError 'object is not a function'

---
 lib/pg.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/pg.js b/lib/pg.js
index 695f8c6..f79b2a6 100644
--- a/lib/pg.js
+++ b/lib/pg.js
@@ -19,7 +19,8 @@ var once = require('once');
 var pg = require('pg');
 require('pg-parse-float')(pg);
 var pooling = require('pooling');
-var VError = require('verror');
+var verror = require('verror');
+var VError = verror.VError;
 
 var dtrace = require('./dtrace');
 
-- 
2.21.0

