From 62386d11214db3feac1507b54ecd6862aa34b247 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Thu, 24 Jan 2019 12:56:54 -0700
Subject: [PATCH] TRITON-1101 add disklayout support for large storage servers

---
 scripts/joysetup.sh | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/scripts/joysetup.sh b/scripts/joysetup.sh
index 2b4e6cfb..1423a6dd 100755
--- a/scripts/joysetup.sh
+++ b/scripts/joysetup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright 2017 Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 #
@@ -621,12 +621,16 @@ if [[ "$(zpool list)" == "no pools available" ]] \
     if [[ -n "${CONFIG_cache}" && "${CONFIG_cache}" == "false" ]]; then
         dl2="-c"
     fi
+    if [[ -n "${CONFIG_width}" ]]; then
+        dl3="-w"
+        dl4="${CONFIG_width}"
+    fi
     if [[ -n "${CONFIG_layout}" ]]; then
-        dl3="${CONFIG_layout}"
+        dl5="${CONFIG_layout}"
     fi
 
-    if ! /usr/bin/disklayout ${dl0} ${dl1:+"$dl1"} ${dl2} "${dl3}" \
-         > ${POOL_FILE}; then
+    if ! /usr/bin/disklayout ${dl0} ${dl1:+"$dl1"} ${dl2} \
+	${dl3} ${dl4:+"$dl4"} "${dl5}" > ${POOL_FILE}; then
         fatal "disk layout failed"
     fi
 
-- 
2.21.0

