commit e9681ef0d32f404899373e4639f238fa42f35aa6 (refs/changes/79/379/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2016-08-29T23:48:52+00:00 (3 years, 1 month ago)
    
    MANATEE-312 manatee should not set snapdir to visible

diff --git a/tools/mkdevsitters b/tools/mkdevsitters
index a6dc936..78f146a 100755
--- a/tools/mkdevsitters
+++ b/tools/mkdevsitters
@@ -32,8 +32,7 @@ function mksitter
 	mountpoint="$PWD/devconfs/datasets/manatee$which"
 
 	echo -n "peer $which: creating ZFS dataset ... "
-	if ! zfs create -o mountpoint=$mountpoint -o canmount=noauto \
-	    -o snapdir=visible $dataset; then
+	if ! zfs create -o mountpoint=$mountpoint -o canmount=noauto; then
 		fail "failed to create dataset \"$dataset\""
 	fi
 
