commit f7614620a4916bbcc0bd440c2f17f9074ae1015c (refs/changes/79/379/2)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2016-09-08T20:52:27+00:00 (3 years, 1 month ago)
    
    MANATEE-312 manatee should not set snapdir to visible

diff --git a/tools/mkdevsitters b/tools/mkdevsitters
index a6dc936..92b23aa 100755
--- a/tools/mkdevsitters
+++ b/tools/mkdevsitters
@@ -33,7 +33,7 @@ function mksitter
 
 	echo -n "peer $which: creating ZFS dataset ... "
 	if ! zfs create -o mountpoint=$mountpoint -o canmount=noauto \
-	    -o snapdir=visible $dataset; then
+	    $dataset; then
 		fail "failed to create dataset \"$dataset\""
 	fi
 
