From 090d4a764cba06596a2a5554097c67976f342749 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Mon, 30 Jan 2017 18:20:38 -0800
Subject: [PATCH] MORAY-393 document moray env file

---
 env.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/env.sh b/env.sh
index e652223..2bfec26 100644
--- a/env.sh
+++ b/env.sh
@@ -8,6 +8,12 @@
 # Copyright (c) 2017, Joyent, Inc.
 #
 
+#
+# This environment file is intended for use only in development environments.
+# This should set up any environment variables or aliases that are helpful for
+# running Moray directly out of this workspace.  See the README for details.
+#
+
 set -o xtrace
 export PATH=$PWD/build/node/bin:$PWD/node_modules/.bin:node_modules/moray/bin:$PATH
 alias npm='node `which npm`'
-- 
2.21.0

