commit 090d4a764cba06596a2a5554097c67976f342749 (refs/changes/77/1377/1)
Author: David Pacheco <dap@joyent.com>
Date:   2017-01-30T18:20:38-08:00 (2 years, 8 months ago)
    
    MORAY-393 document moray env file

diff --git a/env.sh b/env.sh
index e652223..2bfec26 100644
--- a/env.sh
+++ b/env.sh
@@ -8,6 +8,12 @@
 # Copyright (c) 2017, Joyent, Inc.
 #
 
+#
+# This environment file is intended for use only in development environments.
+# This should set up any environment variables or aliases that are helpful for
+# running Moray directly out of this workspace.  See the README for details.
+#
+
 set -o xtrace
 export PATH=$PWD/build/node/bin:$PWD/node_modules/.bin:node_modules/moray/bin:$PATH
 alias npm='node `which npm`'
