From 9170f582f08bdda05e0477e784c09bf70934b9c3 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Wed, 30 Nov 2016 13:52:04 -0500
Subject: [PATCH] joyent/node-triton#149 resurrected example-get-account

---
 examples/example-get-account.js | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/examples/example-get-account.js b/examples/example-get-account.js
index 043f5d5..e35e725 100755
--- a/examples/example-get-account.js
+++ b/examples/example-get-account.js
@@ -16,16 +16,18 @@ var log = bunyan.createLogger({
     level: 'trace'
 })
 
-var USER = process.env.SDC_USER || process.env.SDC_ACCOUNT || 'bob';
+var ACCOUNT = process.env.SDC_ACCOUNT || 'bob';
+var USER = process.env.SDC_USER;
 var KEY_ID = process.env.SDC_KEY_ID || 'b4:f0:b4:6c:18:3b:44:63:b4:4e:58:22:74:43:d4:bc';
 
 var sign = auth.cliSigner({
     keyId: KEY_ID,
-    user: USER,
+    user: ACCOUNT,
     log: log
 });
 var client = cloudapi.createClient({
     url: 'https://us-sw-1.api.joyent.com',
+    account: ACCOUNT,
     user: USER,
     version: '*',
     sign: sign,
-- 
2.21.0

