commit af2992704f1ce1e972e7a290f46c8497356c91a8 (refs/changes/65/3165/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-01-05T17:38:57-08:00 (1 year, 9 months ago)
    
    HEAD-2386 workflow_recent_jobs "recent" calculation is broken

diff --git a/tools/dump-hourly-sdc-data.sh b/tools/dump-hourly-sdc-data.sh
index 3806e0d..ab622e9 100755
--- a/tools/dump-hourly-sdc-data.sh
+++ b/tools/dump-hourly-sdc-data.sh
@@ -221,7 +221,7 @@ sdc-workflow /workflows >$DUMPDIR/workflow_workflows-$TIMESTAMP.json
 [ $? -ne 0 ] && echo "$0: error: Dumping WFAPI workflows failed" >&2
 # Right now we dump recent jobs (jobs created in the past 2 hours)
 now=$TIMESTAMP
-ago=$(date -u "+%s" -d -2hour)
+ago=$((now - 7200))
 # javascript expects milliseconds
 now=$((now * 1000))
 ago=$((ago * 1000))
