commit 3a4a64bac42d46f1c35c84da204f6b48537effa2 (refs/changes/77/3377/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-02-15T16:28:00-08:00 (1 year, 8 months ago)
    
    TRITON-154 vmapi updates existing nics with the wrong state

diff --git a/test/vmapi.test.js b/test/vmapi.test.js
index 22faa7b..89816ce 100644
--- a/test/vmapi.test.js
+++ b/test/vmapi.test.js
@@ -284,8 +284,8 @@ test('vmapi', function (tt) {
 
             napi.listNetworks({ name: 'external' }, function (err2, nets2) {
                 t.ifError(err2);
-                t.ok(nets1);
-                EXTERNAL_NETWORK = nets1[0].uuid;
+                t.ok(nets2);
+                EXTERNAL_NETWORK = nets2[0].uuid;
                 t.end();
             });
         });
@@ -816,6 +816,7 @@ test('vmapi', function (tt) {
             t.ifError(err);
             t.ok(job);
             JOB_UUID = job.job_uuid;
+            t.ok(JOB_UUID, 'JOB_UUID is ' + JOB_UUID);
             t.end();
         });
     });
