From 2f1074037228cf09814d4a86dd8895616c6b7282 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 3 Oct 2016 12:10:49 -0700
Subject: [PATCH] joyent/node-cueball#33 claim() should double-check state of
 ConnectionFSM

---
 lib/pool.js | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/lib/pool.js b/lib/pool.js
index 98a97ab..9bc4a0b 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -605,9 +605,19 @@ CueBallConnectionPool.prototype.claim = function (options, cb) {
 	Error.captureStackTrace(e);
 
 	/* If there are idle connections sitting around, take one. */
-	if (this.p_idleq.length > 0) {
+	while (this.p_idleq.length > 0) {
 		var fsm = this.p_idleq.shift();
 		delete (fsm.p_idleq_node);
+		/*
+		 * Since 'stateChanged' is emitted async from mooremachine,
+		 * things may be on the idle queue still but not actually idle.
+		 * If we find one, just rip it off the queue (which we've
+		 * already done) and try the next thing. The state mgmt
+		 * callback from addConnection will cope.
+		 */
+		if (!fsm.isInState('idle'))
+			continue;
+
 		fsm.claim(e.stack, function (err, hdl, conn) {
 			if (err) {
 				if (!done)
-- 
2.21.0

