commit 5fe18237bbd3214c4ace75b492905696958d3edf (refs/changes/52/3152/2)
Author: David Pacheco <dap@joyent.com>
Date:   2018-01-05T15:56:40-08:00 (1 year, 9 months ago)
    
    joyent/node-fast#15 server leaks memory on socket errors

diff --git a/CHANGES.md b/CHANGES.md
index 024fd84..40b5d8a 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,11 @@
 
 ## Not yet released.
 
+None yet.
+
+## v2.3.2
+
+* #15 server leaks memory on socket errors
 * #12 want support for node v6.12.0
 
 ## v2.3.1
diff --git a/lib/fast_server.js b/lib/fast_server.js
index 66e0485..620f476 100644
--- a/lib/fast_server.js
+++ b/lib/fast_server.js
@@ -565,15 +565,25 @@ FastServer.prototype.connDrain = function (conn)
 		    return ([ self.fs_dtid, conn.fc_connid ]);
 		});
 
-		/*
-		 * As long as we didn't see a socket error and didn't already
-		 * terminate the socket, destroy the socket.  We could try to
-		 * end it gracefully, but we don't actually want to wait for the
-		 * client to shut it down cleanly.  If we already saw an error,
-		 * then there's nothing else to do.
-		 */
-		if (conn.fc_socket_error === null &&
-		    conn.fc_server_error === null) {
+		if (conn.fc_socket_error !== null) {
+			/*
+			 * If we did see a socket error, then we must explicitly
+			 * tear down the pipeline.  Otherwise, these streams
+			 * will maintain references to each other, resulting in
+			 * a leak.
+			 */
+			conn.fc_socket.unpipe();
+			conn.fc_ckddecoder.unpipe();
+			conn.fc_rawdecoder.unpipe();
+		} else if (conn.fc_server_error === null) {
+			/*
+			 * As long as we didn't see a socket error and didn't
+			 * already terminate the socket, destroy the socket.  We
+			 * could try to end it gracefully, but we don't actually
+			 * want to wait for the client to shut it down cleanly.
+			 * If we already saw an error, then there's nothing else
+			 * to do.
+			 */
 			conn.fc_socket.destroy();
 		}
 	}
diff --git a/package.json b/package.json
index 6065578..4b4ef7d 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
 	"name": "fast",
 	"description": "streaming JSON RPC over TCP",
-	"version": "2.3.1",
+	"version": "2.3.2",
 	"main": "./lib/fast.js",
 	"repository": {
 		"type": "git",
