commit d0798bfdb4ada9435e16510be5f1735f527fbbf1 (refs/changes/90/1090/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-12-07T21:17:10-08:00 (2 years, 10 months ago)
    
    joyent/manta-thoth#144 bump dependency versions and release 1.0.0

diff --git a/bin/thoth b/bin/thoth
index e9bd6c4..f962b5d 100755
--- a/bin/thoth
+++ b/bin/thoth
@@ -10,7 +10,7 @@ var mod_fs = require('fs');
 var util = require('util');
 var mod_crypto = require('crypto');
 var mod_bunyan = require('bunyan');
-var mod_restify = require('restify');
+var mod_restify_clients = require('restify-clients');
 var mod_manta = require('manta');
 var mod_zlib = require('zlib');
 var mod_child = require('child_process');
@@ -3435,7 +3435,7 @@ var connect = function ()
 		name: mod_path.basename(process.argv[1]),
 		level: (process.env.LOG_LEVEL || 'info'),
 		stream: process.stderr,
-		serializers: mod_restify.bunyan.serializers
+		serializers: mod_restify_clients.bunyan.serializers
 	});
 
 	return (mod_manta.createBinClient(opts));
diff --git a/package.json b/package.json
index 5abd9a2..c176ccb 100644
--- a/package.json
+++ b/package.json
@@ -1,15 +1,15 @@
 {
 	"name": "manta-thoth",
 	"description": "Manta-based system for core and crash dump analysis",
-	"version": "0.2.1",
+	"version": "1.0.0",
 	"author": "Joyent (joyent.com)",
 	"dependencies": {
-		"manta": "1.6.0",
+		"manta": "^4.0.0",
 		"ctype": ">=0.5.2",
 		"sprintf": "0.1.1",
-		"bunyan": "0.21.3",
-		"jsprim": "^0.8.0",
-		"restify": "2.5.0",
+		"bunyan": "^1.8.5",
+		"jsprim": "^1.3.1",
+		"restify-clients": "^1.4.1",
 		"rethinkdb": "2.0.0",
 		"readable-stream": ">=1.1.9"
 	},
@@ -19,6 +19,5 @@
 	"repository": {
 		"type" : "git",
 		"url" : "http://github.com/joyent/manta-thoth.git"
-	},
-	"engines": { "node": ">=0.8" }
+	}
 }
