commit d5de5fb65369bed1787c640ba21b3a222f05bacb (refs/changes/90/4590/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-07-26T17:01:34-07:00 (1 year, 2 months ago)
    
    TRITON-638 Triton isn't ready for VMAPI using node v6 yet

diff --git a/Makefile b/Makefile
index e617edb..0fc9722 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,7 @@
 
 # The prebuilt sdcnode version we want. See
 # "tools/mk/Makefile.node_prebuilt.targ" for details.
-NODE_PREBUILT_VERSION=v6.14.0
+NODE_PREBUILT_VERSION=v4.9.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than sdc-minimal-multiarch-lts@15.4.1
diff --git a/package.json b/package.json
index 77c4a3a..0395b9c 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.6.0",
+  "version": "9.6.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
