commit 5234f16c65e5b30151f33fff23d4f90373eefe25 (refs/changes/14/1114/1)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2016-12-14T09:21:09-08:00 (2 years, 10 months ago)
    
    CNAPI-683 CNAPI should log waitlist requests

diff --git a/lib/models/waitlist.js b/lib/models/waitlist.js
index be13ae9..8d94332 100644
--- a/lib/models/waitlist.js
+++ b/lib/models/waitlist.js
@@ -187,6 +187,14 @@ WaitlistDirector.prototype.onUpdate = function (timestamp, tickets) {
                 delete self.callbacks[ticket.uuid];
                 fecb();
                 return;
+            } else if (
+                ticket.status === 'active' &&
+                !self.callbacks[ticket.uuid])
+            {
+                self.log.info(
+                    { ticket: ticket },
+                    'onUpdate: ticket %s active but no callbacks found',
+                    ticket.uuid);
             } else {
                 self.log.info(
                     { ticket: ticket }, 'nothing to do for ticket %s onUpdate',
@@ -207,6 +215,8 @@ WaitlistDirector.prototype.onUpdate = function (timestamp, tickets) {
 WaitlistDirector.prototype.waitForTicketByUuid = function (uuid, callback) {
     var self = this;
 
+    self.log.info('waitForTicketByUuid: ticket %s', uuid)k
+
     ModelWaitlist.getTicket(uuid, function (error, t) {
         if (error) {
             callback(new VError('fetching ticket %s', uuid));
@@ -602,7 +612,6 @@ ModelWaitlist.prototype.ensureServerQueue = function (callback) {
     var serverqueue = null;
     var etag = null;
     var moray = ModelWaitlist.getMoray();
-    var attempts = 0;
 
     vasync.waterfall([
         function (wfcb) {
@@ -640,11 +649,9 @@ ModelWaitlist.prototype.ensureServerQueue = function (callback) {
                        (VError.hasCauseWithName(err, 'EtagConflictError') ||
                        (VError.hasCauseWithName(err, 'UniqueAttributeError'))))
                     {
-                        attempts++;
                         self.log.warn({ err: err },
-                            'waitlist collision on queue ' +
-                            'initialization after %s attempts, retrying',
-                            attempts);
+                            'waitlist collision on queue initialization, retrying');
+                            
                         process.nextTick(function () {
                             self.ensureServerQueue(wfcb);
                         });
@@ -895,15 +902,19 @@ function (ticket, serverqueue, etag, callback) {
     }
 
     ticket.status = 'finished';
-    ticket.updated_at = (new Date()).toISOString();
-    ticket.updated_at =
-    serverqueue.updated_at =
-        (new Date()).toISOString();
+    ticket.updated_at = serverqueue.updated_at = (new Date()).toISOString();
 
     var queue = serverqueue.tickets[key];
     var ticketIdx = queue.indexOf(ticket.uuid);
 
-    // This is the payload we will submit to moray.batch
+    /**
+     * Craft the payload we will submit to moray.batch()
+     *
+     * This includes:
+     *   - modifying the metadata on the finished ticket
+     *   - remove finished ticket from the queue it's in
+     *   - modifying the metadata on the next ticket in the queue
+     */
     var data = [
         {
             bucket: buckets.waitlist_tickets.name,
@@ -971,6 +982,11 @@ function (ticket, serverqueue, etag, callback) {
             self.log.info({ batch: data },
                           'doing batch write after finishing %s', ticket.uuid);
             moray.batch(data, function (err, meta) {
+                if (err) {
+                    self.log.warn({ err: err },
+                        'batch error after finishing %s', ticket.uuid);
+                }
+
                 if (err &&
                    (VError.hasCauseWithName(err, 'EtagConflictError') ||
                    (VError.hasCauseWithName(err, 'UniqueAttributeError'))))
@@ -981,6 +997,7 @@ function (ticket, serverqueue, etag, callback) {
                                 wfcb(err2);
                                 return;
                             }
+                            self.log.info({ err: err }, 'retrying self.finishTicketUpdateQueueActivateNext %s', ticket.uuid);
                             self.finishTicketUpdateQueueActivateNext(
                                 ticket,
                                 resp.serverqueue,
@@ -995,6 +1012,9 @@ function (ticket, serverqueue, etag, callback) {
         }
     ],
     function (wferr) {
+        self.log.info(
+            { err: err }, 'error self.finishTicketUpdateQueueActivateNext %s',
+            ticket.uuid);
         callback(wferr);
     });
 };
diff --git a/lib/validation/endpoints.js b/lib/validation/endpoints.js
index fad2296..5b37e45 100644
--- a/lib/validation/endpoints.js
+++ b/lib/validation/endpoints.js
@@ -93,7 +93,11 @@ function ensureParamsValid(req, res, paramRules, opts) {
 
                     if (!req.params.hasOwnProperty(paramName)) {
                         skip = true;
-                        req.params[paramName] = rule[1];
+
+                        if (rule.length > 1) {
+                            req.params[paramName] = rule[1];
+                        }
+
                         break;
                     }
                     break;
