commit 7de6dc28eef880db2c0ded60080d54dc48c52471 (refs/changes/47/4247/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-06-18T17:08:23+02:00 (1 year, 4 months ago)
    
    OS-7031 panic in ppt_bar_crawl()

diff --git a/usr/src/uts/i86pc/io/vmm/io/ppt.c b/usr/src/uts/i86pc/io/vmm/io/ppt.c
index 89e9f35c10..0c5815278e 100644
--- a/usr/src/uts/i86pc/io/vmm/io/ppt.c
+++ b/usr/src/uts/i86pc/io/vmm/io/ppt.c
@@ -406,7 +406,7 @@ ppt_bar_crawl(struct pptdev *ppt)
 	}
 
 	VERIFY3S(rlen, >, 0);
-	rcount = (rlen * sizeof (int)) / sizeof (pci_regspec_t);
+	rcount = rlen / sizeof (pci_regspec_t);
 	for (i = 0; i < rcount; i++) {
 		pci_regspec_t *reg = &regs[i];
 		struct pptbar *pbar;
