From f5e7f20341c8cb71b241b919bd4e979612a480d0 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Thu, 22 Feb 2018 22:54:41 +0000
Subject: [PATCH] TRITON-176 Move FWAPI to new Triton origin image

---
 targets.json.in | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 57760f2..97cd66a 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -812,10 +812,9 @@ cat <<EOF
     "image_name": "fwapi",
     "image_description": "SDC FWAPI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
-    "pkgsrc": [
-      "openldap-client-2.4.24"
-    ],
+    "// image_uuid": "triton-origin-multiarch-15.4.1@1.0.1",
+    "image_uuid": "04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
+    "pkgsrc": [],
     "repos": [
         {"url": "git@github.com:joyent/sdc-fwapi.git"}
     ],
-- 
2.21.0

