commit 660fbaf055c38abc47bdce9134e9f32d496c2df3 (refs/changes/01/5501/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-06T18:30:46+00:00 (8 months ago)
    
    TOOLS-2174 agent-cache 'clone' should use '--' when switching branches

diff --git a/deps/eng b/deps/eng
index 4443fcd..7c472bc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 4443fcd38fe1682556e1851df97c7d652d078f3d
+Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
