commit aa95965cf54a84c039e7ef179e745bbc800930d1 (refs/changes/26/1226/1)
Author: Yang Yong <yy1111.yang@samsung.com>
Date:   2017-01-09T15:26:03+09:00 (2 years, 9 months ago)
    
    joyent/node-triton#162 triton make test is not ok

diff --git a/test/integration/cli-basics.test.js b/test/integration/cli-basics.test.js
index b1a285a..2eca291 100644
--- a/test/integration/cli-basics.test.js
+++ b/test/integration/cli-basics.test.js
@@ -29,7 +29,7 @@ test('triton (basics)', function (tt) {
         h.triton('--version', function (err, stdout, stderr) {
             if (h.ifErr(t, err, 'triton --version'))
                 return t.end();
-            t.ok(/^triton \d+\.\d+\.\d+/.test(stdout));
+            t.ok(/^Triton CLI \d+\.\d+\.\d+/.test(stdout));
             t.end();
         });
     });
