commit f118f1816eb77e582d692e242d6dbea6344f6727 (refs/changes/80/3980/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-05-18T15:37:30-07:00 (1 year, 5 months ago)
    
    TRITON-52 x-DC image copy

diff --git a/CHANGES.md b/CHANGES.md
index fa5e924..e7355bc 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,11 @@
 
 ## not yet released
 
+## 10.6.0
+
+- TRITON-52 Add imgapi.importImageFromDc method, which can be used to copy an
+  image between different DCs.
+
 ## 10.5.0
 
 - DOCKER-524 Implement docker push
diff --git a/lib/imgapi.js b/lib/imgapi.js
index 0431eed..666f1f7 100644
--- a/lib/imgapi.js
+++ b/lib/imgapi.js
@@ -2529,6 +2529,94 @@ function deleteImage(uuid, account, options, callback) {
 };
 
 
+/**
+ * Imports an image from the provided datatcenter.
+ *
+ * @param {String} uuid : the UUID of the image.
+ * @param {UUID} account : The UUID of the account on behalf of whom
+ *      this request is being made.
+ * @param {Object} options: Required.
+ *      - dc {String} Required. The datacenter from which to import from.
+ *      - headers {Object} Optional. Additional request headers.
+ * @param {Function} callback : `function (err, job, res)`
+ */
+IMGAPI.prototype.importImageFromDc =
+function importImageFromDc(uuid, account, options, callback) {
+    var self = this;
+    assert.uuid(uuid, 'uuid');
+    assert.uuid(account, 'account');
+    assert.object(options, 'options');
+    assert.string(options.dc, 'options.dc');
+    assert.optionalObject(options.headers, 'options.headers');
+    assert.func(callback, 'callback');
+
+    var path = self._path(format('/images/%s', uuid), {
+        account: account,
+        action: 'import-from-dc',
+        channel: self.channel,
+        dc: options.dc
+    });
+    self._getAuthHeaders(function (hErr, headers) {
+        if (hErr) {
+            callback(hErr);
+            return;
+        }
+        if (options.headers) {
+            simpleMerge(headers, options.headers);
+        }
+        var reqOpts = {
+            path: path,
+            headers: headers
+        };
+        self.client.post(reqOpts, function (err, req, res, obj) {
+            if (err) {
+                callback(err, null, res);
+            } else {
+                callback(null, obj, res);
+            }
+        });
+    });
+};
+
+
+/**
+ * Imports an image from the provided datatcenter.
+ *
+ * @param {String} uuid : the UUID of the image.
+ * @param {UUID} account : The UUID of the account on behalf of whom
+ *      this request is being made.
+ * @param {Object} options: Required.
+ *      - dc {String} Required. The datacenter from which to import from.
+ *      - headers {Object} Optional. Additional request headers.
+ * @param {Function} callback : `function (err, image, res)`
+ */
+IMGAPI.prototype.importImageFromDcAndWait =
+function importImageFromDcAndWait(uuid, account, options, callback) {
+    var self = this;
+
+    self.importImageFromDc(uuid, account, options, function (err, job, res) {
+        if (err) {
+            callback(err, null, res);
+            return;
+        }
+
+        var wfapiUrl = res.headers['workflow-api'];
+
+        assert.string(wfapiUrl, 'wfapiUrl');
+        assert.string(job['job_uuid'], 'job_uuid');
+        assert.string(job['image_uuid'], 'image_uuid');
+
+        waitForJob(wfapiUrl, job['job_uuid'], function (jErr) {
+            if (jErr) {
+                callback(jErr);
+                return;
+            }
+            self.getImage(job['image_uuid'], callback);
+        });
+    });
+};
+
+
 /**
  * ListChannels
  * <https://mo.joyent.com/docs/imgapi/master/#ListChannels>
diff --git a/package.json b/package.json
index a319985..48cf29e 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "10.5.0",
+  "version": "10.6.0",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
