From f0e2d1382bf7a985910eac85e6de5b77ed4eae4b Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Wed, 12 Oct 2016 11:32:48 +0000
Subject: [PATCH] 7458 I219 Kaby Lake Support

---
 overlay/generic/etc/driver_aliases | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 93c6bf39..99ec4d58 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -362,6 +362,10 @@ e1000g "pci8086,15a3"
 e1000g "pci8086,15b7"
 e1000g "pci8086,15b8"
 e1000g "pci8086,15b9"
+e1000g "pci8086,15d6"
+e1000g "pci8086,15d7"
+e1000g "pci8086,15d8"
+e1000g "pci8086,15e3"
 e1000g "pci8086,294c"
 e1000g "pci8086,f0fe"
 e1000g "pciex8086,1049"
@@ -426,6 +430,10 @@ e1000g "pciex8086,15a3"
 e1000g "pciex8086,15b7"
 e1000g "pciex8086,15b8"
 e1000g "pciex8086,15b9"
+e1000g "pciex8086,15d6"
+e1000g "pciex8086,15d7"
+e1000g "pciex8086,15d8"
+e1000g "pciex8086,15e3"
 e1000g "pciex8086,294c"
 e1000g "pciex8086,f0fe"
 ecpp "lp"
-- 
2.21.0

