commit f0e2d1382bf7a985910eac85e6de5b77ed4eae4b (refs/changes/57/657/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2016-10-12T11:32:48+00:00 (3 years ago)
    
    7458 I219 Kaby Lake Support

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 93c6bf39..99ec4d58 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -362,6 +362,10 @@ e1000g "pci8086,15a3"
 e1000g "pci8086,15b7"
 e1000g "pci8086,15b8"
 e1000g "pci8086,15b9"
+e1000g "pci8086,15d6"
+e1000g "pci8086,15d7"
+e1000g "pci8086,15d8"
+e1000g "pci8086,15e3"
 e1000g "pci8086,294c"
 e1000g "pci8086,f0fe"
 e1000g "pciex8086,1049"
@@ -426,6 +430,10 @@ e1000g "pciex8086,15a3"
 e1000g "pciex8086,15b7"
 e1000g "pciex8086,15b8"
 e1000g "pciex8086,15b9"
+e1000g "pciex8086,15d6"
+e1000g "pciex8086,15d7"
+e1000g "pciex8086,15d8"
+e1000g "pciex8086,15e3"
 e1000g "pciex8086,294c"
 e1000g "pciex8086,f0fe"
 ecpp "lp"
