commit 3960a8df7b6b0710328aeb36f328a8755278267c (refs/changes/24/4624/3)
Author: dyep49 <dyep49@gmail.com>
Date:   2018-08-02T12:38:26-07:00 (1 year, 2 months ago)
    
    TRITON-488 Add metricPorts metadata to CloudAPI for cmon-agent

diff --git a/boot/setup.sh b/boot/setup.sh
index 0775d3f..6dd8cdb 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -7,7 +7,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -162,6 +162,9 @@ sdc_log_rotation_add cloudapi-8083 /var/svc/log/*cloudapi-8083.log 1g
 sdc_log_rotation_add cloudapi-8084 /var/svc/log/*cloudapi-8084.log 1g
 sdc_log_rotation_setup_end
 
+# Add metadata for cmon-agent discovery
+mdata-put metricPorts 8881,8882,8883,8884
+
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
 
diff --git a/package.json b/package.json
index b370216..011e611 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "SmartDataCenter CloudAPI",
-    "version": "9.2.2",
+    "version": "9.2.3",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
