From 520274a3a73c9b7988b0e5acfb1738a3fe1972c9 Mon Sep 17 00:00:00 2001
From: dyep49 <dyep49@gmail.com>
Date: Thu, 2 Aug 2018 12:09:34 -0700
Subject: [PATCH] TRITON-488 Add metricPorts metadata to CloudAPI for
 cmon-agent

---
 boot/setup.sh | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index 0775d3f..6dd8cdb 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -7,7 +7,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -162,6 +162,9 @@ sdc_log_rotation_add cloudapi-8083 /var/svc/log/*cloudapi-8083.log 1g
 sdc_log_rotation_add cloudapi-8084 /var/svc/log/*cloudapi-8084.log 1g
 sdc_log_rotation_setup_end
 
+# Add metadata for cmon-agent discovery
+mdata-put metricPorts 8881,8882,8883,8884
+
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
 
-- 
2.21.0

