commit a04cf78b5477f025b52b9bd95a90f0414e431807 (refs/changes/60/2060/1)
Author: David Pacheco <dap@joyent.com>
Date:   2017-06-08T14:39:14-07:00 (2 years, 4 months ago)
    
    MANATEE-333 manatee-adm could report sync replay lag

diff --git a/lib/adm.js b/lib/adm.js
index 303322e..dd822ef 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -379,7 +379,7 @@ function _addPostgresStatus(_, _cb) {
                 } else {
                     entry.online = true;
                     entry.repl = res.rows[0] ? res.rows[0] : {};
-                    if (peer !== 'primary' && peer !== 'sync') {
+                    if (peer !== 'primary') {
                         queryPg(pgUrl, PG_REPL_LAG,
                                 function (err2, res2) {
                                     if (err2) {
@@ -597,11 +597,10 @@ function _setActiveZkNode(opts, _cb) {
  *                      this peer is the upstream.  This object will contain
  *                      fields from postgres's "pg_stat_replication" view.
  *
- *          pgp_lag     if this peer is the target of async replication,
- *                      indicates how long since data was received from
- *                      upstream.  Note that if no writes are being committed
- *                      upstream, this may grow large, and that's not
- *                      necessarily a problem.
+ *          pgp_lag     if this peer is the target of replication, indicates how
+ *          		long since data was replayed from upstream.  Note that
+ *          		if no writes are being committed upstream, this may grow
+ *          		large, and that's not necessarily a problem.
  *
  *     pgs_generation   current cluster generation number
  *
