From bf900d5f9b47322e414a4d07fbccfa33b8b0b3f3 Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Mon, 28 Nov 2016 17:26:01 -0800
Subject: [PATCH] HEAD-2332 Revert HEAD-2167 to include sdcsso logs for hermes
 upload

---
 etc/logsets.json | 1 +
 1 file changed, 1 insertion(+)

diff --git a/etc/logsets.json b/etc/logsets.json
index 7f92990..f019b96 100644
--- a/etc/logsets.json
+++ b/etc/logsets.json
@@ -230,6 +230,7 @@
       "redis",
       "sapi",
       "sdc",
+      "sdcsso",
       "ufds",
       "vmapi",
       "workflow"
-- 
2.21.0

