From 04bd767b4e6f0edeb51b68435e1f7e706b631aa6 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 5 Mar 2019 14:49:37 -0500
Subject: [PATCH] TOOLS-2186 sdcnode updates for Feburary 2019 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fa4e8ef..0ea0047 100644
--- a/Makefile
+++ b/Makefile
@@ -37,7 +37,7 @@ JSSTYLE_FLAGS =		-o indent=4,doxygen,unparenthesized-return=0
 # doesn't know about (@@ENABLED@@)
 # SMF_MANIFESTS_IN = smf/manifests/cn-agent.xml.in
 
-NODE_PREBUILT_VERSION =	v6.15.1
+NODE_PREBUILT_VERSION =	v6.17.0
 NODE_PREBUILT_TAG =	gz
 ifeq ($(shell uname -s),SunOS)
 NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
-- 
2.21.0

