commit 04bd767b4e6f0edeb51b68435e1f7e706b631aa6
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T14:49:37-05:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/Makefile b/Makefile
index fa4e8ef..0ea0047 100644
--- a/Makefile
+++ b/Makefile
@@ -37,7 +37,7 @@ JSSTYLE_FLAGS =		-o indent=4,doxygen,unparenthesized-return=0
 # doesn't know about (@@ENABLED@@)
 # SMF_MANIFESTS_IN = smf/manifests/cn-agent.xml.in
 
-NODE_PREBUILT_VERSION =	v6.15.1
+NODE_PREBUILT_VERSION =	v6.17.0
 NODE_PREBUILT_TAG =	gz
 ifeq ($(shell uname -s),SunOS)
 NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
