commit af4178a93a684e9b81264e32c195892dc4622082 (refs/changes/59/859/1)
Author: Cameron Watters <cameron.watters@joyent.com>
Date:   2016-11-08T09:02:13-07:00 (2 years, 11 months ago)
    
    DOCKER-967: properly order extraHosts entries in /etc/hosts

diff --git a/test/integration/cli-run-add-host.test.js b/test/integration/cli-run-add-host.test.js
index 2aff700..6096d75 100644
--- a/test/integration/cli-run-add-host.test.js
+++ b/test/integration/cli-run-add-host.test.js
@@ -66,7 +66,7 @@ test('docker run --add-host foo:1.2.3.4', function (tt) {
         var cmd = 'exec ' + containerName + ' grep foo /etc/hosts';
         cli.docker(cmd, function (err, stdout, stderr) {
             t.ifErr(err, 'docker CONTAINER grep foo /etc/hosts');
-            t.equal(stdout, 'foo\t1.2.3.4\n');
+            t.equal(stdout, '1.2.3.4\tfoo\n');
             t.end();
         });
     });
