commit 9a0f8d2380215535172b9298ff1f465aeb01e646 (refs/changes/53/1053/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-12-16T18:38:29+00:00 (2 years, 10 months ago)
    
    MORAY-363 Validate index names

diff --git a/lib/buckets/common.js b/lib/buckets/common.js
index fdc8d75..9898366 100644
--- a/lib/buckets/common.js
+++ b/lib/buckets/common.js
@@ -40,8 +40,24 @@ var INDEX_TYPES = {
 // <= 63 characters in length
 var BUCKET_NAME_RE = /^[a-zA-Z]\w{0,62}$/;
 
+var INDEX_NAME_RE =
+    /^(_[a-zA-Z0-9]+|[a-zA-Z][_a-zA-Z0-9]*[a-zA-Z0-9]|[a-zA-Z])$/;
+
 var RESERVED_BUCKETS = ['moray', 'search'];
 
+var RESERVED_INDEXES = [
+    '_etag',
+    '_id',
+    '_key',
+    '_atime',
+    '_ctime',
+    '_mtime',
+    '_rver',
+    '_txn_snap',
+    '_value',
+    '_vnode'
+];
+
 
 ///--- API
 
@@ -131,6 +147,23 @@ function validateIndexes(schema) {
                                                'objects');
         }
 
+        if (!INDEX_NAME_RE.test(k)) {
+            throw new InvalidBucketConfigError(
+                'invalid index name "' + k + '"');
+        }
+
+        var kLC = k.toLowerCase();
+
+        if (RESERVED_INDEXES.indexOf(kLC) !== -1) {
+            throw new InvalidBucketConfigError(
+                'index name "' + k + '" is reserved for use by Moray');
+        }
+
+        if (/^_*moray/.test(kLC)) {
+            throw new InvalidBucketConfigError(
+                'index names cannot start with "moray": "' + k + '"');
+        }
+
         sub = schema[k];
         subKeys = Object.keys(sub);
         for (j = 0; j < subKeys.length; j++) {
