commit cae0fbeb551c8dce7c3154b3e8b92605b6fff64b
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-08-23T14:34:52+00:00 (6 weeks ago)
    
    OS-7966 smartos-live configure should no longer download Studio 12

diff --git a/README.md b/README.md
index 8784a3ec..159ebf59 100644
--- a/README.md
+++ b/README.md
@@ -432,7 +432,6 @@ If this file does not exist, the following defaults are set by `configure`:
 ```
 PUBLISHER="joyent"
 RELEASE_VER="joyent_147"
-SUNW_SPRO12_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/SunStudio.tar.bz2"
 ON_CLOSED_BINS_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/illumos/on-closed-bins.i386.tar.bz2"
 ON_CLOSED_BINS_ND_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/illumos/on-closed-bins-nd.i386.tar.bz2"
 ILLUMOS_ADJUNCT_TARBALL_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/adjuncts/illumos-adjunct.20130131.tgz"
diff --git a/configure b/configure
index ecfa0135..a22f37e0 100755
--- a/configure
+++ b/configure
@@ -263,18 +263,6 @@ function install_packages
 	[[ $? -eq 0 ]] || fatal "failed to install packages"
 }
 
-function fetch_studio
-{
-	[[ -f "/opt/SUNWspro/prod/bin/cc" ]] && return
-	[[ -z "$SUNW_SPRO12_URL" ]] && fatal \
-	    "SUNW_SPRO12_URL not set in configure.*"
-	curl "$SUNW_SPRO12_URL" | bzcat | $conf_priv tar xf - -C /opt
-	[[ $? -eq 0 ]] || fatal "failed to fetch Studio"
-	[[ -e /opt/SUNspro/sunstudio12.1 ]] && return
-	$conf_priv ln -s /opt/SUNWspro /opt/SUNWspro/sunstudio12.1 || fatal \
-	    "cannot create Studio symlink"
-}
-
 function fetch_adjuncts
 {
 	local tgz
@@ -542,9 +530,6 @@ echo "done."
 echo "Installing packages ... \c "
 install_packages
 echo "done."
-echo "Fetching studio... \c "
-fetch_studio
-echo "done."
 echo "Fetching adjuncts tgz... \c "
 fetch_adjuncts
 echo "done."
diff --git a/default.configure-build b/default.configure-build
index 2e55b740..74ef484f 100644
--- a/default.configure-build
+++ b/default.configure-build
@@ -1,6 +1,5 @@
 PUBLISHER="joyent"
 RELEASE_VER="joyent_147"
-SUNW_SPRO12_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/SunStudio.tar.bz2"
 ON_CLOSED_BINS_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/illumos/on-closed-bins.i386.tar.bz2"
 ON_CLOSED_BINS_ND_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/illumos/on-closed-bins-nd.i386.tar.bz2"
 ILLUMOS_ADJUNCT_TARBALL_URL="https://us-east.manta.joyent.com/Joyent_Dev/public/releng/adjuncts/illumos-adjunct.20130131.tgz"
