From a84c21c62a83a279853288875c5b11aab6d64c1c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 9 Aug 2019 15:52:52 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index 3bdc30e..9058562 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 3bdc30e8bf3e10e6b6b283d216806d8e5d333c1a
+Subproject commit 9058562f2f59892795e470ce25443ff636c90478
-- 
2.21.0

