commit e718631d1232971137a3f39a8e7dca0b2beffe6e (refs/changes/94/1494/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2017-02-10T16:51:37-05:00 (2 years, 8 months ago)
    
    OS-5955 make check should fail on cstyle

diff --git a/src/Makefile b/src/Makefile
index acd26880..bd81a5fa 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -540,6 +540,7 @@ check: $(JSLINT) $(SUBDIRS)
 	@(for file in $(C_CHECK_TARGETS); do \
 		echo $(PWD)/$$file; \
 		$(CSTYLE) $$file; \
+		[[ $$? == "0" ]] || exit 1; \
 	done)
 	@printf "\nC files ok!\n"
 	@printf "\n==> Running JavaScriptLint...\n"
