From e718631d1232971137a3f39a8e7dca0b2beffe6e Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Fri, 10 Feb 2017 16:51:37 -0500
Subject: [PATCH] OS-5955 make check should fail on cstyle

---
 src/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/Makefile b/src/Makefile
index acd26880..bd81a5fa 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -540,6 +540,7 @@ check: $(JSLINT) $(SUBDIRS)
 	@(for file in $(C_CHECK_TARGETS); do \
 		echo $(PWD)/$$file; \
 		$(CSTYLE) $$file; \
+		[[ $$? == "0" ]] || exit 1; \
 	done)
 	@printf "\nC files ok!\n"
 	@printf "\n==> Running JavaScriptLint...\n"
-- 
2.21.0

