From 3cbb9a5b3dd171a470d355ae7e9ff4fb3c2daa24 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 15 Nov 2016 17:11:15 +0000
Subject: [PATCH] OS-5793 illumos build failures point to wrong files

---
 tools/build_illumos | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/build_illumos b/tools/build_illumos
index b84389d1..c362cb80 100755
--- a/tools/build_illumos
+++ b/tools/build_illumos
@@ -65,8 +65,8 @@ time ./nightly illumos.sh
 status=$?
 if [[ $status -ne 0 ]]; then
 	log_dir=$(ls -1t ./log | head -n1)
-	mail_msg=$PWD/$log_dir/mail_msg
-	nightly=$PWD/$log_dir/nightl.log
+	mail_msg=$PWD/log/$log_dir/mail_msg
+	nightly=$PWD/log/$log_dir/nightly.log
 
 	#
 	# Note, the blank lines at the start and end of the heredoc are
-- 
2.21.0

