commit 3cbb9a5b3dd171a470d355ae7e9ff4fb3c2daa24 (refs/changes/04/904/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-11-15T17:11:15+00:00 (2 years, 11 months ago)
    
    OS-5793 illumos build failures point to wrong files

diff --git a/tools/build_illumos b/tools/build_illumos
index b84389d1..c362cb80 100755
--- a/tools/build_illumos
+++ b/tools/build_illumos
@@ -65,8 +65,8 @@ time ./nightly illumos.sh
 status=$?
 if [[ $status -ne 0 ]]; then
 	log_dir=$(ls -1t ./log | head -n1)
-	mail_msg=$PWD/$log_dir/mail_msg
-	nightly=$PWD/$log_dir/nightl.log
+	mail_msg=$PWD/log/$log_dir/mail_msg
+	nightly=$PWD/log/$log_dir/nightly.log
 
 	#
 	# Note, the blank lines at the start and end of the heredoc are
