commit 6e5b135bbe3dbcb903ba28b874b3ef5c186b730c
Author: Kody A Kantor <kody@kkantor.com>
Date:   2019-04-09T18:35:05+00:00 (6 months ago)
    
    OS-7709 metaslab load/unload counter kstats leak memory

diff --git a/usr/src/uts/common/fs/zfs/metaslab.c b/usr/src/uts/common/fs/zfs/metaslab.c
index bcc705ac9e..90dfc48678 100644
--- a/usr/src/uts/common/fs/zfs/metaslab.c
+++ b/usr/src/uts/common/fs/zfs/metaslab.c
@@ -23,6 +23,7 @@
  * Copyright (c) 2011, 2018 by Delphix. All rights reserved.
  * Copyright (c) 2013 by Saso Kiselkov. All rights reserved.
  * Copyright (c) 2014 Integros [integros.com]
+ * Copyright 2019 Joyent, Inc.
  */
 
 #include <sys/zfs_context.h>
@@ -850,7 +851,9 @@ metaslab_group_passivate(metaslab_group_t *mg)
 	mg->mg_next = NULL;
 
 	if (mg->mg_kstat != NULL) {
+		metaslab_group_kstat_t *data = mg->mg_kstat->ks_data;
 		kstat_delete(mg->mg_kstat);
+		kmem_free(data, sizeof (metaslab_group_kstat_t));
 	}
 	mutex_destroy(&mg->mg_kstat_lock);
 }
