From 9206f31b57d0bfc70b403dc7d6fdd59a4473995c Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 14 Dec 2018 10:54:26 -0800
Subject: [PATCH] TRITON-1044 cns should use newer redis server and client,
 paginate AXFRs

---
 targets.json.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 8e33820..2e41d6d 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -181,9 +181,9 @@ cat <<EOF
     "image_name": "cns",
     "image_description": "Triton Container Naming Service",
     "image_version": "1.0.0",
-    "image_uuid": "b4bdc598-8939-11e3-bea4-8341f6861379",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
     "pkgsrc": [
-      "redis-2.6.16"
+      "redis-3.0.5"
     ],
     "repos": [
         {"url": "$MG_GIT_PREFIX/triton-cns.git", "dir": "triton-cns"}
-- 
2.21.0

