commit 38d322aa04bb0758bd70ae6f3ba3ecfa732d124e (refs/changes/16/616/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-10-06T17:46:54-04:00 (3 years ago)
    
    RELENG-612 prefer (git+)https urls

diff --git a/.gitmodules b/.gitmodules
index 287c42a..00d5b76 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "deps/eng"]
 	path = deps/eng
-	url = git@github.com:joyent/eng.git
+	url = https://github.com/joyent/eng.git
 [submodule "deps/manta-scripts"]
 	path = deps/manta-scripts
-	url = git@github.com:joyent/manta-scripts.git
+	url = https://github.com/joyent/manta-scripts.git
diff --git a/package.json b/package.json
index 458c89e..a1a4b15 100644
--- a/package.json
+++ b/package.json
@@ -8,7 +8,7 @@
 		"assert-plus": "^1.0.0",
 		"bunyan": "^1.6.0",
 		"manta": "^2.0.5",
-		"moray": "git+ssh://git@github.com:joyent/node-moray.git#master",
+		"moray": "git+https://github.com/joyent/node-moray.git#master",
 		"once": "1.1.1",
 		"pty.js": "0.3.0",
 		"restify": "^2.8.5",
