commit ed0bf1b6ccc6fed71077790289aca6e421f3bb1e (refs/changes/08/1008/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    instance {start,stop,reboot,delete}

diff --git a/lib/do_instance/gen_do_ACTION.js b/lib/do_instance/gen_do_ACTION.js
index 1e86fc7..d39f254 100644
--- a/lib/do_instance/gen_do_ACTION.js
+++ b/lib/do_instance/gen_do_ACTION.js
@@ -83,8 +83,6 @@ function gen_do_ACTION(opts) {
 function _doTheAction(action, subcmd, opts, args, callback) {
     var self = this;
 
-    var now = Date.now();
-
     var command, state;
     switch (action) {
         case 'start':
@@ -116,7 +114,17 @@ function _doTheAction(action, subcmd, opts, args, callback) {
         callback(new errors.UsageError('missing INST arg(s)'));
         return;
     }
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            callback(setupErr);
+        }
+        _doOnEachInstance(self, action, command, state, args, opts, callback);
+    });
+}
 
+function _doOnEachInstance(self, action, command, state, instances,
+                           opts, callback) {
+    var now = Date.now();
     vasync.forEachParallel({
         func: function (arg, cb) {
             var alias, uuid;
@@ -190,7 +198,7 @@ function _doTheAction(action, subcmd, opts, args, callback) {
                 });
             }
         },
-        inputs: args
+        inputs: instances
     }, function (err, results) {
         var e = err ? (new Error('command failure')) : null;
         callback(e);
