commit 71e791bc51d46e3f124a3fc05a346fbfb6a286e5 (refs/changes/50/2950/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2017-11-15T14:47:43-08:00 (1 year, 11 months ago)
    
    joyent/manta-thoth#153 Want build release

diff --git a/.gitignore b/.gitignore
index c2658d7..2ab6019 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1 +1,2 @@
 node_modules/
+manta-thoth-*.tar.gz
diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..5cc98e6
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,19 @@
+UNAME=$(shell uname -s)
+VER=$(shell json -f package.json version)
+GH_USER=$(git config github.user)
+
+node_modules: package.json
+	npm install --production
+
+.PHONY: release github-release
+release: manta-thoth-$(UNAME)-$(VER).tar.gz
+
+publish: release
+	./tools/publish.sh "$(VER)"
+
+manta-thoth-$(UNAME)-$(VER).tar.gz: node_modules
+	./tools/build-release.sh
+
+clean:
+	rm -rf node_modules
+	rm -f manta-thoth-$(UNAME)-$(VER).tar.gz
diff --git a/tools/build-release.sh b/tools/build-release.sh
new file mode 100755
index 0000000..1b77472
--- /dev/null
+++ b/tools/build-release.sh
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -o errexit
+
+outdir=$(mktemp -d -t manta-thoth-XXXXXX)
+ver=$(json -f package.json version)
+uname=$(uname -s)
+tar="${outdir}/manta-thoth-${uname}-${ver}.tar.gz"
+
+tar zcf "$tar" "../$(basename $PWD)"
+mv "$tar" .
+rmdir "$outdir"
diff --git a/tools/publish.sh b/tools/publish.sh
new file mode 100755
index 0000000..9a8cc84
--- /dev/null
+++ b/tools/publish.sh
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+ver="$1"
+uname=$(uname -s)
+name="manta-thoth-${uname}-${ver:?}.tar.gz"
+latest="manta-thoth-${uname}-latest.tar.gz"
+
+mput -f "$name" "/thoth/public/$name"
+mln "/thoth/public/$name" "/thoth/public/$latest"
