From 8c4eb975925d6b5d365a0fbea2b3a842e7d0b38b Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Mon, 27 May 2019 11:14:44 +0200
Subject: [PATCH] TRITON-1702 TRITON-1321 fwapi log boolean rules broke tests

---
 Makefile             | 7 +++++--
 deps/fw/package.json | 4 ++--
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 0ced4f2..ac34d1b 100644
--- a/Makefile
+++ b/Makefile
@@ -41,8 +41,11 @@ endif
 ifeq ($(shell uname -s),SunOS)
 	include ./deps/eng/tools/mk/Makefile.node_prebuilt.defs
 else
-	NPM_EXEC :=
-	NPM = npm
+	# Good enough for non-SmartOS dev.
+	NPM=npm
+	NODE=node
+	NPM_EXEC=$(shell which npm)
+	NODE_EXEC=$(shell which node)
 endif
 include ./deps/eng/tools/mk/Makefile.smf.defs
 
diff --git a/deps/fw/package.json b/deps/fw/package.json
index 9131120..7e828ea 100644
--- a/deps/fw/package.json
+++ b/deps/fw/package.json
@@ -1,7 +1,7 @@
 {
   "name": "fw",
   "description": "Administrative tool for managing SmartOS VM firewalls",
-  "version": "1.1.1",
+  "version": "1.2.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "main": "./lib/fw.js",
@@ -16,7 +16,7 @@
     "clone": "0.1.4",
     "cmdln": "4.1.1",
     "extsprintf": "1.0.2",
-    "fwrule": "2.0.0",
+    "fwrule": "2.1.0",
     "ip6addr": "0.2.2",
     "mkdirp": "0.3.4",
     "uuid": "3.2.1",
-- 
2.21.0

