From 608697149c8f141837d28fec707949386f52216b Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 27 Mar 2018 17:01:43 -0700
Subject: [PATCH] TRITON-282 drop obsolete gsed'ing of IMGAPI SMF manifest at
 zone setup

---
 boot/setup.sh | 1 -
 1 file changed, 1 deletion(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index 565a253..8f1fb0c 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -40,7 +40,6 @@ if [[ ! -d $LOCAL_BASE_DIR ]]; then
 fi
 chown nobody:nobody $LOCAL_BASE_DIR
 
-$(/opt/local/bin/gsed -i"" -e "s/@@PREFIX@@/\/opt\/smartdc\/imgapi/g" /opt/smartdc/imgapi/smf/manifests/imgapi.xml)
 /usr/sbin/svccfg import /opt/smartdc/imgapi/smf/manifests/imgapi.xml
 
 # Log rotation.
-- 
2.21.0

