commit 1c7bb2e05be486680d7e9b21e4b8c4b4b3eed5fc
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2019-07-03T21:09:03+00:00 (3 months ago)
    
    TRITON-1794 firewaller's configuration migration SMF service cannot be imported

diff --git a/Makefile b/Makefile
index d5d17f3..ce8e692 100644
--- a/Makefile
+++ b/Makefile
@@ -18,7 +18,7 @@
 
 
 BASH_FILES  := npm/postinstall.sh npm/postuninstall.sh
-JS_FILES        := $(shell find lib test -name '*.js') main.js
+JS_FILES        := $(shell find lib test -name '*.js') $(wildcard ./*.js)
 JSON_FILES       = package.json config.json
 JSL_CONF_NODE    = tools/jsl.node.conf
 JSL_FILES_NODE   = $(JS_FILES)
@@ -101,6 +101,7 @@ release: all docs $(SMF_MANIFESTS)
 	(git symbolic-ref HEAD | awk -F/ '{print $$3}' && git describe) > $(DSTDIR)/describe
 	cp -r \
     $(TOP)/config.json \
+    $(TOP)/config-migration.js \
     $(TOP)/main.js \
     $(TOP)/lib \
     $(TOP)/node_modules \
diff --git a/config-migration.js b/config-migration.js
index 3c6ee11..c889c8d 100644
--- a/config-migration.js
+++ b/config-migration.js
@@ -185,6 +185,9 @@ vasync.pipeline({
 }, function pipeCb(pipeErr) {
     if (pipeErr) {
         console.error(pipeErr);
+        process.exit(1);
+    } else {
+        process.exit(0);
     }
 });
 
diff --git a/smf/manifests/firewaller-config-migration.xml.in b/smf/manifests/firewaller-config-migration.xml.in
index a7f5982..ebb8c96 100644
--- a/smf/manifests/firewaller-config-migration.xml.in
+++ b/smf/manifests/firewaller-config-migration.xml.in
@@ -20,15 +20,18 @@
       <service_fmri value="svc:/network/physical"/>
       <service_fmri value="svc:/system/filesystem/local"/>
     </dependency>
+    <dependency name="vminfod_when_present" grouping="optional_all" restart_on="error" type="service">
+      <service_fmri value="svc:/system/smartdc/vminfod"/>
+    </dependency>
     <!--
-        Make vmadmd and zones services depend on this service so that firewall
-        rules are correct before those services cause them to be read.
+        We make the zones service depend on this service so that the firewall
+        rules are correct before the zones (and other services like vmadmd)
+        cause them to be read.
     -->
-    <dependent name="smartdc_vmadmd_and_system_zones" grouping="optional_all" restart_on="none">
-      <service_fmri value="svc:/system/smartdc/vmadmd:default"/>
+    <dependent name="system_zones" grouping="optional_all" restart_on="none">
       <service_fmri value="svc:/system/zones:default"/>
     </dependent>
-    <exec_method type="method" name="start" exec="/usr/bin/ctrun -l child -o noorphan @@ROOT@@/node/bin/node --abort_on_uncaught_exception @@ROOT@@/config-migration.js &amp;" timeout_seconds="60">
+    <exec_method type="method" name="start" exec="/usr/bin/ctrun -l child -o noorphan @@ROOT@@/node/bin/node --abort_on_uncaught_exception @@ROOT@@/config-migration.js" timeout_seconds="0">
       <method_context>
         <method_credential user="root" group="staff"/>
         <method_environment>
@@ -36,16 +39,14 @@
         </method_environment>
       </method_context>
     </exec_method>
-    <exec_method type="method" name="restart" exec=":kill" timeout_seconds="60">
-      <method_context>
-        <method_credential user="root" group="staff"/>
-      </method_context>
-    </exec_method>
-    <exec_method type="method" name="stop" exec=":kill" timeout_seconds="60">
+    <exec_method type="method" name="stop" exec=":true" timeout_seconds="60">
       <method_context>
         <method_credential user="root" group="staff"/>
       </method_context>
     </exec_method>
+    <property_group name='startd' type='framework'>
+      <propval name='duration' type='astring' value='transient' />
+    </property_group>
     <instance name="default" enabled="true"/>
     <stability value="Unstable"/>
     <template>
