From 9b280495e36fd5d3fe67149fefec2df8e32a4cad Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike.zeller@joyent.com>
Date: Tue, 2 Jan 2018 16:09:45 -0800
Subject: [PATCH] ZAPI-820 Cleanup error messages in network validation

---
 lib/common/validation.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/common/validation.js b/lib/common/validation.js
index 547fd47..d6cb5f9 100644
--- a/lib/common/validation.js
+++ b/lib/common/validation.js
@@ -2321,7 +2321,7 @@ exports.validateNetworks = function (params) {
     }
 
     if (errs.length) {
-        throw new errors.ValidationFailedError('Invalid VM update parameters',
+        throw new errors.ValidationFailedError('Invalid network parameters',
             errs);
     }
 
@@ -2553,4 +2553,4 @@ function hasPolymorphicParamWithName(fieldName, params) {
 
     return false;
 }
-exports.hasPolymorphicParamWithName = hasPolymorphicParamWithName;
\ No newline at end of file
+exports.hasPolymorphicParamWithName = hasPolymorphicParamWithName;
-- 
2.21.0

