commit 9b280495e36fd5d3fe67149fefec2df8e32a4cad (refs/changes/50/3150/1)
Author: Mike Zeller <mike.zeller@joyent.com>
Date:   2018-01-02T16:09:45-08:00 (1 year, 9 months ago)
    
    ZAPI-820 Cleanup error messages in network validation

diff --git a/lib/common/validation.js b/lib/common/validation.js
index 547fd47..d6cb5f9 100644
--- a/lib/common/validation.js
+++ b/lib/common/validation.js
@@ -2321,7 +2321,7 @@ exports.validateNetworks = function (params) {
     }
 
     if (errs.length) {
-        throw new errors.ValidationFailedError('Invalid VM update parameters',
+        throw new errors.ValidationFailedError('Invalid network parameters',
             errs);
     }
 
@@ -2553,4 +2553,4 @@ function hasPolymorphicParamWithName(fieldName, params) {
 
     return false;
 }
-exports.hasPolymorphicParamWithName = hasPolymorphicParamWithName;
\ No newline at end of file
+exports.hasPolymorphicParamWithName = hasPolymorphicParamWithName;
