commit 2504a8340ba52ba277498c5d36eed824126112cf (refs/changes/58/2958/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-11-16T15:33:43-08:00 (1 year, 11 months ago)
    
    joyent/node-watershed#2 Use crypto.randomBytes() for key generation

diff --git a/lib/watershed.js b/lib/watershed.js
index fd58fdc..ade164e 100644
--- a/lib/watershed.js
+++ b/lib/watershed.js
@@ -113,11 +113,7 @@ Watershed()
 Watershed.prototype.generateKey = function
 generateKey()
 {
-	var nonce = new Buffer(NONCE_LENGTH);
-	for (var i = 0; i < nonce.length; i++) {
-		nonce[i] = Math.floor(Math.random() * 256);
-	}
-	return (nonce.toString('base64'));
+	return (crypto.randomBytes(NONCE_LENGTH).toString('base64'));
 };
 
 /*
@@ -551,10 +547,7 @@ _ws_writeFrameCommon(opcode, data)
 	 * According to the RFC, the client MUST mask their outgoing frames.
 	 */
 	if (this._options.localShouldMask) {
-		maskbuf = new Buffer(4);
-		for (var i = 0; i < maskbuf.length; i++) {
-			maskbuf[i] = Math.floor(Math.random * 256);
-		}
+		maskbuf = crypto.randomBytes(4);
 		for (var j = 0; j < data.length; j++) {
 			data[j] = data[j] ^ maskbuf[j % maskbuf.length];
 		}
