commit 48a865e41d503edfa9c2d442ca4e86450c908eb6 (refs/changes/26/2126/2)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-07-17T15:05:29-07:00 (2 years, 3 months ago)
    
    OS-6192 `pigz` (parallel gzip) support in the platform

diff --git a/Makefile b/Makefile
index 78e49f8..3c3b703 100644
--- a/Makefile
+++ b/Makefile
@@ -59,6 +59,7 @@ SUBDIRS = \
 	openssh \
 	pbzip2 \
 	perl \
+	pigz \
 	rsync \
 	rsyslog \
 	screen \
@@ -117,6 +118,7 @@ curl: libz openssl1x libidn2
 gzip: libz
 node.js: openssl1x
 dialog: ncurses
+pigz: libz
 socat: openssl1x
 wget: openssl1x libidn
 openldap: openssl1x
diff --git a/manifest b/manifest
index 3399611..ac422c5 100644
--- a/manifest
+++ b/manifest
@@ -2987,6 +2987,8 @@ s usr/perl5/bin/perl=../5.12/bin/perl
 d usr/perl5/site_perl 0755 root bin
 d usr/perl5/site_perl/5.12 0755 root bin
 d usr/perl5/site_perl/5.12/i86pc-solaris-64int 0755 root bin
+# pigz
+f usr/img/sbin/pigz 0555 root bin
 # rsync
 f usr/bin/rsync 0555 root bin
 f usr/share/man/man1/rsync.1 0444 root bin
diff --git a/pigz/Makefile b/pigz/Makefile
new file mode 100644
index 0000000..723eb66
--- /dev/null
+++ b/pigz/Makefile
@@ -0,0 +1,32 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright (c) 2017, Joyent, Inc.
+#
+
+VER =	pigz-2.3.4
+
+include ../Makefile.defs
+
+BINDIR =	$(DESTDIR)/usr/img/sbin/
+
+PATCHES = \
+	pigz-configure.patch \
+	pigz-Makefile.patch
+
+include ../Makefile.targ
+
+all: all_autoconf
+
+install: all
+	mkdir -p $(BINDIR)
+	ginstall -m 0555 $(BASE)/$(VER.32)/pigz $(BINDIR)/pigz
diff --git a/pigz/pigz-2.3.4.tar.gz b/pigz/pigz-2.3.4.tar.gz
new file mode 100644
index 0000000..65fa66a
Binary files /dev/null and b/pigz/pigz-2.3.4.tar.gz differ
diff --git a/pigz/pigz-Makefile.patch b/pigz/pigz-Makefile.patch
new file mode 100644
index 0000000..48ab00b
--- /dev/null
+++ b/pigz/pigz-Makefile.patch
@@ -0,0 +1,9 @@
+--- pigz-2.3.4/Makefile	2016-05-29 22:09:41.000000000 -0700
++++ pigz-2.3.4/Makefile	2017-06-14 17:04:57.000000000 -0700
+@@ -1,5 +1,6 @@
+ CC=cc
+ CFLAGS=-O3 -Wall -Wextra
++CFLAGS+=-std=gnu99
+ LDFLAGS=
+ LIBS=-lm -lpthread -lz
+ ZOPFLI=zopfli/src/zopfli/
diff --git a/pigz/pigz-configure.patch b/pigz/pigz-configure.patch
new file mode 100644
index 0000000..750f7ae
--- /dev/null
+++ b/pigz/pigz-configure.patch
@@ -0,0 +1,6 @@
+This patch creates an empty pigz configure file, as the illumos-extra build
+system expects a configure file to exist in each component subdirectory.
+exist the illumos-extra build system to function correctly.
+diff --git 2.3.4/configure 2.3.4/configure
+new file mode 100644
+index 0000000..e69de29
