commit cf8d93c364beb4a958c7986f282a392833ae5523 (refs/changes/44/2144/5)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-06-26T21:48:48+00:00 (2 years, 3 months ago)
    
    DOCKER-1072 Work around OS-6193 in tools/check-docs.sh

diff --git a/deps/javascriptlint b/deps/javascriptlint
index e1bd0ab..ad52812 160000
--- a/deps/javascriptlint
+++ b/deps/javascriptlint
@@ -1 +1 @@
-Subproject commit e1bd0abfd424811af469d1ece3af131d95443924
+Subproject commit ad52812e77bdfb1e90fb71a1201adb2b665a27e6
diff --git a/tools/check-docs.sh b/tools/check-docs.sh
index cbba338..4c44f16 100755
--- a/tools/check-docs.sh
+++ b/tools/check-docs.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -35,12 +35,11 @@ TOP=$(cd $(dirname $0)/../; pwd)
 #---- mainline
 
 hits=0
-for file in $(find $TOP/docs -name "*.md"); do
-    lastchar=$(tail -c1 $file | od -a | head -1 | awk '{print $2}')
-    if [[ $lastchar != 'nl' ]]; then
+while read -d $'\0' -r file; do
+    if [[ `tail -1c "$file" | tr ' \n' '_ '` != ' ' ]]; then
         echo "$file: does not end with a newline" >&2
-        hits=$(( $hits + 1 ))
+        hits=1
     fi
-done
+done < <(find $TOP/docs -name "*.md" -print0)
 
 exit $hits
