commit e386a9d90417b8ed898a9880411c854ba5de1a4a
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2019-03-06T16:04:34-08:00 (7 months ago)
    
    joyent/node-triton#258 triton picks disabled image

diff --git a/CHANGES.md b/CHANGES.md
index e171277..65145de 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -6,6 +6,11 @@ Known issues:
 
 ## not yet released
 
+## 6.4.0
+
+- [joyent/node-triton#258] Getting images by name now returns the newest
+  *active* image, rather than the newest image.
+
 ## 6.3.0
 
 - [joyent/node-triton#259] Added basic support for use of SSH bastion hosts
diff --git a/lib/tritonapi.js b/lib/tritonapi.js
index 6e26c1d..91cd0d7 100644
--- a/lib/tritonapi.js
+++ b/lib/tritonapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /* BEGIN JSSTYLED */
@@ -788,8 +788,25 @@ TritonApi.prototype.getImage = function getImage(opts, cb) {
             if (nameMatches.length === 1) {
                 cb(null, nameMatches[0]);
             } else if (nameMatches.length > 1) {
-                tabula.sortArrayOfObjects(nameMatches, 'published_at');
-                cb(null, nameMatches[nameMatches.length - 1]);
+                /**
+                 *
+                 * If we've reached here, the following conditions are met:
+                 *
+                 * - the user has specified an image by name
+                 * - there are multiple matching images
+                 *
+                 * In this case, we want to filter disabled images because
+                 * it is most likely they are searching for a "suitable"
+                 * image, for some value of "suitable". In almost any case
+                 * returning an image in any state other than active will
+                 * result in sadness.
+                 *
+                 **/
+                var activeNameMatches = nameMatches.filter(function (o) {
+                    return o.state === "active";
+                });
+                tabula.sortArrayOfObjects(activeNameMatches, 'published_at');
+                cb(null, activeNameMatches[activeNameMatches.length - 1]);
             } else if (shortIdMatches.length === 1) {
                 cb(null, shortIdMatches[0]);
             } else if (shortIdMatches.length === 0) {
diff --git a/package.json b/package.json
index 78ca693..c61bac1 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "triton",
   "description": "Joyent Triton CLI and client (https://www.joyent.com/triton)",
-  "version": "6.3.0",
+  "version": "6.4.0",
   "author": "Joyent (joyent.com)",
   "homepage": "https://github.com/joyent/node-triton",
   "dependencies": {
