commit d98af348f84863d9468b8d7cee73536b06551ef0
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2019-03-06T17:10:14-08:00 (7 months ago)
    
    joyent/node-triton#258 triton picks disabled image

diff --git a/CHANGES.md b/CHANGES.md
index e171277..daef950 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -6,6 +6,10 @@ Known issues:
 
 ## not yet released
 
+## 6.3.1
+
+- [joyent/node-triton#258] `triton create` now will only select active images.
+
 ## 6.3.0
 
 - [joyent/node-triton#259] Added basic support for use of SSH bastion hosts
diff --git a/lib/do_instance/do_create.js b/lib/do_instance/do_create.js
index cbedb35..9bf8c45 100644
--- a/lib/do_instance/do_create.js
+++ b/lib/do_instance/do_create.js
@@ -203,6 +203,7 @@ function do_create(subcmd, opts, args, cb) {
         function getImg(ctx, next) {
             var _opts = {
                 name: args[0],
+                state: 'active',
                 useCache: true
             };
             tritonapi.getImage(_opts, function (err, img) {
diff --git a/lib/tritonapi.js b/lib/tritonapi.js
index 6e26c1d..da94a4b 100644
--- a/lib/tritonapi.js
+++ b/lib/tritonapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /* BEGIN JSSTYLED */
@@ -757,7 +757,7 @@ TritonApi.prototype.getImage = function getImage(opts, cb) {
 
         var listOpts = {
             // Explicitly include inactive images.
-            state: 'all'
+            state: opts.state || 'all'
         };
         if (version) {
             nameSelector = name + '@' + version;
diff --git a/package.json b/package.json
index 78ca693..2392437 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "triton",
   "description": "Joyent Triton CLI and client (https://www.joyent.com/triton)",
-  "version": "6.3.0",
+  "version": "6.3.1",
   "author": "Joyent (joyent.com)",
   "homepage": "https://github.com/joyent/node-triton",
   "dependencies": {
