From 700ef93aad710428add4246739e2cdf6b3104aa6 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Tue, 17 Oct 2017 00:35:57 +0000
Subject: [PATCH] MORAY-446 apply connection limit to "moray" role

---
 boot/setup.sh | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index 7e2d88e..6838125 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -204,6 +204,19 @@ function sdc_moray_setup {
 
 }
 
+function set_moray_role_connlimit {
+    local reserve_conns=32
+    local primary_ip=$0
+    local pg_max_conns=$(psql -U postgres -h $primary_ip -p 5432 \
+        -c 'SHOW  max_connections' | tail +3 | head -1)
+    [[ $? -eq 0 && $pg_max_conns =~ '^[0-9]+$' ]] || fatal \
+        "Unable to retrieve postgres max_connections"
+
+    psql -U postgres -h $primary_ip -p 5432 -c "ALTER ROLE $PG_USER " + \
+        "WITH CONNECTION LIMIT $(expr $pg_max_conns - $reserve_conns)"
+    [[ $? -eq 0 ]] || fatal "Unable to set moray role connection limit"
+}
+
 function manta_setup_moray_config {
     #.bashrc
     echo 'function req() { grep "$@" `svcs -L moray` | bunyan ;}' >> $PROFILE
@@ -223,6 +236,7 @@ function manta_setup_moray_config {
     # exists, so we don't check error, subsequent pg requests will fail with
     # this user if it dne.
     createuser -U postgres -h $primary_ip -p 5432 -d -S -R $PG_USER
+    set_moray_role_connlimit $primary_ip
 
     # Postgres sucks at return codes, so we basically have no choice but to
     # ignore the error code here since we can't conditionally create the DB
@@ -301,6 +315,7 @@ function sdc_moray_createdb {
       -h ${POSTGRES_HOST} ${role} -c "\du"|grep ${role}) ]]; then
       echo "User ${role} does not exist. Creating it"
       /opt/local/bin/createuser -U postgres -h ${POSTGRES_HOST} -d -S -R $PG_USER
+      set_moray_role_connlimit ${POSTGRES_HOST}
     else
       echo "User ${role} already exists."
     fi
-- 
2.21.0

