From fbe75be27edab45b73379c4abcb1d21b8e84aff8 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Tue, 17 Oct 2017 00:35:57 +0000
Subject: [PATCH] MORAY-446 apply connection limit to "moray" role

---
 boot/setup.sh | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index 7e2d88e..e21c2fd 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -204,6 +204,17 @@ function sdc_moray_setup {
 
 }
 
+function manta_set_moray_role_connlimit {
+    local reserve_conns=32
+    local primary_ip=$1
+    local pg_max_conns=$(psql -U postgres -h $primary_ip -p 5432 \
+        -c 'SHOW max_connections' | tail +3 | head -1 | tr -d '[:space:]')
+    [[ $? -eq 0 ]] || fatal "Unable to retrieve postgres max_connections"
+
+    psql -U postgres -h $primary_ip -p 5432 -c "ALTER ROLE $PG_USER WITH CONNECTION LIMIT $(expr $pg_max_conns - $reserve_conns)"
+    [[ $? -eq 0 ]] || fatal "Unable to set moray role connection limit"
+}
+
 function manta_setup_moray_config {
     #.bashrc
     echo 'function req() { grep "$@" `svcs -L moray` | bunyan ;}' >> $PROFILE
@@ -223,6 +234,7 @@ function manta_setup_moray_config {
     # exists, so we don't check error, subsequent pg requests will fail with
     # this user if it dne.
     createuser -U postgres -h $primary_ip -p 5432 -d -S -R $PG_USER
+    manta_set_moray_role_connlimit $primary_ip
 
     # Postgres sucks at return codes, so we basically have no choice but to
     # ignore the error code here since we can't conditionally create the DB
-- 
2.21.0

