commit cc41b091f62a12be9efa3db0feae66390fd0af74 (refs/changes/89/4489/2)
Author: Jason King <jason.king@joyent.com>
Date:   2018-07-08T19:09:39+00:00 (1 year, 3 months ago)
    
    Fix VL2 refleak

diff --git a/usr/src/uts/common/io/overlay/overlay_target.c b/usr/src/uts/common/io/overlay/overlay_target.c
index a174a1f7a5..ef9a0b8f1a 100644
--- a/usr/src/uts/common/io/overlay/overlay_target.c
+++ b/usr/src/uts/common/io/overlay/overlay_target.c
@@ -1354,6 +1354,10 @@ overlay_target_lookup_respond_vl3(const overlay_targ_resp_t *otr,
 		vl2_entry->ote_flags |= OVERLAY_ENTRY_F_VALID;
 	}
 	mutex_exit(&vl2_entry->ote_lock);
+
+	mutex_enter(&ott->ott_lock);
+	sarc_rele(mhash, vl2_entry);
+	mutex_exit(&ott->ott_lock);
 }
 
 static int
