commit e4318349347a2c5e0b89e38f2eab6d46b005af1a (refs/changes/89/4489/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-07-05T21:14:11-05:00 (1 year, 3 months ago)
    
    Fix VL2 refleak

diff --git a/usr/src/uts/common/io/overlay/overlay_target.c b/usr/src/uts/common/io/overlay/overlay_target.c
index 66bb918ee3..cc9dd315e6 100644
--- a/usr/src/uts/common/io/overlay/overlay_target.c
+++ b/usr/src/uts/common/io/overlay/overlay_target.c
@@ -1354,6 +1354,10 @@ overlay_target_lookup_respond_vl3(const overlay_targ_resp_t *otr,
 		vl2_entry->ote_flags |= OVERLAY_ENTRY_F_VALID;
 	}
 	mutex_exit(&vl2_entry->ote_lock);
+
+	mutex_enter(&ott->ott_lock);
+	sarc_rele(mhash, vl2_entry);
+	mutex_exit(&ott->ott_lock);
 }
 
 static int
