From 01167cf507789b801051cd3875023c30a6f6ec95 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Wed, 21 Nov 2018 15:07:33 -0800
Subject: [PATCH] TRITON-766 cainstsvc ctf module relies on CTF type ordering

---
 deps/node-ctype | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/node-ctype b/deps/node-ctype
index 4d272d3..e5b7e26 160000
--- a/deps/node-ctype
+++ b/deps/node-ctype
@@ -1 +1 @@
-Subproject commit 4d272d324bd95834212d27376a2875d45e71aef1
+Subproject commit e5b7e2673e9009e3b43e3d0f96b0aec97d7a7b16
-- 
2.21.0

