commit 01167cf507789b801051cd3875023c30a6f6ec95 (refs/changes/94/5094/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-11-21T15:07:33-08:00 (11 months ago)
    
    TRITON-766 cainstsvc ctf module relies on CTF type ordering

diff --git a/deps/node-ctype b/deps/node-ctype
index 4d272d3..e5b7e26 160000
--- a/deps/node-ctype
+++ b/deps/node-ctype
@@ -1 +1 @@
-Subproject commit 4d272d324bd95834212d27376a2875d45e71aef1
+Subproject commit e5b7e2673e9009e3b43e3d0f96b0aec97d7a7b16
