commit 7b98f7e902f8a0736900b8df2ba9f13d59ec0089 (refs/changes/94/1794/2)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2017-04-13T23:01:36-07:00 (2 years, 6 months ago)
    
    TOOLS-1753 sdcnode inexplicably doing `npm install` before building node

diff --git a/Makefile b/Makefile
index 39a1b64..7359e06 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -36,7 +36,11 @@ all: build/src nodes bits
 
 .PHONY: npmdeps
 npmdeps:
+ifeq ($(SKIP_NPM),)
 	npm install
+else
+	/bin/true # skipping npm
+endif
 
 build/src:
 	git clone https://github.com/nodejs/node.git build/src
