From 7b98f7e902f8a0736900b8df2ba9f13d59ec0089 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <jwilsdon@joyent.com>
Date: Thu, 13 Apr 2017 23:01:36 -0700
Subject: [PATCH] TOOLS-1753 sdcnode inexplicably doing `npm install` before
 building node

---
 Makefile | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 39a1b64..7359e06 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -36,7 +36,11 @@ all: build/src nodes bits
 
 .PHONY: npmdeps
 npmdeps:
+ifeq ($(SKIP_NPM),)
 	npm install
+else
+	/bin/true # skipping npm
+endif
 
 build/src:
 	git clone https://github.com/nodejs/node.git build/src
-- 
2.21.0

