From 6486a3326181e958e5884b30065a1ef550a22f68 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <jwilsdon@joyent.com>
Date: Thu, 13 Apr 2017 23:03:20 -0700
Subject: [PATCH] TOOLS-1753 sdcnode inexplicably doing `npm install` before
 building node

---
 Makefile  | 6 +++++-
 README.md | 6 ++++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 39a1b64..7359e06 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -36,7 +36,11 @@ all: build/src nodes bits
 
 .PHONY: npmdeps
 npmdeps:
+ifeq ($(SKIP_NPM),)
 	npm install
+else
+	/bin/true # skipping npm
+endif
 
 build/src:
 	git clone https://github.com/nodejs/node.git build/src
diff --git a/README.md b/README.md
index 735b57b..621f899 100644
--- a/README.md
+++ b/README.md
@@ -37,6 +37,12 @@ To build all sdcnode distro configurations:
 Before commiting/pushing run `make prepush` and, if possible, get a code
 review.
 
+If you're doing a bootstrap build and don't already have a node use:
+
+    make SKIP_NPM=1 all
+
+instead of `make all` to avoid needing npm to build.
+
 
 # Dev Notes
 
-- 
2.21.0

