commit 897bffcb1fa1cf8937b202a1ac628efc5725f541 (refs/changes/66/2866/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-10-25T12:13:52-07:00 (1 year, 11 months ago)
    
    TOOLS-1898 sdcnode updates for Node.js Oct 2017 vulns

diff --git a/Makefile b/Makefile
index b901ca5..4ea3e41 100644
--- a/Makefile
+++ b/Makefile
@@ -33,7 +33,7 @@ SMF_MANIFESTS = $(shell ls smf/manifests/*.xml)
 NODEUNIT	:= ./node_modules/.bin/nodeunit
 CLEAN_FILES += ./node_modules
 
-NODE_PREBUILT_VERSION=v4.8.4
+NODE_PREBUILT_VERSION=v4.8.5
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
