From 329aa68bcba230039635949997d4bf189e1bf416 Mon Sep 17 00:00:00 2001
From: Rob Johnston <rob.johnston@joyent.com>
Date: Fri, 2 Feb 2018 03:55:39 +0000
Subject: [PATCH] OS-6583 add missing DTDs to manlint exception list

---
 exception_lists/manlint | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/exception_lists/manlint b/exception_lists/manlint
index 5982ac5cec..0d4ea0f464 100644
--- a/exception_lists/manlint
+++ b/exception_lists/manlint
@@ -10,10 +10,17 @@
 #
 #
 # Copyright 2016 Toomas Soome <tsoome@me.com>
+# Copyright (c) 2018, Joyent, Inc.
 #
 usr/src/boot/*
 # Not actually a manual page
+usr/src/cmd/hal/fdi/fdi.dtd.1
+usr/src/cmd/isns/isnsd/xml_def/isnsdata.dtd.1
 usr/src/cmd/svc/dtd/service_bundle.dtd.1
+usr/src/lib/fm/topo/maps/common/topology.dtd.1
+usr/src/lib/libbrand/dtd/brand.dtd.1
+usr/src/lib/libbrand/dtd/zone_platform.dtd.1
 usr/src/lib/libbsm/adt_record.dtd.1
 usr/src/lib/libbsm/adt_record.xsl.1
+usr/src/lib/libpool/dtd/rm_pool.dtd.1
 usr/src/lib/libzonecfg/dtd/zonecfg.dtd.1
-- 
2.21.0

