commit 436100affea5bc87b24ddba212ccfb199ce1cb9b (refs/changes/92/3292/1)
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2018-02-02T03:55:39+00:00 (1 year, 8 months ago)
    
    OS-6583 add topology.dtd.1 to manlint exception list

diff --git a/exception_lists/manlint b/exception_lists/manlint
index 5982ac5cec..ebdafa2a84 100644
--- a/exception_lists/manlint
+++ b/exception_lists/manlint
@@ -10,10 +10,12 @@
 #
 #
 # Copyright 2016 Toomas Soome <tsoome@me.com>
+# Copyright (c) 2018, Joyent, Inc.
 #
 usr/src/boot/*
 # Not actually a manual page
 usr/src/cmd/svc/dtd/service_bundle.dtd.1
+usr/src/lib/fm/topo/maps/common/topology.dtd.1
 usr/src/lib/libbsm/adt_record.dtd.1
 usr/src/lib/libbsm/adt_record.xsl.1
 usr/src/lib/libzonecfg/dtd/zonecfg.dtd.1
