From 3ca0044fd000a1afdbbae87e07d5e2fae901184f Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Thu, 13 Jun 2019 13:05:02 -0400
Subject: [PATCH] TRITON-1851 update nat base image to
 triton-origin-x86_64-18.4.0

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 569f811..0add8f4 100644
--- a/Makefile
+++ b/Makefile
@@ -19,7 +19,7 @@ TOP ?= $(error Unable to access eng.git submodule Makefiles.)
 RELEASE_TARBALL:=$(NAME)-pkg-$(STAMP).tar.gz
 RELSTAGEDIR:=/tmp/$(NAME)-$(STAMP)
 
-BASE_IMAGE_UUID = de411e86-548d-11e4-a4b7-3bb60478632a
+BASE_IMAGE_UUID = a9368831-958e-432d-a031-f8ce6768d190
 BUILDIMAGE_NAME = $(NAME)
 BUILDIMAGE_DESC	= SmartDataCenter per-user NAT zone
 
-- 
2.21.0

