From 93a5d82f37ed08d209b264b96e1d839c9529bc91 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Wed, 15 May 2019 14:23:41 -0400
Subject: [PATCH] MANTA-4279 fix make check for boray, add back in rust checks

---
 Makefile | 10 ++--------
 deps/eng |  2 +-
 2 files changed, 3 insertions(+), 9 deletions(-)

diff --git a/Makefile b/Makefile
index b514038..73a76a6 100644
--- a/Makefile
+++ b/Makefile
@@ -4,6 +4,8 @@
 
 NAME = manta-boray
 
+RUST_CODE = 1
+
 SMF_MANIFESTS_IN = smf/manifests/boray.xml.in
 
 ENGBLD_USE_BUILDIMAGE =	true
@@ -23,9 +25,6 @@ include ./deps/eng/tools/mk/Makefile.smf.defs
 # TODO: Use this to download or verify install of expected rust version
 RUST_PREBUILT_VERSION =		1.33.0
 
-# TODO: Generalize this to the eng repo
-CARGO = $(shell which cargo)
-
 RELEASE_TARBALL :=	$(NAME)-pkg-$(STAMP).tar.gz
 ROOT :=			$(shell pwd)
 RELSTAGEDIR :=		/tmp/$(NAME)-$(STAMP)
@@ -82,11 +81,6 @@ publish: release
 build-boray:
 	$(CARGO) build --release
 
-.PHONY: rust-check
-rust-check:
-	cargo check
-
-
 include ./deps/eng/tools/mk/Makefile.deps
 include ./deps/eng/tools/mk/Makefile.agent_prebuilt.targ
 include ./deps/eng/tools/mk/Makefile.smf.targ
diff --git a/deps/eng b/deps/eng
index dcf15a6..d25b8fc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit dcf15a6507556313d7925c6914c0040fbc52161e
+Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
-- 
2.21.0

