From 3398ec1998a89a7cca128a8505ee9d14a2969eb4 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Wed, 6 Mar 2019 13:41:51 -0500
Subject: [PATCH] joyent/boray#6 want `make check`

---
 Makefile | 7 +++++++
 1 file changed, 7 insertions(+)
 create mode 100644 Makefile

diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..06ff1d7
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,7 @@
+#
+# Copyright 2019 Joyent, Inc
+#
+
+.PHONY: check
+check:
+	cargo check
-- 
2.21.0

