commit d71ca6a9cc50b4877cbfbd5d2a6b72830417ee61 (refs/changes/76/5276/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-12-20T23:55:51+00:00 (10 months ago)
    
    OS-7464 uninitialized boolean from OS-7419 is too truthy

diff --git a/usr/src/cmd/bhyve/bhyverun.c b/usr/src/cmd/bhyve/bhyverun.c
index c85c834e9f..7afc47111b 100644
--- a/usr/src/cmd/bhyve/bhyverun.c
+++ b/usr/src/cmd/bhyve/bhyverun.c
@@ -1018,7 +1018,7 @@ main(int argc, char *argv[])
 	int rtc_localtime;
 	bool gdb_stop;
 #ifndef __FreeBSD__
-	bool suspend;
+	bool suspend = false;
 #endif
 	struct vmctx *ctx;
 	uint64_t rip;
