From 6acf9dfbd4ff23f440aaaa9c34bfaba72f620ae8 Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Wed, 20 Jul 2016 11:00:54 +0200
Subject: [PATCH] TOOLS-1492 `sdcadm insts` exceptions on SAPI instances for
 removed servers

---
 lib/sdcadm.js | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 717d84c..21ba686 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -815,13 +815,21 @@ SdcAdm.prototype.listInsts = function listInsts(opts, cb) {
                             if (ins.params && ins.params.server_uuid) {
                                 var id = ins.params.server_uuid;
                                 d.server = id;
+                                // TOOLS-1492: A server may have been deleted
+                                // leaving the SAPI instance present and
+                                // associated to nothing:
+                                if (!ctx.serverFromUuid[id]) {
+                                    return null;
+                                }
                                 d.hostname = ctx.serverFromUuid[id].hostname;
                                 d.server_ip = ctx.serverAdminIpFromUuid[id];
                             }
                             return (d);
+                        }).filter(function (ins) {
+                            return (ins !== null);
                         });
                         ctx.insts = ctx.insts.concat(dlInsts);
-                        return next();
+                        next();
                     });
                 });
             });
-- 
2.21.0

