commit ad691e51cc8488fff7f680eb39fb5b784070c4fe (refs/changes/22/4222/1)
Author: dyep49 <dyep49@gmail.com>
Date:   2018-06-14T12:23:10-07:00 (1 year, 4 months ago)
    
    TRITON-492 Add metricPorts metadata to workflow for cmon-agent

diff --git a/boot/setup.sh b/boot/setup.sh
index 7f31f4a..2b66ae9 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -7,7 +7,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -52,6 +52,9 @@ sdc_log_rotation_add wf-runner /var/svc/log/*wf-runner*.log 1g
 sdc_log_rotation_add wf-backfill /var/svc/log/*wf-backfill*.log 1g
 sdc_log_rotation_setup_end
 
+# Add metricsPorts metadata for cmon-agent discovery
+mdata-put metricPorts 8881,8882
+
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
 
diff --git a/package.json b/package.json
index 168e655..a63c0a2 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "workflow",
   "description": "SmartDataCenter Workflow API and Runners",
-  "version": "1.3.0",
+  "version": "1.3.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "repository": {
