commit 0fcaadf4ce6f28607cb64c8f80f86a63a279a099 (refs/changes/22/2422/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-17T19:21:48+00:00 (2 years, 2 months ago)
    
    joyent/node-cueball#124 new getPool api method for CueBallAgent

diff --git a/lib/agent.js b/lib/agent.js
index 4c58c65..5f0630d 100644
--- a/lib/agent.js
+++ b/lib/agent.js
@@ -423,6 +423,12 @@ CueBallAgent.prototype.checkSocket = function (host, handle, socket) {
 	req.end();
 };
 
+/* Used by muskie to check for recent shark connections */
+CueBallAgent.prototype.getPool = function (host) {
+    mod_assert.string(host, 'hostname');
+    return (this.pools[host]);
+};
+
 /* Only used by UNIX socket connections with "socketPath" set. */
 CueBallAgent.prototype.createConnection = function (options, connectListener) {
 	throw (new Error('UNIX domain sockets not supported'));
