commit 40b433e7caa1ab48e0760841dd1c7c6a9d0fab39
Author: Trent Mick <trentm@gmail.com>
Date:   2019-07-25T10:02:18-07:00 (2 months ago)
    
    TOOLS-2298 check-copyright fails for non-master Jenkins builds

diff --git a/tools/check-copyright b/tools/check-copyright
index b260b93..d1f3fac 100755
--- a/tools/check-copyright
+++ b/tools/check-copyright
@@ -145,7 +145,13 @@ done
 shift $((OPTIND - 1))
 
 # Gather the files to possibly check.
-committedFiles=$(git diff --name-only master..)
+committedFiles=
+if git branch -a | grep '^  remotes/origin/master$' >/dev/null 2>/dev/null; then
+    # The intent for committed files is to compare against the ultimate
+    # merge branch target. We assume that is "origin/master", the common case
+    # for Joyent eng repos.
+    committedFiles=$(git diff --name-only master..)
+fi
 stagedFiles=$(git diff --name-only --staged)
 unstagedFiles=$(git diff --name-only)
 uniqFiles=$(echo "
