From a6460a99145f7e8222f47c9b1a8f56d0a0772c9c Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Fri, 23 Aug 2019 13:54:56 +0000
Subject: [PATCH] sysinfo.js should handle partial failure gracefully

---
 lib/backends/linux/sysinfo.js | 56 ++++++++++++++++++++++-------------
 1 file changed, 36 insertions(+), 20 deletions(-)

diff --git a/lib/backends/linux/sysinfo.js b/lib/backends/linux/sysinfo.js
index 07d2fa8..065f78f 100644
--- a/lib/backends/linux/sysinfo.js
+++ b/lib/backends/linux/sysinfo.js
@@ -31,7 +31,7 @@ var smartdc_config = require('./smartdc-config');
 
 var LOG = null;
 
-function getBasicInfo(callback)
+function getBasicInfo(log, callback)
 {
     var sysinfo = {
         'SDC Version': '7.0',
@@ -62,7 +62,7 @@ function getBasicInfo(callback)
  * 'CPU Total Cores'        Also misleading: # of threads
  * 'CPU Virtualization'     "vmx" (Intel), "svm" (AMD), or none
  */
-function getCpuInfo(callback)
+function getCpuInfo(log, callback)
 {
     fs.readFile('/proc/cpuinfo', function readCpuInfo(err, data) {
         if (err) {
@@ -130,7 +130,7 @@ function getCpuInfo(callback)
  *
  * 'MiB of Memory'
  */
-function getMemInfo(callback)
+function getMemInfo(log, callback)
 {
     fs.readFile('/proc/meminfo', function readMemInfo(err, data) {
         if (err) {
@@ -164,7 +164,7 @@ function getMemInfo(callback)
  *     'sdd': {'Size in GB': 447},
  *  }
  */
-function getDiskInfo(callback)
+function getDiskInfo(log, callback)
 {
     execFile('/bin/lsblk', ['-Jbd', '-o', 'name,size'],
         function lsBlk(err, stdout, stderr) {
@@ -204,7 +204,7 @@ function getDiskInfo(callback)
  * 'Zpool Creation'     When it was created
  * 'Zpool Size in GiB'  zfs used + available
  */
-function getZpoolInfo(callback)
+function getZpoolInfo(log, callback)
 {
     var sysinfo = {};
 
@@ -240,8 +240,8 @@ function getZpoolInfo(callback)
                     var pool = vals[0];
                     dsprops[pool] = {};
 
-                    for (var i in props) {
-                        dsprops[pool][props[i]] = vals[i];
+                    for (var prop in props) {
+                        dsprops[pool][props[prop]] = vals[prop];
                     }
                 }
 
@@ -430,7 +430,7 @@ function getZpoolInfo(callback)
  *  }
  * 
  */
-function getNetInfo(callback)
+function getNetInfo(log, callback)
 {
     var sysinfo;
 
@@ -519,7 +519,7 @@ function ipAddrToNetInfo(nets)
     return ({ nics: nics, vnics: vnics, aggrs: aggrs });
 }
 
-function getSmbiosInfo(callback)
+function getSmbiosInfo(log, callback)
 {
     var sysinfo;
 
@@ -556,10 +556,10 @@ function getSmbiosInfo(callback)
     });
 }
 
-function sysinfo(callback)
+function sysInfo(log, callback)
 {
-    vasync.parallel({
-        'funcs': [
+    vasync.forEachParallel({
+        'inputs': [
             getBasicInfo,
             getCpuInfo,
             getMemInfo,
@@ -567,19 +567,35 @@ function sysinfo(callback)
             getZpoolInfo,
             getNetInfo,
             getSmbiosInfo
-        ]
+        ],
+        'func': function callInfoFunc(func, cb) {
+            // Do not blow up when something goes wrong.  Return as many results
+            // as possible.  The caller can decide whether errors are fatal.
+            try {
+                func(log, cb);
+            } catch(error) {
+                cb(error);
+            }
+        }
     }, function sysinfoDone(err, results) {
-        var si = {};
+        var sysinfo = {};
 
-        if (!err) {
-            results.successes.unshift(si);
-            Object.assign.apply(this, results.successes);
+        for (var frag of results.successes) {
+            if (frag) {
+                Object.assign(sysinfo, frag);
+            }
         }
-        callback(err, si);
+        callback(err, sysinfo);
     });
 }
 
 module.exports = {
-    log: LOG,
-    sysinfo: sysinfo
+    sysInfo: sysInfo,
+    getBasicInfo: getBasicInfo,
+    getCpuInfo: getCpuInfo,
+    getMemInfo: getMemInfo,
+    getDiskInfo: getDiskInfo,
+    getZpoolInfo: getZpoolInfo,
+    getNetInfo: getNetInfo,
+    getSmbiosInfo: getSmbiosInfo
 };
-- 
2.17.2 (Apple Git-113)

