From 3c606d0ac275521c8e3cb1d0b4e32c3559108f53 Mon Sep 17 00:00:00 2001
From: Your Name <you@example.com>
Date: Tue, 10 Jul 2018 19:01:09 +0000
Subject: [PATCH] joyent/node-fast#20 allow specifying ports used for fastbench
 CLI

---
 bin/fastbench | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/bin/fastbench b/bin/fastbench
index 644fedf..eaf1e29 100755
--- a/bin/fastbench
+++ b/bin/fastbench
@@ -76,6 +76,8 @@ function fbUsageMessage()
 	    '    -i | --interval NSECONDS     Report every NSECONDS seconds.',
 	    '    -n | --nrequests NREQUESTS   Stop after completing ' +
 		'NREQUESTS requests.',
+	    '    -p | --artediport PORT       Use PORT for metrics endpoint.',
+	    '    -k | --kangport PORT         Use PORT for kang endpoint.',
 	    ''
 	].join('\n');
 
@@ -106,7 +108,9 @@ function main()
 	});
 
 	parser = new mod_getopt.BasicParser(
-	    'c:(count)d:(duration)i:(interval)n:(nrequests)', process.argv);
+	    'c:(count)d:(duration)i:(interval)n:(nrequests)' +
+		'p:(artediport)k:(kangport)',
+	     process.argv);
 	while ((option = parser.getopt()) !== undefined) {
 		switch (option.option) {
 		case 'c':
@@ -126,6 +130,14 @@ function main()
 			config.fbNRequests = parseOptInt(option, 1);
 			break;
 
+		case 'p':
+			config.fbArtediPort = parseOptInt(option, 1);
+			break;
+
+		case 'k':
+			config.fbKangPort = parseOptInt(option, 1);
+			break;
+
 		default:
 			/* error message already emitted */
 			mod_assertplus.equal('?', option.option);
-- 
2.21.0

