commit 18116463351ce94cc61e06ed7d6519446848e10c (refs/changes/88/1088/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-12-08T00:40:20+00:00 (2 years, 10 months ago)
    
    OS-1151 need support for USB v3.0

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index e35fc7f5..35dfbf9d 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -1283,6 +1283,7 @@ vr "pci1106,3106"
 xge "pci17d5,5731"
 xge "pci17d5,5831"
 xge "pci17d5,5832"
+xhci "pciclass,0c0330"
 xnbe "xnb,ioemu"
 xnbo "xnb"
 xnbo "xnb,SUNW_mac"
diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index cd69fc36..eeb69a15 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -295,3 +295,4 @@ pvscsi 298
 vmxnet3s 299
 smrt 300
 iwn 301
+xhci 302
