commit 182c9e2fb0ddafd31fe8448b094df3c1c50e2657
Author: Sven Meys <sven.meys@gmail.com>
Date:   2019-08-20T18:55:53+00:00 (7 weeks ago)
    
    joyent/node-bunyan-syslog#54 Support building for node v12

diff --git a/README.md b/README.md
index f520800..1744153 100644
--- a/README.md
+++ b/README.md
@@ -77,7 +77,7 @@ This module maps bunyan levels to syslog levels as follows:
 You can run the test suite using the provided `rsyslog` configuration:
 
 ```
-$ rsyslogd -f ./test/rsyslog.conf -i ./test.pid -u $USER
+$ rsyslogd -f ./test/rsyslog.conf -i $PWD/test.pid -u $USER
 $ make test
 $ kill $(cat test.pid)
 ```
diff --git a/package.json b/package.json
index 35f5106..3e373b1 100755
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "bunyan-syslog",
   "description": "Syslog Stream for Bunyan",
-  "version": "0.3.2",
+  "version": "0.3.3",
   "author": "Mark Cavage",
   "license": "MIT",
   "main": "./lib/index.js",
@@ -11,13 +11,13 @@
   },
   "dependencies": {
     "assert-plus": "0.1.5",
-    "nan": "^2.10.0"
+    "nan": "^2.14.0"
   },
   "devDependencies": {
     "bunyan": "^1.8.12",
     "istanbul": "^0.3.6",
     "faucet": "0.0.1",
-    "tape": "^3.5.0",
+    "tape": "^4.11.0",
     "eslint": "^4.13.1",
     "eslint-plugin-joyent": "~2.1.0"
   },
diff --git a/src/syslog.cc b/src/syslog.cc
index 07e269e..9544d4e 100644
--- a/src/syslog.cc
+++ b/src/syslog.cc
@@ -14,42 +14,27 @@ using namespace v8;
 #define RETURN_ARGS_EXCEPTION(MSG)                                      \
 	Nan::ThrowError(MSG)
 
-#define RETURN_ERRNO_EXCEPTION(RC, API, MSG)				\
-	Nan::ThrowError(node::ErrnoException(RC, API, MSG))
-
-#define RETURN_OOM_EXCEPTION()						\
-	RETURN_ERRNO_EXCEPTION(ENOMEM, "malloc", strerror(ENOMEM))
-
 #define REQUIRE_ARGS(ARGS)					\
 	if (ARGS.Length() == 0)					\
 		RETURN_ARGS_EXCEPTION("missing arguments");
 
 #define REQUIRE_INT_ARG(ARGS, I, VAR)                                   \
 	REQUIRE_ARGS(ARGS);						\
-	if (ARGS.Length() <= (I) || !ARGS[I]->IsNumber())		\
+	if (ARGS.Length() <= (I))					\
+		RETURN_ARGS_EXCEPTION("argument " #I " must be an Integer"); \
+	Nan::MaybeLocal<v8::Integer> _ ## VAR(Nan::To<v8::Integer>(ARGS[I])); \
+	if (_ ## VAR.IsEmpty())					\
 		RETURN_ARGS_EXCEPTION("argument " #I " must be an Integer"); \
-	Local<Integer> _ ## VAR(ARGS[I]->ToInteger());			\
-	int VAR = _ ## VAR->Value();
+	int VAR = _ ## VAR.ToLocalChecked()->Value();
 
 #define REQUIRE_STRING_ARG(ARGS, I, VAR)				\
 	REQUIRE_ARGS(ARGS);						\
-	if (ARGS.Length() <= (I) || !ARGS[I]->IsString())		\
+	if (ARGS.Length() <= (I))					\
 		RETURN_ARGS_EXCEPTION("argument " #I " must be a String"); \
-	String::Utf8Value VAR(ARGS[I]->ToString());
-
-#define REQUIRE_FUNCTION_ARG(ARGS, I, VAR)                              \
-	REQUIRE_ARGS(ARGS);						\
-	if (ARGS.Length() <= (I) || !ARGS[I]->IsFunction())		\
-		RETURN_EXCEPTION("argument " #I " must be a Function");	\
-	Local<Function> VAR = Local<Function>::Cast(ARGS[I]);
-
-
-#define REQUIRE_OBJECT_ARG(ARGS, I, VAR)				\
-	REQUIRE_ARGS(ARGS);						\
-	if (ARGS.Length() <= (I) || !ARGS[I]->IsObject())		\
-		RETURN_EXCEPTION("argument " #I " must be an Object");	\
-	Local<Object> VAR(ARGS[I]->ToObject());
-
+	Nan::MaybeLocal<v8::String> _ ## VAR(Nan::To<v8::String>(ARGS[I])); \
+	if (_ ## VAR.IsEmpty())					\
+		RETURN_ARGS_EXCEPTION("argument " #I " must be a String"); \
+	Nan::Utf8String VAR(_ ## VAR.ToLocalChecked());
 
 
 ///--- API
@@ -95,7 +80,7 @@ NAN_METHOD(Mask) {
 	info.GetReturnValue().Set(scope.Escape(Nan::New<Integer>(mask)));
 }
 
-void init(Handle<Object> target) {
+NAN_MODULE_INIT(init) {
 	Nan::SetMethod(target, "openlog", Open);
 	Nan::SetMethod(target, "syslog", Log);
 	Nan::SetMethod(target, "closelog", Close);
