commit ae371c7750e22f68d18d5381676d7bc26f81b594 (refs/changes/82/3082/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-12-08T17:00:57-08:00 (1 year, 10 months ago)
    
    RELENG-782 baffling  failures in dhcpd build

diff --git a/Makefile b/Makefile
index 981d313..e14fdeb 100644
--- a/Makefile
+++ b/Makefile
@@ -130,7 +130,7 @@ pkg: all
 release: $(RELEASE_TARBALL)
 
 $(RELEASE_TARBALL): pkg
-	(cd $(PKG_DIR); tar -jcf $(TOP)/$(RELEASE_TARBALL) root)
+	(cd $(PKG_DIR); $(TAR) -jcf $(TOP)/$(RELEASE_TARBALL) root)
 
 publish:
 	@if [[ -z "$(BITS_DIR)" ]]; then \
