From ae371c7750e22f68d18d5381676d7bc26f81b594 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 8 Dec 2017 17:00:57 -0800
Subject: [PATCH] RELENG-782 baffling  failures in dhcpd build

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 981d313..e14fdeb 100644
--- a/Makefile
+++ b/Makefile
@@ -130,7 +130,7 @@ pkg: all
 release: $(RELEASE_TARBALL)
 
 $(RELEASE_TARBALL): pkg
-	(cd $(PKG_DIR); tar -jcf $(TOP)/$(RELEASE_TARBALL) root)
+	(cd $(PKG_DIR); $(TAR) -jcf $(TOP)/$(RELEASE_TARBALL) root)
 
 publish:
 	@if [[ -z "$(BITS_DIR)" ]]; then \
-- 
2.21.0

