From db22e7b9bbf40c42fc4ed92e675f46a5116deb3d Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Thu, 10 May 2018 20:49:54 +0000
Subject: [PATCH] OS-6934 need UEFI (non-CSM) firmware for bhyve

---
 Makefile           |  1 +
 manifest           |  3 ++-
 uefi-edk2/Makefile | 36 ++++++++++++++++++++++++++++++------
 3 files changed, 33 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index 165f999..7692a74 100644
--- a/Makefile
+++ b/Makefile
@@ -73,6 +73,7 @@ SUBDIRS = \
 	screen \
 	socat \
 	tun \
+	uefi-edk2 \
 	uuid \
 	vim \
 	wget \
diff --git a/manifest b/manifest
index c40d74b..fe98fc3 100644
--- a/manifest
+++ b/manifest
@@ -3022,7 +3022,8 @@ f usr/share/man/man1/screen.1 0444 root bin
 f usr/bin/socat 0555 root bin
 f usr/share/man/man1/socat.1 0444 root bin
 # uefi-edk2
-# f usr/share/bhyve/uefi-csm-rom.bin 0644 root bin
+f usr/share/bhyve/uefi-csm-rom.bin 0644 root bin
+f usr/share/bhyve/uefi-rom.bin 0644 root bin
 # uuid
 f usr/bin/uuid 0755 root bin
 f usr/share/man/man1/uuid.1 0644 root bin
diff --git a/uefi-edk2/Makefile b/uefi-edk2/Makefile
index 49af187..f4e62a5 100644
--- a/uefi-edk2/Makefile
+++ b/uefi-edk2/Makefile
@@ -34,8 +34,13 @@ MAKE_ARGS+=	AS=$(STRAPPROTO)/usr/gnu/bin/gas \
 ARCH=		X64
 UEFI_TARGET=	RELEASE
 
+ROMS=		uefi uefi-csm
 
-all: $(VER.64)/$(UNPACK_SENTINEL)
+.NOTPARALLEL: $(ROMS) csm16
+
+all: $(ROMS)
+
+basetools: $(VER.64)/$(UNPACK_SENTINEL)
 	mkdir -p $(VER.64)/Build
 	ln -sf $(STRAPPROTO)/usr/bin/gcc $(VER.64)/Build/gcc
 	ln -sf $(STRAPPROTO)/usr/gnu/bin/gld $(VER.64)/Build/ld
@@ -44,11 +49,30 @@ all: $(VER.64)/$(UNPACK_SENTINEL)
 	ln -sf $(STRAPPROTO)/usr/gnu/bin/gobjcopy $(VER.64)/Build/objcopy
 	ln -sf /opt/local/bin/nasm $(VER.64)/Build/nasm
 	$(MAKE) $(MAKE_ARGS) ARCH=$(ARCH) -C $(VER.64)/BaseTools
-	bash -c "cd $(VER.64); source edksetup.sh; $(MAKE) $(MAKE_ARGS) ARCH=$(ARCH) -C BhyvePkg/Csm/BhyveCsm16"
-	bash -c "cd $(VER.64); source edksetup.sh; export ILGCC_BIN=$(PWD)/$(VER.64)/Build/; build -t ILGCC -a $(ARCH) -b $(UEFI_TARGET) -p BhyvePkg/BhyvePkgX64.dsc -DDEBUG_ON_SERIAL_PORT=TRUE -DFD_SIZE_2MB -DCSM_ENABLE=TRUE"
-	mv $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/BHYVE.fd $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-csm-rom.bin
+
+csm16:
+	bash -c "cd $(VER.64); \
+		source edksetup.sh; \
+		$(MAKE) $(MAKE_ARGS) ARCH=$(ARCH) -C BhyvePkg/Csm/BhyveCsm16"
+
+uefi-csm: csm16
+
+$(ROMS): basetools
+	rm -rf $(VER.64)/Build/BhyveX64
+	bash -c 'cd $(VER.64); \
+		source edksetup.sh; \
+		export ILGCC_BIN=$(PWD)/$(VER.64)/Build/; \
+		[[ $@ == uefi-csm ]] && extra_cflags=-DCSM_ENABLE=TRUE; \
+		build -t ILGCC -a $(ARCH) -b $(UEFI_TARGET) \
+			-p BhyvePkg/BhyvePkgX64.dsc \
+			-DDEBUG_ON_SERIAL_PORT=TRUE -DFD_SIZE_2MB \
+			$${extra_cflags}'
+	mv $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/BHYVE.fd \
+		$(VER.64)/Build/$@-rom.bin
 
 install: all
 	mkdir -p $(DESTDIR)/usr/share/bhyve
-	/usr/sbin/install -m 0755 -f $(DESTDIR)/usr/share/bhyve $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-csm-rom.bin
-
+	for rom in $(ROMS); do \
+		/usr/sbin/install -m 0755 -f $(DESTDIR)/usr/share/bhyve \
+		    $(VER.64)/Build/$$rom-rom.bin; \
+	done
-- 
2.21.0

