From 84d475ce50e3fedb1ade1905eb2b1219bbf3be8e Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody@kkantor.com>
Date: Fri, 20 Apr 2018 14:41:35 +0000
Subject: [PATCH] MANTA-3624 create pgstatsmon image

---
 targets.json.in | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/targets.json.in b/targets.json.in
index accbc08..40eeff2 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1438,9 +1438,11 @@ cat <<EOF
     ],
     "public": true,
     "deps": [
+      "amon",
       "config-agent"
     ],
     "tarballs": [
+      {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
       {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
     ]
   },
-- 
2.21.0

