commit 41a3bf2405d2b398166ba7a9318ecfb18bbae073 (refs/changes/49/3849/1)
Author: Kody A Kantor <kody.kantor@gmail.com>
Date:   2018-04-20T14:30:53+00:00 (1 year, 6 months ago)
    
    MANTA-3624 create pgstatsmon image

diff --git a/targets.json.in b/targets.json.in
index accbc08..40eeff2 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1438,9 +1438,11 @@ cat <<EOF
     ],
     "public": true,
     "deps": [
+      "amon",
       "config-agent"
     ],
     "tarballs": [
+      {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
       {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
     ]
   },
