From 73f60c0617725bcc8b32b7cd0545e53c2f51b38e Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 2 May 2018 17:12:23 +0200
Subject: [PATCH] TRITON-351 "sdcadm self-update" dumps core

---
 lib/sdcadm.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 72729ee..47cee2c 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -436,7 +436,7 @@ SdcAdm.prototype.fini = function fini() {
     if (self.sadm_ur !== null) {
         self.sadm_ur.close();
     }
-    if (self._cueballAgent !== undefined) {
+    if (self._cueballAgent !== undefined && !self._cueballAgent.cba_stopped) {
         self.cueballAgent.stop();
     }
 };
-- 
2.21.0

