commit 2760d783858c5fd76ed9fe91df067ed232989552 (refs/changes/93/3893/3)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-05-02T19:22:43+02:00 (1 year, 5 months ago)
    
    TRITON-351 "sdcadm self-update" dumps core

diff --git a/lib/cli/do_self_update.js b/lib/cli/do_self_update.js
index 0805324..29ba71e 100644
--- a/lib/cli/do_self_update.js
+++ b/lib/cli/do_self_update.js
@@ -28,11 +28,13 @@ function do_self_update(subcmd, opts, args, cb) {
         cb(new errors.UsageError(
         'Please provide an image UUID or use `sdcadm self-update --latest`\n' +
         'in order to update to the latest available image.'));
+        return;
     }
 
     if (image === 'help') {
         cb(new errors.UsageError(
         'Please use `sdcadm help self-update` instead'));
+        return;
     }
 
 
diff --git a/lib/default-fabric.js b/lib/default-fabric.js
index 97027e0..c834349 100644
--- a/lib/default-fabric.js
+++ b/lib/default-fabric.js
@@ -128,7 +128,8 @@ function do_default_fabric(subcmd, opts, args, cb) {
         this.do_help('help', {}, [ subcmd ], cb);
         return;
     } else if (args.length !== 1) {
-        return cb(new errors.UsageError('Must specify account'));
+        cb(new errors.UsageError('Must specify account'));
+        return;
     }
 
     vasync.pipeline({arg: {}, funcs: [
diff --git a/lib/platform.js b/lib/platform.js
index c72d317..284615c 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1725,6 +1725,7 @@ function do_assign(subcmd, opts, args, cb) {
     if (platform === 'help') {
         cb(new errors.UsageError(
         'Please use `sdcadm platform help assign` instead'));
+        return;
     }
     var server = args.length ? args : null;
     var assignOpts;
@@ -2334,6 +2335,7 @@ function do_set_default(subcmd, opts, args, cb) {
     if (opts.platform === 'help') {
         cb(new errors.UsageError(
         'Please use `sdcadm platform help set-default` instead'));
+        return;
     }
 
 
