commit b6c6c7e59e55139a1aa4a87ee9c0620b1d0d82ca (refs/changes/34/3634/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-03-14T20:59:51+00:00 (1 year, 7 months ago)
    
    OS-6711 ship bhyve bootrom binary

diff --git a/overlay/generic/manifest b/overlay/generic/manifest
index 8683f59c..64d0f21b 100644
--- a/overlay/generic/manifest
+++ b/overlay/generic/manifest
@@ -174,6 +174,8 @@ f usr/lib/localedef/extensions/UTF-8.x 0444 root bin
 f usr/sbin/fssstat 0555 root bin
 f usr/sbin/zonememstat 0555 root bin
 f usr/sbin/zonemon 0555 root bin
+d usr/share/bhyve 0555 root bin
+f usr/share/bhyve/uefi-csm-rom.bin 0444 root bin
 f usr/share/man/man1/json.1 0444 root bin
 f usr/share/man/man1m/fssstat.1m 0444 root bin
 d var/db 0755 root sys
diff --git a/overlay/generic/usr/share/bhyve/uefi-csm-rom.bin b/overlay/generic/usr/share/bhyve/uefi-csm-rom.bin
new file mode 100755
index 00000000..02f11ea5
Binary files /dev/null and b/overlay/generic/usr/share/bhyve/uefi-csm-rom.bin differ
