commit 7c5249bce3c8c948f586f1bb282ef06b5cea95e0 (refs/changes/85/3585/3)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-03-12T19:01:09+00:00 (1 year, 7 months ago)
    
    OS-6686 add FreeBSD compat & contrib sources to cscope

diff --git a/usr/src/req.flg b/usr/src/req.flg
index 83e2936a36..d62f4ac12c 100644
--- a/usr/src/req.flg
+++ b/usr/src/req.flg
@@ -22,6 +22,7 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
+# Copyright 2018 Joyent, Inc
 #
 
 echo_file usr/src/Makefile
@@ -32,3 +33,5 @@ echo_file usr/src/Makefile.master.64
 echo_file usr/src/Makefile.msg.targ
 echo_file usr/src/Makefile.psm
 echo_file usr/src/Makefile.psm.targ
+
+find_files "s.*" usr/contrib/freebsd
diff --git a/usr/src/tools/scripts/build_cscope.conf b/usr/src/tools/scripts/build_cscope.conf
index 859b5137d6..298db1281b 100644
--- a/usr/src/tools/scripts/build_cscope.conf
+++ b/usr/src/tools/scripts/build_cscope.conf
@@ -22,8 +22,8 @@
 #
 # Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
+# Copyright 2018 Joyent, Inc.
 #
-# ident	"%Z%%M%	%I%	%E% SMI"
 #
 # This file configures the set of cross-references built by build_cscope.
 # The format is:
@@ -35,6 +35,6 @@
 # directories.
 #
 
-complete  -f   .
+complete  ""   .
 uts	  ""   uts uts/sun4u uts/sun4v uts/i86pc
 psm	  ""   psm/stand psm/stand/boot psm/stand/boot/sparcv9/sun4u psm/stand/boot/sparcv9/sun4v
diff --git a/usr/src/uts/req.flg b/usr/src/uts/req.flg
index ffbaa3f643..15085a486d 100644
--- a/usr/src/uts/req.flg
+++ b/usr/src/uts/req.flg
@@ -22,6 +22,7 @@
 #
 # Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
+# Copyright 2018 Joyent, Inc.
 #
 
 #
@@ -37,6 +38,7 @@ echo_file usr/src/Makefile
 
 # For full builds (open and closed), we want both etc/certs and
 # etc/keys.  For an open source build, there's no etc/keys directory.
+find_files "s.*" usr/contrib/freebsd
 find_files "s.*" usr/src/cmd/cmd-crypto/etc
 find_files "s.*" usr/src/common/acl
 find_files "s.*" usr/src/common/atomic
@@ -56,4 +58,5 @@ find_files "s.*" usr/src/common/smbios
 find_files "s.*" usr/src/common/tsol
 find_files "s.*" usr/src/common/util
 find_files "s.*" usr/src/common/zfs
+find_files "s.*" usr/src/compat/freebsd
 find_files "s.*" usr/src/psm/promif
