From 8f91b4c38f775423fd5b461eec7e709fc4cb337e Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Tue, 7 May 2019 18:22:03 +0000
Subject: [PATCH] MANTA-4145 Add metricPorts mdata variable to
 manta-garbage-collector

---
 boot/firstboot.sh | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/boot/firstboot.sh b/boot/firstboot.sh
index 48d6740..b8ffa8f 100644
--- a/boot/firstboot.sh
+++ b/boot/firstboot.sh
@@ -21,6 +21,7 @@ NAME=manta-garbage-collector
 #
 
 SVC_ROOT="/opt/smartdc/$NAME"
+SAPI_CONFIG="$SVC_ROOT/etc/config.json"
 
 #
 # Build PATH from this list of directories.  This PATH will be used both in the
@@ -79,4 +80,18 @@ if ! svccfg import "/opt/smartdc/$NAME/smf/manifests/garbage-collector.xml"; the
 	fatal 'could not import SMF service'
 fi
 
+#
+# Get the port from the sapi config file, then add 1000 to it to get the
+# metricPort to allow scraping by cmon-agent.
+#
+# Note that the config file is guaranteed to have been written by config-agent
+# at this point, because we've already called manta_common_setup, which calls
+# manta_enable_config_agent, which enables the config-agent service
+# synchronously.
+#
+if [[ ! -f $SAPI_CONFIG ]]; then
+	fatal 'SAPI config not found'
+fi
+mdata-put metricPorts $(expr $(< "$SAPI_CONFIG" json 'port') + 1000)
+
 manta_common_setup_end
-- 
2.21.0

