From 367b9a278e4d1edbfb42c074ceec21f8bc46780d Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Thu, 18 Apr 2019 20:16:19 +0000
Subject: [PATCH] OS-7860 Install libsunw_crypto.a to proto area during build

---
 openssl1x/install-sfw-64 | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/openssl1x/install-sfw-64 b/openssl1x/install-sfw-64
index f619d2d..d33d1df 100644
--- a/openssl1x/install-sfw-64
+++ b/openssl1x/install-sfw-64
@@ -24,7 +24,7 @@
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# Copyright (c) 2010 Joyent Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 set -o errexit
@@ -67,4 +67,13 @@ _install L ../../../lib/amd64/libsunw_ssl.so.${LIBVER} \
     ${USRLIBDIR}/libsunw_ssl.so.${LIBVER}
 _install L ../../../lib/amd64/libsunw_ssl.so ${USRLIBDIR}/libsunw_ssl.so
 
+# The pivy tools as well as kbmd deliberately statically
+# link against libsunw_crypto.a to prevent sharing of code pages
+# to mitigate against certain timing attacks.  We install the .a
+# into the proto area in a location that won't be seen (and
+# unintentionally used) by other software, but do not
+# deliver the static library in the platform image.
+mkdir -p ${DESTDIR}/.build
+_install D libcrypto.a ${DESTDIR}/.build/libsunw_crypto.a 0755
+
 exit 0
-- 
2.21.0

