From 2c37bee06430c5d7dd737e1087bc5420425d9d5f Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Thu, 18 Apr 2019 20:16:19 +0000
Subject: [PATCH] OS-7860 Install libsunw_crypto.a to proto area during build

---
 openssl1x/install-sfw-64 | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/openssl1x/install-sfw-64 b/openssl1x/install-sfw-64
index f619d2d..11ea04f 100644
--- a/openssl1x/install-sfw-64
+++ b/openssl1x/install-sfw-64
@@ -24,7 +24,7 @@
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# Copyright (c) 2010 Joyent Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 set -o errexit
@@ -67,4 +67,8 @@ _install L ../../../lib/amd64/libsunw_ssl.so.${LIBVER} \
     ${USRLIBDIR}/libsunw_ssl.so.${LIBVER}
 _install L ../../../lib/amd64/libsunw_ssl.so ${USRLIBDIR}/libsunw_ssl.so
 
+# This isn't delivered in the PI image, only installed
+# into the proto area so kbmd can statically link against it
+_install D libcrypto.a ${LIBDIR}/libsunw_crypto.a 0755
+
 exit 0
-- 
2.21.0

