commit 1812c1030bb0b39273b07d266831e992b8c0f956 (refs/changes/80/2280/3)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2017-07-25T23:04:17+00:00 (2 years, 2 months ago)
    
    DOCKER-1077 docker tests should expect invalid params error

diff --git a/test/integration/api-create.test.js b/test/integration/api-create.test.js
index 9ea4069..9c1ba12 100644
--- a/test/integration/api-create.test.js
+++ b/test/integration/api-create.test.js
@@ -649,12 +649,13 @@ test('create with NetworkMode (docker run --net=)', function (tt) {
     });
 
     tt.test('create with a network that doesn\'t exist', function (t) {
+        var uuid = libuuid.create();
         h.createDockerContainer({
             vmapiClient: VMAPI,
             dockerClient: DOCKER_ALICE,
             test: t,
-            extra: { 'HostConfig.NetworkMode': 'netmodefoobar' },
-            expectedErr: '(Error) network netmodefoobar not found',
+            extra: { 'HostConfig.NetworkMode': uuid },
+            expectedErr: '(Error) network ' + uuid + ' not found',
             start: true
         }, oncreate);
 
@@ -778,33 +779,18 @@ test('create with NetworkMode (docker run --net=)', function (tt) {
         }
     });
 
-    tt.test('fail to create with admin network', function (t) {
-        h.createDockerContainer({
-            vmapiClient: VMAPI,
-            dockerClient: DOCKER_ALICE,
-            test: t,
-            extra: { 'HostConfig.NetworkMode': adminNetwork.name },
-            start: false,
-            expectedErr: '(Error) network admin not found'
-        }, oncreate);
-
-        function oncreate(err, result) {
-            t.ok(err, 'Expecting error');
-            t.end();
-        }
-    });
-
-    tt.test('fail to create with admin network, publish ports', function (t) {
+    tt.test('fail to create with a network that doesn\'t exit, publish ports', function (t) {
+        var uuid = libuuid.create();
         h.createDockerContainer({
             vmapiClient: VMAPI,
             dockerClient: DOCKER_ALICE,
             test: t,
             extra: {
-                'HostConfig.NetworkMode': adminNetwork.name,
+                'HostConfig.NetworkMode': uuid,
                 'HostConfig.PublishAllPorts': true
             },
             start: false,
-            expectedErr: '(Error) network admin not found'
+            expectedErr: '(Error) network ' + uuid + ' not found'
         }, oncreate);
 
         function oncreate(err, result) {
