commit 57c19300b3f5d67d3b45124fb5c71be40c38e6ff (refs/changes/80/2280/2)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2017-07-25T21:59:41+00:00 (2 years, 2 months ago)
    
    DOCKER-1077 docker tests should expect invalid params error

diff --git a/test/integration/api-create.test.js b/test/integration/api-create.test.js
index 9ea4069..91628ee 100644
--- a/test/integration/api-create.test.js
+++ b/test/integration/api-create.test.js
@@ -649,12 +649,13 @@ test('create with NetworkMode (docker run --net=)', function (tt) {
     });
 
     tt.test('create with a network that doesn\'t exist', function (t) {
+        var uuid = libuuid.create();
         h.createDockerContainer({
             vmapiClient: VMAPI,
             dockerClient: DOCKER_ALICE,
             test: t,
-            extra: { 'HostConfig.NetworkMode': 'netmodefoobar' },
-            expectedErr: '(Error) network netmodefoobar not found',
+            extra: { 'HostConfig.NetworkMode': uuid },
+            expectedErr: '(Error) network ' + uuid + ' not found',
             start: true
         }, oncreate);
 
@@ -785,7 +786,7 @@ test('create with NetworkMode (docker run --net=)', function (tt) {
             test: t,
             extra: { 'HostConfig.NetworkMode': adminNetwork.name },
             start: false,
-            expectedErr: '(Error) network admin not found'
+            expectedErr: '(InvalidParameters) Invalid parameters'
         }, oncreate);
 
         function oncreate(err, result) {
@@ -804,7 +805,7 @@ test('create with NetworkMode (docker run --net=)', function (tt) {
                 'HostConfig.PublishAllPorts': true
             },
             start: false,
-            expectedErr: '(Error) network admin not found'
+            expectedErr: '(InvalidParameters) Invalid parameters'
         }, oncreate);
 
         function oncreate(err, result) {
