commit 44aaeb7975746b8c3e2384c8aac9e73d807c4231 (refs/changes/87/3887/1)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2018-04-30T13:09:56-07:00 (1 year, 5 months ago)
    
    TRITON-356 VMAPI should use node v6.x

diff --git a/Makefile b/Makefile
index e8bbe11..e617edb 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -24,7 +24,7 @@
 
 # The prebuilt sdcnode version we want. See
 # "tools/mk/Makefile.node_prebuilt.targ" for details.
-NODE_PREBUILT_VERSION=v4.6.1
+NODE_PREBUILT_VERSION=v6.14.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than sdc-minimal-multiarch-lts@15.4.1
diff --git a/package.json b/package.json
index fad45ac..d4e04b3 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.5.1",
+  "version": "9.5.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
