commit e28382d9696db5cf9c6e182e1722f005578acedf (refs/changes/30/1430/2)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2017-02-03T15:37:37-08:00 (2 years, 8 months ago)
    
    AGENT-1059 cn-agent should not log docker_build payloads

diff --git a/lib/app.js b/lib/app.js
index 4f74c26..9443dda 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var async = require('async');
@@ -327,11 +327,24 @@ App.prototype.start = function () {
             onhttpmsg: createHttpTaskDispatchFn(agent, taskspath),
             tasks: [
                 'docker_exec',
-                'docker_build',
                 'docker_copy',
                 'docker_stats'
             ]
         },
+        {
+            name: 'docker_build_task',
+            onhttpmsg: createHttpTaskDispatchFn(agent, taskspath),
+
+            /**
+             * `docker build` payloads, particularly those with a large range
+             * of exposed ports, tend to be HUGE and negatively impact the
+             * service when logged.
+             */
+            log_params: false,
+            tasks: [
+                'docker_build'
+            ]
+        },
         {
             name: 'server_nic_tasks',
             onhttpmsg: createHttpTaskDispatchFn(agent, taskspath),
diff --git a/lib/task_agent/task_agent.js b/lib/task_agent/task_agent.js
index ffc8739..2377d1a 100644
--- a/lib/task_agent/task_agent.js
+++ b/lib/task_agent/task_agent.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var util = require('util');
@@ -88,14 +88,25 @@ TaskAgent.prototype.setupTaskRoutes = function (defns) {
         }
 
         var dispatch = {};
+        var taskName = req.params.task;
+        var logParams = true;
 
         self.queueDefns.forEach(function (i) {
             i.tasks.forEach(function (j) {
+                if (j === taskName && i.log_params === false) {
+                    logParams = false;
+                }
+
                 dispatch[j] = i.onhttpmsg;
             });
         });
 
-        req.log.info({ task: req.params }, 'task params');
+        if (logParams) {
+            req.log.info({ task: req.params }, '%s task params', taskName);
+        } else {
+            req.log.info(
+                'not logging task params for %s (log_params=false)', taskName);
+        }
 
         var value, error;
 
