commit e257fa2621343c26e1422297fecf826ba93db76c (refs/changes/77/5177/3)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-14T14:39:36-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index b571858..7a78de8 100644
--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,7 @@ JSSTYLE_FLAGS	= -o indent=4,doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/cnapi.xml.in
 SMF_DTD		= $(REPO_ROOT)/tools/service_bundle.dtd.1
 
-NODE_PREBUILT_VERSION=v6.14.0
+NODE_PREBUILT_VERSION=v6.15.1
 NODE_PREBUILT_TAG=zone
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
diff --git a/package.json b/package.json
index 10bcb5a..f61b720 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.17.1",
+  "version": "1.17.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
