From 00ec357362fbc7c8de18a72ebbf5563651189098 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Wed, 11 Jan 2017 10:22:05 -0800
Subject: [PATCH] MANTA-3080 muskie picker only picks from 200 nodes

---
 lib/picker.js | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/picker.js b/lib/picker.js
index 5f72e72..c32197d 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -82,6 +82,7 @@ var fetch = function fetch_moray(opts, cb) {
     cb = once(cb);
 
     var count = 0;
+    var recs = 0;
     var f = sprintf('(&(percentUsed<=%d)(timestamp>=%d)%s)',
                     opts.utilization || 90,
                     Date.now() - opts.lag,
@@ -103,10 +104,11 @@ var fetch = function fetch_moray(opts, cb) {
         values.push(data.value);
         count = data._count;
         marker = data._id;
+        recs++;
     });
 
     req.once('end', function () {
-        if (values.length < count) {
+        if (recs < count) {
             var next = {
                 lag: opts.lag,
                 limit: opts.limit,
-- 
2.21.0

