commit 941c31052d33800a5a0e77e29b4244376f19cef5 (refs/changes/50/650/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-10-11T13:24:26-07:00 (3 years ago)
    
    MANTA-3000 node-mahi should support "agent" options

diff --git a/lib/client.js b/lib/client.js
index 358d423..126c6e5 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -43,6 +43,7 @@ var ANONYMOUS_USER = 'anonymous';
 function MahiClient(opts) {
     assert.object(opts, 'opts');
     assert.string(opts.url, 'url');
+    assert.optionalObject(opts.agent, 'agent');
     assert.optionalObject(opts.typeTable, 'typeTable');
     assert.optionalNumber(opts.maxAuthCacheSize, 'maxAuthCacheSize');
     assert.optionalNumber(opts.maxAuthCacheAgeMs, 'maxAuthCacheAgeMs');
@@ -66,7 +67,8 @@ function MahiClient(opts) {
     });
 
     var clientOpts = {
-        url: opts.url
+        url: opts.url,
+        agent: opts.agent
     };
 
     if (opts.log) {
diff --git a/package.json b/package.json
index d112952..9c816c9 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "mahi",
     "description": "Mahi client",
-    "version": "2.0.0",
+    "version": "2.0.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "main": "index.js",
