commit e54810720958e817107928df5f949d2b6cf4d77a (refs/changes/26/5026/2)
Author: Jeff Emershaw <me@jeffemershaw.com>
Date:   2018-11-05T14:20:28-05:00 (11 months ago)
    
    mchmod issues with parseCmdOptions

diff --git a/CHANGES.md b/CHANGES.md
index 204ca2d..7d9a1a8 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,7 @@ See `CONTRIBUTING.md` for details on how to update this file
 
 ## not yet released
 
+- [#358](https://github.com/joyent/node-manta/issues/358) mchmod command stopped working after v5.1.1
 - [#349](https://github.com/joyent/node-manta/issues/349) issue numbers in
   CHANGES.md should link to GitHub issues
 - [#335](https://github.com/joyent/node-manta/issues/335) Want option to confim
diff --git a/lib/options.js b/lib/options.js
index 3f7fdef..ebe11f0 100644
--- a/lib/options.js
+++ b/lib/options.js
@@ -84,12 +84,13 @@ function parseOptions(args) {
                         assert.ifError(utils.assertPath(p, true));
                         return (client.path(p, true));
                     };
-    opts.paths = opts._args.map(getPath);
 
     if (args.parseCmdOptions) {
         args.parseCmdOptions(opts, parser);
     }
 
+    opts.paths = opts._args.map(getPath);
+
     try {
         cc.checkBinEnv(opts);
     } catch (e) {
