commit 911ae54186a356b2e814bab18b9e5d91d696175a (refs/changes/26/5026/1)
Author: Jeff Emershaw <me@jeffemershaw.com>
Date:   2018-11-03T00:16:47-04:00 (11 months ago)
    
    mchmod issues with parseCmdOptions

diff --git a/lib/options.js b/lib/options.js
index 3f7fdef..ebe11f0 100644
--- a/lib/options.js
+++ b/lib/options.js
@@ -84,12 +84,13 @@ function parseOptions(args) {
                         assert.ifError(utils.assertPath(p, true));
                         return (client.path(p, true));
                     };
-    opts.paths = opts._args.map(getPath);
 
     if (args.parseCmdOptions) {
         args.parseCmdOptions(opts, parser);
     }
 
+    opts.paths = opts._args.map(getPath);
+
     try {
         cc.checkBinEnv(opts);
     } catch (e) {
