From d5250d33e746a36a2c1e735dc56e6334b26ee31b Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Mon, 19 Dec 2016 18:12:49 -0800
Subject: [PATCH] fix error messages

---
 lib/cmd.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/cmd.js b/lib/cmd.js
index f754124..49266b6 100644
--- a/lib/cmd.js
+++ b/lib/cmd.js
@@ -194,7 +194,7 @@ function parseCommonCliOption(errstream, options, option) {
     case 'p':
         p = parseTcpPort(option.optarg);
         if (p === null) {
-            fprintf(errstream, '-p/--port: expected valid TCP port');
+            fprintf(errstream, '-p/--port: expected valid TCP port\n');
             return (false);
         }
         options.port = p;
@@ -236,7 +236,7 @@ function finalizeCliOptions(errstream, options, env) {
         /* The user specified -s/--service. */
         if (options.port !== undefined || options.host !== undefined) {
             fprintf(errstream, '-S/--service cannot be combined with ' +
-                '-h/--host or -p/--port');
+                '-h/--host or -p/--port\n');
             return (false);
         }
 
@@ -279,7 +279,7 @@ function finalizeCliOptions(errstream, options, env) {
 function validateSrvDomain(errstream, domain) {
     if (net.isIP(domain)) {
         fprintf(errstream,
-            'cannot use an IP address with -S/--service/MORAY_SERVICE');
+            'cannot use an IP address with -S/--service/MORAY_SERVICE\n');
         return (false);
     }
 
@@ -311,7 +311,7 @@ function populateDirectArguments(errstream, options, env) {
                 p = parseTcpPort(u['port']);
                 if (p === null) {
                     fprintf(errstream,
-                        'port in MORAY_URL is not a valid TCP port');
+                        'port in MORAY_URL is not a valid TCP port\n');
                     return (false);
                 }
 
-- 
2.21.0

