commit 4977bd0a57154801c45f12add34648553fa51c4e (refs/changes/21/1921/2)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-05-18T22:21:05+00:00 (2 years, 5 months ago)
    
    DOCKER-1051 fix docker ps --filter since/before filtering

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index ee9966d..95ce635 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -2666,6 +2666,13 @@ function getContainers(opts, callback) {
                 stash.objects = objects;
                 cb();
             });
+
+        // DEPRECATED API v1.24 -- start
+        //
+        // Note that the 'before' and 'since' query param filtering is
+        // deprecated and has been removed in docker client 1.12 (API v1.24).
+        // The 'before' and 'since' handing is now done through the docker ps
+        // --filter options instead.
         }, function _filterSince(stash, cb) {
             var match;
             var objects = stash.objects;
@@ -2705,6 +2712,8 @@ function getContainers(opts, callback) {
             stash.objects = objects.slice(objects.indexOf(match) + 1);
 
             cb();
+        // DEPRECATED API v1.24 -- end
+
         }, function _preloadPackages(stash, cb) {
             var filteringOnPkgs = false;
             var i;
@@ -2780,6 +2789,7 @@ function getContainers(opts, callback) {
 
             Object.keys(filters).forEach(function _filterField(field) {
                 var labelFilters;
+                var match;
                 var val = filters[field];
                 var val_to_sdc_status;
 
@@ -2903,6 +2913,19 @@ function getContainers(opts, callback) {
                             return false;
                         });
                     }
+                } else if (field === 'before' || field === 'since') {
+                    match = findContainerIdMatch(val[0], objects);
+                    if (!match) {
+                        filterErr = new errors.DockerError(
+                            'Could not find container with name or id '
+                            + val[0]);
+                        return;
+                    }
+                    if (field === 'before') {
+                        objects = objects.slice(objects.indexOf(match) + 1);
+                    } else {
+                        objects = objects.slice(0, objects.indexOf(match));
+                    }
                 } else {
                     log.warn('Unhandled docker filter name:', field);
                 }
