From 5a825744c54d39da39887c864d7a68fd35add360 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 2 Aug 2016 06:19:38 +0000
Subject: [PATCH] DOCKER-903 "docker build" broken by DOCKER-585

---
 lib/backends/sdc/containers.js | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 596b776..248274d 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -1959,6 +1959,15 @@ function buildVmPayload(opts, container, callback) {
         },
 
         function addExternalNet(_, cb) {
+            /*
+             * If fabrics aren't enabled, then addNetworksToPayload()'s
+             * already added an external network.
+             */
+            if (!opts.config.overlay.enabled) {
+                cb();
+                return;
+            }
+
             if (publishingPorts(container) || opts.fabricRequireExternal) {
                 assert.object(opts.config.overlay, 'opts.config.overlay');
                 assert.string(opts.config.overlay.externalPool,
-- 
2.21.0

