commit 5a825744c54d39da39887c864d7a68fd35add360 (refs/changes/90/190/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-08-02T06:19:38+00:00 (3 years, 2 months ago)
    
    DOCKER-903 "docker build" broken by DOCKER-585

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 596b776..248274d 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -1959,6 +1959,15 @@ function buildVmPayload(opts, container, callback) {
         },
 
         function addExternalNet(_, cb) {
+            /*
+             * If fabrics aren't enabled, then addNetworksToPayload()'s
+             * already added an external network.
+             */
+            if (!opts.config.overlay.enabled) {
+                cb();
+                return;
+            }
+
             if (publishingPorts(container) || opts.fabricRequireExternal) {
                 assert.object(opts.config.overlay, 'opts.config.overlay');
                 assert.string(opts.config.overlay.externalPool,
