From e99601f8dcdb0504e29b1eb561398449627f2e2f Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 20 Jul 2018 15:22:14 -0700
Subject: [PATCH] TRITON-586 long live the new mockcloud

---
 targets.json.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/targets.json.in b/targets.json.in
index 2ed8c56..67c637a 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -236,7 +236,7 @@ cat <<EOF
     "image_uuid": "b6ea7cb4-6b90-48c0-99e7-1d34c2895248",
     "pkgsrc": [],
     "repos": [
-        {"url": "git@github.com:joyent/sdc-mockcloud.git", "dir": "mockcloud"}
+        {"url": "git@github.com:joyent/triton-mockcloud.git", "dir": "mockcloud"}
     ],
     "public": false,
     "deps": [],
-- 
2.21.0

