From a57e0eaa198414a91d2c1594cc99f06f7832bca1 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 27 Jan 2017 12:59:25 +0100
Subject: [PATCH] TOOLS-1664 sdcadm post-setup cmon does not complete setup

---
 lib/post-setup/cmon.js | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/lib/post-setup/cmon.js b/lib/post-setup/cmon.js
index 47efb8f..a4aecaf 100644
--- a/lib/post-setup/cmon.js
+++ b/lib/post-setup/cmon.js
@@ -178,10 +178,12 @@ function do_cmon(subcmd, opts, args, cb) {
                     function (err, img_) {
                 if (err && err.body && err.body.code === 'ResourceNotFound') {
                     ctx.imgsToDownload.push(ctx.cmonImg);
+                    next();
                 } else if (err) {
-                    return next(err);
+                    next(err);
+                } else {
+                    next();
                 }
-                next();
             });
         },
 
@@ -191,11 +193,12 @@ function do_cmon(subcmd, opts, args, cb) {
             }, function (err, images) {
                 if (err) {
                     next(err);
+                    return;
                 } else if (images && images.length) {
                     // TODO presuming sorted
                     ctx.cmonAgentImg = images[images.length - 1];
-                    next();
                 }
+                next();
             });
         },
 
-- 
2.21.0

