commit 6c344f598b3b4fc89decd5115ed2b0049796f01c (refs/changes/23/3423/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-02-19T15:58:45-08:00 (1 year, 8 months ago)
    
    TRITON-146 modernize SAPI: recent node, triton-origin image, restify 4.x

diff --git a/targets.json.in b/targets.json.in
index aa23beb..3130c10 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -863,7 +863,8 @@ cat <<EOF
     "image_name": "sapi",
     "image_description": "SDC SAPI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
+    "// image_uuid": "triton-origin-multiarch-15.4.1@1.0.1",
+    "image_uuid": "04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
     "pkgsrc": [],
     "repos": [
         {"url": "git@github.com:joyent/sdc-sapi.git"}
diff --git a/tools/launch-build b/tools/launch-build
index f4e3e55..5bd1fa4 100755
--- a/tools/launch-build
+++ b/tools/launch-build
@@ -16,6 +16,13 @@
 # or the -u option, in the form <user>:<api token>.
 #
 
+if [[ -n "$TRACE" ]]; then
+	export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+	set -o xtrace
+fi
+set -o errexit
+set -o pipefail
+
 PROGNAME=$0
 
 function usage() {
