From 8d7f95b677c5b0dda2023bc8e70268e2a3c10f96 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Thu, 16 Nov 2017 16:41:28 -0800
Subject: [PATCH] joyent/manta-thoth#157 building in manta is broken

---
 tools/manta-thoth-build.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/manta-thoth-build.sh b/tools/manta-thoth-build.sh
index 64bbc0e..f2bd9d6 100755
--- a/tools/manta-thoth-build.sh
+++ b/tools/manta-thoth-build.sh
@@ -3,7 +3,7 @@
 set -o errexit
 
 job=$(
-mjob create --close -m 'set -o errexit ;
+mjob create --close -r 'set -o errexit ;
 cd /var/tmp ;
 git clone https://github.com/joyent/manta-thoth ;
 cd manta-thoth ;
-- 
2.21.0

