commit 8d7f95b677c5b0dda2023bc8e70268e2a3c10f96 (refs/changes/61/2961/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2017-11-16T16:41:28-08:00 (1 year, 11 months ago)
    
    joyent/manta-thoth#157 building in manta is broken

diff --git a/tools/manta-thoth-build.sh b/tools/manta-thoth-build.sh
index 64bbc0e..f2bd9d6 100755
--- a/tools/manta-thoth-build.sh
+++ b/tools/manta-thoth-build.sh
@@ -3,7 +3,7 @@
 set -o errexit
 
 job=$(
-mjob create --close -m 'set -o errexit ;
+mjob create --close -r 'set -o errexit ;
 cd /var/tmp ;
 git clone https://github.com/joyent/manta-thoth ;
 cd manta-thoth ;
