commit a4ddbb2acb0ac1fdf133a37856d7d63da9102956 (refs/changes/80/180/1)
Author: Matt Smillie <matt.smillie@joyent.com>
Date:   2016-07-31T01:02:22-07:00 (3 years, 2 months ago)
    
    DOCKER-901: `docker -p` causes assertion failure without fabrics enabled

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index 596b776..ec5d46e 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -1959,7 +1959,12 @@ function buildVmPayload(opts, container, callback) {
         },
 
         function addExternalNet(_, cb) {
-            if (publishingPorts(container) || opts.fabricRequireExternal) {
+            assert.object(opts.config, 'config');
+            assert.object(opts.config.overlay, 'config.overlay');
+            assert.object(opts.config.overlay.enabled,
+                'config.overlay.enabled');
+            if (opts.config.overlay.enabled
+                && (publishingPorts(container) || opts.fabricRequireExternal)) {
                 assert.object(opts.config.overlay, 'opts.config.overlay');
                 assert.string(opts.config.overlay.externalPool,
                     'opts.config.overlay.externalPool');
