commit 9edb934fbbe849a1a32e68d6602f1c72597639d8 (refs/changes/65/2865/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-10-25T12:10:12-07:00 (1 year, 11 months ago)
    
    TOOLS-1898 sdcnode updates for Node.js Oct 2017 vulns

diff --git a/Makefile b/Makefile
index 259b6e5..3e67c0b 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,7 @@ JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 SMF_MANIFESTS_IN = smf/manifests/docker.xml.in
 CLEAN_FILES += ./node_modules
 
-NODE_PREBUILT_VERSION=v4.6.1
+NODE_PREBUILT_VERSION=v4.8.5
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
