commit f62920ae4ef4c9f40b14975a75eb706e7a0b62f2 (refs/changes/47/347/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-08-25T22:35:45+00:00 (3 years, 1 month ago)
    
    OS-5621 platform build should run check_rtime

diff --git a/Makefile b/Makefile
index 500967b5..fb825e39 100644
--- a/Makefile
+++ b/Makefile
@@ -10,6 +10,7 @@ BOOT_MPROTO =	$(ROOT)/boot.manifest.d
 BOOT_PROTO =	$(ROOT)/proto.boot
 IMAGES_PROTO =	$(ROOT)/proto.images
 MCPROTO =	$(ROOT)/mancheck.conf.d
+EXPROTO =	$(ROOT)/ex_list.d
 
 # On Darwin/OS X we support running 'make check'
 ifeq ($(shell uname -s),Darwin)
@@ -78,6 +79,15 @@ WORLD_MANCHECK_CONFS := \
 	$(MCPROTO)/live.mancheck.conf \
 	$(MCPROTO)/illumos-extra.mancheck.conf
 
+WORLD_EXLIST_RTIME := \
+	$(EXPROTO)/illumos.check_rtime.ex \
+	$(EXPROTO)/live.check_rtime.ex \
+	$(EXPROTO)/illumos-extra.check_rtime.ex
+
+ILLUMOS_EXLIST :=	projects/illumos/exception_lists/check_rtime
+EXTRA_EXLIST := 	projects/illumos-extra/exception_lists/check_rtime.ex
+RTIME_EXLIST_GEN :=	$(EXPROTO)/check_rtime.ex.gen
+
 BOOT_MANIFESTS := \
 	$(BOOT_MPROTO)/illumos.manifest
 
@@ -89,6 +99,9 @@ SUBDIR_MANCHECK_CONFS := \
 OVERLAY_MANCHECK_CONFS := \
 	$(OVERLAYS:$(ROOT)/overlay/%=$(MCPROTO)/%.ov.mancheck.conf)
 
+SUBDIR_EXLIST_RTIME_CONFS := \
+	$(LOCAL_SUBDIRS:%=$(EXPROTO)/%.sd.check_rtime.ex)
+
 BOOT_VERSION :=	boot-$(shell [[ -f $(ROOT)/configure-buildver ]] && \
     echo $$(head -n1 $(ROOT)/configure-buildver)-)$(shell head -n1 $(STAMPFILE))
 BOOT_TARBALL :=	output/$(BOOT_VERSION).tgz
@@ -176,7 +189,7 @@ dump_mancheck_conf: manifest mancheck_conf $(MANCHECK)
 	    args="$$args -c $$x"; done; \
 	    $(MANCHECK) -f manifest.gen -s -D $$args
 
-$(MPROTO) $(BOOT_MPROTO) $(MCPROTO):
+$(MPROTO) $(BOOT_MPROTO) $(MCPROTO) $(EXPROTO):
 	mkdir -p $@
 
 $(MPROTO)/live.manifest: src/manifest | $(MPROTO)
@@ -187,12 +200,18 @@ $(MCPROTO)/live.mancheck.conf: src/mancheck.conf | $(MCPROTO)
 	gmake DESTDIR=$(MCPROTO) DESTNAME=live.mancheck.conf \
 	    -C src mancheck_conf
 
+$(EXPROTO)/live.check_rtime.ex: src/check_rtime.ex | $(EXPROTO)
+	cp src/check_rtime.ex $@
+
 $(MPROTO)/illumos.manifest: projects/illumos/manifest | $(MPROTO)
 	cp projects/illumos/manifest $(MPROTO)/illumos.manifest
 
 $(MCPROTO)/illumos.mancheck.conf: projects/illumos/mancheck.conf | $(MCPROTO)
 	cp projects/illumos/mancheck.conf $(MCPROTO)/illumos.mancheck.conf
 
+$(EXPROTO)/illumos.check_rtime.ex: $(ILLUMOS_EXLIST) | $(EXPROTO)
+	cp $(ILLUMOS_EXLIST) $@
+
 $(BOOT_MPROTO)/illumos.manifest: projects/illumos/manifest | $(BOOT_MPROTO)
 	cp projects/illumos/boot.manifest $(BOOT_MPROTO)/illumos.manifest
 
@@ -204,6 +223,9 @@ $(MCPROTO)/illumos-extra.mancheck.conf: FRC | 1-extra-stamp $(MCPROTO)
 	gmake DESTDIR=$(MCPROTO) DESTNAME=illumos-extra.mancheck.conf \
 	    -C projects/illumos-extra mancheck_conf; \
 
+$(EXPROTO)/illumos-extra.check_rtime.ex: $(EXTRA_EXLIST) | $(EXPROTO)
+	cp $(EXTRA_EXLIST) $@
+
 $(MPROTO)/%.sd.manifest: projects/local/%/Makefile projects/local/%/manifest
 	cd $(ROOT)/projects/local/$* && \
 	    if [[ -f Makefile.joyent ]]; then \
@@ -224,6 +246,14 @@ $(MCPROTO)/%.sd.mancheck.conf: FRC | $(MCPROTO)
 		    mancheck_conf; \
 	    fi
 
+$(EXPROTO)/%.sd.check_rtime.ex: FRC | $(EXPROTO)
+	cd $(ROOT)/projects/local/$* && \
+	    if [[ -f check_rtime.ex ]]; then \
+		cp check_rtime.ex $@; \
+	    else \
+		/bin/true; \
+	    fi
+
 $(MPROTO)/%.ov.manifest: $(MPROTO) $(ROOT)/overlay/%/manifest
 	cp $(ROOT)/overlay/$*/manifest $@
 
@@ -278,7 +308,10 @@ update-base:
 	    (cd projects/devpro && gmake DESTDIR=$(PROTO) install)
 	touch $@
 
-0-illumos-stamp: 0-extra-stamp
+$(RTIME_EXLIST_GEN): $(WORLD_EXLIST_RTIME) $(SUBDIR_EXLIST_RTIME_CONFS)
+	cat $(EXPROTO)/*.ex > $@
+
+0-illumos-stamp: 0-extra-stamp $(RTIME_EXLIST_GEN)
 	(cd $(ROOT) && MAX_JOBS=$(MAX_JOBS) ./tools/build_illumos)
 	touch $@
 
diff --git a/configure b/configure
index ffed9f51..9ee09099 100755
--- a/configure
+++ b/configure
@@ -221,7 +221,7 @@ function fetch_closed
 
 function generate_env
 {
-	local nopts="-CimMNnt"
+	local nopts="-CimMNntr"
 	local lint debug lprefix
 
 	if [[ -n "$ILLUMOS_ENABLE_LINT" ]]; then
@@ -315,6 +315,7 @@ YACC=/opt/local/bin/yacc;			export YACC
 RPCGEN=/opt/local/bin/rpcgen;			export RPCGEN
 ASTBINDIR=/opt/local/ast/bin;			export ASTBINDIR
 LD_TOXIC_PATH="\$ROOT/lib:\$ROOT/usr/lib";	export LD_TOXIC_PATH
+CHECK_RTIME_EXLIST="$conf_root/ex_list.d/check_rtime.ex.gen"; export CHECK_RTIME_EXLIST
 EOF
 	[[ $? -eq 0 ]] || fatal "failed to write illumos nightly env file"
 }
diff --git a/src/check_rtime.ex b/src/check_rtime.ex
new file mode 100644
index 00000000..93b3fa1d
--- /dev/null
+++ b/src/check_rtime.ex
@@ -0,0 +1,57 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2016 Joyent, Inc.
+#
+
+#
+# Exceptions for check_rtime logic. This list should be whittled down
+# over time. For an explanation, see check_rtime(1ONBLD) or
+# exception_lists/check_rtime in illumos-joyent.
+#
+
+SKIP	^smartdc/bin/qemu-exec
+SKIP	^smartdc/lib/sdc-on-tty
+SKIP	^usr/bin/bootparams
+
+#
+# It's not clear why we have these node_modules or the nomknod bits
+# there either.
+#
+SKIP	^usr/bin/.*\.node$
+SKIP	^usr/bin/nomknod\.so\.32$
+SKIP	^usr/bin/nomknod\.so\.64$
+SKIP	^usr/bin/sysinfo_mod\.so$
+
+SKIP	^usr/bin/disk_size$
+SKIP	^usr/bin/diskinfo$
+SKIP	^usr/bin/removable_disk$
+SKIP	^usr/bin/sysevent$
+SKIP	^usr/bin/vmunbundle$
+SKIP	^usr/bin/zfs_recv$
+SKIP	^usr/bin/zfs_send$
+SKIP	^usr/bin/zoneevent$
+SKIP	^usr/lib/cryptpass$
+SKIP	^usr/MACH(lib)/libC\.so\.5$
+SKIP	^usr/MACH(lib)/libCrun\.so\.1$
+SKIP	^usr/MACH(lib)/libCstd\.so\.1$
+SKIP	^usr/MACH(lib)/libdemangle\.so\.1$
+SKIP	^usr/MACH(lib)/libiostream\.so\.1$
+
+SKIP	^usr/lib/brand/lx/routeinfo$
+SKIP	^usr/lib/measure_terminal$
+SKIP	^usr/lib/sysevent/modules/sysinfo_mod.so$
+
+SKIP	^usr/fw
+SKIP	^usr/img
+SKIP	^usr/node
+SKIP	^usr/vm
