commit eee22bb8f8cbc397ed308bd59ee3a9f7719a8ede (refs/changes/04/2804/1)
Author: Jordan Hendricks <jordan.hendricks@joyent.com>
Date:   2017-09-27T23:45:26+00:00 (2 years ago)
    
    Added MPU moray cleaner tests

diff --git a/Makefile b/Makefile
index 50ff1c0..6053521 100644
--- a/Makefile
+++ b/Makefile
@@ -25,7 +25,7 @@
 #
 # Tools
 #
-NODEUNIT        := ./node_modules/.bin/nodeunit
+NODEUNIT        := ./node_modules/.bin/nodeunit --reporter=tap
 NPM             := npm
 
 #
diff --git a/lib/mpu/mpuMorayCleanerStream.js b/lib/mpu/mpuMorayCleanerStream.js
index 44d0853..b28532f 100644
--- a/lib/mpu/mpuMorayCleanerStream.js
+++ b/lib/mpu/mpuMorayCleanerStream.js
@@ -83,17 +83,20 @@ MpuMorayCleanerStream.prototype._write = function mmcsWrite(batch, _, cb) {
         if (!self.dryRun) {
                 self.deleteFinalizingRecord(fr.shard, fr.key, function (err) {
                         if (err) {
+                                /*
+                                 * We don't want to throw an error here in case
+                                 * this is an isolated problem, so log an error
+                                 * and continue.
+                                 */
                                 self.log.error({
                                         uploadId: batch.uploadId,
                                         err: err
                                 }, 'mpu moray cleaner stream failure');
-                                cb(err);
-                        } else {
-                                cb();
                         }
+
+                        cb();
                 });
         } else {
-                //self.log.info('mmcs cb: ' + fr.key);
                 cb();
         }
 };
diff --git a/test/mpu/mpuUnlinkLiveRecordStream.test.js b/test/mpu/mpuUnlinkLiveRecordStream.test.js
index 30fe86f..592e824 100644
--- a/test/mpu/mpuUnlinkLiveRecordStream.test.js
+++ b/test/mpu/mpuUnlinkLiveRecordStream.test.js
@@ -766,7 +766,7 @@ test('parts: one batch (3 parts)', function (t) {
         testMpuUnlinkLiveRecordStream(args);
 });
 
-test('parts: multiple batches (3 parts, 1', function (t) {
+test('parts: multiple batches (1 part, 3 parts, 0 parts)', function (t) {
         var input = [
                 {
                         uploadId: inputs.ID_0,
diff --git a/test/mpu/testInputs.js b/test/mpu/testInputs.js
index 9c605c6..37288d7 100644
--- a/test/mpu/testInputs.js
+++ b/test/mpu/testInputs.js
@@ -149,7 +149,7 @@ var PATH_UR2 = '/' + ACCT_LOGIN_2 +
 
 var ID_2 = '38aecc30-9a8c-63a4-f906-e512f02f5915';
 var DATE_2 =  new Date(Date.now()).toISOString(); // TODO is this right
-var SHARD_2 =  MORAY_1;
+var SHARD_2 =  MORAY_2;
 /* BEGIN JSSTYLED */
 var KEY_FR2 = '38aecc30-9a8c-63a4-f906-e512f02f5915:/88af09d7-4845-e09a-8998-d7d04a88b879/stor/batch2';
 var KEY_UR2 = '/88af09d7-4845-e09a-8998-d7d04a88b879/uploads/3/38aecc30-9a8c-63a4-f906-e512f02f5915';
