commit 83c52b609a8bb397051184f6db9d60d75aeb00bd (refs/changes/64/3964/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2018-05-16T22:20:48+00:00 (1 year, 5 months ago)
    
    TRITON-378 Drop smartos-overlay from mountain-gorilla

diff --git a/targets.json.in b/targets.json.in
index 7b5da7e..92f2e10 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1557,7 +1557,6 @@ cat <<EOF
         {"url": "git@github.com:joyent/illumos-extra.git"},
         {"url": "git@github.com:joyent/illumos-kvm.git"},
         {"url": "git@github.com:joyent/illumos-kvm-cmd.git"},
-        {"url": "git@github.com:joyent/smartos-overlay.git"},
         {"url": "git@github.com:joyent/mdata-client.git"}
     ],
     "public": true,
