From d882a304c4b2338a0bd9754db4fb23f4360c4925 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 17 Jan 2017 22:55:08 +0000
Subject: [PATCH] MANTA-3095 mako test suite never reads its HTTP streams

---
 test/api.test.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/test/api.test.js b/test/api.test.js
index 5e86f06..acbf24d 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -54,6 +54,7 @@ var getNonexistentObject = function (t) {
         var req = http.request(options, function (res) {
                 console.log('STATUS: ' + res.statusCode);
                 console.log('HEADERS: ' + JSON.stringify(res.headers));
+                res.resume();
                 t.equal(res.statusCode, 404);
                 t.end();
         });
@@ -84,6 +85,7 @@ test('put 10 MiB object', function (t) {
         var req = http.request(options, function (res) {
                 console.log('STATUS: ' + res.statusCode);
                 console.log('HEADERS: ' + JSON.stringify(res.headers));
+                res.resume();
                 t.equal(res.statusCode, 201);
                 t.end();
         });
@@ -155,6 +157,7 @@ test('100s of small files', function (t) {
 
                         var req = http.request(options,
                             function (res) {
+                                res.resume();
                                 t.equal(res.statusCode, 201);
                                 return (subcb(null));
                         });
-- 
2.21.0

