commit 2ea4ba2424413850892bf7d3301fea682a3d5b8b (refs/changes/82/5282/1)
Author: Rui Loura <rui@joyent.com>
Date:   2018-12-21T17:18:51+00:00 (10 months ago)
    
    TRITON-1058 cnapi server uuid validation should handle "default"

diff --git a/lib/endpoints/index.js b/lib/endpoints/index.js
index 318b691..072d155 100644
--- a/lib/endpoints/index.js
+++ b/lib/endpoints/index.js
@@ -119,13 +119,15 @@ function prepopulateServer(opts, req, res, next) {
     assert.object(req, 'req');
     assert.object(req.params, 'req.params');
 
-    if (!isUuid(req.params.server_uuid)) {
+    var server_uuid = req.params.server_uuid;
+
+    if (server_uuid !== 'default' && !isUuid(server_uuid)) {
         next(new restify.InvalidArgumentError('Invalid server_uuid'));
         return;
     }
 
     ModelServer.get(
-        req.params.server_uuid,
+        server_uuid,
         function (err, servermodel, server) {
             if (err) {
                 req.log.error(err);
@@ -135,8 +137,7 @@ function prepopulateServer(opts, req, res, next) {
 
             // Check if any servers were returned
             if (!server) {
-                var errorMsg = 'Server ' + req.params.server_uuid
-                    + ' not found';
+                var errorMsg = 'Server ' + server_uuid + ' not found';
                 next(new restify.ResourceNotFoundError(errorMsg));
                 return;
             }
