From 1c786eaacf85acfc7335cc42e5e02bba50f33300 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Wed, 25 Oct 2017 09:46:06 -0600
Subject: [PATCH] MANTA-3474 Use faucet for muskie test output formatting

---
 Makefile     | 3 ++-
 package.json | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 9544f59..7763891 100644
--- a/Makefile
+++ b/Makefile
@@ -37,6 +37,7 @@ NAME 			= muskie
 NODE_PREBUILT_TAG       = zone
 NODE_PREBUILT_VERSION	:= v0.10.48
 NODE_PREBUILT_IMAGE     = fd2cc906-8938-11e3-beab-4359c665ac99
+FAUCET       = ./node_modules/.bin/faucet
 
 include ./tools/mk/Makefile.defs
 include ./tools/mk/Makefile.node_prebuilt.defs
@@ -64,7 +65,7 @@ manta-scripts: deps/manta-scripts/.git
 .PHONY: test
 test: $(STAMP_NODE_MODULES)
 	$(NODE) ./node_modules/.bin/nodeunit --reporter=tap \
-	    test/*.test.js test/mpu/*.test.js
+		test/*.test.js test/mpu/*.test.js | $(FAUCET)
 
 .PHONY: release
 release: all docs
diff --git a/package.json b/package.json
index 490ff4a..93df50d 100644
--- a/package.json
+++ b/package.json
@@ -21,6 +21,7 @@
         "dashdash": "1.3.2",
         "deep-equal": "0.0.0",
         "dtrace-provider": "0.2.8",
+        "faucet": "0.0.1",
         "http-signature": "1.1.0",
         "jsprim": "^1.3.1",
         "kang": "1.1.0",
-- 
2.21.0

