commit 140c0429c86b50d58847a61891bb6263c45f701c (refs/changes/06/4406/1)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2018-06-19T18:46:58+00:00 (1 year, 4 months ago)
    
    TRITON-373 disable directory listing on assets zone nginx

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index c5775c7..030f7a2 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -45,7 +45,6 @@ http {
 	send_timeout	300s;
 
 	server {
-		autoindex 	on;
 		listen		80 so_keepalive=10s:30s:10;
 		root		/manta;
 		server_name	localhost;
