commit 387c3e136ebcde4cfd63561c0ae74049a40ecda1 (refs/changes/54/1754/2)
Author: David Pacheco <dap@joyent.com>
Date:   2017-04-07T13:38:19-07:00 (2 years, 6 months ago)
    
    MANTA-3111 marlin agent cueball reports "read ECONNRESET" and 408 errors

diff --git a/agent/lib/agent/agent.js b/agent/lib/agent/agent.js
index c4a37ab..2a89209 100644
--- a/agent/lib/agent/agent.js
+++ b/agent/lib/agent/agent.js
@@ -257,6 +257,8 @@ var maRequestTimeout 		= 300000;	/* local long poll timeout */
 var maUpstreamSpareConnsDefault		= 16;
 var maUpstreamMaxConnsDefault		= 512;
 var maUpstreamTcpKeepAliveDelayDefault	= 15000;
+var maUpstreamPingPath			= '/';
+var maUpstreamPingInterval		= 60000;
 
 var maUpstreamSrvMaxTimeout		= 5000;
 
@@ -464,11 +466,17 @@ function mAgent(filename)
 	    maUpstreamSpareConnsDefault;
 	this.ma_http_keepalive_delay = tunables['httpTcpKeepAliveDelay'] ||
 	    maUpstreamTcpKeepAliveDelayDefault;
+	this.ma_http_ping_path = tunables['httpPingPath'] ||
+	    maUpstreamPingPath;
+	this.ma_http_ping_interval = tunables['httpPingInterval'] ||
+	    maUpstreamPingInterval;
 	this.ma_log.info({
 	    'config': this.ma_conf,
 	    'http_nmaxconns': this.ma_http_nmaxconns,
 	    'http_nspares': this.ma_http_nspares,
 	    'http_keepalive_delay': this.ma_http_keepalive_delay,
+	    'http_ping_path': this.ma_http_ping_path,
+	    'http_ping_interval': this.ma_http_ping_interval,
 	    'cueball_recovery': this.ma_cueball_recovery
 	}, 'configuration');
 
@@ -518,6 +526,9 @@ function mAgent(filename)
 	    'log': this.ma_log.child({ 'component': 'CueballAgent' }),
 	    'spares': this.ma_http_nspares,
 	    'maximum': this.ma_http_nmaxconns,
+	    'ping': this.ma_http_ping_path,
+	    'pingInterval': this.ma_http_ping_path ?
+	        this.ma_http_ping_interval : undefined,
 	    'initialDomains': [ this.ma_manta_host ],
 	    'tcpKeepAliveInitialDelay': this.ma_http_keepalive_delay
 	});
diff --git a/dev/package.json b/dev/package.json
index 3368300..46d11e4 100644
--- a/dev/package.json
+++ b/dev/package.json
@@ -16,7 +16,7 @@
 		"assert-plus": "^1.0.0",
 		"bunyan": "0.20.0",
 		"catstreams": "0.5.0",
-		"cueball": "^1.1.4",
+		"cueball": "2.2.4",
 		"dtrace-provider": "0.6.0",
 		"extsprintf": "^1.3.0",
 		"fuzzstream": "0.0.1",
