commit b5c601a4198702b14d2d7c4b86d4144b65473465
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-08-07T13:27:09+00:00 (9 weeks ago)
    
    8564 developer/build/onbld shouldn't require ghostscript [fix local pkg]

diff --git a/pkgsrc/onbld/packlist b/pkgsrc/onbld/packlist
index eaf4d9b3..3c90695d 100644
--- a/pkgsrc/onbld/packlist
+++ b/pkgsrc/onbld/packlist
@@ -15,7 +15,6 @@
 ./bin/git-pbchk
 ./bin/hdrchk
 ./bin/i386/aw
-./bin/i386/codereview
 ./bin/i386/cscope-fast
 ./bin/i386/ctfconvert
 ./bin/i386/ctfdiff
@@ -106,7 +105,6 @@
 ./man/man1onbld/cddlchk.1onbld
 ./man/man1onbld/check_rtime.1onbld
 ./man/man1onbld/checkpaths.1onbld
-./man/man1onbld/codereview.1onbld
 ./man/man1onbld/cstyle.1onbld
 ./man/man1onbld/cw.1onbld
 ./man/man1onbld/find_elf.1onbld
