From 23bcf3c2a1194a53ab9f1a985b7577a9564cd4a7 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Tue, 29 Aug 2017 00:22:40 +1200
Subject: [PATCH] PUBAPI-1424: cloudapi tests cannot succeed when user has no
 default_network

---
 test/common.js | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/test/common.js b/test/common.js
index 766c2dc..ac03b25 100644
--- a/test/common.js
+++ b/test/common.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -248,7 +248,7 @@ function clientDataTeardown(client, cb) {
                 }
 
                 if (!sub) {
-                    ufds.deleteDcLocalConfig(id, dc, pollConfigDeletion);
+                    pollConfigDeletion();
                 } else {
                     ufds.deleteUser(account, cb);
                 }
@@ -260,12 +260,7 @@ function clientDataTeardown(client, cb) {
     }
 
     var pollConfigCount = 10;
-    function pollConfigDeletion(err) {
-        if (err) {
-            cb(err);
-            return;
-        }
-
+    function pollConfigDeletion() {
         --pollConfigCount;
         if (pollConfigCount === 0) {
             cb(new Error('Config failed to delete in time'));
@@ -527,6 +522,11 @@ function waitForAccountConfigReady(client, cb) {
     assert.object(client, 'client');
     assert.func(cb, 'callback');
 
+    if (!CONFIG.fabrics_enabled) {
+        cb();
+        return;
+    }
+
     var nbTries = 0;
     var MAX_NB_TRIES = 20;
     var TRY_DELAY_IN_MS = 1000;
-- 
2.21.0

