From 098d3ddc06ac24ab2abd3614a65fc26649593709 Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Mon, 31 Jul 2017 16:13:54 -0700
Subject: [PATCH] VOLAPI-64 rename smartdc_role nfsserver to nfsvolumestorage

---
 lib/volumes.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/volumes.js b/lib/volumes.js
index a2fc477..5d65d0b 100644
--- a/lib/volumes.js
+++ b/lib/volumes.js
@@ -13,7 +13,7 @@ var assert = require('assert-plus');
 var NFS_SHARED_VOLUME_EXPORTS_BASEDIR = '/exports';
 var NFS_SHARED_VOLUME_EXPORTS_DIRNAME = 'data';
 var NFS_SHARED_VOLUME_VM_ALIAS_PREFIX = 'nfs-shared-volume';
-var NFS_SHARED_VOLUME_SMARTDC_ROLE = 'nfsserver';
+var NFS_SHARED_VOLUME_SMARTDC_ROLE = 'nfsvolumestorage';
 
 module.exports = {
     NFS_SHARED_VOLUME_EXPORTS_BASEDIR: NFS_SHARED_VOLUME_EXPORTS_BASEDIR,
-- 
2.21.0

