commit 098d3ddc06ac24ab2abd3614a65fc26649593709 (refs/changes/09/2309/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-07-31T16:13:54-07:00 (2 years, 2 months ago)
    
    VOLAPI-64 rename smartdc_role nfsserver to nfsvolumestorage

diff --git a/lib/volumes.js b/lib/volumes.js
index a2fc477..5d65d0b 100644
--- a/lib/volumes.js
+++ b/lib/volumes.js
@@ -13,7 +13,7 @@ var assert = require('assert-plus');
 var NFS_SHARED_VOLUME_EXPORTS_BASEDIR = '/exports';
 var NFS_SHARED_VOLUME_EXPORTS_DIRNAME = 'data';
 var NFS_SHARED_VOLUME_VM_ALIAS_PREFIX = 'nfs-shared-volume';
-var NFS_SHARED_VOLUME_SMARTDC_ROLE = 'nfsserver';
+var NFS_SHARED_VOLUME_SMARTDC_ROLE = 'nfsvolumestorage';
 
 module.exports = {
     NFS_SHARED_VOLUME_EXPORTS_BASEDIR: NFS_SHARED_VOLUME_EXPORTS_BASEDIR,
