commit fcf35996e7cd89e5341d4b06277802f9f796e29c (refs/changes/94/3194/3)
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2018-01-11T22:16:01+00:00 (1 year, 9 months ago)
    
    OS-6557 fmdump misinterprets retval of nvlist_print_json()

diff --git a/usr/src/cmd/fm/fmdump/common/nvlrender.c b/usr/src/cmd/fm/fmdump/common/nvlrender.c
index 2c2f5ca662..99f027a77d 100644
--- a/usr/src/cmd/fm/fmdump/common/nvlrender.c
+++ b/usr/src/cmd/fm/fmdump/common/nvlrender.c
@@ -83,7 +83,7 @@ fmdump_render_nvlist(nvlist_prtctl_t pctl, void *private, nvlist_t *nvl,
 int
 fmdump_print_json(fmd_log_t *lp, const fmd_log_record_t *rp, FILE *fp)
 {
-	if (nvlist_print_json(fp, rp->rec_nvl) != 0 || fprintf(fp, "\n") < 0 ||
+	if (nvlist_print_json(fp, rp->rec_nvl) < 0 || fprintf(fp, "\n") < 0 ||
 	    fflush(fp) != 0)
 		return (-1);
 
diff --git a/usr/src/test/util-tests/tests/libnvpair_json/print_json.c b/usr/src/test/util-tests/tests/libnvpair_json/print_json.c
index e34ae8f7b1..9ac19b1bd1 100644
--- a/usr/src/test/util-tests/tests/libnvpair_json/print_json.c
+++ b/usr/src/test/util-tests/tests/libnvpair_json/print_json.c
@@ -814,7 +814,7 @@ main(int argc, char **argv)
 	/*
 	 * Print the resultant list, and a terminating newline:
 	 */
-	if (nvlist_print_json(stdout, lw->lw_nvl[0]) != 0 ||
+	if (nvlist_print_json(stdout, lw->lw_nvl[0]) < 0 ||
 	    fprintf(stdout, "\n") < 0)
 		goto out;
 
