commit 7409c22605e1c81bbe40e58161909b3e0b6fd2bd (refs/changes/95/2095/1)
Author: Sebastian Wiedenroth <wiedi@frubar.net>
Date:   2017-06-14T23:53:37+02:00 (2 years, 4 months ago)
    
    add timeout option for imgadm create

diff --git a/src/img/lib/cli.js b/src/img/lib/cli.js
index f42b6e25..76de8cb0 100644
--- a/src/img/lib/cli.js
+++ b/src/img/lib/cli.js
@@ -1847,6 +1847,7 @@ CLI.prototype.do_create = function do_create(subcmd, opts, args, cb) {
             compression: opts.compression,
             incremental: opts.incremental,
             prepareScript: opts.s && fs.readFileSync(opts.s, 'utf8'),
+            prepareTimeout: opts.timeout || 300,
             savePrefix: savePrefix,
             logCb: console.log,
             quiet: opts.quiet,
@@ -2017,6 +2018,12 @@ CLI.prototype.do_create.options = [
             + 'prepare-image script must follow. See the "PREPARE '
             + 'IMAGE SCRIPT" section in "man imgadm".'
     },
+    {
+        names: ['timeout'],
+        type: 'positiveInteger',
+        helpArg: '<timeout>',
+        help: 'Timeout in seconds. Default is 300 (5 minutes).'
+    },
     {
         group: ''
     },
diff --git a/src/img/man/imgadm.1m.md b/src/img/man/imgadm.1m.md
index b53ecd00..55a2ac20 100644
--- a/src/img/man/imgadm.1m.md
+++ b/src/img/man/imgadm.1m.md
@@ -374,6 +374,9 @@ UUID.
                            prepare-image script must follow. See the "PREPARE
                            IMAGE SCRIPT" section in "man imgadm".
 
+            --timeout <timeout>
+                           Timeout in seconds. Default is 300 (5 minutes).
+
             -p <url>, --publish <url>
                            Publish directly to the given image source
                            (an IMGAPI server). You may not specify both
