commit dc642c7e4a3fafc3c0a51c72efe3b65b5164ca9a (refs/changes/75/2875/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-10-26T19:36:12+00:00 (1 year, 11 months ago)
    
    OS-6420 panic in prgetattr

diff --git a/usr/src/uts/common/fs/proc/prvnops.c b/usr/src/uts/common/fs/proc/prvnops.c
index 7d6cb037c1..fa5f637f05 100644
--- a/usr/src/uts/common/fs/proc/prvnops.c
+++ b/usr/src/uts/common/fs/proc/prvnops.c
@@ -1636,6 +1636,11 @@ pr_read_spymaster(prnode_t *pnp, uio_t *uiop)
 	if ((error = prlock(pnp, ZNO)) != 0)
 		return (error);
 
+	if (pnp->pr_common->prc_thread == NULL) {
+		prunlock(pnp);
+		return (0);
+	}
+
 	lwp = pnp->pr_common->prc_thread->t_lwp;
 
 	if (lwp->lwp_spymaster == NULL) {
@@ -2666,6 +2671,11 @@ pr_read_spymaster_32(prnode_t *pnp, uio_t *uiop)
 	if ((error = prlock(pnp, ZNO)) != 0)
 		return (error);
 
+	if (pnp->pr_common->prc_thread == NULL) {
+		prunlock(pnp);
+		return (0);
+	}
+
 	lwp = pnp->pr_common->prc_thread->t_lwp;
 
 	if (lwp->lwp_spymaster == NULL) {
@@ -3299,7 +3309,8 @@ prgetattr(vnode_t *vp, vattr_t *vap, int flags, cred_t *cr,
 			vap->va_size = 0;
 		break;
 	case PR_SPYMASTER:
-		if (pnp->pr_common->prc_thread->t_lwp->lwp_spymaster != NULL) {
+		if (pnp->pr_common->prc_thread != NULL &&
+		    pnp->pr_common->prc_thread->t_lwp->lwp_spymaster != NULL) {
 			vap->va_size = PR_OBJSIZE(psinfo32_t, psinfo_t);
 		} else {
 			vap->va_size = 0;
