From 5d8e8ba54fc724b9b8aae2a46dd016f7c537289f Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Thu, 13 Dec 2018 19:01:48 +0000
Subject: [PATCH] TRITON-1033 adminIpFromSysinfo should check for missing NIC

---
 lib/index.js | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/index.js b/lib/index.js
index f9a8686..f0ae562 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -83,7 +83,11 @@ function adminIpFromSysinfo(sysinfo) {
 
     var nic = adminNicFromSysinfo(sysinfo);
 
-    return nic.ip4addr;
+    if (nic) {
+        return nic.ip4addr;
+    }
+
+    return null;
 }
 
 // ---- From VM Metadata
-- 
2.21.0

