commit 5d8e8ba54fc724b9b8aae2a46dd016f7c537289f (refs/changes/36/5236/1)
Author: Rui Loura <rui@joyent.com>
Date:   2018-12-13T19:01:48+00:00 (10 months ago)
    
    TRITON-1033 adminIpFromSysinfo should check for missing NIC

diff --git a/lib/index.js b/lib/index.js
index f9a8686..f0ae562 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -83,7 +83,11 @@ function adminIpFromSysinfo(sysinfo) {
 
     var nic = adminNicFromSysinfo(sysinfo);
 
-    return nic.ip4addr;
+    if (nic) {
+        return nic.ip4addr;
+    }
+
+    return null;
 }
 
 // ---- From VM Metadata
