commit 2af2fcfaa6e93fae40e5b0cc9fd96a025141258e (refs/changes/55/1555/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-02-22T14:28:02-08:00 (2 years, 7 months ago)
    
    DOCKER-893 Improve docker pull reliability

diff --git a/lib/backends/sdc/images.js b/lib/backends/sdc/images.js
index 31ab93b..3f0d6f4 100644
--- a/lib/backends/sdc/images.js
+++ b/lib/backends/sdc/images.js
@@ -1392,8 +1392,10 @@ function pullImage(opts, callback) {
      *
      * 1. In some cases the `err.code` is set to a meaningful error code.
      *    This comes via the `JOB.chain_results[-1].error.name`, set by
-     *    pull-image.js error handling.
-     *
+     *    pull-image.js error handling. Note that some older versions of IMGAPI
+     *    can send back docker-registry-client errors (as well as IMGAPI errors)
+     *    that have a different err.code string (drc code will use an extra
+     *    `Error` suffix on the err.code).
      * 2. If there is an error in the Docker Registry v2 API, then `err.message`
      *    equals a JSON string per:
      *        https://docs.docker.com/registry/spec/api/#errors
@@ -1419,7 +1421,14 @@ function pullImage(opts, callback) {
         var recognized = true;
 
         var errmsg;
-        if (err.code === 'ENOTFOUND') {
+        if (
+            /*
+             * `RemoteSourceError` code comes from newer IMGAPI.
+             * `ENOTFOUND` code comes from docker-registry-client - through
+             * old IMGAPI versions.
+             */
+            err.code === 'RemoteSourceError' || err.code === 'ENOTFOUND')
+        {
             /* BEGIN JSSTYLED */
             /*
              * Docker-docker:
@@ -1450,13 +1459,28 @@ function pullImage(opts, callback) {
              * --- NotFoundError
              * E.g.: `docker pull quay.io/no-such-user`
              */
-            err.code === 'UnauthorizedError' || err.code === 'NotFoundError')
+            err.code === 'UnauthorizedError' || err.code === 'NotFoundError'
+                || err.code === 'UNAUTHORIZED')
         {
             errmsg = format('Error: image %s not found', imgName);
-        } else if (err.code === 'DownloadError') {
-            // E.g. A `DownloadError` from docker-registry-client.
+        } else if (
+            /*
+             * `Download` code comes from newer IMGAPI.
+             * `DownloadError` code comes from docker-registry-client - through
+             * old IMGAPI versions.
+             */
+            err.code === 'Download' || err.code === 'DownloadError')
+        {
             errmsg = format('Error downloading %s: %s', imgName, err.message);
-        } else if (err.code === 'ConnectTimeoutError') {
+        } else if (
+            /*
+             * `RemoteSourceError` code comes from newer IMGAPI.
+             * `ConnectTimeoutError` code comes from docker-registry-client -
+             * through old IMGAPI versions.
+             */
+            err.code === 'ConnectTimeoutError'
+            || err.code === 'RemoteSourceError')
+        {
             errmsg = format('Timeout connecting to host %s',
                 opts.rat.index.name);
         } else if (err.code === 'NotImplemented') {
