commit 65799bd7bb92c203ea32d60501fa093c9962012d (refs/changes/28/3428/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-02-20T22:30:14+00:00 (1 year, 8 months ago)
    
    OS-6651 root fs needs a bit more space for bhyve

diff --git a/tools/build_live b/tools/build_live
index 77564714..569cf012 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -788,7 +788,7 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_root_size=272000
+bi_root_size=276000
 if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
     grep "DEBUG enabled" >/dev/null; then
 	#
