From 04face87f33644b04664ae92b8929066ec91d85f Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 20 Jul 2017 17:15:19 +0200
Subject: [PATCH] OS-6250 nvmeadm: nvme_dskname() string handling is buggy

---
 usr/src/cmd/nvmeadm/nvmeadm.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/usr/src/cmd/nvmeadm/nvmeadm.c b/usr/src/cmd/nvmeadm/nvmeadm.c
index 13cace3ead..64760c2b43 100644
--- a/usr/src/cmd/nvmeadm/nvmeadm.c
+++ b/usr/src/cmd/nvmeadm/nvmeadm.c
@@ -443,20 +443,21 @@ nvme_dskname(const nvme_process_arg_t *npa)
 		path = di_dim_path_dev(dim, di_driver_name(child),
 		    di_instance(child), "c");
 
-		if (path != NULL) {
+		if (path != NULL && strlen(path) > 2) {
 			path[strlen(path) - 2] = '\0';
-			path = strrchr(path, '/') + 1;
-			if (path != NULL) {
-				path = strdup(path);
-				if (path == NULL)
-					err(-1, "nvme_dskname");
-			}
+			path = strrchr(path, '/');
+			if (path != NULL)
+				path = strdup(path + 1);
 		}
 
+		if (path == NULL)
+			err(-1, "nvme_dskname");
+
 		break;
 	}
 
 	di_dim_fini(dim);
+
 	return (path);
 }
 
-- 
2.21.0

