From 3023546ee421b24ab6078f96fd8d0d4c50542e17 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 20 Jul 2017 17:15:19 +0200
Subject: [PATCH] OS-6250 nvmeadm: nvme_dskname() string handling is buggy

---
 usr/src/cmd/nvmeadm/nvmeadm.c | 26 +++++++++++++++++---------
 1 file changed, 17 insertions(+), 9 deletions(-)

diff --git a/usr/src/cmd/nvmeadm/nvmeadm.c b/usr/src/cmd/nvmeadm/nvmeadm.c
index 6fcaff6b89..a08fa59258 100644
--- a/usr/src/cmd/nvmeadm/nvmeadm.c
+++ b/usr/src/cmd/nvmeadm/nvmeadm.c
@@ -431,21 +431,29 @@ nvme_dskname(const nvme_process_arg_t *npa)
 		path = di_dim_path_dev(dim, di_driver_name(child),
 		    di_instance(child), "c");
 
-		if (path != NULL) {
-			path[strlen(path) - 2] = '\0';
-			path = strrchr(path, '/') + 1;
-			if (path != NULL) {
-				path = strdup(path);
-				if (path == NULL)
-					err(-1, "nvme_dskname");
-			}
-		}
+		/*
+		 * Error out if we didn't get a path, or if it's too short for
+		 * the following operations to be safe.
+		 */
+		if (path == NULL || strlen(path) < 2)
+			goto fail;
+
+		/* Chop off 's0' and get everything past the last '/' */
+		path[strlen(path) - 2] = '\0';
+		path = strrchr(path, '/');
+		if (path == NULL)
+			goto fail;
+		path++;
 
 		break;
 	}
 
 	di_dim_fini(dim);
+
 	return (path);
+
+fail:
+	err(-1, "nvme_dskname");
 }
 
 static int
-- 
2.21.0

