From cc77145a564a256cf0d3575f96eb5d42deeaf00f Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike@mikezeller.net>
Date: Wed, 17 Jul 2019 19:31:03 +0000
Subject: [PATCH] TRITON-1820 cfwlogd tries to change into non existent
 directory

---
 cfwlogd/src/main.rs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/cfwlogd/src/main.rs b/cfwlogd/src/main.rs
index 8793511..2fc731f 100644
--- a/cfwlogd/src/main.rs
+++ b/cfwlogd/src/main.rs
@@ -145,7 +145,6 @@ fn cfwlogd_daemonize() {
         .stdout(stdout)
         .stderr(stderr)
         .group("daemon")
-        .working_directory(LOG_DIR)
         .umask(0o022)
         .start()
     {
@@ -212,6 +211,7 @@ fn cfwlogd_handle_signals(s: c_int, loggers: &Loggers) -> bool {
 fn cfwlogd_chroot<P: Into<PathBuf>>(p: P) -> io::Result<()> {
     let path = p.into();
     std::fs::create_dir_all(&path)?;
+    std::env::set_current_dir(&path)?;
     let cstring = CString::new(path.into_os_string().into_vec())
         .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "path contained nuls"))?;
     unsafe {
-- 
2.21.0

