From 19cb39f4475529cf0d3df768e0f4cc8b5af32d9a Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 14 Mar 2019 15:46:39 -0700
Subject: [PATCH] TRITON-884 remove cloud analytics (ca) from Triton

---
 etc/logsets.json | 1 -
 1 file changed, 1 deletion(-)

diff --git a/etc/logsets.json b/etc/logsets.json
index 3086781..887abb1 100644
--- a/etc/logsets.json
+++ b/etc/logsets.json
@@ -227,7 +227,6 @@
       "amon",
       "amonredis",
       "binder",
-      "ca",
       "cloudapi",
       "cmon",
       "cnapi",
-- 
2.21.0

