commit dab18c10165ea95eba4c15ce881ef6ea6bd962b7 (refs/changes/27/27/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-07-05T18:46:06-07:00 (3 years, 3 months ago)
    
    joyent/node-zkstream#2 printfs are left in from debugging

diff --git a/lib/zk-streams.js b/lib/zk-streams.js
index 6999ade..e721fed 100644
--- a/lib/zk-streams.js
+++ b/lib/zk-streams.js
@@ -84,7 +84,6 @@ ZKDecodeStream.prototype._transform = function (chunk, enc, cb) {
 				    e.name + ': ' + e.message));
 				return;
 			}
-			console.log(" < %j", pkt);
 			self.push(pkt);
 		}
 	}
@@ -111,8 +110,6 @@ ZKEncodeStream.prototype._transform = function (pkt, enc, cb) {
 	mod_assert.object(pkt);
 	var zkb = new ZKBuffer({});
 
-	console.log(" > %j", pkt);
-
 	if (this.ze_fsm.getState() === 'connecting') {
 		zkb.writeLengthPrefixed(function (sub) {
 			sub.writeConnectRequest(pkt);
diff --git a/package.json b/package.json
index 8bd1de2..e3a9cc4 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "zkstream",
-  "version": "0.1.0",
+  "version": "0.1.1",
   "description": "",
   "main": "lib/index.js",
   "dependencies": {
