From a9f3d38a54b1bdb4d666d92d0aebbbbd1f6397d8 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 5 Jul 2016 18:43:53 -0700
Subject: [PATCH] joyent/node-zkstream#2 printfs are left in from debugging

---
 lib/zk-streams.js | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/lib/zk-streams.js b/lib/zk-streams.js
index 6999ade..e721fed 100644
--- a/lib/zk-streams.js
+++ b/lib/zk-streams.js
@@ -84,7 +84,6 @@ ZKDecodeStream.prototype._transform = function (chunk, enc, cb) {
 				    e.name + ': ' + e.message));
 				return;
 			}
-			console.log(" < %j", pkt);
 			self.push(pkt);
 		}
 	}
@@ -111,8 +110,6 @@ ZKEncodeStream.prototype._transform = function (pkt, enc, cb) {
 	mod_assert.object(pkt);
 	var zkb = new ZKBuffer({});
 
-	console.log(" > %j", pkt);
-
 	if (this.ze_fsm.getState() === 'connecting') {
 		zkb.writeLengthPrefixed(function (sub) {
 			sub.writeConnectRequest(pkt);
-- 
2.21.0

