commit 1981ba1a05406d01db5aacf236d15b6373165e6d (refs/changes/65/4165/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-06-08T13:00:56+02:00 (1 year, 4 months ago)
    
    TRITON-480 Use @smaller/tap in papi

diff --git a/Makefile b/Makefile
index f6563d5..8e7c872 100644
--- a/Makefile
+++ b/Makefile
@@ -25,7 +25,7 @@ NAME		:= papi
 #
 # Tools
 #
-TAP		:= ./node_modules/.bin/tape
+TAP		:= ./node_modules/.bin/tap
 
 #
 # Files
diff --git a/package.json b/package.json
index 250b4dc..918be9e 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-papi",
   "description": "Packages API",
-  "version": "7.1.0",
+  "version": "7.1.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "repository": {
@@ -21,7 +21,7 @@
     "jsprim": "^1.3.1"
   },
   "devDependencies": {
-    "tape": "3.5.0"
+    "@smaller/tap": "11.1.4-1.0.0"
   },
   "scripts": {
     "start": "node ./server.js",
diff --git a/test/api.test.js b/test/api.test.js
index 0264569..5d1dfa4 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -23,7 +23,7 @@ var jsprim = require('jsprim');
 var libuuid = require('libuuid');
 var Logger  = require('bunyan');
 var restify = require('restify');
-var test    = require('tape').test;
+var test    = require('@smaller/tap').test;
 var VError = require('verror');
 
 var papi = require('../lib/papi');
diff --git a/test/runtests b/test/runtests
index 7218f45..918a7b3 100755
--- a/test/runtests
+++ b/test/runtests
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -72,7 +72,7 @@ start_time=$(date +%s)
 
 NODE_INSTALL=$TOP/build/node
 PATH=$TOP/build/node/bin:$TOP/node_modules/.bin:$PATH
-TAPE=./node_modules/.bin/tape
+TAPE=./node_modules/.bin/tap
 
 # Options.
 opt_test_pattern=
@@ -96,8 +96,8 @@ done
 # Setup a clean output dir.
 OUTPUT_DIR=/var/tmp/papitest
 echo "# Setup a clean output dir ($OUTPUT_DIR)."
-rm -rf /var/tmp/papitest
-mkdir -p /var/tmp/papitest
+rm -rf $OUTPUT_DIR
+mkdir -p $OUTPUT_DIR
 
 # Gather DC setup info for the test files.
 echo "# Datacenter config:"
@@ -112,7 +112,6 @@ if [[ -n "$opt_test_pattern" ]]; then
     echo "# Running filtered set of test files: $test_files"
 fi
 
-failed_tests=0
 failed_test_names=
 set +o errexit
 set -o pipefail
@@ -120,7 +119,6 @@ for t in ${test_files}; do
     PATH=${NODE_INSTALL}/bin:${PATH} SDC_SETUP_TESTS=1 ${NODE_INSTALL}/bin/node \
         --abort_on_uncaught_exception ${t} | tee -a ${OUTPUT_DIR}/papi.tap
     if [[ $? != 0 ]]; then
-        failed_tests=$((${failed_tests} + 1))
         failed_test_names="${failed_test_names} ${t}"
     fi
 done
@@ -130,18 +128,15 @@ set -o errexit
 end_time=$(date +%s)
 elapsed=$((${end_time} - ${start_time}))
 
-tests=$(grep "^# tests [0-9]" ${OUTPUT_DIR}/papi.tap | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
-passed=$(grep "^# pass  [0-9]" ${OUTPUT_DIR}/papi.tap | tr -s ' ' | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
-skipped=$(grep "^# skip  [0-9]" ${OUTPUT_DIR}/papi.tap | tr -s ' ' | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
+tests=$(grep "^1\.\.[0-9]" ${OUTPUT_DIR}/*.tap | cut -d '.' -f3 | xargs | tr ' ' '+' | bc)
+failed=$(grep "^# failed [0-9]" ${OUTPUT_DIR}/*.tap | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
+skipped=$(grep "^# skip  [0-9]" ${OUTPUT_DIR}/*.tap | cut -d ' ' -f3 | xargs | tr ' ' '+' | bc)
 
 [[ -z ${tests} ]] && tests=0
-[[ -z ${passed} ]] && passed=0
+[[ -z ${failed} ]] && failed=0
 [[ -z ${skipped} ]] && skipped=0
 
-fail=$((${tests} - ${passed} - ${skipped}))
-if [[ ${failed_tests} != 0 ]]; then
-    fail=$((${fail} + ${failed_tests}))
-fi
+passed=$((${tests} - ${failed} - ${skipped}))
 
 if [[ -t 1 ]]; then
     # We're on a terminal, so use color
@@ -165,8 +160,8 @@ echo -e "#   ${COLOR_GREEN}PASS: ${passed} / ${tests}${COLOR_NORMAL}"
 if [[ ${skipped} -gt 0 ]]; then
     echo -e "#   ${COLOR_ORANGE}SKIP: ${skipped} / ${tests}${COLOR_NORMAL}"
 fi
-if [[ ${fail} -gt 0 ]]; then
-    echo -e "#   ${COLOR_RED}FAIL: ${fail} / ${tests}${COLOR_NORMAL}"
+if [[ ${failed} -gt 0 ]]; then
+    echo -e "#   ${COLOR_RED}FAIL: ${failed} / ${tests}${COLOR_NORMAL}"
     echo "#"
     echo "# FAILED TESTS:"
     echo "#"
@@ -176,7 +171,7 @@ if [[ ${fail} -gt 0 ]]; then
 fi
 echo "#"
 
-if [[ ${fail} -gt 0 ]]; then
+if [[ ${failed} -gt 0 ]]; then
     exit 1
 fi
 
