From a0c8b2f4f86257bdb3fbed26aaa2a72c6a16dfaf Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 27 Nov 2018 16:32:03 -0800
Subject: [PATCH] MANTA-4036 Unauthenticated requests broken by MANTA-3983

---
 lib/auth.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/auth.js b/lib/auth.js
index 0718bce..37205e5 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -782,7 +782,7 @@ function getActiveRoles(req, res, next) {
      * status, if present).
      */
     var account = req.caller.account;
-    if (account.groups.indexOf('role-operators') !== -1) {
+    if (account.groups && account.groups.indexOf('role-operators') !== -1) {
         /*
          * The req.caller object is cached and potentially shared between
          * multiple requests. We're going to alter the isOperator flag on
-- 
2.21.0

