commit a0c8b2f4f86257bdb3fbed26aaa2a72c6a16dfaf (refs/changes/46/5146/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-11-27T16:32:03-08:00 (11 months ago)
    
    MANTA-4036 Unauthenticated requests broken by MANTA-3983

diff --git a/lib/auth.js b/lib/auth.js
index 0718bce..37205e5 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -782,7 +782,7 @@ function getActiveRoles(req, res, next) {
      * status, if present).
      */
     var account = req.caller.account;
-    if (account.groups.indexOf('role-operators') !== -1) {
+    if (account.groups && account.groups.indexOf('role-operators') !== -1) {
         /*
          * The req.caller object is cached and potentially shared between
          * multiple requests. We're going to alter the isOperator flag on
