commit 394e2811ccf2d7c9ac31e772c45928a3ffa09178 (refs/changes/32/3432/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-02-21T16:05:58+00:00 (1 year, 8 months ago)
    
    OS-6654 allow brands to override CPU bursting setup

diff --git a/overlay/generic/usr/lib/brand/jcommon/statechange b/overlay/generic/usr/lib/brand/jcommon/statechange
index 5efd3e26..2678b0fa 100644
--- a/overlay/generic/usr/lib/brand/jcommon/statechange
+++ b/overlay/generic/usr/lib/brand/jcommon/statechange
@@ -632,6 +632,9 @@ setup_snapshots()
 #
 setup_cpu_baseline()
 {
+	# A brand can override the setup of bursting.
+	[ -n "$NO_BURSTING" ] && return
+
 	# If there is already a baseline, don't set one heuristically
 	curr_base=`prctl -P -n zone.cpu-baseline -i zone $ZONENAME | nawk '{
 		if ($2 == "privileged") print $3
