commit 9b3b0750a33262d43a07a8902a190524ed15633a (refs/changes/11/1511/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-02-15T08:42:59-08:00 (2 years, 8 months ago)
    
    DOCKER-1004 test-docker "test ps filtering on package" failures

diff --git a/test/integration/cli-packagelabels.test.js b/test/integration/cli-packagelabels.test.js
index a47c5d2..7253735 100644
--- a/test/integration/cli-packagelabels.test.js
+++ b/test/integration/cli-packagelabels.test.js
@@ -198,7 +198,8 @@ test('test ps filtering on package', opts, function (tt) {
     vasync.forEachPipeline({
         inputs: Object.keys(expectedResults),
         func: function _performFilteredPs(pkg, cb) {
-            var argstring = '--filter "label=com.joyent.package=' + pkg + '"';
+            var argstring = '--filter "label=com.joyent.package=' + pkg + '"'
+                + ' --all';
 
             cli.ps(tt, {
                 args: argstring
@@ -331,7 +332,8 @@ test('test lookup w/ invalid package names', opts, function (tt) {
         inputs: labels,
         func: function _lookupContainer(label, cb) {
             var argstring = '--format "{{.ID}}:\t{{.Labels}}" '
-                + '--filter "label=com.joyent.package=' + label + '"';
+                + '--filter "label=com.joyent.package=' + label + '" '
+                + ' --all';
 
             cli.ps(tt, {
                 args: argstring,
@@ -387,7 +389,7 @@ test('test creation w/ two package labels', opts, function (tt) {
 
     cli.create(tt, {args: cmdline}, function (err, id) {
         var argstring = '--filter "label=com.joyent.package=' + packageB.name
-            + '" --format '
+            + '" --all --format '
             + '\"{{.ID}},{{.Label \\\"com.joyent.package\\\"}}\"';
         var shortId;
 
