From 99b1281dbd59c3e91250f367a3afc0adc5d2781d Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 4 Jul 2019 12:26:09 +0100
Subject: [PATCH] TOOLS-2287 bits-upload gets confused when building a gerrit
 branch

---
 tools/mk/Makefile.targ | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index a6f42d7..755107f 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -535,7 +535,7 @@ ifndef ENGBLD_BITS_UPLOAD_OVERRIDE
 .PHONY: bits-upload
 bits-upload: publish $(BUILDIMAGE_TARG)
 	$(TOP)/deps/eng/tools/bits-upload.sh \
-		-b $(BRANCH) \
+		-b "$(BRANCH)" \
 		-B "$(TRY_BRANCH)" \
 		$(BITS_UPLOAD_LOCAL_ARG) \
 		$(BITS_UPLOAD_IMGAPI_ARG) \
@@ -552,7 +552,7 @@ bits-upload: publish $(BUILDIMAGE_TARG)
 .PHONY: bits-upload-latest
 bits-upload-latest:
 	$(TOP)/deps/eng/tools/bits-upload.sh \
-		-b $(BRANCH) \
+		-b "$(BRANCH)" \
 		-B "$(TRY_BRANCH)" \
 		$(BITS_UPLOAD_LOCAL_ARG) \
 		$(BITS_UPLOAD_IMGAPI_ARG) \
-- 
2.21.0

