commit 6446325aa31816015cd4c59bfcd227692edb7b17 (refs/changes/40/4740/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-08-24T20:25:30+02:00 (1 year, 1 month ago)
    
    TRITON-591 refreservation not set properly on bhyve disks

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index ac43db55..d22dbf7a 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -2074,7 +2074,7 @@ function createVolume(volume, opts, callback)
 
     if (volume.hasOwnProperty('refreservation')) {
         refreserv = volume.refreservation;
-    } else if (volume.hasOwnProperty('image_uuid')) {
+    } else if (volume.hasOwnProperty('image_uuid') && brand !== 'bhyve') {
         // We can't set refreserv > volsize when making a clone, so set it
         // to volsize instead.
         refreserv = size;
@@ -2146,6 +2146,8 @@ function createVolume(volume, opts, callback)
                 }
                 if (refreserv !== 'auto') {
                     args.push('-o', 'refreservation=' + refreserv + 'M');
+                } else if (brand === 'bhyve' && refreserv === 'auto') {
+                    args.push('-o', 'refreservation=auto');
                 }
                 args.push(snapshot, target);
                 zfs(args, log, function (e) {
