commit a7a295a4be076ac6e87485be762cd56dd7adbd2d (refs/changes/40/4740/4)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-08-30T14:59:22+00:00 (1 year, 1 month ago)
    
    TRITON-591 refreservation not set properly on bhyve disks

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 5c6d76f3..27a98a32 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -2106,7 +2106,6 @@ function createVolume(volume, opts, callback)
     assert.string(opts.brand, 'opts.brand');
     assert.object(opts.log, 'opts.log');
 
-    var brand = opts.brand;
     var log = opts.log;
     var refreserv;
     var size;
@@ -2133,23 +2132,7 @@ function createVolume(volume, opts, callback)
 
     if (volume.hasOwnProperty('refreservation')) {
         refreserv = volume.refreservation;
-    } else if (volume.hasOwnProperty('image_uuid')) {
-        // We can't set refreserv > volsize when making a clone, so set it
-        // to volsize instead.
-        refreserv = size;
-        log.debug({
-            size: size
-        }, 'refreservation unset, setting refreserv=size for cloned disk');
-    } else if (brand === 'kvm') {
-        // For KVM we keep the pre-OS-6753 for now to not break compatibility
-        // but could change this to match bhyve (default case below) in the
-        // future.
-        refreserv = size;
-        log.debug({
-            size: size
-        }, 'refreservation unset for kvm, setting refreserv=size for new disk');
     } else {
-        // We default to letting SmartOS set refreservation for bhyve: OS-6753
         log.debug('refreservation unset, letting system determine');
         refreserv = 'auto';
     }
@@ -2205,6 +2188,8 @@ function createVolume(volume, opts, callback)
                 }
                 if (refreserv !== 'auto') {
                     args.push('-o', 'refreservation=' + refreserv + 'M');
+                } else if (refreserv === 'auto') {
+                    args.push('-o', 'refreservation=auto');
                 }
                 args.push(snapshot, target);
                 zfs(args, log, function (e) {
