From 4dadc8f75196f3aa78ed3d5c767570b7121ada19 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 30 Nov 2016 14:46:33 -0800
Subject: [PATCH] TOOLS-1624 sdcadm crashes on platform install image not in
 channel

---
 lib/platform.js | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/platform.js b/lib/platform.js
index 74342ee..f9eba36 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -429,6 +429,7 @@ Platform.prototype.install = function install(opts, callback) {
     var self = this;
     var localdir = '/var/tmp';
     var deleteOnFinish = true;
+    var defaultChannel;
     var filepath;
     var image;
     var progress = self.progress;
@@ -588,6 +589,7 @@ Platform.prototype.install = function install(opts, callback) {
                 if (err) {
                     return next();
                 }
+                defaultChannel = channel;
                 if (!fs.existsSync(opts.image)) {
                     progress('Using channel %s', channel);
                 }
@@ -727,7 +729,7 @@ Platform.prototype.install = function install(opts, callback) {
             } else if (imgNotFoundError) {
                 progress('Unable to find the given Platform Image \'%s\' ' +
                         'in channel \'%s\'', opts.image,
-                        self.sdcadm.updates.channel);
+                        self.sdcadm.updates.channel || defaultChannel);
             } else {
                 progress('In order not to have to re-download image, ' +
                          '%s has been left behind.', filepath);
-- 
2.21.0

