From 3f9171f9734d4c837601c5925d93c03c016f09ea Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 6 Feb 2019 16:50:08 +0000
Subject: [PATCH] TOOLS-2174 agent-cache 'clone' should use '--' when switching
 branches

---
 tools/agent-prebuilt.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/agent-prebuilt.sh b/tools/agent-prebuilt.sh
index c36b736..f1e0119 100755
--- a/tools/agent-prebuilt.sh
+++ b/tools/agent-prebuilt.sh
@@ -90,7 +90,7 @@ function do_clone {
             echo "the branch the existing repository is checked out to doesn't"
             echo "exist upstream."
         fi
-        git checkout $agent_branch
+        git checkout $agent_branch --
         if [[ $? -ne 0 ]]; then
             echo "Checking out $agent_branch failed, falling back to $branch"
             git checkout $branch
@@ -99,7 +99,7 @@ function do_clone {
             # at this point, failures are really fatal.
             set -o errexit
             echo "Checking out $branch also failed, falling back to master"
-            git checkout master
+            git checkout master --
             git_exit=$?
         fi
     else
-- 
2.21.0

