From 08601a26aa40d4dbca2778205885fd7df9eeafbc Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Mon, 6 Feb 2017 22:04:27 +0000
Subject: [PATCH] TOOLS-1641 hermes-proxy cannot deal with lots of Manta load
 balancers

---
 deps/hermes | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/hermes b/deps/hermes
index 417ef53..3702adf 160000
--- a/deps/hermes
+++ b/deps/hermes
@@ -1 +1 @@
-Subproject commit 417ef53987bc1ab264f470d6fb1c5d59761e9e89
+Subproject commit 3702adf367e3ea2a6b0c27889a0e9e0b4ac93ebc
-- 
2.21.0

