From aee784218f26c8b04ba636bf8cc49289db69be16 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?K=C3=A9vin=20Berthommier?= <kevin.bertho@gmail.com>
Date: Tue, 3 Jul 2018 22:58:30 +0200
Subject: [PATCH] add missing throws

---
 lib/verror.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/verror.js b/lib/verror.js
index e5d0bf3..1d6e16c 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -53,8 +53,8 @@ function parseConstructorArguments(args)
 	var argv, options, sprintf_args, shortmessage, k;
 
 	if (!isObject(args)) throw new mod_AssertionError('args (object) is required');
-	if (!isBoolean(args.strict)) new mod_AssertionError('args.strict (bool) is required');
-	if (!Array.isArray(args.argv)) new mod_AssertionError('args.argv (array) is required');
+	if (!isBoolean(args.strict)) throw new mod_AssertionError('args.strict (bool) is required');
+	if (!Array.isArray(args.argv)) throw new mod_AssertionError('args.argv (array) is required');
 	argv = args.argv;
 
 	/*
-- 
2.17.2 (Apple Git-113)

