commit aee784218f26c8b04ba636bf8cc49289db69be16 (refs/changes/76/4476/1)
Author: Kévin Berthommier <kevin.bertho@gmail.com>
Date:   2018-07-03T22:58:30+02:00 (1 year, 3 months ago)
    
    add missing throws

diff --git a/lib/verror.js b/lib/verror.js
index e5d0bf3..1d6e16c 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -53,8 +53,8 @@ function parseConstructorArguments(args)
 	var argv, options, sprintf_args, shortmessage, k;
 
 	if (!isObject(args)) throw new mod_AssertionError('args (object) is required');
-	if (!isBoolean(args.strict)) new mod_AssertionError('args.strict (bool) is required');
-	if (!Array.isArray(args.argv)) new mod_AssertionError('args.argv (array) is required');
+	if (!isBoolean(args.strict)) throw new mod_AssertionError('args.strict (bool) is required');
+	if (!Array.isArray(args.argv)) throw new mod_AssertionError('args.argv (array) is required');
 	argv = args.argv;
 
 	/*
