From 8405755e573eed75f4c14b5a79c0371be80b0b34 Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody.kantor@gmail.com>
Date: Tue, 6 Feb 2018 19:08:52 +0000
Subject: [PATCH] MANTA-3575 increase config-agent poll interval to 120 seconds

---
 services.sh     | 4 ++--
 update/repos.sh | 1 +
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/services.sh b/services.sh
index a04dd6f..fb9662f 100644
--- a/services.sh
+++ b/services.sh
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -34,7 +34,7 @@ function manta_setup_config_agent {
     cat >${prefix}/etc/config.json <<EOF
 {
     "logLevel": "info",
-    "pollInterval": 10000,
+    "pollInterval": 120000,
     "sapi": {
         "url": "${SAPI_URL}"
     },
diff --git a/update/repos.sh b/update/repos.sh
index b8763e5..8281754 100755
--- a/update/repos.sh
+++ b/update/repos.sh
@@ -34,6 +34,7 @@ REPOS=(
     manta-mola
     manta-muskie
     manta-propeller
+    manta-reshard
     manta-wrasse
     moray
     muppet
-- 
2.21.0

