commit f96dfcfbd2cdf28be71f19e311ce4b4e53f09d97 (refs/changes/06/3306/2)
Author: Kody A Kantor <kody.kantor@gmail.com>
Date:   2018-02-07T14:59:20+00:00 (1 year, 8 months ago)
    
    MANTA-3575 revisit config-agent pollInterval

diff --git a/services.sh b/services.sh
index a04dd6f..fb9662f 100644
--- a/services.sh
+++ b/services.sh
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -34,7 +34,7 @@ function manta_setup_config_agent {
     cat >${prefix}/etc/config.json <<EOF
 {
     "logLevel": "info",
-    "pollInterval": 10000,
+    "pollInterval": 120000,
     "sapi": {
         "url": "${SAPI_URL}"
     },
diff --git a/update/repos.sh b/update/repos.sh
index b8763e5..8281754 100755
--- a/update/repos.sh
+++ b/update/repos.sh
@@ -34,6 +34,7 @@ REPOS=(
     manta-mola
     manta-muskie
     manta-propeller
+    manta-reshard
     manta-wrasse
     moray
     muppet
