commit 481b015e1963cce82c53d4b0a0f5f1ffab2fdf53 (refs/changes/06/3306/1)
Author: Kody A Kantor <kody.kantor@gmail.com>
Date:   2018-02-06T19:08:52+00:00 (1 year, 8 months ago)
    
    MANTA-3575 revisit config-agent pollInterval

diff --git a/services.sh b/services.sh
index a04dd6f..fb9662f 100644
--- a/services.sh
+++ b/services.sh
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -34,7 +34,7 @@ function manta_setup_config_agent {
     cat >${prefix}/etc/config.json <<EOF
 {
     "logLevel": "info",
-    "pollInterval": 10000,
+    "pollInterval": 120000,
     "sapi": {
         "url": "${SAPI_URL}"
     },
