commit f960e8322e039f56d52ffdbe1c353357acc851a0 (refs/changes/62/3562/2)
Author: Jeff Emershaw <me@jeffemershaw.com>
Date:   2018-06-20T16:40:45+02:00 (1 year, 4 months ago)
    
    PUBAPI-1378 Provide UpdateFabricNetwork endpoint

diff --git a/CHANGES.md b/CHANGES.md
index 1cd45e2..f952948 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,10 @@
 
 ## not yet released
 
+## 11.1.0
+
+- PUBAPI-1378 Provide UpdateFabricNetwork endpoint
+
 ## 11.0.0
 
 - [Backward incompatible] Make `options.version` a required param to the SAPI
diff --git a/deps/jsstyle b/deps/jsstyle
index d75b7ca..ccb145b 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit d75b7ca8308be17c80e2b120f2a01d4a0c20d8a8
+Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
diff --git a/lib/napi.js b/lib/napi.js
index 6fca6a5..40eca91 100644
--- a/lib/napi.js
+++ b/lib/napi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 /*
@@ -1107,6 +1107,45 @@ NAPI.prototype.createFabricNetwork =
     return this.post(opts, params, callback);
 };
 
+/**
+ * Update a Network on a VLAN, scoped to a user's fabric.
+ *
+ * @param {String} owner: The Owner UUID for this network.
+ * @param {String} id : The VLAN ID containing this network.
+ * @param {String} network : The network UUID of this network.
+ * @param {Object} params : Required Network parameters.
+ * @param {Object} options : Optional request options.
+ * @param {Function} callback : of the form f(err, res).
+ */
+NAPI.prototype.updateFabricNetwork =
+        function (owner, id, network, params, options, callback) {
+    var opts;
+
+    assert.string(owner, 'owner');
+    assert.number(id, 'id');
+    assert.string(network, 'network');
+    assert.object(params, 'params');
+
+    if (typeof (options) === 'function') {
+        callback = options;
+        options = {};
+    }
+
+    assert.object(options, 'options');
+    assert.func(callback, 'callback');
+
+    opts = {
+         path: format('/fabrics/%s/vlans/%s/networks/%s', owner, id, network),
+         query: params
+    };
+
+    if (options && options.headers) {
+        opts.headers = options.headers;
+    }
+
+    return this.put(opts, params, callback);
+};
+
 /**
  * Get a Network on a VLAN, scoped to a user's fabric.
  *
diff --git a/package.json b/package.json
index fd16e8f..98ccb10 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "11.0.0",
+  "version": "11.1.0",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
