commit 0e827ea89d90fd67e8d0dcbb05bcf05374e35f50 (refs/changes/24/3924/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-05-09T14:55:58-05:00 (1 year, 5 months ago)
    
    OS-6943 varpd not listed as an install_h target

diff --git a/usr/src/lib/Makefile b/usr/src/lib/Makefile
index 7d337e7175..9efd87bc10 100644
--- a/usr/src/lib/Makefile
+++ b/usr/src/lib/Makefile
@@ -506,6 +506,7 @@ HDRSUBDIRS=				\
 	smbsrv		\
 	smhba		\
 	udapl		\
+	varpd		\
 	$($(MACH)_HDRSUBDIRS)
 
 i386_HDRSUBDIRS=	\
