commit 7392f9c688d6373406b29c2244c33a4233c0edc0 (refs/changes/38/2838/2)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2017-10-25T01:24:05+00:00 (1 year, 11 months ago)
    
    ADMINUI-2389 Typo on Timed Out Message

diff --git a/www/js/router.js b/www/js/router.js
index ffd68c48..05c5fae5 100644
--- a/www/js/router.js
+++ b/www/js/router.js
@@ -152,7 +152,7 @@ module.exports = Backbone.Marionette.AppRouter.extend({
                     reject();
                     self.showSignin();
                     if (t === "timeout") {
-                        window.alert("One more the services required for Authentication Timed out.");
+                        window.alert("Some of the services required for authentication timed out.");
                     }
                 }).done(function () {
                     resolve();
