From 7392f9c688d6373406b29c2244c33a4233c0edc0 Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Sun, 22 Oct 2017 21:55:42 -0700
Subject: [PATCH] ADMINUI-2389 Typo on Timed Out Message

---
 www/js/router.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/www/js/router.js b/www/js/router.js
index ffd68c48..05c5fae5 100644
--- a/www/js/router.js
+++ b/www/js/router.js
@@ -152,7 +152,7 @@ module.exports = Backbone.Marionette.AppRouter.extend({
                     reject();
                     self.showSignin();
                     if (t === "timeout") {
-                        window.alert("One more the services required for Authentication Timed out.");
+                        window.alert("Some of the services required for authentication timed out.");
                     }
                 }).done(function () {
                     resolve();
-- 
2.21.0

