commit 2388903a699a3231c2dc5e554aa873aae720d292 (refs/changes/38/2838/1)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2017-10-22T21:55:42-07:00 (2 years ago)
    
    ADMINUI-2389 Typo on Timed Out Message

diff --git a/www/js/router.js b/www/js/router.js
index ffd68c48..05c5fae5 100644
--- a/www/js/router.js
+++ b/www/js/router.js
@@ -152,7 +152,7 @@ module.exports = Backbone.Marionette.AppRouter.extend({
                     reject();
                     self.showSignin();
                     if (t === "timeout") {
-                        window.alert("One more the services required for Authentication Timed out.");
+                        window.alert("Some of the services required for authentication timed out.");
                     }
                 }).done(function () {
                     resolve();
