From e95af30f30575d5475c77171549988cb24ba4286 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 11 Oct 2017 01:03:30 +0000
Subject: [PATCH] MORAY-443 Clean shutdown broken by MORAY-421

---
 lib/server.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index ff11000..c765bb2 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -246,8 +246,8 @@ MorayServer.prototype.close = function close() {
     barrier.on('drain', function () {
         self.fast_server.close();
 
-        if (self.kang_server !== null) {
-            self.kang_server.close();
+        if (self.monitor_server !== null) {
+            self.monitor_server.close();
         }
 
         self.emit('close');
-- 
2.21.0

