commit 22fb73b5fb83d7853203b0562d8a1bc8a260c83f (refs/changes/40/2940/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-11-10T18:28:55-08:00 (1 year, 11 months ago)
    
    VOLAPI-91 want experimental NFS volumes tests in nightly-1

diff --git a/Makefile b/Makefile
index 7205cea..f525f5f 100644
--- a/Makefile
+++ b/Makefile
@@ -73,6 +73,7 @@ release: all deps docs $(SMF_MANIFESTS)
     $(ROOT)/package.json \
     $(ROOT)/sapi_manifests \
     $(ROOT)/smf \
+    $(ROOT)/test \
     $(ROOT)/tools \
     $(RELSTAGEDIR)/root/opt/smartdc/$(SERVICE_NAME)/
 	mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/boot
