From 22fb73b5fb83d7853203b0562d8a1bc8a260c83f Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Fri, 10 Nov 2017 18:28:55 -0800
Subject: [PATCH] VOLAPI-91 want experimental NFS volumes tests in nightly-1

---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 7205cea..f525f5f 100644
--- a/Makefile
+++ b/Makefile
@@ -73,6 +73,7 @@ release: all deps docs $(SMF_MANIFESTS)
     $(ROOT)/package.json \
     $(ROOT)/sapi_manifests \
     $(ROOT)/smf \
+    $(ROOT)/test \
     $(ROOT)/tools \
     $(RELSTAGEDIR)/root/opt/smartdc/$(SERVICE_NAME)/
 	mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/boot
-- 
2.21.0

