commit 02a4427b43718fa9c5ea844cc617dd1abb51b411 (refs/changes/22/4422/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-06-24T12:47:04-07:00 (1 year, 3 months ago)
    
    TRITON-542 cn-agent should close connection after posting sysinfo

diff --git a/lib/app.js b/lib/app.js
index 059c019..45d12e0 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -358,6 +358,7 @@ App.prototype.registerServer = function registerServer(callback) {
     var urlPath = path.join('/servers', self.uuid, 'sysinfo');
 
     restifyOptions = {
+        agent: false,
         connectTimeout: 5000,
         requestTimeout: 5000,
         url: url,
diff --git a/package.json b/package.json
index 4508124..c91faf1 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "2.2.1",
+  "version": "2.2.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
