commit 118bb08e7bfbcf35da14b4e18b4b964ac8e64332 (refs/changes/07/3907/2)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-05-08T15:30:09+00:00 (1 year, 5 months ago)
    
    OS-6928 recursive mutex in pci_xhci

diff --git a/usr/src/cmd/bhyve/pci_xhci.c b/usr/src/cmd/bhyve/pci_xhci.c
index f178468108..06d46b981f 100644
--- a/usr/src/cmd/bhyve/pci_xhci.c
+++ b/usr/src/cmd/bhyve/pci_xhci.c
@@ -1661,7 +1661,9 @@ pci_xhci_try_usb_xfer(struct pci_xhci_softc *sc,
 	do_intr = 0;
 
 	xfer = devep->ep_xfer;
+#ifdef __FreeBSD__
 	USB_DATA_XFER_LOCK(xfer);
+#endif
 
 	/* outstanding requests queued up */
 	if (dev->dev_ue->ue_data != NULL) {
@@ -1684,8 +1686,9 @@ pci_xhci_try_usb_xfer(struct pci_xhci_softc *sc,
 		}
 	}
 
+#ifdef __FreeBSD__
 	USB_DATA_XFER_UNLOCK(xfer);
-
+#endif
 
 	return (err);
 }
@@ -1917,7 +1920,13 @@ pci_xhci_device_doorbell(struct pci_xhci_softc *sc, uint32_t slot,
 
 	/* handle pending transfers */
 	if (devep->ep_xfer->ndata > 0) {
+#ifndef __FreeBSD__
+		USB_DATA_XFER_LOCK(devep->ep_xfer);
+#endif
 		pci_xhci_try_usb_xfer(sc, dev, devep, ep_ctx, slot, epid);
+#ifndef __FreeBSD__
+		USB_DATA_XFER_UNLOCK(devep->ep_xfer);
+#endif
 		return;
 	}
 
