commit 5bc6922be3074d62fc1d3b2a7a549193a389f389 (refs/changes/07/3907/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-05-03T19:33:18+00:00 (1 year, 5 months ago)
    
    OS-6928 recursive mutex in pci_xhci

diff --git a/usr/src/cmd/bhyve/pci_xhci.c b/usr/src/cmd/bhyve/pci_xhci.c
index f178468108..8edeff7ae6 100644
--- a/usr/src/cmd/bhyve/pci_xhci.c
+++ b/usr/src/cmd/bhyve/pci_xhci.c
@@ -1661,7 +1661,6 @@ pci_xhci_try_usb_xfer(struct pci_xhci_softc *sc,
 	do_intr = 0;
 
 	xfer = devep->ep_xfer;
-	USB_DATA_XFER_LOCK(xfer);
 
 	/* outstanding requests queued up */
 	if (dev->dev_ue->ue_data != NULL) {
@@ -1684,9 +1683,6 @@ pci_xhci_try_usb_xfer(struct pci_xhci_softc *sc,
 		}
 	}
 
-	USB_DATA_XFER_UNLOCK(xfer);
-
-
 	return (err);
 }
 
@@ -1917,7 +1913,9 @@ pci_xhci_device_doorbell(struct pci_xhci_softc *sc, uint32_t slot,
 
 	/* handle pending transfers */
 	if (devep->ep_xfer->ndata > 0) {
+		USB_DATA_XFER_LOCK(devep->ep_xfer);
 		pci_xhci_try_usb_xfer(sc, dev, devep, ep_ctx, slot, epid);
+		USB_DATA_XFER_UNLOCK(devep->ep_xfer);
 		return;
 	}
 
