From f977c714c5442503e7994edf143ea3efb51b8fd9 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 1 Nov 2016 12:29:40 -0700
Subject: [PATCH] TOOLS-1599 sdcadm image "version" should include the
 buildstamp

---
 tools/mk-shar | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tools/mk-shar b/tools/mk-shar
index 356eda1..fb4e2e8 100755
--- a/tools/mk-shar
+++ b/tools/mk-shar
@@ -177,7 +177,7 @@ cat <<EOM >$OUTMANIFEST
     "uuid": "$(uuid)",
     "owner": "00000000-0000-0000-0000-000000000000",
     "name": "sdcadm",
-    "version": "$VERSION",
+    "version": "$VERSION-$STAMP",
     "description": "Administer SmartDataCenter standups",
     "disabled": false,
     "public": false,
@@ -190,4 +190,3 @@ cat <<EOM >$OUTMANIFEST
 }
 EOM
 echo created $OUTMANIFEST
-
-- 
2.21.0

