From e2b6e4e83960a56269920b06d9fde7a4fc5c0a0f Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 23 Nov 2018 10:37:06 -0800
Subject: [PATCH] MANTA-4033 get lagging Manta components on the latest v0.10.x

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 5016887..1e87737 100644
--- a/Makefile
+++ b/Makefile
@@ -35,7 +35,7 @@ JSSTYLE_FILES	 = $(JS_FILES)
 # Variables
 #
 NAME			= marlin-dashboard
-NODE_PREBUILT_VERSION	= v0.10.25
+NODE_PREBUILT_VERSION	= v0.10.48
 NODE_PREBUILT_TAG	= zone
 NODE_PREBUILT_IMAGE	= fd2cc906-8938-11e3-beab-4359c665ac99
 
-- 
2.21.0

