commit 745d7272c276a7cd3f4fd3367b31bccb1d41ec29 (refs/changes/09/5609/1)
Author: John Levon <john.levon@joyent.com>
Date:   2019-02-19T23:00:59+00:00 (8 months ago)
    
    OS-7594 Disable efinet support in Loader to workaround TRITON-1191

diff --git a/usr/src/boot/sys/boot/efi/loader/Makefile.com b/usr/src/boot/sys/boot/efi/loader/Makefile.com
index be81482ae1..011ec84d52 100644
--- a/usr/src/boot/sys/boot/efi/loader/Makefile.com
+++ b/usr/src/boot/sys/boot/efi/loader/Makefile.com
@@ -12,6 +12,8 @@
 #
 # Copyright 2016 Toomas Soome <tsoome@me.com>
 #
+# Copyright (c) 2019, Joyent, Inc.
+#
 
 include $(SRC)/Makefile.master
 include $(SRC)/boot/Makefile.version
@@ -87,6 +89,13 @@ CPPFLAGS +=	-I$(SRC)/uts/intel/sys/acpi
 CPPFLAGS +=	-I$(PNGLITE)
 CPPFLAGS +=	-DNO_PCI -DEFI
 
+#
+# Using SNP from loader causes issues when chain-loading iPXE, as described in
+# TRITON-1191.  While the exact problem is not known, we have no use for SNP, so
+# we'll just disable it.
+#
+CPPFLAGS +=	-DLOADER_DISABLE_SNP
+
 # Export serial numbers, UUID, and asset tag from loader.
 smbios.o := CPPFLAGS += -DSMBIOS_SERIAL_NUMBERS
 # Use little-endian UUID format as defined in SMBIOS 2.6.
diff --git a/usr/src/boot/sys/boot/efi/loader/conf.c b/usr/src/boot/sys/boot/efi/loader/conf.c
index 5a7bfede26..854230a204 100644
--- a/usr/src/boot/sys/boot/efi/loader/conf.c
+++ b/usr/src/boot/sys/boot/efi/loader/conf.c
@@ -24,6 +24,10 @@
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+/*
+ * Copyright (c) 2019, Joyent, Inc.
+ */
+
 #include <sys/cdefs.h>
 
 #include <stand.h>
@@ -36,7 +40,9 @@ struct devsw *devsw[] = {
 	&efipart_fddev,
 	&efipart_cddev,
 	&efipart_hddev,
+#ifndef LOADER_DISABLE_SNP
 	&efinet_dev,
+#endif
 	&zfs_dev,
 	NULL
 };
@@ -57,7 +63,9 @@ struct fs_ops *file_system[] = {
 };
 
 struct netif_driver *netif_drivers[] = {
+#ifndef LOADER_DISABLE_SNP
 	&efinetif,
+#endif
 	NULL
 };
 
