commit 7c25c542bab5644c300274f2d953d760b3afa2ec (refs/changes/51/751/2)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2016-11-17T13:38:33-08:00 (2 years, 11 months ago)
    
    DOCKER-947 update to node v4

diff --git a/targets.json.in b/targets.json.in
index 1c74359..232bd40 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -415,7 +415,7 @@ cat <<EOF
     "image_name": "docker",
     "image_description": "SDC Docker Engine",
     "image_version": "1.0.0",
-    "image_uuid": "de411e86-548d-11e4-a4b7-3bb60478632a",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
     "repos": [
       {"url": "git@github.com:joyent/sdc-docker.git"}
     ],
@@ -426,7 +426,13 @@ cat <<EOF
       "registrar",
       "https://download.joyent.com/pub/build/sdcnode"
     ],
-    "pkgsrc": [],
+    "pkgsrc": [
+        "coreutils-8.23nb2",
+        "curl-7.47.1",
+        "gsed-4.2.2nb4",
+        "patch-2.7.5",
+        "sudo-1.8.15"
+    ],
     "tarballs": [
       {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
