From 5d51b051212d62a2588f28563646fdcecc531796 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Thu, 18 May 2017 09:09:18 +0000
Subject: [PATCH] MANTA-3273 "manta-adm zk fixup" broken with VMs in remote DCs

---
 lib/adm.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/adm.js b/lib/adm.js
index 984a7e0..f3a470b 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -2428,7 +2428,7 @@ maAdm.prototype.auditZkServers = function ()
 
 		serverUuid = zkinstance.params['server_uuid'];
 		if (!self.ma_vms.hasOwnProperty(uuid)) {
-			if (self.ma_cns[serverUuid] !== null) {
+			if (self.ma_cns.hasOwnProperty(serverUuid)) {
 				validationErrors.push(new VError(
 				    'nameservice instance "%s": VM appears ' +
 				    'to have been provisioned in this ' +
-- 
2.21.0

