From 7d3a92941994addecd73c17b2ae3d6766324612b Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Thu, 18 May 2017 09:59:39 +0000
Subject: [PATCH] MANTA-3273 "manta-adm zk fixup" broken with VMs in remote DCs

---
 lib/adm.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/adm.js b/lib/adm.js
index 984a7e0..578d2f0 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -2428,7 +2428,8 @@ maAdm.prototype.auditZkServers = function ()
 
 		serverUuid = zkinstance.params['server_uuid'];
 		if (!self.ma_vms.hasOwnProperty(uuid)) {
-			if (self.ma_cns[serverUuid] !== null) {
+			if (self.ma_cns.hasOwnProperty(serverUuid) &&
+			    self.ma_cns[serverUuid] !== null) {
 				validationErrors.push(new VError(
 				    'nameservice instance "%s": VM appears ' +
 				    'to have been provisioned in this ' +
-- 
2.21.0

