commit febff4312acd3737133f17dd983ae5ae857c4267 (refs/changes/65/1165/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-12-21T15:03:25-05:00 (2 years, 10 months ago)
    
    SDC_TESTING falseyness

diff --git a/CHANGES.md b/CHANGES.md
index a0410af..a3b6b12 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -7,6 +7,12 @@ Known issues:
 
 ## not yet released
 
+- PUBAPI-1205/PUBAPI-1171/PUBAPI-1351 The handling of legacy `SDC_*`
+  environmental variables has been cleaned up.  These environment
+  variables are used for compatibility with the node-smartdc toolset.
+   * `SDC_TESTING` is now evaluated the same way as noe-smartdc.  Any
+     set value but the empty string is true.
+
 - [joyent/node-triton#137] Improve the handling for the getting started case
   when a user may not have envvars or a profile setup.
 
diff --git a/lib/config.js b/lib/config.js
index 10f22c2..df790b7 100644
--- a/lib/config.js
+++ b/lib/config.js
@@ -296,7 +296,7 @@ function _loadEnvProfile() {
         envProfile.insecure = common.boolFromString(
             process.env.SDC_TLS_INSECURE);
     } else if (process.env.SDC_TESTING) { // deprecated
-        envProfile.insecure = common.boolFromString(process.env.SDC_TESTING);
+        envProfile.insecure = true;
     }
 
     validateProfile(envProfile, 'environment variables');
