From c78e5f3230dbd29e886b86fe5c9b3436af51028c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 24 Oct 2019 13:50:21 +0100
Subject: [PATCH] OS-8033 smartos-only-release target should include tests-tar

---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 54655f47..66f08cfa 100644
--- a/Makefile
+++ b/Makefile
@@ -681,6 +681,7 @@ triton-smartos-release: \
 
 .PHONY: smartos-only-release
 smartos-only-release: \
+    tests-tar \
     common-platform-publish \
     smartos-build \
     smartos-publish \
-- 
2.17.2 (Apple Git-113)

