From bdb999a3041c0399fb4dac82167f30eead1ed44b Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Fri, 21 Oct 2016 22:09:26 +0000
Subject: [PATCH] OS-5737 psecflags missing non-isaexec binaries

---
 manifest | 1 +
 1 file changed, 1 insertion(+)

diff --git a/manifest b/manifest
index a732ea6e15..25c15ede2f 100644
--- a/manifest
+++ b/manifest
@@ -1939,6 +1939,7 @@ f usr/bin/amd64/preap 0555 root bin
 f usr/bin/amd64/prstat 0555 root bin
 f usr/bin/amd64/prun 0555 root bin
 f usr/bin/amd64/ps 0555 root bin
+f usr/bin/amd64/psecflags 0555 root bin
 f usr/bin/amd64/psig 0555 root bin
 f usr/bin/amd64/pstack 0555 root bin
 f usr/bin/amd64/pstop 0555 root bin
-- 
2.21.0

