From 89dc917c18fc3e9739376e9eca229dedd229003e Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 11 May 2017 14:51:06 -0700
Subject: [PATCH] MANTA-3257 mackerel build needs to specify
 NODE_PREBUILT_IMAGE

---
 Makefile | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/Makefile b/Makefile
index 6b5939a..00f283a 100644
--- a/Makefile
+++ b/Makefile
@@ -45,6 +45,11 @@ JSSTYLE_FLAGS    = -f tools/jsstyle.conf
 
 NODE_PREBUILT_VERSION=v0.10.40
 NODE_PREBUILT_TAG=zone
+# We want the sdc-smartos@1.6.3 builds of sdcnode. This should match the
+# origin image used for "mola", in which mackerel is run. See the compatibility
+# matrix here:
+# 	https://github.com/joyent/triton-origin-image/blob/master/README.md#sdcnode-compatibility-with-triton-origin-images
+NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
 
 
 include ./tools/mk/Makefile.defs
-- 
2.21.0

