commit 13152c9cd51a3813b3eace93dad4b98855e26abb (refs/changes/09/4709/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-08-16T17:20:08-07:00 (1 year, 2 months ago)
    
    TRITON-702 cnapi test-model-server.js failure after TRITON-557

diff --git a/test/lib/mock.js b/test/lib/mock.js
index adf168a..48c9edc 100644
--- a/test/lib/mock.js
+++ b/test/lib/mock.js
@@ -15,7 +15,11 @@ var App = require('../../lib/app');
 
 var configFilename = path.join(__dirname, '..', '..', 'config', 'test.json');
 var mockedMetricsManager = {
-    collectRestifyMetrics: function _collectRestifyMetrics() {}
+    collectRestifyMetrics: function _collectRestifyMetrics() {},
+    collector: {
+        gauge: function _createGauge(opts) { return {}; }
+    },
+    addPreCollectFunc: function _addPreCollectFunc() {}
 };
 
 var MockLogger = {
diff --git a/test/model/test-model-server.js b/test/model/test-model-server.js
index 7d47a6d..3ef9d7d 100644
--- a/test/model/test-model-server.js
+++ b/test/model/test-model-server.js
@@ -67,12 +67,16 @@ function testListServersAll(test) {
                 {
                     uuid: '372bdb58-f8dd-11e1-8038-0b6dbddc5e58',
                     ram: '12345',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 },
                 {
                     uuid: '6e8eb888-f8e0-11e1-b1a8-5f74056f9365',
                     ram: '56789',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 }
             ];
 
@@ -127,12 +131,16 @@ function testListServersByUuids(test) {
                 {
                     uuid: '372bdb58-f8dd-11e1-8038-0b6dbddc5e58',
                     ram: '12345',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 },
                 {
                     uuid: 'b31695ce-f8e6-11e1-b252-fb742866284b',
                     ram: '56789',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 }
             ];
 
@@ -204,12 +212,16 @@ function testListServersSetup(test) {
                 {
                     uuid: '372bdb58-f8dd-11e1-8038-0b6dbddc5e58',
                     ram: '12345',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 },
                 {
                     uuid: '6e8eb888-f8e0-11e1-b1a8-5f74056f9365',
                     ram: '56789',
-                    sysinfo: { setup: true }
+                    sysinfo: { setup: true },
+                    last_heartbeat: null,
+                    status: 'unknown'
                 }
             ];
 
