From e87ca9d21bf962cbf737bf55679af3371c81317b Mon Sep 17 00:00:00 2001
From: Dan McDonald <danmcd@joyent.com>
Date: Mon, 22 Jan 2018 10:00:37 -0500
Subject: [PATCH] NAPI-440 IP address object needs belongs_to_type of router

---
 lib/models/nic/common.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/models/nic/common.js b/lib/models/nic/common.js
index 71f4c7e..3b17731 100644
--- a/lib/models/nic/common.js
+++ b/lib/models/nic/common.js
@@ -30,7 +30,7 @@ var validate = require('../../util/validate');
 // --- Globals
 
 var BUCKET = require('./bucket').BUCKET;
-var BELONGS_TO_TYPES = [ 'other', 'server', 'zone' ];
+var BELONGS_TO_TYPES = [ 'other', 'server', 'zone', 'router' ];
 var VALID_NIC_STATES = [ 'provisioning', 'stopped', 'running' ];
 
 var getPoolIntersections = util_intersect.getPoolIntersections;
-- 
2.21.0

