commit 1fe32763ced2c04096684bde7f69d0d1c90fe235 (refs/changes/99/5099/3)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-11-28T01:00:26+00:00 (11 months ago)
    
    TRITON-779 store affinity rules when provisioning

diff --git a/lib/metadata.js b/lib/metadata.js
index 7ca7744..3b2de23 100644
--- a/lib/metadata.js
+++ b/lib/metadata.js
@@ -170,10 +170,10 @@ function list(req, res, next) {
                 }
 
                 if (Object.keys(md2).length) {
+                    delete md2.locality;
                     Object.keys(md2).forEach(function (k) {
                         if (/_pw$/.test(k)) {
-                            md2[k.replace(/_pw$/, '')] =
-                                md2[k];
+                            md2[k.replace(/_pw$/, '')] = md2[k];
                             delete md2[k];
                         }
                     });
@@ -220,10 +220,10 @@ function get(req, res, next) {
             } catch (e) {}
 
             if (Object.keys(md).length) {
+                delete md.locality;
                 Object.keys(md).forEach(function (k) {
                     if (/_pw$/.test(k)) {
-                        md[k.replace(/_pw$/, '')] =
-                            md[k];
+                        md[k.replace(/_pw$/, '')] = md[k];
                         delete md[k];
                     }
                 });
