From ba5187096fa7f7364e76701674dcb93b07565384 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Thu, 22 Nov 2018 17:00:44 +1300
Subject: [PATCH] TRITON-779 store affinity rules when provisioning

---
 lib/metadata.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/metadata.js b/lib/metadata.js
index 7ca7744..3b2de23 100644
--- a/lib/metadata.js
+++ b/lib/metadata.js
@@ -170,10 +170,10 @@ function list(req, res, next) {
                 }
 
                 if (Object.keys(md2).length) {
+                    delete md2.locality;
                     Object.keys(md2).forEach(function (k) {
                         if (/_pw$/.test(k)) {
-                            md2[k.replace(/_pw$/, '')] =
-                                md2[k];
+                            md2[k.replace(/_pw$/, '')] = md2[k];
                             delete md2[k];
                         }
                     });
@@ -220,10 +220,10 @@ function get(req, res, next) {
             } catch (e) {}
 
             if (Object.keys(md).length) {
+                delete md.locality;
                 Object.keys(md).forEach(function (k) {
                     if (/_pw$/.test(k)) {
-                        md[k.replace(/_pw$/, '')] =
-                            md[k];
+                        md[k.replace(/_pw$/, '')] = md[k];
                         delete md[k];
                     }
                 });
-- 
2.21.0

