From c771e0e6494cb85511c73ab0ca26eb2bc5acd805 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 12 Feb 2019 11:01:35 -0500
Subject: [PATCH] TRITON-1220 node-tap versions between 12.0.1 and 12.5.2 can
 emit invalid TAP

---
 package.json | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/package.json b/package.json
index 4d84ff9..e8ee531 100644
--- a/package.json
+++ b/package.json
@@ -39,7 +39,7 @@
     "eslint": "4.13.1",
     "eslint-plugin-joyent": "~2.0.0",
     "mock-fs": "4.5.x",
-    "tap": "^12.0.1",
+    "tap": "^12.5.2",
     "zkstream": "0.11.2"
   },
   "engines": {
-- 
2.21.0

