commit 0706e3604a9818c5d67308c4ed965c9698b82027 (refs/changes/80/880/1)
Author: Yang Yong <yy1111.yang@samsung.com>
Date:   2016-11-11T12:42:45+09:00 (2 years, 11 months ago)
    
    joyent/node-triton#129 triton reboot --wait INST' doesn't wait

diff --git a/lib/do_instance/gen_do_ACTION.js b/lib/do_instance/gen_do_ACTION.js
index 1e86fc7..ef8d932 100644
--- a/lib/do_instance/gen_do_ACTION.js
+++ b/lib/do_instance/gen_do_ACTION.js
@@ -90,19 +90,19 @@ function _doTheAction(action, subcmd, opts, args, callback) {
         case 'start':
             command = opts.snapshot ? 'startMachineFromSnapshot' :
                                       'startMachine';
-            state = 'running';
+            state = ['running'];
             break;
         case 'stop':
             command = 'stopMachine';
-            state = 'stopped';
+            state = ['stopped'];
             break;
         case 'reboot':
             command = 'rebootMachine';
-            state = 'running';
+            state = ['stopped', 'stopping'];
             break;
         case 'delete':
             command = 'deleteMachine';
-            state = 'deleted';
+            state = ['deleted'];
             break;
         default:
             callback(new Error('unknown action: ' + action));
@@ -165,7 +165,7 @@ function _doTheAction(action, subcmd, opts, args, callback) {
 
                     self.top.tritonapi.cloudapi.waitForMachineStates({
                         id: uuid,
-                        states: [state]
+                        states: state
                     }, function (err2, inst2, res2) {
                         if (action === 'delete' &&
                             res2 && res2.statusCode === 410) {
@@ -176,19 +176,32 @@ function _doTheAction(action, subcmd, opts, args, callback) {
                             cb(err2);
                             return;
                         }
-
-                        var dur = common.humanDurationFromMs(Date.now() - now);
-                        if (alias)
-                            console.log('%s instance %s (%s, %s)',
-                                common.capitalize(action), alias, uuid, dur);
-                        else
-                            console.log('%s instance %s (%s)',
-                                common.capitalize(action), uuid, dur);
-
-                        cb();
+                        // check again machine state for reboot finish
+                        if (action === 'reboot') {
+                            self.top.tritonapi.cloudapi.waitForMachineStates({
+                                id: uuid,
+                                states: ['running']
+                            }, function () {
+                                waitFinishConsole();
+                            });
+                        } else {
+                            waitFinishConsole();
+                        }
                     });
                 });
             }
+
+            function waitFinishConsole() {
+                var dur = common.humanDurationFromMs(Date.now() - now);
+                if (alias)
+                    console.log('%s instance %s (%s, %s)',
+                        common.capitalize(action), alias, uuid, dur);
+                else
+                     console.log('%s instance %s (%s)',
+                         common.capitalize(action), uuid, dur);
+
+                cb();
+            }
         },
         inputs: args
     }, function (err, results) {
