From 647ab53dfa0b7128befd1789c6e33db7202018fa Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 2 Aug 2016 22:04:50 +0000
Subject: [PATCH] OS-5543 Deliver the ACPI debugging tools [fix mancheck]

---
 usr/src/man/man1m/acpixtract.1m | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/usr/src/man/man1m/acpixtract.1m b/usr/src/man/man1m/acpixtract.1m
index 1bd62f7645..c4fc222273 100644
--- a/usr/src/man/man1m/acpixtract.1m
+++ b/usr/src/man/man1m/acpixtract.1m
@@ -13,11 +13,11 @@
 .Dd Aug 2, 2016
 .Dt ACPIXTRACT 1M
 .Os
-.Sh NAME 
+.Sh NAME
 .Nm acpixtract
 .Nd extract binary ACPI tables from a dump file
 .Sh SYNOPSIS
-.Nm 
+.Nm
 .Op Fl ahlmv
 .Op Fl s Ar signature
 .Ar file
@@ -56,5 +56,6 @@ tables.
 Get the table with the specified signature.
 .It Fl v
 Print the version.
+.El
 .Sh SEE ALSO
 .Xr acpidump 1M
-- 
2.21.0

