commit 1dd055e94463396a500c4b72344811cd949e7b3d (refs/changes/84/5684/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-03-04T19:24:26+01:00 (7 months ago)
    
    TRITON-1272 sdcadm test failures due to outdated tape version

diff --git a/test/instances.test.js b/test/instances.test.js
index 8e3981f..2006695 100644
--- a/test/instances.test.js
+++ b/test/instances.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2018 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 
@@ -57,8 +57,18 @@ function checkInstancesDetails(t, instancesDetails) {
             version: item[3],
             alias: item[4]
         });
+    }).filter(function (inst) {
+        // Versions for these components are not the using the traditional
+        // branch-releasedate-gitsha approach, which will make the img version
+        // vs. inst version check fail
+        return [
+            'mockcloud',
+            'grafana',
+            'prometheus']
+        .indexOf(inst.service) === -1;
     });
 
+
     common.checkInsts(t, {
         inputs: instancesDetails,
         serviceNamesFromUUID: serviceNamesFromUUID,
diff --git a/test/update-gz-tools.test.js b/test/update-gz-tools.test.js
index 17072f1..336c1bf 100644
--- a/test/update-gz-tools.test.js
+++ b/test/update-gz-tools.test.js
@@ -108,7 +108,6 @@ test('update-gz-tools --latest --just-download', function (t) {
     exec(cmd, function (err, stdout, stderr) {
         t.ifError(err, 'Update gz-tools error');
         var findStrings = [
-            'Downloading gz-tools',
             'Using channel',
             'Updated gz-tools successfully'
         ];
@@ -118,6 +117,12 @@ test('update-gz-tools --latest --just-download', function (t) {
                 util.format('check just-download string present %s', str));
         });
 
+        if (stdout.indexOf('Downloading gz-tools') === -1) {
+            t.notEqual(stdout.indexOf('from previous download'), -1,
+                util.format('check just-download string present %s',
+                    'from previous download'));
+        }
+
         var findNotStrings = [
             'Decompressing gz-tools tarball',
             'Validating gz-tools tarball files',
