From 76eb674794f33f194b858b0fdb96bd13e2475a6f Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Mon, 18 Jun 2018 15:20:57 +0000
Subject: [PATCH] OS-7032 mismatch between sysinfo (json) and sysinfo -p

---
 src/sysinfo | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/sysinfo b/src/sysinfo
index c28215f2..2a889c77 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -937,7 +937,7 @@ END
         ipv4_var="Virtual_Network_Interface_${iface}_IPv4_Address"
         link_status_var="Virtual_Network_Interface_${iface}_Link_Status"
         vlan_var="Virtual_Network_Interface_${iface}_VLAN"
-        parent_var="Virtual_Network_Interface_${iface}_Host_Interface"
+        host_var="Virtual_Network_Interface_${iface}_Host_Interface"
 
         eval "mac=\${${mac_var}}"
         eval "ipv4=\${${ipv4_var}}"
-- 
2.21.0

