From 69f3b24f172467426c0c97e58eed080eba6e6c19 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Thu, 15 Nov 2018 14:18:15 -0500
Subject: [PATCH] TRITON-968 hagfish-watcher dependencies have bitrotted into
 oblivion

---
 Makefile     | 10 ----------
 package.json |  1 -
 2 files changed, 11 deletions(-)

diff --git a/Makefile b/Makefile
index 14fbba6..3c3a554 100644
--- a/Makefile
+++ b/Makefile
@@ -54,7 +54,6 @@ NAME :=			hagfish-watcher
 RELEASE_TARBALL :=	$(NAME)-$(STAMP).tgz
 RELEASE_MANIFEST :=	$(NAME)-$(STAMP).manifest
 RELSTAGEDIR :=		/tmp/$(STAMP)
-NODEUNIT =		$(TOP)/node_modules/.bin/nodeunit
 
 #
 # Repo-specific targets
@@ -63,15 +62,6 @@ NODEUNIT =		$(TOP)/node_modules/.bin/nodeunit
 all: $(SMF_MANIFESTS) | $(NODE_EXEC) $(REPO_DEPS)
 	$(NPM_EXEC) install
 
-$(NODEUNIT): | $(NODE_EXEC)
-	$(NPM_EXEC) install
-
-CLEAN_FILES += $(NODEUNIT) ./node_modules/tap
-
-.PHONY: test
-test: $(NODEUNIT)
-	$(NODEUNIT) --reporter=tap test/test-*.js
-
 $(TOP)/downloads/$(NODE_TARBALL):
 	@echo "downloading node $(NODE_VERSION) ..."
 	mkdir -p $(TOP)/downloads
diff --git a/package.json b/package.json
index b38ff1c..b8ecc89 100644
--- a/package.json
+++ b/package.json
@@ -10,7 +10,6 @@
         "mkdirp": "0.3.5",
         "kstat": "git://github.com/orlandov/node-kstat.git#8c2643153c38f045cfd9dbbde24b644b492974ae",
         "zutil": "git://github.com/orlandov/node-zutil.git#9257881d575dd4496f5283930e6b07ed59685809",
-        "nodeunit": "0.7.4",
         "extsprintf": "1.0.3",
         "vasync": "1.5.0",
         "jsprim": "0.5.1",
-- 
2.21.0

