commit c1bb48a7f77de0ebe657cec60bb62e859b96b1c5
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-18T15:37:28+00:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index c74f8e2..855677b 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -3,7 +3,7 @@
 	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
 	url = https://github.com/joyent/eng.git
diff --git a/deps/jsstyle b/deps/jsstyle
index b94eb62..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit b94eb629ee524d21b563c8b4461caa2bdf1becc6
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
