commit 8b09a444fffcde282603db95c61c25b460597fe5 (refs/changes/34/2434/3)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-24T23:18:11+00:00 (2 years, 1 month ago)
    
    MANTA-3368 MPU commit does not return the computed MD5

diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index ded74d8..856e1a7 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -516,6 +516,7 @@ function commit(req, res, next) {
             }, 'commit: completed');
 
             res.setHeader('Location', p);
+            res.setHeader('Computed-MD5', md5);
             res.send(201);
             next();
         }
diff --git a/package.json b/package.json
index 22ed74d..efabd1a 100644
--- a/package.json
+++ b/package.json
@@ -33,14 +33,14 @@
         "moray": "3.1.1",
         "once": "1.3.0",
         "restify": "2.6.3",
-        "vasync": "1.4.3",
+        "vasync": "^1.5.0",
         "verror": "^1.9.0",
         "watershed": "0.3.0",
         "xtend": "2.1.1"
     },
     "devDependencies": {
         "smartdc": "7.3.1",
-        "manta": "4.3.0",
+        "manta": "file:../node-manta",
         "nodeunit": "0.9.1",
         "node-uuid": "1.4.1"
     },
diff --git a/test/mpu/commit.test.js b/test/mpu/commit.test.js
index 51f454d..6ef3e76 100644
--- a/test/mpu/commit.test.js
+++ b/test/mpu/commit.test.js
@@ -112,6 +112,61 @@ test('commit upload: one part', function (t) {
     });
 });
 
+test('commit upload: commit content md5 match', function (t) {
+    var self = this;
+    vasync.waterfall([
+        function (callback) {
+            self.createUpload(self.path, null, function (err) {
+                if (ifErr(t, err, 'commit md5')) {
+                    callback(err);
+                } else {
+                    callback(null);
+                }
+            });
+        },
+        function (callback) {
+            self.writeTestObject(self.uploadId, 0, function (err, res) {
+                if (ifErr(t, err, 'upload part')) {
+                    callback(err);
+                } else {
+                    t.ok(res);
+                    t.checkResponse(res, 204);
+                    callback(null, res.headers.etag);
+                }
+            });
+        },
+        function (etag, callback) {
+            self.commitUpload(self.uploadId, [etag], function (err, res) {
+                if (ifErr(t, err, 'commit upload')) {
+                    callback(err);
+                } else {
+                    t.ok(res);
+                    t.ok(res.headers);
+                    t.ok(res.headers['computed-md5']);
+                    callback(null, res.headers['computed-md5']);
+                }
+            });
+        },
+        function (computedMd5, callback) {
+            self.client.info(self.path, function (err, info) {
+                if (ifErr(t, err, 'get object info')) {
+                    callback(err);
+                } else {
+                    t.ok(info);
+                    if (info) {
+                        var headers = info.headers || {};
+                        t.equal(computedMd5, helper.TEXT_MD5);
+                        t.equal(computedMd5, headers['content-md5']);
+                    }
+                    callback(null);
+                }
+            });
+        }
+    ], function () {
+        t.done();
+    });
+});
+
 
 test('commit upload: already commited, same set of parts', function (t) {
     var self = this;
diff --git a/test/mpu/helper.js b/test/mpu/helper.js
index 7a71f53..33d3b68 100644
--- a/test/mpu/helper.js
+++ b/test/mpu/helper.js
@@ -337,12 +337,12 @@ function commitUploadHelper(id, etags, subuser, cb) {
         account: self.client.user
     };
 
-    client.commitUpload(id, etags, opts, function (err) {
+    client.commitUpload(id, etags, opts, function (err, res) {
         if (err) {
-            cb(err);
+            cb(err, null);
         } else {
             self.uploadFinalized = true;
-            cb();
+            cb(null, res);
         }
     });
 }
