commit 3ee93841e356b31d325a68070b6a054462694c7e (refs/changes/34/2434/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-08-19T00:30:38+00:00 (2 years, 2 months ago)
    
    MANTA-3368 MPU commit does not return the computed MD5

diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index ded74d8..b12392c 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -516,6 +516,7 @@ function commit(req, res, next) {
             }, 'commit: completed');
 
             res.setHeader('Location', p);
+            res.setHeader('Computed-MD5', partsMD5);
             res.send(201);
             next();
         }
