From 3ee93841e356b31d325a68070b6a054462694c7e Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Sat, 19 Aug 2017 00:30:38 +0000
Subject: [PATCH] MANTA-3368 MPU commit does not return the computed MD5

---
 lib/uploads/commit.js | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index ded74d8..b12392c 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -516,6 +516,7 @@ function commit(req, res, next) {
             }, 'commit: completed');
 
             res.setHeader('Location', p);
+            res.setHeader('Computed-MD5', partsMD5);
             res.send(201);
             next();
         }
-- 
2.21.0

