From a5de0a5f1f6f7bd5dd05e3073b81bb81eca03953 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 6 Jul 2016 15:03:06 -0700
Subject: [PATCH] latest sdc-scripts

---
 deps/sdc-scripts | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/sdc-scripts b/deps/sdc-scripts
index 9a78955..6db4018 160000
--- a/deps/sdc-scripts
+++ b/deps/sdc-scripts
@@ -1 +1 @@
-Subproject commit 9a78955bf23c8b4d3c0dd083c69f062a456f33a5
+Subproject commit 6db4018c4db1e86aa14c6a0d4d1e7e50faaeeb4b
-- 
2.21.0

