From 57184d837a95d1da95466cb6cefb2ab7945da869 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 22 Mar 2019 16:39:05 -0700
Subject: [PATCH] TRITON-1347 switch node-triton test runner to node-tap

---
 Makefile     | 16 +++++++---------
 package.json |  2 +-
 2 files changed, 8 insertions(+), 10 deletions(-)

diff --git a/Makefile b/Makefile
index b2fac48..79bf7f0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 # Makefile for node-triton
 #
@@ -14,6 +14,8 @@ JSL_FILES_NODE	 = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 CLEAN_FILES += ./node_modules
+TAP_EXEC = ./node_modules/.bin/tap
+TEST_UNIT_JOBS ?= 4
 
 include ./tools/mk/Makefile.defs
 
@@ -29,17 +31,13 @@ test: test-unit test-integration
 
 .PHONY: test-unit
 test-unit:
-	NODE_NDEBUG= ./node_modules/.bin/tape test/unit/*.test.js
+	NODE_NDEBUG= $(TAP_EXEC) --timeout 600 -j $(TEST_UNIT_JOBS) \
+		-o ./test-unit.tap test/unit/*.test.js
 
 .PHONY: test-integration
 test-integration:
-	NODE_NDEBUG= ./node_modules/.bin/tape test/integration/*.test.js
-
-.PHONY: test-in-parallel
-test-in-parallel:
-	ls test/unit/*.test.js test/integration/*.test.js \
-	    | parallel ./node_modules/.bin/tape \
-	    | ./node_modules/.bin/tap-summary --no-ansi --no-progress
+	NODE_NDEBUG= $(TAP_EXEC) --timeout 600 -j $(TEST_UNIT_JOBS) \
+		-o ./test-integration.tap test/integration/*.test.js
 
 .PHONY: clean
 clean::
diff --git a/package.json b/package.json
index 499091a..58f9658 100644
--- a/package.json
+++ b/package.json
@@ -32,8 +32,8 @@
     "wordwrap": "1.0.0"
   },
   "devDependencies": {
+    "tap": "11.1.5",
     "tape": "4.2.0",
-    "tap-summary": "3.0.2",
     "uuid": "3.2.1"
   },
   "main": "./lib",
-- 
2.21.0

