From 7e500ce517eb558b0240f351f815a6cb5739a553 Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Wed, 27 Mar 2019 20:28:45 +0000
Subject: [PATCH] TRITON-1353 Triton systems should ship with HT disabled

---
 lib/models/server.js | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/lib/models/server.js b/lib/models/server.js
index 08f21ac..bf87b5a 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1258,6 +1258,21 @@ ModelServer.getBootParamsDefault = function (callback) {
                 params.kernel_flags[k] = server.kernel_flags[k];
             }
 
+            /*
+             * We want to default new / factory-reset CNs to disabling HT. The
+             * default boot params is originally generated via
+             * initializeBuckets(), but we want this policy to apply on upgrade,
+             * not just a completely fresh install. So we will mix in the HT
+             * setting here, if it's not been set by an admin already.
+             *
+             * The net effect here is that a new CN will have HT disabled, but
+             * a system that is merely upgraded will already have its own
+             * bootparams, and hence not see this change in behavior.
+             */
+            if (!params.kernel_args.hasOwnProperty('ht_enabled')) {
+                params.kernel_args['ht_enabled'] = 'false';
+            }
+
             params.boot_modules = server.boot_modules;
 
             params.default_console = server.default_console;
-- 
2.21.0

