commit 41bee0e1a44c245628da8f91c4517c93c6d3e0a5
Author: John Levon <john.levon@joyent.com>
Date:   2019-03-28T16:08:10+00:00 (6 months ago)
    
    TRITON-1353 Triton systems should ship with HT disabled

diff --git a/lib/models/server.js b/lib/models/server.js
index 08f21ac..b58d729 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1258,6 +1258,23 @@ ModelServer.getBootParamsDefault = function (callback) {
                 params.kernel_flags[k] = server.kernel_flags[k];
             }
 
+            /*
+             * We want to default new / factory-reset CNs to disabling HT. The
+             * default boot params is originally generated via
+             * initializeBuckets(), but we want this policy to apply on upgrade,
+             * not just a completely fresh install. So we will mix in the HT
+             * default setting from SAPI here.
+             *
+             * The net effect here is that a new CN will have HT disabled, but
+             * a system that is merely upgraded will already have its own
+             * bootparams, and hence not see this change in behavior.
+             * FIXME: update static.md
+             */
+            if (!params.kernel_args.hasOwnProperty('ht_enabled')) {
+                params.kernel_args['ht_enabled'] =
+                    ModelServer.getConfig().cnapi.ht_enabled_default;
+            }
+
             params.boot_modules = server.boot_modules;
 
             params.default_console = server.default_console;
diff --git a/sapi_manifests/cnapi/template b/sapi_manifests/cnapi/template
index 2dc3f6b..606cabf 100644
--- a/sapi_manifests/cnapi/template
+++ b/sapi_manifests/cnapi/template
@@ -43,7 +43,8 @@
 		"url": "http://{{{assets_admin_ips}}}"
 	},
 	"cnapi": {
-		"url": "http://{{{auto.ADMIN_IP}}}"
+		"url": "http://{{{auto.ADMIN_IP}}}",
+                "ht_enabled_default": {{{HT_ENABLED_DEFAULT}}}{{^HT_ENABLED_DEFAULT}}false{{/HT_ENABLED_DEFAULT}}
 	},
 	"imgapi": {
 		"url": "http://{{{IMGAPI_SERVICE}}}"
