commit 64f43d0ba30c112e0336689bb9fbae538453bbef (refs/changes/55/4555/3)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-08-27T14:51:27-07:00 (1 year, 1 month ago)
    
    MANTA-3830 mime ReDoS vuln that affects muskie

diff --git a/lib/server.js b/lib/server.js
index a7a3780..0e26d5d 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -193,8 +193,10 @@ function createServer(options, clients, name) {
     server.pre(function cleanupContentType(req, res, next) {
         var ct = req.headers['content-type'];
         /* JSSTYLED */
-        if (ct && !/.*\/.*/.test(ct))
-            req.headers['content-type'] = mime.lookup(ct);
+        if (ct && !/.*\/.*/.test(ct)) {
+            req.headers['content-type'] = mime.getType(ct) ||
+                'application/octet-stream';
+        }
         next();
     });
 
diff --git a/package.json b/package.json
index ec2a59a..e4ecfa5 100644
--- a/package.json
+++ b/package.json
@@ -33,7 +33,7 @@
         "lstream": "0.0.4",
         "mahi": "2.1.0",
         "marlin": "git+https://github.com/joyent/manta-marlin.git#master",
-        "mime": "1.2.11",
+        "mime": "2.3.1",
         "moray": "3.4.1",
         "moray-filter": "1.0.0",
         "once": "1.3.0",
