commit 6e5ad0805f860dd4ebeb7326a27949ddb70fb26b
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2019-07-16T17:29:59+00:00 (3 months ago)
    
    joyent/bugview#35 JSON endpoint broken by joyent/bugview#33

diff --git a/jirapub.js b/jirapub.js
index 8ac0f3a..740e206 100644
--- a/jirapub.js
+++ b/jirapub.js
@@ -408,7 +408,7 @@ handle_issue_json(req, res, next)
 			return;
 		}
 
-		mod_assert.arrayOfStrings(issue.fields.labels, 'labels');
+		mod_assert.arrayOfString(issue.fields.labels, 'labels');
 
 		if (!UNRESTRICTED &&
 		    issue.fields.labels.indexOf(CONFIG.label) === -1) {
