From 6e5ad0805f860dd4ebeb7326a27949ddb70fb26b Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Tue, 16 Jul 2019 17:29:59 +0000
Subject: [PATCH] joyent/bugview#35 JSON endpoint broken by joyent/bugview#33

---
 jirapub.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/jirapub.js b/jirapub.js
index 8ac0f3a..740e206 100644
--- a/jirapub.js
+++ b/jirapub.js
@@ -408,7 +408,7 @@ handle_issue_json(req, res, next)
 			return;
 		}
 
-		mod_assert.arrayOfStrings(issue.fields.labels, 'labels');
+		mod_assert.arrayOfString(issue.fields.labels, 'labels');
 
 		if (!UNRESTRICTED &&
 		    issue.fields.labels.indexOf(CONFIG.label) === -1) {
-- 
2.21.0

