commit 2eda7f0d4a57885472b0ad9ca289a7a64d4df1f9 (refs/changes/53/1353/1)
Author: David Pacheco <dap@joyent.com>
Date:   2017-01-26T16:27:46-08:00 (2 years, 8 months ago)
    
    MANTA-3127 update minnow to moray v3

diff --git a/Makefile b/Makefile
index a997378..71dd6b8 100644
--- a/Makefile
+++ b/Makefile
@@ -32,6 +32,10 @@ JSL_FILES_NODE   = $(JS_FILES)
 JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS    = -f tools/jsstyle.conf
 SMF_MANIFESTS_IN = smf/manifests/minnow.xml.in
+JSON_FILES	 = \
+    etc/config.coal.json \
+    package.json \
+    sapi_manifests/minnow/template
 
 #
 # Variables
diff --git a/etc/config.coal.json b/etc/config.coal.json
index bc25e29..4da60ef 100644
--- a/etc/config.coal.json
+++ b/etc/config.coal.json
@@ -13,14 +13,14 @@
                 "manta_storage_id": { "type": "string" }
             }
         },
-        "connectTimeout": 200,
-        "retry": {
-            "retries": null,
-            "minTimeout": 1000,
-            "maxTimeout": 60000
-        },
-        "host": "1.moray.coal.joyent.us",
-        "port": 2020
+        "morayConfig": {
+            "srvDomain": "{{STORAGE_MORAY_SHARD}}",
+            "cueballOptions": {
+                    "target": 1,
+                    "maximum": 1,
+                    "resolvers": [ "nameservice.{{DOMAIN_NAME}}" ]
+            }
+        }
     },
     "pingPort": 3030,
     "datacenter": "coal",
diff --git a/etc/config.lab.json b/etc/config.lab.json
deleted file mode 100644
index 106ae93..0000000
--- a/etc/config.lab.json
+++ /dev/null
@@ -1,34 +0,0 @@
-{
-    "moray": {
-        "bucket": {
-            "name": "manta_storage",
-            "index": {
-                "hostname": { "type": "string" },
-                "availableMB": { "type": "number" },
-                "percentUsed": { "type": "number" },
-                "server_uuid": { "type": "string" },
-                "timestamp": { "type": "number" },
-                "zone_uuid": { "type": "string" },
-                "manta_compute_id": { "type": "string" },
-                "manta_storage_id": { "type": "string" }
-            }
-        },
-        "connectTimeout": 200,
-        "maxConnections": 1,
-        "retry": {
-            "minTimeout": 1000,
-            "maxTimeout": 60000
-        },
-        "host": "1.moray.bh1.joyent.us",
-        "port": 2020
-    },
-    "pingPort": 3030,
-    "datacenter": "bh1",
-    "domain": "test.bh1.joyent.us",
-    "objectRoot": "/",
-    "server_uuid": "e05a6490-bcad-11e1-afa7-0800200c9a66",
-    "zone_uuid": "e91001d0-bcad-11e1-afa7-0800200c9a66",
-    "manta_compute_id": "compute-1",
-    "manta_storage_id": "storage-1",
-    "interval": 5000
-}
diff --git a/main.js b/main.js
index f77b381..4ad7820 100644
--- a/main.js
+++ b/main.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -111,6 +111,7 @@ function configure() {
     }
 
     assert.object(cfg.moray, 'config.moray');
+    assert.object(cfg.moray.morayConfig, 'config.moray.morayConfig');
     assertNonEmptyString(cfg.moray.bucket.name, 'cfg.moray.bucket.name');
     assertNonEmptyString(cfg.datacenter, 'cfg.datacenter');
     assertNonEmptyString(cfg.domain, 'cfg.domain');
@@ -120,8 +121,6 @@ function configure() {
     assertNonEmptyString(cfg.manta_compute_id, 'cfg.manta_compute_id');
     assertNonEmptyString(cfg.manta_storage_id, 'cfg.manta_storage_id');
 
-    cfg.moray.log = LOG;
-
     return (cfg);
 }
 
@@ -131,23 +130,13 @@ function configure() {
 
 function createMorayClient(opts, cb) {
     assert.object(opts, 'options');
-    assert.object(opts.log, 'options.log');
-    assert.optionalObject(opts.retry, 'options.retry');
     assert.func(cb, 'callback');
 
     cb = once(cb);
 
-    var retry = opts.retry || {};
-    retry.minTimeout = retry.minTimeout || 2000;
-    retry.maxTimeout = retry.maxTimeout || 120000;
-    var client = moray.createClient({
-        connectTimeout: opts.connectTimeout,
-        log: opts.log,
-        maxConnections: opts.maxConnections,
-        host: opts.host,
-        port: opts.port,
-        retry: retry
-    });
+    var moraycfg = opts.morayConfig;
+    moraycfg.log = LOG.child({ 'component': 'MorayClient' });
+    var client = moray.createClient(moraycfg);
 
     function onConnectSetup() {
         var bname = opts.bucket.name;
diff --git a/package.json b/package.json
index d934fc4..3791bb3 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
         "bunyan": "0.21.4",
         "dashdash": "1.3.2",
         "libmanta": "git+ssh://git@github.com:joyent/node-libmanta.git#master",
-        "moray": "^2.0.0",
+        "moray": "3.0.0",
         "node-uuid": "1.4.0",
         "once": "1.2.0",
         "statvfs": "2.0.0"
diff --git a/sapi_manifests/minnow/template b/sapi_manifests/minnow/template
index 93437ae..763fd7f 100644
--- a/sapi_manifests/minnow/template
+++ b/sapi_manifests/minnow/template
@@ -13,14 +13,14 @@
                                 "manta_storage_id": { "type": "string" }
                         }
                 },
-                "connectTimeout": 2000,
-                "maxConnections": 1,
-                "retry": {
-                        "minTimeout": 2000,
-                        "maxTimeout": 120000
-                },
-                "host": "{{STORAGE_MORAY_SHARD}}",
-                "port": 2020
+                "morayConfig": {
+                        "srvDomain": "{{STORAGE_MORAY_SHARD}}",
+                        "cueballOptions": {
+                                "target": 1,
+                                "maximum": 1,
+                                "resolvers": [ "nameservice.{{DOMAIN_NAME}}" ]
+                        }
+                }
         },
         "pingPort": 3030,
         "datacenter": "{{DATACENTER}}",
