commit 4a326e26a9c0c6ca6ed4bdc31041c5b8d29e2947
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-15T15:07:28-04:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index 9254389..c2d54ab 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -3,7 +3,7 @@
 	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/restdown"]
 	path = deps/restdown
 	url = https://github.com/trentm/restdown.git
diff --git a/deps/jsstyle b/deps/jsstyle
index b9f5092..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit b9f50929ee2df140fd2d111f053b20593e068cdc
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
