commit 4c24477f91dce301295ba4a8020d78d00c327240 (refs/changes/02/2702/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-10-03T17:59:11+00:00 (2 years ago)
    
    MANATEE-372 disable wal sender timeout

diff --git a/etc/postgresql.conf b/etc/postgresql.conf
index 33a8134..37f8395 100644
--- a/etc/postgresql.conf
+++ b/etc/postgresql.conf
@@ -210,6 +210,7 @@ full_page_writes = off      # recover from partial page writes
 max_wal_senders = 15    # max number of walsender processes
         # (change requires restart)
 wal_keep_segments = 1000    # in logfile segments, 16MB each; 0 disables
+wal_sender_timeout = 0      # don't timeout inactive replication connections
 #replication_timeout = 60s  # in milliseconds; 0 disables
 
 # - Master Server -
