commit 927553a13413aa4431888958b0a20c5eb257dea0 (refs/changes/30/30/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-07-05T19:07:31-07:00 (3 years, 3 months ago)
    
    joyent/node-zkstream#3 travis-ci is broken

diff --git a/test/basic.test.js b/test/basic.test.js
index 9155e0e..5d21003 100644
--- a/test/basic.test.js
+++ b/test/basic.test.js
@@ -58,7 +58,7 @@ mod_tape.test('find the test object', function (t) {
 		var req = zkc.list('/');
 		req.once('reply', function (pkt) {
 			t.strictEqual(pkt.opcode, 'GET_CHILDREN');
-			t.deepEqual(pkt.children, ['zookeeper', 'foo']);
+			t.deepEqual(pkt.children.sort(), ['foo', 'zookeeper']);
 
 			var req2 = zkc.get('/foo');
 			req2.once('reply', function (pkt2) {
diff --git a/test/zkserver.js b/test/zkserver.js
index 3368ac8..2536556 100644
--- a/test/zkserver.js
+++ b/test/zkserver.js
@@ -99,8 +99,11 @@ ZKServer.prototype.state_starting = function (on) {
 
 ZKServer.prototype.state_spawning = function (on) {
 	var self = this;
+	var opts = {};
+	opts.env = {};
+	opts.env.ZOO_LOG_DIR = this.zk_tmpdir;
 	this.zk_kid = mod_cproc.spawn(this.zk_cmd, ['start-foreground',
-	    this.zk_config]);
+	    this.zk_config], opts);
 	on(this.zk_kid, 'error', function (err) {
 		if (err.code === 'ENOENT') {
 			self.gotoState('starting');
