commit e14bc7dea34a543ef2c814d3b5bf392dd25b1f66 (refs/changes/34/3134/1)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2017-12-22T11:38:06-08:00 (1 year, 10 months ago)
    
    ADMINUI-2396 Login consistently fails due to UFDS connection timeout

diff --git a/lib/ufds.js b/lib/ufds.js
index 62f2a754..8a98688c 100644
--- a/lib/ufds.js
+++ b/lib/ufds.js
@@ -12,7 +12,7 @@ var sdc = require('sdc-clients');
 var UFDS = require('ufds');
 var assert = require('assert-plus');
 
-var UFDS_DEFAULT_CONNECT_TIMEOUT = 4000;
+var UFDS_DEFAULT_CONNECT_TIMEOUT = 10000;
 var UFDS_DEFAULT_CLIENT_TIMEOUT = 2000;
 var UFDS_DEFAULT_IDLE_TIMEOUT = 10000;
 
