From 262fe3bb44b97c1d3ce7aab95fd0ca0d0888556c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 1 Feb 2019 14:05:46 +0000
Subject: [PATCH] MANTA-4093 nginx should not modify its source tree during the
 build

---
 deps/nginx | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/nginx b/deps/nginx
index f054a7d..4d88576 160000
--- a/deps/nginx
+++ b/deps/nginx
@@ -1 +1 @@
-Subproject commit f054a7df29b99932cbe50c1216d545029b2e769c
+Subproject commit 4d885764f41be161c907cb8049d79cf0d6558981
-- 
2.21.0

