commit 34a3377d989931ef4c0c1338005506b10db359db
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2019-05-24T09:58:59-07:00 (5 months ago)
    
    TRITON-1704 new test for TRITON-1691 broke teardown

diff --git a/test/common.js b/test/common.js
index 0c18c1ff..ddb50d39 100644
--- a/test/common.js
+++ b/test/common.js
@@ -829,6 +829,12 @@ function teardown(clients, server, cb) {
         function (_, next) {
             ufds.deletePolicy(accUuid, userClient.policy.uuid, next);
         },
+        function (_, next) {
+            ufds.deleteRole(accUuid, userClient.xrole.uuid, next);
+        },
+        function (_, next) {
+            ufds.deletePolicy(accUuid, userClient.xpolicy.uuid, next);
+        },
         function (_, next) {
             deletePackage(userClient, SDC_128_PACKAGE, next);
         },
