From 462ffa19f23aedcd7493a9b1c7b0622bd8499e93 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Mon, 3 Apr 2017 15:12:13 -0700
Subject: [PATCH] TOOLS-1738 test-system is failing: pubapi-1003.test.js

---
 test/pubapi-1003.test.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/test/pubapi-1003.test.js b/test/pubapi-1003.test.js
index da03b90..00872e9 100644
--- a/test/pubapi-1003.test.js
+++ b/test/pubapi-1003.test.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -27,7 +27,7 @@ var test = helper.test;
 
 
 test('PUBAPI-1003', function (t) {
-    exec('openssl s_client -ssl3 -no_tls1 -connect '
+    exec('openssl s_client -ssl3 -connect '
             + '$(vmadm lookup -j alias=cloudapi0 | json -a nics '
             + '| json -c \'this.nic_tag==="external"\' 0.ip):443',
         function (err, stdout, stderr) {
-- 
2.21.0

