commit e23ab05382a68b98693cc9a288610455c480925e (refs/changes/79/1979/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-05-19T18:47:36+02:00 (2 years, 5 months ago)
    
    TOOLS-1782 sdcadm regression after TOOLS-1774 merge

diff --git a/lib/cli/do_update_agents.js b/lib/cli/do_update_agents.js
index 21ce9f6..b04760e 100644
--- a/lib/cli/do_update_agents.js
+++ b/lib/cli/do_update_agents.js
@@ -410,7 +410,7 @@ function updateAgents(options, callback) {
             });
         },
 
-        steps.agentServicesEnsureCreated,
+        steps.sapiEnsureAgentServices,
 
         function downloadAgentsshar(ctx, next) {
             if (filepath) {
