From 38343945783c7d47337979a442e877d4d958b4fb Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 22 Dec 2017 17:19:37 +0100
Subject: [PATCH] HEAD-2384 sapi0 instance should be created with the proper
 data

---
 scripts/headnode.sh | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/scripts/headnode.sh b/scripts/headnode.sh
index b34f7e44..86b0581d 100755
--- a/scripts/headnode.sh
+++ b/scripts/headnode.sh
@@ -645,6 +645,7 @@ HERE
   {"set_customer_metadata": {"dns_domain": "${CONFIG_dns_domain}"}}
 EOF
         fi
+
         setup_state_add "sapi_bootstrapped"
     fi
 }
@@ -693,6 +694,7 @@ function sapi_adopt()
 
     if [[ "${service_name}" == "sapi" ]]; then
       local sapi_url=http://${CONFIG_sapi_admin_ips}
+      local sapi_alias=$(vmadm get ${uuid}|json alias)
     else
       local sapi_url=http://${CONFIG_sapi_domain}
     fi
@@ -716,10 +718,17 @@ function sapi_adopt()
 
     i=0
     while [[ -z ${sapi_instance} && ${i} -lt 48 ]]; do
-        sapi_instance=$(curl ${sapi_url}/instances -sS -X POST \
-            -H content-type:application/json \
-            -d "{ \"service_uuid\" : \"${service_uuid}\", \"uuid\" : \"${uuid}\" }" \
-        | json -H uuid || true)
+        if [[ -n ${sapi_alias} ]]; then
+            sapi_instance=$(curl ${sapi_url}/instances -sS -X POST \
+                -H content-type:application/json \
+                -d "{ \"service_uuid\" : \"${service_uuid}\", \"uuid\" : \"${uuid}\", \"params\": { \"alias\": \"${sapi_alias}\"} }" \
+            | json -H uuid || true)
+        else
+            sapi_instance=$(curl ${sapi_url}/instances -sS -X POST \
+                -H content-type:application/json \
+                -d "{ \"service_uuid\" : \"${service_uuid}\", \"uuid\" : \"${uuid}\" }" \
+            | json -H uuid || true)
+        fi
         if [[ -z ${sapi_instance} ]]; then
             echo "Unable to adopt ${service_name} ${uuid} into sapi yet.  Sleeping..."
             sleep 5
-- 
2.21.0

