commit 5bd09d4ba3915661d3d9d053079386f1dad9a722 (refs/changes/02/2202/1)
Author: Jared Morrow <jm@joyent.com>
Date:   2017-07-10T15:55:13-06:00 (2 years, 3 months ago)
    
    MANTA-3109 manta log uploader retries multiple times in parallel

diff --git a/backup.sh b/backup.sh
old mode 100644
new mode 100755
index 3b143bf..c84a192
--- a/backup.sh
+++ b/backup.sh
@@ -31,6 +31,7 @@ export PATH=/opt/local/bin:$PATH
 # Immutables
 
 SSH_KEY=/root/.ssh/id_rsa
+BACKUP_LOCKFILE=/tmp/backup_lockfile
 
 MANTA_KEY_ID=$(ssh-keygen -l -f $SSH_KEY.pub | awk '{print $2}')
 MANTA_URL=$(json -f /opt/smartdc/common/etc/config.json manta.url)
@@ -68,6 +69,54 @@ function sign() {
 }
 
 
+# $1 -> lockfile to create
+# $$ pid of this script
+function create_lockfile()
+{
+    # Creating the tempfile can race, but
+    # ln is atomic, so that's the true locking
+    # operation
+    TEMPFILE="$1.$$"
+    LOCKFILE="$1.lock"
+    echo $$ > $TEMPFILE 2>/dev/null || {
+        echo "You don't have permission to access `dirname $TEMPFILE`"
+        return 1
+    }
+
+    # Create lock, remove TEMPFILE
+    ln $TEMPFILE $LOCKFILE 2>/dev/null && {
+        rm -f $TEMPFILE
+        return 0
+    }
+
+    STALE_PID=`< $LOCKFILE`
+    test "$STALE_PID" -gt "0" >/dev/null || {
+        rm -f $TEMPFILE
+        return 1
+    }
+
+    # Test if PID from lockfile is running
+    # If it is still running, the function will return here
+    kill -0 $STALE_PID 2>/dev/null && {
+        rm -f $TEMPFILE
+        return 1
+    }
+
+    # PID was stale, remove it
+    rm $LOCKFILE 2>/dev/null && {
+        echo "Removed stale lock file of process $STALE_PID"
+    }
+
+    ln $TEMPFILE $LOCKFILE 2>/dev/null && {
+        rm -f $TEMPFILE
+        return 0
+    }
+
+    rm -f $TEMPFILE
+    return 1
+}
+
+
 function manta_put() {
     sign || fail "unable to sign"
     curl -fisSk \
@@ -105,6 +154,14 @@ function mkdirp() {
 # And we transform them to this in manta:
 #     /poseidon/stor/logs/muskie/2012/10/17/20/0db94777.log
 
+# Do not run if this script is being run already
+create_lockfile $BACKUP_LOCKFILE || {
+    # Rare cases the tempfile would not get removed by create_lockfile on fail
+    # so be extra sure here
+    rm -f $BACKUP_LOCKFILE
+    fail "backup is already running"
+}
+
 for f in $(ls /var/log/manta/upload/*.log)
 do
     service=$(echo $f | cut -d _ -f 1 | cut -d / -f 6)
@@ -116,3 +173,6 @@ do
     manta_put "$key" "$LOG_TYPE" "-T $f"
     rm $f
 done
+
+# Remove lockfile
+rm -f $BACKUP_LOCKFILE.lock
