From 8835bee2c6f18f8f3da2296a5be63c42b84db907 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Fri, 8 Dec 2017 17:14:08 -0700
Subject: [PATCH] MANTA-3157 muskie picker ought to start even when there is no
 space

---
 lib/picker.js | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/lib/picker.js b/lib/picker.js
index 271fdbc..3ca1625 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -242,16 +242,15 @@ function poll() {
                 count++;
             });
 
-            // Don't replace if we got an empty "DB"
-            if (count > 0) {
-                self.datacenters = dcs;
-                self.db = obj;
-                self.emit('topology', self.db);
-            } else {
+            if (count === 0) {
                 self.log.warn('Picker.poll: could not find any minnow ' +
                     'instances');
             }
 
+            self.datacenters = dcs;
+            self.db = obj;
+            self.emit('topology', self.db);
+
             self.log.trace('Picker.poll: done');
         });
     });
-- 
2.21.0

