From 86700e200166defcc2ef3b36e4cee2ea6d01b30e Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Mon, 24 Apr 2017 11:48:23 +0200
Subject: [PATCH] OS-6072 increase ramdisk size by 3M for DEBUG kernels

---
 tools/build_live | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tools/build_live b/tools/build_live
index e8c98150..547e1edf 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -66,6 +66,11 @@ if [ "$(id -u)" != "0" ]; then
   exit 1
 fi
 
+# Increase RSIZE by 3M if we have a DEBUG kernel
+if strings ${ROOT}/proto/kernel/$ARCH64/genunix | grep -q "DEBUG enabled"; then
+    RSIZE=$(expr ${RSIZE} + 3000)
+fi
+
 # Send a copy of stdout and stderr to LOGFILE
 exec > >(tee ${LOGFILE}) 2>&1
 
-- 
2.21.0

