commit 4357a7a060589aba050f85ee71f7d16be71f1b1b (refs/changes/28/1828/3)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2017-05-22T12:26:07+02:00 (2 years, 5 months ago)
    
    OS-6072 increase ramdisk size by 3M for DEBUG kernels

diff --git a/tools/build_live b/tools/build_live
index 97bf6ee9..7981618c 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -65,6 +65,8 @@ bi_file_usr=
 bi_mnt_root=
 bi_mnt_usr=
 
+bi_root_size=
+
 bi_log=
 
 
@@ -786,7 +788,15 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_create_ramdisk 272000 "$bi_file_root"
+bi_root_size=272000
+if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
+    grep "DEBUG enabled" >/dev/null; then
+	#
+	# Increase root size by 3M if we have a DEBUG kernel.
+	#
+	bi_root_size=$(( bi_root_size + 3000))
+fi
+bi_create_ramdisk "$bi_root_size" "$bi_file_root"
 bi_lofi_root="$bi_lofi_last"
 bi_mnt_root="$bi_tmpdir/a"
 bi_create_ufs 12248 "$bi_lofi_root" "$bi_mnt_root" "/"
