From ec79905fb214a3430f213b9e1c2725c697781477 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Mon, 27 Mar 2017 13:52:24 +0200
Subject: [PATCH] TOOLS-1730 Add adminui to the list of HA-eligible services

---
 lib/procedures/index.js | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/procedures/index.js b/lib/procedures/index.js
index e07cadd..f81f7d2 100644
--- a/lib/procedures/index.js
+++ b/lib/procedures/index.js
@@ -27,6 +27,7 @@ var errors = require('../errors'),
 
 // --- HA ready services
 var HA_READY_SVCS = [
+    'adminui',
     'cloudapi',
     'cmon',
     'mahi',
-- 
2.21.0

