commit eced1d56ebc7a7e477b3238d98f85f0444a2eb29 (refs/changes/65/4765/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-08-29T17:18:36+00:00 (1 year, 1 month ago)
    
    OS-7157 lx_audit_emit_user_msg needs bounds check

diff --git a/usr/src/uts/common/brand/lx/os/lx_audit.c b/usr/src/uts/common/brand/lx/os/lx_audit.c
index 65b8d35e41..de68bcafb1 100644
--- a/usr/src/uts/common/brand/lx/os/lx_audit.c
+++ b/usr/src/uts/common/brand/lx/os/lx_audit.c
@@ -1234,6 +1234,9 @@ lx_audit_emit_user_msg(uint_t mtype, uint_t len, char *datap)
 	    lxzd->lxzd_audit_state == NULL)
 		return;
 
+	if (len >= sizeof (msg))
+		len = sizeof (msg) - 1;
+
 	mutex_enter(&p->p_splock);
 	sessid = p->p_sessp->s_sid;
 	mutex_exit(&p->p_splock);
