commit 3c7efbc85025b5d60ea6bc7311d0276e0957b5a6 (refs/changes/14/314/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-08-18T11:37:27-07:00 (3 years, 2 months ago)
    
    IMGAPI-581 update imgapi git submodules to latest

diff --git a/deps/javascriptlint b/deps/javascriptlint
index e1bd0ab..040bf5e 160000
--- a/deps/javascriptlint
+++ b/deps/javascriptlint
@@ -1 +1 @@
-Subproject commit e1bd0abfd424811af469d1ece3af131d95443924
+Subproject commit 040bf5e429969ae42d976b570fd64c9e17fab20f
diff --git a/deps/jsstyle b/deps/jsstyle
index da42b50..b94eb62 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit da42b50ceb12d431437b32efd4c411a8e2fac0c8
+Subproject commit b94eb629ee524d21b563c8b4461caa2bdf1becc6
diff --git a/deps/sdc-scripts b/deps/sdc-scripts
index 17fc5c8..d01c308 160000
--- a/deps/sdc-scripts
+++ b/deps/sdc-scripts
@@ -1 +1 @@
-Subproject commit 17fc5c8b83623078388b00260e5ad4be4f7348ac
+Subproject commit d01c308f63ec6044092b40973e51a9afe110a8dd
