commit f1b5ec66ac6a573765e42ff191df718d73237d6b (refs/changes/75/3975/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-05-18T18:18:38+02:00 (1 year, 5 months ago)
    
    TRITON-396 TRITON-348 breaks sdcadm call config-agent sync

diff --git a/lib/common.js b/lib/common.js
index 9533f9c..675690a 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -1114,7 +1114,7 @@ function callConfigAgentSync(opts, cb) {
         opts.server,
         /* JSSTYLED */
         format('/usr/sbin/zlogin %s ' +
-            '"/opt/smartdc/config-agent/build/node/bin/node' +
+            '"/opt/smartdc/config-agent/build/node/bin/node ' +
             '/opt/smartdc/config-agent/agent.js -f ' +
             '/opt/smartdc/config-agent/etc/config.json -s"', opts.vm)
     ];
