From 6e9799b0d5254a5188f889f2a20cf5c4146bd05f Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 9 Nov 2016 18:18:07 +0100
Subject: [PATCH] HEAD-2326 `sdc-server setup` support for disk layout

---
 tools/bin/sdc-server | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/tools/bin/sdc-server b/tools/bin/sdc-server
index 153a62f3..c1a66b46 100755
--- a/tools/bin/sdc-server
+++ b/tools/bin/sdc-server
@@ -1180,6 +1180,12 @@ function dispatchCommand(command, args, callback)
                     setupOpts.params.hostname = param[1];
                 } else if (param[0] === 'postsetup_script') {
                     setupOpts.params.postsetup_script = param[1];
+                } else if (param[0] === 'spares') {
+                    setupOpts.params.disk_spares = param[1];
+                } else if (param[0] === 'cache') {
+                    setupOpts.params.disk_cache = param[1];
+                } else if (param[0] === 'layout') {
+                    setupOpts.params.disk_layout = param[1];
                 } else {
                     var match = param[0].match(/^(.+)_nic$/);
                     if (!match) {
-- 
2.21.0

