commit 708435e140fde963582446501c53f772853d5fc3
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2019-04-05T15:55:50-07:00 (6 months ago)
    
    TRITON-1372 Server can get stuck in state "unknown"

diff --git a/lib/app.js b/lib/app.js
index 0d08827..6ffc1c8 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -549,7 +549,15 @@ function _setSysinfoTimer(serverUuid, lastSeen) {
     self.unsetupServers[serverUuid].lastUrSysinfo = lastSeen;
     self.unsetupServers[serverUuid].sysinfoTimer =
         setTimeout(function _markUnknown() {
-            self.log.debug('No sysinfo for %d seconds, ' +
+            if (self.observedHeartbeats[serverUuid]) {
+                self.log.info({
+                    serverUuid: serverUuid
+                }, 'Server is heartbeating. No longer relying on Ur sysinfo ' +
+                    'messages to set status');
+                return;
+            }
+
+            self.log.warn('No sysinfo for %d seconds, ' +
                 'marking status="unknown"',
                 UNSETUP_UR_SYSINFO_TIMEOUT_SECONDS);
 
@@ -605,6 +613,14 @@ function onSysinfoReceivedUr(message, routingKey) {
             sysinfo: sysinfo
         }, 'Got sysinfo from Ur and attempted update');
 
+        if (self.observedHeartbeats[serverUuid]) {
+            self.log.debug({
+                serverUuid: serverUuid
+            }, 'Server is heartbeating. Not setting status from Ur sysinfo ' +
+                'message');
+            return;
+        }
+
         if (sysinfo['Setup'].toString() !== 'true') {
             ModelServer.upsert(serverUuid, {status: 'running'}, {
                 etagRetries: 0
diff --git a/package.json b/package.json
index 3713afb..a638b0e 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "Triton Compute Node API",
-  "version": "1.23.0",
+  "version": "1.23.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
