From 1a6c621f1ef077e1cf6c719af6c1fc8063c3c80a Mon Sep 17 00:00:00 2001
From: Andrey Yatskov <andrey.yatskov@silvertreesystems.com>
Date: Tue, 26 Jul 2016 20:42:59 +0300
Subject: [PATCH] PORTAL-2165 The ability to append fields to the log record

---
 lib/cloudapi.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/cloudapi.js b/lib/cloudapi.js
index 3a415c1..b7f739e 100644
--- a/lib/cloudapi.js
+++ b/lib/cloudapi.js
@@ -170,6 +170,7 @@ function _addToQuery(req, obj) {
  * @param {Object} options object (required):
  *        - {String} url (required) CloudAPI location.
  *        - {String} account (optional) the login name to use (default my).
+ *        - {Object} logFields (optional) the additional fields that will be included in log records.
  *        - {Number} logLevel (optional) an enum value for the logging level.
  *        - {String} version (optional) api version (default ~7.2).
  *        - {Function} sign (required) callback function to use for signing
@@ -188,6 +189,8 @@ function CloudAPI(options) {
     assert.func(options.sign, 'options.sign');
     assert.optionalString(options.account, 'options.account');
 
+    if (options.logFields)
+        log = log.child(options.logFields);
     if (options.logLevel)
         log.level(options.logLevel);
     options.log = log;
-- 
2.21.0

