From 1364511dc1da77ae19ce6cbd10b5be88b0331c68 Mon Sep 17 00:00:00 2001
From: Nick Zivkovic <nick.zivkovic@joyent.com>
Date: Wed, 24 Aug 2016 23:59:33 +0000
Subject: [PATCH] MANATEE-312 manatee should not set snapdir to visible

---
 boot/setup.sh | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index cb69ec6..b946d5a 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -86,6 +86,9 @@ function common_manatee_setup {
     echo "creating manatee dataset"
     zfs create -o canmount=noauto $DATASET
 
+    echo "make snapdir property match the ancestor's"
+    zfs inherit -r snapdir $DATASET
+
     # create postgres user
     echo "creating postgres user"
     useradd postgres
@@ -114,10 +117,6 @@ function common_manatee_setup {
     chown postgres $PG_DIR
     chmod 700 $PG_DIR
 
-    # make .zfs dir visible for snapshots
-    echo "make snapshot dir visible"
-    zfs set snapdir=visible $DATASET
-
     # add pg log dir
     mkdir -p $PG_LOG_DIR
     chown -R postgres $PG_LOG_DIR
-- 
2.21.0

