commit 7e020707ef9e496afe4cc8453afc60da2c75b891 (refs/changes/33/2533/2)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2017-09-08T13:33:21-07:00 (2 years, 1 month ago)
    
    DOCKER-1064 docker cp client exiting can cause stuck zlogin

diff --git a/lib/docker-stdio.js b/lib/docker-stdio.js
index 03cfbb5..1bdd060 100644
--- a/lib/docker-stdio.js
+++ b/lib/docker-stdio.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -32,6 +32,8 @@ var vmadm = require('vmadm');
 var LineStream = require('lstream');
 var wait_flag = require('./update-wait-flag');
 
+var DATA_TIMEOUT_MS = 5000;
+
 var commands = {};
 var CTRL_P = '\u0010';
 var CTRL_Q = '\u0011';
@@ -199,6 +201,7 @@ function createDockerFileReadStreamServer(opts, callback) {
     var zonepath = path.join('/zones', opts.uuid);
     var root = path.join(zonepath, 'root');
     var abspath = path.join(root, norm);
+    var dataTimeout;
 
     var error;
     var containerPathStat;
@@ -289,6 +292,25 @@ function createDockerFileReadStreamServer(opts, callback) {
         callback(null, tcpServer, { containerPathStat: containerPathStat });
     });
 
+    function resetDataTimeout(streamProc) {
+        clearTimeout(dataTimeout);
+        dataTimeout = setTimeout(function () {
+            log.error('onConnectionZoneRunning(read): ' +
+                    'data timeout; terminating archive process');
+            streamProc.kill();
+        }, DATA_TIMEOUT_MS).unref();
+    }
+
+    function monitorForDataTimeout(streamProc) {
+        streamProc.stdout.on('data', function () {
+            resetDataTimeout();
+        });
+
+        streamProc.stdout.on('end', function () {
+            clearTimeout(dataTimeout);
+        });
+    }
+
     function onConnectionZoneRunning(socket) {
         clearTimeout(serverTimeout);
 
@@ -317,13 +339,15 @@ function createDockerFileReadStreamServer(opts, callback) {
                       'zlogin error output');
         });
 
-        streamProc.on('end', function (err) {
+        streamProc.on('exit', function (err) {
+            clearTimeout(dataTimeout);
+            tcpServer.close();
             if (err) {
                 log.error(err);
             }
-            tcpServer.close();
         });
 
+        monitorForDataTimeout(streamProc);
         streamProc.stdout.pipe(socket);
     }
 
@@ -358,13 +382,15 @@ function createDockerFileReadStreamServer(opts, callback) {
                       'chroot-gtar error output');
         });
 
-        streamProc.on('end', function (err) {
+        streamProc.on('exit', function (err) {
+            clearTimeout(dataTimeout);
+            tcpServer.close();
             if (err) {
                 log.error(err);
             }
-            tcpServer.close();
         });
 
+        monitorForDataTimeout(streamProc);
         streamProc.stdout.pipe(socket);
     }
 }
@@ -510,7 +536,7 @@ function createDockerFileWriteStreamServer(opts, callback) {
             log.error({ errorOutput: data.toString() }, 'zlogin error output');
         });
 
-        streamProc.on('end', function (err) {
+        streamProc.on('exit', function (err) {
             if (err) {
                 log.error(err);
             }
@@ -555,7 +581,7 @@ function createDockerFileWriteStreamServer(opts, callback) {
             console.log(data.toString());
         });
 
-        streamProc.on('end', function (err) {
+        streamProc.on('exit', function (err) {
             if (err) {
                 log.error(err);
             }
