commit 4566f02dc395f532b848223f58d678f7a834b538 (refs/changes/96/996/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton image list

diff --git a/lib/do_image/do_list.js b/lib/do_image/do_list.js
index d9b917c..cc00b40 100644
--- a/lib/do_image/do_list.js
+++ b/lib/do_image/do_list.js
@@ -63,42 +63,48 @@ function do_list(subcmd, opts, args, callback) {
         listOpts.state = 'all';
     }
 
-    this.top.tritonapi.listImages(listOpts, function onRes(err, imgs, res) {
-        if (err) {
-            return callback(err);
+    var tritonapi = this.top.tritonapi;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            callback(setupErr);
         }
+        tritonapi.listImages(listOpts, function onRes(err, imgs, res) {
+            if (err) {
+                return callback(err);
+            }
 
-        if (opts.json) {
-            common.jsonStream(imgs);
-        } else {
-            // Add some convenience fields
-            // Added fields taken from imgapi-cli.git.
-            for (var i = 0; i < imgs.length; i++) {
-                var img = imgs[i];
-                img.shortid = img.id.split('-', 1)[0];
-                if (img.published_at) {
-                    // Just the date.
-                    img.pubdate = img.published_at.slice(0, 10);
-                    // Normalize on no milliseconds.
-                    img.pub = img.published_at.replace(/\.\d+Z$/, 'Z');
-                }
-                if (img.files && img.files[0]) {
-                    img.size = img.files[0].size;
+            if (opts.json) {
+                common.jsonStream(imgs);
+            } else {
+                // Add some convenience fields
+                // Added fields taken from imgapi-cli.git.
+                for (var i = 0; i < imgs.length; i++) {
+                    var img = imgs[i];
+                    img.shortid = img.id.split('-', 1)[0];
+                    if (img.published_at) {
+                        // Just the date.
+                        img.pubdate = img.published_at.slice(0, 10);
+                        // Normalize on no milliseconds.
+                        img.pub = img.published_at.replace(/\.\d+Z$/, 'Z');
+                    }
+                    if (img.files && img.files[0]) {
+                        img.size = img.files[0].size;
+                    }
+                    var flags = [];
+                    if (img.origin) flags.push('I');
+                    if (img['public']) flags.push('P');
+                    if (img.state !== 'active') flags.push('X');
+                    img.flags = flags.length ? flags.join('') : undefined;
                 }
-                var flags = [];
-                if (img.origin) flags.push('I');
-                if (img['public']) flags.push('P');
-                if (img.state !== 'active') flags.push('X');
-                img.flags = flags.length ? flags.join('') : undefined;
-            }
 
-            tabula(imgs, {
-                skipHeader: opts.H,
-                columns: columns,
-                sort: sort
-            });
-        }
-        callback();
+                tabula(imgs, {
+                    skipHeader: opts.H,
+                    columns: columns,
+                    sort: sort
+                });
+            }
+            callback();
+        });
     });
 }
 
