From 31042d6eb1d03b4e54662f51ef3366f9346617a0 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 26 Apr 2019 10:42:46 +0200
Subject: [PATCH] TRITON-1299 need logarchiver-agent service to run along side
 hermes-actor

---
 deps/hermes | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/hermes b/deps/hermes
index f607e55..1cf50fa 160000
--- a/deps/hermes
+++ b/deps/hermes
@@ -1 +1 @@
-Subproject commit f607e55d1902dca6a3e97b1d7d72071a3be0ec93
+Subproject commit 1cf50fa80a674a5a427df733ac58eb162b7aadba
-- 
2.21.0

