From 1a69709b8e83d53c722f91117c9535d8274fa07d Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 11 Jul 2016 14:42:23 -0700
Subject: [PATCH] OS-5508 rsyslog should not send kernel NOTICE to the console

---
 overlay/generic/etc/rsyslog.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/overlay/generic/etc/rsyslog.conf b/overlay/generic/etc/rsyslog.conf
index cc37dfb9..f60451ca 100644
--- a/overlay/generic/etc/rsyslog.conf
+++ b/overlay/generic/etc/rsyslog.conf
@@ -8,7 +8,7 @@ $ModLoad imsolaris
 $ModLoad imtcp
 $ModLoad imudp
 
-*.err;kern.notice;auth.notice			/dev/sysmsg
+*.err;kern.warning;auth.notice			/dev/sysmsg
 *.err;kern.debug;daemon.notice;mail.crit	/var/adm/messages
 
 *.alert;kern.err;daemon.err			operator
-- 
2.21.0

