commit 700e3afc66a391a384cb14ae19c0893d0689ac9c (refs/changes/26/4726/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-08-20T15:35:14-07:00 (1 year, 2 months ago)
    
    TRITON-708 CNS redis pool cap too high

diff --git a/server.js b/server.js
index 40d4c07..84be232 100644
--- a/server.js
+++ b/server.js
@@ -40,10 +40,10 @@ var redisPool = new cueball.ConnectionPool({
 	defaultPort: 6379,
 	log: log,
 	spares: 4,
-	maximum: 100,
+	maximum: 12,
 	recovery: {
 		default: {
-			timeout: 1000,
+			timeout: 2000,
 			delay: 500,
 			retries: 5
 		}
