commit 77195c1816fbcb90ba4c2cbe0d039b437a819f51 (refs/changes/03/2303/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-07-31T13:07:02+00:00 (2 years, 2 months ago)
    
    8528 Want Chelsio T6 support

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 187fb9a1..f9d9080e 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -1152,6 +1152,28 @@ t4nex "pciex1425,5499"
 t4nex "pciex1425,549a"
 t4nex "pciex1425,549b"
 t4nex "pciex1425,549c"
+t4nex "pciex1425,549d"
+t4nex "pciex1425,549e"
+t4nex "pciex1425,549f"
+t4nex "pciex1425,54a0"
+t4nex "pciex1425,54a1"
+t4nex "pciex1425,6400"
+t4nex "pciex1425,6401"
+t4nex "pciex1425,6402"
+t4nex "pciex1425,6403"
+t4nex "pciex1425,6404"
+t4nex "pciex1425,6404"
+t4nex "pciex1425,6406"
+t4nex "pciex1425,6407"
+t4nex "pciex1425,6408"
+t4nex "pciex1425,6409"
+t4nex "pciex1425,640d"
+t4nex "pciex1425,6410"
+t4nex "pciex1425,6411"
+t4nex "pciex1425,6414"
+t4nex "pciex1425,6415"
+t4nex "pciex1425,6480"
+t4nex "pciex1425,6481"
 tavor "pci15b3,5a44"
 tavor "pci15b3,5a45"
 tavor "pci15b3,6278"
