commit 9d648f91991f8e2d377e490e33079c4b4a392d3a (refs/changes/77/877/1)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2016-11-10T14:53:02-08:00 (2 years, 11 months ago)
    
    HEAD-2330 sdc-useradm should not hardcode UFDS passwords

diff --git a/lib/sdc-useradm.js b/lib/sdc-useradm.js
index 83d87f9..828fff6 100755
--- a/lib/sdc-useradm.js
+++ b/lib/sdc-useradm.js
@@ -198,7 +198,7 @@ CLI.prototype.getLocalUfdsClient = function (callback) {
 
     var options = {
         bindDN: 'cn=root',
-        bindPassword: 'secret',
+        bindPassword: config.ufds_ldap_root_pw,
         log: log.child({ufds: 'local'}, true),
         url: 'ldaps://' + config.ufds_domain,
         connectTimeout: 15000,
@@ -223,7 +223,7 @@ CLI.prototype.getMasterUfdsClient = function (callback) {
 
     var options = {
         bindDN: 'cn=root',
-        bindPassword: 'secret',
+        bindPassword: config.ufds_remote_ldap_root_pw || config.ufds_ldap_root_pw,
         log: log.child({ufds: 'master'}, true),
         url: 'ldaps://' + config.ufds_remote_ip,
         connectTimeout: 10000,
