commit 7ec72234afd22982a19425563ba9bd54e31f4647 (refs/changes/77/877/2)
Author: Angela Fong <angela.fong@joyent.com>
Date:   2016-11-10T16:31:06-08:00 (2 years, 11 months ago)
    
    HEAD-2330 sdc-useradm should not hardcode UFDS passwords

diff --git a/lib/sdc-useradm.js b/lib/sdc-useradm.js
index 83d87f9..535ac24 100755
--- a/lib/sdc-useradm.js
+++ b/lib/sdc-useradm.js
@@ -198,7 +198,7 @@ CLI.prototype.getLocalUfdsClient = function (callback) {
 
     var options = {
         bindDN: 'cn=root',
-        bindPassword: 'secret',
+        bindPassword: config.ufds_ldap_root_pw,
         log: log.child({ufds: 'local'}, true),
         url: 'ldaps://' + config.ufds_domain,
         connectTimeout: 15000,
@@ -223,7 +223,7 @@ CLI.prototype.getMasterUfdsClient = function (callback) {
 
     var options = {
         bindDN: 'cn=root',
-        bindPassword: 'secret',
+        bindPassword: config.ufds_remote_ldap_root_pw,
         log: log.child({ufds: 'master'}, true),
         url: 'ldaps://' + config.ufds_remote_ip,
         connectTimeout: 10000,
diff --git a/sapi_manifests/sdc/template b/sapi_manifests/sdc/template
index b7bc722..a171415 100644
--- a/sapi_manifests/sdc/template
+++ b/sapi_manifests/sdc/template
@@ -21,6 +21,12 @@
     "ufds_ldap_root_dn": "{{{ufds_ldap_root_dn}}}",
     "ufds_ldap_root_pw": "{{{ufds_ldap_root_pw}}}",
     "ufds_remote_ip": "{{{ufds_remote_ip}}}",
+    {{^ufds_remote_ldap_root_pw}}
+    "ufds_remote_ldap_root_pw": "{{{ufds_ldap_root_pw}}}",
+    {{/ufds_remote_ldap_root_pw}}
+    {{#ufds_remote_ldap_root_pw}}
+    "ufds_remote_ldap_root_pw": "{{{ufds_remote_ldap_root_pw}}}",
+    {{/ufds_remote_ldap_root_pw}}
     "vmapi_domain": "{{{vmapi_domain}}}",
     "workflow_domain": "{{{workflow_domain}}}"
 }
