From ed164c7b0ccf7c0c31442904b42169ab0b46c2a2 Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Mon, 8 May 2017 10:49:54 -0700
Subject: [PATCH] TOOLS-1768 sdcadm update to exclude piranha (portal)

---
 lib/cli/index.js | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/lib/cli/index.js b/lib/cli/index.js
index e887624..e3a5b5a 100644
--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -377,7 +377,10 @@ CLI.prototype._specFromArgs = function _specFromArgs(opts, args, cb) {
                 // HEAD-2167 sdcsso has been removed, but the service still
                 // exists in some SDC installs (e.g. JPC), so drop update
                 // attempts of it.
-                if (name === 'sdcsso') {
+                // PORTAL-2801 sdcsso and piranha service config are now
+                // managed with SAPI. Some SDC installs may have them but
+                // update is still managed outside of sdcadm.
+                if (name === 'sdcsso' || name === 'piranha') {
                     return false;
                 }
                 // ditto for vcapi
-- 
2.21.0

