commit dc394b158f9650c62d0ddc8f16f2a68b5ff69ed3 (refs/changes/15/1115/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-12-14T10:59:36-08:00 (2 years, 10 months ago)
    
    joyent/node-triton#154 #108 changes broke `triton cloudapi ...`

diff --git a/lib/do_cloudapi.js b/lib/do_cloudapi.js
index e0ed906..201f3fd 100644
--- a/lib/do_cloudapi.js
+++ b/lib/do_cloudapi.js
@@ -11,7 +11,9 @@
  */
 
 var http = require('http');
+var vasync = require('vasync');
 
+var common = require('./common');
 var errors = require('./errors');
 
 
@@ -62,27 +64,34 @@ function do_cloudapi(subcmd, opts, args, callback) {
         }
     }
 
-    this.tritonapi.cloudapi._request(reqOpts, function (err, req, res, body) {
-        if (err) {
-            callback(err);
-            return;
-        }
-        if (opts.headers || reqOpts.method === 'head') {
-            console.error('%s/%s %d %s',
-                req.connection.encrypted ? 'HTTPS' : 'HTTP',
-                res.httpVersion,
-                res.statusCode,
-                http.STATUS_CODES[res.statusCode]);
-            Object.keys(res.headers).forEach(function (key) {
-                console.error('%s: %s', key, res.headers[key]);
+    vasync.pipeline({arg: {cli: this}, funcs: [
+        common.cliSetupTritonApi,
+
+        function callCloudapi(ctx, next) {
+            var cloudapi = ctx.cli.tritonapi.cloudapi;
+            cloudapi._request(reqOpts, function (err, req, res, body) {
+                if (err) {
+                    next(err);
+                    return;
+                }
+                if (opts.headers || reqOpts.method === 'head') {
+                    console.error('%s/%s %d %s',
+                        req.connection.encrypted ? 'HTTPS' : 'HTTP',
+                        res.httpVersion,
+                        res.statusCode,
+                        http.STATUS_CODES[res.statusCode]);
+                    Object.keys(res.headers).forEach(function (key) {
+                        console.error('%s: %s', key, res.headers[key]);
+                    });
+                    console.error();
+                }
+
+                if (reqOpts.method !== 'head')
+                    console.log(JSON.stringify(body, null, 4));
+                next();
             });
-            console.error();
         }
-
-        if (reqOpts.method !== 'head')
-            console.log(JSON.stringify(body, null, 4));
-        callback();
-    });
+    ]}, callback);
 }
 
 do_cloudapi.options = [
