commit b77f7705266280415eaf1846ae53bfa4156c4a02 (refs/changes/39/639/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-10-10T17:12:39-04:00 (3 years ago)
    
    RELENG-612 prefer (git+)https urls

diff --git a/.gitmodules b/.gitmodules
index d65f40b..62eb074 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
-	url = git://github.com/davepacheco/javascriptlint.git
+	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/davepacheco/jsstyle.git
