commit 6e66ca04a4805c929080eca11fb7ae4268f00630
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-05-07T08:00:54-07:00 (5 months ago)
    
    TRITON-1647 need to save vmapi runtests tap output

diff --git a/test/runtests b/test/runtests
index 9960a0c..555a2f1 100755
--- a/test/runtests
+++ b/test/runtests
@@ -210,5 +210,5 @@ fi
 if [[ -n "$test_files" ]]; then
     VMAPI_IP=$VMAPI_IP CNAPI_IP=$CNAPI_IP NAPI_IP=$NAPI_IP \
     PATH=$NODE_INSTALL/bin:$PATH $NODEUNIT --reporter \
-    $opt_reporter $test_files
+    $opt_reporter $test_files | tee $OUTPUT_DIR/vmapi.tap
 fi
