commit eeef88d789016ea9d3a2b69f45b470dc164ee6f9 (refs/changes/42/5142/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2018-11-27T16:49:30+00:00 (11 months ago)
    
    OS-7393 9903 merge lost OS-3752 modifications

diff --git a/usr/src/uts/common/sys/stream.h b/usr/src/uts/common/sys/stream.h
index 7dd1cf4723..efcd2a9194 100644
--- a/usr/src/uts/common/sys/stream.h
+++ b/usr/src/uts/common/sys/stream.h
@@ -651,7 +651,7 @@ struct stroptions {
 struct struiod {
 	mblk_t		*d_mp;		/* pointer to mblk (chain) */
 	uio_t		d_uio;		/* uio info */
-	iovec_t d_iov[DEF_IOV_MAX];	/* iov referenced by uio */
+	iovec_t		*d_iov;		/* iov referenced by uio */
 };
 
 /*
