commit 0d6075d4205f98d9180bb879b1241a166766dd30
Author: Trent Mick <trentm@gmail.com>
Date:   2019-03-15T10:14:04-07:00 (7 months ago)
    
    TRITON-1 Track AUTHAPI rejection implementation among Triton components

diff --git a/Makefile b/Makefile
index 4d998e5..91deb2f 100644
--- a/Makefile
+++ b/Makefile
@@ -4,6 +4,8 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 #
 
+# hi from trent, testing
+
 #
 # Copyright 2019, Joyent, Inc.
 #
