From 0d6075d4205f98d9180bb879b1241a166766dd30 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 15 Mar 2019 10:14:04 -0700
Subject: [PATCH] TRITON-1 Track AUTHAPI rejection implementation among Triton
 components

---
 Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile b/Makefile
index 4d998e5..91deb2f 100644
--- a/Makefile
+++ b/Makefile
@@ -4,6 +4,8 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 #
 
+# hi from trent, testing
+
 #
 # Copyright 2019, Joyent, Inc.
 #
-- 
2.21.0

