commit eb786d929f116a7192fff2db990c5fc873e74f65 (refs/changes/67/2967/1)
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   2017-11-20T16:28:04+00:00 (1 year, 11 months ago)
    
    OS-6463 Clearing a thread name before setting causes panic

diff --git a/usr/src/uts/common/brand/lx/syscall/lx_prctl.c b/usr/src/uts/common/brand/lx/syscall/lx_prctl.c
index 4aeb2b191e..2662826281 100644
--- a/usr/src/uts/common/brand/lx/syscall/lx_prctl.c
+++ b/usr/src/uts/common/brand/lx/syscall/lx_prctl.c
@@ -200,8 +200,10 @@ lx_prctl(int opt, uintptr_t data)
 		 * Linux.
 		 */
 		mutex_enter(&p->p_lock);
-		(void) strncpy(p->p_user.u_comm, t->t_name, MAXCOMLEN + 1);
-		(void) strncpy(p->p_user.u_psargs, t->t_name, PSARGSZ);
+		(void) strncpy(p->p_user.u_comm,
+		    t->t_name != NULL ? t->t_name : name, MAXCOMLEN + 1);
+		(void) strncpy(p->p_user.u_psargs,
+		    t->t_name != NULL ? t->t_name : name, PSARGSZ);
 		mutex_exit(&p->p_lock);
 		return (0);
 	}
