From 0400849cd69b5b9058eb0ebccc8277d97fe708a4 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 29 Jul 2019 19:32:24 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index dcf15a65..01f669da 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit dcf15a6507556313d7925c6914c0040fbc52161e
+Subproject commit 01f669da1f5cb1439d573fc42efd2dc43a7c6184
-- 
2.21.0

