commit 86a27399126ee20e768d4a0b070d93ad743e7723 (refs/changes/01/4001/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-05-24T17:43:35+02:00 (1 year, 5 months ago)
    
    OS-6912 bhyve virtio-blk should sync when needed

diff --git a/usr/src/cmd/bhyve/block_if.c b/usr/src/cmd/bhyve/block_if.c
index 86978bbda2..0bc9952a56 100644
--- a/usr/src/cmd/bhyve/block_if.c
+++ b/usr/src/cmd/bhyve/block_if.c
@@ -301,12 +301,13 @@ blockif_proc(struct blockif_ctxt *bc, struct blockif_elem *be, uint8_t *buf)
 		}
 		break;
 	case BOP_FLUSH:
-		if (bc->bc_ischr) {
 #ifdef	__FreeBSD__
+		if (bc->bc_ischr) {
 			if (ioctl(bc->bc_fd, DIOCGFLUSH))
 				err = errno;
+		} else
 #endif
-		} else if (fsync(bc->bc_fd))
+		if (fsync(bc->bc_fd))
 			err = errno;
 		break;
 	case BOP_DELETE:
diff --git a/usr/src/cmd/bhyve/pci_virtio_block.c b/usr/src/cmd/bhyve/pci_virtio_block.c
index 0c72c1f503..145412e82b 100644
--- a/usr/src/cmd/bhyve/pci_virtio_block.c
+++ b/usr/src/cmd/bhyve/pci_virtio_block.c
@@ -214,6 +214,11 @@ pci_vtblk_done(struct blockif_req *br, int err)
 	pthread_mutex_unlock(&sc->vsc_mtx);
 }
 
+static void
+pci_vtblk_nullcb(struct blockif_req *br, int err)
+{
+}
+
 static void
 pci_vtblk_proc(struct pci_vtblk_softc *sc, struct vqueue_info *vq)
 {
@@ -280,6 +285,14 @@ pci_vtblk_proc(struct pci_vtblk_softc *sc, struct vqueue_info *vq)
 		break;
 	case VBH_OP_WRITE:
 		err = blockif_write(sc->bc, &io->io_req);
+#ifndef __FreeBSD__
+		if (err == 0 &&
+		    (sc->vbsc_vs.vs_negotiated_caps & VTBLK_F_FLUSH) == 0) {
+			static struct blockif_req breq =
+			    { .br_callback = pci_vtblk_nullcb };
+			(void) blockif_flush(sc->bc, &breq);
+		}
+#endif
 		break;
 	case VBH_OP_FLUSH:
 	case VBH_OP_FLUSH_OUT:
