From e1ddc83db339f8f5103f912b7cb6bf788b73d9b0 Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Wed, 4 Apr 2018 10:00:53 +0000
Subject: [PATCH] OS-6727 bhyve should provide zone UUID to guest via SMBIOS

---
 src/sysinfo | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/src/sysinfo b/src/sysinfo
index 67873b78..f6fddc55 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -101,16 +101,27 @@ function normalize_mac()
 
 function get_smbios_system_info()
 {
+
     # This puts the variables we're pulling out into the local environment
     eval $(smbios -t SMB_TYPE_SYSTEM | tr -d "\"" \
-        | egrep "Manufacturer: |Product: |Serial Number: |UUID: |SKU Number: |Version: |Family: " \
+	| egrep "Manufacturer: |Product: |Serial Number: |UUID (Endian-corrected): |UUID: |SKU Number: |Version: |Family: " \
         | sed -e 's/^ *//' \
         -e 's/: /="/' \
         -e 's/ *$/"/' \
         -e 's/Serial Number/Serial_Number/' \
         -e 's/SKU Number/SKU_Number/' \
         -e 's/Version/HW_Version/' \
-        -e 's/Family/HW_Family/')
+        -e 's/Family/HW_Family/' \
+	-e 's/UUID (Endian-corrected)/Fixed_UUID/')
+
+    vendor=$(smbios -t SMB_TYPE_BIOS | grep "Vendor: " \
+        | sed -e 's/^.*Vendor: //;s/ *$//')
+
+    # If we are a bhyve guest, we can use the correct UUID format; otherwise we
+    # use the network-endian one for compatibility reasons.
+    if [ "$vendor" = "BHYVE" -a -n "$Fixed_UUID" ]; then
+         UUID=$Fixed_UUID
+    fi
 
     # overwrite UUID if config dictates otherwise
     tmp_uuid=$(/usr/bin/bootparams | grep "^override_uuid=" | cut -f2 -d'=')
-- 
2.21.0

