From b1631a99319b635d2a6843a1d7eba4d1e78080f1 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Fri, 11 Jan 2019 15:36:52 +0000
Subject: [PATCH] 9979 Support python3 for in-gate tools [fix more broken build
 issues]

---
 configure | 1 +
 1 file changed, 1 insertion(+)

diff --git a/configure b/configure
index baec0161..14e15d8c 100755
--- a/configure
+++ b/configure
@@ -320,6 +320,7 @@ FLEX=/opt/local/bin/flex;			export FLEX
 GNUXGETTEXT=/opt/local/bin/xgettext;		export GNUXGETTEXT
 PYTHON=/opt/local/bin/python2.7;		export PYTHON
 BUILDPY3='#';					export BUILDPY3
+BUILDPY3TOOLS='#';				export BUILDPY3TOOLS
 #
 # Note that this isn't the strap perl: that only really exists for the benefit
 # of projects/illumos-extra/openssl
-- 
2.21.0

