commit 79e13e2ceabcef7b1c61d651d4aff761faeb0b29 (refs/changes/84/4984/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-10-20T21:01:02+00:00 (12 months ago)
    
    OS-7317 vmadm "remove_filesystems" is both broken and undocumented

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 5cbe43f5..dfc7d02f 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -5990,7 +5990,7 @@ function buildFilesystemZonecfg(vmobj, payload, options)
     // remove is a list of disk paths, add a remove for each now.
     for (filesystem in remove) {
         filesystem = remove[filesystem];
-        zcfg = zcfg + 'remove fs match=' + filesystem + '\n';
+        zcfg = zcfg + 'remove fs dir=' + filesystem + '\n';
     }
 
     for (filesystem in add) {
@@ -6519,7 +6519,7 @@ function buildZonecfgUpdate(vmobj, payload, log)
     if (payload.hasOwnProperty('filesystems')
         || payload.hasOwnProperty('add_filesystems')
         || payload.hasOwnProperty('update_filesystems')
-        || payload.hasOwnProperty('add_filesystems')) {
+        || payload.hasOwnProperty('remove_filesystems')) {
 
         zcfg = zcfg + buildFilesystemZonecfg(vmobj, payload);
     }
