From cce113bc338b0ee9801fe27828231818949cb705 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Mon, 17 Oct 2016 17:08:09 +0100
Subject: [PATCH] AZ information now properly checked in networking config

---
 networking/validate.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/networking/validate.js b/networking/validate.js
index 68f2aa9..3adfbf9 100644
--- a/networking/validate.js
+++ b/networking/validate.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2016, Joyent, Inc.
  */
 
 /*
@@ -108,7 +108,7 @@ function main()
 	if (!('network' in spec['marlin']))
 		fatal('marlin section missing network name');
 
-	for (i = 0; i < spec['azs'].lenght; i++) {
+	for (i = 0; i < spec['azs'].length; i++) {
 		az = spec['azs'][i];
 		if (!(az in spec['admin']))
 			fatal('missing information for az ' + az +
-- 
2.21.0

