From 9d30d7e96bc42cd4587f015cd19eff36face3998 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 20 Oct 2016 16:32:41 -0700
Subject: [PATCH] CNS-171 log rotation setup is incomplete

---
 boot/setup.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index 731356f..3199515 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -72,6 +72,7 @@ svcadm enable cns-updater
 # add log rotation entries for services
 sdc_log_rotation_add cns-updater /var/svc/log/*cns-updater*.log 1g
 sdc_log_rotation_add cns-server /var/svc/log/*cns-server*.log 1g
+sdc_log_rotation_setup_end
 
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
-- 
2.21.0

