From c1014818af969ef33c67ca15ae3ff3b3bd204187 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 16 Aug 2018 17:52:45 -0700
Subject: [PATCH] TRITON-681 node-imgmanifest support for requirements.bootrom
 (add changelog entry)

---
 CHANGES.md | 4 ++++
 Makefile   | 2 ++
 2 files changed, 6 insertions(+)

diff --git a/CHANGES.md b/CHANGES.md
index e77a2e1..b99ac00 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,10 @@
 
 (nothing yet)
 
+## 3.1.0
+
+- TRITON-681 Add optional 'requirements.bootrom' validation.
+
 ## 3.0.0
 
 - BREAKING change, DOCKER-929 Support registry 2.0 docker images.
diff --git a/Makefile b/Makefile
index f4778ba..1ba565a 100644
--- a/Makefile
+++ b/Makefile
@@ -46,6 +46,8 @@ versioncheck:
 	@echo version is: $(shell cat package.json | json version)
 	[[ `cat package.json | json version` == `grep '^## ' CHANGES.md | head -2 | tail -1 | awk '{print $$2}'` ]]
 
+check: versioncheck
+
 .PHONY: cutarelease
 cutarelease: versioncheck
 	[[ -z `git status --short` ]]  # If this fails, the working dir is dirty.
-- 
2.21.0

