commit f484b8a6deaf8fc5a6660cf41433c22ea4901464 (refs/changes/23/1323/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-01-24T16:26:19+01:00 (2 years, 8 months ago)
    
    GH-76 EventEmitter listeners limit should be increased

diff --git a/lib/resolver.js b/lib/resolver.js
index 3ee2fb3..a5feb4c 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -65,6 +65,7 @@ function CueBallResolver(fsm, options) {
 	mod_assert.object(options, 'options');
 
 	mod_assert.object(fsm, 'fsm');
+	fsm.setMaxListeners(30);
 	this.r_fsm = fsm;
 
 	this.r_fsm.on('added', this.emit.bind(this, 'added'));
