commit 60dd2ea020e87b23555d24a6ccc0fddacd142879
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-03-30T21:58:33+01:00 (6 months ago)
    
    TRITON-1358 Logarchiver logset manta path is incorrect

diff --git a/Makefile b/Makefile
index 8f250a0..8c99595 100644
--- a/Makefile
+++ b/Makefile
@@ -79,21 +79,21 @@ hermes:
 .PHONY: release
 release: all hermes
 	@echo "Building $(RELEASE_TARBALL)"
-	@mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/$(NAME)
+	@mkdir -p $(RELSTAGEDIR)/root/opt/triton/$(NAME)
 	@mkdir -p $(RELSTAGEDIR)/site
 	@touch $(RELSTAGEDIR)/site/.do-not-delete-me
 	cp -r $(ROOT)/node_modules \
 		$(ROOT)/package.json \
 		$(ROOT)/build \
-		$(RELSTAGEDIR)/root/opt/smartdc/$(NAME)/
+		$(RELSTAGEDIR)/root/opt/triton/$(NAME)/
 	mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/boot
 	cp -R $(ROOT)/deps/sdc-scripts/* $(RELSTAGEDIR)/root/opt/smartdc/boot
 	cp -R $(ROOT)/boot/* $(RELSTAGEDIR)/root/opt/smartdc/boot
 	cp -r $(TOP)/build/hermes/opt/smartdc/hermes \
-		$(RELSTAGEDIR)/root/opt/smartdc/hermes
-	mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/hermes/etc
+		$(RELSTAGEDIR)/root/opt/triton/hermes
+	mkdir -p $(RELSTAGEDIR)/root/opt/triton/hermes/etc
 	cp $(TOP)/etc/logsets.json \
-		$(RELSTAGEDIR)/root/opt/smartdc/hermes/etc
+		$(RELSTAGEDIR)/root/opt/triton/hermes/etc
 	(cd $(RELSTAGEDIR) && $(TAR) -I pigz -cf $(ROOT)/$(RELEASE_TARBALL) root site)
 	@rm -rf $(RELSTAGEDIR)
 
diff --git a/boot/setup.sh b/boot/setup.sh
index c06fe57..3c50921 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -32,8 +32,8 @@ source /opt/smartdc/boot/lib/util.sh
 sdc_common_setup
 
 # set up services
-svccfg import /opt/smartdc/hermes/smf/hermes.xml
-svccfg import /opt/smartdc/hermes/smf/hermes-proxy.xml
+svccfg import $SVC_ROOT/smf/hermes.xml
+svccfg import $SVC_ROOT/smf/hermes-proxy.xml
 
 # add log rotation entries for services
 sdc_log_rotation_add amon-agent /var/svc/log/*amon-agent*.log 1g
