commit b093e03bee25fe8a43f0d62a574b21fb55144325
Author: Jason King <jason.king@joyent.com>
Date:   2019-04-01T17:15:05+00:00 (6 months ago)
    
    OS-7695 mdb_dem_process should handle ENOTSUP

diff --git a/usr/src/cmd/mdb/common/mdb/mdb_demangle.c b/usr/src/cmd/mdb/common/mdb/mdb_demangle.c
index b016297b31..24e0eece1d 100644
--- a/usr/src/cmd/mdb/common/mdb/mdb_demangle.c
+++ b/usr/src/cmd/mdb/common/mdb/mdb_demangle.c
@@ -27,7 +27,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.  All rights reserved.
+ * Copyright (c) 2019, Joyent, Inc.  All rights reserved.
  */
 
 #include <mdb/mdb_modapi.h>
@@ -210,7 +210,17 @@ mdb_dem_process(mdb_demangler_t *dmp, const char *name)
 
 	res = sysdemangle(name + prefixlen, dmp->dm_lang, &mdb_dem_demops);
 	if (res == NULL) {
-		if (errno != EINVAL)
+		/*
+		 * EINVAL indicates the name is not a properly mangled name
+		 * (or perhaps is truncated so it cannot be correctly
+		 * demangled) while ENOTSUP means sysdemangle could not
+		 * determine which language was used to mangle the name
+		 * (or again, truncation might also prevent correct detection).
+		 *
+		 * In either case, we do not want to emit a warning -- just
+		 * let the caller display the original name.
+		 */
+		if (errno != EINVAL && errno != ENOTSUP)
 			mdb_warn("Error while demangling");
 		return (-1);
 	}
