commit f20da5297e9f49619eb7b5e01b6c4767ddf72e81 (refs/changes/67/5167/2)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-11-30T22:30:35+00:00 (10 months ago)
    
    OS-7413 metadata daemon crashes on bhyve or instance deletion

diff --git a/src/vm/lib/metadata/agent.js b/src/vm/lib/metadata/agent.js
index eb774ef6..9800b4ec 100644
--- a/src/vm/lib/metadata/agent.js
+++ b/src/vm/lib/metadata/agent.js
@@ -353,6 +353,9 @@ MetadataAgent.prototype.purgeZoneCache = function purgeZoneCache(zonename) {
 
     if (self.zoneConnections.hasOwnProperty(zonename)) {
         if (self.zoneConnections[zonename]) {
+            // If deleting the instance before the metadata socket connected, we
+            // must cancel reconnection attempts.
+            self.stopKvmReconnTimer(zonename);
             // it's not undefined, so attempt to close it
             closeZoneConnection(self.zoneConnections[zonename]);
         }
@@ -365,8 +368,8 @@ MetadataAgent.prototype.stopKvmReconnTimer =
 function stopKvmReconnTimer(zonename) {
     var self = this;
 
-    self.log.warn({zonename: zonename},
-        'clearing connection retries for KVM VM.');
+    self.log.info({zonename: zonename},
+        'clearing connection retries for HVM instance.');
 
     if (self.zoneKvmReconnTimers.hasOwnProperty(zonename)) {
         clearTimeout(self.zoneKvmReconnTimers[zonename]);
@@ -642,6 +645,8 @@ MetadataAgent.prototype.createKVMServer = function (zopts, callback) {
 
         handler = self.makeMetadataHandler(zopts.zone, kvmstream);
 
+        assert.object(self.zoneConnections[zopts.zone],
+            'zone connection initialized and not yet reaped');
         self.zoneConnections[zopts.zone].conn = kvmstream;
 
         kvmstream.on('connect', function _onConnect() {
