From 24461cf928ac3fa9e5ad2aacfe3d1bd6e0660201 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 13 Mar 2018 12:22:20 +0000
Subject: [PATCH] OS-6638 graceful shutdown for bhyve zones

---
 usr/src/cmd/bhyve/pm.c | 35 +++++++++++++++++++++++++++++++++++
 1 file changed, 35 insertions(+)

diff --git a/usr/src/cmd/bhyve/pm.c b/usr/src/cmd/bhyve/pm.c
index 1313fbfb1d..dc4ed6b227 100644
--- a/usr/src/cmd/bhyve/pm.c
+++ b/usr/src/cmd/bhyve/pm.c
@@ -24,6 +24,9 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
+/*
+ * Copyright 2018 Joyent, Inc.
+ */
 
 #include <sys/cdefs.h>
 __FBSDID("$FreeBSD$");
@@ -52,6 +55,8 @@ static pthread_mutex_t pm_lock = PTHREAD_MUTEX_INITIALIZER;
 #ifdef	__FreeBSD__
 static struct mevent *power_button;
 static sig_t old_power_handler;
+#else
+struct vmctx *pwr_ctx;
 #endif
 
 /*
@@ -219,6 +224,22 @@ power_button_handler(int signal, enum ev_type type, void *arg)
 	}
 	pthread_mutex_unlock(&pm_lock);
 }
+
+#else
+/*
+ * Initiate graceful power off.
+ */
+/*ARGSUSED*/
+static void
+power_button_handler(int signal, siginfo_t *type, void *cp)
+{
+	pthread_mutex_lock(&pm_lock);
+	if (!(pm1_status & PM1_PWRBTN_STS)) {
+		pm1_status |= PM1_PWRBTN_STS;
+		sci_update(pwr_ctx);
+	}
+	pthread_mutex_unlock(&pm_lock);
+}
 #endif
 
 /*
@@ -326,4 +347,18 @@ sci_init(struct vmctx *ctx)
 	 */
 	pci_irq_use(SCI_INT);
 	vm_isa_set_irq_trigger(ctx, SCI_INT, LEVEL_TRIGGER);
+
+#ifndef	__FreeBSD__
+	{
+		/*
+		 * Install SIGTERM signal handler for graceful power off.
+		 */
+		struct sigaction act;
+
+		pwr_ctx = ctx;
+		act.sa_flags = 0;
+		act.sa_sigaction = power_button_handler;
+		(void) sigaction(SIGTERM, &act, NULL);
+	}
+#endif
 }
-- 
2.21.0

