commit f82d912a1cf214c647d38636e1e57d37f7734f80 (refs/changes/50/1650/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-03-10T14:56:05-08:00 (2 years, 7 months ago)
    
    joyent/node-mname#17 crash when receiving unknown query types

diff --git a/lib/server.js b/lib/server.js
index 1871b6c..3b3ea1f 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -181,7 +181,8 @@ Server.prototype.listenTcp = function listenTcp(opts, callback) {
                                 log = log.child({
                                         qId: query.id,
                                         qName: query.name(),
-                                        qType: query.type()
+                                        qType: query.type(),
+                                        qOp: query.operation()
                                 });
                                 query._log = log;
                         } catch (e) {
@@ -309,7 +310,8 @@ Server.prototype.listenUdp = function listenUdp(opts, callback) {
                         log = log.child({
                                 qId: query.id,
                                 qName: query.name(),
-                                qType: query.type()
+                                qType: query.type(),
+                                qOp: query.operation()
                         });
                         query._log = log;
                 } catch (e) {
diff --git a/package.json b/package.json
index f7386af..46f6721 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "mname",
   "description": "DNS server library for node.js",
-  "version": "1.3.4",
+  "version": "1.3.5",
   "author": "arekinath <alex@cooperi.net>",
   "contributors": [
     "Mark Cavage",
