commit 1e5ce717acaef1adb988885b8d634bfbd7e9a19d (refs/changes/43/1743/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-04-04T15:16:39+00:00 (2 years, 6 months ago)
    
    5157 remove ELFSIGN leftovers from nightly

diff --git a/pkgsrc/onbld/packlist b/pkgsrc/onbld/packlist
index 7697095d..807b1ab0 100644
--- a/pkgsrc/onbld/packlist
+++ b/pkgsrc/onbld/packlist
@@ -8,7 +8,6 @@
 ./bin/copyrightchk
 ./bin/cstyle
 ./bin/elfcmp
-./bin/elfsigncmp
 ./bin/find_elf
 ./bin/findcrypto
 ./bin/flg.flp
@@ -31,7 +30,6 @@
 ./bin/i386/ctfstrip
 ./bin/i386/cw
 ./bin/i386/elfextract
-./bin/i386/elfsign
 ./bin/i386/findunref
 ./bin/i386/install
 ./bin/i386/install.bin
