From 1e5ce717acaef1adb988885b8d634bfbd7e9a19d Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 4 Apr 2017 15:16:39 +0000
Subject: [PATCH] 5157 remove ELFSIGN leftovers from nightly

---
 pkgsrc/onbld/packlist | 2 --
 1 file changed, 2 deletions(-)

diff --git a/pkgsrc/onbld/packlist b/pkgsrc/onbld/packlist
index 7697095d..807b1ab0 100644
--- a/pkgsrc/onbld/packlist
+++ b/pkgsrc/onbld/packlist
@@ -8,7 +8,6 @@
 ./bin/copyrightchk
 ./bin/cstyle
 ./bin/elfcmp
-./bin/elfsigncmp
 ./bin/find_elf
 ./bin/findcrypto
 ./bin/flg.flp
@@ -31,7 +30,6 @@
 ./bin/i386/ctfstrip
 ./bin/i386/cw
 ./bin/i386/elfextract
-./bin/i386/elfsign
 ./bin/i386/findunref
 ./bin/i386/install
 ./bin/i386/install.bin
-- 
2.21.0

