From e7f384667df0534d001d1539267745ae5ed082cd Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 23 May 2017 21:11:23 +0000
Subject: [PATCH] This is a different second commit

---
 README.md | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/README.md b/README.md
index d51b47e..d37194e 100644
--- a/README.md
+++ b/README.md
@@ -4,3 +4,5 @@
 Can we rewind things?
 
 This is an update
+
+This is a different second update
-- 
2.21.0

