From acca68c2832608f712329636aec97fd8bb5cd829 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 6 Nov 2017 14:07:49 +0000
Subject: [PATCH] 8747 Intel Wifi Link 5300 (pciex8086,4236) works well with
 iwn(7d)

---
 overlay/generic/etc/driver_aliases | 1 +
 1 file changed, 1 insertion(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 8e29dede..0c11a3d9 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -727,6 +727,7 @@ iwn "pci8086,897"
 iwn "pci8086,8ae"
 iwn "pci8086,422b"
 iwn "pci8086,422c"
+iwn "pci8086,4236"
 iwn "pci8086,4238"
 iwn "pci8086,4239"
 ixgb "pci8086,1048"
-- 
2.21.0

