commit 8833cad29b1c8440cbabaf368b6e540e2d3219b1
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-03-07T09:57:30-08:00 (7 months ago)
    
    TRITON-897 RFD 34 Triton instance migration

diff --git a/docs/index.md b/docs/index.md
index 07a2880..f429f6d 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -1486,9 +1486,10 @@ Query the server for the VM's details.
 
 ### Inputs
 
-| Param | Type   | Description                               |
-| ----- | ------ | ----------------------------------------- |
-| jobid | String | Post information to workflow with this id |
+| Param       | Type    | Description                                   |
+| ----------- | ------- | --------------------------------------------- |
+| jobid       | String  | Post information to workflow with this id     |
+| include_dni | Boolean | Allow a VM with the do_not_inventory flag set |
 
 
 ### Responses
diff --git a/lib/apis/moray.js b/lib/apis/moray.js
index baa4c03..f0f9119 100644
--- a/lib/apis/moray.js
+++ b/lib/apis/moray.js
@@ -87,6 +87,17 @@ var BUCKETS = {
                 last_heartbeat: { type: 'string' }
             }
         }
+    },
+    'migrations': {
+        name: 'cnapi_migrations',
+        bucket: {
+            index: {
+                id:  { type: 'string' },
+                vm_uuid: { type: 'string' },
+                source_server_uuid: { type: 'string' },
+                target_server_uuid: { type: 'string' }
+            }
+        }
     }
 };
 
diff --git a/lib/endpoints/vms.js b/lib/endpoints/vms.js
index 7d2ec1c..190f1d7 100644
--- a/lib/endpoints/vms.js
+++ b/lib/endpoints/vms.js
@@ -21,6 +21,7 @@ var VError = require('verror');
 var sdcClients = require('sdc-clients');
 
 var validation = require('../validation/endpoints');
+var migrationValidation = require('../validation/migration');
 var errors = require('../errors');
 var ModelServer = require('../models/server');
 var ModelVM = require('../models/vm');
@@ -37,7 +38,6 @@ var vmValidationRules = {
 };
 
 
-
 /**
  * Query the server for a list of VMs.
  *
@@ -1207,6 +1207,67 @@ VM.dockerBuild = function handlerVmDockerBuild(req, res, next) {
 };
 
 
+/**
+ * Start migrating the instance to a new server.
+ *
+ * @name VmMigrate
+ * @endpoint POST /servers/:server\_uuid/vms/:uuid/migrate
+ * @section Virtual Machine API
+ *
+ * @response 204 None Task was sent to server
+ * @response 404 Error No such server
+ * @response 500 Error Error encountered while attempting to fulfill request
+ *
+ * @example POST /servers/<server-uuid>/vms/<uuid>/migrate
+ */
+
+VM.migrate = function handlerVmMigrate(req, res, next) {
+    if (!migrationValidation.validateReq(req, res)) {
+        next();
+        return;
+    }
+
+    req.stash.vm.performVmTask('machine_migrate', true, req, res, next);
+};
+
+
+/**
+ * Setup migration receiver on this server.
+ *
+ * @name VmMigrateReceive
+ * @endpoint POST /servers/:server\_uuid/vms/:uuid/migrateReceive
+ * @section Virtual Machine API
+ *
+ * @response 204 None Task was sent to server
+ * @response 404 Error No such server
+ * @response 500 Error Error encountered while attempting to fulfill request
+ *
+ * @example POST /servers/<server-uuid>/vms/<uuid>/migrate
+ */
+
+VM.migrateReceive = function handlerVmMigrateReceive(req, res, next) {
+    var self = this;
+
+    if (!migrationValidation.validateReq(req, res)) {
+        next();
+        return;
+    }
+
+    // No req.stash.vm for this task, so we can't use performVmTask():
+    req.stash.server.sendTaskRequest({
+        task: 'machine_migrate_receive',
+        params: req.params,
+        req_id: req.getId(),
+        req: req,
+        evcb: ModelServer.createComputeNodeAgentHandler(self, req.params.jobid),
+        cb: function _vmMigrateReceiveTaskCb(error, task) {
+            // XXX Orlando - what to do on error here?
+            res.send({ id: task.id });
+            next();
+        }
+    });
+};
+
 
 function attachTo(http, app) {
     VM.init();
@@ -1387,6 +1448,19 @@ function attachTo(http, app) {
         }),
         VM.destroy);
 
+    // Migrate VM
+    http.post(
+        { path: '/servers/:server_uuid/vms/:uuid/migrate',
+            name: 'VmMigrate' },
+        ensure({
+            connectionTimeoutSeconds: 60 * 60,
+            app: app,
+            serverRunning: true,
+            prepopulate: ['server', 'vm'],
+            connected: ['moray']
+        }),
+        VM.migrate);
+
 
     /**
      *
diff --git a/lib/models/vm.js b/lib/models/vm.js
index 62cd629..3d2441b 100644
--- a/lib/models/vm.js
+++ b/lib/models/vm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -60,6 +60,12 @@ ModelVM.init = function (app) {
 ModelVM.prototype.load = function (opts, callback) {
     var self = this;
 
+    var params = {uuid: self.uuid};
+
+    if (opts.req && opts.req.params && opts.req.params.include_dni) {
+        params.include_dni = (opts.req.params.include_dni === 'true');
+    }
+
     ModelServer.get(self.serverUuid, function (err, servermodel) {
         var request = {
             persist: false,
@@ -72,7 +78,7 @@ ModelVM.prototype.load = function (opts, callback) {
             },
             req_id: opts.req_id,
             req: opts.req,
-            params: { uuid: self.uuid }
+            params: params
         };
 
         servermodel.sendTaskRequest(request);
diff --git a/lib/validation/endpoints.js b/lib/validation/endpoints.js
index f779c79..21c0e64 100644
--- a/lib/validation/endpoints.js
+++ b/lib/validation/endpoints.js
@@ -45,13 +45,14 @@ restify_validator.Validator.prototype.check = function (str, fail_msg) {
 };
 
 function ensureParamsValid(req, res, paramRules, opts) {
+    var params = opts && opts.params || req.params;
     var rule;
     var skip = false;
 
     opts = opts || {};
 
     if (opts.strict) {
-        for (var param in req.params) {
+        for (var param in params) {
             if (!paramRules.hasOwnProperty(param)) {
                 var error = new restify.InvalidArgumentError(
                     param + ' is not an explicitly defined parameter');
@@ -90,11 +91,11 @@ function ensureParamsValid(req, res, paramRules, opts) {
                             ruleIdx);
                     }
 
-                    if (!req.params.hasOwnProperty(paramName) ||
-                        req.params[paramName] === undefined) {
+                    if (!params.hasOwnProperty(paramName) ||
+                        params[paramName] === undefined) {
 
                         skip = true;
-                        req.params[paramName] = rule[1];
+                        params[paramName] = rule[1];
                         break;
                     }
                     break;
diff --git a/lib/validation/migration.js b/lib/validation/migration.js
new file mode 100644
index 0000000..49fe319
--- /dev/null
+++ b/lib/validation/migration.js
@@ -0,0 +1,49 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+/*
+ * Overview: HTTP migration endpoint validation.
+ */
+
+var validation = require('./endpoints');
+
+
+var vmMigrationRules = {
+    jobid: ['optional', 'isStringType'],
+    migrationTask: ['isObjectType'],
+    uuid: ['isStringType'],
+    'migrationTask.action': ['isStringType'],
+    'migrationTask.record': ['isObjectType']
+};
+var migrationTaskRules = {
+};
+
+
+/**
+ * Validate that the action requested is valid for the vm migration state.
+ */
+function validateReq(req, res) {
+    if (validation.ensureParamsValid(req, res, vmMigrationRules)) {
+        return false;
+    }
+
+    // var migrationTask = req.params.migrationTask;
+
+    // if (validation.ensureParamsValid(req, res, migrationTaskRules,
+    //         {params: migrationTask})) {
+    //     return false;
+    // }
+
+    return true;
+}
+
+module.exports = {
+    validateReq: validateReq
+};
