commit df226516edab517ce2b188773ac210e02b4d9e3b (refs/changes/83/4583/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-07-26T12:09:02+00:00 (1 year, 2 months ago)
    
    OS-7090 GDT limit reset to 0xffff

diff --git a/kvm_vmx.c b/kvm_vmx.c
index e585b20..ffac0c6 100644
--- a/kvm_vmx.c
+++ b/kvm_vmx.c
@@ -813,6 +813,8 @@ __vmx_load_host_state(struct vcpu_vmx *vmx)
 		rdmsrl(MSR_KERNEL_GS_BASE, vmx->msr_guest_kernel_gs_base);
 		wrmsrl(MSR_KERNEL_GS_BASE, vmx->msr_host_kernel_gs_base);
 	}
+
+	reset_gdtr_limit();
 }
 
 static void
