From 2ab2c0c7596ac003e2169b634924784b8f6a268a Mon Sep 17 00:00:00 2001
From: Jared Morrow <jm@joyent.com>
Date: Fri, 15 Jun 2018 08:13:57 -0600
Subject: [PATCH] MANTA-3764 Create manta_fastdelete_queue bucket in moray for
 fast GC

---
 lib/moray.js | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/lib/moray.js b/lib/moray.js
index 99d8f4f..e89b7b6 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -69,6 +69,10 @@ var DELETE_LOG_SCHEMA = {
 };
 var DELETE_LOG_VERSION = 1;
 
+var FASTDELETE_QUEUE_BUCKET = process.env.MANTA_FASTDELETE_QUEUE_BUCKET ||
+    'manta_fastdelete_queue';
+var FASTDELETE_QUEUE_VERSION = 1;
+
 var DIR_COUNT_BUCKET = 'manta_directory_counts';
 var DIR_COUNT_SCHEMA = {
     entries: {
@@ -130,6 +134,15 @@ function setupMantaBuckets(log, client, cb) {
                 }
             },
             log: log
+        }, {
+            client: client,
+            bucket: FASTDELETE_QUEUE_BUCKET,
+            opts: {
+                options: {
+                    version: FASTDELETE_QUEUE_VERSION
+                }
+            },
+            log: log
         }, {
             client: client,
             bucket: DIR_COUNT_BUCKET,
-- 
2.21.0

