commit 9ab269bb1a27e6cfd718e3e43e942b8f985386b8 (refs/changes/80/5180/3)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-14T14:36:52-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 2b5e694..1635409 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS    = -t 4 -o doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/config-agent.xml.in
 
-NODE_PREBUILT_VERSION=v6.14.3
+NODE_PREBUILT_VERSION=v6.15.1
 ifeq ($(shell uname -s),SunOS)
 	# config-agent runs in zones *and* in the GZ, so we need to make sure we use
 	# a node runtime that is able to run in the GZ (those runtimes can run in
diff --git a/package.json b/package.json
index c5400ee..5430382 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "config-agent",
     "description": "SmartDataCenter Config Agent",
-    "version": "1.8.1",
+    "version": "1.8.2",
     "author": "Joyent (joyent.com)",
     "dependencies": {
         "assert-plus": "1.0.0",
