From 8ac6c120395be62ea2064a929333e533cbc7ca40 Mon Sep 17 00:00:00 2001
From: Andrey Yatskov <andrey.yatskov@silvertreesystems.com>
Date: Tue, 26 Jul 2016 20:42:59 +0300
Subject: [PATCH] PORTAL-2165 The ability to append fields to the log record

---
 lib/cloudapi.js | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/lib/cloudapi.js b/lib/cloudapi.js
index ffd3581..f41614d 100644
--- a/lib/cloudapi.js
+++ b/lib/cloudapi.js
@@ -192,20 +192,16 @@ function CloudAPI(options) {
     assert.optionalObject(options.logFields, 'options.logFields');
 
     if (options.logFields) {
-        options.log = log.child(options.logFields);
+        log = log.child(options.logFields);
     } else {
-        options.log = log.child();
+        log = log.child();
     }
 
     if (options.logLevel) {
-        options.log.level(options.logLevel);
+        log.level(options.logLevel);
     }
-
-    if (options.api_version)
-        options.version = options.api_version;
-
-    if (!options.version)
-        options.version = API_VERSION;
+    options.log = log;
+    options.version = options.api_version || API_VERSION;
 
     this.account = options.account || 'my';
     this.sign = options.sign;
-- 
2.21.0

