From 9bf7e24bf045cbe1d5caa447c49ee8babd8d2501 Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Wed, 11 Apr 2018 11:42:32 -0700
Subject: [PATCH] MANTA-3621 Mako error logs are big and have low SNR

---
 sapi_manifests/mako/template | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index c5775c7..aaafd9c 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -13,7 +13,7 @@ thread_pool mpu threads=8;
 {{/MAKO_THREAD_POOL_SIZE}}
 
 
-error_log /var/log/mako-error.log info;
+error_log /var/log/mako-error.log notice;
 
 events {
 	worker_connections  1024;
-- 
2.21.0

