commit 7b8a7d91454e22645f33c1348cf4bcaa59702102
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-06-03T14:23:20+01:00 (4 months ago)
    
    OS-7775 platform builds on base-64 need gcc4.x to bootstrap

diff --git a/Makefile.defs b/Makefile.defs
index 75b445e..ecf61a7 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -10,7 +10,7 @@
 #
 
 #
-# Copyright (c) 2019, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 BASE =		$(PWD)
@@ -47,11 +47,18 @@ MAKE_VERBOSE =	V=1
 GCC.64 =	$(GCC) -m64
 GXX.64 =	$(GXX) -m64
 
-GCC.host =	/opt/local/bin/gcc
-GXX.host =	/opt/local/bin/g++
+# The version of gcc in /opt/local/bin varies depending on the pkgsrc release
+GCC.host_prefix ?= /opt/local/bin
+GCC.host =	$(GCC.host_prefix)/gcc
+GXX.host =	$(GCC.host_prefix)/g++
 GCC.host.64 =	$(GCC.host) -m64
 GXX.host.64 =	$(GXX.host) -m64
 
+# We need gcc 4.x in order to build our copy of gcc 4.4. This is set by
+# 'configure' in smartos-live.
+GCC4.host =	$(HOST_GCC4_PREFIX)/gcc
+GXX4.host =	$(HOST_GCC4_PREFIX)/g++
+
 LEX = 		/opt/local/bin/lex
 YACC = 		/opt/local/bin/yacc
 GPATCH =	/opt/local/bin/gpatch
diff --git a/Makefile.gcc b/Makefile.gcc
index b1e698a..bb1617b 100644
--- a/Makefile.gcc
+++ b/Makefile.gcc
@@ -8,7 +8,7 @@
 # source.  A copy of the CDDL is also available via the Internet at
 # http://www.illumos.org/license/CDDL.
 #
-# Copyright (c) 2019, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 
 #
 # Shared makefile for building GCC versions. For the most part, the build is the
@@ -44,7 +44,9 @@ AUTOCONF_OPTS += \
 
 AUTOCONF_ENV += \
 	DESTDIR=$(DESTDIR) \
-	MAKE=$(MAKE)
+	MAKE=$(MAKE) \
+	GCC=$(GCC.host) \
+	GXX=$(GXX.host)
 
 OVERRIDES += \
 	$(AUTOCONF_CFLAGS) \
diff --git a/gcc4/Makefile b/gcc4/Makefile
index ffaa7e3..701c57f 100644
--- a/gcc4/Makefile
+++ b/gcc4/Makefile
@@ -20,7 +20,7 @@
 #
 # Copyright 2010 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright (c) 2018, Joyent, Inc.  All rights reserved.
+# Copyright 2019 Joyent, Inc.  All rights reserved.
 #
 
 include ../Makefile.defs
@@ -34,3 +34,6 @@ GXX = /bin/false
 CFLAGS = -nostdinc -isystem /usr/include -g -O2
 
 include ../Makefile.gcc
+
+# We specifically need GCC 4.x in order to build gcc4
+AUTOCONF_ENV += CC=$(GCC4.host) CXX=$(GXX4.host)
