commit d7085dc6c9327e6eeb31d08490455d0843103d14 (refs/changes/59/3859/2)
Author: Mike Zeller <mike@mikezeller.net>
Date:   2018-04-24T16:02:19+00:00 (1 year, 5 months ago)
    
    OS-6904 Want validMacAddress for proptable

diff --git a/src/vm/man/vmadm.1m.md b/src/vm/man/vmadm.1m.md
index fe8ff157..69ac8db6 100644
--- a/src/vm/man/vmadm.1m.md
+++ b/src/vm/man/vmadm.1m.md
@@ -1725,21 +1725,25 @@ tab-complete UUIDs rather than having to type them out for every command.
 
     routes:
 
-        This is a key-value object that maps destinations to gateways. These
-        will be set as static routes in the VM. The destinations can be either
-        IPs or subnets in CIDR form. The gateways can either be IP addresses, or
-        can be of the form "nics[0]", which specifies a link-local route on the
-        numbered nic in that VM's nics array (the first nic is 0).  As an
-        example:
+	This is a key-value object that maps destinations to gateways. These
+	will be set as static routes in the VM. The destinations can be either
+	IPs or subnets in CIDR form. The gateways can either be IP addresses,
+	or can be of the form "nics[0] or macs[aa:bb:cc:12:34:56]". Using
+	nics[] or macs[] specifies a link-local route. When using nics[] the IP
+	of the numbered nic in that VM's nics array (the first nic is 0) is
+	used. When using macs[] the IP of the nic with the matching mac address
+	in that VM's nic array is used. As an example:
 
             {
                 "10.2.2.0/24": "10.2.1.1",
-                "10.3.0.1": "nics[1]"
+                "10.3.0.1": "nics[1]",
+                "10.4.0.1": "macs[aa:bb:cc:12:34:56]"
             }
 
-        This sets two static routes: to the 10.2.2.0/24 subnet with a gateway
-        of 10.2.1.1, and a link-local route to the host 10.3.0.1 over the VM's
-        second nic.
+	This sets two static routes: to the 10.2.2.0/24 subnet with a gateway
+	of 10.2.1.1, a link-local route to the host 10.3.0.1 over the VM's
+	second nic, and a link-local route to the host 10.4.0.1 over the VM's
+	nic with the corresponding mac address.
 
         type: object
         vmtype: OS
diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 4b04d0ba..b302f0ab 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -614,6 +614,20 @@ function validateProperty(brand, prop, value, action, data, errors, log)
 
                 errors.bad_values.push(prop);
             }
+
+            // if a validator was defined, pass value through that
+            if (PAYLOAD_PROPERTIES[prop]
+                .hasOwnProperty('pr_valueValidator')) {
+
+                if (!PAYLOAD_PROPERTIES[prop]
+                    .pr_valueValidator(value)) {
+
+                    if (errors.bad_values.indexOf(prop) === -1) {
+                        errors.bad_values.push(prop);
+                    }
+                    break;
+                }
+            }
             break;
         case 'integer':
             var nval;
diff --git a/src/vm/node_modules/proptable.js b/src/vm/node_modules/proptable.js
index fd698311..f0e93167 100644
--- a/src/vm/node_modules/proptable.js
+++ b/src/vm/node_modules/proptable.js
@@ -179,6 +179,10 @@ exports.KVM_MEM_OVERHEAD = KVM_MEM_OVERHEAD;
  *      'required': (optional) an object with brand names as keys and a list
  *                  of actions this payload property *must* be specified with
  *      'type': (required) the type of this property (for validation)
+ *      'valueValidator': (optional) validate this property with a function.
+ *                  Currently only string and list types check for the presence
+ *                  of a validator. If you need to validate something else,
+ *                  then the case statement in VM.js needs to be updated.
  *
  *    Valid property types are:
  *
@@ -1286,7 +1290,8 @@ exports.properties = {
                 'kvm': ['add', 'update'],
                 'lx': ['add', 'update']
             },
-            type: 'string'
+            type: 'string',
+            valueValidator: 'utils.validMacAddress'
         },
         updatable: true,
         zonexml: 'zone.network.mac-addr'
diff --git a/src/vm/node_modules/utils.js b/src/vm/node_modules/utils.js
index 7befd9de..bbd26429 100644
--- a/src/vm/node_modules/utils.js
+++ b/src/vm/node_modules/utils.js
@@ -27,6 +27,7 @@
 var assert = require('assert');
 var fs = require('fs');
 var ipaddr = require('/usr/vm/node_modules/ip');
+var macaddr = require('/usr/vm/node_modules/macaddr');
 var net = require('net');
 var sprintf = require('/usr/node/node_modules/sprintf').sprintf;
 
@@ -240,6 +241,17 @@ function validResolver(str)
     return Boolean(net.isIP(str));
 }
 
+function validMacAddress(str)
+{
+    try {
+        macaddr.parse(str);
+    } catch (e) {
+        return false;
+    }
+
+    return /^[0-9a-fA-F:]+$/.test(str);
+}
+
 function ltrim(str, chars)
 {
     chars = chars || '\\s';
@@ -316,5 +328,6 @@ module.exports = {
     unmangleMem: unmangleMem,
     validAttrValue: validAttrValue,
     validResolver: validResolver,
+    validMacAddress: validMacAddress,
     vrrpMAC: vrrpMAC
 };
