commit 38a8d4690379484163643161aa2bb3c7aba54dd8 (refs/changes/04/4404/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-06-21T14:59:34-05:00 (1 year, 4 months ago)
    
    Fix non-shared VL2 target init

diff --git a/usr/src/uts/common/io/overlay/overlay_target.c b/usr/src/uts/common/io/overlay/overlay_target.c
index e743faf0c7..5e33d36043 100644
--- a/usr/src/uts/common/io/overlay/overlay_target.c
+++ b/usr/src/uts/common/io/overlay/overlay_target.c
@@ -1244,10 +1244,21 @@ overlay_target_lookup_respond_vl3(const overlay_targ_resp_t *otr,
 		vl2_entry = kmem_cache_alloc(overlay_entry_cache,
 		    KM_NOSLEEP | KM_NORMALPRI);
 		if (vl2_entry == NULL) {
-			/* XXX: drop */
+			/*
+			 * If we can't allocate a VL2 entry for the VL3
+			 * destination, we just give up for now and drain
+			 * any queued packets.  New packets will retry this
+			 * allocation, so if the memory pressure lets up, we
+			 * should recover.
+			 */
+			freemsgchain(entry->ote_chead);
+			entry->ote_chead = entry->ote_ctail = NULL;
 			return;
 		}
 
+		vl2_entry->ote_ott = ott;
+		vl2_entry->ote_odd = entry->ote_odd;
+
 		bcopy(&otr->otr_answer, &vl2_entry->ote_u.ote_vl2.otvl2_dest,
 		    sizeof (overlay_target_point_t));
 		bcopy(&otr->otr_mac, &vl2_entry->ote_u.ote_vl2.otvl2_mac,
