From 5a6f5402e2ab3c75215a26d1bb43f27002c6ebed Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 3 Jul 2019 04:34:30 +0000
Subject: [PATCH] TRITON-1793 Add more fabrics permission tests

---
 test/integration/fabrics.test.js | 44 +++++++++++++++++++++++++++++++-
 1 file changed, 43 insertions(+), 1 deletion(-)

diff --git a/test/integration/fabrics.test.js b/test/integration/fabrics.test.js
index bab7b8e..12b882a 100644
--- a/test/integration/fabrics.test.js
+++ b/test/integration/fabrics.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 /*
@@ -863,6 +863,48 @@ test('list networks', function (t) {
 
 });
 
+test('network owner checks', function (t) {
+    t.test('get NETS[2]: correct owner (OWNERS[0])', function (t2) {
+        mod_fabric_net.get(t2, {
+            params: {
+                vlan_id: VLANS[1].vlan_id,
+                uuid: NETS[2].uuid,
+                owner_uuid: OWNERS[0]
+            },
+            exp: NETS[2]
+        });
+    });
+
+    t.test('get NETS[2]: incorrect owner (OWNERS[1])', function (t2) {
+        mod_fabric_net.get(t2, {
+            params: {
+                vlan_id: VLANS[1].vlan_id,
+                uuid: NETS[2].uuid,
+                owner_uuid: OWNERS[1]
+            },
+            expCode: 404,
+            expErr: {
+                code: 'ResourceNotFound',
+                message: 'network not found'
+            }
+        });
+    });
+
+    t.test('delete NETS[2]: incorrect owner (OWNERS[1])', function (t2) {
+        mod_fabric_net.del(t2, {
+            params: {
+                vlan_id: VLANS[1].vlan_id,
+                uuid: NETS[2].uuid,
+                owner_uuid: OWNERS[1]
+            },
+            expCode: 404,
+            expErr: {
+                code: 'ResourceNotFound',
+                message: 'network not found'
+            }
+        });
+    });
+});
 
 /*
  * Test that we can create the same network (particularly with the same name)
-- 
2.21.0

