From 4c17ccf683dff0415feaee727eaf22e95d8507a3 Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Tue, 9 Jan 2018 15:49:37 +0000
Subject: [PATCH] MANTA-3546 lab deployments are created with too many vnodes

---
 bin/manta-deploy-dev | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bin/manta-deploy-dev b/bin/manta-deploy-dev
index adfb379..4e5615c 100755
--- a/bin/manta-deploy-dev
+++ b/bin/manta-deploy-dev
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -62,7 +62,7 @@ case "$1" in
 		;;
 	lab)
 		mdl_genconfig="lab"
-		mdl_topo_size=10000000
+		mdl_topo_size=131072
 		mdl_index_shards="1.moray 2.moray"
 		;;
 	*)
-- 
2.21.0

