commit 4c17ccf683dff0415feaee727eaf22e95d8507a3 (refs/changes/82/3182/2)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-01-09T23:49:37+00:00 (1 year, 9 months ago)
    
    MANTA-3546 lab deployments are created with too many vnodes

diff --git a/bin/manta-deploy-dev b/bin/manta-deploy-dev
index adfb379..4e5615c 100755
--- a/bin/manta-deploy-dev
+++ b/bin/manta-deploy-dev
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -62,7 +62,7 @@ case "$1" in
 		;;
 	lab)
 		mdl_genconfig="lab"
-		mdl_topo_size=10000000
+		mdl_topo_size=131072
 		mdl_index_shards="1.moray 2.moray"
 		;;
 	*)
