From f3df6c78f885f9cd8ef8798c2654d790c39e70cd Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Thu, 16 Aug 2018 14:12:07 -0400
Subject: [PATCH] joyent/node-manta#303 mfind should probably require at least
 one path argument

---
 bin/mfind | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/bin/mfind b/bin/mfind
index 820582b..6d27942 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -104,11 +104,7 @@ function parseCmdOptions(opts, parser) {
         opts.name = new RegExp(opts.name);
     }
 
-    if (!opts._args.length) {
-        opts.paths = ['/' + opts.account + '/stor'];
-    } else {
-        opts.paths = opts._args;
-    }
+    opts.paths = opts._args;
 
     return (opts);
 }
@@ -156,6 +152,7 @@ function printEntry(opts, obj) {
         }
     });
 
+    assert(options.paths.length >= 1, 'no paths specified');
     options.paths.forEach(function (p) {
         barrier.start(p);
         client.ftw(p, options, function (err, res) {
-- 
2.21.0

