From ab47b2d8121896d7763f6b52a7c9b9103465fb98 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 21 Nov 2018 16:39:35 +0000
Subject: [PATCH] MANTA-4021 wrasse's nodunit dep no longer installable with
 prebuilt node

---
 Makefile     | 4 ++--
 package.json | 3 ++-
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index b4fd460..4eee55b 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -70,7 +70,7 @@ CLEAN_FILES += node_modules
 
 .PHONY: test
 test: $(NODEUNIT)
-	$(NODEUNIT) test/*.test.js
+	$(NODE_EXEC) $(NODEUNIT) test/*.test.js
 
 .PHONY: scripts
 scripts: deps/manta-scripts/.git
diff --git a/package.json b/package.json
index 97f3018..87579a7 100644
--- a/package.json
+++ b/package.json
@@ -24,7 +24,8 @@
         "vasync": "1.4.3"
     },
     "devDependencies": {
-        "nodeunit": "0.8.6"
+        "nodeunit": "0.10.1",
+        "tap": "10.7.2"
     },
     "scripts": {
         "test": "nodeunit test/*.test.js"
-- 
2.21.0

