commit d6fe917c7c279c82501b77e31caff850434960c7 (refs/changes/45/1445/1)
Author: David Pacheco <dap@joyent.com>
Date:   2017-02-06T14:10:01-08:00 (2 years, 8 months ago)
    
    MANTA-3143 moray A record ttls could be higher

diff --git a/sapi_manifests/registrar/template.in b/sapi_manifests/registrar/template.in
index fc8e746..c6c46ff 100644
--- a/sapi_manifests/registrar/template.in
+++ b/sapi_manifests/registrar/template.in
@@ -11,7 +11,7 @@
       },
       "ttl": 60
     },
-    "ttl": 30,
+    "ttl": 60,
     "ports": [@@PORTS@@]
   },
 
