From d6fe917c7c279c82501b77e31caff850434960c7 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Mon, 6 Feb 2017 14:10:01 -0800
Subject: [PATCH] MANTA-3143 moray A record ttls could be higher

---
 sapi_manifests/registrar/template.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sapi_manifests/registrar/template.in b/sapi_manifests/registrar/template.in
index fc8e746..c6c46ff 100644
--- a/sapi_manifests/registrar/template.in
+++ b/sapi_manifests/registrar/template.in
@@ -11,7 +11,7 @@
       },
       "ttl": 60
     },
-    "ttl": 30,
+    "ttl": 60,
     "ports": [@@PORTS@@]
   },
 
-- 
2.21.0

