From fcc3000470f6622ed05a74bcdbc0d28226c39369 Mon Sep 17 00:00:00 2001
From: Rob Johnston <rob.johnston@joyent.com>
Date: Tue, 4 Jun 2019 15:24:56 -0700
Subject: [PATCH] TRITON-1651 sdc-usbkey: set-variable should enclose value in
 quotes

---
 tools/lib/usbkey.js | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/tools/lib/usbkey.js b/tools/lib/usbkey.js
index 6c46adef..c8e3ef55 100644
--- a/tools/lib/usbkey.js
+++ b/tools/lib/usbkey.js
@@ -1193,6 +1193,17 @@ set_variable_loader(mountpoint, name, value, callback)
     mod_assert.func(callback, 'callback');
 
     var search = '^\\s*' + name + '\\s*=\\s*.*$';
+
+    /*
+     * We need the value to be enclosed in quotes (") in case the value
+     * contains embedded whitespace.  Otherwise, loader will fail to parse
+     * it correctly.  So check whether the value is bare and, if so, enclose
+     * it in quotes.
+     */
+    if (value.match(/^\".*\"/) === null) {
+        value = '"' + value + '"';
+    }
+
     var replace = name + '=' + value;
 
     sedfile(mountpoint + '/boot/loader.conf', search, replace, callback);
-- 
2.21.0

