commit f79cefc01ffcbf68151164a47a2ff375e8ce8472 (refs/changes/31/3931/3)
Author: Robert Bogart <robert.bogart@joyent.com>
Date:   2018-05-11T00:10:23+00:00 (1 year, 5 months ago)
    
    MANTA-3676 Improve binder observability

diff --git a/lib/server.js b/lib/server.js
index 7c950db..389c238 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -538,6 +538,8 @@ Server.prototype.send = function send(res, opts) {
 
                 var type = 'socket';
                 var s = this._socket;
+                res.bytesSent = len;
+
                 if (opts && opts.balancer) {
                         s = opts.balancer;
                         type = 'balancer';
@@ -564,6 +566,12 @@ Server.prototype.send = function send(res, opts) {
                         break;
                 }
 
+                /*
+                 * Per RFC 7766, the message is prepended with the two-octet
+                 * length field.
+                 */
+                res.bytesSent = buf.length + 2;
+
                 log.trace({len: len},
                     'send: writing DNS message to TCP socket');
                 var lenbuf = new Buffer(2);
diff --git a/package.json b/package.json
index 290072a..7f54d8c 100644
--- a/package.json
+++ b/package.json
@@ -1,14 +1,15 @@
 {
   "name": "mname",
   "description": "DNS server library for node.js",
-  "version": "1.4.0",
+  "version": "1.5.0",
   "author": "arekinath <alex@cooperi.net>",
   "contributors": [
     "Mark Cavage",
     "Alex Wilson",
     "Lorenz Brun",
     "Trevor Orsztynowicz",
-    "Joshua M. Clulow"
+    "Joshua M. Clulow",
+    "Robert Bogart"
   ],
   "repository": {
     "type": "git",
