From 467f52f512f4a44ca8f6db903f06861a882f35a1 Mon Sep 17 00:00:00 2001
From: Robert Bogart <robert.bogart@joyent.com>
Date: Thu, 10 May 2018 02:32:16 +0000
Subject: [PATCH] MANTA-3676 Improve binder observability

---
 lib/server.js | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/lib/server.js b/lib/server.js
index 7c950db..fb9649a 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -538,6 +538,8 @@ Server.prototype.send = function send(res, opts) {
 
                 var type = 'socket';
                 var s = this._socket;
+                res._bytes_sent = len;
+
                 if (opts && opts.balancer) {
                         s = opts.balancer;
                         type = 'balancer';
@@ -564,6 +566,12 @@ Server.prototype.send = function send(res, opts) {
                         break;
                 }
 
+                /*
+                 * Per RFC 7766, the message is prepended with the two-octet
+                 * length field.
+                 */
+                res._bytes_sent = buf.length + 2;
+
                 log.trace({len: len},
                     'send: writing DNS message to TCP socket');
                 var lenbuf = new Buffer(2);
-- 
2.21.0

