From b15df144ae4afc7c588ac207480f71ed38eb5872 Mon Sep 17 00:00:00 2001
From: Robert Bogart <robert.bogart@joyent.com>
Date: Thu, 10 May 2018 02:32:16 +0000
Subject: [PATCH] MANTA-3676 Improve binder observability

---
 lib/server.js | 8 ++++++++
 package.json  | 5 +++--
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index 7c950db..389c238 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -538,6 +538,8 @@ Server.prototype.send = function send(res, opts) {
 
                 var type = 'socket';
                 var s = this._socket;
+                res.bytesSent = len;
+
                 if (opts && opts.balancer) {
                         s = opts.balancer;
                         type = 'balancer';
@@ -564,6 +566,12 @@ Server.prototype.send = function send(res, opts) {
                         break;
                 }
 
+                /*
+                 * Per RFC 7766, the message is prepended with the two-octet
+                 * length field.
+                 */
+                res.bytesSent = buf.length + 2;
+
                 log.trace({len: len},
                     'send: writing DNS message to TCP socket');
                 var lenbuf = new Buffer(2);
diff --git a/package.json b/package.json
index 290072a..54e846a 100644
--- a/package.json
+++ b/package.json
@@ -1,14 +1,15 @@
 {
   "name": "mname",
   "description": "DNS server library for node.js",
-  "version": "1.4.0",
+  "version": "1.4.1",
   "author": "arekinath <alex@cooperi.net>",
   "contributors": [
     "Mark Cavage",
     "Alex Wilson",
     "Lorenz Brun",
     "Trevor Orsztynowicz",
-    "Joshua M. Clulow"
+    "Joshua M. Clulow",
+    "Robert Bogart"
   ],
   "repository": {
     "type": "git",
-- 
2.21.0

