commit e6d97a3384f2a34805c6c1291ca658136029fb5f (refs/changes/30/2630/24)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-01-03T20:45:05+00:00 (1 year, 9 months ago)
    
    MANTA-3284 want operator-configurable throttle on muskie requests

diff --git a/Makefile b/Makefile
index 9544f59..9503048 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/README.md b/README.md
index 550c41b..ee49b66 100644
--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2017, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 # manta-muskie: The Manta WebAPI
@@ -35,23 +35,28 @@ prerequisites in your development environment:
    to Manta.  The manta-deployment zone includes a tool called add-dev-user that
    can be used to do this.  You can find it at
    /opt/smartdc/manta-deployment/tools/add-dev-user.
-   Note: You should not use the "poseidon" user for running the tests.
 2. The ssh key that you use to authenticate as this account should be
    passwordless.  It must also be stored locally, with the public key being
    called $HOME/.ssh/id\_rsa.pub.
-3. Your SDC and Manta environment variables should point at the SDC and Manta
+3. Some tests also require an operator account to test with.  By default, the
+   tests will use the "poseidon" account, but you must provide a valid private
+   key for the poseidon account at `$HOME/.ssh/id_rsa_poseidon`.
+   Optionally, you can provide a separate operator SDC account by setting
+   `MUSKIETEST_OPERATOR_USER` and the location of its private key at
+   `MUSKIETEST_OPERATOR_KEYFILE` in your environment.
+4. Your SDC and Manta environment variables should point at the SDC and Manta
    instances that you're testing with.  The SDC and Manta variables should refer
    to the same user account, and they should both refer to the ssh key stored in
    $HOME/.ssh/id\_rsa.pub mentioned above.
-4. Before running the tests, you must set the `MUSKIE_SALT`, `MUSKIE_KEY`, and
-   `MUSKIE_IV` environment variables to the same values being used for the
-   muskie instances in your existing Manta installation.  You can find these
-   values in SAPI, using:
+5. Before running the tests, you must set the `MUSKIE_SALT`, `MUSKIE_KEY`, and
+   `MUSKIE_IV` environment variables to the same values being used for the muskie
+   instances in your existing Manta installation.  You can find these values in
+   SAPI, using:
 
         sdc-sapi /services?application_uuid="$(sdc-sapi \
             /applications?name=manta | json -H application_uuid)&name=webapi" |
             json -H -a metadata
-5. You'll need to create a muskie configuration file that's appropriate for your
+6. You'll need to create a muskie configuration file that's appropriate for your
    environment.  The easiest way to do this is to copy "etc/config.coal.json" in
    this repo into a new file "config.json".  Then:
 
@@ -81,6 +86,8 @@ In summary, you should make sure these environment variables are set properly:
 | `MANTA_USER`              | refers to your non-operator user created above |
 | `MANTA_KEY_ID`            | refers to a passwordless ssh key in $HOME/.ssh/id\_rsa |
 | `MANTA_TLS_INSECURE `     | usually 1 in an environment with self-signed certificates |
+| `MUSKIETEST_OPERATOR_USER`| operator account for testing (optional, "poseidon" used by default) |
+| `MUSKIETEST_OPERATOR_KEYFILE` | path to a passwordless ssh key for `MUSKIETEST_OPERATOR_USER` (optional, `$HOME/.ssh/id_rsa_poseidon` used by default) |
 | `SDC_URL`                 | points to the SDC deployment that you're using to test |
 | `SDC_ACCOUNT`             | same value as `MANTA_USER` |
 | `SDC_KEY_ID`              | same value as `MANTA_KEY_ID` |
@@ -157,7 +164,7 @@ check first before reporting problems:
    configuration is managed by SAPI, which includes any zones deployed using
    `manta-adm`, you will need to set this variable using sapiadm:
 
-        $ sapiadm update (sdc-sapi /services?name=webapi | json -Ha uuid) metadata."MPU_ENABLE"=true
+        $ sapiadm update $(sdc-sapi /services?name=webapi | json -Ha uuid) metadata."MPU_ENABLE"=true
 
    Note that ANY value (including `false`) set for `MPU_ENABLE` will write a
    value of `true` to the muskie configuarion file. So, once you have done that,
@@ -219,3 +226,55 @@ enumerate the Muskie instances using DNS, nor is there a way to add that without
 changing the DNS name for webapi instances, which would be a flag day for
 Muppet.**  (This may explain why [muppet](https://github.com/joyent/muppet) is a
 ZooKeeper consumer rather than just a DNS client.)
+
+## Dtrace Probes
+
+Muskie has two dtrace providers. The first, `muskie`, has the following probes:
+* `client_close`: `json`. Fires if a client uploading an object or part closes
+  before data has been streamed to mako. Also fires if the client closes the
+  connection while the stream is in progress. The argument json object has the
+  following format:
+  ```
+  {
+      id: restify uuid, or x-request-id/request-id http header (string)
+      method: request http method (string)
+      headers: http headers specified by the client (object)
+      url: http request url (string)
+      bytes_sent: number of bytes streamed to mako before client close (int)
+      bytes_expected: number of bytes that should have been streamed (int)
+  }
+  ```
+* `socket_timeout`: `json`. Fires when the timeout limit is reached on a
+  connection to a client. This timeout can be configured either by setting the
+  `SOCKET_TIMEOUT` environment variable. The default is 120 seconds. The object
+  passed has the same fields to the `client_close` dtrace probe, except for the
+  `bytes_sent` and `bytes_expected`. These parameters are only present if muskie
+  is able to determine the last request sent on this socket.
+
+The second provider, `muskie-throttle`, has the following probes, which will not
+fire if the throttle is disabled:
+* `request_throttled`: `int`, `int`, `char *`, `char *` - slots occupied, queued
+  requests, url, method. Fires when a request has been throttled.
+* `request_handled`: `int`, `int`, `char *`, `char *` - slots occupied, queued
+  requests, url, method. Fires after a request has been handled.
+Internally, the muskie throttle is implemented with a vasync-queue. A "slot"
+in the above description refers to one of `concurrency` possible spaces
+allotted for concurrently scheduled request-handling callbacks. If all slots are
+occupied, incoming requests will be "queued", which indicates that they are
+waiting for slots to free up.
+* `queue_enter`: `char *` - restify request uuid. This probe fires as a request
+enters the queue.
+* `queue_leave`: `char *` - restify request uuid. This probe fires as a request
+is dequeued, before it is handled. The purpose of these probes is to make it
+easy to write d scripts that measure the latency impact the throttle has on
+individual requests.
+
+The script `bin/throttlestat.d` is implemented as an analog to `moraystat.d`
+with the `queue_enter` and `queue_leave` probes. It is a good starting point for
+gaining insight into both how actively a muskie process is being throttled and
+how much stress it is under.
+
+The throttle probes are provided in a separate provider to prevent coupling the
+throttle implementation with muskie itself. Future work may involve making the
+throttle a generic module that can be included in any service with minimal code
+modification.
diff --git a/bin/throttlestat.d b/bin/throttlestat.d
new file mode 100755
index 0000000..e6f4f3c
--- /dev/null
+++ b/bin/throttlestat.d
@@ -0,0 +1,57 @@
+#!/usr/sbin/dtrace -Cs
+
+#pragma D option quiet
+
+/*
+ * @throttled_by_method - an aggregation that counts requests throttled by HTTP
+ * method over time.
+ */
+
+int latencies[char *];
+
+
+muskie-throttle*:::queue_enter
+{
+	latencies[copyinstr(arg0)] = timestamp;
+}
+
+muskie-throttle*:::queue_leave
+/latencies[copyinstr(arg0)]/
+{
+	latencies[copyinstr(arg0)] = timestamp - latencies[copyinstr(arg0)];
+	@avg_latency[pid] = avg(latencies[copyinstr(arg0)] / 1000000);
+	latencies[copyinstr(arg0)] = 0;
+}
+
+muskie-throttle*:::request_throttled
+{
+	@throttled[pid] = count();
+}
+
+muskie-throttle*:::request_handled
+{
+	@qlen[pid] = max(arg1);
+	@qrunning[pid] = max(arg0);
+}
+
+profile:::tick-1sec
+/lines < 1/
+{
+	printf("THROTTLED-PER-SEC | AVG-LATENCY-MS | MAX-QLEN | MAX-RUNNING\n");
+	printf("------------------+----------------+----------+------------\n");
+	lines = 5;
+}
+
+
+profile:::tick-1sec
+/lines > 0/
+{
+	lines -= 1;
+	printa("%@4u              %@4u            %@4u       %@4u\n", @throttled,
+            @avg_latency, @qlen, @qrunning);
+
+	clear(@throttled);
+	clear(@avg_latency);
+	clear(@qlen);
+	clear(@qrunning);
+}
diff --git a/boot/setup.sh b/boot/setup.sh
index 98dcc6e..738ce5f 100644
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -7,7 +7,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 set -o xtrace
diff --git a/docs/README.md b/docs/README.md
index 377ef77..f76c379 100644
--- a/docs/README.md
+++ b/docs/README.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 # MANTA DOCUMENTATION MOVED
diff --git a/docs/bluejoy/footer.html.in b/docs/bluejoy/footer.html.in
index 8d27019..fabb9c0 100644
--- a/docs/bluejoy/footer.html.in
+++ b/docs/bluejoy/footer.html.in
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <script type="text/javascript" charset="utf-8">
diff --git a/docs/bluejoy/header.html.in b/docs/bluejoy/header.html.in
index 0862b2a..bda735b 100644
--- a/docs/bluejoy/header.html.in
+++ b/docs/bluejoy/header.html.in
@@ -6,7 +6,7 @@
 -->
 
 <!--
-    Copyright (c) 2017, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <html lang="en">
diff --git a/docs/bluejoy/media/css/restdown.css b/docs/bluejoy/media/css/restdown.css
index b564c40..8b1d01c 100644
--- a/docs/bluejoy/media/css/restdown.css
+++ b/docs/bluejoy/media/css/restdown.css
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /* ---- general styles */
diff --git a/docs/index.md b/docs/index.md
index c80befb..d1a7337 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -10,7 +10,7 @@ apisections: Directories, Objects, Links, Compute, Jobs, SnapLinks
 -->
 
 <!--
-    Copyright (c) 2015, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 # REST API
diff --git a/docs/internal/acuserguide.md b/docs/internal/acuserguide.md
index 6a287fe..cb83e15 100644
--- a/docs/internal/acuserguide.md
+++ b/docs/internal/acuserguide.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 # Access Control User Guide
diff --git a/docs/internal/design.md b/docs/internal/design.md
index 41a43ff..ddf8a6a 100644
--- a/docs/internal/design.md
+++ b/docs/internal/design.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2017, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 
diff --git a/docs/internal/jobs.md b/docs/internal/jobs.md
index 75f4267..4cf88b2 100644
--- a/docs/internal/jobs.md
+++ b/docs/internal/jobs.md
@@ -5,7 +5,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 # Abstract
diff --git a/etc/config.coal.json b/etc/config.coal.json
index ea3521f..b57ceb8 100644
--- a/etc/config.coal.json
+++ b/etc/config.coal.json
@@ -7,6 +7,11 @@
             "type": "udp"
         }
     },
+    "throttle": {
+        "enabled": false,
+        "concurrency": 50,
+        "queueTolerance": 25
+    },
     "maxObjectCopies": 6,
     "maxRequestAge": 600,
     "enableMPU": true,
diff --git a/lib/audit.js b/lib/audit.js
index f45fa40..43d6259 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/auth.js b/lib/auth.js
index 68fe898..4e9c5bf 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 //
diff --git a/lib/check_stream.js b/lib/check_stream.js
index 256f411..06e9b6f 100644
--- a/lib/check_stream.js
+++ b/lib/check_stream.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/lib/common.js b/lib/common.js
index fcac826..114337d 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var EventEmitter = require('events').EventEmitter;
diff --git a/lib/dir.js b/lib/dir.js
index 01f2fd0..59ebafb 100644
--- a/lib/dir.js
+++ b/lib/dir.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var util = require('util');
diff --git a/lib/errors.js b/lib/errors.js
index 8e1e22d..d730095 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert');
@@ -574,11 +574,12 @@ function ResourceNotFoundError(p) {
 util.inherits(ResourceNotFoundError, MuskieError);
 
 
-function ServiceUnavailableError(req) {
+function ServiceUnavailableError(req, cause) {
     MuskieError.call(this, {
         restCode: 'ServiceUnavailable',
         statusCode: 503,
-        message: 'manta is unable to serve this request'
+        message: 'manta is unable to serve this request',
+        cause: cause
     });
 }
 util.inherits(ServiceUnavailableError, MuskieError);
@@ -593,6 +594,15 @@ function SSLRequiredError() {
 }
 util.inherits(SSLRequiredError, MuskieError);
 
+function ThrottledError() {
+    MuskieError.call(this, {
+        restCode: 'ThrottledError',
+        statusCode: 503,
+        message: 'manta throttled this request'
+    });
+    this.name = this.constructor.name;
+}
+util.inherits(ThrottledError, MuskieError);
 
 function UploadAbandonedError() {
     MuskieError.call(this, {
@@ -638,6 +648,27 @@ function translateError(err, req) {
         return (err);
 
     var cause;
+
+    // A NoDatabasePeersError with a context object that has a 'name' and a
+    // 'message' field sent from Moray indicates a transient availability error
+    // due to request overload. We report these types of errors as 503s to
+    // indicate to the client that it can ameliorate the situation with
+    // sufficient back-off. A NoDatabasePeersError without a context object
+    // of this form still indicates an internal error (500) that should be
+    // investigated as a bug or other more serious problem in Manta.
+    cause = VError.findCauseByName(err, 'NoDatabasePeersError');
+    if (cause !== null && cause.context && cause.context.name &&
+        cause.context.message && cause.context.name === 'OverloadedError') {
+        return (new ServiceUnavailableError(req, new VError({
+            'name': cause.context.name
+        }, '%s', cause.context.message)));
+    }
+
+    cause = VError.findCauseByName(err, 'ThrottledError');
+    if (cause !== null) {
+        return (cause);
+    }
+
     cause = VError.findCauseByName(err, 'ObjectNotFoundError');
     if (cause !== null) {
         return (new restify.ResourceNotFoundError(cause,
diff --git a/lib/index.js b/lib/index.js
index c2cf6a3..5348697 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var server = require('./server');
diff --git a/lib/jobs/common.js b/lib/jobs/common.js
index 53aa0c5..f4e6bbf 100644
--- a/lib/jobs/common.js
+++ b/lib/jobs/common.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/jobs/create.js b/lib/jobs/create.js
index 42ac15b..b5c4a43 100644
--- a/lib/jobs/create.js
+++ b/lib/jobs/create.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var util = require('util');
diff --git a/lib/jobs/error.js b/lib/jobs/error.js
index 486d809..fa3ab35 100644
--- a/lib/jobs/error.js
+++ b/lib/jobs/error.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var auth = require('../auth');
diff --git a/lib/jobs/fail.js b/lib/jobs/fail.js
index 12353ab..9f45247 100644
--- a/lib/jobs/fail.js
+++ b/lib/jobs/fail.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var auth = require('../auth');
diff --git a/lib/jobs/get.js b/lib/jobs/get.js
index 2c01971..f592884 100644
--- a/lib/jobs/get.js
+++ b/lib/jobs/get.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var libmanta = require('libmanta');
diff --git a/lib/jobs/index.js b/lib/jobs/index.js
index a7d4eb3..47adca7 100644
--- a/lib/jobs/index.js
+++ b/lib/jobs/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 ///--- Helpers
diff --git a/lib/jobs/input.js b/lib/jobs/input.js
index 678efb8..15c1c7b 100644
--- a/lib/jobs/input.js
+++ b/lib/jobs/input.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var stream = require('stream');
diff --git a/lib/jobs/list.js b/lib/jobs/list.js
index cf4a991..0e0c346 100644
--- a/lib/jobs/list.js
+++ b/lib/jobs/list.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/jobs/output.js b/lib/jobs/output.js
index 7415a4f..bb1c449 100644
--- a/lib/jobs/output.js
+++ b/lib/jobs/output.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var auth = require('../auth');
diff --git a/lib/jobs/post.js b/lib/jobs/post.js
index a61f996..2a8c240 100644
--- a/lib/jobs/post.js
+++ b/lib/jobs/post.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var auth = require('../auth');
diff --git a/lib/link.js b/lib/link.js
index 9c3315d..a0bfa8c 100644
--- a/lib/link.js
+++ b/lib/link.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var url = require('url');
diff --git a/lib/medusa/connector.js b/lib/medusa/connector.js
index 5fa5f9e..8b19e55 100644
--- a/lib/medusa/connector.js
+++ b/lib/medusa/connector.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/medusa/index.js b/lib/medusa/index.js
index c4873e7..79a23c6 100644
--- a/lib/medusa/index.js
+++ b/lib/medusa/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 ///--- Helpers
diff --git a/lib/medusa/routes.js b/lib/medusa/routes.js
index e972865..4810d83 100644
--- a/lib/medusa/routes.js
+++ b/lib/medusa/routes.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 // vim: set ts=8 sts=8 sw=8 et:
diff --git a/lib/obj.js b/lib/obj.js
index 23082b5..1fc7863 100644
--- a/lib/obj.js
+++ b/lib/obj.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 //
diff --git a/lib/other.js b/lib/other.js
index 4b20013..89fd35d 100644
--- a/lib/other.js
+++ b/lib/other.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var os = require('os');
diff --git a/lib/picker.js b/lib/picker.js
index 06aafa9..4b07726 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 //
diff --git a/lib/server.js b/lib/server.js
index b04eb4b..ff6d4a7 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,13 +5,14 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
 var fs = require('fs');
 var path = require('path');
 var url = require('url');
+var verror = require('verror');
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
@@ -29,6 +30,7 @@ var obj = require('./obj');
 var other = require('./other');
 var picker = require('./picker');
 var uploads = require('./uploads');
+var throttle = require('./throttle');
 
 // injects into the global namespace
 require('./errors');
@@ -88,19 +90,16 @@ function formatJSON(req, res, body) {
  * Wrapper over restify's createServer to make testing and
  * configuration handling easier.
  *
- * The returned server object will have a '.start()' method on it, which
- * wraps up the port/host settings for you.
- *
- * @arg {object} options      - options object.
- * @arg {string} options.file - configuration file to read from.
- * @arg {object} options.log  - bunyan logger.
- * @arg {function} callback   - of the form f(err, server).
+ * @param {object} options            - options object.
+ * @param {object} options.log        - bunyan logger.
+ * @param {object} options.collector  - artedi metric collector.
  * @throws {TypeError} on bad input.
  */
-function createServer(options, clearProxy) {
+function createServer(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
     assert.object(options.collector, 'options.collector');
+    assert.object(options.throttle, 'options.throttle');
 
     options.formatters = {
         'application/json': formatJSON,
@@ -117,6 +116,32 @@ function createServer(options, clearProxy) {
     }, true);
     var server = restify.createServer(options);
 
+    /* Initialize metric collectors for use in handlers and audit logger. */
+    // A counter to track the number of HTTP requests serviced.
+    options.collector.counter({
+        name: common.METRIC_REQUEST_COUNTER,
+        help: 'count of Muskie requests completed'
+    });
+    // A histogram to track the time to first byte.
+    options.collector.histogram({
+        name: common.METRIC_LATENCY_HISTOGRAM,
+        help: 'time-to-first-byte of Muskie requests'
+    });
+    // A histogram to track the time it took to fully process each HTTP request.
+    options.collector.histogram({
+        name: common.METRIC_DURATION_HISTOGRAM,
+        help: 'total time to process Muskie requests'
+    });
+    // A pair of counters to track inbound and outbound throughput.
+    options.collector.counter({
+        name: common.METRIC_INBOUND_DATA_COUNTER,
+        help: 'count of object bytes streamed from client to storage'
+    });
+    options.collector.counter({
+        name: common.METRIC_OUTBOUND_DATA_COUNTER,
+        help: 'count of object bytes streamed from storage to client'
+    });
+
     var _timeout = parseInt((process.env.SOCKET_TIMEOUT || 120), 10) * 1000;
     server.server.setTimeout(_timeout, function onTimeout(socket) {
         var l = (((socket._httpMessage || {}).req || {}).log || log);
@@ -185,31 +210,43 @@ function createServer(options, clearProxy) {
 
     server.use(function ensureDependencies(req, res, next) {
         var ok = true;
+        var errors = [];
 
         if (!options.picker()) {
+            errors.push(new Error('picker unavailable'));
             req.log.error('picker unavailable');
             ok = false;
         } else if (!options.moray()) {
+            errors.push(new Error('index moray unavailable'));
             req.log.error('index moray unavailable');
             ok = false;
         } else if (!options.mahi()) {
+            errors.push(new Error('mahi unavailable'));
             req.log.error('mahi unavailable');
             ok = false;
         } else if (!options.marlin()) {
+            errors.push(new Error('marlin available'));
             req.log.error('marlin unavailable');
             ok = !req.isMarlinRequest();
         } else if (!options.medusa()) {
+            errors.push(new Error('medusa unavailable'));
             req.log.error('medusa unavailable');
             ok = !req.isMedusaRequest();
         }
 
         if (!ok) {
-            next(new ServiceUnavailableError());
+            next(new ServiceUnavailableError(req,
+                        new verror.MultiError(errors)));
         } else {
             next();
         }
     });
 
+    if (options.throttle.enabled) {
+        options.throttle.log = options.log;
+        var throttleHandle = throttle.createThrottle(options.throttle);
+        server.use(throttle.throttleHandler(throttleHandle));
+    }
     server.use(auth.authenticationHandler({
         log: log,
         mahi: options.mahi,
@@ -217,6 +254,8 @@ function createServer(options, clearProxy) {
     }));
 
     server.use(auth.gatherContext);
+
+    // Add various fields to the 'req' object before the handlers get called.
     server.use(common.setupHandler(options));
 
     // Compute jobs
@@ -444,13 +483,15 @@ function createServer(options, clearProxy) {
     return (server);
 }
 
+
 function forbiddenHandler(req, res, next) {
-        req.log.info('Method ' + req.method + ' disallowed for ' + req.url);
-        res.send(405);
-        next(false);
+    req.log.debug('Method ' + req.method + ' disallowed for ' + req.url);
+    res.send(405);
+    next(false);
 }
 
 
+
 /*
  * This adds the routes for the majority of multipart upload API endpoints,
  * including:
@@ -541,24 +582,6 @@ function addMultipartUploadRoutes(server) {
         contentType: '*/*'
     }, uploads.redirectHandler());
 
-    /*
-     * Path: /:account/uploads/[0-f]/:id
-     *
-     * Allowed: GET (list-parts), HEAD
-     * Disallowed: PUT, POST, DELETE
-     */
-    server.put({
-        path: '/:account/uploads/[0-f]+/:id'
-    }, forbiddenHandler);
-
-    server.post({
-        path: '/:account/uploads/[0-f]+/:id'
-    }, forbiddenHandler);
-
-    server.del({
-        path: '/:account/uploads/[0-f]+/:id'
-    }, forbiddenHandler);
-
     /*
      * Path: /:account/uploads/[0-f]/:id/state
      *
@@ -648,12 +671,36 @@ function addMultipartUploadRoutes(server) {
  * API -- that is, the uploading of parts.
  */
 function addMultipartUploadDataPlaneRoutes(server) {
+    /*
+     * Path: /:account/uploads/[0-f]/:id
+     *
+     * Allowed: GET (list-parts), HEAD
+     * Disallowed: PUT, POST, DELETE (except with override query param)
+     */
+    server.del({
+        path: '/:account/uploads/[0-f]+/:id',
+        name: 'DelUploadDir'
+    }, uploads.delUploadDirHandler());
+
+    server.put({
+        path: '/:account/uploads/[0-f]+/:id'
+    }, forbiddenHandler);
+
+    server.post({
+        path: '/:account/uploads/[0-f]+/:id'
+    }, forbiddenHandler);
+
     /*
      * Path: /:account/uploads/[0-f]/:id/:partNum
      *
      * Allowed: PUT (upload-part), HEAD
-     * Disallowed: GET, POST, DELETE
+     * Disallowed: GET, POST, DELETE (except with override query param)
      */
+    server.del({
+        path: '/:account/uploads/[0-f]+/:id/:partNum',
+        name: 'DelPart'
+    }, uploads.delPartHandler());
+
     server.put({
         path: '/:account/uploads/[0-f]+/:id/:partNum',
         name: 'UploadPart',
@@ -667,12 +714,9 @@ function addMultipartUploadDataPlaneRoutes(server) {
     server.post({
         path: '/:account/uploads/[0-f]+/:id/:partNum'
     }, forbiddenHandler);
-
-    server.del({
-        path: '/:account/uploads/[0-f]+/:id/:partNum'
-    }, forbiddenHandler);
 }
 
+
 ///--- Exports
 
 module.exports = {
diff --git a/lib/shark_client.js b/lib/shark_client.js
index a4dc1d8..ba73c09 100644
--- a/lib/shark_client.js
+++ b/lib/shark_client.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var EventEmitter = require('events').EventEmitter;
diff --git a/lib/throttle.js b/lib/throttle.js
new file mode 100644
index 0000000..82cd5b4
--- /dev/null
+++ b/lib/throttle.js
@@ -0,0 +1,205 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var bunyan = require('bunyan');
+var vasync = require('vasync');
+var uuid = require('node-uuid');
+var util = require('util');
+var mod_url = require('url');
+var fs = require('fs');
+var dtrace = require('dtrace-provider');
+var restify = require('restify');
+var jsprim = require('jsprim');
+var VError = require('verror');
+
+require('./errors');
+
+/*
+ * High Level Operation
+ *
+ * This module exports a 'wait' method, which serves as the entry point
+ * to all throttling operations the module provides. Users of the module
+ * simply call the 'wait' function in a request-processing code path,
+ * passing in a callback representing the work required to handle the
+ * request. Currently, 'req' and 'res' arguments are also supplied
+ * because the throttle is plugged in as restify middleware in muskie.
+ * Future iterations of this throttle will aim to be generic across all
+ * communication protocols, requiring only an argumentless work function
+ * as input.
+ *
+ * The operation of wait() can be summarized as follows:
+ *  - If the number of queued requests exceeds 'queueTolerance' and
+ *    the throttle is enabled, throttle the incoming request and
+ *    return.
+ *  - If the throttle is enabled, put the incoming request-processing
+ *    function on the request queue. This will result in either the
+ *    request callback being scheduled immediately or, if all slots
+ *    are occupied, being put on the request queue.
+ *  - If the throttle is disabled, simply call the request-processing
+ *    work function.
+ *
+ * Overview of Tunables and Tradeoffs
+ *
+ * The following parameters are implemented as SAPI tunables. Muskie
+ * restart is required for any tunable modification to take effect.
+ *
+ * queueTolerance - the number of requests the throttle can queue before
+ * it starts sending indications that requests have been throttled to
+ * clients. There is one 'global' queue to which incoming request are
+ * added if all 'concurrency' slots in a vasync queue are occupied.
+ *
+ * Higher 'queueTolerance' values make it less likely that the throttle
+ * will reject incoming requests and will increase muskies memory footprint
+ * during period of high load. Lower 'queueTolerance' values make it more
+ * likely that muskie will reject incoming requests.
+ *
+ * concurrency - the number of slots the request queue has for scheduling
+ * request-handling worker callbacks concurrently. When all the slots are
+ * filled, the throttle will queue up to 'queueTolerance' callbacks before
+ * throttling requests.
+ *
+ * Higher 'concurrency' values allow Manta to handle more requests
+ * concurrently and also makes it less likely that requests will spend time in
+ * the queue and/or be throttled. Lower 'concurrency' values restrict
+ * the number of requests manta can handle at once and make it more likely
+ * that requests will spend time in the queue and/or be throttled.
+ *
+ * To prevent dropping incoming traffic needlessly, it is recommended that
+ * lower 'concurrency' values be accompanied by proportionally higher
+ * 'queueTolerance' values. Higher 'concurrency' values will result in
+ * more requests be handled concurrently, and thus fewer requests being
+ * queued (assuming the same load as in the previous scenario). This is
+ * effectively a CPU/memory trade-off.
+ *
+ * enabled - a boolean value describing whether the throttle should queue
+ * and throttle requests as designed.
+ *
+ * If enabled is false, the throttle will invoke the request-processing
+ * callback immediately. The operation of muskie with 'enabled' set to
+ * false is identical to the operation of muskie prior to the change
+ * introducing this throttle.
+ *
+ * If enabled is true, the throttle will queue and throttle requests as
+ * necessary.
+ */
+
+// Used for nanosecond to second conversion
+const NANOSEC_PER_SEC = Math.pow(10, 9);
+
+/*
+ * The throttle object maintains all the state used by the throttle. This state
+ * consists of the tunables described above in addition to dtrace probes that
+ * help to describe the runtime operation of the throttle. Structuring the
+ * throttle as an object allows us to potentially instantiate multiple
+ * throttles for different communication abstractions in the same service.
+ */
+function Throttle(options) {
+    assert.number(options.concurrency, 'options.concurrency');
+    assert.ok(options.concurrency > 0, 'concurrency must be positive');
+    assert.bool(options.enabled, 'options.enabled');
+    assert.ok(options.log, 'options.log');
+    assert.number(options.queueTolerance, 'options.queueTolerance');
+    assert.ok(options.queueTolerance > 0, 'queueTolerance must be positive');
+
+    this.log = options.log.child({ component: 'throttle'}, true);
+
+    this.dtp = dtrace.createDTraceProvider('muskie-throttle');
+    this.throttle_probes = {
+        // number of occupied slots, number of queued requests,
+        // request rate, url, method
+        request_throttled: this.dtp.addProbe('request_throttled', 'int', 'int',
+                'char *', 'char *'),
+        // number of occupied slots, number of queued requests
+        request_handled: this.dtp.addProbe('request_handled', 'int', 'int',
+                'char *', 'char *'),
+        // request id
+        queue_enter: this.dtp.addProbe('queue_enter', 'char *'),
+        // request id
+        queue_leave: this.dtp.addProbe('queue_leave', 'char *')
+
+    };
+    this.dtp.enable();
+
+    this.enabled = options.enabled;
+    this.concurrency = options.concurrency;
+    this.queueTolerance = options.queueTolerance;
+
+    this.requestQueue = vasync.queue(function (task, callback) {
+        task(callback);
+    }, this.concurrency);
+}
+
+Throttle.prototype.wait = function wait(req, res, next) {
+    var self = this;
+
+    if (self.requestQueue.length() >= self.queueTolerance) {
+        self.throttle_probes.request_throttled.fire(function () {
+            return ([self.requestQueue.npending, self.requestQueue.length(),
+                req.url, req.method]);
+        });
+        /*
+         * Wrap the ThrottledError in a VError so that the relevant fields
+         * appear in the audit entry of the throttled request. We translate the
+         * error to its cause before sending the response, as we don't want to
+         * expose the queue length or the queue tolerance of the throttle to end
+         * users.
+         */
+        var state = {
+            queuedRequests: self.requestQueue.npending,
+            inFlightRequests: self.requestQueue.length()
+        };
+        var cfg = {
+            queueTolerance: self.queueTolerance,
+            concurrency: self.concurrency
+        };
+        next(new VError(new ThrottledError(), 'muskie throttled this ' +
+                    'request. observed: %j, configured with: %j', state,
+                    cfg));
+        return;
+    }
+
+    var req_id = req.getId();
+
+    self.throttle_probes.queue_enter.fire(function () {
+        return ([req_id]);
+    });
+
+    self.requestQueue.push(function (cb) {
+        self.throttle_probes.queue_leave.fire(function () {
+            return ([req_id]);
+        });
+        next();
+        cb();
+    });
+
+    self.throttle_probes.request_handled.fire(function () {
+        return ([self.requestQueue.npending, self.requestQueue.length(),
+            req.url, req.method]);
+    });
+};
+
+
+///--- Exports
+
+module.exports = {
+
+    createThrottle: function createThrottle(options) {
+        return (new Throttle(options));
+    },
+
+    throttleHandler: function (throttle) {
+        function throttleRequest(req, res, next) {
+            throttle.wait(req, res, next);
+        }
+        return (throttleRequest);
+    }
+
+};
diff --git a/lib/uploads/abort.js b/lib/uploads/abort.js
index a41de3d..ea59b71 100644
--- a/lib/uploads/abort.js
+++ b/lib/uploads/abort.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index 7c01bdb..e70b512 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/uploads/common.js b/lib/uploads/common.js
index f432788..d28ed55 100644
--- a/lib/uploads/common.js
+++ b/lib/uploads/common.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var util = require('util');
diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 1fb859d..9c3c7a6 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/lib/uploads/get.js b/lib/uploads/get.js
index 89d43a2..5adc017 100644
--- a/lib/uploads/get.js
+++ b/lib/uploads/get.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var auth = require('../auth');
diff --git a/lib/uploads/index.js b/lib/uploads/index.js
index da67ab5..573e4bc 100644
--- a/lib/uploads/index.js
+++ b/lib/uploads/index.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 ///--- Helpers
diff --git a/lib/uploads/redirect.js b/lib/uploads/redirect.js
index f4c3606..90c45c7 100644
--- a/lib/uploads/redirect.js
+++ b/lib/uploads/redirect.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var uploadsCommon = require('./common');
diff --git a/lib/uploads/upload.js b/lib/uploads/upload.js
index 32e5a28..11a88b9 100644
--- a/lib/uploads/upload.js
+++ b/lib/uploads/upload.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var jsprim = require('jsprim');
diff --git a/lib/utils.js b/lib/utils.js
index 434ac7e..df98694 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 //
@@ -79,4 +79,4 @@ module.exports = {
     interleave: interleave,
     shallowCopy: shallowCopy,
     shuffle: shuffle
-};
\ No newline at end of file
+};
diff --git a/main.js b/main.js
index 8f1b83e..68d699f 100644
--- a/main.js
+++ b/main.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var fs = require('fs');
diff --git a/sapi_manifests/muskie/template b/sapi_manifests/muskie/template
index 998e3ed..04b156e 100644
--- a/sapi_manifests/muskie/template
+++ b/sapi_manifests/muskie/template
@@ -7,6 +7,26 @@
       "type": "udp"
     }
   },
+  "throttle": {
+    {{#MUSKIE_THROTTLE_ENABLED}}
+    "enabled": {{MUSKIE_THROTTLE_ENABLED}},
+    {{/MUSKIE_THROTTLE_ENABLED}}
+    {{^MUSKIE_THROTTLE_ENABLED}}
+    "enabled": false,
+    {{/MUSKIE_THROTTLE_ENABLED}}
+    {{#MUSKIE_THROTTLE_CONCURRENCY}}
+    "concurrency": {{MUSKIE_THROTTLE_CONCURRENCY}},
+    {{/MUSKIE_THROTTLE_CONCURRENCY}}
+    {{^MUSKIE_THROTTLE_CONCURRENCY}}
+    "concurrency": 50,
+    {{/MUSKIE_THROTTLE_CONCURRENCY}}
+    {{#MUSKIE_THROTTLE_QUEUE_TOLERANCE}}
+    "queueTolerance": {{MUSKIE_THROTTLE_QUEUE_TOLERANCE}}
+    {{/MUSKIE_THROTTLE_QUEUE_TOLERANCE}}
+    {{^MUSKIE_THROTTLE_QUEUE_TOLERANCE}}
+    "queueTolerance": 25
+    {{/MUSKIE_THROTTLE_QUEUE_TOLERANCE}}
+  },
   "maxObjectCopies": 6,
   "maxRequestAge": 600,
   "numWorkers": 4,
@@ -182,7 +202,8 @@
   "storage": {
     "lag": 60000,
     "multiDC": {{MUSKIE_MULTI_DC}}{{#MUSKIE_DEFAULT_MAX_STREAMING_SIZE_MB}},
-    "defaultMaxStreamingSizeMB": {{MUSKIE_DEFAULT_MAX_STREAMING_SIZE_MB}}{{/MUSKIE_DEFAULT_MAX_STREAMING_SIZE_MB}},
+    "defaultMaxStreamingSizeMB": {{MUSKIE_DEFAULT_MAX_STREAMING_SIZE_MB}}{{/MUSKIE_DEFAULT_MAX_STREAMING_SIZE_MB}}{{#MUSKIE_MAX_UTILIZATION_PCT}},
+    "maxUtilizationPct": {{MUSKIE_MAX_UTILIZATION_PCT}}{{/MUSKIE_MAX_UTILIZATION_PCT}},
     "moray": {
         "srvDomain": "{{STORAGE_MORAY_SHARD}}",
         "cueballOptions": {
diff --git a/test/ac.test.js b/test/ac.test.js
index c44a8a9..f1e36c7 100644
--- a/test/ac.test.js
+++ b/test/ac.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var _helper = __dirname + '/helper.js';
diff --git a/test/acsetup.js b/test/acsetup.js
index 8e9e91b..679fd61 100644
--- a/test/acsetup.js
+++ b/test/acsetup.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var _helper = __dirname + '/helper.js';
diff --git a/test/auth.test.js b/test/auth.test.js
index 5f7b74f..d1a32ec 100644
--- a/test/auth.test.js
+++ b/test/auth.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var util = require('util');
diff --git a/test/cors.test.js b/test/cors.test.js
index 027db0a..aab7b65 100644
--- a/test/cors.test.js
+++ b/test/cors.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/dir.test.js b/test/dir.test.js
index 29b1e21..1f90c5e 100644
--- a/test/dir.test.js
+++ b/test/dir.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var MemoryStream = require('stream').PassThrough;
diff --git a/test/helper.js b/test/helper.js
index 7f67563..51a2db2 100644
--- a/test/helper.js
+++ b/test/helper.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var domain = require('domain');
diff --git a/test/integ.test.js b/test/integ.test.js
index 9f122b4..01d1e29 100644
--- a/test/integ.test.js
+++ b/test/integ.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/jobs.test.js b/test/jobs.test.js
index 3a97974..20308f2 100644
--- a/test/jobs.test.js
+++ b/test/jobs.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/link.test.js b/test/link.test.js
index 6a70780..38f688e 100644
--- a/test/link.test.js
+++ b/test/link.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/monitoring.test.js b/test/monitoring.test.js
index fa36ec3..2d3dfb7 100644
--- a/test/monitoring.test.js
+++ b/test/monitoring.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 if (require.cache[__dirname + '/helper.js'])
diff --git a/test/mpu/abort.test.js b/test/mpu/abort.test.js
index 0f1a301..b8bafb9 100644
--- a/test/mpu/abort.test.js
+++ b/test/mpu/abort.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var uuid = require('node-uuid');
diff --git a/test/mpu/auth.test.js b/test/mpu/auth.test.js
index c8afd55..eea4052 100644
--- a/test/mpu/auth.test.js
+++ b/test/mpu/auth.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/mpu/commit.test.js b/test/mpu/commit.test.js
index 9490fef..cffb6be 100644
--- a/test/mpu/commit.test.js
+++ b/test/mpu/commit.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var MemoryStream = require('stream').PassThrough;
diff --git a/test/mpu/create.test.js b/test/mpu/create.test.js
index a1d752b..9832e6e 100644
--- a/test/mpu/create.test.js
+++ b/test/mpu/create.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var path = require('path');
diff --git a/test/mpu/get.test.js b/test/mpu/get.test.js
index 0bc48b8..37a28cf 100644
--- a/test/mpu/get.test.js
+++ b/test/mpu/get.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var path = require('path');
diff --git a/test/mpu/helper.js b/test/mpu/helper.js
index cada050..fd823b6 100644
--- a/test/mpu/helper.js
+++ b/test/mpu/helper.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
diff --git a/test/mpu/redirect.test.js b/test/mpu/redirect.test.js
index 4462e53..89f08c3 100644
--- a/test/mpu/redirect.test.js
+++ b/test/mpu/redirect.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var manta = require('manta');
diff --git a/test/mpu/upload.test.js b/test/mpu/upload.test.js
index ce940f6..d8608ac 100644
--- a/test/mpu/upload.test.js
+++ b/test/mpu/upload.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/test/obj.test.js b/test/obj.test.js
index c4863a1..a4cb3ad 100644
--- a/test/obj.test.js
+++ b/test/obj.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var crypto = require('crypto');
diff --git a/tools/jsl.node.conf b/tools/jsl.node.conf
index 76a1fc1..293bc4c 100644
--- a/tools/jsl.node.conf
+++ b/tools/jsl.node.conf
@@ -185,6 +185,7 @@
 +define ServiceUnavailableError
 +define SharksExhaustedError
 +define SSLRequiredError
++define ThrottledError
 +define UploadAbandonedError
 +define UploadTimeoutError
 +define UserDoesNotExistError
diff --git a/tools/mk/Makefile.defs b/tools/mk/Makefile.defs
index 8c6be90..39a4714 100644
--- a/tools/mk/Makefile.defs
+++ b/tools/mk/Makefile.defs
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.deps b/tools/mk/Makefile.deps
index 1cffbe7..3e82e0c 100644
--- a/tools/mk/Makefile.deps
+++ b/tools/mk/Makefile.deps
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node.defs b/tools/mk/Makefile.node.defs
index 487824d..cff9304 100644
--- a/tools/mk/Makefile.node.defs
+++ b/tools/mk/Makefile.node.defs
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node.targ b/tools/mk/Makefile.node.targ
index bf53f78..580a6c3 100644
--- a/tools/mk/Makefile.node.targ
+++ b/tools/mk/Makefile.node.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node_modules.defs b/tools/mk/Makefile.node_modules.defs
index ec8cc8e..ad9fb6d 100644
--- a/tools/mk/Makefile.node_modules.defs
+++ b/tools/mk/Makefile.node_modules.defs
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node_modules.targ b/tools/mk/Makefile.node_modules.targ
index 0156bce..762f3bf 100644
--- a/tools/mk/Makefile.node_modules.targ
+++ b/tools/mk/Makefile.node_modules.targ
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node_prebuilt.defs b/tools/mk/Makefile.node_prebuilt.defs
index 2129742..9edc396 100644
--- a/tools/mk/Makefile.node_prebuilt.defs
+++ b/tools/mk/Makefile.node_prebuilt.defs
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node_prebuilt.targ b/tools/mk/Makefile.node_prebuilt.targ
index 6877333..0e720a8 100644
--- a/tools/mk/Makefile.node_prebuilt.targ
+++ b/tools/mk/Makefile.node_prebuilt.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.smf.defs b/tools/mk/Makefile.smf.defs
index b988bbe..e70569d 100644
--- a/tools/mk/Makefile.smf.defs
+++ b/tools/mk/Makefile.smf.defs
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.smf.targ b/tools/mk/Makefile.smf.targ
index f78de96..15ec11b 100644
--- a/tools/mk/Makefile.smf.targ
+++ b/tools/mk/Makefile.smf.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index 54edf0e..099aeb9 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
