From 1ce22085763f61c0c1f8c6df66f5b5b52bcc6fa3 Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Mon, 28 Oct 2019 22:36:44 +0000
Subject: [PATCH] OS-8036 'gmake clean' doesn't cleanup ucodecheck

---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 66f08cfa..3016ecde 100644
--- a/Makefile
+++ b/Makefile
@@ -450,6 +450,7 @@ clean:
 	(cd tools/mancheck && gmake clean)
 	(cd tools/mancf && gmake clean)
 	(cd tools/tzcheck && gmake clean)
+	(cd tools/ucodecheck && gmake clean)
 	(cd man/sdc && gmake clean)
 	rm -f 0-*-stamp 1-*-stamp
 
-- 
2.17.2 (Apple Git-113)

