commit 154797e9603e5559581f67238db7a3cc6d23e5be (refs/changes/66/3966/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-05-17T19:16:55+00:00 (1 year, 5 months ago)
    
    TRITON-392 AdminUI boot script sed confusion

diff --git a/boot/configure.sh b/boot/configure.sh
index 3db8e218..43c4c7c2 100755
--- a/boot/configure.sh
+++ b/boot/configure.sh
@@ -1,5 +1,4 @@
 #!/bin/bash
-# -*- mode: shell-script; fill-column: 80; -*-
 #
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
@@ -7,18 +6,10 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
-export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
 set -o xtrace
 
-echo "Updating adminui service manifest"
-$(/usr/bin/sed -i"" \
-  -e "s/@@PREFIX@@/\/opt\/smartdc\/adminui/g" \
-  /opt/smartdc/adminui/smf/manifests/adminui.xml)
-
-echo "Importing adminui service manifest"
-svccfg import /opt/smartdc/adminui/smf/manifests/adminui.xml
-
 exit 0
diff --git a/boot/setup.sh b/boot/setup.sh
index 65ece44f..2379f506 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -1,5 +1,4 @@
 #!/usr/bin/bash
-# -*- mode: shell-script; fill-column: 80; -*-
 #
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
@@ -7,44 +6,76 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
-export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
+export PS4
 set -o xtrace
 
-PATH=/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
+PATH='/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin'
 
-role=adminui
+role='adminui'
+prefix="/opt/smartdc/$role"
 
-# Include common utility functions (then run the boilerplate)
+#
+# Include common utility functions, then run the boilerplate.
+#
 source /opt/smartdc/boot/lib/util.sh
-CONFIG_AGENT_LOCAL_MANIFESTS_DIRS=/opt/smartdc/$role
+CONFIG_AGENT_LOCAL_MANIFESTS_DIRS="$prefix"
 sdc_common_setup
 
-# Identify this as a smartdc zone
-mkdir -p /var/smartdc/adminui
-mkdir -p /opt/smartdc/adminui
+#
+# Identify this as a smartdc zone:
+#
+mkdir -p "/var/smartdc/$role"
+mkdir -p "$prefix"
 
-# We need to generate our own self signed certificate for Nginx.
-mkdir -p /opt/smartdc/adminui/etc/ssl
+#
+# We need to generate our own self signed certificate for nginx.
+#
+mkdir -p "$prefix/etc/ssl"
 echo "Generating SSL Certificate"
 /opt/local/bin/openssl req -x509 -nodes -subj '/CN=*' \
     -newkey rsa:4096 -sha256 -days 365 \
-    -keyout /opt/smartdc/adminui/etc/ssl/default.pem \
-    -out /opt/smartdc/adminui/etc/ssl/default.pem
+    -keyout "$prefix/etc/ssl/default.pem" \
+    -out "$prefix/etc/ssl/default.pem"
 
-# Extend the PATH for convenience.
-echo -e "\nexport PATH=\$PATH:/opt/smartdc/adminui/node_modules/.bin:/opt/smartdc/adminui/build/node/bin" >> /root/.bashrc
+#
+# Amend the PATH in the stock root user profile with the location of tools for
+# this zone.
+#
+extra_path=(
+    '$PATH'
+    "$prefix/node_modules/.bin"
+    "$prefix/build/node/bin"
+)
+printf '\nexport PATH="%s"\n' "$(IFS=':'; printf '%s' "${extra_path[*]}")" \
+  >> /root/.bashrc
 
+#
 # Log rotation.
+#
 sdc_log_rotation_add amon-agent /var/svc/log/*amon-agent*.log 1g
 sdc_log_rotation_add config-agent /var/svc/log/*config-agent*.log 1g
 sdc_log_rotation_add registrar /var/svc/log/*registrar*.log 1g
-sdc_log_rotation_add $role /var/svc/log/*$role*.log 1g
+sdc_log_rotation_add "$role" /var/svc/log/*$role*.log 1g
 sdc_log_rotation_setup_end
 
-# All done, run boilerplate end-of-setup
+#
+# All done, run boilerplate end-of-setup:
+#
 sdc_setup_complete
 
+#
+# Process and import the SMF manifest:
+#
+manifest="$prefix/smf/manifests/$role.xml"
+
+echo "Updating AdminUI service manifest"
+/usr/bin/sed -i '' -e "s/@@PREFIX@@/$prefix/g" "$manifest"
+
+echo "Importing AdminUI service manifest"
+/usr/sbin/svccfg import "$manifest"
+
 exit 0
