From 3ad5646c19bd0fc20e54edb91bc84cc88b223db8 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Fri, 10 Aug 2018 20:13:30 +0000
Subject: [PATCH] joyent/manta-garbage-collector#48 repo cleanup/improvements

---
 lib/mako_instruction_uploader.js | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/mako_instruction_uploader.js b/lib/mako_instruction_uploader.js
index f18ea3d..59a8cbb 100644
--- a/lib/mako_instruction_uploader.js
+++ b/lib/mako_instruction_uploader.js
@@ -142,12 +142,12 @@ _listen_for_instructions()
 
 		var keys = [];
 		var lines = [];
-		var marked_space = [];
+		var marked_bytes = 0;
 
 		instruction.lines.forEach(function (elem) {
 			keys.push(elem.key);
 			lines.push(elem.line);
-			marked_space += elem.size;
+			marked_bytes += elem.size;
 		});
 
 		self.mu_log.debug({
@@ -195,7 +195,7 @@ _listen_for_instructions()
 				});
 				self._get_collector().getCollector(
 					'gc_bytes_marked_for_delete').observe(
-					marked_space, {
+					marked_bytes, {
 					manta_storage_id: storage_id
 				});
 			}
-- 
2.21.0

