commit 3ad5646c19bd0fc20e54edb91bc84cc88b223db8 (refs/changes/62/4662/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-08-10T21:07:06+00:00 (1 year, 2 months ago)
    
    joyent/manta-garbage-collector#48 repo cleanup/improvements

diff --git a/lib/mako_instruction_uploader.js b/lib/mako_instruction_uploader.js
index f18ea3d..59a8cbb 100644
--- a/lib/mako_instruction_uploader.js
+++ b/lib/mako_instruction_uploader.js
@@ -142,12 +142,12 @@ _listen_for_instructions()
 
 		var keys = [];
 		var lines = [];
-		var marked_space = [];
+		var marked_bytes = 0;
 
 		instruction.lines.forEach(function (elem) {
 			keys.push(elem.key);
 			lines.push(elem.line);
-			marked_space += elem.size;
+			marked_bytes += elem.size;
 		});
 
 		self.mu_log.debug({
@@ -195,7 +195,7 @@ _listen_for_instructions()
 				});
 				self._get_collector().getCollector(
 					'gc_bytes_marked_for_delete').observe(
-					marked_space, {
+					marked_bytes, {
 					manta_storage_id: storage_id
 				});
 			}
