commit d4c849dce7ae749310fa771ec942c7364f5764bd (refs/changes/51/2551/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-09-25T19:36:51+00:00 (2 years ago)
    
    MORAY-432 test/client.test.js sporadically fails

diff --git a/test/client.test.js b/test/client.test.js
index 37dae22..9e5dd09 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -91,7 +91,7 @@ test('close() with outstanding callback request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
@@ -118,7 +118,7 @@ test('close() with outstanding event-emitter request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
