commit 135cb84cf355b91397fd60689084be97d67ecf79 (refs/changes/51/2551/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-09-12T23:02:39+00:00 (2 years, 1 month ago)
    
    MORAY-432 test/client.test.js sporadically fails

diff --git a/test/client.test.js b/test/client.test.js
index 37dae22..9e5dd09 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -91,7 +91,7 @@ test('close() with outstanding callback request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
@@ -118,7 +118,7 @@ test('close() with outstanding event-emitter request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
