From 135cb84cf355b91397fd60689084be97d67ecf79 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 12 Sep 2017 19:05:50 +0000
Subject: [PATCH] MORAY-432 test/client.test.js sporadically fails

---
 test/client.test.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/test/client.test.js b/test/client.test.js
index 37dae22..9e5dd09 100644
--- a/test/client.test.js
+++ b/test/client.test.js
@@ -91,7 +91,7 @@ test('close() with outstanding callback request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
@@ -118,7 +118,7 @@ test('close() with outstanding event-emitter request', function (t) {
             t.end();
         });
 
-        setImmediate(function () { c2.close(); });
+        process.nextTick(function () { c2.close(); });
     });
 });
 
-- 
2.21.0

