commit 2f0146a56acd387ce30cdf01cead2842f757ec65 (refs/changes/36/2936/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2017-11-10T16:18:21-05:00 (1 year, 11 months ago)
    
    NET-381 use triton-origin for booter/dhcpd

diff --git a/targets.json.in b/targets.json.in
index d8ff174..f2009ce 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -269,9 +269,10 @@ cat <<EOF
     "image_name": "dhcpd",
     "image_description": "SDC DHCPD",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
+    "// image_uuid": "triton-origin-multiarch-15.4.1@1.0.1",
+    "image_uuid": "04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
     "pkgsrc": [
-      "nginx-1.0.10",
+      "nginx-1.10.1",
       "tftp-hpa-5.2"
     ],
     "repos": [
