From e1f8e48dd6ad7356a864fcfc031de15b870eb2d2 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <josh@wilsdon.ca>
Date: Wed, 2 Aug 2017 16:51:41 -0700
Subject: [PATCH] VOLAPI-66 VOLAPI-19 broke volapi

---
 lib/endpoints/ping.js | 1 -
 1 file changed, 1 deletion(-)

diff --git a/lib/endpoints/ping.js b/lib/endpoints/ping.js
index 51be764..47bb0af 100644
--- a/lib/endpoints/ping.js
+++ b/lib/endpoints/ping.js
@@ -11,7 +11,6 @@
 var assert = require('assert');
 var restify = require('restify');
 
-var common = require('./common');
 var errors = require('../errors');
 var validationUtils = require('../validation/utils');
 
-- 
2.21.0

