commit e1f8e48dd6ad7356a864fcfc031de15b870eb2d2 (refs/changes/25/2325/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2017-08-02T16:51:41-07:00 (2 years, 2 months ago)
    
    VOLAPI-66 VOLAPI-19 broke volapi

diff --git a/lib/endpoints/ping.js b/lib/endpoints/ping.js
index 51be764..47bb0af 100644
--- a/lib/endpoints/ping.js
+++ b/lib/endpoints/ping.js
@@ -11,7 +11,6 @@
 var assert = require('assert');
 var restify = require('restify');
 
-var common = require('./common');
 var errors = require('../errors');
 var validationUtils = require('../validation/utils');
 
