commit 02903085fdcd37f35b712d4f2012eae108051838 (refs/changes/82/2482/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-08-31T15:27:20+00:00 (2 years, 1 month ago)
    
    8572 ccompile.h: rename __GNU_UNUSED to __unused [fix broken build]

diff --git a/src/dockerinit/json-nvlist/json-nvlist.c b/src/dockerinit/json-nvlist/json-nvlist.c
index 9daa05fe..dedacb81 100644
--- a/src/dockerinit/json-nvlist/json-nvlist.c
+++ b/src/dockerinit/json-nvlist/json-nvlist.c
@@ -332,7 +332,7 @@ hdlr_bareword(state_t *s)
 /* ARGSUSED */
 static int
 collect_number(state_t *s, boolean_t *isint, int32_t *result,
-    double *fresult __GNU_UNUSED)
+    double *fresult __unused)
 {
 	boolean_t neg = B_FALSE;
 	int t;
diff --git a/src/dockerinit/src/docker-common.c b/src/dockerinit/src/docker-common.c
index b9ab418d..c10b7ba1 100644
--- a/src/dockerinit/src/docker-common.c
+++ b/src/dockerinit/src/docker-common.c
@@ -293,8 +293,8 @@ mdataDelete(const char *keyname)
  * list entry with the same variable name.
  */
 void
-cbEachEnvEntry(const char *array_name __GNU_UNUSED, unsigned int idx
-  __GNU_UNUSED, const char *val, void *arg0, void *arg1 __GNU_UNUSED)
+cbEachEnvEntry(const char *array_name __unused, unsigned int idx
+  __unused, const char *val, void *arg0, void *arg1 __unused)
 {
     strlist_t *env = arg0;
     char *env_name = NULL;
@@ -862,8 +862,8 @@ getPathList(strlist_t *env, strlist_t *path, const char *working_directory)
  * into the next available slot in the combined command string list.
  */
 void
-cbEachCmdEntry(const char *array_name __GNU_UNUSED, unsigned int idx
-  __GNU_UNUSED, const char *val, void *arg0, void *arg1)
+cbEachCmdEntry(const char *array_name __unused, unsigned int idx
+  __unused, const char *val, void *arg0, void *arg1)
 {
     strlist_t *cmdline = arg0;
     const char *typ = arg1;
