From 83ddb6ae89918f69df1903ccd0cb5512f5af7fd7 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 16 Oct 2017 12:28:48 +0000
Subject: [PATCH] 8665 Need support for LSI 9305-24i

---
 overlay/generic/etc/driver_aliases | 1 +
 1 file changed, 1 insertion(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index c3836113..f32881c6 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -927,6 +927,7 @@ mpt_sas "pciex1000,95"
 mpt_sas "pciex1000,96"
 mpt_sas "pciex1000,97"
 mpt_sas "pciex1000,c4"
+mpt_sas "pciex1000,c9"
 mr_sas "pciex1000,52"
 mr_sas "pciex1000,53"
 mr_sas "pciex1000,5b"
-- 
2.21.0

