From 3737071f74dbe668d162b4e7c8fc1bc4df682a32 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Fri, 2 Jun 2017 22:26:07 +0000
Subject: [PATCH] joyent/bugview#19 Show related links

---
 Makefile     | 10 +++++++--
 jirapub.js   | 61 ++++++++++++++++++++++++++++++++++++++++++++++++++--
 package.json |  5 ++++-
 3 files changed, 71 insertions(+), 5 deletions(-)

diff --git a/Makefile b/Makefile
index 37ac8c3..b3ce882 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,6 @@
 
+JSHINT = ./node_modules/.bin/jshint
+
 JS_FILES = \
 	jirapub.js
 
@@ -6,8 +8,12 @@ JS_FILES = \
 all: 0-npm-stamp
 
 .PHONY: check
-check:
-	jshint $(JS_FILES)
+check: $(JSHINT)
+	$(JSHINT) $(JS_FILES)
+
+$(JSHINT):
+	npm install \
+	    jshint@`json -f package.json devDependencies.jshint`
 
 0-npm-stamp:
 	npm install
diff --git a/jirapub.js b/jirapub.js
index 6c108d7..3c22f06 100644
--- a/jirapub.js
+++ b/jirapub.js
@@ -21,6 +21,11 @@ var HEADING_LEVELS = [ 1, 2, 3, 4, 5, 6 ].map(function (l) {
 
 var TEMPLATES = {};
 
+var ALLOWED_DOMAINS = [
+	'cr.joyent.us',
+	'github.com'
+];
+
 var CONFIG = read_config(LOG);
 
 var JIRA;
@@ -874,6 +879,8 @@ format_issue(opts, callback)
 	mod_assert.object(opts.log, 'opts.log');
 	mod_assert.func(callback, 'callback');
 
+	var rls;
+
 	var issue = opts.issue;
 	var log = opts.log;
 
@@ -944,14 +951,39 @@ format_issue(opts, callback)
 			next(err);
 		});
 
+	}, function get_remote_links(next) {
+		var url = CONFIG.url.path +
+		    '/issue/' + issue.id + '/remotelink';
+
+		JIRA.get(url, function (_err, _req, _res, links) {
+			if (_err) {
+				next(_err);
+				return;
+			}
+
+			mod_assert.array(links, 'links');
+
+			rls = links.filter(function (rl) {
+				var parsed = mod_url.parse(rl.object.url);
+				var domain = parsed.hostname;
+
+				if (domain === null) {
+					return false;
+				}
+
+				return (ALLOWED_DOMAINS.indexOf(domain) !== -1);
+			});
+
+			next(null);
+		});
 	}, function do_format(next) {
-		next(null, format_issue_finalise(issue, other_issues));
+		next(null, format_issue_finalise(issue, rls, other_issues));
 
 	} ], callback);
 }
 
 function
-format_issue_finalise(issue, other_issues)
+format_issue_finalise(issue, remotelinks, other_issues)
 {
 	mod_assert.object(issue, 'issue');
 	mod_assert.object(other_issues, 'other_issues');
@@ -1014,6 +1046,21 @@ format_issue_finalise(issue, other_issues)
 		}
 	}
 
+	if (remotelinks.length > 0) {
+		out += '<h2>Related Links</h2>\n';
+		out += '<p><ul>\n';
+
+		for (i = 0; i < remotelinks.length; i++) {
+			var rl = remotelinks[i].object;
+
+			out += '<li>';
+			out += format_remote_link(rl.url, rl.title);
+			out += '</li>\n';
+		}
+
+		out += '</ul></p>\n';
+	}
+
 	if (issue.fields.description) {
 		out += '<h2>Description</h2>\n';
 		out += '<div>';
@@ -1069,6 +1116,16 @@ format_issue_finalise(issue, other_issues)
 	return (out);
 }
 
+
+function
+format_remote_link(link, text)
+{
+	var anchor = '<a rel="noopener noreferrer" target="_blank" href="' +
+	    link + '">' + text + '</a>';
+
+	return (anchor);
+}
+
 /*
  * Main:
  */
diff --git a/package.json b/package.json
index 9817e26..c508f11 100644
--- a/package.json
+++ b/package.json
@@ -7,10 +7,13 @@
   "license": "MIT",
   "private": true,
   "dependencies": {
-    "assert-plus": "^0.1.5",
+    "assert-plus": "^1.0.0",
     "bunyan": "~0.22.1",
     "ent": "~2.0.0",
     "restify": "~2.6.1",
     "vasync": "^1.6.4"
+  },
+  "devDependencies": {
+    "jshint": "^2.5.6"
   }
 }
-- 
2.21.0

