commit 04f96c7c1184e2fbd58cbe6000f30efd58fbf650 (refs/changes/09/5309/3)
Author: Rui Loura <rui@joyent.com>
Date:   2019-01-03T21:44:42+00:00 (9 months ago)
    
    TRITON-721 joyent-schemas should support RAN sysinfo keys

diff --git a/lib/cnapi/server.js b/lib/cnapi/server.js
index 088c2fc..3124ed6 100644
--- a/lib/cnapi/server.js
+++ b/lib/cnapi/server.js
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Joyent, Inc.  All rights reserved.
+ * Copyright 2019 Joyent, Inc.  All rights reserved.
  */
 
 var schema_util = require('../util.js');
@@ -204,6 +204,8 @@ var s_sysinfo = {
 		'CPU Type': base.str.nonempty.req,
 		'CPU Virtualization': base.str.req,
 		'CPU Physical Cores': base.intg.pos.req,
+		'Admin NIC Tag': base.str.opt, // Added by OS-7075 (20180719)
+		'Admin IP': base.str.v4addr.opt, // Added by TRITON-841 (20181206)
 		'UUID': base.str.uuid.req,
 		'Hostname': base.str.hostname.opt,
 		'HVM API': base.bool.opt,
diff --git a/lib/dapi/server.js b/lib/dapi/server.js
index 3ee2db1..99f2969 100644
--- a/lib/dapi/server.js
+++ b/lib/dapi/server.js
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Joyent, Inc.  All rights reserved.
+ * Copyright 2019 Joyent, Inc.  All rights reserved.
  */
 
 var schema_util = require('../util.js');
@@ -50,6 +50,8 @@ var s_sysinfo = {
 		'Bhyve Capable': base.bool.opt,
 		'Bhyve Max Vcpus': base.intg.nonneg.opt,
 		'CPU Total Cores': base.intg.pos.req,
+		'Admin NIC Tag': base.str.opt, // Added by OS-7075 (20180719)
+		'Admin IP': base.str.v4addr.opt, // Added by TRITON-841 (20181206)
 		'HVM API': base.bool.opt,
 		'Network Interfaces': s_si_nics_r,
 		'Live Image': base.str.nonempty.req,
