From 0b2f81a50b074843fcfbdd2b3bfc89b6a637a57f Mon Sep 17 00:00:00 2001
From: Orlando Vazquez <ovazquez@gmail.com>
Date: Thu, 26 Jan 2017 12:52:26 -0800
Subject: [PATCH] CNAPI-690 CNAPI needs to be selective with logging request
 payload

---
 lib/server.js | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index 7276d2c..db59ae5 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -13,6 +13,7 @@ var restify = require('restify');
 var restifyValidator = require('restify-validator');
 var trace_event = require('trace-event');
 var formatJSON = require('restify/lib/formatters/json');
+var jsrim = require('jsprim');
 
 var endpoints = require('./endpoints/index');
 var request_seq_id = 0;
@@ -32,6 +33,7 @@ function createServer(options) {
         'servereventheartbeat': true,
         'servereventvmsupdate': true
     };
+
     cnapi.use(function setTracing(req, res, next) {
         req.trace = trace_event.createBunyanTracer({
             log: req.log
@@ -72,18 +74,27 @@ function createServer(options) {
         'servereventheartbeat': true,
         'servereventvmsupdate': true
     };
+
+    var AUDIT_OMIT_BODY_ROUTES = {
+        'vmdockerbuild': true
+    };
+
     cnapi.on('after', function (req, res, route, err) {
         if (req.route && AUDIT_SKIP_ROUTES[req.route.name]) {
             return;
         }
 
         // Successful GET res bodies are uninteresting and *big*.
-        var body = req.method !== 'GET' &&
-                   res.statusCode !== 404 &&
-                   Math.floor(res.statusCode/100) !== 2;
+        var isOmitBodyRoute = (jsprim.hasKey(GET_NO_LOG_PAYLOAD, req.route.name)
+                    && AUDIT_NO_LOG_BODY[req.route.name]);
+
+        var includeBody = !isOmitBodyRoute || (req.method !== 'GET' &&
+                        res.statusCode !== 404 &&
+                        Math.floor(res.statusCode/100) !== 2);
+
         restify.auditLogger({
             log: req.log.child({ route: route && route.name }, true),
-            body: body
+            body: includeBody
         })(req, res, route, err);
     });
 
-- 
2.21.0

