From 25c4488665b35932fea625f19a77633e504b4458 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Thu, 28 Jul 2016 12:23:01 +0000
Subject: [PATCH] 6064 ixgbe needs X550 support

---
 overlay/generic/etc/driver_aliases | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 4b1804e2..a2ad1d27 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -748,7 +748,17 @@ ixgbe "pciex8086,1517"
 ixgbe "pciex8086,151c"
 ixgbe "pciex8086,1528"
 ixgbe "pciex8086,154d"
+ixgbe "pciex8086,154a"
 ixgbe "pciex8086,1557"
+ixgbe "pciex8086,1558"
+ixgbe "pciex8086,1560"
+ixgbe "pciex8086,1563"
+ixgbe "pciex8086,15aa"
+ixgbe "pciex8086,15ab"
+ixgbe "pciex8086,15ac"
+ixgbe "pciex8086,15ad"
+ixgbe "pciex8086,15ae"
+ixgbe "pciex8086,15d1"
 kb8042 "pnpPNP,303"
 lsimega "pci1000,1960"
 lsimega "pci1000,407"
-- 
2.21.0

