From 63d3dfc5cef9322f4e14f05f6a5bf71d7cac9145 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 24 Mar 2017 19:24:00 +0100
Subject: [PATCH] TOOLS-1726 sdcadm create should not request usage of
 '--force-data-path'

---
 lib/sdcadm.js | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 90ae0dc..3b99fbd 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -3635,6 +3635,10 @@ SdcAdm.prototype.genUpdatePlan = function genUpdatePlan(options, cb) {
 
             for (var i = 0; i < changes.length; i++) {
                 var ch = changes[i];
+                if (ch.type !== 'update-instance' &&
+                    ch.type !== 'update-service') {
+                    continue;
+                }
                 if (ch.service && dataPath.indexOf(ch.service.name) !== -1 &&
                     !options.forceDataPath)
                 {
-- 
2.21.0

