From 4ec0d7cbf77eaa3415f0f28a239b16c5af5afcda Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 22 Mar 2017 20:06:17 +0100
Subject: [PATCH] TOOLS-1726 sdcadm create should not request usage of
 '--force-data-path'

---
 lib/sdcadm.js | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 2cca542..47f5408 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -3588,6 +3588,10 @@ SdcAdm.prototype.genUpdatePlan = function genUpdatePlan(options, cb) {
 
             for (var i = 0; i < changes.length; i++) {
                 var ch = changes[i];
+                if (ch.type !== 'update-instance' &&
+                    ch.type !== 'update.service') {
+                    continue;
+                }
                 if (ch.service && dataPath.indexOf(ch.service.name) !== -1 &&
                     !options.forceDataPath)
                 {
-- 
2.21.0

