From db39437ba4ae5311c95581158048d2054288ef53 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Fri, 17 Nov 2017 13:13:08 +0000
Subject: [PATCH] 8804 nvme: add alias for pciclass,010802

---
 overlay/generic/etc/driver_aliases | 1 +
 1 file changed, 1 insertion(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 0c11a3d9..b0b10436 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -987,6 +987,7 @@ nv_sata "pci10de,3f6"
 nv_sata "pci10de,3f7"
 nv_sata "pci10de,54"
 nv_sata "pci10de,55"
+nvme "pciclass,010802"
 nvme "pciexclass,010802"
 nxge "SUNW,niusl"
 nxge "pciex108e,abcd"
-- 
2.21.0

