commit 56c5490e60c80e4643635b36f93dda18119cc574 (refs/changes/20/5520/2)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-21T15:13:25+00:00 (8 months ago)
    
    TRITON-1118 convert sdc-assets to engbld framework

diff --git a/.gitignore b/.gitignore
index 89cb514..1274f22 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1 +1,3 @@
-/assets-pkg-master-*.tar.bz2
+bits
+/make_stamps
+/assets-pkg-master-*.tar.gz
diff --git a/.gitmodules b/.gitmodules
index 48ab5ca..15c4a41 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,3 +1,9 @@
 [submodule "deps/sdc-scripts"]
 	path = deps/sdc-scripts
 	url = https://github.com/joyent/sdc-scripts.git
+[submodule "eng"]
+	path = eng
+	url = /Users/timf/projects/submodule/eng.git/
+[submodule "deps/eng"]
+	path = deps/eng
+	url = https://github.com/joyent/eng.git
diff --git a/Makefile b/Makefile
index 0147a1a..268f8db 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 NAME=assets
@@ -14,7 +14,7 @@ ifeq ($(VERSION), "")
     @echo "Use gmake"
 endif
 
-TAR = tar
+TAR = gtar
 
 ifeq ($(TIMESTAMP),)
     TIMESTAMP=$(shell date -u "+%Y%m%dT%H%M%SZ")
@@ -24,7 +24,29 @@ ASSETS_PUBLISH_VERSION := $(shell git symbolic-ref HEAD | \
       awk -F / '{print $$3}')-$(TIMESTAMP)-g$(shell \
                 git describe --all --long | awk -F '-g' '{print $$NF}')
 
-RELEASE_TARBALL=assets-pkg-$(ASSETS_PUBLISH_VERSION).tar.bz2
+RELEASE_TARBALL=assets-pkg-$(ASSETS_PUBLISH_VERSION).tar.gz
+
+#
+# Stuff used for buildimage
+#
+BASE_IMAGE_UUID		= 04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f
+BUILDIMAGE_NAME		= assets
+BUILDIMAGE_DESC		= SDC Assets
+BUILDIMAGE_PKGSRC	= nginx-1.11.1nb1
+
+AGENTS = amon
+
+DISTCLEAN_FILES += root
+
+ENGBLD_USE_BUILDIMAGE	= true
+ENGBLD_REQUIRE		:= $(shell git submodule update --init deps/eng)
+include ./deps/eng/tools/mk/Makefile.defs
+TOP ?= $(error Unable to access eng.git submodule Makefiles.)
+
+ifeq ($(shell uname -s),SunOS)
+	include ./deps/eng/tools/mk/Makefile.agent_prebuilt.defs
+endif
+
 
 .PHONY: all
 
@@ -36,12 +58,10 @@ $(RELEASE_TARBALL):
 	TAR=$(TAR) bash package.sh $(RELEASE_TARBALL)
 
 publish:
-	@if [[ -z "$(BITS_DIR)" ]]; then \
-      echo "error: 'BITS_DIR' must be set for 'publish' target"; \
-      exit 1; \
-    fi
-	mkdir -p $(BITS_DIR)/assets
-	cp $(RELEASE_TARBALL) $(BITS_DIR)/assets/$(RELEASE_TARBALL)
-
-clean:
-	rm -fr assets-*.tar.bz2
+	mkdir -p $(ENGBLD_BITS_DIR)/assets
+	cp $(RELEASE_TARBALL) $(ENGBLD_BITS_DIR)/assets/$(RELEASE_TARBALL)
+
+include ./deps/eng/tools/mk/Makefile.targ
+ifeq ($(shell uname -s),SunOS)
+	include ./deps/eng/tools/mk/Makefile.agent_prebuilt.targ
+endif
diff --git a/deps/eng b/deps/eng
new file mode 160000
index 0000000..7c472bc
--- /dev/null
+++ b/deps/eng
@@ -0,0 +1 @@
+Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
diff --git a/package.sh b/package.sh
index 2b5cde6..15c1ee9 100755
--- a/package.sh
+++ b/package.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 set -o xtrace
@@ -15,6 +15,10 @@ set -o errexit
 RELEASE_TARBALL=$1
 echo "Building ${RELEASE_TARBALL}"
 
+if [[ -z "$TAR" ]]; then
+	TAR=gtar
+fi
+
 ROOT=$(pwd)
 
 tmpdir="/tmp/assets.$$"
@@ -29,6 +33,6 @@ mkdir -p ${tmpdir}/root/opt/smartdc/boot
 cp -R ${ROOT}/deps/sdc-scripts/* ${tmpdir}/root/opt/smartdc/boot/
 cp -R ${ROOT}/boot/* ${tmpdir}/root/opt/smartdc/boot/
 
-( cd ${tmpdir}; tar -jcf ${ROOT}/${RELEASE_TARBALL} root site)
+( cd ${tmpdir}; ${TAR} -I pigz -cf ${ROOT}/${RELEASE_TARBALL} root site)
 
 rm -rf ${tmpdir}
