commit e4225f058549308ad3719634e3159dc0e977b789 (refs/changes/15/3715/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-03-23T19:53:49+00:00 (1 year, 7 months ago)
    
    OS-6799 hvm exclusion failures are silent

diff --git a/usr/src/uts/i86pc/os/pc_hvm.c b/usr/src/uts/i86pc/os/pc_hvm.c
index b74ceb17bf..f04c031e0d 100644
--- a/usr/src/uts/i86pc/os/pc_hvm.c
+++ b/usr/src/uts/i86pc/os/pc_hvm.c
@@ -10,9 +10,12 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
+#include <sys/cmn_err.h>
+#include <sys/ddi.h>
+#include <sys/sunddi.h>
 #include <sys/param.h>
 #include <sys/types.h>
 #include <sys/mutex.h>
@@ -36,14 +39,22 @@ boolean_t
 hvm_excl_hold(const char *consumer)
 {
 	boolean_t res = B_FALSE;
+	char holder[32];
 
 	mutex_enter(&hvm_excl_lock);
 	if (hvm_excl_holder == NULL) {
 		hvm_excl_holder = consumer;
 		res = B_TRUE;
+	} else {
+		(void) strlcpy(holder, hvm_excl_holder, sizeof (holder));
 	}
 	mutex_exit(&hvm_excl_lock);
 
+	if (!res) {
+		cmn_err(CE_WARN, "%s: '%s' cannot take hvm_xcl_lock: "
+		    "held by '%s'", __func__, consumer, holder);
+	}
+
 	return (res);
 }
 
