From 320ce5087607ddf07b5532a99fdc2b719f31abc2 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Thu, 27 Oct 2016 09:52:37 +0100
Subject: [PATCH] MANTA-2799 "manta-adm update" should fail when given
 non-existent service

---
 lib/adm.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/adm.js b/lib/adm.js
index 94e2b1a..1ab4146 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -1390,6 +1390,11 @@ maAdm.prototype.generatePlan = function (callback, service, noreprovision)
 	var self = this;
 	var log = this.ma_log;
 
+	if (service && !svcs.serviceNameIsValid(service)) {
+		callback(new VError('unrecognized service: "%s"', service));
+		return;
+	}
+
 	log.info('generating plan');
 	log.trace('generating plan', this.ma_instances_wanted,
 	    this.ma_config_bycfg);
-- 
2.21.0

