From 6f8f95a447a2e4ea8c30601bb67825267e81bcbc Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Wed, 26 Oct 2016 10:51:34 +0100
Subject: [PATCH] MANTA-2799 "manta-adm update" should fail when given
 non-existent service

---
 lib/adm.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/adm.js b/lib/adm.js
index 94e2b1a..fce45db 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -1383,6 +1383,11 @@ maAdm.prototype.generatePlan = function (callback, service, noreprovision)
 	assert.ok(this.ma_instances_wanted !== null);
 	assert.ok(this.ma_plan === null);
 
+	if(service && !svcs.serviceNameIsValid(service)){
+		callback(new VError('unrecognized service: "%s"', service));
+		return;
+	}
+
 	var cnid, dcnconf, svcconfig, svcid, svcname, actual;
 	var svcname2uuids = {};
 	var usedany = false;
-- 
2.21.0

