commit 774a4c4b8d5f3c59645b2f7e991ca3d481d9e745 (refs/changes/96/3796/3)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-04-11T13:38:38+00:00 (1 year, 6 months ago)
    
    OS-6863 presetate should enforce zlog-name

diff --git a/overlay/generic/usr/lib/brand/jcommon/statechange b/overlay/generic/usr/lib/brand/jcommon/statechange
index 2678b0fa..4f61a02a 100644
--- a/overlay/generic/usr/lib/brand/jcommon/statechange
+++ b/overlay/generic/usr/lib/brand/jcommon/statechange
@@ -794,27 +794,76 @@ cleanup_mount()
 	    id_gz_sockholder /var/run/.smartdc-amon.sock
 }
 
+function fix_forced_attrs {
+	typeset attr
+
+	for attr in ${!FORCED_ATTRS[@]}; do
+		typeset nval=${FORCED_ATTRS["$attr"]}
+		typeset -n envvar=_ZONECFG_attr_${attr//-/_}
+		typeset cval=$envvar
+
+		if [[ $cval == $nval ]]; then
+			# In most cases, $nval and $cval will be the same and
+			# nothing needs to be done. This includes the case where
+			# $nval and $cval are "".
+			continue
+		elif [[ -z $nval ]]; then
+			logger -p daemon.error "[zone $ZONENAME]" \
+			    "Illegal value for attr '$attr': '$cval'." \
+			    "Removing attr '$attr'"
+			zonecfg -z "$ZONENAME" "remove -F attr name=$attr"
+
+			unset ${!envvar}
+		else
+			logger -p daemon.error "[zone $ZONENAME]" \
+			    "Illegal value for attr '$attr': '$cval'." \
+			    "Setting to '$nval'"
+			zonecfg -z "$ZONENAME" "remove -F attr name=$attr;" \
+			    "add attr; set type=string;" \
+			    "set name=$attr; set value=\"$nval\"; end;"
+
+			export ${!envvar}="$nval"
+		fi
+	done
+}
+
 #
 # Main
 #
 
-[[ "$subcommand" == "pre" && $cmd == 0 ]] && setup_fs
-if [[ "$subcommand" == "post" && $cmd == 0 ]]; then
-	[[ -n "$jst_showsnap" ]] && setup_snapshots
-	setup_net
-	setup_fw
-fi
-
-# We can't set a rctl until we have a process in the zone to grab
-[[ "$subcommand" == "post" && $cmd == 1 ]] && setup_cpu_baseline
-
-# Zone halt is hung unmounting, try to recover
-[[ "$subcommand" == "post" && $state == 6 && $cmd == 8 ]] && \
-    cleanup_mount "$6"
-
-if [[ "$subcommand" == "pre" && $cmd == 4 ]]; then
-	[[ -n "$jst_snowsnap" ]] && cleanup_snapshots
-	cleanup_net
-fi
+case $subcommand in
+pre)
+	case $cmd in
+	0)	# pre-ready
+		fix_forced_attrs
+		setup_fs
+		;;
+	4)	# pre-halt
+		[[ -n "$jst_snowsnap" ]] && cleanup_snapshots
+		cleanup_net
+		;;
+	esac
+	;;
+post)
+	case $cmd in
+	0)	# post-ready
+		[[ -n "$jst_showsnap" ]] && setup_snapshots
+		setup_net
+		setup_fw
+		;;
+	1)	# post-boot
+		# We can't set a rctl until we have a process in the zone to
+		# grab
+		setup_cpu_baseline
+		;;
+	8)	# post-unmount
+		# Zone halt is hung unmounting, try to recover
+		if [[ $state == 6 ]]; then
+			cleanup_mount "$6"
+		fi
+		;;
+	esac
+	;;
+esac
 
 exit 0
diff --git a/overlay/generic/usr/lib/brand/joyent-minimal/statechange b/overlay/generic/usr/lib/brand/joyent-minimal/statechange
index 22731cf6..5e2eeff0 100755
--- a/overlay/generic/usr/lib/brand/joyent-minimal/statechange
+++ b/overlay/generic/usr/lib/brand/joyent-minimal/statechange
@@ -12,7 +12,7 @@
 #
 
 #
-# Copyright (c) 2014 Joyent, Inc.  All rights reserved.
+# Copyright (c) 2018 Joyent, Inc.  All rights reserved.
 #
 
 # Do we support vrrp?
@@ -33,4 +33,10 @@ jst_showsnap="yes"
 # Where does the mdata socket live?
 jst_mdatapath="/.zonecontrol/"
 
+# Ensure docker zfd(7D) log is not put in the wrong place
+if [[ "$_ZONECFG_attr_docker" == true ]]; then
+	typeset -A FORCED_ATTRS
+	FORCED_ATTRS["zlog-name"]=
+fi
+
 . /usr/lib/brand/jcommon/statechange
diff --git a/overlay/generic/usr/lib/brand/joyent/statechange b/overlay/generic/usr/lib/brand/joyent/statechange
index 22731cf6..5e2eeff0 100755
--- a/overlay/generic/usr/lib/brand/joyent/statechange
+++ b/overlay/generic/usr/lib/brand/joyent/statechange
@@ -12,7 +12,7 @@
 #
 
 #
-# Copyright (c) 2014 Joyent, Inc.  All rights reserved.
+# Copyright (c) 2018 Joyent, Inc.  All rights reserved.
 #
 
 # Do we support vrrp?
@@ -33,4 +33,10 @@ jst_showsnap="yes"
 # Where does the mdata socket live?
 jst_mdatapath="/.zonecontrol/"
 
+# Ensure docker zfd(7D) log is not put in the wrong place
+if [[ "$_ZONECFG_attr_docker" == true ]]; then
+	typeset -A FORCED_ATTRS
+	FORCED_ATTRS["zlog-name"]=
+fi
+
 . /usr/lib/brand/jcommon/statechange
diff --git a/overlay/generic/usr/lib/brand/lx/statechange b/overlay/generic/usr/lib/brand/lx/statechange
index d4c434e8..d0bc11ee 100755
--- a/overlay/generic/usr/lib/brand/lx/statechange
+++ b/overlay/generic/usr/lib/brand/lx/statechange
@@ -11,7 +11,7 @@
 #
 
 #
-# Copyright (c) 2014 Joyent, Inc.  All rights reserved.
+# Copyright (c) 2018 Joyent, Inc.  All rights reserved.
 #
 
 # Do we support vrrp?
@@ -32,4 +32,10 @@ jst_showsnap="yes"
 # Where does the mdata socket live?
 jst_mdatapath="/native/.zonecontrol/"
 
+# Ensure docker zfd(7D) log is not put in the wrong place
+if [[ "$_ZONECFG_attr_docker" == true ]]; then
+	typeset -A FORCED_ATTRS
+	FORCED_ATTRS["zlog-name"]=
+fi
+
 . /usr/lib/brand/jcommon/statechange
