commit affcf28f23b303949b11e9b420b368676f78c37b (refs/changes/38/1738/1)
Author: Kody Kantor <Kody.Kantor@gmail.com>
Date:   2017-04-03T16:29:10-05:00 (2 years, 6 months ago)
    
    MANTA-2980 electric-moray should report moray shard

diff --git a/lib/audit.js b/lib/audit.js
index 93b380b..284fe21 100644
--- a/lib/audit.js
+++ b/lib/audit.js
@@ -233,6 +233,7 @@ function auditLogger(options) {
         obj.sharksContacted = req.sharksContacted;
         obj.entryShard = req.entryShard;
         obj.parentShard = req.parentShard;
+        obj.morayContacted = req.morayContacted;
 
         if (req._timeToLastByte !== undefined &&
             req._totalBytes !== undefined) {
diff --git a/lib/obj.js b/lib/obj.js
index 6686ece..ecbb4b7 100644
--- a/lib/obj.js
+++ b/lib/obj.js
@@ -591,13 +591,14 @@ function saveMetadata(req, res, next) {
         log.debug({
             options: opts
         }, 'saveMetadata: entered');
-        req.moray.putMetadata(opts, function (err2) {
+        req.moray.putMetadata(opts, function (err2, obj) {
             req.sharks = null;
             if (err2) {
                 log.debug(err2, 'saveMetadata: failed');
                 next(err2);
             } else {
                 log.debug('saveMetadata: done');
+                req.morayContacted = obj._node; // for debugging
                 if (req.headers['origin']) {
                     res.header('Access-Control-Allow-Origin',
                                req.headers['origin']);
