commit 91599e804f9be049a4387f2ca9ff0179fc8b537f (refs/changes/38/1738/2)
Author: Kody Kantor <Kody.Kantor@gmail.com>
Date:   2017-04-13T15:27:56-05:00 (2 years, 6 months ago)
    
    MANTA-2980 electric-moray should report moray shard

diff --git a/lib/errors.js b/lib/errors.js
index bd5c156..fd28651 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -625,7 +625,7 @@ function translateError(err, req) {
         return (new ConcurrentRequestError(req.path()));
     }
 
-    return (new InternalError());
+    return (new InternalError(err));
 }
 
 
diff --git a/lib/obj.js b/lib/obj.js
index 6686ece..2a7ac74 100644
--- a/lib/obj.js
+++ b/lib/obj.js
@@ -53,7 +53,6 @@
 // requests, the number of copies, the size, etc.
 
 var crypto = require('crypto');
-var util = require('util');
 
 var assert = require('assert-plus');
 var libmanta = require('libmanta');
@@ -75,7 +74,6 @@ require('./errors');
 
 var clone = utils.shallowCopy;
 var httpDate = restify.httpDate;
-var sprintf = util.format;
 
 var DATA_TIMEOUT = parseInt(process.env.MUSKIE_DATA_TIMEOUT || 45000, 10);
 var DEF_MAX_LEN = 53687091200;
@@ -458,7 +456,7 @@ function sharkStreams(req, res, next) {
                 muskieMd5: check.digest('base64'),
                 makoMd5: _md5s
             }, 'mako didnt recieve what muskie sent');
-            var m = sprintf('muskie md5 %s and mako md5 ' +
+            var m = new VError('muskie md5 %s and mako md5 ' +
                             '%s don\'t match', check.digest('base64'),
                             _md5s.join());
             next_err(new InternalError(m));
@@ -511,7 +509,9 @@ function sharkStreams(req, res, next) {
                         client_res: sres,
                         body: body
                     }, 'mako: response error');
-                    next_err(new InternalError());
+                    var m = new VError('muskie response error, storage id (%s)',
+                                        s._shark.manta_storage_id);
+                    next_err(new InternalError(m));
                 });
                 sres.once('error', function (err) {
                     next_err(new InternalError(err));
@@ -591,8 +591,11 @@ function saveMetadata(req, res, next) {
         log.debug({
             options: opts
         }, 'saveMetadata: entered');
-        req.moray.putMetadata(opts, function (err2) {
+        req.moray.putMetadata(opts, function (err2, obj) {
             req.sharks = null;
+            if (obj) {
+              req.entryShard = obj._node.pnode; // for debugging
+            }
             if (err2) {
                 log.debug(err2, 'saveMetadata: failed');
                 next(err2);
