commit 246a6c3d11c5d371b2d7ecea3bead59be9a57026 (refs/changes/65/965/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2016-11-29T23:48:37+13:00 (2 years, 10 months ago)
    
    PUBAPI-1291: primaryIp's logic could be better

diff --git a/lib/machines.js b/lib/machines.js
index 201dc98..4584b91 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -164,17 +164,29 @@ function translate(machine, req)  {
             }
         });
 
-        msg.primaryIp = machine.nics.reduce(function (acc, nic) {
-            return (nic.nic_tag === 'external') ? nic.ip : acc;
-        }, '');
+        var primaryNic = machine.nics.filter(function (nic) {
+            return nic.nic_tag === 'external';
+        }).pop();
+
+        if (!primaryNic) {
+            primaryNic = machine.nics.filter(function (nic) {
+                return nic.primary;
+            }).pop();
+        }
 
         // PUBAPI-727: If a machine has only internal networks, use that
         // as the primary IP
-        if (msg.primaryIp === '') {
-            msg.primaryIp = machine.nics.reduce(function (acc, nic) {
-                return (nic.nic_tag === 'internal') ? nic.ip : acc;
-            }, '');
+        if (!primaryNic) {
+            primaryNic = machine.nics.filter(function (nic) {
+                return nic.nic_tag === 'internal';
+            }).pop();
+        }
+
+        if (!primaryNic) {
+            primaryNic = machine.nics[0];
         }
+
+        msg.primaryIp = primaryNic ? primaryNic.ip : '';
     }
 
     if (credentials && machine.internal_metadata &&
diff --git a/test/machines/common.js b/test/machines/common.js
index 58a0765..d3fd836 100644
--- a/test/machines/common.js
+++ b/test/machines/common.js
@@ -81,6 +81,10 @@ function checkMachine(t, m) {
     t.ok(m.metadata, 'checkMachine metadata ok');
     t.notOk(m.docker, 'checkMachine docker attr not set');
 
+    if (m.state === 'running') {
+        t.ok(m.primaryIp, 'checkMachine primaryIp ok');
+    }
+
     // Sometimes test suites from other applications create zones with a
     // 00000000-0000-0000-0000-000000000000 billing_id, which is changed by
     // cloudapi to '' since it's not an actual package UUID. Alas, we work
