commit a40bacdfacd95b593479a5e9496b5b90febf80d0 (refs/changes/80/3880/1)
Author: Kody A Kantor <kody@kkantor.com>
Date:   2018-04-26T19:38:43+00:00 (1 year, 5 months ago)
    
    MANTA-3523 make muskie monitoring tests optional, improve port wrangling

diff --git a/.gitmodules b/.gitmodules
index 721ad95..df5ac00 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/restdown"]
 	path = deps/restdown
 	url = https://github.com/trentm/restdown.git
diff --git a/deps/jsstyle b/deps/jsstyle
index d75b7ca..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit d75b7ca8308be17c80e2b120f2a01d4a0c20d8a8
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
diff --git a/package.json b/package.json
index 7115b69..10a45c8 100644
--- a/package.json
+++ b/package.json
@@ -12,7 +12,7 @@
         "ajv": "5.2.2",
         "ajv-keywords": "2.1.0",
         "aperture-config": "git+https://github.com/joyent/aperture-config.git#master",
-        "artedi": "1.1.1",
+        "artedi": "1.3.0",
         "assert-plus": "0.1.5",
         "backoff": "2.3.0",
         "bunyan": "0.22.1",
diff --git a/test/monitoring.test.js b/test/monitoring.test.js
index fa36ec3..53365ef 100644
--- a/test/monitoring.test.js
+++ b/test/monitoring.test.js
@@ -14,6 +14,7 @@ var helper = require('./helper.js');
 
 var restify = require('restify');
 var url = require('url');
+var net = require('net');
 var jsprim = require('jsprim');
 
 var before = helper.before;
@@ -21,30 +22,57 @@ var test = helper.test;
 
 ///--- Tests
 
+
 before(function (cb) {
     this.kangPath = '/kang/snapshot';
     this.metricsPath = '/metrics';
 
+    var port;
+
     /*
      * Since the monitoring server runs on a separate port (http_port + 800),
      * we need to do some slicing and dicing of the user-provided MANTA_URL.
      *
-     * Note: This requires that the user follows the README and set MANTA_URL
-     * to point to a running Muskie instance. This will fail if MANTA_URL
+     * Note: To run the monitoring tests, the user must set MANTA_URL
+     * to point to a running Muskie instance. The tests will fail if MANTA_URL
      * points to a loadbalancer, or doesn't include a port number.
+     *
+     * The MANTA_URL needs to have an IP address and port number. It could still
+     * be the case that the MANTA_URL points to a loadbalancer's IP and port,
+     * which will cause the monitoring tests to fail.
      */
 
     var parsed_url = url.parse(process.env.MANTA_URL);
     if (typeof (parsed_url.port) !== 'string') {
-        cb(new Error('MANTA_URL must include a valid port number'));
+        this.skip = true;
+        console.log('skipping monitoring tests: port number not found on' +
+            ' MANTA_URL');
+        cb();
+        return;
+    }
+
+    /*
+     * If the hostname isn't an IP then we're _probably_ not pointing at a
+     * muskie process.
+     */
+    if (! net.isIP(parsed_url.hostname) &&
+            parsed_url.hostname !== 'localhost') {
+
+        this.skip = true;
+        console.log('skipping monitoring tests: MANTA_URL is not an IP' +
+                ' address and port');
+        cb();
         return;
     }
 
     // Take the :port section of the URL and add 800.
-    var port = jsprim.parseInteger(parsed_url.port);
+    port = jsprim.parseInteger(parsed_url.port);
     if (typeof (port) !== 'number') {
         // parseInteger() returned an error, not a number.
-        cb(new Error('error parsing MANTA_URL port: ' +  port.message));
+        this.skip = true;
+        console.log('skipping monitoring test: error parsing port number on' +
+                ' MANTA_URL');
+        cb();
         return;
     }
     port += 800;
@@ -59,6 +87,10 @@ before(function (cb) {
 });
 
 test('kang handler running', function (t) {
+    if (this.skip) {
+        t.end();
+        return;
+    }
     var client = restify.createJsonClient({
         connectTimeout: 250,
         rejectUnauthorized: false,
@@ -75,6 +107,10 @@ test('kang handler running', function (t) {
 });
 
 test('metric handler running', function (t) {
+    if (this.skip) {
+        t.end();
+        return;
+    }
     var client = restify.createStringClient({
         connectTimeout: 250,
         rejectUnauthorized: false,
