commit 06f60d52b30e5d82fdae0b6bfeee028caea6a627 (refs/changes/10/4610/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-07-31T14:27:39-07:00 (1 year, 2 months ago)
    
    TRITON-647 vmapi NAT zone destruction should use SAPI DeleteInstance

diff --git a/lib/workflows/fabric-common.js b/lib/workflows/fabric-common.js
index 71a304e..578d663 100644
--- a/lib/workflows/fabric-common.js
+++ b/lib/workflows/fabric-common.js
@@ -164,6 +164,7 @@ function provisionFabricNats(job, cb) {
     var sapi = new sdcClients.SAPI({
         log: job.log.child({ component: 'sapi' }),
         url: sapiUrl,
+        version: '~2||~1',
         headers: { 'x-request-id': job.params['x-request-id'] }
     });
 
@@ -486,10 +487,6 @@ function destroyFabricNats(job, cb) {
         url: napiUrl,
         headers: headers
     });
-    var vmapi = new sdcClients.VMAPI({
-        url: vmapiUrl,
-        headers: headers
-    });
 
     /* Find the NAT VM and destroy it. */
     function destroyNatZone(fabricNic, done) {
@@ -497,7 +494,14 @@ function destroyFabricNats(job, cb) {
             alias: 'nat-' + fabricNic.network_uuid,
             state: 'running'
         };
+        var vmapi = new sdcClients.VMAPI({
+            url: vmapiUrl,
+            headers: headers
+        });
+
         vmapi.listVms(listVmParams, function onGetVm(err, vms) {
+            var sapi;
+
             if (err) {
                 done(err);
                 return;
@@ -519,11 +523,13 @@ function destroyFabricNats(job, cb) {
                 'Destroying NAT zone');
 
             /* Synchronously destroy the NAT vm. */
-            var deleteVmParams = {
-                sync: true,
-                uuid: vms[0].uuid
-            };
-            vmapi.deleteVm(deleteVmParams, function (dErr) {
+            sapi = new sdcClients.SAPI({
+                url: sapiUrl,
+                headers: headers,
+                version: '~2||~1',
+                log: job.log.child({ component: 'sapi' })
+            });
+            sapi.deleteInstance(vms[0].uuid, function (dErr) {
                 if (dErr && dErr.statusCode !== 404) {
                     done(dErr);
                     return;
