commit 526972d813bf926bd2366493214297d9e11624bb (refs/changes/26/726/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-10-18T09:16:08-07:00 (3 years ago)
    
    RELENG-728 purge-mg-builds: support verbose mrm

diff --git a/tools/purge-mg-builds b/tools/purge-mg-builds
index 9839083..0a6a5f9 100755
--- a/tools/purge-mg-builds
+++ b/tools/purge-mg-builds
@@ -233,7 +233,11 @@ function purge_dir
     dir="$1"
     log "mrm -r $dir"
     if [[ "$opt_dryrun" == "no" ]]; then
-        mrm -r "$dir"
+        if [[ -n "$TRACE" ]]; then
+            mrm -rv "$dir"
+        else
+            mrm -r "$dir"
+        fi
     fi
 }
 
