From 5e5731b3a41ab7bc3c6395a28f8964cddbe23e43 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Fri, 30 Sep 2016 11:55:31 +0000
Subject: [PATCH] 7382 basic NVMe 1.1 support

---
 overlay/generic/etc/driver_aliases | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 33880bd4..93c6bf39 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -935,7 +935,7 @@ nv_sata "pci10de,3f6"
 nv_sata "pci10de,3f7"
 nv_sata "pci10de,54"
 nv_sata "pci10de,55"
-nvme "pciex8086,953"
+nvme "pciexclass,010802"
 nxge "SUNW,niusl"
 nxge "pciex108e,abcd"
 oce "pciex19a2,700"
-- 
2.21.0

