From 325a7ce7f93688d2d1d028a687beae6e6327feaa Mon Sep 17 00:00:00 2001
From: andrew <andrew@whartonlabs.com>
Date: Fri, 3 Feb 2017 11:40:02 +1100
Subject: [PATCH] Added resize command to the list of instance commands.

---
 lib/do_instance/index.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/do_instance/index.js b/lib/do_instance/index.js
index f2554bd..d8f1513 100644
--- a/lib/do_instance/index.js
+++ b/lib/do_instance/index.js
@@ -34,6 +34,7 @@ function InstanceCLI(top) {
             'get',
             'create',
             'delete',
+            'resize',
             'rename',
             { group: '' },
             'start',
@@ -64,6 +65,7 @@ InstanceCLI.prototype.do_list = require('./do_list');
 InstanceCLI.prototype.do_get = require('./do_get');
 InstanceCLI.prototype.do_create = require('./do_create');
 InstanceCLI.prototype.do_delete = require('./do_delete');
+InstanceCLI.prototype.do_resize = require('./do_resize');
 InstanceCLI.prototype.do_rename = require('./do_rename');
 
 InstanceCLI.prototype.do_start = require('./do_start');
-- 
2.21.0

