commit d995e36d49642ee9807193486308a30f7d22e833 (refs/changes/44/3044/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2017-12-05T16:11:26-08:00 (1 year, 10 months ago)
    
    joyent/manta-thoth#166 Add commit hash to build info

diff --git a/tools/build-release.sh b/tools/build-release.sh
index d8fcd0d..f961a4f 100755
--- a/tools/build-release.sh
+++ b/tools/build-release.sh
@@ -10,13 +10,14 @@ set -o errexit
 node_ver='0.10.38'
 thoth_ver=$(json -f package.json version)
 gcc_ver=$(pkg_info -E gcc4? | cut -d - -f 1)
+thoth_commit=$(git rev-parse --short HEAD)
 uname=$(uname -s | tr "[:upper:]" "[:lower:]")
 case "$uname" in
     darwin) node_arch='x64';;
     sunos) node_arch='x86';;
     *) printf 'Sorry, not ready for you'; exit 1;;
 esac
-tar="thoth-${uname}-${thoth_ver}.tar.gz"
+tar="thoth-${uname}-${thoth_ver}-${thoth_commit}.tar.gz"
 node_dir="node-v${node_ver}-${uname}-${node_arch}"
 node_tar="${node_dir}.tar.gz"
 node_location="https://nodejs.org/download/release/v${node_ver}"
diff --git a/tools/publish.sh b/tools/publish.sh
index e4ba56b..09770cb 100755
--- a/tools/publish.sh
+++ b/tools/publish.sh
@@ -2,7 +2,8 @@
 
 ver="$1"
 uname=$(uname -s | tr "[:upper:]" "[:lower:]")
-name="thoth-${uname}-${ver:?}.tar.gz"
+commit=$(git rev-parse --short HEAD)
+name="thoth-${uname}-${ver:?}-${commit:?}.tar.gz"
 latest="thoth-${uname}-latest.tar.gz"
 
 mput -f "$name" "/thoth/public/$name"
