commit a4a38d29af153f6766d2575aa8518f737a456b12 (refs/changes/54/2354/1)
Author: Daniel Kontsek <daniel@kontsek.sk>
Date:   2017-08-09T13:24:49+02:00 (2 years, 2 months ago)
    
    Fixed JS linter issues

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 7e963133..92b4e1f5 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -1680,7 +1680,7 @@ function lookupConflicts(macs, ips, ipNics, vrids, vnc_port, log, callback) {
     var tracers_obj;
 
     load_fields = ['brand', 'state', 'nics', 'uuid', 'zonename', 'zone_state',
-                   'vnc_port'];
+        'vnc_port'];
     load_opts = {fields: load_fields, log: log};
 
     assert(log, 'no logger passed to lookupConflicts()');
@@ -1751,9 +1751,9 @@ function lookupConflicts(macs, ips, ipNics, vrids, vnc_port, log, callback) {
         }
 
         if (vm.vnc_port && vnc_port && vm.vnc_port === vnc_port) {
-          log.error('Found conflict: ' + vm.uuid + ' already has VNC port '
-              + vnc_port);
-          conflict = true;
+            log.error('Found conflict: ' + vm.uuid + ' already has VNC port '
+                + vnc_port);
+            conflict = true;
         }
 
         cb(conflict);
@@ -5010,7 +5010,7 @@ function checkPayloadProperties(payload, vmobj, log, callback)
 
     // Fetch vnc_port for uniqueness check in lookupConflicts()
     if (payload.hasOwnProperty('vnc_port') && payload.vnc_port > 0) {
-      vnc_port = payload.vnc_port;
+        vnc_port = payload.vnc_port;
     }
 
     props = ['zfs_root_recsize', 'zfs_data_recsize'];
