commit 15a769c7b3e66235ee6c03cc8fe1e822845faa8a (refs/changes/18/3018/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-11-30T21:25:24+01:00 (1 year, 10 months ago)
    
    HEAD-2378 Add dns_domain to sapi metadata

diff --git a/scripts/headnode.sh b/scripts/headnode.sh
index ce58f72b..06af40a8 100755
--- a/scripts/headnode.sh
+++ b/scripts/headnode.sh
@@ -638,6 +638,7 @@ upload_values
 download_metadata
 write_initial_config
 registrar_setup
+mdata-put dns_domain ${CONFIG_dns_domain}
 HERE
         setup_state_add "sapi_bootstrapped"
     fi
