commit 35d2eb341057cbc32c322b77ad2ddf8f70910b66 (refs/changes/18/3018/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-12-15T18:47:41+01:00 (1 year, 10 months ago)
    
    HEAD-2378 Add dns_domain to sapi metadata

diff --git a/scripts/headnode.sh b/scripts/headnode.sh
index ce58f72b..06af40a8 100755
--- a/scripts/headnode.sh
+++ b/scripts/headnode.sh
@@ -638,6 +638,7 @@ upload_values
 download_metadata
 write_initial_config
 registrar_setup
+mdata-put dns_domain ${CONFIG_dns_domain}
 HERE
         setup_state_add "sapi_bootstrapped"
     fi
