commit 7b7195eb63aef56b5d36bd9b19d034c9f74edf8d (refs/changes/98/2998/1)
Author: Jonathan Perkin <jperkin@joyent.com>
Date:   2017-11-27T15:22:37+00:00 (1 year, 10 months ago)
    
    OS-6469 CTF incorrect data object handling

diff --git a/usr/src/common/ctf/ctf_types.c b/usr/src/common/ctf/ctf_types.c
index 2ef4f42d6b..bb4cbdb724 100644
--- a/usr/src/common/ctf/ctf_types.c
+++ b/usr/src/common/ctf/ctf_types.c
@@ -973,6 +973,8 @@ ctf_object_iter(ctf_file_t *fp, ctf_object_f *func, void *arg)
 			continue;
 		id = *(ushort_t *)((uintptr_t)fp->ctf_buf +
 		    fp->ctf_sxlate[i]);
+		if (id == 0)
+			continue;
 
 		/*
 		 * Validate whether or not we're looking at a data object as
