From 27385a38a4578ad0b0247a76fc72a075c86b88cb Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Mon, 17 Oct 2016 14:34:10 +0200
Subject: [PATCH] TOOLS-1575 sdcadm platform install does not exit non-zero on
 failure

---
 lib/platform.js | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/lib/platform.js b/lib/platform.js
index 49e7804..a4298e7 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -734,18 +734,23 @@ Platform.prototype.install = function install(opts, callback) {
                 progress('After correcting above problem, rerun ' +
                          '`sdcadm platform install %s`.', filepath);
             }
-            return updateHistory(hist, callback);
+            updateHistory(hist, function (_histErr) {
+                callback(err);
+            });
+            return;
         }
 
         if (deleteOnFinish) {
-            return cleanup(callback);
+            cleanup(callback);
+            return;
         } else {
             progress('Platform image explicitly specified; ' +
                      'will not delete %s', filepath);
             progress('Installation complete');
         }
 
-        return updateHistory(hist, callback);
+        updateHistory(hist, callback);
+        return;
     });
 };
 
-- 
2.21.0

