commit 1a4e2a8581210b3b5b97425c9ff3c09206484088 (refs/changes/72/1372/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-01-30T22:09:06+00:00 (2 years, 8 months ago)
    
    joyent/mdb_v8#84 start working on v1.1.5

diff --git a/version b/version
index 2f1034b..5f84e78 100644
--- a/version
+++ b/version
@@ -8,4 +8,4 @@
 #
 # This value is also used by the "publish" target in the build process.
 #
-1.1.4
+1.1.5
