From 1a4e2a8581210b3b5b97425c9ff3c09206484088 Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Mon, 30 Jan 2017 22:08:39 +0000
Subject: [PATCH] joyent/mdb_v8#84 start working on v1.1.5

---
 version | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/version b/version
index 2f1034b..5f84e78 100644
--- a/version
+++ b/version
@@ -8,4 +8,4 @@
 #
 # This value is also used by the "publish" target in the build process.
 #
-1.1.4
+1.1.5
-- 
2.21.0

