From 5115d04b28b4066e0e02801bc96c9b8970ccfcce Mon Sep 17 00:00:00 2001
From: Kody Kantor <Kody.Kantor@gmail.com>
Date: Mon, 3 Apr 2017 16:26:33 -0500
Subject: [PATCH] MANTA-2980 electric-moray should report moray shard

---
 lib/moray.js | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/lib/moray.js b/lib/moray.js
index 43396de..b88927b 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -569,7 +569,7 @@ Moray.prototype.putMetadata = function putMetadata(options, callback) {
         headers: opts.headers
     }, 'Moray.putMetadata: entered');
     (function put() {
-        client.putObject(BUCKET, key, md, opts, function (err) {
+        client.putObject(BUCKET, key, md, opts, function (err, data) {
             if (err) {
                 log.debug({
                     err: err,
@@ -589,6 +589,9 @@ Moray.prototype.putMetadata = function putMetadata(options, callback) {
                     key: key,
                     requestId: opts.requestId
                 }, 'Moray.putMetadata: done');
+                if (data) { // add the shard information
+                    md._node = data._node;
+                }
                 callback(null, md);
             }
         });
-- 
2.21.0

