From 1ceb818368291359f621b99c540c4563bb8c0ed8 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 19 Oct 2016 15:13:53 -0700
Subject: [PATCH] IMGAPI-595 Update node version for CVE-2016-5180

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 061d9d1..b8cdf2e 100644
--- a/Makefile
+++ b/Makefile
@@ -33,7 +33,7 @@ SMF_MANIFESTS = $(shell ls smf/manifests/*.xml)
 NODEUNIT	:= ./node_modules/.bin/nodeunit
 CLEAN_FILES += ./node_modules
 
-NODE_PREBUILT_VERSION=v4.6.0
+NODE_PREBUILT_VERSION=v4.6.1
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
-- 
2.21.0

