commit ee456f59269a611450b84f25799322baedcb721e (refs/changes/52/5352/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-01-11T14:34:02+00:00 (9 months ago)
    
    9979 Support python3 for in-gate tools [fix broken build]

diff --git a/configure b/configure
index 15854252..baec0161 100755
--- a/configure
+++ b/configure
@@ -12,7 +12,7 @@
 #
 
 #
-# Copyright 2018 Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 shopt -s xpg_echo
 unalias -a
@@ -319,6 +319,7 @@ JAVA_ROOT=/opt/local/java/openjdk7;		export JAVA_ROOT
 FLEX=/opt/local/bin/flex;			export FLEX
 GNUXGETTEXT=/opt/local/bin/xgettext;		export GNUXGETTEXT
 PYTHON=/opt/local/bin/python2.7;		export PYTHON
+BUILDPY3='#';					export BUILDPY3
 #
 # Note that this isn't the strap perl: that only really exists for the benefit
 # of projects/illumos-extra/openssl
