From bee15b629949bcfa3e0cedcb357534ef10c75998 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Wed, 13 Feb 2019 17:11:32 -0700
Subject: [PATCH] MANTA-4117 Log files for pg_prefaulter should be stored in
 manta

---
 boot/setup.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index 926ae76..0abb9e0 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -168,6 +168,12 @@ function manta_setup_manatee_env {
     manta_add_logadm_entry "manatee-snapshotter"
     manta_add_logadm_entry "postgresql" "/var/pg"
     manta_add_logadm_entry "pgdump" "/var/log/manatee"
+    # The following entry does not use the manta_add_logadm_entry function
+    # because the underscore in the pg_prefaulter name breaks the upload script
+    logadm -w pg_prefaulter -C 48 -c -p 1h \
+        -t "/var/log/manta/upload/pg-prefaulter_\$nodename_%FT%H:00:00.log" \
+        "/var/svc/log/*pg_prefaulter*.log" || fatal \
+        "unable to create logadm entry"
 }
 
 
-- 
2.21.0

