From 33924cbbb914f4c52064fecdd549bed0342d263f Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 19 Jul 2019 10:00:39 -0700
Subject: [PATCH] TRITON-1824 make mockcloud and sdc-system-tests builds upload
 to public builds area

---
 Makefile | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 9004c06..90b533e 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2019, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 #
@@ -31,9 +31,6 @@ else
 	NPM	:= $(shell which npm)
 endif
 
-# sdc-system-tests is not a public module, so override
-ENGBLD_DEST_OUT_PATH ?= /stor/builds
-
 ENGBLD_REQUIRE := $(shell git submodule update --init deps/eng)
 include ./deps/eng/tools/mk/Makefile.defs
 TOP ?= $(error Unable to access eng.git submodule Makefiles.)
-- 
2.21.0

