commit 32bc8908b4d906179817af9a418d7718a2407d6c
Author: Jon Anderson <jon.anderson@joyent.com>
Date:   2019-03-15T17:48:55-04:00 (7 months ago)
    
    MANTA-4133 Invalid content-type resulted in 500 response

diff --git a/lib/server.js b/lib/server.js
index c23229d..a0535ca 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -16,7 +16,6 @@ var verror = require('verror');
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
-var mime = require('mime');
 var restify = require('restify');
 
 var audit = require('./audit');
@@ -42,7 +41,7 @@ require('./errors');
 var JOBS_ROOT_RE = /^\/([a-zA-Z][a-zA-Z0-9_\.@%]+)\/jobs\/?$/;
 /* JSSTYLED */
 var JOBS_LIVE_RE = /(state|status|name)=/i;
-
+var VALID_CONTENT_TYPE_RE = /^\w+\/.*/;
 
 ///--- Helpers
 
@@ -192,9 +191,13 @@ function createServer(options, clients, name) {
     server.pre(restify.pre.sanitizePath());
     server.pre(function cleanupContentType(req, res, next) {
         var ct = req.headers['content-type'];
-        /* JSSTYLED */
-        if (ct && !/.*\/.*/.test(ct))
-            req.headers['content-type'] = mime.lookup(ct);
+        /* content-type must have a type, '/' and optional sub-type, or return
+           an UnsupportedMediaType.
+           https://www.w3.org/Protocols/rfc1341/4_Content-Type.html
+        */
+        if (ct && !VALID_CONTENT_TYPE_RE.test(ct)) {
+            sendUnsupportedMediaTypeStatus(req, res, next);
+        }
         next();
     });
 
@@ -498,6 +501,12 @@ function forbiddenHandler(req, res, next) {
     next(false);
 }
 
+function sendUnsupportedMediaTypeStatus(req, res, next) {
+    req.log.debug('content-type: ' + req['content-type'] + ' unsupported');
+    res.send(415);
+    next(false);
+}
+
 
 
 /*
diff --git a/test/dir.test.js b/test/dir.test.js
index 1545716..7e4534b 100644
--- a/test/dir.test.js
+++ b/test/dir.test.js
@@ -141,7 +141,7 @@ test('mkdir, chattr: content-type (ignore)', function (t) {
     var k = this.key;
     var opts = {
         headers: {
-            'content-type': 'jpg'
+            'content-type': 'image/jpg'
         }
     };
     var self = this;
diff --git a/test/obj.test.js b/test/obj.test.js
index c4863a1..5bd1dd1 100644
--- a/test/obj.test.js
+++ b/test/obj.test.js
@@ -122,6 +122,23 @@ after(function (cb) {
 });
 
 
+test('MANTA-4133 (invalid content-type)', function (t) {
+    var opts = {
+        headers: {
+            'content-type': '/*'
+        }
+    };
+    var self = this;
+
+    this.putObject(t, function () {
+        self.client.chattr(self.key, opts, function (err, res) {
+            t.equal(err.name, 'UnsupportedMediaTypeError');
+            t.end();
+        });
+    });
+});
+
+
 test('put object', function (t) {
     var stream = new MemoryStream();
     var text = 'The lazy brown fox \nsomething \nsomething foo';
@@ -321,7 +338,7 @@ test('chattr: m- headers', function (t) {
 test('chattr: content-type', function (t) {
     var opts = {
         headers: {
-            'content-type': 'jpg'
+            'content-type': 'image/jpg'
         }
     };
     var self = this;
@@ -336,7 +353,7 @@ test('chattr: content-type', function (t) {
                 t.ok(info);
                 if (info) {
                     var h = info.headers || {};
-                    t.equal(h['content-type'], 'image/jpeg');
+                    t.equal(h['content-type'], 'image/jpg');
                 }
                 t.end();
             });
@@ -345,6 +362,7 @@ test('chattr: content-type', function (t) {
 });
 
 
+
 test('chattr: durability-level (not ok)', function (t) {
     var opts = {
         headers: {
