commit 5e84e46f52544f760ace92d073b857b36d4b0887
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-07-03T17:46:53+00:00 (3 months ago)
    
    OS-7879 zfs recv is broken

diff --git a/usr/src/lib/libzfs/common/libzfs_sendrecv.c b/usr/src/lib/libzfs/common/libzfs_sendrecv.c
index 27f05f101e..2bf7c160de 100644
--- a/usr/src/lib/libzfs/common/libzfs_sendrecv.c
+++ b/usr/src/lib/libzfs/common/libzfs_sendrecv.c
@@ -3754,7 +3754,6 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 		}
 
 		newfs = B_TRUE;
-		*cp = '/';
 	}
 
 	zc.zc_begin_record = *drr_noswap;
