From 67fa640800b91005b249eef6486f8eae0382f2b1 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Mon, 7 Nov 2016 14:10:31 -0800
Subject: [PATCH] MORAY-371 moray-test-suite should default to stock npm
 node-moray

---
 tools/configure | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/tools/configure b/tools/configure
index f7e39b8..044f114 100755
--- a/tools/configure
+++ b/tools/configure
@@ -37,10 +37,8 @@ var WORLD_EXECUTABLE = 0x1;
 
 /*
  * This is the "npm install" argument we use to get the "stock" Moray client.
- * TODO This should likely eventually become "moray" (i.e., the latest published
- * one), but we haven't published v2 yet, so we're using this.
  */
-var DFL_MORAY_CLIENT_TARGET = 'git://github.com/joyent/node-moray.git';
+var DFL_MORAY_CLIENT_TARGET = 'moray';
 var DFL_MORAY_SERVER_TARGET = 'git://github.com/joyent/moray.git';
 
 var usageMessage = [
-- 
2.21.0

