From e77a5fe6eec99139feb24956612cdc5214f53bca Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 19 Mar 2019 06:24:15 +0000
Subject: [PATCH] OS-7670 kvm: Remove unused thermal MSRs

---
 msr-index.h | 14 --------------
 1 file changed, 14 deletions(-)

diff --git a/msr-index.h b/msr-index.h
index bea940f..f059650 100644
--- a/msr-index.h
+++ b/msr-index.h
@@ -240,20 +240,6 @@
 #define MSR_IA32_MPERF			0x000000e7
 #define MSR_IA32_APERF			0x000000e8
 
-#define MSR_IA32_THERM_CONTROL		0x0000019a
-#define MSR_IA32_THERM_INTERRUPT	0x0000019b
-
-#define THERM_INT_LOW_ENABLE		(1 << 0)
-#define THERM_INT_HIGH_ENABLE		(1 << 1)
-
-#define MSR_IA32_THERM_STATUS		0x0000019c
-
-#define THERM_STATUS_PROCHOT		(1 << 0)
-
-#define MSR_THERM2_CTL			0x0000019d
-
-#define MSR_THERM2_CTL_TM_SELECT	(1ULL << 16)
-
 #define MSR_IA32_MISC_ENABLE		0x000001a0
 
 /* MISC_ENABLE bits: architectural */
-- 
2.21.0

