From b9fbdb3ed61f6831e5f0279dcbd7c844943935b7 Mon Sep 17 00:00:00 2001
From: Peter Marton <pmarton@netflix.com>
Date: Mon, 31 Dec 2018 10:24:57 -0800
Subject: [PATCH] joyent/node-verror#63 do not call sprintf with one argument

---
 lib/verror.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/verror.js b/lib/verror.js
index 8663dde..45e4ac1 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -114,8 +114,13 @@ function parseConstructorArguments(args)
 		});
 	}
 
+	/**
+	 * Do not call sprintf with one argument, performance optimization
+	 */
 	if (sprintf_args.length === 0) {
 		shortmessage = '';
+	} else if (sprintf_args.length === 1) {
+		shortmessage = sprintf_args[0];
 	} else {
 		shortmessage = sprintf.apply(null, sprintf_args);
 	}
-- 
2.21.0

