From dc853cdd2fb08808d168a1ef73fd214598070815 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 5 Mar 2019 14:40:09 -0500
Subject: [PATCH] TOOLS-2186 sdcnode updates for Feburary 2019 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index feae48b..568998c 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS    = -t 4 -o doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/config-agent.xml.in
 
-NODE_PREBUILT_VERSION=v6.15.1
+NODE_PREBUILT_VERSION=v6.17.0
 ifeq ($(shell uname -s),SunOS)
 	# config-agent runs in zones *and* in the GZ, so we need to make sure we use
 	# a node runtime that is able to run in the GZ (those runtimes can run in
-- 
2.21.0

