commit 793202faa1952b963ca95d1bd1be9fa8455cc7b2 (refs/changes/58/3758/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-04-02T16:15:06-07:00 (1 year, 6 months ago)
    
    TRITON-288 sdcnode updates for March 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 34a6e1f..2682ab3 100644
--- a/Makefile
+++ b/Makefile
@@ -48,7 +48,7 @@ JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS    = -o indent=4,doxygen,unparenthesized-return=0
 SMF_MANIFESTS_IN = smf/manifests/wf-api.xml.in smf/manifests/wf-runner.xml.in smf/manifests/wf-backfill.xml.in
 
-NODE_PREBUILT_VERSION=v4.8.7
+NODE_PREBUILT_VERSION=v4.9.0
 
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
