commit 1b05014bd93863eebe4bc37f98e5747521370c81 (refs/changes/27/5627/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-02-21T23:58:32+00:00 (8 months ago)
    
    OS-7592 viona doesn't zero ULP checksum for HCK_FULLCKSUM

diff --git a/usr/src/uts/i86pc/io/viona/viona.c b/usr/src/uts/i86pc/io/viona/viona.c
index 4ae0f7317f..7fe4943a7e 100644
--- a/usr/src/uts/i86pc/io/viona/viona.c
+++ b/usr/src/uts/i86pc/io/viona/viona.c
@@ -2661,6 +2661,9 @@ viona_tx_csum(viona_vring_t *ring, const struct virtio_net_hdr *hdr,
 	if (ftype == ETHERTYPE_IP) {
 		if ((link->l_cap_csum & HCKSUM_INET_FULL_V4) != 0 &&
 		    (ipproto == IPPROTO_TCP || ipproto == IPPROTO_UDP)) {
+			uint16_t *csump = (uint16_t *)(mp->b_rptr +
+			    hdr->vrh_csum_start + hdr->vrh_csum_offset);
+			*csump = 0;
 			flags |= HCK_FULLCKSUM;
 			mac_hcksum_set(mp, 0, 0, 0, 0, flags);
 			return (B_TRUE);
@@ -2673,6 +2676,9 @@ viona_tx_csum(viona_vring_t *ring, const struct virtio_net_hdr *hdr,
 	} else if (ftype == ETHERTYPE_IPV6) {
 		if ((link->l_cap_csum & HCKSUM_INET_FULL_V6) != 0 &&
 		    (ipproto == IPPROTO_TCP || ipproto == IPPROTO_UDP)) {
+			uint16_t *csump = (uint16_t *)(mp->b_rptr +
+			    hdr->vrh_csum_start + hdr->vrh_csum_offset);
+			*csump = 0;
 			flags |= HCK_FULLCKSUM;
 			mac_hcksum_set(mp, 0, 0, 0, 0, flags);
 			return (B_TRUE);
