From b7272e9a8d0bc947e9e85b9ebae9237a8024a55d Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 24 Apr 2017 12:21:15 +0000
Subject: [PATCH] 8088 Add support for LSI Intruder and Cutlass cards.

---
 overlay/generic/etc/driver_aliases | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 75896e89..187fb9a1 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -920,11 +920,15 @@ mpt_sas "pciex1000,94"
 mpt_sas "pciex1000,95"
 mpt_sas "pciex1000,96"
 mpt_sas "pciex1000,97"
+mr_sas "pciex1000,52"
+mr_sas "pciex1000,53"
 mr_sas "pciex1000,5b"
 mr_sas "pciex1000,5d"
 mr_sas "pciex1000,5f"
 mr_sas "pciex1000,71"
 mr_sas "pciex1000,73"
+mr_sas "pciex1000,ce"
+mr_sas "pciex1000,cf"
 mwl "pci11ab,2a0a"
 mwl "pci11ab,2a24"
 mxfe "pci10d9,512"
-- 
2.21.0

