commit b46407cada39ee777bad6b9f3cc95fcfe79a1c0f (refs/changes/10/3610/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-03-13T13:20:49-04:00 (1 year, 7 months ago)
    
    TRITON-240 zoneevent should be called with an identifier string

diff --git a/lib/vm.js b/lib/vm.js
index 5715c40..76f5afa 100644
--- a/lib/vm.js
+++ b/lib/vm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -204,7 +204,7 @@ EventWatcher.prototype.startZoneWatcher = function () {
 
     var watcher = this.zoneeventWatcher = spawn(
         '/usr/vm/sbin/zoneevent',
-        [],
+        ['net-agent'],
         {'customFds': [-1, -1, -1]});
 
     log.info('zoneevent running with pid ' + watcher.pid);
