From b46407cada39ee777bad6b9f3cc95fcfe79a1c0f Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Tue, 13 Mar 2018 13:20:49 -0400
Subject: [PATCH] TRITON-240 zoneevent should be called with an identifier
 string

---
 lib/vm.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/vm.js b/lib/vm.js
index 5715c40..76f5afa 100644
--- a/lib/vm.js
+++ b/lib/vm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -204,7 +204,7 @@ EventWatcher.prototype.startZoneWatcher = function () {
 
     var watcher = this.zoneeventWatcher = spawn(
         '/usr/vm/sbin/zoneevent',
-        [],
+        ['net-agent'],
         {'customFds': [-1, -1, -1]});
 
     log.info('zoneevent running with pid ' + watcher.pid);
-- 
2.21.0

