commit c81e016ce01a4154d600c4caff607b33e65ca63e (refs/changes/43/5143/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-11-27T18:08:20+01:00 (11 months ago)
    
    bhyve: fix vm_suspend_cpu/vm_resume_cpu

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index 04f5409afa..adb09149e5 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -290,6 +290,24 @@ vmmdev_do_ioctl(vmm_softc_t *sc, int cmd, intptr_t arg, int md,
 	 * Some VMM ioctls can operate only on vcpus that are not running.
 	 */
 	switch (cmd) {
+	case VM_SUSPEND_CPU:
+	case VM_RESUME_CPU:
+		/*
+		 * Copy in the ID of the vCPU chosen for this operation.
+		 * Since a nefarious caller could update their struct between
+		 * this locking and when the rest of the ioctl data is copied
+		 * in, it is _critical_ that this local 'vcpu' variable be used
+		 * rather than the in-struct one when performing the ioctl.
+		 */
+		if (ddi_copyin(datap, &vcpu, sizeof (vcpu), md)) {
+			return (EFAULT);
+		}
+		if (vcpu < -1 || vcpu >= VM_MAXCPU) {
+			error = EINVAL;
+			goto done;
+		}
+		break;
+
 	case VM_RUN:
 	case VM_GET_REGISTER:
 	case VM_SET_REGISTER:
