commit fa909a7232607d3c65566c805df58c8f0528bca9 (refs/changes/81/181/1)
Author: Matt Smillie <matt.smillie@joyent.com>
Date:   2016-07-31T17:33:13-07:00 (3 years, 2 months ago)
    
    DOCKER-901: `docker -p` crashes sdc-docker without farics enabled.

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index ec5d46e..7dce192 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -1961,7 +1961,7 @@ function buildVmPayload(opts, container, callback) {
         function addExternalNet(_, cb) {
             assert.object(opts.config, 'config');
             assert.object(opts.config.overlay, 'config.overlay');
-            assert.object(opts.config.overlay.enabled,
+            assert.bool(opts.config.overlay.enabled,
                 'config.overlay.enabled');
             if (opts.config.overlay.enabled
                 && (publishingPorts(container) || opts.fabricRequireExternal)) {
