commit 6395a8e09b24dcc549145eca8a1d02119cce199e (refs/changes/68/3668/1)
Author: Mike Zeller <mike@mikezeller.net>
Date:   2018-03-16T22:00:33+00:00 (1 year, 7 months ago)
    
    overlay_target_entry needs a spot for vlan

diff --git a/usr/src/uts/common/sys/overlay_impl.h b/usr/src/uts/common/sys/overlay_impl.h
index 5645cbfbb0..7d42bfca19 100644
--- a/usr/src/uts/common/sys/overlay_impl.h
+++ b/usr/src/uts/common/sys/overlay_impl.h
@@ -169,6 +169,7 @@ typedef struct overlay_target_entry {
 	list_node_t		ote_qlink;
 	overlay_target_entry_flags_t ote_flags;	/* RW: state flags */
 	uint32_t		ote_dcid;
+	uint16_t		ote_vlan;	/* RO: VL3 vlan id */
 	uint8_t			ote_addr[ETHERADDRL];	/* RO: mac addr */
 	struct in6_addr		ote_ip;		/* RO: VL3 IP */
 	overlay_target_t	*ote_ott;	/* RO */
