commit 5be6ebb83e5c523952a05ad5446850f24600b0cf (refs/changes/18/4118/1)
Author: sjorge <sjorge@blackdot.be>
Date:   2018-06-02T10:19:45+02:00 (1 year, 4 months ago)
    
    OS-6984 bhyve bootrom should be configurable via vmadm

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index a475ba9f..5ee8e9a1 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -6243,6 +6243,7 @@ function buildZonecfgUpdate(vmobj, payload, log)
         }
         setAttr('com1', 'com1');
         setAttr('com2', 'com2');
+        setAttr('bootrom', 'bootrom');
 
         setAttr('bhyve-extra-opts', 'bhyve_extra_opts');
 
