From 193ea85464ac040391cfd32e305aa49ffdf75577 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 12 Dec 2018 16:51:50 +0000
Subject: [PATCH] TOOLS-2133 MG needs to look for .tar.gz agent tarballs after
 TOOLS-2063

---
 Makefile        |  12 ++--
 targets.json.in | 176 ++++++++++++++++++++++++------------------------
 2 files changed, 94 insertions(+), 94 deletions(-)

diff --git a/Makefile b/Makefile
index 7042bf7..f603348 100644
--- a/Makefile
+++ b/Makefile
@@ -126,7 +126,7 @@ clean_smartlogin:
 #---- amon
 
 _amon_stamp=$(SDC_AMON_BRANCH)-$(TIMESTAMP)-g$(SDC_AMON_SHA)
-AMON_BITS=$(BITS_DIR)/amon/amon-pkg-$(_amon_stamp).tar.bz2 \
+AMON_BITS=$(BITS_DIR)/amon/amon-pkg-$(_amon_stamp).tar.gz \
 	$(BITS_DIR)/amon/amon-relay-$(_amon_stamp).tgz \
 	$(BITS_DIR)/amon/amon-agent-$(_amon_stamp).tgz
 AMON_BITS_0=$(shell echo $(AMON_BITS) | awk '{print $$1}')
@@ -1285,7 +1285,7 @@ clean_cn_agent:
 #---- Configuration Agent
 
 _config_agent_stamp=$(SDC_CONFIG_AGENT_BRANCH)-$(TIMESTAMP)-g$(SDC_CONFIG_AGENT_SHA)
-CONFIG_AGENT_BIT=$(BITS_DIR)/config-agent/config-agent-pkg-$(_config_agent_stamp).tar.bz2
+CONFIG_AGENT_BIT=$(BITS_DIR)/config-agent/config-agent-pkg-$(_config_agent_stamp).tar.gz
 CONFIG_AGENT_MANIFEST_BIT=$(BITS_DIR)/config-agent/config-agent-pkg-$(_config_agent_stamp).manifest
 
 .PHONY: config-agent
@@ -2302,7 +2302,7 @@ clean_wrasse:
 #---- Registrar
 
 _registrar_stamp=$(REGISTRAR_BRANCH)-$(TIMESTAMP)-g$(REGISTRAR_SHA)
-REGISTRAR_BITS=$(BITS_DIR)/registrar/registrar-pkg-$(_registrar_stamp).tar.bz2
+REGISTRAR_BITS=$(BITS_DIR)/registrar/registrar-pkg-$(_registrar_stamp).tar.gz
 
 .PHONY: registrar
 registrar: $(REGISTRAR_BITS)
@@ -2323,7 +2323,7 @@ clean_registrar:
 #---- waferlock
 
 _waferlock_stamp=$(WAFERLOCK_BRANCH)-$(TIMESTAMP)-g$(WAFERLOCK_SHA)
-WAFERLOCK_BITS=$(BITS_DIR)/waferlock/waferlock-pkg-$(_waferlock_stamp).tar.bz2
+WAFERLOCK_BITS=$(BITS_DIR)/waferlock/waferlock-pkg-$(_waferlock_stamp).tar.gz
 
 .PHONY: waferlock
 waferlock: $(WAFERLOCK_BITS)
@@ -2343,7 +2343,7 @@ clean_waferlock:
 #---- mackerel
 
 _mackerel_stamp=$(MANTA_MACKEREL_BRANCH)-$(TIMESTAMP)-g$(MANTA_MACKEREL_SHA)
-MACKEREL_BITS=$(BITS_DIR)/mackerel/mackerel-pkg-$(_mackerel_stamp).tar.bz2
+MACKEREL_BITS=$(BITS_DIR)/mackerel/mackerel-pkg-$(_mackerel_stamp).tar.gz
 
 .PHONY: mackerel
 mackerel: $(MACKEREL_BITS)
@@ -2445,7 +2445,7 @@ clean_muppet:
 #---- Minnow
 
 _minnow_stamp=$(MANTA_MINNOW_BRANCH)-$(TIMESTAMP)-g$(MANTA_MINNOW_SHA)
-MINNOW_BITS=$(BITS_DIR)/minnow/minnow-pkg-$(_minnow_stamp).tar.bz2
+MINNOW_BITS=$(BITS_DIR)/minnow/minnow-pkg-$(_minnow_stamp).tar.gz
 
 .PHONY: minnow
 minnow: $(MINNOW_BITS)
diff --git a/targets.json.in b/targets.json.in
index 2e41d6d..880ad0c 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -125,9 +125,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "amon": {
@@ -148,9 +148,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "assets": {
@@ -196,9 +196,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "adminui": {
@@ -221,9 +221,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "mockcloud": {
@@ -283,9 +283,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "amonredis": {
@@ -307,9 +307,9 @@ cat <<EOF
       "registrar"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "rabbitmq": {
@@ -339,9 +339,9 @@ cat <<EOF
       "registrar"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "cloudapi": {
@@ -368,9 +368,9 @@ cat <<EOF
       "haproxy-1.6.2"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "nat": {
@@ -408,9 +408,9 @@ cat <<EOF
     ],
     "pkgsrc": [],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "volapi": {
@@ -433,9 +433,9 @@ cat <<EOF
     ],
     "pkgsrc": [],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
   "portolan": {
@@ -457,9 +457,9 @@ cat <<EOF
     ],
     "pkgsrc": [],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -485,9 +485,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -513,9 +513,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -537,8 +537,8 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -572,9 +572,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -598,9 +598,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -631,9 +631,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -658,7 +658,7 @@ cat <<EOF
     ],
     "tarballs": [
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -759,9 +759,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -785,9 +785,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -810,9 +810,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -836,9 +836,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -908,9 +908,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -936,10 +936,10 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
-      {"tarball": "waferlock/waferlock-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
+      {"tarball": "waferlock/waferlock-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -965,10 +965,10 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
-      {"tarball": "waferlock/waferlock-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
+      {"tarball": "waferlock/waferlock-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -992,9 +992,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1019,9 +1019,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1047,9 +1047,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1074,9 +1074,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1103,9 +1103,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1130,9 +1130,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1158,10 +1158,10 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
-      {"tarball": "mackerel/mackerel-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
+      {"tarball": "mackerel/mackerel-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1187,9 +1187,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1212,9 +1212,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1242,10 +1242,10 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "minnow/minnow-pkg-*.tar.bz2", "sysroot": "/"},
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "minnow/minnow-pkg-*.tar.gz", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1269,9 +1269,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1295,9 +1295,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1323,9 +1323,9 @@ cat <<EOF
       "config-agent"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1360,9 +1360,9 @@ cat <<EOF
       "https://download.joyent.com/pub/build/sdcnode"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1418,9 +1418,9 @@ cat <<EOF
       "registrar"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1446,8 +1446,8 @@ cat <<EOF
     ],
     "tarballs": [
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"},
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"}
     ]
   },
 
@@ -1468,9 +1468,9 @@ cat <<EOF
         "registrar"
     ],
     "tarballs": [
-      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "registrar/registrar-pkg-*.tar.gz", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
@@ -1496,7 +1496,7 @@ cat <<EOF
     ],
     "tarballs": [
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
-      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
+      {"tarball": "config-agent/config-agent-*.tar.gz", "sysroot": "/opt/smartdc"}
     ]
   },
 
-- 
2.21.0

