commit 70f2d8d9fcd5cb05b5e7f6e24280433149b9bed7
Author: Trent Mick <trentm@gmail.com>
Date:   2019-03-12T19:44:54-07:00 (7 months ago)
    
    TRITON-884 remove cloud analytics (ca) from Triton

diff --git a/CHANGES.md b/CHANGES.md
index fd420d3..fdcda9c 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -6,7 +6,11 @@ Known issues:
 
 ## not yet released
 
-(nothing)
+## 7.0.1
+
+- [TRITON-884] Drop support for Cloud Analytics-related RBAC "resources"
+  for role-tags. Cloud Analytics has been removed from Triton. See
+  [TRITON-884](https://smartos.org/bugview/TRITON-884) for details.
 
 ## 7.0.0
 
diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index 512e65e..b5fb949 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -2838,9 +2838,7 @@ CloudApi.prototype.getRoleTags = function getRoleTags(opts, cb) {
         'policies',
         // TODO: validate, test
         'keys',
-        'datacenters',
-        'analytics',
-        'instrumentations'
+        'datacenters'
     ];
     var parts = opts.resource.split('/');
     if (validResources.indexOf(parts[2]) === -1) {
@@ -2916,9 +2914,7 @@ CloudApi.prototype.setRoleTags = function setRoleTags(opts, cb) {
         'policies',
         // TODO: validate, test
         'keys',
-        'datacenters',
-        'analytics',
-        'instrumentations'
+        'datacenters'
     ];
     var parts = opts.resource.split('/');
     if (validResources.indexOf(parts[2]) === -1) {
diff --git a/package.json b/package.json
index bc3c39b..499091a 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "triton",
   "description": "Joyent Triton CLI and client (https://www.joyent.com/triton)",
-  "version": "7.0.0",
+  "version": "7.0.1",
   "author": "Joyent (joyent.com)",
   "homepage": "https://github.com/joyent/node-triton",
   "dependencies": {
