From 376fd6a7e0b69e6678372a3e41ca1872cc11a0c2 Mon Sep 17 00:00:00 2001
From: Bryan Cantrill <bryan@joyent.com>
Date: Sat, 5 Nov 2016 05:45:34 +0000
Subject: [PATCH] TOOLS-1602 v8plus dies when C constructors generate
 exceptions

---
 v8plus_objectwrap.cc | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/v8plus_objectwrap.cc b/v8plus_objectwrap.cc
index eadd345..62cd05a 100644
--- a/v8plus_objectwrap.cc
+++ b/v8plus_objectwrap.cc
@@ -234,6 +234,16 @@ v8plus::ObjectWrap::~ObjectWrap()
 	v8plus_module_defn_t *mdp =
 	    reinterpret_cast<v8plus_module_defn_t *>(_defn);
 
+	if (_c_impl == NULL) {
+		/*
+		 * If the pointer is NULL, it means we are in the ObjectWrap
+		 * destructor after the C constructor indicated an exception.
+		 * In this case, we don't call the C destructor -- and indeed
+		 * _defn is uninitialized so we must not attempt to use it.
+		 */
+		return;
+	}
+
 	mdp->vmd_dtor(_c_impl);
 	(void) _objhash.erase(_c_impl);
 }
-- 
2.21.0

