commit 12c5f124b020bb8a5ac04b48e09e2b293053c8ce (refs/changes/28/628/1)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2016-10-07T16:57:39-07:00 (3 years ago)
    
    AGENT-1039 sdc-vm-agent should use node 4.x

diff --git a/Makefile b/Makefile
index 2fa3a1c..f52b705 100644
--- a/Makefile
+++ b/Makefile
@@ -33,10 +33,10 @@ JSSTYLE_FLAGS =
 
 # Should be the same version as the platform's /usr/node/bin/node.
 NODE_PREBUILT_TAG =	gz
-NODE_PREBUILT_VERSION =	v0.12.9
+NODE_PREBUILT_VERSION =	v4.6.0
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-smartos/1.6.3.
-	NODE_PREBUILT_IMAGE =	fd2cc906-8938-11e3-beab-4359c665ac99
+	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
 endif
 
 # Included definitions
diff --git a/npm/preinstall.sh b/npm/preinstall.sh
new file mode 100755
index 0000000..6098ba9
--- /dev/null
+++ b/npm/preinstall.sh
@@ -0,0 +1,24 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright 2016 Joyent, Inc.
+#
+
+if [[ "${SDC_AGENT_SKIP_LIFECYCLE:-no}" = "yes" ]]; then
+    printf 'Running during package build; skipping lifecycle script.\n' >&2
+    exit 0
+fi
+
+# min-platform is now 20141030T081701Z, so refuse to install if someone is
+# trying to load on something older.
+if [[ $(uname -v | cut -d '_' -f2 | tr -d '[A-Z]') -lt 20141030081701 ]]; then
+    echo "FATAL: min-platform for this agent is 20141030T081701Z"
+    exit 2
+fi
+
+exit 0
diff --git a/package.json b/package.json
index 4625d1f..6e11445 100644
--- a/package.json
+++ b/package.json
@@ -22,6 +22,7 @@
         "config-agent": ">=1.3.0"
     },
     "scripts": {
+        "preinstall": "npm/preinstall.sh",
         "postinstall": "npm/postinstall.sh",
         "postuninstall": "npm/postuninstall.sh"
     },
