From ed46c6bb86277682a8360219921444351040e852 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 25 Oct 2016 12:51:29 +0200
Subject: [PATCH] TOOLS-1580 `sdcadm create binder` should refuse to work

---
 lib/procedures/index.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/procedures/index.js b/lib/procedures/index.js
index 89d357b..ddd8ea7 100644
--- a/lib/procedures/index.js
+++ b/lib/procedures/index.js
@@ -607,7 +607,7 @@ function coordinatePlan(opts, cb) {
         function createServiceInstance(_, next) {
             // Any instance which should never be created using
             // this tool should be here:
-            var avoid = ['zookeeper', 'manatee', 'rabbitmq', 'sdc'];
+            var avoid = ['binder', 'manatee', 'rabbitmq', 'sdc', 'zookeeper'];
             // Anything which can be created w/o --skip-ha-check flag should
             // be on this list:
             var allow = ['moray', 'workflow', 'cloudapi', 'mahi', 'nat',
-- 
2.21.0

