commit b14eab09a02aa518b6ea4c3997fce96e133f5e24 (refs/changes/01/1401/1)
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2017-02-01T23:45:10-08:00 (2 years, 8 months ago)
    
    HEAD-2346 remove vestigial redis zone

diff --git a/build.spec b/build.spec
index f3cfae13..760cc5aa 100644
--- a/build.spec
+++ b/build.spec
@@ -49,7 +49,6 @@
     "napi": {},
     "papi": {},
     "rabbitmq": {},
-    "redis": {},
     "sapi": {},
     "sdc": {},
     "ufds": {},
diff --git a/buildtools/novus/cmd/checker.js b/buildtools/novus/cmd/checker.js
index f93e0e2d..23ddc32c 100644
--- a/buildtools/novus/cmd/checker.js
+++ b/buildtools/novus/cmd/checker.js
@@ -6,7 +6,7 @@
  */
 
 /*
- * Copyright 2015 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var mod_fs = require('fs');
@@ -225,7 +225,6 @@ check_old_image_specs()
 		'napi',
 		'papi',
 		'rabbitmq',
-		'redis',
 		'sapi',
 		'sdc',
 		'ufds',
diff --git a/config/sapi/services/redis/service.json b/config/sapi/services/redis/service.json
deleted file mode 100644
index 3ee9f93b..00000000
--- a/config/sapi/services/redis/service.json
+++ /dev/null
@@ -1,20 +0,0 @@
-{
-    "name": "redis",
-    "params": {
-        "archive_on_delete": true,
-        "delegate_dataset": true,
-        "package_name": "sdc_1024",
-        "image_uuid": "IMAGE_UUID",
-        "maintain_resolvers": true,
-        "networks": [ "admin" ],
-        "tags": {
-            "smartdc_role": "redis",
-            "smartdc_type": "core"
-        }
-    },
-    "metadata": {
-        "SERVICE_NAME": "redis"
-    },
-    "manifests": {
-    }
-}
diff --git a/scripts/headnode.sh b/scripts/headnode.sh
index 196da9ef..0c881a83 100755
--- a/scripts/headnode.sh
+++ b/scripts/headnode.sh
@@ -805,7 +805,6 @@ if setup_state_not_seen "setup_complete" \
     create_zone manatee
     create_zone moray
     create_zone amonredis
-    create_zone redis
     create_zone ufds
     create_zone workflow
     create_zone amon
diff --git a/scripts/prompt-config.sh b/scripts/prompt-config.sh
index b99bd5b3..bed13be2 100755
--- a/scripts/prompt-config.sh
+++ b/scripts/prompt-config.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 # XXX - TODO
@@ -1326,10 +1326,6 @@ next_addr=$(($next_addr + 1))
 num_to_ip $next_addr
 amonredis_admin_ip="$ip_addr"
 
-next_addr=$(($next_addr + 1))
-num_to_ip $next_addr
-redis_admin_ip="$ip_addr"
-
 next_addr=$(($next_addr + 1))
 num_to_ip $next_addr
 amon_admin_ip="$ip_addr"
@@ -1594,11 +1590,6 @@ echo "amonredis_root_pw='$escaped_zone_admin_pw'" >>$tmp_config
 echo "amonredis_domain=amonredis.${datacenter_name}.${dns_domain}" >>$tmp_config
 echo >>$tmp_config
 
-echo "redis_admin_ips=$redis_admin_ip" >>$tmp_config
-echo "redis_root_pw='$escaped_zone_admin_pw'" >>$tmp_config
-echo "redis_domain=redis.${datacenter_name}.${dns_domain}" >>$tmp_config
-echo >>$tmp_config
-
 # NOTE: we add admin_ip and admin_ips here because some stuff is hardcoded to
 #       use admin_ip.  When this is cleaned up we can just keep ips.
 echo "assets_admin_ip=$assets_admin_ip" >>$tmp_config
diff --git a/tools/bin/sdc-restore b/tools/bin/sdc-restore
index 8d5550be..4fe245a6 100755
--- a/tools/bin/sdc-restore
+++ b/tools/bin/sdc-restore
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -42,7 +42,7 @@ echo "(debug log is /tmp/restorelog.$$)"
 
 HEADNODE_UUID=$(sysinfo | json UUID)
 
-core_zones="assets binder manatee moray redis ufds workflow amon napi"
+core_zones="assets binder manatee moray ufds workflow amon napi"
 core_zones="$core_zones rabbitmq imgapi cnapi dhcpd fwapi vmapi ca"
 core_zones="$core_zones adminui sapi"
 
diff --git a/zones/redis/backup b/zones/redis/backup
deleted file mode 100755
index b08dd943..00000000
--- a/zones/redis/backup
+++ /dev/null
@@ -1,58 +0,0 @@
-#!/usr/bin/bash
-#
-# This Source Code Form is subject to the terms of the Mozilla Public
-# License, v. 2.0. If a copy of the MPL was not distributed with this
-# file, You can obtain one at http://mozilla.org/MPL/2.0/.
-#
-
-#
-# Copyright (c) 2014, Joyent, Inc.
-#
-
-PATH=/usr/bin:/usr/sbin
-export PATH
-
-if [[ $# != 2 ]]; then
-    echo "Usage: $0 <zone_name> <target_directory>"
-    exit 1
-fi
-
-ZONE=$1
-TARGET_DIR=$2
-ROLE="redis"
-
-STAMP=$(date +'%F-%H-%M-%S-%Z')
-BACKUP_VERSION="${ROLE}-${STAMP}"
-
-if [[ ! -e "${TARGET_DIR}" ]]; then
-    mkdir -p ${TARGET_DIR}
-fi
-
-DATA_DATASET=$(zfs list -H -o name|grep "${ZONE}/data$")
-if [[ -z $DATA_DATASET ]]; then
-    echo "FATAL: Cannot find '${ROLE}' data dataset"
-    exit 103
-fi
-
-echo "==> Creating snapshot of '${ZONE}/data' dataset"
-zfs snapshot zones/${ZONE}/data@${BACKUP_VERSION} 2>&1
-if [[ $? -gt 0 ]]; then
-    echo "FATAL: Unable to snapshot data dataset"
-    exit 106
-fi
-
-echo "==> Creating backup directory '${TARGET_DIR}/${ROLE}'"
-mkdir -p "${TARGET_DIR}/${ROLE}"
-
-echo "==> Saving data dataset"
-zfs send "zones/${ZONE}/data@${BACKUP_VERSION}" \
-    > "${TARGET_DIR}/${ROLE}/${ROLE}-data.zfs" 2>&1
-if [[ $? -gt 0 ]]; then
-    echo "Unable to zfs send data dataset"
-    exit 108
-fi
-
-echo "==> Removing temporary snapshot of '${ZONE}'"
-zfs destroy "zones/${ZONE}/data@${BACKUP_VERSION}"
-
-exit 0
diff --git a/zones/redis/restore b/zones/redis/restore
deleted file mode 100755
index 293f46fe..00000000
--- a/zones/redis/restore
+++ /dev/null
@@ -1,69 +0,0 @@
-#!/usr/bin/bash
-#
-# This Source Code Form is subject to the terms of the Mozilla Public
-# License, v. 2.0. If a copy of the MPL was not distributed with this
-# file, You can obtain one at http://mozilla.org/MPL/2.0/.
-#
-
-#
-# Copyright (c) 2014, Joyent, Inc.
-#
-
-PATH=/usr/bin:/usr/sbin
-export PATH
-
-if [[ $# != 2 ]]; then
-    echo "Usage: $0 <zone_name> <target_directory>"
-    exit 1
-fi
-
-ZONE=$1
-TARGET_DIR=$2
-ROLE="redis"
-
-# Don't want redis running while swapping its data.
-svcadm disable redis
-
-# Destroy previous data dataset.
-DATA_DATASET=$(zfs list -H -o name|grep "${ZONE}/data$")
-if [[ -z "${DATA_DATASET}" ]]; then
-    echo "FATAL: Missing '${DATA_DATASET}' dataset"
-    exit 105
-fi
-echo "==> Destroying dataset '${DATA_DATASET}'"
-zfs destroy -r "${DATA_DATASET}"
-if [[ $? -gt 0 ]]; then
-    echo "FATAL: Unable to zfs destroy '${DATA_DATASET}' dataset"
-    exit 106
-fi
-
-echo "==> Receiving '${TARGET_DIR}/${ROLE}/${ROLE}-data.zfs'"
-zfs receive -v "${DATA_DATASET}" < "${TARGET_DIR}/${ROLE}/${ROLE}-data.zfs"
-if [[ $? -gt 0 ]]; then
-    echo "FATAL: Unable to zfs receive data dataset"
-    exit 108
-fi
-
-svcadm enable redis
-
-
-# Reboot the zone to set mountpoint for the new dataset properly. Then
-# double check.
-echo "==> Booting '${ZONE}' zone"
-/usr/sbin/zoneadm -z ${ZONE} boot
-if [[ "$(zfs get -H -o value mountpoint ${DATA_DATASET})" != "/data" ]]; then
-    echo "==> Setting mountpoint for dataset '${DATA_DATASET}'"
-    zlogin ${ZONE} /usr/sbin/zfs set mountpoint=/data "${DATA_DATASET}"
-    if [[ $? -gt 0 ]]; then
-        echo "FATAL: Unable to set mountpoint for data dataset into '${ZONE}' zone"
-        exit 112
-    fi
-fi
-echo "==> Waiting for 10 seconds while the zone services are running"
-sleep 10
-
-
-echo "==> Halting '${ZONE}' zone"
-/usr/sbin/zoneadm -z ${ZONE} halt
-
-exit 0
