commit 90d063e5545c4d91362bbbe0bd2cb977913435a1
Author: Rui Loura <rui@joyent.com>
Date:   2019-03-15T18:34:25+00:00 (7 months ago)
    
    TRITON-815 triton-cns needs to be rack aware

diff --git a/lib/api-server.js b/lib/api-server.js
index e8ea993..6de31c6 100644
--- a/lib/api-server.js
+++ b/lib/api-server.js
@@ -26,6 +26,7 @@ var fs = require('fs');
 var cueball = require('cueball');
 var getSuffixes = require('./vm-to-zones').getSuffixes;
 var createUfdsPool = require('./ufds-pool');
+var netconfig = require('triton-netconfig');
 
 var EventEmitter = require('events').EventEmitter;
 var UfdsFilter = require('./ufds-filter');
@@ -80,7 +81,7 @@ function APIServer(opts) {
 			var nics = JSON.parse(stdout);
 			assert.arrayOfObject(nics);
 			var adminIps = nics.filter(function (n) {
-				if (n.nic_tag !== 'admin')
+				if (netconfig.isNicAdmin(n))
 					return (false);
 				if (!n.ip)
 					return (false);
diff --git a/lib/cn-filter.js b/lib/cn-filter.js
index 02d2a3f..fc89974 100644
--- a/lib/cn-filter.js
+++ b/lib/cn-filter.js
@@ -166,7 +166,7 @@ UpdateCNCacheFSM.prototype.state_retry = function (S) {
 UpdateCNCacheFSM.prototype.state_update = function (S) {
 	var self = this;
 	var cnf = this.cnfilter;
-	cnf.client.get('/servers?setup=true&extras=status',
+	cnf.client.get('/servers?setup=true&extras=status,sysinfo',
 	    S.callback(function (err, req, res, objs) {
 		if (err) {
 			cnf.log.warn(err, 'failed to update server cache');
@@ -308,7 +308,7 @@ function makeFakeVM(server) {
 			'smartdc_role': server.hostname
 		},
 		nics: [ {
-			nic_tag: 'admin',
+			nic_tag: server.admin_tag,
 			ip: '192.0.2.1',
 			vlan_id: 0,
 			mac: mac
@@ -366,6 +366,11 @@ function cutServerObj(obj) {
 	cutObj.last_boot = new Date(obj.last_boot);
 	cutObj.last_boot_age = (new Date()) - cutObj.last_boot;
 
+	cutObj.admin_tag = 'admin';
+	if (obj.sysinfo['Admin NIC Tag']) {
+		cutObj.admin_tag = obj.sysinfo['Admin NIC Tag'];
+	}
+
 	/*
 	 * Don't bother looking for a heartbeat from a server that isn't
 	 * set up yet. These are always "down".
diff --git a/lib/net-filter.js b/lib/net-filter.js
index c0afaca..c4d35b7 100644
--- a/lib/net-filter.js
+++ b/lib/net-filter.js
@@ -15,6 +15,7 @@ var utils = require('./utils');
 var bunyan = require('bunyan');
 var vasync = require('vasync');
 var LRUCache = require('lru-cache');
+var netconfig = require('triton-netconfig');
 
 var consts = require('./consts');
 
@@ -51,7 +52,7 @@ NetFilter.prototype._transform = function (vm, enc, cb) {
 		 * If you change this, make sure to visit cn-filter.js and
 		 * think about what happens to the fake CN VMs.
 		 */
-		if (nic.nic_tag === 'admin' && (nic.vlan_id === 0 ||
+		if (netconfig.isNicAdmin(nic) && (nic.vlan_id === 0 ||
 		    nic.vlan_id === undefined)) {
 			return;
 		}
diff --git a/package.json b/package.json
index 3c3a467..c85fc06 100644
--- a/package.json
+++ b/package.json
@@ -40,6 +40,7 @@
     "restify-errors": "4.3.0",
     "sdc-scripts": "git+https://github.com/joyent/sdc-scripts.git#deefaef587ed3bee2706cb6e53ee3468e682932e",
     "sprintf-js": "1.0.3",
+    "triton-netconfig": "1.3.0",
     "triton-tags": "1.3.0",
     "uuid": "3.3.2",
     "vasync": "2.2.0"
