From ec960238e6efdd0c90033a79abc2c53c70bcb3db Mon Sep 17 00:00:00 2001
From: Mark Brooks <mark.brooks@joyent.com>
Date: Sat, 12 Aug 2017 13:00:45 -0400
Subject: [PATCH] OS-6283: service may be temporarily disabled

---
 usr/src/man/man1m/svcadm.1m | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/usr/src/man/man1m/svcadm.1m b/usr/src/man/man1m/svcadm.1m
index 1d647304b2..3dd7439e6c 100644
--- a/usr/src/man/man1m/svcadm.1m
+++ b/usr/src/man/man1m/svcadm.1m
@@ -15,7 +15,7 @@ svcadm \- manipulate service instances
 
 .LP
 .nf
-\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-s\fR]
+\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-st\fR]
      [{\fIFMRI\fR | \fIpattern\fR}...]
 .fi
 
-- 
2.21.0

