commit ec960238e6efdd0c90033a79abc2c53c70bcb3db (refs/changes/85/2385/1)
Author: Mark Brooks <mark.brooks@joyent.com>
Date:   2017-08-12T13:00:45-04:00 (2 years, 2 months ago)
    
    OS-6283: service may be temporarily disabled

diff --git a/usr/src/man/man1m/svcadm.1m b/usr/src/man/man1m/svcadm.1m
index 1d647304b2..3dd7439e6c 100644
--- a/usr/src/man/man1m/svcadm.1m
+++ b/usr/src/man/man1m/svcadm.1m
@@ -15,7 +15,7 @@ svcadm \- manipulate service instances
 
 .LP
 .nf
-\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-s\fR]
+\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-st\fR]
      [{\fIFMRI\fR | \fIpattern\fR}...]
 .fi
 
