commit 056bde83852071ddbb30283a8b35c594a1e94230 (refs/changes/85/2385/2)
Author: Mark Brooks <mark.brooks@joyent.com>
Date:   2017-09-14T14:30:43-04:00 (2 years, 1 month ago)
    
    OS-6283 Fix disable option summary in svcadm(1m)

diff --git a/usr/src/man/man1m/svcadm.1m b/usr/src/man/man1m/svcadm.1m
index 1d647304b2..59b22acb6f 100644
--- a/usr/src/man/man1m/svcadm.1m
+++ b/usr/src/man/man1m/svcadm.1m
@@ -3,7 +3,7 @@
 .\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
 .\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
 .\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
-.TH SVCADM 1M "May 9, 2008"
+.TH SVCADM 1M "Sep 14, 2017"
 .SH NAME
 svcadm \- manipulate service instances
 .SH SYNOPSIS
@@ -15,7 +15,7 @@ svcadm \- manipulate service instances
 
 .LP
 .nf
-\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-s\fR]
+\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-st\fR]
      [{\fIFMRI\fR | \fIpattern\fR}...]
 .fi
 
