From ad997d964a62abfea7992becb59bef56c235d7a9 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 24 Jul 2018 10:48:19 -0700
Subject: [PATCH] MANTA-3850 allow 'make check' in sdc-manta.git to work on
 non-SmartOS

---
 Makefile | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index ffff0c1..8284738 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -63,7 +63,15 @@ NODE_PREBUILT_VERSION=v0.10.48
 NODE_PREBUILT_TAG=gz
 # sdc-minimal-multiarch-lts 15.4.1
 NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
-include ./tools/mk/Makefile.node_prebuilt.defs
+
+ifeq ($(shell uname -s),SunOS)
+	include ./tools/mk/Makefile.node_prebuilt.defs
+else
+	NPM=npm
+	NODE=node
+	NPM_EXEC=$(shell which npm)
+	NODE_EXEC=$(shell which node)
+endif
 
 MAN_INROOT	 = docs/man
 MAN_OUTROOT	 = man
-- 
2.21.0

