From a7c3aea27896db5ecad20961f5b7084a4be75098 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 20 Oct 2016 11:39:56 -0700
Subject: [PATCH] IMGAPI-597 imgapi crash when a docker v2 layer download fails

---
 lib/images.js | 16 ----------------
 1 file changed, 16 deletions(-)

diff --git a/lib/images.js b/lib/images.js
index 179f24e..9b5e4e8 100644
--- a/lib/images.js
+++ b/lib/images.js
@@ -2681,22 +2681,6 @@ function _dockerDownloadAndImportImage(opts, callback) {
             var shasum = crypto.createHash('sha1');
             var md5sum = crypto.createHash('md5');
 
-            stream.on('end', function () {
-                req.log.debug('fs layer completed for %s', imgId);
-                ctx.resMessage({
-                    type: 'progress',
-                    payload: {
-                        id: imgId.substr(0, 12),
-                        status: 'Downloading',
-                        progressDetail: {
-                            current: size,
-                            total: size,
-                            start: startTs
-                        }
-                    }
-                });
-            });
-
             stream.on('data', function (chunk) {
                 if (ctx.downloadsCanceled) {
                     stream.destroy();
-- 
2.21.0

