From 7df1eb96e5300a77d27768612edb70cf1b5b65b9 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Fri, 14 Jun 2019 14:17:32 +0200
Subject: [PATCH] TRITON-1741 Flexible Disk cannot be sized down

---
 lib/tritonapi.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/tritonapi.js b/lib/tritonapi.js
index dfa3918..a3406e0 100644
--- a/lib/tritonapi.js
+++ b/lib/tritonapi.js
@@ -4107,7 +4107,7 @@ function resizeInstanceDisk(opts, cb) {
                 id: arg.instId,
                 diskId: arg.diskId,
                 size: opts.size,
-                dangerousAllowSrhink: opts.dangerousAllowShrink
+                dangerousAllowShrink: opts.dangerousAllowShrink
             }, function (err, _, _res) {
                 res = _res;
                 res.instId = arg.instId; // gross hack, in case caller needs it
-- 
2.21.0

