commit 37d57d487d956e74bf332341b3867d173eb90c22
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2019-08-21T18:00:40+02:00 (7 weeks ago)
    
    TRITON-1876 Sanity check provided dns domain during triton install

diff --git a/answers.json.tmpl b/answers.json.tmpl
index b264a920..6f3a6456 100644
--- a/answers.json.tmpl
+++ b/answers.json.tmpl
@@ -15,6 +15,7 @@
     "admin_gateway": "none",
     "setup_external_network": false,
     "skip_ntp_check": true,
+    "skip_dns_check": true,
     "headnode_default_gateway": "<default>",
     "dns_resolver1": "<default>",
     "dns_resolver2": "<default>",
diff --git a/answers.json.tmpl.external b/answers.json.tmpl.external
index aac46362..629e5cb4 100644
--- a/answers.json.tmpl.external
+++ b/answers.json.tmpl.external
@@ -4,6 +4,7 @@
     "simple_headers": true,
     "skip_final_confirm": true,
     "skip_edit_config": true,
+    "skip_dns_check": true,
     "datacenter_company_name": "Joyent",
     "region_name": "coal",
     "datacenter_name": "coal",
diff --git a/scripts/prompt-config.sh b/scripts/prompt-config.sh
index d1220743..efdc949b 100755
--- a/scripts/prompt-config.sh
+++ b/scripts/prompt-config.sh
@@ -186,7 +186,7 @@ isdigit ()
 	[ $# -eq 1 ] || return 1
 
 	case $1 in
-  	*[!0-9]*|"") return 1;;
+	*[!0-9]*|"") return 1;;
 	*) return 0;;
 	esac
 }
@@ -349,9 +349,9 @@ promptval()
 		# Forward and back quotes not allowed
 		echo $val | nawk '{
 		    if (index($0, "\047") != 0)
-		        exit 1
+			exit 1
 		    if (index($0, "`") != 0)
-		        exit 1
+			exit 1
 		}'
 		if [ $? != 0 ]; then
 			echo "Single quotes are not allowed."
@@ -1162,6 +1162,23 @@ provisioned on the 'external' network.\n\n"
 	promptval "Default DNS search domain" "$dns_domain" "dns_search"
 	dns_domain="$val"
 
+
+	# DNS domain checking. Make sure the provided domain name is not yet
+	# resolvable by the given DNS servers. Note this is not checked unless
+	# it's explicitly required by a config setting:
+	skip_dns_check=$(getanswer "skip_dns_check")
+	if [[ -n "${skip_dns_check}" && ${skip_dns_check} != "true" ]]; then
+		domain="binder.${datacenter_name}.${val}"
+		echo "DNS domain: $domain"
+		for resolver in "$dns_resolver1" "$dns_resolver2"; do
+			dig "@$resolver" +noall +answer +nocomments +short \
+			    "$domain" >/dev/null 2>&1 && continue
+			warning="Provided domain name '$domain' is resolvable "
+			warning+="by '$resolver' but should not be resolvable."
+			print_warning "$warning"
+		done
+	fi
+
 	message="
 By default the headnode acts as an NTP server for the admin network. You can
 set the headnode to be an NTP client to synchronize to another NTP server.\n"
