commit 8831584492180acfb5f4f7d8f6b2a75eeef45307 (refs/changes/86/986/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    account get

diff --git a/lib/do_account/do_get.js b/lib/do_account/do_get.js
index 5901db1..d2619e7 100644
--- a/lib/do_account/do_get.js
+++ b/lib/do_account/do_get.js
@@ -21,28 +21,34 @@ function do_get(subcmd, opts, args, callback) {
         return;
     }
 
-    this.top.tritonapi.cloudapi.getAccount(function (err, account) {
-        if (err) {
-            callback(err);
-            return;
+    var tritonapi = this.top.tritonapi;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            callback(setupErr);
         }
+        tritonapi.cloudapi.getAccount(function (err, account) {
+            if (err) {
+                callback(err);
+                return;
+            }
 
-        if (opts.json) {
-            console.log(JSON.stringify(account));
-        } else {
-            // pretty print
-            var dates = ['updated', 'created'];
-            Object.keys(account).forEach(function (key) {
-                var val = account[key];
-                if (dates.indexOf(key) >= 0) {
-                    console.log('%s: %s (%s)', key, val,
-                        common.longAgo(new Date(val)));
-                } else {
-                    console.log('%s: %s', key, val);
-                }
-            });
-        }
-        callback();
+            if (opts.json) {
+                console.log(JSON.stringify(account));
+            } else {
+                // pretty print
+                var dates = ['updated', 'created'];
+                Object.keys(account).forEach(function (key) {
+                    var val = account[key];
+                    if (dates.indexOf(key) >= 0) {
+                        console.log('%s: %s (%s)', key, val,
+                                    common.longAgo(new Date(val)));
+                    } else {
+                        console.log('%s: %s', key, val);
+                    }
+                });
+            }
+            callback();
+        });
     });
 }
 
