From eeb145fd73de0ec6ef384fc8379ff1764def5c96 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Tue, 21 Mar 2017 21:34:57 +1300
Subject: [PATCH] PAPI-136: Add a test to cover multiple wildcards

---
 test/api.test.js | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/test/api.test.js b/test/api.test.js
index bb62c95..322399c 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -584,6 +584,18 @@ test('GET /packages (Search by wildcard)', function (t) {
 
 
 
+test('GET /packages (Search by multiple wildcard)', function (t) {
+    var query = '/packages?name=*pi_test_*';
+
+    var testFilter = function (p) {
+        return /^api_test_/.test(p.name);
+    };
+
+    searchAndCheckPkgs(t, query, testFilter);
+});
+
+
+
 test('GET /packages (Search by multiple fields)', function (t) {
     var query = '/packages?name=api_test_*&owner_uuids=' + uuid();
 
-- 
2.21.0

