commit 8aaf141d160e239b32642f9671b504935e85018e (refs/changes/91/1691/2)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-03-22T23:38:26+13:00 (2 years, 7 months ago)
    
    PAPI-136: Add a test to cover multiple wildcards

diff --git a/test/api.test.js b/test/api.test.js
index 802f5ed..522e5c0 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -584,6 +584,18 @@ test('GET /packages (Search by wildcard)', function (t) {
 
 
 
+test('GET /packages (Search by multiple wildcard)', function (t) {
+    var query = '/packages?name=*pi_test_*';
+
+    var testFilter = function (p) {
+        return /^api_test_/.test(p.name);
+    };
+
+    searchAndCheckPkgs(t, query, testFilter);
+});
+
+
+
 test('GET /packages (Search by multiple fields)', function (t) {
     var query = '/packages?name=api_test_*&owner_uuids=' + uuid();
 
