commit 66d68c6f0f4134619f1a68f56d2a2805e89dce29 (refs/changes/83/3783/1)
Author: dyep49 <dyep49@gmail.com>
Date:   2018-04-06T17:46:15-07:00 (1 year, 6 months ago)
    
    WORKFLOW-222 artedi metrics for job times

diff --git a/lib/workflow-moray-backend.js b/lib/workflow-moray-backend.js
index 1c5151d..92c96b8 100644
--- a/lib/workflow-moray-backend.js
+++ b/lib/workflow-moray-backend.js
@@ -5,9 +5,10 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
+var assert = require('assert-plus');
 var util = require('util');
 var moray = require('moray');
 var vasync = require('vasync');
@@ -45,6 +46,8 @@ var WorkflowMorayBackend = module.exports = function (config) {
         throw new TypeError('\'config\' (Object) required');
     }
 
+    assert.object(config.metricsManager, 'config.metricsManager');
+
     var log;
     if (config.log) {
         log = config.log.child({component: 'wf-moray-backend'});
@@ -66,6 +69,7 @@ var WorkflowMorayBackend = module.exports = function (config) {
         log = new Logger(config.logger);
     }
     var client;
+    var metricsManager = config.metricsManager;
 
     var definitions = bu.definitions;
     var buckets = [];
@@ -1030,6 +1034,9 @@ var WorkflowMorayBackend = module.exports = function (config) {
                 if (err) {
                     return callback(err);
                 }
+
+                metricsManager.collectMetrics('jobMetrics', theJob);
+
                 if (typeof (theJob.locks) !== 'undefined') {
                     client.delObject('wf_locked_targets', theJob.uuid,
                         function (err1) {
diff --git a/package.json b/package.json
index f8bd5f3..1b02cf8 100644
--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "author": "Pedro Palazón Candel <kusorbox@gmail.com> (http://www.joyent.com)",
   "name": "wf-moray-backend",
   "description": "A backend for wf built over Moray",
-  "version": "1.1.0",
+  "version": "2.0.0",
   "private": true,
   "homepage": "https://github.com/joyent/node-workflow-moray-backend",
   "repository": {
@@ -17,16 +17,17 @@
     "node": ">=0.8"
   },
   "dependencies": {
+    "assert-plus": "1.0.0",
     "bunyan": "1.8.1",
+    "jsprim": "^1.3.1",
     "moray": "3.1.1",
     "uuid": "3.0.1",
-    "wf": "git+https://github.com/joyent/node-workflow.git#aac927e4",
     "vasync": "1.6.4",
-    "jsprim": "^1.3.1",
-    "verror": "1.10.0"
+    "verror": "1.10.0",
+    "wf": "git+https://github.com/joyent/node-workflow.git#f7f0e2fc"
   },
   "optionalDependencies": {
-      "dtrace-provider": "0.8.2"
+    "dtrace-provider": "0.8.2"
   },
   "devDependencies": {
     "tap": "~0.3"
diff --git a/test/helper.js b/test/helper.js
index 385c511..665ced9 100644
--- a/test/helper.js
+++ b/test/helper.js
@@ -32,8 +32,13 @@ module.exports = {
             unique: false
         };
 
+        config.backend.opts.metricsManager = {
+            collectMetrics: function () {}
+        };
+
         config.api.job_extra_params.push('foo');
         config.api.job_extra_params.push('bar');
+
         return config;
     }
 };
