commit 0b545cb7cc2792ac50a1f715c1e788163bc277f8 (refs/changes/51/3751/1)
Author: Mike Zeller <mike@mikezeller.net>
Date:   2018-04-02T13:35:03-04:00 (1 year, 6 months ago)
    
    TRITON-266 Assign NIC indexes for "linklocal" routes from NAPI

diff --git a/lib/workflows/provision.js b/lib/workflows/provision.js
index 0818e0f..8e07118 100644
--- a/lib/workflows/provision.js
+++ b/lib/workflows/provision.js
@@ -278,7 +278,11 @@ function preparePayload(job, cb) {
             typeof (nic['routes']) === 'object') {
             for (var r in nic['routes']) {
                 if (!routes.hasOwnProperty(r)) {
-                    routes[r] = nic['routes'][r];
+                    if (nic['routes'][r] === 'linklocal') {
+                        routes[r] = 'nics[' + i.toString() + ']';
+                    } else {
+                        routes[r] = nic['routes'][r];
+                    }
                 }
             }
         }
