From 0b545cb7cc2792ac50a1f715c1e788163bc277f8 Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike@mikezeller.net>
Date: Wed, 21 Mar 2018 16:38:41 -0400
Subject: [PATCH] TRITON-266 Assign NIC indexes for "linklocal" routes from
 NAPI

---
 lib/workflows/provision.js | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/workflows/provision.js b/lib/workflows/provision.js
index 0818e0f..8e07118 100644
--- a/lib/workflows/provision.js
+++ b/lib/workflows/provision.js
@@ -278,7 +278,11 @@ function preparePayload(job, cb) {
             typeof (nic['routes']) === 'object') {
             for (var r in nic['routes']) {
                 if (!routes.hasOwnProperty(r)) {
-                    routes[r] = nic['routes'][r];
+                    if (nic['routes'][r] === 'linklocal') {
+                        routes[r] = 'nics[' + i.toString() + ']';
+                    } else {
+                        routes[r] = nic['routes'][r];
+                    }
                 }
             }
         }
-- 
2.21.0

