From 3130da8850f2ec0e56b3f47ae074f2460868d17e Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 9 Aug 2019 15:37:57 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index d25b8fc..9058562 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit d25b8fc60fb8c649559924870fe3aaf75e7421d5
+Subproject commit 9058562f2f59892795e470ce25443ff636c90478
-- 
2.21.0

