commit b1742b8ea18d4afd7052c1972541739d5c7334f6
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2019-08-16T13:34:00+00:00 (7 weeks ago)
    
    TRITON-1873 TRITON-1692 broke linux-prepare-image for Alpine Linux

diff --git a/tools/prepare-image/linux-prepare-image b/tools/prepare-image/linux-prepare-image
index 6e355ad..b6ab08a 100755
--- a/tools/prepare-image/linux-prepare-image
+++ b/tools/prepare-image/linux-prepare-image
@@ -98,9 +98,14 @@ function cleanup_root() {
     history -c
     history -w || true
 
-    # Remove or set password for root.  If root-password-hash is not set, this
-    # is equivalent to "passwd -d root".
-    usermod -p "$(mdata_get_pi root-password-hash)" root
+    # Set or remove the root password. prepare-image:root-password-hash is only
+    # supported on those distros that support "usermod -p <hash> <user>".
+    local pw="$(mdata_get_pi root-password-hash)"
+    if [[ -n "$pw" ]]; then
+        usermod -p "$pw" root
+    else
+        passwd -d root
+    fi
 
     # Clean up history, ssh keys, etc.
     cleanup_homedir /root
