commit 4dc2b2a983e8612e59464ef23f237b19d59c3daa (refs/changes/50/5250/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-12-14T16:22:30-08:00 (10 months ago)
    
    MANTA-4035 retire propeller

diff --git a/sapi_manifests/madtom_coal/template b/sapi_manifests/madtom_coal/template
index cb3f72c..72baef9 100644
--- a/sapi_manifests/madtom_coal/template
+++ b/sapi_manifests/madtom_coal/template
@@ -93,12 +93,6 @@
         { "processType": "marlin-dashboard", "port": 80 }
       ]
     },
-    {
-      "hostType": "propeller",
-      "processes": [
-        { "processType": "propeller", "port": 80 }
-      ]
-    },
     {
       "hostType": "workflow",
       "processes": [
@@ -151,12 +145,6 @@
       "path": "/index.htm",
       "secure": false
     },
-    {
-      "processType": "propeller",
-      "checkerType": "http",
-      "path": "/ping",
-      "secure": false
-    },
     {
       "processType": "medusa",
       "checkerType": "http",
diff --git a/sapi_manifests/madtom_lab/template b/sapi_manifests/madtom_lab/template
index b77117b..65f567e 100644
--- a/sapi_manifests/madtom_lab/template
+++ b/sapi_manifests/madtom_lab/template
@@ -102,12 +102,6 @@
         { "processType": "marlin-dashboard", "port": 80 }
       ]
     },
-    {
-      "hostType": "propeller",
-      "processes": [
-        { "processType": "propeller", "port": 80 }
-      ]
-    },
     {
       "hostType": "workflow",
       "processes": [
@@ -160,12 +154,6 @@
       "path": "/index.htm",
       "secure": false
     },
-    {
-      "processType": "propeller",
-      "checkerType": "http",
-      "path": "/ping",
-      "secure": false
-    },
     {
       "processType": "medusa",
       "checkerType": "http",
diff --git a/sapi_manifests/madtom_production/template b/sapi_manifests/madtom_production/template
index 4f9a1da..d1b00ae 100644
--- a/sapi_manifests/madtom_production/template
+++ b/sapi_manifests/madtom_production/template
@@ -114,12 +114,6 @@
         { "processType": "marlin-dashboard", "port": 80 }
       ]
     },
-    {
-      "hostType": "propeller",
-      "processes": [
-        { "processType": "propeller", "port": 80 }
-      ]
-    },
     {
       "hostType": "workflow",
       "processes": [
@@ -172,12 +166,6 @@
       "path": "/index.htm",
       "secure": false
     },
-    {
-      "processType": "propeller",
-      "checkerType": "http",
-      "path": "/ping",
-      "secure": false
-    },
     {
       "processType": "medusa",
       "checkerType": "http",
