commit 479921920e666b1b165dc6d470e8df97d5164158 (refs/changes/89/2389/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-08-14T21:35:33+00:00 (2 years, 2 months ago)
    
    OS-6295 t4nex driver missing pciex1425,6405

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index f9d9080e..eda1c7b9 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -1162,7 +1162,7 @@ t4nex "pciex1425,6401"
 t4nex "pciex1425,6402"
 t4nex "pciex1425,6403"
 t4nex "pciex1425,6404"
-t4nex "pciex1425,6404"
+t4nex "pciex1425,6405"
 t4nex "pciex1425,6406"
 t4nex "pciex1425,6407"
 t4nex "pciex1425,6408"
