commit 5e7e14b31db9c4485d34aac1161f74c366a5584b (refs/changes/75/3775/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-04-04T13:23:01-07:00 (1 year, 6 months ago)
    
    TRITON-306 remove SAPI's uncaughtException restify handler

diff --git a/lib/server/sapi.js b/lib/server/sapi.js
index 8247fd4..7e15415 100644
--- a/lib/server/sapi.js
+++ b/lib/server/sapi.js
@@ -198,20 +198,6 @@ function createServer(options) {
 
     endpoints.attachTo(server, options.model);
 
-    server.on('uncaughtException', function (req, res, route, err) {
-        req.log.error({
-            req: req,
-            res: res,
-            route: route,
-            err: err
-        }, 'uncaught exception');
-        if (!res.headersSent) {
-            req.log.error('sending error response from uncaught ' +
-                'exception');
-            res.send(err);
-        }
-    });
-
     return (server);
 }
 
