From 171e846648f8ab62f141971e283c96bd9f9abddf Mon Sep 17 00:00:00 2001
From: Jared Morrow <jm@joyent.com>
Date: Fri, 9 Jun 2017 10:40:00 -0600
Subject: [PATCH] MANTA-2900 mlocate should not need to setupMantaBuckets

---
 bin/mlocate | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bin/mlocate b/bin/mlocate
index d592789..993e949 100755
--- a/bin/mlocate
+++ b/bin/mlocate
@@ -81,6 +81,7 @@ function createMorayClient(opts, cb) {
 
     var log = LOG.child({component: 'moray'}, true);
     opts.log = log;
+    opts.readOnly = true;
 
     var client = new libmanta.createMorayClient(opts);
 
-- 
2.21.0

