From dd88af412f2b763fb16695131df0f80dae8d1a1a Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 5 Mar 2019 17:26:41 -0500
Subject: [PATCH] TRITON-1287 agentsshar no longer able to assemble agent
 branches

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8c52a3b..655afaa 100644
--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@ deps:
 	PATH=/opt/tools/bin:$(PATH) /opt/tools/bin/npm install
 
 shar: deps
-	PATH=/opt/tools/bin:$(PATH) ./mk-agents-shar -b $(BUILDNAME)
+	PATH=/opt/tools/bin:$(PATH) ./mk-agents-shar -b "$(BUILDNAME)"
 
 publish: shar
 	mkdir -p $(ENGBLD_BITS_DIR)/$(NAME)
-- 
2.21.0

