commit dd88af412f2b763fb16695131df0f80dae8d1a1a
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T17:26:41-05:00 (7 months ago)
    
    TRITON-1287 agentsshar no longer able to assemble agent branches

diff --git a/Makefile b/Makefile
index 8c52a3b..655afaa 100644
--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@ deps:
 	PATH=/opt/tools/bin:$(PATH) /opt/tools/bin/npm install
 
 shar: deps
-	PATH=/opt/tools/bin:$(PATH) ./mk-agents-shar -b $(BUILDNAME)
+	PATH=/opt/tools/bin:$(PATH) ./mk-agents-shar -b "$(BUILDNAME)"
 
 publish: shar
 	mkdir -p $(ENGBLD_BITS_DIR)/$(NAME)
