commit c37c632cdb2a4cfe18be0767c03a504e98ac1223 (refs/changes/85/2485/1)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2017-08-31T21:48:26+00:00 (2 years, 1 month ago)
    
    TOOLS-1848 some makefile defaults have old copyright years

diff --git a/tools/mk/Makefile.deps b/tools/mk/Makefile.deps
index 1cffbe7..66c7cfc 100644
--- a/tools/mk/Makefile.deps
+++ b/tools/mk/Makefile.deps
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.manpages.defs b/tools/mk/Makefile.manpages.defs
index 6da7876..85c5488 100644
--- a/tools/mk/Makefile.manpages.defs
+++ b/tools/mk/Makefile.manpages.defs
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2016, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.manpages.targ b/tools/mk/Makefile.manpages.targ
index 11f242b..6f00caf 100644
--- a/tools/mk/Makefile.manpages.targ
+++ b/tools/mk/Makefile.manpages.targ
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2016, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node.targ b/tools/mk/Makefile.node.targ
index bf53f78..8cb1c33 100644
--- a/tools/mk/Makefile.node.targ
+++ b/tools/mk/Makefile.node.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.node_prebuilt.targ b/tools/mk/Makefile.node_prebuilt.targ
index 6877333..fc6a767 100644
--- a/tools/mk/Makefile.node_prebuilt.targ
+++ b/tools/mk/Makefile.node_prebuilt.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.smf.defs b/tools/mk/Makefile.smf.defs
index b988bbe..3c0c4d0 100644
--- a/tools/mk/Makefile.smf.defs
+++ b/tools/mk/Makefile.smf.defs
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
diff --git a/tools/mk/Makefile.smf.targ b/tools/mk/Makefile.smf.targ
index f78de96..9554aa0 100644
--- a/tools/mk/Makefile.smf.targ
+++ b/tools/mk/Makefile.smf.targ
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
