commit 7d6159ada4ed26db468663ada4e201b41e557b6f (refs/changes/33/2333/1)
Author: Tim Kordas <tim.kordas@joyent.com>
Date:   2017-08-04T14:35:04-07:00 (2 years, 2 months ago)
    
    MANTA-3091: consume data in http-checker, instead of leaking

diff --git a/lib/checkers/http_checker.js b/lib/checkers/http_checker.js
index c34c82d..8f42e17 100644
--- a/lib/checkers/http_checker.js
+++ b/lib/checkers/http_checker.js
@@ -60,8 +60,11 @@ HttpChecker.prototype.check = function (cb) {
                         error = new Error('status code is >= 300');
                         error.code = res.statusCode;
                 }
-                cb(error, {
-                        'httpStatusCode': res.statusCode
+                res.on('data', function(d) { /* consume */});
+                res.on('end', function() {
+                       cb(error, {
+                               'httpStatusCode': res.statusCode
+                       });
                 });
         });
         req.once('error', function (err) {
