From 7d6159ada4ed26db468663ada4e201b41e557b6f Mon Sep 17 00:00:00 2001
From: Tim Kordas <tim.kordas@joyent.com>
Date: Fri, 4 Aug 2017 14:35:04 -0700
Subject: [PATCH] MANTA-3091: consume data in http-checker, instead of leaking

---
 lib/checkers/http_checker.js | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/lib/checkers/http_checker.js b/lib/checkers/http_checker.js
index c34c82d..8f42e17 100644
--- a/lib/checkers/http_checker.js
+++ b/lib/checkers/http_checker.js
@@ -60,8 +60,11 @@ HttpChecker.prototype.check = function (cb) {
                         error = new Error('status code is >= 300');
                         error.code = res.statusCode;
                 }
-                cb(error, {
-                        'httpStatusCode': res.statusCode
+                res.on('data', function(d) { /* consume */});
+                res.on('end', function() {
+                       cb(error, {
+                               'httpStatusCode': res.statusCode
+                       });
                 });
         });
         req.once('error', function (err) {
-- 
2.21.0

