commit 03d6b5a02b7e9505661e5dcf5d1eabd94cc91591 (refs/changes/73/5573/2)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2019-02-13T02:21:38+00:00 (8 months ago)
    
    MANTA-2948 mako's tombstone grace period should be configurable

diff --git a/config/application.json b/config/application.json
index 16906d9..a5f9261 100644
--- a/config/application.json
+++ b/config/application.json
@@ -2,7 +2,5 @@
 	"name": "manta",
 	"params": {
 		"delegate_dataset": true
-	},
-	"metadata": {
 	}
 }
diff --git a/config/services/storage/service.json b/config/services/storage/service.json
index cbb0e94..591cc83 100644
--- a/config/services/storage/service.json
+++ b/config/services/storage/service.json
@@ -4,5 +4,8 @@
 		"ram": 256,
 		"quota": 0,
 		"zfs_io_priority": 0
+	},
+	"metadata": {
+		"STORAGE_MAX_TOMBSTONE_DIRECTORIES": 21
 	}
 }
diff --git a/config/services/storage/service.json.lab b/config/services/storage/service.json.lab
index 7c87a2f..6135106 100644
--- a/config/services/storage/service.json.lab
+++ b/config/services/storage/service.json.lab
@@ -1,5 +1,8 @@
 {
 	"params": {
                 "cpu_shares": 2048
+	},
+	"metadata": {
+		"STORAGE_MAX_TOMBSTONE_DIRECTORIES": 21
 	}
 }
diff --git a/config/services/storage/service.json.production b/config/services/storage/service.json.production
index 79689cf..6079a27 100644
--- a/config/services/storage/service.json.production
+++ b/config/services/storage/service.json.production
@@ -2,5 +2,8 @@
 	"params": {
 		"cpu_shares": 2048,
 		"ram": 8192
+	},
+	"metadata": {
+		"STORAGE_MAX_TOMBSTONE_DIRECTORIES": 21
 	}
 }
