From 7c8bb17b5d865ebf7e4aa678669203adc9ce3a94 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Tue, 12 Feb 2019 22:38:35 +0000
Subject: [PATCH] MANTA-2948 mako's tombstone grace period should be
 configurable

---
 config/application.json | 1 +
 1 file changed, 1 insertion(+)

diff --git a/config/application.json b/config/application.json
index 16906d9..76d1ae6 100644
--- a/config/application.json
+++ b/config/application.json
@@ -4,5 +4,6 @@
 		"delegate_dataset": true
 	},
 	"metadata": {
+		"STORAGE_MAX_TOMBSTONE_DIRECTORIES": 21
 	}
 }
-- 
2.21.0

