commit 557bd1abe7fc299fdaee93c1d787668387853e9f (refs/changes/61/1761/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-04-06T22:40:30+00:00 (2 years, 6 months ago)
    
    OS-5565 Need smrt HBA mode support

diff --git a/overlay/generic/etc/driver_classes b/overlay/generic/etc/driver_classes
index 8608747a..3124f9a2 100644
--- a/overlay/generic/etc/driver_classes
+++ b/overlay/generic/etc/driver_classes
@@ -27,4 +27,4 @@ qlc	fibre-channel
 si3124	scsi-self-identifying
 dr_sas	scsi-self-identifying
 skd	disk
-smrt	scsi
+smrt	scsi-self-identifying
