commit 22726217ae96bbb0b2962038347fa545ad042b85
Author: John Levon <john.levon@joyent.com>
Date:   2019-08-20T12:51:21+00:00 (7 weeks ago)
    
    MANTA-4489 ops is under-sized in a lab deployment

diff --git a/config/services/ops/service.json.lab b/config/services/ops/service.json.lab
new file mode 100644
index 0000000..2a27ab7
--- /dev/null
+++ b/config/services/ops/service.json.lab
@@ -0,0 +1,5 @@
+{
+	"params": {
+		"ram": 4096
+	}
+}
