From 45d6fb0d9e65fc310f93da69f9ac141ae2503f7a Mon Sep 17 00:00:00 2001
From: Jordan Hendricks <jordan.hendricks@joyent.com>
Date: Fri, 23 Mar 2018 00:35:19 +0000
Subject: [PATCH] MANTA-3410 Audit job reports missing MPU parts

---
 lib/uploads/commit.js | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/lib/uploads/commit.js b/lib/uploads/commit.js
index 08f737b..a5c56a8 100644
--- a/lib/uploads/commit.js
+++ b/lib/uploads/commit.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -430,14 +430,26 @@ function finalizeUpload(req, res, next) {
         inputs: sharks
     },  function (err, results) {
             log.debug('mako-finalize: all sharks contacted');
-
             if (err) {
                 results.operations.forEach(function (r) {
                     log.warn({
                         err: r.err,
                         shark: r.result.shark
-                    }, 'error with shark');
+                    }, 'error from mako');
                 });
+
+                /*
+                 * If we get a SharkResponseError (as opposed to a
+                 * ConnectTimeoutError), we log information about it at the
+                 * 'error' level so that an amon alarm fires.
+                 */
+                if (verror.hasCauseWithName(err, 'SharkResponseError')) {
+                    log.error({
+                        message: err.cause().message,
+                        uploadId: id
+                    });
+                }
+
                 next(new SharksExhaustedError());
             } else {
                 var md5 = null;
-- 
2.21.0

