commit 7b699bdacaf6606595d0282d9a3f42db646f2286 (refs/changes/18/418/1)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2016-09-06T09:28:57-07:00 (3 years, 1 month ago)
    
    DOCKER-925 createContainer error: "Cannot read property 'sort' of null"

diff --git a/lib/endpoints/allocator.js b/lib/endpoints/allocator.js
index f050d5a..d83f348 100644
--- a/lib/endpoints/allocator.js
+++ b/lib/endpoints/allocator.js
@@ -489,7 +489,7 @@ function getOpenProvisioningTickets(log, cb) {
 
     var start = new Date();
 
-    ModelWaitlist.query(filter, null, function (err, tickets) {
+    ModelWaitlist.query(filter, {}, function (err, tickets) {
         var delta = new Date() - start;
         log.debug('Open provisioning tickets search took ' + delta + ' ms');
 
diff --git a/lib/models/waitlist.js b/lib/models/waitlist.js
index 0cd27b5..6a92aeb 100644
--- a/lib/models/waitlist.js
+++ b/lib/models/waitlist.js
@@ -482,6 +482,10 @@ ModelWaitlist.query = function (filter, findOpts, callback) {
         findOpts = {};
     }
 
+    if (!findOpts) {
+        findOpts = {};
+    }
+
     var findParams = jsprim.deepCopy(findOpts);
     var sort = findParams.sort;
     findParams.sort = jsprim.deepCopy(defaultSort);
