From dff17b076bce62352e7a70a6c70cc42a8797ae41 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 13 Jun 2018 10:42:28 -0700
Subject: [PATCH] MANTA-3758 ops zone missing reverse lookup data in binder

---
 lib/server.js | 1 +
 lib/zk.js     | 1 +
 2 files changed, 2 insertions(+)

diff --git a/lib/server.js b/lib/server.js
index 09272ed..078b4d3 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -302,6 +302,7 @@ function resolve(options, query, cb) {
         case 'load_balancer':
         case 'moray_host':
         case 'redis_host':
+        case 'ops_host':
         case 'rr_host':
                 addr = record[record.type].address;
                 query.addAnswer(domain, new ARecord(addr), ttl);
diff --git a/lib/zk.js b/lib/zk.js
index b7b998f..655efb2 100644
--- a/lib/zk.js
+++ b/lib/zk.js
@@ -175,6 +175,7 @@ TreeNode.prototype.onDataChanged = function (zk, data, stat) {
         case 'load_balancer':
         case 'moray_host':
         case 'redis_host':
+        case 'ops_host':
         case 'rr_host':
                 var record = parsedData[parsedData.type];
                 if (typeof (record) !== 'object' || record === null)
-- 
2.21.0

