From c6f1b9eaec8e85d1c8133722852bc77644a9de8e Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Wed, 26 Apr 2017 00:38:11 +0000
Subject: [PATCH] OS-6078 mac_transceiver_info(9F) man page links are incorrect

---
 usr/src/man/man9f/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/usr/src/man/man9f/Makefile b/usr/src/man/man9f/Makefile
index 2fba3735d0..dcea9314f9 100644
--- a/usr/src/man/man9f/Makefile
+++ b/usr/src/man/man9f/Makefile
@@ -1832,8 +1832,8 @@ mac_prop_info_set_default_uint64.9f	:= LINKSRC = mac_prop_info.9f
 mac_prop_info_set_perm.9f		:= LINKSRC = mac_prop_info.9f
 mac_prop_info_set_range_uint32.9f	:= LINKSRC = mac_prop_info.9f
 
-mac_transceiver_info_set_present.9f	:= LINKSRC = mac_transceiver.info.9f
-mac_transceiver_info_set_usable.9f	:= LINKSRC = mac_transceiver.info.9f
+mac_transceiver_info_set_present.9f	:= LINKSRC = mac_transceiver_info.9f
+mac_transceiver_info_set_usable.9f	:= LINKSRC = mac_transceiver_info.9f
 
 mac_unregister.9f			:= LINKSRC = mac_register.9f
 
-- 
2.21.0

