commit ab1ee3f8ae68475a06fdbcddd0a2bfe13f3e46a2 (refs/changes/99/5399/2)
Author: Andy Fiddaman <omnios@citrus-it.co.uk>
Date:   2019-01-25T11:27:00+01:00 (9 months ago)
    
    OS-7528 update bhyve pkg manifest from OmniOS

diff --git a/usr/src/pkg/manifests/system-bhyve.mf b/usr/src/pkg/manifests/system-bhyve.mf
index fe19fb21b4..002bef64cc 100644
--- a/usr/src/pkg/manifests/system-bhyve.mf
+++ b/usr/src/pkg/manifests/system-bhyve.mf
@@ -15,6 +15,7 @@
 
 #
 # Copyright 2018 Joyent, Inc.
+# Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
 #
 
 #
@@ -30,31 +31,27 @@ set name=info.classification \
     value=org.opensolaris.category.2008:System/Virtualization
 set name=variant.arch value=i386
 dir path=kernel group=sys
-dir path=lib group=bin
-dir path=lib/$(ARCH64) group=bin
 dir path=usr group=sys
 dir path=usr/kernel/drv group=sys
 dir path=usr/kernel/drv/$(ARCH64) group=sys
-dir path=usr/lib group=bin
+dir path=usr/sbin
 dir path=usr/share
 dir path=usr/share/man
 dir path=usr/share/man/man1m
-dir path=usr/sbin
 driver name=ppt
 driver name=viona
 driver name=vmm
-file path=lib/$(ARCH64)/libvmmapi.so.1
-link path=lib/$(ARCH64)/libvmmapi.so target=./libvmmapi.so.1
 file path=usr/kernel/drv/$(ARCH64)/ppt
 file path=usr/kernel/drv/$(ARCH64)/viona
 file path=usr/kernel/drv/$(ARCH64)/vmm
 file path=usr/kernel/drv/ppt.conf
 file path=usr/kernel/drv/viona.conf
 file path=usr/kernel/drv/vmm.conf
-file path=usr/lib/libppt.so.1
-file path=usr/lib/$(ARCH64)/libppt.so.1
 file path=usr/sbin/bhyve mode=0555
 file path=usr/sbin/bhyvectl mode=0555
 file path=usr/sbin/pptadm mode=0555
 file path=usr/share/man/man1m/pptadm.1m
 license lic_CDDL license=lic_CDDL
+depend fmri=developer/acpi type=require
+depend fmri=system/bhyve/firmware type=require
+depend fmri=system/library/bhyve type=require
diff --git a/usr/src/pkg/manifests/system-library-bhyve.mf b/usr/src/pkg/manifests/system-library-bhyve.mf
new file mode 100644
index 0000000000..989be155e5
--- /dev/null
+++ b/usr/src/pkg/manifests/system-library-bhyve.mf
@@ -0,0 +1,38 @@
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
+#
+
+set name=pkg.fmri value=pkg:/system/library/bhyve@$(PKGVERS)
+set name=pkg.description value="BSD hypervisor (libraries)"
+set name=pkg.summary value="BSD hypervisor (libraries)"
+set name=variant.arch value=i386
+dir path=lib group=bin
+dir path=lib/$(ARCH64) group=bin
+dir path=usr group=sys
+dir path=usr/lib group=bin
+file path=lib/$(ARCH64)/libvmm.so.1
+file path=lib/$(ARCH64)/libvmmapi.so.1
+file path=lib/$(ARCH64)/llib-lvmmapi.ln
+file path=usr/include/libppt.h
+file path=usr/include/libvmm.h
+file path=usr/lib/$(ARCH64)/libppt.so.1
+file path=usr/lib/$(ARCH64)/llib-lppt.ln
+file path=usr/lib/libppt.so.1
+file path=usr/lib/llib-lppt
+file path=usr/lib/llib-lppt.ln
+license lic_CDDL license=lic_CDDL
