From 43b9312960f93777e93b73d5ce69e0f1d5e79e2c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 18 Jul 2018 08:58:47 +0000
Subject: [PATCH] TRITON-554 sdc-dockerlogger build fails when cloned to
 anything other than 'dockerlogger'

---
 Makefile | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/Makefile b/Makefile
index e1dd2fb..db244ab 100644
--- a/Makefile
+++ b/Makefile
@@ -5,13 +5,16 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 ELFEDIT=/usr/bin/elfedit
 GOPATH=$(PWD)/vendor
 TARGETS=dockerlogger dockerlogger.smartos
 
+BRANCH ?= $(shell git branch 2> /dev/null | sed -e '/^[^*]/d' -e 's/* \(.*\)/\1/')
+DESTDIR ?= .
+
 ifeq ($(shell uname -s),Darwin)
 	TARGET = dockerlogger
 else
@@ -25,16 +28,14 @@ dockerlogger.smartos: dockerlogger
 	/usr/bin/elfedit -e 'ehdr:ei_osabi ELFOSABI_SOLARIS' dockerlogger dockerlogger.smartos
 
 dockerlogger: dockerlogger.go
-	GOPATH=$(GOPATH) go build
+	GOPATH=$(GOPATH) go build $<
 
 .PHONY: fmt
 fmt:
 	gofmt -w dockerlogger.go
 
 .PHONY: pkg
-pkg: dockerlogger.smartos
-	@[[ -n "$(BRANCH)" ]] || (echo "missing BRANCH="; exit 1)
-	@[[ -n "$(DESTDIR)" ]] || (echo "missing DESTDIR="; exit 1)
+pkg: $(DESTDIR) dockerlogger.smartos
 	./tools/mk-shar -b $(BRANCH) -o $(DESTDIR)
 
 .PHONY: check
@@ -51,4 +52,10 @@ test: $(TARGET) node_modules/tape/bin/tape
 .PHONY: clean
 clean:
 	rm -f $(TARGETS)
+	rm -f $(DESTDIR)/dockerlogger-*.manifest
+	rm -f $(DESTDIR)/dockerlogger-*.md5sum
+	rm -f $(DESTDIR)/dockerlogger-*.sh
 	rm -rf build
+
+$(DESTDIR):
+	mkdir -p $@
-- 
2.21.0

