From 488d2c2aeaf1c2257ac708a0b244374ec29f749a Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 8 Feb 2017 18:47:40 -0800
Subject: [PATCH] joyent/node-cueball#93 kang payload missing "options.domain"
 etc

---
 lib/pool-monitor.js | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/lib/pool-monitor.js b/lib/pool-monitor.js
index 51718fe..40dad3f 100644
--- a/lib/pool-monitor.js
+++ b/lib/pool-monitor.js
@@ -102,9 +102,12 @@ CueBallPoolMonitor.prototype.toKangOptions = function () {
 		obj.state = pool.getState();
 		obj.counters = pool.p_counters;
 		obj.options = {};
-		obj.options.domain = pool.p_resolver.r_domain;
-		obj.options.service = pool.p_resolver.r_service;
-		obj.options.defaultPort = pool.p_resolver.r_defport;
+		var res = pool.p_resolver;
+		if (res.r_fsm !== undefined)
+			res = res.r_fsm;
+		obj.options.domain = res.r_domain || pool.p_domain;
+		obj.options.service = res.r_service;
+		obj.options.defaultPort = res.r_defport;
 		obj.options.spares = pool.p_spares;
 		obj.options.maximum = pool.p_max;
 		return (obj);
@@ -144,9 +147,12 @@ CueBallPoolMonitor.prototype.toKangOptions = function () {
 		obj.target = cset.cs_target;
 		obj.maximum = cset.cs_max;
 		obj.options = {};
-		obj.options.domain = cset.cs_resolver.r_domain;
-		obj.options.service = cset.cs_resolver.r_service;
-		obj.options.defaultPort = cset.cs_resolver.r_defport;
+		var res = cset.cs_resolver;
+		if (res.r_fsm !== undefined)
+			res = res.r_fsm;
+		obj.options.domain = res.r_domain || cset.cs_domain;
+		obj.options.service = res.r_service;
+		obj.options.defaultPort = res.r_defport;
 		return (obj);
 	}
 
-- 
2.21.0

