commit e80388bef054d31e104a7892eda40e26941f9693 (refs/changes/17/4017/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-05-29T09:46:28-05:00 (1 year, 4 months ago)
    
    OS-6980 small leak in libvarpd

diff --git a/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c b/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
index f557f4cbc6..28b4005dbe 100644
--- a/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
+++ b/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
@@ -522,6 +522,8 @@ libvarpd_plugin_query_reply(varpd_query_handle_t *vqh, int action)
 	else
 		libvarpd_panic("plugin %s passed in an invalid action: %d",
 		    vqp->vq_instance->vri_plugin->vpp_name, action);
+
+	umem_cache_free(vqp->vq_instance->vri_impl->vdi_qcache, vqp);
 }
 
 void
