commit a2b6fb8480b78b215e9cda7c9bff37527bba8b83 (refs/changes/18/1018/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:06-05:00 (2 years, 10 months ago)
    
    ./bin/triton network get 2d035b65

diff --git a/lib/do_network/do_get.js b/lib/do_network/do_get.js
index 8d62ed3..02531d2 100644
--- a/lib/do_network/do_get.js
+++ b/lib/do_network/do_get.js
@@ -25,17 +25,24 @@ function do_get(subcmd, opts, args, cb) {
             'incorrect number of args (%d)', args.length)));
     }
 
-    this.top.tritonapi.getNetwork(args[0], function (err, net) {
-        if (err) {
-            return cb(err);
-        }
+    var tritonapi = this.top.tritonapi;
 
-        if (opts.json) {
-            console.log(JSON.stringify(net));
-        } else {
-            console.log(JSON.stringify(net, null, 4));
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-        cb();
+        tritonapi.getNetwork(args[0], function (err, net) {
+            if (err) {
+                return cb(err);
+            }
+
+            if (opts.json) {
+                console.log(JSON.stringify(net));
+            } else {
+                console.log(JSON.stringify(net, null, 4));
+            }
+            cb();
+        });
     });
 }
 
