commit ed3cf3a044164924be014bfaf660da75f29f059c (refs/changes/82/3282/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-01-31T20:22:40+00:00 (1 year, 8 months ago)
    
    OS-6582: libfakekernel changes break SmartOS platform build

diff --git a/nss-nspr/Makefile b/nss-nspr/Makefile
index b006810..69e6c2e 100644
--- a/nss-nspr/Makefile
+++ b/nss-nspr/Makefile
@@ -10,7 +10,7 @@
 #
 
 #
-# Copyright 2016 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 #
 
 VER =	nss-3.25
@@ -49,3 +49,10 @@ all_64: $(VER.64)/$(UNPACK_SENTINEL)
 install: all
 	DESTDIR=$(DESTDIR) ksh93 ./install-nss $(VER.32)
 	DESTDIR=$(DESTDIR) MACH64=amd64 ksh93 ./install-nss-64 $(VER.64)
+	# Apology in advance; see OS-6582 for why we're doing this.
+	if [[ ! -f $(DESTDIR)/usr/include/zconf.h.orig ]]; then \
+		cp $(DESTDIR)/usr/include/zconf.h \
+		   $(DESTDIR)/usr/include/zconf.h.orig; \
+		$(GPATCH) -d $(DESTDIR)/usr/include -p$(PATCHSTRIP) \
+		    <zconf.patch; \
+	fi
diff --git a/nss-nspr/zconf.patch b/nss-nspr/zconf.patch
new file mode 100644
index 0000000..c759e97
--- /dev/null
+++ b/nss-nspr/zconf.patch
@@ -0,0 +1,21 @@
+--- a/zconf.h	Wed Jan 31 18:16:08 2018
++++ b/zconf.h	Wed Jan 31 18:17:11 2018
+@@ -8,6 +8,8 @@
+ #ifndef ZCONF_H
+ #define ZCONF_H
+ 
++#include <sys/machelf.h>
++
+ /*
+  * If you *really* need a unique prefix for all types and library functions,
+  * compile with -DZ_PREFIX. The "standard" zlib should be compiled without it.
+@@ -257,9 +259,6 @@
+ #  define FAR
+ #endif
+ 
+-#if !defined(__MACTYPES__)
+-typedef unsigned char  Byte;  /* 8 bits */
+-#endif
+ typedef unsigned int   uInt;  /* 16 bits or more */
+ typedef unsigned long  uLong; /* 32 bits or more */
+ 
