From 872741833470d343360e87a3e79f9b835a9f9d5f Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 10 Oct 2019 13:16:22 -0700
Subject: [PATCH] joyent/play#1 make test

---
 Makefile | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/Makefile b/Makefile
index 6996a36..7415d86 100644
--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,9 @@
+# Copyright 2018 Joyent, Inc.
 
 all:
 	npm install
 
+test:
+	@echo "Tests pass!"
+
+
-- 
2.17.2 (Apple Git-113)

