From 9acaf3feb0ebd0987945bd6663f379491ecd1613 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Tue, 18 Jul 2017 10:35:16 +0000
Subject: [PATCH] MANATEE-304 manatee doesn't always come up with very large
 datasets

---
 lib/postgresMgr.js | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/lib/postgresMgr.js b/lib/postgresMgr.js
index 4efe456..b51b6af 100644
--- a/lib/postgresMgr.js
+++ b/lib/postgresMgr.js
@@ -1419,11 +1419,7 @@ PostgresMgr.prototype._start = function _start(cb) {
                         opsTimeout: self._opsTimeout,
                         postgresPath: self._postgresPath,
                         dataDir: self._dataDir
-                    }, 'PostgresMgr.start: start timeout');
-
-                    self._stop(function () {
-                        return callback(err, postgres);
-                    });
+                    }, 'PostgresMgr.start: start is taking a long time');
                 }
             } else {
                 log.info({
-- 
2.21.0

