From d43de46a016e9a9cf2e11de44a60090d7710a40e Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Fri, 4 Nov 2016 23:27:34 +1300
Subject: [PATCH] PUBAPI-1345: Server hangs on close when started from some
 tests

---
 lib/app.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/app.js b/lib/app.js
index 319c2c4..86f59fa 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -691,6 +691,8 @@ module.exports = {
                 server.on('close', function () {
                     clearInterval(globalAgentInterval);
 
+                    kangServer.close();
+
                     clients.vmapi.close();
                     clients.napi.close();
                     clients.fwapi.close();
-- 
2.21.0

