commit 39ba6c002d647278e08ded23a21efbebddba02c7 (refs/changes/47/1647/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-03-10T09:31:39-08:00 (2 years, 7 months ago)
    
    DOCKER-1013 broke purpose-less certificates in DOCKER-1005

diff --git a/lib/auth.js b/lib/auth.js
index 5ab41d7..68fb2c7 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -73,7 +73,7 @@ function authTls(req, res, next) {
      * the expiry time and use the issuer's CN to look up the real account
      * key.
      */
-    if (cert.purposes.indexOf('joyentDocker') !== -1) {
+    if (cert.purposes && cert.purposes.indexOf('joyentDocker') !== -1) {
         log.trace('found "joyentDocker" certificate purpose, will treat as '
             + 'new-style certificate');
 
