commit e04d04ab1e3d56bedab82c5340d060daae091568 (refs/changes/15/1015/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:06-05:00 (2 years, 10 months ago)
    
    ./bin/triton key add -f ~/.ssh/id_rsa_atest.pub

diff --git a/lib/do_key/do_add.js b/lib/do_key/do_add.js
index 15b5d5a..b868b53 100644
--- a/lib/do_key/do_add.js
+++ b/lib/do_key/do_add.js
@@ -40,7 +40,8 @@ function do_add(subcmd, opts, args, cb) {
     var filePath = args[0];
     var cli = this.top;
 
-    vasync.pipeline({arg: {}, funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
         function gatherDataStdin(ctx, next) {
             if (filePath !== '-') {
                 return next();
