commit b5fb299fb0d1d7c8baec88142caa6943db502a40 (refs/changes/92/4192/2)
Author: Jason King <jason.king@joyent.com>
Date:   2018-06-12T18:55:11-05:00 (1 year, 4 months ago)
    
    libvard_inject_varpd needs to know about routers

diff --git a/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c b/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
index 28b4005dbe..bd5ea51763 100644
--- a/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
+++ b/usr/src/lib/varpd/libvarpd/common/libvarpd_overlay.c
@@ -542,6 +542,8 @@ libvarpd_inject_varp(varpd_provider_handle_t *vph, const uint8_t *mac,
 
 	otc.otc_linkid = inst->vri_linkid;
 	otc.otc_entry.otce_flags = 0;
+	if (IN6_IS_ADDR_UNSPECIFIED(&otp->otp_ip) && otp->otp_port == 0)
+		otc.otc_entry.otce_flags |= OVERLAY_TARGET_CACHE_ROUTER;
 	bcopy(mac, otc.otc_entry.otce_mac.otm_mac, ETHERADDRL);
 	bcopy(otp, &otc.otc_entry.otce_dest, sizeof (overlay_target_point_t));
 
