commit 07338a7e5b5e046620dfbb7c81e33babcb31cfc5 (refs/changes/41/4841/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-09-14T20:16:49+00:00 (1 year, 1 month ago)
    
    OS-7250 undo c99 change in dkio.h

diff --git a/usr/src/cmd/allocate/Makefile b/usr/src/cmd/allocate/Makefile
index 7ec0367c70..41339cdde6 100644
--- a/usr/src/cmd/allocate/Makefile
+++ b/usr/src/cmd/allocate/Makefile
@@ -21,7 +21,6 @@
 
 #
 # Copyright (c) 1992, 2010, Oracle and/or its affiliates. All rights reserved.
-# Copyright 2018 Joyent, Inc.
 #
 
 ETCTSOLFILES =	devalloc_defaults
@@ -183,11 +182,7 @@ clean :
 		$(RM) $(PROG) $(RTLCKS) $(OBJS) \
 		$(SCRIPTS) $(ALLSCRIPTS) $(POFILE) $(POFILES)
 
-lint :
-		$(LINT.c) allocate.c allocate3.c $(LDLIBS)
-		$(LINT.c) -Xc99=%all mkdevalloc.c $(LDLIBS)
-		$(LINT.c) dminfo.c $(LDLIBS)
-		$(LINT.c) add_allocatable.c $(LDLIBS)
+lint :		lint_SRCS
 
 strip :
 		$(STRIP) $(PROG)
diff --git a/usr/src/cmd/cdrw/Makefile b/usr/src/cmd/cdrw/Makefile
index 5f8527ad3c..2b626aeb59 100644
--- a/usr/src/cmd/cdrw/Makefile
+++ b/usr/src/cmd/cdrw/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG= cdrw
@@ -42,7 +41,6 @@ CERRWARN +=	-_gcc=-Wno-uninitialized
 CERRWARN +=	-_gcc=-Wno-parentheses
 CPPFLAGS +=	-D_REENTRANT -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 $(HAL_DBUS_CPPFLAGS)
 LINTFLAGS += 	-um
-C99LMODE=	-Xc99=%all
 
 $(ROOTBIN)/cdrw := FILEMODE = 04755
 
diff --git a/usr/src/cmd/devfsadm/Makefile.com b/usr/src/cmd/devfsadm/Makefile.com
index 95591b2a95..1585db2894 100644
--- a/usr/src/cmd/devfsadm/Makefile.com
+++ b/usr/src/cmd/devfsadm/Makefile.com
@@ -20,7 +20,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 # This target builds both a command (daemon) and various shared objects.  This
@@ -96,8 +95,6 @@ LINTFLAGS += -erroff=E_NAME_USED_NOT_DEF2
 LINTFLAGS += -erroff=E_NAME_DEF_NOT_USED2
 LINTFLAGS += -erroff=E_NAME_MULTIPLY_DEF2
 
-C99LMODE=	-Xc99=%all
-
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-char-subscripts
 CERRWARN += -_gcc=-Wno-parentheses
diff --git a/usr/src/cmd/devinfo/Makefile b/usr/src/cmd/devinfo/Makefile
index 360f4b01c0..58b5c56131 100644
--- a/usr/src/cmd/devinfo/Makefile
+++ b/usr/src/cmd/devinfo/Makefile
@@ -22,7 +22,6 @@
 #
 # Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 # cmd/devinfo/Makefile
 #
@@ -34,7 +33,6 @@ include ../Makefile.cmd
 LDLIBS += -ladm -lefi
 CFLAGS += $(CCVERBOSE)
 CERRWARN += -_gcc=-Wno-uninitialized
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/diskscan/Makefile b/usr/src/cmd/diskscan/Makefile
index dacf787745..fc82a22936 100644
--- a/usr/src/cmd/diskscan/Makefile
+++ b/usr/src/cmd/diskscan/Makefile
@@ -20,17 +20,15 @@
 # CDDL HEADER END
 #
 #
+#ident	"%Z%%M%	%I%	%E% SMI"
 #
 # Copyright (c) 1989, 2001 by Sun Microsystems, Inc.
 # All rights reserved.
-# Copyright 2018 Joyent, Inc.
 
 PROG= diskscan
 
 include ../Makefile.cmd
 
-C99LMODE=	-Xc99=%all
-
 .KEEP_STATE:
 
 all: $(PROG)
diff --git a/usr/src/cmd/eject/Makefile b/usr/src/cmd/eject/Makefile
index d47ddbf29c..33ae2987a6 100644
--- a/usr/src/cmd/eject/Makefile
+++ b/usr/src/cmd/eject/Makefile
@@ -20,7 +20,6 @@
 #
 #
 # Copyright (c) 2006, 2010, Oracle and/or its affiliates. All rights reserved.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG=		eject
@@ -32,7 +31,6 @@ LDLIBS += -lvolmgt -ladm
 CFLAGS += $(CCVERBOSE)
 
 LINTFLAGS += -um
-C99LMODE= -Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/fdformat/Makefile b/usr/src/cmd/fdformat/Makefile
index 28c1ebff2d..8a04df934e 100644
--- a/usr/src/cmd/fdformat/Makefile
+++ b/usr/src/cmd/fdformat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 # cmd/fdformat/Makefile
 #
@@ -32,7 +31,6 @@ include ../Makefile.cmd
 
 LDLIBS +=	-lvolmgt
 LINTFLAGS +=    -um
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/fdisk/Makefile b/usr/src/cmd/fdisk/Makefile
index 6a75b4f4dd..ff449b4de5 100644
--- a/usr/src/cmd/fdisk/Makefile
+++ b/usr/src/cmd/fdisk/Makefile
@@ -20,7 +20,6 @@
 #
 #
 # Copyright (c) 1992, 2010, Oracle and/or its affiliates. All rights reserved.
-# Copyright 2018 Joyent, Inc.
 #
 #
 
@@ -43,7 +42,6 @@ i386_CFLAGS += -D_LARGEFILE64_SOURCE
 i386_CFLAGS += -D_FILE_OFFSET_BITS=64
 
 LINTFLAGS += -erroff=E_BAD_PTR_CAST_ALIGN -erroff=E_NAME_DEF_NOT_USED2
-C99LMODE= -Xc99=%all
 
 all:	$(ROOTFS_PROG)
 
diff --git a/usr/src/cmd/fmthard/Makefile b/usr/src/cmd/fmthard/Makefile
index 40047f2e82..4f1205de7a 100644
--- a/usr/src/cmd/fmthard/Makefile
+++ b/usr/src/cmd/fmthard/Makefile
@@ -20,7 +20,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 # cmd/fmthard/Makefile
 #
@@ -33,7 +32,6 @@ RELUSRSBIN=     ../usr/sbin
 ROOTSYMLINK=    $(ROOTETC)/$(PROG)
 LDLIBS +=	-ladm -lefi
 lint := LINTFLAGS = -u -erroff=E_BAD_FORMAT_ARG_TYPE2
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/format/Makefile b/usr/src/cmd/format/Makefile
index 4835294a0b..34f25b7cc2 100644
--- a/usr/src/cmd/format/Makefile
+++ b/usr/src/cmd/format/Makefile
@@ -23,7 +23,6 @@
 # Use is subject to license terms.
 #
 # Copyright 2015 Nexenta Systems, Inc. All rights reserved.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG= format
@@ -64,7 +63,6 @@ LDLIBS_sparc=
 LDLIBS +=	-ladm -lefi -ldiskmgt -lnvpair -ldevid -ldevinfo $(LDLIBS_$(MACH))
 
 LINTFLAGS += -xerroff=E_NAME_DEF_NOT_USED2
-C99LMODE=	-Xc99=%all
 CPPFLAGS    += -D_EXTVTOC
 
 .KEEP_STATE:
diff --git a/usr/src/cmd/lofiadm/Makefile b/usr/src/cmd/lofiadm/Makefile
index 32323488e0..07d01206a2 100644
--- a/usr/src/cmd/lofiadm/Makefile
+++ b/usr/src/cmd/lofiadm/Makefile
@@ -22,7 +22,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG= lofiadm
@@ -42,7 +41,6 @@ CERRWARN += -_gcc=-Wno-parentheses
 CERRWARN += -_gcc=-Wno-uninitialized
 
 LINTFLAGS += -erroff=E_NAME_USED_NOT_DEF2
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/prtvtoc/Makefile b/usr/src/cmd/prtvtoc/Makefile
index 6520df4125..59db9914bd 100644
--- a/usr/src/cmd/prtvtoc/Makefile
+++ b/usr/src/cmd/prtvtoc/Makefile
@@ -20,7 +20,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 # cmd/prtvtoc/Makefile
 #
@@ -36,8 +35,6 @@ LDLIBS +=	-ladm -lefi
 CERRWARN +=	-_gcc=-Wno-parentheses
 CERRWARN +=	-_gcc=-Wno-uninitialized
 
-C99LMODE=	-Xc99=%all
-
 .KEEP_STATE:
 
 all: $(PROG) 
diff --git a/usr/src/cmd/ramdiskadm/Makefile b/usr/src/cmd/ramdiskadm/Makefile
index d166e088df..7dae74826e 100644
--- a/usr/src/cmd/ramdiskadm/Makefile
+++ b/usr/src/cmd/ramdiskadm/Makefile
@@ -22,7 +22,6 @@
 #
 # Copyright 2002-2003 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG= ramdiskadm
@@ -35,8 +34,6 @@ include ../Makefile.cmd
 LDLIBS	+= -ldevinfo
 CERRWARN += -_gcc=-Wno-uninitialized
 
-C99LMODE=	-Xc99=%all
-
 .KEEP_STATE:
 
 all: $(PROG)
diff --git a/usr/src/cmd/rmformat/Makefile b/usr/src/cmd/rmformat/Makefile
index 6650d79024..91634d31fa 100644
--- a/usr/src/cmd/rmformat/Makefile
+++ b/usr/src/cmd/rmformat/Makefile
@@ -22,7 +22,6 @@
 #
 # Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG= rmformat
@@ -38,7 +37,6 @@ LDLIBS +=	-lsmedia -lvolmgt -ladm -lefi
 CERRWARN += -_gcc=-Wno-uninitialized
 
 LINTFLAGS += -u
-C99LMODE=	-Xc99=%all
 CPPFLAGS += -D_FILE_OFFSET_BITS=64
 
 $(ROOTBIN)/rmformat := FILEMODE = 04555
diff --git a/usr/src/cmd/smserverd/Makefile b/usr/src/cmd/smserverd/Makefile
index 467aa81d37..f138639025 100644
--- a/usr/src/cmd/smserverd/Makefile
+++ b/usr/src/cmd/smserverd/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG =	rpc.smserverd
@@ -49,7 +48,6 @@ CFLAGS += $(CCVERBOSE)
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-unused-variable
 LINTFLAGS += -u
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/cmd/stat/fsstat/Makefile b/usr/src/cmd/stat/fsstat/Makefile
index 5b17d933de..b86e2cc8f7 100644
--- a/usr/src/cmd/stat/fsstat/Makefile
+++ b/usr/src/cmd/stat/fsstat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG = fsstat
@@ -37,7 +36,6 @@ COMMON_SRCS = $(COMMON_OBJS:%.o=$(STATCOMMONDIR)/%.c)
 LDLIBS += -lkstat -lcmdutils
 CFLAGS += $(CCVERBOSE) -I${STATCOMMONDIR}
 CERRWARN += -_gcc=-Wno-parentheses
-C99LMODE=	-Xc99=%all
 FILEMODE= 0555
 
 lint := LINTFLAGS = -muxs -I$(STATCOMMONDIR)
diff --git a/usr/src/cmd/stat/iostat/Makefile b/usr/src/cmd/stat/iostat/Makefile
index 80f6e31a67..36142d5a8a 100644
--- a/usr/src/cmd/stat/iostat/Makefile
+++ b/usr/src/cmd/stat/iostat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG = iostat
@@ -33,7 +32,6 @@ include $(SRC)/cmd/stat/Makefile.stat
 
 LDLIBS += -lkstat -ldevinfo -lavl
 CFLAGS += $(CCVERBOSE) -I${STATCOMMONDIR}
-C99LMODE=	-Xc99=%all
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-switch
 CERRWARN += -_gcc=-Wno-parentheses
diff --git a/usr/src/cmd/stat/kstat/Makefile b/usr/src/cmd/stat/kstat/Makefile
index fd5ac35021..97b1863e08 100644
--- a/usr/src/cmd/stat/kstat/Makefile
+++ b/usr/src/cmd/stat/kstat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG = kstat
@@ -33,7 +32,6 @@ include $(SRC)/cmd/stat/Makefile.stat
 
 LDLIBS += -lavl -lcmdutils -ldevinfo -lgen -lkstat
 CFLAGS += $(CCVERBOSE) -I${STATCOMMONDIR}
-C99LMODE=	-Xc99=%all
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-switch
 CERRWARN += -_gcc=-Wno-parentheses
diff --git a/usr/src/cmd/stat/mpstat/Makefile b/usr/src/cmd/stat/mpstat/Makefile
index 27eed29282..fd3d57dafb 100644
--- a/usr/src/cmd/stat/mpstat/Makefile
+++ b/usr/src/cmd/stat/mpstat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG = mpstat
@@ -33,7 +32,6 @@ include $(SRC)/cmd/stat/Makefile.stat
 
 LDLIBS += -ldevinfo -lkstat -lavl
 CFLAGS += $(CCVERBOSE) -I${STATCOMMONDIR}
-C99LMODE=	-Xc99=%all
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-switch
 CERRWARN += -_gcc=-Wno-parentheses
diff --git a/usr/src/cmd/stat/vfsstat/Makefile b/usr/src/cmd/stat/vfsstat/Makefile
index a08eed8a52..04b5085243 100644
--- a/usr/src/cmd/stat/vfsstat/Makefile
+++ b/usr/src/cmd/stat/vfsstat/Makefile
@@ -18,7 +18,7 @@
 #
 # CDDL HEADER END
 #
-# Copyright (c) 2018, Joyent, Inc. All rights reserved.
+# Copyright (c) 2011, Joyent, Inc. All rights reserved.
 #
 
 include $(SRC)/cmd/Makefile.cmd
@@ -27,8 +27,6 @@ PROG=	vfsstat
 
 .KEEP_STATE:
 
-C99LMODE=	-Xc99=%all
-
 all: $(PROG)
 
 install: all .WAIT $(ROOTPROG)
diff --git a/usr/src/cmd/stat/vmstat/Makefile b/usr/src/cmd/stat/vmstat/Makefile
index 092537ccdd..d6f65b98d0 100644
--- a/usr/src/cmd/stat/vmstat/Makefile
+++ b/usr/src/cmd/stat/vmstat/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 PROG = vmstat
@@ -33,7 +32,6 @@ include $(SRC)/cmd/stat/Makefile.stat
 
 LDLIBS += -ldevinfo -lkstat -lavl
 CFLAGS += $(CCVERBOSE) -I${STATCOMMONDIR}
-C99LMODE=	-Xc99=%all
 CERRWARN += -_gcc=-Wno-switch
 CERRWARN += -_gcc=-Wno-uninitialized
 CERRWARN += -_gcc=-Wno-parentheses
diff --git a/usr/src/cmd/stmsboot/Makefile b/usr/src/cmd/stmsboot/Makefile
index cda3ad09d1..8e97f0f467 100644
--- a/usr/src/cmd/stmsboot/Makefile
+++ b/usr/src/cmd/stmsboot/Makefile
@@ -21,7 +21,6 @@
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 MANIFEST = mpxio-upgrade.xml
@@ -42,7 +41,6 @@ ROOTFSLIB_MPXIO = $(ROOT)/lib/mpxio
 ROOTMANIFESTDIR = $(ROOTSVCSYSTEMDEVICE)
 LDLIBS += -ldevinfo -ldevid -lnvpair -lumem
 CERRWARN += -_gcc=-Wno-uninitialized
-C99LMODE=	-Xc99=%all
 
 .KEEP_STATE:
 
diff --git a/usr/src/lib/libadm/Makefile.com b/usr/src/lib/libadm/Makefile.com
index a496655a3f..4631c50bf4 100644
--- a/usr/src/lib/libadm/Makefile.com
+++ b/usr/src/lib/libadm/Makefile.com
@@ -20,7 +20,6 @@
 #
 #
 # Copyright (c) 1997, 2010, Oracle and/or its affiliates. All rights reserved.
-# Copyright 2018 Joyent, Inc.
 #
 
 LIBRARY=	libadm.a
@@ -43,7 +42,6 @@ LIBS=		$(DYNLIB) $(LINTLIB)
 SRCDIR=		../common
 
 CPPFLAGS +=	-I ../inc
-CSTD =		$(CSTD_GNU99)
 
 CERRWARN +=	-_gcc=-Wno-parentheses
 CERRWARN +=	-_gcc=-Wno-uninitialized
diff --git a/usr/src/lib/libdevid/Makefile.com b/usr/src/lib/libdevid/Makefile.com
index fe065fc59a..cfef8fb21f 100644
--- a/usr/src/lib/libdevid/Makefile.com
+++ b/usr/src/lib/libdevid/Makefile.com
@@ -43,8 +43,6 @@ LDLIBS +=	-ldevinfo -lc
 
 CFLAGS +=	$(CCVERBOSE)
 
-CSTD =	$(CSTD_GNU99)
-
 .KEEP_STATE:
 
 all: $(LIBS)
diff --git a/usr/src/lib/libdiskmgt/Makefile.com b/usr/src/lib/libdiskmgt/Makefile.com
index 03a779d11e..4a8ee2069c 100644
--- a/usr/src/lib/libdiskmgt/Makefile.com
+++ b/usr/src/lib/libdiskmgt/Makefile.com
@@ -66,8 +66,6 @@ CERRWARN +=	-_gcc=-Wno-parentheses
 CERRWARN +=	-_gcc=-Wno-uninitialized
 CPPFLAGS +=	-D_REENTRANT -I$(SRC)/lib/libdiskmgt/common
 
-CSTD =	$(CSTD_GNU99)
-
 .KEEP_STATE:
 
 all: $(LIBS)
diff --git a/usr/src/lib/libefi/Makefile.com b/usr/src/lib/libefi/Makefile.com
index 97a4d579f4..3fc915acdd 100644
--- a/usr/src/lib/libefi/Makefile.com
+++ b/usr/src/lib/libefi/Makefile.com
@@ -22,7 +22,6 @@
 # Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
 # Copyright 2018 OmniOS Community Edition (OmniOSce) Association.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
 
 LIBRARY =	libefi.a
@@ -41,7 +40,6 @@ LDLIBS +=	-luuid -lsmbios -lc
 $(LINTLIB) :=	SRCS = $(SRCDIR)/$(LINTSRC)
 CFLAGS +=	$(CCVERBOSE)
 
-CSTD =		$(CSTD_GNU99)
 LINTFLAGS64 +=	-errchk=longptr64
 
 .KEEP_STATE:
diff --git a/usr/src/lib/libfdisk/Makefile.com b/usr/src/lib/libfdisk/Makefile.com
index 198669d353..d2ac8d3d39 100644
--- a/usr/src/lib/libfdisk/Makefile.com
+++ b/usr/src/lib/libfdisk/Makefile.com
@@ -50,8 +50,6 @@ CFLAGS += -D_FILE_OFFSET_BITS=64
 CFLAGS64 += -D_LARGEFILE64_SOURCE
 CFLAGS64 += -D_FILE_OFFSET_BITS=64
 
-CSTD =	$(CSTD_GNU99)
-
 LINTFLAGS +=    -erroff=E_BAD_PTR_CAST_ALIGN
 LINTFLAGS64 +=    -erroff=E_BAD_PTR_CAST_ALIGN
 
diff --git a/usr/src/lib/libsmedia/library/Makefile.com b/usr/src/lib/libsmedia/library/Makefile.com
index bcdb37b2e7..4807392165 100644
--- a/usr/src/lib/libsmedia/library/Makefile.com
+++ b/usr/src/lib/libsmedia/library/Makefile.com
@@ -56,8 +56,6 @@ CFLAGS64 +=	$(CCVERBOSE)
 
 CERRWARN +=	-_gcc=-Wno-unused-variable
 
-CSTD =	$(CSTD_GNU99)
-
 LDLIBS +=	-lnsl -lc
 
 .KEEP_STATE:
diff --git a/usr/src/lib/libsmedia/plugins/Makefile.com b/usr/src/lib/libsmedia/plugins/Makefile.com
index d31896364e..f1fa1c19e2 100644
--- a/usr/src/lib/libsmedia/plugins/Makefile.com
+++ b/usr/src/lib/libsmedia/plugins/Makefile.com
@@ -21,8 +21,8 @@
 #
 # Copyright 2006 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
-# Copyright 2018 Joyent, Inc.
 #
+# ident	"%Z%%M%	%I%	%E% SMI"
 #
 
 LIBS =		$(DYNLIB)
@@ -43,7 +43,6 @@ CLEANFILES=	$(LINTOUT) $(LINTPLUGIN)
 
 SRCDIR =	../common
 
-CSTD =		$(CSTD_GNU99)
 LINTFLAGS += -xu
 LINTFLAGS64 += -xu
 $(LINTPLUGIN) :=	SRCS=$(OBJECTS:%.o=../common/%.c)
diff --git a/usr/src/uts/common/sys/dkio.h b/usr/src/uts/common/sys/dkio.h
index dcb817fc72..7a9bbc3f15 100644
--- a/usr/src/uts/common/sys/dkio.h
+++ b/usr/src/uts/common/sys/dkio.h
@@ -537,11 +537,11 @@ typedef struct dkioc_free_list_s {
 	uint64_t		dfl_flags;
 	uint64_t		dfl_num_exts;
 	uint64_t		dfl_offset;
-	dkioc_free_list_ext_t	dfl_exts[];
+	dkioc_free_list_ext_t	dfl_exts[1];
 } dkioc_free_list_t;
 #define	DFL_SZ(num_exts) \
 	(sizeof (dkioc_free_list_t) + \
-	(num_exts) * sizeof (dkioc_free_list_ext_t))
+	(num_exts - 1) * sizeof (dkioc_free_list_ext_t))
 
 #ifdef	__cplusplus
 }
