From f92cec52486d748c35d8d95074cccf33187171e1 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Mon, 29 Aug 2016 11:58:58 -0700
Subject: [PATCH] DOCKER-921 Get sdc-docker-build 'make check' running on
 SmartOS

---
 Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile b/Makefile
index 12046ae..f47a322 100644
--- a/Makefile
+++ b/Makefile
@@ -19,6 +19,8 @@ JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 CLEAN_FILES += ./node_modules
 
+NODE_PREBUILT_VERSION=v0.10.46
+
 include ./tools/mk/Makefile.defs
 ifeq ($(shell uname -s),SunOS)
 	include ./tools/mk/Makefile.node_prebuilt.defs
-- 
2.21.0

