commit fb6319c5f124413e120b0ad6ae7244704eda7c8d (refs/changes/83/4483/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-07-04T15:37:22-07:00 (1 year, 3 months ago)
    
    TRITON-568 CNAPI should stop using restify.authorizationParser()

diff --git a/lib/server.js b/lib/server.js
index aed3911..acd0644 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var os = require('os');
@@ -52,7 +52,6 @@ function createServer(options) {
     });
 
     cnapi.use(restify.acceptParser(cnapi.acceptable));
-    cnapi.use(restify.authorizationParser());
     cnapi.use(restify.queryParser({
         mapParams: true,
         allowDots: false,
diff --git a/package.json b/package.json
index 6b55537..905b83f 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.10.0",
+  "version": "1.10.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
