From a4e5b0286ce503a27872f5817eae0605859f109c Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Thu, 10 Oct 2019 10:45:37 +0000
Subject: [PATCH] OS-8020 fswatcher can't find its own errno

---
 src/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/Makefile b/src/Makefile
index 5125aceb..b7392529 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -64,6 +64,7 @@ SMARTDCLIBDIR =	$(DESTDIR)/smartdc/lib
 #
 bootparams :	LIBS +=		-ldevinfo
 bootparams :	WARN_FLAGS +=	-Wno-unused
+fswatcher :	CPPFLAGS +=	-D_REENTRANT
 fswatcher :	LIBS +=		-lthread -lnvpair -lavl
 zfs_recv :	LIBS +=		-lsocket
 zfs_send :	LIBS +=		-lsocket
-- 
2.17.2 (Apple Git-113)

