commit 446a1749a48fc89cbb5e2859aadef948ea76b1dd (refs/changes/45/4445/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-06-26T16:47:13-05:00 (1 year, 3 months ago)
    
    Fix LOG vlan field ordering

diff --git a/usr/src/lib/varpd/svp/common/libvarpd_svp_prot.h b/usr/src/lib/varpd/svp/common/libvarpd_svp_prot.h
index 08dfb99f41..25a626afd1 100644
--- a/usr/src/lib/varpd/svp/common/libvarpd_svp_prot.h
+++ b/usr/src/lib/varpd/svp/common/libvarpd_svp_prot.h
@@ -213,8 +213,8 @@ typedef struct svp_log_route {
 	uint32_t	svlr_dcid;	/* Remote/dest Data Center ID. */
 	uint8_t		svlr_srcip[16];	/* Source IP address base. */
 	uint8_t		svlr_dstip[16];	/* Destination IP address base. */
-	uint16_t	svlr_dst_vlan;	/* Source VLAN id. */
-	uint16_t	svlr_src_vlan;	/* Destination VLAN id. */
+	uint16_t	svlr_src_vlan;	/* Source VLAN id. */
+	uint16_t	svlr_dst_vlan;	/* Destination VLAN id. */
 	uint8_t		svlr_src_prefixlen;	/* Source IP prefix length. */
 	uint8_t		svlr_dst_prefixlen;	/* Dest. IP prefix length. */
 	uint16_t	svlr_pad;	/* So we can be aligned... */
