commit 851d68d83ed1dc6d5a037968716176aa4e8d9c4d (refs/changes/48/5548/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-08T11:29:34+00:00 (8 months ago)
    
    TRITON-1152 convert sdcboot to engbld framework

diff --git a/.gitignore b/.gitignore
index ee60d7b..7274c63 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,3 +4,5 @@ src/int14.o
 src/int14.elf
 src/int14.com
 tools/mtools-4.0.18
+ipxe
+bits
diff --git a/.gitmodules b/.gitmodules
index 4e27616..d0ea914 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -7,3 +7,6 @@
 [submodule "ipxe"]
 	path = ipxe
 	url = https://github.com/joyent/ipxe
+[submodule "deps/eng"]
+	path = deps/eng
+	url = git@github.com:joyent/eng.git
diff --git a/GNUmakefile b/Makefile
similarity index 89%
rename from GNUmakefile
rename to Makefile
index c9b58d3..010beb4 100644
--- a/GNUmakefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 NAME =	sdcboot
@@ -19,7 +19,7 @@ BOOT_ROOT =	$(ROOT)/boot
 RELEASE_TARBALL =	$(NAME)-$(STAMP).tgz
 CLEAN_FILES += \
 	$(ROOT) \
-	$(NAME)-pkg-*.tar.gz \
+	$(NAME)-*.tgz \
 	src/*.o \
 	src/*.elf \
 	src/*.com \
@@ -59,7 +59,7 @@ IPXE_ENV = \
 MAPFILE =	mapfile-dos
 CC =		/opt/local/bin/gcc
 LD =		/usr/bin/ld
-TAR =		/opt/local/bin/tar
+TAR =		/usr/bin/gtar
 MKDIR =		/usr/bin/mkdir
 MKFILE =	/usr/sbin/mkfile
 CP =		/usr/bin/cp
@@ -196,7 +196,8 @@ ROOT_FREEDOS = \
 
 ROOT_BOOT =	$(ROOT_BOOT_BINS)
 
-include ./tools/mk/Makefile.defs
+ENGBLD_REQUIRE	:= $(shell git submodule update --init deps/eng)
+include ./deps/eng/tools/mk/Makefile.defs
 
 $(FREEDOS_ROOT)/autoexec.bat :	FILEMODE = 755
 $(FREEDOS_ROOT)/joyent/%.com :	FILEMODE = 755
@@ -300,25 +301,22 @@ ipxe.clean:
 release: $(RELEASE_TARBALL)
 
 $(RELEASE_TARBALL): pkg
-	(cd $(ROOT); $(TAR) czf $(TOP)/$(RELEASE_TARBALL) .)
+	(cd $(ROOT); $(TAR) -I pigz -cf $(TOP)/$(RELEASE_TARBALL) .)
 
-publish: prepublish $(BITS_DIR)/$(NAME)/$(RELEASE_TARBALL)
-
-.PHONY: prepublish
-prepublish:
-	@if [[ -z "$(BITS_DIR)" ]]; then \
-		echo "error: 'BITS_DIR' must be set for 'publish' target"; \
+publish: $(ENGBLD_BITS_DIR)/$(NAME)/$(RELEASE_TARBALL)
+	@if [[ -z "$(ENGBLD_BITS_DIR)" ]]; then \
+		echo "error: 'ENGBLD_BITS_DIR' must be set for 'publish' target"; \
 		exit 1; \
 	fi
-	@if [[ ! -d "$(BITS_DIR)" ]]; then \
-		echo "error: $(BITS_DIR) is not a directory"; \
+	@if [[ ! -d "$(ENGBLD_BITS_DIR)" ]]; then \
+		echo "error: $(ENGBLD_BITS_DIR) is not a directory"; \
 		exit 1; \
 	fi
 
-$(BITS_DIR)/$(NAME)/$(RELEASE_TARBALL): $(RELEASE_TARBALL) | $(BITS_DIR)/$(NAME)
+$(ENGBLD_BITS_DIR)/$(NAME)/$(RELEASE_TARBALL): $(RELEASE_TARBALL) | $(ENGBLD_BITS_DIR)/$(NAME)
 	$(INS.file)
 
-$(BITS_DIR)/$(NAME):
+$(ENGBLD_BITS_DIR)/$(NAME):
 	$(INS.dir)
 
-include ./tools/mk/Makefile.targ
+include ./deps/eng/tools/mk/Makefile.targ
diff --git a/deps/eng b/deps/eng
new file mode 160000
index 0000000..7c472bc
--- /dev/null
+++ b/deps/eng
@@ -0,0 +1 @@
+Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
