commit 74e67520351c92c2a81bc02ac1ec72439ed48cd6 (refs/changes/00/3500/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-02-27T21:46:21+00:00 (1 year, 7 months ago)
    
    OS-6709 acsolaris.h should avoid non-portable __unused

diff --git a/usr/src/uts/intel/sys/acpi/platform/acsolaris.h b/usr/src/uts/intel/sys/acpi/platform/acsolaris.h
index 3931916d77..b19007c589 100644
--- a/usr/src/uts/intel/sys/acpi/platform/acsolaris.h
+++ b/usr/src/uts/intel/sys/acpi/platform/acsolaris.h
@@ -70,7 +70,7 @@ uint32_t acpi_strtoul(const char *, char **, int);
  * we need to have private definitions here.
  */
 #define	ACPI_PRINTF_LIKE(c)	__PRINTFLIKE(c)
-#define	ACPI_UNUSED_VAR		__unused
+#define	ACPI_UNUSED_VAR		__attribute__((__unused__))
 #define	ACPI_USE_NATIVE_DIVIDE
 #define	ACPI_FLUSH_CPU_CACHE()	(__acpi_wbinvd())
 
