From 75418baa2ba6a178e92a7e7e18a16546858e1700 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Wed, 16 Nov 2016 14:19:34 -0500
Subject: [PATCH] conventions

---
 lib/cli.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/cli.js b/lib/cli.js
index 08af61f..f9ecb68 100644
--- a/lib/cli.js
+++ b/lib/cli.js
@@ -304,8 +304,8 @@ CLI.prototype.init = function (opts, args, callback) {
     });
 
     this.__defineGetter__('tritonapi', function getTritonapi() {
-        if (self._tritonapi_err !== undefined) {
-            throw self._tritonapi_err;
+        if (self._tritonapiErr !== undefined) {
+            throw self._tritonapiErr;
         }
         return self._tritonapi;
     });
@@ -318,14 +318,14 @@ CLI.prototype.init = function (opts, args, callback) {
             getPassphrase: getpass.getPass
         });
     } catch (err) {
-        self._tritonapi_err = err;
+        self._tritonapiErr = err;
         initBase();
         return;
     }
 
     tc.init(function (err) {
         if (err) {
-            self._tritonapi_err = err;
+            self._tritonapiErr = err;
             initBase();
             return;
         }
-- 
2.21.0

