From 8777840b0e774bc54ff9c0bece164662c4a55b0c Mon Sep 17 00:00:00 2001
From: andrew <andrew@whartonlabs.com>
Date: Fri, 16 Dec 2016 17:15:34 +1100
Subject: [PATCH] Added new resizeMachine endpoint to CloudApi

---
 lib/cloudapi2.js | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index f28624c..90ebecf 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -772,6 +772,31 @@ CloudApi.prototype.getMachine = function getMachine(opts, cb) {
     });
 };
 
+/**
+ * resize a machine by id.
+ *
+ * @param {Object} opts
+ *      - id {UUID} Required. The machine id.
+ *      - {String} package. A package id, as returned from listPackages
+ * @param {Function} callback of the form `function (err, res)`
+ */
+CloudApi.prototype.resizeMachine = function resizeMachine(opts, callback) {
+    assert.uuid(opts.id, 'opts.id');
+    assert.string(opts.package, 'opts.package');
+    var data = {
+        action: 'resize',
+        package: opts.package
+    };
+
+    this._request({
+        method: 'POST',
+        path: format('/%s/machines/%s', this.account, opts.id),
+        data: data
+    }, function (err, req, res, body) {
+        callback(err, res);
+    });
+};
+
 /**
  * rename a machine by id.
  *
-- 
2.21.0

