commit 8777840b0e774bc54ff9c0bece164662c4a55b0c (refs/changes/11/1411/1)
Author: andrew <andrew@whartonlabs.com>
Date:   2017-02-02T20:05:44+11:00 (2 years, 8 months ago)
    
    Added new resizeMachine endpoint to CloudApi

diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index f28624c..90ebecf 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -772,6 +772,31 @@ CloudApi.prototype.getMachine = function getMachine(opts, cb) {
     });
 };
 
+/**
+ * resize a machine by id.
+ *
+ * @param {Object} opts
+ *      - id {UUID} Required. The machine id.
+ *      - {String} package. A package id, as returned from listPackages
+ * @param {Function} callback of the form `function (err, res)`
+ */
+CloudApi.prototype.resizeMachine = function resizeMachine(opts, callback) {
+    assert.uuid(opts.id, 'opts.id');
+    assert.string(opts.package, 'opts.package');
+    var data = {
+        action: 'resize',
+        package: opts.package
+    };
+
+    this._request({
+        method: 'POST',
+        path: format('/%s/machines/%s', this.account, opts.id),
+        data: data
+    }, function (err, req, res, body) {
+        callback(err, res);
+    });
+};
+
 /**
  * rename a machine by id.
  *
