commit e28c2a73841fd87d069340c8d331f1566958a235 (refs/changes/36/2536/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2017-09-08T23:59:09+00:00 (2 years, 1 month ago)
    
    OS-6339 lint error due to xsaveopt merge

diff --git a/usr/src/uts/intel/ia32/ml/float.s b/usr/src/uts/intel/ia32/ml/float.s
index 0d33365764..f154a96851 100644
--- a/usr/src/uts/intel/ia32/ml/float.s
+++ b/usr/src/uts/intel/ia32/ml/float.s
@@ -277,10 +277,6 @@ void
 xsave_ctxt(void *arg)
 {}
 
-void
-xsaveopt_ctxt(void *arg)
-{}
-
 /*ARGSUSED*/
 void
 xsaveopt_ctxt(void *arg)
