From 9eaa9140fcbc0cdba36c5f8580f9fdca889d3d76 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Wed, 14 Nov 2018 00:41:59 +0000
Subject: [PATCH] MANTA-4014 store snaplinkDisabled flag as ufds attribute

---
 lib/replicator/transforms/sdcperson.js | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/lib/replicator/transforms/sdcperson.js b/lib/replicator/transforms/sdcperson.js
index 4c3983a..3faa883 100644
--- a/lib/replicator/transforms/sdcperson.js
+++ b/lib/replicator/transforms/sdcperson.js
@@ -37,13 +37,18 @@ function add(opts, cb) {
         changes.triton_cns_enabled[0] &&
         changes.triton_cns_enabled[0] === 'true';
 
+    var snaplinks_disabled = changes.snaplinks_disabled &&
+        changes.snaplinks_disabled[0] &&
+        changes.snaplinks_disabled[0] === 'true';
+
     var payload = {
         type: 'account',
         uuid: uuid,
         login: login,
         groups: [],
         approved_for_provisioning: approved,
-        triton_cns_enabled: tcns
+        triton_cns_enabled: tcns,
+        snaplinks_disabled: snaplinks_disabled
     };
 
     batch.set(sprintf('/uuid/%s', uuid), JSON.stringify(payload));
@@ -119,7 +124,8 @@ function modify(opts, cb) {
         changes.forEach(function (change) {
             var type = change.modification.type;
             if (type === 'approved_for_provisioning' ||
-                type === 'triton_cns_enabled') {
+                type === 'triton_cns_enabled' ||
+                type === 'snaplinks_disabled') {
 
                 log.debug({
                     change: change
-- 
2.17.2 (Apple Git-113)

