commit 14509c7be8e7c3c6077a4795ce50296e80f468f0 (refs/changes/66/5166/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2018-11-30T13:14:48+00:00 (10 months ago)
    
    9950 Need support for Intel I219 v6-v9

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 2ebfc9cb..98b3e5c5 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -362,9 +362,17 @@ e1000g "pci8086,15a3"
 e1000g "pci8086,15b7"
 e1000g "pci8086,15b8"
 e1000g "pci8086,15b9"
+e1000g "pci8086,15bb"
+e1000g "pci8086,15bc"
+e1000g "pci8086,15bd"
+e1000g "pci8086,15be"
 e1000g "pci8086,15d6"
 e1000g "pci8086,15d7"
 e1000g "pci8086,15d8"
+e1000g "pci8086,15df"
+e1000g "pci8086,15e0"
+e1000g "pci8086,15e1"
+e1000g "pci8086,15e2"
 e1000g "pci8086,15e3"
 e1000g "pci8086,294c"
 e1000g "pci8086,f0fe"
@@ -430,9 +438,17 @@ e1000g "pciex8086,15a3"
 e1000g "pciex8086,15b7"
 e1000g "pciex8086,15b8"
 e1000g "pciex8086,15b9"
+e1000g "pciex8086,15bb"
+e1000g "pciex8086,15bc"
+e1000g "pciex8086,15bd"
+e1000g "pciex8086,15be"
 e1000g "pciex8086,15d6"
 e1000g "pciex8086,15d7"
 e1000g "pciex8086,15d8"
+e1000g "pciex8086,15df"
+e1000g "pciex8086,15e0"
+e1000g "pciex8086,15e1"
+e1000g "pciex8086,15e2"
 e1000g "pciex8086,15e3"
 e1000g "pciex8086,294c"
 e1000g "pciex8086,f0fe"
