commit 98f3b65ad838ea4089ed26f81ccc745c4d1388a9 (refs/changes/75/5175/2)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-13T10:28:10-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index b179410..c1f4238 100644
--- a/Makefile
+++ b/Makefile
@@ -47,7 +47,7 @@ ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-*-multiarch 15.4.1.
 	NODE_PREBUILT_IMAGE	= 18b094b0-eb01-11e5-80c1-175dac7ddf02
 	NODE_PREBUILT_TAG	= zone
-	NODE_PREBUILT_VERSION	:= v6.14.3
+	NODE_PREBUILT_VERSION	:= v6.15.1
 endif
 
 # RELENG-341: no npm cache is making builds unreliable
