From 11d06ea4db3c24e3e867e688301406c6e7d58dc2 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 16 Sep 2019 13:09:36 +0000
Subject: [PATCH] TOOLS-2319 eng.git never passes the gerrit bot 'make check'

---
 Makefile               | 2 ++
 tools/mk/Makefile.targ | 5 ++++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 523eaef..e0b5eef 100644
--- a/Makefile
+++ b/Makefile
@@ -129,6 +129,8 @@ include ./deps/eng/tools/mk/Makefile.smf.defs
 NODE_PREBUILT_VERSION =	v6.17.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG = zone64
+	# Use the sdcnode build for minimal-64-lts@18.4.0
+	NODE_PREBUILT_IMAGE = c2c31b00-1d60-11e9-9a77-ff9f06554b0f
 	include ./deps/eng/tools/mk/Makefile.node_prebuilt.defs
 else
 	NPM=npm
diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index 755107f..424a53b 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -235,8 +235,11 @@ deps/%/.git:
 # would invoke bash with zero files), and you can check individual files from
 # the command line with "make filename.bashchk".
 #
+# Since bashstyle is a node program, we need our node runtime as a dependency
+# for running $(BASHSTYLE)
+#
 .PHONY: check-bash
-check-bash: $(BASH_FILES:%=%.bashchk) $(BASH_FILES:%=%.bashstyle)
+check-bash: $(NODE_EXEC) $(BASH_FILES:%=%.bashchk) $(BASH_FILES:%=%.bashstyle)
 
 %.bashchk: %
 	$(BASH) -n $^
-- 
2.21.0

