From 694498e6b8c9190340d68d20ad6372483e02b51e Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 12 Sep 2019 11:25:32 +0100
Subject: [PATCH] TOOLS-2319 eng.git never passes the gerrit bot 'make check'

---
 Makefile | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/Makefile b/Makefile
index 523eaef..53bf488 100644
--- a/Makefile
+++ b/Makefile
@@ -129,6 +129,8 @@ include ./deps/eng/tools/mk/Makefile.smf.defs
 NODE_PREBUILT_VERSION =	v6.17.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG = zone64
+	# Use the sdcnode build for minimal-64-lts@18.4.0
+	NODE_PREBUILT_IMAGE = c2c31b00-1d60-11e9-9a77-ff9f06554b0f
 	include ./deps/eng/tools/mk/Makefile.node_prebuilt.defs
 else
 	NPM=npm
@@ -272,6 +274,13 @@ CLEAN_FILES += $(HELLOCTF_OBJDIR) helloctf
 test_ctf: helloctf $(STAMP_CTF_TOOLS)
 	src/testctf.sh $(CTFDUMP) ./helloctf
 
+#
+# This is only needed for eng.git itself to ensure the 'check' target
+# runs correctly when making changes to eng.git. It should not be
+# copied into repositories using this Makefile as a template.
+#
+check:: $(NODE_EXEC)
+
 #
 # Target definitions.  This is where we include the target Makefiles for
 # the "defs" Makefiles we included above.
-- 
2.21.0

