commit 5251ae8da519a4d0851330ffba398d77c931ea6c (refs/changes/46/346/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2016-08-25T15:35:18-07:00 (3 years, 1 month ago)
    
    TOOLS-1535 gerrit could be better with names and faces

diff --git a/images/appserver/etc/gerrit.config.base b/images/appserver/etc/gerrit.config.base
index d1e6159..70e32b8 100644
--- a/images/appserver/etc/gerrit.config.base
+++ b/images/appserver/etc/gerrit.config.base
@@ -37,6 +37,16 @@
 [cache]
 	directory = cache
 
+#
+# The "web_sessions" cache tracks open user sessions.  These effectively map a
+# cookie in your browser back to an identity.  The maximum age of entries in
+# this cache determines how often a user must sign in again.  A value of 0
+# allows entries to remain in the cache until it fills, at which point the
+# oldest entries will be purged.
+#
+[cache "web_sessions"]
+	maxAge = 0
+
 [plugins]
 	allowRemoteAdmin = true
 
