commit dc35bcc17419ae9e3c35d705fd059608659cca69
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2019-03-20T11:26:45-07:00 (7 months ago)
    
    TRITON-1336 cnapi broken wrt setting hostname on setup

diff --git a/lib/models/server.js b/lib/models/server.js
index a4aa450..08f21ac 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1158,7 +1158,8 @@ ModelServer.prototype.setup = function (params, callback) {
         function _optionallySetHostname(_, cb) {
             if (params.hasOwnProperty('hostname') && params.hostname) {
                 ModelServer.upsert(self.uuid, {hostname: params.hostname}, {
-                    etagRetries: 0
+                    etagRetries: 0,
+                    overrideNonUpdatable: true
                 }, cb);
             } else {
                 cb();
@@ -1879,6 +1880,7 @@ function _attemptUpsert(
 
     assert.object(opts, 'opts');
     assert.optionalBool(opts.allowCreate, 'opts.allowCreate');
+    assert.optionalBool(opts.overrideNonUpdatable, 'opts.overrideNonUpdatable');
     assert.object(opts.log, 'opts.log');
     assert.uuid(serverUuid, 'serverUuid');
     assert.object(properties, 'properties');
@@ -1988,6 +1990,12 @@ function _attemptUpsert(
                     return true;
                 }
 
+                // For server-setup we might need to update ordinarily
+                // non-updatable properties (e.g. hostname)
+                if (opts.overrideNonUpdatable === true) {
+                    return true;
+                }
+
                 if (NON_UPDATABLE_KEYS.indexOf(k) === -1) {
                     // not in blacklist
                     return true;
@@ -2259,7 +2267,8 @@ ModelServer.upsert = function upsert(serverUuid, properties, opts, callback) {
 
     ModelServer._attemptUpsert({
         allowCreate: opts.allowCreate,
-        log: self.log
+        log: self.log,
+        overrideNonUpdatable: opts.overrideNonUpdatable
     }, serverUuid, properties, opts.etagRetries || 0, {stats: {}}, callback);
 };
 
diff --git a/package.json b/package.json
index 324f915..8bce5dc 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.21.8",
+  "version": "1.21.9",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
