From 92ecf80064f25c8e4d177e40885fd2215945085b Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 3 Dec 2018 15:53:30 -0500
Subject: [PATCH] TOOLS-2117 sdcnode updates for November 2018 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 80a09f2..9279e74 100644
--- a/Makefile
+++ b/Makefile
@@ -41,7 +41,7 @@ JSON_FILES  := package.json config.json.sample
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-*-multiarch 15.4.1.
 	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
-	NODE_PREBUILT_VERSION=v6.14.3
+	NODE_PREBUILT_VERSION=v6.15.0
 	NODE_PREBUILT_TAG := zone
 endif
 
-- 
2.21.0

