From 01f669da1f5cb1439d573fc42efd2dc43a7c6184 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 26 Jul 2019 18:20:33 +0100
Subject: [PATCH] TOOLS-2300 buildimage should set "published_at" in image
 manifests

---
 tools/buildimage/bin/buildimage | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tools/buildimage/bin/buildimage b/tools/buildimage/bin/buildimage
index 9dfcc16..bb04477 100755
--- a/tools/buildimage/bin/buildimage
+++ b/tools/buildimage/bin/buildimage
@@ -782,6 +782,9 @@ function buildImage(opts, callback) {
         }
         opts.manifest.requirements.min_platform['7.0'] = vers;
     }
+    if (!opts.manifest.published_at) {
+        opts.manifest.published_at = new Date().toISOString();
+    }
 
     logLine('Starting build for %s (%s)',
         opts.manifest.name, opts.manifest.version);
-- 
2.21.0

