From d87c0946496066765e158284fb4ed7c954954848 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Thu, 21 Jun 2018 13:32:09 -0700
Subject: [PATCH] joyent/node-manta#343 Want `mjob wait` as alias for `mjob
 watch`

---
 bin/mjob | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/bin/mjob b/bin/mjob
index b8196de..c3bfe4f 100755
--- a/bin/mjob
+++ b/bin/mjob
@@ -1,7 +1,7 @@
 #!/usr/bin/env node
 // -*- mode: js -*-
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 var EventEmitter = require('events').EventEmitter;
@@ -1364,6 +1364,7 @@ MJob.prototype.do_watch.help = (
         '{{options}}'
 );
 
+MJob.prototype.do_watch.aliases = ['wait'];
 
 
 //-- Close
-- 
2.21.0

