From 7ba50fed3e670700e263bcbb47385f7727f9e21f Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Thu, 14 Jul 2016 04:25:41 +0000
Subject: [PATCH] joyent/node-forkexec#4 NPM package should not ship
 development files

---
 .npmignore | 4 ++++
 1 file changed, 4 insertions(+)
 create mode 100644 .npmignore

diff --git a/.npmignore b/.npmignore
new file mode 100644
index 0000000..db5dad8
--- /dev/null
+++ b/.npmignore
@@ -0,0 +1,4 @@
+examples/
+tools/
+test/
+Makefile*
-- 
2.21.0

