commit 35a34fbc2bcb1b8cec07b8ba89c04eb0b8750834 (refs/changes/46/2846/6)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-02-19T14:09:31+01:00 (1 year, 8 months ago)
    
    rename uefi-rom.bin to uefi-csm-rom.bin

diff --git a/manifest b/manifest
index 89916b6..e2568f6 100644
--- a/manifest
+++ b/manifest
@@ -3022,7 +3022,7 @@ f usr/share/man/man1/screen.1 0444 root bin
 f usr/bin/socat 0555 root bin
 f usr/share/man/man1/socat.1 0444 root bin
 # uefi-edk2
-f usr/share/bhyve/uefi-rom.bin 0644 root bin
+f usr/share/bhyve/uefi-csm-rom.bin 0644 root bin
 # uuid
 f usr/bin/uuid 0755 root bin
 f usr/share/man/man1/uuid.1 0644 root bin
diff --git a/uefi-edk2/Makefile b/uefi-edk2/Makefile
index 41d975d..49af187 100644
--- a/uefi-edk2/Makefile
+++ b/uefi-edk2/Makefile
@@ -46,9 +46,9 @@ all: $(VER.64)/$(UNPACK_SENTINEL)
 	$(MAKE) $(MAKE_ARGS) ARCH=$(ARCH) -C $(VER.64)/BaseTools
 	bash -c "cd $(VER.64); source edksetup.sh; $(MAKE) $(MAKE_ARGS) ARCH=$(ARCH) -C BhyvePkg/Csm/BhyveCsm16"
 	bash -c "cd $(VER.64); source edksetup.sh; export ILGCC_BIN=$(PWD)/$(VER.64)/Build/; build -t ILGCC -a $(ARCH) -b $(UEFI_TARGET) -p BhyvePkg/BhyvePkgX64.dsc -DDEBUG_ON_SERIAL_PORT=TRUE -DFD_SIZE_2MB -DCSM_ENABLE=TRUE"
-	mv $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/BHYVE.fd $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-rom.bin
+	mv $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/BHYVE.fd $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-csm-rom.bin
 
 install: all
 	mkdir -p $(DESTDIR)/usr/share/bhyve
-	/usr/sbin/install -m 0755 -f $(DESTDIR)/usr/share/bhyve $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-rom.bin
+	/usr/sbin/install -m 0755 -f $(DESTDIR)/usr/share/bhyve $(VER.64)/Build/BhyveX64/RELEASE_ILGCC/FV/uefi-csm-rom.bin
 
