commit eeddd78588bfd0a61b521858c72925546a8b20ff (refs/changes/94/4994/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-10-23T18:00:13+02:00 (12 months ago)
    
    OS-7320 quick workaround for OS-7313

diff --git a/overlay/generic/kernel/drv/sd.conf b/overlay/generic/kernel/drv/sd.conf
index fce0d72c..dd45b4d2 100644
--- a/overlay/generic/kernel/drv/sd.conf
+++ b/overlay/generic/kernel/drv/sd.conf
@@ -73,7 +73,7 @@ ddi-devid-registrant=1;
 # annecdotally solves that problem.
 #
 sd-config-list=
-	"", "retries-timeout:1,retries-busy:1,retries-reset:1,retries-victim:2",
+	"", "power-condition:false,retries-timeout:1,retries-busy:1,retries-reset:1,retries-victim:2",
 	"DELL    PERC H710", "cache-nonvolatile:true",
 	"DELL    PERC H700", "cache-nonvolatile:true",
 	"DELL    PERC/6i", "cache-nonvolatile:true",
