commit 4ec0640c2a1a39009a5f460ecd1a0991d68d6e27
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-05-09T22:11:41+00:00 (5 months ago)
    
    OS-7786 xhci dma segment comment needs work

diff --git a/usr/src/uts/common/io/usb/hcd/xhci/xhci_dma.c b/usr/src/uts/common/io/usb/hcd/xhci/xhci_dma.c
index e7d028a384..712643aabe 100644
--- a/usr/src/uts/common/io/usb/hcd/xhci/xhci_dma.c
+++ b/usr/src/uts/common/io/usb/hcd/xhci/xhci_dma.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright 2019, Joyent, Inc.
  */
 
 /*
@@ -78,9 +78,10 @@ xhci_dma_transfer_attr(xhci_t *xhcip, ddi_dma_attr_t *attrp, uint_t sgl)
 	/*
 	 * The alignment and segment are related. The alignment describes the
 	 * alignment of the PA. The segment describes a boundary that the DMA
-	 * allocation cannot cross. In other words, for a given chunk of memory
-	 * it cannot cross a 64-byte boundary. However, the physical address
-	 * only needs to be aligned to align bytes.
+	 * allocation cannot cross. In other words, for a given chunk of
+	 * allocated DMA memory, the allocated buffer cannot cross a 64k aligned
+	 * boundary. However, the physical address only needs to be aligned to
+	 * 64 bytes.
 	 */
 	attrp->dma_attr_align = XHCI_DMA_ALIGN;
 	attrp->dma_attr_seg = XHCI_TRB_MAX_TRANSFER - 1;
