commit 66fd70c0720b5a5a37c5c28e3188d392b9a7b41d
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-03-12T15:58:15-06:00 (7 months ago)
    
    MANTA-4171 Update boray to Rust 2018

diff --git a/Cargo.lock b/Cargo.lock
index 7beaf19..b2dfff8 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -81,7 +81,7 @@ dependencies = [
 ]
 
 [[package]]
-name = "buckets-demo"
+name = "boray"
 version = "0.1.0"
 dependencies = [
  "base64 0.10.1 (registry+https://github.com/rust-lang/crates.io-index)",
diff --git a/Cargo.toml b/Cargo.toml
index 4994018..22b5350 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,7 +1,8 @@
 [package]
-name = "buckets-demo"
+name = "boray"
 version = "0.1.0"
 authors = ["Kelly McLaughlin <kelly.mclaughlin@joyent.com>"]
+edition = "2018"
 
 [dependencies]
 
diff --git a/src/bucket.rs b/src/bucket.rs
index 9d3d6ac..456d85c 100644
--- a/src/bucket.rs
+++ b/src/bucket.rs
@@ -10,8 +10,9 @@ use r2d2::Pool;
 use postgres::Result as PostgresResult;
 use postgres::rows::Rows;
 use r2d2_postgres::PostgresConnectionManager;
-use serde_json::Value;
-use slog::Logger;
+use serde_derive::{Deserialize, Serialize};
+use serde_json::{json, Value};
+use slog::{Logger, debug};
 use uuid::Uuid;
 
 use rust_fast::protocol::{FastMessage, FastMessageData};
diff --git a/src/main.rs b/src/main.rs
index a4aa192..359ba96 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -2,31 +2,6 @@
  * Copyright 2019 Joyent, Inc.
  */
 
-
-extern crate base64;
-extern crate chrono;
-#[macro_use]
-extern crate clap;
-extern crate hyper;
-#[macro_use]
-extern crate lazy_static;
-extern crate md5;
-extern crate postgres;
-#[macro_use]
-extern crate prometheus;
-extern crate r2d2;
-extern crate r2d2_postgres;
-extern crate rust_fast;
-#[macro_use]
-extern crate serde_derive;
-#[macro_use]
-extern crate serde_json;
-#[macro_use]
-extern crate slog;
-extern crate slog_bunyan;
-extern crate tokio;
-extern crate uuid;
-
 mod bucket;
 mod metrics;
 mod object;
@@ -37,14 +12,16 @@ use std::net::SocketAddr;
 use std::sync::{Arc, Mutex};
 use std::thread;
 
+use clap::{crate_version, value_t};
 use r2d2::Pool;
 use r2d2_postgres::{TlsMode, PostgresConnectionManager};
 use serde_json::Value;
-use slog::{Drain, Level, LevelFilter, Logger};
+use slog::{Drain, Level, LevelFilter, Logger, o};
 use tokio::net::TcpListener;
 use tokio::prelude::*;
 use rust_fast::protocol::FastMessage;
 use rust_fast::server;
+use slog::{error, info, trace};
 
 static APP: &'static str = "buckets-demo";
 
diff --git a/src/metrics.rs b/src/metrics.rs
index 0324d21..752b20e 100644
--- a/src/metrics.rs
+++ b/src/metrics.rs
@@ -11,8 +11,9 @@ use hyper::rt::{self, Future};
 use hyper::server::Server;
 use hyper::service::service_fn_ok;
 use hyper::StatusCode;
-use prometheus::{Counter, Encoder, TextEncoder};
-use slog::{Logger, info};
+use lazy_static::lazy_static;
+use prometheus::{Counter, Encoder, TextEncoder, labels, opts, register_counter};
+use slog::{Logger, error, info};
 
 
 lazy_static! {
diff --git a/src/object.rs b/src/object.rs
index 5f078bc..a821c36 100644
--- a/src/object.rs
+++ b/src/object.rs
@@ -12,8 +12,9 @@ use postgres::Result as PostgresResult;
 use postgres::rows::Rows;
 use r2d2::Pool;
 use r2d2_postgres::PostgresConnectionManager;
-use serde_json::Value;
-use slog::Logger;
+use serde_derive::{Deserialize, Serialize};
+use serde_json::{Value, json};
+use slog::{Logger, debug};
 use uuid::Uuid;
 
 use rust_fast::protocol::{FastMessage, FastMessageData};
diff --git a/src/opts.rs b/src/opts.rs
index b21cdc2..203f974 100644
--- a/src/opts.rs
+++ b/src/opts.rs
@@ -2,7 +2,7 @@
  * Copyright 2019 Joyent, Inc.
  */
 
-use clap::{App, Arg, ArgMatches};
+use clap::{App, Arg, ArgMatches, crate_version};
 
 pub fn parse<'a, 'b>(app: String) -> ArgMatches<'a> {
     App::new(app)
