commit 45c5e184ec60a68cc02516a8f9e62f0318f54b9e (refs/changes/53/2653/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2017-09-26T00:51:09-07:00 (2 years ago)
    
    VOLAPI-81 volapi setup script should enable cloudapi-automount

diff --git a/tools/setup/setup.sh b/tools/setup/setup.sh
index cefca55..f74309b 100644
--- a/tools/setup/setup.sh
+++ b/tools/setup/setup.sh
@@ -307,11 +307,11 @@ upgrade_core_service_to_latest_branch_image "volapi" "master"
 upgrade_gz_tools_to_latest_branch_image tritonnfs
 
 # Now enable the flags in SAPI to enable NFS volumes support for docker and
-# CloudAPI, as well as the ability for docker containers to automount NFS
-# volumes
+# CloudAPI, as well as the ability to automount NFS volumes
 sdcadm experimental nfs-volumes docker
 sdcadm experimental nfs-volumes docker-automount
 sdcadm experimental nfs-volumes cloudapi
+sdcadm experimental nfs-volumes cloudapi-automount
 EOS
 
-echo "VOLAPI setup done, reboot of all CNs is needed before it can be used"
\ No newline at end of file
+echo "VOLAPI setup done, reboot of all CNs is needed before it can be used"
