From b05513cb6ebc6783a6adb7f4e96e5903f18e3a77 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 28 Sep 2016 17:27:16 -0700
Subject: [PATCH] ADMINUI-2336 Update sdcnode version for OpenSSL 1.0.2j/1.0.1u
 vulnerability fixes

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8fff66be..f8039bf7 100644
--- a/Makefile
+++ b/Makefile
@@ -39,7 +39,7 @@ JSSTYLE_FLAGS    = -o "indent=2,doxygen,unparenthesized-return=0,line-length=120
 REPO_MODULES	 =
 SMF_MANIFESTS_IN = smf/manifests/adminui.xml.in
 
-NODE_PREBUILT_VERSION=v0.10.45
+NODE_PREBUILT_VERSION=v0.10.47
 NODE_PREBUILT_TAG=zone
 NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
 
-- 
2.21.0

