From 5f54edaee9686523532856f368268508a8ee4110 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Mon, 17 Jul 2017 00:53:20 +1200
Subject: [PATCH] ADMINUI-2314: Upgrade Node.js version from 0.10 to 4.x

---
 targets.json.in | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 8d8da58..d8ff174 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -206,8 +206,11 @@ cat <<EOF
     "image_name": "adminui",
     "image_description": "SDC AdminUI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
-    "pkgsrc": [],
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "patch-2.7.5"
+    ],
     "repos": [
         {"url": "git@github.com:joyent/sdc-adminui.git", "dir": "adminui"}
     ],
-- 
2.21.0

