commit 709bc2cb59d89e8fd86179f50971667f4e5a98aa (refs/changes/68/1968/1)
Author: Chris Burroughs <chris.burroughs@gmail.com>
Date:   2017-05-18T11:02:05-04:00 (2 years, 5 months ago)
    
    TOOLS-1777 support configuring MG:jenkins-build output directory

diff --git a/tools/jenkins-build b/tools/jenkins-build
index 8d46b8b..4ee66bd 100755
--- a/tools/jenkins-build
+++ b/tools/jenkins-build
@@ -140,11 +140,13 @@ fi
 if [[ ! -f targets.json ]]; then
     bash < targets.json.in | json > targets.json
 fi
-IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
-if [[ ${IS_PUBLIC} == "true" ]]; then
-    JENKINS_OUTPUT_BASE="/public/builds"
-else
-    JENKINS_OUTPUT_BASE="/stor/builds"
+if [[ -z ${JENKINS_OUTPUT_BASE} ]]; then
+    IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
+    if [[ ${IS_PUBLIC} == "true" ]]; then
+        JENKINS_OUTPUT_BASE="/public/builds"
+    else
+        JENKINS_OUTPUT_BASE="/stor/builds"
+    fi
 fi
 
 if [[ -z ${JOYENT_BUILD} || ${JOYENT_BUILD} == "true" ]]; then
