From 9e24fb4685d755bcdc2fe1db560d5e050aa879b1 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@gmail.com>
Date: Thu, 18 May 2017 11:02:05 -0400
Subject: [PATCH] TOOLS-1777 support configuring MG:jenkins-build output
 directory

---
 tools/jenkins-build | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/tools/jenkins-build b/tools/jenkins-build
index 8d46b8b..b650eb7 100755
--- a/tools/jenkins-build
+++ b/tools/jenkins-build
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2017, Joyent, Inc.
 #
 
 #
@@ -140,11 +140,13 @@ fi
 if [[ ! -f targets.json ]]; then
     bash < targets.json.in | json > targets.json
 fi
-IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
-if [[ ${IS_PUBLIC} == "true" ]]; then
-    JENKINS_OUTPUT_BASE="/public/builds"
-else
-    JENKINS_OUTPUT_BASE="/stor/builds"
+if [[ -z ${JENKINS_OUTPUT_BASE} ]]; then
+    IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
+    if [[ ${IS_PUBLIC} == "true" ]]; then
+        JENKINS_OUTPUT_BASE="/public/builds"
+    else
+        JENKINS_OUTPUT_BASE="/stor/builds"
+    fi
 fi
 
 if [[ -z ${JOYENT_BUILD} || ${JOYENT_BUILD} == "true" ]]; then
-- 
2.21.0

