From 709bc2cb59d89e8fd86179f50971667f4e5a98aa Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@gmail.com>
Date: Thu, 18 May 2017 11:02:05 -0400
Subject: [PATCH] TOOLS-1777 support configuring MG:jenkins-build output
 directory

---
 tools/jenkins-build | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/tools/jenkins-build b/tools/jenkins-build
index 8d46b8b..4ee66bd 100755
--- a/tools/jenkins-build
+++ b/tools/jenkins-build
@@ -140,11 +140,13 @@ fi
 if [[ ! -f targets.json ]]; then
     bash < targets.json.in | json > targets.json
 fi
-IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
-if [[ ${IS_PUBLIC} == "true" ]]; then
-    JENKINS_OUTPUT_BASE="/public/builds"
-else
-    JENKINS_OUTPUT_BASE="/stor/builds"
+if [[ -z ${JENKINS_OUTPUT_BASE} ]]; then
+    IS_PUBLIC=$(json < targets.json ${JOB_NAME}.public)
+    if [[ ${IS_PUBLIC} == "true" ]]; then
+        JENKINS_OUTPUT_BASE="/public/builds"
+    else
+        JENKINS_OUTPUT_BASE="/stor/builds"
+    fi
 fi
 
 if [[ -z ${JOYENT_BUILD} || ${JOYENT_BUILD} == "true" ]]; then
-- 
2.21.0

