commit 895306af5af35e0362744e34980d4c50416f6c0c (refs/changes/74/3474/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-02-24T12:19:05+01:00 (1 year, 7 months ago)
    
    OS-6692 need acpica-utils to build OS-6416

diff --git a/configure b/configure
index 777b9bb9..b8824a34 100755
--- a/configure
+++ b/configure
@@ -92,6 +92,13 @@ function install_pkgin
 	pkglist="$pkglist coreutils gsed pkg_alternatives cdrtools"
 	pkglist="$pkglist py27-sqlite3 nasm"
 
+	#
+	# Newer images have /usr/bin/iasl. Install acpica-utils if it is missing.
+	#
+	if [ ! -x /usr/sbin/iasl ]; then
+		pkglist="$pkglist acpica-utils"
+	fi
+
 	for pkg in $pkglist; do
 		if ! pkg_info -qe $pkg; then
 			$conf_priv pkgin -y install $pkg || fatal \
