commit 3cec3a09779ed0e78a8fd871d6a11c587a3403c9 (refs/changes/74/3474/2)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-02-26T11:49:44+01:00 (1 year, 7 months ago)
    
    OS-6692 need acpica-utils to build OS-6416

diff --git a/configure b/configure
index 777b9bb9..4bf16a37 100755
--- a/configure
+++ b/configure
@@ -90,7 +90,7 @@ function install_pkgin
 	pkglist="$pkglist flex libxslt openjdk7 nodejs"
 	pkglist="$pkglist p5-XML-Parser gettext python27 py27-expat"
 	pkglist="$pkglist coreutils gsed pkg_alternatives cdrtools"
-	pkglist="$pkglist py27-sqlite3 nasm"
+	pkglist="$pkglist py27-sqlite3 nasm acpica-utils"
 
 	for pkg in $pkglist; do
 		if ! pkg_info -qe $pkg; then
