From e714bd56728caa461ac4e2f021557bd825e88028 Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody.kantor@gmail.com>
Date: Tue, 10 Apr 2018 20:26:53 +0000
Subject: [PATCH] MANTA-3624 create pgstatsmon image

---
 Makefile        | 40 ++++++++++++++++++++++++++++++++++++++++
 targets.json.in | 24 ++++++++++++++++++++++++
 2 files changed, 64 insertions(+)

diff --git a/Makefile b/Makefile
index 439893c..5d9d469 100644
--- a/Makefile
+++ b/Makefile
@@ -1785,6 +1785,46 @@ clean_manta-reshard:
 	$(RM) -rf $(BITS_DIR)/manta-reshard
 	cd build/manta-reshard && gmake distclean
 
+#---- pgstatsmon
+
+_pgstatsmon_stamp=$(PGSTATSMON_BRANCH)-$(TIMESTAMP)-g$(PGSTATSMON_SHA)
+PGSTATSMON_BITS=$(BITS_DIR)/pgstatsmon/pgstatsmon-pkg-$(_pgstatsmon_stamp).tar.bz2
+PGSTATSMON_IMAGE_BIT=$(BITS_DIR)/pgstatsmon/pgstatsmon-zfs-$(_pgstatsmon_stamp).zfs.gz
+PGSTATSMON_MANIFEST_BIT=$(BITS_DIR)/pgstatsmon/pgstatsmon-zfs-$(_pgstatsmon_stamp).imgmanifest
+
+.PHONY: pgstatsmon
+pgstatsmon: $(PGSTATSMON_BITS) pgstatsmon_image
+
+$(PGSTATSMON_BITS): build/pgstatsmon
+       @echo "# Build pgstatsmon: branch $(PGSTATSMON_BRANCH), sha $(PGSTATSMON_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
+       mkdir -p $(BITS_DIR)
+       cd build/pgstatsmon && NPM_CONFIG_CACHE=$(MG_CACHE_DIR)/npm TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) gmake release publish
+       @echo "# Created pgstatsmon bits (time `date -u +%Y%m%dT%H%M%SZ`):"
+       @ls -l $(PGSTATSMON_BITS)
+       @echo ""
+
+.PHONY: pgstatsmon-image
+pgstatsmon_image: $(PGSTATSMON_IMAGE_BIT)
+
+$(PGSTATSMON_IMAGE_BIT): $(PGSTATSMON_BITS)
+       @echo "# Build pgstatsmon_image: branch $(PGSTATSMON_BRANCH), sha $(PGSTATSMON_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
+       ./tools/prep_dataset_in_jpc.sh -i "$(PGSTATSMON_IMAGE_UUID)" -t $(PGSTATSMON_BITS) \
+               -b "pgstatsmon" \
+               -o "$(PGSTATSMON_IMAGE_BIT)" -p $(PGSTATSMON_PKGSRC) -O "$(MG_OUT_PATH)" \
+               -t $(PGSTATSMON_EXTRA_TARBALLS) -n $(PGSTATSMON_IMAGE_NAME) \
+               -v $(_pgstatsmon_stamp) -d $(PGSTATSMON_IMAGE_DESCRIPTION)
+       @echo "# Created pgstatsmon image (time `date -u +%Y%m%dT%H%M%SZ`):"
+       @ls -l $$(dirname $(PGSTATSMON_IMAGE_BIT))
+       @echo ""
+
+pgstatsmon_publish_image: $(PGSTATSMON_IMAGE_BIT)
+       @echo "# Publish pgstatsmon image to SDC Updates repo."
+       $(UPDATES_IMGADM) import -ddd -m $(PGSTATSMON_MANIFEST_BIT) -f $(PGSTATSMON_IMAGE_BIT)
+
+
+clean_pgstatsmon:
+       $(RM) -rf $(BITS_DIR)/pgstatsmon
+       cd build/pgstatsmon && gmake distclean
 
 #---- Mola
 
diff --git a/targets.json.in b/targets.json.in
index 65997e7..c779f51 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1426,6 +1426,30 @@ cat <<EOF
     ]
   },
 
+  "pgstatsmon": {
+    "appliance": "true",
+    "build_platform": "20151126T062538Z",
+    "image_name": "manta-pgstatsmon",
+    "image_description": "Postgres Monitoring Service",
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "coreutils-8.23nb2"
+    ],
+    "repos": [
+        {"url": "git@github.com:joyent/pgstatsmon.git"}
+    ],
+    "public": true,
+    "deps": [
+      "config-agent",
+      "registrar"
+    ],
+    "tarballs": [
+      {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"},
+      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"}
+    ]
+  },
+
   "manta-reshard": {
     "appliance": "true",
     "build_platform": "20151126T062538Z",
-- 
2.21.0

