commit c057ed69447b6125a3fbe214eedd6bbd688d9f9c (refs/changes/32/3932/1)
Author: Jorge Schrauwen <sjorge@blackdot.be>
Date:   2018-05-10T10:57:09+00:00 (1 year, 5 months ago)
    
    Support different disk models for bhyve

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 416316d5..36d896c0 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -280,11 +280,15 @@ function ensureLogging(side_effects)
     start_logging();
 }
 
-exports.DISK_MODELS = [
+exports.KVM_DISK_MODELS = [
     'virtio',
     'ide',
     'scsi'
 ];
+exports.BHYVE_DISK_MODELS = [
+    'virtio',
+    'ahci'
+];
 
 exports.VGA_TYPES = [
     'cirrus',
@@ -4591,6 +4595,9 @@ function checkPayloadProperties(payload, vmobj, log, callback)
         for (disk in payload.add_disks) {
             if (payload.add_disks.hasOwnProperty(disk)) {
                 zvol = payload.add_disks[disk];
+                disk_models = (brand == 'bhyve') ?
+                    VM.BHYVE_DISK_MODELS :
+                    VM.KVM_DISK_MODELS;
 
                 // path is only allowed in 2 cases when adding a disk:
                 //
@@ -4648,10 +4655,10 @@ function checkPayloadProperties(payload, vmobj, log, callback)
                             + 'disk: ' + JSON.stringify(zvol)));
                         return;
                     }
-                } else if (VM.DISK_MODELS.indexOf(zvol.model) === -1) {
+                } else if (disk_models.indexOf(zvol.model) === -1) {
                     callback(new Error('"' + zvol.model + '"'
-                        + ' is not a valid disk model. Valid are: '
-                        + VM.DISK_MODELS.join(',')));
+                        + ' is not a valid disk model for ' + brand + ' brand.'
+                        + ' Valid are: ' + disk_models.join(',')));
                     return;
                 }
             }
@@ -4662,6 +4669,9 @@ function checkPayloadProperties(payload, vmobj, log, callback)
         for (disk in payload.update_disks) {
             if (payload.update_disks.hasOwnProperty(disk)) {
                 zvol = payload.update_disks[disk];
+                disk_models = (brand == 'bhyve') ?
+                    VM.BHYVE_DISK_MODELS :
+                    VM.KVM_DISK_MODELS;
 
                 // Disks cannot have refreservation=auto on update.
                 // Only create.
@@ -4677,6 +4687,7 @@ function checkPayloadProperties(payload, vmobj, log, callback)
                             + 'disk, must be one of: '
                             + VM.COMPRESSION_TYPES.join(', ')));
                     }
+                    return;
                 }
 
                 if (zvol.hasOwnProperty('block_size')) {
@@ -4684,6 +4695,16 @@ function checkPayloadProperties(payload, vmobj, log, callback)
                         + 'after creation'));
                     return;
                 }
+
+                // Disk model must also be checked for update_disks
+                if (zvol.hasOwnProperty('model')) {
+                    if (disk_models.indexOf(zvol.model) === -1) {
+                        callback(new Error('"' + zvol.model + '"'
+                            + ' is not a valid disk model for ' + brand + ' brand.'
+                            + ' Valid are: ' + disk_models.join(',')));
+                        return;
+                    }
+                }
             }
         }
     }
diff --git a/src/vm/sbin/vmadm.js b/src/vm/sbin/vmadm.js
index f6043a54..3b0d2340 100755
--- a/src/vm/sbin/vmadm.js
+++ b/src/vm/sbin/vmadm.js
@@ -333,9 +333,10 @@ function parseStartArgs(args)
                 if (!model || !p || p.length === 0 || model.length === 0) {
                     usage('Parameter to ' + key + ' must be: path,model');
                 }
-                if (VM.DISK_MODELS.indexOf(model) === -1) {
+                // only check for KVM_DISK_MODELS as we do not support this (yet) for bhyve
+                if (VM.KVM_DISK_MODELS.indexOf(model) === -1) {
                     usage('Invalid model "' + model + '": model must be one '
-                        + 'of: ' + VM.DISK_MODELS.join(','));
+                        + 'of: ' + VM.KVM_DISK_MODELS.join(','));
                 }
                 if (!extra.disks) {
                     extra.disks = [];
