commit faedc65f4112904d028e1add2c8838a2d5c8cef7 (refs/changes/69/5169/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-11-30T23:25:46+00:00 (10 months ago)
    
    OS-7415 metadata agent should ignore init_restarts "added" actions

diff --git a/src/vm/lib/metadata/agent.js b/src/vm/lib/metadata/agent.js
index eb774ef6..2b0ad41a 100644
--- a/src/vm/lib/metadata/agent.js
+++ b/src/vm/lib/metadata/agent.js
@@ -465,7 +465,8 @@ MetadataAgent.prototype.start = function start() {
          * For bhyve, the bhyve process recreates the socket every time it
          * starts. This happens when the zone first starts and every time the
          * guest reboots. The 'running' state change will catch the first one.
-         * Guest reboots can be detected by 'init_restarts' incrementing.
+         * Guest reboots can be detected by 'init_restarts' incrementing, which
+         * is seen when "action" is "changed" - not "added".
          */
         if (ev.vm.brand !== 'kvm' && ev.vm.brand !== 'bhyve') {
             return;
@@ -473,7 +474,8 @@ MetadataAgent.prototype.start = function start() {
 
         var restarts = ev.changes.filter(function (change) {
             return (change.path.length === 1
-                && change.path[0] === 'init_restarts');
+                && change.path[0] === 'init_restarts'
+                && change.action === 'changed');
         });
         if (restarts.length !== 0) {
             // The previous zoneConnection should have already been cleaned up
