commit e42597639bd5b35d0ed1036fd1d59eb6d6b63b24 (refs/changes/71/5171/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-12-03T15:19:44+00:00 (10 months ago)
    
    OS-7410 vmm_zsd assert tripped on zone shutdown

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_zsd.c b/usr/src/uts/i86pc/io/vmm/vmm_zsd.c
index 8313dfbb50..3c73bd83d2 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_zsd.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_zsd.c
@@ -114,7 +114,7 @@ vmm_zsd_create(zoneid_t zid)
 	zsd->vz_zoneid = zid;
 
 	mutex_init(&zsd->vz_lock, NULL, MUTEX_DEFAULT, NULL);
-	zsd->vz_active = B_TRUE;
+	zsd->vz_active = zone_status_get(curzone) < ZONE_IS_SHUTTING_DOWN;
 
 	mutex_enter(&vmm_zsd_lock);
 	list_insert_tail(&vmm_zsd_list, zsd);
@@ -134,6 +134,12 @@ vmm_zsd_shutdown(zoneid_t zid, void *data)
 	vmm_softc_t *sc;
 
 	mutex_enter(&zsd->vz_lock);
+	if (!zsd->vz_active) {
+		/* vmm_zsd_create() called during zone shutdown */
+		mutex_exit(&zsd->vz_lock);
+		return;
+	}
+
 	ASSERT(zsd->vz_active);
 	zsd->vz_active = B_FALSE;
 
