commit 38c6bd0164fca76da0fb3021a46ad4be7752ceb0 (refs/changes/96/1196/1)
Author: Dillon Amburgey <dillona@dillona.com>
Date:   2016-12-31T16:36:33-05:00 (2 years, 9 months ago)
    
    joyent/triton#222 Support coal on CentOS 7 KVM

diff --git a/tools/coal-linux-kvm-setup b/tools/coal-linux-kvm-setup
index b7ef212..350afc7 100755
--- a/tools/coal-linux-kvm-setup
+++ b/tools/coal-linux-kvm-setup
@@ -19,6 +19,8 @@
 #
 # This script is highly experimental and is community supported (it is not used by anyone at 
 # Joyent). This script has been tested on the following Linux distributions: 
+# * CentOS 7
+#   * Install required packages: yum install qemu-kvm libvirt virt-manager
 # * Fedora 23 
 #   * Install required packages: dnf install libvirt virt-manager
 # * Ubuntu 14.04, 16.04
@@ -63,6 +65,8 @@ if [ -e "/usr/bin/qemu-kvm" ]; then
     EMULATOR=/usr/bin/qemu-kvm
 elif [ -e "/usr/bin/kvm" ]; then
     EMULATOR=/usr/bin/kvm
+elif [ -e "/usr/libexec/qemu-kvm" ]; then
+    EMULATOR=/usr/libexec/qemu-kvm
 else
     echo "KVM cannot be found. Exiting."
     exit 1
