commit 02dde56cc801432d90cab1ce376859cbbdaacdef
Author: Jason King <jason.king@joyent.com>
Date:   2019-07-24T21:08:00+00:00 (3 months ago)
    
    OS-7903 Document mdb(1) ::dis flags

diff --git a/usr/src/cmd/mdb/common/mdb/mdb_cmds.c b/usr/src/cmd/mdb/common/mdb/mdb_cmds.c
index 878728251c..94b302e75a 100644
--- a/usr/src/cmd/mdb/common/mdb/mdb_cmds.c
+++ b/usr/src/cmd/mdb/common/mdb/mdb_cmds.c
@@ -26,7 +26,7 @@
 
 /*
  * Copyright (c) 2012 by Delphix. All rights reserved.
- * Copyright (c) 2019 Joyent, Inc. All rights reserved.
+ * Copyright 2019 Joyent, Inc.
  * Copyright (c) 2013 Josef 'Jeff' Sipek <jeffpc@josefsipek.net>
  * Copyright (c) 2015, 2017 by Delphix. All rights reserved.
  * Copyright 2018 OmniOS Community Edition (OmniOSce) Association.
@@ -2165,6 +2165,25 @@ cmd_dis(uintptr_t addr, uint_t flags, int argc, const mdb_arg_t *argv)
 	return (DCMD_OK);
 }
 
+static void
+dis_help(void)
+{
+	mdb_printf("Disassembles instructions starting at the final argument\n"
+	    "or the current value of dot.  If the address is the start of a\n"
+	    "function, the entire function is disassembled, or else a window\n"
+	    "of instructions before and after the disassembled address are\n"
+	    "displayed\n\n");
+
+	mdb_printf(
+	    "  -a\t\tPrint addresses as numeric values instead of symbolic "
+	    "values.\n"
+	    "  -b\t\tPrint addresses as both numeric and symbolic values.\n"
+	    "  -f\t\tRead instructions from the target's object file\n"
+	    "\t\tinstead of the target's virtual address space.\n"
+	    "  -n instr\tDisplay 'instr' instructions\n"
+	    "  -w\t\tForce window behavior, even at the start of a function\n");
+}
+
 /*ARGSUSED*/
 static int
 walk_step(uintptr_t addr, const void *data, void *private)
@@ -3105,7 +3124,8 @@ const mdb_dcmd_t mdb_dcmd_builtins[] = {
 	{ "dcmds", "[[-n] pattern]",
 	    "list available debugger commands", cmd_dcmds, cmd_dcmds_help },
 	{ "delete", "?[id|all]", "delete traced software events", cmd_delete },
-	{ "dis", "?[-abfw] [-n cnt] [addr]", "disassemble near addr", cmd_dis },
+	{ "dis", "?[-abfw] [-n cnt] [addr]", "disassemble near addr", cmd_dis,
+	    dis_help },
 	{ "disasms", NULL, "list available disassemblers", cmd_disasms },
 	{ "dismode", "[mode]", "get/set disassembly mode", cmd_dismode },
 	{ "dmods", "[-l] [mod]", "list loaded debugger modules", cmd_dmods },
