From dffa6f5ecd1f8b7211a65b762e3201a318a23538 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 21 Nov 2016 15:18:43 -0800
Subject: [PATCH] joyent/node-cueball#51 pool should never call .close() on a
 busy backend

---
 lib/pool.js | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/pool.js b/lib/pool.js
index 9103136..5c705d0 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -447,7 +447,12 @@ CueBallConnectionPool.prototype._rebalance = function () {
 		 */
 		var fsmIdx = self.p_connections[fsm.cf_backend.key].
 		    indexOf(fsm);
-		if (fsm.isInState('idle') || fsmIdx > 0 ||
+		if (fsm.isInState('idle')) {
+			fsm.close();
+			--total;
+		} else if (fsm.isInState('busy')) {
+			fsm.closeAfterRelease();
+		} else if (fsmIdx > 0 ||
 		    self.p_keys.indexOf(fsm.cf_backend.key) === -1) {
 			fsm.close();
 			--total;
-- 
2.21.0

