From 342854481eefbe825e63fde1d14dd614344a8ea8 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 25 Apr 2017 12:22:03 -0700
Subject: [PATCH] TOOLS-1757 improve MG's tools/launch-build help output

---
 tools/launch-build | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/launch-build b/tools/launch-build
index c0d2c69..c3fe11a 100755
--- a/tools/launch-build
+++ b/tools/launch-build
@@ -22,7 +22,7 @@ function usage() {
 	if [[ -n $1 ]]; then
 		printf "%s: %s\n" "$PROGNAME" "$1" >&2
 	fi
-	printf "usage: $PROGNAME [-v] [-H url] [-b BRANCH] [-t TRY_BRANCH] [-u auth] project\n" >&2
+	printf "usage:\n    $PROGNAME [-v] [-H url] [-b BRANCH] [-t TRY_BRANCH] [-u auth] project\n" >&2
 	exit 2
 }
 
@@ -50,7 +50,7 @@ if [[ -z $JENKINS_URL ]]; then
 fi
 
 if [[ -z $JENKINS_AUTH ]]; then
-	usage "JENKINS_AUTH must be set to <user>:<api token>"
+	usage "JENKINS_AUTH must be set to <user>:<api token> (get it here: https://jenkins.joyent.us/me/configure)"
 fi
 
 if [[ -n $BRANCH && -n $TRY_BRANCH ]]; then
-- 
2.21.0

