commit a59925ba26bfff72534e2b5653a7a816bf977970 (refs/changes/36/1836/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2017-04-25T12:09:50-07:00 (2 years, 5 months ago)
    
    TOOLS-1757 improve MG's tools/launch-build help output

diff --git a/tools/launch-build b/tools/launch-build
index c0d2c69..c3fe11a 100755
--- a/tools/launch-build
+++ b/tools/launch-build
@@ -22,7 +22,7 @@ function usage() {
 	if [[ -n $1 ]]; then
 		printf "%s: %s\n" "$PROGNAME" "$1" >&2
 	fi
-	printf "usage: $PROGNAME [-v] [-H url] [-b BRANCH] [-t TRY_BRANCH] [-u auth] project\n" >&2
+	printf "usage:\n    $PROGNAME [-v] [-H url] [-b BRANCH] [-t TRY_BRANCH] [-u auth] project\n" >&2
 	exit 2
 }
 
@@ -50,7 +50,7 @@ if [[ -z $JENKINS_URL ]]; then
 fi
 
 if [[ -z $JENKINS_AUTH ]]; then
-	usage "JENKINS_AUTH must be set to <user>:<api token>"
+	usage "JENKINS_AUTH must be set to <user>:<api token> (get it here: https://jenkins.joyent.us/me/configure)"
 fi
 
 if [[ -n $BRANCH && -n $TRY_BRANCH ]]; then
