commit 0443050980954cbc6d5f7580f07690eaab6b93ab (refs/changes/14/2014/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-05-30T12:42:30+02:00 (2 years, 4 months ago)
    
    WORKFLOW-218 Update workflow to node-moray v3

diff --git a/etc/config.json.in b/etc/config.json.in
index 85da203..ef2ed65 100644
--- a/etc/config.json.in
+++ b/etc/config.json.in
@@ -3,7 +3,12 @@
     "backend": {
         "module": "wf-moray-backend",
         "opts": {
-            "url": "http://{{{MORAY_SERVICE}}}:2020",
+            "moray": {
+                "srvDomain": "{{{MORAY_SERVICE}}}",
+                "cueballOptions": {
+                    "resolvers": [ "{{{BINDER_SERVICE}}}" ]
+                }
+            },
             "extra_fields": {
                 "wf_workflows": {},
                 "wf_jobs": {
diff --git a/package.json b/package.json
index 704cb7a..0df2396 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "workflow",
   "description": "SmartDataCenter Workflow API and Runners",
-  "version": "1.0.3",
+  "version": "1.1.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "repository": {
@@ -9,7 +9,7 @@
     "url": "git+https://github.com/joyent/sdc-workflow.git"
   },
   "dependencies": {
-    "wf-moray-backend": "git+https://github.com/joyent/node-workflow-moray-backend.git#376532",
+    "wf-moray-backend": "git+https://github.com/joyent/node-workflow-moray-backend.git#c3504876",
     "wf": "git+https://github.com/joyent/node-workflow.git#cac32eec0fa906adeba683f7f9436a9247740f11",
     "bunyan": "1.3.4",
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
diff --git a/sapi_manifests/workflow/template b/sapi_manifests/workflow/template
index 3064ddd..0635a2d 100644
--- a/sapi_manifests/workflow/template
+++ b/sapi_manifests/workflow/template
@@ -2,7 +2,12 @@
     "backend": {
         "module": "wf-moray-backend",
         "opts": {
-            "url": "http://{{{MORAY_SERVICE}}}:2020",
+            "moray": {
+                "srvDomain": "{{{MORAY_SERVICE}}}",
+                "cueballOptions": {
+                    "resolvers": [ "{{{BINDER_SERVICE}}}" ]
+                }
+            },
             "extra_fields": {
                 "wf_workflows": {},
                 "wf_jobs": {
