From 250077dfedf478e242088dba85578158b33236bc Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Wed, 9 Aug 2017 21:28:23 -0700
Subject: [PATCH] ADMINUI-2384 AdminUI should handle API calls with undefined
 route name

---
 lib/adminui.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/adminui.js b/lib/adminui.js
index e66c2e1c..768ab268 100644
--- a/lib/adminui.js
+++ b/lib/adminui.js
@@ -903,7 +903,7 @@ ADMINUI.prototype.createServer = function () {
         }
         var method = req.method;
         var body =
-            !(method === 'POST' && route.name === 'authenticate') &&
+            !(method === 'POST' && route && route.name === 'authenticate') &&
             !(method === 'GET' && Math.floor(res.statusCode/100) === 2);
         auditLogger({
             log: req.log.child({ component: 'audit', route: route && route.name }, true),
-- 
2.21.0

