commit 1a0b0f0cb2ace0ff60f44170b973a64e4362230b (refs/changes/74/5274/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-12-20T18:44:08+00:00 (10 months ago)
    
    OS-XXXX fix build after OS-6969

diff --git a/usr/src/lib/libvmm/Makefile b/usr/src/lib/libvmm/Makefile
index 2a0de6c5dc..66bd60eb46 100644
--- a/usr/src/lib/libvmm/Makefile
+++ b/usr/src/lib/libvmm/Makefile
@@ -13,45 +13,31 @@
 # Copyright 2018 Joyent, Inc.
 #
 
-LIBRARY =	libvmm.a
-VERS =		.1
-
-OBJECTS =	libvmm.o list.o
-
 include ../Makefile.lib
-include ../Makefile.lib.64
 include ../Makefile.rootfs
 
-SRCDIR =	.
-COMDIR =	$(SRC)/common/list
-
-LIBS = $(DYNLIB)
-
-# The FreeBSD compat and contrib headers need to be first in the search
-# path, hence we can't just append them to CPPFLAGS. So we assign CPPFLAGS
-# directly and pull in CPPFLAGS.master at the appropriate place.
-CPPFLAGS =	-I$(COMPAT)/freebsd -I$(CONTRIB)/freebsd \
-		-I$(COMPAT)/freebsd/amd64 -I$(CONTRIB)/freebsd/amd64 \
-		$(CPPFLAGS.master) -I$(SRC)/uts/i86pc
-
-LDLIBS +=	-lc -lvmmapi
+$(BUILD64)SUBDIRS += $(MACH64)
 
 HDRS =		libvmm.h
 HDRDIR =	.
 CHECKHDRS =	$(HDRS:%.h=%.check)
 
-.KEEP_STATE:
+all:=		TARGET= all
+install:=	TARGET= install
+clean:=		TARGET= clean
+clobber:=	TARGET= clobber
+lint:=		TARGET= lint
 
-all: $(LIBS)
+.KEEP_STATE:
 
-install_h: $(ROOTHDRS)
+all install clean clobber lint: $(SUBDIRS)
 
-check: $(CHECKHDRS)
+install_h:	$(ROOTHDRS)
+check:		$(CHECKHDRS)
 
-pics/%.o: $(COMDIR)/%.c
-	$(COMPILE.c) -o $@ $<
-	$(POST_PROCESS_O)
+$(SUBDIRS): FRC
+	cd $@; pwd; $(MAKE) $(TARGET)
 
-install: install_h all $(ROOTLIBS64)
+FRC:
 
 include ../Makefile.targ
diff --git a/usr/src/lib/libvmm/Makefile.com b/usr/src/lib/libvmm/Makefile.com
new file mode 100644
index 0000000000..d85abae8ce
--- /dev/null
+++ b/usr/src/lib/libvmm/Makefile.com
@@ -0,0 +1,51 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2018 Joyent, Inc.
+#
+
+LIBRARY =	libvmm.a
+VERS =		.1
+OBJECTS =	libvmm.o list.o
+
+SRCDIR =	.
+
+include ../../Makefile.lib
+include ../../Makefile.rootfs
+
+LIBS		= $(DYNLIB)
+
+# The FreeBSD compat and contrib headers need to be first in the search
+# path, hence we can't just append them to CPPFLAGS. So we assign CPPFLAGS
+# directly and pull in CPPFLAGS.master at the appropriate place.
+CPPFLAGS =	-I$(COMPAT)/freebsd -I$(CONTRIB)/freebsd \
+		-I$(COMPAT)/freebsd/amd64 -I$(CONTRIB)/freebsd/amd64 \
+		$(CPPFLAGS.master) -I$(SRC)/uts/i86pc
+
+LDLIBS +=	-lc -lvmmapi
+
+.KEEP_STATE:
+
+all: $(LIBS)
+
+lint: lintcheck
+
+pics/%.o: $(SRC)/common/list/%.c
+	$(COMPILE.c) -o $@ $<
+	$(POST_PROCESS_O)
+
+pics/%.o: ../%.c
+	$(COMPILE.c) -o $@ $<
+	$(POST_PROCESS_O)
+
+# include library targets
+include ../../Makefile.targ
diff --git a/usr/src/lib/libvmm/amd64/Makefile b/usr/src/lib/libvmm/amd64/Makefile
new file mode 100644
index 0000000000..5ba4f14479
--- /dev/null
+++ b/usr/src/lib/libvmm/amd64/Makefile
@@ -0,0 +1,19 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2018 Joyent, Inc.
+#
+
+include ../Makefile.com
+include ../../Makefile.lib.64
+
+install: all $(ROOTLIBS64) $(ROOTLINKS64)
