commit 33a8665864a41fce16b88ae3d27761414c720a03
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-19T13:09:36+00:00 (7 months ago)
    
    TRITON-1235 binder runtests should save TAP output

diff --git a/test/runtests b/test/runtests
index 6e8b609..300cc7d 100755
--- a/test/runtests
+++ b/test/runtests
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2018, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 #
@@ -101,6 +101,12 @@ do
     esac
 done
 
+# Setup a clean output dir.
+OUTPUT_DIR=/var/tmp/bindertest
+echo "# Setup a clean output dir ($OUTPUT_DIR)."
+rm -rf $OUTPUT_DIR
+mkdir -p $OUTPUT_DIR
+
 cd ${TOP}
 
 test_files=$(ls -1 test/*.test.js)
@@ -114,4 +120,4 @@ fi
 echo "# Test log at $LOG"
 ./build/node/bin/node ./node_modules/.bin/nodeunit \
     --reporter=$opt_reporter \
-    ${test_files} 2>$LOG
+    ${test_files}  2>$LOG | tee $OUTPUT_DIR/binder.tap
