commit 372f0a0f40c0079f6b22a1f1bd88afc0b859f9be
Author: Josh Wilsdon <jwilsdon@joyent.com>
Date:   2019-07-18T11:30:37-07:00 (3 months ago)
    
    TRITON-1650 prometheus image should discover global zones

diff --git a/CHANGES.md b/CHANGES.md
index 3655edc..185e6d7 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # triton-prometheus changes
 
+## 1.1.0
+
+Add support for global zone discovery on Triton.
+
 ## 1.0.0
 
 First version.
diff --git a/bin/update_global_zones.sh b/bin/update_global_zones.sh
new file mode 100755
index 0000000..5906d3b
--- /dev/null
+++ b/bin/update_global_zones.sh
@@ -0,0 +1,71 @@
+#!/bin/bash
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+#
+# This generates the global_zones.json file that's used by prometheus with the
+# files_sd service discovery in order that new CNs will automatically be
+# discovered by the config.
+#
+
+set -o errexit
+set -o pipefail
+
+ROOT_DIR=/opt/triton/prometheus
+CMON_AUTH_DIR=${ROOT_DIR}/keys
+CMON_KEY_FILE=${CMON_AUTH_DIR}/prometheus.key.pem
+CMON_CERT_FILE=${CMON_AUTH_DIR}/prometheus.cert.pem
+CONF_DIR=${ROOT_DIR}/etc
+OUTPUT_FILE=${CONF_DIR}/global_zones.json
+OUTPUT_FILE_TMP=${OUTPUT_FILE}.tmp.$$
+CONFIG_JSON=${CONF_DIR}/config.json
+DATACENTER_NAME=$(json -f "${CONFIG_JSON}" datacenter)
+DNS_DOMAIN=$(json -f "${CONFIG_JSON}" dns_domain)
+CMON_HOST="cmon.${DATACENTER_NAME}.cns.${DNS_DOMAIN}"
+
+function cleanup {
+    exit_code=$?
+    if [[ ${exit_code} -ne 0 ]]; then
+        echo "Failed to update global_zones.json: code ${exit_code}" >&2
+    fi
+    rm -f ${OUTPUT_FILE_TMP}
+}
+trap cleanup EXIT
+
+
+echo "[$(date -u)] Updating global_zones.json"
+
+#
+# We have to do the lookup ourselves since we want to use the local resolver,
+# but neither the pkgsrc nor the platform curl supports the --dns-servers
+# option.
+#
+# This command does the lookup and returns the IPs comma separated.
+#
+CMON_HOST_IPS=$(dig +short ${CMON_HOST} @127.0.0.1 | awk '{printf "%s%s",sep,$1; sep=","} END{print ""}')
+if [[ -z ${CMON_HOST_IPS} ]]; then
+    echo "Failed to get IPs for CMON host ${CMON_HOST}" >&2
+    exit 1
+fi
+
+curl -sS -k --max-time 120 \
+    --resolve ${CMON_HOST}:9163:${CMON_HOST_IPS} \
+    -E ${CMON_CERT_FILE} \
+    --key ${CMON_KEY_FILE} \
+    https://${CMON_HOST}:9163/v1/gz/discover \
+    | json cns \
+    | json -e "this.targets=[this.server_uuid+'.${CMON_HOST}:9163']; this.server_uuid=undefined; this.labels={'job': 'global_zones_${DATACENTER_NAME}'}" \
+    > ${OUTPUT_FILE_TMP}
+
+# If there was an error we'll go to cleanup() thanks to errexit & pipefail
+
+chown nobody:nobody ${OUTPUT_FILE_TMP}
+mv ${OUTPUT_FILE_TMP} ${OUTPUT_FILE}
+echo "Found $(json -f ${OUTPUT_FILE} length) GZs"
+
+exit 0
diff --git a/boot/setup.sh b/boot/setup.sh
index 5e259f6..ed2cc97 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -173,6 +173,27 @@ function prometheus_setup_prometheus {
     TRACE=1 /opt/triton/prometheus/bin/prometheus-configure
 }
 
+function prometheus_initialize_global_zones_json {
+    # Start with an empty file
+    echo "[]" > /opt/triton/prometheus/etc/global_zones.json
+}
+
+function prometheus_setup_crontab {
+    # Setup crontab
+    tmp_crontab=/tmp/prometheus-$$.cron
+    minute=$((RANDOM % 60))
+    crontab -l > $tmp_crontab
+    [[ $? -eq 0 ]] || fatal "Unable to write to $tmp_crontab"
+    echo '' >>$tmp_crontab
+    echo '# update the global_zones.json file' >>$tmp_crontab
+    # BASHSTYLED
+    echo "$minute * * * * /opt/triton/prometheus/bin/update_global_zones.sh >>/var/log/update_global_zones.log 2>&1" >>$tmp_crontab
+    crontab $tmp_crontab
+    [[ $? -eq 0 ]] || fatal "Unable import crontab"
+    rm -f $tmp_crontab
+}
+
+
 # ---- mainline
 
 prometheus_setup_delegate_dataset
@@ -199,6 +220,8 @@ else # "$FLAVOR" == "triton"
     source /opt/smartdc/boot/lib/util.sh
     sdc_common_setup
 
+    prometheus_initialize_global_zones_json
+    prometheus_setup_crontab
     prometheus_setup_named
     prometheus_setup_prometheus
 
@@ -208,6 +231,10 @@ else # "$FLAVOR" == "triton"
     sdc_log_rotation_add prometheus /var/svc/log/*prometheus*.log 1g
     sdc_log_rotation_setup_end
 
+    # Update the global_zones.json the first time
+    /opt/triton/prometheus/bin/update_global_zones.sh \
+        >>/var/log/update_global_zones.log 2>&1
+
     sdc_setup_complete
 fi
 
diff --git a/etc/prometheus.yml.in b/etc/prometheus.yml.in
index 719aa47..700efbe 100644
--- a/etc/prometheus.yml.in
+++ b/etc/prometheus.yml.in
@@ -37,6 +37,16 @@ scrape_configs:
           cert_file: %%CMON_CERT_FILE%%
           key_file: %%CMON_KEY_FILE%%
           insecure_skip_verify: %%CMON_INSECURE%%
+  - job_name: 'global_zones_%%DATACENTER_NAME%%'
+    scheme: https
+    tls_config:
+      cert_file: %%CMON_CERT_FILE%%
+      key_file: %%CMON_KEY_FILE%%
+      insecure_skip_verify: %%CMON_INSECURE%%
+    file_sd_configs:
+      - files:
+        # generated by the update_global_zones.sh cronjob
+        - '/opt/triton/prometheus/etc/global_zones.json'
   - job_name: 'prometheus_%%DATACENTER_NAME%%'
     static_configs:
     - targets:
diff --git a/package.json b/package.json
index 855921d..3bd4cea 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "triton-prometheus",
     "description": "Prometheus for TritonDC",
-    "version": "1.0.0",
+    "version": "1.1.0",
     "author": "Joyent (joyent.com)",
     "private": true,
     "repository": {
