commit 4e612f532ea578a44a0258c6abf9115a04e5fc29 (refs/changes/65/3065/3)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-02-07T12:13:18-08:00 (1 year, 8 months ago)
    
    joyent/node-docker-registry-client#27 make test can hang

diff --git a/test/v2.amazonecr.test.js b/test/v2.amazonecr.test.js
index 616d004..8a585af 100644
--- a/test/v2.amazonecr.test.js
+++ b/test/v2.amazonecr.test.js
@@ -188,6 +188,7 @@ test('v2 amazonecr', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -211,6 +212,7 @@ test('v2 amazonecr', function (tt) {
             //      "message": "Your Authorization Token is invalid."
             // }]}
             t.equal(err.statusCode, 400);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.dockerio.test.js b/test/v2.dockerio.test.js
index 9e82794..b005c16 100644
--- a/test/v2.dockerio.test.js
+++ b/test/v2.dockerio.test.js
@@ -230,6 +230,7 @@ test('v2 docker.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -247,6 +248,7 @@ test('v2 docker.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.gcrio.test.js b/test/v2.gcrio.test.js
index 8b65829..9125df5 100644
--- a/test/v2.gcrio.test.js
+++ b/test/v2.gcrio.test.js
@@ -150,6 +150,7 @@ test('v2 gcr.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -167,6 +168,7 @@ test('v2 gcr.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.gitlab.test.js b/test/v2.gitlab.test.js
index 6500c82..458fb6d 100644
--- a/test/v2.gitlab.test.js
+++ b/test/v2.gitlab.test.js
@@ -188,6 +188,7 @@ test('v2 registry.gitlab.com', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -205,6 +206,7 @@ test('v2 registry.gitlab.com', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.jfrogartifactory.test.js b/test/v2.jfrogartifactory.test.js
index c078753..5eb9bdc 100644
--- a/test/v2.jfrogartifactory.test.js
+++ b/test/v2.jfrogartifactory.test.js
@@ -229,6 +229,7 @@ test('v2 jfrog artifactory private repo (' + CONFIG.repo + ')', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -246,6 +247,7 @@ test('v2 jfrog artifactory private repo (' + CONFIG.repo + ')', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.quayio.test.js b/test/v2.quayio.test.js
index 04cf327..e2118a4 100644
--- a/test/v2.quayio.test.js
+++ b/test/v2.quayio.test.js
@@ -151,6 +151,7 @@ test('v2 quay.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -168,6 +169,7 @@ test('v2 quay.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
