commit d91cc15a266535e1e7fbbdf9bc19ff24fd0ae452 (refs/changes/65/3065/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-12-08T09:08:13-08:00 (1 year, 10 months ago)
    
    joyent/node-docker-registry-client#27 make test can hang

diff --git a/test/v2.amazonecr.test.js b/test/v2.amazonecr.test.js
index 4f78a72..706f3de 100644
--- a/test/v2.amazonecr.test.js
+++ b/test/v2.amazonecr.test.js
@@ -185,6 +185,7 @@ test('v2 amazonecr', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -208,6 +209,7 @@ test('v2 amazonecr', function (tt) {
             //      "message": "Your Authorization Token is invalid."
             // }]}
             t.equal(err.statusCode, 400);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.dockerio.test.js b/test/v2.dockerio.test.js
index 9e82794..b005c16 100644
--- a/test/v2.dockerio.test.js
+++ b/test/v2.dockerio.test.js
@@ -230,6 +230,7 @@ test('v2 docker.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -247,6 +248,7 @@ test('v2 docker.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.gcrio.test.js b/test/v2.gcrio.test.js
index 814583d..8c809c6 100644
--- a/test/v2.gcrio.test.js
+++ b/test/v2.gcrio.test.js
@@ -146,6 +146,7 @@ test('v2 gcr.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -163,6 +164,7 @@ test('v2 gcr.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.gitlab.test.js b/test/v2.gitlab.test.js
index 6500c82..458fb6d 100644
--- a/test/v2.gitlab.test.js
+++ b/test/v2.gitlab.test.js
@@ -188,6 +188,7 @@ test('v2 registry.gitlab.com', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -205,6 +206,7 @@ test('v2 registry.gitlab.com', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.jfrogartifactory.test.js b/test/v2.jfrogartifactory.test.js
index c078753..5eb9bdc 100644
--- a/test/v2.jfrogartifactory.test.js
+++ b/test/v2.jfrogartifactory.test.js
@@ -229,6 +229,7 @@ test('v2 jfrog artifactory private repo (' + CONFIG.repo + ')', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -246,6 +247,7 @@ test('v2 jfrog artifactory private repo (' + CONFIG.repo + ')', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
diff --git a/test/v2.quayio.test.js b/test/v2.quayio.test.js
index 04cf327..e2118a4 100644
--- a/test/v2.quayio.test.js
+++ b/test/v2.quayio.test.js
@@ -151,6 +151,7 @@ test('v2 quay.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 404);
+            badRepoClient.close();
             t.end();
         });
     });
@@ -168,6 +169,7 @@ test('v2 quay.io', function (tt) {
             t.ok(err, 'Expected an error on a missing repo');
             t.notOk(manifest_);
             t.equal(err.statusCode, 401);
+            badUserClient.close();
             t.end();
         });
     });
