From eb601d05ab5abe3be2eef51d433aef087fa7d3d3 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 25 Apr 2018 22:59:36 +0000
Subject: [PATCH] MORAY-466 Move "electric-moray" to new Triton origin image

---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index c9fb1e0..19fbec0 100644
--- a/Makefile
+++ b/Makefile
@@ -27,8 +27,8 @@ NODEUNIT_TESTS =	$(notdir $(wildcard test/*.test.js))
 
 
 NODE_PREBUILT_VERSION =	v0.10.48
-# Allow emoray builds on a VM other than sdc-multiarch/13.3.1.
-NODE_PREBUILT_IMAGE =	b4bdc598-8939-11e3-bea4-8341f6861379
+# Allow building on a SmartOS image other than sdc-*-multiarch 15.4.1.
+NODE_PREBUILT_IMAGE	= 18b094b0-eb01-11e5-80c1-175dac7ddf02
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG =	zone64
 endif
-- 
2.21.0

