commit 3897de3003d41dcd6c79af6433706fdb70e842d0
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-06-20T10:30:40-07:00 (4 months ago)
    
    TRITON-1721 migration needs a cleanup command

diff --git a/CHANGES.md b/CHANGES.md
index ba130ac..b314cff 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,8 +1,17 @@
 # sdc (SDC ops core zone) Changelog
 
+## 1.8.2
+
+- TRITON-1721 sdc-migrate needs a cleanup command.
+  Implemented the `sdc-migrate finalize` command.
+
+## 1.8.1
+
+- TRITON-1718 sdc-migrate is missing 'abort' help docs
+
 ## 1.8.0
 
-- TRITON-XXX Add sdc-migrate command to headnode's GZ
+- TRITON-1233 Add sdc-migrate command to headnode's GZ
 
 ## 1.7.1
 
diff --git a/bin/sdc-migrate b/bin/sdc-migrate
index 6e7f923..08e6b04 100755
--- a/bin/sdc-migrate
+++ b/bin/sdc-migrate
@@ -136,6 +136,7 @@ Sub-commands:
     switch VM_UUID               - switch control over to the migrated instance
     pause VM_UUID                - pause a migration operation
     abort VM_UUID                - abort the migration attempt
+    finalize VM_UUID             - cleanup, removes the original source instance
     watch VM_UUID                - watch an ongoing migration operation
 
 EOF
@@ -544,6 +545,38 @@ EOF
 }
 
 
+function migration_finalize() {
+    local result
+    local url
+    local vm_uuid=$1
+
+   # Handle the help argument.
+    if [[ $vm_uuid == "-h" || $vm_uuid == "--help" ]]; then
+        cat <<EOF
+Usage:
+        $0 finalize vm_uuid
+
+Finalizes the migration. Removes the original source instance and removes the
+migration record. The target instance remains untouched.
+EOF
+        return 1;
+    fi
+
+    if [[ -z ${vm_uuid} ]]; then
+        fatal "${action}: no instance uuid provided"
+    fi
+
+    url="/vms/${vm_uuid}?action=migrate&migration_action=${action}"
+
+    result=$(sdc-vmapi "$url" -X POST | $JSON -q -H)
+    if [[ -z $result ]]; then
+        result="Done - the migration is finished."
+    fi
+
+    echo "$result"
+}
+
+
 function migration_watch() {
     local vm_uuid=$1
     local need_newline=
@@ -593,7 +626,7 @@ function migration_watch() {
             fi
 
             if [[ $state == "successful" ]]; then
-                echo "Migration was successful ${message}"
+                echo "OK - switch was successful"
             elif [[ $state == "paused" ]]; then
                 if [[ $phase == "begin" ]]; then
                     echo "OK - ready for migration sync"
@@ -688,6 +721,10 @@ case ${action} in
         migration_estimate "$@"
         exit $?
         ;;
+    finalize)
+        migration_finalize "$@"
+        exit $?
+        ;;
     get)
         migration_get "$@"
         exit $?
diff --git a/package.json b/package.json
index cf8c3fd..71252cb 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "sdc",
     "description": "SmartDataCenter 'sdc' zone to hold ops/admin tools",
-    "version": "1.8.1",
+    "version": "1.8.2",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
