commit dc8970483fa4633a9de9f6affd11224bd8b16dac (refs/changes/74/4474/1)
Author: Kévin Berthommier <kevin.bertho@gmail.com>
Date:   2018-07-03T21:29:51+02:00 (1 year, 3 months ago)
    
    fix MultiError tests

diff --git a/test/tst.multierror.js b/test/tst.multierror.js
index bc8bac7..5128cf1 100644
--- a/test/tst.multierror.js
+++ b/test/tst.multierror.js
@@ -60,23 +60,23 @@ function main()
 	/* errorFromList */
 	mod_assert.throws(function () {
 		console.error(errorFromList());
-	}, /^AssertionError: errors \(\[object\]\) is required$/);
+	}, /errors \(\[object\]\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorFromList(null));
-	}, /^AssertionError: errors \(\[object\]\) is required$/);
+	}, /errors \(\[object\]\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorFromList({}));
-	}, /^AssertionError: errors \(\[object\]\) is required$/);
+	}, /errors \(\[object\]\) is required$/);
 
 	mod_assert.throws(function () {
 		console.error(errorFromList('asdf'));
-	}, /^AssertionError: errors \(\[object\]\) is required$/);
+	}, /errors \(\[object\]\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorFromList([ new Error(), 17 ]));
-	}, /^AssertionError: errors \(\[object\]\) is required$/);
+	}, /errors \(\[object\]\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorFromList([ new Error(), {} ]));
@@ -95,23 +95,23 @@ function main()
 	/* errorForEach */
 	mod_assert.throws(function () {
 		console.error(errorForEach());
-	}, /^AssertionError: err must be an Error$/);
+	}, /err must be an Error/);
 
 	mod_assert.throws(function () {
 		console.error(errorForEach(null));
-	}, /^AssertionError: err must be an Error$/);
+	}, /err must be an Error/);
 
 	mod_assert.throws(function () {
 		console.error(errorForEach(err1));
-	}, /^AssertionError: func \(func\) is required$/);
+	}, /func \(func\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorForEach(err1, {}));
-	}, /^AssertionError: func \(func\) is required$/);
+	}, /func \(func\) is required/);
 
 	mod_assert.throws(function () {
 		console.error(errorForEach({}, function () {}));
-	}, /^AssertionError: err must be an Error$/);
+	}, /err must be an Error/);
 
 	accum = [];
 	doAccum = function (e) { accum.push(e); };
