commit 7644dbf8bc15c2e927accb9dd1a2e2ff812cb63b
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-08-12T15:51:12+01:00 (8 weeks ago)
    
    OS-7921 SmartOS bash_completions should complete alias names

diff --git a/src/vm/etc/vmadm.completion b/src/vm/etc/vmadm.completion
index 89871e13..6d9f7f85 100644
--- a/src/vm/etc/vmadm.completion
+++ b/src/vm/etc/vmadm.completion
@@ -27,6 +27,9 @@ _vmadm()
         vms_uuids=$(zoneadm list -cp | grep -v ':global:' | cut -d':' -f5 | \
             sort | uniq)
         COMPREPLY=( $(compgen -W "${vms_uuids}" -- ${cur}) )
+        if [[ -z "$COMPREPLY" ]]; then
+            _zone_alias
+        fi
 
     elif [[ ${prev} == 'info'
           || ${prev} == 'sysrq' ]]; then
@@ -36,6 +39,9 @@ _vmadm()
             grep ":[bh]*[ky]v[em]:excl:" | \
             cut -d':' -f5 | sort | uniq)
         COMPREPLY=( $(compgen -W "${vms_uuids}" -- ${cur}) )
+        if [[ -z "$COMPREPLY" ]]; then
+            _zone_alias
+        fi
 
     elif [[ ${prev} == 'create-snapshot'
         || ${prev} == 'delete-snapshot'
@@ -44,6 +50,9 @@ _vmadm()
         vms_uuids=$(zoneadm list -cp |
 	    nawk -F: '$2 != "global" && $6 != "kvm" { print $5 }' | sort -u)
         COMPREPLY=( $(compgen -W "${vms_uuids}" -- ${cur}) )
+        if [[ -z "$COMPREPLY" ]]; then
+            _zone_alias
+        fi
 
     elif [[ ${prev} == 'validate' ]]; then
 
