From 1dc69d4d0a568055eaa1254dcf5088362264d946 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Fri, 7 Oct 2016 15:58:37 -0700
Subject: [PATCH] DOCKER-663 update changelog

---
 CHANGES.md | 1 +
 1 file changed, 1 insertion(+)

diff --git a/CHANGES.md b/CHANGES.md
index ad9ca33..7d2fff7 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,7 @@
 
 ## not yet released
 
+- DOCKER-663 Support Amazon ECR Registry
 
 ## 3.2.0
 
-- 
2.21.0

