commit 1dc69d4d0a568055eaa1254dcf5088362264d946 (refs/changes/26/626/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2016-10-07T15:58:37-07:00 (3 years ago)
    
    DOCKER-663 update changelog

diff --git a/CHANGES.md b/CHANGES.md
index ad9ca33..7d2fff7 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -2,6 +2,7 @@
 
 ## not yet released
 
+- DOCKER-663 Support Amazon ECR Registry
 
 ## 3.2.0
 
