From f05ce33518eedf49718297716dbc677aadb6817f Mon Sep 17 00:00:00 2001
From: Mike Zeller <mike@mikezeller.net>
Date: Fri, 4 Oct 2019 13:42:49 -0400
Subject: [PATCH] OS-8008 bhyve on AMD should report SVM as disabled

---
 usr/src/cmd/bhyve/xmsr.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/usr/src/cmd/bhyve/xmsr.c b/usr/src/cmd/bhyve/xmsr.c
index 7fe4804a2e..994445b3e3 100644
--- a/usr/src/cmd/bhyve/xmsr.c
+++ b/usr/src/cmd/bhyve/xmsr.c
@@ -205,6 +205,17 @@ emulate_rdmsr(struct vmctx *ctx, int vcpu, uint32_t num, uint64_t *val)
 			*val = 1;
 			break;
 
+#ifndef	__FreeBSD__
+		case MSR_VM_CR:
+			/*
+			 * We currently don't support nested virt.
+			 * Windows seems to ignore the cpuid bits and reads this
+			 * MSR anyways.
+			 */
+			*val = VM_CR_SVMDIS;
+			break;
+#endif
+
 		default:
 			error = -1;
 			break;
-- 
2.21.0

