From e712cc1ed427be67ad88f183f02f9121e3d60882 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 12 Sep 2017 00:53:14 +0000
Subject: [PATCH] OS-5858 Want support for X722

---
 overlay/generic/etc/driver_aliases | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 8959edc1..c3836113 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -499,6 +499,12 @@ i40e "pciex8086,1586"
 i40e "pciex8086,1589"
 i40e "pciex8086,158a"
 i40e "pciex8086,158b"
+i40e "pciex8086,37ce"
+i40e "pciex8086,37cf"
+i40e "pciex8086,37d0"
+i40e "pciex8086,37d1"
+i40e "pciex8086,37d2"
+i40e "pciex8086,37d3"
 i915 "pci8086,2562"
 i915 "pci8086,2572"
 i915 "pci8086,2582"
-- 
2.21.0

