commit f0c6f327b1254b298abe42fc3249aa88159151d9 (refs/changes/44/2544/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-09-12T00:53:14+00:00 (2 years, 1 month ago)
    
    OS-5858 Want support for X722

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index eda1c7b9..e3970be6 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -499,6 +499,12 @@ i40e "pciex8086,1586"
 i40e "pciex8086,1589"
 i40e "pciex8086,158a"
 i40e "pciex8086,158b"
+i40e "pciex8086,37ce"
+i40e "pciex8086,37cf"
+i40e "pciex8086,37d0"
+i40e "pciex8086,37d1"
+i40e "pciex8086,37d2"
+i40e "pciex8086,37d3"
 i915 "pci8086,2562"
 i915 "pci8086,2572"
 i915 "pci8086,2582"
