From cbe469a4a944935a3215abd6d1f3c696e3759b3b Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <jwilsdon@joyent.com>
Date: Thu, 27 Oct 2016 21:34:06 -0700
Subject: [PATCH] ZAPI-753 Update VMAPI to node v4.x

---
 targets.json.in | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 8e79d85..40e9c35 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -522,8 +522,15 @@ cat <<EOF
     "image_name": "vmapi",
     "image_description": "SDC VMAPI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
-    "pkgsrc": [],
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "coreutils-8.23nb2",
+      "curl-7.47.1",
+      "gsed-4.2.2nb4",
+      "patch-2.7.5",
+      "sudo-1.8.15"
+    ],
     "repos": [
         {"url": "git@github.com:joyent/sdc-vmapi.git"}
     ],
-- 
2.21.0

