commit 4799a3665a36dcf29e419b0c3989ba4622d5ae31 (refs/changes/95/2295/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-07-27T13:55:18-07:00 (2 years, 2 months ago)
    
    IMGAPI-636 docker pull retry not working for v2.1 images

diff --git a/lib/docker.js b/lib/docker.js
index 392e04f..7e43df3 100644
--- a/lib/docker.js
+++ b/lib/docker.js
@@ -465,7 +465,7 @@ function dockerDownloadAndImportLayer(opts, callback) {
 
         stream.on('data', function (chunk) {
             if (ctx.downloadsCanceled) {
-                stream.destroy();
+                resp.destroy();
                 progressStatus('Aborted');
                 return;
             }
@@ -494,6 +494,16 @@ function dockerDownloadAndImportLayer(opts, callback) {
             finish(errors.wrapErrorFromDrc(streamErr));
         });
 
+        if (resp !== stream) {
+            // IMGAPI-636: Listen for errors on the original response object,
+            // as the stream is a buffer-peek-stream (for v2.1 image layers),
+            // and drc will emit the error on the original stream (from the
+            // createBlobReadStream call).
+            resp.on('error', function (streamErr) {
+                finish(errors.wrapErrorFromDrc(streamErr));
+            });
+        }
+
         stor = app.chooseStor(newImage);
         stor.storeFileFromStream({
             image: newImage,
