commit 39b5aa973fbd1a9f69289fd2e269c7ab1914fbdc (refs/changes/11/3111/1)
Author: Jared Morrow <jm@joyent.com>
Date:   2017-10-19T15:43:25-06:00 (2 years ago)
    
    MANTA-3473 Update Muppet to node v4

diff --git a/targets.json.in b/targets.json.in
index d8ff174..9fdc3d4 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1112,12 +1112,12 @@ cat <<EOF
     "build_platform": "20141030T081701Z",
     "image_name": "manta-loadbalancer",
     "image_description": "Manta loadbalancer",
-    "image_uuid": "de411e86-548d-11e4-a4b7-3bb60478632a",
+    "image_uuid": "04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
     "pkgsrc": [
       "py27-curses",
-      "openssl-1.0.1u",
-      "stud-0.3p53nb2",
-      "zookeeper-client-3.4.6"
+      "openssl-1.0.2k",
+      "stud-0.3p53nb5",
+      "libzookeeper-3.4.6"
     ],
     "repos": [
       {"url": "git@github.com:joyent/muppet.git"}
