commit c9422bbeaa6b10c8d0b398f4678ccdc8877757ab (refs/changes/87/5387/1)
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2019-01-18T21:39:37+00:00 (9 months ago)
    
    OS-7521 disk enumerator should make protocol/label prop mutable

diff --git a/usr/src/lib/fm/topo/modules/common/disk/disk_common.c b/usr/src/lib/fm/topo/modules/common/disk/disk_common.c
index b1ddbd06e9..8088e79f02 100644
--- a/usr/src/lib/fm/topo/modules/common/disk/disk_common.c
+++ b/usr/src/lib/fm/topo/modules/common/disk/disk_common.c
@@ -21,7 +21,7 @@
 
 /*
  * Copyright (c) 2008, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -154,7 +154,8 @@ disk_set_props(topo_mod_t *mod, tnode_t *parent,
 			    "label error %s\n", topo_strerror(err));
 			goto error;
 		}
-	} else if (topo_node_label_set(dtn, label, &err) != 0) {
+	} else if (topo_prop_set_string(dtn, TOPO_PGROUP_PROTOCOL,
+	    TOPO_PROP_LABEL, TOPO_PROP_MUTABLE, label, &err) != 0) {
 		topo_mod_dprintf(mod, "disk_set_props: "
 		    "label_set error %s\n", topo_strerror(err));
 		goto error;
