commit bc2741229cf91a304a42a20a3ad436bb83f1daea (refs/changes/89/3789/1)
Author: Dan McDonald <danmcd@joyent.com>
Date:   2018-04-09T18:07:37-04:00 (1 year, 6 months ago)
    
    OS-6872 mac deadlock in aggrs

diff --git a/usr/src/uts/common/io/aggr/aggr_port.c b/usr/src/uts/common/io/aggr/aggr_port.c
index 74f34a3cd5..c8dbe00336 100644
--- a/usr/src/uts/common/io/aggr/aggr_port.c
+++ b/usr/src/uts/common/io/aggr/aggr_port.c
@@ -715,9 +715,10 @@ aggr_port_addvlan(aggr_port_t *port, uint_t idx, uint16_t vid)
 	 * nothing to do.
 	 */
 	if (port->lp_hwghs[idx] == NULL)
-		return (0);
+		err = 0;
+	else
+		err = mac_hwgroup_addvlan(port->lp_hwghs[idx], vid);
 
-	err = mac_hwgroup_addvlan(port->lp_hwghs[idx], vid);
 	mac_perim_exit(pmph);
 	return (err);
 }
@@ -737,9 +738,10 @@ aggr_port_remvlan(aggr_port_t *port, uint_t idx, uint16_t vid)
 		idx = 0;
 
 	if (port->lp_hwghs[idx] == NULL)
-		return (0);
+		err = 0;
+	else
+		err = mac_hwgroup_remvlan(port->lp_hwghs[idx], vid);
 
-	err = mac_hwgroup_remvlan(port->lp_hwghs[idx], vid);
 	mac_perim_exit(pmph);
 	return (err);
 }
