From 37dea895515c7f59d36385011bd09ab32acfd4fc Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 2 Oct 2018 17:25:39 -0700
Subject: [PATCH] TRITON-825 cns-server not forgiving enough of slow updates on
 replicas

---
 lib/dns-server.js | 32 +++++++++++++++++++++++++-------
 1 file changed, 25 insertions(+), 7 deletions(-)

diff --git a/lib/dns-server.js b/lib/dns-server.js
index ce099cf..8e231ab 100644
--- a/lib/dns-server.js
+++ b/lib/dns-server.js
@@ -59,6 +59,7 @@ function DNSServer(opts) {
 		this.notifyInterval = consts.DEFAULT_MIN_NOTIFY_INT;
 	this.checkAndNotify();
 	this.unblacklistTimers = {};
+	this.checkTimers = {};
 
 	this.peerCounters = {};
 
@@ -1072,20 +1073,33 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 			q.addAnswer(z, soa, TTL);
 
 			var log  = q.log;
-			setTimeout(function () {
+			if (addr !== '127.0.0.1' && addr !== '::1')
+				ensurePeerCheckTimer();
+			q.send();
+			cb();
+			q.log.info('responded ok');
+
+			function ensurePeerCheckTimer() {
+				if (!self.checkTimers[normalizeIP(addr)]) {
+					self.checkTimers[normalizeIP(addr)] =
+					    setTimeout(doPeerChecks, 2000);
+				}
+			}
+
+			function doPeerChecks() {
 				var arg = {};
 				vasync.pipeline({
 					funcs: [claim, checkSOA, checkVer],
 					arg: arg
 				}, function () {
+					delete (self.checkTimers[
+					    normalizeIP(addr)]);
+					if (arg.rearm)
+						ensurePeerCheckTimer();
 					if (arg.handle)
 						arg.handle.release();
 				});
-			}, 1000);
-
-			q.send();
-			cb();
-			q.log.info('responded ok');
+			}
 
 			function claim(_, ccb) {
 				self.redisPool.claim(
@@ -1113,7 +1127,11 @@ DNSServer.prototype.handleTransfer = function (q, z, cb) {
 						    'xfer follow-up showed ' +
 						    'peer still stuck on ' +
 						    'serial %d', qsoa.serial);
-						ccb(new Error('Stuck'));
+						_.rearm = true;
+						_.redis.hset(
+						    'peer:' + normalizeIP(addr),
+						    z, String(qsoa.serial),
+						    ccb);
 						return;
 					}
 					self.incrPeerCounter(addr, 'goodxfer');
-- 
2.21.0

