From ecd3409a5d9915667cbe72f12779ebfadc36f09d Mon Sep 17 00:00:00 2001
From: Suk Won Choi <sw00.choi@samsung.com>
Date: Mon, 14 Nov 2016 08:00:54 +0000
Subject: [PATCH] joyent/sdc-docker#95 untagging image used by a container not
 working

---
 lib/backends/sdc/images.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/backends/sdc/images.js b/lib/backends/sdc/images.js
index 835cdb2..b092a66 100644
--- a/lib/backends/sdc/images.js
+++ b/lib/backends/sdc/images.js
@@ -1095,6 +1095,11 @@ function deleteImage(opts, callback) {
     }
 
     function verifyNotInUse(ctx, cb) {
+        if (dontDeleteImages) {
+            cb();
+            return;
+        }
+
         // If force === false this function will return an error if there is at
         // least a running or stopped VM using the image.
         // If force === true, it will return an error if there is at least a
-- 
2.21.0

