From ad4daaa99e2ced10b58b895e97e2f874df79399e Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 22 Mar 2017 17:33:08 +0000
Subject: [PATCH] MORAY-409 Moray client needs to pass timeout to server for
 sql()

---
 lib/meta.js | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/lib/meta.js b/lib/meta.js
index 70cc641..acae4bb 100644
--- a/lib/meta.js
+++ b/lib/meta.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -142,8 +142,12 @@ function sql(rpcctx, statement, values, options) {
     assert.string(statement, 'statement');
     assert.ok(Array.isArray(values));
     assert.object(options, 'options');
+    assert.optionalNumber(options.timeout, 'options.timeout');
 
-    opts = { req_id: options.req_id || libuuid.create() };
+    opts = {
+        req_id: options.req_id || libuuid.create(),
+        timeout: options.timeout
+    };
     log = rpc.childLogger(rpcctx, opts);
     res = new events.EventEmitter();
 
-- 
2.21.0

