commit 09f8b2b5e58131f8dc5d94e327cb6c46f22e104d
Author: John Levon <john.levon@joyent.com>
Date:   2019-05-09T10:29:54+00:00 (5 months ago)
    
    OS-7792 git pbchk copyright check could be more explicit

diff --git a/usr/src/tools/onbld/Checks/Copyright.py b/usr/src/tools/onbld/Checks/Copyright.py
index 00ca8829b9..b358484c2b 100644
--- a/usr/src/tools/onbld/Checks/Copyright.py
+++ b/usr/src/tools/onbld/Checks/Copyright.py
@@ -57,7 +57,8 @@ def copyright(fh, filename=None, output=sys.stderr):
 		err(output, "no copyright message found", filename)
 		ret = 1
 	elif goodrights == 0:
-		err(output, "no correct Joyent copyright found", filename)
+		err(output, "'Copyright %s Joyent, Inc.' not found" %
+		    time.strftime('%Y'), filename)
 		ret = 1
 
 	return ret
