From dc33d4ae36b39035c3e616325e8f7f4ee351f304 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Thu, 1 Sep 2016 11:42:57 -0700
Subject: [PATCH] TOOLS-1540 mg README and help output refer to removed targets

---
 README.md | 19 +++++--------------
 configure |  4 +---
 2 files changed, 6 insertions(+), 17 deletions(-)

diff --git a/README.md b/README.md
index 4465a76..3898e32 100644
--- a/README.md
+++ b/README.md
@@ -61,21 +61,12 @@ cloudapi:
 
 Likewise for any target (`cat targets.json | json --keys`)
 
-There is also a special target to build everything from scratch *except* the
-platform (because the platform is by far the longest part of the build):
-
-    ./configure -t all-except-platform
-    make all-except-platform
-
 To include ancillary closed repositories in a usb-headnode build (which the
-'all', 'all-except-platform', 'usb-headnode', and 'usb-headnode-debug'
-targets will build), you need to pass the `-j` option to configure.  Note
-that this will cause your build to fail if you do not have access to the
-private ancillary repositories, so it should be used only when building
-Joyent products:
-
-    ./configure -j -t all
-    gmake all
+'usb-headnode' and 'usb-headnode-debug' targets will build), you need to pass
+the `-j` option to configure.  Note that this will cause your build to fail if
+you do not have access to the private ancillary repositories, so it should be
+used only when building Joyent products.
+
 
 # Prerequisites
 
diff --git a/configure b/configure
index 681634c..6e8c387 100755
--- a/configure
+++ b/configure
@@ -529,9 +529,7 @@ function print_help() {
     echo "               a feature branch on one of the many repos used for"
     echo "               a target."
     echo "  -t TARGET    Prepare to build only this target. Valid targets are"
-    echo "               the keys of 'targets.json' plus a special value"
-    echo "               'all-except-platform' to build all targets except the"
-    echo "               slow-to-build platform."
+    echo "               the keys of 'targets.json'."
     echo "  -r           Regenerate config.mk. This doesn't touch repos in"
     echo "               'build/' or preload in 'bits/'."
     echo "  -c CACHE-DIR Specify a cache directory. Default: './cache'."
-- 
2.21.0

