commit 8aafd08707cd3ec7c63887c4ed90229066c874a6
Author: John Levon <john.levon@joyent.com>
Date:   2019-03-27T20:28:06+00:00 (6 months ago)
    
    TRITON-1353 Triton systems should ship with HT disabled

diff --git a/bin/build-tar-image b/bin/build-tar-image
index 245fba70..288d59b1 100755
--- a/bin/build-tar-image
+++ b/bin/build-tar-image
@@ -179,6 +179,8 @@ function generate_grub_menu
     console=$(build_spec console)
     local default_boot_option
     default_boot_option=$(build_spec default-boot-option)
+    local ht_enabled
+    ht_enabled=$(build_spec ht_enabled)
 
     # Feature flag for HEAD-2093
     local enable_dr
@@ -191,6 +193,7 @@ function generate_grub_menu
     [[ -z "${serial_dev}" ]] && serial_dev=ttyb
     [[ -z "${console}" ]] && console="serial"
     [[ -z "${default_boot_option}" ]] && default_boot_option=0
+    [[ -z "${ht_enabled}" ]] && ht_enabled="false"
 
     #
     # This section describes the serial-dev and console parameters.  These
@@ -260,10 +263,12 @@ function generate_grub_menu
     esac
 
     serial_string="--speed=115200 --unit=${unit} --word=8 --parity=no --stop=1"
+    bootargs="headnode=true,ht_enabled=${ht_enabled}"
 
     sed \
         -e "s/^#SERIAL/serial ${serial_string}/" \
         -e "s/DEFAULT_CONSOLE/${console}/g" \
+        -e "s/BOOTARGS/${bootargs}/g" \
         -e "s/^default.*$/default ${default_boot_option}/" \
         -e "s/^#DR /${DR_VAL}/" \
         boot/grub/menu.lst.tmpl \
diff --git a/boot/grub/menu.lst.tmpl b/boot/grub/menu.lst.tmpl
index 19cf10cc..2463bca2 100755
--- a/boot/grub/menu.lst.tmpl
+++ b/boot/grub/menu.lst.tmpl
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright 2019, Joyent, Inc.
 #
 
 default 0
@@ -14,33 +14,34 @@ min_mem64 1024
 #SERIAL
 terminal composite
 variable os_console DEFAULT_CONSOLE
+variable bootargs BOOTARGS
 
 title Compute Node (PXE)
    kernel /boot/ipxe.lkrn
    module /boot/default.ipxe
 
 title Live 64-bit
-   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",headnode=true
+   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",${bootargs}
    module /PLATFORM/i86pc/amd64/boot_archive
    module /PLATFORM/i86pc/amd64/boot_archive.hash
 
 #PREV title Live 64-bit (rollback to PREV_PLATFORM_VERSION)
-#PREV    kernel$ /PREV_PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",headnode=true
+#PREV    kernel$ /PREV_PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",${bootargs}
 #PREV    module /PREV_PLATFORM/i86pc/amd64/boot_archive
 #PREV    module /PREV_PLATFORM/i86pc/amd64/boot_archive.hash
 
 title Live 64-bit Rescue (no importing zpool)
-   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",headnode=true,noimport=true
+   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",noimport=true,${bootargs}
    module /PLATFORM/i86pc/amd64/boot_archive
    module /PLATFORM/i86pc/amd64/boot_archive.hash
 
 title Live 64-bit +kmdb
-   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -kd -B console=${os_console},${os_console}-mode="115200,8,n,1,-",headnode=true
+   kernel$ /PLATFORM/i86pc/kernel/amd64/unix -kd -B console=${os_console},${os_console}-mode="115200,8,n,1,-",${bootargs}
    module /PLATFORM/i86pc/amd64/boot_archive
    module /PLATFORM/i86pc/amd64/boot_archive.hash
 
 #DR title Live 64-bit (Disaster recovery mode)
-#DR    kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",headnode=true,DISASTER_RECOVERY=true
+#DR    kernel$ /PLATFORM/i86pc/kernel/amd64/unix -B console=${os_console},${os_console}-mode="115200,8,n,1,-",DISASTER_RECOVERY=true,${bootargs}
 #DR    module /PLATFORM/i86pc/amd64/boot_archive
 #DR    module /PLATFORM/i86pc/amd64/boot_archive.hash
 
diff --git a/build.spec b/build.spec
index 3883dfa2..41d44f19 100644
--- a/build.spec
+++ b/build.spec
@@ -5,6 +5,7 @@
   "coal-enable-serial": true,
   "no-rabbit": true,
   "clean-cache": true,
+  "ht_enabled": false,
   "// joyent-build": "set to true to enable ancillary repository use",
 
   "features": {
