commit b39e1fa51ed4cd72732d6996612bf64f0287a1f7 (refs/changes/26/5126/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-11-23T10:37:06-08:00 (11 months ago)
    
    MANTA-4033 get lagging Manta components on the latest v0.10.x

diff --git a/Makefile b/Makefile
index 0992d47..c044040 100644
--- a/Makefile
+++ b/Makefile
@@ -38,7 +38,7 @@ JSSTYLE_FLAGS    = -f tools/jsstyle.conf
 SMF_MANIFESTS_IN = smf/manifests/wrasse.xml.in
 
 NAME=wrasse
-NODE_PREBUILT_VERSION=v0.10.25
+NODE_PREBUILT_VERSION=v0.10.48
 NODE_PREBUILT_TAG=zone
 NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
 
