commit c6437101ef4d25f01f470958bb117f9e29256189 (refs/changes/67/3267/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-01-25T17:15:42-08:00 (1 year, 8 months ago)
    
    TRITON-86 booter should allow .local override for packetOpts.file

diff --git a/lib/dhcpd.js b/lib/dhcpd.js
index fd974ae..2eb0454 100755
--- a/lib/dhcpd.js
+++ b/lib/dhcpd.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 /*
@@ -21,6 +21,7 @@ var dgram = require('dgram');
 var dhcp = require('./dhcp');
 var mod_boot_files = require('./boot-files');
 var mod_clients = require('./clients');
+var path = require('path');
 var sprintf = require('sprintf').sprintf;
 var uuid = require('node-uuid');
 
@@ -133,7 +134,29 @@ DHCPD.prototype.handleMessage = function (msg, peer) {
             return;
         }
 
+        var localFile;
         var packetOpts = self.buildPacketOpts(packet, params.bootParams, log);
+
+        // Allow local override of boot files, for testing
+        if (packetOpts.file && self.config.tftpRoot) {
+            localFile = packetOpts.file + '.local';
+
+            log.trace({
+                fileOpt: packetOpts.file,
+                mac: mac,
+                tftpRoot: self.config.tftpRoot
+            }, 'checking for local override');
+
+            if (fs.existsSync(path.join(self.config.tftpRoot, localFile))) {
+                log.warn({
+                    generatedFile: packetOpts.file,
+                    mac: mac,
+                    overrideFile: localFile
+                }, 'local override exists, using instead of generated file');
+                packetOpts.file = localFile;
+            }
+        }
+
         self.sendReply(packet, packetOpts, log, function (sendErr) {
             if (sendErr) {
                 log.error(sendErr, 'Error sending reply');
diff --git a/package.json b/package.json
index 2077e73..16e6bbc 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "booter",
   "description": "DHCP server for booting SDC compute nodes",
-  "version": "1.3.0",
+  "version": "1.3.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
