commit e9d69d53af37fdefca1141ab7e67cb086186ef12 (refs/changes/06/1006/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton instance ip 516a51c5-f9cc-4145-93a9-71d0a518ae8d

diff --git a/lib/do_instance/do_ip.js b/lib/do_instance/do_ip.js
index 9dd5af7..6bfdea7 100644
--- a/lib/do_instance/do_ip.js
+++ b/lib/do_instance/do_ip.js
@@ -12,6 +12,7 @@
 
 var format = require('util').format;
 
+var common = require('../common');
 var errors = require('../errors');
 
 
@@ -29,20 +30,25 @@ function do_ip(subcmd, opts, args, cb) {
 
     var cli = this.top;
 
-    cli.tritonapi.getInstance(args[0], function (err, inst) {
-        if (err) {
-            cb(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
+        cli.tritonapi.getInstance(args[0], function (err, inst) {
+            if (err) {
+                cb(err);
+                return;
+            }
 
-        if (!inst.primaryIp) {
-            cb(new errors.TritonError(format(
-                'primaryIp not found for instance "%s"', args[0])));
-            return;
-        }
+            if (!inst.primaryIp) {
+                cb(new errors.TritonError(format(
+                    'primaryIp not found for instance "%s"', args[0])));
+                return;
+            }
 
-        console.log(inst.primaryIp);
-        cb();
+            console.log(inst.primaryIp);
+            cb();
+        });
     });
 }
 
