commit 1852397f582d5cc8b5212d4e58078ad233b3d751 (refs/changes/86/3586/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-03-12T17:26:23+00:00 (1 year, 7 months ago)
    
    OS-6678 need bhyve modules in name_to_major

diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index 7f49def4..cfab8bff 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -298,3 +298,6 @@ iwn 301
 xhci 302
 qede 303
 buildstamp 304
+vmm 305
+viona 306
+ppt 307
