From 62f6e6b49d6b466512ddf87cb674100b1408a5cb Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex@uq.edu.au>
Date: Tue, 1 Oct 2019 15:14:48 +1000
Subject: [PATCH] joyent/sdc-ufds#3 want smartlogin to support administrator
 role

---
 lib/machines.js | 70 ++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 58 insertions(+), 12 deletions(-)

diff --git a/lib/machines.js b/lib/machines.js
index 3acba6b..9bf9cc2 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -6,6 +6,7 @@
 
 /*
  * Copyright 2019 Joyent, Inc.
+ * Copyright 2019 The University of Queensland
  */
 
 /*
@@ -1348,22 +1349,67 @@ function loadSSHKeys(req, res, next) {
             return next(err);
         }
 
-        if (!keys || !keys.length) {
-            return next(new KeyRequiredError());
-        }
+        var roleFilter = '(&(objectclass=sdcaccountrole)(name=administrator))';
+        req.sdc.ufds.listRoles(req.account.uuid, roleFilter,
+            function (err2, roles) {
+
+            if (err2) {
+                return next(err2);
+            }
 
-        req.root_authorized_keys = '';
-        req.keys = keys;
-        keys.forEach(function (k) {
-            req.root_authorized_keys += k.openssh + '\n';
+            var adminRole = roles[0];
+            if (!adminRole || !adminRole.uniquememberdefault) {
+                return gotKeys();
+            }
+
+            var members = adminRole.uniquememberdefault;
+            if (!Array.isArray(members)) {
+                members = [members];
+            }
+
+            vasync.forEachParallel({
+                func: addUserKeys,
+                inputs: members
+            }, gotKeys);
+
+            function addUserKeys(dn, ccb) {
+                req.sdc.ufds.search(dn, {
+                    scope: 'sub',
+                    filter: '(objectclass=sdckey)'
+                }, function (err4, userKeys) {
+                    if (err4) {
+                        return ccb(err4);
+                    }
+                    userKeys.forEach(function (key) {
+                        keys.push(key);
+                    });
+                    return ccb();
+                });
+            }
         });
 
-        log.debug({
-            customer: req.account.uuid,
-            root_authorized_keys: req.root_authorized_keys
-        }, 'Loaded keys for %s', req.account.login);
+        function gotKeys(err3) {
+            if (err3) {
+                return next(err3);
+            }
+
+            if (!keys || !keys.length) {
+                return next(new KeyRequiredError());
+            }
 
-        return next();
+            req.root_authorized_keys = '';
+            req.keys = keys;
+            keys.forEach(function (k) {
+                req.root_authorized_keys += k.openssh + '\n';
+            });
+
+            log.debug({
+                customer: req.account.uuid,
+                root_authorized_keys: req.root_authorized_keys
+            }, 'Loaded keys for %s', req.account.login);
+
+            return next();
+        }
     });
 }
 
-- 
2.17.2 (Apple Git-113)

