From 3fec61bbece368e15156e7083fb192c41fd07053 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 6 Sep 2017 19:14:59 +0000
Subject: [PATCH] MORAY-426 Accept the "offset" option as a string

---
 lib/schema.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/schema.js b/lib/schema.js
index f7fe64f..b1904e6 100644
--- a/lib/schema.js
+++ b/lib/schema.js
@@ -178,7 +178,8 @@ AJV_ENV.addSchema({
         },
         'offset': {
             description: 'An offset into a query result set',
-            type: 'number',
+            type: [ 'integer', 'string' ],
+            pattern: '^[0-9]+$',
             minimum: 0
         },
         'sort': {
-- 
2.21.0

