commit 3fec61bbece368e15156e7083fb192c41fd07053 (refs/changes/10/2510/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-09-06T19:14:59+00:00 (2 years, 1 month ago)
    
    MORAY-426 Accept the "offset" option as a string

diff --git a/lib/schema.js b/lib/schema.js
index f7fe64f..b1904e6 100644
--- a/lib/schema.js
+++ b/lib/schema.js
@@ -178,7 +178,8 @@ AJV_ENV.addSchema({
         },
         'offset': {
             description: 'An offset into a query result set',
-            type: 'number',
+            type: [ 'integer', 'string' ],
+            pattern: '^[0-9]+$',
             minimum: 0
         },
         'sort': {
