From 8391a075c8d83f3fcaf0dda6fe60bfb9200a1b4f Mon Sep 17 00:00:00 2001
From: andrew <andrew@whartonlabs.com>
Date: Fri, 3 Feb 2017 11:42:40 +1100
Subject: [PATCH] Fixed line length > 80 chars which 'make check' was
 complaining about

---
 lib/tritonapi.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/tritonapi.js b/lib/tritonapi.js
index 848f5f7..b8b0bd2 100644
--- a/lib/tritonapi.js
+++ b/lib/tritonapi.js
@@ -2287,7 +2287,8 @@ TritonApi.prototype.deletePolicy = function deletePolicy(opts, cb) {
  *
  * @param {Object} opts
  *      - {String} id: Required. The instance name, short id, or id (a UUID).
- *      - {String} package: Required. The new package name, shortId or id (a UUID).
+ *      - {String} package: Required. The new package name, shortId
+ *        or id (a UUID).
  *      - {Boolean} wait: Wait (via polling) until the rename is complete.
  *        Warning: A concurrent resize of the same instance can result in this
  *        polling being unable to notice the change. Use `waitTimeout` to
-- 
2.21.0

