From c860d91293351c3ed53e4fbd13552fde102409aa Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Wed, 11 Apr 2018 14:20:21 +0000
Subject: [PATCH] OS-6886 SMB libraries are missing

---
 manifest | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/manifest b/manifest
index 24dec079c9..d1af0be4b1 100644
--- a/manifest
+++ b/manifest
@@ -10011,7 +10011,11 @@ f usr/lib/smbsrv/dtrace/smbnode.d 0555 root bin
 f usr/lib/smbsrv/dtrace/smbsrv.d 0555 root bin
 f usr/lib/smbsrv/dtrace/smbvfs.d 0555 root bin
 f usr/lib/smbsrv/libmlsvc.so.1 0755 root bin
+s usr/lib/smbsrv/libmlsvc.so=libmlsvc.so.1
+f usr/lib/smbsrv/libsmb.so.1 0755 root bin
+s usr/lib/smbsrv/libsmb.so=libsmb.so.1
 f usr/lib/smbsrv/libsmbns.so.1 0755 root bin
+s usr/lib/smbsrv/libsmbns.so=libsmbns.so.1
 f usr/lib/smbsrv/smbd 0755 root bin
 d usr/lib/smedia 0755 root bin
 d usr/lib/smedia/amd64 0755 root bin
-- 
2.21.0

