From 8443f59f28d5089597e2fc2e0800d661b00b32b3 Mon Sep 17 00:00:00 2001
From: Josh Wilsdon <jwilsdon@joyent.com>
Date: Thu, 27 Oct 2016 15:47:07 -0700
Subject: [PATCH] PAPI-132 PAPI should be updated to node v4

---
 targets.json.in | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 33eeabb..8e79d85 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -546,8 +546,15 @@ cat <<EOF
     "image_name": "papi",
     "image_description": "SDC PAPI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
-    "pkgsrc": [],
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "coreutils-8.23nb2",
+      "curl-7.47.1",
+      "gsed-4.2.2nb4",
+      "patch-2.7.5",
+      "sudo-1.8.15"
+    ],
     "repos": [
         {"url": "git@github.com:joyent/sdc-papi.git"}
     ],
-- 
2.21.0

