commit 57f9a0b1892b57846e529cac354065ae451bb708
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-10-07T09:43:37+01:00 (15 hours ago)
    
    OS-6146 Update copyright date in /etc/release

diff --git a/.gitignore b/.gitignore
index d1aedc4c..88aa72af 100644
--- a/.gitignore
+++ b/.gitignore
@@ -58,6 +58,7 @@
 /tools/pwgen-2.06/configure
 boot.manifest.d/
 boot.manifest.gen
+etc_release.template.gen
 man/man/man1m/diskinfo.1m
 man/man/man1m/disklayout.1m
 man/man/man1m/dsadm.1m
diff --git a/src/Makefile b/src/Makefile
index f1a5b45e..af1469e2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -351,10 +351,11 @@ LIBSDC_TARGETS = \
 
 ETC_TARGETS = \
 	etc/nodename \
-	etc/notices/COPYRIGHT \
 	etc/rtc_config \
 	etc/ssh/sshd_config
 
+ETC_COPYRIGHT_TARGET = etc/notices/COPYRIGHT
+
 BHYVE_TARGETS = \
 	bhyve/uefi-csm-rom.bin \
 	bhyve/uefi-rom.bin \
@@ -375,7 +376,7 @@ check : TARGET = check
 
 all: $(TARGETS) $(USR_LIB_TARGETS) $(SMARTDC_TARGETS) \
     $(SMARTDC_LIB_TARGETS) $(LIBSDC_TARGETS) $(ETC_TARGETS) \
-    $(SUBDIRS) $(BHYVE_TARGETS)
+    $(SUBDIRS) $(BHYVE_TARGETS) $(ETC_COPYRIGHT_TARGET)
 
 
 $(SUBDIRS): FRC
@@ -499,6 +500,8 @@ install: all $(SUBDIRS)
 	mv $(NOMKNOD_ROOT64)/$(NOMKNOD_64) $(NOMKNOD_ROOT64)/$(NOMKNOD_SO)
 	mkdir -p $(DESTDIR)/usr/share/bhyve
 	cp $(BHYVE_TARGETS) $(DESTDIR)/usr/share/bhyve
+	sed -e "s/%%YEAR%%/$$(date +%Y)/1" $(ETC_COPYRIGHT_TARGET) \
+		> $(DESTDIR)/$(ETC_COPYRIGHT_TARGET)
 
 check: $(JSLINT) $(SUBDIRS)
 	@echo "==> Running cstyle..."
diff --git a/src/etc/notices/COPYRIGHT b/src/etc/notices/COPYRIGHT
index 1834bac7..a98f3199 100644
--- a/src/etc/notices/COPYRIGHT
+++ b/src/etc/notices/COPYRIGHT
@@ -1,5 +1,4 @@
-Copyright © 2011 Joyent, Inc., 345 California Street, Suite 2000
-San Francisco, California 94104 U.S.A. All rights reserved.
+Copyright %%YEAR%% Joyent, Inc.
 
 Copyright © 2010 Oracle, Inc., All rights reserved.
 
diff --git a/tools/build_live b/tools/build_live
index ed5528d4..dad9ad66 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -414,13 +414,20 @@ function bi_create_password
 #
 function bi_gen_version_files
 {
-	local etcrelease_template="$bi_wsroot/tools/etc_release.template.txt"
+	local etcrelease_template="$bi_wsroot/etc_release.template.gen"
 	local build_etcrelease="$bi_wsroot/tools/build_etcrelease"
+	local etcrelease_year=$(date +%Y)
 	local bvfile="$bi_mnt_root/etc/versions/build"
 	local gitstatus="$bi_out_dir/gitstatus.json"
 
 	bi_emit_start 'Building version files'
 
+	cat > "$etcrelease_template" << EOF
+                      SmartOS %%VERSION%% x86_64
+                        Copyright $etcrelease_year Joyent, Inc.
+
+   Built with the following components:
+EOF
 	"$build_etcrelease" >"$gitstatus" ||
 	    fail "could not generate $gitstatus"
 
@@ -440,7 +447,6 @@ function bi_gen_version_files
 	"$build_etcrelease" "$bi_buildstamp" "$etcrelease_template" |
 	    pfexec /usr/bin/tee "$bi_mnt_root/etc/release" >/dev/null ||
 	     fail 'failed to build "/etc/release" file'
-
 	bi_emit_done
 }
 
