From d061f6499a072fe2405f84bbabc41c55f29095ff Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Fri, 9 Jun 2017 20:20:06 +0000
Subject: [PATCH] This is an example commit

---
 README.md | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/README.md b/README.md
index d91cefe..02774a3 100644
--- a/README.md
+++ b/README.md
@@ -3,3 +3,5 @@
 This is a dummy repository for verifying cr.joyent.us behavior.
 
 Amend this file to try new CRs.
+
+This is a new sentence.
-- 
2.21.0

