From 3b6e2ad548280bc0931026b939fc19ff7835b9c7 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Fri, 9 Mar 2018 23:56:19 +0000
Subject: [PATCH] TRITON-224 Release sdc-fwrule@2.0.0

---
 lib/rule.js  | 2 +-
 package.json | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/rule.js b/lib/rule.js
index 99cdecf..8032a3f 100644
--- a/lib/rule.js
+++ b/lib/rule.js
@@ -29,7 +29,7 @@
 'use strict';
 
 var mod_net = require('net');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var sprintf = require('extsprintf').sprintf;
 var util = require('util');
 var validators = require('./validators');
diff --git a/package.json b/package.json
index 35976ed..960f1f5 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "fwrule",
   "description": "Parser for Joyent firewall rules",
-  "version": "1.4.1",
+  "version": "2.0.0",
   "author": "Joyent (joyent.com)",
   "keywords": [ "smartos", "triton" ],
   "repository": {
@@ -10,7 +10,7 @@
   },
   "dependencies": {
     "ip6addr": "0.2.2",
-    "node-uuid": "1.2.0",
+    "uuid": "3.0.0",
     "extsprintf": "1.0.1",
     "verror": "1.3.4"
   },
-- 
2.21.0

