From 02228074b6441f1aa291f5f133b5ce7ac38a26b8 Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Sat, 2 Mar 2019 01:09:11 +0000
Subject: [PATCH] MANTA-4144 Add metricPorts mdata variable to pgstatsmon

---
 boot/firstboot.sh | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/boot/firstboot.sh b/boot/firstboot.sh
index de42ae7..b8a688d 100644
--- a/boot/firstboot.sh
+++ b/boot/firstboot.sh
@@ -20,6 +20,7 @@ NAME=pgstatsmon
 #
 
 SVC_ROOT="/opt/smartdc/$NAME"
+SAPI_CONFIG="$SVC_ROOT/etc/config.json"
 
 #
 # Build PATH from this list of directories.  This PATH will be used both in the
@@ -79,4 +80,15 @@ if ! svccfg import "/opt/smartdc/$NAME/smf/manifests/$NAME.xml"; then
 	fatal 'could not import SMF service'
 fi
 
+# Get the metricPort from the sapi config file to allow scraping by cmon-agent.
+#
+# Note that the config file is guaranteed to have been written by config-agent
+# at this point, because we've already called manta_common_setup, which calls
+# manta_enable_config_agent, which enables the config-agent service
+# synchronously.
+if [[ ! -f $SAPI_CONFIG ]]; then
+	fatal 'SAPI config not found'
+fi
+mdata-put metricPorts $(< "$SAPI_CONFIG" json 'target.port')
+
 manta_common_setup_end
-- 
2.21.0

