commit bc042921dc21e57e2c8c573f4c4c439da966311f
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-03-02T00:49:07+00:00 (7 months ago)
    
    MANTA-4143 Add metricPorts mdata variable to muskie

diff --git a/boot/setup.sh b/boot/setup.sh
index 98dcc6e..ef9a756 100644
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -61,9 +61,11 @@ function manta_setup_muskie {
     #Build the list of ports.  That'll be used for everything else.
     local ports
     local insecure_ports
+    local metric_ports
     for (( i=1; i<=$num_instances; i++ )); do
         ports[$i]=`expr 8080 + $i`
         insecure_ports[$i]=`expr 9080 + $i`
+        metric_ports[$i]=`expr ${ports[i]} + 800`
     done
 
     #To preserve whitespace in echo commands...
@@ -106,6 +108,10 @@ function manta_setup_muskie {
     # Setup haproxy after the muskie's are kicked up
     svcadm enable "manta/haproxy" || fatal "unable to start haproxy"
 
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    mdata-put metricPorts $(IFS=','; echo "${metric_ports[*]}")
+
     unset IFS
 
     # add manatee metadata backup cron
