From 469ff683a0fa504eb1378be13cc6db021a42d64a Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Tue, 31 Jul 2018 21:43:16 +0000
Subject: [PATCH] OS-7100 Build fails in libvscan after merge

---
 usr/src/lib/libvscan/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/usr/src/lib/libvscan/Makefile b/usr/src/lib/libvscan/Makefile
index 1cf8f5effa..1212cc6a73 100644
--- a/usr/src/lib/libvscan/Makefile
+++ b/usr/src/lib/libvscan/Makefile
@@ -42,7 +42,7 @@ MSGFILES =	common/libvscan.c
 
 .KEEP_STATE:
 
-all clean clobber install lint _msg : $(SUBDIRS)
+all clean clobber install lint: $(SUBDIRS)
 
 $(POFILE):	$(MSGFILES)
 	$(BUILDPO.msgfiles)
-- 
2.21.0

