From a4850eff5d07e123b4557fe94d3dbc97ae2df22c Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Tue, 31 Jul 2018 21:43:16 +0000
Subject: [PATCH] OS-7100 Build fails in libvscan after merge

---
 usr/src/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/usr/src/Makefile b/usr/src/Makefile
index ce8962609a..734f46a128 100644
--- a/usr/src/Makefile
+++ b/usr/src/Makefile
@@ -115,7 +115,7 @@ all: mapfiles closedbins sgs .WAIT $(SUBDIRS) pkg
 # packaging to be pulled from $(SRC) and $(CLOSED) and staged in
 # $(ROOT)/licenses.
 #
-install: install1 install2 _msg
+install: install1 install2 .WAIT _msg
 	@cd msg; pwd; $(MAKE) _msg
 	@rm -rf "$(ROOT)/catalog"
 
-- 
2.21.0

