commit 683bd99cd7f4dcc7f3fc6cb1f155982101dcc1d7 (refs/changes/32/1332/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2017-01-25T01:00:34+00:00 (2 years, 8 months ago)
    
    MANTA-3116 disable pwritev in mako

diff --git a/auto/unix b/auto/unix
index 1316e54a..01242b1c 100755
--- a/auto/unix
+++ b/auto/unix
@@ -643,7 +643,14 @@ ngx_feature_test="char buf[1]; struct iovec vec[1]; ssize_t n;
                   vec[0].iov_len = 1;
                   n = pwritev(1, vec, 1, 0);
                   if (n == -1) return 1"
-. auto/feature
+#
+# Unfortuantely, there are bugs in the implementation of pwritev on
+# illumos when building a 32-bit version with LFS support. Effectively
+# we'll spuriously return EINVAL when writing at offsets between 2 GiB
+# and 4 GiB. When we have rolled out a fix for OS-5904 everywhere, then
+# we can remove this change.
+#
+# . auto/feature
 
 
 ngx_feature="sys_nerr"
