commit 6778ef6f2d3ebdb40d34b773688a651d302ce36c (refs/changes/51/4151/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2018-06-07T07:25:32-07:00 (1 year, 4 months ago)
    
    MANATEE-398 manatee-adm history -j doesn't understand v1 timestamps

diff --git a/bin/manatee-adm b/bin/manatee-adm
index 7e34493..13dbd22 100755
--- a/bin/manatee-adm
+++ b/bin/manatee-adm
@@ -719,7 +719,7 @@ ManateeAdm.prototype.do_history = function (subcmd, opts, args, cb) {
 
             if (evt.action) {
                 return ({
-                    'TIME': evt.date.toISOString(),
+                    'TIME': evt.time.toISOString(),
                     'G#': '-',
                     'MODE': '-',
                     'FRZ': '-',
