commit 5e98d3c162b5220204a7987004e23bf5673c4260 (refs/changes/16/1716/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-03-24T16:00:57-07:00 (2 years, 7 months ago)
    
    MANTA-3203 binder serving NOTIMP responses on empty services

diff --git a/lib/server.js b/lib/server.js
index 2d214e8..249796a 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -261,7 +261,12 @@ function resolve(options, query, cb) {
                                         query.setError('nxdomain');
                                         break;
                                 }
-                                // Inefficient, but easy to reason about.
+                                /*
+                                 * Make sure we set noerror here, otherwise we
+                                 * would respond with NOTIMP to a query about
+                                 * a service with no children.
+                                 */
+                                query.setError('noerror');
                                 var recs = record.children.filter(
                                     function (sub) {
                                         return (sub.type === 'load_balancer' ||
