From 92aa3d31cb881b45a572559395dbc27033f0f3b8 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 24 Mar 2017 15:58:41 -0700
Subject: [PATCH] MANTA-3203 binder serving NOTIMP responses on empty services

---
 lib/server.js | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/server.js b/lib/server.js
index 2d214e8..a003357 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -261,7 +261,12 @@ function resolve(options, query, cb) {
                                         query.setError('nxdomain');
                                         break;
                                 }
-                                // Inefficient, but easy to reason about.
+				/*
+				 * Make sure we set noerror here, otherwise we
+				 * would respond with NOTIMP to a query about
+				 * a service with no children.
+				 */
+				query.setError('noerror');
                                 var recs = record.children.filter(
                                     function (sub) {
                                         return (sub.type === 'load_balancer' ||
-- 
2.21.0

