commit 14d8c81464e738963d217a4f143f9d3e4bb2bf54 (refs/changes/63/4663/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-08-10T21:12:23+00:00 (1 year, 2 months ago)
    
    OS-7119 windows requires lpc in slot 31

diff --git a/usr/src/lib/brand/bhyve/zone/boot.c b/usr/src/lib/brand/bhyve/zone/boot.c
index e29616a748..c0e91d56e6 100644
--- a/usr/src/lib/brand/bhyve/zone/boot.c
+++ b/usr/src/lib/brand/bhyve/zone/boot.c
@@ -42,12 +42,12 @@
 
 typedef enum {
 	PCI_SLOT_HOSTBRIDGE = 0,	/* Not used here, but reserved */
-	PCI_SLOT_LPC,
-	PCI_SLOT_CD,
+	PCI_SLOT_CD = 2,		/* LPC was at 1; preserve ids */
 	PCI_SLOT_BOOT_DISK,
 	PCI_SLOT_OTHER_DISKS,
 	PCI_SLOT_NICS,
 	PCI_SLOT_FBUF = 30,
+	PCI_SLOT_LPC = 31,
 } pci_slot_t;
 
 static boolean_t debug;
