commit 11f7c8c57ecb77eda284b208ac0fc8f10622915d (refs/changes/46/5446/1)
Author: Rui Loura <rui@joyent.com>
Date:   2019-01-28T16:54:03+00:00 (8 months ago)
    
    TRITON-1163 cnapi should use agentIpFromSysinfo() for mockcloud support

diff --git a/lib/models/server.js b/lib/models/server.js
index 8ac9892..0d399a4 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1884,13 +1884,14 @@ function (opts) {
             });
         },
         function getCnapiLocationFromSysinfo(wfcb) {
-            if (sysinfo.hasOwnProperty('CN Agent IP')) {
-                // Allow sysinfo to include an IP that we'll use to connect to
-                // that might be different from the "Admin IP".
-                serverAdminIp = sysinfo['CN Agent IP'];
-            } else {
-                serverAdminIp = netconfig.adminIpFromSysinfo(sysinfo);
-            }
+            /*
+             * agentIpFromSysinfo() is intended for use with mockcloud which
+             * requires the use of a different IP for contacting agents on a
+             * mock CN.  If mockcloud support is not a requirement,
+             * adminIpFromSysinfo() should be used to get the admin IP for a
+             * CN.
+             */
+            serverAdminIp = netconfig.agentIpFromSysinfo(sysinfo);
             if (!serverAdminIp) {
                 wfcb(new VError('parsing server ip address in sendTaskReq '
                     + '(No admin NICs detected.)'));
diff --git a/package.json b/package.json
index 5cc1d16..88f84fb 100644
--- a/package.json
+++ b/package.json
@@ -26,7 +26,7 @@
     "sprintf": "0.1.5",
     "trace-event": "1.3.0",
     "triton-metrics": "1.0.0",
-    "triton-netconfig": "1.1.0",
+    "triton-netconfig": "git+https://github.com/joyent/node-triton-netconfig.git#TRITON-1094",
     "vasync": "1.6.4",
     "verror": "1.9.0",
     "wf-client": "0.2.1",
