From 5fc353d039ca32785a16616d761a6597e9d37422 Mon Sep 17 00:00:00 2001
From: Mark Brooks <mark.brooks@joyent.com>
Date: Tue, 18 Oct 2016 14:01:07 -0400
Subject: [PATCH] TOOLS-1572 fix minor typo in fabrics messaging

---
 lib/post-setup/fabrics.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/post-setup/fabrics.js b/lib/post-setup/fabrics.js
index a93e821..9c58721 100644
--- a/lib/post-setup/fabrics.js
+++ b/lib/post-setup/fabrics.js
@@ -675,7 +675,7 @@ Fabrics.prototype.updateSapi = function updateSapi(cb) {
 
     if (self.sdcadm.sdc.metadata.fabric_cfg &&
             jsprim.deepEqual(self.sdcadm.sdc.metadata.fabric_cfg, self.data)) {
-        self.progress('Exaclty the same fabric configuration already in SAPI');
+        self.progress('Exactly the same fabric configuration already in SAPI');
         return cb(null);
     }
     self.configChanged = true;
-- 
2.21.0

