commit 6830a77a008797cf609a0a311b96a7250fe6b93d (refs/changes/06/4606/1)
Author: Jared Morrow <jm@joyent.com>
Date:   2018-07-31T07:24:09-06:00 (1 year, 2 months ago)
    
    MANTA-1387 libmanta should log and ignore BucketVersionError

diff --git a/lib/moray.js b/lib/moray.js
index e89b7b6..2cadd37 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -403,7 +403,11 @@ function createBucket(opts, cb) {
     var client = opts.client;
 
     client.putBucket(bucket, opts.opts, function (err) {
-        if (err) {
+        if (err && VError.findCauseByName(err, 'BucketVersionError') !== null) {
+            opts.log.warn(err, 'bucket schema out of date');
+            opts.log.debug(opts.opts, 'Moray.createBucket done');
+            cb();
+        } else if (err) {
             err.bucket = opts.bucket;
             err.opts = opts.opts;
             cb(err);
