From 7a8bcd22488feca4e1943e7bad4807e10c56f5f7 Mon Sep 17 00:00:00 2001
From: Jared Morrow <jm@joyent.com>
Date: Tue, 31 Jul 2018 07:24:09 -0600
Subject: [PATCH] MANTA-1387 libmanta should log and ignore BucketVersionError

---
 lib/moray.js | 46 +++++++++++++++++++++++++++++++++++++++++++---
 package.json |  2 +-
 2 files changed, 44 insertions(+), 4 deletions(-)

diff --git a/lib/moray.js b/lib/moray.js
index e89b7b6..612ecb4 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -403,10 +403,50 @@ function createBucket(opts, cb) {
     var client = opts.client;
 
     client.putBucket(bucket, opts.opts, function (err) {
+        var realErrors = [];
+
+        /*
+         * Filters out errors we want to ignore that would otherwise
+         * cause startup to fail. This will still push real errors
+         * up to the callback.
+         */
+        function pushErrIfValid(tryError) {
+            function checkAndPushError(sErr) {
+                if (VError.findCauseByName(sErr, 'BucketVersionError')
+                    !== null) {
+                    opts.log.warn(sErr, 'ignoring bucket schema out of date');
+                } else {
+                    realErrors.push(sErr);
+                }
+            }
+
+            if (VError.hasCauseWithName(tryError, 'MultiError')) {
+                // Just pull this apart by hand b/c .errors() is not available
+                var errs = tryError.ase_errors;
+                errs.forEach(checkAndPushError);
+            } else {
+                checkAndPushError(tryError);
+            }
+        }
+
+        /*
+         * Entry point to the above filtering
+         * errorForEach() will handle both cases of a MultiError or a
+         * single error.
+         */
         if (err) {
-            err.bucket = opts.bucket;
-            err.opts = opts.opts;
-            cb(err);
+            VError.errorForEach(err, pushErrIfValid);
+        }
+
+        /*
+         * Collect any important errors we found and rebuild them into
+         * a MultiError
+         */
+        if (realErrors.length > 0) {
+            var err2 = new VError.errorFromList(realErrors);
+            err2.bucket = bucket;
+            err2.opts = opts.opts;
+            cb(err2);
         } else {
             opts.log.debug(opts.opts, 'Moray.createBucket done');
             cb();
diff --git a/package.json b/package.json
index 82e16e9..ad7051f 100644
--- a/package.json
+++ b/package.json
@@ -20,7 +20,7 @@
         "restify": "2.6.3",
         "redis": "0.10.1",
         "vasync": "1.6.3",
-        "verror": "1.6.0",
+        "verror": "^1.10.0",
         "xtend": "2.1.2"
     },
     "devDependencies": {
-- 
2.21.0

