commit 0f8e8f1ad3908008ffdb711589bd467949f279f6 (refs/changes/06/4606/2)
Author: Jared Morrow <jm@joyent.com>
Date:   2018-07-31T08:35:52-06:00 (1 year, 2 months ago)
    
    MANTA-1387 libmanta should log and ignore BucketVersionError

diff --git a/lib/moray.js b/lib/moray.js
index e89b7b6..2cadd37 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -403,7 +403,11 @@ function createBucket(opts, cb) {
     var client = opts.client;
 
     client.putBucket(bucket, opts.opts, function (err) {
-        if (err) {
+        if (err && VError.findCauseByName(err, 'BucketVersionError') !== null) {
+            opts.log.warn(err, 'bucket schema out of date');
+            opts.log.debug(opts.opts, 'Moray.createBucket done');
+            cb();
+        } else if (err) {
             err.bucket = opts.bucket;
             err.opts = opts.opts;
             cb(err);
diff --git a/package.json b/package.json
index 82e16e9..28805e6 100644
--- a/package.json
+++ b/package.json
@@ -20,7 +20,7 @@
         "restify": "2.6.3",
         "redis": "0.10.1",
         "vasync": "1.6.3",
-        "verror": "1.6.0",
+        "verror": "^1.8.1",
         "xtend": "2.1.2"
     },
     "devDependencies": {
