From 6830a77a008797cf609a0a311b96a7250fe6b93d Mon Sep 17 00:00:00 2001
From: Jared Morrow <jm@joyent.com>
Date: Tue, 31 Jul 2018 07:24:09 -0600
Subject: [PATCH] MANTA-1387 libmanta should log and ignore BucketVersionError

---
 lib/moray.js | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/moray.js b/lib/moray.js
index e89b7b6..2cadd37 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -403,7 +403,11 @@ function createBucket(opts, cb) {
     var client = opts.client;
 
     client.putBucket(bucket, opts.opts, function (err) {
-        if (err) {
+        if (err && VError.findCauseByName(err, 'BucketVersionError') !== null) {
+            opts.log.warn(err, 'bucket schema out of date');
+            opts.log.debug(opts.opts, 'Moray.createBucket done');
+            cb();
+        } else if (err) {
             err.bucket = opts.bucket;
             err.opts = opts.opts;
             cb(err);
-- 
2.21.0

