commit 6ca228169e9dc523911dfd42521f0dc9eff481ee (refs/changes/55/1155/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-21T11:33:14-08:00 (2 years, 10 months ago)
    
    MANTA-3064 binder not serving host TTLs properly

diff --git a/lib/server.js b/lib/server.js
index c517713..abe29a2 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -173,13 +173,15 @@ function resolve(options, query, cb) {
                 } else {
                         var addr;
 
-                        var ttl = (record[record.type] || {}).ttl;
                         /*
                          * Default the TTL to 30 seconds (the default ZK
                          * session timeout)
                          */
-                        if (ttl === undefined)
-                                ttl = 30;
+                        var ttl = 30;
+                        if (record.ttl !== undefined)
+                            ttl = record.ttl;
+                        if (record[record.type].ttl !== undefined)
+                            ttl = record[record.type].ttl;
 
                         if (service !== undefined &&
                             record.type !== 'service') {
@@ -214,6 +216,8 @@ function resolve(options, query, cb) {
 
                         case 'service':
                                 var s = record.service.service;
+                                if (s.ttl !== undefined)
+                                    ttl = s.ttl;
                                 if (service !== undefined &&
                                     (service !== s.srvce ||
                                     protocol !== s.proto)) {
@@ -257,7 +261,11 @@ function resolve(options, query, cb) {
                                             ports.length < 1)
                                                 ports = [s.port];
                                         var ar, sr, nm;
-                                        var rttl = host.ttl || 30;
+                                        var rttl = 30;
+                                        if (host.ttl !== undefined)
+                                            rttl = host.ttl;
+                                        if (host[host.type].ttl !== undefined)
+                                            rttl = host[host.type].ttl;
                                         if (service !== undefined) {
                                                 nm = host.name + '.' + domain;
                                                 ports.forEach(function (p) {
