commit c3a662d2ea016edd123e39dd7f94180cb525f2ec (refs/changes/05/1205/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-01-04T12:11:34-08:00 (2 years, 9 months ago)
    
    MANTA-3070 want cueball kang on muskie

diff --git a/main.js b/main.js
index b22766d..735124a 100644
--- a/main.js
+++ b/main.js
@@ -29,6 +29,7 @@ var vasync = require('vasync');
 var medusa = require('./lib/medusa');
 var keyapi = require('keyapi');
 var cueball = require('cueball');
+var kang = require('kang');
 
 var app = require('./lib');
 
@@ -205,9 +206,11 @@ function usage(parser, message)
     process.exit(2);
 }
 
-function createCueballHttpAgent(cfg, sharkCfg) {
+function createCueballHttpAgent(cfg) {
+    var sharkCfg = cfg.sharkConfig;
+
     /* Used for connections to mahi and other services. */
-    AGENT = new cueball.HttpAgent(cfg);
+    AGENT = new cueball.HttpAgent(cfg.cueballHttpAgent);
 
     /* Used only for connections to sharks. */
     var sharkCueball = {
@@ -233,6 +236,18 @@ function createCueballHttpAgent(cfg, sharkCfg) {
         }
     };
     SHARKAGENT = new cueball.HttpAgent(sharkCueball);
+
+    var kangOpts = cueball.poolMonitor.toKangOptions();
+    kangOpts.port = cfg.port + 800;
+    /*
+     * Note that we can't use kang.knStartServer here, as kang's restify
+     * version does not match ours and the two will clobber each other.
+     */
+    var kangServer = restify.createServer({ serverName: 'Kang' });
+    kangServer.get(new RegExp('.*'), kang.knRestifyHandler(kangOpts));
+    kangServer.listen(kangOpts.port, '127.0.0.1', function () {
+        LOG.info('cueball kang monitor started on port %d', kangOpts.port);
+    });
 }
 
 function createPickerClient(cfg) {
@@ -432,7 +447,7 @@ function version() {
 (function main() {
     var cfg = configure();
 
-    createCueballHttpAgent(cfg.cueballHttpAgent, cfg.sharkConfig);
+    createCueballHttpAgent(cfg);
     createMarlinClient(cfg.marlin);
     createPickerClient(cfg.storage);
     createAuthCacheClient(cfg.auth);
diff --git a/package.json b/package.json
index b9a9eae..3bdaebd 100644
--- a/package.json
+++ b/package.json
@@ -19,6 +19,7 @@
         "deep-equal": "0.0.0",
         "dtrace-provider": "0.2.8",
         "http-signature": "1.1.0",
+        "kang": "1.1.0",
         "keep-alive-agent": "0.0.1",
         "keyapi": "git+ssh://git@github.com:joyent/keyapi.git#c30dd2710ad2175095dc0e96479686fa774b8063",
         "libmanta": "git+ssh://git@github.com:joyent/node-libmanta.git#master",
