commit c490c4191309ef79a68d213684ce96ae8ceae21c (refs/changes/96/5496/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-06T18:29:13+00:00 (8 months ago)
    
    TOOLS-2174 agent-cache 'clone' should use '--' when switching branches

diff --git a/deps/eng b/deps/eng
index 4443fcd..7c472bc 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 4443fcd38fe1682556e1851df97c7d652d078f3d
+Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
