commit d5c0b5a238157cc087e6d0d90df791a98e9e708a (refs/changes/21/2921/3)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-11-08T19:25:18+01:00 (1 year, 11 months ago)
    
    SAPI-295 SAPI crashes if MASTER_MORAY_IP is a hostname

diff --git a/lib/server/stor/moray.js b/lib/server/stor/moray.js
index cb26743..69aa31b 100644
--- a/lib/server/stor/moray.js
+++ b/lib/server/stor/moray.js
@@ -111,13 +111,11 @@ MorayStorage.prototype.init = function init(cb) {
                 port: config.master_port
             });
         } else {
-           this.master = moray.createClient({
-                connectTimeout: config.connectTimeout,
+            assert.equal(2020, config.master_port, 'config.master_port');
+            this.master = moray.createClient({
                 log: log,
                 srvDomain: config.master_host,
-                cueballOptions: {
-                    defaultPort: config.master_port
-                }
+                connectTimeout: config.connectTimeout
             });
         }
 
