From d5c0b5a238157cc087e6d0d90df791a98e9e708a Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Wed, 8 Nov 2017 19:25:18 +0100
Subject: [PATCH] SAPI-295 SAPI crashes if MASTER_MORAY_IP is a hostname

---
 lib/server/stor/moray.js | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/lib/server/stor/moray.js b/lib/server/stor/moray.js
index cb26743..69aa31b 100644
--- a/lib/server/stor/moray.js
+++ b/lib/server/stor/moray.js
@@ -111,13 +111,11 @@ MorayStorage.prototype.init = function init(cb) {
                 port: config.master_port
             });
         } else {
-           this.master = moray.createClient({
-                connectTimeout: config.connectTimeout,
+            assert.equal(2020, config.master_port, 'config.master_port');
+            this.master = moray.createClient({
                 log: log,
                 srvDomain: config.master_host,
-                cueballOptions: {
-                    defaultPort: config.master_port
-                }
+                connectTimeout: config.connectTimeout
             });
         }
 
-- 
2.21.0

