commit 8c602ed8dae4a0f0bfa2ff8f99024ef0523b5b04 (refs/changes/21/2921/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-11-07T18:58:26+01:00 (1 year, 11 months ago)
    
    SAPI-295 SAPI crashes if MASTER_MORAY_IP is a hostname

diff --git a/lib/server/stor/moray.js b/lib/server/stor/moray.js
index cb26743..35a45b7 100644
--- a/lib/server/stor/moray.js
+++ b/lib/server/stor/moray.js
@@ -112,7 +112,6 @@ MorayStorage.prototype.init = function init(cb) {
             });
         } else {
            this.master = moray.createClient({
-                connectTimeout: config.connectTimeout,
                 log: log,
                 srvDomain: config.master_host,
                 cueballOptions: {
