From 596ab85d038d9fc58ac00e4e930d71210b12cea2 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Wed, 14 Mar 2018 20:35:29 +0000
Subject: [PATCH] MANTA-3600 binder could have more trace logs

---
 lib/zk.js | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/lib/zk.js b/lib/zk.js
index 5982c3b..52bef67 100644
--- a/lib/zk.js
+++ b/lib/zk.js
@@ -145,8 +145,19 @@ TreeNode.prototype.onDataChanged = function (zk, data, stat) {
         }
         this.tn_data = parsedData;
 
-        if (parsedData === null || typeof (parsedData.type) !== 'string')
+        if (parsedData === null || typeof (parsedData.type) !== 'string') {
+                this.tn_log.trace({
+                    path: this.tn_path,
+                    parsedData: parsedData,
+                    typeField: typeof (parsedData.type)
+                }, 'onDataChanged: unexpected data');
                 return;
+        }
+
+        this.tn_log.trace({
+            path: this.tn_path,
+            type: parsedData.type
+        }, 'onDataChanged');
 
         switch (parsedData.type) {
         case 'db_host':
@@ -172,6 +183,7 @@ TreeNode.prototype.onDataChanged = function (zk, data, stat) {
 };
 TreeNode.prototype.unbind = function () {
         var self = this;
+        this.tn_log.trace('unbinding node at "%s"', this.tn_path);
         if (this.tn_watcher) {
                 this.tn_watcher.removeAllListeners('childrenChanged');
                 this.tn_watcher.removeAllListeners('dataChanged');
-- 
2.21.0

