commit 15620afac5352e410440ae361bfd010631435e33 (refs/changes/71/271/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-08-11T10:31:32-07:00 (3 years, 2 months ago)
    
    CNS-157 bump reaper timeout

diff --git a/updater.js b/updater.js
index 74a6d01..610ab88 100644
--- a/updater.js
+++ b/updater.js
@@ -35,7 +35,7 @@ if (confPath === undefined)
 	confPath = path.join(__dirname, 'etc', 'config.json');
 var conf = config.parse(confPath);
 
-var CF_REAP_TIME = 3600;
+var CF_REAP_TIME = 3600 * 12;
 var FALLBACK_REAP_TIME = 40;
 
 var client = redis.createClient(conf.redis_opts);
