commit 356a61cf5136486b3573d10bcab3f5a2d21c4d25 (refs/changes/11/3611/2)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-03-15T17:05:46-04:00 (1 year, 7 months ago)
    
    TRITON-240 zoneevent should be called with an identifier string

diff --git a/agent/lib/zoneeventwatcher.js b/agent/lib/zoneeventwatcher.js
index 5649e1a..ceef1fe 100644
--- a/agent/lib/zoneeventwatcher.js
+++ b/agent/lib/zoneeventwatcher.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -45,7 +45,8 @@ util.inherits(ZoneEventWatcher, process.EventEmitter);
 ZoneEventWatcher.prototype.start = function () {
     var self = this;
     self.log.trace('spawn zoneevent');
-    var zoneevent = self.child = spawn('/usr/vm/sbin/zoneevent');
+    var zoneevent = self.child = spawn('/usr/vm/sbin/zoneevent',
+        ['-i', 'amon-agent']);
     self.stopped = false;
 
     zoneevent.stdout.setEncoding('utf8');
diff --git a/relay/lib/zoneeventwatcher.js b/relay/lib/zoneeventwatcher.js
index c65ee70..aa94e03 100644
--- a/relay/lib/zoneeventwatcher.js
+++ b/relay/lib/zoneeventwatcher.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -94,7 +94,7 @@ function ZoneEventWatcher(log) {
         process.exit(1);
     }
 
-    var zoneevent = spawn('/usr/vm/sbin/zoneevent');
+    var zoneevent = spawn('/usr/vm/sbin/zoneevent', ['-i', 'amon-relay']);
     zoneevent.stdout.setEncoding('utf8');
     var leftover = '';  // Left-over partial line from last chunk.
     zoneevent.stdout.on('data', function (chunk) {
diff --git a/relay/smf/manifests/amon-zoneevents.xml.in b/relay/smf/manifests/amon-zoneevents.xml.in
index 65910f0..9266151 100644
--- a/relay/smf/manifests/amon-zoneevents.xml.in
+++ b/relay/smf/manifests/amon-zoneevents.xml.in
@@ -7,7 +7,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <service_bundle type='manifest' name='export'>
@@ -20,7 +20,7 @@
     <dependency name='filesystem' grouping='require_all' restart_on='error' type='service'>
       <service_fmri value='svc:/system/filesystem/local'/>
     </dependency>
-    <exec_method name='start' type='method' exec='/usr/vm/sbin/zoneevent &amp;' timeout_seconds='10'>
+    <exec_method name='start' type='method' exec='/usr/vm/sbin/zoneevent -i amon-zoneevents &amp;' timeout_seconds='10'>
       <method_context>
         <method_credential user='root' group='staff'/>
       </method_context>
