From 4b37e6b181aa434a816a96d777616d561f8c74bc Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody@kkantor.com>
Date: Tue, 13 Nov 2018 20:37:32 +0000
Subject: [PATCH] MANTA-4018 manta-login tool should sort by shard number

---
 bin/manta-login | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/bin/manta-login b/bin/manta-login
index bfc2a1e..13806d8 100755
--- a/bin/manta-login
+++ b/bin/manta-login
@@ -53,8 +53,7 @@ elif [[ "$1" == "-h" || "$1" == "--help" ]]; then
 fi
 
 IFS=$'\n'
-results=($(manta-adm show -H -o service,shard,zonename,gz_admin_ip | \
-    grep $1 | sort))
+results=($(manta-adm show -H -o service,shard,zonename,gz_admin_ip | grep $1))
 if [[ ${#results[@]} -gt 1 ]]; then
     c=0
     if [[ -n $2 ]]; then
-- 
2.21.0

