From 83204908dac258f2d6503706ea67746209d8241a Mon Sep 17 00:00:00 2001
From: Orlando Vazquez <ovazquez@gmail.com>
Date: Thu, 6 Apr 2017 16:39:10 -0700
Subject: [PATCH] CNAPI-702 move test suite to tape and TAP output

---
 test/runtests | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/runtests b/test/runtests
index 17e2419..ff8ffdb 100755
--- a/test/runtests
+++ b/test/runtests
@@ -53,7 +53,7 @@ function fatal
     exit 1
 }
 
-opt_reporter=junit
+opt_reporter=tap
 
 function usage
 {
-- 
2.21.0

