commit 862c08e840f0411ec6ddde3bc20e2c5c6e1a5293 (refs/changes/63/1763/2)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2017-04-06T23:39:24+00:00 (2 years, 6 months ago)
    
    CNAPI-702 move test suite to tape and TAP output

diff --git a/test/runtests b/test/runtests
index 17e2419..ff8ffdb 100755
--- a/test/runtests
+++ b/test/runtests
@@ -53,7 +53,7 @@ function fatal
     exit 1
 }
 
-opt_reporter=junit
+opt_reporter=tap
 
 function usage
 {
