From f51309e03b0e54cc6471fc4bcc775ecd91552855 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Tue, 5 Feb 2019 17:43:08 +0000
Subject: [PATCH] TRITON-1178 sdcadm man page build expects node in $PATH

---
 CHANGES.md   | 6 +++++-
 Makefile     | 5 +++--
 package.json | 2 +-
 3 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index 3d5dfb9..2d01589 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -5,11 +5,15 @@
 -->
 
 <!--
-    Copyright (c) 2019 Joyent, Inc.
+    Copyright (c) 2019, Joyent, Inc.
 -->
 
 # sdcadm Changelog
 
+## 1.24.4
+
+- TRITON-1178 sdcadm man page build expects node in $PATH
+
 ## 1.24.3
 
 - TRITON-754 lib/sdcadm.js is not admin rack aware
diff --git a/Makefile b/Makefile
index 42372f9..dc0a58a 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright 2018 Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 #
@@ -118,7 +118,8 @@ check:: check-version
 .PHONY: man
 man: man/man1/sdcadm.1.ronn
 	rm -f man/man1/sdcadm.1
-	./node_modules/.bin/marked-man --input man/man1/sdcadm.1.ronn \
+	$(NODE_EXEC) ./node_modules/.bin/marked-man \
+		--input man/man1/sdcadm.1.ronn \
 		--date `git log -1 --pretty=format:%cd --date=short` \
 		--output man/man1/sdcadm.1
 	chmod 444 man/man1/sdcadm.1
diff --git a/package.json b/package.json
index 2608504..31bec76 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdcadm",
   "description": "Administer a SmartDataCenter (SDC) standup",
-  "version": "1.24.3",
+  "version": "1.24.4",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

