commit 27730f7b1d83c5cc7292d878a7daf7b1f0280689 (refs/changes/70/5470/1)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-03T12:44:14+00:00 (8 months ago)
    
    TRITON-1178 sdcadm man page build expects node in $PATH

diff --git a/Makefile b/Makefile
index 42372f9..dc0a58a 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright 2018 Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 #
@@ -118,7 +118,8 @@ check:: check-version
 .PHONY: man
 man: man/man1/sdcadm.1.ronn
 	rm -f man/man1/sdcadm.1
-	./node_modules/.bin/marked-man --input man/man1/sdcadm.1.ronn \
+	$(NODE_EXEC) ./node_modules/.bin/marked-man \
+		--input man/man1/sdcadm.1.ronn \
 		--date `git log -1 --pretty=format:%cd --date=short` \
 		--output man/man1/sdcadm.1
 	chmod 444 man/man1/sdcadm.1
