From 1b77458c5a13030f7516c90344e33431ec654223 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 26 Sep 2018 11:57:59 -0700
Subject: [PATCH] MANTA-3966 MAKO_HTTP_KEEPALIVE_TIMEOUT should be set by
 default

---
 cmd/manta-init.js | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/cmd/manta-init.js b/cmd/manta-init.js
index 651e90b..87355c8 100755
--- a/cmd/manta-init.js
+++ b/cmd/manta-init.js
@@ -575,6 +575,19 @@ async.waterfall([
 		extra.metadata['MANTA_TLS_INSECURE'] = '1';
 		extra.metadata['MUSKIE_MULTI_DC'] = false;
 
+		/*
+		 * Because of a series of unfortunate bugs, there was a flag
+		 * day between mako (storage) and muskie (webapi) related to
+		 * HTTP keepalives. It's unsafe to enable keepalives at the mako
+		 * end until the muskies are sufficiently up to date. As a
+		 * result, this SAPI metadata key has to be set to do so. For
+		 * new deploys, we assume the muskie image used will be new
+		 * enough and we can safely set it here.
+		 *
+		 * See also MANTA-3966, MANTA-3083, MANTA-3084
+		 */
+		extra.metadata['MAKO_HTTP_KEEPALIVE_TIMEOUT'] = 86400;
+
 		// This is filled in when marlin is deployed.
 		extra.metadata['SERVER_COMPUTE_ID_MAPPING'] = {};
 
-- 
2.21.0

