commit 1ffe32ae973e2b93816bbf031b1da3197dc2620c (refs/changes/15/2215/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-07-13T16:33:46-07:00 (2 years, 3 months ago)
    
    VOLAPI-57 validating network key in ListVolumes' predicate throws

diff --git a/lib/validation/volumes.js b/lib/validation/volumes.js
index 007aae7..89d253f 100644
--- a/lib/validation/volumes.js
+++ b/lib/validation/volumes.js
@@ -156,6 +156,17 @@ function validateVolumeUuid(volumeUuid) {
     return err;
 }
 
+function validateVolumeNetwork(networkUuid) {
+    var err;
+    var validNetworkUuid = validUuid(networkUuid);
+
+    if (!validNetworkUuid) {
+        err = new Error(networkUuid + ' is not a valid network UUID');
+    }
+
+    return err;
+}
+
 module.exports = {
     validateOwnerUuid: validateOwnerUuid,
     validateVolumeName: validateVolumeName,
