From 1ffe32ae973e2b93816bbf031b1da3197dc2620c Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Thu, 13 Jul 2017 16:33:46 -0700
Subject: [PATCH] VOLAPI-57 validating network key in ListVolumes' predicate
 throws

---
 lib/validation/volumes.js | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/lib/validation/volumes.js b/lib/validation/volumes.js
index 007aae7..89d253f 100644
--- a/lib/validation/volumes.js
+++ b/lib/validation/volumes.js
@@ -156,6 +156,17 @@ function validateVolumeUuid(volumeUuid) {
     return err;
 }
 
+function validateVolumeNetwork(networkUuid) {
+    var err;
+    var validNetworkUuid = validUuid(networkUuid);
+
+    if (!validNetworkUuid) {
+        err = new Error(networkUuid + ' is not a valid network UUID');
+    }
+
+    return err;
+}
+
 module.exports = {
     validateOwnerUuid: validateOwnerUuid,
     validateVolumeName: validateVolumeName,
-- 
2.21.0

