commit a2f3b06e5a2e6a24a9fc77392f368bdc8f851a39 (refs/changes/89/5589/1)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2019-02-14T12:23:01-08:00 (8 months ago)
    
    TRITON-1231 Fix failing server model test

diff --git a/test/model/test-model-server.js b/test/model/test-model-server.js
index b806b63..3d9225e 100644
--- a/test/model/test-model-server.js
+++ b/test/model/test-model-server.js
@@ -418,7 +418,7 @@ function testModifyServer(test) {
 }
 
 function testModifyServerWithEtag(test) {
-    test.expect(7);
+    test.expect(6);
 
     var uuid = uuids[0];
     var etag = 'etag-1234';
