From a525df06b279539a708aa125c6f71b29cba35f6a Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Tue, 14 May 2019 07:30:21 -0600
Subject: [PATCH] MANTA-4272 Boray does not pass postgres IP and port info to
 cueball

---
 src/main.rs | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/main.rs b/src/main.rs
index cb77b26..443772f 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -68,8 +68,8 @@ fn main() {
     let pg_config = PostgresConnectionConfig {
         user: Some(user.into()),
         password: None,
-        host: None,
-        port: None,
+        host: Some(pg_ip.to_string()),
+        port: Some(pg_port.into()),
         database: Some(pg_db.into()),
         application_name: Some(application_name.into())
     };
-- 
2.21.0

