commit ab301c67844f8c55e618e47136e078391a7ebc5b (refs/changes/34/3534/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-03-03T23:15:02+00:00 (1 year, 7 months ago)
    
    OS-6736 include git source information in crash dumps

diff --git a/Makefile b/Makefile
index 00098ebb..c05357aa 100644
--- a/Makefile
+++ b/Makefile
@@ -124,8 +124,9 @@ live: world manifest mancheck_conf boot sdcman $(TOOLS_TARGETS) $(MANCF_FILE)
 	@echo $(OVERLAY_MANIFESTS)
 	@echo $(SUBDIR_MANIFESTS)
 	mkdir -p ${ROOT}/log
-	./tools/build_live -m $(ROOT)/$(MANIFEST) -o $(ROOT)/output \
-	    $(OVERLAYS) $(ROOT)/proto $(ROOT)/man/man
+	ALTCTFCONVERT=$(ALTCTFCONVERT) ./tools/build_live \
+	    -m $(ROOT)/$(MANIFEST) -o $(ROOT)/output $(OVERLAYS) $(ROOT)/proto \
+	    $(ROOT)/man/man
 
 boot: $(BOOT_TARBALL)
 
diff --git a/overlay/generic/etc/system b/overlay/generic/etc/system
index 3951d945..62b96ef6 100644
--- a/overlay/generic/etc/system
+++ b/overlay/generic/etc/system
@@ -154,3 +154,10 @@ set sd:sd_io_time=10
 
 * Use hires tick to improve some scheduling latency issues
 set hires_tick=1
+
+*
+* The "buildstamp" module contains information about the git sources used to
+* build the operating system image.  Once loaded, the driver will refuse to
+* detach.  We load it early to ensure it is in a crash dump wherever possible.
+*
+forceload: drv/buildstamp
diff --git a/overlay/generic/kernel/drv/buildstamp.conf b/overlay/generic/kernel/drv/buildstamp.conf
new file mode 100644
index 00000000..0193b620
--- /dev/null
+++ b/overlay/generic/kernel/drv/buildstamp.conf
@@ -0,0 +1 @@
+name="buildstamp" parent="pseudo" instance=0;
diff --git a/overlay/generic/manifest b/overlay/generic/manifest
index 2b621fe5..8683f59c 100644
--- a/overlay/generic/manifest
+++ b/overlay/generic/manifest
@@ -58,6 +58,7 @@ f etc/zones/SUNWdefault.xml 0444 root bin
 f etc/zshrc 0644 root bin
 d kernel/drv 0755 root sys
 f kernel/drv/amd64/bnx 0755 root sys
+f kernel/drv/buildstamp.conf 0644 root sys
 f kernel/drv/mpt_sas.conf 0644 root sys
 f kernel/drv/sd.conf 0644 root sys
 d lib/sdc 0755 root bin
diff --git a/src/buildstamp.c b/src/buildstamp.c
new file mode 100644
index 00000000..3512fe7f
--- /dev/null
+++ b/src/buildstamp.c
@@ -0,0 +1,120 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+/*
+ * BUILDSTAMP KERNEL MODULE
+ *
+ * This kernel module carries the contents of the "boot_archive.gitstatus" file
+ * in such a way that it will be included in kernel crash dumps.  It will be
+ * forceloaded at boot, will refuse to detach, and has no other moving parts.
+ *
+ * The git information can be read back with "mdb", either on the live system
+ * or from a crash dump; e.g.,
+ *
+ *	# mdb -ke 'gitstatus_start/s'
+ *	gitstatus_start:
+ *	gitstatus_start:[
+ *		{
+ *			"repo": "smartos-live",
+ *			"branch": "master",
+ *			"commit_date": "489715200",
+ *			"rev": "fdf15f1d3549138b3d208a52186471fe24eab4b5",
+ *			"url": "https://github.com/joyent/smartos-live.git"
+ *		},
+ *		...
+ */
+
+#include <sys/ddi.h>
+#include <sys/sunddi.h>
+#include <sys/modctl.h>
+
+int buildstamp_no_detach = 1;
+
+static int
+buildstamp_attach(dev_info_t *dip, ddi_attach_cmd_t cmd)
+{
+	switch (cmd) {
+	case DDI_ATTACH:
+	case DDI_RESUME:
+		return (DDI_SUCCESS);
+
+	default:
+		return (DDI_FAILURE);
+	}
+}
+
+static int
+buildstamp_detach(dev_info_t *dip, ddi_detach_cmd_t cmd)
+{
+
+	switch (cmd) {
+	case DDI_DETACH:
+		if (buildstamp_no_detach) {
+			return (DDI_FAILURE);
+		}
+		/* FALLTHRU */
+	case DDI_SUSPEND:
+		return (DDI_SUCCESS);
+
+	default:
+		return (DDI_FAILURE);
+	}
+}
+
+static struct dev_ops buildstamp_dev_ops = {
+	.devo_rev =		DEVO_REV,
+	.devo_refcnt =		0,
+	.devo_getinfo =		nodev,
+	.devo_identify =	nulldev,
+	.devo_probe =		nulldev,
+
+	.devo_attach =		buildstamp_attach,
+	.devo_detach =		buildstamp_detach,
+
+	.devo_reset =		nodev,
+	.devo_cb_ops =		NULL,
+	.devo_bus_ops =		NULL,
+	.devo_power =		nodev,
+	.devo_quiesce =		ddi_quiesce_not_needed,
+};
+
+static struct modldrv buildstamp_md = {
+	.drv_modops =		&mod_driverops,
+	.drv_linkinfo =		"SmartOS buildstamp",
+	.drv_dev_ops =		&buildstamp_dev_ops,
+};
+
+static struct modlinkage buildstamp_ml = {
+	.ml_rev =		MODREV_1,
+	.ml_linkage =		{ &buildstamp_md, NULL }
+};
+
+int
+_init(void)
+{
+	return (mod_install(&buildstamp_ml));
+}
+
+int
+_info(struct modinfo *mi)
+{
+	return (mod_info(&buildstamp_ml, mi));
+}
+
+int
+_fini(void)
+{
+	return (mod_remove(&buildstamp_ml));
+}
diff --git a/tools/build_live b/tools/build_live
index 569cf012..53f737a5 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -11,7 +11,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -414,6 +414,88 @@ function bi_gen_etcrelease
 	bi_emit_done
 }
 
+function bi_gen_buildstamp_module
+{
+	local kernel_source="$bi_wsroot/projects/illumos"
+	local cflags=(
+		'-std=c99'
+		'-m64'
+		'-mcmodel=kernel'
+		'-gdwarf-2'
+		"-I$kernel_source/usr/src/uts/common"
+		"-I$kernel_source/usr/src/uts/intel"
+		"-I$kernel_source/usr/src/uts/i86pc"
+
+		'-nodefaultlibs'
+
+		'-fident'
+		'-fno-builtin'
+		'-fno-asm'
+		'-fno-inline'
+		'-fno-inline-functions'
+		'-ffreestanding'
+		'-mno-red-zone'
+
+		'-Wall'
+		'-Wextra'
+		'-Werror'
+		'-Wno-unknown-pragmas'
+		'-Wno-unused-parameter'
+
+		'-D__sun'
+		'-D_KERNEL'
+	)
+	local tmp_gitstatus="$bi_tmpdir/gitstatus"
+	local tmp_gitstatus_o="$bi_tmpdir/gitstatus.o"
+	local tmp_module_o="$bi_tmpdir/buildstamp.o"
+	local tmp_module="$bi_tmpdir/buildstamp"
+	local tmp_module_img="$bi_mnt_root/kernel/drv/amd64/buildstamp"
+
+	bi_emit_start 'Building "buildstamp" module'
+
+	#
+	# The elfwrap(1) program uses the basename of the file to generate the
+	# symbol names in the target object.  So that the git status
+	# information appears at "gitstatus_start" in the kernel, we make a
+	# copy of the gitstatus file with the desired name.
+	#
+	/usr/bin/rm -f "$tmp_gitstatus"
+	if ! /usr/bin/cp "$bi_archive.gitstatus" "$tmp_gitstatus"; then
+		fail "could not copy gitstatus file"
+	fi
+
+	if ! /usr/bin/elfwrap -64 -z target=x86 -o "$tmp_gitstatus_o" \
+	    "$tmp_gitstatus"; then
+		fail "could not elfwrap the gitstatus file"
+	fi
+
+	if ! gcc -c "${cflags[@]}" -o "$tmp_module_o" \
+	    "$bi_wsroot/src/buildstamp.c"; then
+		fail "could not compile buildstamp.c"
+	fi
+
+	if ! /usr/bin/ld -r -o "$tmp_module" "$tmp_module_o" \
+	    "$tmp_gitstatus_o"; then
+		fail "could not link buildstamp module"
+	fi
+
+	if ! $ALTCTFCONVERT "$tmp_module"; then
+		fail "could not add CTF to buildstamp module"
+	fi
+
+	if ! /usr/bin/strip -x "$tmp_module"; then
+		fail "could not strip buildstamp module"
+	fi
+
+	if ! pfexec /usr/bin/cp "$tmp_module" "$tmp_module_img" ||
+	    ! pfexec /usr/bin/chown 0:3 "$tmp_module_img" ||
+	    ! pfexec /usr/bin/chmod 0755 "$tmp_module_img"; then
+		fail "could not copy buildstamp module into image"
+	fi
+
+	bi_emit_done
+}
+
 function bi_ship_manifest
 {
 	local fl_manifest="$bi_archive.manifest"
@@ -836,6 +918,7 @@ bi_smf_import
 bi_smf_seeds
 bi_create_password
 bi_gen_etcrelease
+bi_gen_buildstamp_module
 bi_ship_manifest
 bi_fill_templates
 bi_file_fixups
