commit 6e068b953b44071a14236d13a1f7d799564fc6b9 (refs/changes/84/3984/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-05-21T20:07:18+00:00 (1 year, 5 months ago)
    
    MORAY-472 moray handler-error dtrace probe is not useful

diff --git a/lib/control.js b/lib/control.js
index 3f6cd02..f01adb5 100644
--- a/lib/control.js
+++ b/lib/control.js
@@ -161,7 +161,7 @@ function handlerPipeline(options) {
                 if (err) {
                     dtrace.fire('handler-error', function () {
                         return ([req.msgid, req.req_id, route, handler.name,
-                            err.toString]);
+                            err.toString()]);
                     });
                 }
                 dtrace.fire('handler-done', function () {
