commit 44550e1e6cab5dbb9d692d8b23980fbc31918d26 (refs/changes/68/5268/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2018-12-20T00:39:05+00:00 (10 months ago)
    
    MANTA-3779 Add 'singlePath' field to new object metadata

diff --git a/lib/moray_delete_record_reader.js b/lib/moray_delete_record_reader.js
index 96c643d..0f3c32c 100644
--- a/lib/moray_delete_record_reader.js
+++ b/lib/moray_delete_record_reader.js
@@ -309,7 +309,18 @@ state_running(S)
 		 */
 		var creator = record.value.creator || record.value.owner;
 
-		if (self._is_allowed_creator(creator)) {
+		/*
+		 * There are two ways objects may enter the accelerated
+		 * garbage-collection pipeline:
+		 *
+		 * 1. The objects are owned by a snaplink disabled account
+		 * 2. The objects have the `singlePath` property set to true
+		 *
+		 * (2) implies that the object has never been snaplinked, so it
+		 * is safe to get rid of its only metadata record.
+		 */
+		if (self._is_allowed_creator(creator) ||
+		    record.value.singlePath === true) {
 			self.mr_listener.emit('record', record);
 			record_keys.push(record.key);
 		}
