commit 99f56bf4762c1224533229938d6d3016aa29e150 (refs/changes/08/3208/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-01-16T20:39:34+01:00 (1 year, 9 months ago)
    
    HEAD-2387 Add "dns_domain" to sapi service metadata

diff --git a/scripts/build-payload.js b/scripts/build-payload.js
index bd49f910..5019cf4c 100755
--- a/scripts/build-payload.js
+++ b/scripts/build-payload.js
@@ -314,6 +314,7 @@ async.series([
         if (zone !== 'sapi') {
             return cb();
         }
+        obj.customer_metadata['dns_domain'] = config['dns_domain'];
 
         // Explicitly put this first sapi in proto mode.
         obj.customer_metadata['SAPI_PROTO_MODE'] = true;
diff --git a/scripts/sdc-init.js b/scripts/sdc-init.js
index d4bf00b9..28a28921 100755
--- a/scripts/sdc-init.js
+++ b/scripts/sdc-init.js
@@ -473,6 +473,9 @@ function filterServices(serviceList, cb) {
             }
             extras.metadata['assets-ip'] = self.config.assets_admin_ip;
             extras.metadata['user-script'] = data.toString();
+            if (service === 'sapi') {
+                extras.metadata['dns_domain'] = self.config.dns_domain;
+            }
 
             // There's no need to pass the service defintion to
             // getOrCreateService() below, so remove it.
