From 3f1246ff20666c9bef1b98f1a4bb3c27ef55fe56 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 27 Dec 2016 14:52:38 -0800
Subject: [PATCH] CNS-188 stop serving service records for fabric NAT zones

---
 lib/flag-filter.js | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/lib/flag-filter.js b/lib/flag-filter.js
index f3e11b4..07e6cbd 100644
--- a/lib/flag-filter.js
+++ b/lib/flag-filter.js
@@ -27,6 +27,13 @@ var svcTagParser = require('triton-tags/lib/cns-svc-tag');
 /*JSSTYLED*/
 var PTR_REGEX = /^[a-z0-9][a-z0-9\-]{0,62}(?:\.[a-z0-9][a-z0-9\-]{0,62})*$/i;
 
+/*
+ * If we see zones with smartdc_role values on this list, ignore them. This is
+ * particularly important for NAT zones, as we may have a lot of them and
+ * listing them in public DNS is silly.
+ */
+var SMARTDC_ROLE_BLACKLIST = ['nat'];
+
 /*
  * The FlagFilter transform stream takes in a VM object that has been annotated
  * with "owner" and "server" objects (by UfdsFilter, CnFilter and NetFilter),
@@ -102,7 +109,8 @@ FlagFilter.prototype._transform = function (vm, enc, cb) {
 			});
 		}
 		tag = vm.tags['smartdc_role'];
-		if (vm.owner.login === 'admin' && tag && tag.length > 0) {
+		if (vm.owner.login === 'admin' && tag && tag.length > 0 &&
+		    SMARTDC_ROLE_BLACKLIST.indexOf(tag) === -1) {
 			vm.services.push({
 				name: tag,
 				ports: []
-- 
2.21.0

