commit 2ee0aabe0c52ac9329f33dd3b2fee135b07b9437 (refs/changes/70/270/1)
Author: Dave Pacheco <dap@joyent.com>
Date:   2016-08-10T09:44:09-07:00 (3 years, 2 months ago)
    
    MANTA-2963 add support for 64-bit marlin image 16.1.0

diff --git a/agent/etc/agentconfig.coal.json b/agent/etc/agentconfig.coal.json
index 5f98e55..ed0decb 100644
--- a/agent/etc/agentconfig.coal.json
+++ b/agent/etc/agentconfig.coal.json
@@ -50,7 +50,7 @@
 		"zoneReservePercent": 12
 	},
 
-	"zoneDefaultImage": "13.3.6",
+	"zoneDefaultImage": "16.1.0",
 
 	"zoneDefaults": {
 		"max_physical_memory": 1024,
diff --git a/dev/test/live/tst.main.js b/dev/test/live/tst.main.js
index 5115070..5d2a4e3 100644
--- a/dev/test/live/tst.main.js
+++ b/dev/test/live/tst.main.js
@@ -247,8 +247,8 @@ var testcases = {
 	'job': {
 	    'phases': [ {
 		'type': 'map',
-		'image': '13.3.6',
-		'exec': 'grep 13.3.6 /etc/motd'
+		'image': '16.1.0',
+		'exec': 'grep 16.1.0 /etc/motd'
 	    } ]
 	},
 	'inputs': [ '/%user%/stor/obj1' ],
diff --git a/jobsupervisor/etc/config.coal.json b/jobsupervisor/etc/config.coal.json
index 6300b7a..7706ae9 100644
--- a/jobsupervisor/etc/config.coal.json
+++ b/jobsupervisor/etc/config.coal.json
@@ -68,6 +68,6 @@
 	},
 
 	"images": [
-		"13.1.0", "13.3.4", "13.3.5", "13.3.6"
+		"13.1.0", "13.3.4", "13.3.5", "13.3.6", "16.1.0"
 	]
 }
diff --git a/jobsupervisor/sapi_manifests/marlin/template b/jobsupervisor/sapi_manifests/marlin/template
index dd0db3d..e3c34c4 100644
--- a/jobsupervisor/sapi_manifests/marlin/template
+++ b/jobsupervisor/sapi_manifests/marlin/template
@@ -63,6 +63,6 @@
 	},
 
 	"images": [
-		"13.1.0", "13.3.4", "13.3.5", "13.3.6"
+		"13.1.0", "13.3.4", "13.3.5", "13.3.6", "16.1.0"
 	]
 }
