From 71a1213e068494a3c2f70cd4d1ccdef209585cf5 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 22 Feb 2018 16:11:27 +0000
Subject: [PATCH] OS-6660 need py27-sqlite3 to build OS-6416

---
 configure | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/configure b/configure
index 07e6a325..c75831ab 100755
--- a/configure
+++ b/configure
@@ -12,7 +12,7 @@
 #
 
 #
-# Copyright 2017 Joyent, Inc.
+# Copyright 2018 Joyent, Inc.
 #
 shopt -s xpg_echo
 unalias -a
@@ -90,6 +90,7 @@ function install_pkgin
 	pkglist="$pkglist flex libxslt openjdk7 nodejs"
 	pkglist="$pkglist p5-XML-Parser gettext python27 py27-expat"
 	pkglist="$pkglist coreutils gsed pkg_alternatives cdrtools"
+	pkglist="$pkglist py27-sqlite3"
 
 	for pkg in $pkglist; do
 		if ! pkg_info -qe $pkg; then
-- 
2.21.0

