commit f0983d3b5d00b381fb3a9bea1cb4024afe04ef0a (refs/changes/39/3439/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-02-22T17:27:47+01:00 (1 year, 7 months ago)
    
    OS-6660 need py27-sqlite3 to build OS-6416

diff --git a/configure b/configure
index 07e6a325..94d7e91e 100755
--- a/configure
+++ b/configure
@@ -90,6 +90,7 @@ function install_pkgin
 	pkglist="$pkglist flex libxslt openjdk7 nodejs"
 	pkglist="$pkglist p5-XML-Parser gettext python27 py27-expat"
 	pkglist="$pkglist coreutils gsed pkg_alternatives cdrtools"
+	pkglist="$pkglist py27-sqlite3"
 
 	for pkg in $pkglist; do
 		if ! pkg_info -qe $pkg; then
