From 69ae5d9e44056a953e33dcd307ce8882dd7da426 Mon Sep 17 00:00:00 2001
From: Robert Bogart <robert.bogart@joyent.com>
Date: Tue, 17 Apr 2018 17:57:05 +0000
Subject: [PATCH] MANTA-2909 muskie logs incorrect bytesTransferred for some
 failed, streaming uploads

---
 lib/obj.js | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/lib/obj.js b/lib/obj.js
index 59993b3..173fee2 100644
--- a/lib/obj.js
+++ b/lib/obj.js
@@ -448,11 +448,21 @@ function sharkStreams(req, res, next) {
         return;
     }
 
+    /*
+     * While in the process of streaming the object out to multiple sharks, if a
+     * failure is experienced on one stream, we will essentially treat it as an
+     * overall failure and abandon the process of streaming this object to all
+     * sharks involved.  Note that `next_err()' is wrapped in the `once()'
+     * method because we need only respond to a failure event once.
+     */
     var next_err = once(function _next_err(err) {
         req.log.debug({
             err: err
         }, 'abandoning request');
 
+        /* Record the number of bytes that we transferred. */
+        req._size = check.bytes;
+
         req.removeListener('end', onEnd);
         req.removeListener('error', next_err);
 
-- 
2.21.0

