commit 6b8a22a77a5877f29c84546219cf1a90687a94c9 (refs/changes/65/2765/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-10-11T00:05:38+00:00 (2 years ago)
    
    MANATEE-334 apply connection limit to "moray" role

diff --git a/lib/postgresMgr.js b/lib/postgresMgr.js
index be7f693..eac0cbc 100644
--- a/lib/postgresMgr.js
+++ b/lib/postgresMgr.js
@@ -152,6 +152,7 @@ var PRIMARY_CONNINFO_STR =
 var PG_STAT_REPLICATION =
     'select * from pg_stat_replication where application_name = \'%s\'';
 
+var PG_ROLES = 'alter role moray set pg_roles.rolconnlimit = %d';
 
 // --- Internal helpers
 
@@ -1065,6 +1066,9 @@ PostgresMgr.prototype._primary = function _primary(stdby, callback) {
         function _restart(_, cb) {
             self._restart(cb);
         },
+        function _setRolConnLimit(_, cb) {
+            self._setRolConnLimit(cb);
+        },
         function _snapshot(_, cb) {
             self._snapShotter.createSnapshot(String(Date.now()), cb);
         },
@@ -1736,6 +1740,41 @@ PostgresMgr.prototype._start = function _start(cb) {
 };
 
 
+PostgresMgr.prototype._setRolConnLimit = function (cb) {
+    var self = this;
+
+    vasync.waterfall([
+        function getMaxConnections(callback) {
+            var maxConnQuery = 'show max_connections';
+            self._queryDb(maxConnQuery, function (err, result) {
+                if (err) {
+                    callback(new verror.VError(err, 'unable to query ' +
+                            'max_connections'));
+                    return;
+                }
+                if (!result || !result.max_connections) {
+                    callback(new verror.VError('missing max_connections'));
+                    return;
+                }
+                callback(null, result.max_connections);
+            });
+        },
+        function setRolConnLimit(maxConns, callback) {
+            var alterRoleQuery = sprintf(PG_ROLES, maxConns);
+            self._queryDb(alterRoleQuery, function (err) {
+                if (err) {
+                    callback(err);
+                    return;
+                }
+                callback();
+            });
+        }
+    ], function (err) {
+        cb(err);
+    });
+};
+
+
 /**
  * Initializes the postgres data directory for a new DB. This can fail if the
  * db has already been initialized - this is okay, as startdb will fail if init
