commit a15a74f0c97f723413aa37012dca4ad0092cddc2 (refs/changes/66/3266/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-01-25T14:48:59-05:00 (1 year, 8 months ago)
    
    RELENG-612 prefer (git+)https urls

diff --git a/.gitmodules b/.gitmodules
index 6db5801..48ab5ca 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,3 +1,3 @@
 [submodule "deps/sdc-scripts"]
 	path = deps/sdc-scripts
-	url = git@github.com:joyent/sdc-scripts.git
+	url = https://github.com/joyent/sdc-scripts.git
