commit 20ef987b5cb422be5924f0025a2c42efa46254f0 (refs/changes/02/1002/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    fix signing of DELETE method requests

diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index 3a82037..d1b8142 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -275,14 +275,8 @@ CloudApi.prototype._request = function _request(opts, cb) {
 
     var method = (opts.method || 'GET').toLowerCase();
     assert.ok(['get', 'post', 'put', 'delete', 'head'].indexOf(method) >= 0,
-        'invalid method given');
-    switch (method) {
-    case 'delete':
-        method = 'del';
-        break;
-    default:
-        break;
-    }
+        'invalid HTTP method given');
+    var clientFnName = (method === 'delete' ? 'del' : method);
 
     if (self.roles && self.roles.length > 0) {
         if (opts.path.indexOf('?') !== -1) {
@@ -305,9 +299,9 @@ CloudApi.prototype._request = function _request(opts, cb) {
             headers: headers
         };
         if (opts.data)
-            self.client[method](reqOpts, opts.data, cb);
+            self.client[clientFnName](reqOpts, opts.data, cb);
         else
-            self.client[method](reqOpts, cb);
+            self.client[clientFnName](reqOpts, cb);
     });
 };
 
