From 67891717ff0775cc510c74f1e2bf1103443072ae Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Mon, 22 Apr 2019 15:05:58 -0700
Subject: [PATCH] OS-7762 vm runtest/runtests do not list new cores if running
 without a terminal

---
 src/vm/runtest | 28 ++++++++++++++++------------
 1 file changed, 16 insertions(+), 12 deletions(-)

diff --git a/src/vm/runtest b/src/vm/runtest
index 326fc38c..d856f985 100755
--- a/src/vm/runtest
+++ b/src/vm/runtest
@@ -205,11 +205,6 @@ if [[ ${TEST_EXIT_CODE} != 0 ]]; then
     tests="?"
 fi
 
-coredir=$(cd $(dirname $(coreadm $$ | awk '{gsub("%Z", "/", $2); print $2;}')) \
-    && pwd -P)
-cores=$(find "$coredir" -type f -newer /tmp/runtest.start.$$)
-rm /tmp/runtest.start.$$
-
 if [[ -t 0 ]]; then
     # We're on a terminal so output the summary
     echo "#"
@@ -224,15 +219,24 @@ if [[ -t 0 ]]; then
     if [[ ${fail} -gt 0 ]]; then
         echo -e "# \033[31mFAIL: ${fail} / ${tests}\033[39m"
     fi
-    if [[ -n "$cores" ]]; then
+    echo "#"
+fi
+
+coredir=$(cd $(dirname $(coreadm $$ | awk '{gsub("%Z", "/", $2); print $2;}')) \
+    && pwd -P)
+cores=$(find "$coredir" -type f -newer /tmp/runtest.start.$$)
+rm /tmp/runtest.start.$$
+
+if [[ -n "$cores" ]]; then
+    if [[ -t 0 ]]; then
         echo -e "# \033[31mNew core files\033[39m"
-        # Send a list of core files to stdout for testers and `runtests` to see
-        # and another in this run's log.
-        for core in $cores; do
-            echo "# new core: $core"
-        done | tee -a /tmp/test.output.$$
     fi
-    echo "#"
+
+    # Send a list of core files to stdout for testers and `runtests` to see
+    # and another in this run's log.
+    for core in $cores; do
+        echo "# new core: $core"
+    done | tee -a /tmp/test.output.$$
 fi
 
 if [[ ${fail} -gt 0 || -n "$cores" ]]; then
-- 
2.21.0

