commit 2ec72bf27c76884522c39d5f3d70c6a9f25aed4c (refs/changes/67/167/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2016-07-28T21:39:48+00:00 (3 years, 2 months ago)
    
    NAPI-360 NAPI needs a new /search/nics endpoint

diff --git a/lib/endpoints/nics.js b/lib/endpoints/nics.js
index 3124cc7..cfa92f3 100644
--- a/lib/endpoints/nics.js
+++ b/lib/endpoints/nics.js
@@ -153,6 +153,8 @@ function _checkNetAgentRequest(req, res, next) {
 function register(http, before) {
     http.get({ path: '/nics', name: 'ListNics' },
         before, listNics);
+    http.get({ path: '/search/nics', name: 'SearchNics' },
+        before, listNics);
     http.head({ path: '/nics', name: 'HeadNics' },
         before, listNics);
     http.post({ path: '/nics', name: 'CreateNic' },
