commit 682dd5885fd69c62fa704c401f58c7eb11fc18d8 (refs/changes/62/2962/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2017-11-16T17:48:52-08:00 (1 year, 11 months ago)
    
    joyent/manta-thoth#157 building in manta is broken

diff --git a/tools/build-release.sh b/tools/build-release.sh
index 2078e51..7549057 100755
--- a/tools/build-release.sh
+++ b/tools/build-release.sh
@@ -35,7 +35,7 @@ if [[ $uname == sunos ]]; then
 fi
 (
     cd opt/custom/thoth
-    npm install smartdc thoth
+    npm install smartdc manta-thoth
 )
 
 tar zcf "$tar" opt
