commit a08eb9690e24832ad92e0e07be55e5e52941d825 (refs/changes/26/5526/2)
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-02-24T19:44:18+00:00 (8 months ago)
    
    TRITON-1144 convert sdc-dockerlogger to engbld framework

diff --git a/.gitignore b/.gitignore
index 9414754..e609ba2 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,9 @@
-build
+/bits
+/build
+/make_stamps
 dockerlogger
 dockerlogger.smartos
+dockerlogger*.manifest
+dockerlogger*.md5sum
+dockerlogger*.sh
 node_modules
diff --git a/.gitmodules b/.gitmodules
new file mode 100644
index 0000000..06f62d4
--- /dev/null
+++ b/.gitmodules
@@ -0,0 +1,3 @@
+[submodule "deps/eng"]
+	path = deps/eng
+	url = git@github.com:joyent/eng.git
diff --git a/Makefile b/Makefile
index 2d7fb5c..ffcd862 100644
--- a/Makefile
+++ b/Makefile
@@ -5,9 +5,12 @@
 #
 
 #
-# Copyright (c) 2018, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
+
+NAME=dockerlogger
+
 ELFEDIT=/usr/bin/elfedit
 GOPATH=$(PWD)/vendor
 TARGETS=dockerlogger dockerlogger.smartos
@@ -17,14 +20,26 @@ _AWK := $(shell (which gawk >/dev/null && echo gawk) \
 	|| echo awk)
 BRANCH := $(shell git symbolic-ref HEAD | $(_AWK) -F/ '{print $$3}')
 
+ENGBLD_REQUIRE := $(shell git submodule update --init deps/eng)
+include ./deps/eng/tools/mk/Makefile.defs
+TOP ?= $(error Unable to access eng.git submodule Makefiles.)
+
 DESTDIR ?= .
 
+GOROOT ?= /root/opt/go
+
 ifeq ($(shell uname -s),Darwin)
 	TARGET = dockerlogger
 else
 	TARGET = dockerlogger.smartos
 endif
 
+CLEAN_FILES = $(TARGETS) \
+	      $(DESTDIR)/dockerlogger-*.manifest \
+	      $(DESTDIR)/dockerlogger-*.md5sum \
+	      $(DESTDIR)/dockerlogger-*.sh \
+	      build
+
 .PHONY: all
 all: $(TARGET)
 
@@ -32,19 +47,23 @@ dockerlogger.smartos: dockerlogger
 	/usr/bin/elfedit -e 'ehdr:ei_osabi ELFOSABI_SOLARIS' dockerlogger dockerlogger.smartos
 
 dockerlogger: dockerlogger.go
-	GOPATH=$(GOPATH) go build $<
+	GOPATH=$(GOPATH) $(GOROOT)/bin/go build $<
 
 .PHONY: fmt
 fmt:
-	gofmt -w dockerlogger.go
+	$(GOROOT)/bin/gofmt -w dockerlogger.go
 
 .PHONY: pkg
 pkg: $(DESTDIR) dockerlogger.smartos
 	./tools/mk-shar -b "$(BRANCH)" -o $(DESTDIR)
 
-.PHONY: check
-check:
-	@echo "No 'make check' here."
+.PHONY: publish
+publish: pkg
+	mkdir -p $(ENGBLD_BITS_DIR)/$(NAME)
+	cp $(DESTDIR)/dockerlogger-*.manifest \
+	    $(DESTDIR)/dockerlogger-*.md5sum \
+	    $(DESTDIR)/dockerlogger-*.sh \
+	    $(ENGBLD_BITS_DIR)/$(NAME)
 
 node_modules/tape/bin/tape:
 	@npm install
@@ -53,13 +72,7 @@ node_modules/tape/bin/tape:
 test: $(TARGET) node_modules/tape/bin/tape
 	@ls -1 ./tests/test_*.js | xargs -L 1 node
 
-.PHONY: clean
-clean:
-	rm -f $(TARGETS)
-	rm -f $(DESTDIR)/dockerlogger-*.manifest
-	rm -f $(DESTDIR)/dockerlogger-*.md5sum
-	rm -f $(DESTDIR)/dockerlogger-*.sh
-	rm -rf build
-
 $(DESTDIR):
 	mkdir -p $@
+
+include ./deps/eng/tools/mk/Makefile.targ
diff --git a/deps/eng b/deps/eng
new file mode 160000
index 0000000..7c472bc
--- /dev/null
+++ b/deps/eng
@@ -0,0 +1 @@
+Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
diff --git a/package.json b/package.json
index 61f6d6e..319c372 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "dockerlogger",
     "description": "Dockerlogger",
-    "version": "1.0.0",
+    "version": "1.0.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
