commit 19c9cfc3f5b4c3cfb69d2f0e584d80f21499abcf (refs/changes/56/756/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-10-20T15:50:45-07:00 (3 years ago)
    
    TOOLS-1583 typo in 'sdcadm platform set-default' handler

diff --git a/lib/platform.js b/lib/platform.js
index a4298e7..74342ee 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -2280,7 +2280,7 @@ function do_set_default(subcmd, opts, args, cb) {
             ' option is not given'));
     }
 
-    if (opts.plaform === 'help') {
+    if (opts.platform === 'help') {
         cb(new errors.UsageError(
         'Please use `sdcadm platform help set-default` instead'));
     }
