commit 01ec58cddb7058900a32f0d83ff270ca53a14b2c
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2019-05-21T07:39:15-05:00 (5 months ago)
    
    TRITON-1685 linux-prepare-image should clear /etc/machine-id

diff --git a/tools/prepare-image/linux-prepare-image b/tools/prepare-image/linux-prepare-image
index 4c5da8a..8346fe7 100755
--- a/tools/prepare-image/linux-prepare-image
+++ b/tools/prepare-image/linux-prepare-image
@@ -129,6 +129,10 @@ function cleanup_hostname() {
     touch /etc/hostname
 }
 
+function cleanup_machineid() {
+    [[ -f /etc/machine-id ]] && cp /dev/null /etc/machine-id
+}
+
 function cleanup_network_devices() {
     if [[ -f /etc/udev/rules.d/70-persistent-net.rules ]] ; then
         rm -f /etc/udev/rules.d/70-persistent-net.rules
@@ -258,6 +262,7 @@ cleanup_ssh
 cleanup_root
 cleanup_metadata
 cleanup_hostname
+cleanup_machineid
 cleanup_network_devices
 cleanup_cloud_init
 
