From f3122299b1730764e6371c0bed71ea270d315767 Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Mon, 20 May 2019 19:43:05 -0500
Subject: [PATCH] TRITON-1685 linux-prepare-image should clear /etc/machine-id

---
 tools/prepare-image/linux-prepare-image | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/tools/prepare-image/linux-prepare-image b/tools/prepare-image/linux-prepare-image
index 4c5da8a..597134c 100755
--- a/tools/prepare-image/linux-prepare-image
+++ b/tools/prepare-image/linux-prepare-image
@@ -129,6 +129,12 @@ function cleanup_hostname() {
     touch /etc/hostname
 }
 
+function cleanup_machineid() {
+    if [[ -f /etc/machine-id ]]; then
+        cp /dev/null /etc/machine-id
+    fi
+}
+
 function cleanup_network_devices() {
     if [[ -f /etc/udev/rules.d/70-persistent-net.rules ]] ; then
         rm -f /etc/udev/rules.d/70-persistent-net.rules
@@ -258,6 +264,7 @@ cleanup_ssh
 cleanup_root
 cleanup_metadata
 cleanup_hostname
+cleanup_machineid
 cleanup_network_devices
 cleanup_cloud_init
 
-- 
2.21.0

