From 71ec89b1c25975de5933203c8677427efce59164 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Fri, 23 Nov 2018 10:37:06 -0800
Subject: [PATCH] MANTA-4033 get lagging Manta components on the latest v0.10.x

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fe9a794..9758edd 100644
--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,7 @@ JSSTYLE_FLAGS    = -f tools/jsstyle.conf
 # Variables
 #
 NAME                  = propeller
-NODE_PREBUILT_VERSION = v0.10.30
+NODE_PREBUILT_VERSION = v0.10.48
 NODE_PREBUILT_TAG     = zone
 NODE_PREBUILT_IMAGE   = fd2cc906-8938-11e3-beab-4359c665ac99
 
-- 
2.21.0

