commit 54555f31be791766ca689a961df40acbfdd4ab38 (refs/changes/08/608/3)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2016-10-06T19:15:20+00:00 (3 years ago)
    
    joyent/node-ip6addr#1 Want a simpler Addr.clone()

diff --git a/ip6addr.js b/ip6addr.js
index 3098909..a63808c 100644
--- a/ip6addr.js
+++ b/ip6addr.js
@@ -9,7 +9,6 @@
  */
 
 var assert = require('assert-plus');
-var clone = require('clone');
 var util = require('util');
 
 
@@ -263,8 +262,10 @@ Addr.prototype.toLong = function toLong() {
 
 Addr.prototype.clone = function cloneAddr() {
   var out = new Addr();
-  out._fields = clone(this._fields);
-  out._attrs = clone(this._attrs);
+  out._fields = this._fields.slice();
+  for (var k in this._attrs) {
+      out._attrs[k] = this._attrs[k];
+  }
   return out;
 };
 
diff --git a/package.json b/package.json
index 37520c3..433ad7f 100644
--- a/package.json
+++ b/package.json
@@ -9,7 +9,6 @@
   },
   "main": "ip6addr.js",
   "dependencies": {
-    "clone": "1.0.0",
     "assert-plus": "0.1.5"
   },
   "devDependencies": {
diff --git a/test/clone.test.js b/test/clone.test.js
new file mode 100644
index 0000000..5b57fdb
--- /dev/null
+++ b/test/clone.test.js
@@ -0,0 +1,50 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2016, Joyent, Inc.
+ */
+
+var test = require('tape').test;
+
+
+///--- Globals
+
+var lib;
+
+
+///--- Tests
+
+test('load library', function (t) {
+  lib = require('../ip6addr.js');
+  t.ok(lib);
+  t.end();
+});
+
+test('clone() produces an equivalent address', function (t) {
+  var addr1 = lib.parse('1.2.3.4');
+  var addr2 = addr1.clone();
+
+  t.deepEqual(addr1, addr2, 'IPv4: addr1 == addr2');
+  t.deepEqual(addr1.toString(), addr2.toString(),
+    'IPv4: addr1.toString() == addr2.toString()');
+
+  addr1 = lib.parse('::ffff:1.2.3.4');
+  addr2 = addr1.clone();
+
+  t.deepEqual(addr1, addr2, 'IPv4 Mapped: addr1 == addr2');
+  t.deepEqual(addr1.toString(), addr2.toString(),
+    'IPv4 Mapped: addr1.toString() == addr2.toString()');
+
+  addr1 = lib.parse('fd00::40e');
+  addr2 = addr1.clone();
+
+  t.deepEqual(addr1, addr2, 'IPv6: addr1 == addr2');
+  t.deepEqual(addr1.toString(), addr2.toString(),
+    'IPv6: addr1.toString() == addr2.toString()');
+
+  t.end();
+});
