From c84a83ec35b20fe048c26f4b50b7436d35ef1ffe Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Fri, 23 Mar 2018 19:23:28 +0000
Subject: [PATCH] OS-6820 bhyve provisioning times out after OS-6720

---
 usr/src/cmd/bhyve/bhyverun.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/usr/src/cmd/bhyve/bhyverun.c b/usr/src/cmd/bhyve/bhyverun.c
index 56240356ae..ca7983edd2 100644
--- a/usr/src/cmd/bhyve/bhyverun.c
+++ b/usr/src/cmd/bhyve/bhyverun.c
@@ -1119,17 +1119,16 @@ main(int argc, char *argv[])
 	 */
 	fbsdrun_addcpu(ctx, BSP, BSP, rip);
 
-	/*
-	 * Head off to the main event dispatch loop
-	 */
-	mevent_dispatch();
-#ifndef __FreeBSD
+#ifndef __FreeBSD__
 	if (vm_started_cb != NULL) {
 		vm_started_cb();
 	}
 #endif
 
-	pthread_exit(NULL);
+	/*
+	 * Head off to the main event dispatch loop
+	 */
+	mevent_dispatch();
 
 	exit(1);
 }
-- 
2.21.0

