commit 92468a45a788494c00a649fb9cc0ac392e6c2f17 (refs/changes/76/3876/1)
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2018-04-26T15:16:46+02:00 (1 year, 5 months ago)
    
    OS-6907 vmm_stat initialization not run, causing 0-length allocation

diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index b4ce5b93b6..8aea5a2f10 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -1984,6 +1984,7 @@ _init(void)
 	}
 
 	vmm_zsd_init();
+	vmm_stat_sysinit();
 
 	error = mod_install(&modlinkage);
 	if (error) {
diff --git a/usr/src/uts/i86pc/io/vmm/vmm_stat.c b/usr/src/uts/i86pc/io/vmm/vmm_stat.c
index c2c2cfe77c..bc3cc20318 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_stat.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_stat.c
@@ -170,3 +170,31 @@ VMM_STAT(VMEXIT_REQIDLE, "number of times idle requested at exit");
 VMM_STAT(VMEXIT_USERSPACE, "number of vm exits handled in userspace");
 VMM_STAT(VMEXIT_RENDEZVOUS, "number of times rendezvous pending at exit");
 VMM_STAT(VMEXIT_EXCEPTION, "number of vm exits due to exceptions");
+
+#ifndef __FreeBSD__
+void
+vmm_stat_sysinit(void)
+{
+	vmm_stat_register(VCPU_MIGRATIONS);
+	vmm_stat_register(VMEXIT_COUNT);
+	vmm_stat_register(VMEXIT_EXTINT);
+	vmm_stat_register(VMEXIT_HLT);
+	vmm_stat_register(VMEXIT_CR_ACCESS);
+	vmm_stat_register(VMEXIT_RDMSR);
+	vmm_stat_register(VMEXIT_WRMSR);
+	vmm_stat_register(VMEXIT_MTRAP);
+	vmm_stat_register(VMEXIT_PAUSE);
+	vmm_stat_register(VMEXIT_INTR_WINDOW);
+	vmm_stat_register(VMEXIT_NMI_WINDOW);
+	vmm_stat_register(VMEXIT_INOUT);
+	vmm_stat_register(VMEXIT_CPUID);
+	vmm_stat_register(VMEXIT_NESTED_FAULT);
+	vmm_stat_register(VMEXIT_INST_EMUL);
+	vmm_stat_register(VMEXIT_UNKNOWN);
+	vmm_stat_register(VMEXIT_ASTPENDING);
+	vmm_stat_register(VMEXIT_REQIDLE);
+	vmm_stat_register(VMEXIT_USERSPACE);
+	vmm_stat_register(VMEXIT_RENDEZVOUS);
+	vmm_stat_register(VMEXIT_EXCEPTION);
+}
+#endif
diff --git a/usr/src/uts/i86pc/io/vmm/vmm_stat.h b/usr/src/uts/i86pc/io/vmm/vmm_stat.h
index 0ea1b49280..a17bd5b06f 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_stat.h
+++ b/usr/src/uts/i86pc/io/vmm/vmm_stat.h
@@ -37,6 +37,10 @@
 #ifndef _VMM_STAT_H_
 #define	_VMM_STAT_H_
 
+#ifndef __FreeBSD__
+#define	VMM_KEEP_STATS
+#endif
+
 struct vm;
 
 #define	MAX_VMM_STAT_ELEMS	64		/* arbitrary */
@@ -59,6 +63,10 @@ struct vmm_stat_type {
 	enum vmm_stat_scope scope;
 };
 
+#ifndef __FreeBSD__
+void	vmm_stat_sysinit(void);
+#endif
+
 void	vmm_stat_register(void *arg);
 
 #define	VMM_STAT_FDEFINE(type, nelems, desc, func, scope)		\
