commit 372d808c3073fe0ef04eaab2c9621d6b0b9e6610 (refs/changes/40/4140/1)
Author: Kody A Kantor <kody@kkantor.com>
Date:   2018-06-05T14:55:51+00:00 (1 year, 4 months ago)
    
    OS-1666 Increase ipdadm packet delay limit to 1s

diff --git a/usr/src/uts/common/inet/ipd/ipd.c b/usr/src/uts/common/inet/ipd/ipd.c
index 553c215984..747a7a814f 100644
--- a/usr/src/uts/common/inet/ipd/ipd.c
+++ b/usr/src/uts/common/inet/ipd/ipd.c
@@ -9,7 +9,7 @@
  * http://www.illumos.org/license/CDDL.
  */
 /*
- * Copyright (c) 2012, Joyent, Inc. All rights reserved.
+ * Copyright (c) 2018, Joyent, Inc. All rights reserved.
  */
 
 /*
@@ -222,7 +222,7 @@ typedef struct ipd_netstack {
 	net_handle_t	ipdn_v6hdl;		/* IPv4 net handle */
 	int		ipdn_hooked;		/* are hooks registered */
 	hook_t		*ipdn_v4in;		/* IPv4 traffic in hook */
-	hook_t		*ipdn_v4out;		/* IPv4 traffice out hook */
+	hook_t		*ipdn_v4out;		/* IPv4 traffic out hook */
 	hook_t		*ipdn_v6in;		/* IPv6 traffic in hook */
 	hook_t		*ipdn_v6out;		/* IPv6 traffic out hook */
 	int		ipdn_enabled;		/* which perturbs are on */
@@ -613,7 +613,7 @@ ipd_toggle_delay(ipd_netstack_t *ins, uint32_t delay)
 	/*
 	 * If ipd_check_hooks_failed, that must mean that we failed to set up
 	 * the hooks, so we are going to effectively zero out and fail the
-	 * request to enable corruption.
+	 * request to enable packet delays.
 	 */
 	if (rval != 0)
 		ins->ipdn_delay = 0;
diff --git a/usr/src/uts/common/sys/ipd.h b/usr/src/uts/common/sys/ipd.h
index bad74f8b81..f21c3fb5af 100644
--- a/usr/src/uts/common/sys/ipd.h
+++ b/usr/src/uts/common/sys/ipd.h
@@ -20,7 +20,7 @@
  */
 
 /*
- * Copyright (c) 2012, Joyent, Inc. All rights reserved.
+ * Copyright (c) 2018, Joyent, Inc. All rights reserved.
  */
 
 /*
@@ -35,7 +35,7 @@ extern "C" {
 #endif
 
 #define	IPD_DEV_PATH	"/dev/ipd"
-#define	IPD_MAX_DELAY	10000		/* 10 ms in us */
+#define	IPD_MAX_DELAY	1000000		/* 1 second in microseconds */
 
 typedef struct ipd_ioc_perturb {
 	zoneid_t	ipip_zoneid;
