commit e10f469cd879b7a323b6dcd76c8de9bb8a187e0c
Author: sjorge <sjorge@blackdot.be>
Date:   2019-09-27T11:19:34+02:00 (11 days ago)
    
    OS-XXXX fix small spelling mistake in libhooks.ksh

diff --git a/usr/src/lib/brand/jcommon/libhooks.ksh b/usr/src/lib/brand/jcommon/libhooks.ksh
index 346356519c..ec49fb9e3d 100644
--- a/usr/src/lib/brand/jcommon/libhooks.ksh
+++ b/usr/src/lib/brand/jcommon/libhooks.ksh
@@ -16,7 +16,7 @@
 #
 
 #
-# This file contains various hooks that are used by nore than a single
+# This file contains various hooks that are used by more than a single
 # brand. This file should be included by the brand-specific files.
 #
 
