From a69940a426f7d9ce0392276f96db03e74f6bcff5 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Wed, 20 Jun 2018 23:34:32 +0000
Subject: [PATCH] OS-7042 illumos-extra should support building optional, extra
 gcc versions

---
 configure | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/configure b/configure
index 777b9bb9..9ca60535 100755
--- a/configure
+++ b/configure
@@ -289,8 +289,8 @@ BUILD_TOOLS="\${CODEMGR_WS}/usr/src/tools/proto/root_\${MACH}-nd/opt";
 						export BUILD_TOOLS
 SPRO_ROOT=/opt/SUNWspro;			export SPRO_ROOT
 SPRO_VROOT=\$SPRO_ROOT;				export SPRO_VROOT
-GNU_ROOT="$conf_root/proto.strap/usr/gnu"		export GNU_ROOT
-GCC_ROOT="$conf_root/proto.strap/usr";		export GCC_ROOT
+GNU_ROOT="$conf_root/proto.strap/usr/gnu"	export GNU_ROOT
+GCC_ROOT="$conf_root/proto.strap/usr/gcc/4";	export GCC_ROOT
 # A temporary hack to allow bootstrap of cw(1)
 CW_GCC_DIR="\${GCC_ROOT}/bin";			export CW_GCC_DIR
 # Use GCC as the primary compiler
-- 
2.21.0

