commit 62659912b84a16244ef74d7217dafd2366646eed (refs/changes/41/1441/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-02-06T09:36:30-08:00 (2 years, 8 months ago)
    
    joyent/sdc-docker#110 Docker ps filtering is not correct

diff --git a/lib/endpoints/containers.js b/lib/endpoints/containers.js
index 43206cd..dfc2bfa 100644
--- a/lib/endpoints/containers.js
+++ b/lib/endpoints/containers.js
@@ -86,15 +86,7 @@ function containerList(req, res, next) {
     var options = {};
 
     options.all = common.boolFromQueryParam(req.query.all);
-    // Note: options.all is implied when using these query params.
-    if (req.query.hasOwnProperty('limit')
-        || req.query.hasOwnProperty('before')
-        || req.query.hasOwnProperty('since'))
-    {
-        options.all = true;
-    }
-
-    options.limit = parseInt(req.query.limit || '0', 10);
+    options.limit = parseInt(req.query.limit || '-1', 10);
     options.before = req.query.before;
     options.since = req.query.since;
     options.size = req.query.size;
@@ -106,6 +98,11 @@ function containerList(req, res, next) {
     options.images = req.images;
     options.clientApiVersion = req.clientApiVersion;
 
+    // Note: options.all is implied when using these query params.
+    if (options.limit > 0 || options.before || options.since) {
+        options.all = true;
+    }
+
     req.backend.getContainers(options, function (err, containers) {
         if (err) {
             if (!(err instanceof errors.DockerError)) {
diff --git a/test/integration/cli-filters.test.js b/test/integration/cli-filters.test.js
index 2702035..27c0377 100644
--- a/test/integration/cli-filters.test.js
+++ b/test/integration/cli-filters.test.js
@@ -129,6 +129,16 @@ test('container filters', function (tt) {
     checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX
         + ' --filter status=running',
         [containerName2]);
+
+    // Filtering using limit - should match running/stopped containers.
+    checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX
+        + ' --last 10',
+        [containerName1, containerName2]);
+
+    // Filtering using negative limit - should only match running containers.
+    checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX
+        + ' --last -10',
+        [containerName2]);
 });
 
 
