commit eda49461587dd1d11dbb76b68f9275d537671879 (refs/changes/14/2314/3)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-11-24T14:11:35-08:00 (1 year, 10 months ago)
    
    DOCKER-524 Implement docker push

diff --git a/CHANGES.md b/CHANGES.md
index 50c277b..8be18fe 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,10 @@
 
 ## not yet released
 
+## 10.5.0
+
+- DOCKER-524 Implement docker push
+
 ## 10.4.1
 
 - DOCKER-1102 docker pull by manifest digest fails with an InternalError
diff --git a/lib/imgapi.js b/lib/imgapi.js
index bf9cead..0431eed 100644
--- a/lib/imgapi.js
+++ b/lib/imgapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -1189,6 +1189,82 @@ IMGAPI.prototype.adminImportDockerImage = function adminImportDockerImage(
 };
 
 
+/**
+ * Push a docker image (which can have multiple layers) to a remote registry.
+ * (operator/admin use only)
+ *
+ * @param {Object} options.
+ *      - image {Object} Required: the sdc-docker ImageV2 model instance.
+ *      - repoAndTag {String} Required. The docker repo, e.g. 'busybox',
+ *        'trentm/busybox:latest', 'myreg.example.com:5000/blah:master'.
+ *      - regAuth {String} Optional. Registry auth info formatted as in
+ *        the 'x-registry-auth' header in `docker` client requests.
+ *      - headers {Object} Optional Additional request headers.
+ * @param {Function} callback : `function (err, res)`
+ */
+IMGAPI.prototype.adminPushDockerImage =
+function adminPushDockerImage(options, callback) {
+    var self = this;
+    if (callback === undefined) {
+        callback = options;
+        options = {};
+    }
+    assert.object(options, 'options');
+    assert.optionalObject(options.headers, 'options.headers');
+    assert.object(options.image, 'options.image');
+    assert.optionalString(options.regAuth, 'options.regAuth');
+    assert.string(options.repoAndTag, 'options.repoAndTag');
+    assert.func(callback, 'callback');
+
+    var path = self._path(format('/images/%s/push', options.image.image_uuid), {
+        channel: self.channel,
+        image: options.image,
+        repoAndTag: options.repoAndTag,
+        public: options.public
+    });
+
+    self._getAuthHeaders(function (hErr, headers) {
+        if (hErr) {
+            callback(hErr);
+            return;
+        }
+        headers['Content-Type'] = 'application/json';
+        if (options.regAuth) {
+            headers['x-registry-auth'] = options.regAuth;
+        }
+        if (options.headers) {
+            simpleMerge(headers, options.headers);
+        }
+        var reqOpts = {
+            path: path,
+            headers: headers
+        };
+
+        self.rawClient.post(reqOpts, function (connectErr, req) {
+            if (connectErr) {
+                callback(connectErr);
+                return;
+            }
+
+            req.on('result', function (resultErr, res) {
+                if (resultErr) {
+                    extendErrFromRawBody(resultErr, res, function () {
+                        callback(resultErr, null, res);
+                    });
+                    return;
+                }
+
+                res.setEncoding('utf8');
+                callback(null, res);
+            });
+
+            req.write(JSON.stringify(options.image));
+            req.end();
+        });
+    });
+};
+
+
 /*
  * Wait for a job to complete.  Returns an error if the job fails with an error
  * other than the (optional) list of expected errors. Taken from SAPI
diff --git a/package.json b/package.json
index 48c9fdf..9e15f80 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "10.4.1",
+  "version": "10.5.0",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
