From 1ec18f3a8d40d2b3ec331f62a2897c5508f2f581 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Thu, 31 Jan 2019 16:54:04 -0500
Subject: [PATCH] TRITON-1092 sdc-system-tests "sdc-system-tests.tap" not in
 TAP format

---
 runtests | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/runtests b/runtests
index 5c9b1f4..7bf2e97 100755
--- a/runtests
+++ b/runtests
@@ -66,7 +66,7 @@ NODEUNIT=./node_modules/nodeunit/bin/nodeunit
 
 # Options.
 opt_test_pattern=
-opt_reporter=default
+opt_reporter=tap
 while getopts "hf:r:" opt
 do
     case "$opt" in
-- 
2.21.0

