From 29bb63e7e2dae0b4d30b1ebb8956302257c0c186 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Mon, 11 Dec 2017 12:11:28 -0800
Subject: [PATCH] TOOLS-1949 sdcnode updates for Dec 2017 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 3e67c0b..e127d9b 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,7 @@ JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 SMF_MANIFESTS_IN = smf/manifests/docker.xml.in
 CLEAN_FILES += ./node_modules
 
-NODE_PREBUILT_VERSION=v4.8.5
+NODE_PREBUILT_VERSION=v4.8.7
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
-- 
2.21.0

