commit 909af066f05548643bb21746b334a40d61801df0 (refs/changes/44/5644/1)
Author: dyep <dyep49@gmail.com>
Date:   2019-02-26T14:48:37-08:00 (7 months ago)
    
    TRITON-1246 CloudAPI parameter disks.*.uuid should be disks.*.id

diff --git a/docs/index.md b/docs/index.md
index bfb5f2e..44e1dcd 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -922,9 +922,13 @@ Note that a `Triton-Datacenter-Name` response header was added in 9.2.0.
 
 The section describes API changes in CloudAPI versions.
 
+## 9.4.1
+- The disks.\*.uuid parameter was renamed to disks.\*.id. Backwards incompatible
+  with version 9.4.0.
+
 ## 9.4.0
 - [CreateMachine](#CreateMachine) may pass disk quantity and size for bhyve
-  instances. [GetMachine](#GetMachine) and [ListMachines](#ListMachcines)
+  instances. [GetMachine](#GetMachine) and [ListMachines](#ListMachines)
   returns information about these disks. [GetMachineSnapshot](#GetMachineSnapshot)
   and [ListMachineSnapshots](#ListMachineSnapshots) includes snapshot size.
 
@@ -4799,15 +4803,15 @@ The `disks` input parameter allows users to specify a list of disks to be provis
 ```
 "disks": [
   {
-    "uuid": "eea4e223-dee6-44dc-a7e1-71f996e534f0",
+    "id": "eea4e223-dee6-44dc-a7e1-71f996e534f0",
     "boot": true
   },
   {
-    "uuid": "dea91a7f-5fe3-4408-b25a-994c97a7975e",
+    "id": "dea91a7f-5fe3-4408-b25a-994c97a7975e",
     "size": 512
   },
   {
-    "uuid": "c41ce11e-bed2-45d2-bdb8-8dc889ed8ced",
+    "id": "c41ce11e-bed2-45d2-bdb8-8dc889ed8ced",
     "size": "remaining"
   }
 ]
@@ -4817,7 +4821,7 @@ Each object of the `disks` array has the following layout
 
 **Field**  | **Type** | **Description**
 ---------- | -------- | ---------------
-UUID       | UUID     | Unique id for this disk
+id         | UUID     | Unique id for this disk
 boot       | Boolean  | If `true`, this is the boot disk
 image      | UUID     | The image from which the disk was created
 size       | Integer  | The size of the disk in mebibytes or "remaining". If "remaining", size will be set to the difference between the package quota and sum of the other disks.
diff --git a/lib/machines.js b/lib/machines.js
index 2596514..b83d272 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -264,7 +264,7 @@ function translate(machine, req)  {
             }
 
             if (disk.uuid) {
-                diskObj.uuid = disk.uuid;
+                diskObj.id = disk.uuid;
             }
 
             return diskObj;
@@ -471,6 +471,12 @@ function getCreateOptions(req) {
                 var imageUuid = params.image || pkg.image;
                 disks[0].image_uuid = imageUuid;
 
+                disks.forEach(function uuidToId(disk) {
+                    disk.uuid = disk.id;
+                    delete disk.id;
+                    return disk;
+                });
+
                 var remainingIdx;
                 var disksSum = disks.reduce(
                     function sumDisk(sum, disk, idx) {
diff --git a/package.json b/package.json
index 065de4e..78aed8d 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "SmartDataCenter CloudAPI",
-    "version": "9.4.0",
+    "version": "9.4.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
diff --git a/test/machines.94.test.js b/test/machines.94.test.js
index 3fb29e4..6109929 100644
--- a/test/machines.94.test.js
+++ b/test/machines.94.test.js
@@ -479,9 +479,9 @@ test('CreateMachine - Disks/flexible disk package', function (t) {
 
     var obj = {
         disks: [
-            { uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0' },
-            { uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 512},
-            { uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 'remaining' }
+            { id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0' },
+            { id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 512},
+            { id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 'remaining' }
         ],
         image: BHYVE_IMAGE_UUID,
         name: 'bhyve-disks-flex-package-test-' + process.pid,
@@ -542,15 +542,15 @@ test('GetMachine has disks - Disks/flexible disk package', function (t) {
             boot: true,
             image: BHYVE_IMAGE_UUID,
             size: BHYVE_IMAGE.image_size,
-            uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
+            id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
         },
         {
             size: 512,
-            uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
+            id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
         },
         {
             size: BHYVE_128_FLEXIBLE.quota - BHYVE_IMAGE.image_size - 512,
-            uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
+            id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
         }
     ];
 
@@ -587,9 +587,9 @@ test('CreateMachine - Disks sum to quota/flex disk package', function (t) {
 
     var obj = {
         disks: [
-            { uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0', size: 14336 },
-            { uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 512},
-            { uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 512 }
+            { id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0', size: 14336 },
+            { id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 512},
+            { id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 512 }
         ],
         image: BHYVE_IMAGE_UUID,
         name: 'bhyve-disks-max-test-' + process.pid,
@@ -651,15 +651,15 @@ test('GetMachine has disks - Disks sum to quota/flex disk package',
             boot: true,
             image: BHYVE_IMAGE_UUID,
             size: 14336,
-            uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
+            id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
         },
         {
             size: 512,
-            uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
+            id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
         },
         {
             size: 512,
-            uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
+            id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
         }
     ];
 
@@ -698,9 +698,9 @@ test('CreateMachine - Disks with remaining/flex disk package',
 
     var obj = {
         disks: [
-            { uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0' },
-            { uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 'remaining'},
-            { uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 512 }
+            { id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0' },
+            { id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e', size: 'remaining'},
+            { id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced', size: 512 }
         ],
         image: BHYVE_IMAGE_UUID,
         name: 'bhyve-disks-flex-package-test-' + process.pid,
@@ -762,15 +762,15 @@ test('GetMachine has disks - Disks with remaining/flex disk package',
             boot: true,
             image: BHYVE_IMAGE_UUID,
             size: BHYVE_IMAGE.image_size,
-            uuid: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
+            id: 'eea4e223-dee6-44dc-a7e1-71f996e534f0'
         },
         {
             size: BHYVE_128_FLEXIBLE.quota - BHYVE_IMAGE.image_size - 512,
-            uuid: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
+            id: 'dea91a7f-5fe3-4408-b25a-994c97a7975e'
         },
         {
             size: 512,
-            uuid: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
+            id: 'c41ce11e-bed2-45d2-bdb8-8dc889ed8ced'
         }
     ];
 
