From 71573aba3535c93478e3c8a3ff8d0fa2096372ec Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 21 Aug 2018 23:03:50 +0000
Subject: [PATCH] MORAY-490 Electric Moray should pass through UpdateBucket and
 DeleteBucket options

---
 lib/server.js | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index c019a11..18deb9a 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -573,7 +573,7 @@ function updateBucket(options) {
         }
 
         options.clients.array.forEach(function (client) {
-            client.updateBucket(name, cfg, function (err2) {
+            client.updateBucket(name, cfg, opts, function (err2) {
                 log.debug({
                     err: err2
                 }, 'updateBucket: returned');
@@ -647,7 +647,7 @@ function delBucket(options) {
         }
 
         options.clients.array.forEach(function (client) {
-            client.delBucket(name, function (err2) {
+            client.delBucket(name, opts, function (err2) {
                 log.debug({
                     err: err2
                 }, 'delBucket: returned');
-- 
2.21.0

