From d65c48ed4a365f6fdf7b34e232fea4ea242e48ea Mon Sep 17 00:00:00 2001
From: June Yi <june.yi@samsung.com>
Date: Thu, 10 Nov 2016 01:32:07 +0900
Subject: [PATCH] joyent/triton#219 sdc-usbkey mount fails in CoaL on KVM

---
 tools/coal-linux-kvm-setup | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/coal-linux-kvm-setup b/tools/coal-linux-kvm-setup
index 43ef238..38c5452 100755
--- a/tools/coal-linux-kvm-setup
+++ b/tools/coal-linux-kvm-setup
@@ -226,8 +226,8 @@ cat > ${VIRSH_DOMAIN_XML} << EOF
         <disk type='file' device='disk'>
             <driver name='qemu' type='${QEMU_IMAGE_FORMAT}' cache='none' io='native'/>
             <source file='/var/lib/libvirt/images/sdc-headnode-usb.${QEMU_IMAGE_FORMAT}'/>
-            <target dev='hda' bus='virtio' />
-            <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
+            <target dev='hda' bus='ide' />
+            <address type='drive' controller='0' bus='1' target='0' unit='0'/>
         </disk>
         <disk type='file' device='disk'>
             <driver name='qemu' type='${QEMU_IMAGE_FORMAT}' cache='none' io='native'/>
-- 
2.21.0

