From 94c729c6489591f1e5b11a82d0752ffe5166706d Mon Sep 17 00:00:00 2001
From: Patrick Mooney <pmooney@pfmooney.com>
Date: Sat, 24 Feb 2018 16:09:46 +0000
Subject: [PATCH] backout OS-6416: needs more work [fix manifest]

---
 manifest | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/manifest b/manifest
index e2568f6..c40d74b 100644
--- a/manifest
+++ b/manifest
@@ -3022,7 +3022,7 @@ f usr/share/man/man1/screen.1 0444 root bin
 f usr/bin/socat 0555 root bin
 f usr/share/man/man1/socat.1 0444 root bin
 # uefi-edk2
-f usr/share/bhyve/uefi-csm-rom.bin 0644 root bin
+# f usr/share/bhyve/uefi-csm-rom.bin 0644 root bin
 # uuid
 f usr/bin/uuid 0755 root bin
 f usr/share/man/man1/uuid.1 0644 root bin
-- 
2.21.0

