commit ddd3462f39293a270369fc7fbeec4ddc208a7a4c (refs/changes/03/3203/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2018-01-10T21:44:28+13:00 (1 year, 9 months ago)
    
    TRITON-19: Triton equivalent to AWS' termination protection

diff --git a/docs/index.md b/docs/index.md
index 109f416..a390bea 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -908,6 +908,12 @@ The set of supported *API versions* is given in the ping endpoint:
 
 The section describes API changes in CloudAPI versions.
 
+## 8.5.1
+
+- Added [Instance Protection](#instance-protection). Adding a special tag to an
+  instance will stop both [DeleteMachine](#DeleteMachine) and SDC Docker from
+  destroying an instance. This remains true until the tag is removed.
+
 ## 8.5.0
 
 - CreateMachine and AddNic now accept specifying a [network
@@ -5983,6 +5989,9 @@ Using node-smartdc:
 
 Allows you to completely destroy an instance.
 
+An instance cannot be destroyed so long as [Instance
+Protection](#instance-protection) is enabled on that instance.
+
 ### Inputs
 
 * None
@@ -5995,10 +6004,11 @@ Allows you to completely destroy an instance.
 
 For all possible errors, see [CloudAPI HTTP Responses](#cloudapi-http-responses).
 
-**Error Code**   | **Description**
----------------- | ---------------
-ResourceNotFound | If `:login` or `:id` does not exist
-InvalidState     | The instance is the wrong state to be deleted
+**Error Code**       | **Description**
+-------------------- | ---------------
+ResourceNotFound     | If `:login` or `:id` does not exist
+InvalidState         | The instance is the wrong state to be deleted
+CannotDestroyMachine | Instance Protection is enabled on this instance
 
 ### CLI Command
 
@@ -6117,6 +6127,38 @@ or
       }, ...]
 
 
+## Instance Protection
+
+If you want to decrease the risk of accidental instance destruction, it is
+possible to make instance destruction (e.g. through
+[DeleteMachine](#DeleteMachine)) a two-step process.
+
+Instances that have the tag `triton.instance.protection` set to boolean `true`
+cannot be deleted, either through CloudAPI or SDC Docker. In order to delete
+such an instance, the above tag needs to be removed first.
+
+Tags can be added during instance creation (see
+[CreateMachine](#CreateMachine)), or added later (see
+[AddMachineTags](#AddMachineTags)). The instance then cannot be destroyed until
+the tag is removed, although all other operations will still work. To destroy
+the instance, first remove the `triton.instance.protection` tag (see
+[DeleteMachineTag](#DeleteMachineTag)).
+
+An example using node-triton:
+
+    $ triton instance tag set eec23902 triton.instance.protection=true
+    {
+        "triton.instance.protection": true
+    }
+    $ triton instance rm eec23902
+    error: Machine has instance-protection tag "triton.instance.protection" enabled, preventing deletion
+    triton instance rm: error: command failure
+    $ triton instance tag rm eec23902 triton.instance.protection --wait
+    Deleted tag triton.instance.protection on instance eec23902
+    $ triton instance rm eec23902
+    Delete (async) instance eec23902 (eec23902-13a2-48ec-ae7d-a8b8a86f053b)
+
+
 
 
 # Analytics
diff --git a/lib/errors.js b/lib/errors.js
index 7e0f1e1..bdd973d 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -240,6 +240,14 @@ CloudApiError.prototype.description = 'Encountered an internal error.';
  */
 
 
+function CannotDestroyMachineError() {
+    _FriendlySigRestError.apply(this, arguments);
+}
+util.inherits(CannotDestroyMachineError, _FriendlySigRestError);
+CannotDestroyMachineError.prototype.restCode = 'MachineUndeletableError';
+CannotDestroyMachineError.prototype.statusCode = 409;
+CannotDestroyMachineError.prototype.description = 'Machine cannot be destroyed';
+
 
 // ---- wrappers for API responses
 
@@ -433,6 +441,7 @@ module.exports = {
     // Custom error classes.
     CloudApiError: CloudApiError,
 
+    CannotDestroyMachineError: CannotDestroyMachineError,
     DefaultFabricNetworkNotConfiguredError:
         DefaultFabricNetworkNotConfiguredError,
 
diff --git a/lib/machines.js b/lib/machines.js
index 0c820c0..cf695a5 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -48,6 +48,7 @@ var EXTERNAL_NIC_TAG = 'external';
 var INTERNAL_NIC_TAG = 'internal';
 // networks used if pkg doesn't have any
 var DEFAULT_NETWORKS = [EXTERNAL_NIC_TAG, INTERNAL_NIC_TAG];
+var INSTANCE_PROTECTION_TAG = 'triton.instance.protection';
 
 var PKG_USED_PARAMS = ['uuid', 'max_physical_memory', 'name', 'version',
         'networks', 'active', 'default', 'owner_uuids', 'fss', 'os'];
@@ -2073,6 +2074,14 @@ function del(req, res, next) {
     var log = req.log;
     var vmapi = req.sdc.vmapi;
 
+    // vmapi does this check too, but checking here lets us avoid the listJobs()
+    // call below
+    if (req.machine.tags[INSTANCE_PROTECTION_TAG]) {
+        return next(new errors.CannotDestroyMachineError(
+            'Instance has "' + INSTANCE_PROTECTION_TAG + '" tag enabled, ' +
+            'preventing deletion'));
+    }
+
     return vmapi.listJobs({
         vm_uuid: machine,
         owner_uuid: customer,
@@ -2124,10 +2133,6 @@ function del(req, res, next) {
             return next();
         });
     });
-
-
-
-
 }
 
 
diff --git a/package.json b/package.json
index 11ad5a3..d40d9c3 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "SmartDataCenter CloudAPI",
-    "version": "8.5.0",
+    "version": "8.5.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
diff --git a/test/machines.test.js b/test/machines.test.js
index 848f19c..966b62f 100644
--- a/test/machines.test.js
+++ b/test/machines.test.js
@@ -1235,7 +1235,8 @@ test('Create Docker machine', function (t) {
             'docker:entrypoint': '[]'
         },
         tags: {
-            'docker:label:com.docker.blah': 'quux'
+            'docker:label:com.docker.blah': 'quux',
+            'triton.instance.protection': true
         },
         autoboot: true, // false
         docker: true,
@@ -1408,6 +1409,44 @@ function (t) {
 });
 
 
+test('Check cannot delete machine with triton.instance.protection tag',
+function (t) {
+    CLIENT.del('/my/machines/' + MACHINE_UUID, function (err, req, res, body) {
+        t.ok(err, 'delete error expected');
+        t.equal(res.statusCode, 409, 'http code');
+        t.deepEqual(body, {
+            code: 'MachineUndeletableError',
+            message: 'Instance has "triton.instance.protection" tag enabled, ' +
+                'preventing deletion'
+        }, 'check error message');
+        t.end();
+    });
+});
+
+
+test('Remove triton.instance.protection tag', function (t) {
+    var url = '/my/machines/' + MACHINE_UUID + '/tags/' +
+        'triton.instance.protection';
+
+    CLIENT.del(url, function (err, req, res) {
+        t.ifError(err, 'delete tag error');
+        t.equal(res.statusCode, 204, 'http code');
+
+        CLIENT.vmapi.listJobs({
+            vm_uuid: MACHINE_UUID,
+            task: 'update'
+        }, function (err2, jobs) {
+            t.ifError(err2, 'list jobs error');
+
+            machinesCommon.waitForJob(CLIENT, jobs[0].uuid, function (err3) {
+                t.ifError(err3, 'wait for job error');
+                t.end();
+            });
+        });
+    });
+});
+
+
 test('Delete Docker machine', deleteMachine);
 
 
