From 95e482e9a9fbb1e0e350515538ea9027d4c376d3 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 11 Jan 2017 18:15:18 -0800
Subject: [PATCH] MANTA-3083 MANTA-1764 not honouring keep-alive

---
 sapi_manifests/mako/template | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index bd42da7..8e87f10 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -14,14 +14,25 @@ http {
 	client_max_body_size 0;
 	default_type	application/octet-stream;
 	include		mime.types;
-	keepalive_timeout 2147483647;
-	keepalive_requests 2147483647;
+
+	{{!
+	  Due to MANTA-3084, we must continue to default to HTTP keepalives
+	  being disabled indefinitely (as older muskies will leak sockets
+	  rapidly and cease functioning if we allow them to use keep-alive).
+	  Once a particular Manta installation has upgraded all of both its
+	  muskies and makos, it can set this tweakable to a reasonable value
+	  (86400 suggested) to enable it. New installs post-MANTA-3084 should
+	  set this during setup.
+	}}
+	keepalive_timeout {{#KEEPALIVE_TIMEOUT}}{{KEEPALIVE_TIMEOUT}}{{/KEEPALIVE_TIMEOUT}}{{^KEEPALIVE_TIMEOUT}}0{{/KEEPALIVE_TIMEOUT}};
+	keepalive_requests {{#KEEPALIVE_TIMEOUT}}1000000{{/KEEPALIVE_TIMEOUT}}{{^KEEPALIVE_TIMEOUT}}0{{/KEEPALIVE_TIMEOUT}};
+
 	sendfile 	on;
 	send_timeout	300s;
 
 	server {
 		autoindex 	on;
-		listen		80;
+		listen		80 so_keepalive=10s:30s:10;
 		root		/manta;
 		server_name	localhost;
 
-- 
2.21.0

