commit 817ae4963f81f6728302d6c13fbce2d28d098897 (refs/changes/38/1238/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-01-11T18:15:18-08:00 (2 years, 9 months ago)
    
    MANTA-3083 MANTA-1764 not honouring keep-alive

diff --git a/sapi_manifests/mako/template b/sapi_manifests/mako/template
index bd42da7..959dd80 100644
--- a/sapi_manifests/mako/template
+++ b/sapi_manifests/mako/template
@@ -14,14 +14,14 @@ http {
 	client_max_body_size 0;
 	default_type	application/octet-stream;
 	include		mime.types;
-	keepalive_timeout 2147483647;
-	keepalive_requests 2147483647;
+	keepalive_timeout 86400;
+	keepalive_requests 1000000;
 	sendfile 	on;
 	send_timeout	300s;
 
 	server {
 		autoindex 	on;
-		listen		80;
+		listen		80 so_keepalive=10s:30s:10;
 		root		/manta;
 		server_name	localhost;
 
