From b7607f7beac72d298a3a3c78eeb9defaa580448e Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 9 Aug 2019 15:48:31 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index ca4a08e..9058562 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit ca4a08e8b7bba7a86d897060338fba27ee4e872b
+Subproject commit 9058562f2f59892795e470ce25443ff636c90478
-- 
2.21.0

