commit 70587b471634be4008a39e561755ed2a0d079e73 (refs/changes/95/895/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2016-11-14T20:07:24+01:00 (2 years, 11 months ago)
    
    TOOLS-1609 sdcadm: recent curl changes broke dockerlogger/agents update

diff --git a/lib/post-setup/docker.js b/lib/post-setup/docker.js
index b05834d..6e76172 100644
--- a/lib/post-setup/docker.js
+++ b/lib/post-setup/docker.js
@@ -580,7 +580,7 @@ function do_docker(subcmd, opts, args, cb) {
                 /*
                  * Exit 22 if cannot download the installer file (curl code)
                  */
-                '/usr/bin/curl -kOsf ' + ip + ':/extra/dockerlogger/' + f,
+                '/usr/bin/curl -kOsf http://' + ip + '/extra/dockerlogger/' + f,
                 'if [[ "$?" -ne "0" ]]; then',
                 '   exit $?',
                 'fi',
diff --git a/lib/procedures/update-dockerlogger.js b/lib/procedures/update-dockerlogger.js
index f814b7d..6f1151b 100644
--- a/lib/procedures/update-dockerlogger.js
+++ b/lib/procedures/update-dockerlogger.js
@@ -350,7 +350,7 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
                     /*
                      * Exit 22 if cannot download the installer file (curl code)
                      */
-                    '/usr/bin/curl -kOsf ' + ip + ':/extra/dockerlogger/' + f,
+                    '/usr/bin/curl -kOsf http://' + ip + '/extra/dockerlogger/' + f,
                     'if [[ "$?" -ne "0" ]]; then',
                     '   exit $?',
                     'fi',
diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 8849c11..4b81de5 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -1702,8 +1702,8 @@ SdcAdm.prototype.updateAgents = function updateAgents(options, callback) {
                 '       mkdir -p /var/tmp/node.config',
                 '   fi',
                 '',
-                '   /usr/bin/curl -ksf ' + ip +
-                        ':/extra/joysetup/node.config' +
+                '   /usr/bin/curl -ksf http://' + ip +
+                        '/extra/joysetup/node.config' +
                         ' -o /var/tmp/node.config/node.config',
                 '   if [[ "$?" -ne "0" ]]; then',
                 '       exit 33',
@@ -1734,7 +1734,7 @@ SdcAdm.prototype.updateAgents = function updateAgents(options, callback) {
                 /*
                  * Exit 22 if cannot download the installer file (curl code)
                  */
-                '/usr/bin/curl -kOsf ' + ip + ':/extra/agents/' + f,
+                '/usr/bin/curl -kOsf http://' + ip + '/extra/agents/' + f,
                 'if [[ "$?" -ne "0" ]]; then',
                 '   exit $?',
                 'fi',
