From af41935cf8988d72baa42f624c941d222993b0db Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Tue, 7 May 2019 01:30:29 +0000
Subject: [PATCH] MANTA-4141 Add metricPorts mdata variable to moray

---
 boot/setup.sh | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index b23a531..65c799c 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -181,6 +181,19 @@ function setup_moray {
             fatal "unable to start $moray_instance"
     done
 
+    #
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    #
+    # The metricPorts values are derived from the moray service's "SIZE"
+    # SAPI metadata. We don't need to worry about keeping the metricPorts
+    # updated if this variable changes, because such a change does not affect
+    # already-provisioned zones. This is because moray zones pull the "SIZE"
+    # variable from /var/tmp/metadata.json, which is only written once, when the
+    # zone is provisioned -- it is not managed by config-agent.
+    #
+    mdata-put metricPorts $(IFS=','; echo "${kangs[*]}")
+
     unset IFS
 }
 
-- 
2.21.0

