commit 097aa7b8469e5d527c6764dab3b75df72d3b6630
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-05-07T01:30:34+00:00 (5 months ago)
    
    MANTA-4141 Add metricPorts mdata variable to moray

diff --git a/boot/setup.sh b/boot/setup.sh
index b23a531..b3ac9d4 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -181,6 +181,18 @@ function setup_moray {
             fatal "unable to start $moray_instance"
     done
 
+    #
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    #
+    # The metricPorts values are derived from the moray service's "SIZE"
+    # SAPI metadata. We don't need to worry about keeping the metricPorts
+    # updated if this variable changes, because such a change does not affect
+    # already-provisioned zones. This is because moray zones pull the "SIZE"
+    # variable from /var/tmp/metadata.json, which is only written once, when the
+    # zone is provisioned -- it is not managed by config-agent.
+    #
+    mdata-put metricPorts $(IFS=','; echo "${kangs[*]}")
     unset IFS
 }
 
