From fe19af7267a9a15221629763a7ccc20d13909de6 Mon Sep 17 00:00:00 2001
From: Isaac Davis <isaac.davis@joyent.com>
Date: Fri, 1 Mar 2019 22:25:27 +0000
Subject: [PATCH] MANTA-4141 Add metricPorts mdata variable to moray

---
 boot/setup.sh | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/boot/setup.sh b/boot/setup.sh
index b23a531..b41efe8 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -181,6 +181,10 @@ function setup_moray {
             fatal "unable to start $moray_instance"
     done
 
+    # We join the metric ports in a comma-separated list, then add this list as
+    # metricPorts mdata to allow scraping by cmon-agent.
+    mdata-put metricPorts $(IFS=','; echo "${kangs[*]}")
+
     unset IFS
 }
 
-- 
2.21.0

