commit c952ad1846a34c748ad29b5fa57f3d3bd4a1decb (refs/changes/13/4513/2)
Author: Jason King <jason.king@joyent.com>
Date:   2018-07-13T16:59:06+00:00 (1 year, 3 months ago)
    
    OS-7076 lx netlink msg_done only sends header

diff --git a/usr/src/uts/common/brand/lx/io/lx_netlink.c b/usr/src/uts/common/brand/lx/io/lx_netlink.c
index bada80ffd9..976a61cd7c 100644
--- a/usr/src/uts/common/brand/lx/io/lx_netlink.c
+++ b/usr/src/uts/common/brand/lx/io/lx_netlink.c
@@ -889,6 +889,7 @@ static void
 lx_netlink_reply_done(lx_netlink_reply_t *reply)
 {
 	lx_netlink_sock_t *lxsock = reply->lxnr_sock;
+	lx_netlink_hdr_t *hdr;
 	mblk_t *mp;
 
 	/*
@@ -900,7 +901,6 @@ lx_netlink_reply_done(lx_netlink_reply_t *reply)
 		/*
 		 * If anything failed, we'll send up an error message.
 		 */
-		lx_netlink_hdr_t *hdr;
 		lx_netlink_err_t *err;
 
 		if (reply->lxnr_mp != NULL) {
@@ -923,6 +923,23 @@ lx_netlink_reply_done(lx_netlink_reply_t *reply)
 		hdr->lxnh_seq = reply->lxnr_hdr.lxnh_seq;
 		hdr->lxnh_pid = lxsock->lxns_port;
 	} else {
+		uint32_t status = 0;
+
+		/*
+		 * More recent versions of the iproute2 utils expect a status
+		 * value after the header, even in the absence of errors.
+		 */
+		lx_netlink_reply_add(reply, &status, sizeof (status));
+
+		/*
+		 * "done" is also the most minimal response possible.  If
+		 * lx_netlink_reply_msg() does not set lxnr_errno, we should
+		 * be guaranteed enough room to hold this (i.e. our
+		 * lx_netlink_reply_add() call should never end up setting
+		 * lxnr_errno).
+		 */
+		VERIFY0(reply->lxnr_errno);
+
 		mp = reply->lxnr_mp;
 		VERIFY(mp != NULL);
 		reply->lxnr_mp = NULL;
