commit 02d6d03286a313f7dd16974fe430d87ff24764fe (refs/changes/13/4513/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-07-13T07:20:39+00:00 (1 year, 3 months ago)
    
    OS-7076 lx netlink msg_done only sends header

diff --git a/usr/src/uts/common/brand/lx/io/lx_netlink.c b/usr/src/uts/common/brand/lx/io/lx_netlink.c
index bada80ffd9..69edb95dc0 100644
--- a/usr/src/uts/common/brand/lx/io/lx_netlink.c
+++ b/usr/src/uts/common/brand/lx/io/lx_netlink.c
@@ -889,6 +889,7 @@ static void
 lx_netlink_reply_done(lx_netlink_reply_t *reply)
 {
 	lx_netlink_sock_t *lxsock = reply->lxnr_sock;
+	lx_netlink_hdr_t *hdr;
 	mblk_t *mp;
 
 	/*
@@ -900,7 +901,6 @@ lx_netlink_reply_done(lx_netlink_reply_t *reply)
 		/*
 		 * If anything failed, we'll send up an error message.
 		 */
-		lx_netlink_hdr_t *hdr;
 		lx_netlink_err_t *err;
 
 		if (reply->lxnr_mp != NULL) {
@@ -926,6 +926,16 @@ lx_netlink_reply_done(lx_netlink_reply_t *reply)
 		mp = reply->lxnr_mp;
 		VERIFY(mp != NULL);
 		reply->lxnr_mp = NULL;
+
+		/*
+		 * More recent versions of the iproute2 utils expect a status
+		 * value after the header, even in the absence of errors
+		 */
+		hdr = (lx_netlink_hdr_t *)mp->b_rptr;
+		hdr->lxnh_len += sizeof (uint32_t);
+
+		bzero(mp->b_wptr, sizeof (uint32_t));
+		mp->b_wptr += sizeof (uint32_t);
 	}
 
 	lx_netlink_reply_sendup(reply, mp, reply->lxnr_mp1);
