From ebb6d3a80ec2c0abbc840da7099f7c97f523c6d7 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 3 Dec 2018 16:16:06 -0500
Subject: [PATCH] TOOLS-2117 sdcnode updates for November 2018 node.js security
 releases

---
 Makefile     | 2 +-
 package.json | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 86040e6..61c55e4 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ ESLINT_FILES := $(JS_FILES)
 
 # Should be the same version as the platform's /usr/node/bin/node.
 NODE_PREBUILT_TAG =	gz
-NODE_PREBUILT_VERSION =	v6.14.0
+NODE_PREBUILT_VERSION =	v6.15.1
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-smartos/1.6.3.
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
diff --git a/package.json b/package.json
index e909684..10a0836 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "vm-agent",
     "description": "SmartDC Node VM Agent",
-    "version": "1.8.0",
+    "version": "1.8.1",
     "author": "Joyent (joyent.com)",
     "private": true,
     "dependencies": {
-- 
2.21.0

