From 8120c92b88b9eaf04c591c75d32907166fd8ca60 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 3 Dec 2018 16:16:06 -0500
Subject: [PATCH] TOOLS-2117 sdcnode updates for November 2018 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 86040e6..01394ba 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ ESLINT_FILES := $(JS_FILES)
 
 # Should be the same version as the platform's /usr/node/bin/node.
 NODE_PREBUILT_TAG =	gz
-NODE_PREBUILT_VERSION =	v6.14.0
+NODE_PREBUILT_VERSION =	v6.15.0
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-smartos/1.6.3.
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
-- 
2.21.0

