commit 8120c92b88b9eaf04c591c75d32907166fd8ca60 (refs/changes/79/5179/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-03T16:16:06-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 86040e6..01394ba 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ ESLINT_FILES := $(JS_FILES)
 
 # Should be the same version as the platform's /usr/node/bin/node.
 NODE_PREBUILT_TAG =	gz
-NODE_PREBUILT_VERSION =	v6.14.0
+NODE_PREBUILT_VERSION =	v6.15.0
 ifeq ($(shell uname -s),SunOS)
 	# Allow building on a SmartOS image other than sdc-smartos/1.6.3.
 	NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
