From 8c2d070d763cb599cce1e3ef64ff6162896f91d7 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Fri, 24 Aug 2018 19:32:41 +0000
Subject: [PATCH] TOOLS-2070 tools/mk/Makefile.targ should conditionally set
 BUILD

---
 tools/mk/Makefile.targ | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index dac448a..1674137 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -126,7 +126,7 @@ endif
 #
 # Defaults for other fixed values.
 #
-BUILD		= build
+BUILD		?= build
 DISTCLEAN_FILES += $(BUILD)
 DOC_BUILD	= $(BUILD)/docs/public
 
-- 
2.21.0

