commit 9016a8e5c6d0caf83841ffc6f548c2fe0e4ab843 (refs/changes/41/4741/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-08-24T22:37:39+00:00 (1 year, 1 month ago)
    
    TOOLS-2070 tools/mk/Makefile.targ should conditionally set BUILD

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index dac448a..8ae89c2 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -73,6 +73,9 @@
 #
 #	BASH		Path to bash (default: "bash")
 #
+#	BUILD		top-level directory for node binaries, generated docs,
+#			and any other build output (default: "build")
+#
 #	CSCOPE_DIRS	Directories to search for source files for the cscope
 #			index. (default: ".")
 #
@@ -126,7 +129,7 @@ endif
 #
 # Defaults for other fixed values.
 #
-BUILD		= build
+BUILD		?= build
 DISTCLEAN_FILES += $(BUILD)
 DOC_BUILD	= $(BUILD)/docs/public
 
