commit 8c2d070d763cb599cce1e3ef64ff6162896f91d7 (refs/changes/41/4741/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-08-24T19:32:41+00:00 (1 year, 1 month ago)
    
    TOOLS-2070 tools/mk/Makefile.targ should conditionally set BUILD

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index dac448a..1674137 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -126,7 +126,7 @@ endif
 #
 # Defaults for other fixed values.
 #
-BUILD		= build
+BUILD		?= build
 DISTCLEAN_FILES += $(BUILD)
 DOC_BUILD	= $(BUILD)/docs/public
 
