commit eacffcc29c3b6ceba5978087a60a65fa623eb80f (refs/changes/00/5000/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-10-25T21:25:54+00:00 (12 months ago)
    
    OS-7328 lx recvmsg should allow cmsgs with no data

diff --git a/usr/src/uts/common/brand/lx/syscall/lx_socket.c b/usr/src/uts/common/brand/lx/syscall/lx_socket.c
index 18bc92433d..e961abdac1 100644
--- a/usr/src/uts/common/brand/lx/syscall/lx_socket.c
+++ b/usr/src/uts/common/brand/lx/syscall/lx_socket.c
@@ -1009,7 +1009,7 @@ stol_cmsgs_copyout(void *input, socklen_t inlen, void *addr,
 		goto finish;
 	}
 
-	VERIFY(inlen > sizeof (struct cmsghdr));
+	VERIFY(inlen >= sizeof (struct cmsghdr));
 
 	/*
 	 * First determine how much space we need for the conversion and
