From 38b69226c966f813d5a958f98482a8f39ce43cc7 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 2 Aug 2016 23:23:29 +0000
Subject: [PATCH] OS-5543 Deliver the ACPI debugging tools [fix mancheck- 2]

---
 usr/src/man/man1m/acpidump.1m | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/usr/src/man/man1m/acpidump.1m b/usr/src/man/man1m/acpidump.1m
index ee385781d3..0c313f927f 100644
--- a/usr/src/man/man1m/acpidump.1m
+++ b/usr/src/man/man1m/acpidump.1m
@@ -13,11 +13,11 @@
 .Dd Aug 2, 2016
 .Dt ACPIDUMP 1M
 .Os
-.Sh NAME 
+.Sh NAME
 .Nm acpidump
 .Nd dump ACPI tables
 .Sh SYNOPSIS
-.Nm 
+.Nm
 .Op Fl bhsvxz
 .Op Fl a Ar address
 .Op Fl c Ar on|off
@@ -64,5 +64,6 @@ Do not use the
 .Sy XSDT.
 .It Fl z
 Verbose.
+.El
 .Sh SEE ALSO
 .Xr acpixtract 1M
-- 
2.21.0

