From 2a623966e2376e693b7fcdb137143814a16c4dd9 Mon Sep 17 00:00:00 2001
From: Orlando Vazquez <ovazquez@gmail.com>
Date: Thu, 15 Feb 2018 09:41:32 -0800
Subject: [PATCH] TRITON-148 Update cn-agent node version to 6.x

---
 Makefile     | 2 +-
 package.json | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 3033115..b89ad62 100644
--- a/Makefile
+++ b/Makefile
@@ -37,7 +37,7 @@ JSSTYLE_FLAGS =		-o indent=4,doxygen,unparenthesized-return=0
 # doesn't know about (@@ENABLED@@)
 # SMF_MANIFESTS_IN = smf/manifests/cn-agent.xml.in
 
-NODE_PREBUILT_VERSION =	v4.6.1
+NODE_PREBUILT_VERSION =	v6.12.2
 NODE_PREBUILT_TAG =	gz
 ifeq ($(shell uname -s),SunOS)
 NODE_PREBUILT_IMAGE =	18b094b0-eb01-11e5-80c1-175dac7ddf02
diff --git a/package.json b/package.json
index d849c7c..fe64b85 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "2.1.0",
+  "version": "2.1.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

