From 0973cafda86b5983423a39fae128f5b4c2c7eba9 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 5 Dec 2018 16:55:07 -0800
Subject: [PATCH] TOOLS-2130 TOOLS-2104 broke the firmware-tools section of
 targets.json

---
 targets.json.in | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/targets.json.in b/targets.json.in
index 5d69bd5..8e33820 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -73,15 +73,15 @@ HEADNODE_TARGETS="${HEADNODE_TARGETS}$(build_headnode_target headnode-debug plat
 
 if [[ "${JOYENT_BUILD}" == "true" ]]; then
   FIRMWARE_TOOLS_DEP='"firmware-tools",'
-  FIRMWARE_TOOLS_ENTRY='
-  "firmware-tools": {
-    "build_platform": "20151126T062538Z",
-    "repos": [
-        {"url": "$MG_GIT_PREFIX/firmware-tools.git"}
-    ],
-    "public": false,
-    "deps": []
-  },'
+  FIRMWARE_TOOLS_ENTRY="
+  \"firmware-tools\": {
+    \"build_platform\": \"20151126T062538Z\",
+    \"repos\": [
+        {\"url\": \"$MG_GIT_PREFIX/firmware-tools.git\"}
+    ],
+    \"public\": false,
+    \"deps\": []
+  },"
   HEADNODE_TARGETS="${HEADNODE_TARGETS}$(build_headnode_target headnode-joyent platform 'firmware-tools')"
   HEADNODE_TARGETS="${HEADNODE_TARGETS}$(build_headnode_target headnode-joyent-debug platform-debug 'firmware-tools')"
 elif [[ "${JOYENT_BUILD}" == "false" ]]; then
-- 
2.21.0

