commit 1f23dc5f604580c9975535722cc9823f3a66789d (refs/changes/14/3114/15)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2018-01-09T17:44:21+00:00 (1 year, 9 months ago)
    
    MANTA-3476 read-only pnodes

diff --git a/boot/setup.sh b/boot/setup.sh
index 7bdcb46..44bb06d 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -143,9 +143,11 @@ function manta_setup_electric_moray {
     #Build the list of ports.  That'll be used for everything else.
     local ports
     local kangs
+    local statuses
     for (( i=1; i<=$ELECTRIC_MORAY_INSTANCES; i++ )); do
         ports[$i]=`expr 2020 + $i`
         kangs[$i]=`expr 3020 + $i`
+        statuses[$i]=`expr 4020 + $i`
     done
 
     #Regenerate the registrar config with the real ports included
@@ -184,10 +186,12 @@ function manta_setup_electric_moray {
     for (( i=1; i<=$ELECTRIC_MORAY_INSTANCES; i++ )); do
         local port=${ports[$i]}
         local kang=${kangs[$i]}
+        local status=${statuses[$i]}
         local electric_moray_instance="electric-moray-$port"
         local electric_moray_xml_out=$SVC_ROOT/smf/manifests/electric-moray-$port.xml
         sed -e "s#@@ELECTRIC-MORAY_PORT@@#$port#g" \
             -e "s#@@KANG_PORT@@#$kang#g" \
+            -e "s#@@STATUS_PORT@@#$status#g" \
             -e "s#@@ELECTRIC-MORAY_INSTANCE_NAME@@#$electric_moray_instance#g" \
             $electric_moray_xml_in  > $electric_moray_xml_out || \
             fatal "could not process $electric_moray_xml_in to $electric_moray_xml_out"
diff --git a/lib/errors.js b/lib/errors.js
index 101abd9..7ccd017 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -5,67 +5,28 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
-var fs = require('fs');
-var util = require('util');
-
 var assert = require('assert-plus');
-var verror = require('verror');
-
-
-
-
-///--- Globals
-
-var WError = verror.WError;
-var VError = verror.VError;
-
-var slice = Function.prototype.call.bind(Array.prototype.slice);
-
-
-
-///--- Helpers
-
-function ISODateString(d) {
-    function pad(n) {
-        return n < 10 ? '0' + n : n;
-    }
-
-    if (typeof (d) === 'string')
-        d = new Date(d);
-
-    return d.getUTCFullYear() + '-'
-        + pad(d.getUTCMonth()+1) + '-'
-        + pad(d.getUTCDate()) + 'T'
-        + pad(d.getUTCHours()) + ':'
-        + pad(d.getUTCMinutes()) + ':'
-        + pad(d.getUTCSeconds()) + 'Z';
-}
-
-
-
-///--- Errors
+var util = require('util');
+var VError = require('verror').VError;
 
 function InvocationError() {
-        VError.apply(this, arguments);
-        this.name = this.constructor.name;
+    VError.apply(this, arguments);
+    this.name = this.constructor.name;
 }
 util.inherits(InvocationError, VError);
 
-function ReadOnlyError(cause) {
-        if (arguments.length === 0) {
-                cause = {};
-        }
-        WError.call(this, cause, 'some vnodes are in read-only mode');
-        this.name = this.constructor.name;
+function ReadOnlyError(pnode) {
+    if (!pnode) {
+        VError.call(this, {}, 'some vnodes are in read-only mode');
+    } else {
+        VError.call(this, 'pnode "%s" is in read-only mode', pnode);
+    }
+    this.name = this.constructor.name;
 }
-util.inherits(ReadOnlyError, WError);
-
-
-
-///--- Exports
+util.inherits(ReadOnlyError, VError);
 
 module.exports = {
     InvocationError: InvocationError,
diff --git a/lib/index.js b/lib/index.js
index 67c2c58..7c1a027 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -5,16 +5,13 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var server = require('./server');
-
-
-///--- Exports
+var status_server = require('./status_server');
 
 module.exports = {
-
-    createServer: server.createServer
-
+    createServer: server.createServer,
+    createStatusServer: status_server.createStatusServer
 };
diff --git a/lib/moray_client.js b/lib/moray_client.js
new file mode 100644
index 0000000..ba32fe3
--- /dev/null
+++ b/lib/moray_client.js
@@ -0,0 +1,90 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var clone = require('clone');
+var fs = require('fs');
+var moray = require('moray');
+var url = require('url');
+var verror = require('verror');
+
+/*
+ * Create moray clients in order to interact with moray instances.  Available
+ * moray clients are listed in the ring configuration in LevelDB, which we
+ * access in electric moray via node-fash.
+ */
+function createClient(options, callback) {
+    assert.object(options, 'options');
+    assert.object(options.log, 'options.log');
+    assert.object(options.ring, 'options.ring');
+    assert.object(options.morayOptions, 'options.morayOptions');
+    assert.func(callback, 'options.callback');
+
+    var log = options.log;
+
+    var clientMap = {};
+    var clientArray = [];
+
+    options.ring.getPnodes(function (err, pnodes) {
+        if (err) {
+            throw new verror.VError(err, 'unable to get pnodes');
+        }
+
+        assert.arrayOfString(pnodes, 'pnodes');
+
+        pnodes.forEach(function (pnode) {
+            var pnodeUrl = url.parse(pnode);
+            assert.string(pnodeUrl.port, 'pnodeUrl.port');
+            assert.string(pnodeUrl.hostname, 'pnodeUrl.hostname');
+
+            log.info({
+                url: pnodeUrl
+            }, 'creating moray client');
+
+            var morayargs = clone(options.morayOptions);
+            if (!morayargs.cueballOptions) {
+                morayargs.cueballOptions = {};
+            }
+            morayargs.unwrapErrors = true;
+            morayargs.srvDomain = pnodeUrl.hostname;
+            morayargs.cueballOptions.defaultPort = parseInt(pnodeUrl.port, 10);
+            morayargs.log = options.log.child({
+                component: 'moray-client-' + pnodeUrl.hostname
+            });
+
+            var client = moray.createClient(morayargs);
+            clientMap[pnode] = client;
+            clientArray.push(client);
+
+            if (clientArray.length === pnodes.length) {
+                // write ready cookie when clients have connected
+                log.info('all moray clients instantiated writing ready cookie');
+                try {
+                    fs.writeFileSync('/var/tmp/electric-moray-ready', null);
+                } catch (e) {
+                    throw new verror.VError(e, 'unable to write ready cookie');
+                }
+            }
+        });
+
+        if (clientArray.length <= 0) {
+            throw new verror.VError('No moray clients exist!');
+        }
+
+        return callback(null, {
+            map: clientMap,
+            array: clientArray
+        });
+    });
+}
+
+module.exports = {
+    createClient: createClient
+};
diff --git a/lib/server.js b/lib/server.js
index 2348b70..fbd7f40 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,25 +5,23 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
 var artedi = require('artedi');
-var clone = require('clone');
 var fast = require('fast');
 var fs = require('fs');
 var kang = require('kang');
-var moray_client = require('moray'); // client
 var net = require('net');
 var os = require('os');
 var restify = require('restify');
 var uuid = require('node-uuid');
-var url = require('url');
 var util = require('util');
 var vasync = require('vasync');
 var verror = require('verror');
 
+var moray_client = require('./moray_client');
 var dtrace = require('./dtrace');
 var errors = require('./errors');
 var schema = require('./schema');
@@ -32,16 +30,11 @@ var ring = require('./ring');
 var InvocationError = errors.InvocationError;
 var ReadOnlyError = errors.ReadOnlyError;
 
-
-///--- GLOBALS
-
 var ALLOWED_BATCH_OPS = [
     'put',
     'delete'
 ];
-
 var READ_ONLY = 'ro';
-
 var KANG_VERSION = '1.2.0';
 
 var B_ARGS_SCHEMA = [
@@ -119,10 +112,10 @@ var UO_ARGS_SCHEMA = [
     { name: 'options', type: 'object' }
 ];
 
-///--- API
 
-function createServer(options) {
+function createServer(options, callback) {
     assert.object(options, 'options');
+    assert.func(callback, 'callback');
     assert.string(options.ringLocation, 'options.ringLocation');
 
     var log = options.log;
@@ -150,7 +143,7 @@ function createServer(options) {
         opts.ring = _ring;
 
         log.info('creating moray clients');
-        createClient({
+        moray_client.createClient({
             ring: opts.ring,
             morayOptions: options.morayOptions,
             log: options.log
@@ -160,6 +153,7 @@ function createServer(options) {
             }
 
             opts.clients = clients;
+            opts.indexShards = options.ringCfg.indexShards;
 
             var collector = artedi.createCollector();
 
@@ -234,6 +228,10 @@ function createServer(options) {
 
             socket.on('listening', function () {
                 log.info('moray listening on %d', options.port);
+                callback(null, {
+                    ring: opts.ring,
+                    clientList: Object.keys(opts.clients.map)
+                });
             });
 
             socket.on('error', function (serr) {
@@ -245,18 +243,6 @@ function createServer(options) {
     });
 }
 
-
-
-///--- Exports
-
-module.exports = {
-    createServer: createServer
-};
-
-
-
-///--- Privates
-
 function invalidArgs(rpc, argv, types) {
     var route = rpc.methodName();
     var len = types.length;
@@ -303,11 +289,83 @@ function invalidArgs(rpc, argv, types) {
     return false;
 }
 
-//TODO; No bucket modification can occur whilst re-sharding.
+/*
+ * Returns a list of pnodes that are not safe for write operations. Parameters
+ * are a bunyan logger for the child process that called this function, and an
+ * options hash containing:
+ *      "indexShards"       an array of index pnodes from our configuration file
+ *      "clients"           a list of moray client tracking objects
+ */
+function listReadOnlyPnodes(log, options) {
+    assert.object(log, 'log');
+    assert.object(options, 'options');
+    assert.arrayOfObject(options.indexShards, 'options.indexShards');
+    assert.object(options.clients, 'options.clients');
+
+    log.debug({
+        indexShards: options.indexShards,
+        clients: options.clients
+    }, 'listReadOnlyPnodes: entered');
+
+    var pnodes = Object.keys(options.clients.map);
+    var readOnlyPnodes = pnodes.filter(function (pnode) {
+        return (isReadOnlyPnode(log, options.indexShards, pnode));
+    });
+
+    return readOnlyPnodes;
+}
+
+/*
+ * Checks if a moray client is set to read-only mode. Parameters are as follows:
+ *      "log"               a bunyan logger
+ *      "indexShards"       an array of index pnodes from our configuration file
+ *      "pnode"             a physical node name
+ */
+function isReadOnlyPnode(log, indexShards, pnode) {
+    assert.object(log, 'log');
+    assert.arrayOfObject(indexShards, 'indexShards');
+    assert.string(pnode, 'pnode');
+
+    log.debug({
+        indexShards: indexShards,
+        pnode: pnode
+    }, 'isReadOnlyPnode: entered');
+
+    var readOnlyStatus = null;
+    for (var i = 0; i < indexShards.length; i++) {
+        var shard = indexShards[i];
+        assert.optionalBool(shard.readOnly, 'shard.readOnly');
+        assert.string(shard.host, 'shard.host');
+        var shard_url = 'tcp://' + shard.host + ':2020';
+        if (shard_url === pnode) {
+            /*
+             * A shard in indexShards may not have a 'readOnly' field, which
+             * would mean it is not set to read-only mode.
+             */
+            readOnlyStatus = !!shard.readOnly;
+            return readOnlyStatus;
+        }
+    }
+
+    /*
+     * If no shard's host field in indexShards matches the pnode passed to this
+     * function, we are in an inconsistent state, and it is safer to assume that
+     * the pnode is read-only than the alternative, which may result in writes
+     * in error.
+     */
+    log.warn({
+        indexShards: indexShards,
+        pnode: pnode
+    }, 'isReadOnlyPnode: pnode not found in indexShards');
+    return true;
+}
+
+
 function createBucket(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
     assert.object(options.clients, 'options.clients');
+    assert.object(options.indexShards, 'options.indexShards');
 
     function _createBucket(rpc) {
         var argv = rpc.argv();
@@ -344,6 +402,12 @@ function createBucket(options) {
         var err = [];
         var done = 0;
 
+        var readOnlyPnodes = listReadOnlyPnodes(log, options);
+        if (readOnlyPnodes.length > 0) {
+            rpc.fail(new ReadOnlyError(readOnlyPnodes.join(', ')));
+            return;
+        }
+
         options.clients.array.forEach(function (client) {
             client.createBucket(name, cfg, opts, function (err2) {
                 log.debug({
@@ -356,7 +420,8 @@ function createBucket(options) {
                 }
 
                 if (++done === options.clients.array.length) {
-                    var multiError = err[0] ? new verror.MultiError(err) : null;
+                    var multiError = err[0]
+                        ? new verror.MultiError(err) : null;
                     log.debug({
                         err: multiError
                     }, 'createBucket: finished all shards');
@@ -492,6 +557,12 @@ function updateBucket(options) {
         var err = [];
         var done = 0;
 
+        var readOnlyPnodes = listReadOnlyPnodes(log, options);
+        if (readOnlyPnodes.length > 0) {
+            rpc.fail(new ReadOnlyError(readOnlyPnodes.join(', ')));
+            return;
+        }
+
         options.clients.array.forEach(function (client) {
             client.updateBucket(name, cfg, function (err2) {
                 log.debug({
@@ -503,7 +574,8 @@ function updateBucket(options) {
                 }
 
                 if (++done === options.clients.array.length) {
-                    var multiError = err[0] ? new verror.MultiError(err) : null;
+                    var multiError = err[0] ?
+                        new verror.MultiError(err) : null;
                     log.debug({
                         err: multiError
                     }, 'updateBucket: finished all shards');
@@ -559,6 +631,12 @@ function delBucket(options) {
         var err = [];
         var done = 0;
 
+        var readOnlyPnodes = listReadOnlyPnodes(log, options);
+        if (readOnlyPnodes.length > 0) {
+            rpc.fail(new ReadOnlyError(readOnlyPnodes.join(', ')));
+            return;
+        }
+
         options.clients.array.forEach(function (client) {
             client.delBucket(name, function (err2) {
                 log.debug({
@@ -570,7 +648,8 @@ function delBucket(options) {
                 }
 
                 if (++done === options.clients.array.length) {
-                    var multiError = err[0] ? new verror.MultiError(err) : null;
+                    var multiError = err[0] ?
+                        new verror.MultiError(err) : null;
                     log.debug({
                         err: multiError
                     }, 'delBucket: finished all shards');
@@ -642,6 +721,12 @@ function putObject(options) {
             v.vnode = node.vnode;
             var pnode = node.pnode;
             var client = options.clients.map[pnode];
+
+            if (isReadOnlyPnode(log, options.indexShards, pnode)) {
+                rpc.fail(new ReadOnlyError(pnode));
+                return;
+            }
+
             client.putObject(b, k, v, opts, function (pErr, meta) {
                 log.debug({
                     err: pErr,
@@ -795,7 +880,15 @@ function delObject(options) {
                 rpc.fail(new ReadOnlyError());
                 return;
             }
-            var client = options.clients.map[node.pnode];
+
+            var pnode = node.pnode;
+            var client = options.clients.map[pnode];
+
+            if (isReadOnlyPnode(log, options.indexShards, pnode)) {
+                rpc.fail(new ReadOnlyError(pnode));
+                return;
+            }
+
             client.delObject(b, k, opts, function (dErr) {
                 log.debug({
                     err: dErr
@@ -950,6 +1043,9 @@ function findObjects(options) {
     return _findObjects;
 }
 
+/*
+ * We don't currently support deleteMany operations.
+ */
 function deleteMany(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
@@ -1021,68 +1117,6 @@ function getTokens(options) {
     return _getTokens;
 }
 
-function createClient(options, callback) {
-    assert.object(options, 'options');
-    assert.object(options.log, 'options.log');
-    assert.object(options.ring, 'options.ring');
-    assert.object(options.morayOptions, 'options.morayOptions');
-    assert.func(callback, 'options.callback');
-
-    var log = options.log;
-
-    var clientMap = {};
-    var clientArray = [];
-
-    options.ring.getPnodes(function (err, pnodes) {
-        if (err) {
-            throw new verror.VError(err, 'unable to get pnodes');
-        }
-        pnodes.forEach(function (pnode) {
-            var pnodeUrl = url.parse(pnode);
-            assert.string(pnodeUrl.port, 'pnodeUrl.port');
-            assert.string(pnodeUrl.hostname, 'pnodeUrl.hostname');
-
-            log.info({
-                url: pnodeUrl
-            }, 'creating moray client');
-
-            var morayargs = clone(options.morayOptions);
-            if (!morayargs.cueballOptions) {
-                morayargs.cueballOptions = {};
-            }
-            morayargs.unwrapErrors = true;
-            morayargs.srvDomain = pnodeUrl.hostname;
-            morayargs.cueballOptions.defaultPort = parseInt(pnodeUrl.port, 10);
-            morayargs.log = options.log.child({
-                component: 'moray-client-' + pnodeUrl.hostname
-            });
-
-            var client = moray_client.createClient(morayargs);
-            clientMap[pnode] = client;
-            clientArray.push(client);
-
-            if (clientArray.length === pnodes.length) {
-                // write ready cookie when clients have connected
-                log.info('all moray clients instantiated writing ready cookie');
-                try {
-                    fs.writeFileSync('/var/tmp/electric-moray-ready', null);
-                } catch (e) {
-                    throw new verror.VError(e, 'unable to write ready cookie');
-                }
-            }
-        });
-
-        if (clientArray.length <= 0) {
-            throw new verror.VError('No moray clients exist!');
-        }
-
-        return callback(null, {
-            map: clientMap,
-            array: clientArray
-        });
-    });
-}
-
 function sql(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
@@ -1137,6 +1171,12 @@ function sql(options) {
             }
         });
 
+        var readOnlyPnodes = listReadOnlyPnodes(log, options);
+        if (readOnlyPnodes.length > 0) {
+            rpc.fail(new ReadOnlyError(readOnlyPnodes.join(', ')));
+            return;
+        }
+
         options.clients.array.forEach(function (client, index) {
             barrier.start(index);
             var req = client.sql(stmt, values, opts);
@@ -1159,7 +1199,9 @@ function sql(options) {
     return _sql;
 }
 
-// we don't currently support update operations
+/*
+ * We don't currently support update operations.
+ */
 function updateObjects(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
@@ -1197,13 +1239,13 @@ function updateObjects(options) {
 
 
 /*
- * Batching is only supported in a very limited case: when all of the requests
- * have keys with the same value after transformation (if appropriate for their
- * bucket), which allows us to be sure that all of the values live on the same
- * shard.
+ * Batching is only supported in a very limited case: for modifications when all
+ * of the requests have keys with the same value after transformation (if
+ * appropriate for their bucket), which allows us to be sure that all of the
+ * values live on the same shard.
  *
- * Operations like 'update' and 'deleteMany' are not allowed since they cannot
- * be guaranteed to only affect values on a single shard.
+ * The operations 'update' and 'deleteMany' are not allowed since they cannot be
+ * guaranteed to only affect values on a single shard.
  */
 function batch(options) {
     assert.object(options, 'options');
@@ -1302,9 +1344,18 @@ function batch(options) {
             var pnode = node.pnode;
             var client = options.clients.map[pnode];
 
+            if (isReadOnlyPnode(log, options.indexShards, pnode)) {
+                rpc.fail(new ReadOnlyError(pnode));
+                return;
+            }
+
             client.batch(requests, opts, done);
         });
     }
 
     return _batch;
 }
+
+module.exports = {
+    createServer: createServer
+};
diff --git a/lib/status_server.js b/lib/status_server.js
new file mode 100644
index 0000000..280e05d
--- /dev/null
+++ b/lib/status_server.js
@@ -0,0 +1,72 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var restify = require('restify');
+
+/*
+ * Serves HTTP requests for electric-moray process state access.
+ */
+function createStatusServer(options, callback) {
+    assert.object(options, 'options');
+    assert.object(options.log, 'options.log');
+    assert.number(options.port, 'options.port');
+    assert.object(options.ring, 'options.ring');
+    assert.arrayOfString(options.clientList, 'options.clientList');
+    assert.func(callback, 'callback');
+
+    var opts = {
+        log: options.log,
+        startTime: (new Date()).toISOString(),
+        clientList: options.clientList,
+        indexShards: options.indexShards
+    };
+
+    /*
+     * REST API server to access the status of an electric-moray instance.
+     */
+    var server = restify.createServer({
+        name: 'electric-moray-status'
+    });
+
+    server.get('/status', createStatusHandler(opts));
+
+    server.on('error', callback);
+
+    server.listen(options.port, function () {
+        opts.log.info('status server listening on port %d', options.port);
+        callback();
+    });
+
+}
+
+/*
+ * Exposes the current configured state of an electric-moray process, providing
+ * some details about this process and SMF identity.  This information enables
+ * other systems to correlate the output of svcs(1) with this status object.
+ */
+function createStatusHandler(opts) {
+    return (function statusHandler(req, res, next) {
+        var body = {
+            smf_fmri: process.env.SMF_FMRI || null,
+            pid: process.pid,
+            start_time: opts.startTime,
+            client_list: opts.clientList,
+            index_shards: opts.indexShards
+        };
+
+        res.send(200, body);
+        next();
+    });
+}
+
+module.exports = {
+    createStatusServer: createStatusServer
+};
diff --git a/main.js b/main.js
index 29bbe3c..42b5710 100644
--- a/main.js
+++ b/main.js
@@ -5,16 +5,14 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
-var fs = require('fs');
-var os = require('os');
-
 var assert = require('assert-plus');
 var bsyslog = require('bunyan-syslog');
 var bunyan = require('bunyan');
 var clone = require('clone');
+var fs = require('fs');
 var jsprim = require('jsprim');
 var getopt = require('posix-getopt');
 var VError = require('verror');
@@ -22,10 +20,6 @@ var extend = require('xtend');
 
 var app = require('./lib');
 
-
-
-///--- Globals
-
 var MIN_PORT = 1;
 var MAX_PORT = 65535;
 
@@ -33,6 +27,7 @@ var DEFAULTS = {
     file: process.cwd() + '/etc/config.json',
     port: 2020,
     monitorPort: 3020,
+    statusPort: 4020,
     bindip: '0.0.0.0'
 };
 var NAME = 'electric-moray';
@@ -50,9 +45,6 @@ var LOG = bunyan.createLogger({
 var LOG_LEVEL_OVERRIDE = false;
 
 
-
-///--- Internal Functions
-
 function setupLogger(config) {
     var cfg_b = config.bunyan;
     assert.object(cfg_b, 'config.bunyan');
@@ -111,7 +103,7 @@ function parsePort(str) {
 function parseOptions() {
     var option;
     var opts = {};
-    var parser = new getopt.BasicParser('cvf:r:p:k:', process.argv);
+    var parser = new getopt.BasicParser('cvf:r:p:k:s:', process.argv);
 
     while ((option = parser.getopt()) !== undefined) {
         switch (option.option) {
@@ -130,6 +122,9 @@ function parseOptions() {
             case 'k':
                 opts.monitorPort = parsePort(option.optarg);
                 break;
+            case 's':
+                opts.statusPort = parsePort(option.optarg);
+                break;
             case 'v':
                 // Allows us to set -vvv -> this little hackery just ensures
                 // that we're never < TRACE
@@ -181,12 +176,31 @@ function run(options) {
     opts.log = LOG;
     opts.name = NAME;
 
-    app.createServer(opts);
-}
-
+    app.createServer(opts, function (err, res) {
+        if (err) {
+            LOG.fatal(err, 'startup failed');
+            process.exit(1);
+        }
 
+        assert.object(res, 'res');
+        assert.object(res.ring, 'res.ring');
+        assert.arrayOfString(res.clientList, 'res.clientList');
+
+        app.createStatusServer({
+            log: LOG.child({ component: 'statusServer' }),
+            ring: res.ring,
+            clientList: res.clientList,
+            indexShards: opts.ringCfg.indexShards,
+            port: opts.statusPort
+        }, function (err2) {
+            if (err2) {
+                LOG.fatal(err2, 'status server startup failed');
+                process.exit(1);
+            }
+        });
+    });
+}
 
-///--- Mainline
 
 (function main() {
     var options = parseOptions();
diff --git a/package.json b/package.json
index 43a9473..7e868f9 100644
--- a/package.json
+++ b/package.json
@@ -24,7 +24,7 @@
         "moray": "^3.4.0",
         "node-uuid": "1.4.0",
         "posix-getopt": "^1.0.0",
-        "restify": "5.2.0",
+        "restify": "6.3.1",
         "strsplit": "1.0.0",
         "vasync": "^1.4.3",
         "verror": "1.9.0",
diff --git a/sapi_manifests/electric-moray/template b/sapi_manifests/electric-moray/template
index 813eb65..a3d0d9b 100644
--- a/sapi_manifests/electric-moray/template
+++ b/sapi_manifests/electric-moray/template
@@ -24,6 +24,21 @@
             "createIfMissing": false,
             "keyEncoding": "utf8",
             "valueEncoding": "json"
-        }
+        },
+        {{! This formatting is necessary to communicate the structure of our
+        indexShards JSON object.  For mustache, our templating engine (and
+        hogan.js, which is mustache for javascript), we have to put the comma in
+        an inverted section, surrounding it with "last" flags, which is a field
+        we use to indicate that mustache/hogan.js should not render a trailing
+        comma, which JSON disallows.}}
+        "indexShards": [ {{#INDEX_MORAY_SHARDS}}
+            {
+                "host": "{{host}}"{{#readOnly}},
+                {{! The readOnly field should have a value of true, or else it
+                should not be included in the object.  Its value is manipulated
+                via manta-shardadm.}}
+                "readOnly": {{readOnly}}{{/readOnly}}
+            }{{^last}},{{/last}}{{/INDEX_MORAY_SHARDS}}
+        ]
     }
 }
diff --git a/smf/manifests/electric-moray.xml.in b/smf/manifests/electric-moray.xml.in
index e810c3e..85d0866 100644
--- a/smf/manifests/electric-moray.xml.in
+++ b/smf/manifests/electric-moray.xml.in
@@ -7,7 +7,7 @@
 -->
 
 <!--
-    Copyright (c) 2014, Joyent, Inc.
+    Copyright (c) 2018, Joyent, Inc.
 -->
 
 <service_bundle type="manifest" name="-electric-moray">
@@ -44,7 +44,7 @@
         <exec_method
             type="method"
             name="start"
-            exec="node --abort-on-uncaught-exception main.js -f etc/config.json -r %{electric-moray/ring-location} -p %{electric-moray/port} -k %{electric-moray/kang} &amp;"
+            exec="node --abort-on-uncaught-exception main.js -f etc/config.json -r %{electric-moray/ring-location} -p %{electric-moray/port} -k %{electric-moray/kang} -s %{electric-moray/status} &amp;"
             timeout_seconds="30">
             <method_context working_directory="/opt/smartdc/electric-moray">
                 <method_credential user="root" group="nobody" />
@@ -69,6 +69,7 @@
             <property_group name="electric-moray" type="application">
                 <propval name="port" type="astring" value="@@ELECTRIC-MORAY_PORT@@" />
                 <propval name="kang" type="astring" value="@@KANG_PORT@@" />
+                <propval name="status" type="astring" value="@@STATUS_PORT@@" />
                 <propval name="ring-location" type="astring" value="/electric-moray/chash/leveldb-@@ELECTRIC-MORAY_PORT@@" />
             </property_group>
         </instance>
diff --git a/test/resharding/roBuckets.test.js b/test/resharding/roBuckets.test.js
index 73d0c45..0407fcc 100644
--- a/test/resharding/roBuckets.test.js
+++ b/test/resharding/roBuckets.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -22,46 +22,10 @@ if (require.cache[__dirname + '/helper.js'])
     delete require.cache[__dirname + '/helper.js'];
 var helper = require('./helper.js');
 
-
-
-///--- Globals
-
 var after = helper.after;
 var before = helper.before;
 var test = helper.test;
 
-var FULL_CFG = {
-    index: {
-        str: {
-            type: 'string'
-        },
-        str_u: {
-            type: 'string',
-            unique: true
-        },
-        num: {
-            type: 'number'
-        },
-        num_u: {
-            type: 'number',
-            unique: true
-        },
-        bool: {
-            type: 'boolean'
-        },
-        bool_u: {
-            type: 'boolean',
-            unique: true
-        }
-    },
-    pre: [function onePre(req, cb) { cb(); }],
-    post: [function onePost(req, cb) { cb(); }],
-    options: {}
-};
-
-
-
-///--- Helpers
 
 function assertBucket(name, t, bucket, cfg) {
     t.ok(bucket);
@@ -88,9 +52,6 @@ function assertBucket(name, t, bucket, cfg) {
 }
 
 
-
-///--- tests
-
 before(function (cb) {
     this.bucket = 'moray_unit_test_' + uuid.v4().substr(0, 7);
     this.assertBucket = assertBucket.bind(null, this.bucket);
