commit bb084ef45b2d820bcf7da062bad60a89e8888a30 (refs/changes/39/239/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2016-08-05T21:49:32+00:00 (3 years, 2 months ago)
    
    OS-5580 disable docs in kvm-cmd

diff --git a/build.sh b/build.sh
index 7c3907b..318f51c 100755
--- a/build.sh
+++ b/build.sh
@@ -51,6 +51,7 @@ XLDFLAGS="${XLDFLAGS} -lz -lm -lc -lvnd"
     --disable-brlapi \
     --disable-curl \
     --enable-debug \
+    --disable-docs \
     --enable-kvm \
     --enable-kvm-pit \
     --enable-vnc-png \
