From 391f2ce6009c230047d2eb2bd2a026a32e109473 Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Fri, 29 Jul 2016 20:59:49 +0000
Subject: [PATCH] CNAPI-658 CNAPI and cn-agent shouldn't try to manage firewall
 rules

---
 lib/cnapi.js | 82 ++++------------------------------------------------
 1 file changed, 5 insertions(+), 77 deletions(-)

diff --git a/lib/cnapi.js b/lib/cnapi.js
index d127a32..fb0691f 100644
--- a/lib/cnapi.js
+++ b/lib/cnapi.js
@@ -32,80 +32,6 @@ var cnapiUrl;
 
 
 
-/**
- * Start a provisioner task with CNAPI on each of the servers to update
- * the firewall data.
- *
- * Inputs:
- * - job.params.fwapiServers {Array} : server UUIDs to send update tasks to
- *
- * Outputs:
- * - job.params.taskIDs {Array}: task objects to poll in a later task, where
- *   each task object looks like:
- *     { server_uuid: uuid, task_id: id }
- */
-function fwUpdate(job, callback) {
-    if (!job.params.fwapiServers || job.params.fwapiServers.length === 0) {
-        return callback(null, 'No remote servers to update');
-    }
-
-    var serversSent = [];
-    var cnapi = new sdcClients.CNAPI({ url: cnapiUrl });
-    var matchingVMs = job.params.fwapiMatchingVMs || [];
-    job.params.taskIDs = [];
-
-    return async.forEachSeries(job.params.fwapiServers, function (uuid, cb) {
-        var endpoint = '/servers/' + uuid + '/fw/update';
-        var firewall = {
-            jobid: job.uuid
-        };
-
-        if (job.params.hasOwnProperty('rule')) {
-            firewall.rules = [ job.params.rule ];
-        }
-
-        var remoteVMs = matchingVMs.filter(function (rvm) {
-            return (rvm.server_uuid != uuid);
-        });
-
-        if (remoteVMs.length) {
-            firewall.remoteVMs = remoteVMs;
-        }
-
-        job.log.debug(firewall, 'Updating rules on server "%s"', uuid);
-        return cnapi.post(endpoint, firewall, function (err, task) {
-            if (err) {
-                if (err.restCode == 'InvalidVersion') {
-                    // This is a 6.5 CN - no firewall data was sent to it
-                    return cb(null);
-                }
-
-                return cb(err);
-            }
-
-            serversSent.push(uuid);
-            job.log.debug(task, 'Server "%s": task', uuid);
-
-            job.params.taskIDs.push({ server_uuid: uuid, task_id: task.id});
-            return cb(null);
-        });
-
-    }, function (err) {
-        if (err) {
-            return callback(err);
-        }
-
-        if (serversSent.length === 0) {
-            return callback(null,
-                    'Only 6.5 servers available: did not send firewall data');
-        }
-
-        return callback(null, 'Started update on servers: '
-            + serversSent.join(', '));
-    });
-}
-
-
 /**
  * Get the server sysinfo objects for the given UUIDs
  *
@@ -219,7 +145,9 @@ function nicUpdate(job, callback) {
  * Poll CNAPI for each of the tasks started in a previous workflow task
  *
  * Inputs:
- * - job.params.taskIDs {Array} : task ID objects as per fwUpdate()
+ * - job.params.taskIDs {Array}: task objects to poll, where each task object
+ *   looks like:
+ *     { server_uuid: uuid, task_id: id }
  *
  * Outputs:
  * - job.params.taskSuccesses {Array} : task ID objects of successes
@@ -280,7 +208,8 @@ function pollTasks(job, callback) {
  * Trigger a sysinfo refresh in CNAPI
  *
  * Inputs:
- * - job.params.taskIDs {Array} : task ID objects as per fwUpdate()
+ * - job.params.server_uuid {String}: UUID of the server whose sysinfo should be
+ *   refreshed.
  *
  * Outputs:
  * - job.params.taskSuccesses {Array} : task ID objects of successes
@@ -304,7 +233,6 @@ function refreshServerSysinfo(job, callback) {
 
 
 module.exports = {
-    fwUpdate: fwUpdate,
     sysinfoList: sysinfoList,
     nicUpdate: nicUpdate,
     pollTasks: pollTasks,
-- 
2.21.0

