commit 88c1ef82a7f8ee8d1a5693343470939a4ffce72c (refs/changes/52/4152/2)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2018-06-07T21:10:01+00:00 (1 year, 4 months ago)
    
    MANATEE-400 Manatee deleterious when dataset will not unmount

diff --git a/boot/setup.sh b/boot/setup.sh
index a1b3ec1..926ae76 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -41,9 +41,6 @@ function manta_manatee_setup {
     manta_setup_manatee_env
 
     PARENT_DATASET=zones/$(zonename)/data
-    DATASET=$PARENT_DATASET/manatee
-    DATASET_MOUNT_DIR=/manatee/pg
-    PG_DIR=/manatee/pg/data
     PG_LOG_DIR=/var/pg
     ZONE_IP=`ifconfig net0 | nawk '{if ($1 == "inet") print $2}'`
     SITTER_CFG_FILE=/opt/smartdc/manatee/etc/sitter.json
@@ -99,13 +96,6 @@ function common_manatee_setup {
     echo "setting recordsize to 8K on manatee dataset"
     zfs set recordsize=8k "$PARENT_DATASET"
 
-    # create manatee dataset
-    echo "creating manatee dataset"
-    zfs create -o canmount=noauto $DATASET
-
-    echo "make snapdir property match the ancestor's"
-    zfs inherit -r snapdir $DATASET
-
     # create postgres group
     echo "creating postgres group (gid=907)"
     groupadd -g 907 postgres
@@ -121,23 +111,6 @@ function common_manatee_setup {
     echo "grant postgres user zfs perms"
     zfs allow -ld postgres create,destroy,diff,hold,release,rename,setuid,rollback,share,snapshot,mount,promote,send,receive,clone,mountpoint,canmount $PARENT_DATASET
 
-    # change dataset perms such that manatee can access the dataset and mount/unmount
-    mkdir -p $DATASET_MOUNT_DIR
-    chown -R postgres $DATASET_MOUNT_DIR
-    chmod 700 -R $DATASET_MOUNT_DIR
-
-    # set mountpoint
-    zfs set mountpoint=$DATASET_MOUNT_DIR $DATASET
-
-    # mount the dataset
-    zfs mount $DATASET
-
-    # make the pg data dir
-    echo "creating $PG_DIR"
-    mkdir -p $PG_DIR
-    chown postgres $PG_DIR
-    chmod 700 $PG_DIR
-
     # add pg log dir
     mkdir -p $PG_LOG_DIR
     chown -R postgres $PG_LOG_DIR
diff --git a/package.json b/package.json
index ed8f2ea..7176fe3 100644
--- a/package.json
+++ b/package.json
@@ -13,7 +13,7 @@
     },
     "dependencies": {
         "manta": "1.2.6",
-        "manatee": "git+https://github.com/joyent/manatee.git#50f8c783",
+        "manatee": "git+https://github.com/joyent/manatee.git#MANATEE-400",
         "carrier": "0.2.0"
     },
     "sdcDependencies": {
