From e1381054e3766ba6b1d47bc6ddb1b9ccd7813220 Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Fri, 23 Mar 2018 21:30:40 +0000
Subject: [PATCH] TOOLS-2006 sdcadm platform usage dumps core

---
 lib/platform.js | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/platform.js b/lib/platform.js
index dbbb9a4..b6fcdc3 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1950,6 +1950,10 @@ function do_usage(subcmd, opts, args, cb) {
             return cb(err);
         }
 
+        if (rows.length === 0) {
+            return cb();
+        }
+
         if (opts.json) {
             console.log(JSON.stringify(rows, null, 4));
             return cb();
-- 
2.21.0

