commit e1381054e3766ba6b1d47bc6ddb1b9ccd7813220 (refs/changes/25/3725/1)
Author: Rui Loura <rui@joyent.com>
Date:   2018-03-23T21:30:40+00:00 (1 year, 7 months ago)
    
    TOOLS-2006 sdcadm platform usage dumps core

diff --git a/lib/platform.js b/lib/platform.js
index dbbb9a4..b6fcdc3 100644
--- a/lib/platform.js
+++ b/lib/platform.js
@@ -1950,6 +1950,10 @@ function do_usage(subcmd, opts, args, cb) {
             return cb(err);
         }
 
+        if (rows.length === 0) {
+            return cb();
+        }
+
         if (opts.json) {
             console.log(JSON.stringify(rows, null, 4));
             return cb();
