commit e385c3024f071fbb2e3639c2f7678c98c59cb2ab (refs/changes/36/3036/1)
Author: Jonathan Perkin <jperkin@joyent.com>
Date:   2017-12-05T14:39:22+00:00 (1 year, 10 months ago)
    
    OS-6486 ctfconvert -i never converts

diff --git a/usr/src/lib/libctf/common/ctf_convert.c b/usr/src/lib/libctf/common/ctf_convert.c
index 1a433d17db..c3b1a004ee 100644
--- a/usr/src/lib/libctf/common/ctf_convert.c
+++ b/usr/src/lib/libctf/common/ctf_convert.c
@@ -66,7 +66,8 @@ ctf_convert_ftypes(Elf *elf, ctf_convert_source_t *types)
 	if ((strdata = elf_getdata(strscn, NULL)) == NULL)
 		return;
 
-	for (i = 0; i < shdr.sh_size / shdr.sh_entsize; i++) {
+	/* Skip past the entry at index 1 which is the output filename */
+	for (i = 2; i < shdr.sh_size / shdr.sh_entsize; i++) {
 		GElf_Sym sym;
 		const char *file;
 		size_t len;
