From f926678196b41b64ea76b116fea33fee04e74a81 Mon Sep 17 00:00:00 2001
From: dyep49 <dyep49@gmail.com>
Date: Thu, 2 Aug 2018 15:45:48 -0700
Subject: [PATCH] TRITON-660 Add metricPorts metadata to CNAPI for cmon-agent

---
 boot/setup.sh | 5 ++++-
 package.json  | 2 +-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index f1b46ed..4ea6f7a 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -7,7 +7,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
@@ -40,6 +40,9 @@ sdc_log_rotation_add registrar /var/svc/log/*registrar*.log 1g
 sdc_log_rotation_add $role /var/svc/log/*$role*.log 1g
 sdc_log_rotation_setup_end
 
+# Add metadata for cmon-agent discovery
+mdata-put metricPorts 8881
+
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
 
diff --git a/package.json b/package.json
index 3d7a2f3..771ed66 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.11.3",
+  "version": "1.11.4",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
-- 
2.21.0

