commit 1fedbe0282f46b38b30868098306769e88912719 (refs/changes/38/838/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-02T13:54:57-04:00 (2 years, 11 months ago)
    
    joyent/node-manta#280 honor options with job expressions

diff --git a/bin/mjob b/bin/mjob
index d08dfa7..e252c56 100755
--- a/bin/mjob
+++ b/bin/mjob
@@ -633,7 +633,10 @@ MJob.prototype.do_create = function create(subcmd, opts, args, cb) {
         function addPhase() {
             var p = {
                 exec: exec,
+                disk: opts.disk,
                 image: opts.image,
+                init: opts.init,
+                memory: opts.memory,
                 type: type
             };
 
