From 1fedbe0282f46b38b30868098306769e88912719 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Wed, 2 Nov 2016 13:54:57 -0400
Subject: [PATCH] joyent/node-manta#280 honor options with job expressions

---
 bin/mjob | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/bin/mjob b/bin/mjob
index d08dfa7..e252c56 100755
--- a/bin/mjob
+++ b/bin/mjob
@@ -633,7 +633,10 @@ MJob.prototype.do_create = function create(subcmd, opts, args, cb) {
         function addPhase() {
             var p = {
                 exec: exec,
+                disk: opts.disk,
                 image: opts.image,
+                init: opts.init,
+                memory: opts.memory,
                 type: type
             };
 
-- 
2.21.0

