commit c24bbe6db20ca408c5f69effa4bca1d45cad6407 (refs/changes/52/2952/3)
Author: Jordan Hendricks <jhendricks@joyent.com>
Date:   2017-11-16T00:56:57+00:00 (1 year, 11 months ago)
    
    joyent/mdb_v8#97 tst.postmortem_jsstack.js fails for node 6.11.2

diff --git a/test/standalone/tst.postmortem_jsstack.js b/test/standalone/tst.postmortem_jsstack.js
index 59de5c6..6eb0b81 100644
--- a/test/standalone/tst.postmortem_jsstack.js
+++ b/test/standalone/tst.postmortem_jsstack.js
@@ -18,6 +18,7 @@ var getRuntimeVersions = require('../lib/runtime-versions').getRuntimeVersions;
 
 var RUNTIME_VERSIONS = getRuntimeVersions();
 var V8_VERSION = RUNTIME_VERSIONS.V8;
+var NODE_VERSION = RUNTIME_VERSIONS.node;
 
 /*
  * Some functions to create a recognizable stack.
@@ -114,6 +115,18 @@ dtrace.on('exit', function (code) {
 		for (var i = 0; i < lines.length; i++) {
 			var line = lines[i];
 
+			/*
+			 * Some later versions of node, beginning with v6, have
+			 * an additional JS stack frame for os.loadavg().
+			 * Ignore this frame.
+			 */
+			if ((line.indexOf(sentinel) !== -1) &&
+				(line.indexOf('loadavg') !== -1)) {
+				assert.ok(NODE_VERSION.major >= 6, 'loadavg ' +
+					'frame found but not expected');
+				continue;
+			}
+
 			if (matched == 1 && line.indexOf(arg1) === 0) {
 				straddr =
 				    line.substr(arg1.length).split(' ')[0];
