From 3aabc08c60d5b274ec22a1254ce2703194d04880 Mon Sep 17 00:00:00 2001
From: Jordan Hendricks <jhendricks@joyent.com>
Date: Thu, 16 Nov 2017 00:38:01 +0000
Subject: [PATCH] joyent/mdb_v8#97 tst.postmortem_jsstack.js fails for node
 6.11.2

---
 test/standalone/tst.postmortem_jsstack.js | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/test/standalone/tst.postmortem_jsstack.js b/test/standalone/tst.postmortem_jsstack.js
index 59de5c6..2640d6d 100644
--- a/test/standalone/tst.postmortem_jsstack.js
+++ b/test/standalone/tst.postmortem_jsstack.js
@@ -18,11 +18,13 @@ var getRuntimeVersions = require('../lib/runtime-versions').getRuntimeVersions;
 
 var RUNTIME_VERSIONS = getRuntimeVersions();
 var V8_VERSION = RUNTIME_VERSIONS.V8;
+var NODE_VERSION = RUNTIME_VERSIONS.node;
 
 /*
  * Some functions to create a recognizable stack.
  */
 var frames = [ 'stalloogle', 'bagnoogle', 'doogle' ];
+
 var expected;
 
 var stalloogle = function (str) {
@@ -114,6 +116,17 @@ dtrace.on('exit', function (code) {
 		for (var i = 0; i < lines.length; i++) {
 			var line = lines[i];
 
+			/*
+			 * Some later versions of node, beginning with version 6, have an
+			 * additional JS stack frame for os.loadavg(). Ignore this frame.
+			 */
+			if ((line.indexOf(sentinel) !== -1) &&
+				(line.indexOf('loadavg') !== -1)) {
+				assert.ok(NODE_VERSION.major >= 6, 'loadavg ' +
+					'frame found but not expected');
+				continue;
+			}
+
 			if (matched == 1 && line.indexOf(arg1) === 0) {
 				straddr =
 				    line.substr(arg1.length).split(' ')[0];
-- 
2.21.0

