commit 38584c27707d53d35aaf7ed3788471c1fe589968 (refs/changes/06/4906/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-10-01T18:57:15-04:00 (1 year ago)
    
    OS-7273 vminfo manpage should be in gitignore

diff --git a/.gitignore b/.gitignore
index 24bf3ef3..ccb2a5ec 100644
--- a/.gitignore
+++ b/.gitignore
@@ -68,6 +68,7 @@ man/man/man1m/sysinfo.1m
 man/man/man1m/diskinfo.1m
 man/man/man1m/disklayout.1m
 man/man/man1m/mkzpool.1m
+man/man/man1m/vminfo.1m
 man/man/man5/fwrule.5
 src/cryptpass
 src/diskinfo
