From 38584c27707d53d35aaf7ed3788471c1fe589968 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Mon, 1 Oct 2018 18:57:15 -0400
Subject: [PATCH] OS-7273 vminfo manpage should be in gitignore

---
 .gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/.gitignore b/.gitignore
index 24bf3ef3..ccb2a5ec 100644
--- a/.gitignore
+++ b/.gitignore
@@ -68,6 +68,7 @@ man/man/man1m/sysinfo.1m
 man/man/man1m/diskinfo.1m
 man/man/man1m/disklayout.1m
 man/man/man1m/mkzpool.1m
+man/man/man1m/vminfo.1m
 man/man/man5/fwrule.5
 src/cryptpass
 src/diskinfo
-- 
2.21.0

