From 2a2732272946bc495758a6b37c3bda3874f6fb6c Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Thu, 25 Jul 2019 14:02:28 +0000
Subject: [PATCH] TRITON-1838 headnode build should insist on answer-file

---
 bin/build-usb-image | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bin/build-usb-image b/bin/build-usb-image
index 4b53ec7c..fc31775c 100755
--- a/bin/build-usb-image
+++ b/bin/build-usb-image
@@ -113,11 +113,11 @@ function copy_config {
         # doc'd for a long time.
         ANSWER_FILE=$(build_spec answers-file)
     fi
-    if [[ -n ${ANSWER_FILE} && -f ${ANSWER_FILE} ]]; then
+    if [[ -n ${ANSWER_FILE} ]]; then
         # BASHSTYLED
         echo "WARNING: Copying in ${ANSWER_FILE}, remove answer-file from build.spec* to disable."
         mkdir -p ${MNT_DIR}/private
-        cp ${ANSWER_FILE} ${MNT_DIR}/private/answers.json
+        cp ${ANSWER_FILE} ${MNT_DIR}/private/answers.json || exit 1
     fi
 }
 
-- 
2.21.0

