commit c3478a208288b82f09b48963004fd618cf3976bf (refs/changes/56/1456/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2017-02-07T08:02:57+00:00 (2 years, 8 months ago)
    
    MANTA-3145 mackerel log records should include job name

diff --git a/lib/meter.js b/lib/meter.js
index c2b3ced..0ae1e79 100644
--- a/lib/meter.js
+++ b/lib/meter.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var mod_assert = require('assert-plus');
@@ -63,7 +63,14 @@ function meter(opts, cb) {
                 date: date
         });
 
-        opts.log.info(jobConfig, 'Job configured.');
+        /*
+         * Include the generated job name in log messages so that we can
+         * more easily correlate related, concurrent activity.
+         */
+        mod_assert.string(jobConfig.job.name, 'jobConfig.job.name');
+        var log = opts.log.child({ jobName: jobConfig.job.name });
+
+        log.info(jobConfig, 'Job configured.');
 
         var assets = generateAssetMap({
                 job: jobConfig.job,
@@ -72,7 +79,7 @@ function meter(opts, cb) {
                 overrides: opts.config.assetOverrides
         });
 
-        opts.log.info(assets, 'Asset map generated.');
+        log.info(assets, 'Asset map generated.');
 
         var keygen = require(__dirname + '/keygen/' + jobConfig.keygen).keygen({
                 client: client,
@@ -85,17 +92,17 @@ function meter(opts, cb) {
                 assets: assets,
                 jobManifest: jobConfig.job,
                 keygen: keygen,
-                log: opts.log,
+                log: log,
                 client: client,
                 monitorBackoff: opts.config.monitorBackoff
         }, function jobDone(err, result) {
                 if (err) {
                         cb(err);
-                        opts.log.fatal(err, 'job error');
+                        log.fatal(err, 'job error');
                         client.close();
                         return;
                 }
-                opts.log.info('Done job');
+                log.info('Done job');
                 if (result.outputs.length && jobConfig.linkPath) {
                         if (jobConfig.linkPath[0] !== '/') {
                                 jobConfig.linkPath =
@@ -107,12 +114,12 @@ function meter(opts, cb) {
 
                                 client.close();
                                 if (lnerr) {
-                                        opts.log.warn(lnerr,
+                                        log.warn(lnerr,
                                                 'Failed to create link.');
                                         cb(lnerr);
                                         return;
                                 }
-                                opts.log.info('Link created ' +
+                                log.info('Link created ' +
                                         jobConfig.linkPath + ' -> ' +
                                         result.outputs[0]);
                                 cb(null, result);
