commit 99e5f8e237d771e9ed16499c35fecdcc02807b4f (refs/changes/05/305/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-08-16T23:08:52+00:00 (3 years, 2 months ago)
    
    https://devhub.joyent.com/jira/browse/OS-5600

diff --git a/configure b/configure
index d8eb7c39..ffed9f51 100755
--- a/configure
+++ b/configure
@@ -221,7 +221,7 @@ function fetch_closed
 
 function generate_env
 {
-	local nopts="-CimNnt"
+	local nopts="-CimMNnt"
 	local lint debug lprefix
 
 	if [[ -n "$ILLUMOS_ENABLE_LINT" ]]; then
