From 13eadc5508219e8e87635bac597bead4c6839793 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 16 Aug 2016 23:08:52 +0000
Subject: [PATCH] OS-5600 nightly shouldn't run pmodes check

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index d8eb7c39..ffed9f51 100755
--- a/configure
+++ b/configure
@@ -221,7 +221,7 @@ function fetch_closed
 
 function generate_env
 {
-	local nopts="-CimNnt"
+	local nopts="-CimMNnt"
 	local lint debug lprefix
 
 	if [[ -n "$ILLUMOS_ENABLE_LINT" ]]; then
-- 
2.21.0

