commit eafabd6590a9aaa6d40e40ba55b59d983d635d5d (refs/changes/05/305/3)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-08-23T16:17:37+00:00 (3 years, 2 months ago)
    
    OS-5600 nightly shouldn't run pmodes check

diff --git a/configure b/configure
index d8eb7c39..ffed9f51 100755
--- a/configure
+++ b/configure
@@ -221,7 +221,7 @@ function fetch_closed
 
 function generate_env
 {
-	local nopts="-CimNnt"
+	local nopts="-CimMNnt"
 	local lint debug lprefix
 
 	if [[ -n "$ILLUMOS_ENABLE_LINT" ]]; then
