commit 759c889548d3a11cd4cea6658a26af358e4c109f (refs/changes/35/3035/1)
Author: Jonathan Perkin <jperkin@joyent.com>
Date:   2017-12-05T11:00:34+00:00 (1 year, 10 months ago)
    
    OS-6458 elfedit failure with large stripped file

diff --git a/usr/src/cmd/sgs/elfedit/common/elfedit_machelf.c b/usr/src/cmd/sgs/elfedit/common/elfedit_machelf.c
index 73653cd6e2..60e25c46b1 100644
--- a/usr/src/cmd/sgs/elfedit/common/elfedit_machelf.c
+++ b/usr/src/cmd/sgs/elfedit/common/elfedit_machelf.c
@@ -276,17 +276,20 @@ elfedit32_init_obj_state(const char *file, int fd, Elf *elf)
 		switch (_cache->sec_shdr->sh_type) {
 		case SHT_SYMTAB_SHNDX:
 			symtab = get_symtab(obj_state, _cache);
-			symtab->symt_xshndx = ndx;
+			if (symtab)
+				symtab->symt_xshndx = ndx;
 			break;
 
 		case SHT_SUNW_syminfo:
 			symtab = get_symtab(obj_state, _cache);
-			symtab->symt_syminfo = ndx;
+			if (symtab)
+				symtab->symt_syminfo = ndx;
 			break;
 
 		case SHT_SUNW_versym:
 			symtab = get_symtab(obj_state, _cache);
-			symtab->symt_versym = ndx;
+			if (symtab)
+				symtab->symt_versym = ndx;
 			break;
 		}
 	}
