commit 32c83956c39093e4c8c214fb12ca7d435b426e41 (refs/changes/89/3289/2)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-02-01T16:15:53-05:00 (1 year, 8 months ago)
    
    TOOLS-1957 sdcadm updateGzTools should limit the number of backups

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 0adc4d8..e99d26d 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 /*
@@ -1523,12 +1523,33 @@ SdcAdm.prototype.updateGzTools = function updateGzTools(options, callback) {
                 common.execFilePlus({argv: argv, log: self.log}, next);
             },
 
+            // keep a finite number cn_tools backups on the usb key
+            function removeOldCNToolsTarballs(_, next) {
+                var backupPath = '/usbkey/extra/joysetup/';
+                var tarballs = fs.readdirSync(backupPath).filter(
+                    function isCNTools(p) {
+                        return (p.startsWith('cn_tools.') &&
+                                p.endsWith('tar.gz') &&
+                                p !== 'cn_tools.tar.gz');
+                    });
+                tarballs.sort();
+                var toDelete = tarballs.slice(4);
+                if (toDelete) {
+                    progress('Removing old cn backups:' + toDelete.join(', '));
+                    toDelete.forEach(function rmBall(fname) {
+                        fs.unlinkSync(path.join(backupPath, fname));
+                    });
+                }
+                next();
+            },
+
             function backupCNToolsFile(_, next) {
                 if (!fs.existsSync('/usbkey/extra/joysetup/cn_tools.tar.gz')) {
                     next();
                     return;
                 }
-                var cnToolsTimestamp = new Date().toISOString();
+                var cnToolsTimestamp = new Date().toISOString()
+                    .split('.')[0].replace(/[:.-]/g, '');
                 fs.rename('/usbkey/extra/joysetup/cn_tools.tar.gz',
                       '/usbkey/extra/joysetup/cn_tools.' + cnToolsTimestamp +
                       '.tar.gz', function (err) {
