From e721ebe8657b5ba36bb97332167b78238e7f124c Mon Sep 17 00:00:00 2001
From: Yang Yong <yy1111.yang@samsung.com>
Date: Wed, 30 Nov 2016 10:40:34 +0900
Subject: [PATCH] joyent/node-triton#143 triton rbac key duplicate output

---
 lib/do_rbac/do_key.js | 7 -------
 1 file changed, 7 deletions(-)

diff --git a/lib/do_rbac/do_key.js b/lib/do_rbac/do_key.js
index c1ca23d..dbff62c 100644
--- a/lib/do_rbac/do_key.js
+++ b/lib/do_rbac/do_key.js
@@ -167,13 +167,6 @@ function _addUserKey(opts, cb) {
                     next(err);
                     return;
                 }
-                if (userKey.name) {
-                    console.log('Added user %s key "%s" (%s)',
-                        opts.userId, userKey.name, userKey.fingerprint);
-                } else {
-                    console.log('Added user %s key %s',
-                        opts.userId, userKey.fingerprint);
-                }
                 var extra = '';
                 if (userKey.name) {
                     extra = format(' (%s)', userKey.name);
-- 
2.21.0

