From 487b756c59b1204ae20e8629126ba2d8dde0dd79 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Mon, 10 Jun 2019 16:55:25 -0700
Subject: [PATCH] TRITON-1719 sdc-migrate abort fails if there's no destination
 CN

---
 bin/sdc-migrate | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/bin/sdc-migrate b/bin/sdc-migrate
index 81eb2a6..6e7f923 100755
--- a/bin/sdc-migrate
+++ b/bin/sdc-migrate
@@ -527,7 +527,17 @@ EOF
 
     local url="/vms/${vm_uuid}?action=migrate&migration_action=abort"
 
-    job_uuid=$(sdc-vmapi "$url" -X POST | $JSON -H job_uuid message)
+    local result=$(sdc-vmapi "$url" -X POST)
+
+    # A 200 status code means the migration was aborted, but it didn't need
+    # a job to delete the target instance (as their was no target instance).
+    local status_code=$(echo "$result" | grep '^HTTP' | head -1 | awk '{print $2}')
+    if [[ $status_code == "200" ]]; then
+        echo "Migration was successfully aborted"
+        return 0
+    fi
+
+    local job_uuid=$($JSON -H job_uuid message <<< "$result")
 
     check_job_uuid "${job_uuid}"
     migration_watch "${vm_uuid}"
-- 
2.21.0

