commit e54647aef85c8d948bfaeab05ea1133c635b9ab9 (refs/changes/04/3504/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-02-27T19:06:27+00:00 (1 year, 7 months ago)
    
    TRITON-180 Move "sdc" zone to new Triton origin image

diff --git a/Makefile b/Makefile
index c6e5f7d..4de5364 100644
--- a/Makefile
+++ b/Makefile
@@ -25,11 +25,11 @@ JSSTYLE_FLAGS	 = -f tools/jsstyle.conf
 RONNJS		 = $(NODE) ./node_modules/.bin/ronn
 PERCENT		:= %
 
-NODE_PREBUILT_VERSION=v0.10.29
 ifeq ($(shell uname -s),SunOS)
+	# Allow building on a SmartOS image other than sdc-*-multiarch 15.4.1.
+	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
+	NODE_PREBUILT_VERSION=v0.10.48
 	NODE_PREBUILT_TAG=zone
-	# Allow building on a SmartOS image other than sdc-smartos@1.6.3.
-	NODE_PREBUILT_IMAGE=fd2cc906-8938-11e3-beab-4359c665ac99
 endif
 
 
