commit 6cb22a1e5c2bc4f2a96317bf476d2442997e932a
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-15T15:07:31-04:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index d9d5697..615e3ce 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -3,7 +3,7 @@
 	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/restdown"]
 	path = deps/restdown
 	url = https://github.com/trentm/restdown.git
diff --git a/deps/jsstyle b/deps/jsstyle
index ccb145b..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
