commit c3689ab5c411aa4e594dc34378283b71a4d693e9
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2019-06-19T03:35:00+00:00 (4 months ago)
    
    OS-7847 vminfod does not sort bhyve disks properly

diff --git a/src/vm/node_modules/vmload/vmload-xml.js b/src/vm/node_modules/vmload/vmload-xml.js
index c460c358..84c0ee75 100644
--- a/src/vm/node_modules/vmload/vmload-xml.js
+++ b/src/vm/node_modules/vmload/vmload-xml.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2018, Joyent, Inc. All rights reserved.
+ * Copyright 2019 Joyent, Inc.
  *
  *
  * vmload-xml: loads VM properties from /etc/zones/<zonename>.xml
@@ -148,7 +148,7 @@ function getVmobjXML(data, options, callback)
 
     // sort the disks + nics + filesystems by index
     if (obj.hasOwnProperty('disks')) {
-        indexSort(obj.disks, 'path', /^.*-disk(\d+)$/);
+        indexSort(obj.disks, 'path', /^.*[\-\/]disk(\d+)$/);
     }
     if (obj.hasOwnProperty('nics')) {
         indexSort(obj.nics, 'interface', /^net(\d+)$/);
