From 1fcc5743c47ccdfb35edc18128f28efe977089c1 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Mon, 30 Sep 2019 15:23:41 +0000
Subject: [PATCH] 7661 illumos should provide 64bit perl modules [fix broken
 build]

---
 configure | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/configure b/configure
index a22f37e0..e5ff5dca 100755
--- a/configure
+++ b/configure
@@ -403,6 +403,12 @@ GNUXGETTEXT=/opt/local/bin/xgettext;		export GNUXGETTEXT
 PYTHON=/opt/local/bin/python2.7;		export PYTHON
 BUILDPY3='#';					export BUILDPY3
 BUILDPY3TOOLS='#';				export BUILDPY3TOOLS
+
+# Disable the split 32/64 perl module builds since smartos doesn't exec perl
+# via isaexec
+# export BUILDPERL32='#'
+export BUILDPERL64='#'
+
 #
 # Note that this isn't the strap perl: that only really exists for the benefit
 # of projects/illumos-extra/openssl
-- 
2.21.0

