commit ce99370ec32c59d43e032297bd1d4ee41c5a25d4
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-09-30T15:11:38+00:00 (7 days ago)
    
    7661 illumos should provide 64bit perl modules [fix broken build]

diff --git a/configure b/configure
index a22f37e0..dad4ab65 100755
--- a/configure
+++ b/configure
@@ -403,6 +403,11 @@ GNUXGETTEXT=/opt/local/bin/xgettext;		export GNUXGETTEXT
 PYTHON=/opt/local/bin/python2.7;		export PYTHON
 BUILDPY3='#';					export BUILDPY3
 BUILDPY3TOOLS='#';				export BUILDPY3TOOLS
+
+# Disable the 32/64-bit perl module builds.
+#export BUILDPERL32='#'
+#export BUILDPERL64='#'
+
 #
 # Note that this isn't the strap perl: that only really exists for the benefit
 # of projects/illumos-extra/openssl
