From 60383b28d050ea1f62d129a8ab477ed32c6bf7cc Mon Sep 17 00:00:00 2001
From: Nick Zivkovic <nick.zivkovic@joyent.com>
Date: Tue, 3 Apr 2018 21:38:15 +0000
Subject: [PATCH] TRITON-261 "triton image get" should show requirements.brand

---
 lib/datasets.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/datasets.js b/lib/datasets.js
index 1a0520c..8723ac1 100644
--- a/lib/datasets.js
+++ b/lib/datasets.js
@@ -149,6 +149,9 @@ function translate(req, dataset) {
             obj.requirements.min_memory = dataset.requirements.min_ram;
             obj.requirements.min_ram = dataset.requirements.min_ram;
         }
+        if (dataset.requirements.brand) {
+            obj.requirements.brand = dataset.requirements.brand;
+        }
     }
 
     var fields = {
-- 
2.21.0

