commit 37dfd5c591d64aa42dc39cdb7af6cff8defbc073 (refs/changes/69/3569/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-03-08T21:19:05+00:00 (1 year, 7 months ago)
    
    OS-6747 make clean removes brand hooks

diff --git a/usr/src/lib/brand/bhyve/zone/Makefile b/usr/src/lib/brand/bhyve/zone/Makefile
index f778fda27a..d7e957072c 100644
--- a/usr/src/lib/brand/bhyve/zone/Makefile
+++ b/usr/src/lib/brand/bhyve/zone/Makefile
@@ -20,7 +20,7 @@ PROGS =		attach detach statechange uninstall
 PROG1 =		boot
 PROG2 =		bhhwcompat
 PROGS +=	$(PROG1) $(PROG2)
-CLEANFILES +=	$(PROGS)
+CLEANFILES +=	$(PROG1) $(PROG2)
 
 $(PROG1) :=	LDLIBS += -lnvpair
 $(PROG2) :=	CPPFLAGS = -I$(COMPAT)/freebsd -I$(CONTRIB)/freebsd \
