commit 4720eb78e35d68a9851b155818c75243b09662f4 (refs/changes/92/3492/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-02-27T15:34:07+00:00 (1 year, 7 months ago)
    
    OS-6696 need iasl in tools (fix build)

diff --git a/usr/src/cmd/acpi/iasl/Makefile.iasl b/usr/src/cmd/acpi/iasl/Makefile.iasl
index c4a68e23f7..167b85eabe 100644
--- a/usr/src/cmd/acpi/iasl/Makefile.iasl
+++ b/usr/src/cmd/acpi/iasl/Makefile.iasl
@@ -123,8 +123,6 @@ LFLAGS = -i -s
 
 .KEEP_STATE:
 
-.PARALLEL: $(OBJS) $(INTERMEDIATES)
-
 all: $(YACC_FILES) .WAIT $(PROG)
 
 $(IASL_OBJ_DIR)/aslcompiler.y: $(IASL_SRC_DIR)/aslparser.y
