From 25d23bdf274f0682e6d4962f9688477bd68981de Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 29 Jul 2019 19:20:30 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index 4443fcd3..01f669da 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 4443fcd38fe1682556e1851df97c7d652d078f3d
+Subproject commit 01f669da1f5cb1439d573fc42efd2dc43a7c6184
-- 
2.21.0

