commit 2292ef40a7ccacfba72beb4265471a1423ecb5c9 (refs/changes/34/4934/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-10-08T16:59:04-07:00 (1 year ago)
    
    MANTA-3983 want support for "sudo" operators group

diff --git a/lib/auth.js b/lib/auth.js
index 2675f3b..0718bce 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -776,6 +776,53 @@ function getActiveRoles(req, res, next) {
         requestedRoles = req.headers['role'];
     }
 
+    /*
+     * Handle the special _operator role if the user is a member of the
+     * "role-operators" group (this overrides the regular isOperator
+     * status, if present).
+     */
+    var account = req.caller.account;
+    if (account.groups.indexOf('role-operators') !== -1) {
+        /*
+         * The req.caller object is cached and potentially shared between
+         * multiple requests. We're going to alter the isOperator flag on
+         * req.caller.account on a per-request basis, so we need a per-request
+         * copy of both req.caller and req.caller.account.
+         *
+         * We can keep sharing all the other child objects of req.caller other
+         * than req.caller.account (i.e. we don't have to do a full deep copy),
+         * because we're not changing those.
+         */
+        var newCaller = {};
+        Object.keys(req.caller).forEach(function (k) {
+            newCaller[k] = req.caller[k];
+        });
+        var newAccount = {};
+        Object.keys(account).forEach(function (k) {
+            newAccount[k] = account[k];
+        });
+        newCaller.account = newAccount;
+        req.caller = newCaller;
+
+        /*
+         * Since they're in role-operators, make them always non-operator
+         * unless the Role header is provided.
+         */
+        newAccount.isOperator = false;
+
+        /*
+         * We treat a Role header value of "_operator" basically as a magic
+         * value. If we have it, we skip all further role processing (since
+         * we're just going to authorize this request using our operator
+         * rights anyway).
+         */
+        if (requestedRoles === '_operator') {
+            newAccount.isOperator = true;
+            setImmediate(next);
+            return;
+        }
+    }
+
     var activeRoles = [];
     var names;
 
