commit baf82c140e4076fb74b69fb62c0404a03175777f (refs/changes/34/4934/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-10-08T15:44:57-07:00 (1 year ago)
    
    MANTA-3983 want support for "sudo" operators group

diff --git a/lib/auth.js b/lib/auth.js
index 2675f3b..73ba162 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -776,6 +776,31 @@ function getActiveRoles(req, res, next) {
         requestedRoles = req.headers['role'];
     }
 
+    /*
+     * Handle the special _operator role if the user is a member of the
+     * "role-operators" group (this overrides the regular isOperator
+     * status, if present).
+     */
+    var account = req.caller.account;
+    if (account.groups.indexOf('role-operators') !== -1) {
+        /*
+         * The req.caller object is cached and potentially shared between
+         * multiple requests, so make a copy just for this request now.
+         */
+        var newAccount = {};
+        req.caller.account = newAccount;
+        Object.keys(account).forEach(function (k) {
+            newAccount[k] = account[k];
+        });
+        newAccount.isOperator = false;
+
+        if (requestedRoles === '_operator') {
+            newAccount.isOperator = true;
+            setImmediate(next);
+            return;
+        }
+    }
+
     var activeRoles = [];
     var names;
 
