commit 6a092b5e19863f3363b89e16c2796ba8f4ca8bc2
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2019-04-10T15:59:11+00:00 (6 months ago)
    
    OS-7702 add zfs mmp tests to smartos run profile

diff --git a/usr/src/test/zfs-tests/include/libtest.shlib b/usr/src/test/zfs-tests/include/libtest.shlib
index 39a60a0900..85aeab10f4 100644
--- a/usr/src/test/zfs-tests/include/libtest.shlib
+++ b/usr/src/test/zfs-tests/include/libtest.shlib
@@ -307,6 +307,13 @@ function default_mirror_setup
 	log_pass
 }
 
+function default_mirror_2way_setup
+{
+	default_mirror_setup_noexit $1 $2
+
+	log_pass
+}
+
 #
 # Given a pair of disks, set up a storage pool and dataset for the mirror
 # @parameters: $1 the primary side of the mirror
@@ -323,7 +330,7 @@ function default_mirror_setup_noexit
 	[[ -z $secondary ]] && \
 		log_fail "$func: No secondary partition passed"
 	[[ -d /$TESTPOOL ]] && rm -rf /$TESTPOOL
-	log_must zpool create -f $TESTPOOL mirror $primary $secondary
+	log_must zpool create -f $TESTPOOL mirror $@
 	log_must zfs create $TESTPOOL/$TESTFS
 	log_must zfs set mountpoint=$TESTDIR $TESTPOOL/$TESTFS
 }
diff --git a/usr/src/test/zfs-tests/runfiles/smartos.run b/usr/src/test/zfs-tests/runfiles/smartos.run
index 1155c29025..49e1bd623c 100644
--- a/usr/src/test/zfs-tests/runfiles/smartos.run
+++ b/usr/src/test/zfs-tests/runfiles/smartos.run
@@ -158,6 +158,11 @@ tests = ['migration_001_pos', 'migration_002_pos', 'migration_003_pos',
 [/opt/zfs-tests/tests/functional/mmap]
 tests = ['mmap_read_001_pos', 'mmap_write_001_pos']
 
+[/opt/zfs-tests/tests/functional/mmp]
+tests = ['mmp_on_thread', 'mmp_on_off', 'mmp_interval',
+    'mmp_active_import', 'mmp_inactive_import', 'mmp_exported_import',
+    'mmp_on_zdb']
+
 [/opt/zfs-tests/tests/functional/mount]
 tests = ['umount_001', 'umountall_001']
 
diff --git a/usr/src/test/zfs-tests/tests/functional/cli_root/zpool_replace/setup.ksh b/usr/src/test/zfs-tests/tests/functional/cli_root/zpool_replace/setup.ksh
index 83337a8949..9a45a363bf 100644
--- a/usr/src/test/zfs-tests/tests/functional/cli_root/zpool_replace/setup.ksh
+++ b/usr/src/test/zfs-tests/tests/functional/cli_root/zpool_replace/setup.ksh
@@ -36,4 +36,4 @@ verify_disk_count "$DISKS" 3
 
 DISK=${DISKS%% *}
 
-default_mirror_setup $DISKS
+default_mirror_2way_setup $DISKS
diff --git a/usr/src/test/zfs-tests/tests/functional/mmp/mmp.cfg b/usr/src/test/zfs-tests/tests/functional/mmp/mmp.cfg
index 52680c275a..4fd94641f1 100644
--- a/usr/src/test/zfs-tests/tests/functional/mmp/mmp.cfg
+++ b/usr/src/test/zfs-tests/tests/functional/mmp/mmp.cfg
@@ -15,10 +15,11 @@
 
 #
 # Copyright (c) 2017 by Lawrence Livermore National Security, LLC.
+# Copyright 2019 Joyent, Inc.
 #
 
-export PREV_UBER="$TEST_BASE_DIR/mmp-uber-prev.txt"
-export CURR_UBER="$TEST_BASE_DIR/mmp-uber-curr.txt"
+export PREV_UBER="/var/tmp/zfs-test/mmp-uber-prev.txt"
+export CURR_UBER="/var/tmp/zfs-test/mmp-uber-curr.txt"
 export DISK=${DISKS%% *}
 
 export HOSTID_FILE="/etc/hostid"
@@ -29,7 +30,7 @@ export TXG_TIMEOUT_LONG=5000
 export TXG_TIMEOUT_DEFAULT=5
 
 export MMP_POOL=mmppool
-export MMP_DIR=$TEST_BASE_DIR/mmp
+export MMP_DIR=/var/tmp/zfs-test/mmp
 export MMP_CACHE=$MMP_DIR/zpool.cache
 export MMP_ZTEST_LOG=$MMP_DIR/ztest.log
 export MMP_HISTORY=100
