commit 4e0e254f2c932f57969398815a8b95763c671fde (refs/changes/92/892/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2016-11-14T08:58:26-07:00 (2 years, 11 months ago)
    
    HEAD-2329 hook cnapi changes for RFD 53 into joysetup

diff --git a/scripts/joysetup.sh b/scripts/joysetup.sh
index 1c0d6b0c..81e7ffb0 100755
--- a/scripts/joysetup.sh
+++ b/scripts/joysetup.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2014, Joyent, Inc.
+# Copyright 2016, Joyent, Inc.
 #
 
 #
@@ -622,7 +622,18 @@ if [[ "$(zpool list)" == "no pools available" ]] \
         POOL_FILE=/mockcn/${MOCKCN_SERVER_UUID}/pool.json
     fi
 
-    if ! /usr/bin/disklayout "${arg_disklayout}" > ${POOL_FILE}; then
+    args_disklayout=""
+    if [[ -n "${CONFIG_spares}" ]]; then
+        args_disklayout="${args_disklayout} -s ${CONFIG_spares}"
+    fi
+    if [[ -n "${CONFIG_cache}" && "${CONFIG_cache}" == "false" ]]; then
+        args_disklayout="${args_disklayout} -c"
+    fi
+    if [[ -n "${CONFIG_layout}" ]]; then
+        args_disklayout="${args_disklayout} ${CONFIG_layout}"
+    fi
+
+    if ! /usr/bin/disklayout ${args_disklayout} > ${POOL_FILE}; then
         fatal "disk layout failed"
     fi
 
