commit 2d0594a0d1d9a76590df1d74566900b434f69b27 (refs/changes/18/5118/2)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-11-23T08:48:08-08:00 (11 months ago)
    
    TRITON-937 CNAPI should have ServerInstallAgentShar endpoint

diff --git a/lib/endpoints/servers.js b/lib/endpoints/servers.js
index 1c32028..a1c00cb 100644
--- a/lib/endpoints/servers.js
+++ b/lib/endpoints/servers.js
@@ -1137,6 +1137,37 @@ Server.installAgent = function handlerServerInstallAgent(req, res, next) {
         }});
 };
 
+/* BEGIN JSSTYLED */
+/**
+ * Instruct server to install given agents shar. Pass in the name of an
+ * agentsshar package to install and server will download the shar and
+ * install.
+ *
+ * @name ServerInstallAgentsShar
+ * @endpoint POST /servers/:server_uuid/install-agentsshar
+ * @section Server API
+ * @param {String} url URL of agentsshar to install
+ *
+ * @response 200 Ok Install task initiated successfully
+ * @response 500 Error Could not process request
+ */
+/* END JSSTYLED */
+
+Server.installAgentsShar =
+function handlerServerInstallAgentsShar(req, res, next) {
+    var self = this;
+
+    req.stash.server.sendTaskRequest({
+        task: 'agent_install_shar',
+        params: req.params,
+        req: req,
+        evcb: ModelServer.createComputeNodeAgentHandler(self, req.params.jobid),
+        cb: function (error, task) {
+            res.send({ id: task.id });
+            next();
+            return;
+        }});
+};
 
 Server.nop = function handlerServerNop(req, res, next) {
     req.params.sleep = parseInt(req.params.sleep, 10) || 0;
@@ -1248,6 +1279,19 @@ function attachTo(http, app) {
         }),
         Server.installAgent);
 
+    // Import an image to the server
+    http.post({
+        path: '/servers/:server_uuid/install-agent-shar',
+        name: 'ServerInstallAgentsShar' },
+        ensure({
+            connectionTimeoutSeconds: 60 * 60,
+            app: app,
+            serverRunning: true,
+            prepopulate: ['server'],
+            connected: ['moray']
+        }),
+        Server.installAgentsShar);
+
     // Refresh server sysinfo
     http.post({
         path: '/servers/:server_uuid/sysinfo-refresh',
diff --git a/package.json b/package.json
index b3196c5..69889d0 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.16.0",
+  "version": "1.17.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
