From 2bd758db22e8167d5f1c97ea0a98826159c3cc28 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 4 Mar 2019 14:13:27 -0500
Subject: [PATCH] TOOLS-2193 TOP Makefile.defs assignment complicates
 overriding

---
 tools/mk/Makefile.defs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/mk/Makefile.defs b/tools/mk/Makefile.defs
index e2e1642..6ee0d6e 100644
--- a/tools/mk/Makefile.defs
+++ b/tools/mk/Makefile.defs
@@ -53,7 +53,7 @@
 # destinations, but sometimes to modify the behavior of the build.
 #
 
-TOP := $(shell pwd)
+TOP ?= $(shell pwd)
 
 #
 # Mountain Gorilla-spec'd versioning.
-- 
2.21.0

