commit df9278c01497d252fcf657d4d5a4c009c8fbf3b4
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T11:46:19-05:00 (7 months ago)
    
    TOOLS-2193 TOP Makefile.defs assignment complicates overriding

diff --git a/tools/mk/Makefile.defs b/tools/mk/Makefile.defs
index fa3569e..8bd6ac9 100644
--- a/tools/mk/Makefile.defs
+++ b/tools/mk/Makefile.defs
@@ -53,7 +53,7 @@
 # destinations, but sometimes to modify the behavior of the build.
 #
 
-TOP := $(shell pwd)
+TOP ?= $(shell pwd)
 
 #
 # Mountain Gorilla-spec'd versioning.
