commit c1289b38c9a85ddcca641f9ccbaa69928ba67592
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-15T15:07:44-04:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index d65f40b..da3ef7d 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -3,4 +3,4 @@
 	url = git://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
diff --git a/deps/jsstyle b/deps/jsstyle
index b94eb62..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit b94eb629ee524d21b563c8b4461caa2bdf1becc6
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
