commit c5b5e26fcb3a6e88b8d9c2bcbe679c47ff5b47f0
Author: bowrocker <jon.anderson@joyent.com>
Date:   2019-08-08T13:36:38+00:00 (9 weeks ago)
    
    MANTA-4477 error building repos that rely on cueball

diff --git a/src/client.rs b/src/client.rs
index 59c64fc..18a97bd 100644
--- a/src/client.rs
+++ b/src/client.rs
@@ -51,6 +51,8 @@ impl MorayClient {
                     claim_timeout: Some(5000),
                     log: log,
                     rebalancer_action_delay: None,
+		    decoherence_interval: None,
+                    decoherence_delay: None,
                 }
             },
             Some(opts) => opts
