commit 1d05ebf28e65e5ac10f62601a1ea65f088883bb4
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T14:46:27-05:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/Makefile b/Makefile
index 0588b32..0399cce 100644
--- a/Makefile
+++ b/Makefile
@@ -43,7 +43,7 @@ JSON_FILES  := package.json config.json.sample
 ifeq ($(shell uname -s),SunOS)
 	# sdc-*-multiarch 15.4.1.
 	NODE_PREBUILT_IMAGE=18b094b0-eb01-11e5-80c1-175dac7ddf02
-	NODE_PREBUILT_VERSION=v6.15.1
+	NODE_PREBUILT_VERSION=v6.17.0
 	NODE_PREBUILT_TAG := zone
 endif
 
