From 8dccea34f933876e22a53ca8d8e0434f1ac178e9 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 1 Apr 2019 16:47:31 +0100
Subject: [PATCH] TOOLS-2227 bits-upload IMAGEFILE regex greedily eats dots

---
 tools/bits-upload.sh | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/tools/bits-upload.sh b/tools/bits-upload.sh
index 135591b..7096cb7 100755
--- a/tools/bits-upload.sh
+++ b/tools/bits-upload.sh
@@ -299,23 +299,25 @@ function publish_to_updates {
         set -o errexit
 
         MF=${file}
-        IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.zfs.gz/g')
+        MF_PREFIX=$(echo ${MF} | \
+            sed -e 's/\.imgmanifest$//g' -e 's/\.manifest$//g')
+        IMAGEFILE=${MF_PREFIX}.zfs.gz
         compression_flag=""
 
         # Some payloads are not zfs-based, look for likely alternatives.
         # This assumes that a single directory with <file>.manifest
         # contains only one of [<file>.zfs.gz, <file>.sh, <file>.tgz,
-	# <file>.tar.gz]
+        # <file>.tar.gz]
         if [[ ! -f "${IMAGEFILE}" ]]; then
-            IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.sh/g')
+            IMAGEFILE=${MF_PREFIX}.sh
             compression_flag="--compression=none"
         fi
         if [[ ! -f "${IMAGEFILE}" ]]; then
-            IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.tgz/g')
+            IMAGEFILE=${MF_PREFIX}.tgz
             compression_flag=""
         fi
         if [[ ! -f "${IMAGEFILE}" ]]; then
-            IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.tar.gz/g')
+            IMAGEFILE=${MF_PREFIX}.tar.gz
             compression_flag=""
         fi
 
-- 
2.21.0

