From 16321d33d718fd8714155de34703a831ca9ef863 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Mon, 29 Aug 2016 12:30:25 -0700
Subject: [PATCH] AGENT-1033: Remove /proc mounting from cn-agent chroot-gtar
 script

---
 bin/chroot-gtar | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/bin/chroot-gtar b/bin/chroot-gtar
index 4f6c525..2cd66f1 100755
--- a/bin/chroot-gtar
+++ b/bin/chroot-gtar
@@ -175,10 +175,8 @@ for lib in $libs; do
 done
 
 # Setup minimum fake filesystem.
-TO_UMOUNT="$TO_UMOUNT $CHROOTDIR/proc $CHROOTDIR/dev/fd"
-TO_REMOVE="$TO_REMOVE $CHROOTDIR/proc $CHROOTDIR/dev/fd"
-mkdir -p $CHROOTDIR/proc
-mount -F proc proc $CHROOTDIR/proc
+TO_UMOUNT="$TO_UMOUNT $CHROOTDIR/dev/fd"
+TO_REMOVE="$TO_REMOVE $CHROOTDIR/dev/fd"
 [[ ! -e $CHROOTDIR/dev ]] && TO_REMOVE="$TO_REMOVE $CHROOTDIR/dev"
 mkdir -p $CHROOTDIR/dev/fd
 mount -F fd fd $CHROOTDIR/dev/fd
-- 
2.21.0

