commit 0e57a585de3b49cbc8327ea60feba87cdd934124
Author: Dave Eddy <dave@daveeddy.com>
Date:   2019-07-12T13:10:32-04:00 (3 months ago)
    
    MANTA-4355 Expose pagination for bucket and object listing

diff --git a/lib/buckets.js b/lib/buckets.js
index 74269aa..5fae450 100644
--- a/lib/buckets.js
+++ b/lib/buckets.js
@@ -273,14 +273,12 @@ function deleteBucketNoVnode(rpcctx, owner, bucket, req_id, callback) {
 /*
  * This function talks to boray
  */
-function listBuckets(rpcctx, owner, order_by, prefix, limit, offset, vnode,
-    req_id) {
+function listBuckets(rpcctx, owner, prefix, limit, marker, vnode, req_id) {
     assert.object(rpcctx, 'rpcctx');
     assert.string(owner, 'owner');
-    assert.string(order_by, 'order_by');
-    assert.string(prefix, 'prefix');
+    assert.optionalString(prefix, 'prefix');
     assert.number(limit, 'limit');
-    assert.number(offset, 'offset');
+    assert.optionalString(marker, 'marker');
     assert.number(vnode, 'vnode');
     assert.string(req_id, 'req_id');
 
@@ -289,10 +287,9 @@ function listBuckets(rpcctx, owner, order_by, prefix, limit, offset, vnode,
 
     var arg = {
         owner: owner,
-        order_by: order_by,
         prefix: prefix,
         limit: limit,
-        offset: offset,
+        marker: marker,
         vnode: vnode,
         request_id: req_id
     };
@@ -324,20 +321,28 @@ function listBuckets(rpcctx, owner, order_by, prefix, limit, offset, vnode,
 /*
  * This function talks to electric-boray
  */
-function listBucketsNoVnode(rpcctx, owner, sorted, order_by, prefix, limit,
+function listBucketsNoVnode(rpcctx, owner, prefix, limit, marker, delimiter,
     req_id) {
+
     assert.object(rpcctx, 'rpcctx');
     assert.string(owner, 'owner');
-    assert.bool(sorted, 'sorted');
-    assert.string(order_by, 'order_by');
-    assert.string(prefix, 'prefix');
+    assert.optionalString(prefix, 'prefix');
     assert.number(limit, 'limit');
+    assert.optionalString(marker, 'marker');
+    assert.optionalString(delimiter, 'delimiter');
     assert.string(req_id, 'req_id');
 
     var opts = {};
     opts.req_id = req_id;
 
-    var arg = [owner, sorted, order_by, prefix, limit, req_id];
+    var arg = [
+        owner,
+        prefix,
+        limit,
+        marker,
+        delimiter,
+        req_id
+    ];
 
     var log = rpc.childLogger(rpcctx, opts);
 
diff --git a/lib/client.js b/lib/client.js
index 329d511..966cda0 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -630,16 +630,14 @@ BorayClient.prototype.getBucketNoVnode =
  * Lists buckets for an account at a specific virtual node
  *
  * @param {String} owner     - Account owner
- * @param {String} order_by  - An ordering clause for the resulting buckets
  * @param {String} prefix    - A prefix to use to group buckets
  * @param {Number} limit     - The maximum number of buckets to return
- * @param {Number} offset    - An starting offset into the buckets set
+ * @param {Number} marker    - An optional string to start listing from
  * @param {Number} vnode     - Virtual node identifier
  * @param {String} req_id    - Request identifier
  */
 BorayClient.prototype.listBuckets =
-    function listBuckets(owner, order_by, prefix, limit, offset, vnode,
-        req_id) {
+    function listBuckets(owner, prefix, limit, marker, vnode, req_id) {
 
     var rpcctx = this.ctxCreateForEmitter();
     var rv;
@@ -648,8 +646,8 @@ BorayClient.prototype.listBuckets =
         return (emitUnavailable());
     }
 
-    rv = buckets.listBuckets(rpcctx, owner, order_by, prefix, limit, offset,
-        vnode, req_id);
+    rv = buckets.listBuckets(rpcctx, owner, prefix, limit, marker, vnode,
+        req_id);
     this.releaseWhenDone(rpcctx, rv);
 
     return (rv);
@@ -659,15 +657,16 @@ BorayClient.prototype.listBuckets =
  * Lists buckets for an account
  *
  * @param {String} owner     - Account owner
- * @param {Boolean} sorted   - Indicates if the results should be sorted
- * @param {String} order_by  - An ordering clause for the resulting buckets
  * @param {String} prefix    - A prefix to use to group buckets
  * @param {Number} limit     - The maximum number of buckets to return
+ * @param {String} marker    - An optional string to start listing from
+ * @param {String} delimiter - An optional string to group listings by
  * @param {String} req_id    - Request identifier
  */
 BorayClient.prototype.listBucketsNoVnode =
-    function listBucketsNoVnode(owner, sorted, order_by, prefix, limit,
+    function listBucketsNoVnode(owner, prefix, limit, marker, delimiter,
         req_id) {
+
     var rpcctx = this.ctxCreateForEmitter();
     var rv;
 
@@ -675,8 +674,8 @@ BorayClient.prototype.listBucketsNoVnode =
         return (emitUnavailable());
     }
 
-    rv = buckets.listBucketsNoVnode(rpcctx, owner, sorted, order_by, prefix,
-        limit, req_id);
+    rv = buckets.listBucketsNoVnode(rpcctx, owner, prefix, limit, marker,
+        delimiter, req_id);
     this.releaseWhenDone(rpcctx, rv);
 
     return (rv);
@@ -977,14 +976,17 @@ BorayClient.prototype.getObjectNoVnode =
  * Lists objects for the bucket belonging to the owner account at a particular
  * virtual node.
  *
- * @param {String} owner           - Account owner
- * @param {String} bucket_id       - Bucket id
- * @param {Number} vnode           - Virtual node identifier
- * @param {String} req_id          - Request identifier
+ * @param {String} owner      - Account owner
+ * @param {String} bucket_id  - Bucket id
+ * @param {String} prefix     - A prefix to use to group buckets
+ * @param {Number} limit      - The maximum number of buckets to return
+ * @param {String} marker     - An optional string to start listing from
+ * @param {Number} vnode      - Virtual node identifier
+ * @param {String} req_id     - Request identifier
  */
 BorayClient.prototype.listObjects =
-    function listObjects(owner, bucket_id, order_by, prefix, limit, offset,
-        vnode, req_id) {
+    function listObjects(owner, bucket_id, prefix, limit, marker, vnode,
+        req_id) {
 
     var rpcctx = this.ctxCreateForEmitter();
     var rv;
@@ -993,8 +995,8 @@ BorayClient.prototype.listObjects =
         return (emitUnavailable());
     }
 
-    rv = objects.listObjects(rpcctx, owner, bucket_id, order_by, prefix, limit,
-        offset, vnode, req_id);
+    rv = objects.listObjects(rpcctx, owner, bucket_id, prefix, limit,
+        marker, vnode, req_id);
     this.releaseWhenDone(rpcctx, rv);
 
     return (rv);
@@ -1003,13 +1005,17 @@ BorayClient.prototype.listObjects =
 /**
  * Lists objects for the bucket belonging to the owner account.
  *
- * @param {String} owner           - Account owner
- * @param {String} bucket_id       - Bucket id
- * @param {String} req_id          - Request identifier
+ * @param {String} owner       - Account owner
+ * @param {String} bucket_id   - Bucket id
+ * @param {String} prefix      - A prefix to use to group buckets
+ * @param {Number} limit       - The maximum number of buckets to return
+ * @param {String} marker      - An optional string to start listing from
+ * @param {String} delimiter   - An optional string to group listings by
+ * @param {String} req_id      - Request identifier
  */
 BorayClient.prototype.listObjectsNoVnode =
-    function listObjectsNoVnode(owner, bucket_id, sorted, order_by, prefix,
-        limit, req_id) {
+    function listObjectsNoVnode(owner, bucket_id, prefix, limit, marker,
+        delimiter, req_id) {
 
     var rpcctx = this.ctxCreateForEmitter();
     var rv;
@@ -1018,8 +1024,8 @@ BorayClient.prototype.listObjectsNoVnode =
         return (emitUnavailable());
     }
 
-    rv = objects.listObjectsNoVnode(rpcctx, owner, bucket_id, sorted, order_by,
-        prefix, limit, req_id);
+    rv = objects.listObjectsNoVnode(rpcctx, owner, bucket_id, prefix, limit,
+        marker, delimiter, req_id);
     this.releaseWhenDone(rpcctx, rv);
 
     return (rv);
diff --git a/lib/objects.js b/lib/objects.js
index 0dcd7c5..0bdf770 100644
--- a/lib/objects.js
+++ b/lib/objects.js
@@ -401,26 +401,24 @@ function deleteObjectNoVnode(rpcctx, owner, bucket_id, name, req_id, callback) {
 }
 
 
-function listObjects(rpcctx, owner, bucket_id, order_by, prefix, limit, offset,
-    vnode, req_id) {
+function listObjects(rpcctx, owner, bucket_id, prefix, limit, marker, vnode,
+    req_id) {
 
     assert.object(rpcctx, 'rpcctx');
     assert.string(owner, 'owner');
     assert.string(bucket_id, 'bucket_id');
-    assert.string(order_by, 'order_by');
-    assert.string(prefix, 'prefix');
+    assert.optionalString(prefix, 'prefix');
     assert.number(limit, 'limit');
-    assert.number(offset, 'offset');
+    assert.optionalString(marker, 'marker');
     assert.number(vnode, 'vnode');
     assert.string(req_id, 'req_id');
 
     var arg = {
         owner: owner,
         bucket_id: bucket_id,
-        order_by: order_by,
         prefix: prefix,
         limit: limit,
-        offset: offset,
+        marker: marker,
         vnode: vnode,
         request_id: req_id
     };
@@ -452,17 +450,18 @@ function listObjects(rpcctx, owner, bucket_id, order_by, prefix, limit, offset,
     return (res);
 }
 
-function listObjectsNoVnode(rpcctx, owner, bucket_id, sorted, order_by, prefix,
-    limit, req_id) {
+function listObjectsNoVnode(rpcctx, owner, bucket_id, prefix, limit, marker,
+    delimiter, req_id) {
+
     var log;
 
     assert.object(rpcctx, 'rpcctx');
     assert.string(owner, 'owner');
     assert.string(bucket_id, 'bucket_id');
-    assert.bool(sorted, 'sorted');
-    assert.string(order_by, 'order_by');
-    assert.string(prefix, 'prefix');
+    assert.optionalString(prefix, 'prefix');
     assert.number(limit, 'limit');
+    assert.optionalString(marker, 'marker');
+    assert.optionalString(delimiter, 'delimiter');
     assert.string(req_id, 'req_id');
 
     var opts = {};
@@ -472,10 +471,20 @@ function listObjectsNoVnode(rpcctx, owner, bucket_id, sorted, order_by, prefix,
 
     var res = new EventEmitter();
 
+    var arg = [
+        owner,
+        bucket_id,
+        prefix,
+        limit,
+        marker,
+        delimiter,
+        req_id
+    ];
+
     var req = rpc.rpcCommon({
         rpcctx: rpcctx,
         rpcmethod: 'listobjects',
-        rpcargs: [owner, bucket_id, sorted, order_by, prefix, limit, req_id],
+        rpcargs: arg,
         log: log
     }, function (err) {
         if (err) {
