From a8760bdbb4be64be4e30494d5a9c3906af18351c Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 24 Oct 2019 17:13:16 +0100
Subject: [PATCH] 11844 add rdmsr utility (fix mismerge)

---
 usr/src/man/man1m/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/usr/src/man/man1m/Makefile b/usr/src/man/man1m/Makefile
index eae46cdb4c..6dcd74a84f 100644
--- a/usr/src/man/man1m/Makefile
+++ b/usr/src/man/man1m/Makefile
@@ -547,7 +547,7 @@ i386_MANFILES=				\
 		acpidump.1m		\
 		acpixtract.1m		\
 		nvmeadm.1m		\
-		pptadm.1m
+		pptadm.1m		\
 		rdmsr.1m
 
 sparc_MANFILES=	dcs.1m			\
-- 
2.17.2 (Apple Git-113)

