commit 8a825aaccdfd1c9f30e653b4c1ac4e3d725aa984
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-07-09T15:15:07-04:00 (3 months ago)
    
    TRITON-1805 volapi over enthusastic about compiling sdcnode

diff --git a/Makefile b/Makefile
index bd7b531..4341345 100644
--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,8 @@ ifeq ($(shell uname -s),SunOS)
 	include ./deps/eng/tools/mk/Makefile.node_prebuilt.defs
 	include ./deps/eng/tools/mk/Makefile.agent_prebuilt.defs
 else
-	include ./deps/eng/tools/mk/Makefile.node.defs
+	NPM_EXEC :=
+	NPM = npm
 endif
 include ./deps/eng/tools/mk/Makefile.smf.defs
 
