commit 1a739a5f80eafdf516ab33e23a3ad0ac51293a4f
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-04-09T17:25:59+00:00 (6 months ago)
    
    OS-7722 mdb crashes on invalid ::walk softstate

diff --git a/usr/src/cmd/mdb/common/modules/genunix/devinfo.c b/usr/src/cmd/mdb/common/modules/genunix/devinfo.c
index 081129a2ce..282d24f537 100644
--- a/usr/src/cmd/mdb/common/modules/genunix/devinfo.c
+++ b/usr/src/cmd/mdb/common/modules/genunix/devinfo.c
@@ -20,7 +20,7 @@
  */
 /*
  * Copyright (c) 2000, 2010, Oracle and/or its affiliates. All rights reserved.
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -1578,6 +1578,16 @@ soft_state_walk_init(mdb_walk_state_t *wsp)
 		return (WALK_ERR);
 	}
 
+	if (sst->ssw_ss.size == 0) {
+		mdb_warn("read invalid softstate: softstate ent size is "
+		    "zero\n");
+		return (WALK_ERR);
+	}
+
+	if (sst->ssw_ss.n_items == 0) {
+		mdb_warn("read invalid softstate: softstate has no entries\n");
+		return (WALK_ERR);
+	}
 
 	/* Read array of pointers to state structs into local storage. */
 	sst->ssw_pointers = mdb_alloc((sst->ssw_ss.n_items * sizeof (void *)),
