commit c851c63cb0f56d03ec46b6e89c30417556906da2 (refs/changes/95/3995/2)
Author: Robert Bogart <robert.bogart@joyent.com>
Date:   2018-05-23T22:43:38+00:00 (1 year, 5 months ago)
    
    MANTA-3596 binder could export metrics via artedi

diff --git a/lib/client.js b/lib/client.js
index d0d1b91..bbe2fa7 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -10,6 +10,7 @@ module.exports = { Client: ZKClient };
 
 const mod_fsm = require('mooremachine');
 const mod_assert = require('assert-plus');
+const mod_artedi = require('artedi');
 const mod_consts = require('./zk-consts');
 const mod_util = require('util');
 const mod_stream = require('stream');
@@ -40,6 +41,18 @@ function ZKClient(opts) {
 		});
 	}
 
+	/*
+	 * If the caller did not supply a collector for tracking metrics, then
+	 * create one internally.  At least for now, it is not a requirement
+	 * for all consumers to provide one.
+	 */
+	mod_assert.optionalObject(opts.collector);
+	if (opts.collector === undefined || opts.collector === null) {
+		this.zc_collector = mod_artedi.createCollector();
+	} else {
+		this.zc_collector = opts.collector;
+	}
+
 	mod_assert.optionalArrayOfObject(opts.servers, 'options.servers');
 	if (opts.servers === undefined || opts.servers === null) {
 		mod_assert.string(opts.address, 'options.address');
@@ -168,7 +181,8 @@ ZKClient.prototype._newSession = function () {
 		return;
 	var s = new mod_zks.ZKSession({
 		timeout: this.zc_sessionTimeout,
-		log: this.zc_log
+		log: this.zc_log,
+		collector: this.zc_collector
 	});
 	this.zc_session = s;
 	function initialHandler(st) {
diff --git a/lib/zk-session.js b/lib/zk-session.js
index a3489d5..abe4563 100644
--- a/lib/zk-session.js
+++ b/lib/zk-session.js
@@ -22,10 +22,14 @@ const mod_events = require('events');
 const mod_bunyan = require('bunyan');
 const mod_verror = require('verror');
 
+var METRIC_ZK_SESSION_COUNTER = 'zookeeper_session_count';
+var METRIC_ZK_NOTIFICATION_COUNTER = 'zookeeper_notifications';
+
 function ZKSession(opts) {
 	mod_assert.object(opts, 'options');
 	mod_assert.number(opts.timeout, 'options.timeout');
 	mod_assert.object(opts.log, 'options.log');
+	mod_assert.object(opts.collector, 'options.collector');
 
 	this.zs_conn = undefined;
 	this.zs_oldConn = undefined;
@@ -37,11 +41,39 @@ function ZKSession(opts) {
 	this.zs_log = opts.log.child({
 		component: 'ZKSession'
 	});
+	this.zs_collector = opts.collector;
 
 	this.zs_lastZxid = new mod_jsbn.BigInteger('0');
 	this.zs_sessionId = new mod_jsbn.BigInteger('0');
 	this.zs_passwd = new Buffer(8).fill(0);
 
+	/*
+	 * If a counter for tracking ZooKeeper notifications does not yet exist
+	 * then that must mean that this is the first session.  Create it now.
+	 */
+	if (this.zs_collector.getCollector(
+	    METRIC_ZK_NOTIFICATION_COUNTER) === null) {
+		this.zs_collector.counter({
+			name: METRIC_ZK_NOTIFICATION_COUNTER,
+			help: 'Notifications received from ZooKeeper'
+		});
+	}
+
+	/*
+	 * The name of this metric can be misleading as the client is
+	 * monogomous.  That is, at any point in time, we only ever have one
+	 * session with a ZooKeeper server.  This is intended to track the total
+	 * number of times that we have had to establish (or re-establish) a
+	 * session with a ZooKeeper server.
+	 */
+	if (this.zs_collector.getCollector(
+	    METRIC_ZK_SESSION_COUNTER) === null) {
+		this.zs_collector.counter({
+			name: METRIC_ZK_SESSION_COUNTER,
+			help: 'Total number of sessions created'
+		});
+	}
+
 	mod_fsm.FSM.call(this, 'detached');
 }
 mod_util.inherits(ZKSession, mod_fsm.FSM);
@@ -158,6 +190,10 @@ ZKSession.prototype.state_attaching = function (S) {
 		self.zs_sessionId = sid;
 		self.zs_passwd = pkt.passwd;
 		self.resetExpiryTimer();
+
+		var session_counter = self.zs_collector.getCollector(
+		    METRIC_ZK_SESSION_COUNTER);
+		session_counter.increment();
 		S.gotoState('attached');
 	});
 
@@ -351,6 +387,11 @@ ZKSession.prototype.processNotification = function (pkt) {
 		zxid: pkt.zxid.toString('hex'),
 		type: pkt.type
 	}, 'notification %s for %s', evt, pkt.path);
+
+	var notifications = this.zs_collector.getCollector(
+	    METRIC_ZK_NOTIFICATION_COUNTER);
+	notifications.increment({event: evt});
+
 	if (watcher)
 		watcher.emit(evt);
 };
diff --git a/package.json b/package.json
index 23daeef..1b970ec 100644
--- a/package.json
+++ b/package.json
@@ -8,6 +8,7 @@
   },
   "main": "lib/index.js",
   "dependencies": {
+    "artedi": "1.1.1",
     "assert-plus": "^1.0.0",
     "bunyan": "^1.8.10",
     "cueball": "^2.2.7",
