From 3eae7167b5b7a2c8c61cdb2de194c9933eb3ebbd Mon Sep 17 00:00:00 2001
From: "Joshua M. Clulow" <jmc@joyent.com>
Date: Fri, 10 Feb 2017 07:19:26 +0000
Subject: [PATCH] MANTA-3157 muskie picker ought to start even when there is no
 space

---
 lib/picker.js | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/lib/picker.js b/lib/picker.js
index 271fdbc..101a7b2 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -242,8 +242,16 @@ function poll() {
                 count++;
             });
 
-            // Don't replace if we got an empty "DB"
-            if (count > 0) {
+            /*
+             * In general, we don't want to replace a non-empty set of sharks
+             * with an empty one.  We make an exception at startup: provided
+             * the query to Moray did not result in any errors, we accept one
+             * empty result set.  This enables us to service requests that will
+             * not ask the Picker for a storage node, while remaining confident
+             * that once storage nodes become available, we'll know about them.
+             */
+            if (count > 0 || self.firstTopology) {
+                self.firstTopology = false;
                 self.datacenters = dcs;
                 self.db = obj;
                 self.emit('topology', self.db);
@@ -352,6 +360,8 @@ function Picker(opts) {
     this.utilization = opts.maxUtilizationPct;
 
     this.client.once('connect', poll.bind(this));
+
+    this.firstTopology = true;
     this.once('topology', this.emit.bind(this, 'connect'));
 }
 util.inherits(Picker, EventEmitter);
-- 
2.21.0

