From c4d77df6a613b73ce1f29e78dd52ac675556eac9 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 21 Nov 2018 18:21:30 +0000
Subject: [PATCH] MANTA-4024 madtom doesn't use its nodeunit dependency

---
 Makefile     | 11 ++++-------
 package.json |  3 ++-
 2 files changed, 6 insertions(+), 8 deletions(-)

diff --git a/Makefile b/Makefile
index 27176bc..9be8549 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -25,7 +25,6 @@
 #
 # Tools
 #
-NODEUNIT        := ./node_modules/.bin/nodeunit
 NPM             := npm
 
 #
@@ -74,17 +73,15 @@ NPM_ENV		 = MAKE_OVERRIDES="CTFCONVERT=/bin/true CTFMERGE=/bin/true"
 # Repo-specific targets
 #
 .PHONY: all
-all: $(SMF_MANIFESTS) | $(NODEUNIT) $(REPO_DEPS) scripts
-	$(NPM) install
-$(NODEUNIT): | $(NPM_EXEC)
+all: $(SMF_MANIFESTS) | $(REPO_DEPS) scripts
 	$(NPM) install
 
 CLEAN_FILES += $(NODEUNIT) ./node_modules/nodeunit
 
 .PHONY: test
-test: $(NODEUNIT)
+test:
 	mkdir -p ./tmp
-	find test/ -name '*.test.js' | xargs -t -L1 ctrun -o noorphan node
+	find test/ -name '*.test.js' | xargs -t -L1 ctrun -o noorphan $(NODE_EXEC)
 
 .PHONY: scripts
 scripts: deps/manta-scripts/.git
diff --git a/package.json b/package.json
index 3992d81..93a91bf 100644
--- a/package.json
+++ b/package.json
@@ -16,8 +16,9 @@
                 "sdc-clients": "~9.1",
                 "vasync": "1.5.0"
         },
+        "//": "One of our modules has a '*' dependency on 'debug' that uses syntax incompatible with our prebuilt node version. Force it to an older version.",
         "devDependencies": {
-                "nodeunit": "0.7.4"
+                "debug": "3.1.0"
         },
         "engines": {
                 "node": "0.8.18"
-- 
2.21.0

