From 4e0568e70ce19b836634a9484fc51c60c084a985 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 21 Nov 2018 17:10:14 +0000
Subject: [PATCH] MANTA-4024 madtom's nodeunit dep cannot install with old node

---
 Makefile     | 4 ++--
 package.json | 3 ++-
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index 27176bc..a84604e 100644
--- a/Makefile
+++ b/Makefile
@@ -5,7 +5,7 @@
 #
 
 #
-# Copyright (c) 2017, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 #
@@ -84,7 +84,7 @@ CLEAN_FILES += $(NODEUNIT) ./node_modules/nodeunit
 .PHONY: test
 test: $(NODEUNIT)
 	mkdir -p ./tmp
-	find test/ -name '*.test.js' | xargs -t -L1 ctrun -o noorphan node
+	find test/ -name '*.test.js' | xargs -t -L1 ctrun -o noorphan $(NODE_EXEC)
 
 .PHONY: scripts
 scripts: deps/manta-scripts/.git
diff --git a/package.json b/package.json
index 3992d81..4024aa4 100644
--- a/package.json
+++ b/package.json
@@ -17,7 +17,8 @@
                 "vasync": "1.5.0"
         },
         "devDependencies": {
-                "nodeunit": "0.7.4"
+                "nodeunit": "0.10.1",
+                "debug": "3.1.0"
         },
         "engines": {
                 "node": "0.8.18"
-- 
2.21.0

