commit bc725baaf70865c43161e71dc3f4aa7e3f2bb5e5
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2019-03-19T16:07:27+01:00 (7 months ago)
    
    OS-7672 recursive mutex enter in ahci

diff --git a/usr/src/uts/common/io/sata/adapters/ahci/ahci.c b/usr/src/uts/common/io/sata/adapters/ahci/ahci.c
index 258e1d742c..202e07b351 100644
--- a/usr/src/uts/common/io/sata/adapters/ahci/ahci.c
+++ b/usr/src/uts/common/io/sata/adapters/ahci/ahci.c
@@ -82,6 +82,7 @@
  */
 
 #include <sys/note.h>
+#include <sys/debug.h>
 #include <sys/scsi/scsi.h>
 #include <sys/pci.h>
 #include <sys/disp.h>
@@ -10847,9 +10848,8 @@ ahci_em_quiesce(ahci_ctl_t *ahci_ctlp)
 {
 	ASSERT(ahci_ctlp->ahcictl_em_flags & AHCI_EM_PRESENT);
 
-	mutex_enter(&ahci_ctlp->ahcictl_mutex);
+	VERIFY(mutex_owned(&ahci_ctlp->ahcictl_mutex));
 	ahci_ctlp->ahcictl_em_flags |= AHCI_EM_QUIESCE;
-	mutex_exit(&ahci_ctlp->ahcictl_mutex);
 
 	ddi_taskq_wait(ahci_ctlp->ahcictl_em_taskq);
 }
@@ -10859,9 +10859,8 @@ ahci_em_suspend(ahci_ctl_t *ahci_ctlp)
 {
 	ahci_em_quiesce(ahci_ctlp);
 
-	mutex_enter(&ahci_ctlp->ahcictl_mutex);
+	VERIFY(mutex_owned(&ahci_ctlp->ahcictl_mutex));
 	ahci_ctlp->ahcictl_em_flags &= ~AHCI_EM_READY;
-	mutex_exit(&ahci_ctlp->ahcictl_mutex);
 }
 
 static void
@@ -10882,7 +10881,10 @@ ahci_em_fini(ahci_ctl_t *ahci_ctlp)
 		return;
 	}
 
+	mutex_enter(&ahci_ctlp->ahcictl_mutex);
 	ahci_em_quiesce(ahci_ctlp);
+	mutex_exit(&ahci_ctlp->ahcictl_mutex);
+
 	ddi_taskq_destroy(ahci_ctlp->ahcictl_em_taskq);
 	ahci_ctlp->ahcictl_em_taskq = NULL;
 }
