commit 18780e073519b501156d0e87bcbc73fe55f3b658 (refs/changes/70/2570/1)
Author: Mark Brooks <mark.brooks@joyent.com>
Date:   2017-09-14T13:18:22-04:00 (2 years, 1 month ago)
    
    OS-6283 Fix disable option summary in svcadm(1m)

diff --git a/usr/src/man/man1m/svcadm.1m b/usr/src/man/man1m/svcadm.1m
index 1d647304b2..3dd7439e6c 100644
--- a/usr/src/man/man1m/svcadm.1m
+++ b/usr/src/man/man1m/svcadm.1m
@@ -15,7 +15,7 @@ svcadm \- manipulate service instances
 
 .LP
 .nf
-\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-s\fR]
+\fB/usr/sbin/svcadm\fR [\fB-S state\fR] [\fB-v\fR] [\fB-Z\fR | \fB-z\fR \fIzone\fR] disable [\fB-st\fR]
      [{\fIFMRI\fR | \fIpattern\fR}...]
 .fi
 
