commit a237676972c9e1de1afe549895748f1426e10c9f
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-05T14:45:31-05:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/Makefile b/Makefile
index 75cce75..afc202c 100644
--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@ CLEAN_FILES += ./node_modules ./build/sdcadm-*.sh ./build/sdcadm-*.imgmanifest .
 TAP_EXEC := ./node_modules/.bin/tap
 TEST_UNIT_JOBS ?= 4
 
-NODE_PREBUILT_VERSION=v6.15.1
+NODE_PREBUILT_VERSION=v6.17.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=gz
 	# sdc-minimal-multiarch-lts 15.4.1
