From 5220813cdbf810c468e13ccb08a030849c8c9625 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Tue, 12 Mar 2019 14:21:57 +0000
Subject: [PATCH] MANTA-4163 improved PG autovacuum msg

---
 deps/postgresql96 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/postgresql96 b/deps/postgresql96
index fda6203..96cfa73 160000
--- a/deps/postgresql96
+++ b/deps/postgresql96
@@ -1 +1 @@
-Subproject commit fda6203fca5c8a4bd3dda2dc725cf02134e8725e
+Subproject commit 96cfa7397d8dc0a9db3f8083ef4ad399be88fc0f
-- 
2.21.0

