commit e8cd2c857a28088ea92e19da47ed68629b3cf2aa (refs/changes/20/4120/4)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-06-05T19:33:39+02:00 (1 year, 4 months ago)
    
    TRITON-399 sdc-clients: remove SAPI's history

diff --git a/CHANGES.md b/CHANGES.md
index fa5e924..a3d58c0 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -8,6 +8,11 @@
 
 ## not yet released
 
+## 11.0.0
+
+- TRITON-399 Get rid of SAPI's history. Require explicit version option given
+  to SAPI client constructor.
+
 ## 10.5.0
 
 - DOCKER-524 Implement docker push
diff --git a/lib/sapi.js b/lib/sapi.js
index c82ff39..0cdd6f7 100644
--- a/lib/sapi.js
+++ b/lib/sapi.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -33,13 +33,16 @@ var RestifyClient = require('./restifyclient');
  * Constructor
  *
  * See the RestifyClient constructor for details
+ *
+ * Note that we're explictly requiring a version from applications
+ * using the client library now. Those can obviously pass '*' as the
+ * version of choice, but it would be desirable to use proper versions
+ * instead.
  */
 function SAPI(options) {
     assert.object(options, 'options');
     assert.object(options.log, 'options.log');
-    if (!options.version) {
-        options.version = '*';
-    }
+    assert.string(options.version, 'options.version');
 
     this.log = options.log;
 
@@ -490,86 +493,6 @@ SAPI.prototype.getPayload = function getPayload(uuid, callback) {
 };
 
 
-// --- History
-
-/**
- * Add history record
- *
- * @param {Function} callback: of the form f(err, history).
- */
-function addHistory(opts, callback) {
-    assert.object(opts, 'opts');
-    assert.string(opts.uuid, 'opts.uuid');
-    assert.number(opts.started, 'opts.started');
-    assert.object(opts.changes, 'opts.changes');
-
-    return (this.post('/history', opts, callback));
-}
-
-SAPI.prototype.addHistory = addHistory;
-
-
-/**
- * Lists all history records
- *
- * @param {Function} callback: of the form f(err, history).
- */
-function listHistory(search_opts, callback) {
-    if (arguments.length === 1) {
-        callback = search_opts;
-        search_opts = {};
-    }
-
-    var uri = '/history?' + qs.stringify(search_opts);
-    return (this.get(uri, callback));
-}
-
-SAPI.prototype.listHistory = listHistory;
-
-
-/**
- * Gets an history record by UUID
- *
- * @param {String} uuid: the UUID of the history record.
- * @param {Function} callback: of the form f(err, history).
- */
-function getHistory(uuid, callback) {
-    return (this.get(sprintf('/history/%s', uuid), callback));
-}
-
-SAPI.prototype.getHistory = getHistory;
-
-
-/**
- * Updates an existing history record
- *
- * @param {String} uuid: the UUID of the history item to update.
- * @param {String} opts: new attributes
- * @param {Function} callback: of the form f(err, history).
- */
-function updateHistory(uuid, opts, callback) {
-    assert.string(uuid, 'uuid');
-    assert.object(opts, 'opts');
-    assert.number(opts.finished, 'opts.finished');
-
-    return (this.put(sprintf('/history/%s', uuid), opts, callback));
-}
-
-SAPI.prototype.updateHistory = updateHistory;
-
-
-/**
- * Deletes an history record by UUID
- *
- * @param {String} uuid: the UUID of the history record.
- * @param {Function} callback : of the form f(err).
- */
-function deleteHistory(uuid, callback) {
-    return (this.del(sprintf('/history/%s', uuid), callback));
-}
-
-SAPI.prototype.deleteHistory = deleteHistory;
-
 module.exports = SAPI;
 
 
diff --git a/package.json b/package.json
index a319985..fd16e8f 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "10.5.0",
+  "version": "11.0.0",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
