commit 33110e222e5498b82d618499e9bc01e9ded502a4 (refs/changes/13/1013/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton key list

diff --git a/lib/do_key/do_list.js b/lib/do_key/do_list.js
index 69f9c4b..b09063d 100644
--- a/lib/do_key/do_list.js
+++ b/lib/do_key/do_list.js
@@ -37,37 +37,42 @@ function do_list(subcmd, opts, args, cb) {
 
     var cli = this.top;
 
-    cli.tritonapi.cloudapi.listKeys({}, function onKeys(err, keys) {
-        if (err) {
-            cb(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            cb(setupErr);
         }
-
-        if (opts.json) {
-            common.jsonStream(keys);
-        } else if (opts.authorized_keys) {
-            keys.forEach(function (key) {
-                console.log(common.chomp(key.key));
-            });
-        } else {
-            var columns = COLUMNS_DEFAULT;
-
-            if (opts.o) {
-                columns = opts.o;
-            } else if (opts.long) {
-                columns = COLUMNS_LONG;
+        cli.tritonapi.cloudapi.listKeys({}, function onKeys(err, keys) {
+            if (err) {
+                cb(err);
+                return;
             }
 
-            columns = columns.split(',');
-            var sort = opts.s.split(',');
-
-            tabula(keys, {
-                skipHeader: false,
-                columns: columns,
-                sort: sort
-            });
-        }
-        cb();
+            if (opts.json) {
+                common.jsonStream(keys);
+            } else if (opts.authorized_keys) {
+                keys.forEach(function (key) {
+                    console.log(common.chomp(key.key));
+                });
+            } else {
+                var columns = COLUMNS_DEFAULT;
+
+                if (opts.o) {
+                    columns = opts.o;
+                } else if (opts.long) {
+                    columns = COLUMNS_LONG;
+                }
+
+                columns = columns.split(',');
+                var sort = opts.s.split(',');
+
+                tabula(keys, {
+                    skipHeader: false,
+                    columns: columns,
+                    sort: sort
+                });
+            }
+            cb();
+        });
     });
 }
 
