From 261bc139226ec5639f354ef73ed100578db39c96 Mon Sep 17 00:00:00 2001
From: Sean Chittenden <sean@chittenden.org>
Date: Fri, 23 Jun 2017 10:47:50 -0700
Subject: [PATCH] Explicitly set the encoding and locale when initdb(1)'ing a
 DB.

---
 lib/postgresMgr.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/postgresMgr.js b/lib/postgresMgr.js
index 620fc9d..a6bc7d6 100644
--- a/lib/postgresMgr.js
+++ b/lib/postgresMgr.js
@@ -1498,7 +1498,7 @@ PostgresMgr.prototype._initDb = function (callback) {
         },
         function _initDb(_, cb) {
             var cmd = 'sudo -u ' + self._dbUser + ' ' +
-                self._pgInitDbPath + ' -D ' + self._dataDir;
+                self._pgInitDbPath + ' --encoding=UTF-8 --locale=C -D ' + self._dataDir;
             log.info({cmd: cmd}, 'PostgresMgr.initDb: initializing db');
             exec(cmd, function (err, stdout, stderr) {
                 // ignore errors since the db could already be initialized
-- 
2.21.0

