commit ae0647008285e51705c568945e7c72a6436a4e70 (refs/changes/46/2146/3)
Author: Sean Chittenden <sean@chittenden.org>
Date:   2017-06-23T12:10:11-07:00 (2 years, 4 months ago)
    
    MANTA-3323: initdb(1) with explicit locale and encoding

diff --git a/lib/postgresMgr.js b/lib/postgresMgr.js
index 620fc9d..d19c1eb 100644
--- a/lib/postgresMgr.js
+++ b/lib/postgresMgr.js
@@ -29,6 +29,7 @@ var fs = require('fs');
 var pg = require('pg');
 var Client = pg.Client;
 var shelljs = require('shelljs');
+var mod_forkexec = require('forkexec');
 var spawn = require('child_process').spawn;
 var exec = require('child_process').exec;
 var once = require('once');
@@ -1497,17 +1498,13 @@ PostgresMgr.prototype._initDb = function (callback) {
 
         },
         function _initDb(_, cb) {
-            var cmd = 'sudo -u ' + self._dbUser + ' ' +
-                self._pgInitDbPath + ' -D ' + self._dataDir;
-            log.info({cmd: cmd}, 'PostgresMgr.initDb: initializing db');
-            exec(cmd, function (err, stdout, stderr) {
+            var args = [ 'sudo', '-u', self._dbUser,
+                         self._pgInitDbPath, '--encoding=UTF-8', '--locale=C',
+                         '-D', self._dataDir ];
+            log.info({ cmd: cmd, arv : args }, 'PostgresMgr.initDb: initializing db');
+            mod_forkexec.forkExecWait({ argv: args }, function (err, info) {
                 // ignore errors since the db could already be initialized
-                log.info({
-                    err: err,
-                    stdout: stdout,
-                    stderr: stderr,
-                    dataDir: self._dataDir
-                }, 'PostgresMgr.initDb: initdb returned');
+                log.info(info, 'PostgresMgr.initDb: initdb returned');
 
                 shelljs.cp('-f', self._hbaConf, self._dataDir + '/pg_hba.conf');
                 log.info({
