From 1e49abe7f3711fa18b42ce346d13134b97e6b311 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Thu, 22 Jun 2017 17:06:16 -0700
Subject: [PATCH] MANTA-3322 want 'severity' column for 'manta-adm alarm list'

---
 lib/adm.js | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/lib/adm.js b/lib/adm.js
index 5876623..1f02814 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -132,7 +132,7 @@ var maAlarmColumns = {
     },
     'dateopened': {
 	'label': 'DATE_OPENED',
-	'width': 10
+	'width': 11
     },
     'timeopened': {
 	'label': 'TIME_OPENED',
@@ -140,7 +140,7 @@ var maAlarmColumns = {
     },
     'dateclosed': {
 	'label': 'DATE_CLOSED',
-	'width': 10
+	'width': 11
     },
     'timeclosed': {
 	'label': 'TIME_CLOSED',
@@ -156,7 +156,7 @@ var maAlarmColumns = {
     },
     'nevents': {
 	'label': 'NEVENTS',
-	'width': 5,
+	'width': 7,
 	'align': 'right'
     },
     'nflts': {
@@ -168,6 +168,10 @@ var maAlarmColumns = {
 	'label': 'NFY',
 	'width': 3
     },
+    'severity': {
+	'label': 'SEVERITY',
+	'width': 8
+    },
     'summary': {
 	'label': 'SUMMARY',
 	'width': 30
@@ -1229,6 +1233,7 @@ maAdm.prototype.alarmsList = function (args)
 		    'NFLTS': alarm.a_faults.length,
 		    'NEVENTS': alarm.a_nevents,
 		    'NFY': alarm.a_suppressed ? 'no' : 'yes',
+		    'SEVERITY': details.ka ? details.ka.ka_severity : '-',
 		    'SUMMARY': details.summary
 		});
 	});
-- 
2.21.0

