From 6107d4d568eda876b7cfe01dca39d36fb8bec354 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Thu, 30 Nov 2017 15:28:00 -0800
Subject: [PATCH] joyent/manta-thoth#163 Publish install/update updates

---
 tools/publish.sh | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tools/publish.sh b/tools/publish.sh
index e4ba56b..6da5e8e 100755
--- a/tools/publish.sh
+++ b/tools/publish.sh
@@ -6,4 +6,7 @@ name="thoth-${uname}-${ver:?}.tar.gz"
 latest="thoth-${uname}-latest.tar.gz"
 
 mput -f "$name" "/thoth/public/$name"
+for c in install update update-all; do
+    mput -f "bin/sdc-thoth-$c" "/thoth/public/sdc-thoth-$c"
+done
 mln "/thoth/public/$name" "/thoth/public/$latest"
-- 
2.21.0

