From 7db9212751ed5cd9ae24b88444e2dfed9262733f Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Sun, 25 Feb 2018 19:16:45 +0000
Subject: [PATCH] OS-6694 bhyve uart emulation does not handle 4 byte requests

---
 usr/src/cmd/bhyve/pci_lpc.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/usr/src/cmd/bhyve/pci_lpc.c b/usr/src/cmd/bhyve/pci_lpc.c
index 2203a00baa..d111b72ca4 100644
--- a/usr/src/cmd/bhyve/pci_lpc.c
+++ b/usr/src/cmd/bhyve/pci_lpc.c
@@ -163,6 +163,19 @@ lpc_uart_io_handler(struct vmctx *ctx, int vcpu, int in, int port, int bytes,
 			uart_write(sc->uart_softc, offset + 1, *eax >> 8);
 		}
 		break;
+	case 4:
+		if (in) {
+			*eax = uart_read(sc->uart_softc, offset);
+			*eax |= uart_read(sc->uart_softc, offset + 1) << 8;
+			*eax |= uart_read(sc->uart_softc, offset + 2) << 16;
+			*eax |= uart_read(sc->uart_softc, offset + 3) << 24;
+		} else {
+			uart_write(sc->uart_softc, offset, *eax);
+			uart_write(sc->uart_softc, offset + 1, *eax >> 8);
+			uart_write(sc->uart_softc, offset + 2, *eax >> 16);
+			uart_write(sc->uart_softc, offset + 3, *eax >> 24);
+		}
+		break;
 	default:
 		return (-1);
 	}
-- 
2.21.0

