commit dfa4030ef61b5d8f8978d91a7a4d8eee53628044 (refs/changes/40/1040/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-12-01T14:56:53-08:00 (2 years, 10 months ago)
    
    OS-5823 'imgadm install' of a type=docker image should work

diff --git a/src/img/CHANGES.md b/src/img/CHANGES.md
index d2b25b90..101bafb0 100644
--- a/src/img/CHANGES.md
+++ b/src/img/CHANGES.md
@@ -6,6 +6,10 @@ Known issues:
   Docker Registry v1, which is now deprecated.
 
 
+## 3.7.1
+
+- OS-5823 Allow 'imgadm install' to be able to install a type=docker image.
+
 ## 3.7.0
 
 - OS-5798 Allow '+' in the image manifest "version" field.
diff --git a/src/img/lib/imgadm.js b/src/img/lib/imgadm.js
index 2763ad4b..8b733c31 100644
--- a/src/img/lib/imgadm.js
+++ b/src/img/lib/imgadm.js
@@ -2340,7 +2340,6 @@ IMGADM.prototype._lockPathFromUuid = function _lockPathFromUuid(uuid) {
 IMGADM.prototype._installDockerImage = function _installDockerImage(ctx, cb) {
     var self = this;
     assert.object(ctx, 'ctx');
-    assert.object(ctx.source, 'ctx.source');
     assert.string(ctx.filePath, 'ctx.filePath');
     assert.string(ctx.dsName, 'ctx.dsName');
     assert.string(ctx.zpool, 'ctx.zpool');
diff --git a/src/img/package.json b/src/img/package.json
index 7cdad364..66b86a93 100644
--- a/src/img/package.json
+++ b/src/img/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgadm",
   "description": "Manage SmartOS virtual machine images.",
-  "version": "3.7.0",
+  "version": "3.7.1",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
