From 86b3777e1a003f346bdcdc01d420a2af8edf0152 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Thu, 27 Sep 2018 12:25:10 -0700
Subject: [PATCH] TRITON-798 CNS needs to support two acme-challenges

---
 lib/vm-to-zones.js | 60 ++++++++++++++++++++++++----------------------
 1 file changed, 32 insertions(+), 28 deletions(-)

diff --git a/lib/vm-to-zones.js b/lib/vm-to-zones.js
index 5fe6d03..e5c81b8 100644
--- a/lib/vm-to-zones.js
+++ b/lib/vm-to-zones.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = buildZonesFromVm;
@@ -186,15 +186,17 @@ function addInstance(zones, vm, ent, config) {
 		var chal = vm.customer_metadata[consts.INST_ACME_KEY];
 		if (chal !== undefined) {
 			name = '_acme-challenge.' + name;
-			if (!zones[ent.zone])
-				zones[ent.zone] = {};
-			if (!zones[ent.zone][name])
-				zones[ent.zone][name] = [];
-			var recs = zones[ent.zone][name];
-			recs.push({
-				constructor: 'TXT',
-				args: [chal]
-			});
+            chal.split(' ').forEach(function(c){
+                if (!zones[ent.zone])
+                    zones[ent.zone] = {};
+                if (!zones[ent.zone][name])
+                    zones[ent.zone][name] = [];
+                var recs = zones[ent.zone][name];
+                recs.push({
+                    constructor: 'TXT',
+                    args: [c]
+                });
+            });
 		}
 	}
 
@@ -308,24 +310,26 @@ function addService(zones, vm, ent, config) {
 			src: vm.uuid
 		});
 	}
-	function addACME(name) {
-		if (vm.customer_metadata === undefined)
-			return;
-		var chal = vm.customer_metadata[consts.INST_ACME_KEY];
-		if (chal !== undefined) {
-			name = '_acme-challenge.' + name;
-			if (!zones[ent.zone])
-				zones[ent.zone] = {};
-			if (!zones[ent.zone][name])
-				zones[ent.zone][name] = [];
-			var recs = zones[ent.zone][name];
-			recs.push({
-				constructor: 'TXT',
-				args: [chal],
-				src: vm.uuid
-			});
-		}
-	}
+    function addACME(name) {
+        if (vm.customer_metadata === undefined)
+            return;
+        var chal = vm.customer_metadata[consts.INST_ACME_KEY];
+        if (chal !== undefined) {
+            name = '_acme-challenge.' + name;
+            chal.split(' ').forEach(function(c){
+                if (!zones[ent.zone])
+                    zones[ent.zone] = {};
+                if (!zones[ent.zone][name])
+                    zones[ent.zone][name] = [];
+                var recs = zones[ent.zone][name];
+                recs.push({
+                    constructor: 'TXT',
+                    args: [c],
+                    src: vm.uuid
+                });
+            });
+        }
+    }
 
 	var doNetworkName = (isWildcard(config, ent.zone) &&
 	    isNetOwned(vm, ent.network));
-- 
2.21.0

