commit 64cc9869e26f7cccbe2bf1fc54398a645c6924d0
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-08-22T20:50:59+01:00 (7 weeks ago)
    
    TOOLS-2247 move sdcnode to manta (fix Makefile)

diff --git a/Makefile b/Makefile
index ae9e2b0..b77f3eb 100644
--- a/Makefile
+++ b/Makefile
@@ -48,7 +48,7 @@ nodes: nodesrc
 
 .PHONY: publish
 publish: prepublish
-	rm -rf $(TOP)/bits
+	rm -rf $(TOP)/bits/sdcnode
 	mkdir -p $(TOP)/bits/sdcnode
 	cp $(TOP)/build/nodes/*/sdcnode-*.tgz $(TOP)/bits/sdcnode
 
