commit 5c1baf965a671402d779a02853f9c282f63f603f
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2019-07-03T21:41:08+00:00 (3 months ago)
    
    OS-7874 gethrtime should use comm page

diff --git a/usr/src/lib/libc/amd64/Makefile b/usr/src/lib/libc/amd64/Makefile
index 579a2b0de7..6789c06377 100644
--- a/usr/src/lib/libc/amd64/Makefile
+++ b/usr/src/lib/libc/amd64/Makefile
@@ -24,7 +24,7 @@
 # Copyright (c) 2013, OmniTI Computer Consulting, Inc. All rights reserved.
 # Copyright 2013 Garrett D'Amore <garrett@damore.org>
 # Copyright 2018 Nexenta Systems, Inc.
-# Copyright (c) 2019, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 # Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
 #
 
@@ -207,7 +207,7 @@ COMSYSOBJS=			\
 	geteuid.o		\
 	getgid.o		\
 	getgroups.o		\
-	gethrtime.o		\
+	gethrvtime.o		\
 	getitimer.o		\
 	getmsg.o		\
 	getpid.o		\
@@ -288,6 +288,7 @@ COMSYSOBJS=			\
 SYSOBJS=			\
 	__clock_gettime.o	\
 	__clock_gettime_sys.o	\
+	__gethrtime_sys.o	\
 	__getcontext.o		\
 	__uadmin.o		\
 	_lwp_mutex_unlock.o	\
@@ -296,6 +297,7 @@ SYSOBJS=			\
 	forkx.o			\
 	forkallx.o		\
 	getcontext.o		\
+	gethrtime.o		\
 	gettimeofday.o		\
 	lwp_private.o		\
 	nuname.o		\
@@ -1189,6 +1191,7 @@ $(PORTI18N_COND:%=pics/%) := \
 pics/arc4random.o :=	CPPFLAGS += -I$(SRC)/common/crypto/chacha
 
 pics/__clock_gettime.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
+pics/gethrtime.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
 pics/gettimeofday.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
 
 .KEEP_STATE:
diff --git a/usr/src/lib/libc/common/sys/gethrtime.s b/usr/src/lib/libc/common/sys/gethrvtime.s
similarity index 78%
rename from usr/src/lib/libc/common/sys/gethrtime.s
rename to usr/src/lib/libc/common/sys/gethrvtime.s
index d66330ba2f..d2669416bc 100644
--- a/usr/src/lib/libc/common/sys/gethrtime.s
+++ b/usr/src/lib/libc/common/sys/gethrvtime.s
@@ -27,25 +27,6 @@
 
 #include "SYS.h"
 
-/*
- * hrtime_t gethrtime(void)
- *
- * Returns the current hi-res real time.
- */
-
-	ENTRY(gethrtime)
-	SYSFASTTRAP(GETHRTIME)
-#if defined(__sparcv9)
-	/*
-	 * Note that the fast trap actually assumes V8 parameter passing
-	 * conventions, so we have to reassemble a 64-bit value here.
-	 */
-	sllx	%o0, 32, %o0
-	or	%o1, %o0, %o0
-#endif
-	RET
-	SET_SIZE(gethrtime)
-
 /*
  * hrtime_t gethrvtime(void)
  *
diff --git a/usr/src/lib/libc/i386/Makefile.com b/usr/src/lib/libc/i386/Makefile.com
index afc7110369..cd394fa0dc 100644
--- a/usr/src/lib/libc/i386/Makefile.com
+++ b/usr/src/lib/libc/i386/Makefile.com
@@ -234,7 +234,7 @@ COMSYSOBJS=			\
 	geteuid.o		\
 	getgid.o		\
 	getgroups.o		\
-	gethrtime.o		\
+	gethrvtime.o		\
 	getitimer.o		\
 	getmsg.o		\
 	getpid.o		\
@@ -315,6 +315,7 @@ COMSYSOBJS=			\
 SYSOBJS=			\
 	__clock_gettime.o	\
 	__clock_gettime_sys.o	\
+	__gethrtime_sys.o	\
 	__getcontext.o		\
 	__uadmin.o		\
 	_lwp_mutex_unlock.o	\
@@ -323,6 +324,7 @@ SYSOBJS=			\
 	forkx.o			\
 	forkallx.o		\
 	getcontext.o		\
+	gethrtime.o		\
 	gettimeofday.o		\
 	lwp_private.o		\
 	nuname.o		\
@@ -1255,6 +1257,7 @@ $(PORTI18N_COND:%=pics/%) := \
 pics/arc4random.o :=	CPPFLAGS += -I$(SRC)/common/crypto/chacha
 
 pics/__clock_gettime.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
+pics/gethrtime.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
 pics/gettimeofday.o := CPPFLAGS += $(COMMPAGE_CPPFLAGS)
 
 .KEEP_STATE:
diff --git a/usr/src/lib/libc/i386/sys/__gethrtime_sys.s b/usr/src/lib/libc/i386/sys/__gethrtime_sys.s
new file mode 100644
index 0000000000..28b07ec632
--- /dev/null
+++ b/usr/src/lib/libc/i386/sys/__gethrtime_sys.s
@@ -0,0 +1,39 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+	.file	"__gethrtime_sys.s"
+
+#include "SYS.h"
+
+/*
+ * hrtime_t __gethrtime_sys(void)
+ *
+ * Returns the current hi-res real time.
+ */
+
+	ENTRY(__gethrtime_sys)
+	SYSFASTTRAP(GETHRTIME)
+	RET
+	SET_SIZE(__gethrtime_sys)
diff --git a/usr/src/lib/libc/i386/sys/gethrtime.c b/usr/src/lib/libc/i386/sys/gethrtime.c
new file mode 100644
index 0000000000..2e266f9afe
--- /dev/null
+++ b/usr/src/lib/libc/i386/sys/gethrtime.c
@@ -0,0 +1,31 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2019 Joyent, Inc.
+ */
+
+
+#include "thr_uberdata.h"
+#include <cp_defs.h>
+
+extern hrtime_t __gethrtime_sys();
+
+hrtime_t
+gethrtime(void)
+{
+	comm_page_t *cp = (comm_page_t *)__uberdata.ub_comm_page;
+
+	if (cp != NULL && __cp_can_gettime(cp) != 0) {
+		return (__cp_gethrtime(cp));
+	}
+	return (__gethrtime_sys());
+}
diff --git a/usr/src/lib/libc/sparc/Makefile.com b/usr/src/lib/libc/sparc/Makefile.com
index 9f68ffd661..9889164f39 100644
--- a/usr/src/lib/libc/sparc/Makefile.com
+++ b/usr/src/lib/libc/sparc/Makefile.com
@@ -24,7 +24,7 @@
 # Copyright (c) 2013, OmniTI Computer Consulting, Inc. All rights reserved.
 # Copyright 2013 Garrett D'Amore <garrett@damore.org>
 # Copyright 2018 Nexenta Systems, Inc.
-# Copyright 2018 Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 # Copyright 2019 OmniOS Community Edition (OmniOSce) Association.
 # Copyright 2019 Peter Tribble.
 #
@@ -250,7 +250,7 @@ COMSYSOBJS=			\
 	geteuid.o		\
 	getgid.o		\
 	getgroups.o		\
-	gethrtime.o		\
+	gethrvtime.o		\
 	getitimer.o		\
 	getmsg.o		\
 	getpid.o		\
@@ -337,6 +337,7 @@ SYSOBJS=			\
 	door.o			\
 	forkx.o			\
 	forkallx.o		\
+	gethrtime.o		\
 	gettimeofday.o		\
 	ptrace.o		\
 	syscall.o		\
diff --git a/usr/src/lib/libc/sparc/sys/gethrtime.s b/usr/src/lib/libc/sparc/sys/gethrtime.s
new file mode 100644
index 0000000000..89bc871d55
--- /dev/null
+++ b/usr/src/lib/libc/sparc/sys/gethrtime.s
@@ -0,0 +1,47 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
+ * Use is subject to license terms.
+ */
+
+	.file	"gethrtime.s"
+
+#include "SYS.h"
+
+/*
+ * hrtime_t gethrtime(void)
+ *
+ * Returns the current hi-res real time.
+ */
+
+	ENTRY(gethrtime)
+	SYSFASTTRAP(GETHRTIME)
+#if defined(__sparcv9)
+	/*
+	 * Note that the fast trap actually assumes V8 parameter passing
+	 * conventions, so we have to reassemble a 64-bit value here.
+	 */
+	sllx	%o0, 32, %o0
+	or	%o1, %o0, %o0
+#endif
+	RET
+	SET_SIZE(gethrtime)
