commit 55a6d4e836df3b559c8afa7680acc6a327624422 (refs/changes/78/5578/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2019-02-13T11:23:24-08:00 (8 months ago)
    
    CNAPI-370 CommandExecute method needs configurable timeout

diff --git a/lib/endpoints/servers.js b/lib/endpoints/servers.js
index 9c8ff31..4e81cf4 100644
--- a/lib/endpoints/servers.js
+++ b/lib/endpoints/servers.js
@@ -1136,6 +1136,7 @@ Server.ensureImage = function handlerServerEnsureImage(req, res, next) {
  * @param {Object} env Object containing environment variables to be passed in
  * @param {String} script Script to be executed. Must have a shebang line
  * @param {Boolean} json Whether to return results as JSON instead of just stdout (default = false)
+ * @param {Integer} timeout Number of ms to wait for command completion before killing task and returning (only supported when using cn-agent, See: FEATURE_USE_CNAGENT_COMMAND_EXECUTE)
  *
  * @response 404 None No such server
  * @response 500 None Error occurred executing script
@@ -1218,6 +1219,8 @@ Server.execute = function handlerCommandExecute(req, res, next) {
             return;
         }
     ]}, function sendRequest(err) {
+        var msg;
+
         if (err) {
             next(err);
             return;
@@ -1225,6 +1228,13 @@ Server.execute = function handlerCommandExecute(req, res, next) {
 
         if (handleWithUr) {
             // Ur's execute functionality
+            if (req.params.hasOwnProperty('timeout')) {
+                msg = 'timeout only supported when using cn-agent (See: ' +
+                    'FEATURE_USE_CNAGENT_COMMAND_EXECUTE)';
+                self.log.error(msg);
+                next(new restify.InvalidArgumentError(msg));
+                return;
+            }
             ur.execute(req, res, next);
         } else {
             // cn-agent's command_execute task
diff --git a/package.json b/package.json
index 0aeb2dc..3f3a6c7 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cnapi",
   "description": "SmartDataCenter Compute Node API",
-  "version": "1.19.2",
+  "version": "1.19.3",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
