commit 678934c17ac0323fb80666abc7139cb94f69e9b0 (refs/changes/74/4874/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-09-25T13:01:36+00:00 (1 year ago)
    
    OS-7262 sysinfo generates invalid JSON due to HVM API

diff --git a/src/sysinfo b/src/sysinfo
index f1dbf980..8f1a3f72 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -211,6 +211,7 @@ function get_bhyve_capability()
 {
     Bhyve_Capable="false"
     Bhyve_Max_Vcpus="0"
+    HVM_API="false"
 
     if [[ -x /usr/lib/brand/bhyve/bhhwcompat ]]; then
         if /usr/lib/brand/bhyve/bhhwcompat; then
