commit 6d039598ea5e3cc295e5cde11ff53350f484afb9
Author: John Levon <john.levon@joyent.com>
Date:   2019-09-04T12:44:12+00:00 (5 weeks ago)
    
    OS-7979 introduce /etc/versions/build

diff --git a/tools/build_live b/tools/build_live
index 927a6ec5..b280ee71 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -268,7 +268,7 @@ function bi_create_ufs
 function bi_remake_buildversion
 {
 	local artefact="$bi_out_dir/gitstatus.json"
-	local bvfile="$bi_wsroot/projects/illumos/usr/src/buildversion"
+	local bvfile="$bi_mnt_root/etc/versions/build"
 
 	#
 	# We include the JSON-only version of the release information as a
@@ -278,19 +278,9 @@ function bi_remake_buildversion
 		fail "could not generate gitstatus file"
 	fi
 
-	if cmp "$artefact" "$bvfile" >/dev/null 2>&1; then
-		return
+	if ! cmp "$artefact" "$bvfile" >/dev/null 2>&1; then
+		cp "$artefact" "$bvfile"
 	fi
-
-	bi_emit_start 'Updating buildversion kernel module'
-
-	cp "$artefact" "$bvfile"
-
-	(cd $bi_wsroot/projects/illumos && ksh93 \
-	    usr/src/tools/scripts/bldenv.sh illumos.sh \
-	    'cd $CODEMGR_WS/usr/src/uts/intel/buildversion && dmake install')
-
-	bi_emit_done
 }
 
 function bi_copy_files
