From d4808db6ef184d9130bd168a0769aba38e52cb37 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Fri, 19 Aug 2016 20:41:35 +0000
Subject: [PATCH] OS-5604 improve tools mandoc Makefile fix

---
 usr/src/tools/mandoc/Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/usr/src/tools/mandoc/Makefile b/usr/src/tools/mandoc/Makefile
index 0aa0dae4ea..5b4c71058a 100644
--- a/usr/src/tools/mandoc/Makefile
+++ b/usr/src/tools/mandoc/Makefile
@@ -19,7 +19,8 @@ CMDDIR=		$(SRC)/cmd/mandoc
 include		../Makefile.tools
 include		$(CMDDIR)/Makefile.common
 
-CPPFLAGS += $(CPPFLAGS.master)
+CPPFLAGS += -_gcc=-nostdinc -I/usr/include -I$(NATIVE_ADJUNCT)/include
+LDFLAGS += -L$(NATIVE_ADJUNCT)/lib
 
 .KEEP_STATE:
 
-- 
2.21.0

