From bfa2d4ad5c0a7bca590ed72c8b6c224f6f8416bd Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Mon, 12 Nov 2018 15:44:01 -0800
Subject: [PATCH] OS-7361 SmartOS initial set up password prompt mangles some
 patterns

---
 scripts/prompt-config.sh | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/scripts/prompt-config.sh b/scripts/prompt-config.sh
index b99bd5b3..31a75717 100755
--- a/scripts/prompt-config.sh
+++ b/scripts/prompt-config.sh
@@ -616,13 +616,19 @@ promptnic()
     val=$mac_addr
 }
 
+readpw()
+{
+	IFS='' read -r -s pw
+	printf '%s' "$val"
+}
+
 promptpw()
 {
 	def="$3"
     key="$4"
 
 	if [[ -n ${key} ]]; then
-		preset_val=$(getanswer "${key}")
+		preset_val="$(getanswer "${key}")"
 	fi
 
 	trap "" SIGINT
@@ -630,17 +636,14 @@ promptpw()
 		val=""
 		while [ -z "$val" ]; do
 			if [[ -n ${preset_val} ]]; then
-				val=${preset_val}
+				val="${preset_val}"
 			else
 				if [ -z "$def" ]; then
 					printf "%s: " "$1"
 				else
-					printf "%s [enter to keep existing]: " \
-					    "$1"
+					printf "%s [enter to keep existing]: " "$1"
 				fi
-				stty -echo
-				read val
-				stty echo
+				val="$(readpw)"
 				echo
 			fi
 			if [ -n "$val" ]; then
@@ -662,7 +665,7 @@ promptpw()
 				fi
 			else
 				if [ -n "$def" ]; then
-					val=$def
+					val="$def"
 					return
 				else
 					echo "A value must be provided."
@@ -673,12 +676,10 @@ promptpw()
 		cval=""
 		while [ -z "$cval" ]; do
 			if [[ -n ${preset_val} ]]; then
-				cval=${preset_val}
+				cval="${preset_val}"
 			else
 				printf "%s: " "Confirm password"
-				stty -echo
-				read cval
-				stty echo
+				cval="$(readpw)"
 				echo
 			fi
 			[ -n "$cval" ] && break
-- 
2.21.0

