From 25114463bcbac6b878cc3b1ccce79bb872b18e48 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Fri, 15 Mar 2019 15:07:29 -0400
Subject: [PATCH] TOOLS-2101 jsstyle submodules too old for old perls

---
 .gitmodules  | 2 +-
 deps/jsstyle | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 251a556..9b8d37f 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -6,4 +6,4 @@
 	url = git://github.com/trentm/restdown
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
diff --git a/deps/jsstyle b/deps/jsstyle
index ccb145b..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
-- 
2.21.0

