commit 52e7981785224180315e16648e84f7acd8f10602 (refs/changes/94/3694/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-03-19T22:33:18-07:00 (1 year, 7 months ago)
    
    TRITON-178 Image creation support for bhyve VM

diff --git a/CHANGES.md b/CHANGES.md
index e93035e..20db701 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,10 @@
 # IMGAPI changelog
 
+## 4.3.0
+
+- TRITON-178 Add support for image creation with bhyve brand (requires platform
+  support)
+
 ## 4.2.2
 
 - TRITON-222 Use node-triton-metrics for IMGAPI metrics collection
diff --git a/lib/images.js b/lib/images.js
index 6905068..4ded88d 100644
--- a/lib/images.js
+++ b/lib/images.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  */
 
 /*
@@ -1880,7 +1880,7 @@ function apiCreateImageFromVm(req, res, callback) {
 
         function getPrepareImageScript(next) {
             var protoImageUuid;
-            if (vm.brand === 'kvm') {
+            if (['bhyve', 'kvm'].indexOf(vm.brand) !== -1) {
                 if (vm.disks) {
                     for (var i = 0; i < vm.disks.length; i++) {
                         if (vm.disks[i].image_uuid) {
diff --git a/package.json b/package.json
index 09101f7..ebd267e 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "imgapi",
   "description": "Image API to manage images for SDC 7",
-  "version": "4.2.2",
+  "version": "4.3.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
