commit 2fb9783e48594fea4f02414bc12a7389548934bd (refs/changes/64/5364/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2019-01-14T15:42:14-08:00 (9 months ago)
    
    TRITON-1087 cnsadm broken after TRITON-1044 jsprim update

diff --git a/bin/cnsadm b/bin/cnsadm
index ebc7a40..0b60a26 100755
--- a/bin/cnsadm
+++ b/bin/cnsadm
@@ -143,7 +143,7 @@ sapi.get('/configs/' + zoneName, function (err, req, res, obj) {
 
 	Object.keys(config).forEach(function (k) {
 		if (obj.metadata.hasOwnProperty(k)) {
-			if (!jsprim.deepEquals(config[k], obj.metadata[k]))
+			if (!jsprim.deepEqual(config[k], obj.metadata[k]))
 				waiting[k] = true;
 			config[k] = obj.metadata[k];
 		}
