From 2fb9783e48594fea4f02414bc12a7389548934bd Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 14 Jan 2019 15:42:14 -0800
Subject: [PATCH] TRITON-1087 cnsadm broken after TRITON-1044 jsprim update

---
 bin/cnsadm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/cnsadm b/bin/cnsadm
index ebc7a40..0b60a26 100755
--- a/bin/cnsadm
+++ b/bin/cnsadm
@@ -143,7 +143,7 @@ sapi.get('/configs/' + zoneName, function (err, req, res, obj) {
 
 	Object.keys(config).forEach(function (k) {
 		if (obj.metadata.hasOwnProperty(k)) {
-			if (!jsprim.deepEquals(config[k], obj.metadata[k]))
+			if (!jsprim.deepEqual(config[k], obj.metadata[k]))
 				waiting[k] = true;
 			config[k] = obj.metadata[k];
 		}
-- 
2.21.0

