commit f2c5dfb162157c3d25a15917b353f4442dc1f3cc
Author: Jason King <jason.king@joyent.com>
Date:   2019-06-18T18:08:55+00:00 (4 months ago)
    
    OS-7698 Add mkzpool -e for creating encrypted zpools

diff --git a/src/mkzpool.js b/src/mkzpool.js
index 1b986881..70a42f0f 100644
--- a/src/mkzpool.js
+++ b/src/mkzpool.js
@@ -1,7 +1,7 @@
 #! /usr/node/bin/node
 
 /*
- * Copyright 2012 Joyent, Inc.  All rights reserved.
+ * Copyright 2019 Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -27,11 +27,16 @@ var config;
 var pool;
 
 var option;
+var options;
+var opt_e = false;
 var opt_f = false;
-var parser = new getopt.BasicParser('f', process.argv);
+var parser = new getopt.BasicParser('ef', process.argv);
 
 while ((option = parser.getopt()) !== undefined && !option.error) {
-	switch (option.option) {
+    switch (option.option) {
+	case 'e':
+		opt_e = true;
+		break;
 	case 'f':
 		opt_f = true;
 		break;
@@ -50,8 +55,12 @@ if (!process.argv[parser.optind()] || !process.argv[parser.optind() + 1])
 pool = process.argv[parser.optind()];
 json = fs.readFileSync(process.argv[parser.optind() + 1], 'utf8');
 config = JSON.parse(json);
+options = {
+	force: opt_f,
+	encrypt: opt_e
+};
 
-zfs.zpool.create(pool, config, opt_f, function (err) {
+zfs.zpool.create(pool, config, options, function mkZpoolCb(err) {
 	if (err) {
 		fatal('pool creation failed: ' + err);
 	}
diff --git a/src/node_modules/zfs.js b/src/node_modules/zfs.js
index 30984e02..7f1f360f 100644
--- a/src/node_modules/zfs.js
+++ b/src/node_modules/zfs.js
@@ -1,8 +1,9 @@
 /*
- * Copyright 2013 Joyent, Inc.  All rights reserved.
+ * Copyright 2019 Joyent, Inc.
  */
 
-var cp = require('child_process'),
+var assert = require('assert-plus'),
+    cp = require('child_process'),
     fs = require('fs');
 
 var execFile = cp.execFile,
@@ -12,6 +13,7 @@ var execFile = cp.execFile,
  * ZFS utilities paths
  */
 exports.paths = {
+    'kbmadm': '/usr/sbin/kbmadm',
     'zpool': '/sbin/zpool',
     'zfs': '/sbin/zfs'
 };
@@ -124,22 +126,31 @@ zpool.status = function (pool, callback) {
  * zpool create command, including log devices, cache devices, and hot spares.
  * The input is an object of the form produced by the disklayout library.
  */
-zpool.create = function (pool, config, force, callback) {
+zpool.create = function (pool, config, options, callback) {
+	var cmd;
 	var args;
 
-	if (arguments.length === 3) {
-		callback = force;
-		force = false;
-	} else if (arguments.length !== 4) {
-		throw Error('Invalid arguments, 3 or 4 arguments required');
-	}
+	assert.string(pool);
+	assert.object(config);
+	assert.object(options);
+	assert.bool(options.force);
+	assert.bool(options.encrypt);
+	assert.func(callback);
 
-	if (force === true) {
-		args = [ 'create', '-f', pool ];
+	if (options.encrypt === true) {
+		cmd = export.paths.kbmadm;
+		args = [ 'create-zpool' ];
 	} else {
-		args = [ 'create', pool ];
+		cmd = export.paths.zpool;
+		args = [ 'create' ];
+	}
+
+	if (options.force === true) {
+		args.push('-f');
 	}
 
+	args.push(pool);
+
 	config.vdevs.forEach(function (vdev) {
 		if (vdev.type)
 			args.push(vdev.type);
@@ -173,7 +184,7 @@ zpool.create = function (pool, config, force, callback) {
 		});
 	}
 
-	execFile(exports.paths.zpool, args, { timeout: timeoutDuration },
+	execFile(cmd, args, { timeout: timeoutDuration },
 	    function (error, stdout, stderr) {
 		if (error)
 			return (callback(stderr.toString()));
