From 9227144bdc9b3322ef666575ac4c3cdc647d8ca3 Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Wed, 8 Mar 2017 14:26:16 +1100
Subject: [PATCH] PUBAPI-1368 - CloudAPI can die on MethodNotAllowed during
 start up

---
 main.js | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/main.js b/main.js
index 60812b9..cce2c52 100644
--- a/main.js
+++ b/main.js
@@ -166,12 +166,9 @@ function createBootstrapServer(port, cb) {
     var bootstrapServer = restify.createServer();
     bootstrapServer.use(restify.fullResponse());
 
-    function bootstrapHandler(req, res, next) {
+    function bootstrapHandler(req, res, next1, next2) {
         res.send(new restify.InternalError('Failure to connect to Moray'));
-
-        if (next) {
-            next();
-        }
+        return next2() || next1(); // method and on() callback signatures differ
     }
 
     ['get', 'post', 'put', 'del', 'head'].forEach(function (method) {
-- 
2.21.0

