commit 9227144bdc9b3322ef666575ac4c3cdc647d8ca3 (refs/changes/24/1624/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-03-08T14:26:16+11:00 (2 years, 7 months ago)
    
    PUBAPI-1368 - CloudAPI can die on MethodNotAllowed during start up

diff --git a/main.js b/main.js
index 60812b9..cce2c52 100644
--- a/main.js
+++ b/main.js
@@ -166,12 +166,9 @@ function createBootstrapServer(port, cb) {
     var bootstrapServer = restify.createServer();
     bootstrapServer.use(restify.fullResponse());
 
-    function bootstrapHandler(req, res, next) {
+    function bootstrapHandler(req, res, next1, next2) {
         res.send(new restify.InternalError('Failure to connect to Moray'));
-
-        if (next) {
-            next();
-        }
+        return next2() || next1(); // method and on() callback signatures differ
     }
 
     ['get', 'post', 'put', 'del', 'head'].forEach(function (method) {
