commit cb0d21dbc4ae203e7a4ae638ed045c6a0bb12195 (refs/changes/64/3764/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2018-04-02T16:22:31-07:00 (1 year, 6 months ago)
    
    TRITON-288 sdcnode updates for March 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 4ea3e41..9553560 100644
--- a/Makefile
+++ b/Makefile
@@ -33,7 +33,7 @@ SMF_MANIFESTS = $(shell ls smf/manifests/*.xml)
 NODEUNIT	:= ./node_modules/.bin/nodeunit
 CLEAN_FILES += ./node_modules
 
-NODE_PREBUILT_VERSION=v4.8.5
+NODE_PREBUILT_VERSION=v4.9.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=zone
 	# Allow building on other than image sdc-minimal-multiarch-lts@15.4.1.
