From 2cbbbe4f9874864b6995423749c3aa7852df0c39 Mon Sep 17 00:00:00 2001
From: Jerry Jelinek <jerry.jelinek@joyent.com>
Date: Fri, 11 Nov 2016 12:25:26 +0000
Subject: [PATCH] 7562 iwn: Leading zeros in /etc/driver_aliases

---
 overlay/generic/etc/driver_aliases | 50 +++++++++++++++---------------
 1 file changed, 25 insertions(+), 25 deletions(-)

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index ba999d32..e35fc7f5 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -692,31 +692,31 @@ iprb "pci8086,2449.8086.3018"
 iprb "pci8086,27dc"
 iprb "pci8086,27dc.8086.308d"
 isa "pciclass,060100"
-iwn "pci8086,0082"
-iwn "pci8086,0083"
-iwn "pci8086,0084"
-iwn "pci8086,0085"
-iwn "pci8086,0087"
-iwn "pci8086,0089"
-iwn "pci8086,008a"
-iwn "pci8086,008b"
-iwn "pci8086,008d"
-iwn "pci8086,008e"
-iwn "pci8086,0090"
-iwn "pci8086,0091"
-iwn "pci8086,0887"
-iwn "pci8086,0888"
-iwn "pci8086,088e"
-iwn "pci8086,088f"
-iwn "pci8086,0890"
-iwn "pci8086,0891"
-iwn "pci8086,0892"
-iwn "pci8086,0893"
-iwn "pci8086,0894"
-iwn "pci8086,0895"
-iwn "pci8086,0896"
-iwn "pci8086,0897"
-iwn "pci8086,08ae"
+iwn "pci8086,82"
+iwn "pci8086,83"
+iwn "pci8086,84"
+iwn "pci8086,85"
+iwn "pci8086,87"
+iwn "pci8086,89"
+iwn "pci8086,8a"
+iwn "pci8086,8b"
+iwn "pci8086,8d"
+iwn "pci8086,8e"
+iwn "pci8086,90"
+iwn "pci8086,91"
+iwn "pci8086,887"
+iwn "pci8086,888"
+iwn "pci8086,88e"
+iwn "pci8086,88f"
+iwn "pci8086,890"
+iwn "pci8086,891"
+iwn "pci8086,892"
+iwn "pci8086,893"
+iwn "pci8086,894"
+iwn "pci8086,895"
+iwn "pci8086,896"
+iwn "pci8086,897"
+iwn "pci8086,8ae"
 iwn "pci8086,422b"
 iwn "pci8086,422c"
 iwn "pci8086,4238"
-- 
2.21.0

