commit 5c065c2741318020e928c1e3eb09d093b1386581 (refs/changes/80/4480/1)
Author: Kévin Berthommier <kevin.bertho@gmail.com>
Date:   2018-07-03T23:29:10+02:00 (1 year, 3 months ago)
    
    fix error

diff --git a/lib/verror.js b/lib/verror.js
index ad4f4eb..3aa5ffd 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -365,7 +365,7 @@ VError.errorForEach = function (err, func)
 	if (!isError(err)) {
 		throw new mod_AssertionError('err must be an Error');
 	}
-	if (typeof arg !== 'function') {
+	if (typeof func !== 'function') {
 		throw new mod_AssertionError('func (func) is required');
 	}
 
