From 5c065c2741318020e928c1e3eb09d093b1386581 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?K=C3=A9vin=20Berthommier?= <kevin.bertho@gmail.com>
Date: Tue, 3 Jul 2018 23:29:10 +0200
Subject: [PATCH] fix error

---
 lib/verror.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/verror.js b/lib/verror.js
index ad4f4eb..3aa5ffd 100644
--- a/lib/verror.js
+++ b/lib/verror.js
@@ -365,7 +365,7 @@ VError.errorForEach = function (err, func)
 	if (!isError(err)) {
 		throw new mod_AssertionError('err must be an Error');
 	}
-	if (typeof arg !== 'function') {
+	if (typeof func !== 'function') {
 		throw new mod_AssertionError('func (func) is required');
 	}
 
-- 
2.17.2 (Apple Git-113)

