commit f0b94fc7ad35406c1b4effb4472edec753879017 (refs/changes/17/4817/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-09-10T20:07:14-05:00 (1 year, 1 month ago)
    
    OS-7200 would like thread name API (fix debug)

diff --git a/usr/src/cmd/mdb/common/modules/genunix/thread.c b/usr/src/cmd/mdb/common/modules/genunix/thread.c
index fadbfec863..410f9973d8 100644
--- a/usr/src/cmd/mdb/common/modules/genunix/thread.c
+++ b/usr/src/cmd/mdb/common/modules/genunix/thread.c
@@ -782,7 +782,6 @@ int
 stackinfo(uintptr_t addr, uint_t flags, int argc, const mdb_arg_t *argv)
 {
 	kthread_t t;
-	proc_t p;
 	uint64_t *ptr;  /* pattern pointer */
 	caddr_t	start;	/* kernel stack start */
 	caddr_t end;	/* kernel stack end */
