From e4bee8ed1944bbb6e77624d3fd1b8efc0e5a6474 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 10 Mar 2017 14:56:54 -0800
Subject: [PATCH] joyent/node-mname#19 infinite loop possible in readName()

---
 lib/dns-buffer.js | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/lib/dns-buffer.js b/lib/dns-buffer.js
index fbe1b94..62bfda2 100644
--- a/lib/dns-buffer.js
+++ b/lib/dns-buffer.js
@@ -92,7 +92,7 @@ DNSBuffer.prototype.readName = function () {
         var finalOff;
 
         rlen = this._buffer.readUInt8(off++);
-        while (rlen !== 0x00) {
+        while (rlen !== 0x00 && name.length < 256) {
                 var meta = rlen & NAME_META_MASK;
 
                 if (meta == NAME_STRING) {
@@ -108,7 +108,9 @@ DNSBuffer.prototype.readName = function () {
                         ptr = ptr | ((rlen & ~(0xC0)) << 8);
 
                         assert.ok(ptr < this._size,
-                            'invalid label pointer (off end of buf)');
+                            'Invalid label pointer (off end of buffer)');
+                        assert.ok(!(ptr >= this._offset && ptr <= off),
+                            'Invalid label pointer (causes a loop)');
 
                         if (finalOff === undefined)
                                 finalOff = off;
@@ -121,6 +123,10 @@ DNSBuffer.prototype.readName = function () {
                 rlen = this._buffer.readUInt8(off++);
         }
 
+        if (name.length > 255) {
+                throw (new Error('Invalid name (maximum length exceeded)'));
+        }
+
         if (finalOff === undefined)
                 finalOff = off;
         this._offset = finalOff;
-- 
2.21.0

