From 1e986ea7338aa5b75185ea28127641c75b2fcb30 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 29 Nov 2016 19:09:45 -0800
Subject: [PATCH] joyent/node-triton#143 triton rbac key duplicate output
 (changelog)

---
 CHANGES.md | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CHANGES.md b/CHANGES.md
index 5ce09c9..a67065a 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -7,7 +7,7 @@ Known issues:
 
 ## not yet released
 
-(nothing yet)
+- [joyent/node-triton#143] Fix duplicate output from 'triton rbac key ...'.
 
 ## 4.15.0
 
-- 
2.21.0

