commit 69ce9f1d88da84df67628e9406b01613dd6762ce (refs/changes/84/2984/5)
Author: Mike Zeller <mike.zeller@joyent.com>
Date:   2017-12-17T22:41:44-08:00 (1 year, 10 months ago)
    
    PUBAPI-1451 Want UpdateNetworkIP

diff --git a/docs/index.md b/docs/index.md
index 74592f3..51aff7e 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -747,11 +747,18 @@ services, here's a simple Bash function you can use to wrap cURL when
 communicating with CloudAPI:
 
     $ function cloudapi() {
-      local now=`date -u "+%a, %d %h %Y %H:%M:%S GMT"`;
-      local signature=`echo ${now} | tr -d '\n' | openssl dgst -sha256 -sign ~/.ssh/id_rsa | openssl enc -e -a | tr -d '\n'`;
-
-      curl -i -H "Accept: application/json" -H "accept-version: ~8" -H "Date: ${now}" -H "Authorization: Signature keyId=\"/$SDC_ACCOUNT/keys/id_rsa\",algorithm=\"rsa-sha256\" ${signature}" --url $SDC_URL$@;
-      echo "";
+        local now=$(date -u '+%a, %d %h %Y %H:%M:%S GMT')
+        local signature=$(echo -n "$now" | openssl dgst -sha256 -sign ~/.ssh/coal_test_user_rsa | openssl enc -e -a | tr -d '\n')
+        local url="$SDC_URL$1"
+        shift
+
+        curl -s -k -i \
+            -H 'Accept: application/json' \
+            -H "accept-version: ~8" \
+            -H "Date: $now" \
+            -H "Authorization: Signature keyId=\"/$SDC_ACCOUNT/keys/laptop_test\",algorithm=\"rsa-sha256\" $signature" \
+            "$@" "$url"
+        echo
     }
 
 You may need to alter the path to your SSH key in the above function, as well as
@@ -901,6 +908,12 @@ The set of supported *API versions* is given in the ping endpoint:
 
 The section describes API changes in CloudAPI versions.
 
+## 8.4.0
+
+- This version adds support for the following new endpoints:
+  [ListNetworkIPs](#ListNetworkIPs), [GetNetworkIP](#GetNetworkIP),
+  and [UpdateNetworkIP](#UpdateNetworkIP).
+
 ## 8.3.0
 
 - CreateMachine supports a new `affinity` field for specifying affinity rules.
@@ -8491,6 +8504,81 @@ ResourceNotFound | If `:login`, `:id`, or `:ip_address` does not exist
       "managed": false
     }
 
+## UpdateNetworkIP (PUT /:login/networks/:id/ips/:ip_address)
+
+Update a network's IP to toggle the `reserved` flag. If `reserved` is set to
+`true` the IP will not be given out automatically at provision time. You cannot
+update an IP on a public network. On private networks you can update an IP that
+is already in use by an instance owned by you, or an IP that is not yet in use
+as long as it's within the network's subnet.
+
+### Inputs
+
+**Field**  | **Type** | **Description**
+---------- | -------- | ---------------
+reserved   | boolean  | If true take the IP out of the provisioning pool
+
+### Returns
+
+An IP object:
+
+**Field**  | **Type**   | **Description**
+---------- | ---------- | ---------------
+ip         | String     | IP Address
+reserved   | Boolean    | Whether this IP is reserved or not
+managed    | Boolean    | True if the user cannot modify the IP via UpdateNetworkIP (example broadcast and gateway IPs)
+owner_uuid | UUID       | UUID of the owner that the instance is associated with (Optional)
+belongs_to_uuid | UUID  | UUID of the instance the IP is associated with (Optional)
+
+### Errors
+
+For all possible errors, see [CloudAPI HTTP Responses](#cloudapi-http-responses).
+
+**Error Code**   | **Description**
+---------------- | ---------------
+ResourceNotFound | If `:login`, `:id`, or `:ip_address` does not exist
+InvalidArgument  | If `:id` is not a UUID, or it is a public network. If `:ip_address` is in use by another user
+MissingParameter | If the `reserved` argument isn't specefied
+
+### CLI Command
+
+    $ triton network ip update daeb93a2-532e-4bd4-8788-b6b30f10ac17 192.168.128.5 reserved=false
+
+#### Example Request
+
+    PUT  /my/networks/daeb93a2-532e-4bd4-8788-b6b30f10ac17/ips/192.168.128.5 HTTP/1.1
+    authorization: Signature keyId="...
+    accept: application/json
+    accept-version: ~8
+    host: api.example.com
+
+    {
+      "reserved": false
+    }
+
+#### Example Response
+
+    HTTP/1.1 200 OK
+    Content-Type: application/json
+    Content-Length: 55
+    Access-Control-Allow-Origin: *
+    Access-Control-Allow-Headers: Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, Api-Version, Response-Time
+    Access-Control-Allow-Methods: PUT, GET, HEAD
+    Access-Control-Expose-Headers: Api-Version, Request-Id, Response-Time
+    Connection: Keep-Alive
+    Content-MD5: DvheKXPZSZw5lav9REKHuw==
+    Date: Wed, 22 Nov 2017 18:53:02 GMT
+    Server: cloudapi/8.3.0
+    Api-Version: 8.0.0
+    Request-Id: a9715a00-6194-4053-9846-fcdf9c5476fb
+    Response-Time: 76
+
+    {
+      "ip": "192.168.128.5",
+      "reserved": false,
+      "managed": false
+    }
+
 # Nics
 
 CloudAPI provides a way to list, add and remove NICs attached to a instance.
diff --git a/lib/endpoints/networks.js b/lib/endpoints/networks.js
index 867498f..8d7cce5 100644
--- a/lib/endpoints/networks.js
+++ b/lib/endpoints/networks.js
@@ -21,6 +21,7 @@ var resources = require('../resources');
 var InternalError = restify.InternalError;
 var InvalidArgumentError = restify.InvalidArgumentError;
 var ResourceNotFoundError = restify.ResourceNotFoundError;
+var MissingParameterError = restify.MissingParameterError;
 
 var FABRIC_VLAN_FIELDS = ['description', 'name', 'vlan_id'];
 var FABRIC_NETWORK_FIELDS = ['description', 'fabric', 'gateway',
@@ -384,6 +385,117 @@ function getNetworkIp(req, res, next) {
     });
 }
 
+/*
+ * Update a network ip under the following conditions:
+ * - The IP is not on a public network
+ * - The IP is not 'managed'
+ * - The IP is anywhere within the subnet (napi enforced)
+ * - The IP is not in use by another account on a shared
+ *   private network
+ */
+function updateNetworkIp(req, res, next) {
+    var napi = req.sdc.napi;
+    assert.ok(napi);
+
+    var adminUuid = req.config.ufds_admin_uuid;
+    assert.uuid(adminUuid, 'admin uuid');
+
+    var networkUuid = req.params.id;
+    var _ip = req.params.ip_address;
+    var userIp;
+
+    if (!Object.prototype.hasOwnProperty.call(req.params, 'reserved')) {
+        return next(new MissingParameterError(
+            'reserved is a required argument'));
+    }
+    var reserved = req.params.reserved;
+
+    /*
+     * cloudapi allows req.param's to come from queryParser and bodyParser.
+     * Since queryParser will give use a string and body parser will give us
+     * a boolean or a string, we need to map this to the right value. We return
+     * an error otherwise.
+     */
+    switch (typeof (reserved)) {
+    case 'boolean':
+        break;
+    case 'string':
+        switch (reserved) {
+            case 'true':
+                reserved = true;
+                break;
+            case 'false':
+                reserved = false;
+                break;
+            default:
+                return next(new InvalidArgumentError(
+                    'reserved must be set to true or false'));
+        }
+        break;
+    default:
+        return next(new InvalidArgumentError(
+            'reserved must be set to true or false'));
+    }
+
+    // validateNetworkForIps should have already validated this exists
+    var net = req.networks.find(function findNetwork(n) {
+        return n.uuid === networkUuid;
+    });
+    assert.object(net, 'net object');
+
+    // If it's a public network we don't allow ip reservations
+    if (!net.hasOwnProperty('owner_uuids')) {
+        return next(new InvalidArgumentError(
+            'cannot update an IP on a public network: ' + networkUuid));
+    }
+
+    var pipeline = [
+        function _getIp(_, cb) {
+            napi.getIP(networkUuid, _ip, function getIP(err, ip) {
+                if (err) {
+                    return cb(translateErr(err));
+                }
+
+                userIp = translateIp(ip, req.account.uuid, adminUuid);
+                cb();
+            });
+        },
+        function _updateIp(_, cb) {
+            if (userIp.managed === true) {
+                return cb(new InvalidArgumentError(
+                    'cannot update managed IP: ' + _ip));
+            }
+
+            if (userIp.hasOwnProperty('owner_uuid') &&
+                userIp.owner_uuid !== req.account.uuid) {
+                return cb(new InvalidArgumentError(
+                    'ip %s on network %s is in use by another account', _ip,
+                        networkUuid));
+            }
+
+            var opts = {reserved: reserved};
+            napi.updateIP(networkUuid, _ip, opts, function updateIp(err, ip) {
+                if (err) {
+                    return cb(translateErr(err));
+                }
+
+                userIp = translateIp(ip, req.account.uuid, adminUuid);
+                cb();
+            });
+        }
+    ]
+
+    return vasync.pipeline({
+        funcs: pipeline
+    }, function (err, results) {
+        if (err) {
+            return next(err);
+        }
+
+        res.send(userIp);
+        return next();
+    });
+}
 
 function listFabricVLANs(req, res, next) {
     assert.ok(req.account);
@@ -761,6 +873,8 @@ function mountNetworks(server, before, pre) {
         name: 'HeadNetwork'
     }, before, pre, getNetwork);
 
+    // --- Network IPs
+
     server.get({
         path: '/:account/networks/:id/ips',
         name: 'ListNetworkIPs'
@@ -771,14 +885,19 @@ function mountNetworks(server, before, pre) {
         name: 'HeadNetworkIPs'
     }, before, pre, validateNetworkForIps, listNetworkIps);
 
+    server.put({
+        path: '/:account/networks/:id/ips/:ip_address',
+        name: 'UpdateNetworkIP'
+    }, before, pre, validateNetworkForIps, updateNetworkIp);
+
     server.get({
         path: '/:account/networks/:id/ips/:ip_address',
-        name: 'GetNetworkIPs'
+        name: 'GetNetworkIP'
     }, before, pre, validateNetworkForIps, getNetworkIp);
 
     server.head({
         path: '/:account/networks/:id/ips/:ip_address',
-        name: 'HeadNetworkIPs'
+        name: 'HeadNetworkIP'
     }, before, pre, validateNetworkForIps, getNetworkIp);
 
     return server;
diff --git a/package.json b/package.json
index 9c037eb..9ae077a 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "cloudapi",
     "description": "SmartDataCenter CloudAPI",
-    "version": "8.3.0",
+    "version": "8.4.0",
     "author": "Joyent (joyent.com)",
     "private": true,
     "engines": {
diff --git a/test/networks.test.js b/test/networks.test.js
index f244c73..bf10779 100644
--- a/test/networks.test.js
+++ b/test/networks.test.js
@@ -33,10 +33,13 @@ var POOL1_NAME = 'sdccloudapitest_networks_pool1';
 // Test variables
 var NO_SUCH_NETWORK_UUID = 'deaddead-c626-11e5-b674-334e7e514480';
 var RESERVED_IP = '10.99.92.25';
+var NETWORK3_MANAGED_IP = '10.99.92.1';
 var ZONE_IP1 = '10.99.90.52';
 var ZONE_IP2 = '10.99.90.53';
+var ZONE_IP3 = '10.99.92.27'; // used for private network test (different owner)
 var ZONE_UUID1 = 'c4311f24-de18-40b9-b57e-249f2aec7533';
 var ZONE_UUID2 = '5dd79db9-3d42-40a3-a600-7fa2984ff48c';
+var ZONE_UUID3 = 'd8fb0a51-4dc3-4338-8bec-de41b82a16bc';
 
 
 // --- Helpers
@@ -127,7 +130,7 @@ function createFixtures(t, cb) {
         },
         function mkTestNetwork3(_, next) {
             var params = {
-                owner_uuids: [ CLIENT.account.uuid ],
+                owner_uuids: [ CLIENT.account.uuid, OTHER.account.uuid ],
                 gateway: '10.99.92.1'
             };
             createTestNetwork(NETWORK3_NAME, 92, params, function (err, net) {
@@ -178,7 +181,19 @@ function createFixtures(t, cb) {
                     function (err, ip) {
                 t.ifError(err, 'createFixtures: zone ip ' + ZONE_IP2);
                 t.ok(ip, 'createFixtures: reserved ip');
-                fixtures.ip3 = ip;
+                next(err);
+            });
+        },
+        function mkTestZoneIP3(_, next) {
+            var params = {
+                owner_uuid: OTHER.account.uuid,
+                belongs_to_type: 'zone',
+                belongs_to_uuid: ZONE_UUID3
+            };
+            CLIENT.napi.updateIP(fixtures.network3.uuid, ZONE_IP3, params,
+                    function (err, ip) {
+                t.ifError(err, 'createFixtures: zone ip ' + ZONE_IP3);
+                t.ok(ip, 'createFixtures: reserved ip');
                 next(err);
             });
         },
@@ -403,11 +418,13 @@ test('networks', function (tt) {
         });
     });
 
+    // IP tests ------------
+
     /*
      *  On public networks we should only see provisioned ips owned
      *  by the specific user.
      */
-    tt.test('  get network ips (public)', function (t) {
+    tt.test('  list network ips (public)', function (t) {
         var out = [
             {
                 ip: ZONE_IP1,
@@ -434,7 +451,7 @@ test('networks', function (tt) {
      *  On private networks we should see provisioned/reserved ips
      *  as well as 'triton_protected' ips such as the broadcast/gateway
      */
-    tt.test('  get network ips (owner)', function (t) {
+    tt.test('  list network ips (owner)', function (t) {
         var out = [
             {
                 ip: '10.99.92.1',
@@ -446,6 +463,12 @@ test('networks', function (tt) {
                 managed: false,
                 reserved: true
             },
+            {
+                ip: ZONE_IP3,
+                managed: false,
+                reserved: false,
+                owner_uuid: OTHER.account.uuid
+            },
             {
                 ip: '10.99.92.255',
                 managed: true,
@@ -528,6 +551,85 @@ test('networks', function (tt) {
         });
     });
 
+    //  Update of IP on a public network returns InvalidArgument
+    tt.test('  update ip on public network', function (t) {
+        var path = format('/my/networks/%s/ips/%s', fixtures.network1.uuid,
+            ZONE_IP2);
+        CLIENT.put(path, {reserved: true}, function (err, req, res, body) {
+            common.checkInvalidArgument(t, err, req, res, body);
+            t.end();
+        });
+    });
+
+    //  Update of a managed IP returns InvalidArgument
+    tt.test('  update managed ip', function (t) {
+        var path = format('/my/networks/%s/ips/%s', fixtures.network3.uuid,
+            NETWORK3_MANAGED_IP);
+        CLIENT.put(path, {reserved: true}, function (err, req, res, body) {
+            common.checkInvalidArgument(t, err, req, res, body);
+            t.end();
+        });
+    });
+
+    tt.test('  update ip on private network (true)', function (t) {
+        var ip = '10.99.92.26';
+        var out = {
+            ip: ip,
+            managed: false,
+            reserved: true
+        };
+        var path = format('/my/networks/%s/ips/%s', fixtures.network3.uuid,
+            ip);
+        CLIENT.put(path, {reserved: true}, function (err, req, res, body) {
+            t.ifError(err, 'PUT /my/networks/' + fixtures.network3.uuid +
+                '/ips/' + ip  + ' error');
+            t.equal(res.statusCode, 200,
+                'PUT /my/networks/:uuid/ips/:ip_address status');
+            common.checkHeaders(t, res.headers);
+            t.ok(body, 'PUT /my/networks/:uuid/ips/:ip_address body');
+            t.deepEqual(body, out, 'UpdateNetworkIP reserved=true works on'
+                + ' private network');
+            t.end();
+        });
+    });
+
+    // Set reservation back to false
+    tt.test('  update ip on private network (false)', function (t) {
+        var ip = '10.99.92.26';
+        var out = {
+            ip: ip,
+            managed: false,
+            reserved: false
+        };
+        var path = format('/my/networks/%s/ips/%s', fixtures.network3.uuid,
+            ip);
+        CLIENT.put(path, {reserved: false}, function (err, req, res, body) {
+            t.ifError(err, 'PUT /my/networks/' + fixtures.network3.uuid +
+                '/ips/' + ip  + ' error');
+            t.equal(res.statusCode, 200,
+                'PUT /my/networks/:uuid/ips/:ip_address status');
+            common.checkHeaders(t, res.headers);
+            t.ok(body, 'PUT /my/networks/:uuid/ips/:ip_address body');
+            t.deepEqual(body, out, 'UpdateNetworkIP reserved=false works on'
+                + ' private network');
+            t.end();
+        });
+    });
+
+    /*
+     * Update of IP on private network in use by another user returns
+     * InvalidArgument
+     */
+    tt.test('  update ip on private network in use by another account',
+        function (t) {
+        var path = format('/my/networks/%s/ips/%s', fixtures.network3.uuid,
+            ZONE_IP3);
+        CLIENT.put(path, {reserved: true}, function (err, req, res, body) {
+            common.checkInvalidArgument(t, err, req, res, body);
+            t.end();
+        });
+    });
+
     tt.test('  teardown', function (t) {
         vasync.pipeline({ funcs: [
             function teardownFixtures(_, next) {
