commit 2c2771cb3996fc57fcf96dea52728fdd76e78046 (refs/changes/22/2822/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-10-19T19:20:24+00:00 (2 years ago)
    
    OS-6338 panic from kvm_load_gs

diff --git a/usr/src/uts/common/brand/lx/os/lx_misc.c b/usr/src/uts/common/brand/lx/os/lx_misc.c
index 53963c0f7d..0990fed430 100644
--- a/usr/src/uts/common/brand/lx/os/lx_misc.c
+++ b/usr/src/uts/common/brand/lx/os/lx_misc.c
@@ -273,6 +273,14 @@ free:
 	}
 	if (sqp)
 		kmem_free(sqp, sizeof (sigqueue_t));
+
+#if defined(__amd64)
+	/*
+	 * We're going to swtch to another process, make sure we don't leave
+	 * any garbage around.
+	 */
+	reset_sregs();
+#endif
 }
 
 void
