commit 746ca6a66f89f466dc70f228846dff27e1bbb73a (refs/changes/22/2822/2)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-10-19T19:37:55+00:00 (2 years ago)
    
    OS-6338 panic from kvm_load_gs

diff --git a/usr/src/uts/common/brand/lx/os/lx_misc.c b/usr/src/uts/common/brand/lx/os/lx_misc.c
index 53963c0f7d..edb2fcd420 100644
--- a/usr/src/uts/common/brand/lx/os/lx_misc.c
+++ b/usr/src/uts/common/brand/lx/os/lx_misc.c
@@ -273,6 +273,12 @@ free:
 	}
 	if (sqp)
 		kmem_free(sqp, sizeof (sigqueue_t));
+
+	/*
+	 * We're going to swtch to another process, make sure we don't leave
+	 * any garbage around.
+	 */
+	lx_save(lwp);
 }
 
 void
