commit 7529fae267947e42ac4bd38f0d731ecae46be1e4 (refs/changes/76/5176/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2018-12-03T16:19:50-05:00 (10 months ago)
    
    TOOLS-2117 sdcnode updates for November 2018 node.js security releases

diff --git a/Makefile b/Makefile
index 29738bf..4f01dd8 100644
--- a/Makefile
+++ b/Makefile
@@ -27,7 +27,7 @@ TAP_EXEC := ./node_modules/.bin/tap
 TEST_UNIT_JOBS ?= 4
 
 
-NODE_PREBUILT_VERSION=v6.14.0
+NODE_PREBUILT_VERSION=v6.15.0
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=gz
 	# sdc-minimal-multiarch-lts 15.4.1
