From c3525d167099561dae86331d087530ae4e4175a2 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 3 Dec 2018 16:19:50 -0500
Subject: [PATCH] TOOLS-2117 sdcnode updates for November 2018 node.js security
 releases

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 29738bf..42372f9 100644
--- a/Makefile
+++ b/Makefile
@@ -27,7 +27,7 @@ TAP_EXEC := ./node_modules/.bin/tap
 TEST_UNIT_JOBS ?= 4
 
 
-NODE_PREBUILT_VERSION=v6.14.0
+NODE_PREBUILT_VERSION=v6.15.1
 ifeq ($(shell uname -s),SunOS)
 	NODE_PREBUILT_TAG=gz
 	# sdc-minimal-multiarch-lts 15.4.1
-- 
2.21.0

