From 44a43411ec05132e06a091ceeaa928a34e6d203b Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Fri, 21 Sep 2018 17:08:57 +0000
Subject: [PATCH] OS-7258 sysinfo needs to indicate if a PI can run kvm and
 bhyve concurrently

---
 src/sysinfo | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/sysinfo b/src/sysinfo
index 106da2ba..79239247 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -219,6 +219,7 @@ function get_bhyve_capability()
             if [[ -n ${vcpus} ]]; then
                 Bhyve_Max_Vcpus=${vcpus}
             fi
+	    HVM_API="true"
         fi
     fi
 }
@@ -826,6 +827,7 @@ END
   "VM Capable": ${VM_Capable},
   "Bhyve Capable": ${Bhyve_Capable},
   "Bhyve Max Vcpus": ${Bhyve_Max_Vcpus},
+  "HVM API": ${HVM_API},
   "CPU Type": "${CPU_Version}",
   "CPU Virtualization": "${CPU_Virtualization}",
   "CPU Physical Cores": ${CPU_Count},
-- 
2.21.0

