From 982def7e7c03bf4611e08ca5550f42b1392f0914 Mon Sep 17 00:00:00 2001
From: Mike Gerdts <mike.gerdts@joyent.com>
Date: Fri, 21 Sep 2018 17:08:57 +0000
Subject: [PATCH] OS-7258 sysinfo needs to indicate if a PI can run kvm and
 bhyve concurrently

---
 src/sysinfo | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/sysinfo b/src/sysinfo
index 106da2ba..f1dbf980 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -219,6 +219,7 @@ function get_bhyve_capability()
             if [[ -n ${vcpus} ]]; then
                 Bhyve_Max_Vcpus=${vcpus}
             fi
+            HVM_API="true"
         fi
     fi
 }
@@ -691,6 +692,7 @@ CPU_Virtualization='${CPU_Virtualization}'
 CPU_Physical_Cores=${CPU_Count}
 Bhyve_Capable='${Bhyve_Capable}'
 Bhyve_Max_Vcpus=${Bhyve_Max_Vcpus}
+HVM_API='${HVM_API}'
 Nic_Tags=${NicTagList}
 Admin_NIC_Tag=${Admin_NIC_Tag}
 Setup='${Setup_complete}'
@@ -826,6 +828,7 @@ END
   "VM Capable": ${VM_Capable},
   "Bhyve Capable": ${Bhyve_Capable},
   "Bhyve Max Vcpus": ${Bhyve_Max_Vcpus},
+  "HVM API": ${HVM_API},
   "CPU Type": "${CPU_Version}",
   "CPU Virtualization": "${CPU_Virtualization}",
   "CPU Physical Cores": ${CPU_Count},
-- 
2.21.0

