From 7ffd2a688d046bdb05165a5d4e0bf72ee85b4af2 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Thu, 20 Jul 2017 17:39:06 -0700
Subject: [PATCH] HEAD-2363 Allow coal vm resizing on Linux

---
 bin/build-coal-image | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/bin/build-coal-image b/bin/build-coal-image
index 746ee871..2134307e 100755
--- a/bin/build-coal-image
+++ b/bin/build-coal-image
@@ -345,8 +345,9 @@ function create_output
     coal_zpool_disk_size=$(build_spec coal-zpool-disk-size)
     # BEGIN BASHSTYLED
     if [[ -n ${coal_zpool_disk_size} ]]; then
-        if [[ -x "/Applications/VMware Fusion.app/Contents/Library/vmware-vdiskmanager" ]]; then
-            /Applications/VMware\ Fusion.app/Contents/Library/vmware-vdiskmanager \
+        vdiskmanager=$(which vmware-vdiskmanager || echo "/Applications/VMware Fusion.app/Contents/Library/vmware-vdiskmanager")
+        if [[ -x "${vdiskmanager}" ]]; then
+            "${vdiskmanager}" \
                 -x ${coal_zpool_disk_size}GB ${STAGE}/USB-headnode.vmwarevm/zpool.vmdk
         else
             echo " !!! WARNING !!! Cannot resize zpool disk with missing vmware-vdiskmanager!" >&2
-- 
2.21.0

