commit 1dad88ef1cdba88672bf7084537ed0673cbfdb64 (refs/changes/83/3383/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-02-16T12:19:35-08:00 (1 year, 8 months ago)
    
    TRITON-124 add node-triton support for bhyve

diff --git a/lib/do_fwrule/do_instances.js b/lib/do_fwrule/do_instances.js
index 53bc6bf..b549f16 100644
--- a/lib/do_fwrule/do_instances.js
+++ b/lib/do_fwrule/do_instances.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  *
  * `triton fwrule instances ...`
  */
@@ -111,6 +111,7 @@ function do_instances(subcmd, opts, args, cb) {
                     common.uuidToShortId(inst.image);
                 inst.shortid = inst.id.split('-', 1)[0];
                 var flags = [];
+                if (inst.brand === 'bhyve') flags.push('B');
                 if (inst.docker) flags.push('D');
                 if (inst.firewall_enabled) flags.push('F');
                 if (inst.brand === 'kvm') flags.push('K');
@@ -159,6 +160,7 @@ do_instances.help = [
     'for convenience):',
     '    shortid*           A short ID prefix.',
     '    flags*             Single letter flags summarizing some fields:',
+    '                           "B" the brand is "bhyve"',
     '                           "D" docker instance',
     '                           "F" firewall is enabled',
     '                           "K" the brand is "kvm"',
diff --git a/lib/do_instance/do_create.js b/lib/do_instance/do_create.js
index 47b7e18..48943c8 100644
--- a/lib/do_instance/do_create.js
+++ b/lib/do_instance/do_create.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2017 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  *
  * `triton instance create ...`
  */
@@ -376,6 +376,9 @@ function do_create(subcmd, opts, args, cb) {
                     function (net) { return net.id; })
             };
 
+            if (opts.brand) {
+                createOpts.brand = opts.brand;
+            }
             if (ctx.volMounts) {
                 createOpts.volumes = ctx.volMounts;
             }
@@ -492,6 +495,12 @@ do_create.options = [
     {
         group: 'Create options'
     },
+    {
+        names: ['brand'],
+        helpArg: 'BRAND',
+        type: 'string',
+        help: 'Override the default brand for this instance.'
+    },
     {
         names: ['name', 'n'],
         helpArg: 'NAME',
diff --git a/lib/do_instance/do_list.js b/lib/do_instance/do_list.js
index 2619a3a..23e7c22 100644
--- a/lib/do_instance/do_list.js
+++ b/lib/do_instance/do_list.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  *
  * `triton instance list ...`
  */
@@ -152,6 +152,7 @@ function do_list(subcmd, opts, args, callback) {
                 var flags = [];
                 if (inst.docker) flags.push('D');
                 if (inst.firewall_enabled) flags.push('F');
+                if (inst.brand === 'bhyve') flags.push('B');
                 if (inst.brand === 'kvm') flags.push('K');
                 inst.flags = flags.length ? flags.join('') : undefined;
             });
@@ -208,6 +209,7 @@ do_list.help = [
     'for convenience):',
     '    shortid*           A short ID prefix.',
     '    flags*             Single letter flags summarizing some fields:',
+    '                           "B" the brand is "bhyve"',
     '                           "D" docker instance',
     '                           "F" firewall is enabled',
     '                           "K" the brand is "kvm"',
diff --git a/lib/do_instance/do_snapshot/do_create.js b/lib/do_instance/do_snapshot/do_create.js
index de7b5cf..62df2f6 100644
--- a/lib/do_instance/do_snapshot/do_create.js
+++ b/lib/do_instance/do_snapshot/do_create.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2018 Joyent, Inc.
  *
  * `triton snapshot create ...`
  */
@@ -133,7 +133,7 @@ do_create.help = [
     '{{usage}}',
     '',
     '{{options}}',
-    'Snapshot do not work for instances of type "kvm".'
+    'Snapshots do not work for instances of type "bhyve" or "kvm".'
 ].join('\n');
 
 do_create.completionArgtypes = ['tritoninstance', 'none'];
diff --git a/package.json b/package.json
index 0d00f3f..2a43292 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "triton",
   "description": "Joyent Triton CLI and client (https://www.joyent.com/triton)",
-  "version": "5.6.1",
+  "version": "5.7.1",
   "author": "Joyent (joyent.com)",
   "homepage": "https://github.com/joyent/node-triton",
   "dependencies": {
