From e3248c2ba7e6d4e7b9647add7861b10223f42b62 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 17 Jan 2017 16:41:58 -0800
Subject: [PATCH] HEAD-2342 sdc-healthcheck's cnapi check doesn't like multiple
 headnodes

---
 tools/bin/sdc-healthcheck | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/tools/bin/sdc-healthcheck b/tools/bin/sdc-healthcheck
index c4727a6f..72f0274a 100755
--- a/tools/bin/sdc-healthcheck
+++ b/tools/bin/sdc-healthcheck
@@ -81,9 +81,9 @@ function check_end_to_end()
       [[ $CLOUDAPI_STATUS -eq 0 ]] && return
       ;;
     cnapi)
-      local got_name=`sdc-cnapi /servers?headnode=true 2>/dev/null | \
-          json -H -a hostname`
-      [[ "$got_name" == "$(hostname)" ]] && return
+      # Allow for multiple headnodes: one of them should be this hostname.
+      sdc-cnapi /servers?headnode=true 2>/dev/null | json -H -a hostname \
+        | grep "^$(hostname)$" >/dev/null && return
       ;;
     fwapi)
       local got_status=`sdc-fwapi /rules 2>/dev/null | \
-- 
2.21.0

