commit af543ffeb1e841b7fa447b3b4e319e433eeb8517 (refs/changes/32/2432/11)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-09-13T22:04:44+00:00 (2 years, 1 month ago)
    
    MANTA-3220 local edits

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 402f6a9..c6e5841 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -49,9 +49,7 @@ function ensureSharkHasPool(req, attempted, shark, callback) {
             shark: shark,
             pool: pool.getState()
         }, 'ensureSharkHasPool: got existing pool');
-    }
-
-    if (!pool) {
+    } else {
         req.sharkAgent.createPool(shark.manta_storage_id);
         pool = req.sharkAgent.getPool(host);
         req.log.debug({
@@ -133,32 +131,30 @@ function chooseSharks(req, res, size, copies, cb) {
             if (err) {
                 cb(err);
             } else {
-                var ndx = 0;
                 var attempted = [];
+                var funcs = [];
 
-                function attempt(inputs) {
+                function tryShark (shark, callback) {
                     vasync.forEachParallel({
                         func: ensureSharkHasPool.bind(null, req, attempted),
-                        inputs: inputs
-                    }, function (err2, results) {
-                        if (err2 || results.successes.length < copies) {
-                            if (ndx < sharks.length) {
-                                attempt(sharks[ndx++]);
-                            } else {
-                                cb(new SharksExhaustedError(res), null);
-                            }
-                            return;
-                        }
-                        log.info({
-                            sharks: attempted
-                        }, 'chooseSharks: unsuccessfully attempted sharks');
-                        log.info({
-                            sharks: results.successes
-                        }, 'chooseSharks: sharks chosen');
-                        cb(null, results.successes);
-                    });
+                        inputs: shark
+                    }, callback);
                 }
-                attempt(sharks[ndx++]);
+
+                sharks.forEach(function (shark) {
+                    funcs.push(tryShark.bind(null, shark));
+                });
+                vasync.tryEach(funcs, function (err, results) {
+                    if (err) {
+                        cb(SharksExhaustedError(res), null);
+                        return;
+                    }
+                    log.debug({
+                        attempted: attempted,
+                        chosen: results.successes
+                    }, 'chooseSharks results');
+                    cb(null, results.successes);
+                });
             }
         });
     }
diff --git a/package.json b/package.json
index 38b5866..150aaa1 100644
--- a/package.json
+++ b/package.json
@@ -34,7 +34,7 @@
         "moray": "3.1.1",
         "once": "1.3.0",
         "restify": "2.6.3",
-        "vasync": "^1.5.0",
+        "vasync": "2.0.0",
         "verror": "^1.9.0",
         "watershed": "0.3.0",
         "xtend": "2.1.1"
