From 8cf945a6fe62d34a58c8ae32402698767a3392c4 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Wed, 19 Oct 2016 13:43:25 -0700
Subject: [PATCH] MANTA-3009 Update muskie node version for CVE-2016-5180

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6fce20d..64b5393 100644
--- a/Makefile
+++ b/Makefile
@@ -55,7 +55,7 @@ CLEAN_FILES += node_modules
 #
 NAME 			= muskie
 NODE_PREBUILT_TAG       = zone
-NODE_PREBUILT_VERSION	:= v0.10.47
+NODE_PREBUILT_VERSION	:= v0.10.48
 NODE_PREBUILT_IMAGE     = fd2cc906-8938-11e3-beab-4359c665ac99
 
 include ./tools/mk/Makefile.defs
-- 
2.21.0

