commit 81a7790e5003a3e01b19eb1f890bce07904e1720 (refs/changes/17/1017/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:06-05:00 (2 years, 10 months ago)
    
     ./bin/triton network list

diff --git a/lib/do_network/do_list.js b/lib/do_network/do_list.js
index b3bd3c9..7cc963c 100644
--- a/lib/do_network/do_list.js
+++ b/lib/do_network/do_list.js
@@ -49,28 +49,34 @@ function do_list(subcmd, opts, args, callback) {
     columns = columns.split(',');
 
     var sort = opts.s.split(',');
+    var tritonapi = this.top.tritonapi;
 
-    this.top.tritonapi.cloudapi.listNetworks(function (err, networks) {
-        if (err) {
-            callback(err);
-            return;
+    common.cliSetupTritonApi({cli: this.top}, function onSetup(setupErr) {
+        if (setupErr) {
+            callback(setupErr);
         }
+        tritonapi.cloudapi.listNetworks(function (err, networks) {
+            if (err) {
+                callback(err);
+                return;
+            }
 
-        if (opts.json) {
-            common.jsonStream(networks);
-        } else {
-            for (var i = 0; i < networks.length; i++) {
-                var net = networks[i];
-                net.shortid = net.id.split('-', 1)[0];
-                net.vlan = net.vlan_id;
+            if (opts.json) {
+                common.jsonStream(networks);
+            } else {
+                for (var i = 0; i < networks.length; i++) {
+                    var net = networks[i];
+                    net.shortid = net.id.split('-', 1)[0];
+                    net.vlan = net.vlan_id;
+                }
+                tabula(networks, {
+                    skipHeader: opts.H,
+                    columns: columns,
+                    sort: sort
+                });
             }
-            tabula(networks, {
-                skipHeader: opts.H,
-                columns: columns,
-                sort: sort
-            });
-        }
-        callback();
+            callback();
+        });
     });
 }
 
