commit 246f3b4591c7b365362f4f6fb5c7c2a39b71df5d (refs/changes/74/1674/1)
Author: Julien Gilli <julien.gilli@joyent.com>
Date:   2017-03-16T10:34:39-07:00 (2 years, 7 months ago)
    
    DOCKER-1019 cli-local-volumes.test.js fails in nightly

diff --git a/test/integration/cli-local-volumes.test.js b/test/integration/cli-local-volumes.test.js
index c7f6670..7c43f22 100644
--- a/test/integration/cli-local-volumes.test.js
+++ b/test/integration/cli-local-volumes.test.js
@@ -51,9 +51,9 @@ test('setup', function (tt) {
 
 test('docker local volumes', function (tt) {
     var containerWithLocalVolName =
-        common.makeResourceName('local-volume-test-container-with-local-vol');
+        common.makeContainerName('local-volume-test-container-with-local-vol');
     var mountingContainerName =
-        common.makeResourceName('local-volume-test-mounting-container');
+        common.makeContainerName('local-volume-test-mounting-container');
 
     tt.test('creating container with local volume should succeed',
         function (t) {
