From 9975fa5919c4bf7ca901be07b261929454d7f694 Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Mon, 8 Jul 2019 09:47:56 +0000
Subject: [PATCH] joyent/node-zkstream#9 use var instead of const so we can run
 in strict mode

---
 README.md | 1 +
 1 file changed, 1 insertion(+)

diff --git a/README.md b/README.md
index ca63117..30c4a3d 100644
--- a/README.md
+++ b/README.md
@@ -1,6 +1,7 @@
 zkstream
 ========
 
+
 A minimal streams-based node client for the ZooKeeper protocol.
 
 API
-- 
2.21.0

