From 97ef72ecc6fba6ce066ea64459b33752480c2f52 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Mon, 23 Jan 2017 16:09:46 +0000
Subject: [PATCH] OS-5912 missing 32-bit libficl-sys.so.4.1.0

---
 manifest | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/manifest b/manifest
index 8e2487b3a8..1a43d98cf0 100644
--- a/manifest
+++ b/manifest
@@ -19,7 +19,7 @@
 # CDDL HEADER END
 #
 #
-# Copyright 2016 Joyent, Inc.
+# Copyright 2017 Joyent, Inc.
 #
 # This lists all the files that illumos-joyent delivers as part of a build.
 # Keep this sorted as by LC_ALL=C sort -u -k2.
@@ -6196,6 +6196,7 @@ f usr/lib/libfcoe.so.1 0755 root bin
 s usr/lib/libfcoe.so=libfcoe.so.1
 s usr/lib/libfdisk.so.1=../../lib/libfdisk.so.1
 s usr/lib/libfdisk.so=../../lib/libfdisk.so.1
+f usr/lib/libficl-sys.so.4.1.0 0755 root bin
 f usr/lib/libform.so.1 0755 root bin
 s usr/lib/libform.so=libform.so.1
 f usr/lib/libfru.so.1 0755 root bin
-- 
2.21.0

