commit 5fd3c7d6a37abb2ce9103565e3107c255c25d912
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-07-05T08:46:31-06:00 (3 months ago)
    
    MANTA-4389 Buckets object deletes failing with 500 error

diff --git a/lib/server.js b/lib/server.js
index a39ce4d..004b074 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -1082,9 +1082,10 @@ function deleteObject(options) {
                     obj: obj
                 }, 'deleteObject: returned');
 
-                // MANTA-1400: set the vnode info for debugging purposes
-                if (obj) {
-                    obj._node = location;
+                if (typeof (obj) !== 'object') {
+                    obj = {
+                        deleted_rows: obj
+                    };
                 }
 
                 log.debug({
@@ -1098,6 +1099,8 @@ function deleteObject(options) {
                 if (gErr) {
                     rpc.fail(gErr);
                 } else {
+                    obj._node = location;
+
                     rpc.write(obj);
                     rpc.end();
                 }
