commit 1c336c8927031764ebc4fd91e564da6fa8c406e8
Author: Robert Mustacchi <rm@joyent.com>
Date:   2019-08-08T00:04:09+00:00 (9 weeks ago)
    
    OS-7925 rw_exit lockstat patch point has garbage %rcx

diff --git a/usr/src/uts/intel/ia32/ml/lock_prim.s b/usr/src/uts/intel/ia32/ml/lock_prim.s
index e67b1ef903..f4d44202aa 100644
--- a/usr/src/uts/intel/ia32/ml/lock_prim.s
+++ b/usr/src/uts/intel/ia32/ml/lock_prim.s
@@ -970,6 +970,7 @@ rw_exit(krwlock_t *lp)
 	jnz	rw_exit_wakeup
 .rw_read_exit_lockstat_patch_point:
 	ret
+	movq	%gs:CPU_THREAD, %rcx		/* rcx = thread ptr */
 	movq	%rdi, %rsi			/* rsi = lock ptr */
 	movl	$LS_RW_EXIT_RELEASE, %edi
 	movl	$RW_READER, %edx
