commit 39b17d38e0f1fa27bf38d18ce52f822a62cb974e (refs/changes/19/4219/2)
Author: dyep49 <dyep49@gmail.com>
Date:   2018-06-14T12:17:12-07:00 (1 year, 4 months ago)
    
    TRITON-490 Add metricPorts metadata to SAPI for cmon-agent

diff --git a/boot/setup.sh b/boot/setup.sh
index 6969b9e..00440d6 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -124,6 +124,9 @@ sdc_log_rotation_add registrar /var/svc/log/*registrar*.log 1g
 sdc_log_rotation_add $role /var/svc/log/*$role*.log 1g
 sdc_log_rotation_setup_end
 
+# Add metricsPorts metadata for cmon-agent discovery
+mdata-put metricPorts 8881
+
 # All done, run boilerplate end-of-setup
 sdc_setup_complete
 
diff --git a/package.json b/package.json
index 7c5974a..ecfea9c 100644
--- a/package.json
+++ b/package.json
@@ -22,7 +22,7 @@
     "sdc-clients": "10.5.0",
     "semver": "5.4.1",
     "sprintf-js": "0.0.7",
-    "triton-metrics": "0.1.0",
+    "triton-metrics": "0.1.1",
     "vasync": "2.2.0",
     "verror": "1.10.0"
   },
