commit 39d1ca5dacefce55792c4ab69842cfc49cd69b84 (refs/changes/57/357/1)
Author: Brian Bennett <brian.bennett@joyent.com>
Date:   2016-08-26T13:48:55-07:00 (3 years, 1 month ago)
    
    OS-5634 system-log SMF method has existential confusion

diff --git a/overlay/generic/lib/svc/method/system-log b/overlay/generic/lib/svc/method/system-log
old mode 100644
new mode 100755
index f8a60a36..e2fc5b2e
--- a/overlay/generic/lib/svc/method/system-log
+++ b/overlay/generic/lib/svc/method/system-log
@@ -103,20 +103,21 @@ fi
 
 if [ -f /etc/rsyslog.conf ]; then
 	if [ ! -x /usr/sbin/rsyslogd ]; then
-		echo "/etc/rsyslog.conf is missing. Exiting."
+		echo "Executable /usr/sbin/rsyslogd is not found. Exiting."
 		exit $SMF_EXIT_ERR_CONFIG
 	fi
 else
 	#
 	# Fall back to old syslogd
 	#
+    echo "/etc/rsyslog.conf is missing. Using syslogd instead."
 	if [ ! -f /etc/syslog.conf ]; then
 		echo "/etc/syslog.conf is missing. Exiting."
 		exit $SMF_EXIT_ERR_CONFIG
 	fi
 
 	if [ ! -x /usr/sbin/syslogd ]; then
-		echo "Executable /usr/sbin/syslogd not found. Exiting"
+		echo "Executable /usr/sbin/syslogd not found. Exiting."
 		exit $SMF_EXIT_ERR_CONFIG
 	fi
 fi
