From 5270f30130c033de15d1104d94c441877bd72496 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 8 Jun 2018 10:40:52 +0000
Subject: [PATCH] TRITON-478 agent builds fail as non-root when postinstall.sh
 writes to /

---
 npm/postinstall.sh | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/npm/postinstall.sh b/npm/postinstall.sh
index fe6116b..a961a0f 100755
--- a/npm/postinstall.sh
+++ b/npm/postinstall.sh
@@ -6,9 +6,14 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
+if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
+    echo "Skipping config-agent postinstall (assuming build-time install)"
+    exit 0
+fi
+
 if [[ -n "$TRACE" ]]; then
     export PS4='[\D{%FT%TZ}] ${BASH_SOURCE}:${LINENO}: ${FUNCNAME[0]:+${FUNCNAME[0]}(): }'
     set -o xtrace
-- 
2.21.0

