From 4219b62aa9d1157d2c0250ac20091ba39b34ed48 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 8 Jun 2018 10:40:52 +0000
Subject: [PATCH] TRITON-478 agent builds fail as non-root when postinstall.sh
 writes to /

---
 npm/postinstall.sh | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/npm/postinstall.sh b/npm/postinstall.sh
index fe6116b..27dd466 100755
--- a/npm/postinstall.sh
+++ b/npm/postinstall.sh
@@ -6,7 +6,7 @@
 #
 
 #
-# Copyright (c) 2015, Joyent, Inc.
+# Copyright (c) 2018, Joyent, Inc.
 #
 
 if [[ -n "$TRACE" ]]; then
@@ -16,6 +16,11 @@ fi
 set -o errexit
 set -o pipefail
 
+if [[ -z "$npm_config_smfdir" || ! -d "$npm_config_smfdir" ]]; then
+    echo "Skipping config-agent postinstall (assuming build-time install)"
+    exit 0
+fi
+
 DIR=`dirname $0`
 ROOT=$(cd `dirname $0`/../ 2>/dev/null && pwd)
 
-- 
2.21.0

