commit d217613cb2a59dfbc5496e36516057d6d2a98612 (refs/changes/97/997/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:56:05-05:00 (2 years, 10 months ago)
    
    ./bin/triton image wait base-32

diff --git a/lib/do_image/do_wait.js b/lib/do_image/do_wait.js
index c887cec..1421ae8 100644
--- a/lib/do_image/do_wait.js
+++ b/lib/do_image/do_wait.js
@@ -12,6 +12,7 @@
 
 var vasync = require('vasync');
 
+var common = require('../common');
 var distractions = require('../distractions');
 var errors = require('../errors');
 
@@ -34,7 +35,8 @@ function do_wait(subcmd, opts, args, cb) {
     var done = 0;
     var imgFromId = {};
 
-    vasync.pipeline({funcs: [
+    vasync.pipeline({arg: {cli: this.top}, funcs: [
+        common.cliSetupTritonApi,
         function getImgs(_, next) {
             vasync.forEachParallel({
                 inputs: ids,
