commit 2a994ded91aaab24499119fc2adc3a6067b1faa5 (refs/changes/22/322/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-08-19T10:29:55-07:00 (3 years, 2 months ago)
    
    TOOLS-1529 crimport: the GitHub team is "Joyent Engineering"

diff --git a/bin/crimport b/bin/crimport
index a2ba1bb..0f70dcc 100755
--- a/bin/crimport
+++ b/bin/crimport
@@ -130,9 +130,9 @@ Success!  The following project has been created on the Gerrit server:
 
     https://$igp_gerrit_host/#/admin/projects/$igp_gerrit_project
 
-To make sure replication to GitHub will work, please make sure that the Joyent
-"Engineering" team on GitHub has write access to the GitHub project by visiting
-this page:
+To make sure replication to GitHub will work, please make sure that the
+"Joyent Engineering" team on GitHub has write access to the GitHub project
+by visiting this page:
 
     https://github.com/$igp_gerrit_project/settings/collaboration
 
