From 6bf6bd83fa9698cf79492a06cba1e8c2ef7636f6 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Thu, 18 Oct 2018 14:41:15 -0400
Subject: [PATCH] OS-7307 vminfo status can show unsafe characters

---
 src/vm/sbin/vminfo.js | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/vm/sbin/vminfo.js b/src/vm/sbin/vminfo.js
index 5e46e9aa..9f7fe5c2 100755
--- a/src/vm/sbin/vminfo.js
+++ b/src/vm/sbin/vminfo.js
@@ -51,6 +51,17 @@ var client = new vminfod.VminfodClient({
     log: log
 });
 
+/*
+ * Safely escape the user-agent value to be printed to the screen.  This was
+ * taken from `escapePath` in node-manta (joyent/node-manta#231).
+ */
+function safeUserAgent(s) {
+    assert.string(s, 'safeUserAgent');
+
+    /* JSSTYLED */
+    return JSON.stringify(s).replace(/^"|"$/g, '').replace(/\\"/g, '"');
+}
+
 function usage() {
     var _args = Array.prototype.slice.call(arguments);
     var msg = f.apply(null, _args);
@@ -183,7 +194,7 @@ function do_status(args) {
         output.push(f('eventsListeners: (%d listeners)', evls.length));
         evls.forEach(function forEachEvLs(uuid) {
             var el = msg.eventsListeners[uuid];
-            output.push(f('  - %s', el.userAgent));
+            output.push(f('  - %s', safeUserAgent(el.userAgent)));
             output.push(f('    %s created %s ago', uuid, el.createdAgo));
             output.push('');
         });
-- 
2.21.0

