commit 58592e65867a8f698478e0afa5a8b496c6981dd4 (refs/changes/25/2625/1)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2017-09-14T23:01:32+00:00 (2 years, 1 month ago)
    
    for testing: MANTA-3279 non-version change

diff --git a/boot/setup.sh b/boot/setup.sh
index 4d6a589..b48f280 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -102,8 +102,8 @@ function manta_setup_leveldb_hash_ring {
     do
         cp -R $leveldb_ring $dir
         [[ $? -eq 0 ]] || fatal "unable to setup leveldb"
-        # test with get_node on the newly created ring
-        $FASH get_node -l $dir -b leveldb yunong
+        # test with get-node on the newly created ring
+        $FASH get-node -l $dir -b leveldb yunong
         [[ $? -eq 0 ]] || fatal "unable to setup leveldb"
     done
     ZFS_SNAPSHOT=$ZFS_DATASET@$(date +%s)000
