From 58592e65867a8f698478e0afa5a8b496c6981dd4 Mon Sep 17 00:00:00 2001
From: Brittany Wald <brittany.wald@joyent.com>
Date: Thu, 14 Sep 2017 22:53:42 +0000
Subject: [PATCH] for testing: MANTA-3279 non-version change

---
 boot/setup.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/boot/setup.sh b/boot/setup.sh
index 4d6a589..b48f280 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -102,8 +102,8 @@ function manta_setup_leveldb_hash_ring {
     do
         cp -R $leveldb_ring $dir
         [[ $? -eq 0 ]] || fatal "unable to setup leveldb"
-        # test with get_node on the newly created ring
-        $FASH get_node -l $dir -b leveldb yunong
+        # test with get-node on the newly created ring
+        $FASH get-node -l $dir -b leveldb yunong
         [[ $? -eq 0 ]] || fatal "unable to setup leveldb"
     done
     ZFS_SNAPSHOT=$ZFS_DATASET@$(date +%s)000
-- 
2.21.0

