commit c834c37108e19db7f6885c382f83946020e8b844 (refs/changes/15/2315/2)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-08-31T18:49:29+00:00 (2 years, 1 month ago)
    
    MORAY-397 Remove query timeouts unless requested by client

diff --git a/test/sql.test.js b/test/sql.test.js
index 1a844fa..b53784c 100644
--- a/test/sql.test.js
+++ b/test/sql.test.js
@@ -102,3 +102,20 @@ test('sql - cleanup', function (t) {
         t.end();
     });
 });
+
+test('sql - client timeout respected', function (t) {
+    q = c.sql('SELECT pg_sleep(5);', [], { timeout: 1000 });
+    q.on('error', function (err) {
+        t.ok(VError.hasCauseWithName(err, 'QueryTimeoutError'),
+            'correct error');
+        t.end();
+    });
+    q.on('record', function (r) {
+        t.fail('no row should be returned');
+        t.deepEqual(r, null, 'no row should be returned - value');
+    });
+    q.on('end', function (r) {
+        t.fail('"error" should have been emitted');
+        t.end();
+    });
+});
