commit c01a0b2f2adaa29b3600a9f444112858c7f784cb
Author: Mohamed Khalfella <mohamed.khalfella@joyent.com>
Date:   2019-09-06T07:49:42+00:00 (5 weeks ago)
    
    OS-7974 bhyve does not persist nic_driver

diff --git a/src/vm/node_modules/proptable.js b/src/vm/node_modules/proptable.js
index d9b7fdec..9042627d 100644
--- a/src/vm/node_modules/proptable.js
+++ b/src/vm/node_modules/proptable.js
@@ -1156,6 +1156,7 @@ exports.properties = {
     }, nic_driver: {
         payload: {
             allowed: {
+                'bhyve': ['create', 'receive', 'update'],
                 'kvm': ['create', 'receive', 'update']
             },
             type: 'string'
