commit e9dfc1a1d1b92b94b3c478887d9be0c416287911 (refs/changes/22/1322/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-01-24T23:19:41+08:00 (2 years, 8 months ago)
    
    PUBAPI-1359: Test failing due to invalid CIDR.

diff --git a/test/machines/firewall-rules.js b/test/machines/firewall-rules.js
index 85f03d6..4cb5b0f 100644
--- a/test/machines/firewall-rules.js
+++ b/test/machines/firewall-rules.js
@@ -115,7 +115,7 @@ module.exports = function (suite, client, other, machine, callback) {
     suite.test('AddRule - other', function (t) {
         other.post(RULES_URL, {
             rule: 'FROM vm ' + machine +
-                ' TO subnet 10.99.99.0/23 ALLOW tcp port 80'
+                ' TO subnet 10.99.98.0/23 ALLOW tcp port 80'
         }, function (err, req, res, body) {
             t.ok(err);
 
