commit 534dacb79ea3874ae5d4f6ceb2354a9fd17daf1a
Author: Tim Foster <tim.foster@joyent.com>
Date:   2019-08-12T15:40:38+01:00 (8 weeks ago)
    
    OS-7921 SmartOS bash_completions should complete alias names

diff --git a/manifest b/manifest
index 19d79fec68..79627bb02f 100644
--- a/manifest
+++ b/manifest
@@ -268,6 +268,7 @@ f etc/bash/bash_completion 0644 root sys
 d etc/bash/bash_completion.d 0755 root sys
 f etc/bash/bash_completion.d/dladm 0644 root sys
 f etc/bash/bash_completion.d/zones 0644 root sys
+f etc/bash/bash_completion.d/zone_alias 0644 root sys
 d etc/brand 0755 root sys
 d etc/certs 0755 root sys
 d etc/cron.d 0755 root sys
diff --git a/usr/src/cmd/nsadmin/Makefile b/usr/src/cmd/nsadmin/Makefile
index 819a69340c..3af33b2a4d 100644
--- a/usr/src/cmd/nsadmin/Makefile
+++ b/usr/src/cmd/nsadmin/Makefile
@@ -28,7 +28,7 @@
 ETCFILES = profile .login ksh.kshrc system zshrc
 ETCSKELFILES = .profile .kshrc .bashrc
 ETCBASHFILES = bash_completion
-ETCBASHCOMPLETIONDFILES = dladm zones
+ETCBASHCOMPLETIONDFILES = dladm zones zone_alias
 ROOTFILES = .profile .bashrc .bash_profile
 
 include ../Makefile.cmd
diff --git a/usr/src/cmd/nsadmin/bash/bash_completion.d/zone_alias b/usr/src/cmd/nsadmin/bash/bash_completion.d/zone_alias
new file mode 100644
index 0000000000..28892da25c
--- /dev/null
+++ b/usr/src/cmd/nsadmin/bash/bash_completion.d/zone_alias
@@ -0,0 +1,39 @@
+_zone_alias()
+{
+    # Attempt alias -> uuid mapping
+    alias="${COMP_WORDS[COMP_CWORD]}"
+    RES=$(vmadm list -H -o alias,uuid | awk "
+        BEGIN { count = 0; alias = \"\"; uuid = \"\"; }
+        /^$alias / { count = 1 ; aliases[1] = \$1; uuid = \$2; exit;}
+        /^$alias/ { count += 1 ; aliases[count]="\$1"; uuid=\$2 }
+        END { if (count == 1) {
+                  print \"smartos_alias_completed \" aliases[1] \" \" uuid;
+              } else {
+                  for (i=1; i <= count; i++) {
+                      print aliases[i]
+                  }
+              }
+        }
+     ")
+     results_arr=( $RES )
+     if [[ "${results_arr[0]}" == "smartos_alias_completed" ]]; then
+         completed_alias="${results_arr[1]}"
+         uuid="${results_arr[2]}"
+         if [[ -n "$SMARTOS_MULTILINE_ALIAS_COMPLETION" ]]; then
+             tput sc
+             tput bold
+             echo "[completed alias: ${alias}]"
+             tput rc
+             tput sgr0
+             COMPREPLY=$uuid
+         elif [[ "$completed_alias" == "$alias" ]]; then
+             COMPREPLY=$uuid
+         else
+             COMPREPLY=( $(compgen -W "$completed_alias" --) )
+             compopt -o nospace
+         fi
+     else
+         COMPREPLY=( $(compgen -W "$RES" -- ${cur}) )
+     fi
+     return 0
+}
diff --git a/usr/src/cmd/nsadmin/bash/bash_completion.d/zones b/usr/src/cmd/nsadmin/bash/bash_completion.d/zones
index 2c68ba366e..e4bc5b4155 100644
--- a/usr/src/cmd/nsadmin/bash/bash_completion.d/zones
+++ b/usr/src/cmd/nsadmin/bash/bash_completion.d/zones
@@ -10,6 +10,11 @@ _zlogin()
         local running=$(zoneadm list | grep -v '^global$')
         COMPREPLY=( $(compgen -W "${running}" -- ${cur}) )
     fi
+    if [[ -n "$COMPREPLY" ]]; then
+      return 0
+    fi
+
+    _zone_alias
     return 0
 }
 
@@ -24,6 +29,10 @@ _dash_z_zone()
     if [[ ${prev} =~ -.*z$ ]]; then
         local zones=$(zoneadm list -c | grep -v '^global$')
         COMPREPLY=( $(compgen -W "${zones}" -- ${cur}) )
+        if [[ -n "$COMPREPLY" ]]; then
+          return 0
+        fi
+        _zone_alias
     fi
     return 0
 }
