commit 4b33312f8d229b3f70b530079e233fd0b7137e42
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-12T17:46:21-04:00 (7 months ago)
    
    TOOLS-2186 sdcnode updates for Feburary 2019 node.js security releases

diff --git a/deps/hermes b/deps/hermes
index f39da64..f607e55 160000
--- a/deps/hermes
+++ b/deps/hermes
@@ -1 +1 @@
-Subproject commit f39da64fb3ce370e0f3b24d8d96a9c5e6ecd0a79
+Subproject commit f607e55d1902dca6a3e97b1d7d72071a3be0ec93
