commit cb24da3e8652cc169f1d5ef4e8b03d35da6e9fbb (refs/changes/36/4636/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-08-06T17:04:36+02:00 (1 year, 2 months ago)
    
    TRITON-666 Fix sdcadm test broken by TRITON-558

diff --git a/test/update.test.js b/test/update.test.js
index 10679db..0fe97ce 100644
--- a/test/update.test.js
+++ b/test/update.test.js
@@ -215,7 +215,6 @@ test('update non-HA moray and SAPI consecutively', function (t) {
             'Reprovisioning VM',
             'Destroying tmp VM',
             'Updating sapi',
-            'Verifying SAPI full mode',
             'Waiting for sapi',
             'Updated successfully'
         ];
