From cb24da3e8652cc169f1d5ef4e8b03d35da6e9fbb Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Mon, 6 Aug 2018 17:04:36 +0200
Subject: [PATCH] TRITON-666 Fix sdcadm test broken by TRITON-558

---
 test/update.test.js | 1 -
 1 file changed, 1 deletion(-)

diff --git a/test/update.test.js b/test/update.test.js
index 10679db..0fe97ce 100644
--- a/test/update.test.js
+++ b/test/update.test.js
@@ -215,7 +215,6 @@ test('update non-HA moray and SAPI consecutively', function (t) {
             'Reprovisioning VM',
             'Destroying tmp VM',
             'Updating sapi',
-            'Verifying SAPI full mode',
             'Waiting for sapi',
             'Updated successfully'
         ];
-- 
2.21.0

