commit ec80800d8e2746939f5906bdb4f9fb253cdbb41d (refs/changes/12/3612/2)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-03-15T17:03:02-04:00 (1 year, 7 months ago)
    
    TRITON-240 zoneevent should be called with an identifier string

diff --git a/lib/backends/smartos/index.js b/lib/backends/smartos/index.js
index beae4fc..7df68b7 100644
--- a/lib/backends/smartos/index.js
+++ b/lib/backends/smartos/index.js
@@ -505,7 +505,8 @@ function startZoneeventWatcher(opts, watchers) {
     assert.object(opts, 'opts');
     assert.object(opts.log, 'opts.log');
 
-    watchers.zoneeventHandle = spawn('/usr/vm/sbin/zoneevent', []);
+    watchers.zoneeventHandle = spawn('/usr/vm/sbin/zoneevent',
+        ['-i', 'cn-agent']);
     pid = watchers.zoneeventHandle.pid;
     opts.log.info('zoneevent[' + pid + '] watcher running');
 
