commit 4225be6ffb53b225a09d320d13169f3f78293731 (refs/changes/12/3612/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-03-13T13:23:14-04:00 (1 year, 7 months ago)
    
    TRITON-240 zoneevent should be called with an identifier string

diff --git a/lib/backends/smartos/index.js b/lib/backends/smartos/index.js
index beae4fc..d29c110 100644
--- a/lib/backends/smartos/index.js
+++ b/lib/backends/smartos/index.js
@@ -505,7 +505,8 @@ function startZoneeventWatcher(opts, watchers) {
     assert.object(opts, 'opts');
     assert.object(opts.log, 'opts.log');
 
-    watchers.zoneeventHandle = spawn('/usr/vm/sbin/zoneevent', []);
+    watchers.zoneeventHandle = spawn('/usr/vm/sbin/zoneevent',
+        ['cn-agent']);
     pid = watchers.zoneeventHandle.pid;
     opts.log.info('zoneevent[' + pid + '] watcher running');
 
