From 6da24020a4ab6b69e51ef8419888ad2b68395733 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Fri, 5 Jul 2019 08:47:32 -0600
Subject: [PATCH] MANTA-4390 Unwrap rpc data array in node-boray's
 deleteObjectNoVnode function

---
 lib/objects.js | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/objects.js b/lib/objects.js
index d7372f3..0dcd7c5 100644
--- a/lib/objects.js
+++ b/lib/objects.js
@@ -391,7 +391,12 @@ function deleteObjectNoVnode(rpcctx, owner, bucket_id, name, req_id, callback) {
             return;
         }
 
-        callback(null, data);
+        if (data.length === 1) {
+            callback(null, data[0]);
+        } else {
+            callback(new VError('expected 1 data messages, found %d',
+                data.length));
+        }
     });
 }
 
-- 
2.21.0

