commit df4de77707eaf4aa00661d97789a3d37648fd784 (refs/changes/80/980/1)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-30T13:53:00-05:00 (2 years, 10 months ago)
    
    suggested copyedits

diff --git a/lib/cloudapi2.js b/lib/cloudapi2.js
index 3c38fab..7139f0a 100644
--- a/lib/cloudapi2.js
+++ b/lib/cloudapi2.js
@@ -81,7 +81,7 @@ var OS_PLATFORM = os.platform();
  *      smartdc-auth KeyPair (the preferred method), or with a signer function
  *      (deprecated, retained for compatibility).
  *
- *      Either:
+ *      Either (prefered):
  *      - {String} account (required) The account login name this cloudapi
  *          client will operate upon.
  *      - {Object} principal (required)
@@ -89,14 +89,13 @@ var OS_PLATFORM = os.platform();
  *              authentication.
  *          - {Object} keyPair (required) A smartdc-auth KeyPair object
  *          - {String} user (optional) RBAC sub-user login name
+ *          - {Array of String} roles (optional) RBAC role(s) to take up.
  *
- *      Or:
+ *      Or (backwards compact):
  *      - {String} account (required) The account login name used both for
  *          authentication and as the account being operated upon.
  *      - {Function} sign (required) An http-signature auth signing function.
  *      - {String} user (optional) The RBAC user login name.
- *
- *      May be added to either of the above:
  *      - {Array of String} roles (optional) RBAC role(s) to take up.
  *
  * @throws {TypeError} on bad input.
