From aadb9a9ceac80448093215508d7f2c2fc3952c36 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 28 Oct 2016 15:33:18 -0700
Subject: [PATCH] joyent/node-mname#7 names with trailing '.' encoded
 incorrectly

---
 lib/dns-buffer.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/dns-buffer.js b/lib/dns-buffer.js
index fbbc197..41c0cfd 100644
--- a/lib/dns-buffer.js
+++ b/lib/dns-buffer.js
@@ -143,6 +143,8 @@ DNSBuffer.prototype.writeName = function (name) {
                 rlen = part.length;
                 assert.ok(rlen < 64, 'segment "' + part + '" of name "' +
                     name + '" is too long');
+                if (rlen === 0 && rparts.length === 0)
+                        break;
                 this.writeUInt8(rlen);
                 this.write(new Buffer(part, 'ascii'));
         }
-- 
2.21.0

