commit 70f103e04748dbe10c8d11b8d4f18f16c8a5b7af (refs/changes/41/3341/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-02-08T15:23:26-08:00 (1 year, 8 months ago)
    
    SSO-204 Jenkins sdcsso build failing in pkgin install

diff --git a/targets.json.in b/targets.json.in
index 5366460..9fa8a9e 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -1375,7 +1375,7 @@ cat <<EOF
     "// image_uuid": "triton-origin-multiarch-15.4.1@1.0.1",
     "image_uuid": "04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
     "pkgsrc": [
-      "openssl-1.0.2k",
+      "openssl-1.0.2nnb1",
       "postfix-3.0.2nb2"
     ],
     "repos": [
diff --git a/tools/prep_dataset_in_jpc.sh b/tools/prep_dataset_in_jpc.sh
index 2561723..cfc0612 100755
--- a/tools/prep_dataset_in_jpc.sh
+++ b/tools/prep_dataset_in_jpc.sh
@@ -321,6 +321,10 @@ if [[ -n "${packages}" ]]; then
     ${SSH} "/opt/local/bin/pkgin -y remove libuuid"
   fi
 
+  # This pkgin can inscrutably fail with a pattern like this:
+  #     pkg_install warnings: 0, errors: 1
+  #     pkg_install error log can be found in /var/db/pkgin/pkg_install-err.log
+  # TODO: It would be nice to include the output of that error log.
   ${SSH} "/opt/local/bin/pkgin -y in ${packages}"
 
   echo "Validating pkgsrc installation"
