From 9edc3d6ee9345b640adc0e9e81f4c20c170afa91 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Thu, 31 Aug 2017 08:59:31 -0600
Subject: [PATCH] MANTA-3338 muskie leaked a connection

---
 lib/server.js | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/lib/server.js b/lib/server.js
index e2056e4..c4ea2b8 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -424,7 +424,20 @@ function createServer(options, clearProxy) {
         _audit(req, res, route, err);
     });
 
-    server.on('after', _audit);
+    server.on('after', function (req, res, route, err) {
+        _audit(req, res, route, err);
+
+        if ((req.method === 'PUT' || req.method === 'POST') &&
+            res.statusCode >= 500) {
+            /*
+             * An error occurred on a PUT or POST request, but there may still
+             * be incoming data on the request stream. Call resume() in order to
+             * dump any remaining request data so the stream emits an 'end' and
+             * the socket resources are not leaked.
+             */
+            req.resume();
+        }
+    });
 
     return (server);
 }
-- 
2.21.0

