commit 17a6ee3272776b0eff4f8da5e9e0fa13ff05ef27 (refs/changes/73/2373/5)
Author: Kelly McLaughlin <kelly@kelly-mclaughlin.com>
Date:   2017-08-16T16:59:38-06:00 (2 years, 2 months ago)
    
    MANTA-3338 muskie leaked a connection

diff --git a/lib/server.js b/lib/server.js
index e2056e4..ac003f4 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -122,17 +122,24 @@ function createServer(options, clearProxy) {
         var req = socket.parser && socket.parser.incoming;
         var res = socket._httpMessage;
 
+        var dobj = req ? {
+            method: req.method,
+            url: req.url,
+            headers: req.headers,
+            id: req._id
+        } : {};
+
+        options.dtrace_probes.socket_timeout.fire(function onFire() {
+            return ([dobj]);
+        });
+
+        /*
+         * Destroy the socket in the following case:
+         *   - A request object exists, the request is complete, and a
+         *     _httpMessage object exists for the socket
+         */
         if (req && req.complete && res) {
             l.warn('socket timeout: destroying connection');
-            options.dtrace_probes.socket_timeout.fire(function onFire() {
-                var dobj = req ? {
-                    method: req.method,
-                    url: req.url,
-                    headers: req.headers,
-                    id: req._id
-                } : {};
-                return ([dobj]);
-            });
             socket.destroy();
         }
     });
@@ -424,7 +431,20 @@ function createServer(options, clearProxy) {
         _audit(req, res, route, err);
     });
 
-    server.on('after', _audit);
+    server.on('after', function (req, res, route, err) {
+        _audit(req, res, route, err);
+
+        if ((req.method === 'PUT' || req.method === 'POST') &&
+            res.statusCode >= 500) {
+            /*
+             * An error occurred on a PUT or POST request, but there may still
+             * be incoming data on the request stream. Call resume() in order to
+             * dump any remaining request data so the stream emits an 'end' and
+             * the socket resources are not leaked.
+             */
+            req.resume();
+        }
+    });
 
     return (server);
 }
