commit e2503df6155f226b756986a7ce913255e38be2b0 (refs/changes/62/4962/4)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2018-10-18T11:20:01-04:00 (12 months ago)
    
    OS-7310 vminfod event stream processes events after stop called

diff --git a/src/vm/node_modules/vminfod/client.js b/src/vm/node_modules/vminfod/client.js
index aaac58a7..605cca78 100644
--- a/src/vm/node_modules/vminfod/client.js
+++ b/src/vm/node_modules/vminfod/client.js
@@ -482,6 +482,14 @@ VminfodEventStream.prototype._transform =
     var parsed;
     var vms;
 
+    if (!self.vs_req) {
+        // we are stopped - ignore any incoming data
+        self.vs_logger.trace({line: line},
+            'stream stopped - ignoring line from vminfod');
+        cb();
+        return;
+    }
+
     try {
         ev = JSON.parse(line);
         ev.date = new Date(ev.date);
