commit 8118e45cd65c133140f8add8b40dad62d438060f (refs/changes/73/5373/1)
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-01-16T13:22:15-07:00 (9 months ago)
    
    MANTA-4069 Muppet should generate core files when it crashes

diff --git a/smf/manifests/muppet.xml.in b/smf/manifests/muppet.xml.in
index cbe4d86..15e5402 100644
--- a/smf/manifests/muppet.xml.in
+++ b/smf/manifests/muppet.xml.in
@@ -7,7 +7,7 @@
 -->
 
 <!--
-    Copyright (c) 2017, Joyent, Inc.
+    Copyright (c) 2019, Joyent, Inc.
 -->
 
 <service_bundle type="manifest" name="muppet">
@@ -45,7 +45,7 @@
 
 	<exec_method type="method"
 		     name="start"
-		     exec="node muppet.js -f ./etc/config.json &amp;"
+		     exec="node --abort-on-uncaught-exception muppet.js -f ./etc/config.json &amp;"
 		     timeout_seconds="30">
 	    <method_context working_directory="/opt/smartdc/muppet">
                 <method_credential user="root"
