From d463aec174a6e98982729051e20281113df87d90 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 16 Jan 2018 20:04:33 +0100
Subject: [PATCH] TOOLS-1974 sdc-scripts refactor sdc_common_setup

---
 lib/util.sh | 25 +++++++++++--------------
 1 file changed, 11 insertions(+), 14 deletions(-)

diff --git a/lib/util.sh b/lib/util.sh
index 8e07285..664ce9f 100644
--- a/lib/util.sh
+++ b/lib/util.sh
@@ -694,21 +694,18 @@ function sdc_common_setup
     _sdc_log_rotation_setup
     _sdc_mdata_rbac_setup
 
-    if [[ ! -f /var/svc/setup_complete ]]; then
-        if [[ ${ZONE_ROLE} != "assets" ]]; then
-            if [[ ${ZONE_ROLE} == "sapi" && "${SAPI_PROTO_MODE}" == \
-              "true" ]]; then
-                echo "Skipping config-agent/SAPI instance setup: 'sapi' " \
-                  "zone in proto mode" >&2
-            else
-                setup_config_agent
-                download_metadata
-                write_initial_config
-                registrar_setup
-            fi
-        fi
+    if [[ -f /var/svc/setup_complete ]]; then
+        echo "Skip config-agent/registrar setup: zone already setup" >&2
+    elif [[ ${ZONE_ROLE} == "assets" ]]; then
+        echo "Skip config-agent/registrar setup: assets zone" >&2
+    elif [[ ${ZONE_ROLE} == "sapi" && "${SAPI_PROTO_MODE}" == "true" ]]; then
+        echo "Skip config-agent/registrar setup: sapi zone in proto mode" >&2
     else
-        echo "Already setup, skipping SAPI and registrar initialization." >&2
+        echo "Setup config-agent/registrar"
+        setup_config_agent
+        download_metadata
+        write_initial_config
+        registrar_setup
     fi
 
     _sdc_enable_cron
-- 
2.21.0

