From 96e5dc1e87465119953372332410917cb5d7335c Mon Sep 17 00:00:00 2001
From: Tim Gross <tim@0x74696d.com>
Date: Mon, 3 Apr 2017 13:37:44 -0400
Subject: [PATCH] remove comments from triton env that break eval

---
 lib/do_env.js | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/lib/do_env.js b/lib/do_env.js
index f6ba08e..209f994 100644
--- a/lib/do_env.js
+++ b/lib/do_env.js
@@ -71,7 +71,6 @@ function do_env(subcmd, opts, args, cb) {
     clientTypes.forEach(function (clientType) {
         switch (clientType) {
         case 'triton':
-            p('# triton');
             if (opts.unset) {
                 [
                     'TRITON_PROFILE',
@@ -88,7 +87,6 @@ function do_env(subcmd, opts, args, cb) {
             }
             break;
         case 'docker':
-            p('# docker');
             var setupJson = path.resolve(self.configDir, 'docker',
                 common.profileSlug(profile), 'setup.json');
             if (fs.existsSync(setupJson)) {
@@ -126,7 +124,6 @@ function do_env(subcmd, opts, args, cb) {
             }
             break;
         case 'smartdc':
-            p('# smartdc');
             if (opts.unset) {
                 [
                     'SDC_URL',
-- 
2.21.0

