commit 96e5dc1e87465119953372332410917cb5d7335c (refs/changes/32/1732/1)
Author: Tim Gross <tim@0x74696d.com>
Date:   2017-04-03T13:37:44-04:00 (2 years, 6 months ago)
    
    remove comments from triton env that break eval

diff --git a/lib/do_env.js b/lib/do_env.js
index f6ba08e..209f994 100644
--- a/lib/do_env.js
+++ b/lib/do_env.js
@@ -71,7 +71,6 @@ function do_env(subcmd, opts, args, cb) {
     clientTypes.forEach(function (clientType) {
         switch (clientType) {
         case 'triton':
-            p('# triton');
             if (opts.unset) {
                 [
                     'TRITON_PROFILE',
@@ -88,7 +87,6 @@ function do_env(subcmd, opts, args, cb) {
             }
             break;
         case 'docker':
-            p('# docker');
             var setupJson = path.resolve(self.configDir, 'docker',
                 common.profileSlug(profile), 'setup.json');
             if (fs.existsSync(setupJson)) {
@@ -126,7 +124,6 @@ function do_env(subcmd, opts, args, cb) {
             }
             break;
         case 'smartdc':
-            p('# smartdc');
             if (opts.unset) {
                 [
                     'SDC_URL',
