commit 938d71196f0dae20fe5cca09cfd4e81dda460fbe
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-15T15:07:45-04:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index 9254389..c2d54ab 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -3,7 +3,7 @@
 	url = https://github.com/davepacheco/javascriptlint.git
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/restdown"]
 	path = deps/restdown
 	url = https://github.com/trentm/restdown.git
diff --git a/deps/jsstyle b/deps/jsstyle
index d75b7ca..52dc973 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit d75b7ca8308be17c80e2b120f2a01d4a0c20d8a8
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
