commit 8717914a5f429f0774608a755e1c4d576de1bc53 (refs/changes/19/2419/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-08-14T17:59:05+00:00 (2 years, 2 months ago)
    
    MANATEE-351 Update to pg_repack v1.4.1

diff --git a/deps/pg_repack b/deps/pg_repack
index cac8179..20f5bf4 160000
--- a/deps/pg_repack
+++ b/deps/pg_repack
@@ -1 +1 @@
-Subproject commit cac817929901796e1ed18f9592694a355464d9b3
+Subproject commit 20f5bf4878317143963f6cbeadbe520b4ef4b195
