commit f3de42b8d72fbb560f04df738b5cc25e50674c7b (refs/changes/04/5004/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-10-26T14:54:44-07:00 (12 months ago)
    
    TRITON-911 support role-operators group in cloudapi

diff --git a/lib/index.js b/lib/index.js
index 6aa585b..4d5e1c6 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -70,6 +70,7 @@ var GROUPS = 'ou=groups, ' + SUFFIX;
 var GROUP_FMT = 'cn=%s, ' + GROUPS;
 var ADMIN_GROUP = sprintf(GROUP_FMT, 'operators');
 var READERS_GROUP = sprintf(GROUP_FMT, 'readers');
+var ROLE_OPER_GROUP = sprintf(GROUP_FMT, 'role-operators');
 
 var USERS = 'ou=users, ' + SUFFIX;
 var USER_FMT = 'uuid=%s, ' + USERS;
@@ -788,6 +789,7 @@ UFDS.prototype._getUser = function _getUser(filter, msg, account, cb, noCache) {
                     // matching groups.
                     var lookup = [
                         ADMIN_GROUP,
+                        ROLE_OPER_GROUP,
                         READERS_GROUP
                     ];
                     var compareDN = result.dn.toString();
diff --git a/package.json b/package.json
index 90bca89..a431941 100644
--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "name": "ufds",
   "author": "Joyent (joyent.com)",
   "description": "SmartDataCenter UFDS Client API",
-  "version": "1.4.0",
+  "version": "1.5.0",
   "homepage": "https://github.com/joyent/sdc",
   "repository": {
     "type": "git",
