From 8c472944183a459ae1a250923775730666b9a18f Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 26 Sep 2017 16:04:15 -0700
Subject: [PATCH] CNS-210 AppFSM returns to cfRunning state, double-pipes
 streams

---
 updater.js | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/updater.js b/updater.js
index 34e6d3e..279db67 100644
--- a/updater.js
+++ b/updater.js
@@ -140,6 +140,8 @@ AppFSM.prototype.state_initial = function (S) {
 AppFSM.prototype.state_cfFirstPoll = function (S) {
 	us.openSerial(false);
 	ps.start();
+	cfl.unpipe();
+	cff.unpipe();
 	S.on(ps, 'pollFinish', function () {
 		log.info('Poll done, committing...');
 		us.closeSerial();
@@ -184,6 +186,8 @@ AppFSM.prototype.state_fallbackFirstPoll = function (S) {
 AppFSM.prototype.state_fallback = function (S) {
 	rs.setReapTime(FALLBACK_REAP_TIME);
 	rs.start();
+	cfl.unpipe();
+	cff.unpipe();
 	S.on(pollTimeEmitter, 'timeout', function () {
 		ps.start();
 	});
-- 
2.21.0

