commit 8c472944183a459ae1a250923775730666b9a18f (refs/changes/58/2658/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-09-26T16:04:15-07:00 (2 years ago)
    
    CNS-210 AppFSM returns to cfRunning state, double-pipes streams

diff --git a/updater.js b/updater.js
index 34e6d3e..279db67 100644
--- a/updater.js
+++ b/updater.js
@@ -140,6 +140,8 @@ AppFSM.prototype.state_initial = function (S) {
 AppFSM.prototype.state_cfFirstPoll = function (S) {
 	us.openSerial(false);
 	ps.start();
+	cfl.unpipe();
+	cff.unpipe();
 	S.on(ps, 'pollFinish', function () {
 		log.info('Poll done, committing...');
 		us.closeSerial();
@@ -184,6 +186,8 @@ AppFSM.prototype.state_fallbackFirstPoll = function (S) {
 AppFSM.prototype.state_fallback = function (S) {
 	rs.setReapTime(FALLBACK_REAP_TIME);
 	rs.start();
+	cfl.unpipe();
+	cff.unpipe();
 	S.on(pollTimeEmitter, 'timeout', function () {
 		ps.start();
 	});
