commit b56784a0ddd5e99f8337e90b2e2a41f9952107c8 (refs/changes/53/1853/2)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2017-04-28T22:03:34+00:00 (2 years, 5 months ago)
    
    MANTA-2904 502 responses should use JSON, HTTP 1.1

diff --git a/etc/haproxy.cfg.in b/etc/haproxy.cfg.in
index 0dd5a07..e1a4212 100644
--- a/etc/haproxy.cfg.in
+++ b/etc/haproxy.cfg.in
@@ -26,6 +26,7 @@ defaults
         errorfile 400 /opt/smartdc/muppet/etc/400.http
         errorfile 408 /opt/smartdc/muppet/etc/408.http
         errorfile 500 /opt/smartdc/muppet/etc/503.http
+        errorfile 502 /opt/smartdc/muppet/etc/503.http
         errorfile 503 /opt/smartdc/muppet/etc/503.http
         errorfile 504 /opt/smartdc/muppet/etc/503.http
 
