From 8fdc4259ac92b8c40ee8d9556ee122fd2d5b252b Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 31 May 2018 20:48:23 +0200
Subject: [PATCH] TOOLS-1926 sdcadm health is stuck and won't exit

---
 lib/sdcadm.js | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/lib/sdcadm.js b/lib/sdcadm.js
index 27f3160..9dabd6d 100644
--- a/lib/sdcadm.js
+++ b/lib/sdcadm.js
@@ -401,6 +401,19 @@ SdcAdm.prototype.ensureSdcApp = function (arg, cb) {
                 appErr = new VError(appErr,
                     'Binder service seems to be down. ' +
                     'Please review it before proceeding');
+                // We need to stop the DNS Resolver we've attempted to create
+                // with Binder for DNS resolution, since binder is down. This
+                // resolver is created by Cueball's `resolverForIpOrDomain`
+                // and stored into CueBallDNSResolver object as a class level
+                // property. If we don't stop this resolver, sdcadm process
+                // will keep hanging and the resolver itself attempting to
+                // connect to binder, given we're using "softexit" to finish
+                // the sdcadm process:
+                var CueBallDNSResolver = require('cueball').DNSResolver;
+                var binderResolver = CueBallDNSResolver.bootstrapResolvers[
+                    format('binder.%s.%s', self.config.datacenter_name,
+                        self.config.dns_domain)];
+                binderResolver.stop();
             } else if (VError.hasCauseWithName(appErr, 'PoolFailedError')) {
                 appErr = new VError(appErr,
                     'SAPI service seems to be down. ' +
-- 
2.21.0

