From 445dd9f369fb6dcb0ef481899ff733a600befb2a Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Fri, 1 Feb 2019 11:54:01 +0000
Subject: [PATCH] TRITON-1177 hermes build looking for node in all the wrong
 places

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7e04473..f4e5f5e 100644
--- a/Makefile
+++ b/Makefile
@@ -167,7 +167,7 @@ $(DESTDIR)$(PREFIX)/sapi_manifests/%: $(PWD)/sapi_manifests/%
 
 0-npm-stamp: $(NODE_EXEC) package.json
 	rm -rf $(PWD)/node_modules
-	PATH=$(PWD)/node/bin:$(PATH) NPM_CONFIG_CACHE= $(NPM_EXEC) install
+	NPM_CONFIG_CACHE= $(NODE_EXEC) $(NPM_EXEC) install
 	touch $@
 
 $(DESTDIR)$(PREFIX)/node_modules: 0-npm-stamp
-- 
2.21.0

