From 3b7a595a5dc7ec160c200ae5ca904026cd0b83bf Mon Sep 17 00:00:00 2001
From: Brittany Wald <brittany.wald@joyent.com>
Date: Wed, 13 Jun 2018 23:42:57 +0000
Subject: [PATCH] TRITON-373 disable directory listing on assets zone nginx

---
 boot/configure.sh | 1 -
 1 file changed, 1 deletion(-)

diff --git a/boot/configure.sh b/boot/configure.sh
index 12d3327..a2445a3 100755
--- a/boot/configure.sh
+++ b/boot/configure.sh
@@ -38,7 +38,6 @@ http {
         location / {
             root   /assets;
             index  index.html index.htm;
-            autoindex on;
         }
 
         error_page   500 502 503 504  /50x.html;
-- 
2.21.0

