commit 3b7a595a5dc7ec160c200ae5ca904026cd0b83bf (refs/changes/97/4397/1)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2018-06-13T23:42:57+00:00 (1 year, 4 months ago)
    
    TRITON-373 disable directory listing on assets zone nginx

diff --git a/boot/configure.sh b/boot/configure.sh
index 12d3327..a2445a3 100755
--- a/boot/configure.sh
+++ b/boot/configure.sh
@@ -38,7 +38,6 @@ http {
         location / {
             root   /assets;
             index  index.html index.htm;
-            autoindex on;
         }
 
         error_page   500 502 503 504  /50x.html;
