From ab9ce2c736141af9c06a51869c731b21d9e9970b Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Thu, 19 Jul 2018 13:11:13 +0000
Subject: [PATCH] TRITON-613 firewaller-agent picks up wrong node during make

---
 tools/mk/Makefile.targ | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/mk/Makefile.targ b/tools/mk/Makefile.targ
index cc5ae95..dac448a 100644
--- a/tools/mk/Makefile.targ
+++ b/tools/mk/Makefile.targ
@@ -101,7 +101,7 @@
 # Defaults for the various tools we use.
 #
 BASH		?= bash
-BASHSTYLE	?= tools/bashstyle
+BASHSTYLE   ?= $(NODE) tools/bashstyle
 CP		?= cp
 CSCOPE		?= cscope
 CSCOPE_DIRS	?= .
-- 
2.21.0

