commit 00ad182d745dc4bd2b46d0dd36f576c06d13debc
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-15T15:07:34-04:00 (7 months ago)
    
    TOOLS-2101 jsstyle submodules too old for old perls

diff --git a/.gitmodules b/.gitmodules
index fec05d10..556c9237 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,6 +1,6 @@
 [submodule "deps/jsstyle"]
   path = deps/jsstyle
-  url = https://github.com/davepacheco/jsstyle.git
+	url = https://github.com/joyent/jsstyle.git
 [submodule "deps/restdown"]
   path = deps/restdown
   url = https://github.com/trentm/restdown.git
diff --git a/deps/jsstyle b/deps/jsstyle
index 07d4f682..52dc973c 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit 07d4f68251063be6496a42dd00a7f5bacd65c5e4
+Subproject commit 52dc973cf64da11834eca7cf46ebce8518e3ee88
