commit 7d1352b1aafbd6c04fe06acdc9441623c652a2e0 (refs/changes/31/5031/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-11-06T00:10:32+00:00 (11 months ago)
    
    OS-7336 lx_acl_getxattr leaks the ACL

diff --git a/usr/src/uts/common/brand/lx/os/lx_acl.c b/usr/src/uts/common/brand/lx/os/lx_acl.c
index 594671f401..482680cf06 100644
--- a/usr/src/uts/common/brand/lx/os/lx_acl.c
+++ b/usr/src/uts/common/brand/lx/os/lx_acl.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright (c) 2017 Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -192,6 +192,16 @@ lx_acl_getxattr(vnode_t *vp, enum lx_acl_type atype, void *data, size_t slen,
 		return (err);
 	}
 
+	if (vsattr.vsa_aclentp != NULL) {
+		kmem_free(vsattr.vsa_aclentp,
+		    vsattr.vsa_aclcnt * sizeof (aclent_t));
+	}
+
+	if (vsattr.vsa_dfaclentp != NULL) {
+		kmem_free(vsattr.vsa_dfaclentp,
+		    vsattr.vsa_dfaclcnt * sizeof (aclent_t));
+	}
+
 	return (ENODATA);
 }
 
