commit 4a9d6952000a864dd8e74700cdfa5abb6657b23f (refs/changes/93/2493/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2017-09-05T16:13:03+00:00 (2 years, 1 month ago)
    
    OS-6320 systemd aborts itself on suse image

diff --git a/usr/src/uts/common/brand/lx/io/lx_netlink.c b/usr/src/uts/common/brand/lx/io/lx_netlink.c
index d6c6c41f23..f5c29a123b 100644
--- a/usr/src/uts/common/brand/lx/io/lx_netlink.c
+++ b/usr/src/uts/common/brand/lx/io/lx_netlink.c
@@ -10,7 +10,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 /*
@@ -292,6 +292,7 @@
  */
 #define	SOL_LX_NETLINK	270
 
+/* See Linux include/uapi/linux/netlink.h */
 #define	LX_NETLINK_SO_ADD_MEMBERSHIP	1
 #define	LX_NETLINK_SO_DROP_MEMBERSHIP	2
 #define	LX_NETLINK_SO_PKTINFO		3
@@ -299,6 +300,9 @@
 #define	LX_NETLINK_SO_NO_ENOBUFS	5
 #define	LX_NETLINK_SO_RX_RING		6
 #define	LX_NETLINK_SO_TX_RING		7
+#define	LX_NETLINK_SO_LISTEN_ALL_NSID	8
+#define	LX_NETLINK_SO_LIST_MEMBERSHIPS	9
+#define	LX_NETLINK_SO_CAP_ACK 		10
 
 /* Internal socket flags */
 #define	LXNLF_RECVUCRED			0x1
@@ -509,6 +513,26 @@ lx_netlink_setsockopt(sock_lower_handle_t handle, int level,
 	}
 }
 
+static int
+lx_netlink_getsockopt(sock_lower_handle_t handle, int level,
+    int option_name, void *optval, socklen_t *optlen, cred_t *cr)
+{
+	t_uscalar_t size = 0;
+
+	if (level != SOL_LX_NETLINK) {
+		return (EOPNOTSUPP);
+	}
+
+	switch (option_name) {
+	case LX_NETLINK_SO_LIST_MEMBERSHIPS:
+		/* Report that we have 0 members to allow systemd to proceed. */
+		bcopy(&size, optlen, sizeof (size));
+		return (0);
+	default:
+		return (EINVAL);
+	}
+}
+
 /*ARGSUSED*/
 static int
 lx_netlink_bind(sock_lower_handle_t handle, struct sockaddr *name,
@@ -1609,7 +1633,7 @@ static sock_downcalls_t sock_lx_netlink_downcalls = {
 	sock_connect_notsupp,		/* sd_connect */
 	sock_getpeername_notsupp,	/* sd_getpeername */
 	lx_netlink_getsockname,		/* sd_getsockname */
-	sock_getsockopt_notsupp,	/* sd_getsockopt */
+	lx_netlink_getsockopt,		/* sd_getsockopt */
 	lx_netlink_setsockopt,		/* sd_setsockopt */
 	lx_netlink_send,		/* sd_send */
 	NULL,				/* sd_send_uio */
