commit 4cfe4266f7348a9ba199bf15f685cce3807db998 (refs/changes/86/5586/1)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2019-02-14T11:57:10-05:00 (8 months ago)
    
    OS-7581 metadata daemon not detecting VM status changes

diff --git a/src/vm/lib/metadata/agent.js b/src/vm/lib/metadata/agent.js
index 53ef0bee..ea5a1a5c 100644
--- a/src/vm/lib/metadata/agent.js
+++ b/src/vm/lib/metadata/agent.js
@@ -20,7 +20,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  *
  *
  * # OVERVIEW
@@ -496,7 +496,9 @@ MetadataAgent.prototype.start = function start() {
         }
         // Only 1 state change event should be seen per vminfod event
         assert.equal(state.length, 1, 'multiple "state" changes seen');
-        state = state[0];
+        assert.object(state[0], 'state[0]');
+        assert.string(state[0].newValue, 'state[0].newValue');
+        state = state[0].newValue;
 
         // If a KVM zone stops while we're trying to reconnect to its metadata
         // socket, stop trying to reconnect.
