From cf5d99eeefb53fef4a9a02ad078c6fda8b81e724 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Thu, 26 Oct 2017 21:00:23 +0000
Subject: [PATCH] MANTA-2412 Allow addition of content-disposition header to
 manta objects

---
 lib/moray.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/moray.js b/lib/moray.js
index ca89840..1097e66 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -337,6 +337,8 @@ function createMetadata(options) {
 
     switch (options.type) {
     case 'object':
+        assert.optionalString(options.contentDisposition,
+                'options.contentDisposition');
         assert.number(options.contentLength, 'options.contentLength');
         assert.string(options.contentMD5, 'options.contentMD5');
         assert.string(options.contentType, 'options.contentType');
@@ -350,6 +352,7 @@ function createMetadata(options) {
             });
         }
 
+        md.contentDisposition = options.contentDisposition;
         md.contentLength = options.contentLength;
         md.contentMD5 = options.contentMD5;
         md.contentType = options.contentType;
-- 
2.21.0

