commit e3c83fc7c1addb6f2779728855007c7a6453f0cb (refs/changes/08/4408/2)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2018-06-22T17:28:01+00:00 (1 year, 4 months ago)
    
    OS-7045 Add Jirisan topo aliases

diff --git a/manifest b/manifest
index fbc076fb9a..508b2dec79 100644
--- a/manifest
+++ b/manifest
@@ -10450,9 +10450,11 @@ f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Compute-Platform-3301-hc-topology.x
 s usr/platform/i86pc/lib/fm/topo/maps/Joyent-Compute-Platform-3302-hc-topology.xml=./Joyent-Compute-Platform-3301-hc-topology.xml
 f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Compute-Platform-330x-chassis-hc-topology.xml 0444 root sys
 f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Compute-Platform-330x-fan-hc-topology.xml 0444 root sys
+s usr/platform/i86pc/lib/fm/topo/maps/Joyent-M12G5-hc-topology.xml=./SSG-2029P-ACR24L-hc-topology.xml
 f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Storage-Platform-7001-hc-topology.xml 0444 root sys
 f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Storage-Platform-7001-chassis-hc-topology.xml 0444 root sys
 f usr/platform/i86pc/lib/fm/topo/maps/Joyent-Storage-Platform-7001-slot-hc-topology.xml 0444 root sys
+s usr/platform/i86pc/lib/fm/topo/maps/Joyent-S10G5-hc-topology.xml=./SSG-6049P-E1CR36L-hc-topology.xml
 f usr/platform/i86pc/lib/fm/topo/maps/SSG-6049P-E1CR36L-chassis-hc-topology.xml 0444 root sys
 f usr/platform/i86pc/lib/fm/topo/maps/SSG-6049P-E1CR36L-hc-topology.xml 0444 root sys
 f usr/platform/i86pc/lib/fm/topo/maps/SSG-6049P-E1CR36L-slot-hc-topology.xml 0444 root sys
diff --git a/usr/src/lib/fm/topo/maps/SMCI,SSG-2029P-ACR24L/Makefile b/usr/src/lib/fm/topo/maps/SMCI,SSG-2029P-ACR24L/Makefile
index f158cf2fe4..0cd1c11a20 100644
--- a/usr/src/lib/fm/topo/maps/SMCI,SSG-2029P-ACR24L/Makefile
+++ b/usr/src/lib/fm/topo/maps/SMCI,SSG-2029P-ACR24L/Makefile
@@ -20,6 +20,10 @@
 ARCH = i86pc
 CLASS = arch
 DTDFILE = topology.dtd.1
+JOYENT_PLATFORMS = \
+	Joyent-M12G5-hc-topology.xml
+
+ROOTJOYENTMAPS = $(JOYENT_PLATFORMS:%=$(arch_ROOTTOPOROOT)/%)
 
 TOPOFILE = \
 	SSG-2029P-ACR24L-hc-topology.xml \
@@ -34,3 +38,14 @@ PLATFORM = SSG-2029P-ACR24L
 CLOBBERFILES +=	SSG-2029P-ACR24L-disk-hc-topology.xml
 
 include ../Makefile.map
+
+install: $(ROOTJOYENTMAPS)
+
+#
+# Note, the Joyent M12G5 is based on the SuperMicro SSG-2029P-ACR24L.
+# Because of that, the topo map used here will work for all such
+# systems.
+#
+$(ROOTJOYENTMAPS): SSG-2029P-ACR24L-hc-topology.xml
+	$(RM) $@
+	$(SYMLINK) ./$? $@
diff --git a/usr/src/lib/fm/topo/maps/SMCI,SSG-6049P-E1CR36L/Makefile b/usr/src/lib/fm/topo/maps/SMCI,SSG-6049P-E1CR36L/Makefile
index 8b701a61f4..eeaf0c9045 100644
--- a/usr/src/lib/fm/topo/maps/SMCI,SSG-6049P-E1CR36L/Makefile
+++ b/usr/src/lib/fm/topo/maps/SMCI,SSG-6049P-E1CR36L/Makefile
@@ -20,6 +20,10 @@
 ARCH = i86pc
 CLASS = arch
 DTDFILE = topology.dtd.1
+JOYENT_PLATFORMS = \
+	Joyent-S10G5-hc-topology.xml
+
+ROOTJOYENTMAPS = $(JOYENT_PLATFORMS:%=$(arch_ROOTTOPOROOT)/%)
 
 TOPOFILE = \
 	SSG-6049P-E1CR36L-hc-topology.xml \
@@ -31,3 +35,13 @@ SRCDIR = ../SMCI,SSG-6049P-E1CR36L
 PLATFORM = SSG-6049P-E1CR36L
 
 include ../Makefile.map
+
+install: $(ROOTJOYENTMAPS)
+#
+# Note, the Joyent S10G5 is based on the SuperMicro SSG-6049P-E1CR36L.
+# Because of that, the topo map used here will work for all such
+# systems.
+#
+$(ROOTJOYENTMAPS): SSG-6049P-E1CR36L-hc-topology.xml
+	$(RM) $@
+	$(SYMLINK) ./$? $@
