commit 3a6da3ba089d7857db4a25b0b3f7f0a7877e6fa7 (refs/changes/72/5572/3)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2019-02-13T22:44:51+00:00 (8 months ago)
    
    MANTA-2948 mako's tombstone grace period should be configurable

diff --git a/bin/mako_gc.sh b/bin/mako_gc.sh
index 3e4dfa6..f8a9cc1 100755
--- a/bin/mako_gc.sh
+++ b/bin/mako_gc.sh
@@ -46,7 +46,8 @@ PID_FILE=/tmp/mako_gc.pid
 TOMB_DATE=$(date "+%Y-%m-%d")
 TOMB_ROOT=/manta/tombstone
 TOMB_DIR=$TOMB_ROOT/$TOMB_DATE
-TOMB_DIRS_TO_KEEP=21
+TOMB_DIRS_TO_KEEP=$(json -f /opt/smartdc/mako/etc/gc_config.json \
+    storage_max_tombstone_directories)
 
 
 # Mutables
diff --git a/sapi_manifests/gc/template b/sapi_manifests/gc/template
index c7cea44..9aefeec 100644
--- a/sapi_manifests/gc/template
+++ b/sapi_manifests/gc/template
@@ -1,4 +1,5 @@
 {
 	"manta_url": "{{MANTA_URL}}",
-	"manta_storage_id": "{{MANTA_STORAGE_ID}}"
+	"manta_storage_id": "{{MANTA_STORAGE_ID}}",
+	"storage_max_tombstone_directories": {{#STORAGE_MAX_TOMBSTONE_DIRECTORIES}}{{STORAGE_MAX_TOMBSTONE_DIRECTORIES}}{{/STORAGE_MAX_TOMBSTONE_DIRECTORIES}}{{^STORAGE_MAX_TOMBSTONE_DIRECTORIES}}21{{/STORAGE_MAX_TOMBSTONE_DIRECTORIES}}
 }
