commit 78707b18782b021a39e1cc231f59780e7065986c (refs/changes/38/38/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-07-06T14:57:37-07:00 (3 years, 3 months ago)
    
    fix triton log uploading to manta

diff --git a/tools/standalone/tritonlogupload.sh b/tools/standalone/tritonlogupload.sh
index e54fbe6..0ef0b8e 100755
--- a/tools/standalone/tritonlogupload.sh
+++ b/tools/standalone/tritonlogupload.sh
@@ -86,7 +86,7 @@ function upload_files() {
         echo "no files to upload"
         return
     fi
-        
+
     nfiles=$(echo "$files" | wc -l)
     echo "upload $nfiles file(s) from $SRCDIR to $dstMdir"
     for f in $files
@@ -101,10 +101,10 @@ function upload_files() {
         hourdir=$($DATE -d \@$(( $($DATE -d $isotime "+%s") - 3600 )) "+%Y/%m/%d/%H")
         targ="$dstMdir/$logname/$hourdir/$nodename.log"
 
-        upload_file $f $targ $dryrun
-        echo "rm $f"
+        upload_file $SRCDIR/$f $targ $dryrun
+        echo "rm $SRCDIR/$f"
         if [[ $dryrun == "no" ]]; then
-            rm $f
+            rm $SRCDIR/$f
         fi
     done
 }
