From 972b68069c03cae7f92e2637ff205b8e1dc1f9dd Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 9 Jul 2019 15:09:53 -0400
Subject: [PATCH] TRITON-1804 booter's node_module vendoring breaks if "build"
 directory does not exist

---
 Makefile | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index efdf2a2..db62eaf 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ TOP := $(shell pwd)
 #
 TAPE := ./node_modules/.bin/tape
 ISTANBUL := ./node_modules/.bin/istanbul
-PACK := ./build/pack-0.0.1.tgz
+PACK := ./$(BUILD)/pack-0.0.1.tgz
 
 #
 # Files
@@ -95,7 +95,8 @@ coverage: | $(ISTANBUL) $(TAPE) node_modules
 
 $(PACK):
 	$(NPM) pack file:$(TOP)/src/node-pack
-	mv pack-0.0.1.tgz build/
+	mkdir -p $(TOP)/$(BUILD)
+	mv pack-0.0.1.tgz $(TOP)/$(BUILD)/
 
 #
 # Packaging targets
-- 
2.21.0

