commit ea6f02e91034e3634fad39398c93a9507bc57d20
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2019-06-12T18:03:11+00:00 (4 months ago)
    
    OS-7835 bhyve modules should have mapfiles

diff --git a/usr/src/uts/i86pc/io/viona/viona.mapfile b/usr/src/uts/i86pc/io/viona/viona.mapfile
new file mode 100644
index 0000000000..cece86348c
--- /dev/null
+++ b/usr/src/uts/i86pc/io/viona/viona.mapfile
@@ -0,0 +1,41 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+#
+# MAPFILE HEADER START
+#
+# WARNING:  STOP NOW.  DO NOT MODIFY THIS FILE.
+# Object versioning must comply with the rules detailed in
+#
+#	usr/src/lib/README.mapfiles
+#
+# You should not be making modifications here until you've read the most current
+# copy of that file. If you need help, contact a gatekeeper for guidance.
+#
+# MAPFILE HEADER END
+#
+
+$mapfile_version 2
+
+SYMBOL_VERSION ILLUMOSprivate {
+    global:
+	# DDI Interfaces
+	_fini;
+	_init;
+	_info;
+
+    local:
+	*;
+};
diff --git a/usr/src/uts/i86pc/io/vmm/io/ppt.c b/usr/src/uts/i86pc/io/vmm/io/ppt.c
index 0c5815278e..64efa954fa 100644
--- a/usr/src/uts/i86pc/io/vmm/io/ppt.c
+++ b/usr/src/uts/i86pc/io/vmm/io/ppt.c
@@ -29,7 +29,7 @@
  */
 
 /*
- * Copyright 2018 Joyent, Inc
+ * Copyright 2019 Joyent, Inc
  */
 
 #include <sys/cdefs.h>
@@ -623,7 +623,7 @@ static struct dev_ops ppt_ops = {
 
 static struct modldrv modldrv = {
 	&mod_driverops,
-	"ppt",
+	"bhyve pci pass-thru",
 	&ppt_ops
 };
 
diff --git a/usr/src/uts/i86pc/io/vmm/io/ppt.mapfile b/usr/src/uts/i86pc/io/vmm/io/ppt.mapfile
new file mode 100644
index 0000000000..aac896e89e
--- /dev/null
+++ b/usr/src/uts/i86pc/io/vmm/io/ppt.mapfile
@@ -0,0 +1,52 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+#
+# MAPFILE HEADER START
+#
+# WARNING:  STOP NOW.  DO NOT MODIFY THIS FILE.
+# Object versioning must comply with the rules detailed in
+#
+#	usr/src/lib/README.mapfiles
+#
+# You should not be making modifications here until you've read the most current
+# copy of that file. If you need help, contact a gatekeeper for guidance.
+#
+# MAPFILE HEADER END
+#
+
+$mapfile_version 2
+
+SYMBOL_VERSION ILLUMOSprivate {
+    global:
+	# DDI Interfaces
+	_fini;
+	_init;
+	_info;
+
+	# PCI pass-thru API for bhyve
+	ppt_assigned_devices;
+	ppt_is_mmio;
+	ppt_assign_device;
+	ppt_unassign_device;
+	ppt_unassign_all;
+	ppt_map_mmio;
+	ppt_setup_msi;
+	ppt_setup_msix;
+	ppt_get_limits;
+
+    local:
+	*;
+};
diff --git a/usr/src/uts/i86pc/io/vmm/vmm.mapfile b/usr/src/uts/i86pc/io/vmm/vmm.mapfile
new file mode 100644
index 0000000000..2059dfcc97
--- /dev/null
+++ b/usr/src/uts/i86pc/io/vmm/vmm.mapfile
@@ -0,0 +1,59 @@
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+#
+# MAPFILE HEADER START
+#
+# WARNING:  STOP NOW.  DO NOT MODIFY THIS FILE.
+# Object versioning must comply with the rules detailed in
+#
+#	usr/src/lib/README.mapfiles
+#
+# You should not be making modifications here until you've read the most current
+# copy of that file. If you need help, contact a gatekeeper for guidance.
+#
+# MAPFILE HEADER END
+#
+
+$mapfile_version 2
+
+SYMBOL_VERSION ILLUMOSprivate {
+    global:
+	# DDI Interfaces
+	_fini;
+	_init;
+	_info;
+
+	# bhyve driver API
+	vmm_drv_hold;
+	vmm_drv_rele;
+	vmm_drv_expired;
+	vmm_drv_gpa2kva;
+	vmm_drv_ioport_hook;
+	vmm_drv_ioport_unhook;
+	vmm_drv_msi;
+
+	# IOMMU API for PCI pass-thru
+	iommu_add_device;
+	iommu_host_domain;
+	iommu_remove_device;
+	lapic_intr_msi;
+	vm_iommu_domain;
+	vm_map_mmio;
+	vm_unmap_mmio;
+
+    local:
+	*;
+};
diff --git a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
index 7d1d0c0bfb..af775d4d7b 100644
--- a/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
+++ b/usr/src/uts/i86pc/io/vmm/vmm_sol_dev.c
@@ -1994,7 +1994,7 @@ static struct dev_ops vmm_ops = {
 
 static struct modldrv modldrv = {
 	&mod_driverops,
-	"vmm",
+	"bhyve vmm",
 	&vmm_ops
 };
 
diff --git a/usr/src/uts/i86pc/ppt/Makefile b/usr/src/uts/i86pc/ppt/Makefile
index 60add02d5a..756561d95c 100644
--- a/usr/src/uts/i86pc/ppt/Makefile
+++ b/usr/src/uts/i86pc/ppt/Makefile
@@ -33,6 +33,9 @@ CONF_SRCDIR	= $(UTSBASE)/i86pc/io/vmm/io
 #
 include $(UTSBASE)/i86pc/Makefile.i86pc
 
+# Override mapfile defined in above include
+MAPFILE		= $(UTSBASE)/i86pc/io/vmm/io/ppt.mapfile
+
 #
 #	Define targets
 #
@@ -51,6 +54,7 @@ INC_PATH	+= -I$(UTSBASE)/i86pc/io/vmm -I$(UTSBASE)/i86pc/io/vmm/io
 AS_INC_PATH	+= -I$(UTSBASE)/i86pc/io/vmm -I$(OBJS_DIR)
 
 LDFLAGS         += -dy -N drv/vmm -N misc/pcie
+LDFLAGS         += -M $(MAPFILE)
 
 $(OBJS_DIR)/ppt.o := CERRWARN	+= -_gcc=-Wno-unused-variable
 
diff --git a/usr/src/uts/i86pc/viona/Makefile b/usr/src/uts/i86pc/viona/Makefile
index 7053d62967..453f37911d 100644
--- a/usr/src/uts/i86pc/viona/Makefile
+++ b/usr/src/uts/i86pc/viona/Makefile
@@ -33,6 +33,9 @@ CONF_SRCDIR	= $(UTSBASE)/i86pc/io/viona
 #
 include $(UTSBASE)/i86pc/Makefile.i86pc
 
+# Override mapfile defined in above include
+MAPFILE		= $(UTSBASE)/i86pc/io/viona/viona.mapfile
+
 #
 # Define targets
 #
@@ -58,6 +61,7 @@ DEF_BUILDS	= $(DEF_BUILDSONLY64)
 CFLAGS		+= $(CCVERBOSE)
 LDFLAGS		+= -dy -Ndrv/dld -Nmisc/mac -Nmisc/dls -Ndrv/vmm -Nmisc/neti
 LDFLAGS		+= -Nmisc/hook
+LDFLAGS		+= -M $(MAPFILE)
 
 #
 #	Default build targets.
diff --git a/usr/src/uts/i86pc/vmm/Makefile b/usr/src/uts/i86pc/vmm/Makefile
index 72090ac159..181e38ae50 100644
--- a/usr/src/uts/i86pc/vmm/Makefile
+++ b/usr/src/uts/i86pc/vmm/Makefile
@@ -33,6 +33,9 @@ CONF_SRCDIR	= $(UTSBASE)/i86pc/io/vmm
 #
 include $(UTSBASE)/i86pc/Makefile.i86pc
 
+# Override mapfile defined in above include
+MAPFILE		= $(UTSBASE)/i86pc/io/vmm/vmm.mapfile
+
 #
 #	Define targets
 #
@@ -109,6 +112,7 @@ $(OBJS_DIR)/vmx.o := CERRWARN += -_gcc=-Wno-unused-variable
 $(OBJS_DIR)/iommu.o := CERRWARN += -_gcc=-Wno-unused-variable
 
 LDFLAGS         += -dy -N misc/acpica -N misc/pcie -N fs/dev
+LDFLAGS         += -M $(MAPFILE)
 
 OFFSETS_VMX	= $(CONF_SRCDIR)/intel/offsets.in
 OFFSETS_SVM	= $(CONF_SRCDIR)/amd/offsets.in
