commit b70846fd46c8b919e33cd00ea30949e2194c42aa
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-04-08T12:00:16-07:00 (6 months ago)
    
    TRITON-1233 add sdc-migrate operator tooling to headnode

diff --git a/Makefile b/Makefile
index 2bd46912..04df789c 100644
--- a/Makefile
+++ b/Makefile
@@ -62,6 +62,7 @@ SDC_ZONE_COMMANDS = \
 	sdc-imgadm \
 	sdc-imgapi \
 	sdc-mahi \
+	sdc-migrate \
 	sdc-napi \
 	sdc-oneachnode \
 	sdc-papi \
diff --git a/package.json b/package.json
index 61e29948..099667e8 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "usb-headnode",
   "description": "Build SDC headnode images",
-  "version": "3.0.1",
+  "version": "3.0.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
