commit ce6dc5926836fe88e7935f23f3bda0803b61f8be (refs/changes/34/2134/2)
Author: Richard Bradley <richard.bradley@joyent.com>
Date:   2017-07-11T08:53:08+00:00 (2 years, 3 months ago)
    
    SAPI-290 getApplicationObjects returns undefined instance list for known services

diff --git a/CHANGES.md b/CHANGES.md
index 4423017..454e672 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -14,6 +14,10 @@
 
 (nothing yet)
 
+## 10.2.1
+
+- SAPI-290 getApplicationObjects returns undefined instance list for known services
+
 ## 10.2.0
 
 - PUBAPI-1380 Cloudapi should support wildcards in ListPackages
diff --git a/lib/sapi.js b/lib/sapi.js
index 0cdaa7b..54f317b 100644
--- a/lib/sapi.js
+++ b/lib/sapi.js
@@ -752,16 +752,16 @@ function getApplicationObjects(app_uuid, opts, cb) {
                     return (subcb(err));
 
                 ret.services = {};
+                ret.instances = {};
                 svcs.forEach(function (svc) {
                     ret.services[svc.uuid] = svc;
+                    ret.instances[svc.uuid] = [];
                 });
 
                 return (subcb());
             });
         },
         function (subcb) {
-            ret.instances = {};
-
             var search_opts = {};
             if (opts.include_master)
                 search_opts.include_master = true;
@@ -780,8 +780,6 @@ function getApplicationObjects(app_uuid, opts, cb) {
                     var svc_uuid = inst.service_uuid;
                     if (!ret.services[svc_uuid])
                         return;
-                    if (!ret.instances[svc_uuid])
-                        ret.instances[svc_uuid] = [];
                     ret.instances[svc_uuid].push(inst);
                 });
 
diff --git a/package.json b/package.json
index 10d796e..074c6b7 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "10.2.0",
+  "version": "10.2.1",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
