commit 660d62668f85b92b046c57b28000f96a47ba9d00 (refs/changes/34/2134/1)
Author: Richard Bradley <richard.bradley@joyent.com>
Date:   2017-06-21T12:28:45+00:00 (2 years, 4 months ago)
    
    SAPI-290 getApplicationObjects returns undefined instance list for known services

diff --git a/lib/sapi.js b/lib/sapi.js
index 0cdaa7b..54f317b 100644
--- a/lib/sapi.js
+++ b/lib/sapi.js
@@ -752,16 +752,16 @@ function getApplicationObjects(app_uuid, opts, cb) {
                     return (subcb(err));
 
                 ret.services = {};
+                ret.instances = {};
                 svcs.forEach(function (svc) {
                     ret.services[svc.uuid] = svc;
+                    ret.instances[svc.uuid] = [];
                 });
 
                 return (subcb());
             });
         },
         function (subcb) {
-            ret.instances = {};
-
             var search_opts = {};
             if (opts.include_master)
                 search_opts.include_master = true;
@@ -780,8 +780,6 @@ function getApplicationObjects(app_uuid, opts, cb) {
                     var svc_uuid = inst.service_uuid;
                     if (!ret.services[svc_uuid])
                         return;
-                    if (!ret.instances[svc_uuid])
-                        ret.instances[svc_uuid] = [];
                     ret.instances[svc_uuid].push(inst);
                 });
 
