From b92895224dd4bff2533e59106554734f685628c4 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 12 Jul 2018 19:32:01 +0200
Subject: [PATCH] TRITON-589 TRITON-558 verbosity breaks some sdcadm tests

---
 lib/procedures/index.js | 1 -
 1 file changed, 1 deletion(-)

diff --git a/lib/procedures/index.js b/lib/procedures/index.js
index e9486cf..948416d 100644
--- a/lib/procedures/index.js
+++ b/lib/procedures/index.js
@@ -131,7 +131,6 @@ function coordinatePlan(opts, cb) {
          * If SAPI isn't in full mode, we'll error out and stop the updates
          */
         function ensureSapiFullMode(_, next) {
-            progress('Verifying SAPI full mode');
             sdcadm.sapi.getMode(function getModeCb(err, mode) {
                 if (err) {
                     next(err);
-- 
2.21.0

