From a1c3a4684ac3cfc9d14eb9ee896d504bbe0c08c2 Mon Sep 17 00:00:00 2001
From: YanChii <janci@binaryparadise.com>
Date: Sat, 9 Mar 2019 16:25:29 +0000
Subject: [PATCH] fix qemu detecting a 64-bit host (related to live migration)

---
 qemu-kvm-x86.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/qemu-kvm-x86.c b/qemu-kvm-x86.c
index cb3dad0..4491ed5 100644
--- a/qemu-kvm-x86.c
+++ b/qemu-kvm-x86.c
@@ -508,7 +508,16 @@ int kvm_arch_qemu_create_context(void)
     struct utsname utsname;
 
     uname(&utsname);
-    lm_capable_kernel = strcmp(utsname.machine, "x86_64") == 0;
+    if(strcmp(utsname.sysname, "SunOS") == 0) {
+        // uname in illumos doesn't provide info about 32/64-bit version of booted kernel
+#if defined(_LP64) || defined (__amd64)
+        lm_capable_kernel = 1;
+#else
+        lm_capable_kernel = 0;
+#endif
+    }
+    else
+        lm_capable_kernel = strcmp(utsname.machine, "x86_64") == 0;
 
     if (kvm_shadow_memory) {
         kvm_set_shadow_pages(kvm_context, kvm_shadow_memory);
-- 
2.21.0

