From 0085d83f30c8db3bcf6fa970bd1111b1041971df Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Mon, 24 Sep 2018 23:59:35 -0400
Subject: [PATCH] OS-7262 sysinfo generates invalid JSON due to HVM API

---
 src/sysinfo | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/sysinfo b/src/sysinfo
index f1dbf980..8f1a3f72 100644
--- a/src/sysinfo
+++ b/src/sysinfo
@@ -211,6 +211,7 @@ function get_bhyve_capability()
 {
     Bhyve_Capable="false"
     Bhyve_Max_Vcpus="0"
+    HVM_API="false"
 
     if [[ -x /usr/lib/brand/bhyve/bhhwcompat ]]; then
         if /usr/lib/brand/bhyve/bhhwcompat; then
-- 
2.21.0

