commit a233430ca63d9da22b28aa52f5e03d9785437e9a (refs/changes/04/3904/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-05-03T21:22:19+00:00 (1 year, 5 months ago)
    
    TRITON-365 NAPI ping endpoint broken by TRITON-269

diff --git a/lib/endpoints/ping.js b/lib/endpoints/ping.js
index 5d7f231..45c9058 100644
--- a/lib/endpoints/ping.js
+++ b/lib/endpoints/ping.js
@@ -37,25 +37,16 @@ function ping(req, res, next) {
         status: 'OK'
     };
 
-    if (!req.app.initialDataLoaded) {
-        stats.status = 'loading initial data';
-        stats.healthy = false;
-    }
-
-    if (req.app.moray) {
-        if (!req.app.morayConnected) {
+    if (req.app.isInState('init')) {
+        if (req.app.isInState('init.moray')) {
             stats.services.moray = 'offline';
-            stats.status = 'moray not connected';
-            stats.healthy = false;
         }
-    } else {
-        stats.services.moray = 'offline';
         stats.status = 'initializing';
         stats.healthy = false;
     }
 
     res.send(200, stats);
-    return next();
+    next();
 }
 
 
diff --git a/lib/napi.js b/lib/napi.js
index 3adac09..3f6ae21 100644
--- a/lib/napi.js
+++ b/lib/napi.js
@@ -64,7 +64,6 @@ function NAPI(opts) {
     var self = this;
     this.log = opts.log;
     this.config = opts.config;
-    this.initialDataLoaded = false;
     constants.UFDS_ADMIN_UUID = opts.config.ufdsAdminUuid;
 
     if (opts.config.overlay.enabled) {
diff --git a/test/unit/migration-1.2.0-networks.test.js b/test/unit/migration-1.2.0-networks.test.js
index 3103b7e..44263b0 100644
--- a/test/unit/migration-1.2.0-networks.test.js
+++ b/test/unit/migration-1.2.0-networks.test.js
@@ -563,12 +563,30 @@ test('setup', function (t) {
                     t.end();
                     return;
                 }
-                NAPI = res;
+                NAPI = res.client;
                 t2.ok(NAPI, 'have NAPI client object');
                 t2.end();
             });
         });
     });
+
+
+    t.test('server ping should no longer be initializing', function (t2) {
+        NAPI.ping(function (err, status) {
+            t2.ifError(err, 'ping() error');
+            t2.deepEqual(status, {
+                config: {
+                    fabrics_enabled: true
+                },
+                healthy: true,
+                services: {
+                    moray: 'online'
+                },
+                status: 'OK'
+            });
+            t2.end();
+        });
+    });
 });
 
 test('networks', function (t) {
