From e26ae5ab3f05d51e2d16eb2d27456a3a5c71037c Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody.kantor@gmail.com>
Date: Thu, 31 Aug 2017 22:33:50 +0000
Subject: [PATCH] MANTA-2980 electric-moray should report moray shard for
 operations like 'putObject'

---
 lib/server.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/server.js b/lib/server.js
index 51f7ce0..323cda0 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -411,6 +411,11 @@ function putObject(options) {
                     meta: meta
                 }, 'putObject: returned');
 
+                // Add shard information to the response.
+                if (meta) {
+                    meta._node = node;
+                }
+
                 dtrace['putobject-done'].fire(function () {
                     return ([res.msgid]);
                 });
-- 
2.21.0

