commit 7bd5e32ad1699bee6d4476c937a47aa3907d9fb5 (refs/changes/37/1737/1)
Author: Kody Kantor <Kody.Kantor@gmail.com>
Date:   2017-04-03T16:27:47-05:00 (2 years, 6 months ago)
    
    MANTA-2980 electric-moray should report moray shard

diff --git a/lib/server.js b/lib/server.js
index 51f7ce0..bbc7890 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -411,6 +411,11 @@ function putObject(options) {
                     meta: meta
                 }, 'putObject: returned');
 
+                // add node information for debugging purposes
+                if (node) {
+                  meta._node = node;
+                }
+
                 dtrace['putobject-done'].fire(function () {
                     return ([res.msgid]);
                 });
