From 7bd5e32ad1699bee6d4476c937a47aa3907d9fb5 Mon Sep 17 00:00:00 2001
From: Kody Kantor <Kody.Kantor@gmail.com>
Date: Mon, 3 Apr 2017 16:27:47 -0500
Subject: [PATCH] MANTA-2980 electric-moray should report moray shard

---
 lib/server.js | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/server.js b/lib/server.js
index 51f7ce0..bbc7890 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -411,6 +411,11 @@ function putObject(options) {
                     meta: meta
                 }, 'putObject: returned');
 
+                // add node information for debugging purposes
+                if (node) {
+                  meta._node = node;
+                }
+
                 dtrace['putobject-done'].fire(function () {
                     return ([res.msgid]);
                 });
-- 
2.21.0

