From 818f0c49839b72b7b3a0c635285deb72aa6843e7 Mon Sep 17 00:00:00 2001
From: Kody A Kantor <kody.kantor@gmail.com>
Date: Thu, 31 Aug 2017 22:33:50 +0000
Subject: [PATCH] MANTA-2980 electric-moray should report moray shard for
 operations like 'putObject'

---
 lib/server.js | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/lib/server.js b/lib/server.js
index 7f6d12c..c019a11 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -749,6 +749,9 @@ function putObject(options) {
                 if (pErr) {
                     rpc.fail(pErr);
                 } else {
+                    // Add shard information to the response.
+                    meta._node = node;
+
                     rpc.write(meta);
                     rpc.end();
                 }
-- 
2.21.0

