From f31324a23dc2301cf2e664118c27b5c6be6bcda0 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Mon, 6 May 2019 18:33:04 +0000
Subject: [PATCH] OS-7787 make clean fails if ./tools/builder is already clean

---
 tools/builder/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/builder/Makefile b/tools/builder/Makefile
index 292d62eb..4b9bce31 100644
--- a/tools/builder/Makefile
+++ b/tools/builder/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2010 Joyent Inc., All rights reserved.
+# Copyright 2019 Joyent, Inc.
 #
 
 CC=gcc
@@ -21,4 +21,4 @@ users.c:
 	./build_users_c.sh $(SRC_ROOT)/overlay/generic > users.c.tmp && mv users.c.tmp users.c
 
 clean:
-	rm $(TARGETS) $(OBJS)
+	rm -f $(TARGETS) $(OBJS)
-- 
2.21.0

