commit 5e3e28c398c2aadb0ab9d5df9eefeb2c4787d2eb (refs/changes/62/862/1)
Author: Jerry Jelinek <jerry.jelinek@joyent.com>
Date:   2016-11-09T15:41:39+00:00 (2 years, 11 months ago)
    
    1237 port Intel WifiLink 4965/100/1000/2000/5000/6000 driver

diff --git a/overlay/generic/etc/driver_aliases b/overlay/generic/etc/driver_aliases
index 99ec4d58..ba999d32 100644
--- a/overlay/generic/etc/driver_aliases
+++ b/overlay/generic/etc/driver_aliases
@@ -692,6 +692,35 @@ iprb "pci8086,2449.8086.3018"
 iprb "pci8086,27dc"
 iprb "pci8086,27dc.8086.308d"
 isa "pciclass,060100"
+iwn "pci8086,0082"
+iwn "pci8086,0083"
+iwn "pci8086,0084"
+iwn "pci8086,0085"
+iwn "pci8086,0087"
+iwn "pci8086,0089"
+iwn "pci8086,008a"
+iwn "pci8086,008b"
+iwn "pci8086,008d"
+iwn "pci8086,008e"
+iwn "pci8086,0090"
+iwn "pci8086,0091"
+iwn "pci8086,0887"
+iwn "pci8086,0888"
+iwn "pci8086,088e"
+iwn "pci8086,088f"
+iwn "pci8086,0890"
+iwn "pci8086,0891"
+iwn "pci8086,0892"
+iwn "pci8086,0893"
+iwn "pci8086,0894"
+iwn "pci8086,0895"
+iwn "pci8086,0896"
+iwn "pci8086,0897"
+iwn "pci8086,08ae"
+iwn "pci8086,422b"
+iwn "pci8086,422c"
+iwn "pci8086,4238"
+iwn "pci8086,4239"
 ixgb "pci8086,1048"
 ixgb "pci8086,109e"
 ixgb "pci8086,1a48"
diff --git a/overlay/generic/etc/minor_perm b/overlay/generic/etc/minor_perm
index f29117fb..0a3e732a 100644
--- a/overlay/generic/etc/minor_perm
+++ b/overlay/generic/etc/minor_perm
@@ -273,3 +273,5 @@ skd:* 0600 root sys
 eventfd:* 0666 root sys
 timerfd:* 0666 root sys
 signalfd:* 0666 root sys
+iwn:* 0666 root sys
+clone:iwn 0666 root sys
diff --git a/overlay/generic/etc/name_to_major b/overlay/generic/etc/name_to_major
index 7475db54..cd69fc36 100644
--- a/overlay/generic/etc/name_to_major
+++ b/overlay/generic/etc/name_to_major
@@ -294,3 +294,4 @@ sfxge 297
 pvscsi 298
 vmxnet3s 299
 smrt 300
+iwn 301
