commit aac066b2f155f1a02f606d80993deac940038258 (refs/changes/70/3970/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-04-30T14:51:41-07:00 (1 year, 5 months ago)
    
    MANTA-3672 want support for cross-account RBAC role membership

diff --git a/lib/auth.js b/lib/auth.js
index 997ca9d..555c24d 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -792,7 +792,7 @@ function getActiveRoles(req, res, next) {
         /* JSSTYLED */
         names = requestedRoles.split(/\s*,\s*/);
         req.mahi.getUuid({
-            account: req.caller.account.login,
+            account: req.owner.account.login,
             type: 'role',
             names: names
         }, function (err, lookup) {
@@ -815,6 +815,12 @@ function getActiveRoles(req, res, next) {
     } else {                            // No explicit roles, use default set
         if (req.caller.user) {
             activeRoles = req.caller.user.defaultRoles || [];
+        } else {
+            activeRoles = req.caller.account.defaultRoles || [];
+            activeRoles = activeRoles.filter(function (role) {
+                return (req.caller.roles[role].account ===
+                    req.owner.account.uuid);
+            });
         }
         req.activeRoles = activeRoles;
         setImmediate(next);
@@ -965,7 +971,10 @@ function authorize(req, res, next) {
             next(new AccountBlockedError(req.caller.account.login));
             return;
         case 'NoMatchingRoleTag':
-            next(new NoMatchingRoleTagError());
+            if (!req.activeRoles || !req.activeRoles.length)
+                next(new AuthorizationError(login, req.path(), e));
+            else
+                next(new NoMatchingRoleTagError());
             return;
         case 'InvalidRole':
             next(new InvalidRoleError(e.message));
diff --git a/package.json b/package.json
index 7115b69..23e40fb 100644
--- a/package.json
+++ b/package.json
@@ -31,7 +31,7 @@
         "libuuid": "0.1.2",
         "lru-cache": "2.3.1",
         "lstream": "0.0.4",
-        "mahi": "2.0.1",
+        "mahi": "git+https://github.com/joyent/node-mahi.git#manta-3672",
         "marlin": "git+https://github.com/joyent/manta-marlin.git#master",
         "mime": "1.2.11",
         "moray": "3.4.1",
