commit 57d25449c46c05cf887ba87c6e1e73c08f38af71 (refs/changes/81/3581/2)
Author: Trent Mick <trentm@gmail.com>
Date:   2018-03-10T08:58:12-08:00 (1 year, 7 months ago)
    
    TRITON-226 TRITON-137 broke sapi proto mode setup (i.e. headnode setup)

diff --git a/boot/setup.sh b/boot/setup.sh
index 007c577..6222817 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -66,6 +66,7 @@ MORAY_ADMIN_IPS=${moray_admin_ips}
 NAPI_ADMIN_IPS=${napi_admin_ips}
 VMAPI_ADMIN_IPS=${vmapi_admin_ips}
 CNAPI_ADMIN_IPS=${cnapi_admin_ips}
+SAPI_ADMIN_IP=${sapi_admin_ips}
 EOF
     )
 
@@ -88,6 +89,10 @@ EOF
     "level": "debug"
   },
   "datacenter_name": "$DATACENTER_NAME",
+  "serviceName": "{{SERVICE_NAME}}",
+  "instanceUuid": "{{auto.ZONENAME}}",
+  "serverUuid": "{{auto.SERVER_UUID}}",
+  "adminIp": "$SAPI_ADMIN_IP",
   "moray": {
     "host": "$MORAY_HOST",
     "port": 2020
