commit a16f5759151aab7b9f73c1b86efed41cb1110a55
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2019-07-09T15:50:51-06:00 (3 months ago)
    
    MANTA-4398 Fix electric-boray haproxy configuration

diff --git a/etc/haproxy.cfg.in b/etc/haproxy.cfg.in
index 5f72e5c..ad1ae9b 100644
--- a/etc/haproxy.cfg.in
+++ b/etc/haproxy.cfg.in
@@ -17,9 +17,9 @@ defaults
         timeout queue   0
         timeout connect 500
 
-frontend electric-moray
+frontend electric-boray
         bind *:2020
         default_backend servers
 
 backend servers
-@@ELECTRIC-MORAY_INSTANCES@@
+@@ELECTRIC-BORAY_INSTANCES@@
