From a16f5759151aab7b9f73c1b86efed41cb1110a55 Mon Sep 17 00:00:00 2001
From: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date: Tue, 9 Jul 2019 15:50:51 -0600
Subject: [PATCH] MANTA-4398 Fix electric-boray haproxy configuration

---
 etc/haproxy.cfg.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/etc/haproxy.cfg.in b/etc/haproxy.cfg.in
index 5f72e5c..ad1ae9b 100644
--- a/etc/haproxy.cfg.in
+++ b/etc/haproxy.cfg.in
@@ -17,9 +17,9 @@ defaults
         timeout queue   0
         timeout connect 500
 
-frontend electric-moray
+frontend electric-boray
         bind *:2020
         default_backend servers
 
 backend servers
-@@ELECTRIC-MORAY_INSTANCES@@
+@@ELECTRIC-BORAY_INSTANCES@@
-- 
2.21.0

