From 5dff4441508fac26272f1907061231989ba451df Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 20 Jul 2017 17:19:34 +0200
Subject: [PATCH] OS-6249 nvmeadm should ignore 0-sized namespaces by default

---
 usr/src/cmd/nvmeadm/nvmeadm.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/usr/src/cmd/nvmeadm/nvmeadm.c b/usr/src/cmd/nvmeadm/nvmeadm.c
index 13cace3ead..875253b7ca 100644
--- a/usr/src/cmd/nvmeadm/nvmeadm.c
+++ b/usr/src/cmd/nvmeadm/nvmeadm.c
@@ -539,6 +539,15 @@ static int
 do_list_nsid(int fd, const nvme_process_arg_t *npa)
 {
 	_NOTE(ARGUNUSED(fd));
+	const uint_t format = npa->npa_idns->id_flbas.lba_format;
+	const uint_t bshift = npa->npa_idns->id_lbaf[format].lbaf_lbads;
+
+	/*
+	 * Some devices have extra namespaces with illegal block sizes and
+	 * zero blocks. Don't list them when verbose operation isn't requested.
+	 */
+	if ((bshift < 9 || npa->npa_idns->id_nsize == 0) && verbose == 0)
+		return (0);
 
 	(void) printf("  %s/%s (%s): ", npa->npa_name,
 	    di_minor_name(npa->npa_minor),
-- 
2.21.0

