From 17baa30338e220577d186b72f68f79c096bacafb Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Fri, 30 Aug 2019 10:42:45 +0000
Subject: [PATCH] OS-7973 fbt entry probes incorrectly placed due to
 -fshrink-wrap

---
 Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index ee2192b..e478192 100644
--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2019, Joyent, Inc.
+# Copyright 2019 Joyent, Inc.
 #
 
 include		$(PWD)/../../../build.env
@@ -108,6 +108,7 @@ KERNEL_CFLAGS = \
 	$(ALWAYS_CFLAGS) \
 	-m64 \
 	-mcmodel=kernel \
+	-fno-shrink-wrap \
 	-g \
 	-O \
 	-fno-inline \
-- 
2.21.0

