From 26685f69c1ced571f8863a64d2034ec2ad1f6df3 Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Mon, 26 Nov 2018 14:03:01 -0800
Subject: [PATCH] TRITON-981 get lagging Triton components on the latest
 v0.10.x

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e0ea01e..aa21e2e 100644
--- a/Makefile
+++ b/Makefile
@@ -28,7 +28,7 @@ SMF_MANIFESTS_IN = smf/manifests/backupserver.xml.in \
 # Variables
 #
 
-NODE_PREBUILT_VERSION   := v0.10.26
+NODE_PREBUILT_VERSION   := v0.10.48
 NODE_PREBUILT_TAG       := zone
 # Allow building on a SmartOS image other than sdc-multiarch/13.3.1.
 NODE_PREBUILT_IMAGE=b4bdc598-8939-11e3-bea4-8341f6861379
-- 
2.21.0

