From 868979386fa186bcee47e3773c437633392fdd72 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 21 May 2019 15:54:32 -0700
Subject: [PATCH] TRITON-1691 want cross-account RBAC in cloudapi

---
 lib/auth.js  | 36 +++++++++++++++++++++++-------------
 package.json |  2 +-
 2 files changed, 24 insertions(+), 14 deletions(-)

diff --git a/lib/auth.js b/lib/auth.js
index 9d95ae2..a96b8fb 100644
--- a/lib/auth.js
+++ b/lib/auth.js
@@ -530,10 +530,6 @@ function authorize(req, res, next) {
         return next();
     }
 
-    if (!req.subuser || !req.accountMgmt) {
-        return next(new NotAuthorizedError(FORBIDDEN, account.login));
-    }
-
     var activeRoles;
     function getRoles(_, cb) {
         var asRoleParam = req.params['as-role'];
@@ -544,7 +540,7 @@ function authorize(req, res, next) {
         // not, we load the user's default roles as activeRoles for this
         // operation instead.
         if (requestedRoles && requestedRoles.length) {
-            caller.roles(function (err, roles) {
+            caller.rolesOnAccount(account.uuid, function (err, roles) {
                 if (err) {
                     return cb(err);
                 }
@@ -567,7 +563,9 @@ function authorize(req, res, next) {
                 return cb();
             });
         } else {
-            caller.defaultRoles(function (err, defaultRoles) {
+            caller.defaultRolesOnAccount(account.uuid,
+                function (err, defaultRoles) {
+
                 if (err) {
                     return cb(err);
                 }
@@ -581,15 +579,27 @@ function authorize(req, res, next) {
 
     var info;
     function getUser(_, cb) {
-        mahi.getUser(caller.login, account.login, function (err, _info) {
-            if (err) {
-                return cb(err);
-            }
+        if (req.subuser) {
+            mahi.getUser(caller.login, account.login, function (err, _info) {
+                if (err) {
+                    return cb(err);
+                }
 
-            info = _info;
+                info = _info;
 
-            return cb();
-        });
+                return cb();
+            });
+        } else {
+            mahi.getAccount(caller.login, function (err, _info) {
+                if (err) {
+                    return cb(err);
+                }
+
+                info = _info;
+
+                return cb();
+            });
+        }
     }
 
     var owner;
diff --git a/package.json b/package.json
index 6700250..7cb1674 100644
--- a/package.json
+++ b/package.json
@@ -39,7 +39,7 @@
         "semver": "5.4.1",
         "triton-metrics": "0.1.0",
         "triton-netconfig": "1.3.0",
-        "ufds": "1.5.0",
+        "ufds": "1.6.0",
         "uuid-by-string": "0.6.0",
         "vasync": "2.2.0",
         "verror": "1.10.0"
-- 
2.21.0

