From 5b21f80ef9f054c8d59cc20444932394ef2f6adc Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Tue, 12 Jun 2018 21:28:10 -0700
Subject: [PATCH] MANTA-3757 pgstatsmon needs a registrar for waferlock

---
 targets.json.in | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/targets.json.in b/targets.json.in
index c9de41a..4fb5521 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -921,11 +921,13 @@ cat <<EOF
     "deps": [
       "https://download.joyent.com/pub/build/sdcnode",
       "registrar",
+      "waferlock",
       "amon",
       "config-agent"
     ],
     "tarballs": [
       {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
+      {"tarball": "waferlock/waferlock-pkg-*.tar.bz2", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
       {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
     ]
@@ -1398,9 +1400,11 @@ cat <<EOF
     "public": true,
     "deps": [
       "amon",
-      "config-agent"
+      "config-agent",
+      "registrar"
     ],
     "tarballs": [
+      {"tarball": "registrar/registrar-pkg-*.tar.bz2", "sysroot": "/"},
       {"tarball": "amon/amon-agent-*.tgz", "sysroot": "/opt"},
       {"tarball": "config-agent/config-agent-*.tar.bz2", "sysroot": "/opt/smartdc"}
     ]
-- 
2.21.0

