From 0e5efeaa6eba14a5278089581209098e91764cb5 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 13 Jun 2018 17:28:06 -0700
Subject: [PATCH] MANTA-3719 Need Firewall Rules for Postgres

---
 Makefile        | 20 ++++++++++++++++++++
 targets.json.in | 11 +++++++++++
 2 files changed, 31 insertions(+)

diff --git a/Makefile b/Makefile
index 81feffa..f0559cb 100644
--- a/Makefile
+++ b/Makefile
@@ -2273,6 +2273,26 @@ clean_registrar:
 	(cd build/registrar && gmake distclean)
 
 
+#---- waferlock
+
+_waferlock_stamp=$(WAFERLOCK_BRANCH)-$(TIMESTAMP)-g$(WAFERLOCK_SHA)
+WAFERLOCK_BITS=$(BITS_DIR)/waferlock/waferlock-pkg-$(_waferlock_stamp).tar.bz2
+
+.PHONY: waferlock
+waferlock: $(WAFERLOCK_BITS)
+
+$(WAFERLOCK_BITS): build/waferlock
+	@echo "# Build waferlock: branch $(WAFERLOCK_BRANCH), sha $(WAFERLOCK_SHA), time `date -u +%Y%m%dT%H%M%SZ`"
+	mkdir -p $(BITS_DIR)
+	(cd build/waferlock && LDFLAGS="-L/opt/local/lib -R/opt/local/lib" NPM_CONFIG_CACHE=$(MG_CACHE_DIR)/npm TIMESTAMP=$(TIMESTAMP) BITS_DIR=$(BITS_DIR) gmake release publish)
+	@echo "# Created waferlock bits (time `date -u +%Y%m%dT%H%M%SZ`):"
+	@ls -l $(WAFERLOCK_BITS)
+	@echo ""
+
+clean_waferlock:
+	$(RM) -rf $(BITS_DIR)/waferlock
+	(cd build/waferlock && gmake distclean)
+
 #---- mackerel
 
 _mackerel_stamp=$(MANTA_MACKEREL_BRANCH)-$(TIMESTAMP)-g$(MANTA_MACKEREL_SHA)
diff --git a/targets.json.in b/targets.json.in
index 7c60df1..d676d52 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -854,6 +854,17 @@ cat <<EOF
     ]
   },
 
+  "waferlock": {
+    "build_platform": "20151126T062538Z",
+    "repos": [
+        {"url": "git@github.com:joyent/waferlock.git"}
+    ],
+    "public": true,
+    "deps": [
+      "https://download.joyent.com/pub/build/sdcnode"
+    ]
+  }
+
   "minnow": {
     "build_platform": "20151126T062538Z",
     "repos": [
-- 
2.21.0

