From af6de540418b7eb55cbd9fb5f7b480920e72beb5 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Mon, 20 Nov 2017 18:04:35 +0000
Subject: [PATCH] MANTA-3491 muskie startup emits error message to stderr

---
 lib/uploads/create.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 805116d..a16fb38 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -18,7 +18,7 @@ var util = require('util');
 var verror = require('verror');
 var Ajv = require('ajv');
 var ajv = new Ajv({allErrors: true});
-require('ajv-keywords')(ajv);
+require('ajv-keywords')(ajv, 'prohibited');
 
 var auth = require('../auth');
 var common = require('../common');
-- 
2.21.0

