From 379463b65ff941d5b5520ae648efd2abf2a0f3c5 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Mon, 20 Nov 2017 18:04:35 +0000
Subject: [PATCH] MANTA-3491 muskie startup emits error message to stderr

---
 lib/uploads/create.js | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 805116d..bf719e6 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -18,7 +18,13 @@ var util = require('util');
 var verror = require('verror');
 var Ajv = require('ajv');
 var ajv = new Ajv({allErrors: true});
-require('ajv-keywords')(ajv);
+
+/*
+ * All keywords provided by ajv-keywords must be explicitly imported here. Any
+ * ajv-keywords used without being imported will be silently ignored during
+ * validation.
+ */
+require('ajv-keywords')(ajv, ['prohibited']);
 
 var auth = require('../auth');
 var common = require('../common');
@@ -49,10 +55,10 @@ var schemaValidator = ajv.compile({
                 'if-none-match',
                 'if-modified-since',
                 'if-unmodified-since'
-            ]
+            ],
         }
     },
-    'required': ['objectPath']
+    'required': ['objectPath'],
 });
 
 ///--- Helpers
-- 
2.21.0

