commit 379463b65ff941d5b5520ae648efd2abf2a0f3c5 (refs/changes/68/2968/2)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-11-28T02:12:52+00:00 (1 year, 10 months ago)
    
    MANTA-3491 muskie startup emits error message to stderr

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 805116d..bf719e6 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -18,7 +18,13 @@ var util = require('util');
 var verror = require('verror');
 var Ajv = require('ajv');
 var ajv = new Ajv({allErrors: true});
-require('ajv-keywords')(ajv);
+
+/*
+ * All keywords provided by ajv-keywords must be explicitly imported here. Any
+ * ajv-keywords used without being imported will be silently ignored during
+ * validation.
+ */
+require('ajv-keywords')(ajv, ['prohibited']);
 
 var auth = require('../auth');
 var common = require('../common');
@@ -49,10 +55,10 @@ var schemaValidator = ajv.compile({
                 'if-none-match',
                 'if-modified-since',
                 'if-unmodified-since'
-            ]
+            ],
         }
     },
-    'required': ['objectPath']
+    'required': ['objectPath'],
 });
 
 ///--- Helpers
