commit 30da2ea268ecc4189f6517317ce43529b24436d1 (refs/changes/68/2968/4)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-12-07T21:55:09+00:00 (1 year, 10 months ago)
    
    MANTA-3491 muskie startup emits error message to stderr

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 805116d..f034135 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -18,7 +18,13 @@ var util = require('util');
 var verror = require('verror');
 var Ajv = require('ajv');
 var ajv = new Ajv({allErrors: true});
-require('ajv-keywords')(ajv);
+
+/*
+ * All keywords provided by ajv-keywords must be explicitly imported here. Any
+ * ajv-keywords used without being imported will be silently ignored during
+ * validation.
+ */
+require('ajv-keywords')(ajv, ['prohibited']);
 
 var auth = require('../auth');
 var common = require('../common');
