commit af6de540418b7eb55cbd9fb5f7b480920e72beb5 (refs/changes/68/2968/1)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-11-20T18:04:35+00:00 (1 year, 11 months ago)
    
    MANTA-3491 muskie startup emits error message to stderr

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 805116d..a16fb38 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -18,7 +18,7 @@ var util = require('util');
 var verror = require('verror');
 var Ajv = require('ajv');
 var ajv = new Ajv({allErrors: true});
-require('ajv-keywords')(ajv);
+require('ajv-keywords')(ajv, 'prohibited');
 
 var auth = require('../auth');
 var common = require('../common');
