From 3c7cfd074d3783bc37b00ecc451fbf749956bc17 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Wed, 27 Jun 2018 16:04:56 +0100
Subject: [PATCH] TRITON-552 components using cp -P during their build fail
 with /usr/bin/cp

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 8aaad6d..e243f30 100644
--- a/Makefile
+++ b/Makefile
@@ -97,7 +97,7 @@ service: all $(SMF_MANIFESTS)
 	    $(TOP)/build/service/root/opt/smartdc/boot/
 	cp -R $(TOP)/boot/* \
 	    $(TOP)/build/service/root/opt/smartdc/boot/
-	cp -P smf/manifests/sapi.xml $(SVC_INSTDIR)/smf/manifests
+	cp smf/manifests/sapi.xml $(SVC_INSTDIR)/smf/manifests
 	cp -r $(TOP)/sapi_manifests $(SVC_INSTDIR)/
 	cp -r $(TOP)/test $(SVC_INSTDIR)/
 	cp -PR $(NODE_INSTALL) $(SVC_INSTDIR)/build/node
-- 
2.21.0

