From 98c3caa1598d6d02fe5aba620a6d2bb5f22e3f7a Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Tue, 9 Jul 2019 16:55:01 -0400
Subject: [PATCH] TRITON-1807 papi makefile missing non-SmartOS dev npm block

---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 44370a7..9377361 100644
--- a/Makefile
+++ b/Makefile
@@ -45,6 +45,10 @@ NODE_PREBUILT_VERSION=v6.17.0
 ifeq ($(shell uname -s),SunOS)
         NODE_PREBUILT_TAG=zone64
         NODE_PREBUILT_IMAGE=c2c31b00-1d60-11e9-9a77-ff9f06554b0f
+else
+	NPM=npm
+	NODE=node
+	NPM_EXEC=$(shell which npm)
 endif
 
 ENGBLD_USE_BUILDIMAGE	= true
-- 
2.21.0

