commit 67adffd4e0c190c4bf6d58af8732964f458fc648 (refs/changes/63/563/1)
Author: Trent Mick <trentm@gmail.com>
Date:   2016-09-30T14:37:21-07:00 (3 years ago)
    
    joyent/node-triton#135 parameterize recent 'make cutarelease' task code

diff --git a/Makefile b/Makefile
index b74b5b4..8129a3e 100644
--- a/Makefile
+++ b/Makefile
@@ -55,17 +55,18 @@ versioncheck:
 .PHONY: cutarelease
 cutarelease: versioncheck
 	[[ -z `git status --short` ]]  # If this fails, the working dir is dirty.
-	@ver=$(shell cat package.json | json version) && \
-	    publishedVer=$(shell npm view -j triton@$(shell json -f package.json version) version 2>/dev/null) && \
+	@which json 2>/dev/null 1>/dev/null && \
+	    ver=$(shell json -f package.json version) && \
+	    name=$(shell json -f package.json name) && \
+	    publishedVer=$(shell npm view -j $(shell json -f package.json name)@$(shell json -f package.json version) version 2>/dev/null) && \
 	    if [[ -n "$$publishedVer" ]]; then \
-		echo "error: triton@$$ver is already published to npm"; \
+		echo "error: $$name@$$ver is already published to npm"; \
 		exit 1; \
-	    fi; \
-	    echo "** Are you sure you want to tag and publish triton@$$ver to npm?"; \
-	    echo "** Enter to continue, Ctrl+C to abort."; \
+	    fi && \
+	    echo "** Are you sure you want to tag and publish $$name@$$ver to npm?" && \
+	    echo "** Enter to continue, Ctrl+C to abort." && \
 	    read
 	ver=$(shell cat package.json | json version) && \
-	    name=$(shell cat package.json | json name) && \
 	    date=$(shell date -u "+%Y-%m-%d") && \
 	    git tag -a "$$ver" -m "version $$ver ($$date)" && \
 	    git push --tags origin && \
