commit 2dc77015d45ca65f848af7e5ec2500ce7d4a1780 (refs/changes/47/5647/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2019-02-27T16:54:45+01:00 (7 months ago)
    
    TRITON-1266 vmapi does not store disk UUID upon create

diff --git a/lib/vmapi.js b/lib/vmapi.js
index e454031..c2b5dcf 100644
--- a/lib/vmapi.js
+++ b/lib/vmapi.js
@@ -788,6 +788,7 @@ VMAPI.prototype.removeNics = function (params, options, callback) {
  * @param {String} params.owner_uuid : Optional, the owner of the VM.
  * @param {Object} params.context : Optional, value to pass as x-context header
  * @param {String} params.pci_slot : Optional, PCI slot to be used by new disk
+ * @param {String} params.disk_uuid : Optional, UUID to be used by new disk
  * @param {Number | String} params.size : Size in MiB, or "remaining"
  * @param {Object} options : Request options.
  * @param {Function} callback : of the form f(err, job).
@@ -802,6 +803,9 @@ VMAPI.prototype.createDisk = function createDisk(params, options, callback) {
         throw new TypeError('UUID is required');
     if (params.pci_slot !== undefined && typeof (params.pci_slot) !== 'string')
         throw new TypeError('pci_slot must be a string');
+    if (params.disk_uuid !== undefined &&
+            typeof (params.disk_uuid) !== 'string')
+        throw new TypeError('disk_uuid must be a string');
     if (isNaN(+params.size) && params.size !== 'remaining')
         throw new TypeError('size is required (number or \'remaining\')');
     if (params.origin)
@@ -828,6 +832,7 @@ VMAPI.prototype.createDisk = function createDisk(params, options, callback) {
     }
 
     var args = {
+        disk_uuid: params.disk_uuid,
         pci_slot: params.pci_slot,
         size: params.size
     };
diff --git a/package.json b/package.json
index 7209e02..7aca2b8 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "sdc-clients",
   "description": "node.js client libraries for Triton core REST APIs.",
-  "version": "12.1.1",
+  "version": "12.1.2",
   "homepage": "http://www.joyent.com",
   "repository": {
     "type": "git",
