From c964e170768aba2b758131bbe3fff2fc0b01c982 Mon Sep 17 00:00:00 2001
From: Patrick Mooney <pmooney@pfmooney.com>
Date: Sat, 24 Feb 2018 02:21:49 +0000
Subject: [PATCH] backout OS-6416: needs more work

---
 Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7692a74..165f999 100644
--- a/Makefile
+++ b/Makefile
@@ -73,7 +73,6 @@ SUBDIRS = \
 	screen \
 	socat \
 	tun \
-	uefi-edk2 \
 	uuid \
 	vim \
 	wget \
-- 
2.21.0

