commit c964e170768aba2b758131bbe3fff2fc0b01c982 (refs/changes/69/3469/1)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-02-24T02:21:49+00:00 (1 year, 7 months ago)
    
    backout OS-6416: needs more work

diff --git a/Makefile b/Makefile
index 7692a74..165f999 100644
--- a/Makefile
+++ b/Makefile
@@ -73,7 +73,6 @@ SUBDIRS = \
 	screen \
 	socat \
 	tun \
-	uefi-edk2 \
 	uuid \
 	vim \
 	wget \
