commit f2de7ad2572d849593730648b3d5bc26816ec7dc
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-10-16T20:38:41+00:00 (2 years ago)
    
    MANTA-3440 Add support for Content-Encoding for objects

diff --git a/lib/moray.js b/lib/moray.js
index ca89840..a973d84 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -337,6 +337,8 @@ function createMetadata(options) {
 
     switch (options.type) {
     case 'object':
+        assert.optionalString(options.contentEncoding,
+                'options.contentEncoding');
         assert.number(options.contentLength, 'options.contentLength');
         assert.string(options.contentMD5, 'options.contentMD5');
         assert.string(options.contentType, 'options.contentType');
@@ -350,6 +352,7 @@ function createMetadata(options) {
             });
         }
 
+        md.contentEncoding = options.contentEncoding;
         md.contentLength = options.contentLength;
         md.contentMD5 = options.contentMD5;
         md.contentType = options.contentType;
