From 297bb82e43763dfd17906a0918f078c1616cc906 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Thu, 20 Sep 2018 18:41:04 +0200
Subject: [PATCH] TRITON-777 increase papi MAX_VCPUS to 48

---
 lib/validations.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/validations.js b/lib/validations.js
index 0f583b0..d6866de 100644
--- a/lib/validations.js
+++ b/lib/validations.js
@@ -67,7 +67,7 @@ var MIN_CPUCAP = 20;
 var MIN_LWPS = 250;
 var MIN_VCPUS = 1;
 
-var MAX_VCPUS = 32;
+var MAX_VCPUS = 48;
 var MAX_ZFSIO = 16383;
 
 
-- 
2.21.0

