commit 297bb82e43763dfd17906a0918f078c1616cc906 (refs/changes/55/4855/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2018-09-20T18:41:04+02:00 (1 year, 1 month ago)
    
    TRITON-777 increase papi MAX_VCPUS to 48

diff --git a/lib/validations.js b/lib/validations.js
index 0f583b0..d6866de 100644
--- a/lib/validations.js
+++ b/lib/validations.js
@@ -67,7 +67,7 @@ var MIN_CPUCAP = 20;
 var MIN_LWPS = 250;
 var MIN_VCPUS = 1;
 
-var MAX_VCPUS = 32;
+var MAX_VCPUS = 48;
 var MAX_ZFSIO = 16383;
 
 
