commit d14cd5815c506f90f13ab371a1fd2da0268a2c50 (refs/changes/76/1376/3)
Author: Jordan Hendricks <jordan.hendricks@joyent.com>
Date:   2017-03-31T00:24:44+00:00 (2 years, 6 months ago)
    
    MANTA-2169 Support multipart upload of a single file to Manta

diff --git a/lib/moray.js b/lib/moray.js
index 43396de..ce444b1 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -78,6 +78,18 @@ var DIR_COUNT_SCHEMA = {
 var DIR_COUNT_VERSION = 1;
 
 
+var MANTA_UPLOADS_BUCKET = process.env.MANTA_UPLOADS_BUCKET ||
+    'manta_uploads';
+var MANTA_UPLOADS_SCHEMA = {
+    finalizingType: {
+        type: 'string'
+    },
+    uploadId: {
+        type: 'string'
+    }
+};
+var MANTA_UPLOADS_VERSION = 1;
+
 
 ///--- Internal Functions
 
@@ -98,6 +110,16 @@ function setupMantaBuckets(log, client, cb) {
                 }
             },
             log: log
+        }, {
+            client: client,
+            bucket: MANTA_UPLOADS_BUCKET,
+            opts: {
+                index: MANTA_UPLOADS_SCHEMA,
+                options: {
+                    version: MANTA_UPLOADS_VERSION
+                }
+            },
+            log: log
         }, {
             client: client,
             bucket: DELETE_LOG_BUCKET,
@@ -292,8 +314,6 @@ function recordDeleteLog(req, cb) {
 
 
 function createMetadata(options) {
-    assert.object(options, 'options');
-    assert.string(options.key, 'options.key');
     assert.string(options.owner, 'options.owner');
     assert.string(options.type, 'options.type');
     assert.optionalObject(options.headers, 'options.headers');
@@ -349,7 +369,9 @@ function createMetadata(options) {
         break;
 
     case 'directory':
-        // noop
+        if (options.upload) {
+            md.upload = options.upload;
+        }
         break;
 
     default:
@@ -377,6 +399,62 @@ function createBucket(opts, cb) {
 }
 
 
+// Helper that calls Moray's `putObject`.
+function put(options, cb) {
+    assert.object(options, 'options');
+    assert.object(options.client, 'options.client');
+    assert.object(options.log, 'options.log');
+    assert.string(options.op, 'options.op');
+    assert.ok(options.op === 'putMetadata' ||
+        options.op === 'putFinalizingMetadata');
+    assert.string(options.bucket, 'options.bucket');
+    assert.string(options.key, 'options.key');
+    assert.object(options.md, 'options.md');
+    assert.object(options.putOptions, 'options.putOptions');
+    assert.func(cb, 'callback');
+
+    var client = options.client;
+    var log = options.log;
+    var op = options.op;
+    var bucket = options.bucket;
+    var key = options.key;
+    var md = options.md;
+    var opts = options.putOptions;
+
+    log.debug({
+        key: key,
+        metadata: md,
+        etag: opts.etag,
+        requestId: opts.requestId,
+        headers: opts.headers
+    }, 'Moray.' + op + ': entered');
+
+    var attempts = 0;
+
+    client.putObject(bucket, key, md, opts, function (err) {
+        if (err) {
+            log.debug({
+                err: err,
+                    key: key,
+                    requestId: opts.requestId
+                }, 'Moray.' + op + ': error writing metadata');
+
+            if ((err.name === 'EtagConflictError' ||
+                err.name === 'UniqueAttributeError') &&
+                opts.etag === undefined && ++attempts < 3) {
+                setImmediate(put, options, cb);
+            } else {
+                cb(err);
+            }
+        } else {
+            log.debug({
+                key: key,
+                requestId: opts.requestId
+            }, 'Moray.' + op + ': done');
+            cb(null, md);
+        }
+    });
+}
 
 ///--- API
 
@@ -532,12 +610,15 @@ Moray.prototype.initAttempt = function initAttempt() {
     });
 };
 
+
 Moray.prototype.putMetadata = function putMetadata(options, callback) {
     assert.object(options, 'options');
     assert.string(options.key, 'options.key');
     assert.string(options.requestId, 'options.requestId');
-    assert.object(options, 'options.previousMetadata');
     assert.func(callback, 'callback');
+    if (!options.upload) {
+        assert.optionalObject(options, 'options.previousMetadata');
+    }
 
     callback = once(callback);
 
@@ -548,51 +629,72 @@ Moray.prototype.putMetadata = function putMetadata(options, callback) {
         return;
     }
 
-    var attempts = 0;
-    var client = this.client;
-    var key = options.key;
-    var log = this.log;
-    var md = createMetadata(options);
-    var opts = {
+    var putOptions = {
         req_id: options.requestId,
-        etag: options._etag,
-        headers: {
+        etag: options._etag
+    };
+
+    if (!options.upload) {
+        putOptions.headers = {
             'x-muskie-prev-metadata': options.previousMetadata
-        }
+        };
+    }
+
+    var opts = {
+        client: this.client,
+        log: this.log,
+        key: options.key,
+        op: 'putMetadata',
+        bucket: BUCKET,
+        md: createMetadata(options),
+        putOptions: putOptions
     };
 
-    log.debug({
-        key: key,
-        metadata: md,
-        etag: opts.etag,
-        requestId: opts.requestId,
-        headers: opts.headers
-    }, 'Moray.putMetadata: entered');
-    (function put() {
-        client.putObject(BUCKET, key, md, opts, function (err) {
-            if (err) {
-                log.debug({
-                    err: err,
-                    key: key,
-                    requestId: opts.requestId
-                }, 'Moray.putMetadata: error writing metadata');
+    put(opts, callback);
+};
 
-                if ((err.name === 'EtagConflictError' ||
-                     err.name === 'UniqueAttributeError') &&
-                    opts.etag === undefined && ++attempts < 3) {
-                    process.nextTick(put);
-                } else {
-                    callback(err);
-                }
-            } else {
-                log.debug({
-                    key: key,
-                    requestId: opts.requestId
-                }, 'Moray.putMetadata: done');
-                callback(null, md);
-            }
+
+/*
+ * Used by the multipart upload API, this function will store a
+ * finalizing record in the special multipart upload bucket.
+ */
+Moray.prototype.putFinalizingMetadata =
+function putFinalizingMetadata(key, options, callback) {
+    assert.string(key, 'key');
+    assert.object(options, 'options');
+    assert.string(options.uploadId, 'options.uploadId');
+    assert.string(options.finalizingType, 'options.finalizingType');
+    assert.string(options.owner, 'options.owner');
+    assert.string(options.requestId, 'options.requestId');
+    assert.string(options.objectPath, 'options.objectPath');
+    assert.string(options.objectId, 'options.objectId');
+    assert.func(callback, 'callback');
+
+    callback = once(callback);
+
+    if (!this.client) {
+        setImmediate(function () {
+            callback(new Error('not connected'));
         });
-    })();
+        return;
+    }
+
+    var putOptions = {
+        req_id: options.requestId,
+        etag: options._etag
+    };
+
+    var opts = {
+        client: this.client,
+        log: this.log,
+        key: key,
+        op: 'putFinalizingMetadata',
+        bucket: MANTA_UPLOADS_BUCKET,
+        md: options,
+        putOptions: putOptions
+    };
+
+    put(opts, callback);
 };
 
 
@@ -643,6 +745,57 @@ Moray.prototype.getMetadata = function getMetadata(options, callback) {
 };
 
 
+/*
+ * Used by the multipart upload API, this function will fetch a
+ * finalizing record from the special multipart upload bucket.
+ */
+Moray.prototype.getFinalizingMetadata =
+function getFinalizingMetadata(options, callback) {
+    assert.object(options, 'options');
+    assert.string(options.key, 'options.key');
+    assert.string(options.requestId, 'options.requestId');
+    assert.func(callback, 'callback');
+
+    if (!this.client) {
+        setImmediate(function () {
+            callback(new Error('not connected'));
+        });
+        return;
+    }
+
+    var client = this.client;
+    var key = options.key;
+    var log = this.log;
+    var opts = {
+        req_id: options.requestId,
+        noCache: true
+    };
+
+    log.debug({
+        key: key,
+        requestId: opts.requestId
+    }, 'Moray.getFinalizingMetadata: entered');
+
+    client.getObject(MANTA_UPLOADS_BUCKET, key, opts, function (err, md) {
+        if (err) {
+            log.debug({
+                err: err,
+                key: key,
+                requestId: opts.requestId
+            }, 'Moray.getFinalizingMetadata: error reading metadata');
+            callback(err);
+        } else {
+            log.debug({
+                key: key,
+                metadata: md.value,
+                requestId: opts.requestId
+            }, 'Moray.getFinalizingMetadata: done');
+            callback(null, md.value, md);
+        }
+    });
+};
+
+
 Moray.prototype.delMetadata = function delMetadata(options, callback) {
     assert.object(options, 'options');
     assert.string(options.key, 'options.key');
@@ -746,6 +899,48 @@ Moray.prototype.getDirectoryCount = function getDirectoryCount(options, cb) {
 };
 
 
+/*
+ * Commits a multipart upload by performing a Moray batch insertion of a
+ * finalizing record in the Manta multipart uploads bucket and an object record
+ * for the target object being committed.
+ */
+Moray.prototype.commitMPU = function commitMPU(options, cb) {
+    assert.object(options, 'options');
+    assert.arrayOfObject(options.requests, 'options.requests');
+    assert.ok(options.requests.length === 2);
+    assert.string(options.requestId, 'options.requestId');
+    assert.func(cb, 'callback');
+
+    cb = once(cb);
+
+    if (!this.client) {
+        setImmediate(function () {
+            cb(new Error('not connected'));
+        });
+        return;
+    }
+
+    var client = this.client;
+    var log = this.log;
+    var opts = {
+        req_id: options.requestId
+    };
+
+    log.debug({
+        requests: options.requests,
+        requestId: opts.requestId
+    }, 'Moray.batch: entered');
+
+    client.batch(options.requests, opts, function (err, meta) {
+        if (err) {
+            cb(err);
+        } else {
+            cb(null, meta);
+        }
+    });
+};
+
+
 Moray.prototype.ping = function ping(opts, cb) {
     if (typeof (opts) === 'function') {
         cb = opts;
