commit 490f579943cbcc14d9ef46054e28fddbdd48f407 (refs/changes/76/1376/1)
Author: Jordan Hendricks <jordan.hendricks@joyent.com>
Date:   2017-01-31T01:52:29+00:00 (2 years, 8 months ago)
    
    MANTA-2169 Support multipart upload of a single file to Manta

diff --git a/lib/moray.js b/lib/moray.js
index 12dbed3..b150366 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 //
@@ -77,6 +77,12 @@ var DIR_COUNT_SCHEMA = {
 var DIR_COUNT_VERSION = 1;
 
 
+var MANTA_UPLOADS_BUCKET = process.env.MANTA_UPLOADS_BUCKET ||
+    'manta_uploads';
+var MANTA_UPLOADS_SCHEMA = {
+};
+var MANTA_UPLOADS_VERSION = 1;
+
 
 ///--- Internal Functions
 
@@ -97,6 +103,16 @@ function setupMantaBuckets(log, client, cb) {
                 }
             },
             log: log
+        }, {
+            client: client,
+            bucket: MANTA_UPLOADS_BUCKET,
+            opts: {
+                index: MANTA_UPLOADS_SCHEMA,
+                options: {
+                    version: MANTA_UPLOADS_VERSION
+                }
+            },
+            log: log
         }, {
             client: client,
             bucket: DELETE_LOG_BUCKET,
