commit 62d7491a4ce7d157e16272d84418c604a0f91a33 (refs/changes/45/3445/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-02-22T12:57:16-08:00 (1 year, 7 months ago)
    
    TRITON-173 CNAPI using wrong origin image

diff --git a/targets.json.in b/targets.json.in
index a0d3dae..bd29611 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -785,8 +785,14 @@ cat <<EOF
     "image_name": "cnapi",
     "image_description": "SDC CNAPI",
     "image_version": "1.0.0",
-    "image_uuid": "fd2cc906-8938-11e3-beab-4359c665ac99",
-    "pkgsrc": [],
+    "// image_uuid": "sdc-minimal-multiarch-lts@15.4.1",
+    "image_uuid": "18b094b0-eb01-11e5-80c1-175dac7ddf02",
+    "pkgsrc": [
+      "coreutils-8.23nb2",
+      "curl-7.47.1",
+      "gsed-4.2.2nb4",
+      "patch-2.7.5"
+    ],
     "repos": [
         {"url": "git@github.com:joyent/sdc-cnapi.git"}
     ],
