commit f77d3acb1765315ab5e1989f1f94e9b47b299a89 (refs/changes/23/4823/2)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-09-24T16:26:40+00:00 (1 year ago)
    
    OS-7238 bhyve zones should add hostbridge device

diff --git a/usr/src/lib/brand/bhyve/zone/boot.c b/usr/src/lib/brand/bhyve/zone/boot.c
index 1adea7b26f..ae6c86ced5 100644
--- a/usr/src/lib/brand/bhyve/zone/boot.c
+++ b/usr/src/lib/brand/bhyve/zone/boot.c
@@ -42,7 +42,7 @@
 #define	DEFAULT_BOOTROM_CSM	"/usr/share/bhyve/uefi-csm-rom.bin"
 
 typedef enum {
-	PCI_SLOT_HOSTBRIDGE = 0,	/* Not used here, but reserved */
+	PCI_SLOT_HOSTBRIDGE = 0,
 	PCI_SLOT_CD = 3,		/* Windows ahci allows slots 3 - 6 */
 	PCI_SLOT_BOOT_DISK,
 	PCI_SLOT_OTHER_DISKS,
@@ -448,6 +448,24 @@ add_lpc(int *argc, char **argv)
 	return (0);
 }
 
+static int
+add_hostbridge(int *argc, char **argv)
+{
+	char conf[MAXPATHLEN];
+
+	/*
+	 * Simply add the default hostbridge for now.  If there is a need for
+	 * it to be configurable, such flexibility can be added later.
+	 */
+	(void) snprintf(conf, sizeof (conf), "%d,hostbridge",
+	    PCI_SLOT_HOSTBRIDGE);
+	if (add_arg(argc, argv, "-s") != 0 ||
+	    add_arg(argc, argv, conf) != 0) {
+		return (-1);
+	}
+	return (0);
+}
+
 static int
 add_bhyve_extra_opts(int *argc, char **argv)
 {
@@ -624,6 +642,7 @@ main(int argc, char **argv)
 
 	if (add_smbios(&zhargc, (char **)&zhargv) != 0 ||
 	    add_lpc(&zhargc, (char **)&zhargv) != 0 ||
+	    add_hostbridge(&zhargc, (char **)&zhargv) != 0 ||
 	    add_cpu(&zhargc, (char **)&zhargv) != 0 ||
 	    add_ram(&zhargc, (char **)&zhargv) != 0 ||
 	    add_devices(&zhargc, (char **)&zhargv) != 0 ||
