commit ff9caea31bf0a8ca62ee52e547a92027370a4122 (refs/changes/75/4775/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-08-31T16:48:37+00:00 (1 year, 1 month ago)
    
    MANATEE-411 ManateePrimaryResolver should handle an empty cluster

diff --git a/manatee.js b/manatee.js
index 4007c8e..8fe9380 100644
--- a/manatee.js
+++ b/manatee.js
@@ -616,6 +616,11 @@ ManateePrimaryResolver.prototype.state_running = function (S) {
             urls: urls
         }, 'manatee topology changed');
 
+        if (urls.length === 0) {
+            /* In a brand-new cluster, there will be nothing yet. */
+            return;
+        }
+
         var primary = mod_url.parse(urls[0]);
 
         assert.strictEqual(primary.protocol, 'tcp:');
