From 7566e3044278d5fb59669026770dbc9bc262d1eb Mon Sep 17 00:00:00 2001
From: Jason King <jason.king@joyent.com>
Date: Fri, 9 Mar 2018 17:35:44 -0600
Subject: [PATCH] OS-6647 smartos build dying in ncurses with syntax error

---
 ncurses/Makefile | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/ncurses/Makefile b/ncurses/Makefile
index 00510c7..e7fd906 100644
--- a/ncurses/Makefile
+++ b/ncurses/Makefile
@@ -18,7 +18,7 @@
 #
 # CDDL HEADER END
 #
-# Copyright (c) 2012, Joyent, Inc.  All rights reserved.
+# Copyright (c) 2018, Joyent, Inc.  All rights reserved.
 #
 
 VER =	ncurses-5.7
@@ -36,6 +36,12 @@ AUTOCONF_OPTS += \
 AUTOCONF_OPTS.64 += \
 	--libdir=/usr/gnu/lib/amd64
 
+# /bin/sh (aka ksh93)'s cat builtin likes to omit the final newline, which
+# causes problems with generating headers.sh using successive here docs.
+# Use bash for now to avoid lines being merged by this.
+AUTOCONF_ENV += 	CONFIG_SHELL=/bin/bash
+AUTOCONF_ENV.64 +=	CONFIG_SHELL=/bin/bash
+
 LDFLAGS +=	-R/usr/gnu/lib
 LDFLAGS.64 +=	-R/usr/gnu/lib/amd64
 
-- 
2.21.0

