commit 19aa3be9430a83034bb1dcbd3213ef4d25355199 (refs/changes/80/3580/1)
Author: Jason King <jason.king@joyent.com>
Date:   2018-03-09T22:35:33-06:00 (1 year, 7 months ago)
    
    OS-6647 smartos build dying in ncurses with syntax error

diff --git a/ncurses/Makefile b/ncurses/Makefile
index 00510c7..e8a26de 100644
--- a/ncurses/Makefile
+++ b/ncurses/Makefile
@@ -36,6 +36,9 @@ AUTOCONF_OPTS += \
 AUTOCONF_OPTS.64 += \
 	--libdir=/usr/gnu/lib/amd64
 
+AUTOCONF_ENV += 	CONFIG_SHELL=/bin/bash
+AUTOCONF_ENV.64 +=	CONFIG_SHELL=/bin/bash
+
 LDFLAGS +=	-R/usr/gnu/lib
 LDFLAGS.64 +=	-R/usr/gnu/lib/amd64
 
