commit f9378b594e11b2cf319fb8d428e8a858afc4c4b9 (refs/changes/80/3580/2)
Author: Jason King <jason.king@joyent.com>
Date:   2018-03-09T22:53:47-06:00 (1 year, 7 months ago)
    
    OS-6647 smartos build dying in ncurses with syntax error

diff --git a/ncurses/Makefile b/ncurses/Makefile
index 00510c7..25ddc00 100644
--- a/ncurses/Makefile
+++ b/ncurses/Makefile
@@ -18,7 +18,7 @@
 #
 # CDDL HEADER END
 #
-# Copyright (c) 2012, Joyent, Inc.  All rights reserved.
+# Copyright (c) 2018, Joyent, Inc.  All rights reserved.
 #
 
 VER =	ncurses-5.7
@@ -36,6 +36,12 @@ AUTOCONF_OPTS += \
 AUTOCONF_OPTS.64 += \
 	--libdir=/usr/gnu/lib/amd64
 
+# /bin/sh (aka ksh93)'s cat builtin likes to omit the final newline, which
+# causes problems with generating headers.sh using successive here docs. 
+# Use bash for now to avoid lines being merged by this.
+AUTOCONF_ENV += 	CONFIG_SHELL=/bin/bash
+AUTOCONF_ENV.64 +=	CONFIG_SHELL=/bin/bash
+
 LDFLAGS +=	-R/usr/gnu/lib
 LDFLAGS.64 +=	-R/usr/gnu/lib/amd64
 
