From 600369fe8194f2cf255e66d600dcd09881f5db0f Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Thu, 21 Mar 2019 13:46:47 +0000
Subject: [PATCH] MANTA-4191 remove deps/node submodule remnants from mola and
 madtom

---
 .gitmodules | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/.gitmodules b/.gitmodules
index 2717da8..c896b7b 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -7,9 +7,6 @@
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
 	url = https://github.com/davepacheco/javascriptlint.git
-[submodule "deps/node"]
-	path = deps/node
-	url = https://github.com/joyent/node.git
 [submodule "deps/manta-scripts"]
 	path = deps/manta-scripts
 	url = https://github.com/joyent/manta-scripts.git
-- 
2.21.0

