From 6e095d5c57bf04f09196075fa02e076eb248094f Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Thu, 6 Dec 2018 11:32:49 +0000
Subject: [PATCH] OS-7431 bump ramdisk size for GCC7

---
 tools/build_live | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/build_live b/tools/build_live
index 280c2dd8..022d134c 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -893,7 +893,7 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_root_size=278000
+bi_root_size=300000
 if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
     grep "DEBUG enabled" >/dev/null; then
 	#
-- 
2.21.0

