commit 6e095d5c57bf04f09196075fa02e076eb248094f (refs/changes/90/5190/2)
Author: John Levon <john.levon@joyent.com>
Date:   2018-12-06T11:32:49+00:00 (10 months ago)
    
    OS-7431 bump ramdisk size for GCC7

diff --git a/tools/build_live b/tools/build_live
index 280c2dd8..022d134c 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -893,7 +893,7 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_root_size=278000
+bi_root_size=300000
 if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
     grep "DEBUG enabled" >/dev/null; then
 	#
