commit 8df27bfd1b17a820fb292da69b7725e3382314a6 (refs/changes/90/5190/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-12-06T11:29:15+00:00 (10 months ago)
    
    OS-7341 remove gcc6

diff --git a/tools/build_live b/tools/build_live
index 280c2dd8..022d134c 100755
--- a/tools/build_live
+++ b/tools/build_live
@@ -893,7 +893,7 @@ bi_setup_work_dir
 # Create and mount the "/" (root) ramdisk image:
 #
 bi_file_root="$bi_tmpdir/root.image"
-bi_root_size=278000
+bi_root_size=300000
 if strings "$bi_wsroot/proto/kernel/amd64/genunix" |
     grep "DEBUG enabled" >/dev/null; then
 	#
