From 7163fe667fa72fbaa86f774c745a38740d8e054a Mon Sep 17 00:00:00 2001
From: Brittany Wald <brittany.wald@joyent.com>
Date: Tue, 5 Sep 2017 23:57:00 +0000
Subject: [PATCH] MANTA-3387 node-fash npm publish cruft

---
 .npmignore | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/.npmignore b/.npmignore
index a56a7ef..3aaf0e8 100644
--- a/.npmignore
+++ b/.npmignore
@@ -1,2 +1,13 @@
 node_modules
-
+.gitmodules
+test/in_memory.test.js
+test/leveldb.test.js
+test/load.test.js
+Makefile
+tools/mk/Makefile.defs
+tools/mk/Makefile.deps
+tools/mk/Makefile.node_modules.defs
+tools/mk/Makefile.node_modules.targ
+tools/mk/Makefile.targ
+tools/jsl.node.conf
+tools/jsstyle.conf
-- 
2.21.0

