commit 0796e1bc691f89266e47a16cb003da88d7ee1638 (refs/changes/98/2498/2)
Author: Brittany Wald <brittany.wald@joyent.com>
Date:   2017-09-06T00:07:13+00:00 (2 years, 1 month ago)
    
    MANTA-3387 node-fash npm publish cruft

diff --git a/.npmignore b/.npmignore
index a56a7ef..3025614 100644
--- a/.npmignore
+++ b/.npmignore
@@ -1,2 +1,6 @@
+deps
 node_modules
-
+test
+tools
+.gitmodules
+Makefile
