From 6ba83321f5e573b80e6941587560ec0162b0323a Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Wed, 7 Feb 2018 17:03:19 -0800
Subject: [PATCH] MORAY-455 moray asks fast to hold onto too many recent RPC
 requests

---
 lib/client.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/client.js b/lib/client.js
index 50bab99..1a2a6da 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -35,7 +35,7 @@ var parseMorayParameters = require('./client_params').parseMorayParameters;
 
 ///--- Default values for function arguments
 
-var fastNRecentRequests         = 30;
+var fastNRecentRequests         = 5;
 var dflClientTcpKeepAliveIdle   = 10000;  /* milliseconds */
 
 /*
-- 
2.21.0

