commit 07379df0c642531d8f76136bf10238107bc52b34 (refs/changes/83/83/1)
Author: Joshua M. Clulow <jmc@joyent.com>
Date:   2016-07-14T17:00:54-07:00 (3 years, 3 months ago)
    
    TOOLS-1486 gerrit signs out users who use multiple tabs

diff --git a/images/appserver/Dockerfile b/images/appserver/Dockerfile
index 0d00b1c..57005f7 100644
--- a/images/appserver/Dockerfile
+++ b/images/appserver/Dockerfile
@@ -11,6 +11,16 @@ FROM java:jre-alpine
 #
 ENV GERRIT_VERSION            2.12.2
 ENV GERRIT_RELEASEURL         https://gerrit-releases.storage.googleapis.com
+#ENV GERRIT_WAR_URL            ${GERRIT_RELEASEURL}/gerrit-${GERRIT_VERSION}.war
+
+#
+# Work around "https://gerrit-review.googlesource.com/#/c/74830/2" by using
+# a custom build of 2.12.2 with the upstream fix for this issue included
+# as well.  Once a release that includes this fix is available, we should
+# switch back to using official Gerrit release artefacts.
+#
+ENV GERRIT_OVERRIDEURL        https://us-east.manta.joyent.com/Joyent_Dev/public/gerrit/custom
+ENV GERRIT_WAR_URL            ${GERRIT_OVERRIDEURL}/gerrit-2.12.2-joyent1.war
 
 ENV PLUGIN_VERSION            stable-2.12
 ENV GERRITFORGE_URL           https://gerrit-ci.gerritforge.com
@@ -57,8 +67,7 @@ RUN  set -o xtrace && \
      mkdir ${GERRIT_HOME}/shipped && \
      mkdir ${GERRIT_HOME}/shipped/plugins && \
      mkdir ${GERRIT_HOME}/shipped/lib && \
-     wget -O $GERRIT_WAR \
-         ${GERRIT_RELEASEURL}/gerrit-${GERRIT_VERSION}.war && \
+     wget -O $GERRIT_WAR "$GERRIT_WAR_URL" && \
      wget -O ${GERRIT_HOME}/shipped/plugins/delete-project.jar \
          ${GERRITFORGE_URL}/job/plugin-delete-project-${PLUGIN_VERSION}/${GERRITFORGE_ARTIFACT_DIR}/delete-project/delete-project.jar && \
      wget -O ${GERRIT_HOME}/shipped/plugins/events-log.jar \
