commit 92fa4fd69606559212b18f906863de4bbaf89872 (refs/changes/39/1539/2)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-03-09T16:20:11+01:00 (2 years, 7 months ago)
    
    WORKFLOW-217 wf-moray-backend leaks Moray clients

diff --git a/lib/workflow-moray-backend.js b/lib/workflow-moray-backend.js
index 69fd7ef..f499b30 100644
--- a/lib/workflow-moray-backend.js
+++ b/lib/workflow-moray-backend.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var util = require('util');
@@ -515,6 +515,9 @@ var WorkflowMorayBackend = module.exports = function (config) {
                 retry.reset();
                 connectCb();
             } else {
+                if (client) {
+                    client.close();
+                }
                 client = makeClient();
                 client.on('connect', onConnect);
                 client.on('error', onError);
