From 92fa4fd69606559212b18f906863de4bbaf89872 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Mon, 20 Feb 2017 16:50:24 +0100
Subject: [PATCH] WORKFLOW-217 wf-moray-backend leaks Moray clients

---
 lib/workflow-moray-backend.js | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/lib/workflow-moray-backend.js b/lib/workflow-moray-backend.js
index 69fd7ef..f499b30 100644
--- a/lib/workflow-moray-backend.js
+++ b/lib/workflow-moray-backend.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var util = require('util');
@@ -515,6 +515,9 @@ var WorkflowMorayBackend = module.exports = function (config) {
                 retry.reset();
                 connectCb();
             } else {
+                if (client) {
+                    client.close();
+                }
                 client = makeClient();
                 client.on('connect', onConnect);
                 client.on('error', onError);
-- 
2.21.0

