commit b0757439778142b8fed91fe6675c6ee6df4466cf (refs/changes/44/144/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-07-26T18:18:44+00:00 (3 years, 2 months ago)
    
    TOOLS-1497 mg shoud build firmware tools from github

diff --git a/targets.json.in b/targets.json.in
index 730d741..6ad7448 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -77,7 +77,7 @@ if [[ "${JOYENT_BUILD}" == "true" ]]; then
   FIRMWARE_TOOLS_ENTRY='
   "firmware-tools": {
     "repos": [
-        {"url": "git@git.joyent.com:firmware-tools.git"}
+        {"url": "git@github.com:joyent/firmware-tools.git"}
     ],
     "public": false,
     "deps": []
