From b0757439778142b8fed91fe6675c6ee6df4466cf Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Tue, 26 Jul 2016 18:18:41 +0000
Subject: [PATCH] TOOLS-1497 mg shoud build firmware tools from github

---
 targets.json.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/targets.json.in b/targets.json.in
index 730d741..6ad7448 100644
--- a/targets.json.in
+++ b/targets.json.in
@@ -77,7 +77,7 @@ if [[ "${JOYENT_BUILD}" == "true" ]]; then
   FIRMWARE_TOOLS_ENTRY='
   "firmware-tools": {
     "repos": [
-        {"url": "git@git.joyent.com:firmware-tools.git"}
+        {"url": "git@github.com:joyent/firmware-tools.git"}
     ],
     "public": false,
     "deps": []
-- 
2.21.0

