commit 49571a9d11b7006d89df19fbdff11cae4eca867b
Author: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
Date:   2017-12-19T14:55:57-07:00 (1 year, 10 months ago)
    
    MANTA-2133 Manta object reads should not require shard 1

diff --git a/lib/server.js b/lib/server.js
index 3e3622e..20a920e 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -214,21 +214,21 @@ function createServer(options, clients, name) {
         var ok = true;
         var errors = [];
 
-        if (!clients.picker) {
-            req.log.error('picker unavailable');
-            ok = false;
-        } else if (!clients.moray) {
+        if (!clients.moray) {
             req.log.error('index moray unavailable');
             ok = false;
         } else if (!clients.mahi) {
             req.log.error('mahi unavailable');
             ok = false;
-        } else if (!clients.marlin) {
+        } else if (!clients.picker && !req.isReadOnly()) {
+            req.log.error('picker unavailable');
+            ok = false;
+        } else if (!clients.marlin && !req.isMarlinRequest()) {
             req.log.error('marlin unavailable');
-            ok = !req.isMarlinRequest();
-        } else if (!clients.medusa) {
+            ok = false;
+        } else if (!clients.medusa && !req.isMedusaRequest()) {
             req.log.error('medusa unavailable');
-            ok = !req.isMedusaRequest();
+            ok = false;
         }
 
         if (!ok) {
