commit c856555956c859ba10714dba6bb5815df78e6f3f (refs/changes/99/2599/1)
Author: Nick Zivkovic <nick.zivkovic@joyent.com>
Date:   2017-09-18T20:29:14+00:00 (2 years, 1 month ago)
    
    DOCKER-1107 Docker needlessly fails an assertion

diff --git a/lib/backends/sdc/networks.js b/lib/backends/sdc/networks.js
index 38fb59e..04eac3a 100644
--- a/lib/backends/sdc/networks.js
+++ b/lib/backends/sdc/networks.js
@@ -384,7 +384,7 @@ function getNetworksOrPools(params, opts, callback) {
             cb(null, {res: []});
         }
     ], function (err, res) {
-        assert(res.err || res.data);
+        assert(res.err || res.res);
         callback(res.err, res.res);
     });
 }
