commit b49bcdc824a271f8e428fe89ac69666eb6e875b9
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-18T10:34:31-04:00 (7 months ago)
    
    TOOLS-2207 eng submodule should use https (not git@)

diff --git a/.gitmodules b/.gitmodules
index 3927626..bc6f366 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -6,4 +6,4 @@
 	url = git://github.com/davepacheco/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
-	url = git@github.com:joyent/eng.git
+	url = https://github.com/joyent/eng.git
