commit 6137f595ae25c90aac8b330fd2d82fc0ffa5a03d (refs/changes/30/130/1)
Author: Dave Pacheco <dap@joyent.com>
Date:   2016-07-21T17:41:16-07:00 (3 years, 3 months ago)
    
    TOOLS-1494 crrestore should pull latest images

diff --git a/bin/crrestore b/bin/crrestore
index a933acf..9a24958 100755
--- a/bin/crrestore
+++ b/bin/crrestore
@@ -118,6 +118,12 @@ function main
 	gr_backup_path="$1"
 	gr_configure || fail "failed to configure"
 
+	echo "Pulling latest copies of cr.joyent.us images ... "
+	if ! docker pull $gr_image_appserver ||
+	   ! docker pull $gr_image_frontdoor; then
+	   	fail "failed"
+	fi
+
 	echo -n "Creating database volume container ... "
 	docker run --name=$gr_name_volume_db \
 	    -v /gerrit-postgres-db-data $gr_image_volumes \
