commit c65735c352e24da53aa7400799ac2e00975f3a0d (refs/changes/63/63/1)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-07-12T15:53:47+00:00 (3 years, 3 months ago)
    
    OS-5492 Tune physical block size for Samsung client SSDs

diff --git a/overlay/generic/kernel/drv/sd.conf b/overlay/generic/kernel/drv/sd.conf
index f57fc874..a977eda3 100644
--- a/overlay/generic/kernel/drv/sd.conf
+++ b/overlay/generic/kernel/drv/sd.conf
@@ -67,8 +67,14 @@ ddi-devid-registrant=1;
 # additional 2 retries for "victim" IOs if a reset is needed.  Retrying is
 # very rarely successful, and it is preferable to let ZFS do it where needed.
 #
+# For the Samsung drives, please see OS-5942 for more information, but
+# the FTL on these are 4K.
+#
 sd-config-list=
 	"", "retries-timeout:1,retries-busy:1,retries-reset:1,retries-victim:2",
 	"DELL    PERC H710", "cache-nonvolatile:true",
 	"DELL    PERC H700", "cache-nonvolatile:true",
-	"DELL    PERC/6i", "cache-nonvolatile:true";
+	"DELL    PERC/6i", "cache-nonvolatile:true",
+	"ATA     Samsung SSD 830", "physical-block-size:4096",
+	"ATA     Samsung SSD 840", "physical-block-size:4096",
+	"ATA     Samsung SSD 850", "physical-block-size:4096";
