commit cd172411929a72540ac359d968c3018f7e86b049 (refs/changes/63/63/4)
Author: Robert Mustacchi <rm@joyent.com>
Date:   2016-07-19T21:13:05+00:00 (3 years, 3 months ago)
    
    OS-5492 Tune physical block size for Samsung client SSDs

diff --git a/overlay/generic/kernel/drv/sd.conf b/overlay/generic/kernel/drv/sd.conf
index f57fc874..fce0d72c 100644
--- a/overlay/generic/kernel/drv/sd.conf
+++ b/overlay/generic/kernel/drv/sd.conf
@@ -67,8 +67,16 @@ ddi-devid-registrant=1;
 # additional 2 retries for "victim" IOs if a reset is needed.  Retrying is
 # very rarely successful, and it is preferable to let ZFS do it where needed.
 #
+# For the Samsung client drives, users have seen data corruption when they use
+# the advertised 512 byte sectors. The actual sector size of the flash
+# translation layer is 4K, so it's relatively safe to make this change which
+# annecdotally solves that problem.
+#
 sd-config-list=
 	"", "retries-timeout:1,retries-busy:1,retries-reset:1,retries-victim:2",
 	"DELL    PERC H710", "cache-nonvolatile:true",
 	"DELL    PERC H700", "cache-nonvolatile:true",
-	"DELL    PERC/6i", "cache-nonvolatile:true";
+	"DELL    PERC/6i", "cache-nonvolatile:true",
+	"ATA     Samsung SSD 830", "physical-block-size:4096",
+	"ATA     Samsung SSD 840", "physical-block-size:4096",
+	"ATA     Samsung SSD 850", "physical-block-size:4096";
