commit 757568d1a11ef0bf484e0f14a6672f3d59391444 (refs/changes/62/1362/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2017-01-27T09:50:11-08:00 (2 years, 8 months ago)
    
    joyent/node-cueball#70 crash in ConnectionSet during reliability testing

diff --git a/lib/pool-monitor.js b/lib/pool-monitor.js
index 779cf32..e6041d4 100644
--- a/lib/pool-monitor.js
+++ b/lib/pool-monitor.js
@@ -78,7 +78,7 @@ CueBallPoolMonitor.prototype.toKangOptions = function () {
 		var obj = {};
 		obj.backends = pool.p_backends;
 		obj.connections = {};
-		var ks = pool.p_keys;
+		var ks = pool.p_keys.slice();
 		Object.keys(pool.p_connections).forEach(function (k) {
 			if (ks.indexOf(k) === -1)
 				ks.push(k);
@@ -118,7 +118,7 @@ CueBallPoolMonitor.prototype.toKangOptions = function () {
 		obj.backends = cset.cs_backends;
 		obj.fsms = {};
 		obj.connections = Object.keys(cset.cs_connections);
-		var ks = cset.cs_keys;
+		var ks = cset.cs_keys.slice();
 		Object.keys(cset.cs_fsms).forEach(function (k) {
 			if (ks.indexOf(k) === -1)
 				ks.push(k);
