commit 265b02625a260823de127980ec84c47159442fe9
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-18T10:34:32-04:00 (7 months ago)
    
    TOOLS-2207 eng submodule should use https (not git@)

diff --git a/.gitmodules b/.gitmodules
index 11410bd..ed0c5c1 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -9,4 +9,4 @@
 	url = git://github.com/davepacheco/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
-	url = git@github.com:joyent/eng.git
+	url = https://github.com/joyent/eng.git
