From 265b02625a260823de127980ec84c47159442fe9 Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Mon, 18 Mar 2019 10:34:32 -0400
Subject: [PATCH] TOOLS-2207 eng submodule should use https (not git@)

---
 .gitmodules | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.gitmodules b/.gitmodules
index 11410bd..ed0c5c1 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -9,4 +9,4 @@
 	url = git://github.com/davepacheco/jsstyle.git
 [submodule "deps/eng"]
 	path = deps/eng
-	url = git@github.com:joyent/eng.git
+	url = https://github.com/joyent/eng.git
-- 
2.21.0

