From 958f3ed5fb6ef82ba3a981534a80c5d8ba20f26e Mon Sep 17 00:00:00 2001
From: Marsell Kukuljevic <marsell@joyent.com>
Date: Wed, 10 Apr 2019 23:13:34 +0200
Subject: [PATCH] TRITON-1388 Remove process.exit() from dapi lib

---
 lib/allocator.js | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/lib/allocator.js b/lib/allocator.js
index d07b872..0d99d64 100644
--- a/lib/allocator.js
+++ b/lib/allocator.js
@@ -475,10 +475,11 @@ Allocator.prototype._createExpression = function (description, algorithmLookup)
 	var self = this;
 	var expression = [];
 
-	var errExit = function (errMsg) {
-		self.log.error('Bad expression given: ' + errMsg);
-		process.exit(1);
-	};
+	function errExit(errStr) {
+		var errMsg = 'Bad expression given: ' + errStr;
+		self.log.error(errMsg);
+		throw new Error(errMsg);
+	}
 
 	var command = description[0];
 
-- 
2.21.0

