commit ba369a0aade039a33fa5c8d0d045525d6ddb16e9
Author: John Levon <john.levon@joyent.com>
Date:   2019-03-29T12:49:42+00:00 (6 months ago)
    
    TRITON-1353 Triton systems should ship with HT disabled

diff --git a/bin/capacity.js b/bin/capacity.js
index 9aa9058..e03a689 100644
--- a/bin/capacity.js
+++ b/bin/capacity.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -60,6 +60,8 @@ calculateCapacity(servers, processServerCap)
 
 		vms = server.vms;
 		cpu = server.sysinfo['CPU Total Cores'] * 100;
+		if (server.sysinfo.hasOwnProperty('CPU Online Count'))
+			cpu = server.sysinfo['CPU Online Count'] * 100;
 		ram = server.memory_total_bytes *
 		    (1 - server.reservation_ratio);
 		disk = server.disk_pool_size_bytes;
diff --git a/bin/check.js b/bin/check.js
index e5d9e97..2108a20 100644
--- a/bin/check.js
+++ b/bin/check.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -58,6 +58,8 @@ servers.forEach(function (server) {
 
 	vms = server.vms;
 	cpu = server.sysinfo['CPU Total Cores'] * 100;
+	if (server.sysinfo.hasOwnProperty('CPU Online Count'))
+		cpu = server.sysinfo['CPU Online Count'] * 100;
 	ram = server.memory_total_bytes * (1 - server.reservation_ratio);
 
 	disk = server.disk_pool_size_bytes -
diff --git a/lib/algorithms/calculate-server-unreserved.js b/lib/algorithms/calculate-server-unreserved.js
index 4431dbf..f326666 100644
--- a/lib/algorithms/calculate-server-unreserved.js
+++ b/lib/algorithms/calculate-server-unreserved.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -69,7 +69,13 @@ calculateServerUnreserved(servers, opts, cb)
 		    DEFAULT_SERVER_OVERPROVISIONING;
 
 		/* also convert to MiB and cpu_cap units */
+
 		server.unreserved_cpu = server.sysinfo['CPU Total Cores'] * 100;
+		if (server.sysinfo.hasOwnProperty('CPU Online Count')) {
+			server.unreserved_cpu = server.sysinfo
+				['CPU Online Count'] * 100;
+		}
+
 		server.unreserved_ram = server.memory_total_bytes / MB *
 		    (1 - server.reservation_ratio);
 		server.unreserved_disk = calcUnreservedDisk(server);
diff --git a/lib/algorithms/hard-filter-hvm.js b/lib/algorithms/hard-filter-hvm.js
index fba981a..6d9aa52 100644
--- a/lib/algorithms/hard-filter-hvm.js
+++ b/lib/algorithms/hard-filter-hvm.js
@@ -64,6 +64,8 @@ function filterHVM(servers, opts, cb) {
 		bhyveMaxVcpus = Number(sysinfo['Bhyve Max Vcpus']);
 		bhyveSupport = sysinfo['Bhyve Capable'];
 		serverTotalCores = Number(sysinfo['CPU Total Cores']);
+		if (sysinfo.hasOwnProperty('CPU Online Count'))
+			serverTotalCores = Number(sysinfo['CPU Online Count']);
 
 		// Older builds did not have 'Bhyve Max Vcpus', and the max
 		// there was 16
