commit 3fbd96c3641ec1bdf56bf3ecb2fbcb5206bbed33
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2019-03-18T10:34:32-04:00 (7 months ago)
    
    TOOLS-2207 eng submodule should use https (not git@)

diff --git a/.gitmodules b/.gitmodules
index d0ea914..ad1ae7a 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -9,4 +9,4 @@
 	url = https://github.com/joyent/ipxe
 [submodule "deps/eng"]
 	path = deps/eng
-	url = git@github.com:joyent/eng.git
+	url = https://github.com/joyent/eng.git
