From 6052cbc52184e52328521ac4b75b0fc9b9524e92 Mon Sep 17 00:00:00 2001
From: Jan Wyszynski <jan.wyszynski@joyent.com>
Date: Mon, 17 Dec 2018 19:52:59 +0000
Subject: [PATCH] MANTA-3779 Add 'singlePath' field to new object metadata

---
 lib/moray.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/moray.js b/lib/moray.js
index 391c079..e637f41 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -380,6 +380,7 @@ function createMetadata(options) {
         assert.string(options.contentType, 'options.contentType');
         assert.string(options.objectId, 'options.objectId');
         assert.arrayOfObject(options.sharks, 'options.sharks');
+        assert.optionalBool(options.singlePath, 'options.singlePath');
 
         if (!process.env.NODE_NDEBUG) {
             options.sharks.forEach(function validateShark(s) {
@@ -394,6 +395,7 @@ function createMetadata(options) {
         md.etag = options.etag || options.objectId;
         md.objectId = options.objectId;
         md.sharks = options.sharks.slice();
+        md.singlePath = options.singlePath;
         break;
     case 'link':
         assert.object(options.link, 'options.link');
-- 
2.21.0

