commit f8fc05b71ebe24e87f95b610550c593a11c8782b (refs/changes/13/713/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2016-10-18T07:53:43+13:00 (3 years ago)
    
    PUBAPI-1310: Upgrade Node.js to v4 before 0.10 EOL in Oct 2016

diff --git a/.gitmodules b/.gitmodules
index 9ae9cfe..58c386b 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,10 +1,10 @@
 [submodule "deps/restdown"]
 	path = deps/restdown
-	url = git://github.com/trentm/restdown.git
+	url = https://github.com/trentm/restdown.git
 [submodule "deps/javascriptlint"]
 	path = deps/javascriptlint
-	url = git://github.com/davepacheco/javascriptlint.git
+	url = https://github.com/davepacheco/javascriptlint.git
 
 [submodule "deps/jsstyle"]
 	path = deps/jsstyle
-	url = git://github.com/davepacheco/jsstyle.git
+	url = https://github.com/davepacheco/jsstyle.git
diff --git a/lib/keylist.js b/lib/keylist.js
index 7880a9c..5930ef6 100644
--- a/lib/keylist.js
+++ b/lib/keylist.js
@@ -20,7 +20,7 @@ var util = require('util');
 var assert = require('assert-plus');
 var retry = require('retry');
 var uuid = require('node-uuid');
-var sdc = require('sdc-clients');
+var UFDS = require('ufds');
 
 
 
@@ -51,7 +51,7 @@ module.exports = Keylist;
 Keylist.prototype.init = function (cb) {
     var self = this;
 
-    var ufds = new sdc.UFDS(self.ufds);
+    var ufds = new UFDS(self.ufds);
 
     var finish = function (entries) {
         self.keys = entries;
diff --git a/package.json b/package.json
index e1297be..1441e73 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "keyapi",
   "description": "SmartDataCenter token API",
-  "version": "1.0.1",
+  "version": "1.1.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "engines": {
@@ -16,12 +16,12 @@
   "dependencies": {
     "assert-plus": "0.1.2",
     "retry": "0.6.0",
-    "sdc-clients": "~9.1",
-    "sdc-securetoken": "git+ssh://git@github.com:joyent/sdc-securetoken.git#8b0e1768681eb9051b5e9523b4e03cc2fd905330",
+    "ufds": "1.2.0",
+    "sdc-securetoken": "git+https://github.com/joyent/sdc-securetoken.git#8b0e1768681eb9051b5e9523b4e03cc2fd905330",
     "node-uuid": "1.4.1"
   },
   "devDependencies": {
-    "bunyan": "0.8.0",
+    "bunyan": "1.8.1",
     "tap": "0.4.8"
   },
   "license": "MPL-2.0"
diff --git a/test/api.test.js b/test/api.test.js
index d5e0050..4b82b03 100644
--- a/test/api.test.js
+++ b/test/api.test.js
@@ -17,7 +17,7 @@
 var test = require('tap').test;
 var Bunyan = require('bunyan');
 var Keyapi = require('../');
-var UFDS = require('sdc-clients').UFDS;
+var UFDS = require('ufds');
 
 
 
