From 78524e47a13dd68a6db1201859023f16338c3995 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Thu, 26 Jan 2017 16:42:13 -0800
Subject: [PATCH] MANTA-3128 update electric-moray to Moray v3

---
 lib/server.js | 8 +++++---
 package.json  | 2 +-
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/lib/server.js b/lib/server.js
index 7975cf5..bff3b77 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var assert = require('assert-plus');
@@ -743,8 +743,10 @@ function createClient(options, callback) {
                 log: options.log.child({
                     component: 'moray-client-' + pnodeUrl.hostname
                 }),
-                host: pnodeUrl.hostname,
-                port: parseInt(pnodeUrl.port, 10)
+                srvDomain: pnodeUrl.hostname,
+                cueballOptions: {
+                    defaultPort: parseInt(pnodeUrl.port, 10)
+                }
             });
 
             clientMap[pnode] = client;
diff --git a/package.json b/package.json
index 85792fb..03f618d 100644
--- a/package.json
+++ b/package.json
@@ -18,7 +18,7 @@
         "fast": "0.3.8",
         "ldapjs": "0.6.3",
         "imgapi-cli": "git+https://github.com/joyent/sdc-imgapi-cli.git#db3efc1",
-        "moray": "^2.0.0",
+        "moray": "^3.0.0",
         "node-uuid": "1.4.0",
         "posix-getopt": "^1.0.0",
         "vasync": "^1.4.3",
-- 
2.21.0

