commit 6b63a4162a2bc9526713eb7aa260440b9befee07 (refs/changes/02/902/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2016-11-15T17:02:29+01:00 (2 years, 11 months ago)
    
    TOOLS-1615 sdcadm TOOLS-1609 changes broke make check

diff --git a/lib/procedures/update-dockerlogger.js b/lib/procedures/update-dockerlogger.js
index 6f1151b..120d4f2 100644
--- a/lib/procedures/update-dockerlogger.js
+++ b/lib/procedures/update-dockerlogger.js
@@ -350,7 +350,8 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
                     /*
                      * Exit 22 if cannot download the installer file (curl code)
                      */
-                    '/usr/bin/curl -kOsf http://' + ip + '/extra/dockerlogger/' + f,
+                    '/usr/bin/curl -kOsf http://' + ip +
+                        '/extra/dockerlogger/' + f,
                     'if [[ "$?" -ne "0" ]]; then',
                     '   exit $?',
                     'fi',
