From 6b63a4162a2bc9526713eb7aa260440b9befee07 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 15 Nov 2016 17:02:29 +0100
Subject: [PATCH] TOOLS-1615 sdcadm TOOLS-1609 changes broke make check

---
 lib/procedures/update-dockerlogger.js | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/procedures/update-dockerlogger.js b/lib/procedures/update-dockerlogger.js
index 6f1151b..120d4f2 100644
--- a/lib/procedures/update-dockerlogger.js
+++ b/lib/procedures/update-dockerlogger.js
@@ -350,7 +350,8 @@ UpdateDockerlogger.prototype.execute = function udExecute(opts, callback) {
                     /*
                      * Exit 22 if cannot download the installer file (curl code)
                      */
-                    '/usr/bin/curl -kOsf http://' + ip + '/extra/dockerlogger/' + f,
+                    '/usr/bin/curl -kOsf http://' + ip +
+                        '/extra/dockerlogger/' + f,
                     'if [[ "$?" -ne "0" ]]; then',
                     '   exit $?',
                     'fi',
-- 
2.21.0

