commit 1dcfaa7888bd337d5e3fd4d4f4745dbeee28ab05 (refs/changes/00/4600/1)
Author: Ryan Zezeski <rpz@joyent.com>
Date:   2018-07-27T22:46:32-06:00 (1 year, 2 months ago)
    
    OS-7098 bufmod sends corrupted LSO packets

diff --git a/usr/src/uts/common/io/bufmod.c b/usr/src/uts/common/io/bufmod.c
index f553c58aff..3af1d5fc4e 100644
--- a/usr/src/uts/common/io/bufmod.c
+++ b/usr/src/uts/common/io/bufmod.c
@@ -22,10 +22,9 @@
 /*
  * Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
  * Use is subject to license terms.
+ * Copyright 2018 Joyent, Inc.
  */
 
-#pragma ident	"%Z%%M%	%I%	%E% SMI"
-
 /*
  * STREAMS Buffering module
  *
@@ -1016,7 +1015,13 @@ sbaddmsg(queue_t *rq, mblk_t *mp)
 
 		pad = Align(hp.sbh_totlen);
 		hp.sbh_totlen += sizeof (hp);
-		hp.sbh_totlen += pad;
+
+		/*
+		 * We pad only when there is a chance another message
+		 * will be added to the chunk.
+		 */
+		if ((hp.sbh_totlen + pad + SMALLEST_MESSAGE) <= sbp->sb_chunk)
+			hp.sbh_totlen += pad;
 
 		/*
 		 * Would the inclusion of this message overflow the current
