From 4c78e48dce3662bf65ae93f82c9f451b51ead49a Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Fri, 16 Mar 2018 18:03:40 -0700
Subject: [PATCH] MANATEE-398 manatee-adm history -j doesn't understand v1
 timestamps

---
 lib/adm.js | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/lib/adm.js b/lib/adm.js
index 411ef5d..92b2636 100644
--- a/lib/adm.js
+++ b/lib/adm.js
@@ -1814,8 +1814,7 @@ function oldHistoryToObj(fNode) {
                      fNode[j] === 'null') ?  '' : fNode[j];
         switch (j) {
         case 0:
-            node.time = entry;
-            node.date = new Date(parseInt(entry, 10));
+            node.time = new Date(parseInt(entry, 10));
             break;
         case 1:
             node.ip = entry;
-- 
2.21.0

