From b4accc35ce3dc56870ba83c077524913a47558b1 Mon Sep 17 00:00:00 2001
From: Brian Bennett <brian.bennett@joyent.com>
Date: Fri, 16 Mar 2018 17:36:37 -0700
Subject: [PATCH] MANATEE-398 manatee-adm history -j doesn't understand v1
 timestamps

---
 bin/manatee-adm | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/bin/manatee-adm b/bin/manatee-adm
index 7e34493..f8daf28 100755
--- a/bin/manatee-adm
+++ b/bin/manatee-adm
@@ -673,9 +673,17 @@ ManateeAdm.prototype.do_history = function (subcmd, opts, args, cb) {
 
         if (opts.json) {
             events.forEach(function (evt) {
+                var ts;
+                // If this is a v1 event, use the date field instead of the
+                // time field. For v1 events time is a string.
+                if (evt.action) {
+                    ts = evt['date'];
+                } else {
+                    ts = evt['time'];
+                }
                 console.log(JSON.stringify({
                     'zkSeq': evt['zkSeq'],
-                    'time': evt['time'].toISOString(),
+                    'time': ts.toISOString(),
                     'state': evt['state']
                 }));
             });
-- 
2.21.0

