commit 25e0d55ca5b6460f6e8e16b946e1ff8231a9421d (refs/changes/07/4007/1)
Author: John Levon <john.levon@joyent.com>
Date:   2018-05-25T15:39:33+00:00 (1 year, 4 months ago)
    
    OS-6977 zoneadmd incorrectly using zerror() in child

diff --git a/usr/src/cmd/zoneadmd/vplat.c b/usr/src/cmd/zoneadmd/vplat.c
index f466836b96..12acee4bfb 100644
--- a/usr/src/cmd/zoneadmd/vplat.c
+++ b/usr/src/cmd/zoneadmd/vplat.c
@@ -1200,9 +1200,12 @@ mount_one_dev(zlog_t *zlogp, char *devpath, zone_mnt_t mount_cmd)
 	while (zonecfg_getdevent(snap_hndl, &ztab) == Z_OK) {
 		char path[MAXPATHLEN];
 
-		if ((err = resolve_device_match(zlogp, &ztab,
-		    path, sizeof (path))) != Z_OK)
+		if ((err = resolve_device_match(&ztab, path,
+		    sizeof (path))) != Z_OK) {
+			zerror(zlogp, B_TRUE, "failed to resolve device '%s'",
+			    ztab.zone_dev_match);
 			goto cleanup;
+		}
 
 		if (di_prof_add_dev(prof, path)) {
 			zerror(zlogp, B_TRUE, "failed to add "
diff --git a/usr/src/cmd/zoneadmd/zoneadmd.c b/usr/src/cmd/zoneadmd/zoneadmd.c
index 51954a9f3e..ed362e4199 100644
--- a/usr/src/cmd/zoneadmd/zoneadmd.c
+++ b/usr/src/cmd/zoneadmd/zoneadmd.c
@@ -836,8 +836,7 @@ set_zonecfg_env(char *rsrc, char *attr, char *name, char *val)
  * configured for PPT already.
  */
 int
-resolve_device_match(zlog_t *zlogp, struct zone_devtab *dtab,
-    char *path, size_t len)
+resolve_device_match(struct zone_devtab *dtab, char *path, size_t len)
 {
 	struct zone_res_attrtab *rap;
 
@@ -854,18 +853,11 @@ resolve_device_match(zlog_t *zlogp, struct zone_devtab *dtab,
 		return (Z_OK);
 	}
 
-	if (strncmp(dtab->zone_dev_match, "/devices",
-	    strlen("/devices")) != 0) {
-		zerror(zlogp, B_FALSE, "invalid passthru match value '%s'",
-		    dtab->zone_dev_match);
+	if (strncmp(dtab->zone_dev_match, "/devices", strlen("/devices")) != 0)
 		return (Z_INVAL);
-	}
 
-	if (ppt_devpath_to_dev(dtab->zone_dev_match, path, len) != 0) {
-		zerror(zlogp, B_TRUE, "failed to resolve passthru device %s",
-		    dtab->zone_dev_match);
-		return (Z_INVAL);
-	}
+	if (ppt_devpath_to_dev(dtab->zone_dev_match, path, len) != 0)
+		return (Z_NO_ENTRY);
 
 	return (Z_OK);
 }
@@ -887,7 +879,7 @@ resolve_device_match(zlog_t *zlogp, struct zone_devtab *dtab,
  * SmartOS.
  */
 static int
-setup_subproc_env(zlog_t *zlogp, boolean_t debug)
+setup_subproc_env(boolean_t debug)
 {
 	int res;
 	struct zone_nwiftab ntab;
@@ -976,9 +968,13 @@ setup_subproc_env(zlog_t *zlogp, boolean_t debug)
 		struct zone_res_attrtab *rap;
 		char path[MAXPATHLEN];
 
-		res = resolve_device_match(zlogp, &dtab, path, sizeof (path));
-		if (res != Z_OK)
+		res = resolve_device_match(&dtab, path, sizeof (path));
+		if (res != Z_OK) {
+			(void) fprintf(stderr,
+			    "failed to resolve device '%s': %s\n", match,
+			    zonecfg_strerror(res));
 			goto done;
+		}
 
 		/*
 		 * Even if not modified, the match path will be mangled in the
@@ -1121,7 +1117,7 @@ do_subproc(zlog_t *zlogp, char *cmdbuf, char **retstr, boolean_t debug)
 		}
 		closefrom(STDERR_FILENO + 1);
 
-		if (setup_subproc_env(zlogp, debug) != Z_OK) {
+		if (setup_subproc_env(debug) != Z_OK) {
 			(void) fprintf(stderr, "failed to setup environment");
 			_exit(127);
 		}
diff --git a/usr/src/cmd/zoneadmd/zoneadmd.h b/usr/src/cmd/zoneadmd/zoneadmd.h
index 4953479e64..4b66cacb48 100644
--- a/usr/src/cmd/zoneadmd/zoneadmd.h
+++ b/usr/src/cmd/zoneadmd/zoneadmd.h
@@ -190,8 +190,7 @@ extern int do_subproc(zlog_t *, char *, char **, boolean_t);
 /*
  * Resource handling.
  */
-extern int resolve_device_match(zlog_t *, struct zone_devtab *,
-    char *, size_t);
+extern int resolve_device_match(struct zone_devtab *, char *, size_t);
 
 #ifdef __cplusplus
 }
