From cc598989d2d952610a82d556d81d0835f78ef494 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Thu, 1 Sep 2016 14:20:28 -0700
Subject: [PATCH] joyent/node-cueball#11 don't create a DNS UDP socket per
 Resolver

---
 lib/resolver.js | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/lib/resolver.js b/lib/resolver.js
index ebc004d..9dbb9fa 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -240,9 +240,14 @@ function CueBallDNSResolver(options) {
 	this.r_bootstrap = undefined;
 	this.r_bootstrapRes = {};
 
-	this.r_nsclient = new mod_nsc.DnsClient({
-		concurrency: this.r_maxres
-	});
+	this.r_nsclient = CueBallDNSResolver.globalNSClients[this.r_maxres];
+	if (this.r_nsclient === undefined) {
+		this.r_nsclient = new mod_nsc.DnsClient({
+			concurrency: this.r_maxres
+		});
+		CueBallDNSResolver.globalNSClients[this.r_maxres] =
+		    this.r_nsclient;
+	}
 
 	this.r_stopping = false;
 
@@ -254,6 +259,8 @@ mod_util.inherits(CueBallDNSResolver, FSM);
 
 CueBallDNSResolver.bootstrapResolvers = {};
 
+CueBallDNSResolver.globalNSClients = {};
+
 CueBallDNSResolver.prototype.start = function () {
 	this.emit('startAsserted');
 };
-- 
2.21.0

