From 487b4dae7a78ac4383481b72320ea5aafafe1580 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Fri, 28 Apr 2017 11:15:18 -0700
Subject: [PATCH] joyent/node-zkstream#5 provide cueball connection compatible
 API

---
 lib/client-fsm.js | 30 ++++++++++++++++--------------
 1 file changed, 16 insertions(+), 14 deletions(-)

diff --git a/lib/client-fsm.js b/lib/client-fsm.js
index 059f39c..e3c3460 100644
--- a/lib/client-fsm.js
+++ b/lib/client-fsm.js
@@ -44,8 +44,13 @@ ZKClientFSM.prototype.connect = function () {
 
 ZKClientFSM.prototype.close = function () {
 	mod_assert.ok(!this.isInState('closed'));
-	var self = this;
-	self.emit('closeAsserted');
+	this.emit('closeAsserted');
+};
+
+ZKClientFSM.prototype.destroy = function () {
+	if (this.isInState('closed'))
+		return;
+	this.emit('closeAsserted');
 };
 
 ZKClientFSM.prototype.nextXid = function () {
@@ -62,18 +67,13 @@ ZKClientFSM.prototype.state_closed = function (S) {
 	S.on(this, 'connectAsserted', function () {
 		S.gotoState('connecting');
 	});
-	if (this.zs_encoder) {
-		this.zs_encoder.end();
-		this.zs_encoder = undefined;
-	}
-	if (this.zs_socket) {
-		this.zs_socket.destroy();
-		this.zs_socket = undefined;
-		this.emit('close');
-	}
-	if (this.zs_decoder) {
-		this.zs_decoder = undefined;
-	}
+
+	this.zs_encoder.end();
+	this.zs_encoder = undefined;
+	this.zs_socket.destroy();
+	this.zs_socket = undefined;
+	this.emit('close');
+	this.zs_decoder = undefined;
 };
 
 ZKClientFSM.prototype.state_connecting = function (S) {
@@ -217,6 +217,8 @@ ZKClientFSM.prototype.state_connected = function (S) {
 		self.zs_lastError = new Error('Ping timeout');
 		S.gotoState('error');
 	});
+
+	this.emit('connect');
 };
 
 ZKClientFSM.prototype.state_error = function (S) {
-- 
2.21.0

