From 5e750d34489ab72a188c517c8176cc2b484b24e9 Mon Sep 17 00:00:00 2001
From: Orlando Vazquez <ovazquez@gmail.com>
Date: Tue, 5 Dec 2017 14:42:25 -0800
Subject: [PATCH] CNAPI-729 Updating server boot params didn't work as expected

---
 lib/models/server.js | 29 +++++++++++++++++------------
 1 file changed, 17 insertions(+), 12 deletions(-)

diff --git a/lib/models/server.js b/lib/models/server.js
index e167280..376f4c7 100644
--- a/lib/models/server.js
+++ b/lib/models/server.js
@@ -1814,21 +1814,26 @@ ModelServer.prototype.getBootParams = function (callback) {
             return;
         }
 
-        dns.resolve(host, function (dnserror, addrs) {
-            if (dnserror) {
-                callback(dnserror);
-                return;
-            }
+        if (host && host.match(/^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/)) {
+            callback(null, params);
+            return;
+        } else {
+            dns.resolve(host, function (dnserror, addrs) {
+                if (dnserror) {
+                    callback(dnserror);
+                    return;
+                }
 
-            host = addrs[Math.floor(Math.random() * addrs.length)];
-            parts[2] = host;
+                host = addrs[Math.floor(Math.random() * addrs.length)];
+                parts[2] = host;
 
-            params.kernel_args.rabbitmq_dns = params.kernel_args.rabbitmq;
-            params.kernel_args.rabbitmq = parts.join(':');
+                params.kernel_args.rabbitmq_dns = params.kernel_args.rabbitmq;
+                params.kernel_args.rabbitmq = parts.join(':');
 
-            callback(null, params);
-            return;
-        });
+                callback(null, params);
+                return;
+            });
+        }
     });
 };
 
-- 
2.21.0

