From 461f16ca910c6d8f8d60bccd986204d4018a23c9 Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 29 Jul 2019 19:25:25 +0100
Subject: [PATCH] TOOLS-2297 update deps/eng for copyright checks and motd work

---
 deps/eng | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/eng b/deps/eng
index 7c472bc..01f669d 160000
--- a/deps/eng
+++ b/deps/eng
@@ -1 +1 @@
-Subproject commit 7c472bc495ba44ac4cdde0af50ff627021942524
+Subproject commit 01f669da1f5cb1439d573fc42efd2dc43a7c6184
-- 
2.21.0

