From 336a5db82eddf55247e462d3fa1045923589706d Mon Sep 17 00:00:00 2001
From: Trent Mick <trentm@gmail.com>
Date: Tue, 20 Feb 2018 09:41:38 -0800
Subject: [PATCH] TRITON-161 clean SAPI-254 debug logging out of SAPI test file

---
 test/instances.test.js | 20 --------------------
 1 file changed, 20 deletions(-)

diff --git a/test/instances.test.js b/test/instances.test.js
index e08797d..5298e5e 100644
--- a/test/instances.test.js
+++ b/test/instances.test.js
@@ -12,9 +12,6 @@
  * test/instances.test.js: test /instances endpoints
  */
 
-var debug = console.error;
-// var debug = function () {};
-
 var async = require('async');
 var jsprim = require('jsprim');
 var node_uuid = require('node-uuid');
@@ -1081,7 +1078,6 @@ test('teardown hooks', function (t) {
     var uri_svc = '/services/' + svc_uuid;
     var uri_inst = '/instances/' + inst.uuid;
 
-    debug('SAPI-254: env:', process.env);
     /*
      * In proto mode, the teardown-hook can't run (since CNAPI is
      * unavailable), so don't run this test.
@@ -1093,15 +1089,12 @@ test('teardown hooks', function (t) {
 
     vasync.pipeline({funcs: [
         function (_, cb) {
-            debug('SAPI-254: createApplication');
             common.createApplication({sapi: self.sapi, uuid: app_uuid}, cb);
         },
         function (_, cb) {
-            debug('SAPI-254: createService');
             common.createService.call(self, app_uuid, svc_uuid, cb);
         },
         function (_, cb) {
-            debug('SAPI-254: post inst');
             client.post(URI, inst, function (err, req, res, obj) {
                 t.ifError(err);
                 t.equal(res.statusCode, 200);
@@ -1109,7 +1102,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: del inst');
             /*
              * Both destroying and reprovisioning an instance should
              * fail when the teardown-hook fails.
@@ -1121,7 +1113,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: upgrade');
             var uri = sprintf('/instances/%s/upgrade', inst.uuid);
 
             var opts = {};
@@ -1134,7 +1125,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: get inst');
             self.client.get(uri_inst, function (err, req, res, obj) {
                 t.ifError(err);
                 t.equal(res.statusCode, 200);
@@ -1142,7 +1132,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: update with teardown-hook');
             var opts = {};
             opts.params = {};
             opts.params['teardown-hook'] = '/bin/true';
@@ -1157,7 +1146,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: post inst without teardown-hook');
             delete inst.params['teardown-hook'];
 
             client.post(URI, inst, function (err, req, res, obj) {
@@ -1167,7 +1155,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: update with teardown-hook=/bin/false');
             var opts = {};
             opts.params = {};
             opts.params['teardown-hook'] = '/bin/false';
@@ -1181,7 +1168,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: del inst');
             /*
              * Both destroying and reprovisioning an instance should
              * fail when the teardown-hook fails.  Note that in this
@@ -1195,7 +1181,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: upgrade inst');
             var uri = sprintf('/instances/%s/upgrade', inst.uuid);
 
             var opts = {};
@@ -1208,7 +1193,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: update inst teardown-hook=/bin/true');
             var opts = {};
             opts.params = {};
             opts.params['teardown-hook'] = '/bin/true';
@@ -1222,7 +1206,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: upgrade inst to NEW_IMAGE');
             var uri = sprintf('/instances/%s/upgrade', inst.uuid);
 
             var opts = {};
@@ -1235,7 +1218,6 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: del inst');
             self.client.del(uri_inst, function (err, req, res, obj) {
                 t.ifError(err);
                 t.equal(res.statusCode, 204);
@@ -1243,13 +1225,11 @@ test('teardown hooks', function (t) {
             });
         },
         function (_, cb) {
-            debug('SAPI-254: deleteService');
             self.sapi.deleteService(svc_uuid, function (err) {
                 cb(err);
             });
         },
         function (_, cb) {
-            debug('SAPI-254: deleteApplication');
             self.sapi.deleteApplication(app_uuid, function (err) {
                 cb(err);
             });
-- 
2.21.0

