From aded360d8380b73339af305cb3a181dd3a0fa700 Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Mon, 19 Aug 2019 11:26:32 -0700
Subject: [PATCH] TRITON-1871 vmapi migration test fail when using uuid
 override

---
 test/lib/migration.js | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/lib/migration.js b/test/lib/migration.js
index 078cff0..9cfd90f 100644
--- a/test/lib/migration.js
+++ b/test/lib/migration.js
@@ -1300,7 +1300,7 @@ function TestMigrationCfg(test, cfg) {
 
         client.post({
             path: format('/vms/%s?action=migrate&migration_action=finalize',
-                targetVm.uuid)
+                sourceVm.uuid)
         }, function onMigrateFinalizeCb(err, req, res) {
             common.ifError(t, err, 'no error from migration finalize call');
             if (!err) {
-- 
2.21.0

